# Conjugations Protocol

**Transfer of Ready-to-Use CRISPR Plasmids into the Target Streptomycetes by Interspecies Conjugation**
**Timing: 2 days**

## 1. Plate the Target Streptomycete Strains

Plate the target streptomycete strains onto MS plates for sporulation. This is typically done days ahead, depending on the growth rate; for example, it requires ~5 days for the S. coelicolor WT strain to fully sporulate on MS plates at 30 °C.

## 2. Transfer Plasmids to E. coli

Transfer 200 ng of the desired plasmids into 50 µL of in-house-made electroporation-competent E. coli ET cells, using the electroporation protocol described previously with small modifications as follows: instead of plating onto selective LB plates, transfer all the reaction to a 50-mL Falcon tube with 20 mL selective LB medium supplemented with 50 µg/mL apramycin, 25 µg/mL kanamycin, and 12.5 µg/mL chloramphenicol. Incubate the tubes overnight at 37 °C with 200 r.p.m. shaking.

**CRITICAL STEP**: It is also possible to plate the transformation reaction onto a selective LB plate and then start the culture from a single colony the next day. Using a pool of transformants lowers the chance of picking a wrong clone and can save at least 1 day.

## 3. Prepare ET Cultures

The following morning, prepare the ET cultures harboring the plasmids of interest by washing twice using the same volume (20 mL) of antibiotic-free LB medium, and harvest the cells by centrifuging at 5,000g for 5 min at room temperature. Then resuspend the cell pellets in 2 mL antibiotic-free LB medium per 20-mL culture.

**CRITICAL STEP**: We recommend using an amount of LB that is 10% (vol/vol) of the initial culture volume for resuspension of the ET cell pellets.

## 4. Collect Streptomycete Spores

Collect spores of the streptomycete of interest by pipetting 10 mL 2× YT onto the surface of the spore lawn of a well-sporulated MS plate.

## 5. Scrape Off the Spores

Use a sterilized cotton swab to scrape off the spores.

## 6. Filter Spore Suspension

Carefully place a sterilized cotton pad above the spore suspension with sterilized tweezers.

## 7. Aspirate Spore Suspension

Use an Integra Pipetboy to aspirate the spore suspension through the cotton pad to remove agar and mycelial debris; then transfer the suspension to a 50-mL Falcon tube.

**CRITICAL STEP**: Repeat the above operation twice to maximize the amount of spores from one plate. For strains that produce only low numbers of spores, it may be necessary to collect spores from more than one plate.

## 8. Resuspend Spores

Collect the spores by centrifuging at 5,000g for 5 min at room temperature. Discard the supernatant carefully, resuspend the spores in 2× YT (typically 2 mL per plate).

## 9. Heat-Shock Spores

Heat-shock the spore suspension for 10 min at 50 °C. The spore suspension is now ready for conjugation.

**CRITICAL STEP**: To achieve higher conjugation efficiency, leave the spore suspension at 4 °C overnight (up to 3 days) for pre-germination.

**PAUSE POINT**: The spore suspension can be used for up to 2 weeks if stored at 4 °C.

## 10. Mix ET Suspension with Spore Suspension

Mix 500 µL ET suspension with 200 µL spore suspension in a sterilized 1.5-mL Eppendorf tube by pipetting.

## 11. Plate Mixture

Plate 100 µL, 200 µL, and 400 µL of the 700-µL mixture onto three different MS plates, air-dry them in a laminar flow hood for 5 min, and incubate the conjugation plates at 30 °C for ~24 h (overnight).

## 12. Overlay Conjugation Plates

Overlay the conjugation plates with 1 mL sterilized H2O containing 1 mg apramycin and 1 mg nalidixic acid. Air-dry the plates in a laminar flow hood for 15 min.

**CAUTION**: We do not recommend using a spreader for the overlay procedure. Instead, spread the 1 mL sterilized H2O containing 1 mg apramycin and 1 mg nalidixic acid by moving the plate. The 1 mL sterilized H2O will form clouds after you add the nalidixic acid stock.

## 13. Incubate Plates

Incubate the plates until exconjugants can be picked with a sterilized wooden toothpick. Typically, it takes 3–5 days if the streptomycete of interest has a normal growth rate. Transfer the picked exconjugants to a fresh ISP2 plate supplemented with 50 µg/mL apramycin and 50 µg/mL nalidixic acid and incubate for 3–5 days at 30 °C.

## 14. Evaluate the Successfully Edited Strains

### Evaluation of Non-Genetically Edited Applications (CRISPRi)

1. **Prepare Seed Cultures**:

   - Inoculate three randomly picked exconjugants for each plasmid and a non-treated control into individual shake flasks, each containing 50 mL selective ISP2 broth, and incubate at 30 °C with 180 r.p.m. shaking for 3 days.

2. **Normalize Start Amount**:

   - Spin down 1-mL cultures at 10,000g at room temperature for 5 min to enable measurement of equal amounts of inocula for the main cultures.

3. **Inoculate Main Cultures**:

   - Inoculate 500-mg cell pellets (wet weight) from seed cultures into a fresh 50-mL selective ISP2 broth–containing shake flask and incubate the flask at 30 °C with 180 r.p.m. shaking for 3–5 days.

4. **Extract and Analyze Endpoint Product**:
   - Extract and analyze the endpoint product accordingly, such as actinorhodin by absorbance measurement.

**CAUTION**: Besides directly analyzing the endpoint product, transcription analysis can also be applied, such as qRT-PCR and RNA-seq.

### Evaluation of Genetically Edited Strains by Colony PCR

1. **Quick Screening**:
   - Scratch ~4 mm2 of mycelia from the ISP2 plates and transfer to a PCR tube containing 20 µL DMSO.

**CRITICAL STEP**: Use the mycelia from a fast-growing stage (before sporulation) for Streptomyces colony PCR.

2. **Incubate**:

   - Incubate for 15 min at 100 °C, with shaking at 1,000 r.p.m. using a shaking/heating block.

3. **Freeze**:
   - Transfer the tube to a −20 °C freezer and freeze for 30 min.

**CRITICAL STEP**: Breaking up streptomycete cells can be improved by repeating the 100 °C−20 °C cycle one more time.

4. **Colony PCR**:
   - Use 1.5 µL of the obtained solution as template DNA for colony PCR. Primers flanking the ~500-bp target regions are used. Perform PCR as follows:

| Component                               | Amount (μL) | Final Concentration |
| --------------------------------------- | ----------- | ------------------- |
| Forward primer                          | 1           | 400 nM              |
| Reverse primer                          | 1           | 400 nM              |
| Template DNA                            | 1.5         | 3% (vol/vol) DMSO   |
| 2× Phusion High-Fidelity PCR Master Mix | 25          | 1×                  |
| ddH2O                                   | 21.5        |                     |
| **Total volume**                        | 50          |                     |

| Cycle no. | Denature     | Anneal      | Extend       | Final       |
| --------- | ------------ | ----------- | ------------ | ----------- |
| 1         | 98 °C, 1 min |             |              |             |
| 2–31      | 98 °C, 10 s  | 65 °C, 30 s | 72 °C, 10 s  |             |
| 32        |              |             | 72 °C, 5 min |             |
| 33        |              |             |              | 10 °C, hold |

**CRITICAL STEP**: We obtained equal efficiencies when using NEB Q5 High-Fidelity 2× Master Mix and 2× Phusion High-Fidelity PCR Master Mix with HF Buffer.

5. **Analyze PCR Products**:

   - Analyze 5 µL of the above PCR reaction along with the GeneRuler 1-kb DNA ladder on an agarose gel (1% (wt/vol)) with 1× TAE buffer. Run the gel at 100 V for 30 min and visualize the bands using a Gel Doc XR+ Gel Documentation System.

6. **Clean Up PCR Products**:

   - Clean up the PCR products with clear bands using a NucleoSpin Gel and PCR Clean-up Kit. Measure the concentration of each fragment using a NanoDrop 2000 spectrophotometer.

7. **Validate Edits by Sanger Sequencing**:
   - Validate the edits by Sanger sequencing of 8–12 PCR products, using the forward primers flanking the target regions.

### Evaluation of Genetically Edited Mutants by Illumina Sequencing

1. **Inoculate Single Colonies**:

   - Inoculate a single colony each of the Sanger sequencing–validated positive strain and a WT parental strain into independent 50-mL shake flasks containing ISP2 broth (with no antibiotics) and incubate at 30 °C with 180 r.p.m. shaking for 3–5 days.

2. **Isolate Genomic DNA**:
   - Isolate genomic DNA from the above strains using the Qiagen Blood & Cell Culture Midi DNA Kit with a modified protocol.

**PAUSE POINT**: The genomic DNA solution can be stored at −20 °C for up to 6 months.

3. **Build Illumina Sequencing Libraries**:
   - Build the Illumina sequencing libraries using a KAPA HyperPlus Kit, aiming for an insert size of ~600 nt, using SeraMag Select SPRI beads.

**CRITICAL STEP**: Using a kit with few PCR cycles is preferable to avoid a high proportion of clonal fragments. Avoid using a transposase-based sequencing library kit due to the high GC content of Streptomyces.

4. **Quantify Libraries**:

   - Quantify the libraries using a fragment analyzer and a Qubit dsDNA HS Assay Kit.

5. **Sequence Samples**:

   - Sequence the samples on a suitable Illumina machine with a paired-end protocol of 2× 150 nt (or longer), such as a NextSeq500 sequencer, to a coverage of ~100.

6. **Trim Reads**:

   - Perform adaptor and quality trimming of the reads using AdapterRemoval2 with the switches --trimns and --trimqualities.

7. **Run FastQC**:

   - Run FastQC to ensure that low-quality reads and sequences have been sufficiently removed from the trimmed dataset.

8. **Estimate Changes Using breseq**:
   - Use breseq with a high-quality reference genome sequence to estimate the changes in both the WT and genome-edited strains. Evaluate the reports and report the predicted mutations from all datasets. Identical predicted ‘mutations’ in the two datasets could reflect mutations that happened before the genome editing experiments. Pay special attention to larger structural changes of the streptomycetes genomes, such as deletions in the chromosome ends.

**CAUTION**: Include the parental strain when evaluating the on/off-target edits from CRISPR genome editing using whole-genome sequencing–based techniques to avoid overestimating off-target effects.
