{
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "# W3-CRISPRi-workflow LLPMBPKK_05956 CRISPRi\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "Transcriptional interference is another powerful tool for assessing the function of genes, by expressing a guide RNA that targets a dead-Cas9/sgRNA complex to bind to the + (Watson) strand of the promoter and 5’-UTR of the targeted gene, thereby sterically blocking binding and progression of RNA polymerases to the promoter region. Workflow 4 uses CRISPR interference (CRISPRi) with ssDNA bridging to reversibly inactivate genes transcriptionally12,29 (Figure 5A). This approach targets regions upstream of the Transcriptional Start Site (TSS), using a dCas9-sgRNA complex positioned near the TSS (default 100bp upstream) to sterically hinder transcription. This allows for the functional study of genes through controlled knockdown. To get started, users can download the pCRISPR-dCas9 plasmid and the S. coelicolor (A3) genome. StreptoAIM will then generate all necessary components, including primers and plasmids."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 1,
      "metadata": {},
      "outputs": [],
      "source": [
        "import sys\n",
        "import os\n",
        "from pydna.dseqrecord import Dseqrecord\n",
        "from datetime import datetime\n",
        "\n",
        "\n",
        "# Ensure the src directory is in the Python path\n",
        "project_root = os.path.abspath(os.path.join(os.getcwd(), '../../'))\n",
        "if project_root not in sys.path:\n",
        "    sys.path.insert(0, project_root)\n",
        "\n",
        "from streptocad.sequence_loading.sequence_loading import (\n",
        "    load_and_process_gene_sequences,\n",
        "    load_and_process_genome_sequences,\n",
        "    load_and_process_plasmid, \n",
        "    check_and_convert_input,\n",
        "    annotate_dseqrecord,\n",
        "    process_specified_gene_sequences_from_record)\n",
        "\n",
        "\n",
        "from streptocad.cloning.ssDNA_bridging import assemble_plasmids_by_ssDNA_bridging, make_ssDNA_oligos\n",
        "from streptocad.crispr.guideRNA_crispri import extract_sgRNAs_for_crispri, SgRNAargs\n",
        "from streptocad.primers.primer_generation import  primers_to_IDT\n",
        "from streptocad.utils import ProjectDirectory,extract_metadata_to_dataframe"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "## INPUT"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 2,
      "metadata": {},
      "outputs": [],
      "source": [
        "# Inputs\n",
        "# 1 Add genome of choice (genbank, fasta)\n",
        "path_to_genome = '../wet_lab_notebooks/data_for_drylab/Go40110_flye_polish_actinoannotPFAM_antismash7/Go40110_flye_polish_actinoannotPFAM.gbk'\n",
        "genome = load_and_process_genome_sequences(path_to_genome)[0]\n",
        "\n",
        "# 2 Add plasmid \n",
        "path_to_plasmid = '../../web_app/assets/pCRISPR-Cas9.gb'\n",
        "clean_plasmid = load_and_process_plasmid(path_to_plasmid)\n",
        "\n",
        "# 3 Choose genes to knock out (list)\n",
        "genes_to_KO = ['LLPMBPKK_05956']#, ''LLPMBPKK_00515', 'LLPMBPKK_00435',', 'SCO5090']\n",
        "\n",
        "\n",
        "#### Advanced settings ####\n",
        "# 4 Filtering metrics for sgRNAs\n",
        "gc_upper = 0.8\n",
        "gc_lower = 0.2\n",
        "off_target_seed = 13\n",
        "off_target_upper = 1\n",
        "cas_type='cas9'\n",
        "number_of_sgRNAs_per_group = 3\n",
        "extension_to_promoter_region=200\n",
        "\n",
        "# 6 Choose overlapping sequences for our plasmid we can use the following\n",
        "#As per the article **\"CRISPR–Cas9, CRISPRi and CRISPR-BEST-mediated genetic manipulation in streptomycetes\"** we need the following oligoes: \n",
        "#CGGTTGGTAGGATCGACGGC **-N20-** GTTTTAGAGCTAGAAATAGC\n",
        "up_homology = Dseqrecord('CGGTTGGTAGGATCGACGGC')\n",
        "dw_homology = Dseqrecord('GTTTTAGAGCTAGAAATAGC')"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 3,
      "metadata": {},
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            ".\n"
          ]
        }
      ],
      "source": [
        "print(clean_plasmid.id)\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "# Computation"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 4,
      "metadata": {},
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "False\n"
          ]
        },
        {
          "data": {
            "text/plain": [
              "18013"
            ]
          },
          "execution_count": 4,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "target_dict, genes_to_KO, annotation_input = check_and_convert_input(genes_to_KO)\n",
        "\n",
        "print(annotation_input)\n",
        "if annotation_input == True:\n",
        "    genome = annotate_dseqrecord(genome, target_dict)\n",
        "\n",
        "\n",
        "len(genome.features)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 5,
      "metadata": {},
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "sgRNA generated were outside the designated border in LLPMBPKK_05956. To incorporate this extent borders. Skipping to next locus tag.\n",
            "sgRNA generated were outside the designated border in LLPMBPKK_05956. To incorporate this extent borders. Skipping to next locus tag.\n",
            "Pam was found outside designated locus_tag: LLPMBPKK_05956. To incorporate this extent borders. Skipping to next locus tag.\n",
            "sgRNA generated were outside the designated border in LLPMBPKK_05956_upstream. To incorporate this extent borders. Skipping to next locus tag.\n"
          ]
        },
        {
          "name": "stderr",
          "output_type": "stream",
          "text": [
            "/Users/lucaslevassor/opt/anaconda3/envs/streptocad/lib/python3.11/site-packages/Bio/SeqFeature.py:257: BiopythonDeprecationWarning: Using the strand argument is deprecated, and will be removed in a future release. Please set it via the location object instead.\n",
            "  warnings.warn(\n"
          ]
        },
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>strain_name</th>\n",
              "      <th>locus_tag</th>\n",
              "      <th>gene_loc</th>\n",
              "      <th>gene_strand</th>\n",
              "      <th>sgrna_strand</th>\n",
              "      <th>sgrna_loc</th>\n",
              "      <th>gc</th>\n",
              "      <th>pam</th>\n",
              "      <th>sgrna</th>\n",
              "      <th>sgrna_seed_sequence</th>\n",
              "      <th>off_target_count</th>\n",
              "      <th>region</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>140</th>\n",
              "      <td>.</td>\n",
              "      <td>LLPMBPKK_05956</td>\n",
              "      <td>6815841</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>-47</td>\n",
              "      <td>0.75</td>\n",
              "      <td>AGG</td>\n",
              "      <td>TGCGGGCAGGGGTGTCCGTG</td>\n",
              "      <td>AGGGGTGTCCGTG</td>\n",
              "      <td>0</td>\n",
              "      <td>upstream</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>145</th>\n",
              "      <td>.</td>\n",
              "      <td>LLPMBPKK_05956</td>\n",
              "      <td>6815841</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>-29</td>\n",
              "      <td>0.80</td>\n",
              "      <td>CGG</td>\n",
              "      <td>TAGCGCGCCGTGGGCGCGTG</td>\n",
              "      <td>CCGTGGGCGCGTG</td>\n",
              "      <td>0</td>\n",
              "      <td>upstream</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>146</th>\n",
              "      <td>.</td>\n",
              "      <td>LLPMBPKK_05956</td>\n",
              "      <td>6815841</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>-20</td>\n",
              "      <td>0.80</td>\n",
              "      <td>GGG</td>\n",
              "      <td>CCACCCGCGTAGCGCGCCGT</td>\n",
              "      <td>CGTAGCGCGCCGT</td>\n",
              "      <td>1</td>\n",
              "      <td>upstream</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>147</th>\n",
              "      <td>.</td>\n",
              "      <td>LLPMBPKK_05956</td>\n",
              "      <td>6815841</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>-19</td>\n",
              "      <td>0.80</td>\n",
              "      <td>TGG</td>\n",
              "      <td>TCCACCCGCGTAGCGCGCCG</td>\n",
              "      <td>GCGTAGCGCGCCG</td>\n",
              "      <td>1</td>\n",
              "      <td>upstream</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>.</td>\n",
              "      <td>LLPMBPKK_05956</td>\n",
              "      <td>6816041</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>60</td>\n",
              "      <td>0.70</td>\n",
              "      <td>CGG</td>\n",
              "      <td>CAGTCCGGTGGCCACCGCAT</td>\n",
              "      <td>GTGGCCACCGCAT</td>\n",
              "      <td>0</td>\n",
              "      <td>CDS</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "    strain_name       locus_tag  gene_loc  gene_strand  sgrna_strand  \\\n",
              "140           .  LLPMBPKK_05956   6815841            1            -1   \n",
              "145           .  LLPMBPKK_05956   6815841            1            -1   \n",
              "146           .  LLPMBPKK_05956   6815841            1            -1   \n",
              "147           .  LLPMBPKK_05956   6815841            1            -1   \n",
              "0             .  LLPMBPKK_05956   6816041            1            -1   \n",
              "\n",
              "     sgrna_loc    gc  pam                 sgrna sgrna_seed_sequence  \\\n",
              "140        -47  0.75  AGG  TGCGGGCAGGGGTGTCCGTG       AGGGGTGTCCGTG   \n",
              "145        -29  0.80  CGG  TAGCGCGCCGTGGGCGCGTG       CCGTGGGCGCGTG   \n",
              "146        -20  0.80  GGG  CCACCCGCGTAGCGCGCCGT       CGTAGCGCGCCGT   \n",
              "147        -19  0.80  TGG  TCCACCCGCGTAGCGCGCCG       GCGTAGCGCGCCG   \n",
              "0           60  0.70  CGG  CAGTCCGGTGGCCACCGCAT       GTGGCCACCGCAT   \n",
              "\n",
              "     off_target_count    region  \n",
              "140                 0  upstream  \n",
              "145                 0  upstream  \n",
              "146                 1  upstream  \n",
              "147                 1  upstream  \n",
              "0                   0       CDS  "
            ]
          },
          "execution_count": 5,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "\n",
        "# Initialize SgRNAArgs with desired parameters\n",
        "args = SgRNAargs(genome, \n",
        "                genes_to_KO,\n",
        "                step=['find', 'filter'],\n",
        "                gc_upper = gc_upper,\n",
        "                gc_lower = gc_lower,\n",
        "                off_target_seed = off_target_seed,\n",
        "                off_target_upper = off_target_upper,\n",
        "                cas_type='cas9',\n",
        "                extension_to_promoter_region=200,\n",
        "                target_non_template_strand = True, \n",
        "                upstream_tss = 70, \n",
        "                dwstream_tss = 60, \n",
        "                  )\n",
        "\n",
        "sgrna_df = extract_sgRNAs_for_crispri(args)\n",
        "sgrna_df"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 6,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>strain_name</th>\n",
              "      <th>locus_tag</th>\n",
              "      <th>gene_loc</th>\n",
              "      <th>gene_strand</th>\n",
              "      <th>sgrna_strand</th>\n",
              "      <th>sgrna_loc</th>\n",
              "      <th>gc</th>\n",
              "      <th>pam</th>\n",
              "      <th>sgrna</th>\n",
              "      <th>sgrna_seed_sequence</th>\n",
              "      <th>off_target_count</th>\n",
              "      <th>region</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>140</th>\n",
              "      <td>.</td>\n",
              "      <td>LLPMBPKK_05956</td>\n",
              "      <td>6815841</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>-47</td>\n",
              "      <td>0.75</td>\n",
              "      <td>AGG</td>\n",
              "      <td>TGCGGGCAGGGGTGTCCGTG</td>\n",
              "      <td>AGGGGTGTCCGTG</td>\n",
              "      <td>0</td>\n",
              "      <td>upstream</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>145</th>\n",
              "      <td>.</td>\n",
              "      <td>LLPMBPKK_05956</td>\n",
              "      <td>6815841</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>-29</td>\n",
              "      <td>0.80</td>\n",
              "      <td>CGG</td>\n",
              "      <td>TAGCGCGCCGTGGGCGCGTG</td>\n",
              "      <td>CCGTGGGCGCGTG</td>\n",
              "      <td>0</td>\n",
              "      <td>upstream</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>146</th>\n",
              "      <td>.</td>\n",
              "      <td>LLPMBPKK_05956</td>\n",
              "      <td>6815841</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>-20</td>\n",
              "      <td>0.80</td>\n",
              "      <td>GGG</td>\n",
              "      <td>CCACCCGCGTAGCGCGCCGT</td>\n",
              "      <td>CGTAGCGCGCCGT</td>\n",
              "      <td>1</td>\n",
              "      <td>upstream</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>147</th>\n",
              "      <td>.</td>\n",
              "      <td>LLPMBPKK_05956</td>\n",
              "      <td>6815841</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>-19</td>\n",
              "      <td>0.80</td>\n",
              "      <td>TGG</td>\n",
              "      <td>TCCACCCGCGTAGCGCGCCG</td>\n",
              "      <td>GCGTAGCGCGCCG</td>\n",
              "      <td>1</td>\n",
              "      <td>upstream</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>.</td>\n",
              "      <td>LLPMBPKK_05956</td>\n",
              "      <td>6816041</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>60</td>\n",
              "      <td>0.70</td>\n",
              "      <td>CGG</td>\n",
              "      <td>CAGTCCGGTGGCCACCGCAT</td>\n",
              "      <td>GTGGCCACCGCAT</td>\n",
              "      <td>0</td>\n",
              "      <td>CDS</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "    strain_name       locus_tag  gene_loc  gene_strand  sgrna_strand  \\\n",
              "140           .  LLPMBPKK_05956   6815841            1            -1   \n",
              "145           .  LLPMBPKK_05956   6815841            1            -1   \n",
              "146           .  LLPMBPKK_05956   6815841            1            -1   \n",
              "147           .  LLPMBPKK_05956   6815841            1            -1   \n",
              "0             .  LLPMBPKK_05956   6816041            1            -1   \n",
              "\n",
              "     sgrna_loc    gc  pam                 sgrna sgrna_seed_sequence  \\\n",
              "140        -47  0.75  AGG  TGCGGGCAGGGGTGTCCGTG       AGGGGTGTCCGTG   \n",
              "145        -29  0.80  CGG  TAGCGCGCCGTGGGCGCGTG       CCGTGGGCGCGTG   \n",
              "146        -20  0.80  GGG  CCACCCGCGTAGCGCGCCGT       CGTAGCGCGCCGT   \n",
              "147        -19  0.80  TGG  TCCACCCGCGTAGCGCGCCG       GCGTAGCGCGCCG   \n",
              "0           60  0.70  CGG  CAGTCCGGTGGCCACCGCAT       GTGGCCACCGCAT   \n",
              "\n",
              "     off_target_count    region  \n",
              "140                 0  upstream  \n",
              "145                 0  upstream  \n",
              "146                 1  upstream  \n",
              "147                 1  upstream  \n",
              "0                   0       CDS  "
            ]
          },
          "execution_count": 6,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "# Filter the DataFrame to retain only up to 5 sgRNA sequences per locus_tag\n",
        "filtered_df = sgrna_df.groupby('locus_tag').sample(n=5, random_state=42).sort_values(by='sgrna_loc')\n",
        "filtered_df"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 7,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>ID</th>\n",
              "      <th>Start</th>\n",
              "      <th>End</th>\n",
              "      <th>Strand</th>\n",
              "      <th>ORF</th>\n",
              "      <th>Sequence</th>\n",
              "      <th>PAM</th>\n",
              "      <th>C to T mutations</th>\n",
              "      <th>A to G mutations</th>\n",
              "      <th>0bp mismatches</th>\n",
              "      <th>1bp mismatches</th>\n",
              "      <th>2bp mismatches</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>CY00000009</td>\n",
              "      <td>29</td>\n",
              "      <td>52</td>\n",
              "      <td>-1</td>\n",
              "      <td>-</td>\n",
              "      <td>TGCGGGCAGGGGTGTCCGTG</td>\n",
              "      <td>AGG</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>0</td>\n",
              "      <td>2</td>\n",
              "      <td>50</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>CY00000014</td>\n",
              "      <td>47</td>\n",
              "      <td>70</td>\n",
              "      <td>-1</td>\n",
              "      <td>-</td>\n",
              "      <td>TAGCGCGCCGTGGGCGCGTG</td>\n",
              "      <td>CGG</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>0</td>\n",
              "      <td>6</td>\n",
              "      <td>151</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>CY00000022</td>\n",
              "      <td>56</td>\n",
              "      <td>79</td>\n",
              "      <td>-1</td>\n",
              "      <td>-</td>\n",
              "      <td>CCACCCGCGTAGCGCGCCGT</td>\n",
              "      <td>GGG</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>1</td>\n",
              "      <td>3</td>\n",
              "      <td>76</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>CY00000023</td>\n",
              "      <td>57</td>\n",
              "      <td>80</td>\n",
              "      <td>-1</td>\n",
              "      <td>-</td>\n",
              "      <td>TCCACCCGCGTAGCGCGCCG</td>\n",
              "      <td>TGG</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>1</td>\n",
              "      <td>9</td>\n",
              "      <td>215</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>CY00000012</td>\n",
              "      <td>136</td>\n",
              "      <td>159</td>\n",
              "      <td>-1</td>\n",
              "      <td>-</td>\n",
              "      <td>CAGTCCGGTGGCCACCGCAT</td>\n",
              "      <td>CGG</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>0</td>\n",
              "      <td>4</td>\n",
              "      <td>66</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "           ID  Start  End  Strand ORF              Sequence  PAM  \\\n",
              "0  CY00000009     29   52      -1   -  TGCGGGCAGGGGTGTCCGTG  AGG   \n",
              "1  CY00000014     47   70      -1   -  TAGCGCGCCGTGGGCGCGTG  CGG   \n",
              "2  CY00000022     56   79      -1   -  CCACCCGCGTAGCGCGCCGT  GGG   \n",
              "3  CY00000023     57   80      -1   -  TCCACCCGCGTAGCGCGCCG  TGG   \n",
              "4  CY00000012    136  159      -1   -  CAGTCCGGTGGCCACCGCAT  CGG   \n",
              "\n",
              "   C to T mutations  A to G mutations  0bp mismatches  1bp mismatches  \\\n",
              "0               NaN               NaN               0               2   \n",
              "1               NaN               NaN               0               6   \n",
              "2               NaN               NaN               1               3   \n",
              "3               NaN               NaN               1               9   \n",
              "4               NaN               NaN               0               4   \n",
              "\n",
              "   2bp mismatches  \n",
              "0              50  \n",
              "1             151  \n",
              "2              76  \n",
              "3             215  \n",
              "4              66  "
            ]
          },
          "execution_count": 7,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "import pandas as pd\n",
        "crispy_web_sgrnas = pd.read_csv('CRISPyweb__goe_targeting_upstream_LLPMBPKK_05956_.csv')\n",
        "crispy_web_sgrnas"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 8,
      "metadata": {},
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "✅ Both tables have the same (Sequence, PAM, Strand) rows.\n"
          ]
        }
      ],
      "source": [
        "# 1) Pull out and align the three columns from each DF\n",
        "df1 = (\n",
        "    filtered_df\n",
        "    [['sgrna','pam','sgrna_strand']]\n",
        "    .rename(columns={\n",
        "        'sgrna': 'Sequence',\n",
        "        'pam':   'PAM',\n",
        "        'sgrna_strand': 'Strand'\n",
        "    })\n",
        ")\n",
        "\n",
        "df2 = crispy_web_sgrnas[['Sequence','PAM','Strand']]\n",
        "\n",
        "# 2) Turn each into a set of tuples\n",
        "set1 = set(df1.itertuples(index=False, name=None))\n",
        "set2 = set(df2.itertuples(index=False, name=None))\n",
        "\n",
        "# 3) Compare\n",
        "if set1 == set2:\n",
        "    print(\"✅ Both tables have the same (Sequence, PAM, Strand) rows.\")\n",
        "else:\n",
        "    only1 = set1 - set2\n",
        "    only2 = set2 - set1\n",
        "    print(\"❌ They differ!\")\n",
        "    if only1:\n",
        "        print(\"\\nRows in filtered_df but not in crispy_web_sgrnas:\")\n",
        "        for row in only1:\n",
        "            print(row)\n",
        "    if only2:\n",
        "        print(\"\\nRows in crispy_web_sgrnas but not in filtered_df:\")\n",
        "        for row in only2:\n",
        "            print(row)"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "## Output"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 9,
      "metadata": {},
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "LLPMBPKK_05956_loc_-47\n"
          ]
        },
        {
          "data": {
            "text/plain": [
              "[Contig(o11279),\n",
              " Contig(o11279),\n",
              " Contig(o11279),\n",
              " Contig(o11279),\n",
              " Contig(o11279)]"
            ]
          },
          "execution_count": 9,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "# MAke oligoes\n",
        "list_of_ssDNAs = make_ssDNA_oligos(filtered_df, upstream_ovh = up_homology,\n",
        "                      downstream_ovh=dw_homology)\n",
        "print(list_of_ssDNAs[0].name)\n",
        "\n",
        "# cut plasmid\n",
        "from Bio.Restriction import NcoI\n",
        "linearized_plasmid = sorted(clean_plasmid.cut(NcoI), key=lambda x: len(x), reverse=True)[0]\n",
        "#print(linearized_plasmid)\n",
        "\n",
        "sgRNA_vectors = assemble_plasmids_by_ssDNA_bridging(list_of_ssDNAs,linearized_plasmid)\n",
        "sgRNA_vectors"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 10,
      "metadata": {},
      "outputs": [],
      "source": [
        "# Constructing a meaningful name, ID, and description for the assembled plasmid using user input\n",
        "targeting_info = []\n",
        "for index, row in filtered_df.iterrows():\n",
        "    formatted_str = f\"CRISPRi_{row['locus_tag']}_p{row['sgrna_loc']}\"\n",
        "    targeting_info.append(formatted_str)\n",
        "\n",
        "for i in range(len(sgRNA_vectors)):\n",
        "    sgRNA_vectors[i].name = f'p{targeting_info[i]}_#{i+1}'\n",
        "    sgRNA_vectors[i].id = sgRNA_vectors[i].name  # Using the same value for ID as for name for simplicity\n",
        "    sgRNA_vectors[i].description = f'Assembled plasmid targeting {\", \".join(genes_to_KO)} for single gene KNOCK-DOWN, assembled using StreptoCAD.'\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 11,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>strain_name</th>\n",
              "      <th>locus_tag</th>\n",
              "      <th>gene_loc</th>\n",
              "      <th>gene_strand</th>\n",
              "      <th>sgrna_strand</th>\n",
              "      <th>sgrna_loc</th>\n",
              "      <th>gc</th>\n",
              "      <th>pam</th>\n",
              "      <th>sgrna</th>\n",
              "      <th>sgrna_seed_sequence</th>\n",
              "      <th>off_target_count</th>\n",
              "      <th>region</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>140</th>\n",
              "      <td>.</td>\n",
              "      <td>LLPMBPKK_05956</td>\n",
              "      <td>6815841</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>-47</td>\n",
              "      <td>0.75</td>\n",
              "      <td>AGG</td>\n",
              "      <td>TGCGGGCAGGGGTGTCCGTG</td>\n",
              "      <td>AGGGGTGTCCGTG</td>\n",
              "      <td>0</td>\n",
              "      <td>upstream</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>145</th>\n",
              "      <td>.</td>\n",
              "      <td>LLPMBPKK_05956</td>\n",
              "      <td>6815841</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>-29</td>\n",
              "      <td>0.80</td>\n",
              "      <td>CGG</td>\n",
              "      <td>TAGCGCGCCGTGGGCGCGTG</td>\n",
              "      <td>CCGTGGGCGCGTG</td>\n",
              "      <td>0</td>\n",
              "      <td>upstream</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>146</th>\n",
              "      <td>.</td>\n",
              "      <td>LLPMBPKK_05956</td>\n",
              "      <td>6815841</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>-20</td>\n",
              "      <td>0.80</td>\n",
              "      <td>GGG</td>\n",
              "      <td>CCACCCGCGTAGCGCGCCGT</td>\n",
              "      <td>CGTAGCGCGCCGT</td>\n",
              "      <td>1</td>\n",
              "      <td>upstream</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>147</th>\n",
              "      <td>.</td>\n",
              "      <td>LLPMBPKK_05956</td>\n",
              "      <td>6815841</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>-19</td>\n",
              "      <td>0.80</td>\n",
              "      <td>TGG</td>\n",
              "      <td>TCCACCCGCGTAGCGCGCCG</td>\n",
              "      <td>GCGTAGCGCGCCG</td>\n",
              "      <td>1</td>\n",
              "      <td>upstream</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>.</td>\n",
              "      <td>LLPMBPKK_05956</td>\n",
              "      <td>6816041</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>60</td>\n",
              "      <td>0.70</td>\n",
              "      <td>CGG</td>\n",
              "      <td>CAGTCCGGTGGCCACCGCAT</td>\n",
              "      <td>GTGGCCACCGCAT</td>\n",
              "      <td>0</td>\n",
              "      <td>CDS</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "    strain_name       locus_tag  gene_loc  gene_strand  sgrna_strand  \\\n",
              "140           .  LLPMBPKK_05956   6815841            1            -1   \n",
              "145           .  LLPMBPKK_05956   6815841            1            -1   \n",
              "146           .  LLPMBPKK_05956   6815841            1            -1   \n",
              "147           .  LLPMBPKK_05956   6815841            1            -1   \n",
              "0             .  LLPMBPKK_05956   6816041            1            -1   \n",
              "\n",
              "     sgrna_loc    gc  pam                 sgrna sgrna_seed_sequence  \\\n",
              "140        -47  0.75  AGG  TGCGGGCAGGGGTGTCCGTG       AGGGGTGTCCGTG   \n",
              "145        -29  0.80  CGG  TAGCGCGCCGTGGGCGCGTG       CCGTGGGCGCGTG   \n",
              "146        -20  0.80  GGG  CCACCCGCGTAGCGCGCCGT       CGTAGCGCGCCGT   \n",
              "147        -19  0.80  TGG  TCCACCCGCGTAGCGCGCCG       GCGTAGCGCGCCG   \n",
              "0           60  0.70  CGG  CAGTCCGGTGGCCACCGCAT       GTGGCCACCGCAT   \n",
              "\n",
              "     off_target_count    region  \n",
              "140                 0  upstream  \n",
              "145                 0  upstream  \n",
              "146                 1  upstream  \n",
              "147                 1  upstream  \n",
              "0                   0       CDS  "
            ]
          },
          "execution_count": 11,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "filtered_df"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 12,
      "metadata": {},
      "outputs": [],
      "source": [
        "# # annotate the plasmids with the sgRNA \n",
        "# from Bio.Seq import Seq\n",
        "\n",
        "# def find_sgRNA_in_plasmid(plasmid_seq, sgRNA_seq, strand):\n",
        "#     \"\"\"\n",
        "#     Find the sgRNA sequence in the plasmid. Returns (start, end, strand) or None if not found.\n",
        "#     - plasmid_seq: Bio.Seq object (can also use str)\n",
        "#     - sgRNA_seq: str (should be 5'-3' sequence)\n",
        "#     - strand: +1 (sense) or -1 (antisense)\n",
        "#     \"\"\"\n",
        "#     # Search on forward strand\n",
        "#     idx = str(plasmid_seq).find(sgRNA_seq)\n",
        "#     if idx != -1 and strand == 1:\n",
        "#         return idx, idx + len(sgRNA_seq), 1\n",
        "#     # Search on reverse complement\n",
        "#     rc_sgRNA_seq = str(Seq(sgRNA_seq).reverse_complement())\n",
        "#     idx = str(plasmid_seq).find(rc_sgRNA_seq)\n",
        "#     if idx != -1 and strand == -1:\n",
        "#         return idx, idx + len(rc_sgRNA_seq), -1\n",
        "#     # Not found\n",
        "#     return None\n",
        "\n",
        "# from Bio.SeqFeature import SeqFeature, FeatureLocation\n",
        "# filtered_df = filtered_df.reset_index(drop=True)\n",
        "\n",
        "# for j, plasmid in enumerate(sgRNA_vectors):\n",
        "#     for i, row in filtered_df.iterrows():\n",
        "#         sgRNA_seq = str(row['sgrna'])\n",
        "#         strand = int(row['sgrna_strand'])\n",
        "#         res = find_sgRNA_in_plasmid(Dseqrecord(plasmid.seq), sgRNA_seq, strand)\n",
        "#         if res:\n",
        "#             start, end, strand = res\n",
        "#             location = FeatureLocation(start, end, strand=strand)\n",
        "#             print(location)\n",
        "#             # Add feature with proper qualifiers...\n",
        "#             # (Add only if this feature doesn't exist yet, to avoid duplicates)\n",
        "#             print(f\"Found {sgRNA_seq} in {plasmid.name} at {start}-{end} (strand {strand})\")\n",
        "#         else:\n",
        "#             print(f\"sgRNA {sgRNA_seq} not found in {plasmid.name}\")\n",
        "\n",
        "\n",
        "#     qualifiers = {\n",
        "#         \"locus_tag\": row[\"locus_tag\"],\n",
        "#         \"pam\": row[\"pam\"],\n",
        "#         \"sgrna_sequence\": row[\"sgrna\"],\n",
        "#         \"seed_sequence\": row[\"sgrna_seed_sequence\"],\n",
        "#         \"off_target_count\": str(row[\"off_target_count\"]),\n",
        "#         \"region\": row[\"region\"],\n",
        "#         \"note\": f\"sgRNA targeting {row['locus_tag']} at {row['sgrna_loc']} ({row['region']})\"\n",
        "#     }\n",
        "\n",
        "#     feature = SeqFeature(\n",
        "#         location=location,\n",
        "#         type=\"sgRNA\",\n",
        "#         qualifiers=qualifiers\n",
        "#     )\n",
        "#     plasmid.features.append(feature)\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 13,
      "metadata": {},
      "outputs": [
        {
          "name": "stderr",
          "output_type": "stream",
          "text": [
            "/Users/lucaslevassor/opt/anaconda3/envs/streptocad/lib/python3.11/site-packages/Bio/SeqIO/InsdcIO.py:728: BiopythonWarning: Increasing length of locus line to allow long name. This will result in fields that are not in usual positions.\n",
            "  warnings.warn(\n"
          ]
        },
        {
          "data": {
            "text/html": [
              "\n",
              "                <table style=\"padding:10px 10px;\n",
              "                word-break:normal;\n",
              "                border-color:#fe0000;\n",
              "                border-collapse:collapse;\n",
              "                border-spacing:1;\n",
              "                font-family:monospace;\n",
              "                font-size:large;\n",
              "                font-weight:bold;\n",
              "                text-align:left;\n",
              "                border: 5px solid red;\">\n",
              "                <thead>\n",
              "                  <tr style=\"color:#0000FF;border: 1px solid;text-align:left;\">\n",
              "                    <th style=\"color:#fe0000;border: 1px solid;text-align:center;font-size:xxx-large;text-align:left;\">&#9888</th>\n",
              "                    <th style=\"color:#f56b00;border: 1px solid;text-align:left;\" colspan=\"2\">Sequence change</th>\n",
              "                  </tr>\n",
              "                </thead>\n",
              "                <tbody>\n",
              "                  <tr style=\"color:#0000FF;border: 1px solid;text-align:left;\">\n",
              "                    <td>Filename</td>\n",
              "                    <td style=\"color:#fe0000;border: 1px solid;text-align:left;\"><a href='plasmids/pCRISPRi_LLPMBPKK_05956_p-47_#1.gb' target='_blank'>plasmids/pCRISPRi_LLPMBPKK_05956_p-47_#1.gb</a></td>\n",
              "                    <td style=\"color:#32cb00;border: 1px solid;text-align:left;\"><a href='plasmids/pCRISPRi_LLPMBPKK_05956_p-47_#1_OLD_1755081159817839.gb' target='_blank'>plasmids/pCRISPRi_LLPMBPKK_05956_p-47_#1_OLD_1755081159817839.gb</a></td>\n",
              "                  </tr>\n",
              "                  <tr style=\"color:#0000FF;border: 1px solid;text-align:left;\">\n",
              "                    <td >Saved</td>\n",
              "                    <td style=\"color:#fe0000;border: 1px solid;text-align:left;\">2025-08-13T12:32:39.845451</td>\n",
              "                    <td style=\"color:#32cb00;border: 1px solid;text-align:left;\">2025-07-24T11:53:20.118708</td>\n",
              "                  </tr>\n",
              "                  <tr style=\"color:#0000FF;border: 1px solid;text-align:left;\">\n",
              "                    <td>Length</td>\n",
              "                    <td style=\"color:#fe0000;border: 1px solid;text-align:left;\">11279</td>\n",
              "                    <td style=\"color:#32cb00;border: 1px solid;text-align:left;\">11276</td>\n",
              "                  </tr>\n",
              "                  <tr style=\"color:#0000FF;border: 1px solid;text-align:left;\">\n",
              "                    <td>uSEGUID</td>\n",
              "                    <td style=\"color:#fe0000;border: 1px solid;text-align:left;\">CvyJ8Vq49Dncq9Ow0uS76_mGiJ4</td>\n",
              "                    <td style=\"color:#32cb00;border: 1px solid;text-align:left;\">3FXsforzoL7ruEBRvpoKmvlwcZw</td>\n",
              "                  </tr>\n",
              "                  <tr style=\"color:#0000FF;border: 1px solid;text-align:left;\">\n",
              "                    <td>cSEGUID</td>\n",
              "                    <td style=\"color:#fe0000;border: 1px solid;text-align:left;\">47ZMiGtX_uPlytF19e91LjohsVU</td>\n",
              "                    <td style=\"color:#32cb00;border: 1px solid;text-align:left;\">FZfohrsBwje4Xxqit-bYc0m9HYc</td>\n",
              "                  </tr>\n",
              "                </tbody>\n",
              "                </table>\n",
              "                "
            ]
          },
          "metadata": {},
          "output_type": "display_data"
        },
        {
          "data": {
            "text/html": [
              "\n",
              "                <table style=\"padding:10px 10px;\n",
              "                word-break:normal;\n",
              "                border-color:#fe0000;\n",
              "                border-collapse:collapse;\n",
              "                border-spacing:1;\n",
              "                font-family:monospace;\n",
              "                font-size:large;\n",
              "                font-weight:bold;\n",
              "                text-align:left;\n",
              "                border: 5px solid red;\">\n",
              "                <thead>\n",
              "                  <tr style=\"color:#0000FF;border: 1px solid;text-align:left;\">\n",
              "                    <th style=\"color:#fe0000;border: 1px solid;text-align:center;font-size:xxx-large;text-align:left;\">&#9888</th>\n",
              "                    <th style=\"color:#f56b00;border: 1px solid;text-align:left;\" colspan=\"2\">Sequence change</th>\n",
              "                  </tr>\n",
              "                </thead>\n",
              "                <tbody>\n",
              "                  <tr style=\"color:#0000FF;border: 1px solid;text-align:left;\">\n",
              "                    <td>Filename</td>\n",
              "                    <td style=\"color:#fe0000;border: 1px solid;text-align:left;\"><a href='plasmids/pCRISPRi_LLPMBPKK_05956_p-29_#2.gb' target='_blank'>plasmids/pCRISPRi_LLPMBPKK_05956_p-29_#2.gb</a></td>\n",
              "                    <td style=\"color:#32cb00;border: 1px solid;text-align:left;\"><a href='plasmids/pCRISPRi_LLPMBPKK_05956_p-29_#2_OLD_1755081159851255.gb' target='_blank'>plasmids/pCRISPRi_LLPMBPKK_05956_p-29_#2_OLD_1755081159851255.gb</a></td>\n",
              "                  </tr>\n",
              "                  <tr style=\"color:#0000FF;border: 1px solid;text-align:left;\">\n",
              "                    <td >Saved</td>\n",
              "                    <td style=\"color:#fe0000;border: 1px solid;text-align:left;\">2025-08-13T12:32:39.869860</td>\n",
              "                    <td style=\"color:#32cb00;border: 1px solid;text-align:left;\">2025-07-24T12:05:28.478516</td>\n",
              "                  </tr>\n",
              "                  <tr style=\"color:#0000FF;border: 1px solid;text-align:left;\">\n",
              "                    <td>Length</td>\n",
              "                    <td style=\"color:#fe0000;border: 1px solid;text-align:left;\">11279</td>\n",
              "                    <td style=\"color:#32cb00;border: 1px solid;text-align:left;\">11276</td>\n",
              "                  </tr>\n",
              "                  <tr style=\"color:#0000FF;border: 1px solid;text-align:left;\">\n",
              "                    <td>uSEGUID</td>\n",
              "                    <td style=\"color:#fe0000;border: 1px solid;text-align:left;\">-BOZ7k5hE44KOAY256INu1t3GkM</td>\n",
              "                    <td style=\"color:#32cb00;border: 1px solid;text-align:left;\">zGzXeG5sBYPY0HC_yHe3bXg_dmw</td>\n",
              "                  </tr>\n",
              "                  <tr style=\"color:#0000FF;border: 1px solid;text-align:left;\">\n",
              "                    <td>cSEGUID</td>\n",
              "                    <td style=\"color:#fe0000;border: 1px solid;text-align:left;\">x92XZbZAfxxj0DEnAKZgUljlaS8</td>\n",
              "                    <td style=\"color:#32cb00;border: 1px solid;text-align:left;\">sbs59QsY5WlraWzr1Y8PoxKR2ec</td>\n",
              "                  </tr>\n",
              "                </tbody>\n",
              "                </table>\n",
              "                "
            ]
          },
          "metadata": {},
          "output_type": "display_data"
        },
        {
          "data": {
            "text/html": [
              "<font face=monospace><a href='plasmids/pCRISPRi_LLPMBPKK_05956_p-20_#3.gb' target='_blank'>plasmids/pCRISPRi_LLPMBPKK_05956_p-20_#3.gb</a></font><br>"
            ]
          },
          "metadata": {},
          "output_type": "display_data"
        },
        {
          "data": {
            "text/html": [
              "<font face=monospace><a href='plasmids/pCRISPRi_LLPMBPKK_05956_p-19_#4.gb' target='_blank'>plasmids/pCRISPRi_LLPMBPKK_05956_p-19_#4.gb</a></font><br>"
            ]
          },
          "metadata": {},
          "output_type": "display_data"
        },
        {
          "data": {
            "text/html": [
              "<font face=monospace><a href='plasmids/pCRISPRi_LLPMBPKK_05956_p60_#5.gb' target='_blank'>plasmids/pCRISPRi_LLPMBPKK_05956_p60_#5.gb</a></font><br>"
            ]
          },
          "metadata": {},
          "output_type": "display_data"
        }
      ],
      "source": [
        "print_plasmids = True\n",
        "\n",
        "if print_plasmids: \n",
        "    for vector in sgRNA_vectors: \n",
        "        vector.write(f\"plasmids/{vector.id}.gb\")"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 14,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>plasmid_name</th>\n",
              "      <th>date</th>\n",
              "      <th>original_plasmid</th>\n",
              "      <th>integration</th>\n",
              "      <th>size</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>pCRISPRi_LLPMBPKK_05956_p-47_#1</td>\n",
              "      <td>2025-08-13</td>\n",
              "      <td>.</td>\n",
              "      <td>sgRNA_LLPMBPKK_05956(-47)</td>\n",
              "      <td>11279</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>pCRISPRi_LLPMBPKK_05956_p-29_#2</td>\n",
              "      <td>2025-08-13</td>\n",
              "      <td>.</td>\n",
              "      <td>sgRNA_LLPMBPKK_05956(-29)</td>\n",
              "      <td>11279</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>pCRISPRi_LLPMBPKK_05956_p-20_#3</td>\n",
              "      <td>2025-08-13</td>\n",
              "      <td>.</td>\n",
              "      <td>sgRNA_LLPMBPKK_05956(-20)</td>\n",
              "      <td>11279</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>pCRISPRi_LLPMBPKK_05956_p-19_#4</td>\n",
              "      <td>2025-08-13</td>\n",
              "      <td>.</td>\n",
              "      <td>sgRNA_LLPMBPKK_05956(-19)</td>\n",
              "      <td>11279</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>pCRISPRi_LLPMBPKK_05956_p60_#5</td>\n",
              "      <td>2025-08-13</td>\n",
              "      <td>.</td>\n",
              "      <td>sgRNA_LLPMBPKK_05956(60)</td>\n",
              "      <td>11279</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "                      plasmid_name        date original_plasmid  \\\n",
              "0  pCRISPRi_LLPMBPKK_05956_p-47_#1  2025-08-13                .   \n",
              "1  pCRISPRi_LLPMBPKK_05956_p-29_#2  2025-08-13                .   \n",
              "2  pCRISPRi_LLPMBPKK_05956_p-20_#3  2025-08-13                .   \n",
              "3  pCRISPRi_LLPMBPKK_05956_p-19_#4  2025-08-13                .   \n",
              "4   pCRISPRi_LLPMBPKK_05956_p60_#5  2025-08-13                .   \n",
              "\n",
              "                 integration   size  \n",
              "0  sgRNA_LLPMBPKK_05956(-47)  11279  \n",
              "1  sgRNA_LLPMBPKK_05956(-29)  11279  \n",
              "2  sgRNA_LLPMBPKK_05956(-20)  11279  \n",
              "3  sgRNA_LLPMBPKK_05956(-19)  11279  \n",
              "4   sgRNA_LLPMBPKK_05956(60)  11279  "
            ]
          },
          "execution_count": 14,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "integration_names = filtered_df.apply(lambda row: f\"sgRNA_{row['locus_tag']}({row['sgrna_loc']})\", axis=1).tolist()\n",
        "plasmid_metadata_df = extract_metadata_to_dataframe(sgRNA_vectors,\n",
        "                                                    clean_plasmid,\n",
        "                                                    integration_names)\n",
        "\n",
        "plasmid_metadata_df"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "### IDT primers"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 15,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>Name</th>\n",
              "      <th>Sequence</th>\n",
              "      <th>Concentration</th>\n",
              "      <th>Purification</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>LLPMBPKK_05956_loc_-47</td>\n",
              "      <td>CGGTTGGTAGGATCGACGGCTGCGGGCAGGGGTGTCCGTGGTTTTA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>LLPMBPKK_05956_loc_-29</td>\n",
              "      <td>CGGTTGGTAGGATCGACGGCTAGCGCGCCGTGGGCGCGTGGTTTTA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>LLPMBPKK_05956_loc_-20</td>\n",
              "      <td>CGGTTGGTAGGATCGACGGCCCACCCGCGTAGCGCGCCGTGTTTTA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>LLPMBPKK_05956_loc_-19</td>\n",
              "      <td>CGGTTGGTAGGATCGACGGCTCCACCCGCGTAGCGCGCCGGTTTTA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>LLPMBPKK_05956_loc_60</td>\n",
              "      <td>CGGTTGGTAGGATCGACGGCCAGTCCGGTGGCCACCGCATGTTTTA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "                     Name                                           Sequence  \\\n",
              "0  LLPMBPKK_05956_loc_-47  CGGTTGGTAGGATCGACGGCTGCGGGCAGGGGTGTCCGTGGTTTTA...   \n",
              "1  LLPMBPKK_05956_loc_-29  CGGTTGGTAGGATCGACGGCTAGCGCGCCGTGGGCGCGTGGTTTTA...   \n",
              "2  LLPMBPKK_05956_loc_-20  CGGTTGGTAGGATCGACGGCCCACCCGCGTAGCGCGCCGTGTTTTA...   \n",
              "3  LLPMBPKK_05956_loc_-19  CGGTTGGTAGGATCGACGGCTCCACCCGCGTAGCGCGCCGGTTTTA...   \n",
              "4   LLPMBPKK_05956_loc_60  CGGTTGGTAGGATCGACGGCCAGTCCGGTGGCCACCGCATGTTTTA...   \n",
              "\n",
              "  Concentration Purification  \n",
              "0          25nm          STD  \n",
              "1          25nm          STD  \n",
              "2          25nm          STD  \n",
              "3          25nm          STD  \n",
              "4          25nm          STD  "
            ]
          },
          "execution_count": 15,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "idt_primers=primers_to_IDT(list_of_ssDNAs)\n",
        "idt_primers"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "## Folder with all the generated I/O"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 16,
      "metadata": {},
      "outputs": [],
      "source": [
        "generate_data_folder = False \n",
        "\n",
        "if generate_data_folder:\n",
        "    input_files = [\n",
        "        {\"name\": \"input_genome.gb\", \"content\": genome},\n",
        "        {\"name\": \"input_plasmid.gb\", \"content\": clean_plasmid}\n",
        "    ]\n",
        "\n",
        "    output_files = [\n",
        "        {\"name\": \"cBEST_w_sgRNAs.gb\", \"content\": sgRNA_vectors}, # LIST OF Dseqrecords\n",
        "        {\"name\": \"full_idt.csv\", \"content\": idt_primers},\n",
        "        {\"name\": \"sgrna_df.csv\", \"content\": sgrna_df},\n",
        "        {\"name\": \"filtered_df.csv\", \"content\": filtered_df},\n",
        "        {\"name\": \"plasmid_metadata_df.csv\", \"content\": plasmid_metadata_df},\n",
        "    ]\n",
        "\n",
        "    input_values = {\n",
        "        \"genes_to_knockout\": genes_to_KO,\n",
        "\n",
        "\n",
        "        \"filtering_metrics\": {\n",
        "            \"gc_upper\": gc_upper,\n",
        "            \"gc_lower\": gc_lower,\n",
        "            \"off_target_seed\": off_target_seed,\n",
        "            \"off_target_upper\": off_target_upper,\n",
        "            \"cas_type\": cas_type,\n",
        "            \"number_of_sgRNAs_per_group\": number_of_sgRNAs_per_group,\n",
        "            'extension_to_promoter_region':extension_to_promoter_region,\n",
        "\n",
        "        },\n",
        "        \"overlapping_sequences\": {\n",
        "            \"up_homology\": str(up_homology),\n",
        "            \"dw_homology\": str(dw_homology)\n",
        "        }\n",
        "    }\n",
        "\n",
        "\n",
        "    # Paths to Markdown files\n",
        "    markdown_file_paths = [\n",
        "        \"../../protocols/conjugation_protcol.md\",\n",
        "        \"../../protocols/single_target_crispr_plasmid_protcol.md\"\n",
        "\n",
        "    ]\n",
        "\n",
        "\n",
        "\n",
        "    # Data and time\n",
        "    timestamp = datetime.utcnow().isoformat()\n",
        "\n",
        "    # Create project directory structure\n",
        "    project_directory = ProjectDirectory(\n",
        "        project_name=f\"CRISPRi_workflow_{timestamp}\",\n",
        "        input_files=input_files,\n",
        "        output_files=output_files,\n",
        "        input_values=input_values,\n",
        "        markdown_file_paths=markdown_file_paths\n",
        "    )\n",
        "\n",
        "\n",
        "    # DO You want to save the folder? \n",
        "    save_zip_folder = True \n",
        "\n",
        "    if save_zip_folder: \n",
        "        # Generate the project directory structure and get the zip content\n",
        "        zip_content = project_directory.create_directory_structure(create_directories=False)\n",
        "\n",
        "        # Save the zip file to disk (optional)\n",
        "        with open(\"project_structure.zip\", \"wb\") as f:\n",
        "            f.write(zip_content)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {},
      "outputs": [],
      "source": []
    }
  ],
  "metadata": {
    "kernelspec": {
      "display_name": "streptocad",
      "language": "python",
      "name": "python3"
    },
    "language_info": {
      "codemirror_mode": {
        "name": "ipython",
        "version": 3
      },
      "file_extension": ".py",
      "mimetype": "text/x-python",
      "name": "python",
      "nbconvert_exporter": "python",
      "pygments_lexer": "ipython3",
      "version": "3.11.11"
    }
  },
  "nbformat": 4,
  "nbformat_minor": 2
}
