{
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "## W2-CRISPR-cBEST workflow\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "Base-editing with CRISPR-BEST is a DSB-free method that uses a single sgRNA to target a specific genomic location with single-nucleotide–resolution(Figure 3A) which can enable in-vivo protein engineering or as it is often used, for introducing early stop-codons10,12. It can be used to introduce targeted point mutations in the genome of Streptomyces, facilitating studies on gene function and protein engineering. To get started quickly users can download the pCRISPR-cBEST plasmid file, the genome file for S.coelicor (A3) and streptoAIM will generate your plasmids based on the genes you chose to target (Figure 3, S5). "
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 1,
      "metadata": {},
      "outputs": [],
      "source": [
        "import sys\n",
        "import os\n",
        "\n",
        "# Ensure the src directory is in the Python path\n",
        "project_root = os.path.abspath(os.path.join(os.getcwd(), '../../'))\n",
        "if project_root not in sys.path:\n",
        "    sys.path.insert(0, project_root)\n",
        "\n",
        "import os\n",
        "import pandas as pd\n",
        "from pydna.dseqrecord import Dseqrecord\n",
        "from datetime import datetime\n",
        "from Bio.SeqRecord import SeqRecord\n",
        "from Bio import SeqIO\n",
        "\n",
        "from streptocad.sequence_loading.sequence_loading import load_and_process_gene_sequences, load_and_process_plasmid, load_and_process_genome_sequences, annotate_dseqrecord, check_and_convert_input, process_specified_gene_sequences_from_record\n",
        "from streptocad.utils import polymerase_dict,ProjectDirectory, extract_metadata_to_dataframe\n",
        "from streptocad.crispr.guideRNAcas3_9_12 import extract_sgRNAs, SgRNAargs\n",
        "from streptocad.cloning.ssDNA_bridging import assemble_plasmids_by_ssDNA_bridging, make_ssDNA_oligos\n",
        "from streptocad.crispr.crispr_best import identify_base_editing_sites, filter_sgrnas_for_base_editing, process_base_editing\n",
        "from streptocad.cloning.plasmid_processing import annotate_plasmid_with_sgrnas\n",
        "from streptocad.primers.primer_generation import find_best_check_primers_from_genome, create_idt_order_dataframe, primers_to_IDT\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "## INPUT"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 2,
      "metadata": {},
      "outputs": [],
      "source": [
        "# Inputs\n",
        "# 1 Add genome of choice (genbank, fasta)\n",
        "path_to_genome = '../../data/genomes/Streptomyces_coelicolor_A3_chromosome.gb'\n",
        "genome = load_and_process_genome_sequences(path_to_genome)[0]\n",
        "\n",
        "# 2 Add plasmid \n",
        "path_to_plasmid = '../../data/plasmids/ pCRISPR-cBEST Sequences.gbk'\n",
        "clean_plasmid = load_and_process_plasmid(path_to_plasmid)\n",
        "\n",
        "# 3 Choose genes to knock out (list)\n",
        "#genes_to_KO = ['SCO5087','SCO5088', 'SCO5089']\n",
        "genes_to_KO = ['80000-100000', '4000-7000', '9000-14000','15000-20000']\n",
        "\n",
        "\n",
        "#### Advanced settings ####\n",
        "# 3 Choose polymerase and target melting temperature FOR CHECKING PRIMERS\n",
        "chosen_polymerase = polymerase_dict['Phusion High-Fidelity DNA Polymerase (GC Buffer)']\n",
        "\n",
        "melting_temperature = 65\n",
        "primer_concentration = 0.4 \n",
        "primer_number_increment = 1\n",
        "flanking_region_number = 500\n",
        "\n",
        "# 4 Filtering metrics for sgRNAs\n",
        "gc_upper = 0.99\n",
        "gc_lower = 0.01\n",
        "off_target_seed = 13\n",
        "off_target_upper = 10\n",
        "cas_type='cas9'\n",
        "number_of_sgRNAs_per_group = 5\n",
        "only_stop_codons = True\n",
        "flanking_region = 500\n",
        "\n",
        "# 6 Choose overlapping sequences for our plasmid we can use the following\n",
        "#As per the article **\"CRISPR–Cas9, CRISPRi and CRISPR-BEST-mediated genetic manipulation in streptomycetes\"** we need the following oligoes: \n",
        "#CGGTTGGTAGGATCGACGGC **-N20-** GTTTTAGAGCTAGAAATAGC\n",
        "up_homology = Dseqrecord('CGGTTGGTAGGATCGACGGC')\n",
        "dw_homology = Dseqrecord('GTTTTAGAGCTAGAAATAGC')"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {},
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "pCRISPR-cBEST(125689)\n"
          ]
        }
      ],
      "source": [
        "print(clean_plasmid.id)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/plain": [
              "Dseqrecord(o12247)"
            ]
          },
          "execution_count": 4,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "clean_plasmid"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {},
      "outputs": [],
      "source": [
        "from teemi.design.fetch_sequences import read_genbank_files     \n",
        "sequences = read_genbank_files(path_to_plasmid)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/plain": [
              "[SeqRecord(seq=Seq('GGCCGTCGACGACGACGCGTCCGCCTGCCTGCTTTTGGCCGGCGGGCCGGGCCG...CCA'), id='pCRISPR-cBEST(125689)', name='pCRISPR-cBEST', description='C to T base editor for actinomycetes', dbxrefs=[])]"
            ]
          },
          "execution_count": 6,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "sequences"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "# Computation"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/plain": [
              "25824"
            ]
          },
          "execution_count": 7,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "len(genome.features)\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {},
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "True\n",
            "this is the target dict [{'region_1_(80000-100000)': [80000, 100000]}, {'region_2_(4000-7000)': [4000, 7000]}, {'region_3_(9000-14000)': [9000, 14000]}, {'region_4_(15000-20000)': [15000, 20000]}]\n",
            "region_1_(80000-100000) [80000, 100000]\n",
            "region_2_(4000-7000) [4000, 7000]\n",
            "region_3_(9000-14000) [9000, 14000]\n",
            "region_4_(15000-20000) [15000, 20000]\n"
          ]
        },
        {
          "data": {
            "text/plain": [
              "25828"
            ]
          },
          "execution_count": 8,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "target_dict, genes_to_KO, annotation_input = check_and_convert_input(genes_to_KO)\n",
        "\n",
        "print(annotation_input)\n",
        "if annotation_input == True:\n",
        "    genome = annotate_dseqrecord(genome, target_dict)\n",
        "\n",
        "\n",
        "len(genome.features)\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/plain": [
              "['region_1_(80000-100000)',\n",
              " 'region_2_(4000-7000)',\n",
              " 'region_3_(9000-14000)',\n",
              " 'region_4_(15000-20000)']"
            ]
          },
          "execution_count": 9,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "genes_to_KO"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "We need to check the input genes so we can find sgRNAs. "
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {},
      "outputs": [],
      "source": [
        "# Initialize SgRNAargs with desired parameters\n",
        "args = SgRNAargs(genome, \n",
        "                genes_to_KO,\n",
        "                step=['find', 'filter'],\n",
        "                gc_upper = gc_upper,\n",
        "                gc_lower = gc_lower,\n",
        "                off_target_seed = off_target_seed,\n",
        "                off_target_upper = off_target_upper,\n",
        "                cas_type='cas9',\n",
        "                )"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {},
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "sgRNA generated were outside the designated border in region_1_(80000-100000). To incorporate this extent borders. Skipping to next locus tag.\n",
            "sgRNA generated were outside the designated border in region_1_(80000-100000). To incorporate this extent borders. Skipping to next locus tag.\n",
            "sgRNA generated were outside the designated border in region_2_(4000-7000). To incorporate this extent borders. Skipping to next locus tag.\n",
            "sgRNA generated were outside the designated border in region_2_(4000-7000). To incorporate this extent borders. Skipping to next locus tag.\n",
            "sgRNA generated were outside the designated border in region_2_(4000-7000). To incorporate this extent borders. Skipping to next locus tag.\n",
            "sgRNA generated were outside the designated border in region_2_(4000-7000). To incorporate this extent borders. Skipping to next locus tag.\n",
            "sgRNA generated were outside the designated border in region_2_(4000-7000). To incorporate this extent borders. Skipping to next locus tag.\n",
            "sgRNA generated were outside the designated border in region_2_(4000-7000). To incorporate this extent borders. Skipping to next locus tag.\n",
            "sgRNA generated were outside the designated border in region_3_(9000-14000). To incorporate this extent borders. Skipping to next locus tag.\n",
            "sgRNA generated were outside the designated border in region_3_(9000-14000). To incorporate this extent borders. Skipping to next locus tag.\n",
            "sgRNA generated were outside the designated border in region_3_(9000-14000). To incorporate this extent borders. Skipping to next locus tag.\n",
            "sgRNA generated were outside the designated border in region_3_(9000-14000). To incorporate this extent borders. Skipping to next locus tag.\n",
            "sgRNA generated were outside the designated border in region_3_(9000-14000). To incorporate this extent borders. Skipping to next locus tag.\n",
            "sgRNA generated were outside the designated border in region_4_(15000-20000). To incorporate this extent borders. Skipping to next locus tag.\n",
            "sgRNA generated were outside the designated border in region_4_(15000-20000). To incorporate this extent borders. Skipping to next locus tag.\n",
            "sgRNA generated were outside the designated border in region_4_(15000-20000). To incorporate this extent borders. Skipping to next locus tag.\n",
            "sgRNA generated were outside the designated border in region_4_(15000-20000). To incorporate this extent borders. Skipping to next locus tag.\n",
            "sgRNA generated were outside the designated border in region_4_(15000-20000). To incorporate this extent borders. Skipping to next locus tag.\n",
            "sgRNA generated were outside the designated border in region_4_(15000-20000). To incorporate this extent borders. Skipping to next locus tag.\n"
          ]
        },
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>strain_name</th>\n",
              "      <th>locus_tag</th>\n",
              "      <th>gene_loc</th>\n",
              "      <th>gene_strand</th>\n",
              "      <th>sgrna_strand</th>\n",
              "      <th>sgrna_loc</th>\n",
              "      <th>gc</th>\n",
              "      <th>pam</th>\n",
              "      <th>sgrna</th>\n",
              "      <th>sgrna_seed_sequence</th>\n",
              "      <th>off_target_count</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>3517</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>region_1_(80000-100000)</td>\n",
              "      <td>80001</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>6574</td>\n",
              "      <td>0.60</td>\n",
              "      <td>TGG</td>\n",
              "      <td>GCCACATTCACTACGACCCC</td>\n",
              "      <td>TCACTACGACCCC</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2538</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>region_1_(80000-100000)</td>\n",
              "      <td>80001</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>16046</td>\n",
              "      <td>0.65</td>\n",
              "      <td>GGG</td>\n",
              "      <td>CCCTGATCCGAGTGTGTGCC</td>\n",
              "      <td>CCGAGTGTGTGCC</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2539</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>region_1_(80000-100000)</td>\n",
              "      <td>80001</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>16045</td>\n",
              "      <td>0.60</td>\n",
              "      <td>CGG</td>\n",
              "      <td>ACCCTGATCCGAGTGTGTGC</td>\n",
              "      <td>TCCGAGTGTGTGC</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2540</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>region_1_(80000-100000)</td>\n",
              "      <td>80001</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>16018</td>\n",
              "      <td>0.55</td>\n",
              "      <td>GGG</td>\n",
              "      <td>ACGACAATCCGGTGATCAGC</td>\n",
              "      <td>TCCGGTGATCAGC</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2541</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>region_1_(80000-100000)</td>\n",
              "      <td>80001</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>16017</td>\n",
              "      <td>0.50</td>\n",
              "      <td>CGG</td>\n",
              "      <td>TACGACAATCCGGTGATCAG</td>\n",
              "      <td>ATCCGGTGATCAG</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>...</th>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>5096</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>region_3_(9000-14000)</td>\n",
              "      <td>9001</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>1710</td>\n",
              "      <td>0.80</td>\n",
              "      <td>AGG</td>\n",
              "      <td>GGAATCGGGCGAGGGCGGCC</td>\n",
              "      <td>GGCGAGGGCGGCC</td>\n",
              "      <td>9</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4158</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>region_1_(80000-100000)</td>\n",
              "      <td>80001</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>137</td>\n",
              "      <td>0.85</td>\n",
              "      <td>TGG</td>\n",
              "      <td>CGAGCACACGGCGCGGGCGC</td>\n",
              "      <td>ACGGCGCGGGCGC</td>\n",
              "      <td>9</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1982</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>region_1_(80000-100000)</td>\n",
              "      <td>80001</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>18880</td>\n",
              "      <td>0.80</td>\n",
              "      <td>TGG</td>\n",
              "      <td>CACCTACGGCGCGCTCGGCG</td>\n",
              "      <td>GGCGCGCTCGGCG</td>\n",
              "      <td>9</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>6674</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>region_4_(15000-20000)</td>\n",
              "      <td>15001</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>3011</td>\n",
              "      <td>0.90</td>\n",
              "      <td>GGG</td>\n",
              "      <td>CGGCGCCGCCGACGCCGACG</td>\n",
              "      <td>GCCGACGCCGACG</td>\n",
              "      <td>10</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2854</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>region_1_(80000-100000)</td>\n",
              "      <td>80001</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>12959</td>\n",
              "      <td>0.85</td>\n",
              "      <td>GGG</td>\n",
              "      <td>CGCTCTGGGCGTCGCCGCCG</td>\n",
              "      <td>GGCGTCGCCGCCG</td>\n",
              "      <td>10</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>7026 rows × 11 columns</p>\n",
              "</div>"
            ],
            "text/plain": [
              "      strain_name                locus_tag  gene_loc  gene_strand  \\\n",
              "3517  NC_003888.3  region_1_(80000-100000)     80001            1   \n",
              "2538  NC_003888.3  region_1_(80000-100000)     80001            1   \n",
              "2539  NC_003888.3  region_1_(80000-100000)     80001            1   \n",
              "2540  NC_003888.3  region_1_(80000-100000)     80001            1   \n",
              "2541  NC_003888.3  region_1_(80000-100000)     80001            1   \n",
              "...           ...                      ...       ...          ...   \n",
              "5096  NC_003888.3    region_3_(9000-14000)      9001            1   \n",
              "4158  NC_003888.3  region_1_(80000-100000)     80001            1   \n",
              "1982  NC_003888.3  region_1_(80000-100000)     80001            1   \n",
              "6674  NC_003888.3   region_4_(15000-20000)     15001            1   \n",
              "2854  NC_003888.3  region_1_(80000-100000)     80001            1   \n",
              "\n",
              "      sgrna_strand  sgrna_loc    gc  pam                 sgrna  \\\n",
              "3517             1       6574  0.60  TGG  GCCACATTCACTACGACCCC   \n",
              "2538             1      16046  0.65  GGG  CCCTGATCCGAGTGTGTGCC   \n",
              "2539             1      16045  0.60  CGG  ACCCTGATCCGAGTGTGTGC   \n",
              "2540             1      16018  0.55  GGG  ACGACAATCCGGTGATCAGC   \n",
              "2541             1      16017  0.50  CGG  TACGACAATCCGGTGATCAG   \n",
              "...            ...        ...   ...  ...                   ...   \n",
              "5096            -1       1710  0.80  AGG  GGAATCGGGCGAGGGCGGCC   \n",
              "4158             1        137  0.85  TGG  CGAGCACACGGCGCGGGCGC   \n",
              "1982            -1      18880  0.80  TGG  CACCTACGGCGCGCTCGGCG   \n",
              "6674             1       3011  0.90  GGG  CGGCGCCGCCGACGCCGACG   \n",
              "2854             1      12959  0.85  GGG  CGCTCTGGGCGTCGCCGCCG   \n",
              "\n",
              "     sgrna_seed_sequence  off_target_count  \n",
              "3517       TCACTACGACCCC                 0  \n",
              "2538       CCGAGTGTGTGCC                 0  \n",
              "2539       TCCGAGTGTGTGC                 0  \n",
              "2540       TCCGGTGATCAGC                 0  \n",
              "2541       ATCCGGTGATCAG                 0  \n",
              "...                  ...               ...  \n",
              "5096       GGCGAGGGCGGCC                 9  \n",
              "4158       ACGGCGCGGGCGC                 9  \n",
              "1982       GGCGCGCTCGGCG                 9  \n",
              "6674       GCCGACGCCGACG                10  \n",
              "2854       GGCGTCGCCGCCG                10  \n",
              "\n",
              "[7026 rows x 11 columns]"
            ]
          },
          "execution_count": 11,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "\n",
        "sgrna_df = extract_sgRNAs(args)\n",
        "sgrna_df"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/plain": [
              "locus_tag\n",
              "region_1_(80000-100000)    4164\n",
              "region_3_(9000-14000)      1131\n",
              "region_4_(15000-20000)     1097\n",
              "region_2_(4000-7000)        634\n",
              "Name: count, dtype: int64"
            ]
          },
          "execution_count": 12,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "value_counts = sgrna_df['locus_tag'].value_counts()\n",
        "value_counts"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/plain": [
              "25828"
            ]
          },
          "execution_count": 13,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "len(genome.features)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>strain_name</th>\n",
              "      <th>locus_tag</th>\n",
              "      <th>gene_loc</th>\n",
              "      <th>gene_strand</th>\n",
              "      <th>sgrna_strand</th>\n",
              "      <th>sgrna_loc</th>\n",
              "      <th>gc</th>\n",
              "      <th>pam</th>\n",
              "      <th>sgrna</th>\n",
              "      <th>sgrna_seed_sequence</th>\n",
              "      <th>off_target_count</th>\n",
              "      <th>editable_cytosines</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>3517</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>region_1_(80000-100000)</td>\n",
              "      <td>80001</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>6574</td>\n",
              "      <td>0.60</td>\n",
              "      <td>TGG</td>\n",
              "      <td>GCCACATTCACTACGACCCC</td>\n",
              "      <td>TCACTACGACCCC</td>\n",
              "      <td>0</td>\n",
              "      <td>3,5,9</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2538</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>region_1_(80000-100000)</td>\n",
              "      <td>80001</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>16046</td>\n",
              "      <td>0.65</td>\n",
              "      <td>GGG</td>\n",
              "      <td>CCCTGATCCGAGTGTGTGCC</td>\n",
              "      <td>CCGAGTGTGTGCC</td>\n",
              "      <td>0</td>\n",
              "      <td>3,8,9</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2539</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>region_1_(80000-100000)</td>\n",
              "      <td>80001</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>16045</td>\n",
              "      <td>0.60</td>\n",
              "      <td>CGG</td>\n",
              "      <td>ACCCTGATCCGAGTGTGTGC</td>\n",
              "      <td>TCCGAGTGTGTGC</td>\n",
              "      <td>0</td>\n",
              "      <td>3,4,9,10</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2540</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>region_1_(80000-100000)</td>\n",
              "      <td>80001</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>16018</td>\n",
              "      <td>0.55</td>\n",
              "      <td>GGG</td>\n",
              "      <td>ACGACAATCCGGTGATCAGC</td>\n",
              "      <td>TCCGGTGATCAGC</td>\n",
              "      <td>0</td>\n",
              "      <td>5,9,10</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2541</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>region_1_(80000-100000)</td>\n",
              "      <td>80001</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>16017</td>\n",
              "      <td>0.50</td>\n",
              "      <td>CGG</td>\n",
              "      <td>TACGACAATCCGGTGATCAG</td>\n",
              "      <td>ATCCGGTGATCAG</td>\n",
              "      <td>0</td>\n",
              "      <td>3,6,10</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>...</th>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>5096</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>region_3_(9000-14000)</td>\n",
              "      <td>9001</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>1710</td>\n",
              "      <td>0.80</td>\n",
              "      <td>AGG</td>\n",
              "      <td>GGAATCGGGCGAGGGCGGCC</td>\n",
              "      <td>GGCGAGGGCGGCC</td>\n",
              "      <td>9</td>\n",
              "      <td>6,10</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4158</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>region_1_(80000-100000)</td>\n",
              "      <td>80001</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>137</td>\n",
              "      <td>0.85</td>\n",
              "      <td>TGG</td>\n",
              "      <td>CGAGCACACGGCGCGGGCGC</td>\n",
              "      <td>ACGGCGCGGGCGC</td>\n",
              "      <td>9</td>\n",
              "      <td>5,7,9</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1982</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>region_1_(80000-100000)</td>\n",
              "      <td>80001</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>18880</td>\n",
              "      <td>0.80</td>\n",
              "      <td>TGG</td>\n",
              "      <td>CACCTACGGCGCGCTCGGCG</td>\n",
              "      <td>GGCGCGCTCGGCG</td>\n",
              "      <td>9</td>\n",
              "      <td>3,4,7,10</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>6674</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>region_4_(15000-20000)</td>\n",
              "      <td>15001</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>3011</td>\n",
              "      <td>0.90</td>\n",
              "      <td>GGG</td>\n",
              "      <td>CGGCGCCGCCGACGCCGACG</td>\n",
              "      <td>GCCGACGCCGACG</td>\n",
              "      <td>10</td>\n",
              "      <td>4,6,7,9,10</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2854</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>region_1_(80000-100000)</td>\n",
              "      <td>80001</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>12959</td>\n",
              "      <td>0.85</td>\n",
              "      <td>GGG</td>\n",
              "      <td>CGCTCTGGGCGTCGCCGCCG</td>\n",
              "      <td>GGCGTCGCCGCCG</td>\n",
              "      <td>10</td>\n",
              "      <td>3,5,10</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>6802 rows × 12 columns</p>\n",
              "</div>"
            ],
            "text/plain": [
              "      strain_name                locus_tag  gene_loc  gene_strand  \\\n",
              "3517  NC_003888.3  region_1_(80000-100000)     80001            1   \n",
              "2538  NC_003888.3  region_1_(80000-100000)     80001            1   \n",
              "2539  NC_003888.3  region_1_(80000-100000)     80001            1   \n",
              "2540  NC_003888.3  region_1_(80000-100000)     80001            1   \n",
              "2541  NC_003888.3  region_1_(80000-100000)     80001            1   \n",
              "...           ...                      ...       ...          ...   \n",
              "5096  NC_003888.3    region_3_(9000-14000)      9001            1   \n",
              "4158  NC_003888.3  region_1_(80000-100000)     80001            1   \n",
              "1982  NC_003888.3  region_1_(80000-100000)     80001            1   \n",
              "6674  NC_003888.3   region_4_(15000-20000)     15001            1   \n",
              "2854  NC_003888.3  region_1_(80000-100000)     80001            1   \n",
              "\n",
              "      sgrna_strand  sgrna_loc    gc  pam                 sgrna  \\\n",
              "3517             1       6574  0.60  TGG  GCCACATTCACTACGACCCC   \n",
              "2538             1      16046  0.65  GGG  CCCTGATCCGAGTGTGTGCC   \n",
              "2539             1      16045  0.60  CGG  ACCCTGATCCGAGTGTGTGC   \n",
              "2540             1      16018  0.55  GGG  ACGACAATCCGGTGATCAGC   \n",
              "2541             1      16017  0.50  CGG  TACGACAATCCGGTGATCAG   \n",
              "...            ...        ...   ...  ...                   ...   \n",
              "5096            -1       1710  0.80  AGG  GGAATCGGGCGAGGGCGGCC   \n",
              "4158             1        137  0.85  TGG  CGAGCACACGGCGCGGGCGC   \n",
              "1982            -1      18880  0.80  TGG  CACCTACGGCGCGCTCGGCG   \n",
              "6674             1       3011  0.90  GGG  CGGCGCCGCCGACGCCGACG   \n",
              "2854             1      12959  0.85  GGG  CGCTCTGGGCGTCGCCGCCG   \n",
              "\n",
              "     sgrna_seed_sequence  off_target_count editable_cytosines  \n",
              "3517       TCACTACGACCCC                 0              3,5,9  \n",
              "2538       CCGAGTGTGTGCC                 0              3,8,9  \n",
              "2539       TCCGAGTGTGTGC                 0           3,4,9,10  \n",
              "2540       TCCGGTGATCAGC                 0             5,9,10  \n",
              "2541       ATCCGGTGATCAG                 0             3,6,10  \n",
              "...                  ...               ...                ...  \n",
              "5096       GGCGAGGGCGGCC                 9               6,10  \n",
              "4158       ACGGCGCGGGCGC                 9              5,7,9  \n",
              "1982       GGCGCGCTCGGCG                 9           3,4,7,10  \n",
              "6674       GCCGACGCCGACG                10         4,6,7,9,10  \n",
              "2854       GGCGTCGCCGCCG                10             3,5,10  \n",
              "\n",
              "[6802 rows x 12 columns]"
            ]
          },
          "execution_count": 14,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "# Load gene sequences from the genome\n",
        "gene_sequences = process_specified_gene_sequences_from_record(genome, genes_to_KO)\n",
        "genes_to_KO_dict = {locus_tag: gene_sequences[locus_tag] for locus_tag in genes_to_KO if locus_tag in gene_sequences}\n",
        "\n",
        "# Identify and annotate base editing sites\n",
        "sgrna_df_with_editing = identify_base_editing_sites(sgrna_df)\n",
        "\n",
        "# filter out only sgRNAs that result in base-editing\n",
        "filtered_sgrna_df_for_base_editing = filter_sgrnas_for_base_editing(sgrna_df_with_editing)\n",
        "filtered_sgrna_df_for_base_editing"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {},
      "outputs": [
        {
          "name": "stderr",
          "output_type": "stream",
          "text": [
            "/Users/lucaslevassor/opt/anaconda3/envs/streptocad/lib/python3.11/site-packages/Bio/Seq.py:3482: BiopythonWarning: Partial codon, len(sequence) not a multiple of three. Explicitly trim the sequence or add trailing N before translation. This may become an error in future.\n",
            "  warnings.warn(\n"
          ]
        },
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>strain_name</th>\n",
              "      <th>locus_tag</th>\n",
              "      <th>gene_loc</th>\n",
              "      <th>gene_strand</th>\n",
              "      <th>sgrna_strand</th>\n",
              "      <th>sgrna_loc</th>\n",
              "      <th>gc</th>\n",
              "      <th>pam</th>\n",
              "      <th>sgrna</th>\n",
              "      <th>sgrna_seed_sequence</th>\n",
              "      <th>off_target_count</th>\n",
              "      <th>editable_cytosines</th>\n",
              "      <th>mutations</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>3517</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>region_1_(80000-100000)</td>\n",
              "      <td>80001</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>6574</td>\n",
              "      <td>0.60</td>\n",
              "      <td>TGG</td>\n",
              "      <td>GCCACATTCACTACGACCCC</td>\n",
              "      <td>TCACTACGACCCC</td>\n",
              "      <td>0</td>\n",
              "      <td>3,5,9</td>\n",
              "      <td>P2186L, H2187Y, S2188L</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2538</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>region_1_(80000-100000)</td>\n",
              "      <td>80001</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>16046</td>\n",
              "      <td>0.65</td>\n",
              "      <td>GGG</td>\n",
              "      <td>CCCTGATCCGAGTGTGTGCC</td>\n",
              "      <td>CCGAGTGTGTGCC</td>\n",
              "      <td>0</td>\n",
              "      <td>3,8,9</td>\n",
              "      <td>S5345F</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2539</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>region_1_(80000-100000)</td>\n",
              "      <td>80001</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>16045</td>\n",
              "      <td>0.60</td>\n",
              "      <td>CGG</td>\n",
              "      <td>ACCCTGATCCGAGTGTGTGC</td>\n",
              "      <td>TCCGAGTGTGTGC</td>\n",
              "      <td>0</td>\n",
              "      <td>3,4,9,10</td>\n",
              "      <td>P5343L, S5345F</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2540</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>region_1_(80000-100000)</td>\n",
              "      <td>80001</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>16018</td>\n",
              "      <td>0.55</td>\n",
              "      <td>GGG</td>\n",
              "      <td>ACGACAATCCGGTGATCAGC</td>\n",
              "      <td>TCCGGTGATCAGC</td>\n",
              "      <td>0</td>\n",
              "      <td>5,9,10</td>\n",
              "      <td>Q5335*, S5336F</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2541</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>region_1_(80000-100000)</td>\n",
              "      <td>80001</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>16017</td>\n",
              "      <td>0.50</td>\n",
              "      <td>CGG</td>\n",
              "      <td>TACGACAATCCGGTGATCAG</td>\n",
              "      <td>ATCCGGTGATCAG</td>\n",
              "      <td>0</td>\n",
              "      <td>3,6,10</td>\n",
              "      <td>R5334*, Q5335*, S5336F</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>...</th>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>5096</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>region_3_(9000-14000)</td>\n",
              "      <td>9001</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>1710</td>\n",
              "      <td>0.80</td>\n",
              "      <td>AGG</td>\n",
              "      <td>GGAATCGGGCGAGGGCGGCC</td>\n",
              "      <td>GGCGAGGGCGGCC</td>\n",
              "      <td>9</td>\n",
              "      <td>6,10</td>\n",
              "      <td>D569N</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4158</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>region_1_(80000-100000)</td>\n",
              "      <td>80001</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>137</td>\n",
              "      <td>0.85</td>\n",
              "      <td>TGG</td>\n",
              "      <td>CGAGCACACGGCGCGGGCGC</td>\n",
              "      <td>ACGGCGCGGGCGC</td>\n",
              "      <td>9</td>\n",
              "      <td>5,7,9</td>\n",
              "      <td>A41V, H42Y</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1982</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>region_1_(80000-100000)</td>\n",
              "      <td>80001</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>18880</td>\n",
              "      <td>0.80</td>\n",
              "      <td>TGG</td>\n",
              "      <td>CACCTACGGCGCGCTCGGCG</td>\n",
              "      <td>GGCGCGCTCGGCG</td>\n",
              "      <td>9</td>\n",
              "      <td>3,4,7,10</td>\n",
              "      <td>A6291T, V6292I, G6293N</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>6674</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>region_4_(15000-20000)</td>\n",
              "      <td>15001</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>3011</td>\n",
              "      <td>0.90</td>\n",
              "      <td>GGG</td>\n",
              "      <td>CGGCGCCGCCGACGCCGACG</td>\n",
              "      <td>GCCGACGCCGACG</td>\n",
              "      <td>10</td>\n",
              "      <td>4,6,7,9,10</td>\n",
              "      <td>R999C, R1000C, R1001*</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2854</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>region_1_(80000-100000)</td>\n",
              "      <td>80001</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>12959</td>\n",
              "      <td>0.85</td>\n",
              "      <td>GGG</td>\n",
              "      <td>CGCTCTGGGCGTCGCCGCCG</td>\n",
              "      <td>GGCGTCGCCGCCG</td>\n",
              "      <td>10</td>\n",
              "      <td>3,5,10</td>\n",
              "      <td>S4315F, R4317C</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>6169 rows × 13 columns</p>\n",
              "</div>"
            ],
            "text/plain": [
              "      strain_name                locus_tag  gene_loc  gene_strand  \\\n",
              "3517  NC_003888.3  region_1_(80000-100000)     80001            1   \n",
              "2538  NC_003888.3  region_1_(80000-100000)     80001            1   \n",
              "2539  NC_003888.3  region_1_(80000-100000)     80001            1   \n",
              "2540  NC_003888.3  region_1_(80000-100000)     80001            1   \n",
              "2541  NC_003888.3  region_1_(80000-100000)     80001            1   \n",
              "...           ...                      ...       ...          ...   \n",
              "5096  NC_003888.3    region_3_(9000-14000)      9001            1   \n",
              "4158  NC_003888.3  region_1_(80000-100000)     80001            1   \n",
              "1982  NC_003888.3  region_1_(80000-100000)     80001            1   \n",
              "6674  NC_003888.3   region_4_(15000-20000)     15001            1   \n",
              "2854  NC_003888.3  region_1_(80000-100000)     80001            1   \n",
              "\n",
              "      sgrna_strand  sgrna_loc    gc  pam                 sgrna  \\\n",
              "3517             1       6574  0.60  TGG  GCCACATTCACTACGACCCC   \n",
              "2538             1      16046  0.65  GGG  CCCTGATCCGAGTGTGTGCC   \n",
              "2539             1      16045  0.60  CGG  ACCCTGATCCGAGTGTGTGC   \n",
              "2540             1      16018  0.55  GGG  ACGACAATCCGGTGATCAGC   \n",
              "2541             1      16017  0.50  CGG  TACGACAATCCGGTGATCAG   \n",
              "...            ...        ...   ...  ...                   ...   \n",
              "5096            -1       1710  0.80  AGG  GGAATCGGGCGAGGGCGGCC   \n",
              "4158             1        137  0.85  TGG  CGAGCACACGGCGCGGGCGC   \n",
              "1982            -1      18880  0.80  TGG  CACCTACGGCGCGCTCGGCG   \n",
              "6674             1       3011  0.90  GGG  CGGCGCCGCCGACGCCGACG   \n",
              "2854             1      12959  0.85  GGG  CGCTCTGGGCGTCGCCGCCG   \n",
              "\n",
              "     sgrna_seed_sequence  off_target_count editable_cytosines  \\\n",
              "3517       TCACTACGACCCC                 0              3,5,9   \n",
              "2538       CCGAGTGTGTGCC                 0              3,8,9   \n",
              "2539       TCCGAGTGTGTGC                 0           3,4,9,10   \n",
              "2540       TCCGGTGATCAGC                 0             5,9,10   \n",
              "2541       ATCCGGTGATCAG                 0             3,6,10   \n",
              "...                  ...               ...                ...   \n",
              "5096       GGCGAGGGCGGCC                 9               6,10   \n",
              "4158       ACGGCGCGGGCGC                 9              5,7,9   \n",
              "1982       GGCGCGCTCGGCG                 9           3,4,7,10   \n",
              "6674       GCCGACGCCGACG                10         4,6,7,9,10   \n",
              "2854       GGCGTCGCCGCCG                10             3,5,10   \n",
              "\n",
              "                   mutations  \n",
              "3517  P2186L, H2187Y, S2188L  \n",
              "2538                  S5345F  \n",
              "2539          P5343L, S5345F  \n",
              "2540          Q5335*, S5336F  \n",
              "2541  R5334*, Q5335*, S5336F  \n",
              "...                      ...  \n",
              "5096                   D569N  \n",
              "4158              A41V, H42Y  \n",
              "1982  A6291T, V6292I, G6293N  \n",
              "6674   R999C, R1000C, R1001*  \n",
              "2854          S4315F, R4317C  \n",
              "\n",
              "[6169 rows x 13 columns]"
            ]
          },
          "execution_count": 15,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "# Process the DataFrame to apply C-to-T mutations\n",
        "mutated_sgrna_df = process_base_editing(filtered_sgrna_df_for_base_editing, \n",
        "                                        genes_to_KO_dict, \n",
        "                                        only_stop_codons = False)\n",
        "mutated_sgrna_df"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {},
      "outputs": [
        {
          "name": "stderr",
          "output_type": "stream",
          "text": [
            "/Users/lucaslevassor/opt/anaconda3/envs/streptocad/lib/python3.11/site-packages/Bio/Seq.py:3482: BiopythonWarning: Partial codon, len(sequence) not a multiple of three. Explicitly trim the sequence or add trailing N before translation. This may become an error in future.\n",
            "  warnings.warn(\n"
          ]
        },
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>strain_name</th>\n",
              "      <th>locus_tag</th>\n",
              "      <th>gene_loc</th>\n",
              "      <th>gene_strand</th>\n",
              "      <th>sgrna_strand</th>\n",
              "      <th>sgrna_loc</th>\n",
              "      <th>gc</th>\n",
              "      <th>pam</th>\n",
              "      <th>sgrna</th>\n",
              "      <th>sgrna_seed_sequence</th>\n",
              "      <th>off_target_count</th>\n",
              "      <th>editable_cytosines</th>\n",
              "      <th>mutations</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>4173</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>region_2_(4000-7000)</td>\n",
              "      <td>4001</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>35</td>\n",
              "      <td>0.75</td>\n",
              "      <td>AGG</td>\n",
              "      <td>CCCCAGACACCACGCACGCC</td>\n",
              "      <td>CACCACGCACGCC</td>\n",
              "      <td>1</td>\n",
              "      <td>3,4,8,10</td>\n",
              "      <td>G9D, V10I, W11*</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4172</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>region_2_(4000-7000)</td>\n",
              "      <td>4001</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>34</td>\n",
              "      <td>0.70</td>\n",
              "      <td>GGG</td>\n",
              "      <td>CCCAGACACCACGCACGCCA</td>\n",
              "      <td>ACCACGCACGCCA</td>\n",
              "      <td>1</td>\n",
              "      <td>3,7,9,10</td>\n",
              "      <td>G9N, V10I, W11*</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>5938</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>region_4_(15000-20000)</td>\n",
              "      <td>15001</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>54</td>\n",
              "      <td>0.70</td>\n",
              "      <td>AGG</td>\n",
              "      <td>GCTCAACCTCCAGGCCGGAC</td>\n",
              "      <td>CTCCAGGCCGGAC</td>\n",
              "      <td>1</td>\n",
              "      <td>4,7,8,10</td>\n",
              "      <td>W15*, R16K</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4166</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>region_1_(80000-100000)</td>\n",
              "      <td>80001</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>80</td>\n",
              "      <td>0.55</td>\n",
              "      <td>AGG</td>\n",
              "      <td>GTTGTCCGTCAGCACTTCGA</td>\n",
              "      <td>GTCAGCACTTCGA</td>\n",
              "      <td>0</td>\n",
              "      <td>6,7,10</td>\n",
              "      <td>R23C, Q24*</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4165</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>region_1_(80000-100000)</td>\n",
              "      <td>80001</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>81</td>\n",
              "      <td>0.50</td>\n",
              "      <td>GGG</td>\n",
              "      <td>TTGTCCGTCAGCACTTCGAA</td>\n",
              "      <td>TCAGCACTTCGAA</td>\n",
              "      <td>1</td>\n",
              "      <td>5,6,9</td>\n",
              "      <td>R23C, Q24*</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>...</th>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2132</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>region_1_(80000-100000)</td>\n",
              "      <td>80001</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>19620</td>\n",
              "      <td>0.85</td>\n",
              "      <td>AGG</td>\n",
              "      <td>GCCGACCCGCGCGCTGGTCG</td>\n",
              "      <td>CGCGCGCTGGTCG</td>\n",
              "      <td>1</td>\n",
              "      <td>3,6,7,8,10</td>\n",
              "      <td>R6535*, P6536F, A6537V</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2133</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>region_1_(80000-100000)</td>\n",
              "      <td>80001</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>19614</td>\n",
              "      <td>0.85</td>\n",
              "      <td>TGG</td>\n",
              "      <td>CGCTGAGCCGACCCGCGCGC</td>\n",
              "      <td>CCGACCCGCGCGC</td>\n",
              "      <td>0</td>\n",
              "      <td>3,8,9</td>\n",
              "      <td>R6535*</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2053</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>region_1_(80000-100000)</td>\n",
              "      <td>80001</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>19624</td>\n",
              "      <td>0.75</td>\n",
              "      <td>CGG</td>\n",
              "      <td>ACCTCGACCAGCGCGCGGGT</td>\n",
              "      <td>CCAGCGCGCGGGT</td>\n",
              "      <td>0</td>\n",
              "      <td>3,5,8,9</td>\n",
              "      <td>W6539*, R6541K</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2065</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>region_1_(80000-100000)</td>\n",
              "      <td>80001</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>19714</td>\n",
              "      <td>0.55</td>\n",
              "      <td>AGG</td>\n",
              "      <td>GCCAGCATCTCGACTTCTTC</td>\n",
              "      <td>TCTCGACTTCTTC</td>\n",
              "      <td>0</td>\n",
              "      <td>3,6,9</td>\n",
              "      <td>R6569K, C6570Y, W6571*</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2068</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>region_1_(80000-100000)</td>\n",
              "      <td>80001</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>19781</td>\n",
              "      <td>0.60</td>\n",
              "      <td>CGG</td>\n",
              "      <td>GTGCACCATAGCGGCAGGTT</td>\n",
              "      <td>ATAGCGGCAGGTT</td>\n",
              "      <td>0</td>\n",
              "      <td>4,6,7</td>\n",
              "      <td>W6592*, C6593Y</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>642 rows × 13 columns</p>\n",
              "</div>"
            ],
            "text/plain": [
              "      strain_name                locus_tag  gene_loc  gene_strand  \\\n",
              "4173  NC_003888.3     region_2_(4000-7000)      4001            1   \n",
              "4172  NC_003888.3     region_2_(4000-7000)      4001            1   \n",
              "5938  NC_003888.3   region_4_(15000-20000)     15001            1   \n",
              "4166  NC_003888.3  region_1_(80000-100000)     80001            1   \n",
              "4165  NC_003888.3  region_1_(80000-100000)     80001            1   \n",
              "...           ...                      ...       ...          ...   \n",
              "2132  NC_003888.3  region_1_(80000-100000)     80001            1   \n",
              "2133  NC_003888.3  region_1_(80000-100000)     80001            1   \n",
              "2053  NC_003888.3  region_1_(80000-100000)     80001            1   \n",
              "2065  NC_003888.3  region_1_(80000-100000)     80001            1   \n",
              "2068  NC_003888.3  region_1_(80000-100000)     80001            1   \n",
              "\n",
              "      sgrna_strand  sgrna_loc    gc  pam                 sgrna  \\\n",
              "4173            -1         35  0.75  AGG  CCCCAGACACCACGCACGCC   \n",
              "4172            -1         34  0.70  GGG  CCCAGACACCACGCACGCCA   \n",
              "5938            -1         54  0.70  AGG  GCTCAACCTCCAGGCCGGAC   \n",
              "4166             1         80  0.55  AGG  GTTGTCCGTCAGCACTTCGA   \n",
              "4165             1         81  0.50  GGG  TTGTCCGTCAGCACTTCGAA   \n",
              "...            ...        ...   ...  ...                   ...   \n",
              "2132             1      19620  0.85  AGG  GCCGACCCGCGCGCTGGTCG   \n",
              "2133             1      19614  0.85  TGG  CGCTGAGCCGACCCGCGCGC   \n",
              "2053            -1      19624  0.75  CGG  ACCTCGACCAGCGCGCGGGT   \n",
              "2065            -1      19714  0.55  AGG  GCCAGCATCTCGACTTCTTC   \n",
              "2068            -1      19781  0.60  CGG  GTGCACCATAGCGGCAGGTT   \n",
              "\n",
              "     sgrna_seed_sequence  off_target_count editable_cytosines  \\\n",
              "4173       CACCACGCACGCC                 1           3,4,8,10   \n",
              "4172       ACCACGCACGCCA                 1           3,7,9,10   \n",
              "5938       CTCCAGGCCGGAC                 1           4,7,8,10   \n",
              "4166       GTCAGCACTTCGA                 0             6,7,10   \n",
              "4165       TCAGCACTTCGAA                 1              5,6,9   \n",
              "...                  ...               ...                ...   \n",
              "2132       CGCGCGCTGGTCG                 1         3,6,7,8,10   \n",
              "2133       CCGACCCGCGCGC                 0              3,8,9   \n",
              "2053       CCAGCGCGCGGGT                 0            3,5,8,9   \n",
              "2065       TCTCGACTTCTTC                 0              3,6,9   \n",
              "2068       ATAGCGGCAGGTT                 0              4,6,7   \n",
              "\n",
              "                   mutations  \n",
              "4173         G9D, V10I, W11*  \n",
              "4172         G9N, V10I, W11*  \n",
              "5938              W15*, R16K  \n",
              "4166              R23C, Q24*  \n",
              "4165              R23C, Q24*  \n",
              "...                      ...  \n",
              "2132  R6535*, P6536F, A6537V  \n",
              "2133                  R6535*  \n",
              "2053          W6539*, R6541K  \n",
              "2065  R6569K, C6570Y, W6571*  \n",
              "2068          W6592*, C6593Y  \n",
              "\n",
              "[642 rows x 13 columns]"
            ]
          },
          "execution_count": 16,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "# Process the DataFrame to apply C-to-T mutations\n",
        "mutated_sgrna_df = process_base_editing(filtered_sgrna_df_for_base_editing, \n",
        "                                        genes_to_KO_dict, \n",
        "                                        only_stop_codons = True)\n",
        "mutated_sgrna_df"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>strain_name</th>\n",
              "      <th>locus_tag</th>\n",
              "      <th>gene_loc</th>\n",
              "      <th>gene_strand</th>\n",
              "      <th>sgrna_strand</th>\n",
              "      <th>sgrna_loc</th>\n",
              "      <th>gc</th>\n",
              "      <th>pam</th>\n",
              "      <th>sgrna</th>\n",
              "      <th>sgrna_seed_sequence</th>\n",
              "      <th>off_target_count</th>\n",
              "      <th>editable_cytosines</th>\n",
              "      <th>mutations</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>4173</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>region_2_(4000-7000)</td>\n",
              "      <td>4001</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>35</td>\n",
              "      <td>0.75</td>\n",
              "      <td>AGG</td>\n",
              "      <td>CCCCAGACACCACGCACGCC</td>\n",
              "      <td>CACCACGCACGCC</td>\n",
              "      <td>1</td>\n",
              "      <td>3,4,8,10</td>\n",
              "      <td>G9D, V10I, W11*</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4172</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>region_2_(4000-7000)</td>\n",
              "      <td>4001</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>34</td>\n",
              "      <td>0.70</td>\n",
              "      <td>GGG</td>\n",
              "      <td>CCCAGACACCACGCACGCCA</td>\n",
              "      <td>ACCACGCACGCCA</td>\n",
              "      <td>1</td>\n",
              "      <td>3,7,9,10</td>\n",
              "      <td>G9N, V10I, W11*</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>5938</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>region_4_(15000-20000)</td>\n",
              "      <td>15001</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>54</td>\n",
              "      <td>0.70</td>\n",
              "      <td>AGG</td>\n",
              "      <td>GCTCAACCTCCAGGCCGGAC</td>\n",
              "      <td>CTCCAGGCCGGAC</td>\n",
              "      <td>1</td>\n",
              "      <td>4,7,8,10</td>\n",
              "      <td>W15*, R16K</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4166</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>region_1_(80000-100000)</td>\n",
              "      <td>80001</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>80</td>\n",
              "      <td>0.55</td>\n",
              "      <td>AGG</td>\n",
              "      <td>GTTGTCCGTCAGCACTTCGA</td>\n",
              "      <td>GTCAGCACTTCGA</td>\n",
              "      <td>0</td>\n",
              "      <td>6,7,10</td>\n",
              "      <td>R23C, Q24*</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4165</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>region_1_(80000-100000)</td>\n",
              "      <td>80001</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>81</td>\n",
              "      <td>0.50</td>\n",
              "      <td>GGG</td>\n",
              "      <td>TTGTCCGTCAGCACTTCGAA</td>\n",
              "      <td>TCAGCACTTCGAA</td>\n",
              "      <td>1</td>\n",
              "      <td>5,6,9</td>\n",
              "      <td>R23C, Q24*</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>5932</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>region_3_(9000-14000)</td>\n",
              "      <td>9001</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>118</td>\n",
              "      <td>0.65</td>\n",
              "      <td>CGG</td>\n",
              "      <td>TGCCCGATCTTCCAGGCCAG</td>\n",
              "      <td>TCTTCCAGGCCAG</td>\n",
              "      <td>1</td>\n",
              "      <td>3,4,5,9</td>\n",
              "      <td>A34V, R35*, S36F</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4160</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>region_1_(80000-100000)</td>\n",
              "      <td>80001</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>130</td>\n",
              "      <td>0.75</td>\n",
              "      <td>CGG</td>\n",
              "      <td>CGAAGGCCGAGCACACGGCG</td>\n",
              "      <td>CGAGCACACGGCG</td>\n",
              "      <td>0</td>\n",
              "      <td>7,8</td>\n",
              "      <td>R40*</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4159</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>region_1_(80000-100000)</td>\n",
              "      <td>80001</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>131</td>\n",
              "      <td>0.75</td>\n",
              "      <td>GGG</td>\n",
              "      <td>GAAGGCCGAGCACACGGCGC</td>\n",
              "      <td>GAGCACACGGCGC</td>\n",
              "      <td>0</td>\n",
              "      <td>6,7</td>\n",
              "      <td>R40*</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>5931</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>region_3_(9000-14000)</td>\n",
              "      <td>9001</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>146</td>\n",
              "      <td>0.60</td>\n",
              "      <td>CGG</td>\n",
              "      <td>ATTCGACGCAGAGACCGCAC</td>\n",
              "      <td>GCAGAGACCGCAC</td>\n",
              "      <td>1</td>\n",
              "      <td>4,7,9</td>\n",
              "      <td>R44*, R45C</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4182</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>region_2_(4000-7000)</td>\n",
              "      <td>4001</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>148</td>\n",
              "      <td>0.70</td>\n",
              "      <td>CGG</td>\n",
              "      <td>CATCCCAGGAAGGCAGCGCG</td>\n",
              "      <td>GGAAGGCAGCGCG</td>\n",
              "      <td>1</td>\n",
              "      <td>4,5,6</td>\n",
              "      <td>W48*, D49N</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4181</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>region_2_(4000-7000)</td>\n",
              "      <td>4001</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>147</td>\n",
              "      <td>0.70</td>\n",
              "      <td>GGG</td>\n",
              "      <td>ATCCCAGGAAGGCAGCGCGC</td>\n",
              "      <td>GAAGGCAGCGCGC</td>\n",
              "      <td>1</td>\n",
              "      <td>3,4,5</td>\n",
              "      <td>W48*, D49N</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>7018</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>region_4_(15000-20000)</td>\n",
              "      <td>15001</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>166</td>\n",
              "      <td>0.70</td>\n",
              "      <td>CGG</td>\n",
              "      <td>GCCACGAGCGGCATCCACTG</td>\n",
              "      <td>GCGGCATCCACTG</td>\n",
              "      <td>1</td>\n",
              "      <td>3,5,9</td>\n",
              "      <td>P50L, R51*, A52V</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>5926</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>region_3_(9000-14000)</td>\n",
              "      <td>9001</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>200</td>\n",
              "      <td>0.60</td>\n",
              "      <td>CGG</td>\n",
              "      <td>CGACCTCATCAGCGTCGACA</td>\n",
              "      <td>ATCAGCGTCGACA</td>\n",
              "      <td>1</td>\n",
              "      <td>4,5,7,10</td>\n",
              "      <td>P62F, H63Y, Q64*</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>5925</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>region_3_(9000-14000)</td>\n",
              "      <td>9001</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>203</td>\n",
              "      <td>0.65</td>\n",
              "      <td>CGG</td>\n",
              "      <td>CCTCATCAGCGTCGACACGG</td>\n",
              "      <td>AGCGTCGACACGG</td>\n",
              "      <td>1</td>\n",
              "      <td>4,7,10</td>\n",
              "      <td>H63Y, Q64*, R65C</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>7009</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>region_4_(15000-20000)</td>\n",
              "      <td>15001</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>204</td>\n",
              "      <td>0.80</td>\n",
              "      <td>CGG</td>\n",
              "      <td>GGGGGCGAGGTTCGGGACGC</td>\n",
              "      <td>AGGTTCGGGACGC</td>\n",
              "      <td>2</td>\n",
              "      <td>6</td>\n",
              "      <td>R64*</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>5924</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>region_3_(9000-14000)</td>\n",
              "      <td>9001</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>213</td>\n",
              "      <td>0.75</td>\n",
              "      <td>CGG</td>\n",
              "      <td>GTCGACACGGCGGCCCAGAC</td>\n",
              "      <td>CGGCGGCCCAGAC</td>\n",
              "      <td>2</td>\n",
              "      <td>3,6,8</td>\n",
              "      <td>R66*, H67Y</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>7004</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>region_4_(15000-20000)</td>\n",
              "      <td>15001</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>254</td>\n",
              "      <td>0.65</td>\n",
              "      <td>AGG</td>\n",
              "      <td>TCACGTCACCGACGGCGTAC</td>\n",
              "      <td>ACCGACGGCGTAC</td>\n",
              "      <td>1</td>\n",
              "      <td>4,7,9,10</td>\n",
              "      <td>R80C, H81Y, R82*</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4147</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>region_1_(80000-100000)</td>\n",
              "      <td>80001</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>256</td>\n",
              "      <td>0.65</td>\n",
              "      <td>AGG</td>\n",
              "      <td>CGACCAGGTCCATCTGCCAC</td>\n",
              "      <td>GTCCATCTGCCAC</td>\n",
              "      <td>0</td>\n",
              "      <td>4,5,10</td>\n",
              "      <td>Q81*</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>7002</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>region_4_(15000-20000)</td>\n",
              "      <td>15001</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>280</td>\n",
              "      <td>0.65</td>\n",
              "      <td>AGG</td>\n",
              "      <td>CGTACAGGTCGCGCATCTCC</td>\n",
              "      <td>GTCGCGCATCTCC</td>\n",
              "      <td>2</td>\n",
              "      <td>5,10</td>\n",
              "      <td>Q89*</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4765</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>region_2_(4000-7000)</td>\n",
              "      <td>4001</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>387</td>\n",
              "      <td>0.70</td>\n",
              "      <td>GGG</td>\n",
              "      <td>AAGGACAGGTGGGCCTCGCC</td>\n",
              "      <td>GGTGGGCCTCGCC</td>\n",
              "      <td>1</td>\n",
              "      <td>6</td>\n",
              "      <td>Q125*</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "      strain_name                locus_tag  gene_loc  gene_strand  \\\n",
              "4173  NC_003888.3     region_2_(4000-7000)      4001            1   \n",
              "4172  NC_003888.3     region_2_(4000-7000)      4001            1   \n",
              "5938  NC_003888.3   region_4_(15000-20000)     15001            1   \n",
              "4166  NC_003888.3  region_1_(80000-100000)     80001            1   \n",
              "4165  NC_003888.3  region_1_(80000-100000)     80001            1   \n",
              "5932  NC_003888.3    region_3_(9000-14000)      9001            1   \n",
              "4160  NC_003888.3  region_1_(80000-100000)     80001            1   \n",
              "4159  NC_003888.3  region_1_(80000-100000)     80001            1   \n",
              "5931  NC_003888.3    region_3_(9000-14000)      9001            1   \n",
              "4182  NC_003888.3     region_2_(4000-7000)      4001            1   \n",
              "4181  NC_003888.3     region_2_(4000-7000)      4001            1   \n",
              "7018  NC_003888.3   region_4_(15000-20000)     15001            1   \n",
              "5926  NC_003888.3    region_3_(9000-14000)      9001            1   \n",
              "5925  NC_003888.3    region_3_(9000-14000)      9001            1   \n",
              "7009  NC_003888.3   region_4_(15000-20000)     15001            1   \n",
              "5924  NC_003888.3    region_3_(9000-14000)      9001            1   \n",
              "7004  NC_003888.3   region_4_(15000-20000)     15001            1   \n",
              "4147  NC_003888.3  region_1_(80000-100000)     80001            1   \n",
              "7002  NC_003888.3   region_4_(15000-20000)     15001            1   \n",
              "4765  NC_003888.3     region_2_(4000-7000)      4001            1   \n",
              "\n",
              "      sgrna_strand  sgrna_loc    gc  pam                 sgrna  \\\n",
              "4173            -1         35  0.75  AGG  CCCCAGACACCACGCACGCC   \n",
              "4172            -1         34  0.70  GGG  CCCAGACACCACGCACGCCA   \n",
              "5938            -1         54  0.70  AGG  GCTCAACCTCCAGGCCGGAC   \n",
              "4166             1         80  0.55  AGG  GTTGTCCGTCAGCACTTCGA   \n",
              "4165             1         81  0.50  GGG  TTGTCCGTCAGCACTTCGAA   \n",
              "5932             1        118  0.65  CGG  TGCCCGATCTTCCAGGCCAG   \n",
              "4160             1        130  0.75  CGG  CGAAGGCCGAGCACACGGCG   \n",
              "4159             1        131  0.75  GGG  GAAGGCCGAGCACACGGCGC   \n",
              "5931             1        146  0.60  CGG  ATTCGACGCAGAGACCGCAC   \n",
              "4182            -1        148  0.70  CGG  CATCCCAGGAAGGCAGCGCG   \n",
              "4181            -1        147  0.70  GGG  ATCCCAGGAAGGCAGCGCGC   \n",
              "7018             1        166  0.70  CGG  GCCACGAGCGGCATCCACTG   \n",
              "5926             1        200  0.60  CGG  CGACCTCATCAGCGTCGACA   \n",
              "5925             1        203  0.65  CGG  CCTCATCAGCGTCGACACGG   \n",
              "7009             1        204  0.80  CGG  GGGGGCGAGGTTCGGGACGC   \n",
              "5924             1        213  0.75  CGG  GTCGACACGGCGGCCCAGAC   \n",
              "7004             1        254  0.65  AGG  TCACGTCACCGACGGCGTAC   \n",
              "4147             1        256  0.65  AGG  CGACCAGGTCCATCTGCCAC   \n",
              "7002             1        280  0.65  AGG  CGTACAGGTCGCGCATCTCC   \n",
              "4765             1        387  0.70  GGG  AAGGACAGGTGGGCCTCGCC   \n",
              "\n",
              "     sgrna_seed_sequence  off_target_count editable_cytosines  \\\n",
              "4173       CACCACGCACGCC                 1           3,4,8,10   \n",
              "4172       ACCACGCACGCCA                 1           3,7,9,10   \n",
              "5938       CTCCAGGCCGGAC                 1           4,7,8,10   \n",
              "4166       GTCAGCACTTCGA                 0             6,7,10   \n",
              "4165       TCAGCACTTCGAA                 1              5,6,9   \n",
              "5932       TCTTCCAGGCCAG                 1            3,4,5,9   \n",
              "4160       CGAGCACACGGCG                 0                7,8   \n",
              "4159       GAGCACACGGCGC                 0                6,7   \n",
              "5931       GCAGAGACCGCAC                 1              4,7,9   \n",
              "4182       GGAAGGCAGCGCG                 1              4,5,6   \n",
              "4181       GAAGGCAGCGCGC                 1              3,4,5   \n",
              "7018       GCGGCATCCACTG                 1              3,5,9   \n",
              "5926       ATCAGCGTCGACA                 1           4,5,7,10   \n",
              "5925       AGCGTCGACACGG                 1             4,7,10   \n",
              "7009       AGGTTCGGGACGC                 2                  6   \n",
              "5924       CGGCGGCCCAGAC                 2              3,6,8   \n",
              "7004       ACCGACGGCGTAC                 1           4,7,9,10   \n",
              "4147       GTCCATCTGCCAC                 0             4,5,10   \n",
              "7002       GTCGCGCATCTCC                 2               5,10   \n",
              "4765       GGTGGGCCTCGCC                 1                  6   \n",
              "\n",
              "             mutations  \n",
              "4173   G9D, V10I, W11*  \n",
              "4172   G9N, V10I, W11*  \n",
              "5938        W15*, R16K  \n",
              "4166        R23C, Q24*  \n",
              "4165        R23C, Q24*  \n",
              "5932  A34V, R35*, S36F  \n",
              "4160              R40*  \n",
              "4159              R40*  \n",
              "5931        R44*, R45C  \n",
              "4182        W48*, D49N  \n",
              "4181        W48*, D49N  \n",
              "7018  P50L, R51*, A52V  \n",
              "5926  P62F, H63Y, Q64*  \n",
              "5925  H63Y, Q64*, R65C  \n",
              "7009              R64*  \n",
              "5924        R66*, H67Y  \n",
              "7004  R80C, H81Y, R82*  \n",
              "4147              Q81*  \n",
              "7002              Q89*  \n",
              "4765             Q125*  "
            ]
          },
          "execution_count": 17,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "# Filter the DataFrame to retain only up to 5 sgRNA sequences per locus_tag\n",
        "filtered_df = mutated_sgrna_df.groupby('locus_tag').head(number_of_sgRNAs_per_group)\n",
        "filtered_df"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "## Output"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {},
      "outputs": [],
      "source": [
        "# MAke oligoes\n",
        "list_of_ssDNAs = make_ssDNA_oligos(filtered_df, upstream_ovh = up_homology,\n",
        "                      downstream_ovh=dw_homology)\n",
        "# cut plasmid\n",
        "from Bio.Restriction import NcoI\n",
        "linearized_plasmid = sorted(clean_plasmid.cut(NcoI), key=lambda x: len(x), reverse=True)[0]\n",
        "\n",
        "# assemble plasmid\n",
        "sgRNA_vectors = assemble_plasmids_by_ssDNA_bridging(list_of_ssDNAs,linearized_plasmid)\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {},
      "outputs": [],
      "source": [
        "# Constructing a meaningful name, ID, and description for the assembled plasmid using user input\n",
        "targeting_info = []\n",
        "for index, row in filtered_df.iterrows():\n",
        "    formatted_str = f\"pCRISPR-BEST_{row['locus_tag']}_p{row['sgrna_loc']}\"\n",
        "    targeting_info.append(formatted_str)\n",
        "\n",
        "for i in range(len(sgRNA_vectors)):\n",
        "    sgRNA_vectors[i].name = f'{targeting_info[i]}_#{i+1}'\n",
        "    sgRNA_vectors[i].id = sgRNA_vectors[i].name  # Using the same value for ID as for name for simplicity\n",
        "    sgRNA_vectors[i].description = f'Assembled plasmid targeting {\", \".join(genes_to_KO)} for base-editing, assembled using StreptoCAD.'\n",
        "\n",
        "# annotate plasmids\n",
        "for plasmid in sgRNA_vectors: \n",
        "    annotate_plasmid_with_sgrnas(plasmid, filtered_df)\n",
        "\n",
        "# PRINT\n",
        "print_plasmids = False\n",
        "if print_plasmids: \n",
        "    for vector in sgRNA_vectors: \n",
        "        vector.write(f\"../../data/plasmids/sgRNA_plasmids_pCRISPR_cBEST/{vector.id}.gb\")"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>plasmid_name</th>\n",
              "      <th>date</th>\n",
              "      <th>original_plasmid</th>\n",
              "      <th>integration</th>\n",
              "      <th>size</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>pCRISPR-BEST_region_2_(4000-7000)_p35_#1</td>\n",
              "      <td>2024-07-23</td>\n",
              "      <td>pCRISPR-cBEST(125689)</td>\n",
              "      <td>sgRNA_region_2_(4000-7000)(35)</td>\n",
              "      <td>12263</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>pCRISPR-BEST_region_2_(4000-7000)_p34_#2</td>\n",
              "      <td>2024-07-23</td>\n",
              "      <td>pCRISPR-cBEST(125689)</td>\n",
              "      <td>sgRNA_region_2_(4000-7000)(34)</td>\n",
              "      <td>12263</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>pCRISPR-BEST_region_4_(15000-20000)_p54_#3</td>\n",
              "      <td>2024-07-23</td>\n",
              "      <td>pCRISPR-cBEST(125689)</td>\n",
              "      <td>sgRNA_region_4_(15000-20000)(54)</td>\n",
              "      <td>12263</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>pCRISPR-BEST_region_1_(80000-100000)_p80_#4</td>\n",
              "      <td>2024-07-23</td>\n",
              "      <td>pCRISPR-cBEST(125689)</td>\n",
              "      <td>sgRNA_region_1_(80000-100000)(80)</td>\n",
              "      <td>12263</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>pCRISPR-BEST_region_1_(80000-100000)_p81_#5</td>\n",
              "      <td>2024-07-23</td>\n",
              "      <td>pCRISPR-cBEST(125689)</td>\n",
              "      <td>sgRNA_region_1_(80000-100000)(81)</td>\n",
              "      <td>12263</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>5</th>\n",
              "      <td>pCRISPR-BEST_region_3_(9000-14000)_p118_#6</td>\n",
              "      <td>2024-07-23</td>\n",
              "      <td>pCRISPR-cBEST(125689)</td>\n",
              "      <td>sgRNA_region_3_(9000-14000)(118)</td>\n",
              "      <td>12263</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>6</th>\n",
              "      <td>pCRISPR-BEST_region_1_(80000-100000)_p130_#7</td>\n",
              "      <td>2024-07-23</td>\n",
              "      <td>pCRISPR-cBEST(125689)</td>\n",
              "      <td>sgRNA_region_1_(80000-100000)(130)</td>\n",
              "      <td>12263</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>7</th>\n",
              "      <td>pCRISPR-BEST_region_1_(80000-100000)_p131_#8</td>\n",
              "      <td>2024-07-23</td>\n",
              "      <td>pCRISPR-cBEST(125689)</td>\n",
              "      <td>sgRNA_region_1_(80000-100000)(131)</td>\n",
              "      <td>12263</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>8</th>\n",
              "      <td>pCRISPR-BEST_region_3_(9000-14000)_p146_#9</td>\n",
              "      <td>2024-07-23</td>\n",
              "      <td>pCRISPR-cBEST(125689)</td>\n",
              "      <td>sgRNA_region_3_(9000-14000)(146)</td>\n",
              "      <td>12263</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>9</th>\n",
              "      <td>pCRISPR-BEST_region_2_(4000-7000)_p148_#10</td>\n",
              "      <td>2024-07-23</td>\n",
              "      <td>pCRISPR-cBEST(125689)</td>\n",
              "      <td>sgRNA_region_2_(4000-7000)(148)</td>\n",
              "      <td>12263</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>10</th>\n",
              "      <td>pCRISPR-BEST_region_2_(4000-7000)_p147_#11</td>\n",
              "      <td>2024-07-23</td>\n",
              "      <td>pCRISPR-cBEST(125689)</td>\n",
              "      <td>sgRNA_region_2_(4000-7000)(147)</td>\n",
              "      <td>12263</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>11</th>\n",
              "      <td>pCRISPR-BEST_region_4_(15000-20000)_p166_#12</td>\n",
              "      <td>2024-07-23</td>\n",
              "      <td>pCRISPR-cBEST(125689)</td>\n",
              "      <td>sgRNA_region_4_(15000-20000)(166)</td>\n",
              "      <td>12263</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>12</th>\n",
              "      <td>pCRISPR-BEST_region_3_(9000-14000)_p200_#13</td>\n",
              "      <td>2024-07-23</td>\n",
              "      <td>pCRISPR-cBEST(125689)</td>\n",
              "      <td>sgRNA_region_3_(9000-14000)(200)</td>\n",
              "      <td>12263</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>13</th>\n",
              "      <td>pCRISPR-BEST_region_3_(9000-14000)_p203_#14</td>\n",
              "      <td>2024-07-23</td>\n",
              "      <td>pCRISPR-cBEST(125689)</td>\n",
              "      <td>sgRNA_region_3_(9000-14000)(203)</td>\n",
              "      <td>12263</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>14</th>\n",
              "      <td>pCRISPR-BEST_region_4_(15000-20000)_p204_#15</td>\n",
              "      <td>2024-07-23</td>\n",
              "      <td>pCRISPR-cBEST(125689)</td>\n",
              "      <td>sgRNA_region_4_(15000-20000)(204)</td>\n",
              "      <td>12263</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>15</th>\n",
              "      <td>pCRISPR-BEST_region_3_(9000-14000)_p213_#16</td>\n",
              "      <td>2024-07-23</td>\n",
              "      <td>pCRISPR-cBEST(125689)</td>\n",
              "      <td>sgRNA_region_3_(9000-14000)(213)</td>\n",
              "      <td>12263</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>16</th>\n",
              "      <td>pCRISPR-BEST_region_4_(15000-20000)_p254_#17</td>\n",
              "      <td>2024-07-23</td>\n",
              "      <td>pCRISPR-cBEST(125689)</td>\n",
              "      <td>sgRNA_region_4_(15000-20000)(254)</td>\n",
              "      <td>12263</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>17</th>\n",
              "      <td>pCRISPR-BEST_region_1_(80000-100000)_p256_#18</td>\n",
              "      <td>2024-07-23</td>\n",
              "      <td>pCRISPR-cBEST(125689)</td>\n",
              "      <td>sgRNA_region_1_(80000-100000)(256)</td>\n",
              "      <td>12263</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>18</th>\n",
              "      <td>pCRISPR-BEST_region_4_(15000-20000)_p280_#19</td>\n",
              "      <td>2024-07-23</td>\n",
              "      <td>pCRISPR-cBEST(125689)</td>\n",
              "      <td>sgRNA_region_4_(15000-20000)(280)</td>\n",
              "      <td>12263</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>19</th>\n",
              "      <td>pCRISPR-BEST_region_2_(4000-7000)_p387_#20</td>\n",
              "      <td>2024-07-23</td>\n",
              "      <td>pCRISPR-cBEST(125689)</td>\n",
              "      <td>sgRNA_region_2_(4000-7000)(387)</td>\n",
              "      <td>12263</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "                                     plasmid_name        date  \\\n",
              "0        pCRISPR-BEST_region_2_(4000-7000)_p35_#1  2024-07-23   \n",
              "1        pCRISPR-BEST_region_2_(4000-7000)_p34_#2  2024-07-23   \n",
              "2      pCRISPR-BEST_region_4_(15000-20000)_p54_#3  2024-07-23   \n",
              "3     pCRISPR-BEST_region_1_(80000-100000)_p80_#4  2024-07-23   \n",
              "4     pCRISPR-BEST_region_1_(80000-100000)_p81_#5  2024-07-23   \n",
              "5      pCRISPR-BEST_region_3_(9000-14000)_p118_#6  2024-07-23   \n",
              "6    pCRISPR-BEST_region_1_(80000-100000)_p130_#7  2024-07-23   \n",
              "7    pCRISPR-BEST_region_1_(80000-100000)_p131_#8  2024-07-23   \n",
              "8      pCRISPR-BEST_region_3_(9000-14000)_p146_#9  2024-07-23   \n",
              "9      pCRISPR-BEST_region_2_(4000-7000)_p148_#10  2024-07-23   \n",
              "10     pCRISPR-BEST_region_2_(4000-7000)_p147_#11  2024-07-23   \n",
              "11   pCRISPR-BEST_region_4_(15000-20000)_p166_#12  2024-07-23   \n",
              "12    pCRISPR-BEST_region_3_(9000-14000)_p200_#13  2024-07-23   \n",
              "13    pCRISPR-BEST_region_3_(9000-14000)_p203_#14  2024-07-23   \n",
              "14   pCRISPR-BEST_region_4_(15000-20000)_p204_#15  2024-07-23   \n",
              "15    pCRISPR-BEST_region_3_(9000-14000)_p213_#16  2024-07-23   \n",
              "16   pCRISPR-BEST_region_4_(15000-20000)_p254_#17  2024-07-23   \n",
              "17  pCRISPR-BEST_region_1_(80000-100000)_p256_#18  2024-07-23   \n",
              "18   pCRISPR-BEST_region_4_(15000-20000)_p280_#19  2024-07-23   \n",
              "19     pCRISPR-BEST_region_2_(4000-7000)_p387_#20  2024-07-23   \n",
              "\n",
              "         original_plasmid                         integration   size  \n",
              "0   pCRISPR-cBEST(125689)      sgRNA_region_2_(4000-7000)(35)  12263  \n",
              "1   pCRISPR-cBEST(125689)      sgRNA_region_2_(4000-7000)(34)  12263  \n",
              "2   pCRISPR-cBEST(125689)    sgRNA_region_4_(15000-20000)(54)  12263  \n",
              "3   pCRISPR-cBEST(125689)   sgRNA_region_1_(80000-100000)(80)  12263  \n",
              "4   pCRISPR-cBEST(125689)   sgRNA_region_1_(80000-100000)(81)  12263  \n",
              "5   pCRISPR-cBEST(125689)    sgRNA_region_3_(9000-14000)(118)  12263  \n",
              "6   pCRISPR-cBEST(125689)  sgRNA_region_1_(80000-100000)(130)  12263  \n",
              "7   pCRISPR-cBEST(125689)  sgRNA_region_1_(80000-100000)(131)  12263  \n",
              "8   pCRISPR-cBEST(125689)    sgRNA_region_3_(9000-14000)(146)  12263  \n",
              "9   pCRISPR-cBEST(125689)     sgRNA_region_2_(4000-7000)(148)  12263  \n",
              "10  pCRISPR-cBEST(125689)     sgRNA_region_2_(4000-7000)(147)  12263  \n",
              "11  pCRISPR-cBEST(125689)   sgRNA_region_4_(15000-20000)(166)  12263  \n",
              "12  pCRISPR-cBEST(125689)    sgRNA_region_3_(9000-14000)(200)  12263  \n",
              "13  pCRISPR-cBEST(125689)    sgRNA_region_3_(9000-14000)(203)  12263  \n",
              "14  pCRISPR-cBEST(125689)   sgRNA_region_4_(15000-20000)(204)  12263  \n",
              "15  pCRISPR-cBEST(125689)    sgRNA_region_3_(9000-14000)(213)  12263  \n",
              "16  pCRISPR-cBEST(125689)   sgRNA_region_4_(15000-20000)(254)  12263  \n",
              "17  pCRISPR-cBEST(125689)  sgRNA_region_1_(80000-100000)(256)  12263  \n",
              "18  pCRISPR-cBEST(125689)   sgRNA_region_4_(15000-20000)(280)  12263  \n",
              "19  pCRISPR-cBEST(125689)     sgRNA_region_2_(4000-7000)(387)  12263  "
            ]
          },
          "execution_count": 20,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "\n",
        "integration_names = filtered_df.apply(lambda row: f\"sgRNA_{row['locus_tag']}({row['sgrna_loc']})\", axis=1).tolist()\n",
        "plasmid_metadata_df = extract_metadata_to_dataframe(sgRNA_vectors,\n",
        "                                                    clean_plasmid,\n",
        "                                                    integration_names)\n",
        "\n",
        "plasmid_metadata_df"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "# IDT primers"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>Name</th>\n",
              "      <th>Sequence</th>\n",
              "      <th>Concentration</th>\n",
              "      <th>Purification</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>region_2_(4000-7000)_loc_35</td>\n",
              "      <td>CGGTTGGTAGGATCGACGGCCCCCAGACACCACGCACGCCGTTTTA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>region_2_(4000-7000)_loc_34</td>\n",
              "      <td>CGGTTGGTAGGATCGACGGCCCCAGACACCACGCACGCCAGTTTTA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>region_4_(15000-20000)_loc_54</td>\n",
              "      <td>CGGTTGGTAGGATCGACGGCGCTCAACCTCCAGGCCGGACGTTTTA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>region_1_(80000-100000)_loc_80</td>\n",
              "      <td>CGGTTGGTAGGATCGACGGCGTTGTCCGTCAGCACTTCGAGTTTTA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>region_1_(80000-100000)_loc_81</td>\n",
              "      <td>CGGTTGGTAGGATCGACGGCTTGTCCGTCAGCACTTCGAAGTTTTA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>5</th>\n",
              "      <td>region_3_(9000-14000)_loc_118</td>\n",
              "      <td>CGGTTGGTAGGATCGACGGCTGCCCGATCTTCCAGGCCAGGTTTTA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>6</th>\n",
              "      <td>region_1_(80000-100000)_loc_130</td>\n",
              "      <td>CGGTTGGTAGGATCGACGGCCGAAGGCCGAGCACACGGCGGTTTTA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>7</th>\n",
              "      <td>region_1_(80000-100000)_loc_131</td>\n",
              "      <td>CGGTTGGTAGGATCGACGGCGAAGGCCGAGCACACGGCGCGTTTTA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>8</th>\n",
              "      <td>region_3_(9000-14000)_loc_146</td>\n",
              "      <td>CGGTTGGTAGGATCGACGGCATTCGACGCAGAGACCGCACGTTTTA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>9</th>\n",
              "      <td>region_2_(4000-7000)_loc_148</td>\n",
              "      <td>CGGTTGGTAGGATCGACGGCCATCCCAGGAAGGCAGCGCGGTTTTA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>10</th>\n",
              "      <td>region_2_(4000-7000)_loc_147</td>\n",
              "      <td>CGGTTGGTAGGATCGACGGCATCCCAGGAAGGCAGCGCGCGTTTTA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>11</th>\n",
              "      <td>region_4_(15000-20000)_loc_166</td>\n",
              "      <td>CGGTTGGTAGGATCGACGGCGCCACGAGCGGCATCCACTGGTTTTA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>12</th>\n",
              "      <td>region_3_(9000-14000)_loc_200</td>\n",
              "      <td>CGGTTGGTAGGATCGACGGCCGACCTCATCAGCGTCGACAGTTTTA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>13</th>\n",
              "      <td>region_3_(9000-14000)_loc_203</td>\n",
              "      <td>CGGTTGGTAGGATCGACGGCCCTCATCAGCGTCGACACGGGTTTTA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>14</th>\n",
              "      <td>region_4_(15000-20000)_loc_204</td>\n",
              "      <td>CGGTTGGTAGGATCGACGGCGGGGGCGAGGTTCGGGACGCGTTTTA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>15</th>\n",
              "      <td>region_3_(9000-14000)_loc_213</td>\n",
              "      <td>CGGTTGGTAGGATCGACGGCGTCGACACGGCGGCCCAGACGTTTTA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>16</th>\n",
              "      <td>region_4_(15000-20000)_loc_254</td>\n",
              "      <td>CGGTTGGTAGGATCGACGGCTCACGTCACCGACGGCGTACGTTTTA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>17</th>\n",
              "      <td>region_1_(80000-100000)_loc_256</td>\n",
              "      <td>CGGTTGGTAGGATCGACGGCCGACCAGGTCCATCTGCCACGTTTTA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>18</th>\n",
              "      <td>region_4_(15000-20000)_loc_280</td>\n",
              "      <td>CGGTTGGTAGGATCGACGGCCGTACAGGTCGCGCATCTCCGTTTTA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>19</th>\n",
              "      <td>region_2_(4000-7000)_loc_387</td>\n",
              "      <td>CGGTTGGTAGGATCGACGGCAAGGACAGGTGGGCCTCGCCGTTTTA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "                               Name  \\\n",
              "0       region_2_(4000-7000)_loc_35   \n",
              "1       region_2_(4000-7000)_loc_34   \n",
              "2     region_4_(15000-20000)_loc_54   \n",
              "3    region_1_(80000-100000)_loc_80   \n",
              "4    region_1_(80000-100000)_loc_81   \n",
              "5     region_3_(9000-14000)_loc_118   \n",
              "6   region_1_(80000-100000)_loc_130   \n",
              "7   region_1_(80000-100000)_loc_131   \n",
              "8     region_3_(9000-14000)_loc_146   \n",
              "9      region_2_(4000-7000)_loc_148   \n",
              "10     region_2_(4000-7000)_loc_147   \n",
              "11   region_4_(15000-20000)_loc_166   \n",
              "12    region_3_(9000-14000)_loc_200   \n",
              "13    region_3_(9000-14000)_loc_203   \n",
              "14   region_4_(15000-20000)_loc_204   \n",
              "15    region_3_(9000-14000)_loc_213   \n",
              "16   region_4_(15000-20000)_loc_254   \n",
              "17  region_1_(80000-100000)_loc_256   \n",
              "18   region_4_(15000-20000)_loc_280   \n",
              "19     region_2_(4000-7000)_loc_387   \n",
              "\n",
              "                                             Sequence Concentration  \\\n",
              "0   CGGTTGGTAGGATCGACGGCCCCCAGACACCACGCACGCCGTTTTA...          25nm   \n",
              "1   CGGTTGGTAGGATCGACGGCCCCAGACACCACGCACGCCAGTTTTA...          25nm   \n",
              "2   CGGTTGGTAGGATCGACGGCGCTCAACCTCCAGGCCGGACGTTTTA...          25nm   \n",
              "3   CGGTTGGTAGGATCGACGGCGTTGTCCGTCAGCACTTCGAGTTTTA...          25nm   \n",
              "4   CGGTTGGTAGGATCGACGGCTTGTCCGTCAGCACTTCGAAGTTTTA...          25nm   \n",
              "5   CGGTTGGTAGGATCGACGGCTGCCCGATCTTCCAGGCCAGGTTTTA...          25nm   \n",
              "6   CGGTTGGTAGGATCGACGGCCGAAGGCCGAGCACACGGCGGTTTTA...          25nm   \n",
              "7   CGGTTGGTAGGATCGACGGCGAAGGCCGAGCACACGGCGCGTTTTA...          25nm   \n",
              "8   CGGTTGGTAGGATCGACGGCATTCGACGCAGAGACCGCACGTTTTA...          25nm   \n",
              "9   CGGTTGGTAGGATCGACGGCCATCCCAGGAAGGCAGCGCGGTTTTA...          25nm   \n",
              "10  CGGTTGGTAGGATCGACGGCATCCCAGGAAGGCAGCGCGCGTTTTA...          25nm   \n",
              "11  CGGTTGGTAGGATCGACGGCGCCACGAGCGGCATCCACTGGTTTTA...          25nm   \n",
              "12  CGGTTGGTAGGATCGACGGCCGACCTCATCAGCGTCGACAGTTTTA...          25nm   \n",
              "13  CGGTTGGTAGGATCGACGGCCCTCATCAGCGTCGACACGGGTTTTA...          25nm   \n",
              "14  CGGTTGGTAGGATCGACGGCGGGGGCGAGGTTCGGGACGCGTTTTA...          25nm   \n",
              "15  CGGTTGGTAGGATCGACGGCGTCGACACGGCGGCCCAGACGTTTTA...          25nm   \n",
              "16  CGGTTGGTAGGATCGACGGCTCACGTCACCGACGGCGTACGTTTTA...          25nm   \n",
              "17  CGGTTGGTAGGATCGACGGCCGACCAGGTCCATCTGCCACGTTTTA...          25nm   \n",
              "18  CGGTTGGTAGGATCGACGGCCGTACAGGTCGCGCATCTCCGTTTTA...          25nm   \n",
              "19  CGGTTGGTAGGATCGACGGCAAGGACAGGTGGGCCTCGCCGTTTTA...          25nm   \n",
              "\n",
              "   Purification  \n",
              "0           STD  \n",
              "1           STD  \n",
              "2           STD  \n",
              "3           STD  \n",
              "4           STD  \n",
              "5           STD  \n",
              "6           STD  \n",
              "7           STD  \n",
              "8           STD  \n",
              "9           STD  \n",
              "10          STD  \n",
              "11          STD  \n",
              "12          STD  \n",
              "13          STD  \n",
              "14          STD  \n",
              "15          STD  \n",
              "16          STD  \n",
              "17          STD  \n",
              "18          STD  \n",
              "19          STD  "
            ]
          },
          "execution_count": 21,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "idt_df1 = primers_to_IDT(list_of_ssDNAs)\n",
        "idt_df1"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>locus tag</th>\n",
              "      <th>f_primer_name</th>\n",
              "      <th>r_primer_name</th>\n",
              "      <th>f_primer_sequences(5-3)</th>\n",
              "      <th>r_primer_sequences(5-3)</th>\n",
              "      <th>f_tm</th>\n",
              "      <th>r_tm</th>\n",
              "      <th>ta</th>\n",
              "      <th>flanking_region</th>\n",
              "      <th>annealing_temperature</th>\n",
              "      <th>...</th>\n",
              "      <th>homodimer_reverse_tm</th>\n",
              "      <th>homodimer_reverse_deltaG (kcal/mol)</th>\n",
              "      <th>heterodimer_tm</th>\n",
              "      <th>heterodimer_deltaG (kcal/mol)</th>\n",
              "      <th>hairpin_forward_structure_found</th>\n",
              "      <th>hairpin_forward_tm</th>\n",
              "      <th>hairpin_forward_deltaG (kcal/mol)</th>\n",
              "      <th>hairpin_reverse_structure_found</th>\n",
              "      <th>hairpin_reverse_tm</th>\n",
              "      <th>hairpin_reverse_deltaG (kcal/mol)</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>region_4_(15000-20000)</td>\n",
              "      <td>region_4_(15000-20000)_fwd_checking_primer</td>\n",
              "      <td>region_4_(15000-20000)_rev_checking_primer</td>\n",
              "      <td>TTGCTGCAACCTGCCTTG</td>\n",
              "      <td>TAGGTGCCTCCGTCGA</td>\n",
              "      <td>61</td>\n",
              "      <td>58</td>\n",
              "      <td>62</td>\n",
              "      <td>500</td>\n",
              "      <td>62</td>\n",
              "      <td>...</td>\n",
              "      <td>-13.794142</td>\n",
              "      <td>-0.897729</td>\n",
              "      <td>-1.321907</td>\n",
              "      <td>0.948660</td>\n",
              "      <td>False</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>False</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>region_2_(4000-7000)</td>\n",
              "      <td>region_2_(4000-7000)_fwd_checking_primer</td>\n",
              "      <td>region_2_(4000-7000)_rev_checking_primer</td>\n",
              "      <td>GACCGCGCTGGGTG</td>\n",
              "      <td>TCCGGCCGCATCG</td>\n",
              "      <td>59</td>\n",
              "      <td>57</td>\n",
              "      <td>60</td>\n",
              "      <td>500</td>\n",
              "      <td>60</td>\n",
              "      <td>...</td>\n",
              "      <td>22.848051</td>\n",
              "      <td>-5.018625</td>\n",
              "      <td>6.535424</td>\n",
              "      <td>-0.061707</td>\n",
              "      <td>False</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>False</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>region_3_(9000-14000)</td>\n",
              "      <td>region_3_(9000-14000)_fwd_checking_primer</td>\n",
              "      <td>region_3_(9000-14000)_rev_checking_primer</td>\n",
              "      <td>ATCCCGCAGCTGCAGG</td>\n",
              "      <td>GAGGAGCGGCAATGGTCATC</td>\n",
              "      <td>61</td>\n",
              "      <td>63</td>\n",
              "      <td>65</td>\n",
              "      <td>500</td>\n",
              "      <td>65</td>\n",
              "      <td>...</td>\n",
              "      <td>1.445721</td>\n",
              "      <td>-1.026208</td>\n",
              "      <td>9.252334</td>\n",
              "      <td>-1.213024</td>\n",
              "      <td>False</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>False</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>region_1_(80000-100000)</td>\n",
              "      <td>region_1_(80000-100000)_fwd_checking_primer</td>\n",
              "      <td>region_1_(80000-100000)_rev_checking_primer</td>\n",
              "      <td>GCCGCTGGCCAAGC</td>\n",
              "      <td>AGGCGTGCGCGTGG</td>\n",
              "      <td>60</td>\n",
              "      <td>62</td>\n",
              "      <td>63</td>\n",
              "      <td>700</td>\n",
              "      <td>63</td>\n",
              "      <td>...</td>\n",
              "      <td>22.176370</td>\n",
              "      <td>-2.921991</td>\n",
              "      <td>11.333376</td>\n",
              "      <td>-0.877576</td>\n",
              "      <td>False</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>False</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>4 rows × 23 columns</p>\n",
              "</div>"
            ],
            "text/plain": [
              "                 locus tag                                f_primer_name  \\\n",
              "0   region_4_(15000-20000)   region_4_(15000-20000)_fwd_checking_primer   \n",
              "1     region_2_(4000-7000)     region_2_(4000-7000)_fwd_checking_primer   \n",
              "2    region_3_(9000-14000)    region_3_(9000-14000)_fwd_checking_primer   \n",
              "3  region_1_(80000-100000)  region_1_(80000-100000)_fwd_checking_primer   \n",
              "\n",
              "                                 r_primer_name f_primer_sequences(5-3)  \\\n",
              "0   region_4_(15000-20000)_rev_checking_primer      TTGCTGCAACCTGCCTTG   \n",
              "1     region_2_(4000-7000)_rev_checking_primer          GACCGCGCTGGGTG   \n",
              "2    region_3_(9000-14000)_rev_checking_primer        ATCCCGCAGCTGCAGG   \n",
              "3  region_1_(80000-100000)_rev_checking_primer          GCCGCTGGCCAAGC   \n",
              "\n",
              "  r_primer_sequences(5-3) f_tm r_tm  ta flanking_region  \\\n",
              "0        TAGGTGCCTCCGTCGA   61   58  62             500   \n",
              "1           TCCGGCCGCATCG   59   57  60             500   \n",
              "2    GAGGAGCGGCAATGGTCATC   61   63  65             500   \n",
              "3          AGGCGTGCGCGTGG   60   62  63             700   \n",
              "\n",
              "   annealing_temperature  ... homodimer_reverse_tm  \\\n",
              "0                     62  ...           -13.794142   \n",
              "1                     60  ...            22.848051   \n",
              "2                     65  ...             1.445721   \n",
              "3                     63  ...            22.176370   \n",
              "\n",
              "   homodimer_reverse_deltaG (kcal/mol)  heterodimer_tm  \\\n",
              "0                            -0.897729       -1.321907   \n",
              "1                            -5.018625        6.535424   \n",
              "2                            -1.026208        9.252334   \n",
              "3                            -2.921991       11.333376   \n",
              "\n",
              "   heterodimer_deltaG (kcal/mol)  hairpin_forward_structure_found  \\\n",
              "0                       0.948660                            False   \n",
              "1                      -0.061707                            False   \n",
              "2                      -1.213024                            False   \n",
              "3                      -0.877576                            False   \n",
              "\n",
              "   hairpin_forward_tm  hairpin_forward_deltaG (kcal/mol)  \\\n",
              "0                 0.0                                0.0   \n",
              "1                 0.0                                0.0   \n",
              "2                 0.0                                0.0   \n",
              "3                 0.0                                0.0   \n",
              "\n",
              "   hairpin_reverse_structure_found  hairpin_reverse_tm  \\\n",
              "0                            False                 0.0   \n",
              "1                            False                 0.0   \n",
              "2                            False                 0.0   \n",
              "3                            False                 0.0   \n",
              "\n",
              "   hairpin_reverse_deltaG (kcal/mol)  \n",
              "0                                0.0  \n",
              "1                                0.0  \n",
              "2                                0.0  \n",
              "3                                0.0  \n",
              "\n",
              "[4 rows x 23 columns]"
            ]
          },
          "execution_count": 22,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "# Getting checking primers\n",
        "checking_primers_df = find_best_check_primers_from_genome(genome, \n",
        "                                       genes_to_KO, \n",
        "                                       flanking_region=flanking_region,\n",
        "                                       target_tm = melting_temperature, \n",
        "                                        primer_concentration = primer_concentration, \n",
        "                                        polymerase = chosen_polymerase)\n",
        "checking_primers_df\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>Name</th>\n",
              "      <th>Sequence</th>\n",
              "      <th>Concentration</th>\n",
              "      <th>Purification</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>region_4_(15000-20000)_fwd_checking_primer</td>\n",
              "      <td>TTGCTGCAACCTGCCTTG</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>region_2_(4000-7000)_fwd_checking_primer</td>\n",
              "      <td>GACCGCGCTGGGTG</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>region_3_(9000-14000)_fwd_checking_primer</td>\n",
              "      <td>ATCCCGCAGCTGCAGG</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>region_1_(80000-100000)_fwd_checking_primer</td>\n",
              "      <td>GCCGCTGGCCAAGC</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>region_4_(15000-20000)_rev_checking_primer</td>\n",
              "      <td>TAGGTGCCTCCGTCGA</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>5</th>\n",
              "      <td>region_2_(4000-7000)_rev_checking_primer</td>\n",
              "      <td>TCCGGCCGCATCG</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>6</th>\n",
              "      <td>region_3_(9000-14000)_rev_checking_primer</td>\n",
              "      <td>GAGGAGCGGCAATGGTCATC</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>7</th>\n",
              "      <td>region_1_(80000-100000)_rev_checking_primer</td>\n",
              "      <td>AGGCGTGCGCGTGG</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "                                          Name              Sequence  \\\n",
              "0   region_4_(15000-20000)_fwd_checking_primer    TTGCTGCAACCTGCCTTG   \n",
              "1     region_2_(4000-7000)_fwd_checking_primer        GACCGCGCTGGGTG   \n",
              "2    region_3_(9000-14000)_fwd_checking_primer      ATCCCGCAGCTGCAGG   \n",
              "3  region_1_(80000-100000)_fwd_checking_primer        GCCGCTGGCCAAGC   \n",
              "4   region_4_(15000-20000)_rev_checking_primer      TAGGTGCCTCCGTCGA   \n",
              "5     region_2_(4000-7000)_rev_checking_primer         TCCGGCCGCATCG   \n",
              "6    region_3_(9000-14000)_rev_checking_primer  GAGGAGCGGCAATGGTCATC   \n",
              "7  region_1_(80000-100000)_rev_checking_primer        AGGCGTGCGCGTGG   \n",
              "\n",
              "  Concentration Purification  \n",
              "0          25nm          STD  \n",
              "1          25nm          STD  \n",
              "2          25nm          STD  \n",
              "3          25nm          STD  \n",
              "4          25nm          STD  \n",
              "5          25nm          STD  \n",
              "6          25nm          STD  \n",
              "7          25nm          STD  "
            ]
          },
          "execution_count": 23,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "idt_df2 = create_idt_order_dataframe(checking_primers_df)\n",
        "idt_df2"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>Name</th>\n",
              "      <th>Sequence</th>\n",
              "      <th>Concentration</th>\n",
              "      <th>Purification</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>region_2_(4000-7000)_loc_35</td>\n",
              "      <td>CGGTTGGTAGGATCGACGGCCCCCAGACACCACGCACGCCGTTTTA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>region_2_(4000-7000)_loc_34</td>\n",
              "      <td>CGGTTGGTAGGATCGACGGCCCCAGACACCACGCACGCCAGTTTTA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>region_4_(15000-20000)_loc_54</td>\n",
              "      <td>CGGTTGGTAGGATCGACGGCGCTCAACCTCCAGGCCGGACGTTTTA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>region_1_(80000-100000)_loc_80</td>\n",
              "      <td>CGGTTGGTAGGATCGACGGCGTTGTCCGTCAGCACTTCGAGTTTTA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>region_1_(80000-100000)_loc_81</td>\n",
              "      <td>CGGTTGGTAGGATCGACGGCTTGTCCGTCAGCACTTCGAAGTTTTA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>5</th>\n",
              "      <td>region_3_(9000-14000)_loc_118</td>\n",
              "      <td>CGGTTGGTAGGATCGACGGCTGCCCGATCTTCCAGGCCAGGTTTTA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>6</th>\n",
              "      <td>region_1_(80000-100000)_loc_130</td>\n",
              "      <td>CGGTTGGTAGGATCGACGGCCGAAGGCCGAGCACACGGCGGTTTTA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>7</th>\n",
              "      <td>region_1_(80000-100000)_loc_131</td>\n",
              "      <td>CGGTTGGTAGGATCGACGGCGAAGGCCGAGCACACGGCGCGTTTTA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>8</th>\n",
              "      <td>region_3_(9000-14000)_loc_146</td>\n",
              "      <td>CGGTTGGTAGGATCGACGGCATTCGACGCAGAGACCGCACGTTTTA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>9</th>\n",
              "      <td>region_2_(4000-7000)_loc_148</td>\n",
              "      <td>CGGTTGGTAGGATCGACGGCCATCCCAGGAAGGCAGCGCGGTTTTA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>10</th>\n",
              "      <td>region_2_(4000-7000)_loc_147</td>\n",
              "      <td>CGGTTGGTAGGATCGACGGCATCCCAGGAAGGCAGCGCGCGTTTTA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>11</th>\n",
              "      <td>region_4_(15000-20000)_loc_166</td>\n",
              "      <td>CGGTTGGTAGGATCGACGGCGCCACGAGCGGCATCCACTGGTTTTA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>12</th>\n",
              "      <td>region_3_(9000-14000)_loc_200</td>\n",
              "      <td>CGGTTGGTAGGATCGACGGCCGACCTCATCAGCGTCGACAGTTTTA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>13</th>\n",
              "      <td>region_3_(9000-14000)_loc_203</td>\n",
              "      <td>CGGTTGGTAGGATCGACGGCCCTCATCAGCGTCGACACGGGTTTTA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>14</th>\n",
              "      <td>region_4_(15000-20000)_loc_204</td>\n",
              "      <td>CGGTTGGTAGGATCGACGGCGGGGGCGAGGTTCGGGACGCGTTTTA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>15</th>\n",
              "      <td>region_3_(9000-14000)_loc_213</td>\n",
              "      <td>CGGTTGGTAGGATCGACGGCGTCGACACGGCGGCCCAGACGTTTTA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>16</th>\n",
              "      <td>region_4_(15000-20000)_loc_254</td>\n",
              "      <td>CGGTTGGTAGGATCGACGGCTCACGTCACCGACGGCGTACGTTTTA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>17</th>\n",
              "      <td>region_1_(80000-100000)_loc_256</td>\n",
              "      <td>CGGTTGGTAGGATCGACGGCCGACCAGGTCCATCTGCCACGTTTTA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>18</th>\n",
              "      <td>region_4_(15000-20000)_loc_280</td>\n",
              "      <td>CGGTTGGTAGGATCGACGGCCGTACAGGTCGCGCATCTCCGTTTTA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>19</th>\n",
              "      <td>region_2_(4000-7000)_loc_387</td>\n",
              "      <td>CGGTTGGTAGGATCGACGGCAAGGACAGGTGGGCCTCGCCGTTTTA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>region_4_(15000-20000)_fwd_checking_primer</td>\n",
              "      <td>TTGCTGCAACCTGCCTTG</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>region_2_(4000-7000)_fwd_checking_primer</td>\n",
              "      <td>GACCGCGCTGGGTG</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>region_3_(9000-14000)_fwd_checking_primer</td>\n",
              "      <td>ATCCCGCAGCTGCAGG</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>region_1_(80000-100000)_fwd_checking_primer</td>\n",
              "      <td>GCCGCTGGCCAAGC</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>region_4_(15000-20000)_rev_checking_primer</td>\n",
              "      <td>TAGGTGCCTCCGTCGA</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>5</th>\n",
              "      <td>region_2_(4000-7000)_rev_checking_primer</td>\n",
              "      <td>TCCGGCCGCATCG</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>6</th>\n",
              "      <td>region_3_(9000-14000)_rev_checking_primer</td>\n",
              "      <td>GAGGAGCGGCAATGGTCATC</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>7</th>\n",
              "      <td>region_1_(80000-100000)_rev_checking_primer</td>\n",
              "      <td>AGGCGTGCGCGTGG</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "                                           Name  \\\n",
              "0                   region_2_(4000-7000)_loc_35   \n",
              "1                   region_2_(4000-7000)_loc_34   \n",
              "2                 region_4_(15000-20000)_loc_54   \n",
              "3                region_1_(80000-100000)_loc_80   \n",
              "4                region_1_(80000-100000)_loc_81   \n",
              "5                 region_3_(9000-14000)_loc_118   \n",
              "6               region_1_(80000-100000)_loc_130   \n",
              "7               region_1_(80000-100000)_loc_131   \n",
              "8                 region_3_(9000-14000)_loc_146   \n",
              "9                  region_2_(4000-7000)_loc_148   \n",
              "10                 region_2_(4000-7000)_loc_147   \n",
              "11               region_4_(15000-20000)_loc_166   \n",
              "12                region_3_(9000-14000)_loc_200   \n",
              "13                region_3_(9000-14000)_loc_203   \n",
              "14               region_4_(15000-20000)_loc_204   \n",
              "15                region_3_(9000-14000)_loc_213   \n",
              "16               region_4_(15000-20000)_loc_254   \n",
              "17              region_1_(80000-100000)_loc_256   \n",
              "18               region_4_(15000-20000)_loc_280   \n",
              "19                 region_2_(4000-7000)_loc_387   \n",
              "0    region_4_(15000-20000)_fwd_checking_primer   \n",
              "1      region_2_(4000-7000)_fwd_checking_primer   \n",
              "2     region_3_(9000-14000)_fwd_checking_primer   \n",
              "3   region_1_(80000-100000)_fwd_checking_primer   \n",
              "4    region_4_(15000-20000)_rev_checking_primer   \n",
              "5      region_2_(4000-7000)_rev_checking_primer   \n",
              "6     region_3_(9000-14000)_rev_checking_primer   \n",
              "7   region_1_(80000-100000)_rev_checking_primer   \n",
              "\n",
              "                                             Sequence Concentration  \\\n",
              "0   CGGTTGGTAGGATCGACGGCCCCCAGACACCACGCACGCCGTTTTA...          25nm   \n",
              "1   CGGTTGGTAGGATCGACGGCCCCAGACACCACGCACGCCAGTTTTA...          25nm   \n",
              "2   CGGTTGGTAGGATCGACGGCGCTCAACCTCCAGGCCGGACGTTTTA...          25nm   \n",
              "3   CGGTTGGTAGGATCGACGGCGTTGTCCGTCAGCACTTCGAGTTTTA...          25nm   \n",
              "4   CGGTTGGTAGGATCGACGGCTTGTCCGTCAGCACTTCGAAGTTTTA...          25nm   \n",
              "5   CGGTTGGTAGGATCGACGGCTGCCCGATCTTCCAGGCCAGGTTTTA...          25nm   \n",
              "6   CGGTTGGTAGGATCGACGGCCGAAGGCCGAGCACACGGCGGTTTTA...          25nm   \n",
              "7   CGGTTGGTAGGATCGACGGCGAAGGCCGAGCACACGGCGCGTTTTA...          25nm   \n",
              "8   CGGTTGGTAGGATCGACGGCATTCGACGCAGAGACCGCACGTTTTA...          25nm   \n",
              "9   CGGTTGGTAGGATCGACGGCCATCCCAGGAAGGCAGCGCGGTTTTA...          25nm   \n",
              "10  CGGTTGGTAGGATCGACGGCATCCCAGGAAGGCAGCGCGCGTTTTA...          25nm   \n",
              "11  CGGTTGGTAGGATCGACGGCGCCACGAGCGGCATCCACTGGTTTTA...          25nm   \n",
              "12  CGGTTGGTAGGATCGACGGCCGACCTCATCAGCGTCGACAGTTTTA...          25nm   \n",
              "13  CGGTTGGTAGGATCGACGGCCCTCATCAGCGTCGACACGGGTTTTA...          25nm   \n",
              "14  CGGTTGGTAGGATCGACGGCGGGGGCGAGGTTCGGGACGCGTTTTA...          25nm   \n",
              "15  CGGTTGGTAGGATCGACGGCGTCGACACGGCGGCCCAGACGTTTTA...          25nm   \n",
              "16  CGGTTGGTAGGATCGACGGCTCACGTCACCGACGGCGTACGTTTTA...          25nm   \n",
              "17  CGGTTGGTAGGATCGACGGCCGACCAGGTCCATCTGCCACGTTTTA...          25nm   \n",
              "18  CGGTTGGTAGGATCGACGGCCGTACAGGTCGCGCATCTCCGTTTTA...          25nm   \n",
              "19  CGGTTGGTAGGATCGACGGCAAGGACAGGTGGGCCTCGCCGTTTTA...          25nm   \n",
              "0                                  TTGCTGCAACCTGCCTTG          25nm   \n",
              "1                                      GACCGCGCTGGGTG          25nm   \n",
              "2                                    ATCCCGCAGCTGCAGG          25nm   \n",
              "3                                      GCCGCTGGCCAAGC          25nm   \n",
              "4                                    TAGGTGCCTCCGTCGA          25nm   \n",
              "5                                       TCCGGCCGCATCG          25nm   \n",
              "6                                GAGGAGCGGCAATGGTCATC          25nm   \n",
              "7                                      AGGCGTGCGCGTGG          25nm   \n",
              "\n",
              "   Purification  \n",
              "0           STD  \n",
              "1           STD  \n",
              "2           STD  \n",
              "3           STD  \n",
              "4           STD  \n",
              "5           STD  \n",
              "6           STD  \n",
              "7           STD  \n",
              "8           STD  \n",
              "9           STD  \n",
              "10          STD  \n",
              "11          STD  \n",
              "12          STD  \n",
              "13          STD  \n",
              "14          STD  \n",
              "15          STD  \n",
              "16          STD  \n",
              "17          STD  \n",
              "18          STD  \n",
              "19          STD  \n",
              "0           STD  \n",
              "1           STD  \n",
              "2           STD  \n",
              "3           STD  \n",
              "4           STD  \n",
              "5           STD  \n",
              "6           STD  \n",
              "7           STD  "
            ]
          },
          "execution_count": 24,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "full_idt = pd.concat([idt_df1, idt_df2])\n",
        "full_idt"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "## Folder with all the generated I/O"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {},
      "outputs": [],
      "source": [
        "\n",
        "input_files = [\n",
        "    {\"name\": \"input_genome.gb\", \"content\": genome},\n",
        "    {\"name\": \"input_plasmid.gb\", \"content\": clean_plasmid}\n",
        "]\n",
        "\n",
        "output_files = [\n",
        "    {\"name\": \"cBEST_w_sgRNAs.gb\", \"content\": sgRNA_vectors}, # LIST OF Dseqrecords\n",
        "    {\"name\": \"primer_df.csv\", \"content\": checking_primers_df},\n",
        "    {\"name\": \"full_idt.csv\", \"content\": full_idt},\n",
        "    {\"name\": \"mutated_sgrna_df.csv\", \"content\": mutated_sgrna_df},\n",
        "    {\"name\": \"filtered_df.csv\", \"content\": filtered_df},\n",
        "    {\"name\": \"plasmid_metadata_df.csv\", \"content\": plasmid_metadata_df},\n",
        "\n",
        "]\n",
        "\n",
        "input_values = {\n",
        "    \"genes_to_knockout\": genes_to_KO,\n",
        "    \"polymerase_settings\": {\n",
        "        \"chosen_polymerase\": chosen_polymerase,\n",
        "        \"melting_temperature\": melting_temperature,\n",
        "        \"primer_concentration\": primer_concentration,\n",
        "        \"primer_number_increment\": primer_number_increment,\n",
        "        \"flanking_region_number\": flanking_region_number\n",
        "    },\n",
        "    \"filtering_metrics\": {\n",
        "        \"gc_upper\": gc_upper,\n",
        "        \"gc_lower\": gc_lower,\n",
        "        \"off_target_seed\": off_target_seed,\n",
        "        \"off_target_upper\": off_target_upper,\n",
        "        \"cas_type\": cas_type,\n",
        "        \"number_of_sgRNAs_per_group\": number_of_sgRNAs_per_group\n",
        "    },\n",
        "    \"overlapping_sequences\": {\n",
        "        \"up_homology\": str(up_homology),\n",
        "        \"dw_homology\": str(dw_homology)\n",
        "    }\n",
        "}\n",
        "\n",
        "\n",
        "# Paths to Markdown files\n",
        "markdown_file_paths = [\n",
        "    \"../../protocols/conjugation_protcol.md\",\n",
        "    \"../../protocols/single_target_crispr_plasmid_protcol.md\"\n",
        "\n",
        "]\n",
        "\n",
        "# Data and time\n",
        "timestamp = datetime.utcnow().isoformat()\n",
        "\n",
        "# Create project directory structure\n",
        "project_directory = ProjectDirectory(\n",
        "    project_name=f\"CRISPR_cBEST_workflow_{timestamp}\",\n",
        "    input_files=input_files,\n",
        "    output_files=output_files,\n",
        "    input_values=input_values,\n",
        "    markdown_file_paths=markdown_file_paths\n",
        ")\n",
        "\n",
        "\n",
        "# DO You want to save the folder? \n",
        "save_zip_folder = False \n",
        "\n",
        "if save_zip_folder: \n",
        "    # Generate the project directory structure and get the zip content\n",
        "    zip_content = project_directory.create_directory_structure(create_directories=False)\n",
        "\n",
        "    # Save the zip file to disk (optional)\n",
        "    with open(\"project_structure.zip\", \"wb\") as f:\n",
        "        f.write(zip_content)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {},
      "outputs": [],
      "source": []
    }
  ],
  "metadata": {
    "kernelspec": {
      "display_name": "constrain",
      "language": "python",
      "name": "python3"
    },
    "language_info": {
      "codemirror_mode": {
        "name": "ipython",
        "version": 3
      },
      "file_extension": ".py",
      "mimetype": "text/x-python",
      "name": "python",
      "nbconvert_exporter": "python",
      "pygments_lexer": "ipython3",
      "version": "3.11.9"
    }
  },
  "nbformat": 4,
  "nbformat_minor": 2
}
