{
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "## W2-CRISPR-cBEST workflow\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "Base-editing with CRISPR-BEST is a DSB-free method that uses a single sgRNA to target a specific genomic location with single-nucleotide–resolution(Figure 3A) which can enable in-vivo protein engineering or as it is often used, for introducing early stop-codons10,12. It can be used to introduce targeted point mutations in the genome of Streptomyces, facilitating studies on gene function and protein engineering. To get started quickly users can download the pCRISPR-cBEST plasmid file, the genome file for S.coelicor (A3) and streptoAIM will generate your plasmids based on the genes you chose to target (Figure 3, S5). "
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 1,
      "metadata": {},
      "outputs": [],
      "source": [
        "import sys\n",
        "import os\n",
        "\n",
        "# Ensure the src directory is in the Python path\n",
        "project_root = os.path.abspath(os.path.join(os.getcwd(), '../../'))\n",
        "if project_root not in sys.path:\n",
        "    sys.path.insert(0, project_root)\n",
        "\n",
        "import os\n",
        "import pandas as pd\n",
        "from pydna.dseqrecord import Dseqrecord\n",
        "from datetime import datetime\n",
        "\n",
        "\n",
        "from streptocad.sequence_loading.sequence_loading import load_and_process_gene_sequences, load_and_process_plasmid, load_and_process_genome_sequences\n",
        "from streptocad.utils import polymerase_dict,generate_project_directory_structure\n",
        "from streptocad.crispr.guideRNAcas3_9_12 import extract_sgRNAs, SgRNAargs\n",
        "from streptocad.cloning.ssDNA_bridging import assemble_plasmids_by_ssDNA_bridging, make_ssDNA_oligos\n",
        "from streptocad.crispr.crispr_best import identify_base_editing_sites, filter_sgrnas_for_base_editing, process_base_editing\n",
        "from streptocad.cloning.plasmid_processing import annotate_plasmid_with_sgrnas\n",
        "from streptocad.primers.primer_generation import checking_primers, create_idt_order_dataframe, primers_to_IDT\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "## INPUT"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 2,
      "metadata": {},
      "outputs": [],
      "source": [
        "# Inputs\n",
        "# 1 Add genome of choice (genbank, fasta)\n",
        "path_to_genome = '../../data/genomes/Streptomyces_coelicolor_A3_chromosome.gb'\n",
        "genome = load_and_process_genome_sequences(path_to_genome)[0]\n",
        "\n",
        "# 2 Add plasmid \n",
        "path_to_plasmid = '../../data/plasmids/ pCRISPR-cBEST Sequences.gbk'\n",
        "clean_plasmid = load_and_process_plasmid(path_to_plasmid)\n",
        "\n",
        "# 3 Choose genes to knock out (list)\n",
        "genes_to_KO = ['SCO5087','SCO5088', 'SCO5089']\n",
        "\n",
        "#### Advanced settings ####\n",
        "# 3 Choose polymerase and target melting temperature FOR CHECKING PRIMERS\n",
        "chosen_polymerase = polymerase_dict['Phusion High-Fidelity DNA Polymerase (GC Buffer)']\n",
        "melting_temperature = 65\n",
        "primer_concentration = 0.4 \n",
        "primer_number_increment = 1\n",
        "flanking_region_number = 500\n",
        "\n",
        "# 4 Filtering metrics for sgRNAs\n",
        "gc_upper = 0.99\n",
        "gc_lower = 0.01\n",
        "off_target_seed = 13\n",
        "off_target_upper = 10\n",
        "cas_type='cas9'\n",
        "number_of_sgRNAs_per_group = 5\n",
        "only_stop_codons = True\n",
        "\n",
        "# 6 Choose overlapping sequences for our plasmid we can use the following\n",
        "#As per the article **\"CRISPR–Cas9, CRISPRi and CRISPR-BEST-mediated genetic manipulation in streptomycetes\"** we need the following oligoes: \n",
        "#CGGTTGGTAGGATCGACGGC **-N20-** GTTTTAGAGCTAGAAATAGC\n",
        "up_homology = Dseqrecord('CGGTTGGTAGGATCGACGGC')\n",
        "dw_homology = Dseqrecord('GTTTTAGAGCTAGAAATAGC')"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "# Computation"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 3,
      "metadata": {},
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "sgRNA generated were outside the designated border in SCO5087. To incorporate this extent borders. Skipping to next locus tag.\n",
            "sgRNA generated were outside the designated border in SCO5087. To incorporate this extent borders. Skipping to next locus tag.\n",
            "sgRNA generated were outside the designated border in SCO5087. To incorporate this extent borders. Skipping to next locus tag.\n",
            "sgRNA generated were outside the designated border in SCO5088. To incorporate this extent borders. Skipping to next locus tag.\n",
            "sgRNA generated were outside the designated border in SCO5088. To incorporate this extent borders. Skipping to next locus tag.\n",
            "sgRNA generated were outside the designated border in SCO5088. To incorporate this extent borders. Skipping to next locus tag.\n",
            "sgRNA generated were outside the designated border in SCO5088. To incorporate this extent borders. Skipping to next locus tag.\n",
            "sgRNA generated were outside the designated border in SCO5089. To incorporate this extent borders. Skipping to next locus tag.\n",
            "sgRNA generated were outside the designated border in SCO5089. To incorporate this extent borders. Skipping to next locus tag.\n"
          ]
        },
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>strain_name</th>\n",
              "      <th>locus_tag</th>\n",
              "      <th>gene_loc</th>\n",
              "      <th>gene_strand</th>\n",
              "      <th>sgrna_strand</th>\n",
              "      <th>sgrna_loc</th>\n",
              "      <th>gc</th>\n",
              "      <th>pam</th>\n",
              "      <th>sgrna</th>\n",
              "      <th>sgrna_seed_sequence</th>\n",
              "      <th>off_target_count</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>597</th>\n",
              "      <td>Streptomyces_coelicolor_A3_chromosome</td>\n",
              "      <td>SCO5088</td>\n",
              "      <td>5531201</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>17</td>\n",
              "      <td>0.60</td>\n",
              "      <td></td>\n",
              "      <td>ATGAGCGTCCTCATCACCGG</td>\n",
              "      <td>TCCTCATCACCGG</td>\n",
              "      <td>-1</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>448</th>\n",
              "      <td>Streptomyces_coelicolor_A3_chromosome</td>\n",
              "      <td>SCO5088</td>\n",
              "      <td>5531201</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>1225</td>\n",
              "      <td>0.60</td>\n",
              "      <td>GG</td>\n",
              "      <td>TTACGGGGTCGGTGCGAAAC</td>\n",
              "      <td>GTCGGTGCGAAAC</td>\n",
              "      <td>-1</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>291</th>\n",
              "      <td>Streptomyces_coelicolor_A3_chromosome</td>\n",
              "      <td>SCO5087</td>\n",
              "      <td>5529801</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>19</td>\n",
              "      <td>0.85</td>\n",
              "      <td>GG</td>\n",
              "      <td>GTGCCGCTGGACGCGGCGCC</td>\n",
              "      <td>TGGACGCGGCGCC</td>\n",
              "      <td>-1</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>402</th>\n",
              "      <td>Streptomyces_coelicolor_A3_chromosome</td>\n",
              "      <td>SCO5088</td>\n",
              "      <td>5531201</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>879</td>\n",
              "      <td>0.70</td>\n",
              "      <td>CGG</td>\n",
              "      <td>GACGTCGACGTCCTCGGGAC</td>\n",
              "      <td>ACGTCCTCGGGAC</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>401</th>\n",
              "      <td>Streptomyces_coelicolor_A3_chromosome</td>\n",
              "      <td>SCO5088</td>\n",
              "      <td>5531201</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>873</td>\n",
              "      <td>0.75</td>\n",
              "      <td>CGG</td>\n",
              "      <td>GACGTCCTCGGGACCGGTCC</td>\n",
              "      <td>TCGGGACCGGTCC</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>...</th>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>39</th>\n",
              "      <td>Streptomyces_coelicolor_A3_chromosome</td>\n",
              "      <td>SCO5087</td>\n",
              "      <td>5529801</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>394</td>\n",
              "      <td>0.80</td>\n",
              "      <td>CGG</td>\n",
              "      <td>CGCTGGCGGCGAACGCCTCC</td>\n",
              "      <td>GGCGAACGCCTCC</td>\n",
              "      <td>6</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>257</th>\n",
              "      <td>Streptomyces_coelicolor_A3_chromosome</td>\n",
              "      <td>SCO5087</td>\n",
              "      <td>5529801</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>307</td>\n",
              "      <td>0.75</td>\n",
              "      <td>AGG</td>\n",
              "      <td>CGACTTCGACCCGGTCGCCG</td>\n",
              "      <td>GACCCGGTCGCCG</td>\n",
              "      <td>7</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>216</th>\n",
              "      <td>Streptomyces_coelicolor_A3_chromosome</td>\n",
              "      <td>SCO5087</td>\n",
              "      <td>5529801</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>604</td>\n",
              "      <td>0.90</td>\n",
              "      <td>AGG</td>\n",
              "      <td>CGCCTGGGCGGTCGGCGCCG</td>\n",
              "      <td>GCGGTCGGCGCCG</td>\n",
              "      <td>7</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>616</th>\n",
              "      <td>Streptomyces_coelicolor_A3_chromosome</td>\n",
              "      <td>SCO5089</td>\n",
              "      <td>5532449</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>203</td>\n",
              "      <td>0.80</td>\n",
              "      <td>GGG</td>\n",
              "      <td>ACGCGGCCGGCGACGTCGTC</td>\n",
              "      <td>CGGCGACGTCGTC</td>\n",
              "      <td>8</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>484</th>\n",
              "      <td>Streptomyces_coelicolor_A3_chromosome</td>\n",
              "      <td>SCO5088</td>\n",
              "      <td>5531201</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>928</td>\n",
              "      <td>0.85</td>\n",
              "      <td>GGG</td>\n",
              "      <td>GCTGGACGCCGCCGAGGCCC</td>\n",
              "      <td>GCCGCCGAGGCCC</td>\n",
              "      <td>8</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>638 rows × 11 columns</p>\n",
              "</div>"
            ],
            "text/plain": [
              "                               strain_name locus_tag  gene_loc  gene_strand  \\\n",
              "597  Streptomyces_coelicolor_A3_chromosome   SCO5088   5531201            1   \n",
              "448  Streptomyces_coelicolor_A3_chromosome   SCO5088   5531201            1   \n",
              "291  Streptomyces_coelicolor_A3_chromosome   SCO5087   5529801            1   \n",
              "402  Streptomyces_coelicolor_A3_chromosome   SCO5088   5531201            1   \n",
              "401  Streptomyces_coelicolor_A3_chromosome   SCO5088   5531201            1   \n",
              "..                                     ...       ...       ...          ...   \n",
              "39   Streptomyces_coelicolor_A3_chromosome   SCO5087   5529801            1   \n",
              "257  Streptomyces_coelicolor_A3_chromosome   SCO5087   5529801            1   \n",
              "216  Streptomyces_coelicolor_A3_chromosome   SCO5087   5529801            1   \n",
              "616  Streptomyces_coelicolor_A3_chromosome   SCO5089   5532449            1   \n",
              "484  Streptomyces_coelicolor_A3_chromosome   SCO5088   5531201            1   \n",
              "\n",
              "     sgrna_strand  sgrna_loc    gc  pam                 sgrna  \\\n",
              "597             1         17  0.60       ATGAGCGTCCTCATCACCGG   \n",
              "448            -1       1225  0.60   GG  TTACGGGGTCGGTGCGAAAC   \n",
              "291             1         19  0.85   GG  GTGCCGCTGGACGCGGCGCC   \n",
              "402            -1        879  0.70  CGG  GACGTCGACGTCCTCGGGAC   \n",
              "401            -1        873  0.75  CGG  GACGTCCTCGGGACCGGTCC   \n",
              "..            ...        ...   ...  ...                   ...   \n",
              "39             -1        394  0.80  CGG  CGCTGGCGGCGAACGCCTCC   \n",
              "257             1        307  0.75  AGG  CGACTTCGACCCGGTCGCCG   \n",
              "216             1        604  0.90  AGG  CGCCTGGGCGGTCGGCGCCG   \n",
              "616            -1        203  0.80  GGG  ACGCGGCCGGCGACGTCGTC   \n",
              "484             1        928  0.85  GGG  GCTGGACGCCGCCGAGGCCC   \n",
              "\n",
              "    sgrna_seed_sequence  off_target_count  \n",
              "597       TCCTCATCACCGG                -1  \n",
              "448       GTCGGTGCGAAAC                -1  \n",
              "291       TGGACGCGGCGCC                -1  \n",
              "402       ACGTCCTCGGGAC                 0  \n",
              "401       TCGGGACCGGTCC                 0  \n",
              "..                  ...               ...  \n",
              "39        GGCGAACGCCTCC                 6  \n",
              "257       GACCCGGTCGCCG                 7  \n",
              "216       GCGGTCGGCGCCG                 7  \n",
              "616       CGGCGACGTCGTC                 8  \n",
              "484       GCCGCCGAGGCCC                 8  \n",
              "\n",
              "[638 rows x 11 columns]"
            ]
          },
          "execution_count": 3,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "\n",
        "# Initialize SgRNAargs with desired parameters\n",
        "args = SgRNAargs(path_to_genome, \n",
        "                genes_to_KO,\n",
        "                step=['find', 'filter'],\n",
        "                gc_upper = gc_upper,\n",
        "                gc_lower = gc_lower,\n",
        "                off_target_seed = off_target_seed,\n",
        "                off_target_upper = off_target_upper,\n",
        "                cas_type='cas9'\n",
        "                )\n",
        "\n",
        "sgrna_df = extract_sgRNAs(args)\n",
        "sgrna_df"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 4,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>strain_name</th>\n",
              "      <th>locus_tag</th>\n",
              "      <th>gene_loc</th>\n",
              "      <th>gene_strand</th>\n",
              "      <th>sgrna_strand</th>\n",
              "      <th>sgrna_loc</th>\n",
              "      <th>gc</th>\n",
              "      <th>pam</th>\n",
              "      <th>sgrna</th>\n",
              "      <th>sgrna_seed_sequence</th>\n",
              "      <th>off_target_count</th>\n",
              "      <th>editable_cytosines</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>597</th>\n",
              "      <td>Streptomyces_coelicolor_A3_chromosome</td>\n",
              "      <td>SCO5088</td>\n",
              "      <td>5531201</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>17</td>\n",
              "      <td>0.60</td>\n",
              "      <td></td>\n",
              "      <td>ATGAGCGTCCTCATCACCGG</td>\n",
              "      <td>TCCTCATCACCGG</td>\n",
              "      <td>-1</td>\n",
              "      <td>6,9,10</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>448</th>\n",
              "      <td>Streptomyces_coelicolor_A3_chromosome</td>\n",
              "      <td>SCO5088</td>\n",
              "      <td>5531201</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>1225</td>\n",
              "      <td>0.60</td>\n",
              "      <td>GG</td>\n",
              "      <td>TTACGGGGTCGGTGCGAAAC</td>\n",
              "      <td>GTCGGTGCGAAAC</td>\n",
              "      <td>-1</td>\n",
              "      <td>4,10</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>291</th>\n",
              "      <td>Streptomyces_coelicolor_A3_chromosome</td>\n",
              "      <td>SCO5087</td>\n",
              "      <td>5529801</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>19</td>\n",
              "      <td>0.85</td>\n",
              "      <td>GG</td>\n",
              "      <td>GTGCCGCTGGACGCGGCGCC</td>\n",
              "      <td>TGGACGCGGCGCC</td>\n",
              "      <td>-1</td>\n",
              "      <td>4,5,7</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>402</th>\n",
              "      <td>Streptomyces_coelicolor_A3_chromosome</td>\n",
              "      <td>SCO5088</td>\n",
              "      <td>5531201</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>879</td>\n",
              "      <td>0.70</td>\n",
              "      <td>CGG</td>\n",
              "      <td>GACGTCGACGTCCTCGGGAC</td>\n",
              "      <td>ACGTCCTCGGGAC</td>\n",
              "      <td>0</td>\n",
              "      <td>3,6,9</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>401</th>\n",
              "      <td>Streptomyces_coelicolor_A3_chromosome</td>\n",
              "      <td>SCO5088</td>\n",
              "      <td>5531201</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>873</td>\n",
              "      <td>0.75</td>\n",
              "      <td>CGG</td>\n",
              "      <td>GACGTCCTCGGGACCGGTCC</td>\n",
              "      <td>TCGGGACCGGTCC</td>\n",
              "      <td>0</td>\n",
              "      <td>3,6,7,9</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>...</th>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>39</th>\n",
              "      <td>Streptomyces_coelicolor_A3_chromosome</td>\n",
              "      <td>SCO5087</td>\n",
              "      <td>5529801</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>394</td>\n",
              "      <td>0.80</td>\n",
              "      <td>CGG</td>\n",
              "      <td>CGCTGGCGGCGAACGCCTCC</td>\n",
              "      <td>GGCGAACGCCTCC</td>\n",
              "      <td>6</td>\n",
              "      <td>3,7,10</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>257</th>\n",
              "      <td>Streptomyces_coelicolor_A3_chromosome</td>\n",
              "      <td>SCO5087</td>\n",
              "      <td>5529801</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>307</td>\n",
              "      <td>0.75</td>\n",
              "      <td>AGG</td>\n",
              "      <td>CGACTTCGACCCGGTCGCCG</td>\n",
              "      <td>GACCCGGTCGCCG</td>\n",
              "      <td>7</td>\n",
              "      <td>4,7,10</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>216</th>\n",
              "      <td>Streptomyces_coelicolor_A3_chromosome</td>\n",
              "      <td>SCO5087</td>\n",
              "      <td>5529801</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>604</td>\n",
              "      <td>0.90</td>\n",
              "      <td>AGG</td>\n",
              "      <td>CGCCTGGGCGGTCGGCGCCG</td>\n",
              "      <td>GCGGTCGGCGCCG</td>\n",
              "      <td>7</td>\n",
              "      <td>3,4,9</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>616</th>\n",
              "      <td>Streptomyces_coelicolor_A3_chromosome</td>\n",
              "      <td>SCO5089</td>\n",
              "      <td>5532449</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>203</td>\n",
              "      <td>0.80</td>\n",
              "      <td>GGG</td>\n",
              "      <td>ACGCGGCCGGCGACGTCGTC</td>\n",
              "      <td>CGGCGACGTCGTC</td>\n",
              "      <td>8</td>\n",
              "      <td>4,7,8</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>484</th>\n",
              "      <td>Streptomyces_coelicolor_A3_chromosome</td>\n",
              "      <td>SCO5088</td>\n",
              "      <td>5531201</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>928</td>\n",
              "      <td>0.85</td>\n",
              "      <td>GGG</td>\n",
              "      <td>GCTGGACGCCGCCGAGGCCC</td>\n",
              "      <td>GCCGCCGAGGCCC</td>\n",
              "      <td>8</td>\n",
              "      <td>7,9,10</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>633 rows × 12 columns</p>\n",
              "</div>"
            ],
            "text/plain": [
              "                               strain_name locus_tag  gene_loc  gene_strand  \\\n",
              "597  Streptomyces_coelicolor_A3_chromosome   SCO5088   5531201            1   \n",
              "448  Streptomyces_coelicolor_A3_chromosome   SCO5088   5531201            1   \n",
              "291  Streptomyces_coelicolor_A3_chromosome   SCO5087   5529801            1   \n",
              "402  Streptomyces_coelicolor_A3_chromosome   SCO5088   5531201            1   \n",
              "401  Streptomyces_coelicolor_A3_chromosome   SCO5088   5531201            1   \n",
              "..                                     ...       ...       ...          ...   \n",
              "39   Streptomyces_coelicolor_A3_chromosome   SCO5087   5529801            1   \n",
              "257  Streptomyces_coelicolor_A3_chromosome   SCO5087   5529801            1   \n",
              "216  Streptomyces_coelicolor_A3_chromosome   SCO5087   5529801            1   \n",
              "616  Streptomyces_coelicolor_A3_chromosome   SCO5089   5532449            1   \n",
              "484  Streptomyces_coelicolor_A3_chromosome   SCO5088   5531201            1   \n",
              "\n",
              "     sgrna_strand  sgrna_loc    gc  pam                 sgrna  \\\n",
              "597             1         17  0.60       ATGAGCGTCCTCATCACCGG   \n",
              "448            -1       1225  0.60   GG  TTACGGGGTCGGTGCGAAAC   \n",
              "291             1         19  0.85   GG  GTGCCGCTGGACGCGGCGCC   \n",
              "402            -1        879  0.70  CGG  GACGTCGACGTCCTCGGGAC   \n",
              "401            -1        873  0.75  CGG  GACGTCCTCGGGACCGGTCC   \n",
              "..            ...        ...   ...  ...                   ...   \n",
              "39             -1        394  0.80  CGG  CGCTGGCGGCGAACGCCTCC   \n",
              "257             1        307  0.75  AGG  CGACTTCGACCCGGTCGCCG   \n",
              "216             1        604  0.90  AGG  CGCCTGGGCGGTCGGCGCCG   \n",
              "616            -1        203  0.80  GGG  ACGCGGCCGGCGACGTCGTC   \n",
              "484             1        928  0.85  GGG  GCTGGACGCCGCCGAGGCCC   \n",
              "\n",
              "    sgrna_seed_sequence  off_target_count editable_cytosines  \n",
              "597       TCCTCATCACCGG                -1             6,9,10  \n",
              "448       GTCGGTGCGAAAC                -1               4,10  \n",
              "291       TGGACGCGGCGCC                -1              4,5,7  \n",
              "402       ACGTCCTCGGGAC                 0              3,6,9  \n",
              "401       TCGGGACCGGTCC                 0            3,6,7,9  \n",
              "..                  ...               ...                ...  \n",
              "39        GGCGAACGCCTCC                 6             3,7,10  \n",
              "257       GACCCGGTCGCCG                 7             4,7,10  \n",
              "216       GCGGTCGGCGCCG                 7              3,4,9  \n",
              "616       CGGCGACGTCGTC                 8              4,7,8  \n",
              "484       GCCGCCGAGGCCC                 8             7,9,10  \n",
              "\n",
              "[633 rows x 12 columns]"
            ]
          },
          "execution_count": 4,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "# Load gene sequences\n",
        "gene_sequences = load_and_process_gene_sequences(path_to_genome)\n",
        "genes_to_KO_dict = {locus_tag: gene_sequences[locus_tag] for locus_tag in genes_to_KO if locus_tag in gene_sequences}\n",
        "\n",
        "# Identify and annotate base editing sites\n",
        "sgrna_df_with_editing = identify_base_editing_sites(sgrna_df)\n",
        "\n",
        "# filter out only sgRNAs that result in base-editing\n",
        "filtered_sgrna_df_for_base_editing = filter_sgrnas_for_base_editing(sgrna_df_with_editing)\n",
        "filtered_sgrna_df_for_base_editing"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 5,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>strain_name</th>\n",
              "      <th>locus_tag</th>\n",
              "      <th>gene_loc</th>\n",
              "      <th>gene_strand</th>\n",
              "      <th>sgrna_strand</th>\n",
              "      <th>sgrna_loc</th>\n",
              "      <th>gc</th>\n",
              "      <th>pam</th>\n",
              "      <th>sgrna</th>\n",
              "      <th>sgrna_seed_sequence</th>\n",
              "      <th>off_target_count</th>\n",
              "      <th>editable_cytosines</th>\n",
              "      <th>mutations</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>291</th>\n",
              "      <td>Streptomyces_coelicolor_A3_chromosome</td>\n",
              "      <td>SCO5087</td>\n",
              "      <td>5529801</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>19</td>\n",
              "      <td>0.85</td>\n",
              "      <td>GG</td>\n",
              "      <td>GTGCCGCTGGACGCGGCGCC</td>\n",
              "      <td>TGGACGCGGCGCC</td>\n",
              "      <td>-1</td>\n",
              "      <td>4,5,7</td>\n",
              "      <td>P2S</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>402</th>\n",
              "      <td>Streptomyces_coelicolor_A3_chromosome</td>\n",
              "      <td>SCO5088</td>\n",
              "      <td>5531201</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>879</td>\n",
              "      <td>0.70</td>\n",
              "      <td>CGG</td>\n",
              "      <td>GACGTCGACGTCCTCGGGAC</td>\n",
              "      <td>ACGTCCTCGGGAC</td>\n",
              "      <td>0</td>\n",
              "      <td>3,6,9</td>\n",
              "      <td>V291I, D292N, V293I</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>401</th>\n",
              "      <td>Streptomyces_coelicolor_A3_chromosome</td>\n",
              "      <td>SCO5088</td>\n",
              "      <td>5531201</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>873</td>\n",
              "      <td>0.75</td>\n",
              "      <td>CGG</td>\n",
              "      <td>GACGTCCTCGGGACCGGTCC</td>\n",
              "      <td>TCGGGACCGGTCC</td>\n",
              "      <td>0</td>\n",
              "      <td>3,6,7,9</td>\n",
              "      <td>E289K, D290N, V291I</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>400</th>\n",
              "      <td>Streptomyces_coelicolor_A3_chromosome</td>\n",
              "      <td>SCO5088</td>\n",
              "      <td>5531201</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>853</td>\n",
              "      <td>0.75</td>\n",
              "      <td>CGG</td>\n",
              "      <td>CGGCGTCGTTCAGCGCCAGC</td>\n",
              "      <td>GTTCAGCGCCAGC</td>\n",
              "      <td>0</td>\n",
              "      <td>4,7</td>\n",
              "      <td>D283N, A284T</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>399</th>\n",
              "      <td>Streptomyces_coelicolor_A3_chromosome</td>\n",
              "      <td>SCO5088</td>\n",
              "      <td>5531201</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>838</td>\n",
              "      <td>0.75</td>\n",
              "      <td>AGG</td>\n",
              "      <td>CCAGCCGGATCGCGCGTTCC</td>\n",
              "      <td>GATCGCGCGTTCC</td>\n",
              "      <td>0</td>\n",
              "      <td>5,6</td>\n",
              "      <td>R278Q</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>...</th>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>546</th>\n",
              "      <td>Streptomyces_coelicolor_A3_chromosome</td>\n",
              "      <td>SCO5088</td>\n",
              "      <td>5531201</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>487</td>\n",
              "      <td>0.80</td>\n",
              "      <td>GGG</td>\n",
              "      <td>CCTGGTCGCCGAGCAGGCCG</td>\n",
              "      <td>GCCGAGCAGGCCG</td>\n",
              "      <td>6</td>\n",
              "      <td>7,9,10</td>\n",
              "      <td>A159V</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>39</th>\n",
              "      <td>Streptomyces_coelicolor_A3_chromosome</td>\n",
              "      <td>SCO5087</td>\n",
              "      <td>5529801</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>394</td>\n",
              "      <td>0.80</td>\n",
              "      <td>CGG</td>\n",
              "      <td>CGCTGGCGGCGAACGCCTCC</td>\n",
              "      <td>GGCGAACGCCTCC</td>\n",
              "      <td>6</td>\n",
              "      <td>3,7,10</td>\n",
              "      <td>A129T, A130T, S131N</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>216</th>\n",
              "      <td>Streptomyces_coelicolor_A3_chromosome</td>\n",
              "      <td>SCO5087</td>\n",
              "      <td>5529801</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>604</td>\n",
              "      <td>0.90</td>\n",
              "      <td>AGG</td>\n",
              "      <td>CGCCTGGGCGGTCGGCGCCG</td>\n",
              "      <td>GCGGTCGGCGCCG</td>\n",
              "      <td>7</td>\n",
              "      <td>3,4,9</td>\n",
              "      <td>A196V, A198V</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>616</th>\n",
              "      <td>Streptomyces_coelicolor_A3_chromosome</td>\n",
              "      <td>SCO5089</td>\n",
              "      <td>5532449</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>203</td>\n",
              "      <td>0.80</td>\n",
              "      <td>GGG</td>\n",
              "      <td>ACGCGGCCGGCGACGTCGTC</td>\n",
              "      <td>CGGCGACGTCGTC</td>\n",
              "      <td>8</td>\n",
              "      <td>4,7,8</td>\n",
              "      <td>G66N, R67H</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>484</th>\n",
              "      <td>Streptomyces_coelicolor_A3_chromosome</td>\n",
              "      <td>SCO5088</td>\n",
              "      <td>5531201</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>928</td>\n",
              "      <td>0.85</td>\n",
              "      <td>GGG</td>\n",
              "      <td>GCTGGACGCCGCCGAGGCCC</td>\n",
              "      <td>GCCGCCGAGGCCC</td>\n",
              "      <td>8</td>\n",
              "      <td>7,9,10</td>\n",
              "      <td>A306V</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>562 rows × 13 columns</p>\n",
              "</div>"
            ],
            "text/plain": [
              "                               strain_name locus_tag  gene_loc  gene_strand  \\\n",
              "291  Streptomyces_coelicolor_A3_chromosome   SCO5087   5529801            1   \n",
              "402  Streptomyces_coelicolor_A3_chromosome   SCO5088   5531201            1   \n",
              "401  Streptomyces_coelicolor_A3_chromosome   SCO5088   5531201            1   \n",
              "400  Streptomyces_coelicolor_A3_chromosome   SCO5088   5531201            1   \n",
              "399  Streptomyces_coelicolor_A3_chromosome   SCO5088   5531201            1   \n",
              "..                                     ...       ...       ...          ...   \n",
              "546  Streptomyces_coelicolor_A3_chromosome   SCO5088   5531201            1   \n",
              "39   Streptomyces_coelicolor_A3_chromosome   SCO5087   5529801            1   \n",
              "216  Streptomyces_coelicolor_A3_chromosome   SCO5087   5529801            1   \n",
              "616  Streptomyces_coelicolor_A3_chromosome   SCO5089   5532449            1   \n",
              "484  Streptomyces_coelicolor_A3_chromosome   SCO5088   5531201            1   \n",
              "\n",
              "     sgrna_strand  sgrna_loc    gc  pam                 sgrna  \\\n",
              "291             1         19  0.85   GG  GTGCCGCTGGACGCGGCGCC   \n",
              "402            -1        879  0.70  CGG  GACGTCGACGTCCTCGGGAC   \n",
              "401            -1        873  0.75  CGG  GACGTCCTCGGGACCGGTCC   \n",
              "400            -1        853  0.75  CGG  CGGCGTCGTTCAGCGCCAGC   \n",
              "399            -1        838  0.75  AGG  CCAGCCGGATCGCGCGTTCC   \n",
              "..            ...        ...   ...  ...                   ...   \n",
              "546             1        487  0.80  GGG  CCTGGTCGCCGAGCAGGCCG   \n",
              "39             -1        394  0.80  CGG  CGCTGGCGGCGAACGCCTCC   \n",
              "216             1        604  0.90  AGG  CGCCTGGGCGGTCGGCGCCG   \n",
              "616            -1        203  0.80  GGG  ACGCGGCCGGCGACGTCGTC   \n",
              "484             1        928  0.85  GGG  GCTGGACGCCGCCGAGGCCC   \n",
              "\n",
              "    sgrna_seed_sequence  off_target_count editable_cytosines  \\\n",
              "291       TGGACGCGGCGCC                -1              4,5,7   \n",
              "402       ACGTCCTCGGGAC                 0              3,6,9   \n",
              "401       TCGGGACCGGTCC                 0            3,6,7,9   \n",
              "400       GTTCAGCGCCAGC                 0                4,7   \n",
              "399       GATCGCGCGTTCC                 0                5,6   \n",
              "..                  ...               ...                ...   \n",
              "546       GCCGAGCAGGCCG                 6             7,9,10   \n",
              "39        GGCGAACGCCTCC                 6             3,7,10   \n",
              "216       GCGGTCGGCGCCG                 7              3,4,9   \n",
              "616       CGGCGACGTCGTC                 8              4,7,8   \n",
              "484       GCCGCCGAGGCCC                 8             7,9,10   \n",
              "\n",
              "               mutations  \n",
              "291                  P2S  \n",
              "402  V291I, D292N, V293I  \n",
              "401  E289K, D290N, V291I  \n",
              "400         D283N, A284T  \n",
              "399                R278Q  \n",
              "..                   ...  \n",
              "546                A159V  \n",
              "39   A129T, A130T, S131N  \n",
              "216         A196V, A198V  \n",
              "616           G66N, R67H  \n",
              "484                A306V  \n",
              "\n",
              "[562 rows x 13 columns]"
            ]
          },
          "execution_count": 5,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "# Process the DataFrame to apply C-to-T mutations\n",
        "mutated_sgrna_df = process_base_editing(filtered_sgrna_df_for_base_editing, \n",
        "                                        genes_to_KO_dict, \n",
        "                                        only_stop_codons = False)\n",
        "mutated_sgrna_df"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 6,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>strain_name</th>\n",
              "      <th>locus_tag</th>\n",
              "      <th>gene_loc</th>\n",
              "      <th>gene_strand</th>\n",
              "      <th>sgrna_strand</th>\n",
              "      <th>sgrna_loc</th>\n",
              "      <th>gc</th>\n",
              "      <th>pam</th>\n",
              "      <th>sgrna</th>\n",
              "      <th>sgrna_seed_sequence</th>\n",
              "      <th>off_target_count</th>\n",
              "      <th>editable_cytosines</th>\n",
              "      <th>mutations</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>297</th>\n",
              "      <td>Streptomyces_coelicolor_A3_chromosome</td>\n",
              "      <td>SCO5088</td>\n",
              "      <td>5531201</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>73</td>\n",
              "      <td>0.80</td>\n",
              "      <td>AGG</td>\n",
              "      <td>CCGCCGACCAGTACGGCGCC</td>\n",
              "      <td>CCAGTACGGCGCC</td>\n",
              "      <td>1</td>\n",
              "      <td>4,5,8,9</td>\n",
              "      <td>W22*, A24T</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>296</th>\n",
              "      <td>Streptomyces_coelicolor_A3_chromosome</td>\n",
              "      <td>SCO5088</td>\n",
              "      <td>5531201</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>67</td>\n",
              "      <td>0.75</td>\n",
              "      <td>AGG</td>\n",
              "      <td>ACCAGTACGGCGCCAGGCCG</td>\n",
              "      <td>CGGCGCCAGGCCG</td>\n",
              "      <td>0</td>\n",
              "      <td>3,8</td>\n",
              "      <td>W22*</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>628</th>\n",
              "      <td>Streptomyces_coelicolor_A3_chromosome</td>\n",
              "      <td>SCO5089</td>\n",
              "      <td>5532449</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>162</td>\n",
              "      <td>0.80</td>\n",
              "      <td>CGG</td>\n",
              "      <td>CCGCGGCGCGACTCGAGAGC</td>\n",
              "      <td>GCGACTCGAGAGC</td>\n",
              "      <td>0</td>\n",
              "      <td>4,7,9</td>\n",
              "      <td>A49V, A50V, R51*</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>627</th>\n",
              "      <td>Streptomyces_coelicolor_A3_chromosome</td>\n",
              "      <td>SCO5089</td>\n",
              "      <td>5532449</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>167</td>\n",
              "      <td>0.70</td>\n",
              "      <td>CGG</td>\n",
              "      <td>GCGCGACTCGAGAGCCGGTA</td>\n",
              "      <td>TCGAGAGCCGGTA</td>\n",
              "      <td>0</td>\n",
              "      <td>4,7,9</td>\n",
              "      <td>R51*, L52F</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>17</th>\n",
              "      <td>Streptomyces_coelicolor_A3_chromosome</td>\n",
              "      <td>SCO5087</td>\n",
              "      <td>5529801</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>210</td>\n",
              "      <td>0.60</td>\n",
              "      <td>GGG</td>\n",
              "      <td>GAGCAGTTCCCAGAACTGCC</td>\n",
              "      <td>TCCCAGAACTGCC</td>\n",
              "      <td>0</td>\n",
              "      <td>4,9,10</td>\n",
              "      <td>W67*, E68K</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>576</th>\n",
              "      <td>Streptomyces_coelicolor_A3_chromosome</td>\n",
              "      <td>SCO5088</td>\n",
              "      <td>5531201</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>216</td>\n",
              "      <td>0.70</td>\n",
              "      <td>CGG</td>\n",
              "      <td>CGCAGACGGACCCCTCCACA</td>\n",
              "      <td>GGACCCCTCCACA</td>\n",
              "      <td>0</td>\n",
              "      <td>3,7</td>\n",
              "      <td>Q67*, T68M</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>324</th>\n",
              "      <td>Streptomyces_coelicolor_A3_chromosome</td>\n",
              "      <td>SCO5088</td>\n",
              "      <td>5531201</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>252</td>\n",
              "      <td>0.80</td>\n",
              "      <td>CGG</td>\n",
              "      <td>CTGGAGGGCCCAGTCCGCGG</td>\n",
              "      <td>GCCCAGTCCGCGG</td>\n",
              "      <td>0</td>\n",
              "      <td>9,10</td>\n",
              "      <td>W81*, A82T</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>570</th>\n",
              "      <td>Streptomyces_coelicolor_A3_chromosome</td>\n",
              "      <td>SCO5088</td>\n",
              "      <td>5531201</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>262</td>\n",
              "      <td>0.75</td>\n",
              "      <td>CGG</td>\n",
              "      <td>GGCCCTCCAGGACGCGAAGG</td>\n",
              "      <td>CAGGACGCGAAGG</td>\n",
              "      <td>0</td>\n",
              "      <td>3,4,5,7,8</td>\n",
              "      <td>A82V, L83F, Q84*</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>259</th>\n",
              "      <td>Streptomyces_coelicolor_A3_chromosome</td>\n",
              "      <td>SCO5087</td>\n",
              "      <td>5529801</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>283</td>\n",
              "      <td>0.75</td>\n",
              "      <td>AGG</td>\n",
              "      <td>CCGTTCACAGGTCGCGGCGG</td>\n",
              "      <td>CAGGTCGCGGCGG</td>\n",
              "      <td>0</td>\n",
              "      <td>6,8</td>\n",
              "      <td>S90L, Q91*</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>337</th>\n",
              "      <td>Streptomyces_coelicolor_A3_chromosome</td>\n",
              "      <td>SCO5088</td>\n",
              "      <td>5531201</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>361</td>\n",
              "      <td>0.60</td>\n",
              "      <td>CGG</td>\n",
              "      <td>ACCACAGCTTGCGGAACTCC</td>\n",
              "      <td>CTTGCGGAACTCC</td>\n",
              "      <td>0</td>\n",
              "      <td>3,5,8</td>\n",
              "      <td>W120*</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>553</th>\n",
              "      <td>Streptomyces_coelicolor_A3_chromosome</td>\n",
              "      <td>SCO5088</td>\n",
              "      <td>5531201</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>438</td>\n",
              "      <td>0.60</td>\n",
              "      <td>CGG</td>\n",
              "      <td>ACACCGGCCAGATCTCCATC</td>\n",
              "      <td>CCAGATCTCCATC</td>\n",
              "      <td>0</td>\n",
              "      <td>4,5,8,9</td>\n",
              "      <td>T141I, Q143*</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>552</th>\n",
              "      <td>Streptomyces_coelicolor_A3_chromosome</td>\n",
              "      <td>SCO5088</td>\n",
              "      <td>5531201</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>443</td>\n",
              "      <td>0.65</td>\n",
              "      <td>CGG</td>\n",
              "      <td>GGCCAGATCTCCATCCGGCA</td>\n",
              "      <td>TCTCCATCCGGCA</td>\n",
              "      <td>0</td>\n",
              "      <td>3,4,9</td>\n",
              "      <td>Q143*</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>544</th>\n",
              "      <td>Streptomyces_coelicolor_A3_chromosome</td>\n",
              "      <td>SCO5088</td>\n",
              "      <td>5531201</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>493</td>\n",
              "      <td>0.90</td>\n",
              "      <td>TGG</td>\n",
              "      <td>CGCCGAGCAGGCCGGGGGCC</td>\n",
              "      <td>CAGGCCGGGGGCC</td>\n",
              "      <td>0</td>\n",
              "      <td>3,4,8</td>\n",
              "      <td>A159V, Q161*</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>54</th>\n",
              "      <td>Streptomyces_coelicolor_A3_chromosome</td>\n",
              "      <td>SCO5087</td>\n",
              "      <td>5529801</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>522</td>\n",
              "      <td>0.70</td>\n",
              "      <td>CGG</td>\n",
              "      <td>GTCGACCTCCCAGTCACGGC</td>\n",
              "      <td>TCCCAGTCACGGC</td>\n",
              "      <td>0</td>\n",
              "      <td>3,6,7,9,10</td>\n",
              "      <td>W171*, E172K, V173I, D174N</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>63</th>\n",
              "      <td>Streptomyces_coelicolor_A3_chromosome</td>\n",
              "      <td>SCO5087</td>\n",
              "      <td>5529801</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>600</td>\n",
              "      <td>0.80</td>\n",
              "      <td>CGG</td>\n",
              "      <td>GCCGACCGCCCAGGCGACCT</td>\n",
              "      <td>GCCCAGGCGACCT</td>\n",
              "      <td>0</td>\n",
              "      <td>3,6,7,9,10</td>\n",
              "      <td>W197*, A198T, V199I, G200S</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>62</th>\n",
              "      <td>Streptomyces_coelicolor_A3_chromosome</td>\n",
              "      <td>SCO5087</td>\n",
              "      <td>5529801</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>596</td>\n",
              "      <td>0.80</td>\n",
              "      <td>CGG</td>\n",
              "      <td>ACCGCCCAGGCGACCTCGGC</td>\n",
              "      <td>AGGCGACCTCGGC</td>\n",
              "      <td>1</td>\n",
              "      <td>3,5,6,7</td>\n",
              "      <td>W197*, A198T</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>527</th>\n",
              "      <td>Streptomyces_coelicolor_A3_chromosome</td>\n",
              "      <td>SCO5088</td>\n",
              "      <td>5531201</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>605</td>\n",
              "      <td>0.65</td>\n",
              "      <td>TGG</td>\n",
              "      <td>TGGGTCTCGCAGATCGCCAG</td>\n",
              "      <td>CGCAGATCGCCAG</td>\n",
              "      <td>0</td>\n",
              "      <td>6,8,10</td>\n",
              "      <td>S198L, Q199*</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>526</th>\n",
              "      <td>Streptomyces_coelicolor_A3_chromosome</td>\n",
              "      <td>SCO5088</td>\n",
              "      <td>5531201</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>609</td>\n",
              "      <td>0.65</td>\n",
              "      <td>CGG</td>\n",
              "      <td>TCTCGCAGATCGCCAGTGGC</td>\n",
              "      <td>GATCGCCAGTGGC</td>\n",
              "      <td>0</td>\n",
              "      <td>4,6</td>\n",
              "      <td>S198L, Q199*</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>444</th>\n",
              "      <td>Streptomyces_coelicolor_A3_chromosome</td>\n",
              "      <td>SCO5088</td>\n",
              "      <td>5531201</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>1180</td>\n",
              "      <td>0.85</td>\n",
              "      <td>AGG</td>\n",
              "      <td>AGCCCCAGCGGCCCCTGGCC</td>\n",
              "      <td>GCGGCCCCTGGCC</td>\n",
              "      <td>0</td>\n",
              "      <td>3,4,5,6,9</td>\n",
              "      <td>R391H, W392*, G393N</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>445</th>\n",
              "      <td>Streptomyces_coelicolor_A3_chromosome</td>\n",
              "      <td>SCO5088</td>\n",
              "      <td>5531201</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>1185</td>\n",
              "      <td>0.75</td>\n",
              "      <td>TGG</td>\n",
              "      <td>GTTGAAGCCCCAGCGGCCCC</td>\n",
              "      <td>CCCCAGCGGCCCC</td>\n",
              "      <td>0</td>\n",
              "      <td>8,9,10</td>\n",
              "      <td>W392*, G393N</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "                               strain_name locus_tag  gene_loc  gene_strand  \\\n",
              "297  Streptomyces_coelicolor_A3_chromosome   SCO5088   5531201            1   \n",
              "296  Streptomyces_coelicolor_A3_chromosome   SCO5088   5531201            1   \n",
              "628  Streptomyces_coelicolor_A3_chromosome   SCO5089   5532449            1   \n",
              "627  Streptomyces_coelicolor_A3_chromosome   SCO5089   5532449            1   \n",
              "17   Streptomyces_coelicolor_A3_chromosome   SCO5087   5529801            1   \n",
              "576  Streptomyces_coelicolor_A3_chromosome   SCO5088   5531201            1   \n",
              "324  Streptomyces_coelicolor_A3_chromosome   SCO5088   5531201            1   \n",
              "570  Streptomyces_coelicolor_A3_chromosome   SCO5088   5531201            1   \n",
              "259  Streptomyces_coelicolor_A3_chromosome   SCO5087   5529801            1   \n",
              "337  Streptomyces_coelicolor_A3_chromosome   SCO5088   5531201            1   \n",
              "553  Streptomyces_coelicolor_A3_chromosome   SCO5088   5531201            1   \n",
              "552  Streptomyces_coelicolor_A3_chromosome   SCO5088   5531201            1   \n",
              "544  Streptomyces_coelicolor_A3_chromosome   SCO5088   5531201            1   \n",
              "54   Streptomyces_coelicolor_A3_chromosome   SCO5087   5529801            1   \n",
              "63   Streptomyces_coelicolor_A3_chromosome   SCO5087   5529801            1   \n",
              "62   Streptomyces_coelicolor_A3_chromosome   SCO5087   5529801            1   \n",
              "527  Streptomyces_coelicolor_A3_chromosome   SCO5088   5531201            1   \n",
              "526  Streptomyces_coelicolor_A3_chromosome   SCO5088   5531201            1   \n",
              "444  Streptomyces_coelicolor_A3_chromosome   SCO5088   5531201            1   \n",
              "445  Streptomyces_coelicolor_A3_chromosome   SCO5088   5531201            1   \n",
              "\n",
              "     sgrna_strand  sgrna_loc    gc  pam                 sgrna  \\\n",
              "297            -1         73  0.80  AGG  CCGCCGACCAGTACGGCGCC   \n",
              "296            -1         67  0.75  AGG  ACCAGTACGGCGCCAGGCCG   \n",
              "628             1        162  0.80  CGG  CCGCGGCGCGACTCGAGAGC   \n",
              "627             1        167  0.70  CGG  GCGCGACTCGAGAGCCGGTA   \n",
              "17             -1        210  0.60  GGG  GAGCAGTTCCCAGAACTGCC   \n",
              "576             1        216  0.70  CGG  CGCAGACGGACCCCTCCACA   \n",
              "324            -1        252  0.80  CGG  CTGGAGGGCCCAGTCCGCGG   \n",
              "570             1        262  0.75  CGG  GGCCCTCCAGGACGCGAAGG   \n",
              "259             1        283  0.75  AGG  CCGTTCACAGGTCGCGGCGG   \n",
              "337            -1        361  0.60  CGG  ACCACAGCTTGCGGAACTCC   \n",
              "553             1        438  0.60  CGG  ACACCGGCCAGATCTCCATC   \n",
              "552             1        443  0.65  CGG  GGCCAGATCTCCATCCGGCA   \n",
              "544             1        493  0.90  TGG  CGCCGAGCAGGCCGGGGGCC   \n",
              "54             -1        522  0.70  CGG  GTCGACCTCCCAGTCACGGC   \n",
              "63             -1        600  0.80  CGG  GCCGACCGCCCAGGCGACCT   \n",
              "62             -1        596  0.80  CGG  ACCGCCCAGGCGACCTCGGC   \n",
              "527             1        605  0.65  TGG  TGGGTCTCGCAGATCGCCAG   \n",
              "526             1        609  0.65  CGG  TCTCGCAGATCGCCAGTGGC   \n",
              "444            -1       1180  0.85  AGG  AGCCCCAGCGGCCCCTGGCC   \n",
              "445            -1       1185  0.75  TGG  GTTGAAGCCCCAGCGGCCCC   \n",
              "\n",
              "    sgrna_seed_sequence  off_target_count editable_cytosines  \\\n",
              "297       CCAGTACGGCGCC                 1            4,5,8,9   \n",
              "296       CGGCGCCAGGCCG                 0                3,8   \n",
              "628       GCGACTCGAGAGC                 0              4,7,9   \n",
              "627       TCGAGAGCCGGTA                 0              4,7,9   \n",
              "17        TCCCAGAACTGCC                 0             4,9,10   \n",
              "576       GGACCCCTCCACA                 0                3,7   \n",
              "324       GCCCAGTCCGCGG                 0               9,10   \n",
              "570       CAGGACGCGAAGG                 0          3,4,5,7,8   \n",
              "259       CAGGTCGCGGCGG                 0                6,8   \n",
              "337       CTTGCGGAACTCC                 0              3,5,8   \n",
              "553       CCAGATCTCCATC                 0            4,5,8,9   \n",
              "552       TCTCCATCCGGCA                 0              3,4,9   \n",
              "544       CAGGCCGGGGGCC                 0              3,4,8   \n",
              "54        TCCCAGTCACGGC                 0         3,6,7,9,10   \n",
              "63        GCCCAGGCGACCT                 0         3,6,7,9,10   \n",
              "62        AGGCGACCTCGGC                 1            3,5,6,7   \n",
              "527       CGCAGATCGCCAG                 0             6,8,10   \n",
              "526       GATCGCCAGTGGC                 0                4,6   \n",
              "444       GCGGCCCCTGGCC                 0          3,4,5,6,9   \n",
              "445       CCCCAGCGGCCCC                 0             8,9,10   \n",
              "\n",
              "                      mutations  \n",
              "297                  W22*, A24T  \n",
              "296                        W22*  \n",
              "628            A49V, A50V, R51*  \n",
              "627                  R51*, L52F  \n",
              "17                   W67*, E68K  \n",
              "576                  Q67*, T68M  \n",
              "324                  W81*, A82T  \n",
              "570            A82V, L83F, Q84*  \n",
              "259                  S90L, Q91*  \n",
              "337                       W120*  \n",
              "553                T141I, Q143*  \n",
              "552                       Q143*  \n",
              "544                A159V, Q161*  \n",
              "54   W171*, E172K, V173I, D174N  \n",
              "63   W197*, A198T, V199I, G200S  \n",
              "62                 W197*, A198T  \n",
              "527                S198L, Q199*  \n",
              "526                S198L, Q199*  \n",
              "444         R391H, W392*, G393N  \n",
              "445                W392*, G393N  "
            ]
          },
          "execution_count": 6,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "# Process the DataFrame to apply C-to-T mutations\n",
        "mutated_sgrna_df = process_base_editing(filtered_sgrna_df_for_base_editing, \n",
        "                                        genes_to_KO_dict, \n",
        "                                        only_stop_codons = True)\n",
        "mutated_sgrna_df"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 7,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>strain_name</th>\n",
              "      <th>locus_tag</th>\n",
              "      <th>gene_loc</th>\n",
              "      <th>gene_strand</th>\n",
              "      <th>sgrna_strand</th>\n",
              "      <th>sgrna_loc</th>\n",
              "      <th>gc</th>\n",
              "      <th>pam</th>\n",
              "      <th>sgrna</th>\n",
              "      <th>sgrna_seed_sequence</th>\n",
              "      <th>off_target_count</th>\n",
              "      <th>editable_cytosines</th>\n",
              "      <th>mutations</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>297</th>\n",
              "      <td>Streptomyces_coelicolor_A3_chromosome</td>\n",
              "      <td>SCO5088</td>\n",
              "      <td>5531201</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>73</td>\n",
              "      <td>0.80</td>\n",
              "      <td>AGG</td>\n",
              "      <td>CCGCCGACCAGTACGGCGCC</td>\n",
              "      <td>CCAGTACGGCGCC</td>\n",
              "      <td>1</td>\n",
              "      <td>4,5,8,9</td>\n",
              "      <td>W22*, A24T</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>296</th>\n",
              "      <td>Streptomyces_coelicolor_A3_chromosome</td>\n",
              "      <td>SCO5088</td>\n",
              "      <td>5531201</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>67</td>\n",
              "      <td>0.75</td>\n",
              "      <td>AGG</td>\n",
              "      <td>ACCAGTACGGCGCCAGGCCG</td>\n",
              "      <td>CGGCGCCAGGCCG</td>\n",
              "      <td>0</td>\n",
              "      <td>3,8</td>\n",
              "      <td>W22*</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>628</th>\n",
              "      <td>Streptomyces_coelicolor_A3_chromosome</td>\n",
              "      <td>SCO5089</td>\n",
              "      <td>5532449</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>162</td>\n",
              "      <td>0.80</td>\n",
              "      <td>CGG</td>\n",
              "      <td>CCGCGGCGCGACTCGAGAGC</td>\n",
              "      <td>GCGACTCGAGAGC</td>\n",
              "      <td>0</td>\n",
              "      <td>4,7,9</td>\n",
              "      <td>A49V, A50V, R51*</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>627</th>\n",
              "      <td>Streptomyces_coelicolor_A3_chromosome</td>\n",
              "      <td>SCO5089</td>\n",
              "      <td>5532449</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>167</td>\n",
              "      <td>0.70</td>\n",
              "      <td>CGG</td>\n",
              "      <td>GCGCGACTCGAGAGCCGGTA</td>\n",
              "      <td>TCGAGAGCCGGTA</td>\n",
              "      <td>0</td>\n",
              "      <td>4,7,9</td>\n",
              "      <td>R51*, L52F</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>17</th>\n",
              "      <td>Streptomyces_coelicolor_A3_chromosome</td>\n",
              "      <td>SCO5087</td>\n",
              "      <td>5529801</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>210</td>\n",
              "      <td>0.60</td>\n",
              "      <td>GGG</td>\n",
              "      <td>GAGCAGTTCCCAGAACTGCC</td>\n",
              "      <td>TCCCAGAACTGCC</td>\n",
              "      <td>0</td>\n",
              "      <td>4,9,10</td>\n",
              "      <td>W67*, E68K</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>576</th>\n",
              "      <td>Streptomyces_coelicolor_A3_chromosome</td>\n",
              "      <td>SCO5088</td>\n",
              "      <td>5531201</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>216</td>\n",
              "      <td>0.70</td>\n",
              "      <td>CGG</td>\n",
              "      <td>CGCAGACGGACCCCTCCACA</td>\n",
              "      <td>GGACCCCTCCACA</td>\n",
              "      <td>0</td>\n",
              "      <td>3,7</td>\n",
              "      <td>Q67*, T68M</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>324</th>\n",
              "      <td>Streptomyces_coelicolor_A3_chromosome</td>\n",
              "      <td>SCO5088</td>\n",
              "      <td>5531201</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>252</td>\n",
              "      <td>0.80</td>\n",
              "      <td>CGG</td>\n",
              "      <td>CTGGAGGGCCCAGTCCGCGG</td>\n",
              "      <td>GCCCAGTCCGCGG</td>\n",
              "      <td>0</td>\n",
              "      <td>9,10</td>\n",
              "      <td>W81*, A82T</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>570</th>\n",
              "      <td>Streptomyces_coelicolor_A3_chromosome</td>\n",
              "      <td>SCO5088</td>\n",
              "      <td>5531201</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>262</td>\n",
              "      <td>0.75</td>\n",
              "      <td>CGG</td>\n",
              "      <td>GGCCCTCCAGGACGCGAAGG</td>\n",
              "      <td>CAGGACGCGAAGG</td>\n",
              "      <td>0</td>\n",
              "      <td>3,4,5,7,8</td>\n",
              "      <td>A82V, L83F, Q84*</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>259</th>\n",
              "      <td>Streptomyces_coelicolor_A3_chromosome</td>\n",
              "      <td>SCO5087</td>\n",
              "      <td>5529801</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>283</td>\n",
              "      <td>0.75</td>\n",
              "      <td>AGG</td>\n",
              "      <td>CCGTTCACAGGTCGCGGCGG</td>\n",
              "      <td>CAGGTCGCGGCGG</td>\n",
              "      <td>0</td>\n",
              "      <td>6,8</td>\n",
              "      <td>S90L, Q91*</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>54</th>\n",
              "      <td>Streptomyces_coelicolor_A3_chromosome</td>\n",
              "      <td>SCO5087</td>\n",
              "      <td>5529801</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>522</td>\n",
              "      <td>0.70</td>\n",
              "      <td>CGG</td>\n",
              "      <td>GTCGACCTCCCAGTCACGGC</td>\n",
              "      <td>TCCCAGTCACGGC</td>\n",
              "      <td>0</td>\n",
              "      <td>3,6,7,9,10</td>\n",
              "      <td>W171*, E172K, V173I, D174N</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>63</th>\n",
              "      <td>Streptomyces_coelicolor_A3_chromosome</td>\n",
              "      <td>SCO5087</td>\n",
              "      <td>5529801</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>600</td>\n",
              "      <td>0.80</td>\n",
              "      <td>CGG</td>\n",
              "      <td>GCCGACCGCCCAGGCGACCT</td>\n",
              "      <td>GCCCAGGCGACCT</td>\n",
              "      <td>0</td>\n",
              "      <td>3,6,7,9,10</td>\n",
              "      <td>W197*, A198T, V199I, G200S</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>62</th>\n",
              "      <td>Streptomyces_coelicolor_A3_chromosome</td>\n",
              "      <td>SCO5087</td>\n",
              "      <td>5529801</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>596</td>\n",
              "      <td>0.80</td>\n",
              "      <td>CGG</td>\n",
              "      <td>ACCGCCCAGGCGACCTCGGC</td>\n",
              "      <td>AGGCGACCTCGGC</td>\n",
              "      <td>1</td>\n",
              "      <td>3,5,6,7</td>\n",
              "      <td>W197*, A198T</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "                               strain_name locus_tag  gene_loc  gene_strand  \\\n",
              "297  Streptomyces_coelicolor_A3_chromosome   SCO5088   5531201            1   \n",
              "296  Streptomyces_coelicolor_A3_chromosome   SCO5088   5531201            1   \n",
              "628  Streptomyces_coelicolor_A3_chromosome   SCO5089   5532449            1   \n",
              "627  Streptomyces_coelicolor_A3_chromosome   SCO5089   5532449            1   \n",
              "17   Streptomyces_coelicolor_A3_chromosome   SCO5087   5529801            1   \n",
              "576  Streptomyces_coelicolor_A3_chromosome   SCO5088   5531201            1   \n",
              "324  Streptomyces_coelicolor_A3_chromosome   SCO5088   5531201            1   \n",
              "570  Streptomyces_coelicolor_A3_chromosome   SCO5088   5531201            1   \n",
              "259  Streptomyces_coelicolor_A3_chromosome   SCO5087   5529801            1   \n",
              "54   Streptomyces_coelicolor_A3_chromosome   SCO5087   5529801            1   \n",
              "63   Streptomyces_coelicolor_A3_chromosome   SCO5087   5529801            1   \n",
              "62   Streptomyces_coelicolor_A3_chromosome   SCO5087   5529801            1   \n",
              "\n",
              "     sgrna_strand  sgrna_loc    gc  pam                 sgrna  \\\n",
              "297            -1         73  0.80  AGG  CCGCCGACCAGTACGGCGCC   \n",
              "296            -1         67  0.75  AGG  ACCAGTACGGCGCCAGGCCG   \n",
              "628             1        162  0.80  CGG  CCGCGGCGCGACTCGAGAGC   \n",
              "627             1        167  0.70  CGG  GCGCGACTCGAGAGCCGGTA   \n",
              "17             -1        210  0.60  GGG  GAGCAGTTCCCAGAACTGCC   \n",
              "576             1        216  0.70  CGG  CGCAGACGGACCCCTCCACA   \n",
              "324            -1        252  0.80  CGG  CTGGAGGGCCCAGTCCGCGG   \n",
              "570             1        262  0.75  CGG  GGCCCTCCAGGACGCGAAGG   \n",
              "259             1        283  0.75  AGG  CCGTTCACAGGTCGCGGCGG   \n",
              "54             -1        522  0.70  CGG  GTCGACCTCCCAGTCACGGC   \n",
              "63             -1        600  0.80  CGG  GCCGACCGCCCAGGCGACCT   \n",
              "62             -1        596  0.80  CGG  ACCGCCCAGGCGACCTCGGC   \n",
              "\n",
              "    sgrna_seed_sequence  off_target_count editable_cytosines  \\\n",
              "297       CCAGTACGGCGCC                 1            4,5,8,9   \n",
              "296       CGGCGCCAGGCCG                 0                3,8   \n",
              "628       GCGACTCGAGAGC                 0              4,7,9   \n",
              "627       TCGAGAGCCGGTA                 0              4,7,9   \n",
              "17        TCCCAGAACTGCC                 0             4,9,10   \n",
              "576       GGACCCCTCCACA                 0                3,7   \n",
              "324       GCCCAGTCCGCGG                 0               9,10   \n",
              "570       CAGGACGCGAAGG                 0          3,4,5,7,8   \n",
              "259       CAGGTCGCGGCGG                 0                6,8   \n",
              "54        TCCCAGTCACGGC                 0         3,6,7,9,10   \n",
              "63        GCCCAGGCGACCT                 0         3,6,7,9,10   \n",
              "62        AGGCGACCTCGGC                 1            3,5,6,7   \n",
              "\n",
              "                      mutations  \n",
              "297                  W22*, A24T  \n",
              "296                        W22*  \n",
              "628            A49V, A50V, R51*  \n",
              "627                  R51*, L52F  \n",
              "17                   W67*, E68K  \n",
              "576                  Q67*, T68M  \n",
              "324                  W81*, A82T  \n",
              "570            A82V, L83F, Q84*  \n",
              "259                  S90L, Q91*  \n",
              "54   W171*, E172K, V173I, D174N  \n",
              "63   W197*, A198T, V199I, G200S  \n",
              "62                 W197*, A198T  "
            ]
          },
          "execution_count": 7,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "# Filter the DataFrame to retain only up to 5 sgRNA sequences per locus_tag\n",
        "filtered_df = mutated_sgrna_df.groupby('locus_tag').head(number_of_sgRNAs_per_group)\n",
        "filtered_df"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "## Output"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 8,
      "metadata": {},
      "outputs": [],
      "source": [
        "# MAke oligoes\n",
        "list_of_ssDNAs = make_ssDNA_oligos(filtered_df, upstream_ovh = up_homology,\n",
        "                      downstream_ovh=dw_homology)\n",
        "# cut plasmid\n",
        "from Bio.Restriction import NcoI\n",
        "linearized_plasmid = sorted(clean_plasmid.cut(NcoI), key=lambda x: len(x), reverse=True)[0]\n",
        "\n",
        "# assemble plasmid\n",
        "sgRNA_vectors = assemble_plasmids_by_ssDNA_bridging(list_of_ssDNAs,linearized_plasmid)\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 9,
      "metadata": {},
      "outputs": [],
      "source": [
        "# Constructing a meaningful name, ID, and description for the assembled plasmid using user input\n",
        "targeting_info = []\n",
        "for index, row in filtered_df.iterrows():\n",
        "    formatted_str = f\"pCRISPR-BEST_{row['locus_tag']}_p{row['sgrna_loc']}\"\n",
        "    targeting_info.append(formatted_str)\n",
        "\n",
        "for i in range(len(sgRNA_vectors)):\n",
        "    sgRNA_vectors[i].name = f'{targeting_info[i]}_#{i+1}'\n",
        "    sgRNA_vectors[i].id = sgRNA_vectors[i].name  # Using the same value for ID as for name for simplicity\n",
        "    sgRNA_vectors[i].description = f'Assembled plasmid targeting {\", \".join(genes_to_KO)} for base-editing, assembled using StreptoAIM.'\n",
        "\n",
        "# annotate plasmids\n",
        "for plasmid in sgRNA_vectors: \n",
        "    annotate_plasmid_with_sgrnas(plasmid, filtered_df)\n",
        "\n",
        "# PRINT\n",
        "print_plasmids = False\n",
        "if print_plasmids: \n",
        "    for vector in sgRNA_vectors: \n",
        "        vector.write(f\"../../data/plasmids/sgRNA_plasmids_pCRISPR_cBEST/{vector.id}.gb\")"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "# IDT primers"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 10,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>Name</th>\n",
              "      <th>Sequence</th>\n",
              "      <th>Concentration</th>\n",
              "      <th>Purification</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>SCO5088_loc_73</td>\n",
              "      <td>CGGTTGGTAGGATCGACGGCCCGCCGACCAGTACGGCGCCGTTTTA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>SCO5088_loc_67</td>\n",
              "      <td>CGGTTGGTAGGATCGACGGCACCAGTACGGCGCCAGGCCGGTTTTA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>SCO5089_loc_162</td>\n",
              "      <td>CGGTTGGTAGGATCGACGGCCCGCGGCGCGACTCGAGAGCGTTTTA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>SCO5089_loc_167</td>\n",
              "      <td>CGGTTGGTAGGATCGACGGCGCGCGACTCGAGAGCCGGTAGTTTTA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>SCO5087_loc_210</td>\n",
              "      <td>CGGTTGGTAGGATCGACGGCGAGCAGTTCCCAGAACTGCCGTTTTA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>5</th>\n",
              "      <td>SCO5088_loc_216</td>\n",
              "      <td>CGGTTGGTAGGATCGACGGCCGCAGACGGACCCCTCCACAGTTTTA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>6</th>\n",
              "      <td>SCO5088_loc_252</td>\n",
              "      <td>CGGTTGGTAGGATCGACGGCCTGGAGGGCCCAGTCCGCGGGTTTTA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>7</th>\n",
              "      <td>SCO5088_loc_262</td>\n",
              "      <td>CGGTTGGTAGGATCGACGGCGGCCCTCCAGGACGCGAAGGGTTTTA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>8</th>\n",
              "      <td>SCO5087_loc_283</td>\n",
              "      <td>CGGTTGGTAGGATCGACGGCCCGTTCACAGGTCGCGGCGGGTTTTA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>9</th>\n",
              "      <td>SCO5087_loc_522</td>\n",
              "      <td>CGGTTGGTAGGATCGACGGCGTCGACCTCCCAGTCACGGCGTTTTA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>10</th>\n",
              "      <td>SCO5087_loc_600</td>\n",
              "      <td>CGGTTGGTAGGATCGACGGCGCCGACCGCCCAGGCGACCTGTTTTA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>11</th>\n",
              "      <td>SCO5087_loc_596</td>\n",
              "      <td>CGGTTGGTAGGATCGACGGCACCGCCCAGGCGACCTCGGCGTTTTA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "               Name                                           Sequence  \\\n",
              "0    SCO5088_loc_73  CGGTTGGTAGGATCGACGGCCCGCCGACCAGTACGGCGCCGTTTTA...   \n",
              "1    SCO5088_loc_67  CGGTTGGTAGGATCGACGGCACCAGTACGGCGCCAGGCCGGTTTTA...   \n",
              "2   SCO5089_loc_162  CGGTTGGTAGGATCGACGGCCCGCGGCGCGACTCGAGAGCGTTTTA...   \n",
              "3   SCO5089_loc_167  CGGTTGGTAGGATCGACGGCGCGCGACTCGAGAGCCGGTAGTTTTA...   \n",
              "4   SCO5087_loc_210  CGGTTGGTAGGATCGACGGCGAGCAGTTCCCAGAACTGCCGTTTTA...   \n",
              "5   SCO5088_loc_216  CGGTTGGTAGGATCGACGGCCGCAGACGGACCCCTCCACAGTTTTA...   \n",
              "6   SCO5088_loc_252  CGGTTGGTAGGATCGACGGCCTGGAGGGCCCAGTCCGCGGGTTTTA...   \n",
              "7   SCO5088_loc_262  CGGTTGGTAGGATCGACGGCGGCCCTCCAGGACGCGAAGGGTTTTA...   \n",
              "8   SCO5087_loc_283  CGGTTGGTAGGATCGACGGCCCGTTCACAGGTCGCGGCGGGTTTTA...   \n",
              "9   SCO5087_loc_522  CGGTTGGTAGGATCGACGGCGTCGACCTCCCAGTCACGGCGTTTTA...   \n",
              "10  SCO5087_loc_600  CGGTTGGTAGGATCGACGGCGCCGACCGCCCAGGCGACCTGTTTTA...   \n",
              "11  SCO5087_loc_596  CGGTTGGTAGGATCGACGGCACCGCCCAGGCGACCTCGGCGTTTTA...   \n",
              "\n",
              "   Concentration Purification  \n",
              "0           25nm          STD  \n",
              "1           25nm          STD  \n",
              "2           25nm          STD  \n",
              "3           25nm          STD  \n",
              "4           25nm          STD  \n",
              "5           25nm          STD  \n",
              "6           25nm          STD  \n",
              "7           25nm          STD  \n",
              "8           25nm          STD  \n",
              "9           25nm          STD  \n",
              "10          25nm          STD  \n",
              "11          25nm          STD  "
            ]
          },
          "execution_count": 10,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "idt_df1 = primers_to_IDT(list_of_ssDNAs)\n",
        "idt_df1"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 11,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>Locus Tag</th>\n",
              "      <th>f_primer_name</th>\n",
              "      <th>r_primer_name</th>\n",
              "      <th>f_primer_sequences(5-3)</th>\n",
              "      <th>r_primer_sequences(5-3)</th>\n",
              "      <th>f_tm</th>\n",
              "      <th>r_tm</th>\n",
              "      <th>ta</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>SCO5087</td>\n",
              "      <td>SCO5087_fwd_checking_primer</td>\n",
              "      <td>SCO5087_rev_checking_primer</td>\n",
              "      <td>GACGATTCGGCCCGTG</td>\n",
              "      <td>CAGGGCGTCCAGGC</td>\n",
              "      <td>59</td>\n",
              "      <td>58</td>\n",
              "      <td>61</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>SCO5088</td>\n",
              "      <td>SCO5088_fwd_checking_primer</td>\n",
              "      <td>SCO5088_rev_checking_primer</td>\n",
              "      <td>CGCATCCACGCCGAGA</td>\n",
              "      <td>CAGCGTCCGGCGG</td>\n",
              "      <td>61</td>\n",
              "      <td>58</td>\n",
              "      <td>62</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>SCO5089</td>\n",
              "      <td>SCO5089_fwd_checking_primer</td>\n",
              "      <td>SCO5089_rev_checking_primer</td>\n",
              "      <td>CGTACGGCGAACTCGC</td>\n",
              "      <td>GTCCCCGGACGGC</td>\n",
              "      <td>59</td>\n",
              "      <td>57</td>\n",
              "      <td>60</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "  Locus Tag                f_primer_name                r_primer_name  \\\n",
              "0   SCO5087  SCO5087_fwd_checking_primer  SCO5087_rev_checking_primer   \n",
              "1   SCO5088  SCO5088_fwd_checking_primer  SCO5088_rev_checking_primer   \n",
              "2   SCO5089  SCO5089_fwd_checking_primer  SCO5089_rev_checking_primer   \n",
              "\n",
              "  f_primer_sequences(5-3) r_primer_sequences(5-3)  f_tm  r_tm  ta  \n",
              "0        GACGATTCGGCCCGTG          CAGGGCGTCCAGGC    59    58  61  \n",
              "1        CGCATCCACGCCGAGA           CAGCGTCCGGCGG    61    58  62  \n",
              "2        CGTACGGCGAACTCGC           GTCCCCGGACGGC    59    57  60  "
            ]
          },
          "execution_count": 11,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "\n",
        "# Getting checking primeres\n",
        "checking_primers_df = checking_primers(path_to_genome, genes_to_KO, \n",
        "                                       flanking_region=flanking_region_number,\n",
        "                                       target_tm = melting_temperature, \n",
        "                                        primer_concentration = primer_concentration, \n",
        "                                        polymerase = chosen_polymerase)\n",
        "checking_primers_df"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 12,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>Name</th>\n",
              "      <th>Sequence</th>\n",
              "      <th>Concentration</th>\n",
              "      <th>Purification</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>SCO5087_fwd_checking_primer</td>\n",
              "      <td>GACGATTCGGCCCGTG</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>SCO5088_fwd_checking_primer</td>\n",
              "      <td>CGCATCCACGCCGAGA</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>SCO5089_fwd_checking_primer</td>\n",
              "      <td>CGTACGGCGAACTCGC</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>SCO5087_rev_checking_primer</td>\n",
              "      <td>CAGGGCGTCCAGGC</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>SCO5088_rev_checking_primer</td>\n",
              "      <td>CAGCGTCCGGCGG</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>5</th>\n",
              "      <td>SCO5089_rev_checking_primer</td>\n",
              "      <td>GTCCCCGGACGGC</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "                          Name          Sequence Concentration Purification\n",
              "0  SCO5087_fwd_checking_primer  GACGATTCGGCCCGTG          25nm          STD\n",
              "1  SCO5088_fwd_checking_primer  CGCATCCACGCCGAGA          25nm          STD\n",
              "2  SCO5089_fwd_checking_primer  CGTACGGCGAACTCGC          25nm          STD\n",
              "3  SCO5087_rev_checking_primer    CAGGGCGTCCAGGC          25nm          STD\n",
              "4  SCO5088_rev_checking_primer     CAGCGTCCGGCGG          25nm          STD\n",
              "5  SCO5089_rev_checking_primer     GTCCCCGGACGGC          25nm          STD"
            ]
          },
          "execution_count": 12,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "idt_df2 = create_idt_order_dataframe(checking_primers_df)\n",
        "idt_df2"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 13,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>Name</th>\n",
              "      <th>Sequence</th>\n",
              "      <th>Concentration</th>\n",
              "      <th>Purification</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>SCO5088_loc_73</td>\n",
              "      <td>CGGTTGGTAGGATCGACGGCCCGCCGACCAGTACGGCGCCGTTTTA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>SCO5088_loc_67</td>\n",
              "      <td>CGGTTGGTAGGATCGACGGCACCAGTACGGCGCCAGGCCGGTTTTA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>SCO5089_loc_162</td>\n",
              "      <td>CGGTTGGTAGGATCGACGGCCCGCGGCGCGACTCGAGAGCGTTTTA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>SCO5089_loc_167</td>\n",
              "      <td>CGGTTGGTAGGATCGACGGCGCGCGACTCGAGAGCCGGTAGTTTTA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>SCO5087_loc_210</td>\n",
              "      <td>CGGTTGGTAGGATCGACGGCGAGCAGTTCCCAGAACTGCCGTTTTA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>5</th>\n",
              "      <td>SCO5088_loc_216</td>\n",
              "      <td>CGGTTGGTAGGATCGACGGCCGCAGACGGACCCCTCCACAGTTTTA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>6</th>\n",
              "      <td>SCO5088_loc_252</td>\n",
              "      <td>CGGTTGGTAGGATCGACGGCCTGGAGGGCCCAGTCCGCGGGTTTTA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>7</th>\n",
              "      <td>SCO5088_loc_262</td>\n",
              "      <td>CGGTTGGTAGGATCGACGGCGGCCCTCCAGGACGCGAAGGGTTTTA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>8</th>\n",
              "      <td>SCO5087_loc_283</td>\n",
              "      <td>CGGTTGGTAGGATCGACGGCCCGTTCACAGGTCGCGGCGGGTTTTA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>9</th>\n",
              "      <td>SCO5087_loc_522</td>\n",
              "      <td>CGGTTGGTAGGATCGACGGCGTCGACCTCCCAGTCACGGCGTTTTA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>10</th>\n",
              "      <td>SCO5087_loc_600</td>\n",
              "      <td>CGGTTGGTAGGATCGACGGCGCCGACCGCCCAGGCGACCTGTTTTA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>11</th>\n",
              "      <td>SCO5087_loc_596</td>\n",
              "      <td>CGGTTGGTAGGATCGACGGCACCGCCCAGGCGACCTCGGCGTTTTA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>SCO5087_fwd_checking_primer</td>\n",
              "      <td>GACGATTCGGCCCGTG</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>SCO5088_fwd_checking_primer</td>\n",
              "      <td>CGCATCCACGCCGAGA</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>SCO5089_fwd_checking_primer</td>\n",
              "      <td>CGTACGGCGAACTCGC</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>SCO5087_rev_checking_primer</td>\n",
              "      <td>CAGGGCGTCCAGGC</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>SCO5088_rev_checking_primer</td>\n",
              "      <td>CAGCGTCCGGCGG</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>5</th>\n",
              "      <td>SCO5089_rev_checking_primer</td>\n",
              "      <td>GTCCCCGGACGGC</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "                           Name  \\\n",
              "0                SCO5088_loc_73   \n",
              "1                SCO5088_loc_67   \n",
              "2               SCO5089_loc_162   \n",
              "3               SCO5089_loc_167   \n",
              "4               SCO5087_loc_210   \n",
              "5               SCO5088_loc_216   \n",
              "6               SCO5088_loc_252   \n",
              "7               SCO5088_loc_262   \n",
              "8               SCO5087_loc_283   \n",
              "9               SCO5087_loc_522   \n",
              "10              SCO5087_loc_600   \n",
              "11              SCO5087_loc_596   \n",
              "0   SCO5087_fwd_checking_primer   \n",
              "1   SCO5088_fwd_checking_primer   \n",
              "2   SCO5089_fwd_checking_primer   \n",
              "3   SCO5087_rev_checking_primer   \n",
              "4   SCO5088_rev_checking_primer   \n",
              "5   SCO5089_rev_checking_primer   \n",
              "\n",
              "                                             Sequence Concentration  \\\n",
              "0   CGGTTGGTAGGATCGACGGCCCGCCGACCAGTACGGCGCCGTTTTA...          25nm   \n",
              "1   CGGTTGGTAGGATCGACGGCACCAGTACGGCGCCAGGCCGGTTTTA...          25nm   \n",
              "2   CGGTTGGTAGGATCGACGGCCCGCGGCGCGACTCGAGAGCGTTTTA...          25nm   \n",
              "3   CGGTTGGTAGGATCGACGGCGCGCGACTCGAGAGCCGGTAGTTTTA...          25nm   \n",
              "4   CGGTTGGTAGGATCGACGGCGAGCAGTTCCCAGAACTGCCGTTTTA...          25nm   \n",
              "5   CGGTTGGTAGGATCGACGGCCGCAGACGGACCCCTCCACAGTTTTA...          25nm   \n",
              "6   CGGTTGGTAGGATCGACGGCCTGGAGGGCCCAGTCCGCGGGTTTTA...          25nm   \n",
              "7   CGGTTGGTAGGATCGACGGCGGCCCTCCAGGACGCGAAGGGTTTTA...          25nm   \n",
              "8   CGGTTGGTAGGATCGACGGCCCGTTCACAGGTCGCGGCGGGTTTTA...          25nm   \n",
              "9   CGGTTGGTAGGATCGACGGCGTCGACCTCCCAGTCACGGCGTTTTA...          25nm   \n",
              "10  CGGTTGGTAGGATCGACGGCGCCGACCGCCCAGGCGACCTGTTTTA...          25nm   \n",
              "11  CGGTTGGTAGGATCGACGGCACCGCCCAGGCGACCTCGGCGTTTTA...          25nm   \n",
              "0                                    GACGATTCGGCCCGTG          25nm   \n",
              "1                                    CGCATCCACGCCGAGA          25nm   \n",
              "2                                    CGTACGGCGAACTCGC          25nm   \n",
              "3                                      CAGGGCGTCCAGGC          25nm   \n",
              "4                                       CAGCGTCCGGCGG          25nm   \n",
              "5                                       GTCCCCGGACGGC          25nm   \n",
              "\n",
              "   Purification  \n",
              "0           STD  \n",
              "1           STD  \n",
              "2           STD  \n",
              "3           STD  \n",
              "4           STD  \n",
              "5           STD  \n",
              "6           STD  \n",
              "7           STD  \n",
              "8           STD  \n",
              "9           STD  \n",
              "10          STD  \n",
              "11          STD  \n",
              "0           STD  \n",
              "1           STD  \n",
              "2           STD  \n",
              "3           STD  \n",
              "4           STD  \n",
              "5           STD  "
            ]
          },
          "execution_count": 13,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "full_idt = pd.concat([idt_df1, idt_df2])\n",
        "full_idt"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 15,
      "metadata": {},
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "Project structure for 'CRISPR_cBEST_workflow_2024-06-28T09:13:21.621756' created successfully.\n"
          ]
        },
        {
          "name": "stderr",
          "output_type": "stream",
          "text": [
            "/Users/lucaslevassor/opt/anaconda3/envs/streptocad/lib/python3.11/site-packages/Bio/SeqIO/InsdcIO.py:728: BiopythonWarning: Increasing length of locus line to allow long name. This will result in fields that are not in usual positions.\n",
            "  warnings.warn(\n"
          ]
        }
      ],
      "source": [
        "\n",
        "\n",
        "input_files = [\n",
        "    {\"name\": \"input_genome.gb\", \"content\": genome},\n",
        "    {\"name\": \"input_plasmid.gb\", \"content\": clean_plasmid}\n",
        "]\n",
        "\n",
        "output_files = [\n",
        "    {\"name\": \"cBEST_w_sgRNAs.gb\", \"content\": sgRNA_vectors}, # LIST OF Dseqrecords\n",
        "    {\"name\": \"primer_df.csv\", \"content\": checking_primers_df},\n",
        "    {\"name\": \"full_idt.csv\", \"content\": full_idt},\n",
        "    {\"name\": \"mutated_sgrna_df.csv\", \"content\": mutated_sgrna_df},\n",
        "    {\"name\": \"filtered_df.csv\", \"content\": filtered_df}\n",
        "]\n",
        "\n",
        "input_values = {\n",
        "    \"genes_to_knockout\": genes_to_KO,\n",
        "    \"polymerase_settings\": {\n",
        "        \"chosen_polymerase\": chosen_polymerase,\n",
        "        \"melting_temperature\": melting_temperature,\n",
        "        \"primer_concentration\": primer_concentration,\n",
        "        \"primer_number_increment\": primer_number_increment,\n",
        "        \"flanking_region_number\": flanking_region_number\n",
        "    },\n",
        "    \"filtering_metrics\": {\n",
        "        \"gc_upper\": gc_upper,\n",
        "        \"gc_lower\": gc_lower,\n",
        "        \"off_target_seed\": off_target_seed,\n",
        "        \"off_target_upper\": off_target_upper,\n",
        "        \"cas_type\": cas_type,\n",
        "        \"number_of_sgRNAs_per_group\": number_of_sgRNAs_per_group\n",
        "    },\n",
        "    \"overlapping_sequences\": {\n",
        "        \"up_homology\": str(up_homology),\n",
        "        \"dw_homology\": str(dw_homology)\n",
        "    }\n",
        "}\n",
        "\n",
        "\n",
        "# Paths to Markdown files\n",
        "markdown_file_paths = [\n",
        "    \"../../protocols/conjugation_protcol.md\",\n",
        "    \"../../protocols/single_target_crispr_plasmid_protcol.md\"\n",
        "\n",
        "]\n",
        "\n",
        "# Data and time\n",
        "timestamp = datetime.utcnow().isoformat()\n",
        "\n",
        "generate_project_directory_structure(\n",
        "    project_name=f\"CRISPR_cBEST_workflow_{timestamp}\",\n",
        "    input_files=input_files,\n",
        "    output_files=output_files,\n",
        "    input_values=input_values,\n",
        "    markdown_file_paths=markdown_file_paths\n",
        "\n",
        ")"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {},
      "outputs": [],
      "source": []
    }
  ],
  "metadata": {
    "kernelspec": {
      "display_name": "constrain",
      "language": "python",
      "name": "python3"
    },
    "language_info": {
      "codemirror_mode": {
        "name": "ipython",
        "version": 3
      },
      "file_extension": ".py",
      "mimetype": "text/x-python",
      "name": "python",
      "nbconvert_exporter": "python",
      "pygments_lexer": "ipython3",
      "version": "3.11.9"
    }
  },
  "nbformat": 4,
  "nbformat_minor": 2
}
