{
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "## W3-CRISPR-mcBEST workflow\n",
        "\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "The workflow allows for the simultaneous targeted engineering of up to 28 individual situated bases in the genome of Streptomyces, as shown by Whitford et. al 2023, thereby enabling efficient single-step multi-target engineering13. Similar to workflow 2, this workflow employs the CRISPR-BEST methodology but instead of performing base-editing on a single gene the method incorporates multiple guide RNAs as an array into the same vector (Figure 4A). This significantly enhances the efficiency of high-throughput genetic studies. The pCRISPR-cBEST plasmid file can be downloaded on the page along with the genome file for S.coelicor (A3) and streptoAIM will generate your plasmids, primers, checking primers and analyze the overhangs the target genes(Figure 4, S6). \n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 1,
      "metadata": {},
      "outputs": [],
      "source": [
        "import sys\n",
        "import os\n",
        "from pydna.dseqrecord import Dseqrecord\n",
        "from teemi.build.PCR import primer_tm_neb\n",
        "import pandas as pd\n",
        "from datetime import datetime\n",
        "\n",
        "\n",
        "# Ensure the src directory is in the Python path\n",
        "project_root = os.path.abspath(os.path.join(os.getcwd(), '../../'))\n",
        "if project_root not in sys.path:\n",
        "    sys.path.insert(0, project_root)\n",
        "\n",
        "from streptocad.sequence_loading.sequence_loading import (\n",
        "    load_and_process_gene_sequences,\n",
        "    load_and_process_plasmid)\n",
        "\n",
        "\n",
        "from streptocad.utils import polymerase_dict, dataframe_to_seqrecords, generate_project_directory_structure\n",
        "from streptocad.crispr.guideRNAcas3_9_12 import extract_sgRNAs, SgRNAargs\n",
        "from streptocad.crispr.crispr_best import (\n",
        "    identify_base_editing_sites,\n",
        "    filter_sgrnas_for_base_editing,\n",
        "    process_base_editing\n",
        ")\n",
        "\n",
        "from streptocad.cloning.golden_gate_cloning import (\n",
        "    GoldenGateCloning,\n",
        "    create_overhang_dataframe,\n",
        "    digest_amplicons_w_BsaI\n",
        ")\n",
        "\n",
        "from streptocad.primers.primer_analysis import analyze_primers_and_hairpins\n",
        "from streptocad.primers.primer_generation import create_idt_order_dataframe, primers_to_IDT\n",
        "from streptocad.wet_lab.gel_simulation import simulate_gel_electrophoresis\n",
        "from streptocad.cloning.plasmid_processing import annotate_plasmid_with_sgrnas\n",
        "from streptocad.primers.primer_generation import create_dseqrecords_from_df,checking_primers\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "# Inputs\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 2,
      "metadata": {},
      "outputs": [],
      "source": [
        "# 1 Add genome of choice (genbank, fasta)\n",
        "path_to_genome = '../../data/genomes/Streptomyces_coelicolor_A3_chromosome.gb'\n",
        "genome = load_and_process_plasmid(path_to_genome)\n",
        "\n",
        "# 2 Add plasmid \n",
        "path_to_plasmid = '../../data/plasmids/pCRISPR-MCBE_Csy4_kasopGFP.gb'\n",
        "clean_plasmid = load_and_process_plasmid(path_to_plasmid)\n",
        "\n",
        "# 3 Choose genes to knock out (list)\n",
        "genes_to_KO = ['SCO5087', 'SCO5089', 'SCO5090']\n",
        "\n",
        "#### Advanced settings ####\n",
        "# 4 Filtering metrics for sgRNAs\n",
        "gc_upper = 0.72\n",
        "gc_lower = 0.2\n",
        "off_target_seed = 13\n",
        "off_target_upper = 10\n",
        "cas_type='cas9'\n",
        "number_of_sgRNAs_per_group = 5\n",
        "\n",
        "# 5 Choose polymerase and target melting temperature\n",
        "chosen_polymerase = polymerase_dict['Phusion High-Fidelity DNA Polymerase (GC Buffer)']\n",
        "melting_temperature = 65\n",
        "primer_concentration = 0.4 \n",
        "primer_number_increment = 23\n",
        "flanking_region=500\n",
        "\n",
        "\n",
        "# 6 Choose overlapping sequences for our plasmid we can use the following\n",
        "restriction_overhang_f = \"GATCGggtctcc\"\n",
        "restriction_overhang_r = \"GATCAGGTCTCg\"\n",
        "backbone_overhang_f = \"cATG\"\n",
        "backbone_overhang_r = \"cTAG\"\n",
        "cys4 = \"gTTCACTGCCGTATAGGCAGCTAAGAAA\"\n",
        "sgRNA_handle_cys4_site = [Dseqrecord('GTTTTAGAGCTAGAAATAGCAAGTTAAAATAAGGCTAGTCCGTTATCAACTTGAAAAAGTGGCACCGAGTCGGTGCTTTTTT', name = 'sgRNA_handle_cys4')]"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "# Computation"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 3,
      "metadata": {},
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "sgRNA generated were outside the designated border in SCO5087. To incorporate this extent borders. Skipping to next locus tag.\n",
            "sgRNA generated were outside the designated border in SCO5087. To incorporate this extent borders. Skipping to next locus tag.\n",
            "sgRNA generated were outside the designated border in SCO5087. To incorporate this extent borders. Skipping to next locus tag.\n",
            "sgRNA generated were outside the designated border in SCO5089. To incorporate this extent borders. Skipping to next locus tag.\n",
            "sgRNA generated were outside the designated border in SCO5089. To incorporate this extent borders. Skipping to next locus tag.\n",
            "sgRNA generated were outside the designated border in SCO5090. To incorporate this extent borders. Skipping to next locus tag.\n",
            "sgRNA generated were outside the designated border in SCO5090. To incorporate this extent borders. Skipping to next locus tag.\n",
            "sgRNA generated were outside the designated border in SCO5090. To incorporate this extent borders. Skipping to next locus tag.\n",
            "sgRNA generated were outside the designated border in SCO5090. To incorporate this extent borders. Skipping to next locus tag.\n",
            "sgRNA generated were outside the designated border in SCO5090. To incorporate this extent borders. Skipping to next locus tag.\n"
          ]
        },
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>strain_name</th>\n",
              "      <th>locus_tag</th>\n",
              "      <th>gene_loc</th>\n",
              "      <th>gene_strand</th>\n",
              "      <th>sgrna_strand</th>\n",
              "      <th>sgrna_loc</th>\n",
              "      <th>gc</th>\n",
              "      <th>pam</th>\n",
              "      <th>sgrna</th>\n",
              "      <th>sgrna_seed_sequence</th>\n",
              "      <th>off_target_count</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>336</th>\n",
              "      <td>Streptomyces_coelicolor_A3_chromosome</td>\n",
              "      <td>SCO5090</td>\n",
              "      <td>5532706</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>32</td>\n",
              "      <td>0.65</td>\n",
              "      <td>CGG</td>\n",
              "      <td>TGGACCACCCGGTGTTCTCC</td>\n",
              "      <td>CCCGGTGTTCTCC</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>332</th>\n",
              "      <td>Streptomyces_coelicolor_A3_chromosome</td>\n",
              "      <td>SCO5089</td>\n",
              "      <td>5532449</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>61</td>\n",
              "      <td>0.70</td>\n",
              "      <td>CGG</td>\n",
              "      <td>CGTGGAGTGCGCCGGTGAGA</td>\n",
              "      <td>TGCGCCGGTGAGA</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>331</th>\n",
              "      <td>Streptomyces_coelicolor_A3_chromosome</td>\n",
              "      <td>SCO5089</td>\n",
              "      <td>5532449</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>65</td>\n",
              "      <td>0.70</td>\n",
              "      <td>CGG</td>\n",
              "      <td>GAGTGCGCCGGTGAGACGGA</td>\n",
              "      <td>CCGGTGAGACGGA</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>330</th>\n",
              "      <td>Streptomyces_coelicolor_A3_chromosome</td>\n",
              "      <td>SCO5089</td>\n",
              "      <td>5532449</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>66</td>\n",
              "      <td>0.70</td>\n",
              "      <td>GGG</td>\n",
              "      <td>AGTGCGCCGGTGAGACGGAC</td>\n",
              "      <td>CGGTGAGACGGAC</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>328</th>\n",
              "      <td>Streptomyces_coelicolor_A3_chromosome</td>\n",
              "      <td>SCO5089</td>\n",
              "      <td>5532449</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>80</td>\n",
              "      <td>0.70</td>\n",
              "      <td>CGG</td>\n",
              "      <td>ACGGACGGGACGGACCTGTC</td>\n",
              "      <td>GGACGGACCTGTC</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>...</th>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>406</th>\n",
              "      <td>Streptomyces_coelicolor_A3_chromosome</td>\n",
              "      <td>SCO5090</td>\n",
              "      <td>5532706</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>691</td>\n",
              "      <td>0.70</td>\n",
              "      <td>TGG</td>\n",
              "      <td>GCCGGATCGAACGGGTGGTG</td>\n",
              "      <td>CGAACGGGTGGTG</td>\n",
              "      <td>2</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>78</th>\n",
              "      <td>Streptomyces_coelicolor_A3_chromosome</td>\n",
              "      <td>SCO5087</td>\n",
              "      <td>5529801</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>745</td>\n",
              "      <td>0.65</td>\n",
              "      <td>GGG</td>\n",
              "      <td>CGACGACGATCGGGGTGATC</td>\n",
              "      <td>GATCGGGGTGATC</td>\n",
              "      <td>2</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>452</th>\n",
              "      <td>Streptomyces_coelicolor_A3_chromosome</td>\n",
              "      <td>SCO5090</td>\n",
              "      <td>5532706</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>890</td>\n",
              "      <td>0.65</td>\n",
              "      <td>CGG</td>\n",
              "      <td>ACCCATCTGCGCGAGGTGCT</td>\n",
              "      <td>TGCGCGAGGTGCT</td>\n",
              "      <td>3</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>192</th>\n",
              "      <td>Streptomyces_coelicolor_A3_chromosome</td>\n",
              "      <td>SCO5087</td>\n",
              "      <td>5529801</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>838</td>\n",
              "      <td>0.70</td>\n",
              "      <td>TGG</td>\n",
              "      <td>TACGCGGGACGGCTTCGTGC</td>\n",
              "      <td>GACGGCTTCGTGC</td>\n",
              "      <td>3</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>201</th>\n",
              "      <td>Streptomyces_coelicolor_A3_chromosome</td>\n",
              "      <td>SCO5087</td>\n",
              "      <td>5529801</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>706</td>\n",
              "      <td>0.65</td>\n",
              "      <td>CGG</td>\n",
              "      <td>CAGCGCCGACGTGATGTTCG</td>\n",
              "      <td>GACGTGATGTTCG</td>\n",
              "      <td>3</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>279 rows × 11 columns</p>\n",
              "</div>"
            ],
            "text/plain": [
              "                               strain_name locus_tag  gene_loc  gene_strand  \\\n",
              "336  Streptomyces_coelicolor_A3_chromosome   SCO5090   5532706            1   \n",
              "332  Streptomyces_coelicolor_A3_chromosome   SCO5089   5532449            1   \n",
              "331  Streptomyces_coelicolor_A3_chromosome   SCO5089   5532449            1   \n",
              "330  Streptomyces_coelicolor_A3_chromosome   SCO5089   5532449            1   \n",
              "328  Streptomyces_coelicolor_A3_chromosome   SCO5089   5532449            1   \n",
              "..                                     ...       ...       ...          ...   \n",
              "406  Streptomyces_coelicolor_A3_chromosome   SCO5090   5532706            1   \n",
              "78   Streptomyces_coelicolor_A3_chromosome   SCO5087   5529801            1   \n",
              "452  Streptomyces_coelicolor_A3_chromosome   SCO5090   5532706            1   \n",
              "192  Streptomyces_coelicolor_A3_chromosome   SCO5087   5529801            1   \n",
              "201  Streptomyces_coelicolor_A3_chromosome   SCO5087   5529801            1   \n",
              "\n",
              "     sgrna_strand  sgrna_loc    gc  pam                 sgrna  \\\n",
              "336            -1         32  0.65  CGG  TGGACCACCCGGTGTTCTCC   \n",
              "332             1         61  0.70  CGG  CGTGGAGTGCGCCGGTGAGA   \n",
              "331             1         65  0.70  CGG  GAGTGCGCCGGTGAGACGGA   \n",
              "330             1         66  0.70  GGG  AGTGCGCCGGTGAGACGGAC   \n",
              "328             1         80  0.70  CGG  ACGGACGGGACGGACCTGTC   \n",
              "..            ...        ...   ...  ...                   ...   \n",
              "406            -1        691  0.70  TGG  GCCGGATCGAACGGGTGGTG   \n",
              "78             -1        745  0.65  GGG  CGACGACGATCGGGGTGATC   \n",
              "452             1        890  0.65  CGG  ACCCATCTGCGCGAGGTGCT   \n",
              "192             1        838  0.70  TGG  TACGCGGGACGGCTTCGTGC   \n",
              "201             1        706  0.65  CGG  CAGCGCCGACGTGATGTTCG   \n",
              "\n",
              "    sgrna_seed_sequence  off_target_count  \n",
              "336       CCCGGTGTTCTCC                 0  \n",
              "332       TGCGCCGGTGAGA                 0  \n",
              "331       CCGGTGAGACGGA                 0  \n",
              "330       CGGTGAGACGGAC                 0  \n",
              "328       GGACGGACCTGTC                 0  \n",
              "..                  ...               ...  \n",
              "406       CGAACGGGTGGTG                 2  \n",
              "78        GATCGGGGTGATC                 2  \n",
              "452       TGCGCGAGGTGCT                 3  \n",
              "192       GACGGCTTCGTGC                 3  \n",
              "201       GACGTGATGTTCG                 3  \n",
              "\n",
              "[279 rows x 11 columns]"
            ]
          },
          "execution_count": 3,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "\n",
        "# Initialize SgRNAargs with desired parameters\n",
        "args = SgRNAargs(path_to_genome, \n",
        "                genes_to_KO,\n",
        "                step=['find', 'filter'],\n",
        "                gc_upper = gc_upper,\n",
        "                gc_lower = gc_lower,\n",
        "                off_target_seed = off_target_seed,\n",
        "                off_target_upper = off_target_upper,\n",
        "                cas_type='cas9'\n",
        "                )\n",
        "\n",
        "sgrna_df = extract_sgRNAs(args)\n",
        "sgrna_df"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 4,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>strain_name</th>\n",
              "      <th>locus_tag</th>\n",
              "      <th>gene_loc</th>\n",
              "      <th>gene_strand</th>\n",
              "      <th>sgrna_strand</th>\n",
              "      <th>sgrna_loc</th>\n",
              "      <th>gc</th>\n",
              "      <th>pam</th>\n",
              "      <th>sgrna</th>\n",
              "      <th>sgrna_seed_sequence</th>\n",
              "      <th>off_target_count</th>\n",
              "      <th>editable_cytosines</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>336</th>\n",
              "      <td>Streptomyces_coelicolor_A3_chromosome</td>\n",
              "      <td>SCO5090</td>\n",
              "      <td>5532706</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>32</td>\n",
              "      <td>0.65</td>\n",
              "      <td>CGG</td>\n",
              "      <td>TGGACCACCCGGTGTTCTCC</td>\n",
              "      <td>CCCGGTGTTCTCC</td>\n",
              "      <td>0</td>\n",
              "      <td>5,6,8,9,10</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>332</th>\n",
              "      <td>Streptomyces_coelicolor_A3_chromosome</td>\n",
              "      <td>SCO5089</td>\n",
              "      <td>5532449</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>61</td>\n",
              "      <td>0.70</td>\n",
              "      <td>CGG</td>\n",
              "      <td>CGTGGAGTGCGCCGGTGAGA</td>\n",
              "      <td>TGCGCCGGTGAGA</td>\n",
              "      <td>0</td>\n",
              "      <td>10</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>331</th>\n",
              "      <td>Streptomyces_coelicolor_A3_chromosome</td>\n",
              "      <td>SCO5089</td>\n",
              "      <td>5532449</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>65</td>\n",
              "      <td>0.70</td>\n",
              "      <td>CGG</td>\n",
              "      <td>GAGTGCGCCGGTGAGACGGA</td>\n",
              "      <td>CCGGTGAGACGGA</td>\n",
              "      <td>0</td>\n",
              "      <td>6,8,9</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>330</th>\n",
              "      <td>Streptomyces_coelicolor_A3_chromosome</td>\n",
              "      <td>SCO5089</td>\n",
              "      <td>5532449</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>66</td>\n",
              "      <td>0.70</td>\n",
              "      <td>GGG</td>\n",
              "      <td>AGTGCGCCGGTGAGACGGAC</td>\n",
              "      <td>CGGTGAGACGGAC</td>\n",
              "      <td>0</td>\n",
              "      <td>5,7,8</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>328</th>\n",
              "      <td>Streptomyces_coelicolor_A3_chromosome</td>\n",
              "      <td>SCO5089</td>\n",
              "      <td>5532449</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>80</td>\n",
              "      <td>0.70</td>\n",
              "      <td>CGG</td>\n",
              "      <td>ACGGACGGGACGGACCTGTC</td>\n",
              "      <td>GGACGGACCTGTC</td>\n",
              "      <td>0</td>\n",
              "      <td>6</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>...</th>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>406</th>\n",
              "      <td>Streptomyces_coelicolor_A3_chromosome</td>\n",
              "      <td>SCO5090</td>\n",
              "      <td>5532706</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>691</td>\n",
              "      <td>0.70</td>\n",
              "      <td>TGG</td>\n",
              "      <td>GCCGGATCGAACGGGTGGTG</td>\n",
              "      <td>CGAACGGGTGGTG</td>\n",
              "      <td>2</td>\n",
              "      <td>3,8</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>78</th>\n",
              "      <td>Streptomyces_coelicolor_A3_chromosome</td>\n",
              "      <td>SCO5087</td>\n",
              "      <td>5529801</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>745</td>\n",
              "      <td>0.65</td>\n",
              "      <td>GGG</td>\n",
              "      <td>CGACGACGATCGGGGTGATC</td>\n",
              "      <td>GATCGGGGTGATC</td>\n",
              "      <td>2</td>\n",
              "      <td>4,7</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>452</th>\n",
              "      <td>Streptomyces_coelicolor_A3_chromosome</td>\n",
              "      <td>SCO5090</td>\n",
              "      <td>5532706</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>890</td>\n",
              "      <td>0.65</td>\n",
              "      <td>CGG</td>\n",
              "      <td>ACCCATCTGCGCGAGGTGCT</td>\n",
              "      <td>TGCGCGAGGTGCT</td>\n",
              "      <td>3</td>\n",
              "      <td>3,4,7,10</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>192</th>\n",
              "      <td>Streptomyces_coelicolor_A3_chromosome</td>\n",
              "      <td>SCO5087</td>\n",
              "      <td>5529801</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>838</td>\n",
              "      <td>0.70</td>\n",
              "      <td>TGG</td>\n",
              "      <td>TACGCGGGACGGCTTCGTGC</td>\n",
              "      <td>GACGGCTTCGTGC</td>\n",
              "      <td>3</td>\n",
              "      <td>3,5,10</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>201</th>\n",
              "      <td>Streptomyces_coelicolor_A3_chromosome</td>\n",
              "      <td>SCO5087</td>\n",
              "      <td>5529801</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>706</td>\n",
              "      <td>0.65</td>\n",
              "      <td>CGG</td>\n",
              "      <td>CAGCGCCGACGTGATGTTCG</td>\n",
              "      <td>GACGTGATGTTCG</td>\n",
              "      <td>3</td>\n",
              "      <td>4,6,7,10</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>272 rows × 12 columns</p>\n",
              "</div>"
            ],
            "text/plain": [
              "                               strain_name locus_tag  gene_loc  gene_strand  \\\n",
              "336  Streptomyces_coelicolor_A3_chromosome   SCO5090   5532706            1   \n",
              "332  Streptomyces_coelicolor_A3_chromosome   SCO5089   5532449            1   \n",
              "331  Streptomyces_coelicolor_A3_chromosome   SCO5089   5532449            1   \n",
              "330  Streptomyces_coelicolor_A3_chromosome   SCO5089   5532449            1   \n",
              "328  Streptomyces_coelicolor_A3_chromosome   SCO5089   5532449            1   \n",
              "..                                     ...       ...       ...          ...   \n",
              "406  Streptomyces_coelicolor_A3_chromosome   SCO5090   5532706            1   \n",
              "78   Streptomyces_coelicolor_A3_chromosome   SCO5087   5529801            1   \n",
              "452  Streptomyces_coelicolor_A3_chromosome   SCO5090   5532706            1   \n",
              "192  Streptomyces_coelicolor_A3_chromosome   SCO5087   5529801            1   \n",
              "201  Streptomyces_coelicolor_A3_chromosome   SCO5087   5529801            1   \n",
              "\n",
              "     sgrna_strand  sgrna_loc    gc  pam                 sgrna  \\\n",
              "336            -1         32  0.65  CGG  TGGACCACCCGGTGTTCTCC   \n",
              "332             1         61  0.70  CGG  CGTGGAGTGCGCCGGTGAGA   \n",
              "331             1         65  0.70  CGG  GAGTGCGCCGGTGAGACGGA   \n",
              "330             1         66  0.70  GGG  AGTGCGCCGGTGAGACGGAC   \n",
              "328             1         80  0.70  CGG  ACGGACGGGACGGACCTGTC   \n",
              "..            ...        ...   ...  ...                   ...   \n",
              "406            -1        691  0.70  TGG  GCCGGATCGAACGGGTGGTG   \n",
              "78             -1        745  0.65  GGG  CGACGACGATCGGGGTGATC   \n",
              "452             1        890  0.65  CGG  ACCCATCTGCGCGAGGTGCT   \n",
              "192             1        838  0.70  TGG  TACGCGGGACGGCTTCGTGC   \n",
              "201             1        706  0.65  CGG  CAGCGCCGACGTGATGTTCG   \n",
              "\n",
              "    sgrna_seed_sequence  off_target_count editable_cytosines  \n",
              "336       CCCGGTGTTCTCC                 0         5,6,8,9,10  \n",
              "332       TGCGCCGGTGAGA                 0                 10  \n",
              "331       CCGGTGAGACGGA                 0              6,8,9  \n",
              "330       CGGTGAGACGGAC                 0              5,7,8  \n",
              "328       GGACGGACCTGTC                 0                  6  \n",
              "..                  ...               ...                ...  \n",
              "406       CGAACGGGTGGTG                 2                3,8  \n",
              "78        GATCGGGGTGATC                 2                4,7  \n",
              "452       TGCGCGAGGTGCT                 3           3,4,7,10  \n",
              "192       GACGGCTTCGTGC                 3             3,5,10  \n",
              "201       GACGTGATGTTCG                 3           4,6,7,10  \n",
              "\n",
              "[272 rows x 12 columns]"
            ]
          },
          "execution_count": 4,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "\n",
        "# Example usage\n",
        "path_to_genome = '../../data/genomes/Streptomyces_coelicolor_A3_chromosome.gb'\n",
        "\n",
        "# Load gene sequences\n",
        "gene_sequences = load_and_process_gene_sequences(path_to_genome)\n",
        "genes_to_KO_dict = {locus_tag: gene_sequences[locus_tag] for locus_tag in genes_to_KO if locus_tag in gene_sequences}\n",
        "\n",
        "# Identify and annotate base editing sites\n",
        "sgrna_df_with_editing = identify_base_editing_sites(sgrna_df)\n",
        "\n",
        "filtered_sgrna_df_for_base_editing = filter_sgrnas_for_base_editing(sgrna_df_with_editing)\n",
        "filtered_sgrna_df_for_base_editing"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 5,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>strain_name</th>\n",
              "      <th>locus_tag</th>\n",
              "      <th>gene_loc</th>\n",
              "      <th>gene_strand</th>\n",
              "      <th>sgrna_strand</th>\n",
              "      <th>sgrna_loc</th>\n",
              "      <th>gc</th>\n",
              "      <th>pam</th>\n",
              "      <th>sgrna</th>\n",
              "      <th>sgrna_seed_sequence</th>\n",
              "      <th>off_target_count</th>\n",
              "      <th>editable_cytosines</th>\n",
              "      <th>mutations</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>336</th>\n",
              "      <td>Streptomyces_coelicolor_A3_chromosome</td>\n",
              "      <td>SCO5090</td>\n",
              "      <td>5532706</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>32</td>\n",
              "      <td>0.65</td>\n",
              "      <td>CGG</td>\n",
              "      <td>TGGACCACCCGGTGTTCTCC</td>\n",
              "      <td>CCCGGTGTTCTCC</td>\n",
              "      <td>0</td>\n",
              "      <td>5,6,8,9,10</td>\n",
              "      <td>R8Q, V9I, V10I</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>331</th>\n",
              "      <td>Streptomyces_coelicolor_A3_chromosome</td>\n",
              "      <td>SCO5089</td>\n",
              "      <td>5532449</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>65</td>\n",
              "      <td>0.70</td>\n",
              "      <td>CGG</td>\n",
              "      <td>GAGTGCGCCGGTGAGACGGA</td>\n",
              "      <td>CCGGTGAGACGGA</td>\n",
              "      <td>0</td>\n",
              "      <td>6,8,9</td>\n",
              "      <td>A18V</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>330</th>\n",
              "      <td>Streptomyces_coelicolor_A3_chromosome</td>\n",
              "      <td>SCO5089</td>\n",
              "      <td>5532449</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>66</td>\n",
              "      <td>0.70</td>\n",
              "      <td>GGG</td>\n",
              "      <td>AGTGCGCCGGTGAGACGGAC</td>\n",
              "      <td>CGGTGAGACGGAC</td>\n",
              "      <td>0</td>\n",
              "      <td>5,7,8</td>\n",
              "      <td>A18V</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>327</th>\n",
              "      <td>Streptomyces_coelicolor_A3_chromosome</td>\n",
              "      <td>SCO5089</td>\n",
              "      <td>5532449</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>106</td>\n",
              "      <td>0.65</td>\n",
              "      <td>AGG</td>\n",
              "      <td>CTTCCTCGACCTCCGCTTCG</td>\n",
              "      <td>GACCTCCGCTTCG</td>\n",
              "      <td>0</td>\n",
              "      <td>4,5,7,10</td>\n",
              "      <td>L31F</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>326</th>\n",
              "      <td>Streptomyces_coelicolor_A3_chromosome</td>\n",
              "      <td>SCO5089</td>\n",
              "      <td>5532449</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>113</td>\n",
              "      <td>0.60</td>\n",
              "      <td>CGG</td>\n",
              "      <td>GACCTCCGCTTCGAGGACAT</td>\n",
              "      <td>GCTTCGAGGACAT</td>\n",
              "      <td>0</td>\n",
              "      <td>3,4,6,7,9</td>\n",
              "      <td>L33F, R34C</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>...</th>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>406</th>\n",
              "      <td>Streptomyces_coelicolor_A3_chromosome</td>\n",
              "      <td>SCO5090</td>\n",
              "      <td>5532706</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>691</td>\n",
              "      <td>0.70</td>\n",
              "      <td>TGG</td>\n",
              "      <td>GCCGGATCGAACGGGTGGTG</td>\n",
              "      <td>CGAACGGGTGGTG</td>\n",
              "      <td>2</td>\n",
              "      <td>3,8</td>\n",
              "      <td>R230Q</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>78</th>\n",
              "      <td>Streptomyces_coelicolor_A3_chromosome</td>\n",
              "      <td>SCO5087</td>\n",
              "      <td>5529801</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>745</td>\n",
              "      <td>0.65</td>\n",
              "      <td>GGG</td>\n",
              "      <td>CGACGACGATCGGGGTGATC</td>\n",
              "      <td>GATCGGGGTGATC</td>\n",
              "      <td>2</td>\n",
              "      <td>4,7</td>\n",
              "      <td>V247I, V248I</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>452</th>\n",
              "      <td>Streptomyces_coelicolor_A3_chromosome</td>\n",
              "      <td>SCO5090</td>\n",
              "      <td>5532706</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>890</td>\n",
              "      <td>0.65</td>\n",
              "      <td>CGG</td>\n",
              "      <td>ACCCATCTGCGCGAGGTGCT</td>\n",
              "      <td>TGCGCGAGGTGCT</td>\n",
              "      <td>3</td>\n",
              "      <td>3,4,7,10</td>\n",
              "      <td>H292Y, R294C</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>192</th>\n",
              "      <td>Streptomyces_coelicolor_A3_chromosome</td>\n",
              "      <td>SCO5087</td>\n",
              "      <td>5529801</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>838</td>\n",
              "      <td>0.70</td>\n",
              "      <td>TGG</td>\n",
              "      <td>TACGCGGGACGGCTTCGTGC</td>\n",
              "      <td>GACGGCTTCGTGC</td>\n",
              "      <td>3</td>\n",
              "      <td>3,5,10</td>\n",
              "      <td>T274M, R275W</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>201</th>\n",
              "      <td>Streptomyces_coelicolor_A3_chromosome</td>\n",
              "      <td>SCO5087</td>\n",
              "      <td>5529801</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>706</td>\n",
              "      <td>0.65</td>\n",
              "      <td>CGG</td>\n",
              "      <td>CAGCGCCGACGTGATGTTCG</td>\n",
              "      <td>GACGTGATGTTCG</td>\n",
              "      <td>3</td>\n",
              "      <td>4,6,7,10</td>\n",
              "      <td>A231V</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>241 rows × 13 columns</p>\n",
              "</div>"
            ],
            "text/plain": [
              "                               strain_name locus_tag  gene_loc  gene_strand  \\\n",
              "336  Streptomyces_coelicolor_A3_chromosome   SCO5090   5532706            1   \n",
              "331  Streptomyces_coelicolor_A3_chromosome   SCO5089   5532449            1   \n",
              "330  Streptomyces_coelicolor_A3_chromosome   SCO5089   5532449            1   \n",
              "327  Streptomyces_coelicolor_A3_chromosome   SCO5089   5532449            1   \n",
              "326  Streptomyces_coelicolor_A3_chromosome   SCO5089   5532449            1   \n",
              "..                                     ...       ...       ...          ...   \n",
              "406  Streptomyces_coelicolor_A3_chromosome   SCO5090   5532706            1   \n",
              "78   Streptomyces_coelicolor_A3_chromosome   SCO5087   5529801            1   \n",
              "452  Streptomyces_coelicolor_A3_chromosome   SCO5090   5532706            1   \n",
              "192  Streptomyces_coelicolor_A3_chromosome   SCO5087   5529801            1   \n",
              "201  Streptomyces_coelicolor_A3_chromosome   SCO5087   5529801            1   \n",
              "\n",
              "     sgrna_strand  sgrna_loc    gc  pam                 sgrna  \\\n",
              "336            -1         32  0.65  CGG  TGGACCACCCGGTGTTCTCC   \n",
              "331             1         65  0.70  CGG  GAGTGCGCCGGTGAGACGGA   \n",
              "330             1         66  0.70  GGG  AGTGCGCCGGTGAGACGGAC   \n",
              "327             1        106  0.65  AGG  CTTCCTCGACCTCCGCTTCG   \n",
              "326             1        113  0.60  CGG  GACCTCCGCTTCGAGGACAT   \n",
              "..            ...        ...   ...  ...                   ...   \n",
              "406            -1        691  0.70  TGG  GCCGGATCGAACGGGTGGTG   \n",
              "78             -1        745  0.65  GGG  CGACGACGATCGGGGTGATC   \n",
              "452             1        890  0.65  CGG  ACCCATCTGCGCGAGGTGCT   \n",
              "192             1        838  0.70  TGG  TACGCGGGACGGCTTCGTGC   \n",
              "201             1        706  0.65  CGG  CAGCGCCGACGTGATGTTCG   \n",
              "\n",
              "    sgrna_seed_sequence  off_target_count editable_cytosines       mutations  \n",
              "336       CCCGGTGTTCTCC                 0         5,6,8,9,10  R8Q, V9I, V10I  \n",
              "331       CCGGTGAGACGGA                 0              6,8,9            A18V  \n",
              "330       CGGTGAGACGGAC                 0              5,7,8            A18V  \n",
              "327       GACCTCCGCTTCG                 0           4,5,7,10            L31F  \n",
              "326       GCTTCGAGGACAT                 0          3,4,6,7,9      L33F, R34C  \n",
              "..                  ...               ...                ...             ...  \n",
              "406       CGAACGGGTGGTG                 2                3,8           R230Q  \n",
              "78        GATCGGGGTGATC                 2                4,7    V247I, V248I  \n",
              "452       TGCGCGAGGTGCT                 3           3,4,7,10    H292Y, R294C  \n",
              "192       GACGGCTTCGTGC                 3             3,5,10    T274M, R275W  \n",
              "201       GACGTGATGTTCG                 3           4,6,7,10           A231V  \n",
              "\n",
              "[241 rows x 13 columns]"
            ]
          },
          "execution_count": 5,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "\n",
        "# Process the DataFrame to apply C-to-T mutations\n",
        "mutated_sgrna_df = process_base_editing(filtered_sgrna_df_for_base_editing, genes_to_KO_dict, only_stop_codons = False)\n",
        "mutated_sgrna_df"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 6,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>strain_name</th>\n",
              "      <th>locus_tag</th>\n",
              "      <th>gene_loc</th>\n",
              "      <th>gene_strand</th>\n",
              "      <th>sgrna_strand</th>\n",
              "      <th>sgrna_loc</th>\n",
              "      <th>gc</th>\n",
              "      <th>pam</th>\n",
              "      <th>sgrna</th>\n",
              "      <th>sgrna_seed_sequence</th>\n",
              "      <th>off_target_count</th>\n",
              "      <th>editable_cytosines</th>\n",
              "      <th>mutations</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>537</th>\n",
              "      <td>Streptomyces_coelicolor_A3_chromosome</td>\n",
              "      <td>SCO5090</td>\n",
              "      <td>5532706</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>140</td>\n",
              "      <td>0.65</td>\n",
              "      <td>CGG</td>\n",
              "      <td>CACGTACAGGTCCTGGAGCG</td>\n",
              "      <td>AGGTCCTGGAGCG</td>\n",
              "      <td>2</td>\n",
              "      <td>3,7</td>\n",
              "      <td>Q43*</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>321</th>\n",
              "      <td>Streptomyces_coelicolor_A3_chromosome</td>\n",
              "      <td>SCO5089</td>\n",
              "      <td>5532449</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>167</td>\n",
              "      <td>0.70</td>\n",
              "      <td>CGG</td>\n",
              "      <td>GCGCGACTCGAGAGCCGGTA</td>\n",
              "      <td>TCGAGAGCCGGTA</td>\n",
              "      <td>0</td>\n",
              "      <td>4,7,9</td>\n",
              "      <td>R51*, L52F</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>354</th>\n",
              "      <td>Streptomyces_coelicolor_A3_chromosome</td>\n",
              "      <td>SCO5090</td>\n",
              "      <td>5532706</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>174</td>\n",
              "      <td>0.65</td>\n",
              "      <td>CGG</td>\n",
              "      <td>CGCCCAGATCCGGAAGCGTT</td>\n",
              "      <td>ATCCGGAAGCGTT</td>\n",
              "      <td>0</td>\n",
              "      <td>3,4,5,10</td>\n",
              "      <td>W57*, A58T</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>356</th>\n",
              "      <td>Streptomyces_coelicolor_A3_chromosome</td>\n",
              "      <td>SCO5090</td>\n",
              "      <td>5532706</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>208</td>\n",
              "      <td>0.65</td>\n",
              "      <td>CGG</td>\n",
              "      <td>GGGACGTCCAGGTCTTCACC</td>\n",
              "      <td>CCAGGTCTTCACC</td>\n",
              "      <td>0</td>\n",
              "      <td>5,8,9</td>\n",
              "      <td>W67*</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>17</th>\n",
              "      <td>Streptomyces_coelicolor_A3_chromosome</td>\n",
              "      <td>SCO5087</td>\n",
              "      <td>5529801</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>210</td>\n",
              "      <td>0.60</td>\n",
              "      <td>GGG</td>\n",
              "      <td>GAGCAGTTCCCAGAACTGCC</td>\n",
              "      <td>TCCCAGAACTGCC</td>\n",
              "      <td>0</td>\n",
              "      <td>4,9,10</td>\n",
              "      <td>W67*, E68K</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>54</th>\n",
              "      <td>Streptomyces_coelicolor_A3_chromosome</td>\n",
              "      <td>SCO5087</td>\n",
              "      <td>5529801</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>522</td>\n",
              "      <td>0.70</td>\n",
              "      <td>CGG</td>\n",
              "      <td>GTCGACCTCCCAGTCACGGC</td>\n",
              "      <td>TCCCAGTCACGGC</td>\n",
              "      <td>0</td>\n",
              "      <td>3,6,7,9,10</td>\n",
              "      <td>W171*, E172K, V173I, D174N</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>474</th>\n",
              "      <td>Streptomyces_coelicolor_A3_chromosome</td>\n",
              "      <td>SCO5090</td>\n",
              "      <td>5532706</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>640</td>\n",
              "      <td>0.60</td>\n",
              "      <td>TGG</td>\n",
              "      <td>GACCGTCCAGGACATGACCA</td>\n",
              "      <td>CAGGACATGACCA</td>\n",
              "      <td>0</td>\n",
              "      <td>3,4,7,8</td>\n",
              "      <td>T208I, Q210*</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "                               strain_name locus_tag  gene_loc  gene_strand  \\\n",
              "537  Streptomyces_coelicolor_A3_chromosome   SCO5090   5532706            1   \n",
              "321  Streptomyces_coelicolor_A3_chromosome   SCO5089   5532449            1   \n",
              "354  Streptomyces_coelicolor_A3_chromosome   SCO5090   5532706            1   \n",
              "356  Streptomyces_coelicolor_A3_chromosome   SCO5090   5532706            1   \n",
              "17   Streptomyces_coelicolor_A3_chromosome   SCO5087   5529801            1   \n",
              "54   Streptomyces_coelicolor_A3_chromosome   SCO5087   5529801            1   \n",
              "474  Streptomyces_coelicolor_A3_chromosome   SCO5090   5532706            1   \n",
              "\n",
              "     sgrna_strand  sgrna_loc    gc  pam                 sgrna  \\\n",
              "537             1        140  0.65  CGG  CACGTACAGGTCCTGGAGCG   \n",
              "321             1        167  0.70  CGG  GCGCGACTCGAGAGCCGGTA   \n",
              "354            -1        174  0.65  CGG  CGCCCAGATCCGGAAGCGTT   \n",
              "356            -1        208  0.65  CGG  GGGACGTCCAGGTCTTCACC   \n",
              "17             -1        210  0.60  GGG  GAGCAGTTCCCAGAACTGCC   \n",
              "54             -1        522  0.70  CGG  GTCGACCTCCCAGTCACGGC   \n",
              "474             1        640  0.60  TGG  GACCGTCCAGGACATGACCA   \n",
              "\n",
              "    sgrna_seed_sequence  off_target_count editable_cytosines  \\\n",
              "537       AGGTCCTGGAGCG                 2                3,7   \n",
              "321       TCGAGAGCCGGTA                 0              4,7,9   \n",
              "354       ATCCGGAAGCGTT                 0           3,4,5,10   \n",
              "356       CCAGGTCTTCACC                 0              5,8,9   \n",
              "17        TCCCAGAACTGCC                 0             4,9,10   \n",
              "54        TCCCAGTCACGGC                 0         3,6,7,9,10   \n",
              "474       CAGGACATGACCA                 0            3,4,7,8   \n",
              "\n",
              "                      mutations  \n",
              "537                        Q43*  \n",
              "321                  R51*, L52F  \n",
              "354                  W57*, A58T  \n",
              "356                        W67*  \n",
              "17                   W67*, E68K  \n",
              "54   W171*, E172K, V173I, D174N  \n",
              "474                T208I, Q210*  "
            ]
          },
          "execution_count": 6,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "# Process the DataFrame to apply C-to-T mutations\n",
        "mutated_sgrna_df = process_base_editing(filtered_sgrna_df_for_base_editing, \n",
        "                                        genes_to_KO_dict, \n",
        "                                        only_stop_codons = True)\n",
        "mutated_sgrna_df"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 7,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>strain_name</th>\n",
              "      <th>locus_tag</th>\n",
              "      <th>gene_loc</th>\n",
              "      <th>gene_strand</th>\n",
              "      <th>sgrna_strand</th>\n",
              "      <th>sgrna_loc</th>\n",
              "      <th>gc</th>\n",
              "      <th>pam</th>\n",
              "      <th>sgrna</th>\n",
              "      <th>sgrna_seed_sequence</th>\n",
              "      <th>off_target_count</th>\n",
              "      <th>editable_cytosines</th>\n",
              "      <th>mutations</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>537</th>\n",
              "      <td>Streptomyces_coelicolor_A3_chromosome</td>\n",
              "      <td>SCO5090</td>\n",
              "      <td>5532706</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>140</td>\n",
              "      <td>0.65</td>\n",
              "      <td>CGG</td>\n",
              "      <td>CACGTACAGGTCCTGGAGCG</td>\n",
              "      <td>AGGTCCTGGAGCG</td>\n",
              "      <td>2</td>\n",
              "      <td>3,7</td>\n",
              "      <td>Q43*</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>321</th>\n",
              "      <td>Streptomyces_coelicolor_A3_chromosome</td>\n",
              "      <td>SCO5089</td>\n",
              "      <td>5532449</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>167</td>\n",
              "      <td>0.70</td>\n",
              "      <td>CGG</td>\n",
              "      <td>GCGCGACTCGAGAGCCGGTA</td>\n",
              "      <td>TCGAGAGCCGGTA</td>\n",
              "      <td>0</td>\n",
              "      <td>4,7,9</td>\n",
              "      <td>R51*, L52F</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>354</th>\n",
              "      <td>Streptomyces_coelicolor_A3_chromosome</td>\n",
              "      <td>SCO5090</td>\n",
              "      <td>5532706</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>174</td>\n",
              "      <td>0.65</td>\n",
              "      <td>CGG</td>\n",
              "      <td>CGCCCAGATCCGGAAGCGTT</td>\n",
              "      <td>ATCCGGAAGCGTT</td>\n",
              "      <td>0</td>\n",
              "      <td>3,4,5,10</td>\n",
              "      <td>W57*, A58T</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>356</th>\n",
              "      <td>Streptomyces_coelicolor_A3_chromosome</td>\n",
              "      <td>SCO5090</td>\n",
              "      <td>5532706</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>208</td>\n",
              "      <td>0.65</td>\n",
              "      <td>CGG</td>\n",
              "      <td>GGGACGTCCAGGTCTTCACC</td>\n",
              "      <td>CCAGGTCTTCACC</td>\n",
              "      <td>0</td>\n",
              "      <td>5,8,9</td>\n",
              "      <td>W67*</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>17</th>\n",
              "      <td>Streptomyces_coelicolor_A3_chromosome</td>\n",
              "      <td>SCO5087</td>\n",
              "      <td>5529801</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>210</td>\n",
              "      <td>0.60</td>\n",
              "      <td>GGG</td>\n",
              "      <td>GAGCAGTTCCCAGAACTGCC</td>\n",
              "      <td>TCCCAGAACTGCC</td>\n",
              "      <td>0</td>\n",
              "      <td>4,9,10</td>\n",
              "      <td>W67*, E68K</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>54</th>\n",
              "      <td>Streptomyces_coelicolor_A3_chromosome</td>\n",
              "      <td>SCO5087</td>\n",
              "      <td>5529801</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>522</td>\n",
              "      <td>0.70</td>\n",
              "      <td>CGG</td>\n",
              "      <td>GTCGACCTCCCAGTCACGGC</td>\n",
              "      <td>TCCCAGTCACGGC</td>\n",
              "      <td>0</td>\n",
              "      <td>3,6,7,9,10</td>\n",
              "      <td>W171*, E172K, V173I, D174N</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>474</th>\n",
              "      <td>Streptomyces_coelicolor_A3_chromosome</td>\n",
              "      <td>SCO5090</td>\n",
              "      <td>5532706</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>640</td>\n",
              "      <td>0.60</td>\n",
              "      <td>TGG</td>\n",
              "      <td>GACCGTCCAGGACATGACCA</td>\n",
              "      <td>CAGGACATGACCA</td>\n",
              "      <td>0</td>\n",
              "      <td>3,4,7,8</td>\n",
              "      <td>T208I, Q210*</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "                               strain_name locus_tag  gene_loc  gene_strand  \\\n",
              "537  Streptomyces_coelicolor_A3_chromosome   SCO5090   5532706            1   \n",
              "321  Streptomyces_coelicolor_A3_chromosome   SCO5089   5532449            1   \n",
              "354  Streptomyces_coelicolor_A3_chromosome   SCO5090   5532706            1   \n",
              "356  Streptomyces_coelicolor_A3_chromosome   SCO5090   5532706            1   \n",
              "17   Streptomyces_coelicolor_A3_chromosome   SCO5087   5529801            1   \n",
              "54   Streptomyces_coelicolor_A3_chromosome   SCO5087   5529801            1   \n",
              "474  Streptomyces_coelicolor_A3_chromosome   SCO5090   5532706            1   \n",
              "\n",
              "     sgrna_strand  sgrna_loc    gc  pam                 sgrna  \\\n",
              "537             1        140  0.65  CGG  CACGTACAGGTCCTGGAGCG   \n",
              "321             1        167  0.70  CGG  GCGCGACTCGAGAGCCGGTA   \n",
              "354            -1        174  0.65  CGG  CGCCCAGATCCGGAAGCGTT   \n",
              "356            -1        208  0.65  CGG  GGGACGTCCAGGTCTTCACC   \n",
              "17             -1        210  0.60  GGG  GAGCAGTTCCCAGAACTGCC   \n",
              "54             -1        522  0.70  CGG  GTCGACCTCCCAGTCACGGC   \n",
              "474             1        640  0.60  TGG  GACCGTCCAGGACATGACCA   \n",
              "\n",
              "    sgrna_seed_sequence  off_target_count editable_cytosines  \\\n",
              "537       AGGTCCTGGAGCG                 2                3,7   \n",
              "321       TCGAGAGCCGGTA                 0              4,7,9   \n",
              "354       ATCCGGAAGCGTT                 0           3,4,5,10   \n",
              "356       CCAGGTCTTCACC                 0              5,8,9   \n",
              "17        TCCCAGAACTGCC                 0             4,9,10   \n",
              "54        TCCCAGTCACGGC                 0         3,6,7,9,10   \n",
              "474       CAGGACATGACCA                 0            3,4,7,8   \n",
              "\n",
              "                      mutations  \n",
              "537                        Q43*  \n",
              "321                  R51*, L52F  \n",
              "354                  W57*, A58T  \n",
              "356                        W67*  \n",
              "17                   W67*, E68K  \n",
              "54   W171*, E172K, V173I, D174N  \n",
              "474                T208I, Q210*  "
            ]
          },
          "execution_count": 7,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "# Filter the DataFrame to retain only up to 5 sgRNA sequences per locus_tag\n",
        "filtered_df = mutated_sgrna_df.groupby('locus_tag').head(number_of_sgRNAs_per_group)\n",
        "filtered_df"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "## Output"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 8,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/plain": [
              "[Dseqrecord(-20),\n",
              " Dseqrecord(-20),\n",
              " Dseqrecord(-20),\n",
              " Dseqrecord(-20),\n",
              " Dseqrecord(-20),\n",
              " Dseqrecord(-20),\n",
              " Dseqrecord(-20)]"
            ]
          },
          "execution_count": 8,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "sgRNA_list = dataframe_to_seqrecords(filtered_df)\n",
        "sgRNA_list"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 9,
      "metadata": {},
      "outputs": [],
      "source": [
        "\n",
        "sgRNA_handle_cys4_sites = sgRNA_handle_cys4_site*len(sgRNA_list)\n",
        "\n",
        "golden_gate = GoldenGateCloning(sgRNA_list,\n",
        "                                sgRNA_handle_cys4_sites, \n",
        "                                target_tm= melting_temperature,\n",
        "                                restriction_overhang_f = \"GATCGggtctcc\",\n",
        "                                restriction_overhang_r = \"GATCAGGTCTCg\",\n",
        "                                backbone_overhang_f = \"cATG\",\n",
        "                                backbone_overhang_r = \"cTAG\",\n",
        "                                cys4 = \"gTTCACTGCCGTATAGGCAGCTAAGAAA\",\n",
        "                                tm_function = primer_tm_neb, \n",
        "                                primer_incrementation = primer_number_increment, \n",
        "                                polymerase = chosen_polymerase ) "
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 10,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>template</th>\n",
              "      <th>f_primer_anneal(5-3)</th>\n",
              "      <th>r_primer_anneal(5-3)</th>\n",
              "      <th>f_tm</th>\n",
              "      <th>r_tm</th>\n",
              "      <th>ta</th>\n",
              "      <th>f_primer_sequences(5-3)</th>\n",
              "      <th>r_primer_sequences(5-3)</th>\n",
              "      <th>f_primer_name</th>\n",
              "      <th>r_primer_name</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>sgRNA_handle_cys4</td>\n",
              "      <td>GTTTTAGAGCTAGAAATAGCAAGTTAAAATAAGGC</td>\n",
              "      <td>AAAAAAGCACCGACTCGGT</td>\n",
              "      <td>63</td>\n",
              "      <td>59</td>\n",
              "      <td>63</td>\n",
              "      <td>GATCGGGTCTCCCATGGTTCACTGCCGTATAGGCAGCTAAGAAACA...</td>\n",
              "      <td>GATCAGGTCTCGGCGCAAAAAAGCACCGACTCGGT</td>\n",
              "      <td>primer_23</td>\n",
              "      <td>primer_24</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>sgRNA_handle_cys4</td>\n",
              "      <td>GTTTTAGAGCTAGAAATAGCAAGTTAAAATAAGGC</td>\n",
              "      <td>AAAAAAGCACCGACTCGGT</td>\n",
              "      <td>63</td>\n",
              "      <td>59</td>\n",
              "      <td>63</td>\n",
              "      <td>GATCGGGTCTCCGCGCGACTCGAGAGCCGGTAGTTTTAGAGCTAGA...</td>\n",
              "      <td>GATCAGGTCTCGGGCGAAAAAAGCACCGACTCGGT</td>\n",
              "      <td>primer_25</td>\n",
              "      <td>primer_26</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>sgRNA_handle_cys4</td>\n",
              "      <td>GTTTTAGAGCTAGAAATAGCAAGTTAAAATAAGGC</td>\n",
              "      <td>AAAAAAGCACCGACTCGGT</td>\n",
              "      <td>63</td>\n",
              "      <td>59</td>\n",
              "      <td>63</td>\n",
              "      <td>GATCGGGTCTCCCGCCCAGATCCGGAAGCGTTGTTTTAGAGCTAGA...</td>\n",
              "      <td>GATCAGGTCTCGTCCCAAAAAAGCACCGACTCGGT</td>\n",
              "      <td>primer_27</td>\n",
              "      <td>primer_28</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>sgRNA_handle_cys4</td>\n",
              "      <td>GTTTTAGAGCTAGAAATAGCAAGTTAAAATAAGGC</td>\n",
              "      <td>AAAAAAGCACCGACTCGGT</td>\n",
              "      <td>63</td>\n",
              "      <td>59</td>\n",
              "      <td>63</td>\n",
              "      <td>GATCGGGTCTCCGGGACGTCCAGGTCTTCACCGTTTTAGAGCTAGA...</td>\n",
              "      <td>GATCAGGTCTCGGCTCAAAAAAGCACCGACTCGGT</td>\n",
              "      <td>primer_29</td>\n",
              "      <td>primer_30</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>sgRNA_handle_cys4</td>\n",
              "      <td>GTTTTAGAGCTAGAAATAGCAAGTTAAAATAAGGC</td>\n",
              "      <td>AAAAAAGCACCGACTCGGT</td>\n",
              "      <td>63</td>\n",
              "      <td>59</td>\n",
              "      <td>63</td>\n",
              "      <td>GATCGGGTCTCCGAGCAGTTCCCAGAACTGCCGTTTTAGAGCTAGA...</td>\n",
              "      <td>GATCAGGTCTCGCGACAAAAAAGCACCGACTCGGT</td>\n",
              "      <td>primer_31</td>\n",
              "      <td>primer_32</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>5</th>\n",
              "      <td>sgRNA_handle_cys4</td>\n",
              "      <td>GTTTTAGAGCTAGAAATAGCAAGTTAAAATAAGGC</td>\n",
              "      <td>AAAAAAGCACCGACTCGGT</td>\n",
              "      <td>63</td>\n",
              "      <td>59</td>\n",
              "      <td>63</td>\n",
              "      <td>GATCGGGTCTCCGTCGACCTCCCAGTCACGGCGTTTTAGAGCTAGA...</td>\n",
              "      <td>GATCAGGTCTCGGGTCAAAAAAGCACCGACTCGGT</td>\n",
              "      <td>primer_33</td>\n",
              "      <td>primer_34</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>6</th>\n",
              "      <td>sgRNA_handle_cys4</td>\n",
              "      <td>GTTTTAGAGCTAGAAATAGCAAGTTAAAATAAGGC</td>\n",
              "      <td>AAAAAAGCACCGACTCGGT</td>\n",
              "      <td>63</td>\n",
              "      <td>59</td>\n",
              "      <td>63</td>\n",
              "      <td>GATCGGGTCTCCGACCGTCCAGGACATGACCAGTTTTAGAGCTAGA...</td>\n",
              "      <td>GATCAGGTCTCGCTAGAAAAAAGCACCGACTCGGT</td>\n",
              "      <td>primer_35</td>\n",
              "      <td>primer_36</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "            template                 f_primer_anneal(5-3)  \\\n",
              "0  sgRNA_handle_cys4  GTTTTAGAGCTAGAAATAGCAAGTTAAAATAAGGC   \n",
              "1  sgRNA_handle_cys4  GTTTTAGAGCTAGAAATAGCAAGTTAAAATAAGGC   \n",
              "2  sgRNA_handle_cys4  GTTTTAGAGCTAGAAATAGCAAGTTAAAATAAGGC   \n",
              "3  sgRNA_handle_cys4  GTTTTAGAGCTAGAAATAGCAAGTTAAAATAAGGC   \n",
              "4  sgRNA_handle_cys4  GTTTTAGAGCTAGAAATAGCAAGTTAAAATAAGGC   \n",
              "5  sgRNA_handle_cys4  GTTTTAGAGCTAGAAATAGCAAGTTAAAATAAGGC   \n",
              "6  sgRNA_handle_cys4  GTTTTAGAGCTAGAAATAGCAAGTTAAAATAAGGC   \n",
              "\n",
              "  r_primer_anneal(5-3)  f_tm  r_tm  ta  \\\n",
              "0  AAAAAAGCACCGACTCGGT    63    59  63   \n",
              "1  AAAAAAGCACCGACTCGGT    63    59  63   \n",
              "2  AAAAAAGCACCGACTCGGT    63    59  63   \n",
              "3  AAAAAAGCACCGACTCGGT    63    59  63   \n",
              "4  AAAAAAGCACCGACTCGGT    63    59  63   \n",
              "5  AAAAAAGCACCGACTCGGT    63    59  63   \n",
              "6  AAAAAAGCACCGACTCGGT    63    59  63   \n",
              "\n",
              "                             f_primer_sequences(5-3)  \\\n",
              "0  GATCGGGTCTCCCATGGTTCACTGCCGTATAGGCAGCTAAGAAACA...   \n",
              "1  GATCGGGTCTCCGCGCGACTCGAGAGCCGGTAGTTTTAGAGCTAGA...   \n",
              "2  GATCGGGTCTCCCGCCCAGATCCGGAAGCGTTGTTTTAGAGCTAGA...   \n",
              "3  GATCGGGTCTCCGGGACGTCCAGGTCTTCACCGTTTTAGAGCTAGA...   \n",
              "4  GATCGGGTCTCCGAGCAGTTCCCAGAACTGCCGTTTTAGAGCTAGA...   \n",
              "5  GATCGGGTCTCCGTCGACCTCCCAGTCACGGCGTTTTAGAGCTAGA...   \n",
              "6  GATCGGGTCTCCGACCGTCCAGGACATGACCAGTTTTAGAGCTAGA...   \n",
              "\n",
              "               r_primer_sequences(5-3) f_primer_name r_primer_name  \n",
              "0  GATCAGGTCTCGGCGCAAAAAAGCACCGACTCGGT     primer_23     primer_24  \n",
              "1  GATCAGGTCTCGGGCGAAAAAAGCACCGACTCGGT     primer_25     primer_26  \n",
              "2  GATCAGGTCTCGTCCCAAAAAAGCACCGACTCGGT     primer_27     primer_28  \n",
              "3  GATCAGGTCTCGGCTCAAAAAAGCACCGACTCGGT     primer_29     primer_30  \n",
              "4  GATCAGGTCTCGCGACAAAAAAGCACCGACTCGGT     primer_31     primer_32  \n",
              "5  GATCAGGTCTCGGGTCAAAAAAGCACCGACTCGGT     primer_33     primer_34  \n",
              "6  GATCAGGTCTCGCTAGAAAAAAGCACCGACTCGGT     primer_35     primer_36  "
            ]
          },
          "execution_count": 10,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "primer_df = golden_gate.generate_primer_dataframe()\n",
        "primer_df"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 11,
      "metadata": {},
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "Skipping analysis for primer pair primer_23 & primer_24 due to length constraints.\n",
            "Skipping analysis for primer pair primer_25 & primer_26 due to length constraints.\n",
            "Skipping analysis for primer pair primer_27 & primer_28 due to length constraints.\n",
            "Skipping analysis for primer pair primer_29 & primer_30 due to length constraints.\n",
            "Skipping analysis for primer pair primer_31 & primer_32 due to length constraints.\n",
            "Skipping analysis for primer pair primer_33 & primer_34 due to length constraints.\n",
            "Skipping analysis for primer pair primer_35 & primer_36 due to length constraints.\n"
          ]
        },
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "Empty DataFrame\n",
              "Columns: []\n",
              "Index: []"
            ]
          },
          "execution_count": 11,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "# Primer analysis\n",
        "analysis_of_primers = analyze_primers_and_hairpins(primer_df)\n",
        "analysis_of_primers"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 12,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>Name</th>\n",
              "      <th>Sequence</th>\n",
              "      <th>Concentration</th>\n",
              "      <th>Purification</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>primer_23</td>\n",
              "      <td>GATCGGGTCTCCCATGGTTCACTGCCGTATAGGCAGCTAAGAAACA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>primer_25</td>\n",
              "      <td>GATCGGGTCTCCGCGCGACTCGAGAGCCGGTAGTTTTAGAGCTAGA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>primer_27</td>\n",
              "      <td>GATCGGGTCTCCCGCCCAGATCCGGAAGCGTTGTTTTAGAGCTAGA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>primer_29</td>\n",
              "      <td>GATCGGGTCTCCGGGACGTCCAGGTCTTCACCGTTTTAGAGCTAGA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>primer_31</td>\n",
              "      <td>GATCGGGTCTCCGAGCAGTTCCCAGAACTGCCGTTTTAGAGCTAGA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>5</th>\n",
              "      <td>primer_33</td>\n",
              "      <td>GATCGGGTCTCCGTCGACCTCCCAGTCACGGCGTTTTAGAGCTAGA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>6</th>\n",
              "      <td>primer_35</td>\n",
              "      <td>GATCGGGTCTCCGACCGTCCAGGACATGACCAGTTTTAGAGCTAGA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>7</th>\n",
              "      <td>primer_24</td>\n",
              "      <td>GATCAGGTCTCGGCGCAAAAAAGCACCGACTCGGT</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>8</th>\n",
              "      <td>primer_26</td>\n",
              "      <td>GATCAGGTCTCGGGCGAAAAAAGCACCGACTCGGT</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>9</th>\n",
              "      <td>primer_28</td>\n",
              "      <td>GATCAGGTCTCGTCCCAAAAAAGCACCGACTCGGT</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>10</th>\n",
              "      <td>primer_30</td>\n",
              "      <td>GATCAGGTCTCGGCTCAAAAAAGCACCGACTCGGT</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>11</th>\n",
              "      <td>primer_32</td>\n",
              "      <td>GATCAGGTCTCGCGACAAAAAAGCACCGACTCGGT</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>12</th>\n",
              "      <td>primer_34</td>\n",
              "      <td>GATCAGGTCTCGGGTCAAAAAAGCACCGACTCGGT</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>13</th>\n",
              "      <td>primer_36</td>\n",
              "      <td>GATCAGGTCTCGCTAGAAAAAAGCACCGACTCGGT</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "         Name                                           Sequence  \\\n",
              "0   primer_23  GATCGGGTCTCCCATGGTTCACTGCCGTATAGGCAGCTAAGAAACA...   \n",
              "1   primer_25  GATCGGGTCTCCGCGCGACTCGAGAGCCGGTAGTTTTAGAGCTAGA...   \n",
              "2   primer_27  GATCGGGTCTCCCGCCCAGATCCGGAAGCGTTGTTTTAGAGCTAGA...   \n",
              "3   primer_29  GATCGGGTCTCCGGGACGTCCAGGTCTTCACCGTTTTAGAGCTAGA...   \n",
              "4   primer_31  GATCGGGTCTCCGAGCAGTTCCCAGAACTGCCGTTTTAGAGCTAGA...   \n",
              "5   primer_33  GATCGGGTCTCCGTCGACCTCCCAGTCACGGCGTTTTAGAGCTAGA...   \n",
              "6   primer_35  GATCGGGTCTCCGACCGTCCAGGACATGACCAGTTTTAGAGCTAGA...   \n",
              "7   primer_24                GATCAGGTCTCGGCGCAAAAAAGCACCGACTCGGT   \n",
              "8   primer_26                GATCAGGTCTCGGGCGAAAAAAGCACCGACTCGGT   \n",
              "9   primer_28                GATCAGGTCTCGTCCCAAAAAAGCACCGACTCGGT   \n",
              "10  primer_30                GATCAGGTCTCGGCTCAAAAAAGCACCGACTCGGT   \n",
              "11  primer_32                GATCAGGTCTCGCGACAAAAAAGCACCGACTCGGT   \n",
              "12  primer_34                GATCAGGTCTCGGGTCAAAAAAGCACCGACTCGGT   \n",
              "13  primer_36                GATCAGGTCTCGCTAGAAAAAAGCACCGACTCGGT   \n",
              "\n",
              "   Concentration Purification  \n",
              "0           25nm          STD  \n",
              "1           25nm          STD  \n",
              "2           25nm          STD  \n",
              "3           25nm          STD  \n",
              "4           25nm          STD  \n",
              "5           25nm          STD  \n",
              "6           25nm          STD  \n",
              "7           25nm          STD  \n",
              "8           25nm          STD  \n",
              "9           25nm          STD  \n",
              "10          25nm          STD  \n",
              "11          25nm          STD  \n",
              "12          25nm          STD  \n",
              "13          25nm          STD  "
            ]
          },
          "execution_count": 12,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "# 2 Ready made IDT format\n",
        "idt_df = create_idt_order_dataframe(primer_df, concentration=\"25nm\", purification=\"STD\")\n",
        "idt_df"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 13,
      "metadata": {},
      "outputs": [],
      "source": [
        "list_of_amplicons = golden_gate.simulate_pcrs()"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 14,
      "metadata": {},
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "SCO5090_140_pcr_products\n",
            "SCO5089_167_pcr_products\n",
            "SCO5090_174_pcr_products\n",
            "SCO5090_208_pcr_products\n",
            "SCO5087_210_pcr_products\n",
            "SCO5087_522_pcr_products\n",
            "SCO5090_640_pcr_products\n"
          ]
        },
        {
          "data": {
            "image/jpeg": "/9j/4AAQSkZJRgABAQAAAQABAAD/2wBDAAgGBgcGBQgHBwcJCQgKDBQNDAsLDBkSEw8UHRofHh0aHBwgJC4nICIsIxwcKDcpLDAxNDQ0Hyc5PTgyPC4zNDL/2wBDAQkJCQwLDBgNDRgyIRwhMjIyMjIyMjIyMjIyMjIyMjIyMjIyMjIyMjIyMjIyMjIyMjIyMjIyMjIyMjIyMjIyMjL/wAARCAJYAhwDASIAAhEBAxEB/8QAHwAAAQUBAQEBAQEAAAAAAAAAAAECAwQFBgcICQoL/8QAtRAAAgEDAwIEAwUFBAQAAAF9AQIDAAQRBRIhMUEGE1FhByJxFDKBkaEII0KxwRVS0fAkM2JyggkKFhcYGRolJicoKSo0NTY3ODk6Q0RFRkdISUpTVFVWV1hZWmNkZWZnaGlqc3R1dnd4eXqDhIWGh4iJipKTlJWWl5iZmqKjpKWmp6ipqrKztLW2t7i5usLDxMXGx8jJytLT1NXW19jZ2uHi4+Tl5ufo6erx8vP09fb3+Pn6/8QAHwEAAwEBAQEBAQEBAQAAAAAAAAECAwQFBgcICQoL/8QAtREAAgECBAQDBAcFBAQAAQJ3AAECAxEEBSExBhJBUQdhcRMiMoEIFEKRobHBCSMzUvAVYnLRChYkNOEl8RcYGRomJygpKjU2Nzg5OkNERUZHSElKU1RVVldYWVpjZGVmZ2hpanN0dXZ3eHl6goOEhYaHiImKkpOUlZaXmJmaoqOkpaanqKmqsrO0tba3uLm6wsPExcbHyMnK0tPU1dbX2Nna4uPk5ebn6Onq8vP09fb3+Pn6/9oADAMBAAIRAxEAPwDwiHTr24tJruGzuJLaD/WzJExSP/eYDA/GnvpOpRfZfM0+7T7Xj7NuhYednGNnHzdR09RXQaL51r4Xu78Xto7iO4t7eze8iRoxIgWaUozBmynyqACSef4QD0japYRapqszajaGPVb+WbT2WdW+zo1tcIhfB/dYaaEfNgjyz2FAHnyaJq0l9JYppd615Eu6S3Fu5kQcclcZA5H51CmnX0ljJfJZ3DWkTbZLgRMY0PHBbGAeR+ddxJJFPoMGiQanZRapbxWhlnN2ixkJLcsVEudrFRNBwCfucZxVjU9W029u7jVbS7t005LXU4GtjIqyNJNJOYsR53EYlhOQMDy+cYoA82oqZ2tjaRqkMwuQT5kjSgow7ALtBB99x/CoaAFCkjpRtPpU8VwEjC7Acd6f9rH/ADzX8qAKu0+lG0+lWvtY/wCea/lR9rH/ADzX8qAKu0+lG0+lWvtY/wCea/lR9rH/ADzX8qAKpBHUUlTzTiVAoUDnPFQUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFadnoc97ouoapHParFYqHeJpf3rguiZVRngGROTgcnGSCK0H8F6is6QrcWjyLKYbkK7f6I4QuRJlR0VHJ27vuN3FAHOUV0C+ErkuXfULCOyKRvHeu0gik8wsqgfJuzlHHKjGxs4xTX8JahDbTPO9vFcxrM4s3ZvNdImKyMMDbhSr9SD8jYBxQBg0UUUAFFTJbSSIGUHBp32OX+6aAK9FWPscv900fY5f7poAr0VY+xy/3TR9jl/umgCvRUskDxLuYYGcVFQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQBp6XqMNlp+tW8iuXvrJbeIqBgMLiGTJ56Yjbpnkj610z+MtMW9vriOG7b+17qS41BWRR5W+GaJljO47v+PiQgnb0UeprhqKAOvfxBotxpkehz/bxpkCQeXOkKea7RvO5ym/AB+0yD7xxhetS3vjGz1GabVJ4bhNT8i9t44kVTEVuXlYsWzkFftEgxtOcLyK4uigCZ7qR7SO2KwiOMllKwoHOfVwNxHsScVDRRQBMlzJGgVScCnfbJf7xqvRQBY+2S/3jR9sl/vGq9FAFj7ZL/eNH2yX+8ar0UASyTvKu1jkZzUVFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRW1puk6bfaVe3MmoXcM9pbtM6fZFaMncFRd/mA5Yso+7xyeQK1H8EL9qmt49S3PZXDW+pFoMC3ZYpJGKfMd4AhlH8PK++aAORorqD4Vs47cajPqskekyJC0NwLUGRmkaRQGj34GDDLk7j93jOawNRsZtL1O70+4x59rM8Em05G5WKnH4igCtRRRQAUVbgtVkhVy6gnsTUn2FP+ei/nQBQoq/9hT/AJ6L+dH2FP8Anov50AUKKv8A2FP+ei/nR9hT/nov50AUKKtXNssMYYMDk44qrQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQBf/tLboX9mRRbN9x588u7PmYXCLjHAXLn33e1b7+N1+1TXEem7XvbhrjUg0+RcM0UkbBPlGwETSn+LlvbFcjRQB1B8VWcluunT6VJJpMaQrDbi6AkVo2kYFpNmDkzS5G0fe4xisW41i9uNcuNYWd4L2ed5zLCxQq7kk4I5HU1RooAVmZ3LuxZmOSSckmkoooAcHYDAPFHmP602igB3mP60eY/rTaKAHeY/rR5j+tNooAcXZhgmm0UUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFby+HlHhhNYkbUDv34ENlvhTB2jfLvG3Jz/CelaD+CF+1TW8epbnsrhrfUi0GBbssUkjFPmO8AQyj+HlffNAHI0V1B8K2cduNRn1WSPSZEhaG4FqDIzSNIoDR78DBhlydx+7xnNYGo2M2l6nd6fcY8+1meCTacjcrFTj8RQBWooooAKK7LQvClvqWjW9293DG0m7Ks3IwxH9K0f+EGtf8An/g/76oA88or0P8A4Qa1/wCf+D/vqj/hBrX/AJ/4P++qAPPKK9D/AOEGtf8An/g/76o/4Qa1/wCf+D/vqgDzyiuq8S+G4dH06O4juYpS0wTahyehOf0rlaACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKAN6x8RRWWliEWczXaW89skpuj5QSUENmIg/NhjgggZwSCRzoP43X7VNcR6bte9uGuNSDT5FwzRSRsE+UbARNKf4uW9sVyNFAHUHxVZyW66dPpUkmkxpCsNuLoCRWjaRgWk2YOTNLkbR97jGKxbjWL241y41hZ3gvZ53nMsLFCruSTgjkdTVGigBWZncu7FmY5JJySaSiigCzFfzwxCNJGCjoAaf/al1/z1b86p0UAXP7Uuv+erfnR/al1/z1b86p0UAXP7Uuv+erfnR/al1/z1b86p0UAWJ7ya4QJI5YA5wTVeiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigC5p2lX2qvcLY27TG2t5LmbBACRIMsxJ9B+fQc1cfwtrMb2yNZ5e4fYirKjFW27sOAf3Z2nOGxxk9jWv4T8RaRpdlcWV7a3KGaC78y4iuAolLW0kcaFfLY9XIBzgF8nIAxqt4l0aPUNUuFvt6a3eS3DhYnzZLJb3EeGyBkhrn+HPEfuBQBycfhbWJLx7ZbeLeqI+83MQjKucLtkLbGyeBgnOD6GsmSN4ZXilRkkRirKwwVI6giu1l1LR7rQ4fDzatHElvHbH7f5MpSRkkuXZQAu/pc4GVGfL7ZFc7quqw3/irUNWFoksFzdy3At5iwGHYkA7SDxnsaAMqilYhnJChQTkKM4HtzSUAWIrSSWMOoODT/ALBL/dNMiu5Iowik4FP+3y/3jQAfYJf7po+wS/3TR9vl/vGj7fL/AHjQAfYJf7po+wS/3TR9vl/vGj7fL/eNAEc1s8KBmGATioKnmuXmQKxyAc1BQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAWotMv57GW+hsbmS0iOJJ0iYxoeOrAYHUfnTn0nUovsvmafdp9rx9m3QsPOzjGzj5uo6eorpNAljtNGubma/tGifT7qFRJPia0kZWGyOIt8wkyoLbSAGPIK5roG1Swi1TVZm1G0Meq38s2nss6t9nRra4RC+D+6w00I+bBHlnsKAPPk0TVpL6SxTS71ryJd0luLdzIg45K4yByPzqh0NegySRT6DBokGp2UWqW8VoZZzdosZCS3LFRLnaxUTQcAn7nGcVyus3mn3ni/U71YpJNPnvZpUjicRMY2diuCVOOCOxoAyKKVtpclAQueATkgfWkoAlS3d0DAHBp32ST+6amgvmhhWMYwM1J/ab+1AFX7JJ/dNH2ST+6atf2m/tR/ab+1AFX7JJ/dNH2ST+6atf2m/tR/ab+1AFOSB41ywwM4qKrVzeNcRhD2Oaq0AFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUVoWuh6je6fLfW8AaCIOSTIoZggDPtUnc20EE7QcA5OKsP4W1mN7ZGs8vcPsRVlRirbd2HAP7s7TnDY4yexoAx6K2I/C2sSXj2y28W9UR95uYhGVc4XbIW2Nk8DBOcH0NZMkbwyvFKjJIjFWVhgqR1BFADaKKKACipo7Z5EDAHBp/2OX+6aAK1FWfscv900fY5f7poArUVZ+xy/3TR9jl/umgCtRUstu8ShmGBnFRUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAdJpM1lZeHLyddWgTVJUkgjt5VmzFEy4cqQhXe4JTkgAZz1BHQN4l0aPUNUuFvt6a3eS3DhYnzZLJb3EeGyBkhrn+HPEfuBXndFAHby6lo91ocPh5tWjiS3jtj9v8mUpIySXLsoAXf0ucDKjPl9siud1XVYb/AMVahqwtElgubuW4FvMWAw7EgHaQeM9jWVRQArEM5IUKCchRnA9uaSiigC/b6g0ECxjGBn+dSf2q/t+VZlFAGn/ar+35Uf2q/t+VZlFAGn/ar+35Uf2q/t+VZlFAFy7vWuYgh7NmqdFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUVt6XpunXvh7WrmWS6/tCygWeJFCiLaZoYzuPJJPmngYxtHJzgbt14OtItSjtxb3aW0SyPc3QvYptwjgaZ1CKgMbEI23cTx2NAHD0V17+H9Ft9Mj1yf7edMnSDy4EmTzUaR50OX2YIH2aQ/dGcr0rCv8ARbq08R3miQRyXVzb3MluFhQs0hRiCQBz2zQBm0UrKyOUdSrKcEEYINJQA9Y2YZAOKPJf+6a9V8G+HfDt/wCFLK5v75Yrl/M3oV6YdgP0Ard/4RHwj/0E0/75oA8N8l/7po8l/wC6a9y/4RHwj/0E0/75o/4RHwj/ANBNP++aAPDfJf8AumjyX/umvcv+ER8I/wDQTT/vmj/hEfCP/QTT/vmgDwxo2UZIIptek/ELQtC0zQIJtMvFmna6VGUDGF2uc/mBXm1ABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFAFuz1GaytdQt41QpfW4t5SwOQolSTI565jXrngn61rJ4vu4Sn2axsLYGQS3CxRsFuW2NH84LEAFZJBhNo+c+2OeooA6BfFtyHKPp9hJZBI0jsnWQxR+WWZSPn3Zy7nljne2c5rFu7qa9vJ7u4cyTzyNJI56szHJP5moaKACiiigC3DqE8ESxpIQo6AGpP7Wuv+erfnVCigC//AGtdf89W/Oj+1rr/AJ6t+dUKKAL/APa11/z1b86P7Wuv+erfnVCigCzcX01zGEkcsAc8mq1FFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFa+n6NBfaFquoHUFS4sIllFqIiS6mWOMktwFGZRjqTtPA4Na7+CF+1TW8epbnsrhrfUi0GBbssUkjFPmO8AQyj+HlffNAHI0V1B8K2cduNRn1WSPSZEhaG4FqDIzSNIoDR78DBhlydx+7xnNYGo2M2l6nd6fcY8+1meCTacjcrFTj8RQBWooooAKK2LHR3urOOYYw2e/oSKsf8I/J7fnQBz9FdB/wj8nt+dH/CPye350Ac/RXQf8I/J7fnR/wj8nt+dAHP0VqalpbWVusrYwXC/of8Ky6ACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKAL+n6l9hstVt/K3/b7Vbfdux5eJopc9Of9VjHHXPbB338br9qmuI9N2ve3DXGpBp8i4ZopI2CfKNgImlP8XLe2K5GigDqD4qs5LddOn0qSTSY0hWG3F0BIrRtIwLSbMHJmlyNo+9xjFYtxrF7ca5caws7wXs87zmWFihV3JJwRyOpqjRQArMzuXdizMckk5JNJRRQBcgv5YIVjViAKk/tSf8Avms+igDQ/tSf++aP7Un/AL5rPooA0P7Un/vmj+1J/wC+az6KALVzeyXMYR2JAOaq0UUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQBp2mgX99YyXlv9keKKJ5nU3kKyBEzuPllt3b054xnIqV/C2sxvbI1nl7h9iKsqMVbbuw4B/dnac4bHGT2NNgvLez8N3MMMmb2+lEcoCkeXAmGxnod77Tx08r3rrm8S6NHqGqXC329NbvJbhwsT5slkt7iPDZAyQ1z/AA54j9wKAOFv9OutMuBBdIquUDqUkV1dT0KspIYe4NVa2dfurWVNLs7W4W5Sws/s7TorKsjGWSQkBgDgeZt5A+7WY88bWkcItYUkQktOpfe+exBbbgewFAENFFFAGrZ6PNdWiTKpKtn+eKn/AOEfn/uGoLPWJrW0SFWIVc/zzU//AAkE/wDfNAB/wj8/9w0f8I/P/cNH/CQT/wB80f8ACQT/AN80AH/CPz/3DR/wj8/9w0f8JBP/AHzR/wAJBP8A3zQBV1DS5bKBZHUgFtv6H/Cs2tLUNUlvYFjdiQG3fof8azaACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooA0E0a8fS/7R/0dLY7tpkuokdtvXahYM3XsDVh/C2sxvbI1nl7h9iKsqMVbbuw4B/dnac4bHGT2Nallqmmnwwlrez2UixW1xGLY2Z+0ea24xssuCAoYqT8y8AjB4rYbxLo0eoapcLfb01u8luHCxPmyWS3uI8NkDJDXP8OeI/cCgDhb/TrrTLgQXSKrlA6lJFdXU9CrKSGHuDVWtnX7q1lTS7O1uFuUsLP7O06KyrIxlkkJAYA4HmbeQPu1mPPG1pHCLWFJEJLTqX3vnsQW24HsBQBDRRRQB2ug+AtQ1rRbfUIIGaOXdggejFf6Vpf8Ku1X/n2f/vmt3wN8V5/DXg6w0hAm238zGVB+9Izf+zV0P/C9Lr0j/wC+RQBwP/CrtV/59n/75o/4Vdqv/Ps//fNd9/wvS69I/wDvkUf8L0uvSP8A75FAHA/8Ku1X/n2f/vmj/hV2q/8APs//AHzXff8AC9Lr0j/75FH/AAvS69I/++RQB5N4o8GXvh3TI7u5hZEeYRAkdyGP9K5OvU/iV8SZvGPhy30+QLiO7WfgAdEdf/Zq8soAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAUMwGAaXe3qaQKSMgUuxvSgA3t6mje3qaNjelGxvSgA3t6mje3qaNjelGxvSgBCxPU0lKVI6ikoAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooA6Lw34dg1W2vrq+mkhhitblrYR4zNPFA8uOf4QFG4/7Sjjdka7+DdMa9vreOa7X+yLqS31BmdT5uyGaVmjG0bf+PeQAHd1U+orn9H8Va1oStHY6hcpA0U0fkec4jHmIyFgoIG4bsg+oB7VafxpqLTpMtvaJI0pmuiqN/pblChMmWPVXcELt++3c0AUdb0+2s/7PubLzha39r9ojSZgzx4keNlLAAH5o2IOBwRVB7O5jtI7t7aZbaUlY5mQhHI6gN0JFWNT1OTVJoWaGGCKCIQwwQhtkaAk4G4k8lmJJJ5JqjQAUUUUAdVo+iw3elQzvKis27IJ9GIq9/wjtv8A89o/zrlINQlghWNWIAz/ADqT+1Zv75oA6f8A4R23/wCe0f50f8I7b/8APaP865j+1Zv75o/tWb++aAOn/wCEdt/+e0f50f8ACO2//PaP865j+1Zv75o/tWb++aANLX9KisbBJUkRiZQuAfY/4VzlW7q9kuYwjsSAc1UoAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKcEYjIFHlt6UANop3lt6UeW3pQA2ineW3pR5belADaKUqQMkUlABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFAHRaLpGmanpV8zvcLd21tLcM4mRVXauUUR4LyZIwSCAoOTwOdd/BumNe31vHNdr/AGRdSW+oMzqfN2QzSs0Y2jb/AMe8gAO7qp9RXPWviK4tNOFqlramVYZbeO6ZW82OKQEOgw20ghnGSpI3HBHGLb+NNRadJlt7RJGlM10VRv8AS3KFCZMsequ4IXb99u5oAo63p9tZ/wBn3Nl5wtb+1+0RpMwZ48SPGylgAD80bEHA4Iqg9ncx2kd29tMttKSsczIQjkdQG6EirGp6nJqk0LNDDBFBEIYYIQ2yNAScDcSeSzEkk8k1RoAKKKKAOv0TQ4LzSIJ3lRWbdkE+jEf0rQ/4Rq2/57R/nXH22py29usSsQFz/Opf7Zn/AL5oA6v/AIRq2/57R/nR/wAI1bf89o/zrlP7Zn/vmj+2Z/75oA6v/hGrb/ntH+dH/CNW3/PaP865T+2Z/wC+aP7Zn/vmgDT8R6RDYafHLHIjEyhcA+x/wrmKu3eoSXUQR2JAbNUqACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigCzDp17cWk13DZ3EltB/rZkiYpH/vMBgfjT30nUovsvmafdp9rx9m3QsPOzjGzj5uo6eoroNF8618L3d+L20dxHcW9vZveRI0YkQLNKUZgzZT5VABJPP8IB6RtUsItU1WZtRtDHqt/LNp7LOrfZ0a2uEQvg/usNNCPmwR5Z7CgDzW6tLmxuXtru3lt50xuimQoy5GeQeahrd8SPGE0e0E0U1xZ2AhuHhkWRd/myOAGUkNhHReCRxjtWQ7WxtI1SGYXIJ8yRpQUYdgF2gg++4/hQBDRRRQBqWmky3VqkyqSGz/PFTf2DP/cP5Vv+H9ft7LQ7e3eJGZN2SRzyxP8AWtL/AISm1/54RflQBx39gz/3D+VH9gz/ANw/lXY/8JTa/wDPCL8qP+Eptf8AnhF+VAHHf2DP/cP5Uf2DP/cP5V2P/CU2v/PCL8qP+Eptf+eEX5UAcJf6ZJZwLI6kAtt/Q/4Vn11/irWodR0uKGONFKzBsqP9lh/WuRVSzYHWgBKKk8h/SjyH9KAI6Kk8h/SjyH9KAI6Kk8h/SjyH9KAI6KUjBIPakoAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigB6yMowDxR5z+tMooAf5z+tHnP60yigB/nP60ec/rTKKAHNIzDBNaXh21S9122t3YKr7sk+yk/0rLqezuGtbtJlOCuf5YoA9H/4Ra1/57R/nR/wi1r/z2j/OuO/t+f8Avn86P7fn/vn86AOx/wCEWtf+e0f50f8ACLWv/PaP8647+35/75/Oj+35/wC+fzoA7H/hFrX/AJ7R/nR/wi1r/wA9o/zrjv7fn/vn86P7fn/vn86AMzUohBqt5CpBEc7qCPZiKrYJ7U+eQy3Ekh6uxY/ia6DS9AkvdNiuFUkPn9CR/SgDnNp9KNp9K6//AIRab+4fyo/4Rab+4fyoA5DafSjafSuv/wCEWm/uH8qP+EWm/uH8qAOQ2n0oIIrr/wDhFpv7h/Ks3W9Fk06zSZlIBkC/oT/SgDB60u0+lW9LtTe6lFbqMl8/oCf6V0n/AAi039w/lQByG0+lG0+ldf8A8ItN/cP5Uf8ACLTf3D+VAHIbT6UbT6V1/wDwi039w/lR/wAItN/cP5UAchtPpSV2H/CLTf3D+VcrdxGC8nhPWORl/I4oAiwT2o2n0ro9L0CS902K4VSQ+f0JH9Kuf8ItN/cP5UAchtPpRtPpXX/8ItN/cP5Uf8ItN/cP5UAchtPpRtPpXX/8ItN/cP5Uf8ItN/cP5UAcgQRSda3tb0WTTrNJmUgGQL+hP9KzdLtTe6lFbqMl8/oCf6UAVNp9KNp9K6//AIRab+4fyo/4Rab+4fyoA5DafSjafSuv/wCEWm/uH8qP+EWm/uH8qAOQ2n0o2n0rr/8AhFpv7h/Kj/hFpv7h/KgDj6XBPapbuIwXk8J6xyMv5HFb2l6BJe6bFcKpIfP6Ej+lAHObT6UbT6V1/wDwi039w/lR/wAItN/cP5UAchtPpRtPpXX/APCLTf3D+VH/AAi039w/lQByG0+lBBFdf/wi039w/lWbreiyadZpMykAyBf0J/pQBg9aXafSrel2pvdSit1GS+f0BP8ASuk/4Rab+4fyoA5DafSjafSuv/4Rab+4fyo/4Rab+4fyoA5DafSjafSuv/4Rab+4fyo/4Rab+4fyoA5DafSkrsP+EWm/uH8q5W7iMF5PCescjL+RxQBDRRRQAUVr2GhrqGn3N1HqtjG9tA88lvIJg4VTgciMpkkqB83VgOKj1bRxpJWN9RtJ7jjzIIRLujyM8lkC8ZxwTzQBmUUUUAFFFFADgjEZAOKPLf8AumvQPDPg2bVvD9rerGSsu/nHo5H9K1v+FeT/APPE/lQB5T5b/wB00eW/9016t/wryf8A54n8qP8AhXk//PE/lQB5T5b/AN00eW/9016t/wAK8n/54n8qP+FeT/8APE/lQB5QVZRkjFJXb+MfCsuh6RFcvGVDTiPOPVWP9K4igAooooAKKKKACiiigArpNL1+Sy02K3ViAmf1JP8AWubpckd6AOv/AOEpm/vn86P+Epm/vn865DcfWjcfWgDr/wDhKZv75/Oj/hKZv75/OuQ3H1o3H1oA6/8A4Smb++fzrN1vWpNRs0hZiQJA36Ef1rC3H1oJJoAt6XdGy1KK4U4KZ/UEf1rpP+Epm/vn864/pS7j60Adf/wlM398/nR/wlM398/nXIbj60bj60Adf/wlM398/nR/wlM398/nXIbj60bj60Adf/wlM398/nXK3cpnvJ5j1kkZvzOai3H1pKAOk0vX5LLTYrdWICZ/Uk/1q5/wlM398/nXIZI70bj60Adf/wAJTN/fP50f8JTN/fP51yG4+tG4+tAHX/8ACUzf3z+dH/CUzf3z+dchuPrRuPrQBu63rUmo2aQsxIEgb9CP61m6XdGy1KK4U4KZ/UEf1qoSTSdKAOw/4Smb++fzo/4Smb++fzrkNx9aNx9aAOv/AOEpm/vn86P+Epm/vn865DcfWjcfWgDr/wDhKZv75/Oj/hKZv75/OuQ3H1o3H1oAlu5TPeTzHrJIzfmc1vaXr8llpsVurEBM/qSf61zdLkjvQB1//CUzf3z+dH/CUzf3z+dchuPrRuPrQB1//CUzf3z+dH/CUzf3z+dchuPrRuPrQB1//CUzf3z+dZut61JqNmkLMSBIG/Qj+tYW4+tBJNAFvS7o2WpRXCnBTP6gj+tdJ/wlM398/nXH9KXcfWgDr/8AhKZv75/Oj/hKZv75/OuQ3H1o3H1oA6//AISmb++fzo/4Smb++fzrkNx9aNx9aAOv/wCEpm/vn865W7lM95PMeskjN+ZzUW4+tJQAUUUUAaSahDB4dksIVcXFzOHuHIGPLQfIoOc8sWJ/3U9K1/EXiW31bSzbi4v7uV7hJ1N4iD7KAjBo4yp5DFgScL9wcZJrlqKAJba4e1uEnjWJnQ5AliWRfxVgQfxFRsxdyxxknJwAB+Q6UlFABRRRQB6F4Z8ZTaT4ftbJZCFi38Z9XJ/rWt/wsOf/AJ7H868pDsBgE4o8x/7xoA9W/wCFhz/89j+dH/Cw5/8AnsfzrynzH/vGjzH/ALxoA9W/4WHP/wA9j+dH/Cw5/wDnsfzrynzH/vGjzH/vGgDtfGPiqXXNIitnkLBZxJjPorD+tcRSlmYYJzSUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQB/9k=",
            "image/png": "iVBORw0KGgoAAAANSUhEUgAAAhwAAAJYCAIAAACmVahRAAAuUElEQVR4Ae3da2xcZXo48Npx7nEcWhvbGKUr5SJQ15XaL1FXVivoYiSybJcslSxVkcWl+YL2Q2rFdVU3H7r+kLrEwkTVwlJt5EVdAmRlrpYXCem/kAqxbKkobkVaEG1AviQGbHLBufr/dmc6djyJ1zNzxn49+fmDeeeZ8z7neX9j8mjOOXPmN37DDwECBAgQIECAAAECBAgQIECAAAECBAgQIECAAAECBAgQILBoAmWLtqecdlRWFmlh09PTOS3ExgQIELihBCriXG15eXmchV2+fDnOwlRFgACBGARmmsqDDz74wAMPVFZWtre3v/baa1VVVX19fdXV1ePj462trZOTk/lF8ltkRcVMYfllKNIsTaVIsNISIFAaAumjTDU1NUePHr3jjju2b9/e399/++23d3d3j46O9vT0tLW1hWc7Ojryi+THtHHjxvwmFnvWl19+WexdyE+AAIHlK5BuKrfddltjY+Pzzz+/fv36jz/++Oabbx4aGmpubh4eHm5oaBgcHAzP5hfJj6auri6/icWeFRptsXchPwECBJavQPoo0we/+gnLuP/++19++eUwqK2tTf0DOjIyEsZ5R5YvjcoJECBAIFeBq05dbNmyZd++feEgWK5Zfu32XV1dTU1Nvb294djar904bHDu3LmFbGYbAgQIEIhKYKapbNiw4dlnn33ooYdOnToVShwbGwvHoMLhr/r6+jDOO5JabWdnZ07LPn/+fE7b25gAAQIEYhBIX7kbPhcSrvU6ePDg22+/nSprYGCgpaUljMPvMA6D/CL5LfJSrD/5LccsAgQI3FgC4WLiM2fO/L9f/bzyyith8eEC4hdeeOHYsWPhdxjnHcnPMTS5OH/yW45ZBAgQIECAAAECBAgQIECAAAECBAgQIECAAAECBAgQIECAAAECBAgQIECAAAECBAgQIECAAAECBAgQIECAAAECBAgQIECAAAECBAgQIECAAAECBAgQIECAAIEEBdLf/JhgxkRShbtJJpIn8STT09OJ55SQAAECJSMw830qUS2poiLSwi5evBgVlGIIECAQlcDMv92bNm167LHH7rvvvsyN7sM3rFRXV4+Pj7e2tk5OToZ4HpH8Vrty5cr8JhZ7lqZSbGH5CRBY1gLpL+kKawhfTf8v//IvmcM7f/3Xf/3GG2+E7wB+8803/+qv/ipskF8kP501sf7ktxyzCBAgcIMIzJy6CF8ePDo6OjExEd6yhMUPDQ01NzeHrxNuaGgYHBxsbGzML5Kf46233prfxGLP+vTTT4u9C/kJECCwfAVm3qmEjjJ7GbW1tanIyMhIGIen8ovMzmlMgAABAqUtMHNOpajr7OrqCkfSent7+/v7F7Kjc+fOLWQz2xAgQIBAVALXbSpjY2PhgFg4/FVfXx/Goej8IqnVdnZ25rTsqampnLa3MQECBAjEIDBz+GtONQMDAy0tLSEYfodxGOQXmZN2gQ/DRVZx/iywfpsRIECAwP8KhBP1KYhwAfELL7xw7Nix8DtzkXEekVS2XH+HDz/G+ZPrQmxPgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAIEqBsiir+o3y8vI4C7ty5UqchamKAAECMQhUxFBEdg0VFZEWduHChexqRQgQIEAgJRDpv92rVq2K8xXSVOJ8XVRFgEAkAummcuTIkbq6ulBT+Nd8+/bt1dXVVVVVfX19YTA+Pt7a2jo5ObmQSFKrWrNmTVKpks1z5syZZBPKRoAAgVISmHtO5eGHH968efP+/fu7u7tHR0d7enra2tpqamo6OjoWEkmK5rd/+7eTSpVsnv/5n/9JNqFsBAgQKCWBq5pKWVnZu+++e/fdd588eXJoaKi5uXl4eLihoWFwcLCxsXEhkaRoNJWkJOUhQIDAYgpcdU7l3nvvfeedd0JHCRXU1taGdyphMDIyEsYLjCRV+tmzZ5NKJQ8BAgQILJrAVU0lHOnas2dPUvvu6upqamrq7e3t7+/PNefU1FSuU2xPgAABAksuMNNUduzYMTExcfz48VRNY2Nj4dR9OPxVX18fxiG4kMjs9XR2ds5+mNPYRVY5cdmYAAECkQjMfMawvb394MGDmbIGBgZaWlrCw/A7jMNgIZHM9AIHl2L9KXBdphMgQOCGENi6detbb701e6nhAuIXXnjh2LFj4XcYh6cWEpmdwZgAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIEEhYoSzhfQunKyiItbHp6OqElSkOAAIESFKiIc00VFZEWdvHixTjFVEWAAIEYBCL9t3vNmjUx6GTXoKlkm4gQIEAgIxBpU6mqqsqUGNXg9OnTUdWjGAIECEQlEGlTqa+vj4opU8ynn36aGRsQIECAwByBdFO56667nnzyyRMnToSnjx071tnZGd4r9PX1VVdXj4+Pt7a2Tk5OLiQyJ3veD7ds2ZL33KJOfOedd4qaX3ICBAgsa4H0RVa7d+9ev379E088kVlMd3f36OhoT09PW1tbTU1NR0fHQiKZ6QUO/vZv/7bADEWavn///iJllpYAAQIlIFCeWkM43DQyMjJ7Pffcc8+RI0dCJPzeuXNnGCwkMjuDMQECBAjcaALpw1+hqWzdunXfvn2ff/753r17P/roo9ra2vBOJXCEZhPGYbCQSFJ8//Ef/5FUKnkIECBAYNEE0k0lfKbvvffe27Nnz65du5566qk777yz8Aq6urqampp6e3v7+/tzzRa6Wq5TbE+AAAECSy6QbiqPP/74J598Eqp58cUXwxn7MBgbG6urqxseHg5vYsJ4gZHZ6wln+2c/zGk851hcTnNtTIAAAQJLJZBuKgcOHHjmmWdCR9mxY8f7778fqhkYGGhpaQkn6sPvMF5gJKllhIvNkkolDwECBAgsmkD66q9t27YdPnz40qVLU1NTjzzySDj6tJALiLO3SarulStXJpUq2Tw+UZ+sp2wECJSYQKT3bXRDyRL7O7McAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIEIhRwF2Kc3tVwldk5jbB1gQIELiRBNJf0hXbksvLy2MrKVXP5cuX4yxMVQQIEIhBINKmEu2XdGkqMfzVqoEAgWgFIm0qq1evjpMsfDNmnIWpigABAjEIzG0q99xzz09/+tO1a9dmf1XwQiJJLSkUkFSqZPNMTk4mm1A2AgQIlJLAVSfqKysrf/azn33961/fuHFjd3f36OhoT09PW1tbTU1NR0fHQiJJ0WzdujWpVMnm+fDDD5NNKBsBAgRKSeCq8+EHDhx47LHHrly5ElYY3rIcOXIkDMLvnTt3LjBSSjTWQoAAAQK5Cswc/mpqarrllluee+65H/7whyFLbW1teKcSBiMjI2G8wEiuu7/e9mfOnLneU+IECBAgEK1AuqmEE+OPPvrorl27Eiy0q6srNKre3t7+/v5c03711Ve5TrE9AQIECCy5QLqpfPe73w0nVH7yk5+EgjZs2PDjH/94bGysrq5ueHi4vr4+jEN8IZHZ6+ns7Jz9MKfx+fPnc9rexgQIECAQqcDExESoLJyW/4u/+IswCCfq/+7v/m6BkbBZIj8rYv1JZHWSECBAoFQFrrr6K7XI0FQ2bdq0kAuIs7dJiqms7BqFJZW8kDxu01KInrkECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIBA6QmUxbmk8vLyOAu7cuVKnIWpigABAjEIVMRQRHYNmkq2iQgBAgTiF4i0qaxcuTJOu0uXLsVZmKoIECAQg0CkTWXNmjUx6GTX8NVXX2UHRQgQIEAgJZBuKuvWrTt8+HBtbW1lZeX+/ftfffXVqqqqvr6+6urq8fHx1tbWycnJhUSSYl27dm1SqZLN88UXXySbUDYCBAiUkkD6RH17e3s4Bf3oo4/W19e/9dZbX/va17q7u0dHR3t6etra2mpqajo6OhYSSYpm+/btSaVKNs9//ud/JptQNgIECJSSQLqp3HTTTWfPnr1w4cI3v/nNH/zgB9u2bRsaGmpubh4eHm5oaBgcHGxsbFxIJCkaTSUpSXkIECCwmALpw1+pozpPP/30rl27vv3tb4cKwqGw8E4lDEZGRsJ4gZGkSj9z5kxSqeQhQIAAgUUTuOpE/e7du48ePRrOoLz++uuFV9DV1dXU1NTb29vf359rNufDcxWzPQECBGIQSDeVQ4cO7d27N1wv+8orr/zoRz8KlY2NjdXV1YXDX+EsSxgvMDJ7SZ2dnbMf5jSemprKaXsbEyBAgEAMAukProcru77zne+Egr7xjW8cP348DAYGBlpaWsIg/A7jBUbCZon8XIz1J5HVSUKAAIFSFUifqN+8eXO4gDh8jj2cq//e9773wQcfLOQC4uxtkmLyifqkJOUhQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgEKlAWZx1lZVFWtj09HScYqoiQIBADAIVMRSRXYOmkm0iQoAAgfgFIm0qFRWRFnbhwoX4X1QVEiBAYKkEIv23e+XKlUslMv9+NZX5fTxLgMANLjDTVB588MEHHnigsrKyvb39tddeq6qq6uvrq66uHh8fb21tnZycXEgkKc01a9YklSrZPGfPnk02oWwECBAoJYH0+fCampqjR4/ecccd27dv7+/vv/3227u7u0dHR3t6etra2sKzHR0dC4kkRbN58+akUiWb58SJE8kmlI0AAQKlJJBuKrfddltjY+Pzzz+/fv36jz/++Oabbx4aGmpubh4eHm5oaBgcHAzPLiSSFM3v/M7vJJUq2Tz//u//nmxC2QgQIFBKAunDXx/86ics7P7773/55ZfDoLa2NrxTCYORkZEwXmAkbOaHAAECBG5YgZlzKoFgy5Yt+/btCwfBEuHo6upqamrq7e0Nx9NyTXj69Olcp9ieAAECBJZcYKapbNiw4dlnn33ooYdOnToVyhobG6urqwuHv+rr68N4gZHZ6+ns7Jz9MKex8+E5cdmYAAECkQikm0r4sGG41uvgwYNvv/12qrKBgYGWlpZwoj78DuMQXEgkqVVNTU0llUoeAgQIEFg0gfSJ+nAx8aFDh375y1+GHZ85c+Zb3/rWQi4gzt4mqbpXrVqVVKpk8/icSrKeshEgUGIC6aYS26rKy8tjKylVz5UrV+IsTFUECBCIQWDmnEoM1WRqcN/GDIUBAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIEIhUoi7OusrJIC5ueno5TTFUECBCIQaAihiKyaygvL88OxhC5fPlyDGWogQABAnEKRNpUKioiLUxTifPvWFUECEQiEOm/3evWrYsEaE4Z58+fnxPxkAABAgQyApE2laqqqkyJUQ2++OKLqOpRDAECBKISiLSp1NfXR8WUKea///u/M2MDAgQIEJgjEGlT2bp165xCI3n41ltvRVKJMggQIBChwExT2bRp02OPPXbfffelDj3dddddTz755IkTJ0LRx44d6+zsDPG+vr7q6urx8fHW1tbJycnsSFIr/N3f/d2kUslDgAABAosmMPNxkDfffPO55577/ve/H7pL2P3u3bvXr1//xBNPZErp7u4eHR3t6elpa2urqanp6OjIjmQ2LnDw85//vMAMRZr+R3/0R0XKLC0BAgRKQGDm4yB/+qd/eujQocySwlmNkZGRzMMwuOeee44cORIG4ffOnTuvGZm9vTEBAgQI3GgCM4e/wruQ2YsPTSWc2Ni3b9/nn3++d+/ejz76qLa2NrVNaDZhHDbOjszOUMj4F7/4RSHTzSVAgACBJRGYaSpzdh/uR/Lee+/t2bNn165dTz311J133jlng1/7sKurq6mpqbe3t7+//9duPGeDf/u3f5sT8ZAAAQIE4he4blN5/PHHP/nkk7CAF198MZyxD4OxsbG6urrh4eHwJiaMrxmZveBwbn/2w5zGH374YU7b25gAAQIEYhC4blM5cODAM888EzrKjh073n///VDrwMBAS0tLOFEffofxNSNJLWnO6Zyk0spDgAABAkUVmLn6K7WbiYmJ1NVf27ZtO3z48KVLl6amph555JFwTiX7AuLsSFK13nTTTUmlSjaPT9Qn6ykbAQIlJjC3qUSyvNWrV0dSyZwy3PtrDoiHBAgQmC1w3cNfszda/HF4h7T4O7VHAgQIEChQINJ3Kr6kq8DX1XQCBAgQIECAAAECBAgQIECAAAECBAgQIECAAAECBAgQIECAAAECBAgQIECAAAECBAgQIECAAAECBAgQIECAAAECBAgQIECAAAECBAgQIECAAAECBAgQIECAAAECBAgQIECAAAECBAgQIECAAAECBAgQIECAAAECBAgQIECAAAECBAgQIECAAAECBAgQIECAAAECBAgQIECAAAECBAgQIECAAAECBAgQIECAAAECBAgQIECAAAECBAgQIECAAAECBAgQIECAAAECBAgQIECAAAECBAgQIECAAAECBAgQIECAAAECBAgQIECAAAECBAgQIECAAAECBAgQIECAAAECBAgQIECAAAECBAgQIECAAAECBAgQIECAAAECBAgQIECAAAECBAgQIECAAAECBAgQIECAAAECBAgQIECAAAECBAgQIECAAAECBAgQIECAAAECBAgQIECAAAECBAgQIECAQPICZcmnTCJjWVmkhU1PTyexPjkIECBQmgIVcS6rvLw8zsIuX74cZ2GqIkCAQAwCkTaVFStWxKCTXYOmkm0iQoAAgYxApE1l1apVmRKjGly4cCGqehRDgACBqAQibSpr166NiilTzJkzZzJjAwIECBCYIzDTVDZt2vTYY4/dd999VVVVYaPwu6+vr7q6enx8vLW1dXJyciGROdnzflhZWZn33KJOPHXqVFHzS06AAIFlLTBzkdWbb7753HPPff/73w/dJSypu7t7dHS0p6enra2tpqamo6NjIZGkLH7v934vqVTJ5vnXf/3XZBPKRoAAgVISmGkqdXV1oYtMTEykmsrQ0FBzc/Pw8HBDQ8Pg4GBjY+NCIknR/PEf/3FSqZLN8/rrryebUDYCBAiUksDMlbuho8xeWG1tbSoyMjISxuGphURmZzAmQIAAgRtNYOacSuIr7+rqampq6u3t7e/vzzX5559/nusU2xMgQIDAkgtct6mMjY2FA2Lh8Fd9fX0Yh0IXEpm9ns7OztkPcxqH6wJy2t7GBAgQIBCDwHWbysDAQEtLSzhRH36Hcah1IZGklnT69OmkUslDgAABAosmMHOiPrXLzIn6hVxAnL1NUnVv2LAhqVTJ5vE5lWQ9ZSNAoMQE5jaVSJbnE/WRvBDKIECAQE4C1z38lVOWxDd2i63ESSUkQIDAIghE+k7Fre8X4bW3CwIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBC4kQTK4lxseXl5nIVduXIlzsJURYAAgRgEKmIoIrsGTSXbRIQAAQLxC0TaVFauXBmn3aVLl+IsTFUECBCIQSDSprJ69eoYdLJr+Oqrr7KDIgQIECCQEoi0qaxbty7OV2hiYiLOwlRFgACBGAQibSobN26MQSe7huHh4eygCAECBAikBNJN5ciRI3V1dSG0atWq7du3V1dXV1VV9fX1hcH4+Hhra+vk5GR2pHiImzZtKl5ymQkQIECgSAJzLyl++OGHN2/evH///u7u7tHR0Z6enra2tpqamo6OjuxIkWoKaXfu3Fm85IVkfvXVVwuZbi4BAgRKW+CqplJWVvbuu+/efffdJ0+eHBoaam5uDkd7GhoaBgcHGxsbsyPFo9FUimcrMwECBIoncNU5lXvvvfedd94JHSXsr7a2NrxTCYORkZEwvmakeGV99tlnxUsuMwECBAgUSeCqphKOdO3Zs6eQPXV1dTU1NfX29vb39xeSx0VWheiZS4AAgaUSmGkqO3bsCP+UHz9+PFXK2NhYOHUfDn/V19eHcQhmR7KL7uzszA7mEfnyyy/zmGUKAQIECCytwExTaW9vP3jwYKaagYGBlpaWcKI+/A7jEM+OZDZOfHDu3LnEc0pIgAABAsUWSJ+o37p169NPP/0Hf/AHmf1lX0CcHclsnPhg7dq1iedMJKFP1CfCKAkBAqUqcNXVX/EssqJi5i1UPFWFStz7K6qXQzEECMQmEOm/3e4wH9sfinoIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgMANJFAW51rLyiItbHp6Ok4xVREgQCAGgYoYisiuYcWKFdnBGCKXLl2KoQw1ECBAIE6BSJvKqlWr4vTSVOJ8XVRFgEAkApE2lfXr10cCNKeMc+fOzYl4SIAAAQIZgUibyqZNmzIlRjU4depUVPUohgABAlEJRNpU6uvro2LKFPNf//VfmbEBAQIECMwRiLSpbNmyZU6hkTx84403IqlEGQQIEIhQIN1U7rrrrieffPLEiROhxGPHjnV2dlZVVfX19VVXV4+Pj7e2tk5OTmZHireexsbG4iWXmQABAgSKJJD+OMju3bvDufEnnngis5vu7u7R0dGenp62traampqOjo7sSGbjxAehsSWeM5GETU1NieSRhAABAqUs0N7e/id/8iezVzg0NHTLLbeESENDw/vvvx8G2ZHZ2yc7Dp8xjPMn2WXKRoAAgRITKE+tJ5wY37lzZ3h/8NJLL6XOZ9TW1oZ3KuHZkZGRMA6D7EiJWVgOAQIECBQokD6nEt4WvPfee3v27Nm1a9dTTz1155135pe3q6srHCDq7e3t7+/PL0Nq1j//8z8XMt1cAgQIEFgSgXRTefzxxz/55JNQwYsvvhjO2IfB2NhYXV3d8PBweBMTxteMZFcczvBnB/OI/OIXv8hjlikECBAgsLQC6aZy4MCBZ555JnSUHTt2pM6gDAwMtLS0hBP14XcYhyqzI8UrPVVD8fLLTIAAAQLFEEhf/bVt27bDhw+HG1tNTU098sgjH330UfYFxNmRYhSUyvmHf/iHxUteSGafUylEz1wCBEpeINI7zIcmFye9T9TH+bqoigCBSATSh78iqSZTxsTERGZsQIAAAQLLRSDSpnL27NnlIqhOAgQIEMgIRNpULly4kCnRgAABAgSWi0Ck51R8nfBy+QNSJwECBAgQIECAAAECBAgQIECAAAECBAgQIECAAAECBAgQIECAAAECBAgQIECAAAECBAgQIECAAAECBAgQIECAAAECBAgQIECAAIHSF3Cbltxe4/C9y7lNsDUBAgRuJIFIbyi5YsWKOF+F8D1mcRamKgIECMQgEGlTqaiItDBNJYa/WjUQIBCtQKT/dq9atSpOsvB1y3EWpioCBAjEIBBpU1m7dm0MOtk1fPnll9lBEQIECBBICUTaVCorK+N8hcbGxuIsTFUECBCIQWBuU7nnnnt++tOfhjcKVVVVfX191dXV4+Pjra2tk5OT2ZHiLSDsq3jJZSZAgACBIglcdUlxeH/ws5/97Otf//rGjRu7u7tHR0d7enra2tpqamo6OjqyI0WqKaRtbm4uXvJCMr/22muFTDeXAAECpS1wVVP5h3/4h5///Oc//OEPN23aNDQ0FP5lHx4ebmhoGBwcbGxszI4Uj+bP/uzPipe8kMz/9E//VMh0cwkQIFDaAuWZ5TU1Nd1yyy3PPfdcKlJbWxveqYTxyMhIGIdBdiQz14AAAQIECASB9DmV1atXP/roo7t27SoQpaurKzSn3t7e/v7+QlKdOnWqkOnmEiBAgMCSCKSbyne/+91wQuUnP/lJKGLDhg0//vGPw2VOdXV14fBXfX196pKn7Eh2xZ2dndnBPCKfffZZHrNMIUCAAIGlFbjqnEqqlImJiXBOZfZp+Ztvvvkv//IvsyPFK33r1q3FS15I5g8//LCQ6eYSIECgtAWu21SyLyDOjhSPJnUWp3j5886cetOW93QTCRAgUNoC12gqMSw4XNMcQxnZNfhEfbaJCAECBDICcz/8mHliaQcXLlxY2gLsnQABAgTyEIi0qbgZcB6vpSkECBBYcoFID3+VlUVamC/pWvI/WQUQIECAAAECBAgQIECAAAECBAgQIECAAAECBAgQIECAAAECBAgQIECAAAECBAgQIECAAAECBAgQIECAAAECBAgQIECAAAECBAgQIECAAAECBAgQIECAAAECBAgQIECAAAECBAgQIBCfQKR3mC8vL4/P6n8runLlSpyFqYoAAQIxCET6JV0rVqyIQSe7Bk0l20SEAAECGYFIm8rKlSszJUY1uHjxYlT1KIYAAQJRCUTaVFavXh0VU6aYc+fOZcYGBAgQIDBHINKmsn79+jmFRvLwiy++iKQSZRAgQCBCgUibysaNGyPEUhIBAgQIzC+Qbirr1q07fPhwbW1tZWXl/v37X3311aqqqr6+vurq6vHx8dbW1snJyezI/KkLefamm24qZLq5BAgQILAkAulLitvb28N1TY8++mh9ff1bb731ta99rbu7e3R0tKenp62traampqOjIztSvIq//e1vFy95IZlfeumlQqabS4AAgdIWSDeV8M7g7NmzFy5c+OY3v/mDH/xg27ZtQ0NDzc3Nw8PDDQ0Ng4ODjY2N2ZHi0Tz88MPFS15I5n/8x38sZLq5BAgQKG2B9OGv1Pnnp59+eteuXal3CeFQWHinEhY/MjISxmGQHSltGqsjQIAAgVwFrjpRv3v37qNHj4YzKK+//nquiVLbd3V1NTU19fb29vf355chNevkyZOFTDeXAAECBJZEIN1UDh06tHfv3kuXLr3yyis/+tGPQiljY2N1dXXh8Fc4yxLG14xkV9zZ2ZkdzCPy2Wef5THLFAIECBBYWoF0UwlXdn3nO98Jb1O+8Y1vHD9+PNQ0MDDQ0tISTtSH32F8zUjxSvdxkOLZykyAAIHiCaRP1G/evDlcQBxu4xjO1X/ve9/74IMPsi8gzo4Ur6xbb721eMkLyfzpp58WMt1cAgQIlLZAuqnEtshoP6fiLVRsfyrqIUAgKoGrTtTHU9n58+fjKUYlBAgQILBAgUibipsBL/D1sxkBAgSiEoi0qVy+fDkqJsUQIECAAAECBAgQIECAAAECBAgQIECAAAECBAgQIECAAAECBAgQIECAAAECBAgQIECAAAECBAgQIECAAAECBAgQIECAAAECBAgQIECAAAECBAgQIECAAAECBAgQIECAAAECBAgQIECAAAECBAgQIECAAAECBAgQIECAAAECN6RAWZyrLiuLtLDp6ek4xVRFgACBGAQi/TrhFStWxKCTXcOlS5eygyIECBAgkBKItKlUVERamKbi/xwCBAjMIxDpv92rV6+ep+glfGpqamoJ927XBAgQiFwg0qaydu3aOOEmJyfjLExVBAgQiEEg0qayYcOGGHTUQIAAAQI5CUTaVDZt2pTTMmxMgAABAjEIzDSVBx988IEHHqisrGxvb3/ttdeqqqr6+vqqq6vHx8dbW1vDYZ/sSPEW8Fu/9VvFSy4zAQIECBRJIP1xkJqamqNHj95xxx3bt2/v7++//fbbu7u7R0dHe3p62trawrMdHR3ZkSLVFNLu3r27eMkLyfz0008XMt1cAgQIlLZAuqncdtttjY2Nzz///Pr16z/++OObb755aGioubl5eHi4oaFhcHAwPJsdKR7Nvn37ipe8kMx///d/X8h0cwkQIFDaAunDXx/86ics9f7773/55ZfDoLa2NrxTCYORkZEwvmYkBP0QIECAAIGMwMw5lRDasmVLeIsQDoJlns510NXV1dTU1NvbG46h5Tp39vapfjY7YkyAAAEC8QvMNJVwFe+zzz770EMPnTp1KtQ9NjZWV1cXDn/V19eH8TUj2cvr7OzMDuYROXnyZB6zlnZKeXn50hZwvb1fuXLlek+JEyBAIFmBdFMJN3AM13odPHjw7bffTu1gYGCgpaUlnKgPv8M4BLMjyZYyO9tnn302++GyGEd7vzJNZVn8/SiSQGkIpE/Uh4uJDx069Mtf/jKs6syZM9/61reyLyDOjhSPYOvWrcVLXkjmDz/88HrT161bd72nljZ+7ty5pS3A3gkQuHEE0k0ltgWHI2+xlZSqZ56TPTfddFOcNX/xxRdxFqYqAgRKT2DmnEpUa/vqq6+iqmchxUT7TkVTWcjLZxsCBBIRiLSpnD9/PpHlLWaScDOCxdydfREgQCBCgUibynL82hL3K4vw71tJBAgsskCkTeXy5cuLDFH47tyvrHBDGQgQWO4CkTaV5fhV8OEOacv9r0H9BAgQKFAg0qZS4KqWZHrqZjZLsms7JUCAQCQCmkokL8TSlLEc7wKg5gT/Vub5YCxnzvkJaCr5uV1jVupmNtd4IuLQcrwLgJoT/IOap6lw5pyfgKaSn9s1ZqXumXaNJyIOVVRE+gdw8eLF67Gp+XoyecQ554GWx5QSc55fINJ/U+YvOs5nl+P9ylatWhUn5jyfflVzgi8Z5wQx50lVYs7zrDQ8panM75PDsxMTEzlsHcema9eujaOQuVWEr6+eG/q/x2r+P4kE/ss5AcQFpCgx5/lXrKnM75PDs6dPn85h6zg2DV/0GUchOVSh5hywCtiUcwF4OUxdjs7zL09Tmd8nh2eX482AN27cmMMK49hUzYvzOnDmnJ+AppKf2zVmLcf7lS3HW8uo+Rp/fEUIcS4C6jVSLkfnayxjVkhTmYVR2HCeCzwKS1zE2b/5m79ZxOzFSa3m4rjOzcp5rkhxHi9H5/klNJX5fXJ41v3KcsAqYNPleI81NRfwgucwlXMOWEXbVFNJjHaez5Eltg+JCBAgELeAphL361Pk6pbjZ2vUXOQ/inR6zpzzE9BU8nMrkVmff/75sluJmhfnJePMOT8BTSU/txKZtRw/sKnmxfnj48w5PwFNJT+3Epn15ZdfLruVqHlxXjLOnPMT0FTycyuRWWfPnl12K1Hz4rxknDnnJxBpU1m5cmV+6yn2rOX4YZR5TOa5z908s5b2KTUvjj9nzvkJRNpUVq9end96ij2rxJrKhQsXii2WeH41J056zYScr8mSeHA5Os+PEGlTifZOtGfOnJkfdHk9e+nSpeVVcKhWzYvzknHmnJ9ApE2lsrIyv/UUe9Zy/CaueUyW410A1DzPC5rgU5wTxJwn1XJ0nmc54alIm0rp3WRt/pdhqZ5djncBUPPi/LVw5pyfQKRNZTnewye/F8AsAgQIlJJApE2ltra2lJSthQABAjeIQLqpHDlypK6uLqw5fAH49u3bq6ur77rrrieffPLEiRMheOzYsc7OzsUUqa+vX8zd2RcBAgQIJCKQbiotLS2pdA8//PDmzZvDOPSY7u7uJ554IpHd5Jrk1ltvzXWK7QkQIEBgyQWuOvxVVlb2yCOP3H333aGs8F7h+PHjS1VfqrEt1d7tlwABAgTyEyifPe3ee+995513Tp48GYKhqezcuTMc+HrppZe2bNkyezNjAgQIECBwTYGr3qm0tbXt2bMntd309PR7770XHu7ateupp5668847rzk/BLu6upqamnp7e/v7+6+3Ta7x1LmcXGfZngABAgSWVmCmqezYsSPc7DpzyOvxxx//5JNPQnEvvvhiOGM/T5XFOIf/6aefzrNHTxEgQIBAnAIzTaW9vf3gwYOZKg8cOPDMM8+EjhKazfvvv5+JL85gZGRkcXZkLwQIECCQoEC6qWzduvWWW2554403Mqn/5m/+5vDhw3v37p2amvrzP//zTHxxBmNjY4uzI3shQIAAgQQFyhLMlWCq3//9308wW4Kp3n333QSzSUWAAIESE5g5/BXVwpbjV5lGBagYAgQILIlApE3l9OnTS8JhpwQIECBQiECkTWU5futcIS+DuQQIECgNgUibyvnz50vD1yoIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQWTeD/A/cQE2Ny7zrmAAAAAElFTkSuQmCC",
            "text/plain": [
              "<PIL.Image.Image image mode=RGB size=540x600>"
            ]
          },
          "execution_count": 14,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "simulate_gel_electrophoresis(list_of_amplicons)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 15,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>Amplicon Name</th>\n",
              "      <th>5' Overhang</th>\n",
              "      <th>5' Duplicate</th>\n",
              "      <th>3' Overhang</th>\n",
              "      <th>3' Duplicate</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>Amplicon_SCO5090_140_pcr_products</td>\n",
              "      <td>CATG</td>\n",
              "      <td>No</td>\n",
              "      <td>GCGC</td>\n",
              "      <td>No</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>Amplicon_SCO5089_167_pcr_products</td>\n",
              "      <td>GCGC</td>\n",
              "      <td>No</td>\n",
              "      <td>CGCC</td>\n",
              "      <td>No</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>Amplicon_SCO5090_174_pcr_products</td>\n",
              "      <td>CGCC</td>\n",
              "      <td>No</td>\n",
              "      <td>GGGA</td>\n",
              "      <td>No</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>Amplicon_SCO5090_208_pcr_products</td>\n",
              "      <td>GGGA</td>\n",
              "      <td>No</td>\n",
              "      <td>GAGC</td>\n",
              "      <td>No</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>Amplicon_SCO5087_210_pcr_products</td>\n",
              "      <td>GAGC</td>\n",
              "      <td>No</td>\n",
              "      <td>GTCG</td>\n",
              "      <td>No</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>5</th>\n",
              "      <td>Amplicon_SCO5087_522_pcr_products</td>\n",
              "      <td>GTCG</td>\n",
              "      <td>No</td>\n",
              "      <td>GACC</td>\n",
              "      <td>No</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>6</th>\n",
              "      <td>Amplicon_SCO5090_640_pcr_products</td>\n",
              "      <td>GACC</td>\n",
              "      <td>No</td>\n",
              "      <td>CTAG</td>\n",
              "      <td>No</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "                       Amplicon Name 5' Overhang 5' Duplicate 3' Overhang  \\\n",
              "0  Amplicon_SCO5090_140_pcr_products        CATG           No        GCGC   \n",
              "1  Amplicon_SCO5089_167_pcr_products        GCGC           No        CGCC   \n",
              "2  Amplicon_SCO5090_174_pcr_products        CGCC           No        GGGA   \n",
              "3  Amplicon_SCO5090_208_pcr_products        GGGA           No        GAGC   \n",
              "4  Amplicon_SCO5087_210_pcr_products        GAGC           No        GTCG   \n",
              "5  Amplicon_SCO5087_522_pcr_products        GTCG           No        GACC   \n",
              "6  Amplicon_SCO5090_640_pcr_products        GACC           No        CTAG   \n",
              "\n",
              "  3' Duplicate  \n",
              "0           No  \n",
              "1           No  \n",
              "2           No  \n",
              "3           No  \n",
              "4           No  \n",
              "5           No  \n",
              "6           No  "
            ]
          },
          "execution_count": 15,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "overhangs = create_overhang_dataframe(list_of_amplicons)\n",
        "overhangs"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 16,
      "metadata": {},
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "Dseqrecord(-138)\n",
            "CATGGTTCACTGCCGTATAGGCAGCTAAGAAACACGTACAGGTCCTGGAGCG\u001b[48;5;11mGTTTTAGAGCTAGAAATAGCAAGTTAAAATAAGGC\u001b[0mTAGTCCGTTATCAACTTGAAAAAGTGGCACCGAGTCGGTGCTTTTTT    \n",
            "    CAAGTGACGGCATATCCGTCGATTCTTTGTGCATGTCCAGGACCTCGCCAAAATCTCGATCTTTATCGTTCAATTTTATTCCGATCAGGCAATAGTTGAACTTTTTCACCGTGGCTCAGCCACGAAAAAACGCG\n",
            "Dseqrecord(-106)\n",
            "GCGCGACTCGAGAGCCGGTA\u001b[48;5;11mGTTTTAGAGCTAGAAATAGCAAGTTAAAATAAGGC\u001b[0mTAGTCCGTTATCAACTTGAAAAAGTGGCACCGAGTCGGTGCTTTTTT    \n",
            "    CTGAGCTCTCGGCCATCAAAATCTCGATCTTTATCGTTCAATTTTATTCCGATCAGGCAATAGTTGAACTTTTTCACCGTGGCTCAGCCACGAAAAAAGCGG\n",
            "Dseqrecord(-106)\n",
            "CGCCCAGATCCGGAAGCGTT\u001b[48;5;11mGTTTTAGAGCTAGAAATAGCAAGTTAAAATAAGGC\u001b[0mTAGTCCGTTATCAACTTGAAAAAGTGGCACCGAGTCGGTGCTTTTTT    \n",
            "    GTCTAGGCCTTCGCAACAAAATCTCGATCTTTATCGTTCAATTTTATTCCGATCAGGCAATAGTTGAACTTTTTCACCGTGGCTCAGCCACGAAAAAACCCT\n",
            "Dseqrecord(-106)\n",
            "GGGACGTCCAGGTCTTCACC\u001b[48;5;11mGTTTTAGAGCTAGAAATAGCAAGTTAAAATAAGGC\u001b[0mTAGTCCGTTATCAACTTGAAAAAGTGGCACCGAGTCGGTGCTTTTTT    \n",
            "    GCAGGTCCAGAAGTGGCAAAATCTCGATCTTTATCGTTCAATTTTATTCCGATCAGGCAATAGTTGAACTTTTTCACCGTGGCTCAGCCACGAAAAAACTCG\n",
            "Dseqrecord(-106)\n",
            "GAGCAGTTCCCAGAACTGCC\u001b[48;5;11mGTTTTAGAGCTAGAAATAGCAAGTTAAAATAAGGC\u001b[0mTAGTCCGTTATCAACTTGAAAAAGTGGCACCGAGTCGGTGCTTTTTT    \n",
            "    TCAAGGGTCTTGACGGCAAAATCTCGATCTTTATCGTTCAATTTTATTCCGATCAGGCAATAGTTGAACTTTTTCACCGTGGCTCAGCCACGAAAAAACAGC\n",
            "Dseqrecord(-106)\n",
            "GTCGACCTCCCAGTCACGGC\u001b[48;5;11mGTTTTAGAGCTAGAAATAGCAAGTTAAAATAAGGC\u001b[0mTAGTCCGTTATCAACTTGAAAAAGTGGCACCGAGTCGGTGCTTTTTT    \n",
            "    TGGAGGGTCAGTGCCGCAAAATCTCGATCTTTATCGTTCAATTTTATTCCGATCAGGCAATAGTTGAACTTTTTCACCGTGGCTCAGCCACGAAAAAACTGG\n",
            "Dseqrecord(-106)\n",
            "GACCGTCCAGGACATGACCA\u001b[48;5;11mGTTTTAGAGCTAGAAATAGCAAGTTAAAATAAGGC\u001b[0mTAGTCCGTTATCAACTTGAAAAAGTGGCACCGAGTCGGTGCTTTTTT    \n",
            "    CAGGTCCTGTACTGGTCAAAATCTCGATCTTTATCGTTCAATTTTATTCCGATCAGGCAATAGTTGAACTTTTTCACCGTGGCTCAGCCACGAAAAAAGATC\n"
          ]
        }
      ],
      "source": [
        "digest_amplicons = digest_amplicons_w_BsaI(list_of_amplicons)\n",
        "for digest in digest_amplicons: \n",
        "    print(digest.figure())"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 17,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/plain": [
              "[Dseqrecord(-138),\n",
              " Dseqrecord(-106),\n",
              " Dseqrecord(-106),\n",
              " Dseqrecord(-106),\n",
              " Dseqrecord(-106),\n",
              " Dseqrecord(-106),\n",
              " Dseqrecord(-106)]"
            ]
          },
          "execution_count": 17,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "digest_amplicons"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 18,
      "metadata": {},
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "Successfully looped the plasmid.\n"
          ]
        },
        {
          "data": {
            "text/plain": [
              "'pCRISPR-mcBEST_w_sgRNAs'"
            ]
          },
          "execution_count": 18,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "from Bio.Restriction import NcoI, NheI\n",
        "\n",
        "try:\n",
        "    linear_plasmid, smaller_frag = sorted(clean_plasmid.cut(NcoI, NheI), key=lambda x: len(x), reverse=True)\n",
        "    for i in range(len(digest_amplicons)):\n",
        "        # Assuming digest_amplicons[i] is a Dseqrecord, concatenate Dseq sequences\n",
        "        linear_plasmid.seq += digest_amplicons[i].seq\n",
        "    \n",
        "    try:\n",
        "        rec_vec = linear_plasmid.looped()\n",
        "        print(\"Successfully looped the plasmid.\")\n",
        "    except Exception as e:\n",
        "        print(f\"An error occurred while trying to loop the plasmid: {e}\")\n",
        "        # Handle the error or provide an alternative action here\n",
        "        rec_vec = None  # Example response; adjust as needed\n",
        "\n",
        "except Exception as e:\n",
        "    print(f\"An error occurred during plasmid digestion or preparation: {e}\")\n",
        "    # Additional error handling or cleanup here\n",
        "    rec_vec = None  # Ensure rec_vec is defined even if an error occurs\n",
        "\n",
        "\n",
        "# Constructing a meaningful name, ID, and description for the assembled plasmid using user input\n",
        "plasmid_components = \"_\".join(genes_to_KO)  # Concatenates gene names from the user's list\n",
        "plasmid_purpose = \"pCRISPR-mcBEST_w_sgRNAs\"  # Example: purpose of the plasmid\n",
        "\n",
        "rec_vec.name = f'{plasmid_purpose}'#_{plasmid_components}'\n",
        "rec_vec.id = rec_vec.name  # Using the same value for ID as for name for simplicity\n",
        "rec_vec.description = f'CRISPR-mcBEST_plasmid {\", \".join(genes_to_KO)} for gene KNOCK-DOWN, assembled using StreptoAIM.'\n",
        "rec_vec.id"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 19,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/plain": [
              "Dseqrecord(o13210)"
            ]
          },
          "execution_count": 19,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "# annotate the plasmid correctly\n",
        "annotate_plasmid_with_sgrnas(rec_vec, filtered_df)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 20,
      "metadata": {},
      "outputs": [],
      "source": [
        "print_plasmid = False\n",
        "\n",
        "if print_plasmid: \n",
        "    rec_vec.write(f\"../../data/plasmids/sgRNA_plasmids_pCRISPR_mcBEST/{rec_vec.id}.gb\")\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 21,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>Locus Tag</th>\n",
              "      <th>f_primer_name</th>\n",
              "      <th>r_primer_name</th>\n",
              "      <th>f_primer_sequences(5-3)</th>\n",
              "      <th>r_primer_sequences(5-3)</th>\n",
              "      <th>f_tm</th>\n",
              "      <th>r_tm</th>\n",
              "      <th>ta</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>SCO5087</td>\n",
              "      <td>SCO5087_fwd_checking_primer</td>\n",
              "      <td>SCO5087_rev_checking_primer</td>\n",
              "      <td>GACGATTCGGCCCGTG</td>\n",
              "      <td>CAGGGCGTCCAGGC</td>\n",
              "      <td>59</td>\n",
              "      <td>58</td>\n",
              "      <td>61</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>SCO5089</td>\n",
              "      <td>SCO5089_fwd_checking_primer</td>\n",
              "      <td>SCO5089_rev_checking_primer</td>\n",
              "      <td>CGTACGGCGAACTCGC</td>\n",
              "      <td>GTCCCCGGACGGC</td>\n",
              "      <td>59</td>\n",
              "      <td>57</td>\n",
              "      <td>60</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>SCO5090</td>\n",
              "      <td>SCO5090_fwd_checking_primer</td>\n",
              "      <td>SCO5090_rev_checking_primer</td>\n",
              "      <td>GGCGGCGGCCC</td>\n",
              "      <td>GACGTCCCCCGCGAAC</td>\n",
              "      <td>57</td>\n",
              "      <td>62</td>\n",
              "      <td>61</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "  Locus Tag                f_primer_name                r_primer_name  \\\n",
              "0   SCO5087  SCO5087_fwd_checking_primer  SCO5087_rev_checking_primer   \n",
              "1   SCO5089  SCO5089_fwd_checking_primer  SCO5089_rev_checking_primer   \n",
              "2   SCO5090  SCO5090_fwd_checking_primer  SCO5090_rev_checking_primer   \n",
              "\n",
              "  f_primer_sequences(5-3) r_primer_sequences(5-3)  f_tm  r_tm  ta  \n",
              "0        GACGATTCGGCCCGTG          CAGGGCGTCCAGGC    59    58  61  \n",
              "1        CGTACGGCGAACTCGC           GTCCCCGGACGGC    59    57  60  \n",
              "2             GGCGGCGGCCC        GACGTCCCCCGCGAAC    57    62  61  "
            ]
          },
          "execution_count": 21,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "# Getting checking primers\n",
        "checking_primers_df = checking_primers(path_to_genome, genes_to_KO, \n",
        "                                       flanking_region=flanking_region,\n",
        "                                       target_tm = melting_temperature, \n",
        "                                        primer_concentration = primer_concentration, \n",
        "                                        polymerase = chosen_polymerase)\n",
        "checking_primers_df"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 22,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>Name</th>\n",
              "      <th>Sequence</th>\n",
              "      <th>Concentration</th>\n",
              "      <th>Purification</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>SCO5087_fwd_checking_primer</td>\n",
              "      <td>GACGATTCGGCCCGTG</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>SCO5089_fwd_checking_primer</td>\n",
              "      <td>CGTACGGCGAACTCGC</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>SCO5090_fwd_checking_primer</td>\n",
              "      <td>GGCGGCGGCCC</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>SCO5087_rev_checking_primer</td>\n",
              "      <td>CAGGGCGTCCAGGC</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>SCO5089_rev_checking_primer</td>\n",
              "      <td>GTCCCCGGACGGC</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>5</th>\n",
              "      <td>SCO5090_rev_checking_primer</td>\n",
              "      <td>GACGTCCCCCGCGAAC</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "                          Name          Sequence Concentration Purification\n",
              "0  SCO5087_fwd_checking_primer  GACGATTCGGCCCGTG          25nm          STD\n",
              "1  SCO5089_fwd_checking_primer  CGTACGGCGAACTCGC          25nm          STD\n",
              "2  SCO5090_fwd_checking_primer       GGCGGCGGCCC          25nm          STD\n",
              "3  SCO5087_rev_checking_primer    CAGGGCGTCCAGGC          25nm          STD\n",
              "4  SCO5089_rev_checking_primer     GTCCCCGGACGGC          25nm          STD\n",
              "5  SCO5090_rev_checking_primer  GACGTCCCCCGCGAAC          25nm          STD"
            ]
          },
          "execution_count": 22,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "# making the primers into dseqrecords\n",
        "checking_primers_df_idt = create_idt_order_dataframe(checking_primers_df)\n",
        "checking_primers_df_idt\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 23,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>Name</th>\n",
              "      <th>Sequence</th>\n",
              "      <th>Concentration</th>\n",
              "      <th>Purification</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>primer_23</td>\n",
              "      <td>GATCGGGTCTCCCATGGTTCACTGCCGTATAGGCAGCTAAGAAACA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>primer_25</td>\n",
              "      <td>GATCGGGTCTCCGCGCGACTCGAGAGCCGGTAGTTTTAGAGCTAGA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>primer_27</td>\n",
              "      <td>GATCGGGTCTCCCGCCCAGATCCGGAAGCGTTGTTTTAGAGCTAGA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>primer_29</td>\n",
              "      <td>GATCGGGTCTCCGGGACGTCCAGGTCTTCACCGTTTTAGAGCTAGA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>primer_31</td>\n",
              "      <td>GATCGGGTCTCCGAGCAGTTCCCAGAACTGCCGTTTTAGAGCTAGA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>5</th>\n",
              "      <td>primer_33</td>\n",
              "      <td>GATCGGGTCTCCGTCGACCTCCCAGTCACGGCGTTTTAGAGCTAGA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>6</th>\n",
              "      <td>primer_35</td>\n",
              "      <td>GATCGGGTCTCCGACCGTCCAGGACATGACCAGTTTTAGAGCTAGA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>7</th>\n",
              "      <td>primer_24</td>\n",
              "      <td>GATCAGGTCTCGGCGCAAAAAAGCACCGACTCGGT</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>8</th>\n",
              "      <td>primer_26</td>\n",
              "      <td>GATCAGGTCTCGGGCGAAAAAAGCACCGACTCGGT</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>9</th>\n",
              "      <td>primer_28</td>\n",
              "      <td>GATCAGGTCTCGTCCCAAAAAAGCACCGACTCGGT</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>10</th>\n",
              "      <td>primer_30</td>\n",
              "      <td>GATCAGGTCTCGGCTCAAAAAAGCACCGACTCGGT</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>11</th>\n",
              "      <td>primer_32</td>\n",
              "      <td>GATCAGGTCTCGCGACAAAAAAGCACCGACTCGGT</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>12</th>\n",
              "      <td>primer_34</td>\n",
              "      <td>GATCAGGTCTCGGGTCAAAAAAGCACCGACTCGGT</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>13</th>\n",
              "      <td>primer_36</td>\n",
              "      <td>GATCAGGTCTCGCTAGAAAAAAGCACCGACTCGGT</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>SCO5087_fwd_checking_primer</td>\n",
              "      <td>GACGATTCGGCCCGTG</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>SCO5089_fwd_checking_primer</td>\n",
              "      <td>CGTACGGCGAACTCGC</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>SCO5090_fwd_checking_primer</td>\n",
              "      <td>GGCGGCGGCCC</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>SCO5087_rev_checking_primer</td>\n",
              "      <td>CAGGGCGTCCAGGC</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>SCO5089_rev_checking_primer</td>\n",
              "      <td>GTCCCCGGACGGC</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>5</th>\n",
              "      <td>SCO5090_rev_checking_primer</td>\n",
              "      <td>GACGTCCCCCGCGAAC</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "                           Name  \\\n",
              "0                     primer_23   \n",
              "1                     primer_25   \n",
              "2                     primer_27   \n",
              "3                     primer_29   \n",
              "4                     primer_31   \n",
              "5                     primer_33   \n",
              "6                     primer_35   \n",
              "7                     primer_24   \n",
              "8                     primer_26   \n",
              "9                     primer_28   \n",
              "10                    primer_30   \n",
              "11                    primer_32   \n",
              "12                    primer_34   \n",
              "13                    primer_36   \n",
              "0   SCO5087_fwd_checking_primer   \n",
              "1   SCO5089_fwd_checking_primer   \n",
              "2   SCO5090_fwd_checking_primer   \n",
              "3   SCO5087_rev_checking_primer   \n",
              "4   SCO5089_rev_checking_primer   \n",
              "5   SCO5090_rev_checking_primer   \n",
              "\n",
              "                                             Sequence Concentration  \\\n",
              "0   GATCGGGTCTCCCATGGTTCACTGCCGTATAGGCAGCTAAGAAACA...          25nm   \n",
              "1   GATCGGGTCTCCGCGCGACTCGAGAGCCGGTAGTTTTAGAGCTAGA...          25nm   \n",
              "2   GATCGGGTCTCCCGCCCAGATCCGGAAGCGTTGTTTTAGAGCTAGA...          25nm   \n",
              "3   GATCGGGTCTCCGGGACGTCCAGGTCTTCACCGTTTTAGAGCTAGA...          25nm   \n",
              "4   GATCGGGTCTCCGAGCAGTTCCCAGAACTGCCGTTTTAGAGCTAGA...          25nm   \n",
              "5   GATCGGGTCTCCGTCGACCTCCCAGTCACGGCGTTTTAGAGCTAGA...          25nm   \n",
              "6   GATCGGGTCTCCGACCGTCCAGGACATGACCAGTTTTAGAGCTAGA...          25nm   \n",
              "7                 GATCAGGTCTCGGCGCAAAAAAGCACCGACTCGGT          25nm   \n",
              "8                 GATCAGGTCTCGGGCGAAAAAAGCACCGACTCGGT          25nm   \n",
              "9                 GATCAGGTCTCGTCCCAAAAAAGCACCGACTCGGT          25nm   \n",
              "10                GATCAGGTCTCGGCTCAAAAAAGCACCGACTCGGT          25nm   \n",
              "11                GATCAGGTCTCGCGACAAAAAAGCACCGACTCGGT          25nm   \n",
              "12                GATCAGGTCTCGGGTCAAAAAAGCACCGACTCGGT          25nm   \n",
              "13                GATCAGGTCTCGCTAGAAAAAAGCACCGACTCGGT          25nm   \n",
              "0                                    GACGATTCGGCCCGTG          25nm   \n",
              "1                                    CGTACGGCGAACTCGC          25nm   \n",
              "2                                         GGCGGCGGCCC          25nm   \n",
              "3                                      CAGGGCGTCCAGGC          25nm   \n",
              "4                                       GTCCCCGGACGGC          25nm   \n",
              "5                                    GACGTCCCCCGCGAAC          25nm   \n",
              "\n",
              "   Purification  \n",
              "0           STD  \n",
              "1           STD  \n",
              "2           STD  \n",
              "3           STD  \n",
              "4           STD  \n",
              "5           STD  \n",
              "6           STD  \n",
              "7           STD  \n",
              "8           STD  \n",
              "9           STD  \n",
              "10          STD  \n",
              "11          STD  \n",
              "12          STD  \n",
              "13          STD  \n",
              "0           STD  \n",
              "1           STD  \n",
              "2           STD  \n",
              "3           STD  \n",
              "4           STD  \n",
              "5           STD  "
            ]
          },
          "execution_count": 23,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "\n",
        "full_idt = pd.concat([idt_df, checking_primers_df_idt])\n",
        "full_idt"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 24,
      "metadata": {},
      "outputs": [],
      "source": [
        "import os\n",
        "import json\n",
        "import pandas as pd\n",
        "from datetime import datetime\n",
        "from Bio.SeqRecord import SeqRecord\n",
        "from Bio import SeqIO\n",
        "from pydna.dseqrecord import Dseqrecord\n",
        "\n",
        "\n",
        "def generate_project_directory_structure(\n",
        "    project_name,\n",
        "    input_files,\n",
        "    output_files,\n",
        "    input_values,\n",
        "    markdown_file_paths=None\n",
        "\n",
        "):\n",
        "    # Define the main project directory\n",
        "    project_dir = f\"./{project_name}\"\n",
        "    \n",
        "    # Define input and output directories\n",
        "    inputs_dir = os.path.join(project_dir, \"inputs\")\n",
        "    outputs_dir = os.path.join(project_dir, \"outputs\")\n",
        "    \n",
        "    # Create directories if they don't exist\n",
        "    os.makedirs(inputs_dir, exist_ok=True)\n",
        "    os.makedirs(outputs_dir, exist_ok=True)\n",
        "    \n",
        "    # Process and save input files\n",
        "    for input_file in input_files:\n",
        "        file_save_path = os.path.join(inputs_dir, input_file['name'])\n",
        "        \n",
        "        if isinstance(input_file['content'], list) and all(isinstance(seq, Dseqrecord) for seq in input_file['content']):\n",
        "            # Save each Dseqrecord in the list as a separate GenBank file\n",
        "            for i, seq in enumerate(input_file['content']):\n",
        "                seq_file_save_path = os.path.join(inputs_dir, f\"{input_file['name'].split('.')[0]}_{i}.gb\")\n",
        "                SeqIO.write(seq, seq_file_save_path, \"genbank\")\n",
        "        elif isinstance(input_file['content'], SeqRecord):\n",
        "            SeqIO.write(input_file['content'], file_save_path, \"genbank\")\n",
        "        else:\n",
        "            with open(file_save_path, 'w') as file:\n",
        "                file.write(input_file['content'])\n",
        "    \n",
        "    \n",
        "    # Save the input values as a JSON file\n",
        "    input_values_path = os.path.join(inputs_dir, \"input_values.json\")\n",
        "    \n",
        "    with open(input_values_path, 'w') as json_file:\n",
        "        json.dump(input_values, json_file, indent=4)\n",
        "    \n",
        "    # Process and save output files\n",
        "    for output_file in output_files:\n",
        "        file_save_path = os.path.join(outputs_dir, output_file['name'])\n",
        "        \n",
        "        if output_file['name'].endswith(\".csv\"):\n",
        "            if isinstance(output_file['content'], pd.DataFrame):\n",
        "                output_file['content'].to_csv(file_save_path, index=False)\n",
        "            else:\n",
        "                raise TypeError(f\"Expected a DataFrame for {output_file['name']}, but got {type(output_file['content'])}\")\n",
        "        elif output_file['name'].endswith(\".gb\"):\n",
        "            if isinstance(output_file['content'], list) and all(isinstance(record, SeqRecord) for record in output_file['content']):\n",
        "                # Save each SeqRecord in the list separately\n",
        "                for i, record in enumerate(output_file['content']):\n",
        "                    record_file_save_path = os.path.join(outputs_dir, f\"{record.id}_{i}.gb\")\n",
        "                    SeqIO.write(record, record_file_save_path, \"genbank\")\n",
        "            elif isinstance(output_file['content'], SeqRecord):\n",
        "                SeqIO.write(output_file['content'], file_save_path, \"genbank\")\n",
        "        else:\n",
        "            with open(file_save_path, 'w') as file:\n",
        "                file.write(output_file['content'])\n",
        "\n",
        "\n",
        "                    # Process and save markdown files from paths\n",
        "    if markdown_file_paths:\n",
        "        for md_file_path in markdown_file_paths:\n",
        "            md_file_name = os.path.basename(md_file_path)\n",
        "            md_file_save_path = os.path.join(outputs_dir, md_file_name)\n",
        "            with open(md_file_path, 'r') as file:\n",
        "                md_content = file.read()\n",
        "            with open(md_file_save_path, 'w') as file:\n",
        "                file.write(md_content)\n",
        "    \n",
        "    \n",
        "    print(f\"Project structure for '{project_name}' created successfully.\")\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "## Folder with all the generated I/O"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 25,
      "metadata": {},
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "Project structure for 'CRISPR_mcBEST_workflow_2024-06-28T09:13:37.723360' created successfully.\n"
          ]
        },
        {
          "name": "stderr",
          "output_type": "stream",
          "text": [
            "/Users/lucaslevassor/opt/anaconda3/envs/streptocad/lib/python3.11/site-packages/Bio/SeqIO/InsdcIO.py:728: BiopythonWarning: Increasing length of locus line to allow long name. This will result in fields that are not in usual positions.\n",
            "  warnings.warn(\n"
          ]
        }
      ],
      "source": [
        "\n",
        "# Example: \n",
        "\n",
        "input_files = [\n",
        "    {\"name\": \"input_genome.gb\", \"content\": genome},\n",
        "    {\"name\": \"input_plasmid.gb\", \"content\": clean_plasmid}\n",
        "]\n",
        "\n",
        "\n",
        "\n",
        "output_files = [\n",
        "    {\"name\": \"mcBEST_w_sgRNAs.gb\", \"content\": rec_vec},\n",
        "    {\"name\": \"overhang_df.csv\", \"content\": overhangs},\n",
        "    {\"name\": \"pcr_df.csv\", \"content\": primer_df},\n",
        "    {\"name\": \"full_idt.csv\", \"content\": full_idt},\n",
        "    {\"name\": \"mutated_sgrna_df.csv\", \"content\": mutated_sgrna_df},\n",
        "    {\"name\": \"filtered_df.csv\", \"content\": filtered_df}\n",
        "]\n",
        "\n",
        "input_values = {\n",
        "    \"genes_to_knockout\": ['SCO5087', 'SCO5089', 'SCO5090'],\n",
        "    \"filtering_metrics\": {\n",
        "        \"gc_upper\": gc_upper,\n",
        "        \"gc_lower\": gc_lower,\n",
        "        \"off_target_seed\": off_target_seed,\n",
        "        \"off_target_upper\": off_target_upper,\n",
        "        \"cas_type\": cas_type,\n",
        "        \"number_of_sgRNAs_per_group\": number_of_sgRNAs_per_group\n",
        "    },\n",
        "    \"polymerase_settings\": {\n",
        "        \"chosen_polymerase\": chosen_polymerase,\n",
        "        \"melting_temperature\": melting_temperature,\n",
        "        \"primer_concentration\": primer_concentration,\n",
        "        \"primer_number_increment\": primer_number_increment,\n",
        "        \"flanking_region\": flanking_region\n",
        "    }, # TODO might want to incorporate this to the app\n",
        "    \"overlapping_sequences\": {\n",
        "        \"restriction_overhang_f\": restriction_overhang_f,\n",
        "        \"restriction_overhang_r\": restriction_overhang_r,\n",
        "        \"backbone_overhang_f\": backbone_overhang_f,\n",
        "        \"backbone_overhang_r\": backbone_overhang_r,\n",
        "        \"cys4\": cys4,\n",
        "        \"sgRNA_handle_cys4_site\": str(sgRNA_handle_cys4_site[0].seq)\n",
        "    }\n",
        "}\n",
        "\n",
        "\n",
        "# Paths to Markdown files\n",
        "markdown_file_paths = [\n",
        "    \"../../protocols/conjugation_protcol.md\",\n",
        "    \"../../protocols/multi_target_crispr_plasmid_protcol.md\"\n",
        "\n",
        "]\n",
        "\n",
        "# Data and time\n",
        "timestamp = datetime.utcnow().isoformat()\n",
        "\n",
        "generate_project_directory_structure(\n",
        "    project_name=f\"CRISPR_mcBEST_workflow_{timestamp}\",\n",
        "    input_files=input_files,\n",
        "    output_files=output_files,\n",
        "    input_values=input_values,\n",
        "    markdown_file_paths=markdown_file_paths\n",
        ")\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {},
      "outputs": [],
      "source": []
    }
  ],
  "metadata": {
    "kernelspec": {
      "display_name": "constrain",
      "language": "python",
      "name": "python3"
    },
    "language_info": {
      "codemirror_mode": {
        "name": "ipython",
        "version": 3
      },
      "file_extension": ".py",
      "mimetype": "text/x-python",
      "name": "python",
      "nbconvert_exporter": "python",
      "pygments_lexer": "ipython3",
      "version": "3.11.9"
    }
  },
  "nbformat": 4,
  "nbformat_minor": 2
}
