{
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "## W3-CRISPR-mcBEST workflow\n",
        "\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "The workflow allows for the simultaneous targeted engineering of up to 28 individual situated bases in the genome of Streptomyces, as shown by Whitford et. al 2023, thereby enabling efficient single-step multi-target engineering13. Similar to workflow 2, this workflow employs the CRISPR-BEST methodology but instead of performing base-editing on a single gene the method incorporates multiple guide RNAs as an array into the same vector (Figure 4A). This significantly enhances the efficiency of high-throughput genetic studies. The pCRISPR-cBEST plasmid file can be downloaded on the page along with the genome file for S.coelicor (A3) and streptoAIM will generate your plasmids, primers, checking primers and analyze the overhangs the target genes(Figure 4, S6). \n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 1,
      "metadata": {},
      "outputs": [],
      "source": [
        "import sys\n",
        "import os\n",
        "from pydna.dseqrecord import Dseqrecord\n",
        "from teemi.build.PCR import primer_tm_neb\n",
        "import pandas as pd\n",
        "from datetime import datetime\n",
        "\n",
        "\n",
        "# Ensure the src directory is in the Python path\n",
        "project_root = os.path.abspath(os.path.join(os.getcwd(), '../../'))\n",
        "if project_root not in sys.path:\n",
        "    sys.path.insert(0, project_root)\n",
        "\n",
        "from streptocad.sequence_loading.sequence_loading import (\n",
        "    load_and_process_gene_sequences,\n",
        "    load_and_process_plasmid, \n",
        "    check_and_convert_input,\n",
        "    annotate_dseqrecord,\n",
        "    process_specified_gene_sequences_from_record)\n",
        "\n",
        "\n",
        "from streptocad.utils import polymerase_dict, dataframe_to_seqrecords, ProjectDirectory, extract_metadata_to_dataframe\n",
        "from streptocad.crispr.guideRNAcas3_9_12 import extract_sgRNAs, SgRNAargs\n",
        "from streptocad.crispr.crispr_best import (\n",
        "    identify_base_editing_sites,\n",
        "    filter_sgrnas_for_base_editing,\n",
        "    process_base_editing\n",
        ")\n",
        "\n",
        "from streptocad.cloning.golden_gate_cloning import (\n",
        "    GoldenGateCloning,\n",
        "    create_overhang_dataframe,\n",
        "    digest_amplicons_w_BsaI\n",
        ")\n",
        "\n",
        "from streptocad.primers.primer_analysis import analyze_primers_and_hairpins\n",
        "from streptocad.primers.primer_generation import create_idt_order_dataframe, primers_to_IDT\n",
        "from streptocad.wet_lab.gel_simulation import simulate_gel_electrophoresis\n",
        "from streptocad.cloning.plasmid_processing import annotate_plasmid_with_sgrnas\n",
        "from streptocad.primers.primer_generation import find_best_check_primers_from_genome\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "# Inputs\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 2,
      "metadata": {},
      "outputs": [],
      "source": [
        "# 1 Add genome of choice (genbank, fasta)\n",
        "path_to_genome = '../../data/genomes/Streptomyces_coelicolor_A3_chromosome.gb'\n",
        "genome = load_and_process_plasmid(path_to_genome)\n",
        "\n",
        "# 2 Add plasmid \n",
        "path_to_plasmid = '../../data/plasmids/pCRISPR-MCBE_Csy4_kasopGFP.gb'\n",
        "clean_plasmid = load_and_process_plasmid(path_to_plasmid)\n",
        "\n",
        "# 3 Choose genes to knock out (list)\n",
        "#genes_to_KO = ['SCO5087', 'SCO5089', 'SCO5090']\n",
        "genes_to_KO = ['80000-100000', '4000-7000', '9000-14000','15000-20000']\n",
        "\n",
        "\n",
        "#### Advanced settings ####\n",
        "# 4 Filtering metrics for sgRNAs\n",
        "gc_upper = 0.72\n",
        "gc_lower = 0.2\n",
        "off_target_seed = 13\n",
        "off_target_upper = 10\n",
        "cas_type='cas9'\n",
        "number_of_sgRNAs_per_group = 5\n",
        "\n",
        "# 5 Choose polymerase and target melting temperature\n",
        "chosen_polymerase = polymerase_dict['Phusion High-Fidelity DNA Polymerase (GC Buffer)']\n",
        "melting_temperature = 65\n",
        "primer_concentration = 0.4 \n",
        "primer_number_increment = 23\n",
        "flanking_region=500\n",
        "\n",
        "\n",
        "# 6 Choose overlapping sequences for our plasmid we can use the following\n",
        "restriction_overhang_f = \"GATCGggtctcc\"\n",
        "restriction_overhang_r = \"GATCAGGTCTCg\"\n",
        "backbone_overhang_f = \"cATG\"\n",
        "backbone_overhang_r = \"cTAG\"\n",
        "cys4 = \"gTTCACTGCCGTATAGGCAGCTAAGAAA\"\n",
        "sgRNA_handle_cys4_site = [Dseqrecord('GTTTTAGAGCTAGAAATAGCAAGTTAAAATAAGGCTAGTCCGTTATCAACTTGAAAAAGTGGCACCGAGTCGGTGCTTTTTT', name = 'sgRNA_handle_cys4')]"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "# Computation"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 3,
      "metadata": {},
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "True\n",
            "this is the target dict [{'chosen_region_1(80000-100000)': [80000, 100000]}, {'chosen_region_2(4000-7000)': [4000, 7000]}, {'chosen_region_3(9000-14000)': [9000, 14000]}, {'chosen_region_4(15000-20000)': [15000, 20000]}]\n",
            "chosen_region_1(80000-100000) [80000, 100000]\n",
            "chosen_region_2(4000-7000) [4000, 7000]\n",
            "chosen_region_3(9000-14000) [9000, 14000]\n",
            "chosen_region_4(15000-20000) [15000, 20000]\n"
          ]
        },
        {
          "data": {
            "text/plain": [
              "25828"
            ]
          },
          "execution_count": 3,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "target_dict, genes_to_KO, annotation_input = check_and_convert_input(genes_to_KO)\n",
        "\n",
        "print(annotation_input)\n",
        "if annotation_input == True:\n",
        "    genome = annotate_dseqrecord(genome, target_dict)\n",
        "\n",
        "\n",
        "len(genome.features)\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 4,
      "metadata": {},
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "sgRNA generated were outside the designated border in chosen_region_1(80000-100000). To incorporate this extent borders. Skipping to next locus tag.\n",
            "sgRNA generated were outside the designated border in chosen_region_1(80000-100000). To incorporate this extent borders. Skipping to next locus tag.\n",
            "sgRNA generated were outside the designated border in chosen_region_2(4000-7000). To incorporate this extent borders. Skipping to next locus tag.\n",
            "sgRNA generated were outside the designated border in chosen_region_2(4000-7000). To incorporate this extent borders. Skipping to next locus tag.\n",
            "sgRNA generated were outside the designated border in chosen_region_2(4000-7000). To incorporate this extent borders. Skipping to next locus tag.\n",
            "sgRNA generated were outside the designated border in chosen_region_2(4000-7000). To incorporate this extent borders. Skipping to next locus tag.\n",
            "sgRNA generated were outside the designated border in chosen_region_2(4000-7000). To incorporate this extent borders. Skipping to next locus tag.\n",
            "sgRNA generated were outside the designated border in chosen_region_2(4000-7000). To incorporate this extent borders. Skipping to next locus tag.\n",
            "sgRNA generated were outside the designated border in chosen_region_3(9000-14000). To incorporate this extent borders. Skipping to next locus tag.\n",
            "sgRNA generated were outside the designated border in chosen_region_3(9000-14000). To incorporate this extent borders. Skipping to next locus tag.\n",
            "sgRNA generated were outside the designated border in chosen_region_3(9000-14000). To incorporate this extent borders. Skipping to next locus tag.\n",
            "sgRNA generated were outside the designated border in chosen_region_3(9000-14000). To incorporate this extent borders. Skipping to next locus tag.\n",
            "sgRNA generated were outside the designated border in chosen_region_3(9000-14000). To incorporate this extent borders. Skipping to next locus tag.\n",
            "sgRNA generated were outside the designated border in chosen_region_4(15000-20000). To incorporate this extent borders. Skipping to next locus tag.\n",
            "sgRNA generated were outside the designated border in chosen_region_4(15000-20000). To incorporate this extent borders. Skipping to next locus tag.\n",
            "sgRNA generated were outside the designated border in chosen_region_4(15000-20000). To incorporate this extent borders. Skipping to next locus tag.\n",
            "sgRNA generated were outside the designated border in chosen_region_4(15000-20000). To incorporate this extent borders. Skipping to next locus tag.\n",
            "sgRNA generated were outside the designated border in chosen_region_4(15000-20000). To incorporate this extent borders. Skipping to next locus tag.\n",
            "sgRNA generated were outside the designated border in chosen_region_4(15000-20000). To incorporate this extent borders. Skipping to next locus tag.\n"
          ]
        },
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>strain_name</th>\n",
              "      <th>locus_tag</th>\n",
              "      <th>gene_loc</th>\n",
              "      <th>gene_strand</th>\n",
              "      <th>sgrna_strand</th>\n",
              "      <th>sgrna_loc</th>\n",
              "      <th>gc</th>\n",
              "      <th>pam</th>\n",
              "      <th>sgrna</th>\n",
              "      <th>sgrna_seed_sequence</th>\n",
              "      <th>off_target_count</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>3517</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>chosen_region_1(80000-100000)</td>\n",
              "      <td>80001</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>6574</td>\n",
              "      <td>0.60</td>\n",
              "      <td>TGG</td>\n",
              "      <td>GCCACATTCACTACGACCCC</td>\n",
              "      <td>TCACTACGACCCC</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2538</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>chosen_region_1(80000-100000)</td>\n",
              "      <td>80001</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>16046</td>\n",
              "      <td>0.65</td>\n",
              "      <td>GGG</td>\n",
              "      <td>CCCTGATCCGAGTGTGTGCC</td>\n",
              "      <td>CCGAGTGTGTGCC</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2539</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>chosen_region_1(80000-100000)</td>\n",
              "      <td>80001</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>16045</td>\n",
              "      <td>0.60</td>\n",
              "      <td>CGG</td>\n",
              "      <td>ACCCTGATCCGAGTGTGTGC</td>\n",
              "      <td>TCCGAGTGTGTGC</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2540</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>chosen_region_1(80000-100000)</td>\n",
              "      <td>80001</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>16018</td>\n",
              "      <td>0.55</td>\n",
              "      <td>GGG</td>\n",
              "      <td>ACGACAATCCGGTGATCAGC</td>\n",
              "      <td>TCCGGTGATCAGC</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2541</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>chosen_region_1(80000-100000)</td>\n",
              "      <td>80001</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>16017</td>\n",
              "      <td>0.50</td>\n",
              "      <td>CGG</td>\n",
              "      <td>TACGACAATCCGGTGATCAG</td>\n",
              "      <td>ATCCGGTGATCAG</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>...</th>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>191</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>chosen_region_1(80000-100000)</td>\n",
              "      <td>80001</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>1988</td>\n",
              "      <td>0.65</td>\n",
              "      <td>CGG</td>\n",
              "      <td>ATCCTCGCCGTGACCGTCAC</td>\n",
              "      <td>CCGTGACCGTCAC</td>\n",
              "      <td>7</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>172</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>chosen_region_1(80000-100000)</td>\n",
              "      <td>80001</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>1765</td>\n",
              "      <td>0.65</td>\n",
              "      <td>TGG</td>\n",
              "      <td>TGGTAGAGCGCAGCATCGGC</td>\n",
              "      <td>GCGCAGCATCGGC</td>\n",
              "      <td>8</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3954</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>chosen_region_1(80000-100000)</td>\n",
              "      <td>80001</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>2235</td>\n",
              "      <td>0.70</td>\n",
              "      <td>GGG</td>\n",
              "      <td>AGTAGACGGTGCCGTGGCCG</td>\n",
              "      <td>GGTGCCGTGGCCG</td>\n",
              "      <td>8</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>179</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>chosen_region_1(80000-100000)</td>\n",
              "      <td>80001</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>1842</td>\n",
              "      <td>0.70</td>\n",
              "      <td>AGG</td>\n",
              "      <td>CACCCTGGGCATCGACGTCG</td>\n",
              "      <td>GGCATCGACGTCG</td>\n",
              "      <td>8</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3996</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>chosen_region_1(80000-100000)</td>\n",
              "      <td>80001</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>1929</td>\n",
              "      <td>0.65</td>\n",
              "      <td>AGG</td>\n",
              "      <td>TCCTCTTGGCCTGGTCGAGG</td>\n",
              "      <td>GGCCTGGTCGAGG</td>\n",
              "      <td>9</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>4297 rows × 11 columns</p>\n",
              "</div>"
            ],
            "text/plain": [
              "      strain_name                      locus_tag  gene_loc  gene_strand  \\\n",
              "3517  NC_003888.3  chosen_region_1(80000-100000)     80001            1   \n",
              "2538  NC_003888.3  chosen_region_1(80000-100000)     80001            1   \n",
              "2539  NC_003888.3  chosen_region_1(80000-100000)     80001            1   \n",
              "2540  NC_003888.3  chosen_region_1(80000-100000)     80001            1   \n",
              "2541  NC_003888.3  chosen_region_1(80000-100000)     80001            1   \n",
              "...           ...                            ...       ...          ...   \n",
              "191   NC_003888.3  chosen_region_1(80000-100000)     80001            1   \n",
              "172   NC_003888.3  chosen_region_1(80000-100000)     80001            1   \n",
              "3954  NC_003888.3  chosen_region_1(80000-100000)     80001            1   \n",
              "179   NC_003888.3  chosen_region_1(80000-100000)     80001            1   \n",
              "3996  NC_003888.3  chosen_region_1(80000-100000)     80001            1   \n",
              "\n",
              "      sgrna_strand  sgrna_loc    gc  pam                 sgrna  \\\n",
              "3517             1       6574  0.60  TGG  GCCACATTCACTACGACCCC   \n",
              "2538             1      16046  0.65  GGG  CCCTGATCCGAGTGTGTGCC   \n",
              "2539             1      16045  0.60  CGG  ACCCTGATCCGAGTGTGTGC   \n",
              "2540             1      16018  0.55  GGG  ACGACAATCCGGTGATCAGC   \n",
              "2541             1      16017  0.50  CGG  TACGACAATCCGGTGATCAG   \n",
              "...            ...        ...   ...  ...                   ...   \n",
              "191             -1       1988  0.65  CGG  ATCCTCGCCGTGACCGTCAC   \n",
              "172             -1       1765  0.65  TGG  TGGTAGAGCGCAGCATCGGC   \n",
              "3954             1       2235  0.70  GGG  AGTAGACGGTGCCGTGGCCG   \n",
              "179             -1       1842  0.70  AGG  CACCCTGGGCATCGACGTCG   \n",
              "3996             1       1929  0.65  AGG  TCCTCTTGGCCTGGTCGAGG   \n",
              "\n",
              "     sgrna_seed_sequence  off_target_count  \n",
              "3517       TCACTACGACCCC                 0  \n",
              "2538       CCGAGTGTGTGCC                 0  \n",
              "2539       TCCGAGTGTGTGC                 0  \n",
              "2540       TCCGGTGATCAGC                 0  \n",
              "2541       ATCCGGTGATCAG                 0  \n",
              "...                  ...               ...  \n",
              "191        CCGTGACCGTCAC                 7  \n",
              "172        GCGCAGCATCGGC                 8  \n",
              "3954       GGTGCCGTGGCCG                 8  \n",
              "179        GGCATCGACGTCG                 8  \n",
              "3996       GGCCTGGTCGAGG                 9  \n",
              "\n",
              "[4297 rows x 11 columns]"
            ]
          },
          "execution_count": 4,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "\n",
        "# Initialize SgRNAargs with desired parameters\n",
        "args = SgRNAargs(genome, \n",
        "                genes_to_KO,\n",
        "                step=['find', 'filter'],\n",
        "                gc_upper = gc_upper,\n",
        "                gc_lower = gc_lower,\n",
        "                off_target_seed = off_target_seed,\n",
        "                off_target_upper = off_target_upper,\n",
        "                cas_type='cas9'\n",
        "                )\n",
        "\n",
        "sgrna_df = extract_sgRNAs(args)\n",
        "sgrna_df"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 5,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>strain_name</th>\n",
              "      <th>locus_tag</th>\n",
              "      <th>gene_loc</th>\n",
              "      <th>gene_strand</th>\n",
              "      <th>sgrna_strand</th>\n",
              "      <th>sgrna_loc</th>\n",
              "      <th>gc</th>\n",
              "      <th>pam</th>\n",
              "      <th>sgrna</th>\n",
              "      <th>sgrna_seed_sequence</th>\n",
              "      <th>off_target_count</th>\n",
              "      <th>editable_cytosines</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>3517</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>chosen_region_1(80000-100000)</td>\n",
              "      <td>80001</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>6574</td>\n",
              "      <td>0.60</td>\n",
              "      <td>TGG</td>\n",
              "      <td>GCCACATTCACTACGACCCC</td>\n",
              "      <td>TCACTACGACCCC</td>\n",
              "      <td>0</td>\n",
              "      <td>3,5,9</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2538</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>chosen_region_1(80000-100000)</td>\n",
              "      <td>80001</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>16046</td>\n",
              "      <td>0.65</td>\n",
              "      <td>GGG</td>\n",
              "      <td>CCCTGATCCGAGTGTGTGCC</td>\n",
              "      <td>CCGAGTGTGTGCC</td>\n",
              "      <td>0</td>\n",
              "      <td>3,8,9</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2539</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>chosen_region_1(80000-100000)</td>\n",
              "      <td>80001</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>16045</td>\n",
              "      <td>0.60</td>\n",
              "      <td>CGG</td>\n",
              "      <td>ACCCTGATCCGAGTGTGTGC</td>\n",
              "      <td>TCCGAGTGTGTGC</td>\n",
              "      <td>0</td>\n",
              "      <td>3,4,9,10</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2540</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>chosen_region_1(80000-100000)</td>\n",
              "      <td>80001</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>16018</td>\n",
              "      <td>0.55</td>\n",
              "      <td>GGG</td>\n",
              "      <td>ACGACAATCCGGTGATCAGC</td>\n",
              "      <td>TCCGGTGATCAGC</td>\n",
              "      <td>0</td>\n",
              "      <td>5,9,10</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2541</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>chosen_region_1(80000-100000)</td>\n",
              "      <td>80001</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>16017</td>\n",
              "      <td>0.50</td>\n",
              "      <td>CGG</td>\n",
              "      <td>TACGACAATCCGGTGATCAG</td>\n",
              "      <td>ATCCGGTGATCAG</td>\n",
              "      <td>0</td>\n",
              "      <td>3,6,10</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>...</th>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>191</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>chosen_region_1(80000-100000)</td>\n",
              "      <td>80001</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>1988</td>\n",
              "      <td>0.65</td>\n",
              "      <td>CGG</td>\n",
              "      <td>ATCCTCGCCGTGACCGTCAC</td>\n",
              "      <td>CCGTGACCGTCAC</td>\n",
              "      <td>7</td>\n",
              "      <td>3,4,6,8,9</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>172</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>chosen_region_1(80000-100000)</td>\n",
              "      <td>80001</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>1765</td>\n",
              "      <td>0.65</td>\n",
              "      <td>TGG</td>\n",
              "      <td>TGGTAGAGCGCAGCATCGGC</td>\n",
              "      <td>GCGCAGCATCGGC</td>\n",
              "      <td>8</td>\n",
              "      <td>9</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3954</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>chosen_region_1(80000-100000)</td>\n",
              "      <td>80001</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>2235</td>\n",
              "      <td>0.70</td>\n",
              "      <td>GGG</td>\n",
              "      <td>AGTAGACGGTGCCGTGGCCG</td>\n",
              "      <td>GGTGCCGTGGCCG</td>\n",
              "      <td>8</td>\n",
              "      <td>7</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>179</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>chosen_region_1(80000-100000)</td>\n",
              "      <td>80001</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>1842</td>\n",
              "      <td>0.70</td>\n",
              "      <td>AGG</td>\n",
              "      <td>CACCCTGGGCATCGACGTCG</td>\n",
              "      <td>GGCATCGACGTCG</td>\n",
              "      <td>8</td>\n",
              "      <td>3,4,5,10</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3996</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>chosen_region_1(80000-100000)</td>\n",
              "      <td>80001</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>1929</td>\n",
              "      <td>0.65</td>\n",
              "      <td>AGG</td>\n",
              "      <td>TCCTCTTGGCCTGGTCGAGG</td>\n",
              "      <td>GGCCTGGTCGAGG</td>\n",
              "      <td>9</td>\n",
              "      <td>3,5,10</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>4115 rows × 12 columns</p>\n",
              "</div>"
            ],
            "text/plain": [
              "      strain_name                      locus_tag  gene_loc  gene_strand  \\\n",
              "3517  NC_003888.3  chosen_region_1(80000-100000)     80001            1   \n",
              "2538  NC_003888.3  chosen_region_1(80000-100000)     80001            1   \n",
              "2539  NC_003888.3  chosen_region_1(80000-100000)     80001            1   \n",
              "2540  NC_003888.3  chosen_region_1(80000-100000)     80001            1   \n",
              "2541  NC_003888.3  chosen_region_1(80000-100000)     80001            1   \n",
              "...           ...                            ...       ...          ...   \n",
              "191   NC_003888.3  chosen_region_1(80000-100000)     80001            1   \n",
              "172   NC_003888.3  chosen_region_1(80000-100000)     80001            1   \n",
              "3954  NC_003888.3  chosen_region_1(80000-100000)     80001            1   \n",
              "179   NC_003888.3  chosen_region_1(80000-100000)     80001            1   \n",
              "3996  NC_003888.3  chosen_region_1(80000-100000)     80001            1   \n",
              "\n",
              "      sgrna_strand  sgrna_loc    gc  pam                 sgrna  \\\n",
              "3517             1       6574  0.60  TGG  GCCACATTCACTACGACCCC   \n",
              "2538             1      16046  0.65  GGG  CCCTGATCCGAGTGTGTGCC   \n",
              "2539             1      16045  0.60  CGG  ACCCTGATCCGAGTGTGTGC   \n",
              "2540             1      16018  0.55  GGG  ACGACAATCCGGTGATCAGC   \n",
              "2541             1      16017  0.50  CGG  TACGACAATCCGGTGATCAG   \n",
              "...            ...        ...   ...  ...                   ...   \n",
              "191             -1       1988  0.65  CGG  ATCCTCGCCGTGACCGTCAC   \n",
              "172             -1       1765  0.65  TGG  TGGTAGAGCGCAGCATCGGC   \n",
              "3954             1       2235  0.70  GGG  AGTAGACGGTGCCGTGGCCG   \n",
              "179             -1       1842  0.70  AGG  CACCCTGGGCATCGACGTCG   \n",
              "3996             1       1929  0.65  AGG  TCCTCTTGGCCTGGTCGAGG   \n",
              "\n",
              "     sgrna_seed_sequence  off_target_count editable_cytosines  \n",
              "3517       TCACTACGACCCC                 0              3,5,9  \n",
              "2538       CCGAGTGTGTGCC                 0              3,8,9  \n",
              "2539       TCCGAGTGTGTGC                 0           3,4,9,10  \n",
              "2540       TCCGGTGATCAGC                 0             5,9,10  \n",
              "2541       ATCCGGTGATCAG                 0             3,6,10  \n",
              "...                  ...               ...                ...  \n",
              "191        CCGTGACCGTCAC                 7          3,4,6,8,9  \n",
              "172        GCGCAGCATCGGC                 8                  9  \n",
              "3954       GGTGCCGTGGCCG                 8                  7  \n",
              "179        GGCATCGACGTCG                 8           3,4,5,10  \n",
              "3996       GGCCTGGTCGAGG                 9             3,5,10  \n",
              "\n",
              "[4115 rows x 12 columns]"
            ]
          },
          "execution_count": 5,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "\n",
        "gene_sequences = process_specified_gene_sequences_from_record(genome, genes_to_KO)\n",
        "genes_to_KO_dict = {locus_tag: gene_sequences[locus_tag] for locus_tag in genes_to_KO if locus_tag in gene_sequences}\n",
        "\n",
        "# Identify and annotate base editing sites\n",
        "sgrna_df_with_editing = identify_base_editing_sites(sgrna_df)\n",
        "\n",
        "filtered_sgrna_df_for_base_editing = filter_sgrnas_for_base_editing(sgrna_df_with_editing)\n",
        "filtered_sgrna_df_for_base_editing"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 6,
      "metadata": {},
      "outputs": [
        {
          "name": "stderr",
          "output_type": "stream",
          "text": [
            "/Users/lucaslevassor/opt/anaconda3/envs/streptocad/lib/python3.11/site-packages/Bio/Seq.py:3482: BiopythonWarning: Partial codon, len(sequence) not a multiple of three. Explicitly trim the sequence or add trailing N before translation. This may become an error in future.\n",
            "  warnings.warn(\n"
          ]
        },
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>strain_name</th>\n",
              "      <th>locus_tag</th>\n",
              "      <th>gene_loc</th>\n",
              "      <th>gene_strand</th>\n",
              "      <th>sgrna_strand</th>\n",
              "      <th>sgrna_loc</th>\n",
              "      <th>gc</th>\n",
              "      <th>pam</th>\n",
              "      <th>sgrna</th>\n",
              "      <th>sgrna_seed_sequence</th>\n",
              "      <th>off_target_count</th>\n",
              "      <th>editable_cytosines</th>\n",
              "      <th>mutations</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>3517</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>chosen_region_1(80000-100000)</td>\n",
              "      <td>80001</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>6574</td>\n",
              "      <td>0.60</td>\n",
              "      <td>TGG</td>\n",
              "      <td>GCCACATTCACTACGACCCC</td>\n",
              "      <td>TCACTACGACCCC</td>\n",
              "      <td>0</td>\n",
              "      <td>3,5,9</td>\n",
              "      <td>P2186L, H2187Y, S2188L</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2538</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>chosen_region_1(80000-100000)</td>\n",
              "      <td>80001</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>16046</td>\n",
              "      <td>0.65</td>\n",
              "      <td>GGG</td>\n",
              "      <td>CCCTGATCCGAGTGTGTGCC</td>\n",
              "      <td>CCGAGTGTGTGCC</td>\n",
              "      <td>0</td>\n",
              "      <td>3,8,9</td>\n",
              "      <td>S5345F</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2539</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>chosen_region_1(80000-100000)</td>\n",
              "      <td>80001</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>16045</td>\n",
              "      <td>0.60</td>\n",
              "      <td>CGG</td>\n",
              "      <td>ACCCTGATCCGAGTGTGTGC</td>\n",
              "      <td>TCCGAGTGTGTGC</td>\n",
              "      <td>0</td>\n",
              "      <td>3,4,9,10</td>\n",
              "      <td>P5343L, S5345F</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2540</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>chosen_region_1(80000-100000)</td>\n",
              "      <td>80001</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>16018</td>\n",
              "      <td>0.55</td>\n",
              "      <td>GGG</td>\n",
              "      <td>ACGACAATCCGGTGATCAGC</td>\n",
              "      <td>TCCGGTGATCAGC</td>\n",
              "      <td>0</td>\n",
              "      <td>5,9,10</td>\n",
              "      <td>Q5335*, S5336F</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2541</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>chosen_region_1(80000-100000)</td>\n",
              "      <td>80001</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>16017</td>\n",
              "      <td>0.50</td>\n",
              "      <td>CGG</td>\n",
              "      <td>TACGACAATCCGGTGATCAG</td>\n",
              "      <td>ATCCGGTGATCAG</td>\n",
              "      <td>0</td>\n",
              "      <td>3,6,10</td>\n",
              "      <td>R5334*, Q5335*, S5336F</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>...</th>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>191</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>chosen_region_1(80000-100000)</td>\n",
              "      <td>80001</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>1988</td>\n",
              "      <td>0.65</td>\n",
              "      <td>CGG</td>\n",
              "      <td>ATCCTCGCCGTGACCGTCAC</td>\n",
              "      <td>CCGTGACCGTCAC</td>\n",
              "      <td>7</td>\n",
              "      <td>3,4,6,8,9</td>\n",
              "      <td>A661T, R662K</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>172</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>chosen_region_1(80000-100000)</td>\n",
              "      <td>80001</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>1765</td>\n",
              "      <td>0.65</td>\n",
              "      <td>TGG</td>\n",
              "      <td>TGGTAGAGCGCAGCATCGGC</td>\n",
              "      <td>GCGCAGCATCGGC</td>\n",
              "      <td>8</td>\n",
              "      <td>9</td>\n",
              "      <td>R586H</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3954</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>chosen_region_1(80000-100000)</td>\n",
              "      <td>80001</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>2235</td>\n",
              "      <td>0.70</td>\n",
              "      <td>GGG</td>\n",
              "      <td>AGTAGACGGTGCCGTGGCCG</td>\n",
              "      <td>GGTGCCGTGGCCG</td>\n",
              "      <td>8</td>\n",
              "      <td>7</td>\n",
              "      <td>T741M</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>179</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>chosen_region_1(80000-100000)</td>\n",
              "      <td>80001</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>1842</td>\n",
              "      <td>0.70</td>\n",
              "      <td>AGG</td>\n",
              "      <td>CACCCTGGGCATCGACGTCG</td>\n",
              "      <td>GGCATCGACGTCG</td>\n",
              "      <td>8</td>\n",
              "      <td>3,4,5,10</td>\n",
              "      <td>M611I, R613K, V614M</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3996</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>chosen_region_1(80000-100000)</td>\n",
              "      <td>80001</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>1929</td>\n",
              "      <td>0.65</td>\n",
              "      <td>AGG</td>\n",
              "      <td>TCCTCTTGGCCTGGTCGAGG</td>\n",
              "      <td>GGCCTGGTCGAGG</td>\n",
              "      <td>9</td>\n",
              "      <td>3,5,10</td>\n",
              "      <td>L638F, A640V</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>3638 rows × 13 columns</p>\n",
              "</div>"
            ],
            "text/plain": [
              "      strain_name                      locus_tag  gene_loc  gene_strand  \\\n",
              "3517  NC_003888.3  chosen_region_1(80000-100000)     80001            1   \n",
              "2538  NC_003888.3  chosen_region_1(80000-100000)     80001            1   \n",
              "2539  NC_003888.3  chosen_region_1(80000-100000)     80001            1   \n",
              "2540  NC_003888.3  chosen_region_1(80000-100000)     80001            1   \n",
              "2541  NC_003888.3  chosen_region_1(80000-100000)     80001            1   \n",
              "...           ...                            ...       ...          ...   \n",
              "191   NC_003888.3  chosen_region_1(80000-100000)     80001            1   \n",
              "172   NC_003888.3  chosen_region_1(80000-100000)     80001            1   \n",
              "3954  NC_003888.3  chosen_region_1(80000-100000)     80001            1   \n",
              "179   NC_003888.3  chosen_region_1(80000-100000)     80001            1   \n",
              "3996  NC_003888.3  chosen_region_1(80000-100000)     80001            1   \n",
              "\n",
              "      sgrna_strand  sgrna_loc    gc  pam                 sgrna  \\\n",
              "3517             1       6574  0.60  TGG  GCCACATTCACTACGACCCC   \n",
              "2538             1      16046  0.65  GGG  CCCTGATCCGAGTGTGTGCC   \n",
              "2539             1      16045  0.60  CGG  ACCCTGATCCGAGTGTGTGC   \n",
              "2540             1      16018  0.55  GGG  ACGACAATCCGGTGATCAGC   \n",
              "2541             1      16017  0.50  CGG  TACGACAATCCGGTGATCAG   \n",
              "...            ...        ...   ...  ...                   ...   \n",
              "191             -1       1988  0.65  CGG  ATCCTCGCCGTGACCGTCAC   \n",
              "172             -1       1765  0.65  TGG  TGGTAGAGCGCAGCATCGGC   \n",
              "3954             1       2235  0.70  GGG  AGTAGACGGTGCCGTGGCCG   \n",
              "179             -1       1842  0.70  AGG  CACCCTGGGCATCGACGTCG   \n",
              "3996             1       1929  0.65  AGG  TCCTCTTGGCCTGGTCGAGG   \n",
              "\n",
              "     sgrna_seed_sequence  off_target_count editable_cytosines  \\\n",
              "3517       TCACTACGACCCC                 0              3,5,9   \n",
              "2538       CCGAGTGTGTGCC                 0              3,8,9   \n",
              "2539       TCCGAGTGTGTGC                 0           3,4,9,10   \n",
              "2540       TCCGGTGATCAGC                 0             5,9,10   \n",
              "2541       ATCCGGTGATCAG                 0             3,6,10   \n",
              "...                  ...               ...                ...   \n",
              "191        CCGTGACCGTCAC                 7          3,4,6,8,9   \n",
              "172        GCGCAGCATCGGC                 8                  9   \n",
              "3954       GGTGCCGTGGCCG                 8                  7   \n",
              "179        GGCATCGACGTCG                 8           3,4,5,10   \n",
              "3996       GGCCTGGTCGAGG                 9             3,5,10   \n",
              "\n",
              "                   mutations  \n",
              "3517  P2186L, H2187Y, S2188L  \n",
              "2538                  S5345F  \n",
              "2539          P5343L, S5345F  \n",
              "2540          Q5335*, S5336F  \n",
              "2541  R5334*, Q5335*, S5336F  \n",
              "...                      ...  \n",
              "191             A661T, R662K  \n",
              "172                    R586H  \n",
              "3954                   T741M  \n",
              "179      M611I, R613K, V614M  \n",
              "3996            L638F, A640V  \n",
              "\n",
              "[3638 rows x 13 columns]"
            ]
          },
          "execution_count": 6,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "\n",
        "# Process the DataFrame to apply C-to-T mutations\n",
        "mutated_sgrna_df = process_base_editing(filtered_sgrna_df_for_base_editing, genes_to_KO_dict, only_stop_codons = False)\n",
        "mutated_sgrna_df"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 7,
      "metadata": {},
      "outputs": [
        {
          "name": "stderr",
          "output_type": "stream",
          "text": [
            "/Users/lucaslevassor/opt/anaconda3/envs/streptocad/lib/python3.11/site-packages/Bio/Seq.py:3482: BiopythonWarning: Partial codon, len(sequence) not a multiple of three. Explicitly trim the sequence or add trailing N before translation. This may become an error in future.\n",
            "  warnings.warn(\n"
          ]
        },
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>strain_name</th>\n",
              "      <th>locus_tag</th>\n",
              "      <th>gene_loc</th>\n",
              "      <th>gene_strand</th>\n",
              "      <th>sgrna_strand</th>\n",
              "      <th>sgrna_loc</th>\n",
              "      <th>gc</th>\n",
              "      <th>pam</th>\n",
              "      <th>sgrna</th>\n",
              "      <th>sgrna_seed_sequence</th>\n",
              "      <th>off_target_count</th>\n",
              "      <th>editable_cytosines</th>\n",
              "      <th>mutations</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>4172</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>chosen_region_2(4000-7000)</td>\n",
              "      <td>4001</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>34</td>\n",
              "      <td>0.70</td>\n",
              "      <td>GGG</td>\n",
              "      <td>CCCAGACACCACGCACGCCA</td>\n",
              "      <td>ACCACGCACGCCA</td>\n",
              "      <td>1</td>\n",
              "      <td>3,7,9,10</td>\n",
              "      <td>G9N, V10I, W11*</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>5938</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>chosen_region_4(15000-20000)</td>\n",
              "      <td>15001</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>54</td>\n",
              "      <td>0.70</td>\n",
              "      <td>AGG</td>\n",
              "      <td>GCTCAACCTCCAGGCCGGAC</td>\n",
              "      <td>CTCCAGGCCGGAC</td>\n",
              "      <td>1</td>\n",
              "      <td>4,7,8,10</td>\n",
              "      <td>W15*, R16K</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4166</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>chosen_region_1(80000-100000)</td>\n",
              "      <td>80001</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>80</td>\n",
              "      <td>0.55</td>\n",
              "      <td>AGG</td>\n",
              "      <td>GTTGTCCGTCAGCACTTCGA</td>\n",
              "      <td>GTCAGCACTTCGA</td>\n",
              "      <td>0</td>\n",
              "      <td>6,7,10</td>\n",
              "      <td>R23C, Q24*</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4165</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>chosen_region_1(80000-100000)</td>\n",
              "      <td>80001</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>81</td>\n",
              "      <td>0.50</td>\n",
              "      <td>GGG</td>\n",
              "      <td>TTGTCCGTCAGCACTTCGAA</td>\n",
              "      <td>TCAGCACTTCGAA</td>\n",
              "      <td>1</td>\n",
              "      <td>5,6,9</td>\n",
              "      <td>R23C, Q24*</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>5932</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>chosen_region_3(9000-14000)</td>\n",
              "      <td>9001</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>118</td>\n",
              "      <td>0.65</td>\n",
              "      <td>CGG</td>\n",
              "      <td>TGCCCGATCTTCCAGGCCAG</td>\n",
              "      <td>TCTTCCAGGCCAG</td>\n",
              "      <td>1</td>\n",
              "      <td>3,4,5,9</td>\n",
              "      <td>A34V, R35*, S36F</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>...</th>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2142</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>chosen_region_1(80000-100000)</td>\n",
              "      <td>80001</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>19520</td>\n",
              "      <td>0.60</td>\n",
              "      <td>GGG</td>\n",
              "      <td>TCCCAGTCGAAGCCCTGGTA</td>\n",
              "      <td>CGAAGCCCTGGTA</td>\n",
              "      <td>0</td>\n",
              "      <td>3,4,8</td>\n",
              "      <td>Q6502*, S6503L</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2043</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>chosen_region_1(80000-100000)</td>\n",
              "      <td>80001</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>19524</td>\n",
              "      <td>0.65</td>\n",
              "      <td>GGG</td>\n",
              "      <td>GCCCTACCAGGGCTTCGACT</td>\n",
              "      <td>CAGGGCTTCGACT</td>\n",
              "      <td>0</td>\n",
              "      <td>3,4,7,8</td>\n",
              "      <td>W6506*, G6508S</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2044</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>chosen_region_1(80000-100000)</td>\n",
              "      <td>80001</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>19525</td>\n",
              "      <td>0.65</td>\n",
              "      <td>TGG</td>\n",
              "      <td>AGCCCTACCAGGGCTTCGAC</td>\n",
              "      <td>CCAGGGCTTCGAC</td>\n",
              "      <td>1</td>\n",
              "      <td>3,4,5,8,9</td>\n",
              "      <td>W6506*, G6508N</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2065</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>chosen_region_1(80000-100000)</td>\n",
              "      <td>80001</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>19714</td>\n",
              "      <td>0.55</td>\n",
              "      <td>AGG</td>\n",
              "      <td>GCCAGCATCTCGACTTCTTC</td>\n",
              "      <td>TCTCGACTTCTTC</td>\n",
              "      <td>0</td>\n",
              "      <td>3,6,9</td>\n",
              "      <td>R6569K, C6570Y, W6571*</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2068</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>chosen_region_1(80000-100000)</td>\n",
              "      <td>80001</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>19781</td>\n",
              "      <td>0.60</td>\n",
              "      <td>CGG</td>\n",
              "      <td>GTGCACCATAGCGGCAGGTT</td>\n",
              "      <td>ATAGCGGCAGGTT</td>\n",
              "      <td>0</td>\n",
              "      <td>4,6,7</td>\n",
              "      <td>W6592*, C6593Y</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>423 rows × 13 columns</p>\n",
              "</div>"
            ],
            "text/plain": [
              "      strain_name                      locus_tag  gene_loc  gene_strand  \\\n",
              "4172  NC_003888.3     chosen_region_2(4000-7000)      4001            1   \n",
              "5938  NC_003888.3   chosen_region_4(15000-20000)     15001            1   \n",
              "4166  NC_003888.3  chosen_region_1(80000-100000)     80001            1   \n",
              "4165  NC_003888.3  chosen_region_1(80000-100000)     80001            1   \n",
              "5932  NC_003888.3    chosen_region_3(9000-14000)      9001            1   \n",
              "...           ...                            ...       ...          ...   \n",
              "2142  NC_003888.3  chosen_region_1(80000-100000)     80001            1   \n",
              "2043  NC_003888.3  chosen_region_1(80000-100000)     80001            1   \n",
              "2044  NC_003888.3  chosen_region_1(80000-100000)     80001            1   \n",
              "2065  NC_003888.3  chosen_region_1(80000-100000)     80001            1   \n",
              "2068  NC_003888.3  chosen_region_1(80000-100000)     80001            1   \n",
              "\n",
              "      sgrna_strand  sgrna_loc    gc  pam                 sgrna  \\\n",
              "4172            -1         34  0.70  GGG  CCCAGACACCACGCACGCCA   \n",
              "5938            -1         54  0.70  AGG  GCTCAACCTCCAGGCCGGAC   \n",
              "4166             1         80  0.55  AGG  GTTGTCCGTCAGCACTTCGA   \n",
              "4165             1         81  0.50  GGG  TTGTCCGTCAGCACTTCGAA   \n",
              "5932             1        118  0.65  CGG  TGCCCGATCTTCCAGGCCAG   \n",
              "...            ...        ...   ...  ...                   ...   \n",
              "2142             1      19520  0.60  GGG  TCCCAGTCGAAGCCCTGGTA   \n",
              "2043            -1      19524  0.65  GGG  GCCCTACCAGGGCTTCGACT   \n",
              "2044            -1      19525  0.65  TGG  AGCCCTACCAGGGCTTCGAC   \n",
              "2065            -1      19714  0.55  AGG  GCCAGCATCTCGACTTCTTC   \n",
              "2068            -1      19781  0.60  CGG  GTGCACCATAGCGGCAGGTT   \n",
              "\n",
              "     sgrna_seed_sequence  off_target_count editable_cytosines  \\\n",
              "4172       ACCACGCACGCCA                 1           3,7,9,10   \n",
              "5938       CTCCAGGCCGGAC                 1           4,7,8,10   \n",
              "4166       GTCAGCACTTCGA                 0             6,7,10   \n",
              "4165       TCAGCACTTCGAA                 1              5,6,9   \n",
              "5932       TCTTCCAGGCCAG                 1            3,4,5,9   \n",
              "...                  ...               ...                ...   \n",
              "2142       CGAAGCCCTGGTA                 0              3,4,8   \n",
              "2043       CAGGGCTTCGACT                 0            3,4,7,8   \n",
              "2044       CCAGGGCTTCGAC                 1          3,4,5,8,9   \n",
              "2065       TCTCGACTTCTTC                 0              3,6,9   \n",
              "2068       ATAGCGGCAGGTT                 0              4,6,7   \n",
              "\n",
              "                   mutations  \n",
              "4172         G9N, V10I, W11*  \n",
              "5938              W15*, R16K  \n",
              "4166              R23C, Q24*  \n",
              "4165              R23C, Q24*  \n",
              "5932        A34V, R35*, S36F  \n",
              "...                      ...  \n",
              "2142          Q6502*, S6503L  \n",
              "2043          W6506*, G6508S  \n",
              "2044          W6506*, G6508N  \n",
              "2065  R6569K, C6570Y, W6571*  \n",
              "2068          W6592*, C6593Y  \n",
              "\n",
              "[423 rows x 13 columns]"
            ]
          },
          "execution_count": 7,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "# Process the DataFrame to apply C-to-T mutations\n",
        "mutated_sgrna_df = process_base_editing(filtered_sgrna_df_for_base_editing, \n",
        "                                        genes_to_KO_dict, \n",
        "                                        only_stop_codons = True)\n",
        "mutated_sgrna_df"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 8,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>strain_name</th>\n",
              "      <th>locus_tag</th>\n",
              "      <th>gene_loc</th>\n",
              "      <th>gene_strand</th>\n",
              "      <th>sgrna_strand</th>\n",
              "      <th>sgrna_loc</th>\n",
              "      <th>gc</th>\n",
              "      <th>pam</th>\n",
              "      <th>sgrna</th>\n",
              "      <th>sgrna_seed_sequence</th>\n",
              "      <th>off_target_count</th>\n",
              "      <th>editable_cytosines</th>\n",
              "      <th>mutations</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>4172</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>chosen_region_2(4000-7000)</td>\n",
              "      <td>4001</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>34</td>\n",
              "      <td>0.70</td>\n",
              "      <td>GGG</td>\n",
              "      <td>CCCAGACACCACGCACGCCA</td>\n",
              "      <td>ACCACGCACGCCA</td>\n",
              "      <td>1</td>\n",
              "      <td>3,7,9,10</td>\n",
              "      <td>G9N, V10I, W11*</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>5938</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>chosen_region_4(15000-20000)</td>\n",
              "      <td>15001</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>54</td>\n",
              "      <td>0.70</td>\n",
              "      <td>AGG</td>\n",
              "      <td>GCTCAACCTCCAGGCCGGAC</td>\n",
              "      <td>CTCCAGGCCGGAC</td>\n",
              "      <td>1</td>\n",
              "      <td>4,7,8,10</td>\n",
              "      <td>W15*, R16K</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4166</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>chosen_region_1(80000-100000)</td>\n",
              "      <td>80001</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>80</td>\n",
              "      <td>0.55</td>\n",
              "      <td>AGG</td>\n",
              "      <td>GTTGTCCGTCAGCACTTCGA</td>\n",
              "      <td>GTCAGCACTTCGA</td>\n",
              "      <td>0</td>\n",
              "      <td>6,7,10</td>\n",
              "      <td>R23C, Q24*</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4165</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>chosen_region_1(80000-100000)</td>\n",
              "      <td>80001</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>81</td>\n",
              "      <td>0.50</td>\n",
              "      <td>GGG</td>\n",
              "      <td>TTGTCCGTCAGCACTTCGAA</td>\n",
              "      <td>TCAGCACTTCGAA</td>\n",
              "      <td>1</td>\n",
              "      <td>5,6,9</td>\n",
              "      <td>R23C, Q24*</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>5932</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>chosen_region_3(9000-14000)</td>\n",
              "      <td>9001</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>118</td>\n",
              "      <td>0.65</td>\n",
              "      <td>CGG</td>\n",
              "      <td>TGCCCGATCTTCCAGGCCAG</td>\n",
              "      <td>TCTTCCAGGCCAG</td>\n",
              "      <td>1</td>\n",
              "      <td>3,4,5,9</td>\n",
              "      <td>A34V, R35*, S36F</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>5931</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>chosen_region_3(9000-14000)</td>\n",
              "      <td>9001</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>146</td>\n",
              "      <td>0.60</td>\n",
              "      <td>CGG</td>\n",
              "      <td>ATTCGACGCAGAGACCGCAC</td>\n",
              "      <td>GCAGAGACCGCAC</td>\n",
              "      <td>1</td>\n",
              "      <td>4,7,9</td>\n",
              "      <td>R44*, R45C</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4181</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>chosen_region_2(4000-7000)</td>\n",
              "      <td>4001</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>147</td>\n",
              "      <td>0.70</td>\n",
              "      <td>GGG</td>\n",
              "      <td>ATCCCAGGAAGGCAGCGCGC</td>\n",
              "      <td>GAAGGCAGCGCGC</td>\n",
              "      <td>1</td>\n",
              "      <td>3,4,5</td>\n",
              "      <td>W48*, D49N</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4182</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>chosen_region_2(4000-7000)</td>\n",
              "      <td>4001</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>148</td>\n",
              "      <td>0.70</td>\n",
              "      <td>CGG</td>\n",
              "      <td>CATCCCAGGAAGGCAGCGCG</td>\n",
              "      <td>GGAAGGCAGCGCG</td>\n",
              "      <td>1</td>\n",
              "      <td>4,5,6</td>\n",
              "      <td>W48*, D49N</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>7018</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>chosen_region_4(15000-20000)</td>\n",
              "      <td>15001</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>166</td>\n",
              "      <td>0.70</td>\n",
              "      <td>CGG</td>\n",
              "      <td>GCCACGAGCGGCATCCACTG</td>\n",
              "      <td>GCGGCATCCACTG</td>\n",
              "      <td>1</td>\n",
              "      <td>3,5,9</td>\n",
              "      <td>P50L, R51*, A52V</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>5926</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>chosen_region_3(9000-14000)</td>\n",
              "      <td>9001</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>200</td>\n",
              "      <td>0.60</td>\n",
              "      <td>CGG</td>\n",
              "      <td>CGACCTCATCAGCGTCGACA</td>\n",
              "      <td>ATCAGCGTCGACA</td>\n",
              "      <td>1</td>\n",
              "      <td>4,5,7,10</td>\n",
              "      <td>P62F, H63Y, Q64*</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>5925</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>chosen_region_3(9000-14000)</td>\n",
              "      <td>9001</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>203</td>\n",
              "      <td>0.65</td>\n",
              "      <td>CGG</td>\n",
              "      <td>CCTCATCAGCGTCGACACGG</td>\n",
              "      <td>AGCGTCGACACGG</td>\n",
              "      <td>1</td>\n",
              "      <td>4,7,10</td>\n",
              "      <td>H63Y, Q64*, R65C</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>7004</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>chosen_region_4(15000-20000)</td>\n",
              "      <td>15001</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>254</td>\n",
              "      <td>0.65</td>\n",
              "      <td>AGG</td>\n",
              "      <td>TCACGTCACCGACGGCGTAC</td>\n",
              "      <td>ACCGACGGCGTAC</td>\n",
              "      <td>1</td>\n",
              "      <td>4,7,9,10</td>\n",
              "      <td>R80C, H81Y, R82*</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4147</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>chosen_region_1(80000-100000)</td>\n",
              "      <td>80001</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>256</td>\n",
              "      <td>0.65</td>\n",
              "      <td>AGG</td>\n",
              "      <td>CGACCAGGTCCATCTGCCAC</td>\n",
              "      <td>GTCCATCTGCCAC</td>\n",
              "      <td>0</td>\n",
              "      <td>4,5,10</td>\n",
              "      <td>Q81*</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>7002</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>chosen_region_4(15000-20000)</td>\n",
              "      <td>15001</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>280</td>\n",
              "      <td>0.65</td>\n",
              "      <td>AGG</td>\n",
              "      <td>CGTACAGGTCGCGCATCTCC</td>\n",
              "      <td>GTCGCGCATCTCC</td>\n",
              "      <td>2</td>\n",
              "      <td>5,10</td>\n",
              "      <td>Q89*</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>7000</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>chosen_region_4(15000-20000)</td>\n",
              "      <td>15001</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>295</td>\n",
              "      <td>0.65</td>\n",
              "      <td>TGG</td>\n",
              "      <td>TCTCCAGGTGGACCGCCATG</td>\n",
              "      <td>GTGGACCGCCATG</td>\n",
              "      <td>1</td>\n",
              "      <td>4,5</td>\n",
              "      <td>Q94*</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4766</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>chosen_region_2(4000-7000)</td>\n",
              "      <td>4001</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>386</td>\n",
              "      <td>0.70</td>\n",
              "      <td>CGG</td>\n",
              "      <td>GAAGGACAGGTGGGCCTCGC</td>\n",
              "      <td>AGGTGGGCCTCGC</td>\n",
              "      <td>1</td>\n",
              "      <td>7</td>\n",
              "      <td>Q125*</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4765</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>chosen_region_2(4000-7000)</td>\n",
              "      <td>4001</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>387</td>\n",
              "      <td>0.70</td>\n",
              "      <td>GGG</td>\n",
              "      <td>AAGGACAGGTGGGCCTCGCC</td>\n",
              "      <td>GGTGGGCCTCGCC</td>\n",
              "      <td>1</td>\n",
              "      <td>6</td>\n",
              "      <td>Q125*</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>5898</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>chosen_region_3(9000-14000)</td>\n",
              "      <td>9001</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>424</td>\n",
              "      <td>0.70</td>\n",
              "      <td>TGG</td>\n",
              "      <td>TGATGCGCAGCCACAGGCGG</td>\n",
              "      <td>CAGCCACAGGCGG</td>\n",
              "      <td>1</td>\n",
              "      <td>6,8</td>\n",
              "      <td>A137V, Q138*</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4122</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>chosen_region_1(80000-100000)</td>\n",
              "      <td>80001</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>455</td>\n",
              "      <td>0.70</td>\n",
              "      <td>CGG</td>\n",
              "      <td>GCGGCGCAGCTCGCAGATCT</td>\n",
              "      <td>AGCTCGCAGATCT</td>\n",
              "      <td>0</td>\n",
              "      <td>5,7,10</td>\n",
              "      <td>A147V, Q148*, L149F</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4109</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>chosen_region_1(80000-100000)</td>\n",
              "      <td>80001</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>547</td>\n",
              "      <td>0.60</td>\n",
              "      <td>CGG</td>\n",
              "      <td>TACCCTCCAGCACGAACCTC</td>\n",
              "      <td>CAGCACGAACCTC</td>\n",
              "      <td>0</td>\n",
              "      <td>3,4,5,7,8</td>\n",
              "      <td>T177I, L178F, Q179*</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "      strain_name                      locus_tag  gene_loc  gene_strand  \\\n",
              "4172  NC_003888.3     chosen_region_2(4000-7000)      4001            1   \n",
              "5938  NC_003888.3   chosen_region_4(15000-20000)     15001            1   \n",
              "4166  NC_003888.3  chosen_region_1(80000-100000)     80001            1   \n",
              "4165  NC_003888.3  chosen_region_1(80000-100000)     80001            1   \n",
              "5932  NC_003888.3    chosen_region_3(9000-14000)      9001            1   \n",
              "5931  NC_003888.3    chosen_region_3(9000-14000)      9001            1   \n",
              "4181  NC_003888.3     chosen_region_2(4000-7000)      4001            1   \n",
              "4182  NC_003888.3     chosen_region_2(4000-7000)      4001            1   \n",
              "7018  NC_003888.3   chosen_region_4(15000-20000)     15001            1   \n",
              "5926  NC_003888.3    chosen_region_3(9000-14000)      9001            1   \n",
              "5925  NC_003888.3    chosen_region_3(9000-14000)      9001            1   \n",
              "7004  NC_003888.3   chosen_region_4(15000-20000)     15001            1   \n",
              "4147  NC_003888.3  chosen_region_1(80000-100000)     80001            1   \n",
              "7002  NC_003888.3   chosen_region_4(15000-20000)     15001            1   \n",
              "7000  NC_003888.3   chosen_region_4(15000-20000)     15001            1   \n",
              "4766  NC_003888.3     chosen_region_2(4000-7000)      4001            1   \n",
              "4765  NC_003888.3     chosen_region_2(4000-7000)      4001            1   \n",
              "5898  NC_003888.3    chosen_region_3(9000-14000)      9001            1   \n",
              "4122  NC_003888.3  chosen_region_1(80000-100000)     80001            1   \n",
              "4109  NC_003888.3  chosen_region_1(80000-100000)     80001            1   \n",
              "\n",
              "      sgrna_strand  sgrna_loc    gc  pam                 sgrna  \\\n",
              "4172            -1         34  0.70  GGG  CCCAGACACCACGCACGCCA   \n",
              "5938            -1         54  0.70  AGG  GCTCAACCTCCAGGCCGGAC   \n",
              "4166             1         80  0.55  AGG  GTTGTCCGTCAGCACTTCGA   \n",
              "4165             1         81  0.50  GGG  TTGTCCGTCAGCACTTCGAA   \n",
              "5932             1        118  0.65  CGG  TGCCCGATCTTCCAGGCCAG   \n",
              "5931             1        146  0.60  CGG  ATTCGACGCAGAGACCGCAC   \n",
              "4181            -1        147  0.70  GGG  ATCCCAGGAAGGCAGCGCGC   \n",
              "4182            -1        148  0.70  CGG  CATCCCAGGAAGGCAGCGCG   \n",
              "7018             1        166  0.70  CGG  GCCACGAGCGGCATCCACTG   \n",
              "5926             1        200  0.60  CGG  CGACCTCATCAGCGTCGACA   \n",
              "5925             1        203  0.65  CGG  CCTCATCAGCGTCGACACGG   \n",
              "7004             1        254  0.65  AGG  TCACGTCACCGACGGCGTAC   \n",
              "4147             1        256  0.65  AGG  CGACCAGGTCCATCTGCCAC   \n",
              "7002             1        280  0.65  AGG  CGTACAGGTCGCGCATCTCC   \n",
              "7000             1        295  0.65  TGG  TCTCCAGGTGGACCGCCATG   \n",
              "4766             1        386  0.70  CGG  GAAGGACAGGTGGGCCTCGC   \n",
              "4765             1        387  0.70  GGG  AAGGACAGGTGGGCCTCGCC   \n",
              "5898             1        424  0.70  TGG  TGATGCGCAGCCACAGGCGG   \n",
              "4122             1        455  0.70  CGG  GCGGCGCAGCTCGCAGATCT   \n",
              "4109             1        547  0.60  CGG  TACCCTCCAGCACGAACCTC   \n",
              "\n",
              "     sgrna_seed_sequence  off_target_count editable_cytosines  \\\n",
              "4172       ACCACGCACGCCA                 1           3,7,9,10   \n",
              "5938       CTCCAGGCCGGAC                 1           4,7,8,10   \n",
              "4166       GTCAGCACTTCGA                 0             6,7,10   \n",
              "4165       TCAGCACTTCGAA                 1              5,6,9   \n",
              "5932       TCTTCCAGGCCAG                 1            3,4,5,9   \n",
              "5931       GCAGAGACCGCAC                 1              4,7,9   \n",
              "4181       GAAGGCAGCGCGC                 1              3,4,5   \n",
              "4182       GGAAGGCAGCGCG                 1              4,5,6   \n",
              "7018       GCGGCATCCACTG                 1              3,5,9   \n",
              "5926       ATCAGCGTCGACA                 1           4,5,7,10   \n",
              "5925       AGCGTCGACACGG                 1             4,7,10   \n",
              "7004       ACCGACGGCGTAC                 1           4,7,9,10   \n",
              "4147       GTCCATCTGCCAC                 0             4,5,10   \n",
              "7002       GTCGCGCATCTCC                 2               5,10   \n",
              "7000       GTGGACCGCCATG                 1                4,5   \n",
              "4766       AGGTGGGCCTCGC                 1                  7   \n",
              "4765       GGTGGGCCTCGCC                 1                  6   \n",
              "5898       CAGCCACAGGCGG                 1                6,8   \n",
              "4122       AGCTCGCAGATCT                 0             5,7,10   \n",
              "4109       CAGCACGAACCTC                 0          3,4,5,7,8   \n",
              "\n",
              "                mutations  \n",
              "4172      G9N, V10I, W11*  \n",
              "5938           W15*, R16K  \n",
              "4166           R23C, Q24*  \n",
              "4165           R23C, Q24*  \n",
              "5932     A34V, R35*, S36F  \n",
              "5931           R44*, R45C  \n",
              "4181           W48*, D49N  \n",
              "4182           W48*, D49N  \n",
              "7018     P50L, R51*, A52V  \n",
              "5926     P62F, H63Y, Q64*  \n",
              "5925     H63Y, Q64*, R65C  \n",
              "7004     R80C, H81Y, R82*  \n",
              "4147                 Q81*  \n",
              "7002                 Q89*  \n",
              "7000                 Q94*  \n",
              "4766                Q125*  \n",
              "4765                Q125*  \n",
              "5898         A137V, Q138*  \n",
              "4122  A147V, Q148*, L149F  \n",
              "4109  T177I, L178F, Q179*  "
            ]
          },
          "execution_count": 8,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "# Filter the DataFrame to retain only up to 5 sgRNA sequences per locus_tag\n",
        "filtered_df = mutated_sgrna_df.groupby('locus_tag').head(number_of_sgRNAs_per_group)\n",
        "filtered_df"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "## Output"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 9,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/plain": [
              "[Dseqrecord(-20),\n",
              " Dseqrecord(-20),\n",
              " Dseqrecord(-20),\n",
              " Dseqrecord(-20),\n",
              " Dseqrecord(-20),\n",
              " Dseqrecord(-20),\n",
              " Dseqrecord(-20),\n",
              " Dseqrecord(-20),\n",
              " Dseqrecord(-20),\n",
              " Dseqrecord(-20),\n",
              " Dseqrecord(-20),\n",
              " Dseqrecord(-20),\n",
              " Dseqrecord(-20),\n",
              " Dseqrecord(-20),\n",
              " Dseqrecord(-20),\n",
              " Dseqrecord(-20),\n",
              " Dseqrecord(-20),\n",
              " Dseqrecord(-20),\n",
              " Dseqrecord(-20),\n",
              " Dseqrecord(-20)]"
            ]
          },
          "execution_count": 9,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "sgRNA_list = dataframe_to_seqrecords(filtered_df)\n",
        "sgRNA_list"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 10,
      "metadata": {},
      "outputs": [],
      "source": [
        "\n",
        "sgRNA_handle_cys4_sites = sgRNA_handle_cys4_site*len(sgRNA_list)\n",
        "\n",
        "golden_gate = GoldenGateCloning(sgRNA_list,\n",
        "                                sgRNA_handle_cys4_sites, \n",
        "                                target_tm= melting_temperature,\n",
        "                                restriction_overhang_f = \"GATCGggtctcc\",\n",
        "                                restriction_overhang_r = \"GATCAGGTCTCg\",\n",
        "                                backbone_overhang_f = \"cATG\",\n",
        "                                backbone_overhang_r = \"cTAG\",\n",
        "                                cys4 = \"gTTCACTGCCGTATAGGCAGCTAAGAAA\",\n",
        "                                tm_function = primer_tm_neb, \n",
        "                                primer_incrementation = primer_number_increment, \n",
        "                                polymerase = chosen_polymerase ) "
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 11,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>Locus Tag</th>\n",
              "      <th>f_primer_name</th>\n",
              "      <th>r_primer_name</th>\n",
              "      <th>f_primer_sequences(5-3)</th>\n",
              "      <th>r_primer_sequences(5-3)</th>\n",
              "      <th>f_tm</th>\n",
              "      <th>r_tm</th>\n",
              "      <th>ta</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>sgRNA_handle_cys4</td>\n",
              "      <td>primer_23</td>\n",
              "      <td>primer_24</td>\n",
              "      <td>GATCGGGTCTCCCATGGTTCACTGCCGTATAGGCAGCTAAGAAACC...</td>\n",
              "      <td>GATCAGGTCTCGGAGCAAAAAAGCACCGACTCGGT</td>\n",
              "      <td>63</td>\n",
              "      <td>59</td>\n",
              "      <td>63</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>sgRNA_handle_cys4</td>\n",
              "      <td>primer_25</td>\n",
              "      <td>primer_26</td>\n",
              "      <td>GATCGGGTCTCCGCTCAACCTCCAGGCCGGACGTTTTAGAGCTAGA...</td>\n",
              "      <td>GATCAGGTCTCGCAACAAAAAAGCACCGACTCGGT</td>\n",
              "      <td>63</td>\n",
              "      <td>59</td>\n",
              "      <td>63</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>sgRNA_handle_cys4</td>\n",
              "      <td>primer_27</td>\n",
              "      <td>primer_28</td>\n",
              "      <td>GATCGGGTCTCCGTTGTCCGTCAGCACTTCGAGTTTTAGAGCTAGA...</td>\n",
              "      <td>GATCAGGTCTCGACAAAAAAAAGCACCGACTCGGT</td>\n",
              "      <td>63</td>\n",
              "      <td>59</td>\n",
              "      <td>63</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>sgRNA_handle_cys4</td>\n",
              "      <td>primer_29</td>\n",
              "      <td>primer_30</td>\n",
              "      <td>GATCGGGTCTCCTTGTCCGTCAGCACTTCGAAGTTTTAGAGCTAGA...</td>\n",
              "      <td>GATCAGGTCTCGGGCAAAAAAAGCACCGACTCGGT</td>\n",
              "      <td>63</td>\n",
              "      <td>59</td>\n",
              "      <td>63</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>sgRNA_handle_cys4</td>\n",
              "      <td>primer_31</td>\n",
              "      <td>primer_32</td>\n",
              "      <td>GATCGGGTCTCCTGCCCGATCTTCCAGGCCAGGTTTTAGAGCTAGA...</td>\n",
              "      <td>GATCAGGTCTCGGAATAAAAAAGCACCGACTCGGT</td>\n",
              "      <td>63</td>\n",
              "      <td>59</td>\n",
              "      <td>63</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>5</th>\n",
              "      <td>sgRNA_handle_cys4</td>\n",
              "      <td>primer_33</td>\n",
              "      <td>primer_34</td>\n",
              "      <td>GATCGGGTCTCCATTCGACGCAGAGACCGCACGTTTTAGAGCTAGA...</td>\n",
              "      <td>GATCAGGTCTCGGGATAAAAAAGCACCGACTCGGT</td>\n",
              "      <td>63</td>\n",
              "      <td>59</td>\n",
              "      <td>63</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>6</th>\n",
              "      <td>sgRNA_handle_cys4</td>\n",
              "      <td>primer_35</td>\n",
              "      <td>primer_36</td>\n",
              "      <td>GATCGGGTCTCCATCCCAGGAAGGCAGCGCGCGTTTTAGAGCTAGA...</td>\n",
              "      <td>GATCAGGTCTCGGATGAAAAAAGCACCGACTCGGT</td>\n",
              "      <td>63</td>\n",
              "      <td>59</td>\n",
              "      <td>63</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>7</th>\n",
              "      <td>sgRNA_handle_cys4</td>\n",
              "      <td>primer_37</td>\n",
              "      <td>primer_38</td>\n",
              "      <td>GATCGGGTCTCCCATCCCAGGAAGGCAGCGCGGTTTTAGAGCTAGA...</td>\n",
              "      <td>GATCAGGTCTCGTGGCAAAAAAGCACCGACTCGGT</td>\n",
              "      <td>63</td>\n",
              "      <td>59</td>\n",
              "      <td>63</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>8</th>\n",
              "      <td>sgRNA_handle_cys4</td>\n",
              "      <td>primer_39</td>\n",
              "      <td>primer_40</td>\n",
              "      <td>GATCGGGTCTCCGCCACGAGCGGCATCCACTGGTTTTAGAGCTAGA...</td>\n",
              "      <td>GATCAGGTCTCGGTCGAAAAAAGCACCGACTCGGT</td>\n",
              "      <td>63</td>\n",
              "      <td>59</td>\n",
              "      <td>63</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>9</th>\n",
              "      <td>sgRNA_handle_cys4</td>\n",
              "      <td>primer_41</td>\n",
              "      <td>primer_42</td>\n",
              "      <td>GATCGGGTCTCCCGACCTCATCAGCGTCGACAGTTTTAGAGCTAGA...</td>\n",
              "      <td>GATCAGGTCTCGGAGGAAAAAAGCACCGACTCGGT</td>\n",
              "      <td>63</td>\n",
              "      <td>59</td>\n",
              "      <td>63</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>10</th>\n",
              "      <td>sgRNA_handle_cys4</td>\n",
              "      <td>primer_43</td>\n",
              "      <td>primer_44</td>\n",
              "      <td>GATCGGGTCTCCCCTCATCAGCGTCGACACGGGTTTTAGAGCTAGA...</td>\n",
              "      <td>GATCAGGTCTCGGTGAAAAAAAGCACCGACTCGGT</td>\n",
              "      <td>63</td>\n",
              "      <td>59</td>\n",
              "      <td>63</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>11</th>\n",
              "      <td>sgRNA_handle_cys4</td>\n",
              "      <td>primer_45</td>\n",
              "      <td>primer_46</td>\n",
              "      <td>GATCGGGTCTCCTCACGTCACCGACGGCGTACGTTTTAGAGCTAGA...</td>\n",
              "      <td>GATCAGGTCTCGGTCGAAAAAAGCACCGACTCGGT</td>\n",
              "      <td>63</td>\n",
              "      <td>59</td>\n",
              "      <td>63</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>12</th>\n",
              "      <td>sgRNA_handle_cys4</td>\n",
              "      <td>primer_47</td>\n",
              "      <td>primer_48</td>\n",
              "      <td>GATCGGGTCTCCCGACCAGGTCCATCTGCCACGTTTTAGAGCTAGA...</td>\n",
              "      <td>GATCAGGTCTCGTACGAAAAAAGCACCGACTCGGT</td>\n",
              "      <td>63</td>\n",
              "      <td>59</td>\n",
              "      <td>63</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>13</th>\n",
              "      <td>sgRNA_handle_cys4</td>\n",
              "      <td>primer_49</td>\n",
              "      <td>primer_50</td>\n",
              "      <td>GATCGGGTCTCCCGTACAGGTCGCGCATCTCCGTTTTAGAGCTAGA...</td>\n",
              "      <td>GATCAGGTCTCGGAGAAAAAAAGCACCGACTCGGT</td>\n",
              "      <td>63</td>\n",
              "      <td>59</td>\n",
              "      <td>63</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>14</th>\n",
              "      <td>sgRNA_handle_cys4</td>\n",
              "      <td>primer_51</td>\n",
              "      <td>primer_52</td>\n",
              "      <td>GATCGGGTCTCCTCTCCAGGTGGACCGCCATGGTTTTAGAGCTAGA...</td>\n",
              "      <td>GATCAGGTCTCGCTTCAAAAAAGCACCGACTCGGT</td>\n",
              "      <td>63</td>\n",
              "      <td>59</td>\n",
              "      <td>63</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>15</th>\n",
              "      <td>sgRNA_handle_cys4</td>\n",
              "      <td>primer_53</td>\n",
              "      <td>primer_54</td>\n",
              "      <td>GATCGGGTCTCCGAAGGACAGGTGGGCCTCGCGTTTTAGAGCTAGA...</td>\n",
              "      <td>GATCAGGTCTCGCCTTAAAAAAGCACCGACTCGGT</td>\n",
              "      <td>63</td>\n",
              "      <td>59</td>\n",
              "      <td>63</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>16</th>\n",
              "      <td>sgRNA_handle_cys4</td>\n",
              "      <td>primer_55</td>\n",
              "      <td>primer_56</td>\n",
              "      <td>GATCGGGTCTCCAAGGACAGGTGGGCCTCGCCGTTTTAGAGCTAGA...</td>\n",
              "      <td>GATCAGGTCTCGATCAAAAAAAGCACCGACTCGGT</td>\n",
              "      <td>63</td>\n",
              "      <td>59</td>\n",
              "      <td>63</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>17</th>\n",
              "      <td>sgRNA_handle_cys4</td>\n",
              "      <td>primer_57</td>\n",
              "      <td>primer_58</td>\n",
              "      <td>GATCGGGTCTCCTGATGCGCAGCCACAGGCGGGTTTTAGAGCTAGA...</td>\n",
              "      <td>GATCAGGTCTCGCCGCAAAAAAGCACCGACTCGGT</td>\n",
              "      <td>63</td>\n",
              "      <td>59</td>\n",
              "      <td>63</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>18</th>\n",
              "      <td>sgRNA_handle_cys4</td>\n",
              "      <td>primer_59</td>\n",
              "      <td>primer_60</td>\n",
              "      <td>GATCGGGTCTCCGCGGCGCAGCTCGCAGATCTGTTTTAGAGCTAGA...</td>\n",
              "      <td>GATCAGGTCTCGGGTAAAAAAAGCACCGACTCGGT</td>\n",
              "      <td>63</td>\n",
              "      <td>59</td>\n",
              "      <td>63</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>19</th>\n",
              "      <td>sgRNA_handle_cys4</td>\n",
              "      <td>primer_61</td>\n",
              "      <td>primer_62</td>\n",
              "      <td>GATCGGGTCTCCTACCCTCCAGCACGAACCTCGTTTTAGAGCTAGA...</td>\n",
              "      <td>GATCAGGTCTCGCTAGAAAAAAGCACCGACTCGGT</td>\n",
              "      <td>63</td>\n",
              "      <td>59</td>\n",
              "      <td>63</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "            Locus Tag f_primer_name r_primer_name  \\\n",
              "0   sgRNA_handle_cys4     primer_23     primer_24   \n",
              "1   sgRNA_handle_cys4     primer_25     primer_26   \n",
              "2   sgRNA_handle_cys4     primer_27     primer_28   \n",
              "3   sgRNA_handle_cys4     primer_29     primer_30   \n",
              "4   sgRNA_handle_cys4     primer_31     primer_32   \n",
              "5   sgRNA_handle_cys4     primer_33     primer_34   \n",
              "6   sgRNA_handle_cys4     primer_35     primer_36   \n",
              "7   sgRNA_handle_cys4     primer_37     primer_38   \n",
              "8   sgRNA_handle_cys4     primer_39     primer_40   \n",
              "9   sgRNA_handle_cys4     primer_41     primer_42   \n",
              "10  sgRNA_handle_cys4     primer_43     primer_44   \n",
              "11  sgRNA_handle_cys4     primer_45     primer_46   \n",
              "12  sgRNA_handle_cys4     primer_47     primer_48   \n",
              "13  sgRNA_handle_cys4     primer_49     primer_50   \n",
              "14  sgRNA_handle_cys4     primer_51     primer_52   \n",
              "15  sgRNA_handle_cys4     primer_53     primer_54   \n",
              "16  sgRNA_handle_cys4     primer_55     primer_56   \n",
              "17  sgRNA_handle_cys4     primer_57     primer_58   \n",
              "18  sgRNA_handle_cys4     primer_59     primer_60   \n",
              "19  sgRNA_handle_cys4     primer_61     primer_62   \n",
              "\n",
              "                              f_primer_sequences(5-3)  \\\n",
              "0   GATCGGGTCTCCCATGGTTCACTGCCGTATAGGCAGCTAAGAAACC...   \n",
              "1   GATCGGGTCTCCGCTCAACCTCCAGGCCGGACGTTTTAGAGCTAGA...   \n",
              "2   GATCGGGTCTCCGTTGTCCGTCAGCACTTCGAGTTTTAGAGCTAGA...   \n",
              "3   GATCGGGTCTCCTTGTCCGTCAGCACTTCGAAGTTTTAGAGCTAGA...   \n",
              "4   GATCGGGTCTCCTGCCCGATCTTCCAGGCCAGGTTTTAGAGCTAGA...   \n",
              "5   GATCGGGTCTCCATTCGACGCAGAGACCGCACGTTTTAGAGCTAGA...   \n",
              "6   GATCGGGTCTCCATCCCAGGAAGGCAGCGCGCGTTTTAGAGCTAGA...   \n",
              "7   GATCGGGTCTCCCATCCCAGGAAGGCAGCGCGGTTTTAGAGCTAGA...   \n",
              "8   GATCGGGTCTCCGCCACGAGCGGCATCCACTGGTTTTAGAGCTAGA...   \n",
              "9   GATCGGGTCTCCCGACCTCATCAGCGTCGACAGTTTTAGAGCTAGA...   \n",
              "10  GATCGGGTCTCCCCTCATCAGCGTCGACACGGGTTTTAGAGCTAGA...   \n",
              "11  GATCGGGTCTCCTCACGTCACCGACGGCGTACGTTTTAGAGCTAGA...   \n",
              "12  GATCGGGTCTCCCGACCAGGTCCATCTGCCACGTTTTAGAGCTAGA...   \n",
              "13  GATCGGGTCTCCCGTACAGGTCGCGCATCTCCGTTTTAGAGCTAGA...   \n",
              "14  GATCGGGTCTCCTCTCCAGGTGGACCGCCATGGTTTTAGAGCTAGA...   \n",
              "15  GATCGGGTCTCCGAAGGACAGGTGGGCCTCGCGTTTTAGAGCTAGA...   \n",
              "16  GATCGGGTCTCCAAGGACAGGTGGGCCTCGCCGTTTTAGAGCTAGA...   \n",
              "17  GATCGGGTCTCCTGATGCGCAGCCACAGGCGGGTTTTAGAGCTAGA...   \n",
              "18  GATCGGGTCTCCGCGGCGCAGCTCGCAGATCTGTTTTAGAGCTAGA...   \n",
              "19  GATCGGGTCTCCTACCCTCCAGCACGAACCTCGTTTTAGAGCTAGA...   \n",
              "\n",
              "                r_primer_sequences(5-3)  f_tm  r_tm  ta  \n",
              "0   GATCAGGTCTCGGAGCAAAAAAGCACCGACTCGGT    63    59  63  \n",
              "1   GATCAGGTCTCGCAACAAAAAAGCACCGACTCGGT    63    59  63  \n",
              "2   GATCAGGTCTCGACAAAAAAAAGCACCGACTCGGT    63    59  63  \n",
              "3   GATCAGGTCTCGGGCAAAAAAAGCACCGACTCGGT    63    59  63  \n",
              "4   GATCAGGTCTCGGAATAAAAAAGCACCGACTCGGT    63    59  63  \n",
              "5   GATCAGGTCTCGGGATAAAAAAGCACCGACTCGGT    63    59  63  \n",
              "6   GATCAGGTCTCGGATGAAAAAAGCACCGACTCGGT    63    59  63  \n",
              "7   GATCAGGTCTCGTGGCAAAAAAGCACCGACTCGGT    63    59  63  \n",
              "8   GATCAGGTCTCGGTCGAAAAAAGCACCGACTCGGT    63    59  63  \n",
              "9   GATCAGGTCTCGGAGGAAAAAAGCACCGACTCGGT    63    59  63  \n",
              "10  GATCAGGTCTCGGTGAAAAAAAGCACCGACTCGGT    63    59  63  \n",
              "11  GATCAGGTCTCGGTCGAAAAAAGCACCGACTCGGT    63    59  63  \n",
              "12  GATCAGGTCTCGTACGAAAAAAGCACCGACTCGGT    63    59  63  \n",
              "13  GATCAGGTCTCGGAGAAAAAAAGCACCGACTCGGT    63    59  63  \n",
              "14  GATCAGGTCTCGCTTCAAAAAAGCACCGACTCGGT    63    59  63  \n",
              "15  GATCAGGTCTCGCCTTAAAAAAGCACCGACTCGGT    63    59  63  \n",
              "16  GATCAGGTCTCGATCAAAAAAAGCACCGACTCGGT    63    59  63  \n",
              "17  GATCAGGTCTCGCCGCAAAAAAGCACCGACTCGGT    63    59  63  \n",
              "18  GATCAGGTCTCGGGTAAAAAAAGCACCGACTCGGT    63    59  63  \n",
              "19  GATCAGGTCTCGCTAGAAAAAAGCACCGACTCGGT    63    59  63  "
            ]
          },
          "execution_count": 11,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "primer_df = golden_gate.generate_primer_dataframe()\n",
        "primer_df"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 12,
      "metadata": {},
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "Skipping analysis for primer pair primer_23 & primer_24 due to length constraints.\n",
            "Skipping analysis for primer pair primer_25 & primer_26 due to length constraints.\n",
            "Skipping analysis for primer pair primer_27 & primer_28 due to length constraints.\n",
            "Skipping analysis for primer pair primer_29 & primer_30 due to length constraints.\n",
            "Skipping analysis for primer pair primer_31 & primer_32 due to length constraints.\n",
            "Skipping analysis for primer pair primer_33 & primer_34 due to length constraints.\n",
            "Skipping analysis for primer pair primer_35 & primer_36 due to length constraints.\n",
            "Skipping analysis for primer pair primer_37 & primer_38 due to length constraints.\n",
            "Skipping analysis for primer pair primer_39 & primer_40 due to length constraints.\n",
            "Skipping analysis for primer pair primer_41 & primer_42 due to length constraints.\n",
            "Skipping analysis for primer pair primer_43 & primer_44 due to length constraints.\n",
            "Skipping analysis for primer pair primer_45 & primer_46 due to length constraints.\n",
            "Skipping analysis for primer pair primer_47 & primer_48 due to length constraints.\n",
            "Skipping analysis for primer pair primer_49 & primer_50 due to length constraints.\n",
            "Skipping analysis for primer pair primer_51 & primer_52 due to length constraints.\n",
            "Skipping analysis for primer pair primer_53 & primer_54 due to length constraints.\n",
            "Skipping analysis for primer pair primer_55 & primer_56 due to length constraints.\n",
            "Skipping analysis for primer pair primer_57 & primer_58 due to length constraints.\n",
            "Skipping analysis for primer pair primer_59 & primer_60 due to length constraints.\n",
            "Skipping analysis for primer pair primer_61 & primer_62 due to length constraints.\n"
          ]
        },
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "Empty DataFrame\n",
              "Columns: []\n",
              "Index: []"
            ]
          },
          "execution_count": 12,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "# Primer analysis\n",
        "analysis_of_primers = analyze_primers_and_hairpins(primer_df)\n",
        "analysis_of_primers"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 13,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>Name</th>\n",
              "      <th>Sequence</th>\n",
              "      <th>Concentration</th>\n",
              "      <th>Purification</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>primer_23</td>\n",
              "      <td>GATCGGGTCTCCCATGGTTCACTGCCGTATAGGCAGCTAAGAAACC...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>primer_25</td>\n",
              "      <td>GATCGGGTCTCCGCTCAACCTCCAGGCCGGACGTTTTAGAGCTAGA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>primer_27</td>\n",
              "      <td>GATCGGGTCTCCGTTGTCCGTCAGCACTTCGAGTTTTAGAGCTAGA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>primer_29</td>\n",
              "      <td>GATCGGGTCTCCTTGTCCGTCAGCACTTCGAAGTTTTAGAGCTAGA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>primer_31</td>\n",
              "      <td>GATCGGGTCTCCTGCCCGATCTTCCAGGCCAGGTTTTAGAGCTAGA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>5</th>\n",
              "      <td>primer_33</td>\n",
              "      <td>GATCGGGTCTCCATTCGACGCAGAGACCGCACGTTTTAGAGCTAGA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>6</th>\n",
              "      <td>primer_35</td>\n",
              "      <td>GATCGGGTCTCCATCCCAGGAAGGCAGCGCGCGTTTTAGAGCTAGA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>7</th>\n",
              "      <td>primer_37</td>\n",
              "      <td>GATCGGGTCTCCCATCCCAGGAAGGCAGCGCGGTTTTAGAGCTAGA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>8</th>\n",
              "      <td>primer_39</td>\n",
              "      <td>GATCGGGTCTCCGCCACGAGCGGCATCCACTGGTTTTAGAGCTAGA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>9</th>\n",
              "      <td>primer_41</td>\n",
              "      <td>GATCGGGTCTCCCGACCTCATCAGCGTCGACAGTTTTAGAGCTAGA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>10</th>\n",
              "      <td>primer_43</td>\n",
              "      <td>GATCGGGTCTCCCCTCATCAGCGTCGACACGGGTTTTAGAGCTAGA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>11</th>\n",
              "      <td>primer_45</td>\n",
              "      <td>GATCGGGTCTCCTCACGTCACCGACGGCGTACGTTTTAGAGCTAGA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>12</th>\n",
              "      <td>primer_47</td>\n",
              "      <td>GATCGGGTCTCCCGACCAGGTCCATCTGCCACGTTTTAGAGCTAGA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>13</th>\n",
              "      <td>primer_49</td>\n",
              "      <td>GATCGGGTCTCCCGTACAGGTCGCGCATCTCCGTTTTAGAGCTAGA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>14</th>\n",
              "      <td>primer_51</td>\n",
              "      <td>GATCGGGTCTCCTCTCCAGGTGGACCGCCATGGTTTTAGAGCTAGA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>15</th>\n",
              "      <td>primer_53</td>\n",
              "      <td>GATCGGGTCTCCGAAGGACAGGTGGGCCTCGCGTTTTAGAGCTAGA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>16</th>\n",
              "      <td>primer_55</td>\n",
              "      <td>GATCGGGTCTCCAAGGACAGGTGGGCCTCGCCGTTTTAGAGCTAGA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>17</th>\n",
              "      <td>primer_57</td>\n",
              "      <td>GATCGGGTCTCCTGATGCGCAGCCACAGGCGGGTTTTAGAGCTAGA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>18</th>\n",
              "      <td>primer_59</td>\n",
              "      <td>GATCGGGTCTCCGCGGCGCAGCTCGCAGATCTGTTTTAGAGCTAGA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>19</th>\n",
              "      <td>primer_61</td>\n",
              "      <td>GATCGGGTCTCCTACCCTCCAGCACGAACCTCGTTTTAGAGCTAGA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>20</th>\n",
              "      <td>primer_24</td>\n",
              "      <td>GATCAGGTCTCGGAGCAAAAAAGCACCGACTCGGT</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>21</th>\n",
              "      <td>primer_26</td>\n",
              "      <td>GATCAGGTCTCGCAACAAAAAAGCACCGACTCGGT</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>22</th>\n",
              "      <td>primer_28</td>\n",
              "      <td>GATCAGGTCTCGACAAAAAAAAGCACCGACTCGGT</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>23</th>\n",
              "      <td>primer_30</td>\n",
              "      <td>GATCAGGTCTCGGGCAAAAAAAGCACCGACTCGGT</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>24</th>\n",
              "      <td>primer_32</td>\n",
              "      <td>GATCAGGTCTCGGAATAAAAAAGCACCGACTCGGT</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>25</th>\n",
              "      <td>primer_34</td>\n",
              "      <td>GATCAGGTCTCGGGATAAAAAAGCACCGACTCGGT</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>26</th>\n",
              "      <td>primer_36</td>\n",
              "      <td>GATCAGGTCTCGGATGAAAAAAGCACCGACTCGGT</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>27</th>\n",
              "      <td>primer_38</td>\n",
              "      <td>GATCAGGTCTCGTGGCAAAAAAGCACCGACTCGGT</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>28</th>\n",
              "      <td>primer_40</td>\n",
              "      <td>GATCAGGTCTCGGTCGAAAAAAGCACCGACTCGGT</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>29</th>\n",
              "      <td>primer_42</td>\n",
              "      <td>GATCAGGTCTCGGAGGAAAAAAGCACCGACTCGGT</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>30</th>\n",
              "      <td>primer_44</td>\n",
              "      <td>GATCAGGTCTCGGTGAAAAAAAGCACCGACTCGGT</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>31</th>\n",
              "      <td>primer_46</td>\n",
              "      <td>GATCAGGTCTCGGTCGAAAAAAGCACCGACTCGGT</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>32</th>\n",
              "      <td>primer_48</td>\n",
              "      <td>GATCAGGTCTCGTACGAAAAAAGCACCGACTCGGT</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>33</th>\n",
              "      <td>primer_50</td>\n",
              "      <td>GATCAGGTCTCGGAGAAAAAAAGCACCGACTCGGT</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>34</th>\n",
              "      <td>primer_52</td>\n",
              "      <td>GATCAGGTCTCGCTTCAAAAAAGCACCGACTCGGT</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>35</th>\n",
              "      <td>primer_54</td>\n",
              "      <td>GATCAGGTCTCGCCTTAAAAAAGCACCGACTCGGT</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>36</th>\n",
              "      <td>primer_56</td>\n",
              "      <td>GATCAGGTCTCGATCAAAAAAAGCACCGACTCGGT</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>37</th>\n",
              "      <td>primer_58</td>\n",
              "      <td>GATCAGGTCTCGCCGCAAAAAAGCACCGACTCGGT</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>38</th>\n",
              "      <td>primer_60</td>\n",
              "      <td>GATCAGGTCTCGGGTAAAAAAAGCACCGACTCGGT</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>39</th>\n",
              "      <td>primer_62</td>\n",
              "      <td>GATCAGGTCTCGCTAGAAAAAAGCACCGACTCGGT</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "         Name                                           Sequence  \\\n",
              "0   primer_23  GATCGGGTCTCCCATGGTTCACTGCCGTATAGGCAGCTAAGAAACC...   \n",
              "1   primer_25  GATCGGGTCTCCGCTCAACCTCCAGGCCGGACGTTTTAGAGCTAGA...   \n",
              "2   primer_27  GATCGGGTCTCCGTTGTCCGTCAGCACTTCGAGTTTTAGAGCTAGA...   \n",
              "3   primer_29  GATCGGGTCTCCTTGTCCGTCAGCACTTCGAAGTTTTAGAGCTAGA...   \n",
              "4   primer_31  GATCGGGTCTCCTGCCCGATCTTCCAGGCCAGGTTTTAGAGCTAGA...   \n",
              "5   primer_33  GATCGGGTCTCCATTCGACGCAGAGACCGCACGTTTTAGAGCTAGA...   \n",
              "6   primer_35  GATCGGGTCTCCATCCCAGGAAGGCAGCGCGCGTTTTAGAGCTAGA...   \n",
              "7   primer_37  GATCGGGTCTCCCATCCCAGGAAGGCAGCGCGGTTTTAGAGCTAGA...   \n",
              "8   primer_39  GATCGGGTCTCCGCCACGAGCGGCATCCACTGGTTTTAGAGCTAGA...   \n",
              "9   primer_41  GATCGGGTCTCCCGACCTCATCAGCGTCGACAGTTTTAGAGCTAGA...   \n",
              "10  primer_43  GATCGGGTCTCCCCTCATCAGCGTCGACACGGGTTTTAGAGCTAGA...   \n",
              "11  primer_45  GATCGGGTCTCCTCACGTCACCGACGGCGTACGTTTTAGAGCTAGA...   \n",
              "12  primer_47  GATCGGGTCTCCCGACCAGGTCCATCTGCCACGTTTTAGAGCTAGA...   \n",
              "13  primer_49  GATCGGGTCTCCCGTACAGGTCGCGCATCTCCGTTTTAGAGCTAGA...   \n",
              "14  primer_51  GATCGGGTCTCCTCTCCAGGTGGACCGCCATGGTTTTAGAGCTAGA...   \n",
              "15  primer_53  GATCGGGTCTCCGAAGGACAGGTGGGCCTCGCGTTTTAGAGCTAGA...   \n",
              "16  primer_55  GATCGGGTCTCCAAGGACAGGTGGGCCTCGCCGTTTTAGAGCTAGA...   \n",
              "17  primer_57  GATCGGGTCTCCTGATGCGCAGCCACAGGCGGGTTTTAGAGCTAGA...   \n",
              "18  primer_59  GATCGGGTCTCCGCGGCGCAGCTCGCAGATCTGTTTTAGAGCTAGA...   \n",
              "19  primer_61  GATCGGGTCTCCTACCCTCCAGCACGAACCTCGTTTTAGAGCTAGA...   \n",
              "20  primer_24                GATCAGGTCTCGGAGCAAAAAAGCACCGACTCGGT   \n",
              "21  primer_26                GATCAGGTCTCGCAACAAAAAAGCACCGACTCGGT   \n",
              "22  primer_28                GATCAGGTCTCGACAAAAAAAAGCACCGACTCGGT   \n",
              "23  primer_30                GATCAGGTCTCGGGCAAAAAAAGCACCGACTCGGT   \n",
              "24  primer_32                GATCAGGTCTCGGAATAAAAAAGCACCGACTCGGT   \n",
              "25  primer_34                GATCAGGTCTCGGGATAAAAAAGCACCGACTCGGT   \n",
              "26  primer_36                GATCAGGTCTCGGATGAAAAAAGCACCGACTCGGT   \n",
              "27  primer_38                GATCAGGTCTCGTGGCAAAAAAGCACCGACTCGGT   \n",
              "28  primer_40                GATCAGGTCTCGGTCGAAAAAAGCACCGACTCGGT   \n",
              "29  primer_42                GATCAGGTCTCGGAGGAAAAAAGCACCGACTCGGT   \n",
              "30  primer_44                GATCAGGTCTCGGTGAAAAAAAGCACCGACTCGGT   \n",
              "31  primer_46                GATCAGGTCTCGGTCGAAAAAAGCACCGACTCGGT   \n",
              "32  primer_48                GATCAGGTCTCGTACGAAAAAAGCACCGACTCGGT   \n",
              "33  primer_50                GATCAGGTCTCGGAGAAAAAAAGCACCGACTCGGT   \n",
              "34  primer_52                GATCAGGTCTCGCTTCAAAAAAGCACCGACTCGGT   \n",
              "35  primer_54                GATCAGGTCTCGCCTTAAAAAAGCACCGACTCGGT   \n",
              "36  primer_56                GATCAGGTCTCGATCAAAAAAAGCACCGACTCGGT   \n",
              "37  primer_58                GATCAGGTCTCGCCGCAAAAAAGCACCGACTCGGT   \n",
              "38  primer_60                GATCAGGTCTCGGGTAAAAAAAGCACCGACTCGGT   \n",
              "39  primer_62                GATCAGGTCTCGCTAGAAAAAAGCACCGACTCGGT   \n",
              "\n",
              "   Concentration Purification  \n",
              "0           25nm          STD  \n",
              "1           25nm          STD  \n",
              "2           25nm          STD  \n",
              "3           25nm          STD  \n",
              "4           25nm          STD  \n",
              "5           25nm          STD  \n",
              "6           25nm          STD  \n",
              "7           25nm          STD  \n",
              "8           25nm          STD  \n",
              "9           25nm          STD  \n",
              "10          25nm          STD  \n",
              "11          25nm          STD  \n",
              "12          25nm          STD  \n",
              "13          25nm          STD  \n",
              "14          25nm          STD  \n",
              "15          25nm          STD  \n",
              "16          25nm          STD  \n",
              "17          25nm          STD  \n",
              "18          25nm          STD  \n",
              "19          25nm          STD  \n",
              "20          25nm          STD  \n",
              "21          25nm          STD  \n",
              "22          25nm          STD  \n",
              "23          25nm          STD  \n",
              "24          25nm          STD  \n",
              "25          25nm          STD  \n",
              "26          25nm          STD  \n",
              "27          25nm          STD  \n",
              "28          25nm          STD  \n",
              "29          25nm          STD  \n",
              "30          25nm          STD  \n",
              "31          25nm          STD  \n",
              "32          25nm          STD  \n",
              "33          25nm          STD  \n",
              "34          25nm          STD  \n",
              "35          25nm          STD  \n",
              "36          25nm          STD  \n",
              "37          25nm          STD  \n",
              "38          25nm          STD  \n",
              "39          25nm          STD  "
            ]
          },
          "execution_count": 13,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "# 2 Ready made IDT format\n",
        "idt_df = create_idt_order_dataframe(primer_df, concentration=\"25nm\", purification=\"STD\")\n",
        "idt_df"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 14,
      "metadata": {},
      "outputs": [],
      "source": [
        "list_of_amplicons = golden_gate.simulate_pcrs()"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 15,
      "metadata": {},
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "chosen_region_2(4000-7000)_34_pcr_products\n",
            "chosen_region_4(15000-20000)_54_pcr_products\n",
            "chosen_region_1(80000-100000)_80_pcr_products\n",
            "chosen_region_1(80000-100000)_81_pcr_products\n",
            "chosen_region_3(9000-14000)_118_pcr_products\n",
            "chosen_region_3(9000-14000)_146_pcr_products\n",
            "chosen_region_2(4000-7000)_147_pcr_products\n",
            "chosen_region_2(4000-7000)_148_pcr_products\n",
            "chosen_region_4(15000-20000)_166_pcr_products\n",
            "chosen_region_3(9000-14000)_200_pcr_products\n",
            "chosen_region_3(9000-14000)_203_pcr_products\n",
            "chosen_region_4(15000-20000)_254_pcr_products\n",
            "chosen_region_1(80000-100000)_256_pcr_products\n",
            "chosen_region_4(15000-20000)_280_pcr_products\n",
            "chosen_region_4(15000-20000)_295_pcr_products\n",
            "chosen_region_2(4000-7000)_386_pcr_products\n",
            "chosen_region_2(4000-7000)_387_pcr_products\n",
            "chosen_region_3(9000-14000)_424_pcr_products\n",
            "chosen_region_1(80000-100000)_455_pcr_products\n",
            "chosen_region_1(80000-100000)_547_pcr_products\n"
          ]
        },
        {
          "data": {
            "image/jpeg": "/9j/4AAQSkZJRgABAQAAAQABAAD/2wBDAAgGBgcGBQgHBwcJCQgKDBQNDAsLDBkSEw8UHRofHh0aHBwgJC4nICIsIxwcKDcpLDAxNDQ0Hyc5PTgyPC4zNDL/2wBDAQkJCQwLDBgNDRgyIRwhMjIyMjIyMjIyMjIyMjIyMjIyMjIyMjIyMjIyMjIyMjIyMjIyMjIyMjIyMjIyMjIyMjL/wAARCAJYBSgDASIAAhEBAxEB/8QAHwAAAQUBAQEBAQEAAAAAAAAAAAECAwQFBgcICQoL/8QAtRAAAgEDAwIEAwUFBAQAAAF9AQIDAAQRBRIhMUEGE1FhByJxFDKBkaEII0KxwRVS0fAkM2JyggkKFhcYGRolJicoKSo0NTY3ODk6Q0RFRkdISUpTVFVWV1hZWmNkZWZnaGlqc3R1dnd4eXqDhIWGh4iJipKTlJWWl5iZmqKjpKWmp6ipqrKztLW2t7i5usLDxMXGx8jJytLT1NXW19jZ2uHi4+Tl5ufo6erx8vP09fb3+Pn6/8QAHwEAAwEBAQEBAQEBAQAAAAAAAAECAwQFBgcICQoL/8QAtREAAgECBAQDBAcFBAQAAQJ3AAECAxEEBSExBhJBUQdhcRMiMoEIFEKRobHBCSMzUvAVYnLRChYkNOEl8RcYGRomJygpKjU2Nzg5OkNERUZHSElKU1RVVldYWVpjZGVmZ2hpanN0dXZ3eHl6goOEhYaHiImKkpOUlZaXmJmaoqOkpaanqKmqsrO0tba3uLm6wsPExcbHyMnK0tPU1dbX2Nna4uPk5ebn6Onq8vP09fb3+Pn6/9oADAMBAAIRAxEAPwDwiHTr24tJruGzuJLaD/WzJExSP/eYDA/GnvpOpRfZfM0+7T7Xj7NuhYednGNnHzdR09RXQaL51r4Xu78Xto7iO4t7eze8iRoxIgWaUozBmynyqACSef4QD0japYRapqszajaGPVb+WbT2WdW+zo1tcIhfB/dYaaEfNgjyz2FAHnyaJq0l9JYppd615Eu6S3Fu5kQcclcZA5H51CmnX0ljJfJZ3DWkTbZLgRMY0PHBbGAeR+ddxJJFPoMGiQanZRapbxWhlnN2ixkJLcsVEudrFRNBwCfucZxVjU9W029u7jVbS7t005LXU4GtjIqyNJNJOYsR53EYlhOQMDy+cYoA82oqZ2tjaRqkMwuQT5kjSgow7ALtBB99x/CoaAFCkjpRtPpU8VwEjC7Acd6f9rH/ADzX8qAKu0+lG0+lWvtY/wCea/lR9rH/ADzX8qAKu0+lG0+lWvtY/wCea/lR9rH/ADzX8qAKpBHUUlTzTiVAoUDnPFQUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRWnZ6HPe6LqGqRz2qxWKh3iaX964LomVUZ4BkTk4HJxkgitB/BeorOkK3Fo8iymG5Cu3+iOELkSZUdFRydu77jdxQBzlFdAvhK5Ll31Cwjsikbx3rtIIpPMLKoHybs5RxyoxsbOMU1/CWoQ20zzvbxXMazOLN2bzXSJisjDA24Uq/Ug/I2AcUAYNFFFABRUyW0kiBlBwad9jl/umgCvRVj7HL/dNH2OX+6aAK9FWPscv900fY5f7poAr0VLJA8S7mGBnFRUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFAGnpeow2Wn61byK5e+slt4ioGAwuIZMnnpiNumeSPrXTP4y0xb2+uI4btv7XupLjUFZFHlb4ZomWM7ju/4+JCCdvRR6muGooA69/EGi3GmR6HP9vGmQJB5c6Qp5rtG87nKb8AH7TIPvHGF61Le+MbPUZptUnhuE1PyL23jiRVMRW5eVixbOQV+0SDG05wvIri6KAJnupHtI7YrCI4yWUrCgc59XA3EexJxUNFFAEyXMkaBVJwKd9sl/vGq9FAFj7ZL/eNH2yX+8ar0UAWPtkv940fbJf7xqvRQBLJO8q7WORnNRUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRW1puk6bfaVe3MmoXcM9pbtM6fZFaMncFRd/mA5Yso+7xyeQK1H8EL9qmt49S3PZXDW+pFoMC3ZYpJGKfMd4AhlH8PK++aAORorqD4Vs47cajPqskekyJC0NwLUGRmkaRQGj34GDDLk7j93jOawNRsZtL1O70+4x59rM8Em05G5WKnH4igCtRRRQAUVbgtVkhVy6gnsTUn2FP+ei/nQBQoq/9hT/AJ6L+dH2FP8Anov50AUKKv8A2FP+ei/nR9hT/nov50AUKKtXNssMYYMDk44qrQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAX/7S26F/ZkUWzfcefPLuz5mFwi4xwFy5993tW+/jdftU1xHpu1724a41INPkXDNFJGwT5RsBE0p/i5b2xXI0UAdQfFVnJbrp0+lSSaTGkKw24ugJFaNpGBaTZg5M0uRtH3uMYrFuNYvbjXLjWFneC9nnecywsUKu5JOCOR1NUaKAFZmdy7sWZjkknJJpKKKAHB2AwDxR5j+tNooAd5j+tHmP602igB3mP60eY/rTaKAHF2YYJptFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUVvL4eUeGE1iRtQO/fgQ2W+FMHaN8u8bcnP8J6VoP4IX7VNbx6lueyuGt9SLQYFuyxSSMU+Y7wBDKP4eV980AcjRXUHwrZx241GfVZI9JkSFobgWoMjNI0igNHvwMGGXJ3H7vGc1gajYzaXqd3p9xjz7WZ4JNpyNysVOPxFAFaiiigAorstC8KW+paNb3b3cMbSbsqzcjDEf0rR/4Qa1/wCf+D/vqgDzyivQ/wDhBrX/AJ/4P++qP+EGtf8An/g/76oA88or0P8A4Qa1/wCf+D/vqj/hBrX/AJ/4P++qAPPKK6rxL4bh0fTo7iO5ilLTBNqHJ6E5/SuVoAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKAN6x8RRWWliEWczXaW89skpuj5QSUENmIg/NhjgggZwSCRzoP43X7VNcR6bte9uGuNSDT5FwzRSRsE+UbARNKf4uW9sVyNFAHUHxVZyW66dPpUkmkxpCsNuLoCRWjaRgWk2YOTNLkbR97jGKxbjWL241y41hZ3gvZ53nMsLFCruSTgjkdTVGigBWZncu7FmY5JJySaSiigCzFfzwxCNJGCjoAaf/al1/z1b86p0UAXP7Uuv+erfnR/al1/z1b86p0UAXP7Uuv+erfnR/al1/z1b86p0UAWJ7ya4QJI5YA5wTVeiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAuadpV9qr3C2Nu0xtreS5mwQAkSDLMSfQfn0HNXH8LazG9sjWeXuH2IqyoxVtu7DgH92dpzhscZPY1r+E/EWkaXZXFle2tyhmgu/MuIrgKJS1tJHGhXy2PVyAc4BfJyAMareJdGj1DVLhb7emt3ktw4WJ82SyW9xHhsgZIa5/hzxH7gUAcnH4W1iS8e2W3i3qiPvNzEIyrnC7ZC2xsngYJzg+hrJkjeGV4pUZJEYqysMFSOoIrtZdS0e60OHw82rRxJbx2x+3+TKUkZJLl2UALv6XOBlRny+2RXO6rqsN/4q1DVhaJLBc3ctwLeYsBh2JAO0g8Z7GgDKopWIZyQoUE5CjOB7c0lAFiK0kljDqDg0/wCwS/3TTIruSKMIpOBT/t8v940AH2CX+6aPsEv900fb5f7xo+3y/wB40AH2CX+6aPsEv900fb5f7xo+3y/3jQBHNbPCgZhgE4qCp5rl5kCscgHNQUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAWotMv57GW+hsbmS0iOJJ0iYxoeOrAYHUfnTn0nUovsvmafdp9rx9m3QsPOzjGzj5uo6eorpNAljtNGubma/tGifT7qFRJPia0kZWGyOIt8wkyoLbSAGPIK5roG1Swi1TVZm1G0Meq38s2nss6t9nRra4RC+D+6w00I+bBHlnsKAPPk0TVpL6SxTS71ryJd0luLdzIg45K4yByPzqh0NegySRT6DBokGp2UWqW8VoZZzdosZCS3LFRLnaxUTQcAn7nGcVyus3mn3ni/U71YpJNPnvZpUjicRMY2diuCVOOCOxoAyKKVtpclAQueATkgfWkoAlS3d0DAHBp32ST+6amgvmhhWMYwM1J/ab+1AFX7JJ/dNH2ST+6atf2m/tR/ab+1AFX7JJ/dNH2ST+6atf2m/tR/ab+1AFOSB41ywwM4qKrVzeNcRhD2Oaq0AFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUVoWuh6je6fLfW8AaCIOSTIoZggDPtUnc20EE7QcA5OKsP4W1mN7ZGs8vcPsRVlRirbd2HAP7s7TnDY4yexoAx6K2I/C2sSXj2y28W9UR95uYhGVc4XbIW2Nk8DBOcH0NZMkbwyvFKjJIjFWVhgqR1BFADaKKKACipo7Z5EDAHBp/2OX+6aAK1FWfscv900fY5f7poArUVZ+xy/3TR9jl/umgCtRUstu8ShmGBnFRUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFAHSaTNZWXhy8nXVoE1SVJII7eVZsxRMuHKkIV3uCU5IAGc9QR0DeJdGj1DVLhb7emt3ktw4WJ82SyW9xHhsgZIa5/hzxH7gV53RQB28upaPdaHD4ebVo4kt47Y/b/JlKSMkly7KAF39LnAyoz5fbIrndV1WG/wDFWoasLRJYLm7luBbzFgMOxIB2kHjPY1lUUAKxDOSFCgnIUZwPbmkoooAv2+oNBAsYxgZ/nUn9qv7flWZRQBp/2q/t+VH9qv7flWZRQBp/2q/t+VH9qv7flWZRQBcu71rmIIezZqnRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUVt6XpunXvh7WrmWS6/tCygWeJFCiLaZoYzuPJJPmngYxtHJzgbt14OtItSjtxb3aW0SyPc3QvYptwjgaZ1CKgMbEI23cTx2NAHD0V17+H9Ft9Mj1yf7edMnSDy4EmTzUaR50OX2YIH2aQ/dGcr0rCv8ARbq08R3miQRyXVzb3MluFhQs0hRiCQBz2zQBm0UrKyOUdSrKcEEYINJQA9Y2YZAOKPJf+6a9V8G+HfDt/wCFLK5v75Yrl/M3oV6YdgP0Ard/4RHwj/0E0/75oA8N8l/7po8l/wC6a9y/4RHwj/0E0/75o/4RHwj/ANBNP++aAPDfJf8AumjyX/umvcv+ER8I/wDQTT/vmj/hEfCP/QTT/vmgDwxo2UZIIptek/ELQtC0zQIJtMvFmna6VGUDGF2uc/mBXm1ABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQBbs9RmsrXULeNUKX1uLeUsDkKJUkyOeuY1654J+tayeL7uEp9msbC2BkEtwsUbBbltjR/OCxABWSQYTaPnPtjnqKAOgXxbchyj6fYSWQSNI7J1kMUfllmUj592cu55Y53tnOaxbu6mvbye7uHMk88jSSOerMxyT+ZqGigAooooAtw6hPBEsaSEKOgBqT+1rr/nq351QooAv/wBrXX/PVvzo/ta6/wCerfnVCigC/wD2tdf89W/Oj+1rr/nq351QooAs3F9NcxhJHLAHPJqtRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRWvp+jQX2harqB1BUuLCJZRaiIkupljjJLcBRmUY6k7TwODWu/ghftU1vHqW57K4a31ItBgW7LFJIxT5jvAEMo/h5X3zQByNFdQfCtnHbjUZ9Vkj0mRIWhuBagyM0jSKA0e/AwYZcncfu8ZzWBqNjNpep3en3GPPtZngk2nI3KxU4/EUAVqKKKACitix0d7qzjmGMNnv6EirH/CPye350Ac/RXQf8I/J7fnR/wj8nt+dAHP0V0H/CPye350f8I/J7fnQBz9FampaW1lbrK2MFwv6H/CsugAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAv6fqX2Gy1W38rf9vtVt927Hl4milz05/1WMcdc9sHffxuv2qa4j03a97cNcakGnyLhmikjYJ8o2AiaU/xct7YrkaKAOoPiqzkt106fSpJNJjSFYbcXQEitG0jAtJswcmaXI2j73GMVi3GsXtxrlxrCzvBezzvOZYWKFXcknBHI6mqNFACszO5d2LMxySTkk0lFFAFyC/lghWNWIAqT+1J/wC+az6KAND+1J/75o/tSf8Avms+igDQ/tSf++aP7Un/AL5rPooAtXN7JcxhHYkA5qrRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFAGnaaBf31jJeW/2R4oonmdTeQrIETO4+WW3dvTnjGcipX8LazG9sjWeXuH2IqyoxVtu7DgH92dpzhscZPY02C8t7Pw3cwwyZvb6URygKR5cCYbGeh3vtPHTyveuubxLo0eoapcLfb01u8luHCxPmyWS3uI8NkDJDXP8ADniP3AoA4W/0660y4EF0iq5QOpSRXV1PQqykhh7g1VrZ1+6tZU0uztbhblLCz+ztOisqyMZZJCQGAOB5m3kD7tZjzxtaRwi1hSRCS06l9757EFtuB7AUAQ0UUUAatno811aJMqkq2f54qf8A4R+f+4ags9YmtbRIVYhVz/PNT/8ACQT/AN80AH/CPz/3DR/wj8/9w0f8JBP/AHzR/wAJBP8A3zQAf8I/P/cNH/CPz/3DR/wkE/8AfNH/AAkE/wDfNAFXUNLlsoFkdSAW2/of8Kza0tQ1SW9gWN2JAbd+h/xrNoAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigDQTRrx9L/tH/R0tju2mS6iR229dqFgzdewNWH8LazG9sjWeXuH2IqyoxVtu7DgH92dpzhscZPY1qWWqaafDCWt7PZSLFbXEYtjZn7R5rbjGyy4IChipPzLwCMHithvEujR6hqlwt9vTW7yW4cLE+bJZLe4jw2QMkNc/w54j9wKAOFv9OutMuBBdIquUDqUkV1dT0KspIYe4NVa2dfurWVNLs7W4W5Sws/s7TorKsjGWSQkBgDgeZt5A+7WY88bWkcItYUkQktOpfe+exBbbgewFAENFFFAHa6D4C1DWtFt9QggZo5d2CB6MV/pWl/wq7Vf+fZ/++a3fA3xXn8NeDrDSECbbfzMZUH70jN/7NXQ/8L0uvSP/AL5FAHA/8Ku1X/n2f/vmj/hV2q/8+z/98133/C9Lr0j/AO+RR/wvS69I/wDvkUAcD/wq7Vf+fZ/++aP+FXar/wA+z/8AfNd9/wAL0uvSP/vkUf8AC9Lr0j/75FAHk3ijwZe+HdMju7mFkR5hECR3IY/0rk69T+JXxJm8Y+HLfT5AuI7tZ+AB0R1/9mryygAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAUMwGAaXe3qaQKSMgUuxvSgA3t6mje3qaNjelGxvSgA3t6mje3qaNjelGxvSgBCxPU0lKVI6ikoAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKAOi8N+HYNVtr66vppIYYrW5a2EeMzTxQPLjn+EBRuP+0o43ZGu/g3TGvb63jmu1/si6kt9QZnU+bshmlZoxtG3/j3kAB3dVPqK5/R/FWtaErR2OoXKQNFNH5HnOIx5iMhYKCBuG7IPqAe1Wn8aai06TLb2iSNKZroqjf6W5QoTJlj1V3BC7fvt3NAFHW9PtrP+z7my84Wt/a/aI0mYM8eJHjZSwAB+aNiDgcEVQezuY7SO7e2mW2lJWOZkIRyOoDdCRVjU9Tk1SaFmhhgigiEMMEIbZGgJOBuJPJZiSSeSao0AFFFFAHVaPosN3pUM7yorNuyCfRiKvf8I7b/APPaP865SDUJYIVjViAM/wA6k/tWb++aAOn/AOEdt/8AntH+dH/CO2//AD2j/OuY/tWb++aP7Vm/vmgDp/8AhHbf/ntH+dH/AAjtv/z2j/OuY/tWb++aP7Vm/vmgDS1/SorGwSVJEYmULgH2P+Fc5Vu6vZLmMI7EgHNVKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAopwRiMgUeW3pQA2ineW3pR5belADaKd5belHlt6UANopSpAyRSUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFAHRaLpGmanpV8zvcLd21tLcM4mRVXauUUR4LyZIwSCAoOTwOdd/BumNe31vHNdr/AGRdSW+oMzqfN2QzSs0Y2jb/AMe8gAO7qp9RXPWviK4tNOFqlramVYZbeO6ZW82OKQEOgw20ghnGSpI3HBHGLb+NNRadJlt7RJGlM10VRv8AS3KFCZMsequ4IXb99u5oAo63p9tZ/wBn3Nl5wtb+1+0RpMwZ48SPGylgAD80bEHA4Iqg9ncx2kd29tMttKSsczIQjkdQG6EirGp6nJqk0LNDDBFBEIYYIQ2yNAScDcSeSzEkk8k1RoAKKKKAOv0TQ4LzSIJ3lRWbdkE+jEf0rQ/4Rq2/57R/nXH22py29usSsQFz/Opf7Zn/AL5oA6v/AIRq2/57R/nR/wAI1bf89o/zrlP7Zn/vmj+2Z/75oA6v/hGrb/ntH+dH/CNW3/PaP865T+2Z/wC+aP7Zn/vmgDT8R6RDYafHLHIjEyhcA+x/wrmKu3eoSXUQR2JAbNUqACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKALMOnXtxaTXcNncSW0H+tmSJikf+8wGB+NPfSdSi+y+Zp92n2vH2bdCw87OMbOPm6jp6iug0XzrXwvd34vbR3Edxb29m95EjRiRAs0pRmDNlPlUAEk8/wgHpG1Swi1TVZm1G0Meq38s2nss6t9nRra4RC+D+6w00I+bBHlnsKAPNbq0ubG5e2u7eW3nTG6KZCjLkZ5B5qGt3xI8YTR7QTRTXFnYCG4eGRZF3+bI4AZSQ2EdF4JHGO1ZDtbG0jVIZhcgnzJGlBRh2AXaCD77j+FAENFFFAGpaaTLdWqTKpIbP88VN/YM/9w/lW/4f1+3stDt7d4kZk3ZJHPLE/wBa0v8AhKbX/nhF+VAHHf2DP/cP5Uf2DP8A3D+Vdj/wlNr/AM8Ivyo/4Sm1/wCeEX5UAcd/YM/9w/lR/YM/9w/lXY/8JTa/88Ivyo/4Sm1/54RflQBwl/pklnAsjqQC239D/hWfXX+Ktah1HS4oY40UrMGyo/2WH9a5FVLNgdaAEoqTyH9KPIf0oAjoqTyH9KPIf0oAjoqTyH9KPIf0oAjopSMEg9qSgAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigB6yMowDxR5z+tMooAf5z+tHnP60yigB/nP60ec/rTKKAHNIzDBNaXh21S9122t3YKr7sk+yk/0rLqezuGtbtJlOCuf5YoA9H/4Ra1/57R/nR/wi1r/z2j/OuO/t+f8Avn86P7fn/vn86AOx/wCEWtf+e0f50f8ACLWv/PaP8647+35/75/Oj+35/wC+fzoA7H/hFrX/AJ7R/nR/wi1r/wA9o/zrjv7fn/vn86P7fn/vn86AMzUohBqt5CpBEc7qCPZiKrYJ7U+eQy3Ekh6uxY/ia6DS9AkvdNiuFUkPn9CR/SgDnNp9KNp9K6//AIRab+4fyo/4Rab+4fyoA5DafSjafSuv/wCEWm/uH8qP+EWm/uH8qAOQ2n0oIIrr/wDhFpv7h/Ks3W9Fk06zSZlIBkC/oT/SgDB60u0+lW9LtTe6lFbqMl8/oCf6V0n/AAi039w/lQByG0+lG0+ldf8A8ItN/cP5Uf8ACLTf3D+VAHIbT6UbT6V1/wDwi039w/lR/wAItN/cP5UAchtPpSV2H/CLTf3D+VcrdxGC8nhPWORl/I4oAiwT2o2n0ro9L0CS902K4VSQ+f0JH9Kuf8ItN/cP5UAchtPpRtPpXX/8ItN/cP5Uf8ItN/cP5UAchtPpRtPpXX/8ItN/cP5Uf8ItN/cP5UAcgQRSda3tb0WTTrNJmUgGQL+hP9KzdLtTe6lFbqMl8/oCf6UAVNp9KNp9K6//AIRab+4fyo/4Rab+4fyoA5DafSjafSuv/wCEWm/uH8qP+EWm/uH8qAOQ2n0o2n0rr/8AhFpv7h/Kj/hFpv7h/KgDj6XBPapbuIwXk8J6xyMv5HFb2l6BJe6bFcKpIfP6Ej+lAHObT6UbT6V1/wDwi039w/lR/wAItN/cP5UAchtPpRtPpXX/APCLTf3D+VH/AAi039w/lQByG0+lBBFdf/wi039w/lWbreiyadZpMykAyBf0J/pQBg9aXafSrel2pvdSit1GS+f0BP8ASuk/4Rab+4fyoA5DafSjafSuv/4Rab+4fyo/4Rab+4fyoA5DafSjafSuv/4Rab+4fyo/4Rab+4fyoA5DafSkrsP+EWm/uH8q5W7iMF5PCescjL+RxQBFgntRtPpXR6XoEl7psVwqkh8/oSP6Vc/4Rab+4fyoA5DafSjafSuv/wCEWm/uH8qP+EWm/uH8qAOQ2n0o2n0rr/8AhFpv7h/Kj/hFpv7h/KgDkCCKTrW9reiyadZpMykAyBf0J/pWbpdqb3UordRkvn9AT/SgCptPpRtPpXX/APCLTf3D+VH/AAi039w/lQByG0+lG0+ldf8A8ItN/cP5Uf8ACLTf3D+VAHIbT6UbT6V1/wDwi039w/lR/wAItN/cP5UAcfS4J7VLdxGC8nhPWORl/I4re0vQJL3TYrhVJD5/Qkf0oA5zafSjafSuv/4Rab+4fyo/4Rab+4fyoA5DafSjafSuv/4Rab+4fyo/4Rab+4fyoA5DafSggiuv/wCEWm/uH8qzdb0WTTrNJmUgGQL+hP8ASgDB60u0+lW9LtTe6lFbqMl8/oCf6V0n/CLTf3D+VAHIbT6UbT6V1/8Awi039w/lR/wi039w/lQByG0+lG0+ldf/AMItN/cP5Uf8ItN/cP5UAchtPpSV2H/CLTf3D+VcrdxGC8nhPWORl/I4oAiwT2o2n0ro9L0CS902K4VSQ+f0JH9Kuf8ACLTf3D+VAHIbT6UbT6V1/wDwi039w/lR/wAItN/cP5UAchtPpRtPpXX/APCLTf3D+VH/AAi039w/lQByBBFJ1re1vRZNOs0mZSAZAv6E/wBKzdLtTe6lFbqMl8/oCf6UAVNp9KNp9K6//hFpv7h/Kj/hFpv7h/KgDkNp9KNp9K6//hFpv7h/Kj/hFpv7h/KgDkNp9KNp9K6//hFpv7h/Kj/hFpv7h/KgDj6XBPapbuIwXk8J6xyMv5HFb2l6BJe6bFcKpIfP6Ej+lAHObT6UbT6V1/8Awi039w/lR/wi039w/lQByG0+lG0+ldf/AMItN/cP5Uf8ItN/cP5UAchtPpQQRXX/APCLTf3D+VZut6LJp1mkzKQDIF/Qn+lAGD1pdp9Kt6Xam91KK3UZL5/QE/0rpP8AhFpv7h/KgDkNp9KNp9K6/wD4Rab+4fyo/wCEWm/uH8qAOQ2n0o2n0rr/APhFpv7h/Kj/AIRab+4fyoA5DafSkrsP+EWm/uH8q5W7iMF5PCescjL+RxQBFgntRtPpXR6XoEl7psVwqkh8/oSP6Vc/4Rab+4fyoA5DafSjafSuv/4Rab+4fyo/4Rab+4fyoA5DafSjafSuv/4Rab+4fyo/4Rab+4fyoA5Agik61va3osmnWaTMpAMgX9Cf6Vm6Xam91KK3UZL5/QE/0oAqbT6UbT6V1/8Awi039w/lR/wi039w/lQByG0+lG0+ldf/AMItN/cP5Uf8ItN/cP5UAchtPpRtPpXX/wDCLTf3D+VH/CLTf3D+VAHH0uCe1S3cRgvJ4T1jkZfyOK3tL0CS902K4VSQ+f0JH9KAOc2n0o2n0rr/APhFpv7h/Kj/AIRab+4fyoA5DafSjafSuv8A+EWm/uH8qP8AhFpv7h/KgDkNp9KCCK6//hFpv7h/Ks3W9Fk06zSZlIBkC/oT/SgDB60u0+lW9LtTe6lFbqMl8/oCf6V0n/CLTf3D+VAHIbT6UbT6V1//AAi039w/lR/wi039w/lQByG0+lG0+ldf/wAItN/cP5Uf8ItN/cP5UAchtPpSV2H/AAi039w/lXK3cRgvJ4T1jkZfyOKAIsE9qNp9K6PS9AkvdNiuFUkPn9CR/Srn/CLTf3D+VAHIbT6UbT6V1/8Awi039w/lR/wi039w/lQByG0+lG0+ldf/AMItN/cP5Uf8ItN/cP5UAcgQRSVva3osmnWaTMpAMgX9Cf6Vg0AFFFFABRWvYaGuoafc3Ueq2Mb20DzyW8gmDhVOByIymSSoHzdWA4qPVtHGklY31G0nuOPMghEu6PIzyWQLxnHBPNAGZRRRQAUUUUAOCMRkA4o8t/7pr0Dwz4Nm1bw/a3qxkrLv5x6OR/Stb/hXk/8AzxP5UAeU+W/900eW/wDdNerf8K8n/wCeJ/Kj/hXk/wDzxP5UAeU+W/8AdNHlv/dNerf8K8n/AOeJ/Kj/AIV5P/zxP5UAeUFWUZIxSV2/jHwrLoekRXLxlQ04jzj1Vj/SuIoAKKKKACiiigAooooAK6TS9fkstNit1YgJn9ST/Wubpckd6AOv/wCEpm/vn86P+Epm/vn865DcfWjcfWgDr/8AhKZv75/Oj/hKZv75/OuQ3H1o3H1oA6//AISmb++fzrN1vWpNRs0hZiQJA36Ef1rC3H1oJJoAt6XdGy1KK4U4KZ/UEf1rpP8AhKZv75/OuP6Uu4+tAHX/APCUzf3z+dH/AAlM398/nXIbj60bj60Adf8A8JTN/fP50f8ACUzf3z+dchuPrRuPrQB1/wDwlM398/nXK3cpnvJ5j1kkZvzOai3H1pKAOk0vX5LLTYrdWICZ/Uk/1q5/wlM398/nXIZI70bj60Adf/wlM398/nR/wlM398/nXIbj60bj60Adf/wlM398/nR/wlM398/nXIbj60bj60Abut61JqNmkLMSBIG/Qj+tZul3RstSiuFOCmf1BH9aqEk0nSgDsP8AhKZv75/Oj/hKZv75/OuQ3H1o3H1oA6//AISmb++fzo/4Smb++fzrkNx9aNx9aAOv/wCEpm/vn86P+Epm/vn865DcfWjcfWgCW7lM95PMeskjN+ZzW9pevyWWmxW6sQEz+pJ/rXN0uSO9AHX/APCUzf3z+dH/AAlM398/nXIbj60bj60Adf8A8JTN/fP50f8ACUzf3z+dchuPrRuPrQB1/wDwlM398/nWbretSajZpCzEgSBv0I/rWFuPrQSTQBb0u6NlqUVwpwUz+oI/rXSf8JTN/fP51x/Sl3H1oA6//hKZv75/Oj/hKZv75/OuQ3H1o3H1oA6//hKZv75/Oj/hKZv75/OuQ3H1o3H1oA6//hKZv75/OuVu5TPeTzHrJIzfmc1FuPrSUAdJpevyWWmxW6sQEz+pJ/rVz/hKZv75/OuQyR3o3H1oA6//AISmb++fzo/4Smb++fzrkNx9aNx9aAOv/wCEpm/vn86P+Epm/vn865DcfWjcfWgDd1vWpNRs0hZiQJA36Ef1rN0u6NlqUVwpwUz+oI/rVQkmk6UAdh/wlM398/nR/wAJTN/fP51yG4+tG4+tAHX/APCUzf3z+dH/AAlM398/nXIbj60bj60Adf8A8JTN/fP50f8ACUzf3z+dchuPrRuPrQBLdyme8nmPWSRm/M5re0vX5LLTYrdWICZ/Uk/1rm6XJHegDr/+Epm/vn86P+Epm/vn865DcfWjcfWgDr/+Epm/vn86P+Epm/vn865DcfWjcfWgDr/+Epm/vn86zdb1qTUbNIWYkCQN+hH9awtx9aCSaALel3RstSiuFOCmf1BH9a6T/hKZv75/OuP6Uu4+tAHX/wDCUzf3z+dH/CUzf3z+dchuPrRuPrQB1/8AwlM398/nR/wlM398/nXIbj60bj60Adf/AMJTN/fP51yt3KZ7yeY9ZJGb8zmotx9aSgDpNL1+Sy02K3ViAmf1JP8AWrn/AAlM398/nXIZI70bj60Adf8A8JTN/fP50f8ACUzf3z+dchuPrRuPrQB1/wDwlM398/nR/wAJTN/fP51yG4+tG4+tAG7retSajZpCzEgSBv0I/rWbpd0bLUorhTgpn9QR/WqhJNJ0oA7D/hKZv75/Oj/hKZv75/OuQ3H1o3H1oA6//hKZv75/Oj/hKZv75/OuQ3H1o3H1oA6//hKZv75/Oj/hKZv75/OuQ3H1o3H1oAlu5TPeTzHrJIzfmc1vaXr8llpsVurEBM/qSf61zdLkjvQB1/8AwlM398/nR/wlM398/nXIbj60bj60Adf/AMJTN/fP50f8JTN/fP51yG4+tG4+tAHX/wDCUzf3z+dZut61JqNmkLMSBIG/Qj+tYW4+tBJNAFvS7o2WpRXCnBTP6gj+tdJ/wlM398/nXH9KXcfWgDr/APhKZv75/Oj/AISmb++fzrkNx9aNx9aAOv8A+Epm/vn86P8AhKZv75/OuQ3H1o3H1oA6/wD4Smb++fzrlbuUz3k8x6ySM35nNRbj60lAHSaXr8llpsVurEBM/qSf61c/4Smb++fzrkMkd6Nx9aAOv/4Smb++fzo/4Smb++fzrkNx9aNx9aAOv/4Smb++fzo/4Smb++fzrkNx9aNx9aAN3W9ak1GzSFmJAkDfoR/Ws3S7o2WpRXCnBTP6gj+tVCSaTpQB2H/CUzf3z+dH/CUzf3z+dchuPrRuPrQB1/8AwlM398/nR/wlM398/nXIbj60bj60Adf/AMJTN/fP50f8JTN/fP51yG4+tG4+tAEt3KZ7yeY9ZJGb8zmt7S9fkstNit1YgJn9ST/Wubpckd6AOv8A+Epm/vn86P8AhKZv75/OuQ3H1o3H1oA6/wD4Smb++fzo/wCEpm/vn865DcfWjcfWgDr/APhKZv75/Os3W9ak1GzSFmJAkDfoR/WsLcfWgkmgC3pd0bLUorhTgpn9QR/Wuk/4Smb++fzrj+lLuPrQB1//AAlM398/nR/wlM398/nXIbj60bj60Adf/wAJTN/fP50f8JTN/fP51yG4+tG4+tAHX/8ACUzf3z+dcrdyme8nmPWSRm/M5qLcfWkoA6TS9fkstNit1YgJn9ST/Wrn/CUzf3z+dchkjvRuPrQB1/8AwlM398/nR/wlM398/nXIbj60bj60Adf/AMJTN/fP50f8JTN/fP51yG4+tG4+tAG7retSajZpCzEgSBv0I/rWDSkk0lABRRRQBpJqEMHh2SwhVxcXM4e4cgY8tB8ig5zyxYn/AHU9K1/EXiW31bSzbi4v7uV7hJ1N4iD7KAjBo4yp5DFgScL9wcZJrlqKAJba4e1uEnjWJnQ5AliWRfxVgQfxFRsxdyxxknJwAB+Q6UlFABRRRQB6F4Z8ZTaT4ftbJZCFi38Z9XJ/rWt/wsOf/nsfzrykOwGATijzH/vGgD1b/hYc/wDz2P50f8LDn/57H868p8x/7xo8x/7xoA9W/wCFhz/89j+dH/Cw5/8AnsfzrynzH/vGjzH/ALxoA7Xxj4ql1zSIrZ5CwWcSYz6Kw/rXEUpZmGCc0lABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAf/2Q==",
            "image/png": "iVBORw0KGgoAAAANSUhEUgAABSgAAAJYCAIAAADudp+QAABJAElEQVR4Ae3dW2xdVXo48B7HScjVptjYxigdKReBOq7UvkQdWa2gQ5DIMB0yjGSpiiwuzQuah9RK6qpuHjp+SF1iYaJqYKgm8qAOATIyV8uDhPQfSIUYplQUtyIdEG1AviQOsXN1Eif+L43dOI5zPOfss4+97Pz8cFjn22t9+1u/k5ePs8/ev/M7/ggQIECAAAECBAgQIECAAAECBAgQIECAAAECBAgQIECAAAECBAgQIECAAAECBAgQIECAAAECBAgQIECAAAECBAgQIECAAAECBAgQIECAAAECBAgQILAIBDJx7iGTibSw8fHxOMVURYAAAQIECBAgQIAAAQJxCpTGWVZJSUmchV2+fDnOwlRFgAABAgQIECBAgAABAnEKTDXejz766COPPLJmzZrdu3e/9dZbZWVlnZ2dFRUVQ0NDjY2NIyMjySLJtl1aOlVYsgxFWqXxLhKstAQIECBAgAABAgQIEFisApNXdFdWVh46dOiee+7ZtGlTV1fX3Xff3dbWNjAw0N7e3tTUFI42NzcniySDW7t2bbKFxV516tSpYp9CfgIECBAgQIAAAQIECBBYTAKTjfddd91VV1f38ssvr1q16vPPP7/99tt7e3u3bNnS19dXW1vb09MTjiaLJMOqrq5OtrDYq8L/jCj2KeQnQIAAAQIECBAgQIAAgcUkMHlF9ye/+Qsbe/jhh19//fUwqKqqmmgy+/v7wzhxZDFh2QsBAgQIECBAgAABAgQIEMhXYNpPqdevX79r165wwXm+WX7r/NbW1vr6+o6OjnAd+2+dHCacO3cul2nmECBAgAABAgQIECBAgACByAWmGu/Vq1e/+OKLjz322PHjx0PRg4OD4XrvcKl5TU1NGCeOTOy/paUlL4gLFy7kNd9kAgQIECBAgAABAgQIECAQp8DkU7vCc7PDPcz37dv3/vvvTxTa3d3d0NAQxuE1jMMgWSTZtsdi/Uu2HasIECBAgAABAgQIECBA4GYXCA8SO3PmzP/7zd8bb7wROMLDw1555ZXDhw+H1zBOHEkmG/5HQJx/ybZjFQECBAgQIECAAAECBAgQIECAAAECBAgQIECAAAECBAgQIECAAAECBAgQIECAAAECBAgQIECAAAECBAgQIECAAAECBAgQIECAAAECBAgQIECAAAECBAgQIECAAAECBAgQIECAAAECBAgQIECAAAECBAgQIECAAAECBAgQIECAAAECBAgQIECAAAECBAgQIECAAAECBAgQIECAAAECBAgQIECAAAECBAgQIECAAAECBAgQIECAAAECBAgQIECAAAECBAgQIECAAAECBAgQIECAAAECBAgQIECAAAECBAgQSCCQSbBmDpZkMpEWNj4+PgfbdwoCBAgQIECAAAECBAgQWDQCpXHupLQ00sIuXboUp5iqCBAgQIAAAQIECBAgQCBOgan+try8/KmnnnrooYfKyspCreG1s7OzoqJiaGiosbFxZGQkWSTZtpcuXZpsYbFXabyLLSw/AQIECBAgQIAAAQIEFplAydX9vP766//2b/929VLqv/3bv33nnXfq6+vffffdv/mbvwnTkkWu5s9rcEusf3ntwmQCBAgQIECAAAECBAgQIDD1U+rq6uqBgYHh4eHw1Xdw6e3t3bJlS19fX21tbU9PT11dXbJIMuI777wz2cJir/ryyy+LfQr5CRAgQIAAAQIECBAgQGAxCUx94x267ms3VlVVNRHp7+8P43AoWeTanMYECBAgQIAAAQIECBAgQOBmE5j6jXdRd97a2hquWu/o6Ojq6srlROfOnctlmjkECBAgQIAAAQIECBAgQCBygayN9+DgYLj4PFxqXlNTE8ZhG8kiE/tvaWnJC2J0dDSv+SYTIECAAAECBAgQIECAAIE4BaYuNb+uvu7u7oaGhhAMr2EcBski16XN8W24eXicfznWbxoBAgQIECBAgAABAgQIELixQLi52sSB8PCwV1555fDhw+H16gPGEkRufJrfFs3E+vfbCnecAAECBAgQIECAAAECBAgQIECAAAECBAgQIECAAAECBAgQIECAAAECBAgQIECAAAECBAgQIECAAAECBAgQIECAAAECBAgQIECAAAECBAgQIECAAAECBAgQIECAAAECBAgQIECAAAECBAgQIECAAAECBAgQIECAAAECBAgQIECAAAECBAgQIECAAAECBAgQIECAAAECBAgQIECAAAECBAgQIECAAAECBAgQIECAAAECBAgQIECAAAECBAgQIECAAAECBAgQIECAAAECBAgQIECAAAECBAgQIECAAAECBAgQIECAAAECBAgQIECAAAECBAgQIECAAAECBAgQIECAAAECBAgQIECAAAECBAgQIECAAAECBAgQIECAAAECBAgQIECAAAECBAgQIECAAAECBAgQIECAAAECBAgQIECAAAECBAgQIECAAAECBAgQIECAAAECBAgQIECAAAECBAgQIECAAAECBAgQIEBgFoHMLMfm8VBJSck8nn2WU1+5cmWWow4RIECAAAECBAgQIECAAIHrBEqvex/J29LSSAu7ePFiJETKIECAAAECBAgQIECAAIEFIRBpf7ts2bI4+TTecX4uqiJAgAABAgQIECBAgEC0ApON98GDB6urq0OVoePdtGlTRUVFWVlZZ2dnGAwNDTU2No6MjOQSSWuft9xyS1qp0s1z5syZdBPKRoAAAQIECBAgQIAAAQKLW+D633g//vjj69at27NnT1tb28DAQHt7e1NTU2VlZXNzcy6RtLB+7/d+L61U6eb53//933QTykaAAAECBAgQIECAAAECi1tgWuOdyWQ+/PDD+++//9ixY729vVu2bOnr66utre3p6amrq8slkhaWxjstSXkIECBAgAABAgQIECBAYH4Fpv3G+8EHH/zggw9C1x1qqqqqCt94h0F/f38Y5xhJazNnz55NK5U8BAgQIECAAAECBAgQIEBgHgWmNd7hqvIdO3akVU1ra2t9fX1HR0dXV1e+OUdHR/NdYj4BAgQIECBAgAABAgQIEIhQYKrx3rx58/Dw8JEjRyaqHBwcDLdbC5ea19TUhHEI5hK5doctLS3Xvs1r7ObheXGZTIAAAQIECBAgQIAAAQLRCpRcrWz37t379u27+ra7u7uhoSG8Da9hHAa5RK4uL3AwFutfgfuynAABAgQIECBAgAABAgRuUoENGza89957124+PDzslVdeOXz4cHgN43Aol8i1GYwJECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQILEiBTJxVZzKRFjY+Ph6nmKoIECBAgAABAgQIECBAIE6B0kjLKo20sEuXLsUppioCBAgQIECAAAECBAgQiFMg0v72lltuidNL4x3n56IqAgQIECBAgAABAgQIRCsQaeNdVlYWJ9np06fjLExVBAgQIECAAAECBAgQIBCnQKSNd01NTZxeX375ZZyFqYoAAQIECBAgQIAAAQIE4hSYbLzvu+++Z5999ujRo6HKw4cPt7S0hO+cOzs7KyoqhoaGGhsbR0ZGcomktcn169enlSrdPB988EG6CWUjQIAAAQIECBAgQIAAgcUtMHnz8O3bt69ateqZZ565utu2traBgYH29vampqbKysrm5uZcIleXFzj4+7//+wIzFGn5nj17ipRZWgIECBAgQIAAAQIECBBYlAIlE7sKl3b39/dfu8MHHnjg4MGDIRJet27dGga5RK7NYEyAAAECBAgQIECAAAECBAhMXmoeGu8NGzbs2rXrq6++2rlz52effVZVVRW+8Q5AoSEP4zDIJZIW6H/913+llUoeAgQIECBAgAABAgQIECAwjwKTjff4+PhHH320Y8eObdu2Pffcc/fee2/hNbW2ttbX13d0dHR1deWbLXT++S4xnwABAgQIECBAgAABAgQIRCgw2Xg//fTTX3zxRajv1VdfDXdZC4PBwcHq6uq+vr7wZXgY5xi5dofhDm3Xvs1rfN1173mtNZkAAQIECBAgQIAAAQIECMQjMNl4792794UXXghd9+bNmz/++ONQX3d3d0NDQ7i5WngN4xwjaW0s3EQ9rVTyECBAgAABAgQIECBAgACBeRSYvKv5xo0bDxw4MDY2Njo6+sQTT4QrvXN5eNjMOWntZOnSpWmlSjfPpUuX0k0oGwECBAgQIECAAAECBAgsboHJxju2TWYykRYWfgwfm5V6CBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQCBvgUzeK+ZkQSYTaWHj4+NzAuAkBAgQIECAAAECBAgQILBIBErj3EdJSUmchV2+fDnOwlRFgAABAgQIECBAgAABAnEKRNp4L126NE4vjXecn4uqCBAgQIAAAQIECBAgEK1ApI338uXL4yQbHR2NszBVESBAgAABAgQIECBAgECcAtc33g888MDPfvazFStWlJWVdXZ2VlRUDA0NNTY2joyM5BJJa5OhgLRSpZsnOKSbUDYCBAgQIECAAAECBAgQWNwC0+5htmbNmp///Odf//rX165d29bWNjAw0N7e3tTUVFlZ2dzcnEskLawNGzaklSrdPJ9++mm6CWUjQIAAAQIECBAgQIAAgcUtMO0eZnv37n3qqaeuXLkS9hy++j548GAYhNetW7fmGFncWHZHgAABAgQIECBAgAABAgTyFZi61Ly+vv6OO+546aWXfvSjH4UsVVVV4RvvMOjv7w/jHCP5nj7b/DNnzmQ7JE6AAAECBAgQIECAAAECBBaQwGTjHW5m9uSTT27bti3F0ltbW0Mz39HR0dXVlW/a8+fP57vEfAIECBAgQIAAAQIECBAgEKHAZOP93e9+N/zA+6c//WkocfXq1T/5yU8GBwerq6v7+vpqamrCOMRziVy7w5aWlmvf5jW+cOFCXvNNJkCAAAECBAgQIECAAAECC0ZgeHg41BpupfZXf/VXYRBurvYP//APOUbCtFT+lsT6l8ruJCFAgAABAgQIECBAgACBm0dg2l3NJ7YdGu/y8vJcHh42c05acJnMDQpLK3khecbHxwtZbi0BAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAoEgCmSLlLTBtSUlJgRmKtPzKlStFyiwtAQIECBAgQIAAAQIECCxKgdI4d6XxjvNzURUBAgQIECBAgAABAgQI5CsQaeO9dOnSfHcyN/PHxsbm5kTOQoAAAQIECBAgQIAAAQKLQyDSxvuWW26J0/f8+fNxFqYqAgQIECBAgAABAgQIEIhTYLLxXrly5YEDB6qqqtasWbNnz54333yzrKyss7OzoqJiaGiosbFxZGQkl0ham1yxYkVaqdLNc/LkyXQTykaAAAECBAgQIECAAAECi1tg8uZqu3fvDrcNe/LJJ2tqat57772vfe1rbW1tAwMD7e3tTU1NlZWVzc3NuUTSwtq0aVNaqdLN89///d/pJpSNAAECBAgQIECAAAECBBa3wGTjfeutt549e/bixYvf/OY3f/jDH27cuLG3t3fLli19fX21tbU9PT11dXW5RNLC0ninJSkPAQIECBAgQIAAAQIECMyvwOSl5hNXUD///PPbtm379re/HWoKl52Hb7zDoL+/P4xzjKS1mTNnzqSVSh4CBAgQIECAAAECBAgQIDCPAtNurrZ9+/ZDhw6FX3S//fbbhdfU2tpaX1/f0dHR1dWVbzb3MMtXzHwCBAgQIECAAAECBAgQiFNgsvHev3//zp07w7Oy3njjjR//+Meh1sHBwerq6nCpefjVdxjnGLl2ky0tLde+zWs8Ojqa13yTCRAgQIAAAQIECBAgQIBAnAIlE2WFO5Z/5zvfCeNvfOMbR44cCYPu7u6GhoYwCK9hnGMkTEvl71Ksf6nsThICBAgQIECAAAECBAgQuHkEJm+utm7duvDwsJKSknB/te9///uffPJJLg8PmzknLbhQSVqp0s0T7v2ebkLZCBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECEwKZOKUyGQiLWx8fDxOMVURIECAAAECBAgQIECAQJwCpXGWpfGO83NRFQECBAgQIECAAAECBAjkKxBp411aGmlhFy9ezJfYfAIECBAgQIAAAQIECBC4mQUi7W+XLl0a56ei8Y7zc1EVAQIECBAgQIAAAQIEohWYarwfffTRRx55ZM2aNbt3737rrbfKyso6OzsrKiqGhoYaGxtHRkZyiaS1z1tuuSWtVOnmOXv2bLoJZSNAgAABAgQIECBAgACBxS0weQ+zysrKQ4cO3XPPPZs2berq6rr77rvb2toGBgba29ubmprC0ebm5lwiaWGtW7curVTp5jl69Gi6CWUjQIAAAQIECBAgQIAAgcUtMNl433XXXXV1dS+//PKqVas+//zz22+/vbe3d8uWLX19fbW1tT09PeFoLpG0sH7/938/rVTp5vnP//zPdBPKRoAAAQIECBAgQIAAAQKLW2DyUvNPfvMXtvrwww+//vrrYVBVVRW+8Q6D/v7+MM4xEqb5I0CAAAECBAgQIECAAAECBK4KTP3GO4TWr1+/a9eucMH51cOFDFpbW+vr6zs6OsK16/nmOX36dL5LzCdAgAABAgQIECBAgAABAhEKTDXeq1evfvHFFx977LHjx4+HQgcHB6urq8Ol5jU1NWGcY+TaHba0tFz7Nq+xe5jlxWUyAQIECBAgQIAAAQIECEQrMNl4ZzKZcA/zffv2vf/++xO1dnd3NzQ0hJurhdcwDsFcImntc3R0NK1U8hAgQIAAAQIECBAgQIAAgXkUmLy5WniQ2P79+3/1q1+FUs6cOfOtb30rl4eHzZyT1k6WLVuWVqp083iOd7qeshEgQIAAAQIECBAgQGDRC0w23rHts6SkJLaSJuq5cuVKnIWpigABAgQIECBAgAABAgTiFJj6jXdU9Y2Pj0dVj2IIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBwY4HMjcPzHc1kIi1sfHx8vm2cnwABAgQIECBAgAABAgQWkkBpnMWWlJTEWdjly5fjLExVBAgQIECAAAECBAgQIBCnQKSNd2lppIVpvOP8d6wqAgQIECBAgAABAgQIRCsQaX+7cuXKOMkuXLgQZ2GqIkCAAAECBAgQIECAAIE4BSJtvMvKyuL0OnnyZJyFqYoAAQIECBAgQIAAAQIE4hSItPGuqamJ0+t//ud/4ixMVQQIECBAgAABAgQIECAQp0CkjfeGDRvi9HrvvffiLExVBAgQIECAAAECBAgQIBCnwFTjXV5e/tRTTz300EMTl3nfd999zz777NGjR0Pdhw8fbmlpCfHOzs6KioqhoaHGxsaRkZGZkbQ2+Qd/8AdppZKHAAECBAgQIECAAAECBAjMo8DU47Lffffdl1566Qc/+EHowENB27dvX7Vq1TPPPHO1uLa2toGBgfb29qampsrKyubm5pmRq5MLHPziF78oMEORlv/pn/5pkTJLS4AAAQIECBAgQIAAAQKLUmDqcdnf+9739u/ff3WT4VfW/f39V9+GwQMPPHDw4MEwCK9bt269YeTa+cYECBAgQIAAAQIECBAgQIDA1KXm4dvsazlC4x1+aL1r166vvvpq586dn332WVVV1cSc0JCHcZg8M3JthkLGv/zlLwtZbi0BAgQIECBAgAABAgQIEIhEYKrxvq6g8fHxjz76aMeOHdu2bXvuuefuvffe6yb81retra319fUdHR1dXV2/dfJ1E/7jP/7juoi3BAgQIECAAAECBAgQIEBgIQpkbbyffvrpL774Imzp1VdfDXdZC4PBwcHq6uq+vr7wZXgY3zByLUG4H9u1b/Maf/rpp3nNN5kAAQIECBAgQIAAAQIECMQpkLXx3rt37wsvvBC67s2bN3/88ceh+u7u7oaGhnBztfAaxjeMpLXJ635enlZaeQgQIECAAAECBAgQIECAwBwLTN3VfOLEw8PDE3c137hx44EDB8bGxkZHR5944onwG++ZDw+bGUmr+ltvvTWtVOnmOXnyZLoJZSNAgAABAgQIECBAgACBxS1wfeMdyW6XL18eSSXXlXHhwoXrIt4SIECAAAECBAgQIECAAIFZBLJeaj7Lmjk4FL5pn4OzOAUBAgQIECBAgAABAgQIECi2QKTfeGcykRYWbvZe7I9EfgIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgcAOBzA1iEYQymUgLGx8fj4BHCQQIECBAgAABAgQIECCwYARK46y0pKQkzsIuX74cZ2GqIkCAAAECBAgQIECAAIE4BSJtvJcsWRKnl8Y7zs9FVQQIECBAgAABAgQIEIhWINLGe9myZXGSXbx4Mc7CVEWAAAECBAgQIECAAAECcQpE2nivWLEiTq8zZ87EWZiqCBAgQIAAAQIECBAgQCBOganGu7y8/KmnnnrooYfKyspCreG1s7OzoqJiaGiosbFxZGQkl0ham1yzZk1aqdLNc/z48XQTykaAAAECBAgQIECAAAECi1tg6ubh77777ksvvfSDH/wgdOBhz21tbQMDA+3t7U1NTZWVlc3NzblE0sL6wz/8w7RSpZvn3//939NNKBsBAgQIECBAgAABAgQILG6Bqca7uro6dNrDw8MTjXdvb++WLVv6+vpqa2t7enrq6upyiaSF9Wd/9mdppUo3z9tvv51uQtkIECBAgAABAgQIECBAYHELTD21K3Td1261qqpqItLf3x/G4VAukWszGBMgQIAAAQIECBAgQIAAAQJTv/FO3aK1tbW+vr6jo6Orqyvf5F999VW+S8wnQIAAAQIECBAgQIAAAQIRCmRtvAcHB8PF5+FS85qamjAOpecSuXaHLS0t177Naxzu5ZbXfJMJECBAgAABAgQIECBAgECcAlkb7+7u7oaGhnBztfAaxqH6XCJpbfL06dNppZKHAAECBAgQIECAAAECBAjMo8DUzdUmirh6c7VcHh42c05aO1m9enVaqdLN4zne6XrKRoAAAQIECBAgQIAAgUUvcH3jHcmGly1bFkkl15Vx8eLF6yLeEiBAgAABAgQIECBAgACBWQSyXmo+y5o5OHT58uU5OItTECBAgAABAgQIECBAgACBYgtE+o13JhNpYePj48X+SOQnQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQKBggUzBGYqSoKSkpCh5C0565cqVgnNIQIAAAQIECBAgQIAAAQI3kUBpnHvVeMf5uaiKAAECBAgQIECAAAECBPIViLTxXrp0ab47mZv5Y2Njc3MiZyFAgAABAgQIECBAgACBxSEQaeO9fPnyOH3Pnz8fZ2GqIkCAAAECBAgQIECAAIE4BSJtvFeuXBmn1/DwcJyFqYoAAQIECBAgQIAAAQIE4hSItPFeu3ZtnF59fX1xFqYqAgQIECBAgAABAgQIEIhTYLLxPnjwYHV1dShx2bJlmzZtqqioKCsr6+zsDIOhoaHGxsaRkZGZkeJtqby8vHjJZSZAgAABAgQIECBAgAABAnMmcP3jxB5//PF169bt2bOnra1tYGCgvb29qampsrKyubl5ZqR4VW7durV4yQvJ/Oabbxay3FoCBAgQIECAAAECBAgQuNkEpjXemUzmww8/vP/++48dO9bb27tly5ZwZXVtbW1PT09dXd3MSPGwNN7Fs5WZAAECBAgQIECAAAECBOZSYNpvvB988MEPPvggdN2hgqqqqvCNdxj09/eH8Q0jxSv0xIkTxUsuMwECBAgQIECAAAECBAgQmDOBaY13uKp8x44dhZy7tbW1vr6+o6Ojq6urkDxuHl6InrUECBAgQIAAAQIECBAgEI/AVOO9efPm0O4eOXJkorjBwcFwu7VwqXlNTU0Yh+DMyMxttLS0zAwmiJw6dSrBKksIECBAgAABAgQIECBAgEBsAlON9+7du/ft23e1vu7u7oaGhnBztfAaxiE+M3J1cuqDc+fOpZ5TQgIECBAgQIAAAQIECBAgMPcCkzdX27Bhw/PPP//Hf/zHVyuY+fCwmZGrk1MfrFixIvWcqSQ8f/58KnkkIUCAAAECBAgQIECAAIGbRGDaXc3j2XNp6dRX8fFUFSoZGxuLqh7FECBAgAABAgQIECBAgEDkApH2t1euXIkcTnkECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIEbiCQuUEsglAmE2lh4+PjEfAogQABAgQIECBAgAABAgQWjEBpnJUuWbIkzsLGxsbiLExVBAgQIECAAAECBAgQIBCnQKSN97Jly+L00njH+bmoigABAgQIECBAgAABAtEKRNp4r1q1Kk6yc+fOxVmYqggQIECAAAECBAgQIEAgToFIG+/y8vI4vY4fPx5nYaoiQIAAAQIECBAgQIAAgTgFIm28a2pq4vT69a9/HWdhqiJAgAABAgQIECBAgACBOAUibbzXr18fp9c777wTZ2GqIkCAAAECBAgQIECAAIE4BSYb7/vuu+/ZZ589evRoqPLw4cMtLS1lZWWdnZ0VFRVDQ0ONjY0jIyMzI8XbUl1dXfGSy0yAAAECBAgQIECAAAECBOZMYPJx2du3bw/3M3vmmWeunritrW1gYKC9vb2pqamysrK5uXlm5Ork1Aeh+U89ZyoJ6+vrU8kjCQECBAgQIECAAAECBAjcXAK7d+/+8z//82v33Nvbe8cdd4RIbW3txx9/HAYzI9fOT3c8HutfutuUjQABAgQIECBAgAABAgQWvUDJxA7Dzcy2bt0avmd+7bXXJn5fXVVVFb7xDkf7+/vDOAxmRha9jg0SIECAAAECBAgQIECAAIECBSZ/4x2+YP7oo4927Nixbdu255577t57702Wt7W1NVyM3dHR0dXVlSzDxKp//dd/LWS5tQQIECBAgAABAgQIECBAIBKBycb76aef/uKLL0JNr776arjLWhgMDg5WV1f39fWFL8PD+IaRmXsId2WbGUwQ+eUvf5lglSUECBAgQIAAAQIECBAgQCA2gcnGe+/evS+88ELoujdv3jzxi+7u7u6GhoZwc7XwGsah7pmR4m1moobi5ZeZAAECBAgQIECAAAECBAjMjcDkXc03btx44MCBsbGx0dHRJ5544rPPPpv58LCZkeKV+Cd/8ifFS15IZs/xLkTPWgIECBAgQIAAAQIECNyEApONd2w7D/8jILaSJur59a9/HWdhqiJAgAABAgQIECBAgACBOAUmLzWPrbjh4eHYSlIPAQIECBAgQIAAAQIECBBIIBBp43327NkEm7GEAAECBAgQIECAAAECBAjEJhBp433x4sXYpNRDgAABAgQIECBAgAABAgQSCET6G+9MJtLCwgPPEyhbQoAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECMyzQGaez5/l9JlMpIWNj49nKVmYAAECBAgQIECAAAECBAjcQKD0BrEIQkuWLImgihuUMDY2doOoEAECBAgQIECAAAECBAgQyCIQaeNdWhppYRrvLP+QhAkQIECAAAECBAgQIEDgxgKR9rfLli27cb3zHR0dHZ3vEpyfAAECBAgQIECAAAECBBaSQKSN94oVK+JUPHXqVJyFqYoAAQIECBAgQIAAAQIE4hSItPFes2ZNnF6Dg4NxFqYqAgQIECBAgAABAgQIEIhT4PrG+4EHHvjZz34WvnAuKyvr7OysqKgYGhpqbGwcGRmZGSnelsK5ipdcZgIECBAgQIAAAQIECBAgMGcC057aFb5n/vnPf/71r3997dq1bW1tAwMD7e3tTU1NlZWVzc3NMyPFq3LLli3FS15I5rfeequQ5dYSIECAAAECBAgQIECAwM0mMK3x/qd/+qdf/OIXP/rRj8rLy3t7e0P329fXV1tb29PTU1dXNzNSPKy/+Iu/KF7yQjL/y7/8SyHLrSVAgAABAgQIECBAgACBm02g5OqG6+vr77jjjpdeemkiUlVVFb7xDuP+/v4wDoOZkatrDQgQIECAAAECBAgQIECAAIEbCkz+xnv58uVPPvnktm3bbjgp92Bra2to4Ds6Orq6unJfNXPm8ePHZwZFCBAgQIAAAQIECBAgQIDAghOYbLy/+93vhh94//SnPw0bWL169U9+8pNw++7q6upwqXlNTc3ErbxnRmbutqWlZWYwQeTEiRMJVllCgAABAgQIECBAgAABAgRiE5j2G++J4oaHh8NvvK+9ldrtt9/+13/91zMjxdvMhg0bipe8kMyffvppIcutJUCAAAECBAgQIECAAIGbTSBr4z3z4WEzI8XDmvhVefHyJ8488eV/4uUWEiBAgAABAgQIECBAgMDNJnCDxjsGgvA8sxjKmFnDqVOnZgZFCBAgQIAAAQIECBAgQIBANoHJ33hnOzxf8YsXL87XqZ2XAAECBAgQIECAAAECBAikKBBp4z02NpbiJqUiQIAAAQIECBAgQIAAAQLzJRDppeaZTKSFjY+Pz9dH5bwECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBAgQIAAAQIECBBYCAKZOIssKSmJs7ArV67EWZiqCBAgQIAAAQIECBAgQCBOgdI4y1qyZEmchWm84/xcVEWAAAECBAgQIECAAIFoBSJtvJcuXRon2aVLl+IsTFUECBAgQIAAAQIECBAgEKdApI338uXL4/Q6d+5cnIWpigABAgQIECBAgAABAgTiFIi08V61alWcXidPnoyzMFURIECAAAECBAgQIECAQJwCkTbea9eujdNLVQQIECBAgAABAgQIECBAIC+BycZ75cqVBw4cqKqqWrNmzZ49e958882ysrLOzs6KioqhoaHGxsaRkZGZkbzOlNfkW2+9Na/5JhMgQIAAAQIECBAgQIAAgTgFJh8ntnv37nC/7ieffLKmpua999772te+1tbWNjAw0N7e3tTUVFlZ2dzcPDNSvC19+9vfLl7yQjK/9tprhSy3lgABAgQIECBAgAABAgRuNoHJxjt8w3z27NmLFy9+85vf/OEPf7hx48be3t4tW7b09fXV1tb29PTU1dXNjBQP6/HHHy9e8kIy//M//3Mhy60lQIAAAQIECBAgQIAAgZtNYPJS84l7hj3//PPbtm2b+LY5XHYevvEOHP39/WEcBjMjNxuW/RIgQIAAAQIECBAgQIAAgXwFpt1cbfv27YcOHQq/6H777bfzTTQxv7W1tb6+vqOjo6urK1mGiVXHjh0rZLm1BAgQIECAAAECBAgQIEAgEoHJxnv//v07d+4cGxt74403fvzjH4fiBgcHq6urw6Xm4VffYXzDyMw9tLS0zAwmiJw4cSLBKksIECBAgAABAgQIECBAgEBsApONd7hj+Xe+853wdfc3vvGNI0eOhCq7u7sbGhrCzdXCaxjfMFK8zXhcdvFsZSZAgAABAgQIECBAgACBuRSYvLnaunXrwsPDSkpKwv3Vvv/973/yySczHx42M1K8Qu+8887iJS8k85dfflnIcmsJECBAgAABAgQIECBA4GYTmGy8Y9t2tM/x9lV8bP9U1EOAAAECBAgQIECAAIHIBabdXC2eWi9cuBBPMSohQIAAAQIECBAgQIAAAQKJBSJtvC9dupR4SxYSIECAAAECBAgQIECAAIF4BCJtvC9fvhyPkUoIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIEEgqkEm6sLjrMplICxsfHy/uzmUnQIAAAQIECBAgQIAAgcUlUBrndpYsWRJnYWNjY3EWpioCBAgQIECAAAECBAgQiFMg0sa7tDTSwjTecf47VhUBAgQIECBAgAABAgSiFYi0v12+fHmcZKOjo3EWpioCBAgQIECAAAECBAgQiFMg0sZ7xYoVcXqNjIzEWZiqCBAgQIAAAQIECBAgQCBOgUgb79WrV8fppSoCBAgQIECAAAECBAgQIJCXQKSNd3l5eV7bMJkAAQIECBAgQIAAAQIECMQpMNV4P/roo4888siaNWt279791ltvlZWVdXZ2VlRUDA0NNTY2hkusZ0aKt6XbbruteMllJkCAAAECBAgQIECAAAECcyYw+bjsysrKQ4cO3XPPPZs2berq6rr77rvb2toGBgba29ubmprC0ebm5pmR4lW5ffv24iUvJPPzzz9fyHJrCRAgQIAAAQIECBAgQOBmE5hsvO+66666urqXX3551apVn3/++e23397b27tly5a+vr7a2tqenp5wdGakeFi7du0qXvJCMv/jP/5jIcutJUCAAAECBAgQIECAAIGbTWDyUvNPfvMXNv/www+//vrrYVBVVRW+8Q6D/v7+ML5hJAT9ESBAgAABAgQIECBAgAABArMITP3GO0xav359+Ko5XHA+y4LZD7W2ttbX13d0dITr1WefOfvRiZ5/9jmOEiBAgAABAgQIECBAgACB+AWmGu/wBK8XX3zxscceO378eKh7cHCwuro6XGpeU1MTxjeMzNxeS0vLzGCCyLFjxxKsmt8lJSUl81tAtrNfuXIl2yFxAgQIECBAgAABAgQIECi2wGTjnclkwj3M9+3b9/7770+csru7u6GhIdxcLbyGcQjOjBSvuBMnThQveZEyL1mypEiZC0yr8S4Q0HICBAgQIECAAAECBAgUIjB5c7XwILH9+/f/6le/CrnOnDnzrW99a+bDw2ZGCjnx7Gs3bNgw+4T5Ovrpp59mO/XKlSuzHZrf+Llz5+a3AGcnQIAAAQIECBAgQIDAzSww2XjHRhCuco+tpIl6Zvnx+a233hpnzSdPnoyzMFURIECAAAECBAgQIEDgZhCY+o13VLs9f/58VPXkUky033hrvHP5+MwhQIAAAQIECBAgQIBAkQQibbwvXLhQpA0XL+2aNWuKl1xmAgQIECBAgAABAgQIEFigApE23mNjYwsOtLy8fMHVrGACBAgQIECAAAECBAgQKLZApI335cuXi73z1PPfdtttqeeUkAABAgQIECBAgAABAgQWukCkjff4+PiCk62srFxwNSuYAAECBAgQIECAAAECBIotEGnjXextFyN/VVVVMdLKSYAAAQIECBAgQIAAAQILWkDjvaA/vkKLLykpKTRFcdZfuXIlW2I1Z5NJEOecAC3BEs4J0BIs4ZwALcESzgnQEizhnAAtwRLOCdASLOGcAC3BEs4J0BIsmcV59mwa79l98jg6ODiYx+w4pi5ZsiSOQq6vYpZ/0Gq+HquA95wLwMtjKec8sAqYyrkAvDyWcs4Dq4CpnAvAy2Mp5zywCpjKuQC8PJZyzgOrgKmzOM+eVeM9u08eR48fP57H7DimlpZG+g/g0qVL2YTUnE0mQZxzArQESzgnQEuwhHMCtARLOCdAS7CEcwK0BEs4J0BLsIRzArQESzgnQEuwZBbn2bNF2nfNXnScR0+cOBFnYbNUtWzZslmOzuOh8+fPZzu7mrPJJIhzToCWYAnnBGgJlnBOgJZgCecEaAmWcE6AlmAJ5wRoCZZwToCWYAnnBGgJlsziPHs2jffsPnkcHR4ezmN2HFNXrFgRRyHXVzEyMnJ96P/eq/n/JFL4L+cUEHNIwTkHpBSmcE4BMYcUnHNASmEK5xQQc0jBOQekFKZwTgExhxScc0BKYcoszrNn13jP7pPH0dOnT+cxO46pq1atiqOQPKpQcx5YBUzlXABeHks554FVwFTOBeDlsZRzHlgFTOVcAF4eSznngVXAVM4F4OWxlHMeWEWbqvFOjfbcuXOp5ZqrRGvXrp2rU6V2HjWnRjlrIs6z8qR2kHNqlLMm4jwrT2oHOadGOWsizrPypHaQc2qUsybiPCtPagc5p0ZZQCKNdwF405deuHBhemABvCsvL18AVU4vUc3TPYr1jnOxZKfn5Tzdo1jvOBdLdnpeztM9ivWOc7Fkp+flPN2jWO84F0t2el7O0z3m553GOzX3xDe4S62C/BP97u/+bv6L5nmFmufmA+DMOZuAfxvZZNKNc07XM1s2ztlk0o1zTtczWzbO2WTSjXNO1zNbtoXonG0vE3GN9+w+eRy9fPlyHrPjmHrbbbfFUUgeVag5D6wCpnIuAC+PpZzzwCpgKucC8PJYyjkPrAKmci4AL4+lnPPAKmAq5wLw8ljKOQ+sok3VeKdGm/hZ6qlVIBEBAgQIECBAgAABAgQIxCeg8Y7vM5nDihbis8fVPDf/QDhzzibg30Y2mXTjnNP1zJaNczaZdOOc0/XMlo1zNpl045zT9cyWbSE6Z9vLRFzjPbvPIj/61VdfLbgdqnluPjLOnLMJ+LeRTSbdOOd0PbNl45xNJt0453Q9s2XjnE0m3TjndD2zZVuIztn2MhHXeM/us8iPDg8PL7gdqnluPjLOnLMJ+LeRTSbdOOd0PbNl45xNJt0453Q9s2XjnE0m3TjndD2zZVuIztn2MhHXeM/us8iPnjp1asHtUM1z85Fx5pxNwL+NbDLpxjmn65ktG+dsMunGOafrmS0b52wy6cY5p+uZLdtCdM62l4m4xnt2n0V+9OzZswtuh2qem4+MM+dsAv5tZJNJN845Xc9s2Thnk0k3zjldz2zZOGeTSTfOOV3PbNkWonO2vUzEI228ly5dOnvd83V0IT6sexar8+fPz3I0zkNqnpvPhTPnbAL+bWSTSTfOOV3PbNk4Z5NJN845Xc9s2Thnk0k3zjldz2zZFqJztr1MxCNtvJcvXz573fN1dJE13hcvXpwvycTnVXNiurwWcs6LK/Fkzonp8lrIOS+uxJM5J6bLayHnvLgST+acmC6vhZzz4ko8mXNiuhQXRtp4r1ixIsVNppjqzJkzKWab91RjY2PzXkO+Bag5X7Fk8zknc8t3Fed8xZLN55zMLd9VnPMVSzafczK3fFdxzlcs2XzOydzyXcU5X7FizI+08V6zZk0xdlt4zuPHjxeeJJ4Mly9fjqeYHCtRc45QBU7jXCBgjss55whV4DTOBQLmuJxzjlAFTuNcIGCOyznnCFXgNM4FAua4nHOOUEWdFmnjXV5eXtRtSz4hcOXKlQVHoea5+cg4c84m4N9GNpl045zT9cyWjXM2mXTjnNP1zJaNczaZdOOc0/XMlm0hOmfby0Q80sb7tttum71uRwkQIECAAAECBAgQIECAwIIQiLTxrqqqWhB8iiRAgAABAgQIECBAgAABArMLTDbeBw8erK6uDlOXLVu2adOmioqK++6779lnnz169GgIHj58uKWlZfZE6R6tqalJN6FsBAgQIECAAAECBAgQIEBgXgQmG++GhoaJ0z/++OPr1q0L49CHt7W1PfPMM/NS1p133jkv53VSAgQIECBAgAABAgQIECCQrsC0S80zmcwTTzxx//33h3OE75yPHDmS7slyzzbR/Oc+30wCBAgQIECAAAECBAgQIBCnQMm1ZT344IMffPDBsWPHQjA03lu3bg0Xmb/22mvr16+/dpoxAQIECBAgQIAAAQIECBAgkKPAtG+8m5qaduzYMbFyfHz8o48+Cm+3bdv23HPP3Xvvvdkytra21tfXd3R0dHV1ZZuTb3zit+X5rjKfAAECBAgQIECAAAECBAjEJjDVeG/evHl4ePjq5eVPP/30F198Ecp99dVXw13WZqm7GPdd+/LLL2c5o0MECBAgQIAAAQIECBAgQGChCEw13rt37963b9/Vuvfu3fvCCy+Erjs05B9//PHV+NwM+vv75+ZEzkKAAAECBAgQIECAAAECBIoqMNl4b9iw4Y477njnnXeunuzv/u7vDhw4sHPnztHR0b/8y7+8Gp+bweDg4NycyFkIECBAgAABAgQIECBAgEBRBTJFzZ44+R/90R8lXlvUhR9++GFR80tOgAABAgQIECBAgAABAotMYOpS86g2Fn5tHlU9iiFAgAABAgQIECBAgAABAskEIm28T58+nWw/VhEgQIAAAQIECBAgQIAAgagEIm28z58/HxWTYggQIECAAAECBAgQIECAQDKBSBvvCxcuJNuPVQQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQIECBAgAABAgQI/FaB/w8WrVd6LocUDgAAAABJRU5ErkJggg==",
            "text/plain": [
              "<PIL.Image.Image image mode=RGB size=1320x600>"
            ]
          },
          "execution_count": 15,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "simulate_gel_electrophoresis(list_of_amplicons)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 16,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>Amplicon Name</th>\n",
              "      <th>5' Overhang</th>\n",
              "      <th>5' Duplicate</th>\n",
              "      <th>3' Overhang</th>\n",
              "      <th>3' Duplicate</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>Amplicon_chosen_region_2(4000-7000)_34_pcr_pro...</td>\n",
              "      <td>CATG</td>\n",
              "      <td>No</td>\n",
              "      <td>GCTC</td>\n",
              "      <td>No</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>Amplicon_chosen_region_4(15000-20000)_54_pcr_p...</td>\n",
              "      <td>GCTC</td>\n",
              "      <td>No</td>\n",
              "      <td>GTTG</td>\n",
              "      <td>No</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>Amplicon_chosen_region_1(80000-100000)_80_pcr_...</td>\n",
              "      <td>GTTG</td>\n",
              "      <td>No</td>\n",
              "      <td>TTGT</td>\n",
              "      <td>No</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>Amplicon_chosen_region_1(80000-100000)_81_pcr_...</td>\n",
              "      <td>TTGT</td>\n",
              "      <td>No</td>\n",
              "      <td>TGCC</td>\n",
              "      <td>No</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>Amplicon_chosen_region_3(9000-14000)_118_pcr_p...</td>\n",
              "      <td>TGCC</td>\n",
              "      <td>No</td>\n",
              "      <td>ATTC</td>\n",
              "      <td>No</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>5</th>\n",
              "      <td>Amplicon_chosen_region_3(9000-14000)_146_pcr_p...</td>\n",
              "      <td>ACGC</td>\n",
              "      <td>No</td>\n",
              "      <td>ATCC</td>\n",
              "      <td>No</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>6</th>\n",
              "      <td>Amplicon_chosen_region_2(4000-7000)_147_pcr_pr...</td>\n",
              "      <td>ATCC</td>\n",
              "      <td>No</td>\n",
              "      <td>CATC</td>\n",
              "      <td>No</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>7</th>\n",
              "      <td>Amplicon_chosen_region_2(4000-7000)_148_pcr_pr...</td>\n",
              "      <td>CATC</td>\n",
              "      <td>No</td>\n",
              "      <td>GCCA</td>\n",
              "      <td>No</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>8</th>\n",
              "      <td>Amplicon_chosen_region_4(15000-20000)_166_pcr_...</td>\n",
              "      <td>GCCA</td>\n",
              "      <td>No</td>\n",
              "      <td>CGAC</td>\n",
              "      <td>Yes</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>9</th>\n",
              "      <td>Amplicon_chosen_region_3(9000-14000)_200_pcr_p...</td>\n",
              "      <td>CGAC</td>\n",
              "      <td>Yes</td>\n",
              "      <td>CCTC</td>\n",
              "      <td>No</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>10</th>\n",
              "      <td>Amplicon_chosen_region_3(9000-14000)_203_pcr_p...</td>\n",
              "      <td>CCTC</td>\n",
              "      <td>No</td>\n",
              "      <td>TCAC</td>\n",
              "      <td>No</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>11</th>\n",
              "      <td>Amplicon_chosen_region_4(15000-20000)_254_pcr_...</td>\n",
              "      <td>TCAC</td>\n",
              "      <td>No</td>\n",
              "      <td>CGAC</td>\n",
              "      <td>Yes</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>12</th>\n",
              "      <td>Amplicon_chosen_region_1(80000-100000)_256_pcr...</td>\n",
              "      <td>CGAC</td>\n",
              "      <td>Yes</td>\n",
              "      <td>CGTA</td>\n",
              "      <td>No</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>13</th>\n",
              "      <td>Amplicon_chosen_region_4(15000-20000)_280_pcr_...</td>\n",
              "      <td>CGTA</td>\n",
              "      <td>No</td>\n",
              "      <td>TCTC</td>\n",
              "      <td>No</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>14</th>\n",
              "      <td>Amplicon_chosen_region_4(15000-20000)_295_pcr_...</td>\n",
              "      <td>TCTC</td>\n",
              "      <td>No</td>\n",
              "      <td>GAAG</td>\n",
              "      <td>No</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>15</th>\n",
              "      <td>Amplicon_chosen_region_2(4000-7000)_386_pcr_pr...</td>\n",
              "      <td>GAAG</td>\n",
              "      <td>No</td>\n",
              "      <td>AAGG</td>\n",
              "      <td>No</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>16</th>\n",
              "      <td>Amplicon_chosen_region_2(4000-7000)_387_pcr_pr...</td>\n",
              "      <td>AAGG</td>\n",
              "      <td>No</td>\n",
              "      <td>TGAT</td>\n",
              "      <td>No</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>17</th>\n",
              "      <td>Amplicon_chosen_region_3(9000-14000)_424_pcr_p...</td>\n",
              "      <td>TGAT</td>\n",
              "      <td>No</td>\n",
              "      <td>GCGG</td>\n",
              "      <td>No</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>18</th>\n",
              "      <td>Amplicon_chosen_region_1(80000-100000)_455_pcr...</td>\n",
              "      <td>GCGG</td>\n",
              "      <td>No</td>\n",
              "      <td>TACC</td>\n",
              "      <td>No</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>19</th>\n",
              "      <td>Amplicon_chosen_region_1(80000-100000)_547_pcr...</td>\n",
              "      <td>TACC</td>\n",
              "      <td>No</td>\n",
              "      <td>CTAG</td>\n",
              "      <td>No</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "                                        Amplicon Name 5' Overhang  \\\n",
              "0   Amplicon_chosen_region_2(4000-7000)_34_pcr_pro...        CATG   \n",
              "1   Amplicon_chosen_region_4(15000-20000)_54_pcr_p...        GCTC   \n",
              "2   Amplicon_chosen_region_1(80000-100000)_80_pcr_...        GTTG   \n",
              "3   Amplicon_chosen_region_1(80000-100000)_81_pcr_...        TTGT   \n",
              "4   Amplicon_chosen_region_3(9000-14000)_118_pcr_p...        TGCC   \n",
              "5   Amplicon_chosen_region_3(9000-14000)_146_pcr_p...        ACGC   \n",
              "6   Amplicon_chosen_region_2(4000-7000)_147_pcr_pr...        ATCC   \n",
              "7   Amplicon_chosen_region_2(4000-7000)_148_pcr_pr...        CATC   \n",
              "8   Amplicon_chosen_region_4(15000-20000)_166_pcr_...        GCCA   \n",
              "9   Amplicon_chosen_region_3(9000-14000)_200_pcr_p...        CGAC   \n",
              "10  Amplicon_chosen_region_3(9000-14000)_203_pcr_p...        CCTC   \n",
              "11  Amplicon_chosen_region_4(15000-20000)_254_pcr_...        TCAC   \n",
              "12  Amplicon_chosen_region_1(80000-100000)_256_pcr...        CGAC   \n",
              "13  Amplicon_chosen_region_4(15000-20000)_280_pcr_...        CGTA   \n",
              "14  Amplicon_chosen_region_4(15000-20000)_295_pcr_...        TCTC   \n",
              "15  Amplicon_chosen_region_2(4000-7000)_386_pcr_pr...        GAAG   \n",
              "16  Amplicon_chosen_region_2(4000-7000)_387_pcr_pr...        AAGG   \n",
              "17  Amplicon_chosen_region_3(9000-14000)_424_pcr_p...        TGAT   \n",
              "18  Amplicon_chosen_region_1(80000-100000)_455_pcr...        GCGG   \n",
              "19  Amplicon_chosen_region_1(80000-100000)_547_pcr...        TACC   \n",
              "\n",
              "   5' Duplicate 3' Overhang 3' Duplicate  \n",
              "0            No        GCTC           No  \n",
              "1            No        GTTG           No  \n",
              "2            No        TTGT           No  \n",
              "3            No        TGCC           No  \n",
              "4            No        ATTC           No  \n",
              "5            No        ATCC           No  \n",
              "6            No        CATC           No  \n",
              "7            No        GCCA           No  \n",
              "8            No        CGAC          Yes  \n",
              "9           Yes        CCTC           No  \n",
              "10           No        TCAC           No  \n",
              "11           No        CGAC          Yes  \n",
              "12          Yes        CGTA           No  \n",
              "13           No        TCTC           No  \n",
              "14           No        GAAG           No  \n",
              "15           No        AAGG           No  \n",
              "16           No        TGAT           No  \n",
              "17           No        GCGG           No  \n",
              "18           No        TACC           No  \n",
              "19           No        CTAG           No  "
            ]
          },
          "execution_count": 16,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "overhangs = create_overhang_dataframe(list_of_amplicons)\n",
        "overhangs"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 17,
      "metadata": {},
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "Dseqrecord(-138)\n",
            "CATGGTTCACTGCCGTATAGGCAGCTAAGAAACCCAGACACCACGCACGCCA\u001b[48;5;11mGTTTTAGAGCTAGAAATAGCAAGTTAAAATAAGGC\u001b[0mTAGTCCGTTATCAACTTGAAAAAGTGGCACCGAGTCGGTGCTTTTTT    \n",
            "    CAAGTGACGGCATATCCGTCGATTCTTTGGGTCTGTGGTGCGTGCGGTCAAAATCTCGATCTTTATCGTTCAATTTTATTCCGATCAGGCAATAGTTGAACTTTTTCACCGTGGCTCAGCCACGAAAAAACGAG\n",
            "Dseqrecord(-106)\n",
            "GCTCAACCTCCAGGCCGGAC\u001b[48;5;11mGTTTTAGAGCTAGAAATAGCAAGTTAAAATAAGGC\u001b[0mTAGTCCGTTATCAACTTGAAAAAGTGGCACCGAGTCGGTGCTTTTTT    \n",
            "    TTGGAGGTCCGGCCTGCAAAATCTCGATCTTTATCGTTCAATTTTATTCCGATCAGGCAATAGTTGAACTTTTTCACCGTGGCTCAGCCACGAAAAAACAAC\n",
            "Dseqrecord(-106)\n",
            "GTTGTCCGTCAGCACTTCGA\u001b[48;5;11mGTTTTAGAGCTAGAAATAGCAAGTTAAAATAAGGC\u001b[0mTAGTCCGTTATCAACTTGAAAAAGTGGCACCGAGTCGGTGCTTTTTT    \n",
            "    AGGCAGTCGTGAAGCTCAAAATCTCGATCTTTATCGTTCAATTTTATTCCGATCAGGCAATAGTTGAACTTTTTCACCGTGGCTCAGCCACGAAAAAAAACA\n",
            "Dseqrecord(-106)\n",
            "TTGTCCGTCAGCACTTCGAA\u001b[48;5;11mGTTTTAGAGCTAGAAATAGCAAGTTAAAATAAGGC\u001b[0mTAGTCCGTTATCAACTTGAAAAAGTGGCACCGAGTCGGTGCTTTTTT    \n",
            "    GGCAGTCGTGAAGCTTCAAAATCTCGATCTTTATCGTTCAATTTTATTCCGATCAGGCAATAGTTGAACTTTTTCACCGTGGCTCAGCCACGAAAAAAACGG\n",
            "Dseqrecord(-106)\n",
            "TGCCCGATCTTCCAGGCCAG\u001b[48;5;11mGTTTTAGAGCTAGAAATAGCAAGTTAAAATAAGGC\u001b[0mTAGTCCGTTATCAACTTGAAAAAGTGGCACCGAGTCGGTGCTTTTTT    \n",
            "    GCTAGAAGGTCCGGTCCAAAATCTCGATCTTTATCGTTCAATTTTATTCCGATCAGGCAATAGTTGAACTTTTTCACCGTGGCTCAGCCACGAAAAAATAAG\n",
            "Dseqrecord(-101)\n",
            "ACGCAGAGACCGCAC\u001b[48;5;11mGTTTTAGAGCTAGAAATAGCAAGTTAAAATAAGGC\u001b[0mTAGTCCGTTATCAACTTGAAAAAGTGGCACCGAGTCGGTGCTTTTTT    \n",
            "    TCTCTGGCGTGCAAAATCTCGATCTTTATCGTTCAATTTTATTCCGATCAGGCAATAGTTGAACTTTTTCACCGTGGCTCAGCCACGAAAAAATAGG\n",
            "Dseqrecord(-106)\n",
            "ATCCCAGGAAGGCAGCGCGC\u001b[48;5;11mGTTTTAGAGCTAGAAATAGCAAGTTAAAATAAGGC\u001b[0mTAGTCCGTTATCAACTTGAAAAAGTGGCACCGAGTCGGTGCTTTTTT    \n",
            "    GTCCTTCCGTCGCGCGCAAAATCTCGATCTTTATCGTTCAATTTTATTCCGATCAGGCAATAGTTGAACTTTTTCACCGTGGCTCAGCCACGAAAAAAGTAG\n",
            "Dseqrecord(-106)\n",
            "CATCCCAGGAAGGCAGCGCG\u001b[48;5;11mGTTTTAGAGCTAGAAATAGCAAGTTAAAATAAGGC\u001b[0mTAGTCCGTTATCAACTTGAAAAAGTGGCACCGAGTCGGTGCTTTTTT    \n",
            "    GGTCCTTCCGTCGCGCCAAAATCTCGATCTTTATCGTTCAATTTTATTCCGATCAGGCAATAGTTGAACTTTTTCACCGTGGCTCAGCCACGAAAAAACGGT\n",
            "Dseqrecord(-106)\n",
            "GCCACGAGCGGCATCCACTG\u001b[48;5;11mGTTTTAGAGCTAGAAATAGCAAGTTAAAATAAGGC\u001b[0mTAGTCCGTTATCAACTTGAAAAAGTGGCACCGAGTCGGTGCTTTTTT    \n",
            "    GCTCGCCGTAGGTGACCAAAATCTCGATCTTTATCGTTCAATTTTATTCCGATCAGGCAATAGTTGAACTTTTTCACCGTGGCTCAGCCACGAAAAAAGCTG\n",
            "Dseqrecord(-106)\n",
            "CGACCTCATCAGCGTCGACA\u001b[48;5;11mGTTTTAGAGCTAGAAATAGCAAGTTAAAATAAGGC\u001b[0mTAGTCCGTTATCAACTTGAAAAAGTGGCACCGAGTCGGTGCTTTTTT    \n",
            "    GAGTAGTCGCAGCTGTCAAAATCTCGATCTTTATCGTTCAATTTTATTCCGATCAGGCAATAGTTGAACTTTTTCACCGTGGCTCAGCCACGAAAAAAGGAG\n",
            "Dseqrecord(-106)\n",
            "CCTCATCAGCGTCGACACGG\u001b[48;5;11mGTTTTAGAGCTAGAAATAGCAAGTTAAAATAAGGC\u001b[0mTAGTCCGTTATCAACTTGAAAAAGTGGCACCGAGTCGGTGCTTTTTT    \n",
            "    TAGTCGCAGCTGTGCCCAAAATCTCGATCTTTATCGTTCAATTTTATTCCGATCAGGCAATAGTTGAACTTTTTCACCGTGGCTCAGCCACGAAAAAAAGTG\n",
            "Dseqrecord(-106)\n",
            "TCACGTCACCGACGGCGTAC\u001b[48;5;11mGTTTTAGAGCTAGAAATAGCAAGTTAAAATAAGGC\u001b[0mTAGTCCGTTATCAACTTGAAAAAGTGGCACCGAGTCGGTGCTTTTTT    \n",
            "    CAGTGGCTGCCGCATGCAAAATCTCGATCTTTATCGTTCAATTTTATTCCGATCAGGCAATAGTTGAACTTTTTCACCGTGGCTCAGCCACGAAAAAAGCTG\n",
            "Dseqrecord(-106)\n",
            "CGACCAGGTCCATCTGCCAC\u001b[48;5;11mGTTTTAGAGCTAGAAATAGCAAGTTAAAATAAGGC\u001b[0mTAGTCCGTTATCAACTTGAAAAAGTGGCACCGAGTCGGTGCTTTTTT    \n",
            "    GTCCAGGTAGACGGTGCAAAATCTCGATCTTTATCGTTCAATTTTATTCCGATCAGGCAATAGTTGAACTTTTTCACCGTGGCTCAGCCACGAAAAAAGCAT\n",
            "Dseqrecord(-106)\n",
            "CGTACAGGTCGCGCATCTCC\u001b[48;5;11mGTTTTAGAGCTAGAAATAGCAAGTTAAAATAAGGC\u001b[0mTAGTCCGTTATCAACTTGAAAAAGTGGCACCGAGTCGGTGCTTTTTT    \n",
            "    GTCCAGCGCGTAGAGGCAAAATCTCGATCTTTATCGTTCAATTTTATTCCGATCAGGCAATAGTTGAACTTTTTCACCGTGGCTCAGCCACGAAAAAAAGAG\n",
            "Dseqrecord(-106)\n",
            "TCTCCAGGTGGACCGCCATG\u001b[48;5;11mGTTTTAGAGCTAGAAATAGCAAGTTAAAATAAGGC\u001b[0mTAGTCCGTTATCAACTTGAAAAAGTGGCACCGAGTCGGTGCTTTTTT    \n",
            "    GTCCACCTGGCGGTACCAAAATCTCGATCTTTATCGTTCAATTTTATTCCGATCAGGCAATAGTTGAACTTTTTCACCGTGGCTCAGCCACGAAAAAACTTC\n",
            "Dseqrecord(-106)\n",
            "GAAGGACAGGTGGGCCTCGC\u001b[48;5;11mGTTTTAGAGCTAGAAATAGCAAGTTAAAATAAGGC\u001b[0mTAGTCCGTTATCAACTTGAAAAAGTGGCACCGAGTCGGTGCTTTTTT    \n",
            "    CTGTCCACCCGGAGCGCAAAATCTCGATCTTTATCGTTCAATTTTATTCCGATCAGGCAATAGTTGAACTTTTTCACCGTGGCTCAGCCACGAAAAAATTCC\n",
            "Dseqrecord(-106)\n",
            "AAGGACAGGTGGGCCTCGCC\u001b[48;5;11mGTTTTAGAGCTAGAAATAGCAAGTTAAAATAAGGC\u001b[0mTAGTCCGTTATCAACTTGAAAAAGTGGCACCGAGTCGGTGCTTTTTT    \n",
            "    TGTCCACCCGGAGCGGCAAAATCTCGATCTTTATCGTTCAATTTTATTCCGATCAGGCAATAGTTGAACTTTTTCACCGTGGCTCAGCCACGAAAAAAACTA\n",
            "Dseqrecord(-106)\n",
            "TGATGCGCAGCCACAGGCGG\u001b[48;5;11mGTTTTAGAGCTAGAAATAGCAAGTTAAAATAAGGC\u001b[0mTAGTCCGTTATCAACTTGAAAAAGTGGCACCGAGTCGGTGCTTTTTT    \n",
            "    CGCGTCGGTGTCCGCCCAAAATCTCGATCTTTATCGTTCAATTTTATTCCGATCAGGCAATAGTTGAACTTTTTCACCGTGGCTCAGCCACGAAAAAACGCC\n",
            "Dseqrecord(-106)\n",
            "GCGGCGCAGCTCGCAGATCT\u001b[48;5;11mGTTTTAGAGCTAGAAATAGCAAGTTAAAATAAGGC\u001b[0mTAGTCCGTTATCAACTTGAAAAAGTGGCACCGAGTCGGTGCTTTTTT    \n",
            "    GCGTCGAGCGTCTAGACAAAATCTCGATCTTTATCGTTCAATTTTATTCCGATCAGGCAATAGTTGAACTTTTTCACCGTGGCTCAGCCACGAAAAAAATGG\n",
            "Dseqrecord(-106)\n",
            "TACCCTCCAGCACGAACCTC\u001b[48;5;11mGTTTTAGAGCTAGAAATAGCAAGTTAAAATAAGGC\u001b[0mTAGTCCGTTATCAACTTGAAAAAGTGGCACCGAGTCGGTGCTTTTTT    \n",
            "    GAGGTCGTGCTTGGAGCAAAATCTCGATCTTTATCGTTCAATTTTATTCCGATCAGGCAATAGTTGAACTTTTTCACCGTGGCTCAGCCACGAAAAAAGATC\n"
          ]
        }
      ],
      "source": [
        "digest_amplicons = digest_amplicons_w_BsaI(list_of_amplicons)\n",
        "for digest in digest_amplicons: \n",
        "    print(digest.figure())"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 18,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/plain": [
              "[Dseqrecord(-138),\n",
              " Dseqrecord(-106),\n",
              " Dseqrecord(-106),\n",
              " Dseqrecord(-106),\n",
              " Dseqrecord(-106),\n",
              " Dseqrecord(-101),\n",
              " Dseqrecord(-106),\n",
              " Dseqrecord(-106),\n",
              " Dseqrecord(-106),\n",
              " Dseqrecord(-106),\n",
              " Dseqrecord(-106),\n",
              " Dseqrecord(-106),\n",
              " Dseqrecord(-106),\n",
              " Dseqrecord(-106),\n",
              " Dseqrecord(-106),\n",
              " Dseqrecord(-106),\n",
              " Dseqrecord(-106),\n",
              " Dseqrecord(-106),\n",
              " Dseqrecord(-106),\n",
              " Dseqrecord(-106)]"
            ]
          },
          "execution_count": 18,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "digest_amplicons"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 19,
      "metadata": {},
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "An error occurred during plasmid digestion or preparation: sticky ends not compatible!\n"
          ]
        },
        {
          "ename": "AttributeError",
          "evalue": "'NoneType' object has no attribute 'name'",
          "output_type": "error",
          "traceback": [
            "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
            "\u001b[0;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
            "Cell \u001b[0;32mIn[19], line 27\u001b[0m\n\u001b[1;32m     24\u001b[0m plasmid_components \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m_\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;241m.\u001b[39mjoin(genes_to_KO)  \u001b[38;5;66;03m# Concatenates gene names from the user's list\u001b[39;00m\n\u001b[1;32m     25\u001b[0m plasmid_purpose \u001b[38;5;241m=\u001b[39m \u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mpCRISPR-mcBEST_w_\u001b[39m\u001b[38;5;132;01m{\u001b[39;00m\u001b[38;5;28mlen\u001b[39m(filtered_df)\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m_sgrnas\u001b[39m\u001b[38;5;124m\"\u001b[39m  \u001b[38;5;66;03m# Example: purpose of the plasmid\u001b[39;00m\n\u001b[0;32m---> 27\u001b[0m \u001b[43mrec_vec\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mname\u001b[49m \u001b[38;5;241m=\u001b[39m \u001b[38;5;124mf\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;132;01m{\u001b[39;00mplasmid_purpose\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m'\u001b[39m\u001b[38;5;66;03m#_{plasmid_components}'\u001b[39;00m\n\u001b[1;32m     28\u001b[0m rec_vec\u001b[38;5;241m.\u001b[39mid \u001b[38;5;241m=\u001b[39m rec_vec\u001b[38;5;241m.\u001b[39mname  \u001b[38;5;66;03m# Using the same value for ID as for name for simplicity\u001b[39;00m\n\u001b[1;32m     29\u001b[0m rec_vec\u001b[38;5;241m.\u001b[39mdescription \u001b[38;5;241m=\u001b[39m \u001b[38;5;124mf\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mCRISPR-mcBEST_plasmid \u001b[39m\u001b[38;5;132;01m{\u001b[39;00m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m, \u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;241m.\u001b[39mjoin(genes_to_KO)\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m for gene KNOCK-DOWN, assembled using StreptoCAD.\u001b[39m\u001b[38;5;124m'\u001b[39m\n",
            "\u001b[0;31mAttributeError\u001b[0m: 'NoneType' object has no attribute 'name'"
          ]
        }
      ],
      "source": [
        "from Bio.Restriction import NcoI, NheI\n",
        "\n",
        "try:\n",
        "    linear_plasmid, smaller_frag = sorted(clean_plasmid.cut(NcoI, NheI), key=lambda x: len(x), reverse=True)\n",
        "    for i in range(len(digest_amplicons)):\n",
        "        # Assuming digest_amplicons[i] is a Dseqrecord, concatenate Dseq sequences\n",
        "        linear_plasmid.seq += digest_amplicons[i].seq\n",
        "    \n",
        "    try:\n",
        "        rec_vec = linear_plasmid.looped()\n",
        "        print(\"Successfully looped the plasmid.\")\n",
        "    except Exception as e:\n",
        "        print(f\"An error occurred while trying to loop the plasmid: {e}\")\n",
        "        # Handle the error or provide an alternative action here\n",
        "        rec_vec = None  # Example response; adjust as needed\n",
        "\n",
        "except Exception as e:\n",
        "    print(f\"An error occurred during plasmid digestion or preparation: {e}\")\n",
        "    # Additional error handling or cleanup here\n",
        "    rec_vec = None  # Ensure rec_vec is defined even if an error occurs\n",
        "\n",
        "\n",
        "# Constructing a meaningful name, ID, and description for the assembled plasmid using user input\n",
        "plasmid_components = \"_\".join(genes_to_KO)  # Concatenates gene names from the user's list\n",
        "plasmid_purpose = f\"pCRISPR-mcBEST_w_{len(filtered_df)}_sgrnas\"  # Example: purpose of the plasmid\n",
        "\n",
        "rec_vec.name = f'{plasmid_purpose}'#_{plasmid_components}'\n",
        "rec_vec.id = rec_vec.name  # Using the same value for ID as for name for simplicity\n",
        "rec_vec.description = f'CRISPR-mcBEST_plasmid {\", \".join(genes_to_KO)} for gene KNOCK-DOWN, assembled using StreptoCAD.'\n",
        "rec_vec.id"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/plain": [
              "Dseqrecord(o13210)"
            ]
          },
          "execution_count": 25,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "# annotate the plasmid correctly\n",
        "annotate_plasmid_with_sgrnas(rec_vec, filtered_df)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {},
      "outputs": [],
      "source": [
        "print_plasmid = False\n",
        "\n",
        "if print_plasmid: \n",
        "    rec_vec.write(f\"../../data/plasmids/sgRNA_plasmids_pCRISPR_mcBEST/{rec_vec.id}.gb\")\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/plain": [
              "Dseqrecord(o13210)"
            ]
          },
          "execution_count": 27,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "rec_vec"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/plain": [
              "'sgRNA_SCO5090(140)_sgRNA_SCO5089(167)_sgRNA_SCO5090(174)_sgRNA_SCO5090(208)_sgRNA_SCO5087(210)_sgRNA_SCO5087(522)_sgRNA_SCO5090(640)_'"
            ]
          },
          "execution_count": 43,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "integration_names = filtered_df.apply(lambda row: f\"sgRNA_{row['locus_tag']}({row['sgrna_loc']})_\", axis=1).tolist()\n",
        "integration_names = ''.join(integration_names)\n",
        "integration_names"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>plasmid_name</th>\n",
              "      <th>date</th>\n",
              "      <th>original_plasmid</th>\n",
              "      <th>integration</th>\n",
              "      <th>size</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>pCRISPR-mcBEST_w_7_sgrnas</td>\n",
              "      <td>2024-07-18</td>\n",
              "      <td>CRISPR-mcBEST</td>\n",
              "      <td>sgRNA_SCO5090(140);sgRNA_SCO5089(167);sgRNA_SC...</td>\n",
              "      <td>13210</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "                plasmid_name        date original_plasmid  \\\n",
              "0  pCRISPR-mcBEST_w_7_sgrnas  2024-07-18    CRISPR-mcBEST   \n",
              "\n",
              "                                         integration   size  \n",
              "0  sgRNA_SCO5090(140);sgRNA_SCO5089(167);sgRNA_SC...  13210  "
            ]
          },
          "execution_count": 45,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "integration_names = filtered_df.apply(lambda row: f\"sgRNA_{row['locus_tag']}({row['sgrna_loc']})\", axis=1).tolist()\n",
        "integration_names = [';'.join(integration_names)]\n",
        "\n",
        "plasmid_metadata_df = extract_metadata_to_dataframe([rec_vec],\n",
        "                                                    'CRISPR-mcBEST',\n",
        "                                                    integration_names)\n",
        "\n",
        "plasmid_metadata_df"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>Locus Tag</th>\n",
              "      <th>f_primer_name</th>\n",
              "      <th>r_primer_name</th>\n",
              "      <th>f_primer_sequences(5-3)</th>\n",
              "      <th>r_primer_sequences(5-3)</th>\n",
              "      <th>f_tm</th>\n",
              "      <th>r_tm</th>\n",
              "      <th>ta</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>SCO5087</td>\n",
              "      <td>SCO5087_fwd_checking_primer</td>\n",
              "      <td>SCO5087_rev_checking_primer</td>\n",
              "      <td>GACGATTCGGCCCGTG</td>\n",
              "      <td>CAGGGCGTCCAGGC</td>\n",
              "      <td>59</td>\n",
              "      <td>58</td>\n",
              "      <td>61</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>SCO5089</td>\n",
              "      <td>SCO5089_fwd_checking_primer</td>\n",
              "      <td>SCO5089_rev_checking_primer</td>\n",
              "      <td>CGTACGGCGAACTCGC</td>\n",
              "      <td>GTCCCCGGACGGC</td>\n",
              "      <td>59</td>\n",
              "      <td>57</td>\n",
              "      <td>60</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>SCO5090</td>\n",
              "      <td>SCO5090_fwd_checking_primer</td>\n",
              "      <td>SCO5090_rev_checking_primer</td>\n",
              "      <td>GGCGGCGGCCC</td>\n",
              "      <td>GACGTCCCCCGCGAAC</td>\n",
              "      <td>57</td>\n",
              "      <td>62</td>\n",
              "      <td>61</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "  Locus Tag                f_primer_name                r_primer_name  \\\n",
              "0   SCO5087  SCO5087_fwd_checking_primer  SCO5087_rev_checking_primer   \n",
              "1   SCO5089  SCO5089_fwd_checking_primer  SCO5089_rev_checking_primer   \n",
              "2   SCO5090  SCO5090_fwd_checking_primer  SCO5090_rev_checking_primer   \n",
              "\n",
              "  f_primer_sequences(5-3) r_primer_sequences(5-3)  f_tm  r_tm  ta  \n",
              "0        GACGATTCGGCCCGTG          CAGGGCGTCCAGGC    59    58  61  \n",
              "1        CGTACGGCGAACTCGC           GTCCCCGGACGGC    59    57  60  \n",
              "2             GGCGGCGGCCC        GACGTCCCCCGCGAAC    57    62  61  "
            ]
          },
          "execution_count": 21,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "# Getting checking primers\n",
        "checking_primers_df = find_best_check_primers_from_genome(genome, \n",
        "                                       genes_to_KO, \n",
        "                                       flanking_region=flanking_region,\n",
        "                                       target_tm = melting_temperature, \n",
        "                                        primer_concentration = primer_concentration, \n",
        "                                        polymerase = chosen_polymerase)\n",
        "checking_primers_df\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>Name</th>\n",
              "      <th>Sequence</th>\n",
              "      <th>Concentration</th>\n",
              "      <th>Purification</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>SCO5087_fwd_checking_primer</td>\n",
              "      <td>GACGATTCGGCCCGTG</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>SCO5089_fwd_checking_primer</td>\n",
              "      <td>CGTACGGCGAACTCGC</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>SCO5090_fwd_checking_primer</td>\n",
              "      <td>GGCGGCGGCCC</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>SCO5087_rev_checking_primer</td>\n",
              "      <td>CAGGGCGTCCAGGC</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>SCO5089_rev_checking_primer</td>\n",
              "      <td>GTCCCCGGACGGC</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>5</th>\n",
              "      <td>SCO5090_rev_checking_primer</td>\n",
              "      <td>GACGTCCCCCGCGAAC</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "                          Name          Sequence Concentration Purification\n",
              "0  SCO5087_fwd_checking_primer  GACGATTCGGCCCGTG          25nm          STD\n",
              "1  SCO5089_fwd_checking_primer  CGTACGGCGAACTCGC          25nm          STD\n",
              "2  SCO5090_fwd_checking_primer       GGCGGCGGCCC          25nm          STD\n",
              "3  SCO5087_rev_checking_primer    CAGGGCGTCCAGGC          25nm          STD\n",
              "4  SCO5089_rev_checking_primer     GTCCCCGGACGGC          25nm          STD\n",
              "5  SCO5090_rev_checking_primer  GACGTCCCCCGCGAAC          25nm          STD"
            ]
          },
          "execution_count": 22,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "# making the primers into dseqrecords\n",
        "checking_primers_df_idt = create_idt_order_dataframe(checking_primers_df)\n",
        "checking_primers_df_idt\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>Name</th>\n",
              "      <th>Sequence</th>\n",
              "      <th>Concentration</th>\n",
              "      <th>Purification</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>primer_23</td>\n",
              "      <td>GATCGGGTCTCCCATGGTTCACTGCCGTATAGGCAGCTAAGAAACA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>primer_25</td>\n",
              "      <td>GATCGGGTCTCCGCGCGACTCGAGAGCCGGTAGTTTTAGAGCTAGA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>primer_27</td>\n",
              "      <td>GATCGGGTCTCCCGCCCAGATCCGGAAGCGTTGTTTTAGAGCTAGA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>primer_29</td>\n",
              "      <td>GATCGGGTCTCCGGGACGTCCAGGTCTTCACCGTTTTAGAGCTAGA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>primer_31</td>\n",
              "      <td>GATCGGGTCTCCGAGCAGTTCCCAGAACTGCCGTTTTAGAGCTAGA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>5</th>\n",
              "      <td>primer_33</td>\n",
              "      <td>GATCGGGTCTCCGTCGACCTCCCAGTCACGGCGTTTTAGAGCTAGA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>6</th>\n",
              "      <td>primer_35</td>\n",
              "      <td>GATCGGGTCTCCGACCGTCCAGGACATGACCAGTTTTAGAGCTAGA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>7</th>\n",
              "      <td>primer_24</td>\n",
              "      <td>GATCAGGTCTCGGCGCAAAAAAGCACCGACTCGGT</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>8</th>\n",
              "      <td>primer_26</td>\n",
              "      <td>GATCAGGTCTCGGGCGAAAAAAGCACCGACTCGGT</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>9</th>\n",
              "      <td>primer_28</td>\n",
              "      <td>GATCAGGTCTCGTCCCAAAAAAGCACCGACTCGGT</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>10</th>\n",
              "      <td>primer_30</td>\n",
              "      <td>GATCAGGTCTCGGCTCAAAAAAGCACCGACTCGGT</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>11</th>\n",
              "      <td>primer_32</td>\n",
              "      <td>GATCAGGTCTCGCGACAAAAAAGCACCGACTCGGT</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>12</th>\n",
              "      <td>primer_34</td>\n",
              "      <td>GATCAGGTCTCGGGTCAAAAAAGCACCGACTCGGT</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>13</th>\n",
              "      <td>primer_36</td>\n",
              "      <td>GATCAGGTCTCGCTAGAAAAAAGCACCGACTCGGT</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>SCO5087_fwd_checking_primer</td>\n",
              "      <td>GACGATTCGGCCCGTG</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>SCO5089_fwd_checking_primer</td>\n",
              "      <td>CGTACGGCGAACTCGC</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>SCO5090_fwd_checking_primer</td>\n",
              "      <td>GGCGGCGGCCC</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>SCO5087_rev_checking_primer</td>\n",
              "      <td>CAGGGCGTCCAGGC</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>SCO5089_rev_checking_primer</td>\n",
              "      <td>GTCCCCGGACGGC</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>5</th>\n",
              "      <td>SCO5090_rev_checking_primer</td>\n",
              "      <td>GACGTCCCCCGCGAAC</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "                           Name  \\\n",
              "0                     primer_23   \n",
              "1                     primer_25   \n",
              "2                     primer_27   \n",
              "3                     primer_29   \n",
              "4                     primer_31   \n",
              "5                     primer_33   \n",
              "6                     primer_35   \n",
              "7                     primer_24   \n",
              "8                     primer_26   \n",
              "9                     primer_28   \n",
              "10                    primer_30   \n",
              "11                    primer_32   \n",
              "12                    primer_34   \n",
              "13                    primer_36   \n",
              "0   SCO5087_fwd_checking_primer   \n",
              "1   SCO5089_fwd_checking_primer   \n",
              "2   SCO5090_fwd_checking_primer   \n",
              "3   SCO5087_rev_checking_primer   \n",
              "4   SCO5089_rev_checking_primer   \n",
              "5   SCO5090_rev_checking_primer   \n",
              "\n",
              "                                             Sequence Concentration  \\\n",
              "0   GATCGGGTCTCCCATGGTTCACTGCCGTATAGGCAGCTAAGAAACA...          25nm   \n",
              "1   GATCGGGTCTCCGCGCGACTCGAGAGCCGGTAGTTTTAGAGCTAGA...          25nm   \n",
              "2   GATCGGGTCTCCCGCCCAGATCCGGAAGCGTTGTTTTAGAGCTAGA...          25nm   \n",
              "3   GATCGGGTCTCCGGGACGTCCAGGTCTTCACCGTTTTAGAGCTAGA...          25nm   \n",
              "4   GATCGGGTCTCCGAGCAGTTCCCAGAACTGCCGTTTTAGAGCTAGA...          25nm   \n",
              "5   GATCGGGTCTCCGTCGACCTCCCAGTCACGGCGTTTTAGAGCTAGA...          25nm   \n",
              "6   GATCGGGTCTCCGACCGTCCAGGACATGACCAGTTTTAGAGCTAGA...          25nm   \n",
              "7                 GATCAGGTCTCGGCGCAAAAAAGCACCGACTCGGT          25nm   \n",
              "8                 GATCAGGTCTCGGGCGAAAAAAGCACCGACTCGGT          25nm   \n",
              "9                 GATCAGGTCTCGTCCCAAAAAAGCACCGACTCGGT          25nm   \n",
              "10                GATCAGGTCTCGGCTCAAAAAAGCACCGACTCGGT          25nm   \n",
              "11                GATCAGGTCTCGCGACAAAAAAGCACCGACTCGGT          25nm   \n",
              "12                GATCAGGTCTCGGGTCAAAAAAGCACCGACTCGGT          25nm   \n",
              "13                GATCAGGTCTCGCTAGAAAAAAGCACCGACTCGGT          25nm   \n",
              "0                                    GACGATTCGGCCCGTG          25nm   \n",
              "1                                    CGTACGGCGAACTCGC          25nm   \n",
              "2                                         GGCGGCGGCCC          25nm   \n",
              "3                                      CAGGGCGTCCAGGC          25nm   \n",
              "4                                       GTCCCCGGACGGC          25nm   \n",
              "5                                    GACGTCCCCCGCGAAC          25nm   \n",
              "\n",
              "   Purification  \n",
              "0           STD  \n",
              "1           STD  \n",
              "2           STD  \n",
              "3           STD  \n",
              "4           STD  \n",
              "5           STD  \n",
              "6           STD  \n",
              "7           STD  \n",
              "8           STD  \n",
              "9           STD  \n",
              "10          STD  \n",
              "11          STD  \n",
              "12          STD  \n",
              "13          STD  \n",
              "0           STD  \n",
              "1           STD  \n",
              "2           STD  \n",
              "3           STD  \n",
              "4           STD  \n",
              "5           STD  "
            ]
          },
          "execution_count": 23,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "full_idt = pd.concat([idt_df, checking_primers_df_idt])\n",
        "full_idt"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "## Folder with all the generated I/O"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {},
      "outputs": [],
      "source": [
        "get_data_folder = False"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {},
      "outputs": [],
      "source": [
        "if get_data_folder: \n",
        "    # Example: \n",
        "\n",
        "    input_files = [\n",
        "        {\"name\": \"input_genome.gb\", \"content\": genome},\n",
        "        {\"name\": \"input_plasmid.gb\", \"content\": clean_plasmid}\n",
        "    ]\n",
        "\n",
        "    output_files = [\n",
        "        {\"name\": \"mcBEST_w_sgRNAs.gb\", \"content\": rec_vec},\n",
        "        {\"name\": \"overhang_df.csv\", \"content\": overhangs},\n",
        "        {\"name\": \"pcr_df.csv\", \"content\": primer_df},\n",
        "        {\"name\": \"full_idt.csv\", \"content\": full_idt},\n",
        "        {\"name\": \"mutated_sgrna_df.csv\", \"content\": mutated_sgrna_df},\n",
        "        {\"name\": \"filtered_df.csv\", \"content\": filtered_df},\n",
        "        {\"name\": \"plasmid_metadata_df.csv\", \"content\": plasmid_metadata_df},\n",
        "    ]\n",
        "\n",
        "    input_values = {\n",
        "        \"genes_to_knockout\": ['SCO5087', 'SCO5089', 'SCO5090'],\n",
        "        \"filtering_metrics\": {\n",
        "            \"gc_upper\": gc_upper,\n",
        "            \"gc_lower\": gc_lower,\n",
        "            \"off_target_seed\": off_target_seed,\n",
        "            \"off_target_upper\": off_target_upper,\n",
        "            \"cas_type\": cas_type,\n",
        "            \"number_of_sgRNAs_per_group\": number_of_sgRNAs_per_group\n",
        "        },\n",
        "        \"polymerase_settings\": {\n",
        "            \"chosen_polymerase\": chosen_polymerase,\n",
        "            \"melting_temperature\": melting_temperature,\n",
        "            \"primer_concentration\": primer_concentration,\n",
        "            \"primer_number_increment\": primer_number_increment,\n",
        "            \"flanking_region\": flanking_region\n",
        "        }, # TODO might want to incorporate this to the app\n",
        "        \"overlapping_sequences\": {\n",
        "            \"restriction_overhang_f\": restriction_overhang_f,\n",
        "            \"restriction_overhang_r\": restriction_overhang_r,\n",
        "            \"backbone_overhang_f\": backbone_overhang_f,\n",
        "            \"backbone_overhang_r\": backbone_overhang_r,\n",
        "            \"cys4\": cys4,\n",
        "            \"sgRNA_handle_cys4_site\": str(sgRNA_handle_cys4_site[0].seq)\n",
        "        }\n",
        "    }\n",
        "\n",
        "\n",
        "    # Paths to Markdown files\n",
        "    markdown_file_paths = [\n",
        "        \"../../protocols/conjugation_protcol.md\",\n",
        "        \"../../protocols/multi_target_crispr_plasmid_protcol.md\"\n",
        "\n",
        "    ]\n",
        "\n",
        "    # Data and time\n",
        "    timestamp = datetime.utcnow().isoformat()\n",
        "    project_name=f\"CRISPR_mcBEST_workflow_{timestamp}\"\n",
        "\n",
        "\n",
        "    # Create project directory structure\n",
        "    project_directory = ProjectDirectory(\n",
        "        project_name=project_name,\n",
        "        input_files=input_files,\n",
        "        output_files=output_files,\n",
        "        input_values=input_values,\n",
        "        markdown_file_paths=markdown_file_paths\n",
        "    )\n",
        "\n",
        "\n",
        "    # DO You want to save the folder? \n",
        "    save_zip_folder = False \n",
        "\n",
        "    if save_zip_folder: \n",
        "        # Generate the project directory structure and get the zip content\n",
        "        zip_content = project_directory.create_directory_structure(create_directories=False)\n",
        "\n",
        "        # Save the zip file to disk (optional)\n",
        "        with open(\"project_structure.zip\", \"wb\") as f:\n",
        "            f.write(zip_content)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {},
      "outputs": [],
      "source": []
    }
  ],
  "metadata": {
    "kernelspec": {
      "display_name": "constrain",
      "language": "python",
      "name": "python3"
    },
    "language_info": {
      "codemirror_mode": {
        "name": "ipython",
        "version": 3
      },
      "file_extension": ".py",
      "mimetype": "text/x-python",
      "name": "python",
      "nbconvert_exporter": "python",
      "pygments_lexer": "ipython3",
      "version": "3.11.9"
    }
  },
  "nbformat": 4,
  "nbformat_minor": 2
}
