{
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "## W3-CRISPR-mcBEST workflow\n",
        "\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "The workflow allows for the simultaneous targeted engineering of up to 28 individual situated bases in the genome of Streptomyces, as shown by Whitford et. al 2023, thereby enabling efficient single-step multi-target engineering13. Similar to workflow 2, this workflow employs the CRISPR-BEST methodology but instead of performing base-editing on a single gene the method incorporates multiple guide RNAs as an array into the same vector (Figure 4A). This significantly enhances the efficiency of high-throughput genetic studies. The pCRISPR-cBEST plasmid file can be downloaded on the page along with the genome file for S.coelicor (A3) and streptoAIM will generate your plasmids, primers, checking primers and analyze the overhangs the target genes(Figure 4, S6). \n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 1,
      "metadata": {},
      "outputs": [],
      "source": [
        "import sys\n",
        "import os\n",
        "from pydna.dseqrecord import Dseqrecord\n",
        "from teemi.build.PCR import primer_tm_neb\n",
        "import pandas as pd\n",
        "from datetime import datetime\n",
        "\n",
        "\n",
        "# Ensure the src directory is in the Python path\n",
        "project_root = os.path.abspath(os.path.join(os.getcwd(), '../../'))\n",
        "if project_root not in sys.path:\n",
        "    sys.path.insert(0, project_root)\n",
        "\n",
        "from streptocad.sequence_loading.sequence_loading import (\n",
        "    load_and_process_plasmid, \n",
        "    check_and_convert_input,\n",
        "    annotate_dseqrecord,\n",
        "    process_specified_gene_sequences_from_record)\n",
        "\n",
        "\n",
        "from streptocad.utils import polymerase_dict, dataframe_to_seqrecords, ProjectDirectory, extract_metadata_to_dataframe\n",
        "from streptocad.crispr.guideRNAcas3_9_12 import extract_sgRNAs, SgRNAargs\n",
        "from streptocad.crispr.crispr_best import (\n",
        "    identify_base_editing_sites,\n",
        "    filter_sgrnas_for_base_editing,\n",
        "    process_base_editing\n",
        ")\n",
        "\n",
        "from streptocad.cloning.golden_gate_cloning import (\n",
        "    GoldenGateCloning,\n",
        "    create_overhang_dataframe,\n",
        "    digest_amplicons_w_BsaI\n",
        ")\n",
        "\n",
        "from streptocad.primers.primer_analysis import analyze_primers_and_hairpins\n",
        "from streptocad.primers.primer_generation import create_idt_order_dataframe\n",
        "from streptocad.cloning.plasmid_processing import annotate_plasmid_with_sgrnas\n",
        "from streptocad.primers.primer_generation import find_best_check_primers_from_genome"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "# Inputs\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 2,
      "metadata": {},
      "outputs": [],
      "source": [
        "# 1 Add genome of choice (genbank, fasta)\n",
        "path_to_genome = '../../data/genomes/Streptomyces_coelicolor_A3_chromosome.gb'\n",
        "genome = load_and_process_plasmid(path_to_genome)\n",
        "\n",
        "# 2 Add plasmid \n",
        "path_to_plasmid = '../../data/plasmids/pCRISPR-MCBE_Csy4_kasopGFP.gb'\n",
        "clean_plasmid = load_and_process_plasmid(path_to_plasmid)\n",
        "\n",
        "# 3 Choose genes to knock out (list)\n",
        "genes_to_KO = ['SCO5087', 'SCO5089', 'SCO5090']\n",
        "#genes_to_KO = ['80000-100000', '4000-7000', '9000-14000','15000-20000']\n",
        "\n",
        "\n",
        "#### Advanced settings ####\n",
        "# 4 Filtering metrics for sgRNAs\n",
        "gc_upper = 0.72\n",
        "gc_lower = 0.2\n",
        "off_target_seed = 13\n",
        "off_target_upper = 10\n",
        "cas_type='cas9'\n",
        "number_of_sgRNAs_per_group = 2\n",
        "\n",
        "# 5 Choose polymerase and target melting temperature\n",
        "chosen_polymerase = polymerase_dict['Phusion High-Fidelity DNA Polymerase (GC Buffer)']\n",
        "melting_temperature = 60\n",
        "primer_concentration = 0.4 \n",
        "flanking_region=500\n",
        "\n",
        "\n",
        "# 6 Choose overlapping sequences for our plasmid we can use the following\n",
        "restriction_overhang_f = \"GATCGggtctcc\"\n",
        "restriction_overhang_r = \"GATCAGGTCTCg\"\n",
        "backbone_overhang_f = \"cATG\"\n",
        "backbone_overhang_r = \"cTAG\"\n",
        "cys4 = \"gTTCACTGCCGTATAGGCAGCTAAGAAA\"\n",
        "sgRNA_handle_cys4_site = [Dseqrecord('GTTTTAGAGCTAGAAATAGCAAGTTAAAATAAGGCTAGTCCGTTATCAACTTGAAAAAGTGGCACCGAGTCGGTGCTTTTTT', name = 'sgRNA_handle_cys4')]"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 3,
      "metadata": {},
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "pCRISPR-mcBEST\n"
          ]
        }
      ],
      "source": [
        "print(clean_plasmid.id)"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "# Computation"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 4,
      "metadata": {},
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "False\n"
          ]
        },
        {
          "data": {
            "text/plain": [
              "25824"
            ]
          },
          "execution_count": 4,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "target_dict, genes_to_KO, annotation_input = check_and_convert_input(genes_to_KO)\n",
        "\n",
        "print(annotation_input)\n",
        "if annotation_input == True:\n",
        "    genome = annotate_dseqrecord(genome, target_dict)\n",
        "\n",
        "\n",
        "len(genome.features)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 5,
      "metadata": {},
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "Pam was found outside designated locus_tag: SCO5087. To incorporate this extent borders. Skipping to next locus tag.\n",
            "sgRNA generated were outside the designated border in SCO5087. To incorporate this extent borders. Skipping to next locus tag.\n",
            "Pam was found outside designated locus_tag: SCO5087. To incorporate this extent borders. Skipping to next locus tag.\n",
            "sgRNA generated were outside the designated border in SCO5087. To incorporate this extent borders. Skipping to next locus tag.\n",
            "sgRNA generated were outside the designated border in SCO5087. To incorporate this extent borders. Skipping to next locus tag.\n",
            "sgRNA generated were outside the designated border in SCO5089. To incorporate this extent borders. Skipping to next locus tag.\n",
            "sgRNA generated were outside the designated border in SCO5089. To incorporate this extent borders. Skipping to next locus tag.\n",
            "Pam was found outside designated locus_tag: SCO5090. To incorporate this extent borders. Skipping to next locus tag.\n",
            "sgRNA generated were outside the designated border in SCO5090. To incorporate this extent borders. Skipping to next locus tag.\n",
            "sgRNA generated were outside the designated border in SCO5090. To incorporate this extent borders. Skipping to next locus tag.\n",
            "sgRNA generated were outside the designated border in SCO5090. To incorporate this extent borders. Skipping to next locus tag.\n",
            "sgRNA generated were outside the designated border in SCO5090. To incorporate this extent borders. Skipping to next locus tag.\n",
            "sgRNA generated were outside the designated border in SCO5090. To incorporate this extent borders. Skipping to next locus tag.\n"
          ]
        },
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>strain_name</th>\n",
              "      <th>locus_tag</th>\n",
              "      <th>gene_loc</th>\n",
              "      <th>gene_strand</th>\n",
              "      <th>sgrna_strand</th>\n",
              "      <th>sgrna_loc</th>\n",
              "      <th>gc</th>\n",
              "      <th>pam</th>\n",
              "      <th>sgrna</th>\n",
              "      <th>sgrna_seed_sequence</th>\n",
              "      <th>off_target_count</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>334</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5090</td>\n",
              "      <td>5532706</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>32</td>\n",
              "      <td>0.65</td>\n",
              "      <td>CGG</td>\n",
              "      <td>TGGACCACCCGGTGTTCTCC</td>\n",
              "      <td>CCCGGTGTTCTCC</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>330</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5089</td>\n",
              "      <td>5532449</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>61</td>\n",
              "      <td>0.70</td>\n",
              "      <td>CGG</td>\n",
              "      <td>CGTGGAGTGCGCCGGTGAGA</td>\n",
              "      <td>TGCGCCGGTGAGA</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>329</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5089</td>\n",
              "      <td>5532449</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>65</td>\n",
              "      <td>0.70</td>\n",
              "      <td>CGG</td>\n",
              "      <td>GAGTGCGCCGGTGAGACGGA</td>\n",
              "      <td>CCGGTGAGACGGA</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>328</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5089</td>\n",
              "      <td>5532449</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>66</td>\n",
              "      <td>0.70</td>\n",
              "      <td>GGG</td>\n",
              "      <td>AGTGCGCCGGTGAGACGGAC</td>\n",
              "      <td>CGGTGAGACGGAC</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>326</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5089</td>\n",
              "      <td>5532449</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>80</td>\n",
              "      <td>0.70</td>\n",
              "      <td>CGG</td>\n",
              "      <td>ACGGACGGGACGGACCTGTC</td>\n",
              "      <td>GGACGGACCTGTC</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>...</th>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>108</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5087</td>\n",
              "      <td>5529801</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>1038</td>\n",
              "      <td>0.70</td>\n",
              "      <td>GGG</td>\n",
              "      <td>GATGTCCGTCGCGTCCGTGC</td>\n",
              "      <td>GTCGCGTCCGTGC</td>\n",
              "      <td>2</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>295</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5089</td>\n",
              "      <td>5532449</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>60</td>\n",
              "      <td>0.70</td>\n",
              "      <td>GGG</td>\n",
              "      <td>CTCACCGGCGCACTCCACGA</td>\n",
              "      <td>GCGCACTCCACGA</td>\n",
              "      <td>2</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>449</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5090</td>\n",
              "      <td>5532706</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>890</td>\n",
              "      <td>0.65</td>\n",
              "      <td>CGG</td>\n",
              "      <td>ACCCATCTGCGCGAGGTGCT</td>\n",
              "      <td>TGCGCGAGGTGCT</td>\n",
              "      <td>3</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>191</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5087</td>\n",
              "      <td>5529801</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>838</td>\n",
              "      <td>0.70</td>\n",
              "      <td>TGG</td>\n",
              "      <td>TACGCGGGACGGCTTCGTGC</td>\n",
              "      <td>GACGGCTTCGTGC</td>\n",
              "      <td>3</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>200</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5087</td>\n",
              "      <td>5529801</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>706</td>\n",
              "      <td>0.65</td>\n",
              "      <td>CGG</td>\n",
              "      <td>CAGCGCCGACGTGATGTTCG</td>\n",
              "      <td>GACGTGATGTTCG</td>\n",
              "      <td>3</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>279 rows × 11 columns</p>\n",
              "</div>"
            ],
            "text/plain": [
              "     strain_name locus_tag  gene_loc  gene_strand  sgrna_strand  sgrna_loc  \\\n",
              "334  NC_003888.3   SCO5090   5532706            1            -1         32   \n",
              "330  NC_003888.3   SCO5089   5532449            1             1         61   \n",
              "329  NC_003888.3   SCO5089   5532449            1             1         65   \n",
              "328  NC_003888.3   SCO5089   5532449            1             1         66   \n",
              "326  NC_003888.3   SCO5089   5532449            1             1         80   \n",
              "..           ...       ...       ...          ...           ...        ...   \n",
              "108  NC_003888.3   SCO5087   5529801            1            -1       1038   \n",
              "295  NC_003888.3   SCO5089   5532449            1            -1         60   \n",
              "449  NC_003888.3   SCO5090   5532706            1             1        890   \n",
              "191  NC_003888.3   SCO5087   5529801            1             1        838   \n",
              "200  NC_003888.3   SCO5087   5529801            1             1        706   \n",
              "\n",
              "       gc  pam                 sgrna sgrna_seed_sequence  off_target_count  \n",
              "334  0.65  CGG  TGGACCACCCGGTGTTCTCC       CCCGGTGTTCTCC                 0  \n",
              "330  0.70  CGG  CGTGGAGTGCGCCGGTGAGA       TGCGCCGGTGAGA                 0  \n",
              "329  0.70  CGG  GAGTGCGCCGGTGAGACGGA       CCGGTGAGACGGA                 0  \n",
              "328  0.70  GGG  AGTGCGCCGGTGAGACGGAC       CGGTGAGACGGAC                 0  \n",
              "326  0.70  CGG  ACGGACGGGACGGACCTGTC       GGACGGACCTGTC                 0  \n",
              "..    ...  ...                   ...                 ...               ...  \n",
              "108  0.70  GGG  GATGTCCGTCGCGTCCGTGC       GTCGCGTCCGTGC                 2  \n",
              "295  0.70  GGG  CTCACCGGCGCACTCCACGA       GCGCACTCCACGA                 2  \n",
              "449  0.65  CGG  ACCCATCTGCGCGAGGTGCT       TGCGCGAGGTGCT                 3  \n",
              "191  0.70  TGG  TACGCGGGACGGCTTCGTGC       GACGGCTTCGTGC                 3  \n",
              "200  0.65  CGG  CAGCGCCGACGTGATGTTCG       GACGTGATGTTCG                 3  \n",
              "\n",
              "[279 rows x 11 columns]"
            ]
          },
          "execution_count": 5,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "# Initialize SgRNAargs with desired parameters\n",
        "args = SgRNAargs(genome, \n",
        "                genes_to_KO,\n",
        "                step=['find', 'filter'],\n",
        "                gc_upper = gc_upper,\n",
        "                gc_lower = gc_lower,\n",
        "                off_target_seed = off_target_seed,\n",
        "                off_target_upper = off_target_upper,\n",
        "                cas_type='cas9'\n",
        "                )\n",
        "\n",
        "sgrna_df = extract_sgRNAs(args)\n",
        "sgrna_df"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 6,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>strain_name</th>\n",
              "      <th>locus_tag</th>\n",
              "      <th>gene_loc</th>\n",
              "      <th>gene_strand</th>\n",
              "      <th>sgrna_strand</th>\n",
              "      <th>sgrna_loc</th>\n",
              "      <th>gc</th>\n",
              "      <th>pam</th>\n",
              "      <th>sgrna</th>\n",
              "      <th>sgrna_seed_sequence</th>\n",
              "      <th>off_target_count</th>\n",
              "      <th>editing_context</th>\n",
              "      <th>editable_cytosines</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>334</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5090</td>\n",
              "      <td>5532706</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>32</td>\n",
              "      <td>0.65</td>\n",
              "      <td>CGG</td>\n",
              "      <td>TGGACCACCCGGTGTTCTCC</td>\n",
              "      <td>CCCGGTGTTCTCC</td>\n",
              "      <td>0</td>\n",
              "      <td>0,0,0,0,0</td>\n",
              "      <td>5,6,8,9,10</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>330</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5089</td>\n",
              "      <td>5532449</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>61</td>\n",
              "      <td>0.70</td>\n",
              "      <td>CGG</td>\n",
              "      <td>CGTGGAGTGCGCCGGTGAGA</td>\n",
              "      <td>TGCGCCGGTGAGA</td>\n",
              "      <td>0</td>\n",
              "      <td>1</td>\n",
              "      <td>10</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>329</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5089</td>\n",
              "      <td>5532449</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>65</td>\n",
              "      <td>0.70</td>\n",
              "      <td>CGG</td>\n",
              "      <td>GAGTGCGCCGGTGAGACGGA</td>\n",
              "      <td>CCGGTGAGACGGA</td>\n",
              "      <td>0</td>\n",
              "      <td>1,1,0</td>\n",
              "      <td>6,8,9</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>328</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5089</td>\n",
              "      <td>5532449</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>66</td>\n",
              "      <td>0.70</td>\n",
              "      <td>GGG</td>\n",
              "      <td>AGTGCGCCGGTGAGACGGAC</td>\n",
              "      <td>CGGTGAGACGGAC</td>\n",
              "      <td>0</td>\n",
              "      <td>1,1,0</td>\n",
              "      <td>5,7,8</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>326</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5089</td>\n",
              "      <td>5532449</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>80</td>\n",
              "      <td>0.70</td>\n",
              "      <td>CGG</td>\n",
              "      <td>ACGGACGGGACGGACCTGTC</td>\n",
              "      <td>GGACGGACCTGTC</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>6</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>...</th>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>108</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5087</td>\n",
              "      <td>5529801</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>1038</td>\n",
              "      <td>0.70</td>\n",
              "      <td>GGG</td>\n",
              "      <td>GATGTCCGTCGCGTCCGTGC</td>\n",
              "      <td>GTCGCGTCCGTGC</td>\n",
              "      <td>2</td>\n",
              "      <td>0,0,0</td>\n",
              "      <td>6,7,10</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>295</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5089</td>\n",
              "      <td>5532449</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>60</td>\n",
              "      <td>0.70</td>\n",
              "      <td>GGG</td>\n",
              "      <td>CTCACCGGCGCACTCCACGA</td>\n",
              "      <td>GCGCACTCCACGA</td>\n",
              "      <td>2</td>\n",
              "      <td>0,0,0,1</td>\n",
              "      <td>3,5,6,9</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>449</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5090</td>\n",
              "      <td>5532706</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>890</td>\n",
              "      <td>0.65</td>\n",
              "      <td>CGG</td>\n",
              "      <td>ACCCATCTGCGCGAGGTGCT</td>\n",
              "      <td>TGCGCGAGGTGCT</td>\n",
              "      <td>3</td>\n",
              "      <td>0,0,0,1</td>\n",
              "      <td>3,4,7,10</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>191</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5087</td>\n",
              "      <td>5529801</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>838</td>\n",
              "      <td>0.70</td>\n",
              "      <td>TGG</td>\n",
              "      <td>TACGCGGGACGGCTTCGTGC</td>\n",
              "      <td>GACGGCTTCGTGC</td>\n",
              "      <td>3</td>\n",
              "      <td>0,1,0</td>\n",
              "      <td>3,5,10</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>200</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5087</td>\n",
              "      <td>5529801</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>706</td>\n",
              "      <td>0.65</td>\n",
              "      <td>CGG</td>\n",
              "      <td>CAGCGCCGACGTGATGTTCG</td>\n",
              "      <td>GACGTGATGTTCG</td>\n",
              "      <td>3</td>\n",
              "      <td>1,1,0,0</td>\n",
              "      <td>4,6,7,10</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>272 rows × 13 columns</p>\n",
              "</div>"
            ],
            "text/plain": [
              "     strain_name locus_tag  gene_loc  gene_strand  sgrna_strand  sgrna_loc  \\\n",
              "334  NC_003888.3   SCO5090   5532706            1            -1         32   \n",
              "330  NC_003888.3   SCO5089   5532449            1             1         61   \n",
              "329  NC_003888.3   SCO5089   5532449            1             1         65   \n",
              "328  NC_003888.3   SCO5089   5532449            1             1         66   \n",
              "326  NC_003888.3   SCO5089   5532449            1             1         80   \n",
              "..           ...       ...       ...          ...           ...        ...   \n",
              "108  NC_003888.3   SCO5087   5529801            1            -1       1038   \n",
              "295  NC_003888.3   SCO5089   5532449            1            -1         60   \n",
              "449  NC_003888.3   SCO5090   5532706            1             1        890   \n",
              "191  NC_003888.3   SCO5087   5529801            1             1        838   \n",
              "200  NC_003888.3   SCO5087   5529801            1             1        706   \n",
              "\n",
              "       gc  pam                 sgrna sgrna_seed_sequence  off_target_count  \\\n",
              "334  0.65  CGG  TGGACCACCCGGTGTTCTCC       CCCGGTGTTCTCC                 0   \n",
              "330  0.70  CGG  CGTGGAGTGCGCCGGTGAGA       TGCGCCGGTGAGA                 0   \n",
              "329  0.70  CGG  GAGTGCGCCGGTGAGACGGA       CCGGTGAGACGGA                 0   \n",
              "328  0.70  GGG  AGTGCGCCGGTGAGACGGAC       CGGTGAGACGGAC                 0   \n",
              "326  0.70  CGG  ACGGACGGGACGGACCTGTC       GGACGGACCTGTC                 0   \n",
              "..    ...  ...                   ...                 ...               ...   \n",
              "108  0.70  GGG  GATGTCCGTCGCGTCCGTGC       GTCGCGTCCGTGC                 2   \n",
              "295  0.70  GGG  CTCACCGGCGCACTCCACGA       GCGCACTCCACGA                 2   \n",
              "449  0.65  CGG  ACCCATCTGCGCGAGGTGCT       TGCGCGAGGTGCT                 3   \n",
              "191  0.70  TGG  TACGCGGGACGGCTTCGTGC       GACGGCTTCGTGC                 3   \n",
              "200  0.65  CGG  CAGCGCCGACGTGATGTTCG       GACGTGATGTTCG                 3   \n",
              "\n",
              "    editing_context editable_cytosines  \n",
              "334       0,0,0,0,0         5,6,8,9,10  \n",
              "330               1                 10  \n",
              "329           1,1,0              6,8,9  \n",
              "328           1,1,0              5,7,8  \n",
              "326               0                  6  \n",
              "..              ...                ...  \n",
              "108           0,0,0             6,7,10  \n",
              "295         0,0,0,1            3,5,6,9  \n",
              "449         0,0,0,1           3,4,7,10  \n",
              "191           0,1,0             3,5,10  \n",
              "200         1,1,0,0           4,6,7,10  \n",
              "\n",
              "[272 rows x 13 columns]"
            ]
          },
          "execution_count": 6,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "\n",
        "gene_sequences = process_specified_gene_sequences_from_record(genome, genes_to_KO)\n",
        "genes_to_KO_dict = {locus_tag: gene_sequences[locus_tag] for locus_tag in genes_to_KO if locus_tag in gene_sequences}\n",
        "\n",
        "# Identify and annotate base editing sites\n",
        "sgrna_df_with_editing = identify_base_editing_sites(sgrna_df)\n",
        "\n",
        "filtered_sgrna_df_for_base_editing = filter_sgrnas_for_base_editing(sgrna_df_with_editing)\n",
        "filtered_sgrna_df_for_base_editing"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 7,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>strain_name</th>\n",
              "      <th>locus_tag</th>\n",
              "      <th>gene_loc</th>\n",
              "      <th>gene_strand</th>\n",
              "      <th>sgrna_strand</th>\n",
              "      <th>sgrna_loc</th>\n",
              "      <th>gc</th>\n",
              "      <th>pam</th>\n",
              "      <th>sgrna</th>\n",
              "      <th>sgrna_seed_sequence</th>\n",
              "      <th>off_target_count</th>\n",
              "      <th>editing_context</th>\n",
              "      <th>editable_cytosines</th>\n",
              "      <th>mutations</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>334</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5090</td>\n",
              "      <td>5532706</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>32</td>\n",
              "      <td>0.65</td>\n",
              "      <td>CGG</td>\n",
              "      <td>TGGACCACCCGGTGTTCTCC</td>\n",
              "      <td>CCCGGTGTTCTCC</td>\n",
              "      <td>0</td>\n",
              "      <td>0,0,0,0,0</td>\n",
              "      <td>5,6,8,9,10</td>\n",
              "      <td>R8Q, V9I, V10I</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>325</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5089</td>\n",
              "      <td>5532449</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>106</td>\n",
              "      <td>0.65</td>\n",
              "      <td>AGG</td>\n",
              "      <td>CTTCCTCGACCTCCGCTTCG</td>\n",
              "      <td>GACCTCCGCTTCG</td>\n",
              "      <td>0</td>\n",
              "      <td>0,0,0,0</td>\n",
              "      <td>4,5,7,10</td>\n",
              "      <td>L31F</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>322</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5089</td>\n",
              "      <td>5532449</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>136</td>\n",
              "      <td>0.65</td>\n",
              "      <td>TGG</td>\n",
              "      <td>GTACGACTCGCTCGCCCTGA</td>\n",
              "      <td>TCGCTCGCCCTGA</td>\n",
              "      <td>0</td>\n",
              "      <td>0,0,0</td>\n",
              "      <td>4,7,9</td>\n",
              "      <td>S42L</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>318</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5089</td>\n",
              "      <td>5532449</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>194</td>\n",
              "      <td>0.65</td>\n",
              "      <td>CGG</td>\n",
              "      <td>TCCATACCCGACGACGTCGC</td>\n",
              "      <td>CCGACGACGTCGC</td>\n",
              "      <td>0</td>\n",
              "      <td>0,0,0,0</td>\n",
              "      <td>3,7,8,9</td>\n",
              "      <td>P61F</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>312</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5089</td>\n",
              "      <td>5532449</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>233</td>\n",
              "      <td>0.60</td>\n",
              "      <td>CGG</td>\n",
              "      <td>ATCAGGTCGAGCAGCTCTCG</td>\n",
              "      <td>CGAGCAGCTCTCG</td>\n",
              "      <td>0</td>\n",
              "      <td>0,0</td>\n",
              "      <td>3,8</td>\n",
              "      <td>D76N</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>...</th>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>534</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5090</td>\n",
              "      <td>5532706</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>140</td>\n",
              "      <td>0.65</td>\n",
              "      <td>CGG</td>\n",
              "      <td>CACGTACAGGTCCTGGAGCG</td>\n",
              "      <td>AGGTCCTGGAGCG</td>\n",
              "      <td>2</td>\n",
              "      <td>0,0</td>\n",
              "      <td>3,7</td>\n",
              "      <td>Q43*</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>78</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5087</td>\n",
              "      <td>5529801</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>745</td>\n",
              "      <td>0.65</td>\n",
              "      <td>GGG</td>\n",
              "      <td>CGACGACGATCGGGGTGATC</td>\n",
              "      <td>GATCGGGGTGATC</td>\n",
              "      <td>2</td>\n",
              "      <td>0,0</td>\n",
              "      <td>4,7</td>\n",
              "      <td>V247I, V248I</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>478</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5090</td>\n",
              "      <td>5532706</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>586</td>\n",
              "      <td>0.65</td>\n",
              "      <td>CGG</td>\n",
              "      <td>GGTGACACTCACCGAGGAGG</td>\n",
              "      <td>CTCACCGAGGAGG</td>\n",
              "      <td>2</td>\n",
              "      <td>0,0,0</td>\n",
              "      <td>6,8,10</td>\n",
              "      <td>T191I, L192F</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>335</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5090</td>\n",
              "      <td>5532706</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>43</td>\n",
              "      <td>0.65</td>\n",
              "      <td>CGG</td>\n",
              "      <td>TCCTCAGGGTGTGGACCACC</td>\n",
              "      <td>GGTGTGGACCACC</td>\n",
              "      <td>2</td>\n",
              "      <td>0,0</td>\n",
              "      <td>3,5</td>\n",
              "      <td>R14K</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>108</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5087</td>\n",
              "      <td>5529801</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>1038</td>\n",
              "      <td>0.70</td>\n",
              "      <td>GGG</td>\n",
              "      <td>GATGTCCGTCGCGTCCGTGC</td>\n",
              "      <td>GTCGCGTCCGTGC</td>\n",
              "      <td>2</td>\n",
              "      <td>0,0,0</td>\n",
              "      <td>6,7,10</td>\n",
              "      <td>D345N</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>112 rows × 14 columns</p>\n",
              "</div>"
            ],
            "text/plain": [
              "     strain_name locus_tag  gene_loc  gene_strand  sgrna_strand  sgrna_loc  \\\n",
              "334  NC_003888.3   SCO5090   5532706            1            -1         32   \n",
              "325  NC_003888.3   SCO5089   5532449            1             1        106   \n",
              "322  NC_003888.3   SCO5089   5532449            1             1        136   \n",
              "318  NC_003888.3   SCO5089   5532449            1             1        194   \n",
              "312  NC_003888.3   SCO5089   5532449            1            -1        233   \n",
              "..           ...       ...       ...          ...           ...        ...   \n",
              "534  NC_003888.3   SCO5090   5532706            1             1        140   \n",
              "78   NC_003888.3   SCO5087   5529801            1            -1        745   \n",
              "478  NC_003888.3   SCO5090   5532706            1             1        586   \n",
              "335  NC_003888.3   SCO5090   5532706            1            -1         43   \n",
              "108  NC_003888.3   SCO5087   5529801            1            -1       1038   \n",
              "\n",
              "       gc  pam                 sgrna sgrna_seed_sequence  off_target_count  \\\n",
              "334  0.65  CGG  TGGACCACCCGGTGTTCTCC       CCCGGTGTTCTCC                 0   \n",
              "325  0.65  AGG  CTTCCTCGACCTCCGCTTCG       GACCTCCGCTTCG                 0   \n",
              "322  0.65  TGG  GTACGACTCGCTCGCCCTGA       TCGCTCGCCCTGA                 0   \n",
              "318  0.65  CGG  TCCATACCCGACGACGTCGC       CCGACGACGTCGC                 0   \n",
              "312  0.60  CGG  ATCAGGTCGAGCAGCTCTCG       CGAGCAGCTCTCG                 0   \n",
              "..    ...  ...                   ...                 ...               ...   \n",
              "534  0.65  CGG  CACGTACAGGTCCTGGAGCG       AGGTCCTGGAGCG                 2   \n",
              "78   0.65  GGG  CGACGACGATCGGGGTGATC       GATCGGGGTGATC                 2   \n",
              "478  0.65  CGG  GGTGACACTCACCGAGGAGG       CTCACCGAGGAGG                 2   \n",
              "335  0.65  CGG  TCCTCAGGGTGTGGACCACC       GGTGTGGACCACC                 2   \n",
              "108  0.70  GGG  GATGTCCGTCGCGTCCGTGC       GTCGCGTCCGTGC                 2   \n",
              "\n",
              "    editing_context editable_cytosines       mutations  \n",
              "334       0,0,0,0,0         5,6,8,9,10  R8Q, V9I, V10I  \n",
              "325         0,0,0,0           4,5,7,10            L31F  \n",
              "322           0,0,0              4,7,9            S42L  \n",
              "318         0,0,0,0            3,7,8,9            P61F  \n",
              "312             0,0                3,8            D76N  \n",
              "..              ...                ...             ...  \n",
              "534             0,0                3,7            Q43*  \n",
              "78              0,0                4,7    V247I, V248I  \n",
              "478           0,0,0             6,8,10    T191I, L192F  \n",
              "335             0,0                3,5            R14K  \n",
              "108           0,0,0             6,7,10           D345N  \n",
              "\n",
              "[112 rows x 14 columns]"
            ]
          },
          "execution_count": 7,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "\n",
        "# Process the DataFrame to apply C-to-T mutations\n",
        "mutated_sgrna_df = process_base_editing(filtered_sgrna_df_for_base_editing, genes_to_KO_dict, only_stop_codons = False)\n",
        "mutated_sgrna_df"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 8,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>strain_name</th>\n",
              "      <th>locus_tag</th>\n",
              "      <th>gene_loc</th>\n",
              "      <th>gene_strand</th>\n",
              "      <th>sgrna_strand</th>\n",
              "      <th>sgrna_loc</th>\n",
              "      <th>gc</th>\n",
              "      <th>pam</th>\n",
              "      <th>sgrna</th>\n",
              "      <th>sgrna_seed_sequence</th>\n",
              "      <th>off_target_count</th>\n",
              "      <th>editing_context</th>\n",
              "      <th>editable_cytosines</th>\n",
              "      <th>mutations</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>534</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5090</td>\n",
              "      <td>5532706</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>140</td>\n",
              "      <td>0.65</td>\n",
              "      <td>CGG</td>\n",
              "      <td>CACGTACAGGTCCTGGAGCG</td>\n",
              "      <td>AGGTCCTGGAGCG</td>\n",
              "      <td>2</td>\n",
              "      <td>0,0</td>\n",
              "      <td>3,7</td>\n",
              "      <td>Q43*</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>319</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5089</td>\n",
              "      <td>5532449</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>167</td>\n",
              "      <td>0.70</td>\n",
              "      <td>CGG</td>\n",
              "      <td>GCGCGACTCGAGAGCCGGTA</td>\n",
              "      <td>TCGAGAGCCGGTA</td>\n",
              "      <td>0</td>\n",
              "      <td>1,0,0</td>\n",
              "      <td>4,7,9</td>\n",
              "      <td>R51*, L52F</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>352</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5090</td>\n",
              "      <td>5532706</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>174</td>\n",
              "      <td>0.65</td>\n",
              "      <td>CGG</td>\n",
              "      <td>CGCCCAGATCCGGAAGCGTT</td>\n",
              "      <td>ATCCGGAAGCGTT</td>\n",
              "      <td>0</td>\n",
              "      <td>1,0,0,0</td>\n",
              "      <td>3,4,5,10</td>\n",
              "      <td>W57*, A58T</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>354</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5090</td>\n",
              "      <td>5532706</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>208</td>\n",
              "      <td>0.65</td>\n",
              "      <td>CGG</td>\n",
              "      <td>GGGACGTCCAGGTCTTCACC</td>\n",
              "      <td>CCAGGTCTTCACC</td>\n",
              "      <td>0</td>\n",
              "      <td>0,0,0</td>\n",
              "      <td>5,8,9</td>\n",
              "      <td>W67*</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>17</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5087</td>\n",
              "      <td>5529801</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>210</td>\n",
              "      <td>0.60</td>\n",
              "      <td>GGG</td>\n",
              "      <td>GAGCAGTTCCCAGAACTGCC</td>\n",
              "      <td>TCCCAGAACTGCC</td>\n",
              "      <td>0</td>\n",
              "      <td>1,0,0</td>\n",
              "      <td>4,9,10</td>\n",
              "      <td>W67*, E68K</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>54</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5087</td>\n",
              "      <td>5529801</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>522</td>\n",
              "      <td>0.70</td>\n",
              "      <td>CGG</td>\n",
              "      <td>GTCGACCTCCCAGTCACGGC</td>\n",
              "      <td>TCCCAGTCACGGC</td>\n",
              "      <td>0</td>\n",
              "      <td>0,0,0,0,0</td>\n",
              "      <td>3,6,7,9,10</td>\n",
              "      <td>W171*, E172K, V173I, D174N</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>471</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5090</td>\n",
              "      <td>5532706</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>640</td>\n",
              "      <td>0.60</td>\n",
              "      <td>TGG</td>\n",
              "      <td>GACCGTCCAGGACATGACCA</td>\n",
              "      <td>CAGGACATGACCA</td>\n",
              "      <td>0</td>\n",
              "      <td>0,0,0,0</td>\n",
              "      <td>3,4,7,8</td>\n",
              "      <td>T208I, Q210*</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "     strain_name locus_tag  gene_loc  gene_strand  sgrna_strand  sgrna_loc  \\\n",
              "534  NC_003888.3   SCO5090   5532706            1             1        140   \n",
              "319  NC_003888.3   SCO5089   5532449            1             1        167   \n",
              "352  NC_003888.3   SCO5090   5532706            1            -1        174   \n",
              "354  NC_003888.3   SCO5090   5532706            1            -1        208   \n",
              "17   NC_003888.3   SCO5087   5529801            1            -1        210   \n",
              "54   NC_003888.3   SCO5087   5529801            1            -1        522   \n",
              "471  NC_003888.3   SCO5090   5532706            1             1        640   \n",
              "\n",
              "       gc  pam                 sgrna sgrna_seed_sequence  off_target_count  \\\n",
              "534  0.65  CGG  CACGTACAGGTCCTGGAGCG       AGGTCCTGGAGCG                 2   \n",
              "319  0.70  CGG  GCGCGACTCGAGAGCCGGTA       TCGAGAGCCGGTA                 0   \n",
              "352  0.65  CGG  CGCCCAGATCCGGAAGCGTT       ATCCGGAAGCGTT                 0   \n",
              "354  0.65  CGG  GGGACGTCCAGGTCTTCACC       CCAGGTCTTCACC                 0   \n",
              "17   0.60  GGG  GAGCAGTTCCCAGAACTGCC       TCCCAGAACTGCC                 0   \n",
              "54   0.70  CGG  GTCGACCTCCCAGTCACGGC       TCCCAGTCACGGC                 0   \n",
              "471  0.60  TGG  GACCGTCCAGGACATGACCA       CAGGACATGACCA                 0   \n",
              "\n",
              "    editing_context editable_cytosines                   mutations  \n",
              "534             0,0                3,7                        Q43*  \n",
              "319           1,0,0              4,7,9                  R51*, L52F  \n",
              "352         1,0,0,0           3,4,5,10                  W57*, A58T  \n",
              "354           0,0,0              5,8,9                        W67*  \n",
              "17            1,0,0             4,9,10                  W67*, E68K  \n",
              "54        0,0,0,0,0         3,6,7,9,10  W171*, E172K, V173I, D174N  \n",
              "471         0,0,0,0            3,4,7,8                T208I, Q210*  "
            ]
          },
          "execution_count": 8,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "# Process the DataFrame to apply C-to-T mutations\n",
        "mutated_sgrna_df = process_base_editing(filtered_sgrna_df_for_base_editing, \n",
        "                                        genes_to_KO_dict, \n",
        "                                        only_stop_codons = True, \n",
        "                                        editing_context= False)\n",
        "mutated_sgrna_df"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 9,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>strain_name</th>\n",
              "      <th>locus_tag</th>\n",
              "      <th>gene_loc</th>\n",
              "      <th>gene_strand</th>\n",
              "      <th>sgrna_strand</th>\n",
              "      <th>sgrna_loc</th>\n",
              "      <th>gc</th>\n",
              "      <th>pam</th>\n",
              "      <th>sgrna</th>\n",
              "      <th>sgrna_seed_sequence</th>\n",
              "      <th>off_target_count</th>\n",
              "      <th>editing_context</th>\n",
              "      <th>editable_cytosines</th>\n",
              "      <th>mutations</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>534</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5090</td>\n",
              "      <td>5532706</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>140</td>\n",
              "      <td>0.65</td>\n",
              "      <td>CGG</td>\n",
              "      <td>CACGTACAGGTCCTGGAGCG</td>\n",
              "      <td>AGGTCCTGGAGCG</td>\n",
              "      <td>2</td>\n",
              "      <td>0,0</td>\n",
              "      <td>3,7</td>\n",
              "      <td>Q43*</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>354</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5090</td>\n",
              "      <td>5532706</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>208</td>\n",
              "      <td>0.65</td>\n",
              "      <td>CGG</td>\n",
              "      <td>GGGACGTCCAGGTCTTCACC</td>\n",
              "      <td>CCAGGTCTTCACC</td>\n",
              "      <td>0</td>\n",
              "      <td>0,0,0</td>\n",
              "      <td>5,8,9</td>\n",
              "      <td>W67*</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>54</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5087</td>\n",
              "      <td>5529801</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>522</td>\n",
              "      <td>0.70</td>\n",
              "      <td>CGG</td>\n",
              "      <td>GTCGACCTCCCAGTCACGGC</td>\n",
              "      <td>TCCCAGTCACGGC</td>\n",
              "      <td>0</td>\n",
              "      <td>0,0,0,0,0</td>\n",
              "      <td>3,6,7,9,10</td>\n",
              "      <td>W171*, E172K, V173I, D174N</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>471</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5090</td>\n",
              "      <td>5532706</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>640</td>\n",
              "      <td>0.60</td>\n",
              "      <td>TGG</td>\n",
              "      <td>GACCGTCCAGGACATGACCA</td>\n",
              "      <td>CAGGACATGACCA</td>\n",
              "      <td>0</td>\n",
              "      <td>0,0,0,0</td>\n",
              "      <td>3,4,7,8</td>\n",
              "      <td>T208I, Q210*</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "     strain_name locus_tag  gene_loc  gene_strand  sgrna_strand  sgrna_loc  \\\n",
              "534  NC_003888.3   SCO5090   5532706            1             1        140   \n",
              "354  NC_003888.3   SCO5090   5532706            1            -1        208   \n",
              "54   NC_003888.3   SCO5087   5529801            1            -1        522   \n",
              "471  NC_003888.3   SCO5090   5532706            1             1        640   \n",
              "\n",
              "       gc  pam                 sgrna sgrna_seed_sequence  off_target_count  \\\n",
              "534  0.65  CGG  CACGTACAGGTCCTGGAGCG       AGGTCCTGGAGCG                 2   \n",
              "354  0.65  CGG  GGGACGTCCAGGTCTTCACC       CCAGGTCTTCACC                 0   \n",
              "54   0.70  CGG  GTCGACCTCCCAGTCACGGC       TCCCAGTCACGGC                 0   \n",
              "471  0.60  TGG  GACCGTCCAGGACATGACCA       CAGGACATGACCA                 0   \n",
              "\n",
              "    editing_context editable_cytosines                   mutations  \n",
              "534             0,0                3,7                        Q43*  \n",
              "354           0,0,0              5,8,9                        W67*  \n",
              "54        0,0,0,0,0         3,6,7,9,10  W171*, E172K, V173I, D174N  \n",
              "471         0,0,0,0            3,4,7,8                T208I, Q210*  "
            ]
          },
          "execution_count": 9,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "# Process the DataFrame to apply C-to-T mutations\n",
        "mutated_sgrna_df = process_base_editing(filtered_sgrna_df_for_base_editing, \n",
        "                                        genes_to_KO_dict, \n",
        "                                        only_stop_codons = True, \n",
        "                                        editing_context= True)\n",
        "mutated_sgrna_df"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 10,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>strain_name</th>\n",
              "      <th>locus_tag</th>\n",
              "      <th>gene_loc</th>\n",
              "      <th>gene_strand</th>\n",
              "      <th>sgrna_strand</th>\n",
              "      <th>sgrna_loc</th>\n",
              "      <th>gc</th>\n",
              "      <th>pam</th>\n",
              "      <th>sgrna</th>\n",
              "      <th>sgrna_seed_sequence</th>\n",
              "      <th>off_target_count</th>\n",
              "      <th>editing_context</th>\n",
              "      <th>editable_cytosines</th>\n",
              "      <th>mutations</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>534</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5090</td>\n",
              "      <td>5532706</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>140</td>\n",
              "      <td>0.65</td>\n",
              "      <td>CGG</td>\n",
              "      <td>CACGTACAGGTCCTGGAGCG</td>\n",
              "      <td>AGGTCCTGGAGCG</td>\n",
              "      <td>2</td>\n",
              "      <td>0,0</td>\n",
              "      <td>3,7</td>\n",
              "      <td>Q43*</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>354</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5090</td>\n",
              "      <td>5532706</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>208</td>\n",
              "      <td>0.65</td>\n",
              "      <td>CGG</td>\n",
              "      <td>GGGACGTCCAGGTCTTCACC</td>\n",
              "      <td>CCAGGTCTTCACC</td>\n",
              "      <td>0</td>\n",
              "      <td>0,0,0</td>\n",
              "      <td>5,8,9</td>\n",
              "      <td>W67*</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>54</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5087</td>\n",
              "      <td>5529801</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>522</td>\n",
              "      <td>0.70</td>\n",
              "      <td>CGG</td>\n",
              "      <td>GTCGACCTCCCAGTCACGGC</td>\n",
              "      <td>TCCCAGTCACGGC</td>\n",
              "      <td>0</td>\n",
              "      <td>0,0,0,0,0</td>\n",
              "      <td>3,6,7,9,10</td>\n",
              "      <td>W171*, E172K, V173I, D174N</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "     strain_name locus_tag  gene_loc  gene_strand  sgrna_strand  sgrna_loc  \\\n",
              "534  NC_003888.3   SCO5090   5532706            1             1        140   \n",
              "354  NC_003888.3   SCO5090   5532706            1            -1        208   \n",
              "54   NC_003888.3   SCO5087   5529801            1            -1        522   \n",
              "\n",
              "       gc  pam                 sgrna sgrna_seed_sequence  off_target_count  \\\n",
              "534  0.65  CGG  CACGTACAGGTCCTGGAGCG       AGGTCCTGGAGCG                 2   \n",
              "354  0.65  CGG  GGGACGTCCAGGTCTTCACC       CCAGGTCTTCACC                 0   \n",
              "54   0.70  CGG  GTCGACCTCCCAGTCACGGC       TCCCAGTCACGGC                 0   \n",
              "\n",
              "    editing_context editable_cytosines                   mutations  \n",
              "534             0,0                3,7                        Q43*  \n",
              "354           0,0,0              5,8,9                        W67*  \n",
              "54        0,0,0,0,0         3,6,7,9,10  W171*, E172K, V173I, D174N  "
            ]
          },
          "execution_count": 10,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "# Filter the DataFrame to retain only up to 5 sgRNA sequences per locus_tag\n",
        "filtered_df = mutated_sgrna_df.groupby('locus_tag').head(number_of_sgRNAs_per_group)\n",
        "filtered_df"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "## Output"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 11,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/plain": [
              "[Dseqrecord(-20), Dseqrecord(-20), Dseqrecord(-20)]"
            ]
          },
          "execution_count": 11,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "sgRNA_list = dataframe_to_seqrecords(filtered_df)\n",
        "sgRNA_list"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 12,
      "metadata": {},
      "outputs": [],
      "source": [
        "sgRNA_handle_cys4_sites = sgRNA_handle_cys4_site*len(sgRNA_list)\n",
        "\n",
        "golden_gate = GoldenGateCloning(sgRNA_list,\n",
        "                                sgRNA_handle_cys4_sites, \n",
        "                                target_tm= melting_temperature,\n",
        "                                restriction_overhang_f = restriction_overhang_f,\n",
        "                                restriction_overhang_r = restriction_overhang_r,\n",
        "                                backbone_overhang_f = backbone_overhang_f,\n",
        "                                backbone_overhang_r = backbone_overhang_r,\n",
        "                                cys4 = cys4,\n",
        "                                tm_function = primer_tm_neb, \n",
        "                                polymerase = chosen_polymerase ) "
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 13,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>Locus Tag</th>\n",
              "      <th>f_primer_name</th>\n",
              "      <th>r_primer_name</th>\n",
              "      <th>f_primer_sequences(5-3)</th>\n",
              "      <th>r_primer_sequences(5-3)</th>\n",
              "      <th>f_tm</th>\n",
              "      <th>r_tm</th>\n",
              "      <th>ta</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>sgRNA_handle_cys4</td>\n",
              "      <td>fwd_primer_SCO5090_140</td>\n",
              "      <td>rev_primer_SCO5090_140</td>\n",
              "      <td>GATCGGGTCTCCCATGGTTCACTGCCGTATAGGCAGCTAAGAAACA...</td>\n",
              "      <td>GATCAGGTCTCGTCCCAAAAAAGCACCGACTCG</td>\n",
              "      <td>56</td>\n",
              "      <td>54</td>\n",
              "      <td>58</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>sgRNA_handle_cys4</td>\n",
              "      <td>fwd_primer_SCO5090_208</td>\n",
              "      <td>rev_primer_SCO5090_208</td>\n",
              "      <td>GATCGGGTCTCCGGGACGTCCAGGTCTTCACCGTTTTAGAGCTAGA...</td>\n",
              "      <td>GATCAGGTCTCGCGACAAAAAAGCACCGACTCG</td>\n",
              "      <td>56</td>\n",
              "      <td>54</td>\n",
              "      <td>58</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>sgRNA_handle_cys4</td>\n",
              "      <td>fwd_primer_SCO5087_522</td>\n",
              "      <td>rev_primer_SCO5087_522</td>\n",
              "      <td>GATCGGGTCTCCGTCGACCTCCCAGTCACGGCGTTTTAGAGCTAGA...</td>\n",
              "      <td>GATCAGGTCTCGCTAGAAAAAAGCACCGACTCG</td>\n",
              "      <td>56</td>\n",
              "      <td>54</td>\n",
              "      <td>58</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "           Locus Tag           f_primer_name           r_primer_name  \\\n",
              "0  sgRNA_handle_cys4  fwd_primer_SCO5090_140  rev_primer_SCO5090_140   \n",
              "1  sgRNA_handle_cys4  fwd_primer_SCO5090_208  rev_primer_SCO5090_208   \n",
              "2  sgRNA_handle_cys4  fwd_primer_SCO5087_522  rev_primer_SCO5087_522   \n",
              "\n",
              "                             f_primer_sequences(5-3)  \\\n",
              "0  GATCGGGTCTCCCATGGTTCACTGCCGTATAGGCAGCTAAGAAACA...   \n",
              "1  GATCGGGTCTCCGGGACGTCCAGGTCTTCACCGTTTTAGAGCTAGA...   \n",
              "2  GATCGGGTCTCCGTCGACCTCCCAGTCACGGCGTTTTAGAGCTAGA...   \n",
              "\n",
              "             r_primer_sequences(5-3)  f_tm  r_tm  ta  \n",
              "0  GATCAGGTCTCGTCCCAAAAAAGCACCGACTCG    56    54  58  \n",
              "1  GATCAGGTCTCGCGACAAAAAAGCACCGACTCG    56    54  58  \n",
              "2  GATCAGGTCTCGCTAGAAAAAAGCACCGACTCG    56    54  58  "
            ]
          },
          "execution_count": 13,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "primer_df = golden_gate.generate_primer_dataframe()\n",
        "primer_df"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 14,
      "metadata": {},
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "Skipping analysis for primer pair fwd_primer_SCO5090_140 & rev_primer_SCO5090_140 due to length constraints.\n"
          ]
        },
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>annealing_temperature</th>\n",
              "      <th>primer_pair</th>\n",
              "      <th>homodimer_forward_tm</th>\n",
              "      <th>homodimer_forward_deltaG (kcal/mol)</th>\n",
              "      <th>homodimer_reverse_tm</th>\n",
              "      <th>homodimer_reverse_deltaG (kcal/mol)</th>\n",
              "      <th>heterodimer_tm</th>\n",
              "      <th>heterodimer_deltaG (kcal/mol)</th>\n",
              "      <th>hairpin_forward_structure_found</th>\n",
              "      <th>hairpin_forward_tm</th>\n",
              "      <th>hairpin_forward_deltaG (kcal/mol)</th>\n",
              "      <th>hairpin_reverse_structure_found</th>\n",
              "      <th>hairpin_reverse_tm</th>\n",
              "      <th>hairpin_reverse_deltaG (kcal/mol)</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>58</td>\n",
              "      <td>fwd_primer_SCO5090_208 &amp; rev_primer_SCO5090_208</td>\n",
              "      <td>38.497138</td>\n",
              "      <td>-1.774563</td>\n",
              "      <td>3.710596</td>\n",
              "      <td>-3.582472</td>\n",
              "      <td>10.978173</td>\n",
              "      <td>1.893397</td>\n",
              "      <td>True</td>\n",
              "      <td>59.149103</td>\n",
              "      <td>-0.053945</td>\n",
              "      <td>False</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>58</td>\n",
              "      <td>fwd_primer_SCO5087_522 &amp; rev_primer_SCO5087_522</td>\n",
              "      <td>25.023322</td>\n",
              "      <td>7.202194</td>\n",
              "      <td>-14.334752</td>\n",
              "      <td>-1.634182</td>\n",
              "      <td>16.105063</td>\n",
              "      <td>3.290762</td>\n",
              "      <td>True</td>\n",
              "      <td>53.974837</td>\n",
              "      <td>0.348222</td>\n",
              "      <td>False</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "   annealing_temperature                                      primer_pair  \\\n",
              "0                     58  fwd_primer_SCO5090_208 & rev_primer_SCO5090_208   \n",
              "1                     58  fwd_primer_SCO5087_522 & rev_primer_SCO5087_522   \n",
              "\n",
              "   homodimer_forward_tm  homodimer_forward_deltaG (kcal/mol)  \\\n",
              "0             38.497138                            -1.774563   \n",
              "1             25.023322                             7.202194   \n",
              "\n",
              "   homodimer_reverse_tm  homodimer_reverse_deltaG (kcal/mol)  heterodimer_tm  \\\n",
              "0              3.710596                            -3.582472       10.978173   \n",
              "1            -14.334752                            -1.634182       16.105063   \n",
              "\n",
              "   heterodimer_deltaG (kcal/mol)  hairpin_forward_structure_found  \\\n",
              "0                       1.893397                             True   \n",
              "1                       3.290762                             True   \n",
              "\n",
              "   hairpin_forward_tm  hairpin_forward_deltaG (kcal/mol)  \\\n",
              "0           59.149103                          -0.053945   \n",
              "1           53.974837                           0.348222   \n",
              "\n",
              "   hairpin_reverse_structure_found  hairpin_reverse_tm  \\\n",
              "0                            False                 0.0   \n",
              "1                            False                 0.0   \n",
              "\n",
              "   hairpin_reverse_deltaG (kcal/mol)  \n",
              "0                                0.0  \n",
              "1                                0.0  "
            ]
          },
          "execution_count": 14,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "# Primer analysis\n",
        "analysis_of_primers = analyze_primers_and_hairpins(primer_df)\n",
        "analysis_of_primers"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 15,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>Name</th>\n",
              "      <th>Sequence</th>\n",
              "      <th>Concentration</th>\n",
              "      <th>Purification</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>fwd_primer_SCO5090_140</td>\n",
              "      <td>GATCGGGTCTCCCATGGTTCACTGCCGTATAGGCAGCTAAGAAACA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>fwd_primer_SCO5090_208</td>\n",
              "      <td>GATCGGGTCTCCGGGACGTCCAGGTCTTCACCGTTTTAGAGCTAGA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>fwd_primer_SCO5087_522</td>\n",
              "      <td>GATCGGGTCTCCGTCGACCTCCCAGTCACGGCGTTTTAGAGCTAGA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>rev_primer_SCO5090_140</td>\n",
              "      <td>GATCAGGTCTCGTCCCAAAAAAGCACCGACTCG</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>rev_primer_SCO5090_208</td>\n",
              "      <td>GATCAGGTCTCGCGACAAAAAAGCACCGACTCG</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>5</th>\n",
              "      <td>rev_primer_SCO5087_522</td>\n",
              "      <td>GATCAGGTCTCGCTAGAAAAAAGCACCGACTCG</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "                     Name                                           Sequence  \\\n",
              "0  fwd_primer_SCO5090_140  GATCGGGTCTCCCATGGTTCACTGCCGTATAGGCAGCTAAGAAACA...   \n",
              "1  fwd_primer_SCO5090_208  GATCGGGTCTCCGGGACGTCCAGGTCTTCACCGTTTTAGAGCTAGA...   \n",
              "2  fwd_primer_SCO5087_522  GATCGGGTCTCCGTCGACCTCCCAGTCACGGCGTTTTAGAGCTAGA...   \n",
              "3  rev_primer_SCO5090_140                  GATCAGGTCTCGTCCCAAAAAAGCACCGACTCG   \n",
              "4  rev_primer_SCO5090_208                  GATCAGGTCTCGCGACAAAAAAGCACCGACTCG   \n",
              "5  rev_primer_SCO5087_522                  GATCAGGTCTCGCTAGAAAAAAGCACCGACTCG   \n",
              "\n",
              "  Concentration Purification  \n",
              "0          25nm          STD  \n",
              "1          25nm          STD  \n",
              "2          25nm          STD  \n",
              "3          25nm          STD  \n",
              "4          25nm          STD  \n",
              "5          25nm          STD  "
            ]
          },
          "execution_count": 15,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "# 2 Ready made IDT format\n",
        "idt_df = create_idt_order_dataframe(primer_df, concentration=\"25nm\", purification=\"STD\")\n",
        "idt_df"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 16,
      "metadata": {},
      "outputs": [],
      "source": [
        "list_of_amplicons = golden_gate.simulate_pcrs()"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 17,
      "metadata": {},
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "SCO5090_140_pcr_products\n",
            "SCO5090_208_pcr_products\n",
            "SCO5087_522_pcr_products\n"
          ]
        },
        {
          "data": {
            "image/jpeg": "/9j/4AAQSkZJRgABAQAAAQABAAD/2wBDAAgGBgcGBQgHBwcJCQgKDBQNDAsLDBkSEw8UHRofHh0aHBwgJC4nICIsIxwcKDcpLDAxNDQ0Hyc5PTgyPC4zNDL/2wBDAQkJCQwLDBgNDRgyIRwhMjIyMjIyMjIyMjIyMjIyMjIyMjIyMjIyMjIyMjIyMjIyMjIyMjIyMjIyMjIyMjIyMjL/wAARCAJYASwDASIAAhEBAxEB/8QAHwAAAQUBAQEBAQEAAAAAAAAAAAECAwQFBgcICQoL/8QAtRAAAgEDAwIEAwUFBAQAAAF9AQIDAAQRBRIhMUEGE1FhByJxFDKBkaEII0KxwRVS0fAkM2JyggkKFhcYGRolJicoKSo0NTY3ODk6Q0RFRkdISUpTVFVWV1hZWmNkZWZnaGlqc3R1dnd4eXqDhIWGh4iJipKTlJWWl5iZmqKjpKWmp6ipqrKztLW2t7i5usLDxMXGx8jJytLT1NXW19jZ2uHi4+Tl5ufo6erx8vP09fb3+Pn6/8QAHwEAAwEBAQEBAQEBAQAAAAAAAAECAwQFBgcICQoL/8QAtREAAgECBAQDBAcFBAQAAQJ3AAECAxEEBSExBhJBUQdhcRMiMoEIFEKRobHBCSMzUvAVYnLRChYkNOEl8RcYGRomJygpKjU2Nzg5OkNERUZHSElKU1RVVldYWVpjZGVmZ2hpanN0dXZ3eHl6goOEhYaHiImKkpOUlZaXmJmaoqOkpaanqKmqsrO0tba3uLm6wsPExcbHyMnK0tPU1dbX2Nna4uPk5ebn6Onq8vP09fb3+Pn6/9oADAMBAAIRAxEAPwDwiHTr24tJruGzuJLaD/WzJExSP/eYDA/GnvpOpRfZfM0+7T7Xj7NuhYednGNnHzdR09RXQaL51r4Xu78Xto7iO4t7eze8iRoxIgWaUozBmynyqACSef4QD0japYRapqszajaGPVb+WbT2WdW+zo1tcIhfB/dYaaEfNgjyz2FAHnyaJq0l9JYppd615Eu6S3Fu5kQcclcZA5H51CmnX0ljJfJZ3DWkTbZLgRMY0PHBbGAeR+ddxJJFPoMGiQanZRapbxWhlnN2ixkJLcsVEudrFRNBwCfucZxVjU9W029u7jVbS7t005LXU4GtjIqyNJNJOYsR53EYlhOQMDy+cYoA82oqZ2tjaRqkMwuQT5kjSgow7ALtBB99x/CoaAFCkjpRtPpU8VwEjC7Acd6f9rH/ADzX8qAKu0+lG0+lWvtY/wCea/lR9rH/ADzX8qAKu0+lG0+lWvtY/wCea/lR9rH/ADzX8qAKpBHUUlTzTiVAoUDnPFQUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFadnoc97ouoapHParFYqHeJpf3rguiZVRngGROTgcnGSCK0H8F6is6QrcWjyLKYbkK7f6I4QuRJlR0VHJ27vuN3FAHOUV0C+ErkuXfULCOyKRvHeu0gik8wsqgfJuzlHHKjGxs4xTX8JahDbTPO9vFcxrM4s3ZvNdImKyMMDbhSr9SD8jYBxQBg0UUUAFFTJbSSIGUHBp32OX+6aAK9FWPscv900fY5f7poAr0VY+xy/3TR9jl/umgCvRUskDxLuYYGcVFQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQBp6XqMNlp+tW8iuXvrJbeIqBgMLiGTJ56Yjbpnkj610z+MtMW9vriOG7b+17qS41BWRR5W+GaJljO47v+PiQgnb0UeprhqKAOvfxBotxpkehz/bxpkCQeXOkKea7RvO5ym/AB+0yD7xxhetS3vjGz1GabVJ4bhNT8i9t44kVTEVuXlYsWzkFftEgxtOcLyK4uigCZ7qR7SO2KwiOMllKwoHOfVwNxHsScVDRRQBMlzJGgVScCnfbJf7xqvRQBY+2S/3jR9sl/vGq9FAFj7ZL/eNH2yX+8ar0UASyTvKu1jkZzUVFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRW1puk6bfaVe3MmoXcM9pbtM6fZFaMncFRd/mA5Yso+7xyeQK1H8EL9qmt49S3PZXDW+pFoMC3ZYpJGKfMd4AhlH8PK++aAORorqD4Vs47cajPqskekyJC0NwLUGRmkaRQGj34GDDLk7j93jOawNRsZtL1O70+4x59rM8Em05G5WKnH4igCtRRRQAUVbgtVkhVy6gnsTUn2FP+ei/nQBQoq/9hT/AJ6L+dH2FP8Anov50AUKKv8A2FP+ei/nR9hT/nov50AUKKtXNssMYYMDk44qrQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQBf/tLboX9mRRbN9x588u7PmYXCLjHAXLn33e1b7+N1+1TXEem7XvbhrjUg0+RcM0UkbBPlGwETSn+LlvbFcjRQB1B8VWcluunT6VJJpMaQrDbi6AkVo2kYFpNmDkzS5G0fe4xisW41i9uNcuNYWd4L2ed5zLCxQq7kk4I5HU1RooAVmZ3LuxZmOSSckmkoooAcHYDAPFHmP602igB3mP60eY/rTaKAHeY/rR5j+tNooAcXZhgmm0UUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFby+HlHhhNYkbUDv34ENlvhTB2jfLvG3Jz/CelaD+CF+1TW8epbnsrhrfUi0GBbssUkjFPmO8AQyj+HlffNAHI0V1B8K2cduNRn1WSPSZEhaG4FqDIzSNIoDR78DBhlydx+7xnNYGo2M2l6nd6fcY8+1meCTacjcrFTj8RQBWooooAKK7LQvClvqWjW9293DG0m7Ks3IwxH9K0f+EGtf8An/g/76oA88or0P8A4Qa1/wCf+D/vqj/hBrX/AJ/4P++qAPPKK9D/AOEGtf8An/g/76o/4Qa1/wCf+D/vqgDzyiuq8S+G4dH06O4juYpS0wTahyehOf0rlaACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKAN6x8RRWWliEWczXaW89skpuj5QSUENmIg/NhjgggZwSCRzoP43X7VNcR6bte9uGuNSDT5FwzRSRsE+UbARNKf4uW9sVyNFAHUHxVZyW66dPpUkmkxpCsNuLoCRWjaRgWk2YOTNLkbR97jGKxbjWL241y41hZ3gvZ53nMsLFCruSTgjkdTVGigBWZncu7FmY5JJySaSiigCzFfzwxCNJGCjoAaf/al1/z1b86p0UAXP7Uuv+erfnR/al1/z1b86p0UAXP7Uuv+erfnR/al1/z1b86p0UAWJ7ya4QJI5YA5wTVeiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigC5p2lX2qvcLY27TG2t5LmbBACRIMsxJ9B+fQc1cfwtrMb2yNZ5e4fYirKjFW27sOAf3Z2nOGxxk9jWv4T8RaRpdlcWV7a3KGaC78y4iuAolLW0kcaFfLY9XIBzgF8nIAxqt4l0aPUNUuFvt6a3eS3DhYnzZLJb3EeGyBkhrn+HPEfuBQBycfhbWJLx7ZbeLeqI+83MQjKucLtkLbGyeBgnOD6GsmSN4ZXilRkkRirKwwVI6giu1l1LR7rQ4fDzatHElvHbH7f5MpSRkkuXZQAu/pc4GVGfL7ZFc7quqw3/irUNWFoksFzdy3At5iwGHYkA7SDxnsaAMqilYhnJChQTkKM4HtzSUAWIrSSWMOoODT/ALBL/dNMiu5Iowik4FP+3y/3jQAfYJf7po+wS/3TR9vl/vGj7fL/AHjQAfYJf7po+wS/3TR9vl/vGj7fL/eNAEc1s8KBmGATioKnmuXmQKxyAc1BQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAWotMv57GW+hsbmS0iOJJ0iYxoeOrAYHUfnTn0nUovsvmafdp9rx9m3QsPOzjGzj5uo6eorpNAljtNGubma/tGifT7qFRJPia0kZWGyOIt8wkyoLbSAGPIK5roG1Swi1TVZm1G0Meq38s2nss6t9nRra4RC+D+6w00I+bBHlnsKAPPk0TVpL6SxTS71ryJd0luLdzIg45K4yByPzqh0NegySRT6DBokGp2UWqW8VoZZzdosZCS3LFRLnaxUTQcAn7nGcVyus3mn3ni/U71YpJNPnvZpUjicRMY2diuCVOOCOxoAyKKVtpclAQueATkgfWkoAlS3d0DAHBp32ST+6amgvmhhWMYwM1J/ab+1AFX7JJ/dNH2ST+6atf2m/tR/ab+1AFX7JJ/dNH2ST+6atf2m/tR/ab+1AFOSB41ywwM4qKrVzeNcRhD2Oaq0AFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUVoWuh6je6fLfW8AaCIOSTIoZggDPtUnc20EE7QcA5OKsP4W1mN7ZGs8vcPsRVlRirbd2HAP7s7TnDY4yexoAx6K2I/C2sSXj2y28W9UR95uYhGVc4XbIW2Nk8DBOcH0NZMkbwyvFKjJIjFWVhgqR1BFADaKKKACipo7Z5EDAHBp/2OX+6aAK1FWfscv900fY5f7poArUVZ+xy/3TR9jl/umgCtRUstu8ShmGBnFRUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAdJpM1lZeHLyddWgTVJUkgjt5VmzFEy4cqQhXe4JTkgAZz1BHQN4l0aPUNUuFvt6a3eS3DhYnzZLJb3EeGyBkhrn+HPEfuBXndFAHby6lo91ocPh5tWjiS3jtj9v8mUpIySXLsoAXf0ucDKjPl9siud1XVYb/AMVahqwtElgubuW4FvMWAw7EgHaQeM9jWVRQArEM5IUKCchRnA9uaSiigC/b6g0ECxjGBn+dSf2q/t+VZlFAGn/ar+35Uf2q/t+VZlFAGn/ar+35Uf2q/t+VZlFAFy7vWuYgh7NmqdFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUVt6XpunXvh7WrmWS6/tCygWeJFCiLaZoYzuPJJPmngYxtHJzgbt14OtItSjtxb3aW0SyPc3QvYptwjgaZ1CKgMbEI23cTx2NAHD0V17+H9Ft9Mj1yf7edMnSDy4EmTzUaR50OX2YIH2aQ/dGcr0rCv8ARbq08R3miQRyXVzb3MluFhQs0hRiCQBz2zQBm0UrKyOUdSrKcEEYINJQA9Y2YZAOKPJf+6a9V8G+HfDt/wCFLK5v75Yrl/M3oV6YdgP0Ard/4RHwj/0E0/75oA8N8l/7po8l/wC6a9y/4RHwj/0E0/75o/4RHwj/ANBNP++aAPDfJf8AumjyX/umvcv+ER8I/wDQTT/vmj/hEfCP/QTT/vmgDwxo2UZIIptek/ELQtC0zQIJtMvFmna6VGUDGF2uc/mBXm1ABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFAFuz1GaytdQt41QpfW4t5SwOQolSTI565jXrngn61rJ4vu4Sn2axsLYGQS3CxRsFuW2NH84LEAFZJBhNo+c+2OeooA6BfFtyHKPp9hJZBI0jsnWQxR+WWZSPn3Zy7nljne2c5rFu7qa9vJ7u4cyTzyNJI56szHJP5moaKACiiigC3DqE8ESxpIQo6AGpP7Wuv+erfnVCigC//AGtdf89W/Oj+1rr/AJ6t+dUKKAL/APa11/z1b86P7Wuv+erfnVCigCzcX01zGEkcsAc8mq1FFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFa+n6NBfaFquoHUFS4sIllFqIiS6mWOMktwFGZRjqTtPA4Na7+CF+1TW8epbnsrhrfUi0GBbssUkjFPmO8AQyj+HlffNAHI0V1B8K2cduNRn1WSPSZEhaG4FqDIzSNIoDR78DBhlydx+7xnNYGo2M2l6nd6fcY8+1meCTacjcrFTj8RQBWooooAKK2LHR3urOOYYw2e/oSKsf8I/J7fnQBz9FdB/wj8nt+dH/CPye350Ac/RXQf8I/J7fnR/wj8nt+dAHP0VqalpbWVusrYwXC/of8Ky6ACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKAL+n6l9hstVt/K3/b7Vbfdux5eJopc9Of9VjHHXPbB338br9qmuI9N2ve3DXGpBp8i4ZopI2CfKNgImlP8XLe2K5GigDqD4qs5LddOn0qSTSY0hWG3F0BIrRtIwLSbMHJmlyNo+9xjFYtxrF7ca5caws7wXs87zmWFihV3JJwRyOpqjRQArMzuXdizMckk5JNJRRQBcgv5YIVjViAKk/tSf8Avms+igDQ/tSf++aP7Un/AL5rPooA0P7Un/vmj+1J/wC+az6KALVzeyXMYR2JAOaq0UUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQBp2mgX99YyXlv9keKKJ5nU3kKyBEzuPllt3b054xnIqV/C2sxvbI1nl7h9iKsqMVbbuw4B/dnac4bHGT2NNgvLez8N3MMMmb2+lEcoCkeXAmGxnod77Tx08r3rrm8S6NHqGqXC329NbvJbhwsT5slkt7iPDZAyQ1z/AA54j9wKAOFv9OutMuBBdIquUDqUkV1dT0KspIYe4NVa2dfurWVNLs7W4W5Sws/s7TorKsjGWSQkBgDgeZt5A+7WY88bWkcItYUkQktOpfe+exBbbgewFAENFFFAGrZ6PNdWiTKpKtn+eKn/AOEfn/uGoLPWJrW0SFWIVc/zzU//AAkE/wDfNAB/wj8/9w0f8I/P/cNH/CQT/wB80f8ACQT/AN80AH/CPz/3DR/wj8/9w0f8JBP/AHzR/wAJBP8A3zQBV1DS5bKBZHUgFtv6H/Cs2tLUNUlvYFjdiQG3fof8azaACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooA0E0a8fS/7R/0dLY7tpkuokdtvXahYM3XsDVh/C2sxvbI1nl7h9iKsqMVbbuw4B/dnac4bHGT2Nallqmmnwwlrez2UixW1xGLY2Z+0ea24xssuCAoYqT8y8AjB4rYbxLo0eoapcLfb01u8luHCxPmyWS3uI8NkDJDXP8OeI/cCgDhb/TrrTLgQXSKrlA6lJFdXU9CrKSGHuDVWtnX7q1lTS7O1uFuUsLP7O06KyrIxlkkJAYA4HmbeQPu1mPPG1pHCLWFJEJLTqX3vnsQW24HsBQBDRRRQB2ug+AtQ1rRbfUIIGaOXdggejFf6Vpf8Ku1X/n2f/vmt3wN8V5/DXg6w0hAm238zGVB+9Izf+zV0P/C9Lr0j/wC+RQBwP/CrtV/59n/75o/4Vdqv/Ps//fNd9/wvS69I/wDvkUf8L0uvSP8A75FAHA/8Ku1X/n2f/vmj/hV2q/8APs//AHzXff8AC9Lr0j/75FH/AAvS69I/++RQB5N4o8GXvh3TI7u5hZEeYRAkdyGP9K5OvU/iV8SZvGPhy30+QLiO7WfgAdEdf/Zq8soAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAUMwGAaXe3qaQKSMgUuxvSgA3t6mje3qaNjelGxvSgA3t6mje3qaNjelGxvSgBCxPU0lKVI6ikoAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooA6Lw34dg1W2vrq+mkhhitblrYR4zNPFA8uOf4QFG4/7Sjjdka7+DdMa9vreOa7X+yLqS31BmdT5uyGaVmjG0bf+PeQAHd1U+orn9H8Va1oStHY6hcpA0U0fkec4jHmIyFgoIG4bsg+oB7VafxpqLTpMtvaJI0pmuiqN/pblChMmWPVXcELt++3c0AUdb0+2s/7PubLzha39r9ojSZgzx4keNlLAAH5o2IOBwRVB7O5jtI7t7aZbaUlY5mQhHI6gN0JFWNT1OTVJoWaGGCKCIQwwQhtkaAk4G4k8lmJJJ5JqjQAUUUUAdVo+iw3elQzvKis27IJ9GIq9/wjtv8A89o/zrlINQlghWNWIAz/ADqT+1Zv75oA6f8A4R23/wCe0f50f8I7b/8APaP865j+1Zv75o/tWb++aAOn/wCEdt/+e0f50f8ACO2//PaP865j+1Zv75o/tWb++aANLX9KisbBJUkRiZQuAfY/4VzlW7q9kuYwjsSAc1UoAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKcEYjIFHlt6UANop3lt6UeW3pQA2ineW3pR5belADaKUqQMkUlABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFAHRaLpGmanpV8zvcLd21tLcM4mRVXauUUR4LyZIwSCAoOTwOdd/BumNe31vHNdr/AGRdSW+oMzqfN2QzSs0Y2jb/AMe8gAO7qp9RXPWviK4tNOFqlramVYZbeO6ZW82OKQEOgw20ghnGSpI3HBHGLb+NNRadJlt7RJGlM10VRv8AS3KFCZMsequ4IXb99u5oAo63p9tZ/wBn3Nl5wtb+1+0RpMwZ48SPGylgAD80bEHA4Iqg9ncx2kd29tMttKSsczIQjkdQG6EirGp6nJqk0LNDDBFBEIYYIQ2yNAScDcSeSzEkk8k1RoAKKKKAOv0TQ4LzSIJ3lRWbdkE+jEf0rQ/4Rq2/57R/nXH22py29usSsQFz/Opf7Zn/AL5oA6v/AIRq2/57R/nR/wAI1bf89o/zrlP7Zn/vmj+2Z/75oA6v/hGrb/ntH+dH/CNW3/PaP865T+2Z/wC+aP7Zn/vmgDT8R6RDYafHLHIjEyhcA+x/wrmKu3eoSXUQR2JAbNUqACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigCzDp17cWk13DZ3EltB/rZkiYpH/vMBgfjT30nUovsvmafdp9rx9m3QsPOzjGzj5uo6eoroNF8618L3d+L20dxHcW9vZveRI0YkQLNKUZgzZT5VABJPP8IB6RtUsItU1WZtRtDHqt/LNp7LOrfZ0a2uEQvg/usNNCPmwR5Z7CgDzW6tLmxuXtru3lt50xuimQoy5GeQeahrd8SPGE0e0E0U1xZ2AhuHhkWRd/myOAGUkNhHReCRxjtWQ7WxtI1SGYXIJ8yRpQUYdgF2gg++4/hQBDRRRQBqWmky3VqkyqSGz/PFTf2DP/cP5Vv+H9ft7LQ7e3eJGZN2SRzyxP8AWtL/AISm1/54RflQBx39gz/3D+VH9gz/ANw/lXY/8JTa/wDPCL8qP+Eptf8AnhF+VAHHf2DP/cP5Uf2DP/cP5V2P/CU2v/PCL8qP+Eptf+eEX5UAcJf6ZJZwLI6kAtt/Q/4Vn11/irWodR0uKGONFKzBsqP9lh/WuRVSzYHWgBKKk8h/SjyH9KAI6Kk8h/SjyH9KAI6Kk8h/SjyH9KAI6KUjBIPakoAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigB6yMowDxR5z+tMooAf5z+tHnP60yigB/nP60ec/rTKKAHNIzDBNaXh21S9122t3YKr7sk+yk/0rLqezuGtbtJlOCuf5YoA9H/4Ra1/57R/nR/wi1r/z2j/OuO/t+f8Avn86P7fn/vn86AOx/wCEWtf+e0f50f8ACLWv/PaP8647+35/75/Oj+35/wC+fzoA7H/hFrX/AJ7R/nR/wi1r/wA9o/zrjv7fn/vn86P7fn/vn86AMzUohBqt5CpBEc7qCPZiKrYJ7U+eQy3Ekh6uxY/ia6DS9AkvdNiuFUkPn9CR/SgDnNp9KNp9K6//AIRab+4fyo/4Rab+4fyoA5DafSjafSuv/wCEWm/uH8qP+EWm/uH8qAOQ2n0oIIrr/wDhFpv7h/Ks3W9Fk06zSZlIBkC/oT/SgDB60u0+lW9LtTe6lFbqMl8/oCf6V0n/AAi039w/lQByG0+lG0+ldf8A8ItN/cP5Uf8ACLTf3D+VAHIbT6UbT6V1/wDwi039w/lR/wAItN/cP5UAchtPpSV2H/CLTf3D+VcrdxGC8nhPWORl/I4oAhooooAKK17DQ11DT7m6j1Wxje2geeS3kEwcKpwORGUySVA+bqwHFR6to40krG+o2k9xx5kEIl3R5GeSyBeM44J5oAzKKKKACiiigBwRiMgHFHlv/dNegeGfBs2reH7W9WMlZd/OPRyP6Vrf8K8n/wCeJ/KgDyny3/umjy3/ALpr1b/hXk//ADxP5Uf8K8n/AOeJ/KgDyny3/umjy3/umvVv+FeT/wDPE/lR/wAK8n/54n8qAPKCrKMkYpK7fxj4Vl0PSIrl4yoacR5x6qx/pXEUAFFFFABRRRQAUUUUAFdJpevyWWmxW6sQEz+pJ/rXN0uSO9AHX/8ACUzf3z+dH/CUzf3z+dchuPrRuPrQB1//AAlM398/nR/wlM398/nXIbj60bj60Adf/wAJTN/fP51m63rUmo2aQsxIEgb9CP61hbj60Ek0AW9LujZalFcKcFM/qCP610n/AAlM398/nXH9KXcfWgDr/wDhKZv75/Oj/hKZv75/OuQ3H1o3H1oA6/8A4Smb++fzo/4Smb++fzrkNx9aNx9aAOv/AOEpm/vn865W7lM95PMeskjN+ZzUW4+tJQAUUUUAaSahDB4dksIVcXFzOHuHIGPLQfIoOc8sWJ/3U9K1/EXiW31bSzbi4v7uV7hJ1N4iD7KAjBo4yp5DFgScL9wcZJrlqKAJba4e1uEnjWJnQ5AliWRfxVgQfxFRsxdyxxknJwAB+Q6UlFABRRRQB6F4Z8ZTaT4ftbJZCFi38Z9XJ/rWt/wsOf8A57H868pDsBgE4o8x/wC8aAPVv+Fhz/8APY/nR/wsOf8A57H868p8x/7xo8x/7xoA9W/4WHP/AM9j+dH/AAsOf/nsfzrynzH/ALxo8x/7xoA7Xxj4ql1zSIrZ5CwWcSYz6Kw/rXEUpZmGCc0lABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAf//Z",
            "image/png": "iVBORw0KGgoAAAANSUhEUgAAASwAAAJYCAIAAABXRExoAAAi2ElEQVR4Ae3da0wV+fnA8T8IuNwEWxAQQzfhEk2XJu0b0w1po10xkXW7ujYhaQjxUt6YfUEJlKbUF11eWLoQ0TSra7OENV2vG+qNsCYm/1Ua47q1sdJGWo0tGi6CCgoIKPB//j0nnAMDh3OZOeeZw5cXZvjNb555fp85j2fOnGF+//M//CCAAAIIIIAAAggggAACCCCAAAIIhEQgIiR7XXSnERFKE5uenl40eTog4JNAlE+9g9Y5MjIyaPvyaUeTk5M+9aczAosKuIpw165dO3fuTExMrKqqunTpUlJSUnNzc0pKysDAQGlp6dDQkH8ti2Ywb4eoKFdi83YIVSNFGCr5MN6v86wvNTX1zJkzGzZsyMvLa2lpWbduXV1dXW9vb0NDQ0VFhaytrq72r8U/uxUrVvi3odVbPXv2zOpdEH+pCTiLcO3atfn5+adPn46Pj79///6qVas6OjoKCwu7u7szMzPb2tpkrX8t/oGmp6f7t6HVW8l/TFbvgvhLTcB51nfnvz8y+B07dpw/f14W0tLSHC+4np4eWfa7ZamBMl4EfBWY9dErOzu7srJSTkp9jbJo/9ra2oKCgsbGRjnXXbSzdBgdHfWmG30QCAMBVxEmJCScPHly9+7d/f39MrC+vj45J5TT0YyMDFn2u8VhVFNT4xPW+Pi4T/3pjIB9BZzfBMj3cnIttL6+/vr1647BtLa2FhcXy7L8K8uy4F+LfzSvtP74Nxy2QmBxAflyYnh4+H//+3PhwgXZQL6Q+NOf/tTe3i7/yrLfLYvve74e8p+Czp/5kqUNAQQQQAABBBBAAAEEEEAAAQQQQAABBBBAAAEEEEAAAQQQQAABBBBAAAEEEEAAAQQQQAABBBDwSkDpQ83k7m2v0g96J562FnTy8N+h6+8JVY1V7YOeXr58qQqKZMJAwFWEycnJBw4c2LZt28wfLoXwaWvR0dE6cSlCncfF1lm5Hu8pj5b5y1/+MnO69atf/erKlSvyTIqrV6/+8pe/lEH61+Kfzmtaf/wbDlsh4EHA9dFLHmYhT3YaHByUt0TZwL9nqxm38rBvD6vWrFnjYW0IVz18+DCEe2fXYSngeiec8zA/nrYWlsebQSkUcH0mtDQ5nrZmKS/BbS2wYBGG9mlrY2NjtmYleQS8F3Cdjs7Zxr9nqxm3mhPWy1/lIqTOHy/zpxsC/gvIhRnHxjxtbd7Hvfkvy5YIIIAAAggggAACCCCAAAIIIIAAAggggAACCCCAAAIIIIAAAggggAACCCCAAAIIIIAAAggggAACCCCAAAIIIIAAAggggAACCCCAAAIIIIAAAggggAACCCCw5ARcT+BWNfTIyAUfAxfaPKempkKbAHsPP4EFnzsa2qGqnZVpYmIitDLsPfwElBZhTEyMTmuKUOdxsXVWziI8ceKETAgjI5FXf15eXkpKijx31I+p0cyykEmZzAplbpzh4WFzAxINgbmfCffs2ZOVlbVv3766ujqZIqahoaGioiI1NbW6utqbFrNAv/Wtb5kVytw4//nPf8wNSDQEZhWhPHP65s2bmzdvfvTokXGSM29azAKlCM2SJI5+gVmfCbdu3Xrjxg2pQMnbv6nRzBrwyMiIWaGIg4BygVlFKGeeZWVlZmXs63Ro7vtlViZ3DZbDW8BVhOvXr5fZYDo7Ox0D9m9qNHesmpoa9199WuYipE9cdLa1gOs78aqqqvr6+pnBGCc586ZlZvMAF15p/QlwXGyOwIICOTk5165dc1/t39Ro7hFYRgABBBBAAAEEEEAAAQQQQAABBBBAAAEEEEAAAQQQQAABBBBAAAEEEEAAAQQQQAABBBBAAAEEEEAAAQQQQAABBBBAAAEEEEAAAQQQQAABBBBAAAEEEAixwKyH/4Y4F7fdy2OI3X5TtDg9Pa0oG1IJCwHXIw9VDUftrEwvX75UBUUyYSCgtAjVTghDEYbBi17bEJQWoTxwUZuUI5/nz5/rTIys7CugtAgzMjJ0mj58+FBnYmRlXwFnEW7atOnIkSNdXV0ykvb2dnmCfWjnJ8zOztZpKhPm6EyMrOwr4LwIWVJSEh8ff/jw4ZmReDMbobHPzOYBLvzmN78JMIJFm8vMjRZFJuySFXDORSGnfz09Pe4KW7Zskel7pUX+LSoqkgVvWtwjsIwAAt4IOE9HpQhlOorKysonT56Ul5ffu3cvtPMT/uMf//Ame/ogEAYCziKU76Bv3bolkxNu37796NGjGzduDHxsgcxPKP8LBJ4AERCwhYCzCA8ePPjgwQPJ+OzZs3KFRhZCOz/hnHNjW1CSJAL+CTiLcP/+/cePH5cKlKlCb9++LbEcsxE2NDQUFxfLspct/iVh3GpoaMjYSAsCYSngvDqam5vb1NQkM3PKPNV79+6Vs8HQfkURHR2tk5s7ZnQeF1tnpfQ+aW7gtvWriuQRQAABBBBAAAEEEEAAAQQQQAABBBBAAAEEEEAAAQQQQAABBBBAAAEEEEAAAQQQQAABBBBAAAEEENAowF9R+HZUeAy+b1709kLA+Ue9XvQMapfISOcTqIK6Vy92Njk56UUvuiDgg4DSIlT7R70UoQ8vLrp6J6C0CJcvX+5d/sHuJU8eCPYu2V+4C8wtQnm46Oeffx4bGxvax1tIAjrlefiNzuNi66xmXZhJTEz84osv3njjjRUrVhifru1Ni1kW8hBUs0KZG+fu3bvmBiQaArOuf8gz1w4cODA1NSUu3jxv29gHUAQQ8FXAdTpaUFCwevXqU6dOffzxxxIltE/gHh4e9nUk9EfApgLOIpQLIR9++KE8ftvEYQTyBO4XL16YmAmhENAs4CzC9957Tz4QfvbZZ5JrQkLCp59+GtoncI+Pj2tWIzcErBUYHByUHchlmJ///OeyUFFR8dvf/tbLFulmys8yrT+mjI4gCLgLzLo66lghRZicnBzaryh4+K/7QWIZAQQQQAABBBBAAAEEEEAAAQQQQAABBBBAAAEEEEAAAQQQQAABBBBAAAEEEEAAAQQQQAABBBBAAAEEEEAAAQQQQAABBBBAAAEEEEAAAQQQQAABBBBAAAEEEEAAAQT8EJjnaWt+RDF9E7XzEzrmCDB9vARcygKux+CrUqAIVR0OkrFUQGkRqp0k9NWrV5YeD4IvQQGlRfjaa6/pPBhMkqHzuNg6K2cRxsXFNTU1yUxMMiPFvn37Ll68GNoncKudJPTp06e2Pt4kr1DAeWGmqqpKLjnIxEwZGRnXrl17/fXXvZkS1NjHrBHm5eWZFcrcOP/85z/NDUg0BJxFuHLlypGRkYmJibfeeuujjz7Kzc3t6OgoLCzs7u7OzMxsa2vLz8/3psUsUIrQLEni6Bdwno46zrKOHTsmUxS+8847kjeThOo/eGQYHgKzLsyUlJScOXOmtLT08uXLgQ+PSUIDNyTCUhBwFuGhQ4fKy8vl+vuFCxc++eQTGXloJwkdGxtbCvqMEQERiHQoyLXQd999V5bffPPNzs5OWWhtbS0uLpYF+VeWvWyRbqb8vNT6Y8roCIKAu4DzwkxWVlZzc7PcpyLXZt5///07d+6E9isK7phxP0gsI4AAAggggAACCCCAAAIIIIAAAggggAACCCCAAAIIIIAAAggggAACCCCAAAIIIIAAAggggAACCCCAAAIIIIAAAggggAACCCCAAAIIIIAAAggggAACCCCAAAIIIIAAAggggAACCCCAAAIIIIAAAggggAACCCCAAAIIIIAAAggggAACCCCAAAIIIIAAAggggAACCCCAAAIIIIAAAggggAACCCCAAAIIIIAAAggggAACCCCAAAIIIIAAAggggAACToEInRIREUoTm56e1ilGVvYViNKZOkWo87iQlRUCSoswKkppYhMTE1YcBmIuZQGlr/Xo6GidR4Ui1HlcbJ2Vqwh37dq1c+fOxMTEqqqqS5cuJSUlNTc3p6SkDAwMlJaWDg0NedNilsVrr71mVihz44yMjJgbkGgIOK9/pKamnjlzZsOGDXl5eS0tLevWraurq+vt7W1oaKioqJC11dXV3rSYBZqVlWVWKHPjdHV1mRuQaAg4i3Dt2rX5+fmnT5+Oj4+/f//+qlWrOjo6CgsLu7u7MzMz29raZK03LWaBfvvb3zYrlLlx/v73v5sbkGgIOE9H7/z3Rzh27Nhx/vx5WUhLS5N3Qlno6emRZS9bpBs/CCDgk4DrM6Fslp2dXVlZKSelPoVYqHNtbW1BQUFjY6Oc3y7UZ6H258+fL7SKdgTCTMBVhAkJCSdPnty9e3d/f78Msq+vLz09XU5HMzIyZNnLFnedmpoa9199Wub6h09cdLa1gLMI5ctxuRZaX19//fp1x3haW1uLi4vlwoz8K8vS6E2LWRZjY2NmhSIOAsoFnBdm5MuJQ4cOff3115Lu8PDw22+/7c0XEsY+Zo02JibGrFDmxuF7QnM9iSYCziLUZhEZGaktJUc+U1NTOhMjK/sKuD4TqhoD90mrOhwkgwACCCCAAAIIIIAAAggggAACCCCAAAIIIIAAAggggAACCCCAAAIIIIAAAggggAACCCCAAAIIIIAAAggggAACCCCAAAIIIIAAAggggAACCCCAAAIIIIAAAggggAACCCCAAAIIIIAAAggggAACCCCAAAIIIIAAAggggAACXgkonZVJ5kv0Kv2gd2KmmqCTh/8Olc7KpHZqtMnJyfB/UTDC4AooLcKoKKWJUYTBfX0uib0pfa3HxcXp5B8fH9eZGFnZV0BpEcpE3DpNnz59qjMxsrKvgNIizMjI0Gn673//W2diZGVfAaVFmJOTo9P02rVrOhMjK/sKuIowOTn5wIED27Ztc5wKbtq06ciRI11dXTK29vb2mpoaaW9ubk5JSRkYGCgtLR0aGjK2mAXxne98x6xQxEFAuYDr67irV6+eOnXqgw8+kGqUpEtKSuLj4w8fPjwzgLq6ut7e3oaGhoqKitTU1OrqamPLTOcAF7788ssAI1i0+Q9/+EOLIhN2yQpEzoz8Jz/5yaFDh2Z+lU9lPT09M7/KwpYtW06cOCEL8m9RUdG8Le79WUYAAW8EXKej8i7nvoEUoXwwq6ysfPLkSXl5+b1799LS0hx9pDhlWTobW9wjBLL81VdfBbI52yJgIwFXEc5JWu7PunXrVllZ2fbt248ePbpx48Y5HRb9tba2tqCgoLGxsaWlZdHOczr87W9/m9PCrwiEq8CCRXjw4MEHDx7IsM+ePStXaGShr68vPT29u7tb3iRled4Wdya5luP+q0/Ld+/e9ak/nRGwr8CCRbh///7jx49LBa5fv/727dsywtbW1uLiYrkwI//K8rwtZkHM+ThqVljiIKBQwHV11JHc4OCg4+pobm5uU1PTq1evxsbG9u7dK58JjV9IGFvMGuHKlSvNCmVuHO6YMdeTaCIwtwiVoCxfvlxJJnPS4N7ROSD8GrjAgqejgYcOJIK8AweyOdsiYCMBpe+E/FGvjV5DpIoAAggggAACCCCAAAIIIIAAAggggAACCCCAAAIIIIAAAggggAACCCCAAAIIIIAAAggggAACCCCAAAIIIIAAAggggAACCCCAAAIIIIAAAggggAACCCCAAAIIIIAAAggggAACCCCAAAIIIIAAAggggAACCCCAAAIIIIAAAggggAACCCCAAAIIIIAAAggggAACCCCAAAIIIIAAAggggAACCCCAAAIIIIAAAggggAACCCCAAAIIIIAAAggggAACCCCAAAIIIIAAAggggAACCCCAAAIIIIAAAghYKxBhbXh/o0dEKE1senra3zGxHQLzC0TN3xzq1sjIyFCnMP/+Jycn519BKwL+CigtwmXLlvk7Imu3owit9V2S0ZUWYUxMjM7DMTExoTMxsrKvgNIijI2N1Wk6PDysMzGysq+AqwiTk5MPHDiwbdu2pKQkGY/829zcnJKSMjAwUFpaOjQ05E2LWRCJiYlmhTI3Tn9/v7kBiYaA6yLk1atXT5069cEHH0g1iktdXV1vb29DQ0NFRUVqamp1dbU3LWaBfve73zUrlLlx/vrXv5obkGgIuIowPT1dqm5wcNBRhB0dHYWFhd3d3ZmZmW1tbfn5+d60mAX6ox/9yKxQ5sa5fPmyuQGJhoDrmwCpQHeOtLQ0R0tPT48syypvWtwjsIwAAt4IuD4TetPbpz61tbUFBQWNjY0tLS0+bSidnzx54usm9EfApgILFmFfX5+coMrpaEZGhizL8LxpcVeoqalx/9WnZbkO5FN/OiNgX4EFi7C1tbW4uFguzMi/siwj9KbFLIjnz5+bFYo4CCgXcF2YcSQ6c2HGmy8kjH3MGm1CQoJZocyNw/eE5noSTQTmFqESFO6YUXIgSCMIAguejgZh3x52wS2aHnBYFWYCSt8J+VOmMHudMRwEEEAAAQQQQAABBBBAAAEEEEAAAQQQQAABBBBAAAEEEEAAAQQQQAABBBBAAAEEEEAAAQQQQAABBBBAAAEEEEAAAQQQQAABBBBAAAEEEEAAAQQQQAABBBBAAAEEEEAAAQQQQAABBBBAAAEEEEAAAQQQQAABBBBAAAEEEEAAAQQQQAABBBBAAAEEEEAAAQQQQAABBBBAAAEEEEAAAQQQQAABBBBAAAEEEEAAAQQQQAABBBBAwDyBCPNCmRkpMjLSzHDmxZqamjIvGJEQ+H+BKJ0MFKHO40JWVggoLcLo6GgrRht4zFevXgUehAgIuAsoLcLly5e7Z6ln+cWLF3qSIZPwEFBahHFxcTp9BwcHdSZGVvYVUFqEK1as0Gna3d2tMzGysq+AswhPnDiRnp4uw4iJicnLy0tJSUlKSmpubpaFgYGB0tLSoaEhY4t1w05OTrYuOJERUCUw9yuKPXv2ZGVl7du3r66urre3t6GhoaKiIjU1tbq62thi3UiKioqsCx5I5IsXLwayOdsiYBSYVYQRERE3b97cvHnzo0ePOjo6CgsL5ewrMzOzra0tPz/f2GIMZ1YLRWiWJHH0C8z6TLh169YbN25IBUreaWlp8k4oCz09PbI8b4t1w3v8+LF1wYmMgCqBWUUoZ55lZWWB5FdbW1tQUNDY2NjS0hJIHC5CBqLHtvYScBXh+vXr5aXf2dnpGEBfX59cqpHT0YyMDFmWRmOLcag1NTXGRj9anj175sdWbIKAHQVcRVhVVVVfXz8zhtbW1uLiYrkwI//KsrQbW2Y6m74wOjpqekwCIqBTwHlhJicn59ixY9///vdnsjR+IWFsmels+kJsbKzpMU0JyB0zpjASxF1g1tVR9xWhXY6Kcr1FhzaTOXvn3tE5IPwauIDS1zp/MRT4oSUCAggggAACCCCAAAIIIIAAAggggAACCCCAAAIIIIAAAggggAACCCCAAAIIIIAAAggggAACCCCAAAIIIIAAAggggAACCCCAAAIIIIAAAggggAACCCCAAAIIIIAAAggggAACCCCAAAIIIIAAAggggAACCCCAAAIIIIAAAggggAACCCCAAAIIIIDAEhJQOkloRITSxKanp5fQq4OhBkVA6SShy5YtC8rwfd4JM/X6TMYGiwkoLcKYmJjFMg/NeoowNO5hvVelRRgfH6+TfXR0VGdiZGVfAaVFmJycrNO0v79fZ2JkZV8BpUWYkZGh0/Rf//qXzsTIyr4CSoswOztbp+mVK1d0JkZW9hVwFuGmTZuOHDnS1dUlI2lvb6+pqUlKSmpubk5JSRkYGCgtLR0aGjK2WDfs/Px864ITGQFVAs6v40pKSuRayOHDh2eSq6ur6+3tbWhoqKioSE1Nra6uNrbMdDZ9Qf4jMD2mKQELCgpMiUMQBOYKVFVV/fjHP3Zv7ejoWL16tbRkZmbevn1bFowt7v3NXZbvxHX+mDtMoiEgApEOBbkQUlRUJO8/586dc3weS0tLk3dCWdvT0yPLsmBscWzLvwggEIiA8zOhvO3cunWrrKxs+/btR48e3bhxo39Ba2tr5YStsbGxpaXFvwiOrf785z8HsjnbImAjAWcRHjx48MGDB5L32bNn5QqNLPT19aWnp3d3d8ubpCzP22Icp1zRMTb60fLVV1/5sRWbIGBHAWcR7t+///jx41KB69evd3wCbG1tLS4ulgsz8q8sy9iMLdYN2JGDdfGJjIAeAefV0dzc3KamJrkxcmxsbO/evffu3TN+IWFssW4YP/jBD6wLHkhkvicMRI9t5xVQ+hdD8p/CvOmGvJE7ZkJ+CMIvAefpqLaBDQ4OakuJfBCwSEBpEY6MjFg0YMIioE1AaRFOTExokyIfBCwSUPqZkMdbWHS8CYsAAggggAACCCCAAAIIIIAAAggggAACCCCAAAIIIIAAAggggAACCCCAAAIIhFiA29Z8OwDyHBDfNqA3AosJKL2Bm1mZFjtwrA8fAaVFGBWlNDFmZQqf176akSh9raudGk0e/6Hm2JFImAgoLcLY2FidwM+ePdOZGFnZV0BpESYmJuo0dTz9UWduZGVTgblFuGXLls8//1zeiIzPVjO2WDdm2Zd1wYmMgCqBWV9RyPvPF1988cYbb6xYscI4/YuxxbqRFBYWWhc8kMiXLl0KZHO2RcAoMKsIf//733/55Zcff/yxTJQr079IJcgTuGVCmLa2NpmrzNhiDGdWy09/+lOzQpkb549//KO5AYmGgHNCGIGQOSRkGqZTp045UIzTvxhb4EMAgcAFnJ8Jly9f/uGHH8psMAFGNGtCGKaGD/BAsLmNBJxF+N5778kHws8++0xST0hI+PTTT0M7Iczjx49thEiqCAQiMOszoSOQPP1aPhO6X4ZZtWrVL37xC2NLIDv2vG1OTo7nDqFae/fu3VDtmv2Gq8CCRWj8QsLYYh2KY1pS6+L7HZnvCf2mY8OFBOYpwoW6BrNdviMJ5u683xd3zHhvRU8vBeZ+We/lZlZ34zH4VgsTX4+A0iLkjxX0vETIxGoBpaejzEVh9YEnPgIIIIAAAggggAACCCCAAAIIIIAAAggggAACCCCAAAIIIIAAAggggAACCCCAAAIIIIAAAggggAACCCCAAALBFVD6p0yRka5nMQYXZJG9TU1NLdKD1Qj4KKD0j3rVzk9IEfr4AqP74gJKizA6Onrx3EPR4+XLl6HYLfsMZwGlRSgPI9apPjo6qjMxsrKvgNIijI+P12n69OlTnYmRlX0FlBah2kce2vdIk7laAWcRxsXFNTU1ySN35WH4+/btu3jxovFRv8YW60a1cuVK64ITGQFVAs6vKKqqquS6n8wJk5GRce3atddff939ofepqanV1dXGFutG8s4771gXPJDI586dC2RztkXAKOAsQnnnGRkZkUfuvvXWWx999FFubq5xNkJjizGcWS179uwxK5S5cf7whz+YG5BoCDhPRx3XG44dOyazoznehYyzERpb4EMAgcAFZl2YKSkpOXPmTGlp6eXLl/0Lbdb8hI8ePfIvAbZCwHYCziI8dOhQeXm5PHz+woULn3zyiQyD+QltdyxJ2KYCziKUK5/vvvuuvA2++eabnZ2dMpjW1tbi4uKGhgb5V5bnbbFuzHwdZ50tkbUJOC/MZGVlNTc3y23Tcm3m/fffv3PnjvELCWOLdYNZs2aNdcEDifzw4cNANmdbBIwCziI0rghti9rvCXmLDu0LIyz3PuvCjJ4Rjo+P60mGTBCwVEBpEfLHCpYedYKrElBahJOTk6qYSAYBBBBAAAEEEEAAAQQQQAABBBBAAAEEEEAAAQQQQAABBBBAAAEEEEAAAQQQQAABBBBAAAEEEEAAAQQQQAABBBBAAAEEEEAAAQQQQMC2Akof9BQRoTSx6elp2x5rElcqoPTxFmqny5bnIys9kqRlWwGlRRgVpTQxitC2L3W9iSt9raudLntsbEzvwSQzewooLcLY2FidnkNDQzoTIyv7CigtwoSEBPuakjkCPgkoLcLk5GSfhkFnBOwr4CrCXbt27dy5U+asl6mzL126ZJz+xdhi3bC/+c1vWhecyAioEnB+HSez0su8aBs2bMjLy2tpaVm3bp1xhnpji3UjkelKrQseSGSZzDiQzdkWAaOAswjXrl2bn59/+vTp+Pj4+/fvr1q1yjhDvbHFGM6slsrKSrNCmRvnd7/7nbkBiYaA83RUJiSUH+HYsWPH+fPnZcE4Q72xBT4EEAhcwPWZUGJlZ2fLW5CclPod16w563t7e/3OgQ0RsJeAqwjlW4GTJ0/u3r27v79fxhDaOesfPXpkL0fJVuY51pnz1NSUzsTIyiHgLEK5YVqmy66vr79+/bpjRWjnrH/8+LHtjpDa+10pQuWvJeeFGfly4tChQ19//bWkOzw8/Pbbbxu/kDC2WDe2nJwc64IHEvnu3bsLbR4XF7fQqtC2j46OhjYB9u5ZwFmEnjsFf216enrwd+rNHj18WF25cqU3EYLf5+nTp8HfKXv0XsD1mdD7bYLQ88WLF0HYi7m7UPtOSBGae6BNj6a0CMfHx00fqtUB5WYjq3dB/LAUUFqEdvyzPe53DcsKCcKglBbh5ORkEAZv7i6439Vcz6UTTWkR2vFRLnL/7dJ53TBSEwWUFqGJIwxaKLmtL2j7YkfhJEARhtPR9HksdrzLx445ez4wFKFnHx/Wyo1+PvTW0dWOd/nYMWfPR5si9Ozjw1rHPbc+bKCgq9qn2r18+XIhHjvmvNBYHO0UoWcfH9ba8X7XmJgYH0YYxK4e7tawY86e5ShCzz4+rB0cHPSht46udnyqnR1z9ny0KULPPj6sff78uQ+9dXSVBynoSMSHLOyYs+fhUYSefXxYa8c/VlixYoUPI9TR1Y45e5ajCD37+LDWjve72vFWOzvm7PllRBF69vFhrYcLej5ECW7Xb3zjG8HdoQl7s2POnodNEXr28WEt97v6gBVA1/C7R5ciDODlMHtTniIx24PfvBWgCL2VCst+dvxu0445e37xUISefcJ87ZMnT2w3Qjvm7BmZIvTsE+Zr7XiDgR1z9vwyogg9+4T52mfPntluhHbM2TMyRejZJ8zXjoyM2G6EdszZM7LSIoyOjvacd6jW2vHLQA9WHu6T9rBVaFfZMWfPYkqLUO2c9WFWhBMTE55fHwrX2jFnz4xKi1DtnfLyeHLPoPZaa8en2tkxZ8+vCqVFqPYZnnb8y10PrwA73uVjx5w9HAJZpbQIw+8mXc+HIVRr7XiXjx1z9nx8lRZh+N0f6PkwsHYpCygtQh4fuJRflEtt7M4iPHHihGMiJHmAR15eXkpKyqZNm44cOdLV1SUi7e3tNTU1waTJyMgI5u7YFwIhFHAWYXFxsSOJPXv2ZGVlybLUZF1d3eHDh0OS3Jo1a0KyX3aKQPAFZp2Oyny9e/fu3bx5s+Qh70WdnZ3BT8ixR8d/BKHaO/tFIJgCs6ZZ37p1640bNxzzxUsRFhUVyYnouXPnsrOzg5kT+0JgSQnMeiesqKgoKytzjF+mZLl165b8un379qNHj27cuHEhl9ra2oKCgsbGxpaWloX6+Nru+Czq61b0R8COAq4iXL9+vfyRyMwp6MGDBx88eCBDOnv2rFyh8TA2K67ZPHz40MMeWYVAOAm4irCqqqq+vn5mbPv37z9+/LhUoBTn7du3Z9qDs9DT0xOcHbEXBEIu4CzCnJyc1atXX7lyZSahX//6101NTeXl5WNjYz/72c9m2oOzYMfJVYIjw17CTyBC55C+973v6Uzs5s2bOhMjK/sKuE5HVY0h/B5hoIqXZFQJKC1CO87roOq4koyNBJQWYfj99bSNXhOkGmQBpUVox3kdgnzk2B0CCCCAAAIIIIAAAggggAACCCCAAAIIIIAAAggggAACCCCAAAIIIIAAAggggAACCCCAAAIIIIAAAggggAACCCCAAAIIIIAAAggggAACCCCAAAIIIIAAAggggAACCCCAAAIIIIAAAggggAACCCCAAAIIIIAAAggggAACCCCAAAIIIIAAAqoE/g+X22DgxjUQ9gAAAABJRU5ErkJggg==",
            "text/plain": [
              "<PIL.Image.Image image mode=RGB size=300x600>"
            ]
          },
          "execution_count": 17,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "from pydna.gel import gel\n",
        "from pydna.ladders import GeneRuler_1kb_plus\n",
        "\n",
        "bands = list_of_amplicons\n",
        "for band in bands:\n",
        "    print(band.name)\n",
        "\n",
        "gel([GeneRuler_1kb_plus, *[[band] for band in bands]])"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 18,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>Amplicon Name</th>\n",
              "      <th>5' Overhang</th>\n",
              "      <th>5' Duplicate</th>\n",
              "      <th>3' Overhang</th>\n",
              "      <th>3' Duplicate</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>Amplicon_SCO5090_140_pcr_products</td>\n",
              "      <td>CATG</td>\n",
              "      <td>No</td>\n",
              "      <td>GGGA</td>\n",
              "      <td>No</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>Amplicon_SCO5090_208_pcr_products</td>\n",
              "      <td>GGGA</td>\n",
              "      <td>No</td>\n",
              "      <td>GTCG</td>\n",
              "      <td>No</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>Amplicon_SCO5087_522_pcr_products</td>\n",
              "      <td>GTCG</td>\n",
              "      <td>No</td>\n",
              "      <td>CTAG</td>\n",
              "      <td>No</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "                       Amplicon Name 5' Overhang 5' Duplicate 3' Overhang  \\\n",
              "0  Amplicon_SCO5090_140_pcr_products        CATG           No        GGGA   \n",
              "1  Amplicon_SCO5090_208_pcr_products        GGGA           No        GTCG   \n",
              "2  Amplicon_SCO5087_522_pcr_products        GTCG           No        CTAG   \n",
              "\n",
              "  3' Duplicate  \n",
              "0           No  \n",
              "1           No  \n",
              "2           No  "
            ]
          },
          "execution_count": 18,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "overhangs = create_overhang_dataframe(list_of_amplicons)\n",
        "overhangs"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 19,
      "metadata": {},
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "Dseqrecord(-138)\n",
            "CATGGTTCACTGCCGTATAGGCAGCTAAGAAACACGTACAGGTCCTGGAGCG\u001b[48;5;11mGTTTTAGAGCTAGAAATAGCAAGT\u001b[0mTAAAATAAGGCTAGTCCGTTATCAACTTGAAAAAGTGGCACCGAGTCGGTGCTTTTTT    \n",
            "    CAAGTGACGGCATATCCGTCGATTCTTTGTGCATGTCCAGGACCTCGCCAAAATCTCGATCTTTATCGTTCAATTTTATTCCGATCAGGCAATAGTTGAACTTTTTCACCGTGGCTCAGCCACGAAAAAACCCT\n",
            "Dseqrecord(-106)\n",
            "GGGACGTCCAGGTCTTCACC\u001b[48;5;11mGTTTTAGAGCTAGAAATAGCAAGT\u001b[0mTAAAATAAGGCTAGTCCGTTATCAACTTGAAAAAGTGGCACCGAGTCGGTGCTTTTTT    \n",
            "    GCAGGTCCAGAAGTGGCAAAATCTCGATCTTTATCGTTCAATTTTATTCCGATCAGGCAATAGTTGAACTTTTTCACCGTGGCTCAGCCACGAAAAAACAGC\n",
            "Dseqrecord(-106)\n",
            "GTCGACCTCCCAGTCACGGC\u001b[48;5;11mGTTTTAGAGCTAGAAATAGCAAGT\u001b[0mTAAAATAAGGCTAGTCCGTTATCAACTTGAAAAAGTGGCACCGAGTCGGTGCTTTTTT    \n",
            "    TGGAGGGTCAGTGCCGCAAAATCTCGATCTTTATCGTTCAATTTTATTCCGATCAGGCAATAGTTGAACTTTTTCACCGTGGCTCAGCCACGAAAAAAGATC\n"
          ]
        }
      ],
      "source": [
        "digest_amplicons = digest_amplicons_w_BsaI(list_of_amplicons)\n",
        "for digest in digest_amplicons: \n",
        "    print(digest.figure())"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 20,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/plain": [
              "[Dseqrecord(-138), Dseqrecord(-106), Dseqrecord(-106)]"
            ]
          },
          "execution_count": 20,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "digest_amplicons"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 21,
      "metadata": {},
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "Successfully looped the plasmid.\n"
          ]
        }
      ],
      "source": [
        "\n",
        "from Bio.Restriction import NcoI, NheI\n",
        "\n",
        "try:\n",
        "    linear_plasmid, smaller_frag = sorted(clean_plasmid.cut(NcoI, NheI), key=lambda x: len(x), reverse=True)\n",
        "    for i in range(len(digest_amplicons)):\n",
        "        # Assuming digest_amplicons[i] is a Dseqrecord, concatenate Dseq sequences\n",
        "        linear_plasmid.seq += digest_amplicons[i].seq\n",
        "    \n",
        "    try:\n",
        "        rec_vec = linear_plasmid.looped()\n",
        "        print(\"Successfully looped the plasmid.\")\n",
        "    except Exception as e:\n",
        "        print(f\"An error occurred while trying to loop the plasmid: {e}\")\n",
        "        # Handle the error or provide an alternative action here\n",
        "        rec_vec = None  # Example response; adjust as needed\n",
        "\n",
        "except Exception as e:\n",
        "    print(f\"An error occurred during plasmid digestion or preparation: {e}\")\n",
        "    # Additional error handling or cleanup here\n",
        "    rec_vec = None  # Ensure rec_vec is defined even if an error occurs"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 22,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/plain": [
              "'pCRISPR-mcBEST_3_sgrnas'"
            ]
          },
          "execution_count": 22,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "# Constructing a meaningful name, ID, and description for the assembled plasmid using user input\n",
        "plasmid_components = \"_\".join(genes_to_KO)  # Concatenates gene names from the user's list\n",
        "plasmid_purpose = f\"pCRISPR-mcBEST_{len(filtered_df)}_sgrnas\"  # Example: purpose of the plasmid\n",
        "\n",
        "rec_vec.name = f'{plasmid_purpose}'#_{plasmid_components}'\n",
        "rec_vec.id = rec_vec.name  # Using the same value for ID as for name for simplicity\n",
        "rec_vec.description = f'CRISPR-mcBEST_plasmid {\", \".join(genes_to_KO)} for gene KNOCK-DOWN, assembled using StreptoCAD.'\n",
        "rec_vec.id"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 23,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/plain": [
              "Dseqrecord(o12802)"
            ]
          },
          "execution_count": 23,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "# annotate the plasmid correctly\n",
        "annotate_plasmid_with_sgrnas(rec_vec, filtered_df)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 24,
      "metadata": {},
      "outputs": [],
      "source": [
        "print_plasmid = False\n",
        "\n",
        "if print_plasmid: \n",
        "    rec_vec.write(f\"../../data/plasmids/sgRNA_plasmids_pCRISPR_mcBEST/{rec_vec.id}.gb\")\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 25,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/plain": [
              "Dseqrecord(o12802)"
            ]
          },
          "execution_count": 25,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "rec_vec"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 26,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/plain": [
              "'sgRNA_SCO5090(140)_sgRNA_SCO5090(208)_sgRNA_SCO5087(522)_'"
            ]
          },
          "execution_count": 26,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "integration_names = filtered_df.apply(lambda row: f\"sgRNA_{row['locus_tag']}({row['sgrna_loc']})_\", axis=1).tolist()\n",
        "integration_names = ''.join(integration_names)\n",
        "integration_names"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 27,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>plasmid_name</th>\n",
              "      <th>date</th>\n",
              "      <th>original_plasmid</th>\n",
              "      <th>integration</th>\n",
              "      <th>size</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>pCRISPR-mcBEST_3_sgrnas</td>\n",
              "      <td>2024-10-25</td>\n",
              "      <td>pCRISPR-mcBEST</td>\n",
              "      <td>sgRNA_SCO5090(140);sgRNA_SCO5090(208);sgRNA_SC...</td>\n",
              "      <td>12802</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "              plasmid_name        date original_plasmid  \\\n",
              "0  pCRISPR-mcBEST_3_sgrnas  2024-10-25   pCRISPR-mcBEST   \n",
              "\n",
              "                                         integration   size  \n",
              "0  sgRNA_SCO5090(140);sgRNA_SCO5090(208);sgRNA_SC...  12802  "
            ]
          },
          "execution_count": 27,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "integration_names = filtered_df.apply(lambda row: f\"sgRNA_{row['locus_tag']}({row['sgrna_loc']})\", axis=1).tolist()\n",
        "integration_names = [';'.join(integration_names)]\n",
        "\n",
        "plasmid_metadata_df = extract_metadata_to_dataframe([rec_vec],\n",
        "                                                    clean_plasmid,\n",
        "                                                    integration_names)\n",
        "\n",
        "plasmid_metadata_df"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 28,
      "metadata": {},
      "outputs": [],
      "source": [
        "pd.set_option('display.max_columns', None)\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 29,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>locus tag</th>\n",
              "      <th>f_primer_name</th>\n",
              "      <th>r_primer_name</th>\n",
              "      <th>f_primer_sequences(5-3)</th>\n",
              "      <th>r_primer_sequences(5-3)</th>\n",
              "      <th>f_tm</th>\n",
              "      <th>r_tm</th>\n",
              "      <th>ta</th>\n",
              "      <th>flanking_region</th>\n",
              "      <th>annealing_temperature</th>\n",
              "      <th>primer_pair</th>\n",
              "      <th>homodimer_forward_tm</th>\n",
              "      <th>homodimer_forward_deltaG (kcal/mol)</th>\n",
              "      <th>homodimer_reverse_tm</th>\n",
              "      <th>homodimer_reverse_deltaG (kcal/mol)</th>\n",
              "      <th>heterodimer_tm</th>\n",
              "      <th>heterodimer_deltaG (kcal/mol)</th>\n",
              "      <th>hairpin_forward_structure_found</th>\n",
              "      <th>hairpin_forward_tm</th>\n",
              "      <th>hairpin_forward_deltaG (kcal/mol)</th>\n",
              "      <th>hairpin_reverse_structure_found</th>\n",
              "      <th>hairpin_reverse_tm</th>\n",
              "      <th>hairpin_reverse_deltaG (kcal/mol)</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>SCO5087</td>\n",
              "      <td>SCO5087_fwd_checking_primer</td>\n",
              "      <td>SCO5087_rev_checking_primer</td>\n",
              "      <td>TAGGGCGCGGC</td>\n",
              "      <td>CCGGAGACCACCAG</td>\n",
              "      <td>51</td>\n",
              "      <td>53</td>\n",
              "      <td>55</td>\n",
              "      <td>550</td>\n",
              "      <td>55</td>\n",
              "      <td>SCO5087_fwd_checking_primer &amp; SCO5087_rev_chec...</td>\n",
              "      <td>26.385240</td>\n",
              "      <td>-5.886397</td>\n",
              "      <td>-15.268430</td>\n",
              "      <td>-3.528618</td>\n",
              "      <td>-34.382170</td>\n",
              "      <td>-1.534614</td>\n",
              "      <td>False</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>False</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>SCO5089</td>\n",
              "      <td>SCO5089_fwd_checking_primer</td>\n",
              "      <td>SCO5089_rev_checking_primer</td>\n",
              "      <td>CCCTTCGACGAGCG</td>\n",
              "      <td>CGGGGCCGGT</td>\n",
              "      <td>54</td>\n",
              "      <td>49</td>\n",
              "      <td>53</td>\n",
              "      <td>600</td>\n",
              "      <td>53</td>\n",
              "      <td>SCO5089_fwd_checking_primer &amp; SCO5089_rev_chec...</td>\n",
              "      <td>-5.338917</td>\n",
              "      <td>-1.468854</td>\n",
              "      <td>1.902964</td>\n",
              "      <td>-3.360252</td>\n",
              "      <td>-36.526236</td>\n",
              "      <td>-1.541041</td>\n",
              "      <td>False</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>False</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>SCO5090</td>\n",
              "      <td>SCO5090_fwd_checking_primer</td>\n",
              "      <td>SCO5090_rev_checking_primer</td>\n",
              "      <td>CGAGCTGGACGCC</td>\n",
              "      <td>CGACGACCACCTCG</td>\n",
              "      <td>54</td>\n",
              "      <td>54</td>\n",
              "      <td>58</td>\n",
              "      <td>600</td>\n",
              "      <td>58</td>\n",
              "      <td>SCO5090_fwd_checking_primer &amp; SCO5090_rev_chec...</td>\n",
              "      <td>-27.967369</td>\n",
              "      <td>-2.788697</td>\n",
              "      <td>-29.701198</td>\n",
              "      <td>-2.176432</td>\n",
              "      <td>10.025082</td>\n",
              "      <td>-1.435609</td>\n",
              "      <td>False</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>False</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "  locus tag                f_primer_name                r_primer_name  \\\n",
              "0   SCO5087  SCO5087_fwd_checking_primer  SCO5087_rev_checking_primer   \n",
              "1   SCO5089  SCO5089_fwd_checking_primer  SCO5089_rev_checking_primer   \n",
              "2   SCO5090  SCO5090_fwd_checking_primer  SCO5090_rev_checking_primer   \n",
              "\n",
              "  f_primer_sequences(5-3) r_primer_sequences(5-3) f_tm r_tm  ta  \\\n",
              "0             TAGGGCGCGGC          CCGGAGACCACCAG   51   53  55   \n",
              "1          CCCTTCGACGAGCG              CGGGGCCGGT   54   49  53   \n",
              "2           CGAGCTGGACGCC          CGACGACCACCTCG   54   54  58   \n",
              "\n",
              "  flanking_region  annealing_temperature  \\\n",
              "0             550                     55   \n",
              "1             600                     53   \n",
              "2             600                     58   \n",
              "\n",
              "                                         primer_pair  homodimer_forward_tm  \\\n",
              "0  SCO5087_fwd_checking_primer & SCO5087_rev_chec...             26.385240   \n",
              "1  SCO5089_fwd_checking_primer & SCO5089_rev_chec...             -5.338917   \n",
              "2  SCO5090_fwd_checking_primer & SCO5090_rev_chec...            -27.967369   \n",
              "\n",
              "   homodimer_forward_deltaG (kcal/mol)  homodimer_reverse_tm  \\\n",
              "0                            -5.886397            -15.268430   \n",
              "1                            -1.468854              1.902964   \n",
              "2                            -2.788697            -29.701198   \n",
              "\n",
              "   homodimer_reverse_deltaG (kcal/mol)  heterodimer_tm  \\\n",
              "0                            -3.528618      -34.382170   \n",
              "1                            -3.360252      -36.526236   \n",
              "2                            -2.176432       10.025082   \n",
              "\n",
              "   heterodimer_deltaG (kcal/mol)  hairpin_forward_structure_found  \\\n",
              "0                      -1.534614                            False   \n",
              "1                      -1.541041                            False   \n",
              "2                      -1.435609                            False   \n",
              "\n",
              "   hairpin_forward_tm  hairpin_forward_deltaG (kcal/mol)  \\\n",
              "0                 0.0                                0.0   \n",
              "1                 0.0                                0.0   \n",
              "2                 0.0                                0.0   \n",
              "\n",
              "   hairpin_reverse_structure_found  hairpin_reverse_tm  \\\n",
              "0                            False                 0.0   \n",
              "1                            False                 0.0   \n",
              "2                            False                 0.0   \n",
              "\n",
              "   hairpin_reverse_deltaG (kcal/mol)  \n",
              "0                                0.0  \n",
              "1                                0.0  \n",
              "2                                0.0  "
            ]
          },
          "execution_count": 29,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "\n",
        "# Getting checking primers\n",
        "checking_primers_df = find_best_check_primers_from_genome(genome, \n",
        "                                       genes_to_KO, \n",
        "                                       flanking_region=flanking_region,\n",
        "                                       target_tm = melting_temperature, \n",
        "                                        primer_concentration = primer_concentration, \n",
        "                                        polymerase = chosen_polymerase)\n",
        "checking_primers_df"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 30,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>Name</th>\n",
              "      <th>Sequence</th>\n",
              "      <th>Concentration</th>\n",
              "      <th>Purification</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>SCO5087_fwd_checking_primer</td>\n",
              "      <td>TAGGGCGCGGC</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>SCO5089_fwd_checking_primer</td>\n",
              "      <td>CCCTTCGACGAGCG</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>SCO5090_fwd_checking_primer</td>\n",
              "      <td>CGAGCTGGACGCC</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>SCO5087_rev_checking_primer</td>\n",
              "      <td>CCGGAGACCACCAG</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>SCO5089_rev_checking_primer</td>\n",
              "      <td>CGGGGCCGGT</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>5</th>\n",
              "      <td>SCO5090_rev_checking_primer</td>\n",
              "      <td>CGACGACCACCTCG</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "                          Name        Sequence Concentration Purification\n",
              "0  SCO5087_fwd_checking_primer     TAGGGCGCGGC          25nm          STD\n",
              "1  SCO5089_fwd_checking_primer  CCCTTCGACGAGCG          25nm          STD\n",
              "2  SCO5090_fwd_checking_primer   CGAGCTGGACGCC          25nm          STD\n",
              "3  SCO5087_rev_checking_primer  CCGGAGACCACCAG          25nm          STD\n",
              "4  SCO5089_rev_checking_primer      CGGGGCCGGT          25nm          STD\n",
              "5  SCO5090_rev_checking_primer  CGACGACCACCTCG          25nm          STD"
            ]
          },
          "execution_count": 30,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "# making the primers into dseqrecords\n",
        "checking_primers_df_idt = create_idt_order_dataframe(checking_primers_df)\n",
        "checking_primers_df_idt\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 31,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>Name</th>\n",
              "      <th>Sequence</th>\n",
              "      <th>Concentration</th>\n",
              "      <th>Purification</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>fwd_primer_SCO5090_140</td>\n",
              "      <td>GATCGGGTCTCCCATGGTTCACTGCCGTATAGGCAGCTAAGAAACA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>fwd_primer_SCO5090_208</td>\n",
              "      <td>GATCGGGTCTCCGGGACGTCCAGGTCTTCACCGTTTTAGAGCTAGA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>fwd_primer_SCO5087_522</td>\n",
              "      <td>GATCGGGTCTCCGTCGACCTCCCAGTCACGGCGTTTTAGAGCTAGA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>rev_primer_SCO5090_140</td>\n",
              "      <td>GATCAGGTCTCGTCCCAAAAAAGCACCGACTCG</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>rev_primer_SCO5090_208</td>\n",
              "      <td>GATCAGGTCTCGCGACAAAAAAGCACCGACTCG</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>5</th>\n",
              "      <td>rev_primer_SCO5087_522</td>\n",
              "      <td>GATCAGGTCTCGCTAGAAAAAAGCACCGACTCG</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>SCO5087_fwd_checking_primer</td>\n",
              "      <td>TAGGGCGCGGC</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>SCO5089_fwd_checking_primer</td>\n",
              "      <td>CCCTTCGACGAGCG</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>SCO5090_fwd_checking_primer</td>\n",
              "      <td>CGAGCTGGACGCC</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>SCO5087_rev_checking_primer</td>\n",
              "      <td>CCGGAGACCACCAG</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>SCO5089_rev_checking_primer</td>\n",
              "      <td>CGGGGCCGGT</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>5</th>\n",
              "      <td>SCO5090_rev_checking_primer</td>\n",
              "      <td>CGACGACCACCTCG</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "                          Name  \\\n",
              "0       fwd_primer_SCO5090_140   \n",
              "1       fwd_primer_SCO5090_208   \n",
              "2       fwd_primer_SCO5087_522   \n",
              "3       rev_primer_SCO5090_140   \n",
              "4       rev_primer_SCO5090_208   \n",
              "5       rev_primer_SCO5087_522   \n",
              "0  SCO5087_fwd_checking_primer   \n",
              "1  SCO5089_fwd_checking_primer   \n",
              "2  SCO5090_fwd_checking_primer   \n",
              "3  SCO5087_rev_checking_primer   \n",
              "4  SCO5089_rev_checking_primer   \n",
              "5  SCO5090_rev_checking_primer   \n",
              "\n",
              "                                            Sequence Concentration  \\\n",
              "0  GATCGGGTCTCCCATGGTTCACTGCCGTATAGGCAGCTAAGAAACA...          25nm   \n",
              "1  GATCGGGTCTCCGGGACGTCCAGGTCTTCACCGTTTTAGAGCTAGA...          25nm   \n",
              "2  GATCGGGTCTCCGTCGACCTCCCAGTCACGGCGTTTTAGAGCTAGA...          25nm   \n",
              "3                  GATCAGGTCTCGTCCCAAAAAAGCACCGACTCG          25nm   \n",
              "4                  GATCAGGTCTCGCGACAAAAAAGCACCGACTCG          25nm   \n",
              "5                  GATCAGGTCTCGCTAGAAAAAAGCACCGACTCG          25nm   \n",
              "0                                        TAGGGCGCGGC          25nm   \n",
              "1                                     CCCTTCGACGAGCG          25nm   \n",
              "2                                      CGAGCTGGACGCC          25nm   \n",
              "3                                     CCGGAGACCACCAG          25nm   \n",
              "4                                         CGGGGCCGGT          25nm   \n",
              "5                                     CGACGACCACCTCG          25nm   \n",
              "\n",
              "  Purification  \n",
              "0          STD  \n",
              "1          STD  \n",
              "2          STD  \n",
              "3          STD  \n",
              "4          STD  \n",
              "5          STD  \n",
              "0          STD  \n",
              "1          STD  \n",
              "2          STD  \n",
              "3          STD  \n",
              "4          STD  \n",
              "5          STD  "
            ]
          },
          "execution_count": 31,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "full_idt = pd.concat([idt_df, checking_primers_df_idt])\n",
        "full_idt"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "## Folder with all the generated I/O"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 32,
      "metadata": {},
      "outputs": [],
      "source": [
        "get_data_folder = False"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 33,
      "metadata": {},
      "outputs": [],
      "source": [
        "if get_data_folder: \n",
        "    # Example: \n",
        "\n",
        "    input_files = [\n",
        "        {\"name\": \"input_genome.gb\", \"content\": genome},\n",
        "        {\"name\": \"input_plasmid.gb\", \"content\": clean_plasmid}\n",
        "    ]\n",
        "\n",
        "    output_files = [\n",
        "        {\"name\": \"mcBEST_w_sgRNAs.gb\", \"content\": rec_vec},\n",
        "        {\"name\": \"overhang_df.csv\", \"content\": overhangs},\n",
        "        {\"name\": \"pcr_df.csv\", \"content\": primer_df},\n",
        "        {\"name\": \"full_idt.csv\", \"content\": full_idt},\n",
        "        {\"name\": \"mutated_sgrna_df.csv\", \"content\": mutated_sgrna_df},\n",
        "        {\"name\": \"filtered_df.csv\", \"content\": filtered_df},\n",
        "        {\"name\": \"plasmid_metadata_df.csv\", \"content\": plasmid_metadata_df},\n",
        "    ]\n",
        "\n",
        "    input_values = {\n",
        "        \"genes_to_knockout\": ['SCO5087', 'SCO5089', 'SCO5090'],\n",
        "        \"filtering_metrics\": {\n",
        "            \"gc_upper\": gc_upper,\n",
        "            \"gc_lower\": gc_lower,\n",
        "            \"off_target_seed\": off_target_seed,\n",
        "            \"off_target_upper\": off_target_upper,\n",
        "            \"cas_type\": cas_type,\n",
        "            \"number_of_sgRNAs_per_group\": number_of_sgRNAs_per_group\n",
        "        },\n",
        "        \"polymerase_settings\": {\n",
        "            \"chosen_polymerase\": chosen_polymerase,\n",
        "            \"melting_temperature\": melting_temperature,\n",
        "            \"primer_concentration\": primer_concentration,\n",
        "            \"flanking_region\": flanking_region\n",
        "        }, # TODO might want to incorporate this to the app\n",
        "        \"overlapping_sequences\": {\n",
        "            \"restriction_overhang_f\": restriction_overhang_f,\n",
        "            \"restriction_overhang_r\": restriction_overhang_r,\n",
        "            \"backbone_overhang_f\": backbone_overhang_f,\n",
        "            \"backbone_overhang_r\": backbone_overhang_r,\n",
        "            \"cys4\": cys4,\n",
        "            \"sgRNA_handle_cys4_site\": str(sgRNA_handle_cys4_site[0].seq)\n",
        "        }\n",
        "    }\n",
        "\n",
        "\n",
        "    # Paths to Markdown files\n",
        "    markdown_file_paths = [\n",
        "        \"../../protocols/conjugation_protcol.md\",\n",
        "        \"../../protocols/multi_target_crispr_plasmid_protcol.md\"\n",
        "\n",
        "    ]\n",
        "\n",
        "    # Data and time\n",
        "    timestamp = datetime.utcnow().isoformat()\n",
        "    project_name=f\"CRISPR_mcBEST_workflow_{timestamp}\"\n",
        "\n",
        "\n",
        "    # Create project directory structure\n",
        "    project_directory = ProjectDirectory(\n",
        "        project_name=project_name,\n",
        "        input_files=input_files,\n",
        "        output_files=output_files,\n",
        "        input_values=input_values,\n",
        "        markdown_file_paths=markdown_file_paths\n",
        "    )\n",
        "\n",
        "\n",
        "    # DO You want to save the folder? \n",
        "    save_zip_folder = False \n",
        "\n",
        "    if save_zip_folder: \n",
        "        # Generate the project directory structure and get the zip content\n",
        "        zip_content = project_directory.create_directory_structure(create_directories=False)\n",
        "\n",
        "        # Save the zip file to disk (optional)\n",
        "        with open(\"project_structure.zip\", \"wb\") as f:\n",
        "            f.write(zip_content)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {},
      "outputs": [],
      "source": []
    }
  ],
  "metadata": {
    "kernelspec": {
      "display_name": "constrain",
      "language": "python",
      "name": "python3"
    },
    "language_info": {
      "codemirror_mode": {
        "name": "ipython",
        "version": 3
      },
      "file_extension": ".py",
      "mimetype": "text/x-python",
      "name": "python",
      "nbconvert_exporter": "python",
      "pygments_lexer": "ipython3",
      "version": "3.11.9"
    }
  },
  "nbformat": 4,
  "nbformat_minor": 2
}
