{
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "## W3-CRISPR-mcBEST workflow\n",
        "\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "The workflow allows for the simultaneous targeted engineering of up to 28 individual situated bases in the genome of Streptomyces, as shown by Whitford et. al 2023, thereby enabling efficient single-step multi-target engineering13. Similar to workflow 2, this workflow employs the CRISPR-BEST methodology but instead of performing base-editing on a single gene the method incorporates multiple guide RNAs as an array into the same vector (Figure 4A). This significantly enhances the efficiency of high-throughput genetic studies. The pCRISPR-cBEST plasmid file can be downloaded on the page along with the genome file for S.coelicor (A3) and streptoAIM will generate your plasmids, primers, checking primers and analyze the overhangs the target genes(Figure 4, S6). \n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 1,
      "metadata": {},
      "outputs": [],
      "source": [
        "import sys\n",
        "import os\n",
        "from pydna.dseqrecord import Dseqrecord\n",
        "from teemi.build.PCR import primer_tm_neb\n",
        "import pandas as pd\n",
        "from datetime import datetime\n",
        "\n",
        "\n",
        "# Ensure the src directory is in the Python path\n",
        "project_root = os.path.abspath(os.path.join(os.getcwd(), '../../'))\n",
        "if project_root not in sys.path:\n",
        "    sys.path.insert(0, project_root)\n",
        "\n",
        "from streptocad.sequence_loading.sequence_loading import (\n",
        "    load_and_process_gene_sequences,\n",
        "    load_and_process_plasmid, \n",
        "    check_and_convert_input,\n",
        "    annotate_dseqrecord,\n",
        "    process_specified_gene_sequences_from_record)\n",
        "\n",
        "\n",
        "from streptocad.utils import polymerase_dict, dataframe_to_seqrecords, ProjectDirectory, extract_metadata_to_dataframe\n",
        "from streptocad.crispr.guideRNAcas3_9_12 import extract_sgRNAs, SgRNAargs\n",
        "from streptocad.crispr.crispr_best import (\n",
        "    identify_base_editing_sites,\n",
        "    filter_sgrnas_for_base_editing,\n",
        "    process_base_editing\n",
        ")\n",
        "\n",
        "from streptocad.cloning.golden_gate_cloning import (\n",
        "    GoldenGateCloning,\n",
        "    create_overhang_dataframe,\n",
        "    digest_amplicons_w_BsaI\n",
        ")\n",
        "\n",
        "from streptocad.primers.primer_analysis import analyze_primers_and_hairpins\n",
        "from streptocad.primers.primer_generation import create_idt_order_dataframe, primers_to_IDT\n",
        "from streptocad.cloning.plasmid_processing import annotate_plasmid_with_sgrnas\n",
        "from streptocad.primers.primer_generation import find_best_check_primers_from_genome"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "# Inputs\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 2,
      "metadata": {},
      "outputs": [],
      "source": [
        "# 1 Add genome of choice (genbank, fasta)\n",
        "path_to_genome = '../../data/genomes/Streptomyces_coelicolor_A3_chromosome.gb'\n",
        "genome = load_and_process_plasmid(path_to_genome)\n",
        "\n",
        "# 2 Add plasmid \n",
        "path_to_plasmid = '../../data/plasmids/pCRISPR-MCBE_Csy4_kasopGFP.gb'\n",
        "clean_plasmid = load_and_process_plasmid(path_to_plasmid)\n",
        "\n",
        "# 3 Choose genes to knock out (list)\n",
        "genes_to_KO = ['SCO5087', 'SCO5089', 'SCO5090']\n",
        "#genes_to_KO = ['80000-100000', '4000-7000', '9000-14000','15000-20000']\n",
        "\n",
        "\n",
        "#### Advanced settings ####\n",
        "# 4 Filtering metrics for sgRNAs\n",
        "gc_upper = 0.72\n",
        "gc_lower = 0.2\n",
        "off_target_seed = 13\n",
        "off_target_upper = 10\n",
        "cas_type='cas9'\n",
        "number_of_sgRNAs_per_group = 5\n",
        "\n",
        "# 5 Choose polymerase and target melting temperature\n",
        "chosen_polymerase = polymerase_dict['Phusion High-Fidelity DNA Polymerase (GC Buffer)']\n",
        "melting_temperature = 65\n",
        "primer_concentration = 0.4 \n",
        "primer_number_increment = 0\n",
        "flanking_region=500\n",
        "\n",
        "\n",
        "# 6 Choose overlapping sequences for our plasmid we can use the following\n",
        "restriction_overhang_f = \"GATCGggtctcc\"\n",
        "restriction_overhang_r = \"GATCAGGTCTCg\"\n",
        "backbone_overhang_f = \"cATG\"\n",
        "backbone_overhang_r = \"cTAG\"\n",
        "cys4 = \"gTTCACTGCCGTATAGGCAGCTAAGAAA\"\n",
        "sgRNA_handle_cys4_site = [Dseqrecord('GTTTTAGAGCTAGAAATAGCAAGTTAAAATAAGGCTAGTCCGTTATCAACTTGAAAAAGTGGCACCGAGTCGGTGCTTTTTT', name = 'sgRNA_handle_cys4')]"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 3,
      "metadata": {},
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "pCRISPR-mcBEST\n"
          ]
        }
      ],
      "source": [
        "print(clean_plasmid.id)"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "# Computation"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 4,
      "metadata": {},
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "False\n"
          ]
        },
        {
          "data": {
            "text/plain": [
              "25824"
            ]
          },
          "execution_count": 4,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "target_dict, genes_to_KO, annotation_input = check_and_convert_input(genes_to_KO)\n",
        "\n",
        "print(annotation_input)\n",
        "if annotation_input == True:\n",
        "    genome = annotate_dseqrecord(genome, target_dict)\n",
        "\n",
        "\n",
        "len(genome.features)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 5,
      "metadata": {},
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "Pam was found outside designated locus_tag: SCO5087. To incorporate this extent borders. Skipping to next locus tag.\n",
            "sgRNA generated were outside the designated border in SCO5087. To incorporate this extent borders. Skipping to next locus tag.\n",
            "Pam was found outside designated locus_tag: SCO5087. To incorporate this extent borders. Skipping to next locus tag.\n",
            "sgRNA generated were outside the designated border in SCO5087. To incorporate this extent borders. Skipping to next locus tag.\n",
            "sgRNA generated were outside the designated border in SCO5087. To incorporate this extent borders. Skipping to next locus tag.\n",
            "sgRNA generated were outside the designated border in SCO5089. To incorporate this extent borders. Skipping to next locus tag.\n",
            "sgRNA generated were outside the designated border in SCO5089. To incorporate this extent borders. Skipping to next locus tag.\n",
            "Pam was found outside designated locus_tag: SCO5090. To incorporate this extent borders. Skipping to next locus tag.\n",
            "sgRNA generated were outside the designated border in SCO5090. To incorporate this extent borders. Skipping to next locus tag.\n",
            "sgRNA generated were outside the designated border in SCO5090. To incorporate this extent borders. Skipping to next locus tag.\n",
            "sgRNA generated were outside the designated border in SCO5090. To incorporate this extent borders. Skipping to next locus tag.\n",
            "sgRNA generated were outside the designated border in SCO5090. To incorporate this extent borders. Skipping to next locus tag.\n",
            "sgRNA generated were outside the designated border in SCO5090. To incorporate this extent borders. Skipping to next locus tag.\n"
          ]
        },
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>strain_name</th>\n",
              "      <th>locus_tag</th>\n",
              "      <th>gene_loc</th>\n",
              "      <th>gene_strand</th>\n",
              "      <th>sgrna_strand</th>\n",
              "      <th>sgrna_loc</th>\n",
              "      <th>gc</th>\n",
              "      <th>pam</th>\n",
              "      <th>sgrna</th>\n",
              "      <th>sgrna_seed_sequence</th>\n",
              "      <th>off_target_count</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>334</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5090</td>\n",
              "      <td>5532706</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>32</td>\n",
              "      <td>0.65</td>\n",
              "      <td>CGG</td>\n",
              "      <td>TGGACCACCCGGTGTTCTCC</td>\n",
              "      <td>CCCGGTGTTCTCC</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>330</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5089</td>\n",
              "      <td>5532449</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>61</td>\n",
              "      <td>0.70</td>\n",
              "      <td>CGG</td>\n",
              "      <td>CGTGGAGTGCGCCGGTGAGA</td>\n",
              "      <td>TGCGCCGGTGAGA</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>329</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5089</td>\n",
              "      <td>5532449</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>65</td>\n",
              "      <td>0.70</td>\n",
              "      <td>CGG</td>\n",
              "      <td>GAGTGCGCCGGTGAGACGGA</td>\n",
              "      <td>CCGGTGAGACGGA</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>328</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5089</td>\n",
              "      <td>5532449</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>66</td>\n",
              "      <td>0.70</td>\n",
              "      <td>GGG</td>\n",
              "      <td>AGTGCGCCGGTGAGACGGAC</td>\n",
              "      <td>CGGTGAGACGGAC</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>326</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5089</td>\n",
              "      <td>5532449</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>80</td>\n",
              "      <td>0.70</td>\n",
              "      <td>CGG</td>\n",
              "      <td>ACGGACGGGACGGACCTGTC</td>\n",
              "      <td>GGACGGACCTGTC</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>...</th>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>108</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5087</td>\n",
              "      <td>5529801</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>1038</td>\n",
              "      <td>0.70</td>\n",
              "      <td>GGG</td>\n",
              "      <td>GATGTCCGTCGCGTCCGTGC</td>\n",
              "      <td>GTCGCGTCCGTGC</td>\n",
              "      <td>2</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>295</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5089</td>\n",
              "      <td>5532449</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>60</td>\n",
              "      <td>0.70</td>\n",
              "      <td>GGG</td>\n",
              "      <td>CTCACCGGCGCACTCCACGA</td>\n",
              "      <td>GCGCACTCCACGA</td>\n",
              "      <td>2</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>449</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5090</td>\n",
              "      <td>5532706</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>890</td>\n",
              "      <td>0.65</td>\n",
              "      <td>CGG</td>\n",
              "      <td>ACCCATCTGCGCGAGGTGCT</td>\n",
              "      <td>TGCGCGAGGTGCT</td>\n",
              "      <td>3</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>191</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5087</td>\n",
              "      <td>5529801</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>838</td>\n",
              "      <td>0.70</td>\n",
              "      <td>TGG</td>\n",
              "      <td>TACGCGGGACGGCTTCGTGC</td>\n",
              "      <td>GACGGCTTCGTGC</td>\n",
              "      <td>3</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>200</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5087</td>\n",
              "      <td>5529801</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>706</td>\n",
              "      <td>0.65</td>\n",
              "      <td>CGG</td>\n",
              "      <td>CAGCGCCGACGTGATGTTCG</td>\n",
              "      <td>GACGTGATGTTCG</td>\n",
              "      <td>3</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>279 rows × 11 columns</p>\n",
              "</div>"
            ],
            "text/plain": [
              "     strain_name locus_tag  gene_loc  gene_strand  sgrna_strand  sgrna_loc  \\\n",
              "334  NC_003888.3   SCO5090   5532706            1            -1         32   \n",
              "330  NC_003888.3   SCO5089   5532449            1             1         61   \n",
              "329  NC_003888.3   SCO5089   5532449            1             1         65   \n",
              "328  NC_003888.3   SCO5089   5532449            1             1         66   \n",
              "326  NC_003888.3   SCO5089   5532449            1             1         80   \n",
              "..           ...       ...       ...          ...           ...        ...   \n",
              "108  NC_003888.3   SCO5087   5529801            1            -1       1038   \n",
              "295  NC_003888.3   SCO5089   5532449            1            -1         60   \n",
              "449  NC_003888.3   SCO5090   5532706            1             1        890   \n",
              "191  NC_003888.3   SCO5087   5529801            1             1        838   \n",
              "200  NC_003888.3   SCO5087   5529801            1             1        706   \n",
              "\n",
              "       gc  pam                 sgrna sgrna_seed_sequence  off_target_count  \n",
              "334  0.65  CGG  TGGACCACCCGGTGTTCTCC       CCCGGTGTTCTCC                 0  \n",
              "330  0.70  CGG  CGTGGAGTGCGCCGGTGAGA       TGCGCCGGTGAGA                 0  \n",
              "329  0.70  CGG  GAGTGCGCCGGTGAGACGGA       CCGGTGAGACGGA                 0  \n",
              "328  0.70  GGG  AGTGCGCCGGTGAGACGGAC       CGGTGAGACGGAC                 0  \n",
              "326  0.70  CGG  ACGGACGGGACGGACCTGTC       GGACGGACCTGTC                 0  \n",
              "..    ...  ...                   ...                 ...               ...  \n",
              "108  0.70  GGG  GATGTCCGTCGCGTCCGTGC       GTCGCGTCCGTGC                 2  \n",
              "295  0.70  GGG  CTCACCGGCGCACTCCACGA       GCGCACTCCACGA                 2  \n",
              "449  0.65  CGG  ACCCATCTGCGCGAGGTGCT       TGCGCGAGGTGCT                 3  \n",
              "191  0.70  TGG  TACGCGGGACGGCTTCGTGC       GACGGCTTCGTGC                 3  \n",
              "200  0.65  CGG  CAGCGCCGACGTGATGTTCG       GACGTGATGTTCG                 3  \n",
              "\n",
              "[279 rows x 11 columns]"
            ]
          },
          "execution_count": 5,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "# Initialize SgRNAargs with desired parameters\n",
        "args = SgRNAargs(genome, \n",
        "                genes_to_KO,\n",
        "                step=['find', 'filter'],\n",
        "                gc_upper = gc_upper,\n",
        "                gc_lower = gc_lower,\n",
        "                off_target_seed = off_target_seed,\n",
        "                off_target_upper = off_target_upper,\n",
        "                cas_type='cas9'\n",
        "                )\n",
        "\n",
        "sgrna_df = extract_sgRNAs(args)\n",
        "sgrna_df"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 6,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>strain_name</th>\n",
              "      <th>locus_tag</th>\n",
              "      <th>gene_loc</th>\n",
              "      <th>gene_strand</th>\n",
              "      <th>sgrna_strand</th>\n",
              "      <th>sgrna_loc</th>\n",
              "      <th>gc</th>\n",
              "      <th>pam</th>\n",
              "      <th>sgrna</th>\n",
              "      <th>sgrna_seed_sequence</th>\n",
              "      <th>off_target_count</th>\n",
              "      <th>editing_context</th>\n",
              "      <th>editable_cytosines</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>334</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5090</td>\n",
              "      <td>5532706</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>32</td>\n",
              "      <td>0.65</td>\n",
              "      <td>CGG</td>\n",
              "      <td>TGGACCACCCGGTGTTCTCC</td>\n",
              "      <td>CCCGGTGTTCTCC</td>\n",
              "      <td>0</td>\n",
              "      <td>0,0,0,0,0</td>\n",
              "      <td>5,6,8,9,10</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>330</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5089</td>\n",
              "      <td>5532449</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>61</td>\n",
              "      <td>0.70</td>\n",
              "      <td>CGG</td>\n",
              "      <td>CGTGGAGTGCGCCGGTGAGA</td>\n",
              "      <td>TGCGCCGGTGAGA</td>\n",
              "      <td>0</td>\n",
              "      <td>1</td>\n",
              "      <td>10</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>329</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5089</td>\n",
              "      <td>5532449</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>65</td>\n",
              "      <td>0.70</td>\n",
              "      <td>CGG</td>\n",
              "      <td>GAGTGCGCCGGTGAGACGGA</td>\n",
              "      <td>CCGGTGAGACGGA</td>\n",
              "      <td>0</td>\n",
              "      <td>1,1,0</td>\n",
              "      <td>6,8,9</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>328</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5089</td>\n",
              "      <td>5532449</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>66</td>\n",
              "      <td>0.70</td>\n",
              "      <td>GGG</td>\n",
              "      <td>AGTGCGCCGGTGAGACGGAC</td>\n",
              "      <td>CGGTGAGACGGAC</td>\n",
              "      <td>0</td>\n",
              "      <td>1,1,0</td>\n",
              "      <td>5,7,8</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>326</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5089</td>\n",
              "      <td>5532449</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>80</td>\n",
              "      <td>0.70</td>\n",
              "      <td>CGG</td>\n",
              "      <td>ACGGACGGGACGGACCTGTC</td>\n",
              "      <td>GGACGGACCTGTC</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>6</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>...</th>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>108</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5087</td>\n",
              "      <td>5529801</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>1038</td>\n",
              "      <td>0.70</td>\n",
              "      <td>GGG</td>\n",
              "      <td>GATGTCCGTCGCGTCCGTGC</td>\n",
              "      <td>GTCGCGTCCGTGC</td>\n",
              "      <td>2</td>\n",
              "      <td>0,0,0</td>\n",
              "      <td>6,7,10</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>295</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5089</td>\n",
              "      <td>5532449</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>60</td>\n",
              "      <td>0.70</td>\n",
              "      <td>GGG</td>\n",
              "      <td>CTCACCGGCGCACTCCACGA</td>\n",
              "      <td>GCGCACTCCACGA</td>\n",
              "      <td>2</td>\n",
              "      <td>0,0,0,1</td>\n",
              "      <td>3,5,6,9</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>449</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5090</td>\n",
              "      <td>5532706</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>890</td>\n",
              "      <td>0.65</td>\n",
              "      <td>CGG</td>\n",
              "      <td>ACCCATCTGCGCGAGGTGCT</td>\n",
              "      <td>TGCGCGAGGTGCT</td>\n",
              "      <td>3</td>\n",
              "      <td>0,0,0,1</td>\n",
              "      <td>3,4,7,10</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>191</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5087</td>\n",
              "      <td>5529801</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>838</td>\n",
              "      <td>0.70</td>\n",
              "      <td>TGG</td>\n",
              "      <td>TACGCGGGACGGCTTCGTGC</td>\n",
              "      <td>GACGGCTTCGTGC</td>\n",
              "      <td>3</td>\n",
              "      <td>0,1,0</td>\n",
              "      <td>3,5,10</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>200</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5087</td>\n",
              "      <td>5529801</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>706</td>\n",
              "      <td>0.65</td>\n",
              "      <td>CGG</td>\n",
              "      <td>CAGCGCCGACGTGATGTTCG</td>\n",
              "      <td>GACGTGATGTTCG</td>\n",
              "      <td>3</td>\n",
              "      <td>1,1,0,0</td>\n",
              "      <td>4,6,7,10</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>272 rows × 13 columns</p>\n",
              "</div>"
            ],
            "text/plain": [
              "     strain_name locus_tag  gene_loc  gene_strand  sgrna_strand  sgrna_loc  \\\n",
              "334  NC_003888.3   SCO5090   5532706            1            -1         32   \n",
              "330  NC_003888.3   SCO5089   5532449            1             1         61   \n",
              "329  NC_003888.3   SCO5089   5532449            1             1         65   \n",
              "328  NC_003888.3   SCO5089   5532449            1             1         66   \n",
              "326  NC_003888.3   SCO5089   5532449            1             1         80   \n",
              "..           ...       ...       ...          ...           ...        ...   \n",
              "108  NC_003888.3   SCO5087   5529801            1            -1       1038   \n",
              "295  NC_003888.3   SCO5089   5532449            1            -1         60   \n",
              "449  NC_003888.3   SCO5090   5532706            1             1        890   \n",
              "191  NC_003888.3   SCO5087   5529801            1             1        838   \n",
              "200  NC_003888.3   SCO5087   5529801            1             1        706   \n",
              "\n",
              "       gc  pam                 sgrna sgrna_seed_sequence  off_target_count  \\\n",
              "334  0.65  CGG  TGGACCACCCGGTGTTCTCC       CCCGGTGTTCTCC                 0   \n",
              "330  0.70  CGG  CGTGGAGTGCGCCGGTGAGA       TGCGCCGGTGAGA                 0   \n",
              "329  0.70  CGG  GAGTGCGCCGGTGAGACGGA       CCGGTGAGACGGA                 0   \n",
              "328  0.70  GGG  AGTGCGCCGGTGAGACGGAC       CGGTGAGACGGAC                 0   \n",
              "326  0.70  CGG  ACGGACGGGACGGACCTGTC       GGACGGACCTGTC                 0   \n",
              "..    ...  ...                   ...                 ...               ...   \n",
              "108  0.70  GGG  GATGTCCGTCGCGTCCGTGC       GTCGCGTCCGTGC                 2   \n",
              "295  0.70  GGG  CTCACCGGCGCACTCCACGA       GCGCACTCCACGA                 2   \n",
              "449  0.65  CGG  ACCCATCTGCGCGAGGTGCT       TGCGCGAGGTGCT                 3   \n",
              "191  0.70  TGG  TACGCGGGACGGCTTCGTGC       GACGGCTTCGTGC                 3   \n",
              "200  0.65  CGG  CAGCGCCGACGTGATGTTCG       GACGTGATGTTCG                 3   \n",
              "\n",
              "    editing_context editable_cytosines  \n",
              "334       0,0,0,0,0         5,6,8,9,10  \n",
              "330               1                 10  \n",
              "329           1,1,0              6,8,9  \n",
              "328           1,1,0              5,7,8  \n",
              "326               0                  6  \n",
              "..              ...                ...  \n",
              "108           0,0,0             6,7,10  \n",
              "295         0,0,0,1            3,5,6,9  \n",
              "449         0,0,0,1           3,4,7,10  \n",
              "191           0,1,0             3,5,10  \n",
              "200         1,1,0,0           4,6,7,10  \n",
              "\n",
              "[272 rows x 13 columns]"
            ]
          },
          "execution_count": 6,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "\n",
        "gene_sequences = process_specified_gene_sequences_from_record(genome, genes_to_KO)\n",
        "genes_to_KO_dict = {locus_tag: gene_sequences[locus_tag] for locus_tag in genes_to_KO if locus_tag in gene_sequences}\n",
        "\n",
        "# Identify and annotate base editing sites\n",
        "sgrna_df_with_editing = identify_base_editing_sites(sgrna_df)\n",
        "\n",
        "filtered_sgrna_df_for_base_editing = filter_sgrnas_for_base_editing(sgrna_df_with_editing)\n",
        "filtered_sgrna_df_for_base_editing"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 7,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>strain_name</th>\n",
              "      <th>locus_tag</th>\n",
              "      <th>gene_loc</th>\n",
              "      <th>gene_strand</th>\n",
              "      <th>sgrna_strand</th>\n",
              "      <th>sgrna_loc</th>\n",
              "      <th>gc</th>\n",
              "      <th>pam</th>\n",
              "      <th>sgrna</th>\n",
              "      <th>sgrna_seed_sequence</th>\n",
              "      <th>off_target_count</th>\n",
              "      <th>editing_context</th>\n",
              "      <th>editable_cytosines</th>\n",
              "      <th>mutations</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>334</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5090</td>\n",
              "      <td>5532706</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>32</td>\n",
              "      <td>0.65</td>\n",
              "      <td>CGG</td>\n",
              "      <td>TGGACCACCCGGTGTTCTCC</td>\n",
              "      <td>CCCGGTGTTCTCC</td>\n",
              "      <td>0</td>\n",
              "      <td>0,0,0,0,0</td>\n",
              "      <td>5,6,8,9,10</td>\n",
              "      <td>R8Q, V9I, V10I</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>325</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5089</td>\n",
              "      <td>5532449</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>106</td>\n",
              "      <td>0.65</td>\n",
              "      <td>AGG</td>\n",
              "      <td>CTTCCTCGACCTCCGCTTCG</td>\n",
              "      <td>GACCTCCGCTTCG</td>\n",
              "      <td>0</td>\n",
              "      <td>0,0,0,0</td>\n",
              "      <td>4,5,7,10</td>\n",
              "      <td>L31F</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>322</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5089</td>\n",
              "      <td>5532449</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>136</td>\n",
              "      <td>0.65</td>\n",
              "      <td>TGG</td>\n",
              "      <td>GTACGACTCGCTCGCCCTGA</td>\n",
              "      <td>TCGCTCGCCCTGA</td>\n",
              "      <td>0</td>\n",
              "      <td>0,0,0</td>\n",
              "      <td>4,7,9</td>\n",
              "      <td>S42L</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>318</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5089</td>\n",
              "      <td>5532449</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>194</td>\n",
              "      <td>0.65</td>\n",
              "      <td>CGG</td>\n",
              "      <td>TCCATACCCGACGACGTCGC</td>\n",
              "      <td>CCGACGACGTCGC</td>\n",
              "      <td>0</td>\n",
              "      <td>0,0,0,0</td>\n",
              "      <td>3,7,8,9</td>\n",
              "      <td>P61F</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>312</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5089</td>\n",
              "      <td>5532449</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>233</td>\n",
              "      <td>0.60</td>\n",
              "      <td>CGG</td>\n",
              "      <td>ATCAGGTCGAGCAGCTCTCG</td>\n",
              "      <td>CGAGCAGCTCTCG</td>\n",
              "      <td>0</td>\n",
              "      <td>0,0</td>\n",
              "      <td>3,8</td>\n",
              "      <td>D76N</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>...</th>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>534</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5090</td>\n",
              "      <td>5532706</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>140</td>\n",
              "      <td>0.65</td>\n",
              "      <td>CGG</td>\n",
              "      <td>CACGTACAGGTCCTGGAGCG</td>\n",
              "      <td>AGGTCCTGGAGCG</td>\n",
              "      <td>2</td>\n",
              "      <td>0,0</td>\n",
              "      <td>3,7</td>\n",
              "      <td>Q43*</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>78</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5087</td>\n",
              "      <td>5529801</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>745</td>\n",
              "      <td>0.65</td>\n",
              "      <td>GGG</td>\n",
              "      <td>CGACGACGATCGGGGTGATC</td>\n",
              "      <td>GATCGGGGTGATC</td>\n",
              "      <td>2</td>\n",
              "      <td>0,0</td>\n",
              "      <td>4,7</td>\n",
              "      <td>V247I, V248I</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>478</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5090</td>\n",
              "      <td>5532706</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>586</td>\n",
              "      <td>0.65</td>\n",
              "      <td>CGG</td>\n",
              "      <td>GGTGACACTCACCGAGGAGG</td>\n",
              "      <td>CTCACCGAGGAGG</td>\n",
              "      <td>2</td>\n",
              "      <td>0,0,0</td>\n",
              "      <td>6,8,10</td>\n",
              "      <td>T191I, L192F</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>335</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5090</td>\n",
              "      <td>5532706</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>43</td>\n",
              "      <td>0.65</td>\n",
              "      <td>CGG</td>\n",
              "      <td>TCCTCAGGGTGTGGACCACC</td>\n",
              "      <td>GGTGTGGACCACC</td>\n",
              "      <td>2</td>\n",
              "      <td>0,0</td>\n",
              "      <td>3,5</td>\n",
              "      <td>R14K</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>108</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5087</td>\n",
              "      <td>5529801</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>1038</td>\n",
              "      <td>0.70</td>\n",
              "      <td>GGG</td>\n",
              "      <td>GATGTCCGTCGCGTCCGTGC</td>\n",
              "      <td>GTCGCGTCCGTGC</td>\n",
              "      <td>2</td>\n",
              "      <td>0,0,0</td>\n",
              "      <td>6,7,10</td>\n",
              "      <td>D345N</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>112 rows × 14 columns</p>\n",
              "</div>"
            ],
            "text/plain": [
              "     strain_name locus_tag  gene_loc  gene_strand  sgrna_strand  sgrna_loc  \\\n",
              "334  NC_003888.3   SCO5090   5532706            1            -1         32   \n",
              "325  NC_003888.3   SCO5089   5532449            1             1        106   \n",
              "322  NC_003888.3   SCO5089   5532449            1             1        136   \n",
              "318  NC_003888.3   SCO5089   5532449            1             1        194   \n",
              "312  NC_003888.3   SCO5089   5532449            1            -1        233   \n",
              "..           ...       ...       ...          ...           ...        ...   \n",
              "534  NC_003888.3   SCO5090   5532706            1             1        140   \n",
              "78   NC_003888.3   SCO5087   5529801            1            -1        745   \n",
              "478  NC_003888.3   SCO5090   5532706            1             1        586   \n",
              "335  NC_003888.3   SCO5090   5532706            1            -1         43   \n",
              "108  NC_003888.3   SCO5087   5529801            1            -1       1038   \n",
              "\n",
              "       gc  pam                 sgrna sgrna_seed_sequence  off_target_count  \\\n",
              "334  0.65  CGG  TGGACCACCCGGTGTTCTCC       CCCGGTGTTCTCC                 0   \n",
              "325  0.65  AGG  CTTCCTCGACCTCCGCTTCG       GACCTCCGCTTCG                 0   \n",
              "322  0.65  TGG  GTACGACTCGCTCGCCCTGA       TCGCTCGCCCTGA                 0   \n",
              "318  0.65  CGG  TCCATACCCGACGACGTCGC       CCGACGACGTCGC                 0   \n",
              "312  0.60  CGG  ATCAGGTCGAGCAGCTCTCG       CGAGCAGCTCTCG                 0   \n",
              "..    ...  ...                   ...                 ...               ...   \n",
              "534  0.65  CGG  CACGTACAGGTCCTGGAGCG       AGGTCCTGGAGCG                 2   \n",
              "78   0.65  GGG  CGACGACGATCGGGGTGATC       GATCGGGGTGATC                 2   \n",
              "478  0.65  CGG  GGTGACACTCACCGAGGAGG       CTCACCGAGGAGG                 2   \n",
              "335  0.65  CGG  TCCTCAGGGTGTGGACCACC       GGTGTGGACCACC                 2   \n",
              "108  0.70  GGG  GATGTCCGTCGCGTCCGTGC       GTCGCGTCCGTGC                 2   \n",
              "\n",
              "    editing_context editable_cytosines       mutations  \n",
              "334       0,0,0,0,0         5,6,8,9,10  R8Q, V9I, V10I  \n",
              "325         0,0,0,0           4,5,7,10            L31F  \n",
              "322           0,0,0              4,7,9            S42L  \n",
              "318         0,0,0,0            3,7,8,9            P61F  \n",
              "312             0,0                3,8            D76N  \n",
              "..              ...                ...             ...  \n",
              "534             0,0                3,7            Q43*  \n",
              "78              0,0                4,7    V247I, V248I  \n",
              "478           0,0,0             6,8,10    T191I, L192F  \n",
              "335             0,0                3,5            R14K  \n",
              "108           0,0,0             6,7,10           D345N  \n",
              "\n",
              "[112 rows x 14 columns]"
            ]
          },
          "execution_count": 7,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "\n",
        "# Process the DataFrame to apply C-to-T mutations\n",
        "mutated_sgrna_df = process_base_editing(filtered_sgrna_df_for_base_editing, genes_to_KO_dict, only_stop_codons = False)\n",
        "mutated_sgrna_df"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 8,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>strain_name</th>\n",
              "      <th>locus_tag</th>\n",
              "      <th>gene_loc</th>\n",
              "      <th>gene_strand</th>\n",
              "      <th>sgrna_strand</th>\n",
              "      <th>sgrna_loc</th>\n",
              "      <th>gc</th>\n",
              "      <th>pam</th>\n",
              "      <th>sgrna</th>\n",
              "      <th>sgrna_seed_sequence</th>\n",
              "      <th>off_target_count</th>\n",
              "      <th>editing_context</th>\n",
              "      <th>editable_cytosines</th>\n",
              "      <th>mutations</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>534</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5090</td>\n",
              "      <td>5532706</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>140</td>\n",
              "      <td>0.65</td>\n",
              "      <td>CGG</td>\n",
              "      <td>CACGTACAGGTCCTGGAGCG</td>\n",
              "      <td>AGGTCCTGGAGCG</td>\n",
              "      <td>2</td>\n",
              "      <td>0,0</td>\n",
              "      <td>3,7</td>\n",
              "      <td>Q43*</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>319</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5089</td>\n",
              "      <td>5532449</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>167</td>\n",
              "      <td>0.70</td>\n",
              "      <td>CGG</td>\n",
              "      <td>GCGCGACTCGAGAGCCGGTA</td>\n",
              "      <td>TCGAGAGCCGGTA</td>\n",
              "      <td>0</td>\n",
              "      <td>1,0,0</td>\n",
              "      <td>4,7,9</td>\n",
              "      <td>R51*, L52F</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>352</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5090</td>\n",
              "      <td>5532706</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>174</td>\n",
              "      <td>0.65</td>\n",
              "      <td>CGG</td>\n",
              "      <td>CGCCCAGATCCGGAAGCGTT</td>\n",
              "      <td>ATCCGGAAGCGTT</td>\n",
              "      <td>0</td>\n",
              "      <td>1,0,0,0</td>\n",
              "      <td>3,4,5,10</td>\n",
              "      <td>W57*, A58T</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>354</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5090</td>\n",
              "      <td>5532706</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>208</td>\n",
              "      <td>0.65</td>\n",
              "      <td>CGG</td>\n",
              "      <td>GGGACGTCCAGGTCTTCACC</td>\n",
              "      <td>CCAGGTCTTCACC</td>\n",
              "      <td>0</td>\n",
              "      <td>0,0,0</td>\n",
              "      <td>5,8,9</td>\n",
              "      <td>W67*</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>17</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5087</td>\n",
              "      <td>5529801</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>210</td>\n",
              "      <td>0.60</td>\n",
              "      <td>GGG</td>\n",
              "      <td>GAGCAGTTCCCAGAACTGCC</td>\n",
              "      <td>TCCCAGAACTGCC</td>\n",
              "      <td>0</td>\n",
              "      <td>1,0,0</td>\n",
              "      <td>4,9,10</td>\n",
              "      <td>W67*, E68K</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>54</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5087</td>\n",
              "      <td>5529801</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>522</td>\n",
              "      <td>0.70</td>\n",
              "      <td>CGG</td>\n",
              "      <td>GTCGACCTCCCAGTCACGGC</td>\n",
              "      <td>TCCCAGTCACGGC</td>\n",
              "      <td>0</td>\n",
              "      <td>0,0,0,0,0</td>\n",
              "      <td>3,6,7,9,10</td>\n",
              "      <td>W171*, E172K, V173I, D174N</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>471</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5090</td>\n",
              "      <td>5532706</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>640</td>\n",
              "      <td>0.60</td>\n",
              "      <td>TGG</td>\n",
              "      <td>GACCGTCCAGGACATGACCA</td>\n",
              "      <td>CAGGACATGACCA</td>\n",
              "      <td>0</td>\n",
              "      <td>0,0,0,0</td>\n",
              "      <td>3,4,7,8</td>\n",
              "      <td>T208I, Q210*</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "     strain_name locus_tag  gene_loc  gene_strand  sgrna_strand  sgrna_loc  \\\n",
              "534  NC_003888.3   SCO5090   5532706            1             1        140   \n",
              "319  NC_003888.3   SCO5089   5532449            1             1        167   \n",
              "352  NC_003888.3   SCO5090   5532706            1            -1        174   \n",
              "354  NC_003888.3   SCO5090   5532706            1            -1        208   \n",
              "17   NC_003888.3   SCO5087   5529801            1            -1        210   \n",
              "54   NC_003888.3   SCO5087   5529801            1            -1        522   \n",
              "471  NC_003888.3   SCO5090   5532706            1             1        640   \n",
              "\n",
              "       gc  pam                 sgrna sgrna_seed_sequence  off_target_count  \\\n",
              "534  0.65  CGG  CACGTACAGGTCCTGGAGCG       AGGTCCTGGAGCG                 2   \n",
              "319  0.70  CGG  GCGCGACTCGAGAGCCGGTA       TCGAGAGCCGGTA                 0   \n",
              "352  0.65  CGG  CGCCCAGATCCGGAAGCGTT       ATCCGGAAGCGTT                 0   \n",
              "354  0.65  CGG  GGGACGTCCAGGTCTTCACC       CCAGGTCTTCACC                 0   \n",
              "17   0.60  GGG  GAGCAGTTCCCAGAACTGCC       TCCCAGAACTGCC                 0   \n",
              "54   0.70  CGG  GTCGACCTCCCAGTCACGGC       TCCCAGTCACGGC                 0   \n",
              "471  0.60  TGG  GACCGTCCAGGACATGACCA       CAGGACATGACCA                 0   \n",
              "\n",
              "    editing_context editable_cytosines                   mutations  \n",
              "534             0,0                3,7                        Q43*  \n",
              "319           1,0,0              4,7,9                  R51*, L52F  \n",
              "352         1,0,0,0           3,4,5,10                  W57*, A58T  \n",
              "354           0,0,0              5,8,9                        W67*  \n",
              "17            1,0,0             4,9,10                  W67*, E68K  \n",
              "54        0,0,0,0,0         3,6,7,9,10  W171*, E172K, V173I, D174N  \n",
              "471         0,0,0,0            3,4,7,8                T208I, Q210*  "
            ]
          },
          "execution_count": 8,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "# Process the DataFrame to apply C-to-T mutations\n",
        "mutated_sgrna_df = process_base_editing(filtered_sgrna_df_for_base_editing, \n",
        "                                        genes_to_KO_dict, \n",
        "                                        only_stop_codons = True, \n",
        "                                        editing_context= False)\n",
        "mutated_sgrna_df"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 9,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>strain_name</th>\n",
              "      <th>locus_tag</th>\n",
              "      <th>gene_loc</th>\n",
              "      <th>gene_strand</th>\n",
              "      <th>sgrna_strand</th>\n",
              "      <th>sgrna_loc</th>\n",
              "      <th>gc</th>\n",
              "      <th>pam</th>\n",
              "      <th>sgrna</th>\n",
              "      <th>sgrna_seed_sequence</th>\n",
              "      <th>off_target_count</th>\n",
              "      <th>editing_context</th>\n",
              "      <th>editable_cytosines</th>\n",
              "      <th>mutations</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>534</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5090</td>\n",
              "      <td>5532706</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>140</td>\n",
              "      <td>0.65</td>\n",
              "      <td>CGG</td>\n",
              "      <td>CACGTACAGGTCCTGGAGCG</td>\n",
              "      <td>AGGTCCTGGAGCG</td>\n",
              "      <td>2</td>\n",
              "      <td>0,0</td>\n",
              "      <td>3,7</td>\n",
              "      <td>Q43*</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>354</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5090</td>\n",
              "      <td>5532706</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>208</td>\n",
              "      <td>0.65</td>\n",
              "      <td>CGG</td>\n",
              "      <td>GGGACGTCCAGGTCTTCACC</td>\n",
              "      <td>CCAGGTCTTCACC</td>\n",
              "      <td>0</td>\n",
              "      <td>0,0,0</td>\n",
              "      <td>5,8,9</td>\n",
              "      <td>W67*</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>54</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5087</td>\n",
              "      <td>5529801</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>522</td>\n",
              "      <td>0.70</td>\n",
              "      <td>CGG</td>\n",
              "      <td>GTCGACCTCCCAGTCACGGC</td>\n",
              "      <td>TCCCAGTCACGGC</td>\n",
              "      <td>0</td>\n",
              "      <td>0,0,0,0,0</td>\n",
              "      <td>3,6,7,9,10</td>\n",
              "      <td>W171*, E172K, V173I, D174N</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>471</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5090</td>\n",
              "      <td>5532706</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>640</td>\n",
              "      <td>0.60</td>\n",
              "      <td>TGG</td>\n",
              "      <td>GACCGTCCAGGACATGACCA</td>\n",
              "      <td>CAGGACATGACCA</td>\n",
              "      <td>0</td>\n",
              "      <td>0,0,0,0</td>\n",
              "      <td>3,4,7,8</td>\n",
              "      <td>T208I, Q210*</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "     strain_name locus_tag  gene_loc  gene_strand  sgrna_strand  sgrna_loc  \\\n",
              "534  NC_003888.3   SCO5090   5532706            1             1        140   \n",
              "354  NC_003888.3   SCO5090   5532706            1            -1        208   \n",
              "54   NC_003888.3   SCO5087   5529801            1            -1        522   \n",
              "471  NC_003888.3   SCO5090   5532706            1             1        640   \n",
              "\n",
              "       gc  pam                 sgrna sgrna_seed_sequence  off_target_count  \\\n",
              "534  0.65  CGG  CACGTACAGGTCCTGGAGCG       AGGTCCTGGAGCG                 2   \n",
              "354  0.65  CGG  GGGACGTCCAGGTCTTCACC       CCAGGTCTTCACC                 0   \n",
              "54   0.70  CGG  GTCGACCTCCCAGTCACGGC       TCCCAGTCACGGC                 0   \n",
              "471  0.60  TGG  GACCGTCCAGGACATGACCA       CAGGACATGACCA                 0   \n",
              "\n",
              "    editing_context editable_cytosines                   mutations  \n",
              "534             0,0                3,7                        Q43*  \n",
              "354           0,0,0              5,8,9                        W67*  \n",
              "54        0,0,0,0,0         3,6,7,9,10  W171*, E172K, V173I, D174N  \n",
              "471         0,0,0,0            3,4,7,8                T208I, Q210*  "
            ]
          },
          "execution_count": 9,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "# Process the DataFrame to apply C-to-T mutations\n",
        "mutated_sgrna_df = process_base_editing(filtered_sgrna_df_for_base_editing, \n",
        "                                        genes_to_KO_dict, \n",
        "                                        only_stop_codons = True, \n",
        "                                        editing_context= True)\n",
        "mutated_sgrna_df"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 10,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>strain_name</th>\n",
              "      <th>locus_tag</th>\n",
              "      <th>gene_loc</th>\n",
              "      <th>gene_strand</th>\n",
              "      <th>sgrna_strand</th>\n",
              "      <th>sgrna_loc</th>\n",
              "      <th>gc</th>\n",
              "      <th>pam</th>\n",
              "      <th>sgrna</th>\n",
              "      <th>sgrna_seed_sequence</th>\n",
              "      <th>off_target_count</th>\n",
              "      <th>editing_context</th>\n",
              "      <th>editable_cytosines</th>\n",
              "      <th>mutations</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>534</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5090</td>\n",
              "      <td>5532706</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>140</td>\n",
              "      <td>0.65</td>\n",
              "      <td>CGG</td>\n",
              "      <td>CACGTACAGGTCCTGGAGCG</td>\n",
              "      <td>AGGTCCTGGAGCG</td>\n",
              "      <td>2</td>\n",
              "      <td>0,0</td>\n",
              "      <td>3,7</td>\n",
              "      <td>Q43*</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>354</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5090</td>\n",
              "      <td>5532706</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>208</td>\n",
              "      <td>0.65</td>\n",
              "      <td>CGG</td>\n",
              "      <td>GGGACGTCCAGGTCTTCACC</td>\n",
              "      <td>CCAGGTCTTCACC</td>\n",
              "      <td>0</td>\n",
              "      <td>0,0,0</td>\n",
              "      <td>5,8,9</td>\n",
              "      <td>W67*</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>54</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5087</td>\n",
              "      <td>5529801</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>522</td>\n",
              "      <td>0.70</td>\n",
              "      <td>CGG</td>\n",
              "      <td>GTCGACCTCCCAGTCACGGC</td>\n",
              "      <td>TCCCAGTCACGGC</td>\n",
              "      <td>0</td>\n",
              "      <td>0,0,0,0,0</td>\n",
              "      <td>3,6,7,9,10</td>\n",
              "      <td>W171*, E172K, V173I, D174N</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>471</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5090</td>\n",
              "      <td>5532706</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>640</td>\n",
              "      <td>0.60</td>\n",
              "      <td>TGG</td>\n",
              "      <td>GACCGTCCAGGACATGACCA</td>\n",
              "      <td>CAGGACATGACCA</td>\n",
              "      <td>0</td>\n",
              "      <td>0,0,0,0</td>\n",
              "      <td>3,4,7,8</td>\n",
              "      <td>T208I, Q210*</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "     strain_name locus_tag  gene_loc  gene_strand  sgrna_strand  sgrna_loc  \\\n",
              "534  NC_003888.3   SCO5090   5532706            1             1        140   \n",
              "354  NC_003888.3   SCO5090   5532706            1            -1        208   \n",
              "54   NC_003888.3   SCO5087   5529801            1            -1        522   \n",
              "471  NC_003888.3   SCO5090   5532706            1             1        640   \n",
              "\n",
              "       gc  pam                 sgrna sgrna_seed_sequence  off_target_count  \\\n",
              "534  0.65  CGG  CACGTACAGGTCCTGGAGCG       AGGTCCTGGAGCG                 2   \n",
              "354  0.65  CGG  GGGACGTCCAGGTCTTCACC       CCAGGTCTTCACC                 0   \n",
              "54   0.70  CGG  GTCGACCTCCCAGTCACGGC       TCCCAGTCACGGC                 0   \n",
              "471  0.60  TGG  GACCGTCCAGGACATGACCA       CAGGACATGACCA                 0   \n",
              "\n",
              "    editing_context editable_cytosines                   mutations  \n",
              "534             0,0                3,7                        Q43*  \n",
              "354           0,0,0              5,8,9                        W67*  \n",
              "54        0,0,0,0,0         3,6,7,9,10  W171*, E172K, V173I, D174N  \n",
              "471         0,0,0,0            3,4,7,8                T208I, Q210*  "
            ]
          },
          "execution_count": 10,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "# Filter the DataFrame to retain only up to 5 sgRNA sequences per locus_tag\n",
        "filtered_df = mutated_sgrna_df.groupby('locus_tag').head(number_of_sgRNAs_per_group)\n",
        "filtered_df"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "## Output"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 11,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/plain": [
              "[Dseqrecord(-20), Dseqrecord(-20), Dseqrecord(-20), Dseqrecord(-20)]"
            ]
          },
          "execution_count": 11,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "sgRNA_list = dataframe_to_seqrecords(filtered_df)\n",
        "sgRNA_list"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 12,
      "metadata": {},
      "outputs": [],
      "source": [
        "sgRNA_handle_cys4_sites = sgRNA_handle_cys4_site*len(sgRNA_list)\n",
        "\n",
        "golden_gate = GoldenGateCloning(sgRNA_list,\n",
        "                                sgRNA_handle_cys4_sites, \n",
        "                                target_tm= melting_temperature,\n",
        "                                restriction_overhang_f = restriction_overhang_f,\n",
        "                                restriction_overhang_r = restriction_overhang_r,\n",
        "                                backbone_overhang_f = backbone_overhang_f,\n",
        "                                backbone_overhang_r = backbone_overhang_r,\n",
        "                                cys4 = cys4,\n",
        "                                tm_function = primer_tm_neb, \n",
        "                                primer_incrementation = primer_number_increment, \n",
        "                                polymerase = chosen_polymerase ) "
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 13,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>Locus Tag</th>\n",
              "      <th>f_primer_name</th>\n",
              "      <th>r_primer_name</th>\n",
              "      <th>f_primer_sequences(5-3)</th>\n",
              "      <th>r_primer_sequences(5-3)</th>\n",
              "      <th>f_tm</th>\n",
              "      <th>r_tm</th>\n",
              "      <th>ta</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>sgRNA_handle_cys4</td>\n",
              "      <td>primer_0</td>\n",
              "      <td>primer_1</td>\n",
              "      <td>GATCGGGTCTCCCATGGTTCACTGCCGTATAGGCAGCTAAGAAACA...</td>\n",
              "      <td>GATCAGGTCTCGTCCCAAAAAAGCACCGACTCGGT</td>\n",
              "      <td>63</td>\n",
              "      <td>59</td>\n",
              "      <td>63</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>sgRNA_handle_cys4</td>\n",
              "      <td>primer_2</td>\n",
              "      <td>primer_3</td>\n",
              "      <td>GATCGGGTCTCCGGGACGTCCAGGTCTTCACCGTTTTAGAGCTAGA...</td>\n",
              "      <td>GATCAGGTCTCGCGACAAAAAAGCACCGACTCGGT</td>\n",
              "      <td>63</td>\n",
              "      <td>59</td>\n",
              "      <td>63</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>sgRNA_handle_cys4</td>\n",
              "      <td>primer_4</td>\n",
              "      <td>primer_5</td>\n",
              "      <td>GATCGGGTCTCCGTCGACCTCCCAGTCACGGCGTTTTAGAGCTAGA...</td>\n",
              "      <td>GATCAGGTCTCGGGTCAAAAAAGCACCGACTCGGT</td>\n",
              "      <td>63</td>\n",
              "      <td>59</td>\n",
              "      <td>63</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>sgRNA_handle_cys4</td>\n",
              "      <td>primer_6</td>\n",
              "      <td>primer_7</td>\n",
              "      <td>GATCGGGTCTCCGACCGTCCAGGACATGACCAGTTTTAGAGCTAGA...</td>\n",
              "      <td>GATCAGGTCTCGCTAGAAAAAAGCACCGACTCGGT</td>\n",
              "      <td>63</td>\n",
              "      <td>59</td>\n",
              "      <td>63</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "           Locus Tag f_primer_name r_primer_name  \\\n",
              "0  sgRNA_handle_cys4      primer_0      primer_1   \n",
              "1  sgRNA_handle_cys4      primer_2      primer_3   \n",
              "2  sgRNA_handle_cys4      primer_4      primer_5   \n",
              "3  sgRNA_handle_cys4      primer_6      primer_7   \n",
              "\n",
              "                             f_primer_sequences(5-3)  \\\n",
              "0  GATCGGGTCTCCCATGGTTCACTGCCGTATAGGCAGCTAAGAAACA...   \n",
              "1  GATCGGGTCTCCGGGACGTCCAGGTCTTCACCGTTTTAGAGCTAGA...   \n",
              "2  GATCGGGTCTCCGTCGACCTCCCAGTCACGGCGTTTTAGAGCTAGA...   \n",
              "3  GATCGGGTCTCCGACCGTCCAGGACATGACCAGTTTTAGAGCTAGA...   \n",
              "\n",
              "               r_primer_sequences(5-3)  f_tm  r_tm  ta  \n",
              "0  GATCAGGTCTCGTCCCAAAAAAGCACCGACTCGGT    63    59  63  \n",
              "1  GATCAGGTCTCGCGACAAAAAAGCACCGACTCGGT    63    59  63  \n",
              "2  GATCAGGTCTCGGGTCAAAAAAGCACCGACTCGGT    63    59  63  \n",
              "3  GATCAGGTCTCGCTAGAAAAAAGCACCGACTCGGT    63    59  63  "
            ]
          },
          "execution_count": 13,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "primer_df = golden_gate.generate_primer_dataframe()\n",
        "primer_df"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 14,
      "metadata": {},
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "Skipping analysis for primer pair primer_0 & primer_1 due to length constraints.\n",
            "Skipping analysis for primer pair primer_2 & primer_3 due to length constraints.\n",
            "Skipping analysis for primer pair primer_4 & primer_5 due to length constraints.\n",
            "Skipping analysis for primer pair primer_6 & primer_7 due to length constraints.\n"
          ]
        },
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "Empty DataFrame\n",
              "Columns: []\n",
              "Index: []"
            ]
          },
          "execution_count": 14,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "# Primer analysis\n",
        "analysis_of_primers = analyze_primers_and_hairpins(primer_df)\n",
        "analysis_of_primers"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 15,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>Name</th>\n",
              "      <th>Sequence</th>\n",
              "      <th>Concentration</th>\n",
              "      <th>Purification</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>primer_0</td>\n",
              "      <td>GATCGGGTCTCCCATGGTTCACTGCCGTATAGGCAGCTAAGAAACA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>primer_2</td>\n",
              "      <td>GATCGGGTCTCCGGGACGTCCAGGTCTTCACCGTTTTAGAGCTAGA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>primer_4</td>\n",
              "      <td>GATCGGGTCTCCGTCGACCTCCCAGTCACGGCGTTTTAGAGCTAGA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>primer_6</td>\n",
              "      <td>GATCGGGTCTCCGACCGTCCAGGACATGACCAGTTTTAGAGCTAGA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>primer_1</td>\n",
              "      <td>GATCAGGTCTCGTCCCAAAAAAGCACCGACTCGGT</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>5</th>\n",
              "      <td>primer_3</td>\n",
              "      <td>GATCAGGTCTCGCGACAAAAAAGCACCGACTCGGT</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>6</th>\n",
              "      <td>primer_5</td>\n",
              "      <td>GATCAGGTCTCGGGTCAAAAAAGCACCGACTCGGT</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>7</th>\n",
              "      <td>primer_7</td>\n",
              "      <td>GATCAGGTCTCGCTAGAAAAAAGCACCGACTCGGT</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "       Name                                           Sequence Concentration  \\\n",
              "0  primer_0  GATCGGGTCTCCCATGGTTCACTGCCGTATAGGCAGCTAAGAAACA...          25nm   \n",
              "1  primer_2  GATCGGGTCTCCGGGACGTCCAGGTCTTCACCGTTTTAGAGCTAGA...          25nm   \n",
              "2  primer_4  GATCGGGTCTCCGTCGACCTCCCAGTCACGGCGTTTTAGAGCTAGA...          25nm   \n",
              "3  primer_6  GATCGGGTCTCCGACCGTCCAGGACATGACCAGTTTTAGAGCTAGA...          25nm   \n",
              "4  primer_1                GATCAGGTCTCGTCCCAAAAAAGCACCGACTCGGT          25nm   \n",
              "5  primer_3                GATCAGGTCTCGCGACAAAAAAGCACCGACTCGGT          25nm   \n",
              "6  primer_5                GATCAGGTCTCGGGTCAAAAAAGCACCGACTCGGT          25nm   \n",
              "7  primer_7                GATCAGGTCTCGCTAGAAAAAAGCACCGACTCGGT          25nm   \n",
              "\n",
              "  Purification  \n",
              "0          STD  \n",
              "1          STD  \n",
              "2          STD  \n",
              "3          STD  \n",
              "4          STD  \n",
              "5          STD  \n",
              "6          STD  \n",
              "7          STD  "
            ]
          },
          "execution_count": 15,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "# 2 Ready made IDT format\n",
        "idt_df = create_idt_order_dataframe(primer_df, concentration=\"25nm\", purification=\"STD\")\n",
        "idt_df"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 16,
      "metadata": {},
      "outputs": [],
      "source": [
        "list_of_amplicons = golden_gate.simulate_pcrs()"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 17,
      "metadata": {},
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "SCO5090_140_pcr_products\n",
            "SCO5090_208_pcr_products\n",
            "SCO5087_522_pcr_products\n",
            "SCO5090_640_pcr_products\n"
          ]
        },
        {
          "data": {
            "image/jpeg": "/9j/4AAQSkZJRgABAQAAAQABAAD/2wBDAAgGBgcGBQgHBwcJCQgKDBQNDAsLDBkSEw8UHRofHh0aHBwgJC4nICIsIxwcKDcpLDAxNDQ0Hyc5PTgyPC4zNDL/2wBDAQkJCQwLDBgNDRgyIRwhMjIyMjIyMjIyMjIyMjIyMjIyMjIyMjIyMjIyMjIyMjIyMjIyMjIyMjIyMjIyMjIyMjL/wAARCAJYAWgDASIAAhEBAxEB/8QAHwAAAQUBAQEBAQEAAAAAAAAAAAECAwQFBgcICQoL/8QAtRAAAgEDAwIEAwUFBAQAAAF9AQIDAAQRBRIhMUEGE1FhByJxFDKBkaEII0KxwRVS0fAkM2JyggkKFhcYGRolJicoKSo0NTY3ODk6Q0RFRkdISUpTVFVWV1hZWmNkZWZnaGlqc3R1dnd4eXqDhIWGh4iJipKTlJWWl5iZmqKjpKWmp6ipqrKztLW2t7i5usLDxMXGx8jJytLT1NXW19jZ2uHi4+Tl5ufo6erx8vP09fb3+Pn6/8QAHwEAAwEBAQEBAQEBAQAAAAAAAAECAwQFBgcICQoL/8QAtREAAgECBAQDBAcFBAQAAQJ3AAECAxEEBSExBhJBUQdhcRMiMoEIFEKRobHBCSMzUvAVYnLRChYkNOEl8RcYGRomJygpKjU2Nzg5OkNERUZHSElKU1RVVldYWVpjZGVmZ2hpanN0dXZ3eHl6goOEhYaHiImKkpOUlZaXmJmaoqOkpaanqKmqsrO0tba3uLm6wsPExcbHyMnK0tPU1dbX2Nna4uPk5ebn6Onq8vP09fb3+Pn6/9oADAMBAAIRAxEAPwDwiHTr24tJruGzuJLaD/WzJExSP/eYDA/GnvpOpRfZfM0+7T7Xj7NuhYednGNnHzdR09RXQaL51r4Xu78Xto7iO4t7eze8iRoxIgWaUozBmynyqACSef4QD0japYRapqszajaGPVb+WbT2WdW+zo1tcIhfB/dYaaEfNgjyz2FAHnyaJq0l9JYppd615Eu6S3Fu5kQcclcZA5H51CmnX0ljJfJZ3DWkTbZLgRMY0PHBbGAeR+ddxJJFPoMGiQanZRapbxWhlnN2ixkJLcsVEudrFRNBwCfucZxVjU9W029u7jVbS7t005LXU4GtjIqyNJNJOYsR53EYlhOQMDy+cYoA82oqZ2tjaRqkMwuQT5kjSgow7ALtBB99x/CoaAFCkjpRtPpU8VwEjC7Acd6f9rH/ADzX8qAKu0+lG0+lWvtY/wCea/lR9rH/ADzX8qAKu0+lG0+lWvtY/wCea/lR9rH/ADzX8qAKpBHUUlTzTiVAoUDnPFQUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRWnZ6HPe6LqGqRz2qxWKh3iaX964LomVUZ4BkTk4HJxkgitB/BeorOkK3Fo8iymG5Cu3+iOELkSZUdFRydu77jdxQBzlFdAvhK5Ll31Cwjsikbx3rtIIpPMLKoHybs5RxyoxsbOMU1/CWoQ20zzvbxXMazOLN2bzXSJisjDA24Uq/Ug/I2AcUAYNFFFABRUyW0kiBlBwad9jl/umgCvRVj7HL/dNH2OX+6aAK9FWPscv900fY5f7poAr0VLJA8S7mGBnFRUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFAGnpeow2Wn61byK5e+slt4ioGAwuIZMnnpiNumeSPrXTP4y0xb2+uI4btv7XupLjUFZFHlb4ZomWM7ju/4+JCCdvRR6muGooA69/EGi3GmR6HP9vGmQJB5c6Qp5rtG87nKb8AH7TIPvHGF61Le+MbPUZptUnhuE1PyL23jiRVMRW5eVixbOQV+0SDG05wvIri6KAJnupHtI7YrCI4yWUrCgc59XA3EexJxUNFFAEyXMkaBVJwKd9sl/vGq9FAFj7ZL/eNH2yX+8ar0UAWPtkv940fbJf7xqvRQBLJO8q7WORnNRUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRW1puk6bfaVe3MmoXcM9pbtM6fZFaMncFRd/mA5Yso+7xyeQK1H8EL9qmt49S3PZXDW+pFoMC3ZYpJGKfMd4AhlH8PK++aAORorqD4Vs47cajPqskekyJC0NwLUGRmkaRQGj34GDDLk7j93jOawNRsZtL1O70+4x59rM8Em05G5WKnH4igCtRRRQAUVbgtVkhVy6gnsTUn2FP+ei/nQBQoq/9hT/AJ6L+dH2FP8Anov50AUKKv8A2FP+ei/nR9hT/nov50AUKKtXNssMYYMDk44qrQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAX/7S26F/ZkUWzfcefPLuz5mFwi4xwFy5993tW+/jdftU1xHpu1724a41INPkXDNFJGwT5RsBE0p/i5b2xXI0UAdQfFVnJbrp0+lSSaTGkKw24ugJFaNpGBaTZg5M0uRtH3uMYrFuNYvbjXLjWFneC9nnecywsUKu5JOCOR1NUaKAFZmdy7sWZjkknJJpKKKAHB2AwDxR5j+tNooAd5j+tHmP602igB3mP60eY/rTaKAHF2YYJptFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUVvL4eUeGE1iRtQO/fgQ2W+FMHaN8u8bcnP8J6VoP4IX7VNbx6lueyuGt9SLQYFuyxSSMU+Y7wBDKP4eV980AcjRXUHwrZx241GfVZI9JkSFobgWoMjNI0igNHvwMGGXJ3H7vGc1gajYzaXqd3p9xjz7WZ4JNpyNysVOPxFAFaiiigAorstC8KW+paNb3b3cMbSbsqzcjDEf0rR/4Qa1/wCf+D/vqgDzyivQ/wDhBrX/AJ/4P++qP+EGtf8An/g/76oA88or0P8A4Qa1/wCf+D/vqj/hBrX/AJ/4P++qAPPKK6rxL4bh0fTo7iO5ilLTBNqHJ6E5/SuVoAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKAN6x8RRWWliEWczXaW89skpuj5QSUENmIg/NhjgggZwSCRzoP43X7VNcR6bte9uGuNSDT5FwzRSRsE+UbARNKf4uW9sVyNFAHUHxVZyW66dPpUkmkxpCsNuLoCRWjaRgWk2YOTNLkbR97jGKxbjWL241y41hZ3gvZ53nMsLFCruSTgjkdTVGigBWZncu7FmY5JJySaSiigCzFfzwxCNJGCjoAaf/al1/z1b86p0UAXP7Uuv+erfnR/al1/z1b86p0UAXP7Uuv+erfnR/al1/z1b86p0UAWJ7ya4QJI5YA5wTVeiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAuadpV9qr3C2Nu0xtreS5mwQAkSDLMSfQfn0HNXH8LazG9sjWeXuH2IqyoxVtu7DgH92dpzhscZPY1r+E/EWkaXZXFle2tyhmgu/MuIrgKJS1tJHGhXy2PVyAc4BfJyAMareJdGj1DVLhb7emt3ktw4WJ82SyW9xHhsgZIa5/hzxH7gUAcnH4W1iS8e2W3i3qiPvNzEIyrnC7ZC2xsngYJzg+hrJkjeGV4pUZJEYqysMFSOoIrtZdS0e60OHw82rRxJbx2x+3+TKUkZJLl2UALv6XOBlRny+2RXO6rqsN/4q1DVhaJLBc3ctwLeYsBh2JAO0g8Z7GgDKopWIZyQoUE5CjOB7c0lAFiK0kljDqDg0/wCwS/3TTIruSKMIpOBT/t8v940AH2CX+6aPsEv900fb5f7xo+3y/wB40AH2CX+6aPsEv900fb5f7xo+3y/3jQBHNbPCgZhgE4qCp5rl5kCscgHNQUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAWotMv57GW+hsbmS0iOJJ0iYxoeOrAYHUfnTn0nUovsvmafdp9rx9m3QsPOzjGzj5uo6eorpNAljtNGubma/tGifT7qFRJPia0kZWGyOIt8wkyoLbSAGPIK5roG1Swi1TVZm1G0Meq38s2nss6t9nRra4RC+D+6w00I+bBHlnsKAPPk0TVpL6SxTS71ryJd0luLdzIg45K4yByPzqh0NegySRT6DBokGp2UWqW8VoZZzdosZCS3LFRLnaxUTQcAn7nGcVyus3mn3ni/U71YpJNPnvZpUjicRMY2diuCVOOCOxoAyKKVtpclAQueATkgfWkoAlS3d0DAHBp32ST+6amgvmhhWMYwM1J/ab+1AFX7JJ/dNH2ST+6atf2m/tR/ab+1AFX7JJ/dNH2ST+6atf2m/tR/ab+1AFOSB41ywwM4qKrVzeNcRhD2Oaq0AFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUVoWuh6je6fLfW8AaCIOSTIoZggDPtUnc20EE7QcA5OKsP4W1mN7ZGs8vcPsRVlRirbd2HAP7s7TnDY4yexoAx6K2I/C2sSXj2y28W9UR95uYhGVc4XbIW2Nk8DBOcH0NZMkbwyvFKjJIjFWVhgqR1BFADaKKKACipo7Z5EDAHBp/2OX+6aAK1FWfscv900fY5f7poArUVZ+xy/3TR9jl/umgCtRUstu8ShmGBnFRUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFAHSaTNZWXhy8nXVoE1SVJII7eVZsxRMuHKkIV3uCU5IAGc9QR0DeJdGj1DVLhb7emt3ktw4WJ82SyW9xHhsgZIa5/hzxH7gV53RQB28upaPdaHD4ebVo4kt47Y/b/JlKSMkly7KAF39LnAyoz5fbIrndV1WG/wDFWoasLRJYLm7luBbzFgMOxIB2kHjPY1lUUAKxDOSFCgnIUZwPbmkoooAv2+oNBAsYxgZ/nUn9qv7flWZRQBp/2q/t+VH9qv7flWZRQBp/2q/t+VH9qv7flWZRQBcu71rmIIezZqnRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUVt6XpunXvh7WrmWS6/tCygWeJFCiLaZoYzuPJJPmngYxtHJzgbt14OtItSjtxb3aW0SyPc3QvYptwjgaZ1CKgMbEI23cTx2NAHD0V17+H9Ft9Mj1yf7edMnSDy4EmTzUaR50OX2YIH2aQ/dGcr0rCv8ARbq08R3miQRyXVzb3MluFhQs0hRiCQBz2zQBm0UrKyOUdSrKcEEYINJQA9Y2YZAOKPJf+6a9V8G+HfDt/wCFLK5v75Yrl/M3oV6YdgP0Ard/4RHwj/0E0/75oA8N8l/7po8l/wC6a9y/4RHwj/0E0/75o/4RHwj/ANBNP++aAPDfJf8AumjyX/umvcv+ER8I/wDQTT/vmj/hEfCP/QTT/vmgDwxo2UZIIptek/ELQtC0zQIJtMvFmna6VGUDGF2uc/mBXm1ABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQBbs9RmsrXULeNUKX1uLeUsDkKJUkyOeuY1654J+tayeL7uEp9msbC2BkEtwsUbBbltjR/OCxABWSQYTaPnPtjnqKAOgXxbchyj6fYSWQSNI7J1kMUfllmUj592cu55Y53tnOaxbu6mvbye7uHMk88jSSOerMxyT+ZqGigAooooAtw6hPBEsaSEKOgBqT+1rr/nq351QooAv/wBrXX/PVvzo/ta6/wCerfnVCigC/wD2tdf89W/Oj+1rr/nq351QooAs3F9NcxhJHLAHPJqtRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRWvp+jQX2harqB1BUuLCJZRaiIkupljjJLcBRmUY6k7TwODWu/ghftU1vHqW57K4a31ItBgW7LFJIxT5jvAEMo/h5X3zQByNFdQfCtnHbjUZ9Vkj0mRIWhuBagyM0jSKA0e/AwYZcncfu8ZzWBqNjNpep3en3GPPtZngk2nI3KxU4/EUAVqKKKACitix0d7qzjmGMNnv6EirH/CPye350Ac/RXQf8I/J7fnR/wj8nt+dAHP0V0H/CPye350f8I/J7fnQBz9FampaW1lbrK2MFwv6H/CsugAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAv6fqX2Gy1W38rf9vtVt927Hl4milz05/1WMcdc9sHffxuv2qa4j03a97cNcakGnyLhmikjYJ8o2AiaU/xct7YrkaKAOoPiqzkt106fSpJNJjSFYbcXQEitG0jAtJswcmaXI2j73GMVi3GsXtxrlxrCzvBezzvOZYWKFXcknBHI6mqNFACszO5d2LMxySTkk0lFFAFyC/lghWNWIAqT+1J/wC+az6KAND+1J/75o/tSf8Avms+igDQ/tSf++aP7Un/AL5rPooAtXN7JcxhHYkA5qrRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFAGnaaBf31jJeW/2R4oonmdTeQrIETO4+WW3dvTnjGcipX8LazG9sjWeXuH2IqyoxVtu7DgH92dpzhscZPY02C8t7Pw3cwwyZvb6URygKR5cCYbGeh3vtPHTyveuubxLo0eoapcLfb01u8luHCxPmyWS3uI8NkDJDXP8ADniP3AoA4W/0660y4EF0iq5QOpSRXV1PQqykhh7g1VrZ1+6tZU0uztbhblLCz+ztOisqyMZZJCQGAOB5m3kD7tZjzxtaRwi1hSRCS06l9757EFtuB7AUAQ0UUUAatno811aJMqkq2f54qf8A4R+f+4ags9YmtbRIVYhVz/PNT/8ACQT/AN80AH/CPz/3DR/wj8/9w0f8JBP/AHzR/wAJBP8A3zQAf8I/P/cNH/CPz/3DR/wkE/8AfNH/AAkE/wDfNAFXUNLlsoFkdSAW2/of8Kza0tQ1SW9gWN2JAbd+h/xrNoAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigDQTRrx9L/tH/R0tju2mS6iR229dqFgzdewNWH8LazG9sjWeXuH2IqyoxVtu7DgH92dpzhscZPY1qWWqaafDCWt7PZSLFbXEYtjZn7R5rbjGyy4IChipPzLwCMHithvEujR6hqlwt9vTW7yW4cLE+bJZLe4jw2QMkNc/w54j9wKAOFv9OutMuBBdIquUDqUkV1dT0KspIYe4NVa2dfurWVNLs7W4W5Sws/s7TorKsjGWSQkBgDgeZt5A+7WY88bWkcItYUkQktOpfe+exBbbgewFAENFFFAHa6D4C1DWtFt9QggZo5d2CB6MV/pWl/wq7Vf+fZ/++a3fA3xXn8NeDrDSECbbfzMZUH70jN/7NXQ/8L0uvSP/AL5FAHA/8Ku1X/n2f/vmj/hV2q/8+z/98133/C9Lr0j/AO+RR/wvS69I/wDvkUAcD/wq7Vf+fZ/++aP+FXar/wA+z/8AfNd9/wAL0uvSP/vkUf8AC9Lr0j/75FAHk3ijwZe+HdMju7mFkR5hECR3IY/0rk69T+JXxJm8Y+HLfT5AuI7tZ+AB0R1/9mryygAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAUMwGAaXe3qaQKSMgUuxvSgA3t6mje3qaNjelGxvSgA3t6mje3qaNjelGxvSgBCxPU0lKVI6ikoAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKAOi8N+HYNVtr66vppIYYrW5a2EeMzTxQPLjn+EBRuP+0o43ZGu/g3TGvb63jmu1/si6kt9QZnU+bshmlZoxtG3/j3kAB3dVPqK5/R/FWtaErR2OoXKQNFNH5HnOIx5iMhYKCBuG7IPqAe1Wn8aai06TLb2iSNKZroqjf6W5QoTJlj1V3BC7fvt3NAFHW9PtrP+z7my84Wt/a/aI0mYM8eJHjZSwAB+aNiDgcEVQezuY7SO7e2mW2lJWOZkIRyOoDdCRVjU9Tk1SaFmhhgigiEMMEIbZGgJOBuJPJZiSSeSao0AFFFFAHVaPosN3pUM7yorNuyCfRiKvf8I7b/APPaP865SDUJYIVjViAM/wA6k/tWb++aAOn/AOEdt/8AntH+dH/CO2//AD2j/OuY/tWb++aP7Vm/vmgDp/8AhHbf/ntH+dH/AAjtv/z2j/OuY/tWb++aP7Vm/vmgDS1/SorGwSVJEYmULgH2P+Fc5Vu6vZLmMI7EgHNVKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAopwRiMgUeW3pQA2ineW3pR5belADaKd5belHlt6UANopSpAyRSUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFAHRaLpGmanpV8zvcLd21tLcM4mRVXauUUR4LyZIwSCAoOTwOdd/BumNe31vHNdr/AGRdSW+oMzqfN2QzSs0Y2jb/AMe8gAO7qp9RXPWviK4tNOFqlramVYZbeO6ZW82OKQEOgw20ghnGSpI3HBHGLb+NNRadJlt7RJGlM10VRv8AS3KFCZMsequ4IXb99u5oAo63p9tZ/wBn3Nl5wtb+1+0RpMwZ48SPGylgAD80bEHA4Iqg9ncx2kd29tMttKSsczIQjkdQG6EirGp6nJqk0LNDDBFBEIYYIQ2yNAScDcSeSzEkk8k1RoAKKKKAOv0TQ4LzSIJ3lRWbdkE+jEf0rQ/4Rq2/57R/nXH22py29usSsQFz/Opf7Zn/AL5oA6v/AIRq2/57R/nR/wAI1bf89o/zrlP7Zn/vmj+2Z/75oA6v/hGrb/ntH+dH/CNW3/PaP865T+2Z/wC+aP7Zn/vmgDT8R6RDYafHLHIjEyhcA+x/wrmKu3eoSXUQR2JAbNUqACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKALMOnXtxaTXcNncSW0H+tmSJikf+8wGB+NPfSdSi+y+Zp92n2vH2bdCw87OMbOPm6jp6iug0XzrXwvd34vbR3Edxb29m95EjRiRAs0pRmDNlPlUAEk8/wgHpG1Swi1TVZm1G0Meq38s2nss6t9nRra4RC+D+6w00I+bBHlnsKAPNbq0ubG5e2u7eW3nTG6KZCjLkZ5B5qGt3xI8YTR7QTRTXFnYCG4eGRZF3+bI4AZSQ2EdF4JHGO1ZDtbG0jVIZhcgnzJGlBRh2AXaCD77j+FAENFFFAGpaaTLdWqTKpIbP88VN/YM/9w/lW/4f1+3stDt7d4kZk3ZJHPLE/wBa0v8AhKbX/nhF+VAHHf2DP/cP5Uf2DP8A3D+Vdj/wlNr/AM8Ivyo/4Sm1/wCeEX5UAcd/YM/9w/lR/YM/9w/lXY/8JTa/88Ivyo/4Sm1/54RflQBwl/pklnAsjqQC239D/hWfXX+Ktah1HS4oY40UrMGyo/2WH9a5FVLNgdaAEoqTyH9KPIf0oAjoqTyH9KPIf0oAjoqTyH9KPIf0oAjopSMEg9qSgAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigB6yMowDxR5z+tMooAf5z+tHnP60yigB/nP60ec/rTKKAHNIzDBNaXh21S9122t3YKr7sk+yk/0rLqezuGtbtJlOCuf5YoA9H/4Ra1/57R/nR/wi1r/z2j/OuO/t+f8Avn86P7fn/vn86AOx/wCEWtf+e0f50f8ACLWv/PaP8647+35/75/Oj+35/wC+fzoA7H/hFrX/AJ7R/nR/wi1r/wA9o/zrjv7fn/vn86P7fn/vn86AMzUohBqt5CpBEc7qCPZiKrYJ7U+eQy3Ekh6uxY/ia6DS9AkvdNiuFUkPn9CR/SgDnNp9KNp9K6//AIRab+4fyo/4Rab+4fyoA5DafSjafSuv/wCEWm/uH8qP+EWm/uH8qAOQ2n0oIIrr/wDhFpv7h/Ks3W9Fk06zSZlIBkC/oT/SgDB60u0+lW9LtTe6lFbqMl8/oCf6V0n/AAi039w/lQByG0+lG0+ldf8A8ItN/cP5Uf8ACLTf3D+VAHIbT6UbT6V1/wDwi039w/lR/wAItN/cP5UAchtPpSV2H/CLTf3D+VcrdxGC8nhPWORl/I4oAiwT2o2n0ro9L0CS902K4VSQ+f0JH9Kuf8ItN/cP5UAchtPpRtPpXX/8ItN/cP5Uf8ItN/cP5UAchtPpRtPpXX/8ItN/cP5Uf8ItN/cP5UAcgQRSVva3osmnWaTMpAMgX9Cf6Vg0AFFFFABRWvYaGuoafc3Ueq2Mb20DzyW8gmDhVOByIymSSoHzdWA4qPVtHGklY31G0nuOPMghEu6PIzyWQLxnHBPNAGZRRRQAUUUUAOCMRkA4o8t/7pr0Dwz4Nm1bw/a3qxkrLv5x6OR/Stb/AIV5P/zxP5UAeU+W/wDdNHlv/dNerf8ACvJ/+eJ/Kj/hXk//ADxP5UAeU+W/900eW/8AdNerf8K8n/54n8qP+FeT/wDPE/lQB5QVZRkjFJXb+MfCsuh6RFcvGVDTiPOPVWP9K4igAooooAKKKKACiiigArpNL1+Sy02K3ViAmf1JP9a5ulyR3oA6/wD4Smb++fzo/wCEpm/vn865DcfWjcfWgDr/APhKZv75/Oj/AISmb++fzrkNx9aNx9aAOv8A+Epm/vn86zdb1qTUbNIWYkCQN+hH9awtx9aCSaALel3RstSiuFOCmf1BH9a6T/hKZv75/OuP6Uu4+tAHX/8ACUzf3z+dH/CUzf3z+dchuPrRuPrQB1//AAlM398/nR/wlM398/nXIbj60bj60Adf/wAJTN/fP51yt3KZ7yeY9ZJGb8zmotx9aSgDpNL1+Sy02K3ViAmf1JP9auf8JTN/fP51yGSO9G4+tAHX/wDCUzf3z+dH/CUzf3z+dchuPrRuPrQB1/8AwlM398/nR/wlM398/nXIbj60bj60Abut61JqNmkLMSBIG/Qj+tYNKSTSUAFFFFAGkmoQweHZLCFXFxczh7hyBjy0HyKDnPLFif8AdT0rX8ReJbfVtLNuLi/u5XuEnU3iIPsoCMGjjKnkMWBJwv3BxkmuWooAltrh7W4SeNYmdDkCWJZF/FWBB/EVGzF3LHGScnAAH5DpSUUAFFFFAHoXhnxlNpPh+1slkIWLfxn1cn+ta3/Cw5/+ex/OvKQ7AYBOKPMf+8aAPVv+Fhz/APPY/nR/wsOf/nsfzrynzH/vGjzH/vGgD1b/AIWHP/z2P50f8LDn/wCex/OvKfMf+8aPMf8AvGgDtfGPiqXXNIitnkLBZxJjPorD+tcRSlmYYJzSUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQB//Z",
            "image/png": "iVBORw0KGgoAAAANSUhEUgAAAWgAAAJYCAIAAADAFeuyAAAm90lEQVR4Ae3dbUxU2f3A8YI8LAKCLSOMGLoJSDRdmrRvSDeTNtoVE123K2uTSRpCfChvzL6wREpT6ovuvLB0mTiaZnVtlrCm6+Nm1ifCmpj8V22M69bGShttNbZqGBBUUFBEHv6/7ExmgCvsnZkzd84MX16QM2fO/Z3f/czwY+6dmXu+9S1+EEAAAQQQQAABBBBAAAEEEEAAAQQQQACBOSiQouc+p6RomtjExISeYmSFgJUCaVZOZn6u1NRU84OtHDk2NmbldMyFgJ4CocKxadOmjRs35ubmNjQ0nDlzJi8vr62traCgoK+vr7a2dmBgILKeyHY7LS2UWGQRYrQVhSNGsIRNLIHAEYHNZjt27NiKFSvKy8u9Xu/y5cubm5u7u7vdbnd9fb3c29jYGFlPZBwLFiyIbMNYb/X48eNYT0F8BPQXCBSOZcuWVVRUHD16NDs7+/bt24sWLers7Kyqqurq6iouLu7o6JB7I+uJjKCoqCiyDWO9lRTTWE9BfAT0FwgcEVz/+kfS3bBhw8mTJ6VRWFjo/yPx+XzSjrhHfwIyRACBcAWmnEooLS3dvn27HLCEG+Ubx7tcLofD4fF45DjoGwfLgKdPn5oZxhgEEIiLQKhw5OTkHD58ePPmzb29vZJKT0+PHC/IoYrdbpd2xD3+vWpqagpr954/fx7WeAYjgICVAoF3PeVzE/IeSktLy6VLl/zTt7e3O51OactvaUsjsp7IdmZU15/IdoetEEhOAXkjdnBw8P++/jl16pTspLz5+tlnn124cEF+Szvinsi8pJDp+RPZ7rAVAggggAACCCCAAAIIIIAAAggggAACCCCAAAIIIIAAAggggAACCCCAAAIIIIAAAggggAACCCCAAAIIIIAAAggggMBLBTS9mLh8w+2l6ca9k6ucx/0hIAEdBELX49Ahm2AO2l6s+MWLF8EkaSAwZwVChSM/P3/Xrl3r168Pfok+jlc5T09P1/MhoXDo+biQlcUCoeVL5FKjf/vb34IvxX/729+eO3dOrvd3/vz53/zmN5JWZD2R7c8ruv5EtjtshUCSCYROJciFAuXqxP39/fLSQ3YysmuaG7eKzGvJkiWRbRjrre7duxfrKYiPgP4CoVcc0y78z1XO9X/wyBCBeAmEznHENAOuch5TXoIjYLHAjIUjvlc5Hx4ethiC6RBAwLxA6FBl2jaRXdPcuNW0sCZvypsXev6YzJ9hCMwtATk56t9hrnL+0susz61nA3uLAAIIIIAAAggggAACCCCAAAIIIIAAAggggAACCCCAAAIIIIAAAggggAACCCCAAAIIIIAAAggggAACCCCAAAIIIIAAAggggAACCCCAAAIIIIAAAggggAACCCCAAAIIIIAAAggggAACCCAwJwRCK7lptbupqTNefj2+eY6Pj8c3AWZHQAeBGddViW9y2q5WPzIyEl8ZZkdABwFNC0dGRoYOOsYcKBxGE3rmoECgcBw6dEgWnZb9l7/Y8vLygoICWVelra1NGn19fbW1tQMDA2Z6VAnKYvWqQqmNMzg4qDYg0RBIRIHp5zi2bNlSUlKyY8eO5uZmWYba7XbX19fbbLbGxkYzPaoIvvvd76oKpTbO//73P7UBiYZAIgpMKRyydtmVK1dWr159//79zs7Oqqqqrq6u4uLijo6OiooKMz2qCCgcqiSJg0AsBKac41i3bt3ly5elashMhYWF8opDGj6fT9ome1SlODQ0pCoUcRBAQLnAlMIhRyV1dXWq5nC5XA6Hw+PxeL3ecGOyWn24YoxHwEqBUOGorKyUFadv3Ljhn76np0dOl8qhit1ul7Z0mumZnHpTU9Pkm2G1efMiLC4GI2CxQOhzVg0NDS0tLcHp29vbnU6n3JTf0paGmZ7g5lE2RnX9iXK/2ByBpBIoKyu7ePHi5F2SN18/++yzCxcuyG9py11meiZHoI0AAggggAACCCCAAAIIIIAAAggggAACCCCAAAIIIIAAAggggAACCCCAAAIIIIAAAggggAACCCCAAAIIIIAAAggggAACCCCAAAIIIIAAAggggAACCCCAAAIIIIAAAggggAACCCCAAAJxEJiyIFMc5p9hSlkaaoZ74tw9MTER5wyYHgENBELLI2iQTCgFbVerf/HiRShLWgjMVQFNC4e2i05TOObqXwr7PUVA08LhX5BhSqZ63Hjy5IkeiZAFAvEU0LRwyPJx8VSZee579+7NfCf3IDBXBAKFY9WqVfv27btz547styzCJKs3yv/8tra2goKCvr6+2tragYEBMz2q2EpLS1WFUhtHFuVWG5BoCCSiQODNi5qamuzs7L179wb3obm5WVard7vdshK1zWZrbGw00xPcPMrG73//+ygjxGjzHTt2xCgyYRFIIIHA2rFyaODz+SbnvWbNmkOHDkmP/F67dq00zPRMjkAbAQSSVSBwqCKFQ5aP3b59+8OHD7dt23br1q3CwkJ5xSG7LQVF2tIw06OK6V//+peqUMRBAAHlAoHCIZ9runr1al1dXXV19f79+1euXBn9TC6Xy+FweDwer9cbbjSpXOFuwngEELBMIFA4du/efffuXZn1+PHjcpZUGj09PUVFRV1dXfJiRNomeybnLWdYJ98Mqz3tuCmsbRmMAAKxFggUjp07dx48eFCqRmVl5bVr12TW9vZ2p9MpJ0flt7RN9qhKV97EURWKOAggoFwg8K7K0qVLW1tbR0dHh4eHt27dKkcKZt58NY5RlV96erqqUGrj8MlRtZ5ES1ABTb9LxpfcEvT5RNoIIIAAAggggAACCCCAAAIIIIAAAggggAACCCCAAAIIIIAAAggggAACCCCAAAIIIIAAAggggAACCCCAAAIIIIAAAggggEBsBfh2bHi+LAEZnhejk1QgcCEf3fYuNTVwFWXdEhsbG9MtJfJBwHoBTQuHthfyoXBY/xxlRg0FNC0cmZmZGmJJSnKFND0TIysErBSYXjhk8ZRPP/00KyvLeFlAMz2qUpcEVIVSG4eLoar1JFqCCkw5OZqbm/v555+/9tprCxYsMLNum3GMKgVZ5EVVKLVxbt68qTYg0RBIRIEp5yDlWue7du0aHx+XPTGzbptxTCISkDMCCIQrEDpUkcWTFi9efOTIkQ8//FCimFm3zTgm3OlnGj84ODjTXfQjgEDcBQKFQ05Gvv/++7KMm8KEolnJ7dmzZwozIRQCCKgVCBSOd955R05wfPLJJxI9Jyfn448/ju9Kbs+fP1e7n0RDAIHYCvT398sEcuLzV7/6lTTq6+v/8Ic/mOyRYUp+5un6o2TvCIJAogtMeVfFvzNSOPLz8828+Woco4qDBZlUSRIHAQQQQAABBBBAAAEEEEAAAQQQQAABBBBAAAEEEEAAAQQQQAABBBBAAAEEEEAAAQQQQAABBBBAAAEEEEAAAQQQQAABBBBAAAEEEEAAAQQQQAABBBBAAAEEEEAAAQQQQAABBBBAAAEEEEAAAQQQQAAB5QIvucq58jkiCJiaOmVtyggixGgT//qYMQpOWAQSRSC0BKRWGVM4tHo4SAaBaQKaFo709PRpiWpyc3R0VJNMSAOBOApoWjheeeWVOKLMMjWL2s6Cw11zRyBQOObPn9/a2iqrz8sKsjt27Dh9+rRxlTYzPargsrKyVIVSG+fRo0dqAxINgUQUCJwcbWhokNN+smC93W6/ePHiq6++KmvHdnd3u91uWTvWZrM1Njaa6VFFUF5eriqU2jj//ve/1QYkGgKJKBAoHAsXLhwaGhoZGXnjjTc++OCDpUuXdnZ2VlVVdXV1FRcXd3R0VFRUmOlRRUDhUCVJHARiIRA4VPG/Aj9w4EB1dfVbb70lM8lhi7zikIbP55O2yR5VKQ4ODqoKRRwEEFAuMOXkaE1NzbFjx2pra8+ePRv9TC6Xy+FweDwer9cbbjTOQYYrxngErBQIFI49e/Zs27ZN3ms8derURx99JBn09PQUFRXJoYqc9ZC2yZ7JqTc1NU2+GVZ7eHg4rPEMRgABKwUCH9CUd0zefvttmfj111+/ceOGNNrb251OpzTkt7RN9sgwJT8vdP1RsncEQSDRBQInR0tKStra2uTzmnJ+9N13371+/bqZN1+NY1Rx8MlRVZLEQQABBBBAAAEEEEAAAQQQQAABBBBAAAEEEEAAAQQQQAABBBBAAAEEEEAAAQQQQAABBBBAAAEEEEAAAQQQQAABBBBAAAEEEEAAAQQQQAABBBBAAAEEEEAAAQQQQAABBBBAAAEEEEAAAQQQQAABBBBAAAEEEEAAAQQQQAABBBBAAAEEEEAAAQQQQAABBBBAAAEEEEAAAQQQQAABBBBAAAEEEEAAAQQQQAABBBBAAAEEEEAAAQQQQAABBBBAAAEEEEAAAQQQQAABBBBAAAEEEEAAAQQQQAABBBBAAAEEEEAAAQQQQCDOAilxnn+G6VNSNE1sYmJihpTpRmAOCaTpua8UDj0fF7JCwC+gaeFIS9M0sZGREZ46CCCg6d9nenq6no8NhUPPx4WsLBYIFY5NmzZt3LgxNze3oaHhzJkzeXl5bW1tBQUFfX19tbW1AwMDZnpUZf/KK6+oCqU2ztDQkNqAREMgEQUC5yBtNtuxY8dWrFhRXl7u9XqXL1/e3Nzc3d3tdrvr6+vl3sbGRjM9qghKSkpUhVIb586dO2oDEg2BRBQIFI5ly5ZVVFQcPXo0Ozv79u3bixYt6uzsrKqq6urqKi4u7ujokHvN9Kgi+N73vqcqlNo4//znP9UGJBoCiSgQOFS5/vWP7MCGDRtOnjwpjcLCQnnFIQ2fzydtkz0yjB8EEEh6gdA5DtnV0tLS7du3ywGLkt12uVwOh8Pj8cixT7gBnzx5Eu4mjEcAAcsEQoUjJyfn8OHDmzdv7u3tlel7enqKiorkUMVut0vbZM/kvJuamibfDKvNOciwuBiMgMUCgcIhH7iS91BaWlouXbrkz6C9vd3pdMrJUfktbek006Mq++HhYVWhiIMAAsoFAidH5Y3YPXv2fPXVVzLB4ODgm2++aebNV+MYVfllZGSoCqU2Dp/jUOtJtAQVCBQO3bJPTU3VLSV/PuPj43omRlYIWCkQOsdh5azfOBffJftGIgYggAACCCCAAAIIIIAAAggggAACCCCAAAIIIIAAAggggAACCCCAAAIIIIAAAggggAACCCCAAAIIIIAAAggggAACCCCAAAIIIIAAAggggAACCCCAAAIIIIAAAggggAACCCCAAAIIIIAAAggggAACCCCAAAIIIIAAAggggAACCCCAAAIIIIAAAggggAACCCCAAAIIIIBAUgloulp9SoqmibEadlI9/dmZSAU0Xa0+NTU10j2K7XZjY2OxnYDoCCSCgKaFIy1N08QoHInwrCbHmAto+vc5f/78mO96RBM8f/48ou3YCIGkEtC0cOTl5enJ/OjRIz0TIysErBTQtHDY7XYrFczP9d///tf8YEYikKwCmhaOsrIyPcUvXryoZ2JkhYCVAqHCkZ+fv2vXrvXr1/sPE1atWrVv3747d+5INhcuXGhqapL+tra2goKCvr6+2tragYEBY4+q1L///e+rCkUcBBBQLhD6uMT58+ePHDny3nvvSQWRaWpqarKzs/fu3Rucsrm5ubu72+1219fX22y2xsZGY09wcJSNL774IsoIMdr8Jz/5SYwiExaBBBIIfVzi5z//+Z49e4Kpy1kGn88XvCmNNWvWHDp0SBrye+3atS/tmTyeNgIIJKtA6FBFXk1M3kkpHHKiYfv27Q8fPty2bdutW7cKCwv9Y6SgSFsGG3smR4im/eWXX0azOdsigEBMBUKFY9o08tnqq1ev1tXVVVdX79+/f+XKldMGfONNl8vlcDg8Ho/X6/3GwdMG/OMf/5jWw00EENBHYMbCsXv37rt370qix48fl7Ok0ujp6SkqKurq6pIXI9J+ac/kHZPzqZNvhtW+efNmWOMZjAACVgrMWDh27tx58OBBqRqVlZXXrl2TnNrb251Op5wcld/SfmmPqtSnnV5RFZY4CCCgRCD0roo/XH9/v/9dlaVLl7a2to6Ojg4PD2/dulXOcRjffDX2KMlJgixcuFBVKLVx+OSoWk+iJajA9MKhyW5kZmZqksm0NPiuyjQQbs5NgRkPVeLLIa904psAsyOAwCwCmr7i4EI+szxm3IUAAggggAACCCCAAAIIIIAAAggggAACCCCAAAIIIIAAAggggAACCCCAAAIIIIAAAggggAACCCCAAAIIIIAAAggggAACCCCAAAIIIIAAAggggAACCCCAAAIIIIAAAggggAACCCCAAAIIIIAAAggggAACCCCAAAIIIIAAAggggAACCCCAAAIIIIAAAggggAACCCCAAAIIIIAAAggggAACCCCAAAIIIIAAAggggAACCCCAAAIIIIAAAggggAACCCCAAAIIIIAAAggggAACCCCAAAIIIIAAAggggAACCCCAAAIIIIAAAggggAACCCCAAAIIIIAAAggggAACCCCAAAIIIIAAAskmkKLnDqWkaJrYxMSEnmJkhYCVAmlWTmZ+rtTUVPODrRw5NjZm5XTMhYCeApoWjnnz5unpReHQ83EhK4sFNC0cGRkZFkOYnG5kZMTkSIYhkMQCmhaOrKwsPdEHBwf1TIysELBSIFQ48vPzd+3atX79+ry8PMlAfre1tRUUFPT19dXW1g4MDJjpUZV6bm6uqlBq4/T29qoNSDQEElEg9ObF+fPnjxw58t5770kFkT1pbm7u7u52u9319fU2m62xsdFMjyqCH/zgB6pCqY3z97//XW1AoiGQiAKhwlFUVCSVor+/3184Ojs7q6qqurq6iouLOzo6KioqzPSoIvjpT3+qKpTaOGfPnlUbkGgIJKJA6F1PqRqTd6CwsNDf4/P5pC13memZHIE2Aggkq0DoHIfyPXS5XA6Hw+PxeL3ecIM/fPgw3E0YjwAClgnMWDh6enrk4EUOVex2u7QlITM9k/NuamqafDOstpyLDWs8gxFAwEqBGQtHe3u70+mUk6PyW9qSk5keVak/efJEVSjiIICAcoHQyVF/6ODJUTNvvhrHqMovJydHVSi1cfgch1pPoiWowPTCoclu8MlRTR4I0kDgpQIzHqq8dLRlnXwlxDJqJkIgAgFNX3HwtfoIHks2QQABBBBAAAEEEEAAAQQQQAABBBBAAAEEEEAAAQQQQAABBBBAAAEEEEAAAQQQQAABBBBAAAEEEEAAAQQQQAABBBBAAAEEEEAAAQQQQAABBBBAAAEEEEAAAQQQQAABBBBAAAEEEEAAAQQQQAABBBBAAAEEEEAAAQQQQAABBBBAAAEEEEAAAQQQQAABBBBAAAEEEEAAAQQQQAABBBBAAAEEEEAAAQQQQAABBBBAAAEEEEAAAQQQQAABBBBAAAEEEEAAAQQQQAABBBBAAAEEEEAAAQQQQAABBBBAIHyBlPA3sWKL1NRUK6YJf47x8fHwN2ILBJJNIE3PHaJw6Pm4kBUCfgFNC0d6erqej9Do6KieiZEVAlYKaFo4MjMzrVQwP9ezZ8/MD2YkAskqoGnhmD9/vp7i/f39eiZGVghYKaBp4ViwYIGVCubn6urqMj+YkQgkq0CgcBw6dKioqEh2MiMjo7y8vKCgIC8vr62tTRp9fX21tbUDAwPGntih5Ofnxy44kRFAIEqB6W/HbtmypaSkZMeOHc3Nzd3d3W63u76+3mazNTY2GnuinHuWzdeuXTvLvXG86/Tp03GcnakR0ERgSuFISUm5cuXK6tWr79+/39nZWVVVJa/Mi4uLOzo6KioqjD2x2wcKR+xsiYxA9AJTznGsW7fu8uXLUjUkbmFhobzikIbP55P2S3uin36mCA8ePJjpLvoRQCDuAlMKhxyV1NXVRZOTy+VyOBwej8fr9UYThzcvotFjWwRiLRAqHJWVlfLneuPGDf+UPT09crpUDlXsdru0pdPYY0yuqanJ2BlBz+PHjyPYik0QQMAagVDhaGhoaGlpCc7a3t7udDrl5Kj8lrb0G3uCg5U3nj59qjwmARFAQJVA4ORoWVnZgQMHfvSjHwXjGt98NfYEBytvZGVlKY+pJCCfHFXCSJBEF5jyroo+O5OWFnoppE9WkgnfVdHq4SCZeAlo+vfJt9fj9YRgXgQQQAABBBBAAAEEEEAAAQQQQAABBBBAAAEEEEAAAQQQQAABBBBAAAEEEEAAAQQQQAABBBBAAAEEEEAAAQQQQAABBBBAAAEEEEAAAQQQQAABBBBAAAEEEEAAAQQQQAABBBBAAAEEEEAAAQQQQAABBBBAAAEEEEAAAQQQQAABBBBAAAEEEEAAAQQQQAABBBBAAAEEEEAAAQQQQAABBBBAAAEEEEAgIJCip0RKiqaJTUxM6ClGVghYKZBm5WTm55o3b575wVaOHB0dtXI65kJATwFNC0dGRoaeXhQOPR8XsrJYQNPCkZ2dbTGEyemePn1qciTDEEhiAU0LR35+vp7ovb29eiZGVghYKaBp4bDb7VYqmJ/rP//5j/nBjEQgWQU0LRylpaV6ip87d07PxMgKASsFAoVj1apV+/btu3Pnjsx94cKFpqamvLy8tra2goKCvr6+2tragYEBY0/sEq2oqIhdcCIjgECUAoGPS9TU1Mj5yL179wbDNTc3d3d3u93u+vp6m83W2Nho7AkOVt6Q4qU8ppKADodDSRyCIJAMAg0NDT/72c8m70lnZ+fixYulp7i4+Nq1a9Iw9kwer7Ytn7PS80ftbhINgQQVSPXnLScj165dK//nT5w44T+/UFhYKK845F6fzydtaRh7EnSfSRsBBKIUCJzjkH/vV69eraurq66u3r9//8qVKyOL63K55MW8x+Pxer2RRfBv9de//jWazdkWAQRiKhAoHLt37757967MdPz4cTlLKo2enp6ioqKuri55MSLtl/YYM5OzqsbOCHq+/PLLCLZiEwQQsEYgUDh27tx58OBBqRqVlZX+Mxrt7e1Op1NOjspvaUs2xp7YpejPIXbxiYwAAtEIBN5VWbp0aWtrq3wRY3h4eOvWrbdu3TK++WrsiWbi2bf98Y9/PPuAeN3L5zjiJc+8Wglo+u11KWRaMQWT4ZOjQQoac1kgcKiiG0F/f79uKZEPAggEBTQtHENDQ8EUaSCAgG4CmhaOkZER3aTIBwEEggKanuPg0oHBR4gGAggggAACCCCAAAIIIIAAAggggAACCCCAAAIIIIAAAggggAACCCCAAAIIIIAAAggggAACCCAQNwE+ch4evVxjMbwNGI1AMgpo+iU3VqtPxicb+5Q8ApoWjrQ0TRNjtfrkee6zJ1EIaPr3mZGREcVOxXBTubRiDKMTGoEEEdC0cGRlZekJ+PjxYz0TIysErBTQtHDk5uZaqWB+Lv9KEebHMxKBpBSYXjjWrFnz6aefyj984zXNjT2xE5G5YhecyAggEKXAlLdj5f/8559//tprry1YsMC4xLSxJ8q5Z9m8qqpqlnvjeNeZM2fiODtTI6CJwJTC8ac//emLL7748MMP8/PzZYlp+euVldxk0emOjo6KigpjT+z24Re/+EXsgkcT+S9/+Us0m7MtAskhEFh0WnZG1nyV5emPHDni3zHjEtPGnuQgYC8QQCBcgcA5jszMzPfff19WnA53+2njVS063dvbOy0yNxFAQB+BQOF455135ATHJ598Ipnl5OR8/PHH8V10+sGDB/oYkQkCCEwTmHKOw3+frKIm5zgmnwpdtGjRr3/9a2PPtFgKb5aVlSmMpjDUzZs3FUYjFAIJKjBj4TC++Wrsid0+y/mU2AWPJjKf44hGj22TRuAlhUOHfZP3g3VIw5gDnxw1mtAzBwWmfwBMEwKWgNTkgSANBF4qoGnh4EuoL3206ERAEwFND1VYO1aT5wdpIIAAAggggAACCCCAAAIIIIAAAggggAACCCCAAAIIIIAAAggggAACCCCAAAIIIIAAAggggAACCCCAAAIIIIAAAggggAACCCCAAAIJLqDp1+pTU0PrNmglPD4+rlU+JINAXAQ0vZDPvHnz4sLxjZNSOL6RiAFzQUDTwpGenq6n/osXL/RMjKwQsFJA08IhC0RZqWB+rqdPn5ofzEgEklVA08KRnZ2tp/ijR4/0TIysELBSQNPCoe3yCFY+NsyFgLYCgcIxf/781tZWWQZJFoLcsWPH6dOnjcsvGXtit1cLFy6MXXAiI4BAlAKBt2MbGhrk/QJZd9put1+8ePHVV1+dvOCjzWZrbGw09kQ59yybv/XWW7PcG8e7Tpw4EcfZmRoBTQQChUP+ww8NDckySG+88cYHH3ywdOnSzs7Oqqqqrq6u4uLijo6OiooKY0/s9mHLli2xCx5N5D//+c/RbM62CCSHQOBQxX/O78CBA9XV1f7/9nLY0t3dLTvp8/n8K7kae5KDgL1AAIFwBaacHK2pqTl27Fhtbe3Zs2fDDeQf73K5HA6Hx+Pxer2RRfBvdf/+/Wg2Z1sEEIipQKBw7NmzZ9u2bbLw4qlTpz766COZUpZlLyoqkkMVOevhX6Ld2GPMrKmpydgZQc+DBw8i2IpNEEDAGoFA4ZB3TN5++215ufH666/fuHFD5m5vb3c6nW63W35L+6U9sUuRj0vEzpbICEQvEDg5WlJS0tbWJl8tk/Oj77777vXr141vvhp7op9+pghLliyZ6a749t+7dy++CTA7AjoIBAqHDqlMzkHbz3HwUmjyw0R7zgpMOTmqj8Lz58/1SYZMEEBgmoCmhYMvoU57nLiJgFYCmhaOsbExrZhIBgEEEEAAAQQQQAABBBBAAAEEEEAAAQQQQAABBBBAAAEEEEAAAQQQQAABBBBAAAEEEEAAAQQQQAABBBBAAAEEEEAAAQQQQAABBBBAAAEEEEAAAQQQQAABBBBAAAFFApperDglRdPEJiYmFMkTBoEEFtD00oHz5s3TE1XWrNIzMbJCwEoBTQtHWpqmiVE4rHx2Mpe2Apr+fWZmZupJNjw8rGdiZIWAlQKaFo6srCwrFczPNTAwYH4wIxFIVgFNC0dOTk6yirNfCCSBgKaFIz8/Pwlw2QUEklUgVDg2bdq0cePG3NzchoaGM2fOGJeYNvbEDuU73/lO7IITGQEEohQIfFzCZrMdO3ZsxYoV5eXlXq93+fLlzc3N3d3dbre7vr5e7m1sbDT2RDn3LJvX1NTMcm8c7zpw4EAcZ2dqBDQRCBSOZcuWVVRUHD16NDs7+/bt24sWLers7Kyqqurq6iouLu7o6JB7jT2x24ft27fHLng0kf/4xz9GsznbIpAcAoFDletf/8gubdiw4eTJk9IoLCyUVxzS8Pl80n5pj3TygwACc1AgdI5Ddr60tFT+1csBS8QQLpfL4XB4PB453ok4iGzor1nRRGBbBBCInUCocMg7oIcPH968eXNvb6/M19PTU1RUJIcqdrtd2i/tMabV1NRk7Iyg5/79+xFsFd9NUlNT45vATLOPj4/PdBf9CEQmECgc8qWytra2lpaWS5cu+QO1t7c7nU45OSq/pS2dxp7IpjSz1YMHD8wM02qMtt+voXBo9TxJjmQCJ0fljdg9e/Z89dVXsleDg4Nvvvmm8c1XY0/sCMrKymIXPJrIN2/enGnz+fPnz3RXfPufPn0a3wSYPfkEAoVDtx2ToyTdUvLnM8vJl4ULF+qZ86NHj/RMjKwSVyB0jkOrfXj27JlW+ZhJRttXHBQOMw8fY8IS0LRwPH/+PKzd0GGwfOhWhzTIAQELBDQtHIl42Qu+X2PB85UpNBHQtHCMjY1pAmQ+Db5fY96KkYkuoGnhSMRLe8o3ehL92UD+CJgU0LRwmMxeq2H+D+ZrlRLJIBAjAQpHjGATI2wiftqVnBU+tyL+cCCFQ9mj4P9gvrJwlgRKxE+7krPCpwaFQyFmhKH83/GJcOM4babt1eRfvHgxEwk5zyQTQf8szrNH4xXH7D5h3JuI36/JyMgIYw8tHDrLJwDJWeHjMIvz7LNQOGb3CePe/v7+MEbrMTQRryZPzgqfOxFftZ/CoexRePLkibJYVgWSC75ZNZWyechZGWUUgSgcUeBN3TQRv4S6YMGCqTuRALfIWYcHicKh7FFIxO/XJOLH5MlZ2VM2ikAUjijwpm4a8QnqqWEsvfXtb3/b0vlUTEbOKhSjjUHhiFYwuD3frwlSxLSRiN8JSsScZ38QKRyz+4Rxb8SfpQljDoYioIcAhUOPxyFOWSTiZ0/IOU5PlinTUjimcMy1Gw8fPky4XSZnHR4yCocOj0LcckjED62Rc9yeLpMmpnBMwph7zcePHyfcTpOzDg8ZhUOHRyFuOQwNDcVt7kgnJudI5VRup2nhSE9PV7mX6mIl4oc1Ztn7iL/jNEvMWN9FzrEWNhNf08KRmZlpJnvrxyRZ4RgZGbHeMMoZyTlKQCWba1o4tP0GpCxzp8RdkyCJeDV5ctbhyaNp4dB2jZJEvFrPLM+zRPy0KznP8oBadpemhSMRv8hk2WOmcKJE/LQrOSt8AkQcStPCkXyf7Y/4EWJDBDQU0LRwsNSAhs8VUkIgKBAoHIcOHfIvEC8XdCwvLy8oKFi1atW+ffvu3LkjQy9cuNDU1BTcxoKG3W63YBamQACByAQChcPpdPq337JlS0lJibSljjQ3N+/duzeyuFFutWTJkigjsDkCCMROYMqhSkpKytatW1evXi3zyf/8GzduxG7i2SP7i9fsY7gXAQTiJZA6eeJ169Zdvnz5/v370imFY+3atXKQcuLEidLS0snDaCOAwBwXmPKKo76+vq6uzi8iyz5fvXpVblZXV+/fv3/lypUzSblcLofD4fF4vF7vTGPC7fefWwl3K8YjgIA1AqHCUVlZKV9YDh6e7N69++7du5LE8ePH5SzpLNnE4rzpvXv3ZpmRuxBAIL4CocLR0NDQ0tISzGbnzp0HDx6UqiEF5dq1a8F+axo+n8+aiZgFAQQiEAgUjrKyssWLF587dy4Y4ne/+11ra+u2bduGh4d/+ctfBvutaSTiAs7WyDALAjoIpOiQhDGHH/7wh8ZOHXquXLmiQxrkgEB8BUKHKvHNY9rsiXh5uGm7wE0EklhA08KRiOuwJvGzhF1DYJqApoUjEa/yNE2WmwgksYCmhSMR12FN4mcJu4YAAggggAACCCCAAAIIIIAAAggggAACCCCAAAIIIIAAAggggAACCCCAAAIIIIAAAggggAACCCCAAAIIIIAAAggggAACCCCAAAIIIIAAAggggAACCCCAAAIIIIAAAggggAACCCCAAAIIIIAAAggggAACCCCAAAIIIIAAAggggAACCCCAAAIIIIAAAggggAACCCCAAAIIIIAAAggggAACSSDw/8TNjX1VHzoQAAAAAElFTkSuQmCC",
            "text/plain": [
              "<PIL.Image.Image image mode=RGB size=360x600>"
            ]
          },
          "execution_count": 17,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "from pydna.gel import gel\n",
        "from pydna.ladders import GeneRuler_1kb_plus\n",
        "\n",
        "bands = list_of_amplicons\n",
        "for band in bands:\n",
        "    print(band.name)\n",
        "\n",
        "gel([GeneRuler_1kb_plus, *[[band] for band in bands]])"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 18,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>Amplicon Name</th>\n",
              "      <th>5' Overhang</th>\n",
              "      <th>5' Duplicate</th>\n",
              "      <th>3' Overhang</th>\n",
              "      <th>3' Duplicate</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>Amplicon_SCO5090_140_pcr_products</td>\n",
              "      <td>CATG</td>\n",
              "      <td>No</td>\n",
              "      <td>GGGA</td>\n",
              "      <td>No</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>Amplicon_SCO5090_208_pcr_products</td>\n",
              "      <td>GGGA</td>\n",
              "      <td>No</td>\n",
              "      <td>GTCG</td>\n",
              "      <td>No</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>Amplicon_SCO5087_522_pcr_products</td>\n",
              "      <td>GTCG</td>\n",
              "      <td>No</td>\n",
              "      <td>GACC</td>\n",
              "      <td>No</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>Amplicon_SCO5090_640_pcr_products</td>\n",
              "      <td>GACC</td>\n",
              "      <td>No</td>\n",
              "      <td>CTAG</td>\n",
              "      <td>No</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "                       Amplicon Name 5' Overhang 5' Duplicate 3' Overhang  \\\n",
              "0  Amplicon_SCO5090_140_pcr_products        CATG           No        GGGA   \n",
              "1  Amplicon_SCO5090_208_pcr_products        GGGA           No        GTCG   \n",
              "2  Amplicon_SCO5087_522_pcr_products        GTCG           No        GACC   \n",
              "3  Amplicon_SCO5090_640_pcr_products        GACC           No        CTAG   \n",
              "\n",
              "  3' Duplicate  \n",
              "0           No  \n",
              "1           No  \n",
              "2           No  \n",
              "3           No  "
            ]
          },
          "execution_count": 18,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "overhangs = create_overhang_dataframe(list_of_amplicons)\n",
        "overhangs"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 19,
      "metadata": {},
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "Dseqrecord(-138)\n",
            "CATGGTTCACTGCCGTATAGGCAGCTAAGAAACACGTACAGGTCCTGGAGCG\u001b[48;5;11mGTTTTAGAGCTAGAAATAGCAAGTTAAAATAAGGC\u001b[0mTAGTCCGTTATCAACTTGAAAAAGTGGCACCGAGTCGGTGCTTTTTT    \n",
            "    CAAGTGACGGCATATCCGTCGATTCTTTGTGCATGTCCAGGACCTCGCCAAAATCTCGATCTTTATCGTTCAATTTTATTCCGATCAGGCAATAGTTGAACTTTTTCACCGTGGCTCAGCCACGAAAAAACCCT\n",
            "Dseqrecord(-106)\n",
            "GGGACGTCCAGGTCTTCACC\u001b[48;5;11mGTTTTAGAGCTAGAAATAGCAAGTTAAAATAAGGC\u001b[0mTAGTCCGTTATCAACTTGAAAAAGTGGCACCGAGTCGGTGCTTTTTT    \n",
            "    GCAGGTCCAGAAGTGGCAAAATCTCGATCTTTATCGTTCAATTTTATTCCGATCAGGCAATAGTTGAACTTTTTCACCGTGGCTCAGCCACGAAAAAACAGC\n",
            "Dseqrecord(-106)\n",
            "GTCGACCTCCCAGTCACGGC\u001b[48;5;11mGTTTTAGAGCTAGAAATAGCAAGTTAAAATAAGGC\u001b[0mTAGTCCGTTATCAACTTGAAAAAGTGGCACCGAGTCGGTGCTTTTTT    \n",
            "    TGGAGGGTCAGTGCCGCAAAATCTCGATCTTTATCGTTCAATTTTATTCCGATCAGGCAATAGTTGAACTTTTTCACCGTGGCTCAGCCACGAAAAAACTGG\n",
            "Dseqrecord(-106)\n",
            "GACCGTCCAGGACATGACCA\u001b[48;5;11mGTTTTAGAGCTAGAAATAGCAAGTTAAAATAAGGC\u001b[0mTAGTCCGTTATCAACTTGAAAAAGTGGCACCGAGTCGGTGCTTTTTT    \n",
            "    CAGGTCCTGTACTGGTCAAAATCTCGATCTTTATCGTTCAATTTTATTCCGATCAGGCAATAGTTGAACTTTTTCACCGTGGCTCAGCCACGAAAAAAGATC\n"
          ]
        }
      ],
      "source": [
        "digest_amplicons = digest_amplicons_w_BsaI(list_of_amplicons)\n",
        "for digest in digest_amplicons: \n",
        "    print(digest.figure())"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 20,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/plain": [
              "[Dseqrecord(-138), Dseqrecord(-106), Dseqrecord(-106), Dseqrecord(-106)]"
            ]
          },
          "execution_count": 20,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "digest_amplicons"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 21,
      "metadata": {},
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "Successfully looped the plasmid.\n"
          ]
        }
      ],
      "source": [
        "\n",
        "from Bio.Restriction import NcoI, NheI\n",
        "\n",
        "try:\n",
        "    linear_plasmid, smaller_frag = sorted(clean_plasmid.cut(NcoI, NheI), key=lambda x: len(x), reverse=True)\n",
        "    for i in range(len(digest_amplicons)):\n",
        "        # Assuming digest_amplicons[i] is a Dseqrecord, concatenate Dseq sequences\n",
        "        linear_plasmid.seq += digest_amplicons[i].seq\n",
        "    \n",
        "    try:\n",
        "        rec_vec = linear_plasmid.looped()\n",
        "        print(\"Successfully looped the plasmid.\")\n",
        "    except Exception as e:\n",
        "        print(f\"An error occurred while trying to loop the plasmid: {e}\")\n",
        "        # Handle the error or provide an alternative action here\n",
        "        rec_vec = None  # Example response; adjust as needed\n",
        "\n",
        "except Exception as e:\n",
        "    print(f\"An error occurred during plasmid digestion or preparation: {e}\")\n",
        "    # Additional error handling or cleanup here\n",
        "    rec_vec = None  # Ensure rec_vec is defined even if an error occurs"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 22,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/plain": [
              "'pCRISPR-mcBEST_4_sgrnas'"
            ]
          },
          "execution_count": 22,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "# Constructing a meaningful name, ID, and description for the assembled plasmid using user input\n",
        "plasmid_components = \"_\".join(genes_to_KO)  # Concatenates gene names from the user's list\n",
        "plasmid_purpose = f\"pCRISPR-mcBEST_{len(filtered_df)}_sgrnas\"  # Example: purpose of the plasmid\n",
        "\n",
        "rec_vec.name = f'{plasmid_purpose}'#_{plasmid_components}'\n",
        "rec_vec.id = rec_vec.name  # Using the same value for ID as for name for simplicity\n",
        "rec_vec.description = f'CRISPR-mcBEST_plasmid {\", \".join(genes_to_KO)} for gene KNOCK-DOWN, assembled using StreptoCAD.'\n",
        "rec_vec.id"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 23,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/plain": [
              "Dseqrecord(o12904)"
            ]
          },
          "execution_count": 23,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "# annotate the plasmid correctly\n",
        "annotate_plasmid_with_sgrnas(rec_vec, filtered_df)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 24,
      "metadata": {},
      "outputs": [],
      "source": [
        "print_plasmid = False\n",
        "\n",
        "if print_plasmid: \n",
        "    rec_vec.write(f\"../../data/plasmids/sgRNA_plasmids_pCRISPR_mcBEST/{rec_vec.id}.gb\")\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 25,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/plain": [
              "Dseqrecord(o12904)"
            ]
          },
          "execution_count": 25,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "rec_vec"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 26,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/plain": [
              "'sgRNA_SCO5090(140)_sgRNA_SCO5090(208)_sgRNA_SCO5087(522)_sgRNA_SCO5090(640)_'"
            ]
          },
          "execution_count": 26,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "integration_names = filtered_df.apply(lambda row: f\"sgRNA_{row['locus_tag']}({row['sgrna_loc']})_\", axis=1).tolist()\n",
        "integration_names = ''.join(integration_names)\n",
        "integration_names"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 27,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>plasmid_name</th>\n",
              "      <th>date</th>\n",
              "      <th>original_plasmid</th>\n",
              "      <th>integration</th>\n",
              "      <th>size</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>pCRISPR-mcBEST_4_sgrnas</td>\n",
              "      <td>2024-08-21</td>\n",
              "      <td>pCRISPR-mcBEST</td>\n",
              "      <td>sgRNA_SCO5090(140);sgRNA_SCO5090(208);sgRNA_SC...</td>\n",
              "      <td>12904</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "              plasmid_name        date original_plasmid  \\\n",
              "0  pCRISPR-mcBEST_4_sgrnas  2024-08-21   pCRISPR-mcBEST   \n",
              "\n",
              "                                         integration   size  \n",
              "0  sgRNA_SCO5090(140);sgRNA_SCO5090(208);sgRNA_SC...  12904  "
            ]
          },
          "execution_count": 27,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "integration_names = filtered_df.apply(lambda row: f\"sgRNA_{row['locus_tag']}({row['sgrna_loc']})\", axis=1).tolist()\n",
        "integration_names = [';'.join(integration_names)]\n",
        "\n",
        "plasmid_metadata_df = extract_metadata_to_dataframe([rec_vec],\n",
        "                                                    clean_plasmid,\n",
        "                                                    integration_names)\n",
        "\n",
        "plasmid_metadata_df"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 28,
      "metadata": {},
      "outputs": [],
      "source": [
        "pd.set_option('display.max_columns', None)\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 29,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>locus tag</th>\n",
              "      <th>f_primer_name</th>\n",
              "      <th>r_primer_name</th>\n",
              "      <th>f_primer_sequences(5-3)</th>\n",
              "      <th>r_primer_sequences(5-3)</th>\n",
              "      <th>f_tm</th>\n",
              "      <th>r_tm</th>\n",
              "      <th>ta</th>\n",
              "      <th>flanking_region</th>\n",
              "      <th>annealing_temperature</th>\n",
              "      <th>primer_pair</th>\n",
              "      <th>homodimer_forward_tm</th>\n",
              "      <th>homodimer_forward_deltaG (kcal/mol)</th>\n",
              "      <th>homodimer_reverse_tm</th>\n",
              "      <th>homodimer_reverse_deltaG (kcal/mol)</th>\n",
              "      <th>heterodimer_tm</th>\n",
              "      <th>heterodimer_deltaG (kcal/mol)</th>\n",
              "      <th>hairpin_forward_structure_found</th>\n",
              "      <th>hairpin_forward_tm</th>\n",
              "      <th>hairpin_forward_deltaG (kcal/mol)</th>\n",
              "      <th>hairpin_reverse_structure_found</th>\n",
              "      <th>hairpin_reverse_tm</th>\n",
              "      <th>hairpin_reverse_deltaG (kcal/mol)</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>SCO5087</td>\n",
              "      <td>SCO5087_fwd_checking_primer</td>\n",
              "      <td>SCO5087_rev_checking_primer</td>\n",
              "      <td>GACGATTCGGCCCGTG</td>\n",
              "      <td>CAGGGCGTCCAGGC</td>\n",
              "      <td>59</td>\n",
              "      <td>58</td>\n",
              "      <td>61</td>\n",
              "      <td>500</td>\n",
              "      <td>61</td>\n",
              "      <td>SCO5087_fwd_checking_primer &amp; SCO5087_rev_chec...</td>\n",
              "      <td>9.596676</td>\n",
              "      <td>-0.339306</td>\n",
              "      <td>9.706246</td>\n",
              "      <td>-3.415427</td>\n",
              "      <td>1.918567</td>\n",
              "      <td>-2.396641</td>\n",
              "      <td>False</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>False</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>SCO5090</td>\n",
              "      <td>SCO5090_fwd_checking_primer</td>\n",
              "      <td>SCO5090_rev_checking_primer</td>\n",
              "      <td>GGCGGCGGCCC</td>\n",
              "      <td>GACGTCCCCCGCGAAC</td>\n",
              "      <td>57</td>\n",
              "      <td>62</td>\n",
              "      <td>61</td>\n",
              "      <td>500</td>\n",
              "      <td>61</td>\n",
              "      <td>SCO5090_fwd_checking_primer &amp; SCO5090_rev_chec...</td>\n",
              "      <td>24.062465</td>\n",
              "      <td>-3.383988</td>\n",
              "      <td>8.234092</td>\n",
              "      <td>-3.870086</td>\n",
              "      <td>-1.478824</td>\n",
              "      <td>-1.481522</td>\n",
              "      <td>False</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>False</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>SCO5089</td>\n",
              "      <td>SCO5089_fwd_checking_primer</td>\n",
              "      <td>SCO5089_rev_checking_primer</td>\n",
              "      <td>CCCTTCGACGAGCGG</td>\n",
              "      <td>CGGGGCCGGTAC</td>\n",
              "      <td>58</td>\n",
              "      <td>52</td>\n",
              "      <td>56</td>\n",
              "      <td>600</td>\n",
              "      <td>56</td>\n",
              "      <td>SCO5089_fwd_checking_primer &amp; SCO5089_rev_chec...</td>\n",
              "      <td>-5.338917</td>\n",
              "      <td>-1.046369</td>\n",
              "      <td>1.902964</td>\n",
              "      <td>-2.973561</td>\n",
              "      <td>-20.141857</td>\n",
              "      <td>-2.152111</td>\n",
              "      <td>False</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>False</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "  locus tag                f_primer_name                r_primer_name  \\\n",
              "0   SCO5087  SCO5087_fwd_checking_primer  SCO5087_rev_checking_primer   \n",
              "1   SCO5090  SCO5090_fwd_checking_primer  SCO5090_rev_checking_primer   \n",
              "2   SCO5089  SCO5089_fwd_checking_primer  SCO5089_rev_checking_primer   \n",
              "\n",
              "  f_primer_sequences(5-3) r_primer_sequences(5-3) f_tm r_tm  ta  \\\n",
              "0        GACGATTCGGCCCGTG          CAGGGCGTCCAGGC   59   58  61   \n",
              "1             GGCGGCGGCCC        GACGTCCCCCGCGAAC   57   62  61   \n",
              "2         CCCTTCGACGAGCGG            CGGGGCCGGTAC   58   52  56   \n",
              "\n",
              "  flanking_region  annealing_temperature  \\\n",
              "0             500                     61   \n",
              "1             500                     61   \n",
              "2             600                     56   \n",
              "\n",
              "                                         primer_pair  homodimer_forward_tm  \\\n",
              "0  SCO5087_fwd_checking_primer & SCO5087_rev_chec...              9.596676   \n",
              "1  SCO5090_fwd_checking_primer & SCO5090_rev_chec...             24.062465   \n",
              "2  SCO5089_fwd_checking_primer & SCO5089_rev_chec...             -5.338917   \n",
              "\n",
              "   homodimer_forward_deltaG (kcal/mol)  homodimer_reverse_tm  \\\n",
              "0                            -0.339306              9.706246   \n",
              "1                            -3.383988              8.234092   \n",
              "2                            -1.046369              1.902964   \n",
              "\n",
              "   homodimer_reverse_deltaG (kcal/mol)  heterodimer_tm  \\\n",
              "0                            -3.415427        1.918567   \n",
              "1                            -3.870086       -1.478824   \n",
              "2                            -2.973561      -20.141857   \n",
              "\n",
              "   heterodimer_deltaG (kcal/mol)  hairpin_forward_structure_found  \\\n",
              "0                      -2.396641                            False   \n",
              "1                      -1.481522                            False   \n",
              "2                      -2.152111                            False   \n",
              "\n",
              "   hairpin_forward_tm  hairpin_forward_deltaG (kcal/mol)  \\\n",
              "0                 0.0                                0.0   \n",
              "1                 0.0                                0.0   \n",
              "2                 0.0                                0.0   \n",
              "\n",
              "   hairpin_reverse_structure_found  hairpin_reverse_tm  \\\n",
              "0                            False                 0.0   \n",
              "1                            False                 0.0   \n",
              "2                            False                 0.0   \n",
              "\n",
              "   hairpin_reverse_deltaG (kcal/mol)  \n",
              "0                                0.0  \n",
              "1                                0.0  \n",
              "2                                0.0  "
            ]
          },
          "execution_count": 29,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "\n",
        "# Getting checking primers\n",
        "checking_primers_df = find_best_check_primers_from_genome(genome, \n",
        "                                       genes_to_KO, \n",
        "                                       flanking_region=flanking_region,\n",
        "                                       target_tm = melting_temperature, \n",
        "                                        primer_concentration = primer_concentration, \n",
        "                                        polymerase = chosen_polymerase)\n",
        "checking_primers_df"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 30,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>Name</th>\n",
              "      <th>Sequence</th>\n",
              "      <th>Concentration</th>\n",
              "      <th>Purification</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>SCO5087_fwd_checking_primer</td>\n",
              "      <td>GACGATTCGGCCCGTG</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>SCO5090_fwd_checking_primer</td>\n",
              "      <td>GGCGGCGGCCC</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>SCO5089_fwd_checking_primer</td>\n",
              "      <td>CCCTTCGACGAGCGG</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>SCO5087_rev_checking_primer</td>\n",
              "      <td>CAGGGCGTCCAGGC</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>SCO5090_rev_checking_primer</td>\n",
              "      <td>GACGTCCCCCGCGAAC</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>5</th>\n",
              "      <td>SCO5089_rev_checking_primer</td>\n",
              "      <td>CGGGGCCGGTAC</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "                          Name          Sequence Concentration Purification\n",
              "0  SCO5087_fwd_checking_primer  GACGATTCGGCCCGTG          25nm          STD\n",
              "1  SCO5090_fwd_checking_primer       GGCGGCGGCCC          25nm          STD\n",
              "2  SCO5089_fwd_checking_primer   CCCTTCGACGAGCGG          25nm          STD\n",
              "3  SCO5087_rev_checking_primer    CAGGGCGTCCAGGC          25nm          STD\n",
              "4  SCO5090_rev_checking_primer  GACGTCCCCCGCGAAC          25nm          STD\n",
              "5  SCO5089_rev_checking_primer      CGGGGCCGGTAC          25nm          STD"
            ]
          },
          "execution_count": 30,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "# making the primers into dseqrecords\n",
        "checking_primers_df_idt = create_idt_order_dataframe(checking_primers_df)\n",
        "checking_primers_df_idt\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 31,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>Name</th>\n",
              "      <th>Sequence</th>\n",
              "      <th>Concentration</th>\n",
              "      <th>Purification</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>primer_0</td>\n",
              "      <td>GATCGGGTCTCCCATGGTTCACTGCCGTATAGGCAGCTAAGAAACA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>primer_2</td>\n",
              "      <td>GATCGGGTCTCCGGGACGTCCAGGTCTTCACCGTTTTAGAGCTAGA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>primer_4</td>\n",
              "      <td>GATCGGGTCTCCGTCGACCTCCCAGTCACGGCGTTTTAGAGCTAGA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>primer_6</td>\n",
              "      <td>GATCGGGTCTCCGACCGTCCAGGACATGACCAGTTTTAGAGCTAGA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>primer_1</td>\n",
              "      <td>GATCAGGTCTCGTCCCAAAAAAGCACCGACTCGGT</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>5</th>\n",
              "      <td>primer_3</td>\n",
              "      <td>GATCAGGTCTCGCGACAAAAAAGCACCGACTCGGT</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>6</th>\n",
              "      <td>primer_5</td>\n",
              "      <td>GATCAGGTCTCGGGTCAAAAAAGCACCGACTCGGT</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>7</th>\n",
              "      <td>primer_7</td>\n",
              "      <td>GATCAGGTCTCGCTAGAAAAAAGCACCGACTCGGT</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>SCO5087_fwd_checking_primer</td>\n",
              "      <td>GACGATTCGGCCCGTG</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>SCO5090_fwd_checking_primer</td>\n",
              "      <td>GGCGGCGGCCC</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>SCO5089_fwd_checking_primer</td>\n",
              "      <td>CCCTTCGACGAGCGG</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>SCO5087_rev_checking_primer</td>\n",
              "      <td>CAGGGCGTCCAGGC</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>SCO5090_rev_checking_primer</td>\n",
              "      <td>GACGTCCCCCGCGAAC</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>5</th>\n",
              "      <td>SCO5089_rev_checking_primer</td>\n",
              "      <td>CGGGGCCGGTAC</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "                          Name  \\\n",
              "0                     primer_0   \n",
              "1                     primer_2   \n",
              "2                     primer_4   \n",
              "3                     primer_6   \n",
              "4                     primer_1   \n",
              "5                     primer_3   \n",
              "6                     primer_5   \n",
              "7                     primer_7   \n",
              "0  SCO5087_fwd_checking_primer   \n",
              "1  SCO5090_fwd_checking_primer   \n",
              "2  SCO5089_fwd_checking_primer   \n",
              "3  SCO5087_rev_checking_primer   \n",
              "4  SCO5090_rev_checking_primer   \n",
              "5  SCO5089_rev_checking_primer   \n",
              "\n",
              "                                            Sequence Concentration  \\\n",
              "0  GATCGGGTCTCCCATGGTTCACTGCCGTATAGGCAGCTAAGAAACA...          25nm   \n",
              "1  GATCGGGTCTCCGGGACGTCCAGGTCTTCACCGTTTTAGAGCTAGA...          25nm   \n",
              "2  GATCGGGTCTCCGTCGACCTCCCAGTCACGGCGTTTTAGAGCTAGA...          25nm   \n",
              "3  GATCGGGTCTCCGACCGTCCAGGACATGACCAGTTTTAGAGCTAGA...          25nm   \n",
              "4                GATCAGGTCTCGTCCCAAAAAAGCACCGACTCGGT          25nm   \n",
              "5                GATCAGGTCTCGCGACAAAAAAGCACCGACTCGGT          25nm   \n",
              "6                GATCAGGTCTCGGGTCAAAAAAGCACCGACTCGGT          25nm   \n",
              "7                GATCAGGTCTCGCTAGAAAAAAGCACCGACTCGGT          25nm   \n",
              "0                                   GACGATTCGGCCCGTG          25nm   \n",
              "1                                        GGCGGCGGCCC          25nm   \n",
              "2                                    CCCTTCGACGAGCGG          25nm   \n",
              "3                                     CAGGGCGTCCAGGC          25nm   \n",
              "4                                   GACGTCCCCCGCGAAC          25nm   \n",
              "5                                       CGGGGCCGGTAC          25nm   \n",
              "\n",
              "  Purification  \n",
              "0          STD  \n",
              "1          STD  \n",
              "2          STD  \n",
              "3          STD  \n",
              "4          STD  \n",
              "5          STD  \n",
              "6          STD  \n",
              "7          STD  \n",
              "0          STD  \n",
              "1          STD  \n",
              "2          STD  \n",
              "3          STD  \n",
              "4          STD  \n",
              "5          STD  "
            ]
          },
          "execution_count": 31,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "full_idt = pd.concat([idt_df, checking_primers_df_idt])\n",
        "full_idt"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "## Folder with all the generated I/O"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 32,
      "metadata": {},
      "outputs": [],
      "source": [
        "get_data_folder = False"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 33,
      "metadata": {},
      "outputs": [],
      "source": [
        "if get_data_folder: \n",
        "    # Example: \n",
        "\n",
        "    input_files = [\n",
        "        {\"name\": \"input_genome.gb\", \"content\": genome},\n",
        "        {\"name\": \"input_plasmid.gb\", \"content\": clean_plasmid}\n",
        "    ]\n",
        "\n",
        "    output_files = [\n",
        "        {\"name\": \"mcBEST_w_sgRNAs.gb\", \"content\": rec_vec},\n",
        "        {\"name\": \"overhang_df.csv\", \"content\": overhangs},\n",
        "        {\"name\": \"pcr_df.csv\", \"content\": primer_df},\n",
        "        {\"name\": \"full_idt.csv\", \"content\": full_idt},\n",
        "        {\"name\": \"mutated_sgrna_df.csv\", \"content\": mutated_sgrna_df},\n",
        "        {\"name\": \"filtered_df.csv\", \"content\": filtered_df},\n",
        "        {\"name\": \"plasmid_metadata_df.csv\", \"content\": plasmid_metadata_df},\n",
        "    ]\n",
        "\n",
        "    input_values = {\n",
        "        \"genes_to_knockout\": ['SCO5087', 'SCO5089', 'SCO5090'],\n",
        "        \"filtering_metrics\": {\n",
        "            \"gc_upper\": gc_upper,\n",
        "            \"gc_lower\": gc_lower,\n",
        "            \"off_target_seed\": off_target_seed,\n",
        "            \"off_target_upper\": off_target_upper,\n",
        "            \"cas_type\": cas_type,\n",
        "            \"number_of_sgRNAs_per_group\": number_of_sgRNAs_per_group\n",
        "        },\n",
        "        \"polymerase_settings\": {\n",
        "            \"chosen_polymerase\": chosen_polymerase,\n",
        "            \"melting_temperature\": melting_temperature,\n",
        "            \"primer_concentration\": primer_concentration,\n",
        "            \"primer_number_increment\": primer_number_increment,\n",
        "            \"flanking_region\": flanking_region\n",
        "        }, # TODO might want to incorporate this to the app\n",
        "        \"overlapping_sequences\": {\n",
        "            \"restriction_overhang_f\": restriction_overhang_f,\n",
        "            \"restriction_overhang_r\": restriction_overhang_r,\n",
        "            \"backbone_overhang_f\": backbone_overhang_f,\n",
        "            \"backbone_overhang_r\": backbone_overhang_r,\n",
        "            \"cys4\": cys4,\n",
        "            \"sgRNA_handle_cys4_site\": str(sgRNA_handle_cys4_site[0].seq)\n",
        "        }\n",
        "    }\n",
        "\n",
        "\n",
        "    # Paths to Markdown files\n",
        "    markdown_file_paths = [\n",
        "        \"../../protocols/conjugation_protcol.md\",\n",
        "        \"../../protocols/multi_target_crispr_plasmid_protcol.md\"\n",
        "\n",
        "    ]\n",
        "\n",
        "    # Data and time\n",
        "    timestamp = datetime.utcnow().isoformat()\n",
        "    project_name=f\"CRISPR_mcBEST_workflow_{timestamp}\"\n",
        "\n",
        "\n",
        "    # Create project directory structure\n",
        "    project_directory = ProjectDirectory(\n",
        "        project_name=project_name,\n",
        "        input_files=input_files,\n",
        "        output_files=output_files,\n",
        "        input_values=input_values,\n",
        "        markdown_file_paths=markdown_file_paths\n",
        "    )\n",
        "\n",
        "\n",
        "    # DO You want to save the folder? \n",
        "    save_zip_folder = False \n",
        "\n",
        "    if save_zip_folder: \n",
        "        # Generate the project directory structure and get the zip content\n",
        "        zip_content = project_directory.create_directory_structure(create_directories=False)\n",
        "\n",
        "        # Save the zip file to disk (optional)\n",
        "        with open(\"project_structure.zip\", \"wb\") as f:\n",
        "            f.write(zip_content)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {},
      "outputs": [],
      "source": []
    }
  ],
  "metadata": {
    "kernelspec": {
      "display_name": "constrain",
      "language": "python",
      "name": "python3"
    },
    "language_info": {
      "codemirror_mode": {
        "name": "ipython",
        "version": 3
      },
      "file_extension": ".py",
      "mimetype": "text/x-python",
      "name": "python",
      "nbconvert_exporter": "python",
      "pygments_lexer": "ipython3",
      "version": "3.11.9"
    }
  },
  "nbformat": 4,
  "nbformat_minor": 2
}
