{
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "# W4-CRISPRi-workflow"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "Transcriptional interference is another powerful tool for assessing the function of genes, by expressing a guide RNA that targets a dead-Cas9/sgRNA complex to bind to the + (Watson) strand of the promoter and 5’-UTR of the targeted gene, thereby sterically blocking binding and progression of RNA polymerases to the promoter region. Workflow 4 uses CRISPR interference (CRISPRi) with ssDNA bridging to reversibly inactivate genes transcriptionally12,29 (Figure 5A). This approach targets regions upstream of the Transcriptional Start Site (TSS), using a dCas9-sgRNA complex positioned near the TSS (default 100bp upstream) to sterically hinder transcription. This allows for the functional study of genes through controlled knockdown. To get started, users can download the pCRISPR-dCas9 plasmid and the S. coelicolor (A3) genome. StreptoAIM will then generate all necessary components, including primers and plasmids."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 13,
      "metadata": {},
      "outputs": [],
      "source": [
        "import sys\n",
        "import os\n",
        "from pydna.dseqrecord import Dseqrecord\n",
        "from datetime import datetime\n",
        "\n",
        "\n",
        "# Ensure the src directory is in the Python path\n",
        "project_root = os.path.abspath(os.path.join(os.getcwd(), '../../'))\n",
        "if project_root not in sys.path:\n",
        "    sys.path.insert(0, project_root)\n",
        "\n",
        "from streptocad.sequence_loading.sequence_loading import (\n",
        "    load_and_process_gene_sequences,\n",
        "    load_and_process_genome_sequences,\n",
        "    load_and_process_plasmid, \n",
        "    check_and_convert_input,\n",
        "    annotate_dseqrecord,\n",
        "    process_specified_gene_sequences_from_record)\n",
        "\n",
        "\n",
        "from streptocad.cloning.ssDNA_bridging import assemble_plasmids_by_ssDNA_bridging, make_ssDNA_oligos\n",
        "from streptocad.crispr.guideRNA_crispri import extract_sgRNAs_for_crispri, SgRNAargs\n",
        "from streptocad.primers.primer_generation import  primers_to_IDT\n",
        "from streptocad.utils import ProjectDirectory,extract_metadata_to_dataframe"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "## INPUT"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 14,
      "metadata": {},
      "outputs": [],
      "source": [
        "# Inputs\n",
        "# 1 Add genome of choice (genbank, fasta)\n",
        "path_to_genome = '../../data/genomes/Streptomyces_coelicolor_A3_chromosome.gb'\n",
        "genome = load_and_process_genome_sequences(path_to_genome)[0]\n",
        "\n",
        "# 2 Add plasmid \n",
        "path_to_plasmid = '../../data/plasmids/pCRISPR–Cas9_plasmid_addgene.gbk'\n",
        "clean_plasmid = load_and_process_plasmid(path_to_plasmid)\n",
        "\n",
        "# 3 Choose genes to knock out (list)\n",
        "genes_to_KO = ['SCO5087', 'SCO5089', 'SCO5090']\n",
        "genes_to_KO = ['80000-100000', '4000-7000', '9000-14000','15000-20000']\n",
        "\n",
        "\n",
        "#### Advanced settings ####\n",
        "# 4 Filtering metrics for sgRNAs\n",
        "gc_upper = 0.72\n",
        "gc_lower = 0.2\n",
        "off_target_seed = 13\n",
        "off_target_upper = 0\n",
        "cas_type='cas9'\n",
        "number_of_sgRNAs_per_group = 5\n",
        "extension_to_promoter_region=200\n",
        "\n",
        "# 6 Choose overlapping sequences for our plasmid we can use the following\n",
        "#As per the article **\"CRISPR–Cas9, CRISPRi and CRISPR-BEST-mediated genetic manipulation in streptomycetes\"** we need the following oligoes: \n",
        "#CGGTTGGTAGGATCGACGGC **-N20-** GTTTTAGAGCTAGAAATAGC\n",
        "up_homology = Dseqrecord('CGGTTGGTAGGATCGACGGC')\n",
        "dw_homology = Dseqrecord('GTTTTAGAGCTAGAAATAGC')"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 15,
      "metadata": {},
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "pCRISPR-Cas9\n"
          ]
        }
      ],
      "source": [
        "print(clean_plasmid.id)"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "# Computation"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 3,
      "metadata": {},
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "True\n",
            "this is the target dict [{'region_1_(80000-100000)': [80000, 100000]}, {'region_2_(4000-7000)': [4000, 7000]}, {'region_3_(9000-14000)': [9000, 14000]}, {'region_4_(15000-20000)': [15000, 20000]}]\n",
            "region_1_(80000-100000) [80000, 100000]\n",
            "region_2_(4000-7000) [4000, 7000]\n",
            "region_3_(9000-14000) [9000, 14000]\n",
            "region_4_(15000-20000) [15000, 20000]\n"
          ]
        },
        {
          "data": {
            "text/plain": [
              "25828"
            ]
          },
          "execution_count": 3,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "target_dict, genes_to_KO, annotation_input = check_and_convert_input(genes_to_KO)\n",
        "\n",
        "print(annotation_input)\n",
        "if annotation_input == True:\n",
        "    genome = annotate_dseqrecord(genome, target_dict)\n",
        "\n",
        "\n",
        "len(genome.features)\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 4,
      "metadata": {},
      "outputs": [],
      "source": [
        "\n",
        "# Initialize SgRNAArgs with desired parameters\n",
        "args = SgRNAargs(genome, \n",
        "                genes_to_KO,\n",
        "                step=['find', 'filter'],\n",
        "                gc_upper = gc_upper,\n",
        "                gc_lower = gc_lower,\n",
        "                off_target_seed = off_target_seed,\n",
        "                off_target_upper = off_target_upper,\n",
        "                cas_type='cas9',\n",
        "                extension_to_promoter_region=200,\n",
        "                target_non_template_strand = True\n",
        "                  )\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 5,
      "metadata": {},
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "sgRNA or PAM generated were outside the designated border in region_1_(80000-100000). Skipping to next locus tag.\n",
            "sgRNA or PAM generated were outside the designated border in region_1_(80000-100000). Skipping to next locus tag.\n",
            "sgRNA or PAM generated were outside the designated border in region_1_(80000-100000). Skipping to next locus tag.\n",
            "sgRNA or PAM generated were outside the designated border in region_1_(80000-100000). Skipping to next locus tag.\n",
            "sgRNA or PAM generated were outside the designated border in region_1_(80000-100000). Skipping to next locus tag.\n",
            "sgRNA or PAM generated were outside the designated border in region_1_(80000-100000). Skipping to next locus tag.\n",
            "sgRNA or PAM generated were outside the designated border in region_1_(80000-100000). Skipping to next locus tag.\n",
            "sgRNA or PAM generated were outside the designated border in region_1_(80000-100000). Skipping to next locus tag.\n",
            "sgRNA or PAM generated were outside the designated border in region_2_(4000-7000). Skipping to next locus tag.\n",
            "sgRNA or PAM generated were outside the designated border in region_2_(4000-7000). Skipping to next locus tag.\n",
            "sgRNA or PAM generated were outside the designated border in region_2_(4000-7000). Skipping to next locus tag.\n",
            "sgRNA or PAM generated were outside the designated border in region_2_(4000-7000). Skipping to next locus tag.\n",
            "sgRNA or PAM generated were outside the designated border in region_2_(4000-7000). Skipping to next locus tag.\n",
            "sgRNA or PAM generated were outside the designated border in region_2_(4000-7000). Skipping to next locus tag.\n",
            "sgRNA or PAM generated were outside the designated border in region_2_(4000-7000). Skipping to next locus tag.\n",
            "sgRNA or PAM generated were outside the designated border in region_3_(9000-14000). Skipping to next locus tag.\n",
            "sgRNA or PAM generated were outside the designated border in region_3_(9000-14000). Skipping to next locus tag.\n",
            "sgRNA or PAM generated were outside the designated border in region_3_(9000-14000). Skipping to next locus tag.\n",
            "sgRNA or PAM generated were outside the designated border in region_3_(9000-14000). Skipping to next locus tag.\n",
            "sgRNA or PAM generated were outside the designated border in region_3_(9000-14000). Skipping to next locus tag.\n",
            "sgRNA or PAM generated were outside the designated border in region_3_(9000-14000). Skipping to next locus tag.\n",
            "sgRNA or PAM generated were outside the designated border in region_3_(9000-14000). Skipping to next locus tag.\n",
            "sgRNA or PAM generated were outside the designated border in region_3_(9000-14000). Skipping to next locus tag.\n",
            "sgRNA or PAM generated were outside the designated border in region_3_(9000-14000). Skipping to next locus tag.\n",
            "sgRNA or PAM generated were outside the designated border in region_4_(15000-20000). Skipping to next locus tag.\n",
            "sgRNA or PAM generated were outside the designated border in region_4_(15000-20000). Skipping to next locus tag.\n",
            "sgRNA or PAM generated were outside the designated border in region_4_(15000-20000). Skipping to next locus tag.\n",
            "sgRNA or PAM generated were outside the designated border in region_4_(15000-20000). Skipping to next locus tag.\n",
            "sgRNA or PAM generated were outside the designated border in region_4_(15000-20000). Skipping to next locus tag.\n",
            "sgRNA or PAM generated were outside the designated border in region_4_(15000-20000). Skipping to next locus tag.\n"
          ]
        },
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>strain_name</th>\n",
              "      <th>locus_tag</th>\n",
              "      <th>gene_loc</th>\n",
              "      <th>gene_strand</th>\n",
              "      <th>sgrna_strand</th>\n",
              "      <th>sgrna_loc</th>\n",
              "      <th>gc</th>\n",
              "      <th>pam</th>\n",
              "      <th>sgrna</th>\n",
              "      <th>sgrna_seed_sequence</th>\n",
              "      <th>off_target_count</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>8</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>region_1_(80000-100000)</td>\n",
              "      <td>80001</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>-22</td>\n",
              "      <td>0.70</td>\n",
              "      <td>CGG</td>\n",
              "      <td>CCCAGCGGTTGACCAAGCCG</td>\n",
              "      <td>GTTGACCAAGCCG</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>9</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>region_1_(80000-100000)</td>\n",
              "      <td>80001</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>-7</td>\n",
              "      <td>0.65</td>\n",
              "      <td>CGG</td>\n",
              "      <td>GGGAGAACGGCATCACCCAG</td>\n",
              "      <td>CGGCATCACCCAG</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>10</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>region_1_(80000-100000)</td>\n",
              "      <td>80001</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>6</td>\n",
              "      <td>0.60</td>\n",
              "      <td>CGG</td>\n",
              "      <td>GAGCGTATCTGCCGGGAGAA</td>\n",
              "      <td>TCTGCCGGGAGAA</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>11</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>region_1_(80000-100000)</td>\n",
              "      <td>80001</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>13</td>\n",
              "      <td>0.60</td>\n",
              "      <td>GGG</td>\n",
              "      <td>GCTCTTCGAGCGTATCTGCC</td>\n",
              "      <td>GAGCGTATCTGCC</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>12</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>region_1_(80000-100000)</td>\n",
              "      <td>80001</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>14</td>\n",
              "      <td>0.55</td>\n",
              "      <td>CGG</td>\n",
              "      <td>TGCTCTTCGAGCGTATCTGC</td>\n",
              "      <td>CGAGCGTATCTGC</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>15</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>region_1_(80000-100000)</td>\n",
              "      <td>80001</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>53</td>\n",
              "      <td>0.65</td>\n",
              "      <td>CGG</td>\n",
              "      <td>AGTTCACCGGCCGGCACCAA</td>\n",
              "      <td>CGGCCGGCACCAA</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>17</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>region_1_(80000-100000)</td>\n",
              "      <td>80001</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>66</td>\n",
              "      <td>0.55</td>\n",
              "      <td>CGG</td>\n",
              "      <td>GACAACGGCAAGCAGTTCAC</td>\n",
              "      <td>GCAAGCAGTTCAC</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>18</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>region_1_(80000-100000)</td>\n",
              "      <td>80001</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>81</td>\n",
              "      <td>0.50</td>\n",
              "      <td>CGG</td>\n",
              "      <td>TTCGAAGTGCTGACGGACAA</td>\n",
              "      <td>TGCTGACGGACAA</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>19</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>region_1_(80000-100000)</td>\n",
              "      <td>80001</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>88</td>\n",
              "      <td>0.60</td>\n",
              "      <td>CGG</td>\n",
              "      <td>CGTGCCCTTCGAAGTGCTGA</td>\n",
              "      <td>TTCGAAGTGCTGA</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "    strain_name                locus_tag  gene_loc  gene_strand  sgrna_strand  \\\n",
              "8   NC_003888.3  region_1_(80000-100000)     80001            1            -1   \n",
              "9   NC_003888.3  region_1_(80000-100000)     80001            1            -1   \n",
              "10  NC_003888.3  region_1_(80000-100000)     80001            1            -1   \n",
              "11  NC_003888.3  region_1_(80000-100000)     80001            1            -1   \n",
              "12  NC_003888.3  region_1_(80000-100000)     80001            1            -1   \n",
              "15  NC_003888.3  region_1_(80000-100000)     80001            1            -1   \n",
              "17  NC_003888.3  region_1_(80000-100000)     80001            1            -1   \n",
              "18  NC_003888.3  region_1_(80000-100000)     80001            1            -1   \n",
              "19  NC_003888.3  region_1_(80000-100000)     80001            1            -1   \n",
              "\n",
              "    sgrna_loc    gc  pam                 sgrna sgrna_seed_sequence  \\\n",
              "8         -22  0.70  CGG  CCCAGCGGTTGACCAAGCCG       GTTGACCAAGCCG   \n",
              "9          -7  0.65  CGG  GGGAGAACGGCATCACCCAG       CGGCATCACCCAG   \n",
              "10          6  0.60  CGG  GAGCGTATCTGCCGGGAGAA       TCTGCCGGGAGAA   \n",
              "11         13  0.60  GGG  GCTCTTCGAGCGTATCTGCC       GAGCGTATCTGCC   \n",
              "12         14  0.55  CGG  TGCTCTTCGAGCGTATCTGC       CGAGCGTATCTGC   \n",
              "15         53  0.65  CGG  AGTTCACCGGCCGGCACCAA       CGGCCGGCACCAA   \n",
              "17         66  0.55  CGG  GACAACGGCAAGCAGTTCAC       GCAAGCAGTTCAC   \n",
              "18         81  0.50  CGG  TTCGAAGTGCTGACGGACAA       TGCTGACGGACAA   \n",
              "19         88  0.60  CGG  CGTGCCCTTCGAAGTGCTGA       TTCGAAGTGCTGA   \n",
              "\n",
              "    off_target_count  \n",
              "8                  0  \n",
              "9                  0  \n",
              "10                 0  \n",
              "11                 0  \n",
              "12                 0  \n",
              "15                 0  \n",
              "17                 0  \n",
              "18                 0  \n",
              "19                 0  "
            ]
          },
          "execution_count": 5,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "\n",
        "sgrna_df = extract_sgRNAs_for_crispri(args)\n",
        "sgrna_df"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 6,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>strain_name</th>\n",
              "      <th>locus_tag</th>\n",
              "      <th>gene_loc</th>\n",
              "      <th>gene_strand</th>\n",
              "      <th>sgrna_strand</th>\n",
              "      <th>sgrna_loc</th>\n",
              "      <th>gc</th>\n",
              "      <th>pam</th>\n",
              "      <th>sgrna</th>\n",
              "      <th>sgrna_seed_sequence</th>\n",
              "      <th>off_target_count</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>8</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>region_1_(80000-100000)</td>\n",
              "      <td>80001</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>-22</td>\n",
              "      <td>0.70</td>\n",
              "      <td>CGG</td>\n",
              "      <td>CCCAGCGGTTGACCAAGCCG</td>\n",
              "      <td>GTTGACCAAGCCG</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>9</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>region_1_(80000-100000)</td>\n",
              "      <td>80001</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>-7</td>\n",
              "      <td>0.65</td>\n",
              "      <td>CGG</td>\n",
              "      <td>GGGAGAACGGCATCACCCAG</td>\n",
              "      <td>CGGCATCACCCAG</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>10</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>region_1_(80000-100000)</td>\n",
              "      <td>80001</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>6</td>\n",
              "      <td>0.60</td>\n",
              "      <td>CGG</td>\n",
              "      <td>GAGCGTATCTGCCGGGAGAA</td>\n",
              "      <td>TCTGCCGGGAGAA</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>11</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>region_1_(80000-100000)</td>\n",
              "      <td>80001</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>13</td>\n",
              "      <td>0.60</td>\n",
              "      <td>GGG</td>\n",
              "      <td>GCTCTTCGAGCGTATCTGCC</td>\n",
              "      <td>GAGCGTATCTGCC</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>12</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>region_1_(80000-100000)</td>\n",
              "      <td>80001</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>14</td>\n",
              "      <td>0.55</td>\n",
              "      <td>CGG</td>\n",
              "      <td>TGCTCTTCGAGCGTATCTGC</td>\n",
              "      <td>CGAGCGTATCTGC</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "    strain_name                locus_tag  gene_loc  gene_strand  sgrna_strand  \\\n",
              "8   NC_003888.3  region_1_(80000-100000)     80001            1            -1   \n",
              "9   NC_003888.3  region_1_(80000-100000)     80001            1            -1   \n",
              "10  NC_003888.3  region_1_(80000-100000)     80001            1            -1   \n",
              "11  NC_003888.3  region_1_(80000-100000)     80001            1            -1   \n",
              "12  NC_003888.3  region_1_(80000-100000)     80001            1            -1   \n",
              "\n",
              "    sgrna_loc    gc  pam                 sgrna sgrna_seed_sequence  \\\n",
              "8         -22  0.70  CGG  CCCAGCGGTTGACCAAGCCG       GTTGACCAAGCCG   \n",
              "9          -7  0.65  CGG  GGGAGAACGGCATCACCCAG       CGGCATCACCCAG   \n",
              "10          6  0.60  CGG  GAGCGTATCTGCCGGGAGAA       TCTGCCGGGAGAA   \n",
              "11         13  0.60  GGG  GCTCTTCGAGCGTATCTGCC       GAGCGTATCTGCC   \n",
              "12         14  0.55  CGG  TGCTCTTCGAGCGTATCTGC       CGAGCGTATCTGC   \n",
              "\n",
              "    off_target_count  \n",
              "8                  0  \n",
              "9                  0  \n",
              "10                 0  \n",
              "11                 0  \n",
              "12                 0  "
            ]
          },
          "execution_count": 6,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "# Filter the DataFrame to retain only up to 5 sgRNA sequences per locus_tag\n",
        "filtered_df = sgrna_df.groupby('locus_tag').head(number_of_sgRNAs_per_group)\n",
        "filtered_df"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "## Output"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 7,
      "metadata": {},
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "region_1_(80000-100000)_loc_-22\n"
          ]
        },
        {
          "data": {
            "text/plain": [
              "[Contig(o11279),\n",
              " Contig(o11279),\n",
              " Contig(o11279),\n",
              " Contig(o11279),\n",
              " Contig(o11279)]"
            ]
          },
          "execution_count": 7,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "# MAke oligoes\n",
        "list_of_ssDNAs = make_ssDNA_oligos(filtered_df, upstream_ovh = up_homology,\n",
        "                      downstream_ovh=dw_homology)\n",
        "print(list_of_ssDNAs[0].name)\n",
        "\n",
        "# cut plasmid\n",
        "from Bio.Restriction import NcoI\n",
        "linearized_plasmid = sorted(clean_plasmid.cut(NcoI), key=lambda x: len(x), reverse=True)[0]\n",
        "#print(linearized_plasmid)\n",
        "\n",
        "sgRNA_vectors = assemble_plasmids_by_ssDNA_bridging(list_of_ssDNAs,linearized_plasmid)\n",
        "sgRNA_vectors"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 8,
      "metadata": {},
      "outputs": [],
      "source": [
        "\n",
        "# Constructing a meaningful name, ID, and description for the assembled plasmid using user input\n",
        "targeting_info = []\n",
        "for index, row in filtered_df.iterrows():\n",
        "    formatted_str = f\"CRISPRi_{row['locus_tag']}_p{row['sgrna_loc']}\"\n",
        "    targeting_info.append(formatted_str)\n",
        "\n",
        "for i in range(len(sgRNA_vectors)):\n",
        "    sgRNA_vectors[i].name = f'p{targeting_info[i]}_#{i+1}'\n",
        "    sgRNA_vectors[i].id = sgRNA_vectors[i].name  # Using the same value for ID as for name for simplicity\n",
        "    sgRNA_vectors[i].description = f'Assembled plasmid targeting {\", \".join(genes_to_KO)} for single gene KNOCK-DOWN, assembled using StreptoCAD.'\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 9,
      "metadata": {},
      "outputs": [],
      "source": [
        "print_plasmids = False\n",
        "\n",
        "if print_plasmids: \n",
        "    for vector in sgRNA_vectors: \n",
        "        vector.write(f\"../../data/plasmids/sgRNA_plasmids_pCRISPR–dCas9/{vector.id}.gb\")"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 10,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>plasmid_name</th>\n",
              "      <th>date</th>\n",
              "      <th>original_plasmid</th>\n",
              "      <th>integration</th>\n",
              "      <th>size</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>pCRISPRi_region_1_(80000-100000)_p-22_#1</td>\n",
              "      <td>2024-07-23</td>\n",
              "      <td>.</td>\n",
              "      <td>sgRNA_region_1_(80000-100000)(-22)</td>\n",
              "      <td>11279</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>pCRISPRi_region_1_(80000-100000)_p-7_#2</td>\n",
              "      <td>2024-07-23</td>\n",
              "      <td>.</td>\n",
              "      <td>sgRNA_region_1_(80000-100000)(-7)</td>\n",
              "      <td>11279</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>pCRISPRi_region_1_(80000-100000)_p6_#3</td>\n",
              "      <td>2024-07-23</td>\n",
              "      <td>.</td>\n",
              "      <td>sgRNA_region_1_(80000-100000)(6)</td>\n",
              "      <td>11279</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>pCRISPRi_region_1_(80000-100000)_p13_#4</td>\n",
              "      <td>2024-07-23</td>\n",
              "      <td>.</td>\n",
              "      <td>sgRNA_region_1_(80000-100000)(13)</td>\n",
              "      <td>11279</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>pCRISPRi_region_1_(80000-100000)_p14_#5</td>\n",
              "      <td>2024-07-23</td>\n",
              "      <td>.</td>\n",
              "      <td>sgRNA_region_1_(80000-100000)(14)</td>\n",
              "      <td>11279</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "                               plasmid_name        date original_plasmid  \\\n",
              "0  pCRISPRi_region_1_(80000-100000)_p-22_#1  2024-07-23                .   \n",
              "1   pCRISPRi_region_1_(80000-100000)_p-7_#2  2024-07-23                .   \n",
              "2    pCRISPRi_region_1_(80000-100000)_p6_#3  2024-07-23                .   \n",
              "3   pCRISPRi_region_1_(80000-100000)_p13_#4  2024-07-23                .   \n",
              "4   pCRISPRi_region_1_(80000-100000)_p14_#5  2024-07-23                .   \n",
              "\n",
              "                          integration   size  \n",
              "0  sgRNA_region_1_(80000-100000)(-22)  11279  \n",
              "1   sgRNA_region_1_(80000-100000)(-7)  11279  \n",
              "2    sgRNA_region_1_(80000-100000)(6)  11279  \n",
              "3   sgRNA_region_1_(80000-100000)(13)  11279  \n",
              "4   sgRNA_region_1_(80000-100000)(14)  11279  "
            ]
          },
          "execution_count": 10,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "integration_names = filtered_df.apply(lambda row: f\"sgRNA_{row['locus_tag']}({row['sgrna_loc']})\", axis=1).tolist()\n",
        "plasmid_metadata_df = extract_metadata_to_dataframe(sgRNA_vectors,\n",
        "                                                    clean_plasmid,\n",
        "                                                    integration_names)\n",
        "\n",
        "plasmid_metadata_df"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "### IDT primers"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 11,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>Name</th>\n",
              "      <th>Sequence</th>\n",
              "      <th>Concentration</th>\n",
              "      <th>Purification</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>region_1_(80000-100000)_loc_-22</td>\n",
              "      <td>CGGTTGGTAGGATCGACGGCCCCAGCGGTTGACCAAGCCGGTTTTA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>region_1_(80000-100000)_loc_-7</td>\n",
              "      <td>CGGTTGGTAGGATCGACGGCGGGAGAACGGCATCACCCAGGTTTTA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>region_1_(80000-100000)_loc_6</td>\n",
              "      <td>CGGTTGGTAGGATCGACGGCGAGCGTATCTGCCGGGAGAAGTTTTA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>region_1_(80000-100000)_loc_13</td>\n",
              "      <td>CGGTTGGTAGGATCGACGGCGCTCTTCGAGCGTATCTGCCGTTTTA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>region_1_(80000-100000)_loc_14</td>\n",
              "      <td>CGGTTGGTAGGATCGACGGCTGCTCTTCGAGCGTATCTGCGTTTTA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "                              Name  \\\n",
              "0  region_1_(80000-100000)_loc_-22   \n",
              "1   region_1_(80000-100000)_loc_-7   \n",
              "2    region_1_(80000-100000)_loc_6   \n",
              "3   region_1_(80000-100000)_loc_13   \n",
              "4   region_1_(80000-100000)_loc_14   \n",
              "\n",
              "                                            Sequence Concentration  \\\n",
              "0  CGGTTGGTAGGATCGACGGCCCCAGCGGTTGACCAAGCCGGTTTTA...          25nm   \n",
              "1  CGGTTGGTAGGATCGACGGCGGGAGAACGGCATCACCCAGGTTTTA...          25nm   \n",
              "2  CGGTTGGTAGGATCGACGGCGAGCGTATCTGCCGGGAGAAGTTTTA...          25nm   \n",
              "3  CGGTTGGTAGGATCGACGGCGCTCTTCGAGCGTATCTGCCGTTTTA...          25nm   \n",
              "4  CGGTTGGTAGGATCGACGGCTGCTCTTCGAGCGTATCTGCGTTTTA...          25nm   \n",
              "\n",
              "  Purification  \n",
              "0          STD  \n",
              "1          STD  \n",
              "2          STD  \n",
              "3          STD  \n",
              "4          STD  "
            ]
          },
          "execution_count": 11,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "idt_primers=primers_to_IDT(list_of_ssDNAs)\n",
        "idt_primers"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "## Folder with all the generated I/O"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 12,
      "metadata": {},
      "outputs": [],
      "source": [
        "generate_data_folder = True \n",
        "\n",
        "if generate_data_folder:\n",
        "    input_files = [\n",
        "        {\"name\": \"input_genome.gb\", \"content\": genome},\n",
        "        {\"name\": \"input_plasmid.gb\", \"content\": clean_plasmid}\n",
        "    ]\n",
        "\n",
        "    output_files = [\n",
        "        {\"name\": \"cBEST_w_sgRNAs.gb\", \"content\": sgRNA_vectors}, # LIST OF Dseqrecords\n",
        "        {\"name\": \"full_idt.csv\", \"content\": idt_primers},\n",
        "        {\"name\": \"sgrna_df.csv\", \"content\": sgrna_df},\n",
        "        {\"name\": \"filtered_df.csv\", \"content\": filtered_df},\n",
        "        {\"name\": \"plasmid_metadata_df.csv\", \"content\": plasmid_metadata_df},\n",
        "    ]\n",
        "\n",
        "    input_values = {\n",
        "        \"genes_to_knockout\": genes_to_KO,\n",
        "\n",
        "\n",
        "        \"filtering_metrics\": {\n",
        "            \"gc_upper\": gc_upper,\n",
        "            \"gc_lower\": gc_lower,\n",
        "            \"off_target_seed\": off_target_seed,\n",
        "            \"off_target_upper\": off_target_upper,\n",
        "            \"cas_type\": cas_type,\n",
        "            \"number_of_sgRNAs_per_group\": number_of_sgRNAs_per_group,\n",
        "            'extension_to_promoter_region':extension_to_promoter_region,\n",
        "\n",
        "        },\n",
        "        \"overlapping_sequences\": {\n",
        "            \"up_homology\": str(up_homology),\n",
        "            \"dw_homology\": str(dw_homology)\n",
        "        }\n",
        "    }\n",
        "\n",
        "\n",
        "    # Paths to Markdown files\n",
        "    markdown_file_paths = [\n",
        "        \"../../protocols/conjugation_protcol.md\",\n",
        "        \"../../protocols/single_target_crispr_plasmid_protcol.md\"\n",
        "\n",
        "    ]\n",
        "\n",
        "\n",
        "\n",
        "    # Data and time\n",
        "    timestamp = datetime.utcnow().isoformat()\n",
        "\n",
        "    # Create project directory structure\n",
        "    project_directory = ProjectDirectory(\n",
        "        project_name=f\"CRISPRi_workflow_{timestamp}\",\n",
        "        input_files=input_files,\n",
        "        output_files=output_files,\n",
        "        input_values=input_values,\n",
        "        markdown_file_paths=markdown_file_paths\n",
        "    )\n",
        "\n",
        "\n",
        "    # DO You want to save the folder? \n",
        "    save_zip_folder = False \n",
        "\n",
        "    if save_zip_folder: \n",
        "        # Generate the project directory structure and get the zip content\n",
        "        zip_content = project_directory.create_directory_structure(create_directories=False)\n",
        "\n",
        "        # Save the zip file to disk (optional)\n",
        "        with open(\"project_structure.zip\", \"wb\") as f:\n",
        "            f.write(zip_content)"
      ]
    }
  ],
  "metadata": {
    "kernelspec": {
      "display_name": "constrain",
      "language": "python",
      "name": "python3"
    },
    "language_info": {
      "codemirror_mode": {
        "name": "ipython",
        "version": 3
      },
      "file_extension": ".py",
      "mimetype": "text/x-python",
      "name": "python",
      "nbconvert_exporter": "python",
      "pygments_lexer": "ipython3",
      "version": "3.11.9"
    }
  },
  "nbformat": 4,
  "nbformat_minor": 2
}
