{
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "# W4-CRISPRi-workflow"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "Transcriptional interference is another powerful tool for assessing the function of genes, by expressing a guide RNA that targets a dead-Cas9/sgRNA complex to bind to the + (Watson) strand of the promoter and 5’-UTR of the targeted gene, thereby sterically blocking binding and progression of RNA polymerases to the promoter region. Workflow 4 uses CRISPR interference (CRISPRi) with ssDNA bridging to reversibly inactivate genes transcriptionally12,29 (Figure 5A). This approach targets regions upstream of the Transcriptional Start Site (TSS), using a dCas9-sgRNA complex positioned near the TSS (default 100bp upstream) to sterically hinder transcription. This allows for the functional study of genes through controlled knockdown. To get started, users can download the pCRISPR-dCas9 plasmid and the S. coelicolor (A3) genome. StreptoAIM will then generate all necessary components, including primers and plasmids."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 1,
      "metadata": {},
      "outputs": [],
      "source": [
        "import sys\n",
        "import os\n",
        "from pydna.dseqrecord import Dseqrecord\n",
        "from datetime import datetime\n",
        "\n",
        "\n",
        "# Ensure the src directory is in the Python path\n",
        "project_root = os.path.abspath(os.path.join(os.getcwd(), '../../'))\n",
        "if project_root not in sys.path:\n",
        "    sys.path.insert(0, project_root)\n",
        "\n",
        "from streptocad.sequence_loading.sequence_loading import (\n",
        "    load_and_process_gene_sequences,\n",
        "    load_and_process_genome_sequences,\n",
        "    load_and_process_plasmid, \n",
        "    check_and_convert_input,\n",
        "    annotate_dseqrecord,\n",
        "    process_specified_gene_sequences_from_record)\n",
        "\n",
        "\n",
        "from streptocad.cloning.ssDNA_bridging import assemble_plasmids_by_ssDNA_bridging, make_ssDNA_oligos\n",
        "from streptocad.crispr.guideRNA_crispri import extract_sgRNAs_for_crispri, SgRNAargs\n",
        "from streptocad.primers.primer_generation import  primers_to_IDT\n",
        "from streptocad.utils import ProjectDirectory,extract_metadata_to_dataframe"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "## INPUT"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 2,
      "metadata": {},
      "outputs": [],
      "source": [
        "# Inputs\n",
        "# 1 Add genome of choice (genbank, fasta)\n",
        "path_to_genome = '../../data/genomes/Streptomyces_coelicolor_A3_chromosome.gb'\n",
        "genome = load_and_process_genome_sequences(path_to_genome)[0]\n",
        "\n",
        "# 2 Add plasmid \n",
        "path_to_plasmid = '../../data/plasmids/pCRISPR–Cas9_plasmid_addgene.gbk'\n",
        "clean_plasmid = load_and_process_plasmid(path_to_plasmid)\n",
        "\n",
        "# 3 Choose genes to knock out (list)\n",
        "genes_to_KO = ['SCO5087']#, 'SCO5089', 'SCO5090']\n",
        "\n",
        "# negative strand example \n",
        "#genes_to_KO =['SCO0007']\n",
        "\n",
        "\n",
        "#### Advanced settings ####\n",
        "# 4 Filtering metrics for sgRNAs\n",
        "gc_upper = 0.9999\n",
        "gc_lower = 0.0001\n",
        "off_target_seed = 13\n",
        "off_target_upper = 10\n",
        "cas_type='cas9'\n",
        "number_of_sgRNAs_per_group = 5\n",
        "extension_to_promoter_region=200\n",
        "\n",
        "# 6 Choose overlapping sequences for our plasmid we can use the following\n",
        "#As per the article **\"CRISPR–Cas9, CRISPRi and CRISPR-BEST-mediated genetic manipulation in streptomycetes\"** we need the following oligoes: \n",
        "#CGGTTGGTAGGATCGACGGC **-N20-** GTTTTAGAGCTAGAAATAGC\n",
        "up_homology = Dseqrecord('CGGTTGGTAGGATCGACGGC')\n",
        "dw_homology = Dseqrecord('GTTTTAGAGCTAGAAATAGC')"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 3,
      "metadata": {},
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "pCRISPR-Cas9\n"
          ]
        }
      ],
      "source": [
        "print(clean_plasmid.id)\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "# Computation"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 4,
      "metadata": {},
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "False\n"
          ]
        },
        {
          "data": {
            "text/plain": [
              "25824"
            ]
          },
          "execution_count": 4,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "target_dict, genes_to_KO, annotation_input = check_and_convert_input(genes_to_KO)\n",
        "\n",
        "print(annotation_input)\n",
        "if annotation_input == True:\n",
        "    genome = annotate_dseqrecord(genome, target_dict)\n",
        "\n",
        "\n",
        "len(genome.features)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 16,
      "metadata": {},
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "Pam was found outside designated locus_tag: SCO5087. To incorporate this extent borders. Skipping to next locus tag.\n",
            "sgRNA generated were outside the designated border in SCO5087. To incorporate this extent borders. Skipping to next locus tag.\n",
            "Pam was found outside designated locus_tag: SCO5087. To incorporate this extent borders. Skipping to next locus tag.\n",
            "sgRNA generated were outside the designated border in SCO5087. To incorporate this extent borders. Skipping to next locus tag.\n",
            "sgRNA generated were outside the designated border in SCO5087. To incorporate this extent borders. Skipping to next locus tag.\n"
          ]
        },
        {
          "name": "stderr",
          "output_type": "stream",
          "text": [
            "/Users/lucaslevassor/opt/anaconda3/envs/streptocad/lib/python3.11/site-packages/Bio/SeqFeature.py:257: BiopythonDeprecationWarning: Using the strand argument is deprecated, and will be removed in a future release. Please set it via the location object instead.\n",
            "  warnings.warn(\n"
          ]
        },
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>strain_name</th>\n",
              "      <th>locus_tag</th>\n",
              "      <th>gene_loc</th>\n",
              "      <th>gene_strand</th>\n",
              "      <th>sgrna_strand</th>\n",
              "      <th>sgrna_loc</th>\n",
              "      <th>gc</th>\n",
              "      <th>pam</th>\n",
              "      <th>sgrna</th>\n",
              "      <th>sgrna_seed_sequence</th>\n",
              "      <th>off_target_count</th>\n",
              "      <th>region</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>304</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5087</td>\n",
              "      <td>5529601</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>-90</td>\n",
              "      <td>0.65</td>\n",
              "      <td>AGG</td>\n",
              "      <td>AGGGGAACACATGGCCACGC</td>\n",
              "      <td>CACATGGCCACGC</td>\n",
              "      <td>0</td>\n",
              "      <td>upstream</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>305</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5087</td>\n",
              "      <td>5529601</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>-81</td>\n",
              "      <td>0.65</td>\n",
              "      <td>TGG</td>\n",
              "      <td>GAGGCAGGGAGGGGAACACA</td>\n",
              "      <td>GGAGGGGAACACA</td>\n",
              "      <td>0</td>\n",
              "      <td>upstream</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>325</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5087</td>\n",
              "      <td>5529601</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>-80</td>\n",
              "      <td>0.70</td>\n",
              "      <td>TGG</td>\n",
              "      <td>GTGTTCCCCTCCCTGCCTCG</td>\n",
              "      <td>CCTCCCTGCCTCG</td>\n",
              "      <td>0</td>\n",
              "      <td>upstream</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>306</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5087</td>\n",
              "      <td>5529601</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>-72</td>\n",
              "      <td>0.70</td>\n",
              "      <td>GGG</td>\n",
              "      <td>AGGGACCACGAGGCAGGGAG</td>\n",
              "      <td>ACGAGGCAGGGAG</td>\n",
              "      <td>0</td>\n",
              "      <td>upstream</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>307</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5087</td>\n",
              "      <td>5529601</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>-71</td>\n",
              "      <td>0.70</td>\n",
              "      <td>GGG</td>\n",
              "      <td>GAGGGACCACGAGGCAGGGA</td>\n",
              "      <td>CACGAGGCAGGGA</td>\n",
              "      <td>0</td>\n",
              "      <td>upstream</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>308</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5087</td>\n",
              "      <td>5529601</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>-70</td>\n",
              "      <td>0.70</td>\n",
              "      <td>AGG</td>\n",
              "      <td>TGAGGGACCACGAGGCAGGG</td>\n",
              "      <td>CCACGAGGCAGGG</td>\n",
              "      <td>0</td>\n",
              "      <td>upstream</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>309</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5087</td>\n",
              "      <td>5529601</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>-67</td>\n",
              "      <td>0.70</td>\n",
              "      <td>GGG</td>\n",
              "      <td>GCGTGAGGGACCACGAGGCA</td>\n",
              "      <td>GGACCACGAGGCA</td>\n",
              "      <td>0</td>\n",
              "      <td>upstream</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>310</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5087</td>\n",
              "      <td>5529601</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>-66</td>\n",
              "      <td>0.75</td>\n",
              "      <td>AGG</td>\n",
              "      <td>CGCGTGAGGGACCACGAGGC</td>\n",
              "      <td>GGGACCACGAGGC</td>\n",
              "      <td>0</td>\n",
              "      <td>upstream</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>311</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5087</td>\n",
              "      <td>5529601</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>-62</td>\n",
              "      <td>0.70</td>\n",
              "      <td>AGG</td>\n",
              "      <td>TGAGCGCGTGAGGGACCACG</td>\n",
              "      <td>GTGAGGGACCACG</td>\n",
              "      <td>0</td>\n",
              "      <td>upstream</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>324</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5087</td>\n",
              "      <td>5529601</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>-58</td>\n",
              "      <td>0.65</td>\n",
              "      <td>TGG</td>\n",
              "      <td>GTCCCTCACGCGCTCAGCTT</td>\n",
              "      <td>ACGCGCTCAGCTT</td>\n",
              "      <td>0</td>\n",
              "      <td>upstream</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>323</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5087</td>\n",
              "      <td>5529601</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>-57</td>\n",
              "      <td>0.60</td>\n",
              "      <td>GGG</td>\n",
              "      <td>TCCCTCACGCGCTCAGCTTT</td>\n",
              "      <td>CGCGCTCAGCTTT</td>\n",
              "      <td>0</td>\n",
              "      <td>upstream</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>312</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5087</td>\n",
              "      <td>5529601</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>-53</td>\n",
              "      <td>0.65</td>\n",
              "      <td>GGG</td>\n",
              "      <td>GCCCAAAGCTGAGCGCGTGA</td>\n",
              "      <td>GCTGAGCGCGTGA</td>\n",
              "      <td>0</td>\n",
              "      <td>upstream</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>313</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5087</td>\n",
              "      <td>5529601</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>-52</td>\n",
              "      <td>0.70</td>\n",
              "      <td>AGG</td>\n",
              "      <td>CGCCCAAAGCTGAGCGCGTG</td>\n",
              "      <td>AGCTGAGCGCGTG</td>\n",
              "      <td>0</td>\n",
              "      <td>upstream</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>322</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5087</td>\n",
              "      <td>5529601</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>-50</td>\n",
              "      <td>0.75</td>\n",
              "      <td>CGG</td>\n",
              "      <td>CGCGCTCAGCTTTGGGCGCC</td>\n",
              "      <td>AGCTTTGGGCGCC</td>\n",
              "      <td>0</td>\n",
              "      <td>upstream</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>321</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5087</td>\n",
              "      <td>5529601</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>-41</td>\n",
              "      <td>0.75</td>\n",
              "      <td>CGG</td>\n",
              "      <td>CTTTGGGCGCCCGGCTCGAG</td>\n",
              "      <td>CGCCCGGCTCGAG</td>\n",
              "      <td>0</td>\n",
              "      <td>upstream</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>320</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5087</td>\n",
              "      <td>5529601</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>-38</td>\n",
              "      <td>0.85</td>\n",
              "      <td>CGG</td>\n",
              "      <td>TGGGCGCCCGGCTCGAGCGG</td>\n",
              "      <td>CCGGCTCGAGCGG</td>\n",
              "      <td>0</td>\n",
              "      <td>upstream</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>319</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5087</td>\n",
              "      <td>5529601</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>-31</td>\n",
              "      <td>0.80</td>\n",
              "      <td>AGG</td>\n",
              "      <td>CCGGCTCGAGCGGCGGTCGA</td>\n",
              "      <td>GAGCGGCGGTCGA</td>\n",
              "      <td>0</td>\n",
              "      <td>upstream</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>318</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5087</td>\n",
              "      <td>5529601</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>-30</td>\n",
              "      <td>0.75</td>\n",
              "      <td>GGG</td>\n",
              "      <td>CGGCTCGAGCGGCGGTCGAA</td>\n",
              "      <td>AGCGGCGGTCGAA</td>\n",
              "      <td>0</td>\n",
              "      <td>upstream</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>314</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5087</td>\n",
              "      <td>5529601</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>-29</td>\n",
              "      <td>0.75</td>\n",
              "      <td>GGG</td>\n",
              "      <td>CTTCGACCGCCGCTCGAGCC</td>\n",
              "      <td>CGCCGCTCGAGCC</td>\n",
              "      <td>1</td>\n",
              "      <td>upstream</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>315</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5087</td>\n",
              "      <td>5529601</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>-28</td>\n",
              "      <td>0.75</td>\n",
              "      <td>CGG</td>\n",
              "      <td>CCTTCGACCGCCGCTCGAGC</td>\n",
              "      <td>CCGCCGCTCGAGC</td>\n",
              "      <td>0</td>\n",
              "      <td>upstream</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>317</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5087</td>\n",
              "      <td>5529601</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>-24</td>\n",
              "      <td>0.70</td>\n",
              "      <td>TGG</td>\n",
              "      <td>GAGCGGCGGTCGAAGGGAGA</td>\n",
              "      <td>GGTCGAAGGGAGA</td>\n",
              "      <td>0</td>\n",
              "      <td>upstream</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>316</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5087</td>\n",
              "      <td>5529601</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>-23</td>\n",
              "      <td>0.65</td>\n",
              "      <td>GGG</td>\n",
              "      <td>AGCGGCGGTCGAAGGGAGAT</td>\n",
              "      <td>GTCGAAGGGAGAT</td>\n",
              "      <td>0</td>\n",
              "      <td>upstream</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>289</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5087</td>\n",
              "      <td>5529801</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>2</td>\n",
              "      <td>0.90</td>\n",
              "      <td>TGG</td>\n",
              "      <td>GCCGCTGGACGCGGCGCCGG</td>\n",
              "      <td>GACGCGGCGCCGG</td>\n",
              "      <td>1</td>\n",
              "      <td>CDS</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5087</td>\n",
              "      <td>5529801</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>6</td>\n",
              "      <td>0.80</td>\n",
              "      <td>CGG</td>\n",
              "      <td>TCCACCGGCGCCGCGTCCAG</td>\n",
              "      <td>GCGCCGCGTCCAG</td>\n",
              "      <td>0</td>\n",
              "      <td>CDS</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>288</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5087</td>\n",
              "      <td>5529801</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>8</td>\n",
              "      <td>0.80</td>\n",
              "      <td>CGG</td>\n",
              "      <td>GGACGCGGCGCCGGTGGATC</td>\n",
              "      <td>GCGCCGGTGGATC</td>\n",
              "      <td>1</td>\n",
              "      <td>CDS</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>287</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5087</td>\n",
              "      <td>5529801</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>16</td>\n",
              "      <td>0.75</td>\n",
              "      <td>AGG</td>\n",
              "      <td>CGCCGGTGGATCCGGCATCG</td>\n",
              "      <td>GGATCCGGCATCG</td>\n",
              "      <td>0</td>\n",
              "      <td>CDS</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>286</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5087</td>\n",
              "      <td>5529801</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>17</td>\n",
              "      <td>0.70</td>\n",
              "      <td>GGG</td>\n",
              "      <td>GCCGGTGGATCCGGCATCGA</td>\n",
              "      <td>GATCCGGCATCGA</td>\n",
              "      <td>0</td>\n",
              "      <td>CDS</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>285</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5087</td>\n",
              "      <td>5529801</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>18</td>\n",
              "      <td>0.70</td>\n",
              "      <td>GGG</td>\n",
              "      <td>CCGGTGGATCCGGCATCGAG</td>\n",
              "      <td>ATCCGGCATCGAG</td>\n",
              "      <td>0</td>\n",
              "      <td>CDS</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5087</td>\n",
              "      <td>5529801</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>21</td>\n",
              "      <td>0.70</td>\n",
              "      <td>CGG</td>\n",
              "      <td>CCCCTCGATGCCGGATCCAC</td>\n",
              "      <td>ATGCCGGATCCAC</td>\n",
              "      <td>0</td>\n",
              "      <td>CDS</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>284</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5087</td>\n",
              "      <td>5529801</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>29</td>\n",
              "      <td>0.65</td>\n",
              "      <td>CGG</td>\n",
              "      <td>GGCATCGAGGGGTCCCGTAT</td>\n",
              "      <td>AGGGGTCCCGTAT</td>\n",
              "      <td>0</td>\n",
              "      <td>CDS</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5087</td>\n",
              "      <td>5529801</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>30</td>\n",
              "      <td>0.65</td>\n",
              "      <td>CGG</td>\n",
              "      <td>GATACGGGACCCCTCGATGC</td>\n",
              "      <td>GACCCCTCGATGC</td>\n",
              "      <td>0</td>\n",
              "      <td>CDS</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5087</td>\n",
              "      <td>5529801</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>45</td>\n",
              "      <td>0.65</td>\n",
              "      <td>GGG</td>\n",
              "      <td>GGAGGCTCGAAGGCCGATAC</td>\n",
              "      <td>CGAAGGCCGATAC</td>\n",
              "      <td>0</td>\n",
              "      <td>CDS</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5087</td>\n",
              "      <td>5529801</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>46</td>\n",
              "      <td>0.60</td>\n",
              "      <td>CGG</td>\n",
              "      <td>AGGAGGCTCGAAGGCCGATA</td>\n",
              "      <td>TCGAAGGCCGATA</td>\n",
              "      <td>0</td>\n",
              "      <td>CDS</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>283</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5087</td>\n",
              "      <td>5529801</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>54</td>\n",
              "      <td>0.60</td>\n",
              "      <td>CGG</td>\n",
              "      <td>TTCGAGCCTCCTTCGAGCCA</td>\n",
              "      <td>CTCCTTCGAGCCA</td>\n",
              "      <td>0</td>\n",
              "      <td>CDS</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>5</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5087</td>\n",
              "      <td>5529801</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>55</td>\n",
              "      <td>0.65</td>\n",
              "      <td>AGG</td>\n",
              "      <td>GTGGCTCGAAGGAGGCTCGA</td>\n",
              "      <td>GAAGGAGGCTCGA</td>\n",
              "      <td>0</td>\n",
              "      <td>CDS</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>282</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5087</td>\n",
              "      <td>5529801</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>55</td>\n",
              "      <td>0.65</td>\n",
              "      <td>GGG</td>\n",
              "      <td>TCGAGCCTCCTTCGAGCCAC</td>\n",
              "      <td>TCCTTCGAGCCAC</td>\n",
              "      <td>0</td>\n",
              "      <td>CDS</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>281</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5087</td>\n",
              "      <td>5529801</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>56</td>\n",
              "      <td>0.70</td>\n",
              "      <td>GGG</td>\n",
              "      <td>CGAGCCTCCTTCGAGCCACG</td>\n",
              "      <td>CCTTCGAGCCACG</td>\n",
              "      <td>0</td>\n",
              "      <td>CDS</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>6</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5087</td>\n",
              "      <td>5529801</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>63</td>\n",
              "      <td>0.75</td>\n",
              "      <td>AGG</td>\n",
              "      <td>TCGGCCCCGTGGCTCGAAGG</td>\n",
              "      <td>CGTGGCTCGAAGG</td>\n",
              "      <td>0</td>\n",
              "      <td>CDS</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>7</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5087</td>\n",
              "      <td>5529801</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>66</td>\n",
              "      <td>0.75</td>\n",
              "      <td>AGG</td>\n",
              "      <td>TCGTCGGCCCCGTGGCTCGA</td>\n",
              "      <td>CCCCGTGGCTCGA</td>\n",
              "      <td>0</td>\n",
              "      <td>CDS</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>8</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5087</td>\n",
              "      <td>5529801</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>74</td>\n",
              "      <td>0.75</td>\n",
              "      <td>TGG</td>\n",
              "      <td>CGTCGTCATCGTCGGCCCCG</td>\n",
              "      <td>ATCGTCGGCCCCG</td>\n",
              "      <td>0</td>\n",
              "      <td>CDS</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>280</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5087</td>\n",
              "      <td>5529801</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>79</td>\n",
              "      <td>0.65</td>\n",
              "      <td>CGG</td>\n",
              "      <td>CCGACGATGACGACGACCAC</td>\n",
              "      <td>TGACGACGACCAC</td>\n",
              "      <td>1</td>\n",
              "      <td>CDS</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "     strain_name locus_tag  gene_loc  gene_strand  sgrna_strand  sgrna_loc  \\\n",
              "304  NC_003888.3   SCO5087   5529601            1            -1        -90   \n",
              "305  NC_003888.3   SCO5087   5529601            1            -1        -81   \n",
              "325  NC_003888.3   SCO5087   5529601            1             1        -80   \n",
              "306  NC_003888.3   SCO5087   5529601            1            -1        -72   \n",
              "307  NC_003888.3   SCO5087   5529601            1            -1        -71   \n",
              "308  NC_003888.3   SCO5087   5529601            1            -1        -70   \n",
              "309  NC_003888.3   SCO5087   5529601            1            -1        -67   \n",
              "310  NC_003888.3   SCO5087   5529601            1            -1        -66   \n",
              "311  NC_003888.3   SCO5087   5529601            1            -1        -62   \n",
              "324  NC_003888.3   SCO5087   5529601            1             1        -58   \n",
              "323  NC_003888.3   SCO5087   5529601            1             1        -57   \n",
              "312  NC_003888.3   SCO5087   5529601            1            -1        -53   \n",
              "313  NC_003888.3   SCO5087   5529601            1            -1        -52   \n",
              "322  NC_003888.3   SCO5087   5529601            1             1        -50   \n",
              "321  NC_003888.3   SCO5087   5529601            1             1        -41   \n",
              "320  NC_003888.3   SCO5087   5529601            1             1        -38   \n",
              "319  NC_003888.3   SCO5087   5529601            1             1        -31   \n",
              "318  NC_003888.3   SCO5087   5529601            1             1        -30   \n",
              "314  NC_003888.3   SCO5087   5529601            1            -1        -29   \n",
              "315  NC_003888.3   SCO5087   5529601            1            -1        -28   \n",
              "317  NC_003888.3   SCO5087   5529601            1             1        -24   \n",
              "316  NC_003888.3   SCO5087   5529601            1             1        -23   \n",
              "289  NC_003888.3   SCO5087   5529801            1             1          2   \n",
              "0    NC_003888.3   SCO5087   5529801            1            -1          6   \n",
              "288  NC_003888.3   SCO5087   5529801            1             1          8   \n",
              "287  NC_003888.3   SCO5087   5529801            1             1         16   \n",
              "286  NC_003888.3   SCO5087   5529801            1             1         17   \n",
              "285  NC_003888.3   SCO5087   5529801            1             1         18   \n",
              "1    NC_003888.3   SCO5087   5529801            1            -1         21   \n",
              "284  NC_003888.3   SCO5087   5529801            1             1         29   \n",
              "2    NC_003888.3   SCO5087   5529801            1            -1         30   \n",
              "3    NC_003888.3   SCO5087   5529801            1            -1         45   \n",
              "4    NC_003888.3   SCO5087   5529801            1            -1         46   \n",
              "283  NC_003888.3   SCO5087   5529801            1             1         54   \n",
              "5    NC_003888.3   SCO5087   5529801            1            -1         55   \n",
              "282  NC_003888.3   SCO5087   5529801            1             1         55   \n",
              "281  NC_003888.3   SCO5087   5529801            1             1         56   \n",
              "6    NC_003888.3   SCO5087   5529801            1            -1         63   \n",
              "7    NC_003888.3   SCO5087   5529801            1            -1         66   \n",
              "8    NC_003888.3   SCO5087   5529801            1            -1         74   \n",
              "280  NC_003888.3   SCO5087   5529801            1             1         79   \n",
              "\n",
              "       gc  pam                 sgrna sgrna_seed_sequence  off_target_count  \\\n",
              "304  0.65  AGG  AGGGGAACACATGGCCACGC       CACATGGCCACGC                 0   \n",
              "305  0.65  TGG  GAGGCAGGGAGGGGAACACA       GGAGGGGAACACA                 0   \n",
              "325  0.70  TGG  GTGTTCCCCTCCCTGCCTCG       CCTCCCTGCCTCG                 0   \n",
              "306  0.70  GGG  AGGGACCACGAGGCAGGGAG       ACGAGGCAGGGAG                 0   \n",
              "307  0.70  GGG  GAGGGACCACGAGGCAGGGA       CACGAGGCAGGGA                 0   \n",
              "308  0.70  AGG  TGAGGGACCACGAGGCAGGG       CCACGAGGCAGGG                 0   \n",
              "309  0.70  GGG  GCGTGAGGGACCACGAGGCA       GGACCACGAGGCA                 0   \n",
              "310  0.75  AGG  CGCGTGAGGGACCACGAGGC       GGGACCACGAGGC                 0   \n",
              "311  0.70  AGG  TGAGCGCGTGAGGGACCACG       GTGAGGGACCACG                 0   \n",
              "324  0.65  TGG  GTCCCTCACGCGCTCAGCTT       ACGCGCTCAGCTT                 0   \n",
              "323  0.60  GGG  TCCCTCACGCGCTCAGCTTT       CGCGCTCAGCTTT                 0   \n",
              "312  0.65  GGG  GCCCAAAGCTGAGCGCGTGA       GCTGAGCGCGTGA                 0   \n",
              "313  0.70  AGG  CGCCCAAAGCTGAGCGCGTG       AGCTGAGCGCGTG                 0   \n",
              "322  0.75  CGG  CGCGCTCAGCTTTGGGCGCC       AGCTTTGGGCGCC                 0   \n",
              "321  0.75  CGG  CTTTGGGCGCCCGGCTCGAG       CGCCCGGCTCGAG                 0   \n",
              "320  0.85  CGG  TGGGCGCCCGGCTCGAGCGG       CCGGCTCGAGCGG                 0   \n",
              "319  0.80  AGG  CCGGCTCGAGCGGCGGTCGA       GAGCGGCGGTCGA                 0   \n",
              "318  0.75  GGG  CGGCTCGAGCGGCGGTCGAA       AGCGGCGGTCGAA                 0   \n",
              "314  0.75  GGG  CTTCGACCGCCGCTCGAGCC       CGCCGCTCGAGCC                 1   \n",
              "315  0.75  CGG  CCTTCGACCGCCGCTCGAGC       CCGCCGCTCGAGC                 0   \n",
              "317  0.70  TGG  GAGCGGCGGTCGAAGGGAGA       GGTCGAAGGGAGA                 0   \n",
              "316  0.65  GGG  AGCGGCGGTCGAAGGGAGAT       GTCGAAGGGAGAT                 0   \n",
              "289  0.90  TGG  GCCGCTGGACGCGGCGCCGG       GACGCGGCGCCGG                 1   \n",
              "0    0.80  CGG  TCCACCGGCGCCGCGTCCAG       GCGCCGCGTCCAG                 0   \n",
              "288  0.80  CGG  GGACGCGGCGCCGGTGGATC       GCGCCGGTGGATC                 1   \n",
              "287  0.75  AGG  CGCCGGTGGATCCGGCATCG       GGATCCGGCATCG                 0   \n",
              "286  0.70  GGG  GCCGGTGGATCCGGCATCGA       GATCCGGCATCGA                 0   \n",
              "285  0.70  GGG  CCGGTGGATCCGGCATCGAG       ATCCGGCATCGAG                 0   \n",
              "1    0.70  CGG  CCCCTCGATGCCGGATCCAC       ATGCCGGATCCAC                 0   \n",
              "284  0.65  CGG  GGCATCGAGGGGTCCCGTAT       AGGGGTCCCGTAT                 0   \n",
              "2    0.65  CGG  GATACGGGACCCCTCGATGC       GACCCCTCGATGC                 0   \n",
              "3    0.65  GGG  GGAGGCTCGAAGGCCGATAC       CGAAGGCCGATAC                 0   \n",
              "4    0.60  CGG  AGGAGGCTCGAAGGCCGATA       TCGAAGGCCGATA                 0   \n",
              "283  0.60  CGG  TTCGAGCCTCCTTCGAGCCA       CTCCTTCGAGCCA                 0   \n",
              "5    0.65  AGG  GTGGCTCGAAGGAGGCTCGA       GAAGGAGGCTCGA                 0   \n",
              "282  0.65  GGG  TCGAGCCTCCTTCGAGCCAC       TCCTTCGAGCCAC                 0   \n",
              "281  0.70  GGG  CGAGCCTCCTTCGAGCCACG       CCTTCGAGCCACG                 0   \n",
              "6    0.75  AGG  TCGGCCCCGTGGCTCGAAGG       CGTGGCTCGAAGG                 0   \n",
              "7    0.75  AGG  TCGTCGGCCCCGTGGCTCGA       CCCCGTGGCTCGA                 0   \n",
              "8    0.75  TGG  CGTCGTCATCGTCGGCCCCG       ATCGTCGGCCCCG                 0   \n",
              "280  0.65  CGG  CCGACGATGACGACGACCAC       TGACGACGACCAC                 1   \n",
              "\n",
              "       region  \n",
              "304  upstream  \n",
              "305  upstream  \n",
              "325  upstream  \n",
              "306  upstream  \n",
              "307  upstream  \n",
              "308  upstream  \n",
              "309  upstream  \n",
              "310  upstream  \n",
              "311  upstream  \n",
              "324  upstream  \n",
              "323  upstream  \n",
              "312  upstream  \n",
              "313  upstream  \n",
              "322  upstream  \n",
              "321  upstream  \n",
              "320  upstream  \n",
              "319  upstream  \n",
              "318  upstream  \n",
              "314  upstream  \n",
              "315  upstream  \n",
              "317  upstream  \n",
              "316  upstream  \n",
              "289       CDS  \n",
              "0         CDS  \n",
              "288       CDS  \n",
              "287       CDS  \n",
              "286       CDS  \n",
              "285       CDS  \n",
              "1         CDS  \n",
              "284       CDS  \n",
              "2         CDS  \n",
              "3         CDS  \n",
              "4         CDS  \n",
              "283       CDS  \n",
              "5         CDS  \n",
              "282       CDS  \n",
              "281       CDS  \n",
              "6         CDS  \n",
              "7         CDS  \n",
              "8         CDS  \n",
              "280       CDS  "
            ]
          },
          "execution_count": 16,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "# Initialize SgRNAArgs with desired parameters\n",
        "args = SgRNAargs(genome, \n",
        "                genes_to_KO,\n",
        "                step=['find', 'filter'],\n",
        "                gc_upper = gc_upper,\n",
        "                gc_lower = gc_lower,\n",
        "                off_target_seed = off_target_seed,\n",
        "                off_target_upper = off_target_upper,\n",
        "                cas_type='cas9',\n",
        "                target_non_template_strand = False, \n",
        "                extension_to_promoter_region=200,\n",
        "                upstream_tss=100,\n",
        "                dwstream_tss=80,\n",
        "                  )\n",
        "\n",
        "sgrna_df = extract_sgRNAs_for_crispri(args)\n",
        "sgrna_df"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 17,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/plain": [
              "41"
            ]
          },
          "execution_count": 17,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "len(sgrna_df)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 7,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>ID</th>\n",
              "      <th>Start</th>\n",
              "      <th>End</th>\n",
              "      <th>Strand</th>\n",
              "      <th>ORF</th>\n",
              "      <th>Sequence</th>\n",
              "      <th>PAM</th>\n",
              "      <th>C to T mutations</th>\n",
              "      <th>A to G mutations</th>\n",
              "      <th>0bp mismatches</th>\n",
              "      <th>1bp mismatches</th>\n",
              "      <th>2bp mismatches</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>CY00000011</td>\n",
              "      <td>6</td>\n",
              "      <td>29</td>\n",
              "      <td>-1</td>\n",
              "      <td>-</td>\n",
              "      <td>AGGGGAACACATGGCCACGC</td>\n",
              "      <td>AGG</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>50</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>CY00000003</td>\n",
              "      <td>15</td>\n",
              "      <td>38</td>\n",
              "      <td>-1</td>\n",
              "      <td>-</td>\n",
              "      <td>GAGGCAGGGAGGGGAACACA</td>\n",
              "      <td>TGG</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>11</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>10</th>\n",
              "      <td>CY00000020</td>\n",
              "      <td>24</td>\n",
              "      <td>47</td>\n",
              "      <td>-1</td>\n",
              "      <td>-</td>\n",
              "      <td>AGGGACCACGAGGCAGGGAG</td>\n",
              "      <td>GGG</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>0</td>\n",
              "      <td>2</td>\n",
              "      <td>30</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>8</th>\n",
              "      <td>CY00000018</td>\n",
              "      <td>25</td>\n",
              "      <td>48</td>\n",
              "      <td>-1</td>\n",
              "      <td>-</td>\n",
              "      <td>GAGGGACCACGAGGCAGGGA</td>\n",
              "      <td>GGG</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>0</td>\n",
              "      <td>1</td>\n",
              "      <td>34</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>CY00000008</td>\n",
              "      <td>26</td>\n",
              "      <td>49</td>\n",
              "      <td>-1</td>\n",
              "      <td>-</td>\n",
              "      <td>TGAGGGACCACGAGGCAGGG</td>\n",
              "      <td>AGG</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>26</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>11</th>\n",
              "      <td>CY00000021</td>\n",
              "      <td>29</td>\n",
              "      <td>52</td>\n",
              "      <td>-1</td>\n",
              "      <td>-</td>\n",
              "      <td>GCGTGAGGGACCACGAGGCA</td>\n",
              "      <td>GGG</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>0</td>\n",
              "      <td>2</td>\n",
              "      <td>36</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>14</th>\n",
              "      <td>CY00000027</td>\n",
              "      <td>30</td>\n",
              "      <td>53</td>\n",
              "      <td>-1</td>\n",
              "      <td>-</td>\n",
              "      <td>CGCGTGAGGGACCACGAGGC</td>\n",
              "      <td>AGG</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>0</td>\n",
              "      <td>5</td>\n",
              "      <td>117</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>13</th>\n",
              "      <td>CY00000026</td>\n",
              "      <td>34</td>\n",
              "      <td>57</td>\n",
              "      <td>-1</td>\n",
              "      <td>-</td>\n",
              "      <td>TGAGCGCGTGAGGGACCACG</td>\n",
              "      <td>AGG</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>0</td>\n",
              "      <td>4</td>\n",
              "      <td>36</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>12</th>\n",
              "      <td>CY00000025</td>\n",
              "      <td>43</td>\n",
              "      <td>66</td>\n",
              "      <td>-1</td>\n",
              "      <td>-</td>\n",
              "      <td>GCCCAAAGCTGAGCGCGTGA</td>\n",
              "      <td>GGG</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>0</td>\n",
              "      <td>3</td>\n",
              "      <td>59</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>CY00000009</td>\n",
              "      <td>44</td>\n",
              "      <td>67</td>\n",
              "      <td>-1</td>\n",
              "      <td>-</td>\n",
              "      <td>CGCCCAAAGCTGAGCGCGTG</td>\n",
              "      <td>AGG</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>29</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>16</th>\n",
              "      <td>CY00000032</td>\n",
              "      <td>68</td>\n",
              "      <td>91</td>\n",
              "      <td>-1</td>\n",
              "      <td>-</td>\n",
              "      <td>CCTTCGACCGCCGCTCGAGC</td>\n",
              "      <td>CGG</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>0</td>\n",
              "      <td>16</td>\n",
              "      <td>340</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>15</th>\n",
              "      <td>CY00000031</td>\n",
              "      <td>102</td>\n",
              "      <td>125</td>\n",
              "      <td>-1</td>\n",
              "      <td>-</td>\n",
              "      <td>TCCACCGGCGCCGCGTCCAG</td>\n",
              "      <td>CGG</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>0</td>\n",
              "      <td>15</td>\n",
              "      <td>231</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>6</th>\n",
              "      <td>CY00000016</td>\n",
              "      <td>117</td>\n",
              "      <td>140</td>\n",
              "      <td>-1</td>\n",
              "      <td>-</td>\n",
              "      <td>CCCCTCGATGCCGGATCCAC</td>\n",
              "      <td>CGG</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>0</td>\n",
              "      <td>1</td>\n",
              "      <td>27</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>9</th>\n",
              "      <td>CY00000019</td>\n",
              "      <td>126</td>\n",
              "      <td>149</td>\n",
              "      <td>-1</td>\n",
              "      <td>-</td>\n",
              "      <td>GATACGGGACCCCTCGATGC</td>\n",
              "      <td>CGG</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>0</td>\n",
              "      <td>1</td>\n",
              "      <td>56</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>7</th>\n",
              "      <td>CY00000017</td>\n",
              "      <td>141</td>\n",
              "      <td>164</td>\n",
              "      <td>-1</td>\n",
              "      <td>-</td>\n",
              "      <td>GGAGGCTCGAAGGCCGATAC</td>\n",
              "      <td>GGG</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>0</td>\n",
              "      <td>1</td>\n",
              "      <td>27</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>5</th>\n",
              "      <td>CY00000013</td>\n",
              "      <td>142</td>\n",
              "      <td>165</td>\n",
              "      <td>-1</td>\n",
              "      <td>-</td>\n",
              "      <td>AGGAGGCTCGAAGGCCGATA</td>\n",
              "      <td>CGG</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>0</td>\n",
              "      <td>1</td>\n",
              "      <td>20</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>CY00000005</td>\n",
              "      <td>151</td>\n",
              "      <td>174</td>\n",
              "      <td>-1</td>\n",
              "      <td>-</td>\n",
              "      <td>GTGGCTCGAAGGAGGCTCGA</td>\n",
              "      <td>AGG</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>11</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "            ID  Start  End  Strand ORF              Sequence  PAM  \\\n",
              "4   CY00000011      6   29      -1   -  AGGGGAACACATGGCCACGC  AGG   \n",
              "0   CY00000003     15   38      -1   -  GAGGCAGGGAGGGGAACACA  TGG   \n",
              "10  CY00000020     24   47      -1   -  AGGGACCACGAGGCAGGGAG  GGG   \n",
              "8   CY00000018     25   48      -1   -  GAGGGACCACGAGGCAGGGA  GGG   \n",
              "2   CY00000008     26   49      -1   -  TGAGGGACCACGAGGCAGGG  AGG   \n",
              "11  CY00000021     29   52      -1   -  GCGTGAGGGACCACGAGGCA  GGG   \n",
              "14  CY00000027     30   53      -1   -  CGCGTGAGGGACCACGAGGC  AGG   \n",
              "13  CY00000026     34   57      -1   -  TGAGCGCGTGAGGGACCACG  AGG   \n",
              "12  CY00000025     43   66      -1   -  GCCCAAAGCTGAGCGCGTGA  GGG   \n",
              "3   CY00000009     44   67      -1   -  CGCCCAAAGCTGAGCGCGTG  AGG   \n",
              "16  CY00000032     68   91      -1   -  CCTTCGACCGCCGCTCGAGC  CGG   \n",
              "15  CY00000031    102  125      -1   -  TCCACCGGCGCCGCGTCCAG  CGG   \n",
              "6   CY00000016    117  140      -1   -  CCCCTCGATGCCGGATCCAC  CGG   \n",
              "9   CY00000019    126  149      -1   -  GATACGGGACCCCTCGATGC  CGG   \n",
              "7   CY00000017    141  164      -1   -  GGAGGCTCGAAGGCCGATAC  GGG   \n",
              "5   CY00000013    142  165      -1   -  AGGAGGCTCGAAGGCCGATA  CGG   \n",
              "1   CY00000005    151  174      -1   -  GTGGCTCGAAGGAGGCTCGA  AGG   \n",
              "\n",
              "    C to T mutations  A to G mutations  0bp mismatches  1bp mismatches  \\\n",
              "4                NaN               NaN               0               0   \n",
              "0                NaN               NaN               0               0   \n",
              "10               NaN               NaN               0               2   \n",
              "8                NaN               NaN               0               1   \n",
              "2                NaN               NaN               0               0   \n",
              "11               NaN               NaN               0               2   \n",
              "14               NaN               NaN               0               5   \n",
              "13               NaN               NaN               0               4   \n",
              "12               NaN               NaN               0               3   \n",
              "3                NaN               NaN               0               0   \n",
              "16               NaN               NaN               0              16   \n",
              "15               NaN               NaN               0              15   \n",
              "6                NaN               NaN               0               1   \n",
              "9                NaN               NaN               0               1   \n",
              "7                NaN               NaN               0               1   \n",
              "5                NaN               NaN               0               1   \n",
              "1                NaN               NaN               0               0   \n",
              "\n",
              "    2bp mismatches  \n",
              "4               50  \n",
              "0               11  \n",
              "10              30  \n",
              "8               34  \n",
              "2               26  \n",
              "11              36  \n",
              "14             117  \n",
              "13              36  \n",
              "12              59  \n",
              "3               29  \n",
              "16             340  \n",
              "15             231  \n",
              "6               27  \n",
              "9               56  \n",
              "7               27  \n",
              "5               20  \n",
              "1               11  "
            ]
          },
          "execution_count": 7,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "# CRISPyweb output\n",
        "import pandas as pd\n",
        "crispy_web_df = pd.read_csv('../../tests/test_files/CRISPy-web_CRISPRi_region_SCO5087_s_coelicolorA3_0mismatch.csv').sort_values(by=['Start'])\n",
        "crispy_web_df"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 8,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/plain": [
              "17"
            ]
          },
          "execution_count": 8,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "len(crispy_web_df)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 9,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>strain_name</th>\n",
              "      <th>locus_tag</th>\n",
              "      <th>gene_loc</th>\n",
              "      <th>gene_strand</th>\n",
              "      <th>sgrna_strand</th>\n",
              "      <th>sgrna_loc</th>\n",
              "      <th>gc</th>\n",
              "      <th>pam</th>\n",
              "      <th>sgrna</th>\n",
              "      <th>sgrna_seed_sequence</th>\n",
              "      <th>off_target_count</th>\n",
              "      <th>region</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>304</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5087</td>\n",
              "      <td>5529601</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>-90</td>\n",
              "      <td>0.65</td>\n",
              "      <td>AGG</td>\n",
              "      <td>AGGGGAACACATGGCCACGC</td>\n",
              "      <td>CACATGGCCACGC</td>\n",
              "      <td>0</td>\n",
              "      <td>upstream</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>305</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5087</td>\n",
              "      <td>5529601</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>-81</td>\n",
              "      <td>0.65</td>\n",
              "      <td>TGG</td>\n",
              "      <td>GAGGCAGGGAGGGGAACACA</td>\n",
              "      <td>GGAGGGGAACACA</td>\n",
              "      <td>0</td>\n",
              "      <td>upstream</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>325</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5087</td>\n",
              "      <td>5529601</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>-80</td>\n",
              "      <td>0.70</td>\n",
              "      <td>TGG</td>\n",
              "      <td>GTGTTCCCCTCCCTGCCTCG</td>\n",
              "      <td>CCTCCCTGCCTCG</td>\n",
              "      <td>0</td>\n",
              "      <td>upstream</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>306</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5087</td>\n",
              "      <td>5529601</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>-72</td>\n",
              "      <td>0.70</td>\n",
              "      <td>GGG</td>\n",
              "      <td>AGGGACCACGAGGCAGGGAG</td>\n",
              "      <td>ACGAGGCAGGGAG</td>\n",
              "      <td>0</td>\n",
              "      <td>upstream</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>307</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5087</td>\n",
              "      <td>5529601</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>-71</td>\n",
              "      <td>0.70</td>\n",
              "      <td>GGG</td>\n",
              "      <td>GAGGGACCACGAGGCAGGGA</td>\n",
              "      <td>CACGAGGCAGGGA</td>\n",
              "      <td>0</td>\n",
              "      <td>upstream</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "     strain_name locus_tag  gene_loc  gene_strand  sgrna_strand  sgrna_loc  \\\n",
              "304  NC_003888.3   SCO5087   5529601            1            -1        -90   \n",
              "305  NC_003888.3   SCO5087   5529601            1            -1        -81   \n",
              "325  NC_003888.3   SCO5087   5529601            1             1        -80   \n",
              "306  NC_003888.3   SCO5087   5529601            1            -1        -72   \n",
              "307  NC_003888.3   SCO5087   5529601            1            -1        -71   \n",
              "\n",
              "       gc  pam                 sgrna sgrna_seed_sequence  off_target_count  \\\n",
              "304  0.65  AGG  AGGGGAACACATGGCCACGC       CACATGGCCACGC                 0   \n",
              "305  0.65  TGG  GAGGCAGGGAGGGGAACACA       GGAGGGGAACACA                 0   \n",
              "325  0.70  TGG  GTGTTCCCCTCCCTGCCTCG       CCTCCCTGCCTCG                 0   \n",
              "306  0.70  GGG  AGGGACCACGAGGCAGGGAG       ACGAGGCAGGGAG                 0   \n",
              "307  0.70  GGG  GAGGGACCACGAGGCAGGGA       CACGAGGCAGGGA                 0   \n",
              "\n",
              "       region  \n",
              "304  upstream  \n",
              "305  upstream  \n",
              "325  upstream  \n",
              "306  upstream  \n",
              "307  upstream  "
            ]
          },
          "execution_count": 9,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "# Filter the DataFrame to retain only up to 5 sgRNA sequences per locus_tag\n",
        "filtered_df = sgrna_df.groupby('locus_tag').head(number_of_sgRNAs_per_group)\n",
        "filtered_df"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "## Output"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 10,
      "metadata": {},
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "SCO5087_loc_-90\n"
          ]
        },
        {
          "data": {
            "text/plain": [
              "[Contig(o11279),\n",
              " Contig(o11279),\n",
              " Contig(o11279),\n",
              " Contig(o11279),\n",
              " Contig(o11279)]"
            ]
          },
          "execution_count": 10,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "# MAke oligoes\n",
        "list_of_ssDNAs = make_ssDNA_oligos(filtered_df, upstream_ovh = up_homology,\n",
        "                      downstream_ovh=dw_homology)\n",
        "print(list_of_ssDNAs[0].name)\n",
        "\n",
        "# cut plasmid\n",
        "from Bio.Restriction import NcoI\n",
        "linearized_plasmid = sorted(clean_plasmid.cut(NcoI), key=lambda x: len(x), reverse=True)[0]\n",
        "#print(linearized_plasmid)\n",
        "\n",
        "sgRNA_vectors = assemble_plasmids_by_ssDNA_bridging(list_of_ssDNAs,linearized_plasmid)\n",
        "sgRNA_vectors"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 11,
      "metadata": {},
      "outputs": [],
      "source": [
        "# Constructing a meaningful name, ID, and description for the assembled plasmid using user input\n",
        "targeting_info = []\n",
        "for index, row in filtered_df.iterrows():\n",
        "    formatted_str = f\"CRISPRi_{row['locus_tag']}_p{row['sgrna_loc']}\"\n",
        "    targeting_info.append(formatted_str)\n",
        "\n",
        "for i in range(len(sgRNA_vectors)):\n",
        "    sgRNA_vectors[i].name = f'p{targeting_info[i]}_#{i+1}'\n",
        "    sgRNA_vectors[i].id = sgRNA_vectors[i].name  # Using the same value for ID as for name for simplicity\n",
        "    sgRNA_vectors[i].description = f'Assembled plasmid targeting {\", \".join(genes_to_KO)} for single gene KNOCK-DOWN, assembled using StreptoCAD.'\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 12,
      "metadata": {},
      "outputs": [],
      "source": [
        "print_plasmids = False\n",
        "\n",
        "if print_plasmids: \n",
        "    for vector in sgRNA_vectors: \n",
        "        vector.write(f\"../../data/plasmids/sgRNA_plasmids_pCRISPR–dCas9/{vector.id}.gb\")"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 13,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>plasmid_name</th>\n",
              "      <th>date</th>\n",
              "      <th>original_plasmid</th>\n",
              "      <th>integration</th>\n",
              "      <th>size</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>pCRISPRi_SCO5087_p-90_#1</td>\n",
              "      <td>2025-05-28</td>\n",
              "      <td>pCRISPR-Cas9</td>\n",
              "      <td>sgRNA_SCO5087(-90)</td>\n",
              "      <td>11279</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>pCRISPRi_SCO5087_p-81_#2</td>\n",
              "      <td>2025-05-28</td>\n",
              "      <td>pCRISPR-Cas9</td>\n",
              "      <td>sgRNA_SCO5087(-81)</td>\n",
              "      <td>11279</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>pCRISPRi_SCO5087_p-80_#3</td>\n",
              "      <td>2025-05-28</td>\n",
              "      <td>pCRISPR-Cas9</td>\n",
              "      <td>sgRNA_SCO5087(-80)</td>\n",
              "      <td>11279</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>pCRISPRi_SCO5087_p-72_#4</td>\n",
              "      <td>2025-05-28</td>\n",
              "      <td>pCRISPR-Cas9</td>\n",
              "      <td>sgRNA_SCO5087(-72)</td>\n",
              "      <td>11279</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>pCRISPRi_SCO5087_p-71_#5</td>\n",
              "      <td>2025-05-28</td>\n",
              "      <td>pCRISPR-Cas9</td>\n",
              "      <td>sgRNA_SCO5087(-71)</td>\n",
              "      <td>11279</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "               plasmid_name        date original_plasmid         integration  \\\n",
              "0  pCRISPRi_SCO5087_p-90_#1  2025-05-28     pCRISPR-Cas9  sgRNA_SCO5087(-90)   \n",
              "1  pCRISPRi_SCO5087_p-81_#2  2025-05-28     pCRISPR-Cas9  sgRNA_SCO5087(-81)   \n",
              "2  pCRISPRi_SCO5087_p-80_#3  2025-05-28     pCRISPR-Cas9  sgRNA_SCO5087(-80)   \n",
              "3  pCRISPRi_SCO5087_p-72_#4  2025-05-28     pCRISPR-Cas9  sgRNA_SCO5087(-72)   \n",
              "4  pCRISPRi_SCO5087_p-71_#5  2025-05-28     pCRISPR-Cas9  sgRNA_SCO5087(-71)   \n",
              "\n",
              "    size  \n",
              "0  11279  \n",
              "1  11279  \n",
              "2  11279  \n",
              "3  11279  \n",
              "4  11279  "
            ]
          },
          "execution_count": 13,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "integration_names = filtered_df.apply(lambda row: f\"sgRNA_{row['locus_tag']}({row['sgrna_loc']})\", axis=1).tolist()\n",
        "plasmid_metadata_df = extract_metadata_to_dataframe(sgRNA_vectors,\n",
        "                                                    clean_plasmid,\n",
        "                                                    integration_names)\n",
        "\n",
        "plasmid_metadata_df"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "### IDT primers"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 14,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>Name</th>\n",
              "      <th>Sequence</th>\n",
              "      <th>Concentration</th>\n",
              "      <th>Purification</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>SCO5087_loc_-90</td>\n",
              "      <td>CGGTTGGTAGGATCGACGGCAGGGGAACACATGGCCACGCGTTTTA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>SCO5087_loc_-81</td>\n",
              "      <td>CGGTTGGTAGGATCGACGGCGAGGCAGGGAGGGGAACACAGTTTTA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>SCO5087_loc_-80</td>\n",
              "      <td>CGGTTGGTAGGATCGACGGCGTGTTCCCCTCCCTGCCTCGGTTTTA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>SCO5087_loc_-72</td>\n",
              "      <td>CGGTTGGTAGGATCGACGGCAGGGACCACGAGGCAGGGAGGTTTTA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>SCO5087_loc_-71</td>\n",
              "      <td>CGGTTGGTAGGATCGACGGCGAGGGACCACGAGGCAGGGAGTTTTA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "              Name                                           Sequence  \\\n",
              "0  SCO5087_loc_-90  CGGTTGGTAGGATCGACGGCAGGGGAACACATGGCCACGCGTTTTA...   \n",
              "1  SCO5087_loc_-81  CGGTTGGTAGGATCGACGGCGAGGCAGGGAGGGGAACACAGTTTTA...   \n",
              "2  SCO5087_loc_-80  CGGTTGGTAGGATCGACGGCGTGTTCCCCTCCCTGCCTCGGTTTTA...   \n",
              "3  SCO5087_loc_-72  CGGTTGGTAGGATCGACGGCAGGGACCACGAGGCAGGGAGGTTTTA...   \n",
              "4  SCO5087_loc_-71  CGGTTGGTAGGATCGACGGCGAGGGACCACGAGGCAGGGAGTTTTA...   \n",
              "\n",
              "  Concentration Purification  \n",
              "0          25nm          STD  \n",
              "1          25nm          STD  \n",
              "2          25nm          STD  \n",
              "3          25nm          STD  \n",
              "4          25nm          STD  "
            ]
          },
          "execution_count": 14,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "idt_primers=primers_to_IDT(list_of_ssDNAs)\n",
        "idt_primers"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "## Folder with all the generated I/O"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 15,
      "metadata": {},
      "outputs": [],
      "source": [
        "generate_data_folder = True \n",
        "\n",
        "if generate_data_folder:\n",
        "    input_files = [\n",
        "        {\"name\": \"input_genome.gb\", \"content\": genome},\n",
        "        {\"name\": \"input_plasmid.gb\", \"content\": clean_plasmid}\n",
        "    ]\n",
        "\n",
        "    output_files = [\n",
        "        {\"name\": \"cBEST_w_sgRNAs.gb\", \"content\": sgRNA_vectors}, # LIST OF Dseqrecords\n",
        "        {\"name\": \"full_idt.csv\", \"content\": idt_primers},\n",
        "        {\"name\": \"sgrna_df.csv\", \"content\": sgrna_df},\n",
        "        {\"name\": \"filtered_df.csv\", \"content\": filtered_df},\n",
        "        {\"name\": \"plasmid_metadata_df.csv\", \"content\": plasmid_metadata_df},\n",
        "    ]\n",
        "\n",
        "    input_values = {\n",
        "        \"genes_to_knockout\": genes_to_KO,\n",
        "\n",
        "\n",
        "        \"filtering_metrics\": {\n",
        "            \"gc_upper\": gc_upper,\n",
        "            \"gc_lower\": gc_lower,\n",
        "            \"off_target_seed\": off_target_seed,\n",
        "            \"off_target_upper\": off_target_upper,\n",
        "            \"cas_type\": cas_type,\n",
        "            \"number_of_sgRNAs_per_group\": number_of_sgRNAs_per_group,\n",
        "            'extension_to_promoter_region':extension_to_promoter_region,\n",
        "\n",
        "        },\n",
        "        \"overlapping_sequences\": {\n",
        "            \"up_homology\": str(up_homology),\n",
        "            \"dw_homology\": str(dw_homology)\n",
        "        }\n",
        "    }\n",
        "\n",
        "\n",
        "    # Paths to Markdown files\n",
        "    markdown_file_paths = [\n",
        "        \"../../protocols/conjugation_protcol.md\",\n",
        "        \"../../protocols/single_target_crispr_plasmid_protcol.md\"\n",
        "\n",
        "    ]\n",
        "\n",
        "\n",
        "\n",
        "    # Data and time\n",
        "    timestamp = datetime.utcnow().isoformat()\n",
        "\n",
        "    # Create project directory structure\n",
        "    project_directory = ProjectDirectory(\n",
        "        project_name=f\"CRISPRi_workflow_{timestamp}\",\n",
        "        input_files=input_files,\n",
        "        output_files=output_files,\n",
        "        input_values=input_values,\n",
        "        markdown_file_paths=markdown_file_paths\n",
        "    )\n",
        "\n",
        "\n",
        "    # DO You want to save the folder? \n",
        "    save_zip_folder = False \n",
        "\n",
        "    if save_zip_folder: \n",
        "        # Generate the project directory structure and get the zip content\n",
        "        zip_content = project_directory.create_directory_structure(create_directories=False)\n",
        "\n",
        "        # Save the zip file to disk (optional)\n",
        "        with open(\"project_structure.zip\", \"wb\") as f:\n",
        "            f.write(zip_content)"
      ]
    }
  ],
  "metadata": {
    "kernelspec": {
      "display_name": "streptocad",
      "language": "python",
      "name": "python3"
    },
    "language_info": {
      "codemirror_mode": {
        "name": "ipython",
        "version": 3
      },
      "file_extension": ".py",
      "mimetype": "text/x-python",
      "name": "python",
      "nbconvert_exporter": "python",
      "pygments_lexer": "ipython3",
      "version": "3.11.11"
    }
  },
  "nbformat": 4,
  "nbformat_minor": 2
}
