{
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "# W4-CRISPRi-workflow"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "Transcriptional interference is another powerful tool for assessing the function of genes, by expressing a guide RNA that targets a dead-Cas9/sgRNA complex to bind to the + (Watson) strand of the promoter and 5’-UTR of the targeted gene, thereby sterically blocking binding and progression of RNA polymerases to the promoter region. Workflow 4 uses CRISPR interference (CRISPRi) with ssDNA bridging to reversibly inactivate genes transcriptionally12,29 (Figure 5A). This approach targets regions upstream of the Transcriptional Start Site (TSS), using a dCas9-sgRNA complex positioned near the TSS (default 100bp upstream) to sterically hinder transcription. This allows for the functional study of genes through controlled knockdown. To get started, users can download the pCRISPR-dCas9 plasmid and the S. coelicolor (A3) genome. StreptoAIM will then generate all necessary components, including primers and plasmids."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 1,
      "metadata": {},
      "outputs": [],
      "source": [
        "import sys\n",
        "import os\n",
        "from pydna.dseqrecord import Dseqrecord\n",
        "from datetime import datetime\n",
        "\n",
        "\n",
        "# Ensure the src directory is in the Python path\n",
        "project_root = os.path.abspath(os.path.join(os.getcwd(), '../../'))\n",
        "if project_root not in sys.path:\n",
        "    sys.path.insert(0, project_root)\n",
        "\n",
        "from streptocad.sequence_loading.sequence_loading import (\n",
        "    load_and_process_plasmid,\n",
        "    load_and_process_genome_sequences\n",
        ")\n",
        "from streptocad.cloning.ssDNA_bridging import assemble_plasmids_by_ssDNA_bridging, make_ssDNA_oligos\n",
        "from streptocad.crispr.guideRNA_crispri import extract_sgRNAs_for_crispri, SgRNAargs\n",
        "from streptocad.primers.primer_generation import  primers_to_IDT\n",
        "from streptocad.utils import generate_project_directory_structure"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "## INPUT"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 2,
      "metadata": {},
      "outputs": [],
      "source": [
        "# Inputs\n",
        "# 1 Add genome of choice (genbank, fasta)\n",
        "path_to_genome = '../../data/genomes/Streptomyces_coelicolor_A3_chromosome.gb'\n",
        "genome = load_and_process_genome_sequences(path_to_genome)[0]\n",
        "\n",
        "# 2 Add plasmid \n",
        "path_to_plasmid = '../../data/plasmids/pCRISPR–Cas9_plasmid_addgene.gbk'\n",
        "clean_plasmid = load_and_process_plasmid(path_to_plasmid)\n",
        "\n",
        "# 3 Choose genes to knock out (list)\n",
        "genes_to_KO = ['SCO5087', 'SCO5089', 'SCO5090']\n",
        "\n",
        "#### Advanced settings ####\n",
        "# 4 Filtering metrics for sgRNAs\n",
        "gc_upper = 0.72\n",
        "gc_lower = 0.2\n",
        "off_target_seed = 13\n",
        "off_target_upper = 0\n",
        "cas_type='cas9'\n",
        "number_of_sgRNAs_per_group = 5\n",
        "extension_to_promoter_region=200\n",
        "\n",
        "# 6 Choose overlapping sequences for our plasmid we can use the following\n",
        "#As per the article **\"CRISPR–Cas9, CRISPRi and CRISPR-BEST-mediated genetic manipulation in streptomycetes\"** we need the following oligoes: \n",
        "#CGGTTGGTAGGATCGACGGC **-N20-** GTTTTAGAGCTAGAAATAGC\n",
        "up_homology = Dseqrecord('CGGTTGGTAGGATCGACGGC')\n",
        "dw_homology = Dseqrecord('GTTTTAGAGCTAGAAATAGC')"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "# Computation"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 3,
      "metadata": {},
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "sgRNA or PAM generated were outside the designated border in SCO5087. Skipping to next locus tag.\n",
            "sgRNA or PAM generated were outside the designated border in SCO5087. Skipping to next locus tag.\n",
            "sgRNA or PAM generated were outside the designated border in SCO5089. Skipping to next locus tag.\n",
            "sgRNA or PAM generated were outside the designated border in SCO5089. Skipping to next locus tag.\n",
            "sgRNA or PAM generated were outside the designated border in SCO5089. Skipping to next locus tag.\n",
            "sgRNA or PAM generated were outside the designated border in SCO5089. Skipping to next locus tag.\n",
            "sgRNA or PAM generated were outside the designated border in SCO5090. Skipping to next locus tag.\n",
            "sgRNA or PAM generated were outside the designated border in SCO5090. Skipping to next locus tag.\n",
            "sgRNA or PAM generated were outside the designated border in SCO5090. Skipping to next locus tag.\n",
            "sgRNA or PAM generated were outside the designated border in SCO5090. Skipping to next locus tag.\n",
            "sgRNA or PAM generated were outside the designated border in SCO5090. Skipping to next locus tag.\n",
            "sgRNA or PAM generated were outside the designated border in SCO5090. Skipping to next locus tag.\n",
            "sgRNA or PAM generated were outside the designated border in SCO5090. Skipping to next locus tag.\n"
          ]
        },
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>strain_name</th>\n",
              "      <th>locus_tag</th>\n",
              "      <th>gene_loc</th>\n",
              "      <th>gene_strand</th>\n",
              "      <th>sgrna_strand</th>\n",
              "      <th>sgrna_loc</th>\n",
              "      <th>gc</th>\n",
              "      <th>pam</th>\n",
              "      <th>sgrna</th>\n",
              "      <th>sgrna_seed_sequence</th>\n",
              "      <th>off_target_count</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>11</th>\n",
              "      <td>Streptomyces_coelicolor_A3</td>\n",
              "      <td>SCO5087</td>\n",
              "      <td>5529801</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>-95</td>\n",
              "      <td>0.60</td>\n",
              "      <td>GGG</td>\n",
              "      <td>TCCGAAGTCGCACTGGTGAC</td>\n",
              "      <td>TCGCACTGGTGAC</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>12</th>\n",
              "      <td>Streptomyces_coelicolor_A3</td>\n",
              "      <td>SCO5087</td>\n",
              "      <td>5529801</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>-94</td>\n",
              "      <td>0.60</td>\n",
              "      <td>CGG</td>\n",
              "      <td>CTCCGAAGTCGCACTGGTGA</td>\n",
              "      <td>GTCGCACTGGTGA</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>13</th>\n",
              "      <td>Streptomyces_coelicolor_A3</td>\n",
              "      <td>SCO5087</td>\n",
              "      <td>5529801</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>-88</td>\n",
              "      <td>0.65</td>\n",
              "      <td>TGG</td>\n",
              "      <td>GCAGGACTCCGAAGTCGCAC</td>\n",
              "      <td>TCCGAAGTCGCAC</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>434</th>\n",
              "      <td>Streptomyces_coelicolor_A3</td>\n",
              "      <td>SCO5090</td>\n",
              "      <td>5532706</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>-73</td>\n",
              "      <td>0.65</td>\n",
              "      <td>CGG</td>\n",
              "      <td>CGGGTATGGACACGCCGTAC</td>\n",
              "      <td>GGACACGCCGTAC</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>14</th>\n",
              "      <td>Streptomyces_coelicolor_A3</td>\n",
              "      <td>SCO5087</td>\n",
              "      <td>5529801</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>-70</td>\n",
              "      <td>0.65</td>\n",
              "      <td>AGG</td>\n",
              "      <td>AGGGGAACACATGGCCACGC</td>\n",
              "      <td>CACATGGCCACGC</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>15</th>\n",
              "      <td>Streptomyces_coelicolor_A3</td>\n",
              "      <td>SCO5087</td>\n",
              "      <td>5529801</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>-61</td>\n",
              "      <td>0.65</td>\n",
              "      <td>TGG</td>\n",
              "      <td>GAGGCAGGGAGGGGAACACA</td>\n",
              "      <td>GGAGGGGAACACA</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>351</th>\n",
              "      <td>Streptomyces_coelicolor_A3</td>\n",
              "      <td>SCO5089</td>\n",
              "      <td>5532449</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>-56</td>\n",
              "      <td>0.70</td>\n",
              "      <td>CGG</td>\n",
              "      <td>CCGCCGAGTTGAAGCCCCAG</td>\n",
              "      <td>GTTGAAGCCCCAG</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>16</th>\n",
              "      <td>Streptomyces_coelicolor_A3</td>\n",
              "      <td>SCO5087</td>\n",
              "      <td>5529801</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>-52</td>\n",
              "      <td>0.70</td>\n",
              "      <td>GGG</td>\n",
              "      <td>AGGGACCACGAGGCAGGGAG</td>\n",
              "      <td>ACGAGGCAGGGAG</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>17</th>\n",
              "      <td>Streptomyces_coelicolor_A3</td>\n",
              "      <td>SCO5087</td>\n",
              "      <td>5529801</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>-51</td>\n",
              "      <td>0.70</td>\n",
              "      <td>GGG</td>\n",
              "      <td>GAGGGACCACGAGGCAGGGA</td>\n",
              "      <td>CACGAGGCAGGGA</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>18</th>\n",
              "      <td>Streptomyces_coelicolor_A3</td>\n",
              "      <td>SCO5087</td>\n",
              "      <td>5529801</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>-50</td>\n",
              "      <td>0.70</td>\n",
              "      <td>AGG</td>\n",
              "      <td>TGAGGGACCACGAGGCAGGG</td>\n",
              "      <td>CCACGAGGCAGGG</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>19</th>\n",
              "      <td>Streptomyces_coelicolor_A3</td>\n",
              "      <td>SCO5087</td>\n",
              "      <td>5529801</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>-47</td>\n",
              "      <td>0.70</td>\n",
              "      <td>GGG</td>\n",
              "      <td>GCGTGAGGGACCACGAGGCA</td>\n",
              "      <td>GGACCACGAGGCA</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>21</th>\n",
              "      <td>Streptomyces_coelicolor_A3</td>\n",
              "      <td>SCO5087</td>\n",
              "      <td>5529801</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>-42</td>\n",
              "      <td>0.70</td>\n",
              "      <td>AGG</td>\n",
              "      <td>TGAGCGCGTGAGGGACCACG</td>\n",
              "      <td>GTGAGGGACCACG</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>22</th>\n",
              "      <td>Streptomyces_coelicolor_A3</td>\n",
              "      <td>SCO5087</td>\n",
              "      <td>5529801</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>-33</td>\n",
              "      <td>0.65</td>\n",
              "      <td>GGG</td>\n",
              "      <td>GCCCAAAGCTGAGCGCGTGA</td>\n",
              "      <td>GCTGAGCGCGTGA</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>23</th>\n",
              "      <td>Streptomyces_coelicolor_A3</td>\n",
              "      <td>SCO5087</td>\n",
              "      <td>5529801</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>-32</td>\n",
              "      <td>0.70</td>\n",
              "      <td>AGG</td>\n",
              "      <td>CGCCCAAAGCTGAGCGCGTG</td>\n",
              "      <td>AGCTGAGCGCGTG</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>440</th>\n",
              "      <td>Streptomyces_coelicolor_A3</td>\n",
              "      <td>SCO5090</td>\n",
              "      <td>5532706</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>-24</td>\n",
              "      <td>0.60</td>\n",
              "      <td>CGG</td>\n",
              "      <td>ATCAGGTCGAGCAGCTCTCG</td>\n",
              "      <td>CGAGCAGCTCTCG</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>353</th>\n",
              "      <td>Streptomyces_coelicolor_A3</td>\n",
              "      <td>SCO5089</td>\n",
              "      <td>5532449</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>-23</td>\n",
              "      <td>0.60</td>\n",
              "      <td>CGG</td>\n",
              "      <td>GTTACGGGGTCGGTGCGAAA</td>\n",
              "      <td>GGTCGGTGCGAAA</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>354</th>\n",
              "      <td>Streptomyces_coelicolor_A3</td>\n",
              "      <td>SCO5089</td>\n",
              "      <td>5532449</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>-13</td>\n",
              "      <td>0.65</td>\n",
              "      <td>CGG</td>\n",
              "      <td>TCTCGGGGTCGTTACGGGGT</td>\n",
              "      <td>GTCGTTACGGGGT</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>355</th>\n",
              "      <td>Streptomyces_coelicolor_A3</td>\n",
              "      <td>SCO5089</td>\n",
              "      <td>5532449</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>-9</td>\n",
              "      <td>0.60</td>\n",
              "      <td>GGG</td>\n",
              "      <td>CAGTTCTCGGGGTCGTTACG</td>\n",
              "      <td>CGGGGTCGTTACG</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>356</th>\n",
              "      <td>Streptomyces_coelicolor_A3</td>\n",
              "      <td>SCO5089</td>\n",
              "      <td>5532449</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>-8</td>\n",
              "      <td>0.60</td>\n",
              "      <td>GGG</td>\n",
              "      <td>CCAGTTCTCGGGGTCGTTAC</td>\n",
              "      <td>TCGGGGTCGTTAC</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>441</th>\n",
              "      <td>Streptomyces_coelicolor_A3</td>\n",
              "      <td>SCO5090</td>\n",
              "      <td>5532706</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>-7</td>\n",
              "      <td>0.70</td>\n",
              "      <td>AGG</td>\n",
              "      <td>CGGCCAGTGCGCCGTTGATC</td>\n",
              "      <td>TGCGCCGTTGATC</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>357</th>\n",
              "      <td>Streptomyces_coelicolor_A3</td>\n",
              "      <td>SCO5089</td>\n",
              "      <td>5532449</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>-7</td>\n",
              "      <td>0.55</td>\n",
              "      <td>CGG</td>\n",
              "      <td>TCCAGTTCTCGGGGTCGTTA</td>\n",
              "      <td>CTCGGGGTCGTTA</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>358</th>\n",
              "      <td>Streptomyces_coelicolor_A3</td>\n",
              "      <td>SCO5089</td>\n",
              "      <td>5532449</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>2</td>\n",
              "      <td>0.50</td>\n",
              "      <td>GGG</td>\n",
              "      <td>ATGGCTTTCTCCAGTTCTCG</td>\n",
              "      <td>TCTCCAGTTCTCG</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>359</th>\n",
              "      <td>Streptomyces_coelicolor_A3</td>\n",
              "      <td>SCO5089</td>\n",
              "      <td>5532449</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>3</td>\n",
              "      <td>0.50</td>\n",
              "      <td>GGG</td>\n",
              "      <td>CATGGCTTTCTCCAGTTCTC</td>\n",
              "      <td>TTCTCCAGTTCTC</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>360</th>\n",
              "      <td>Streptomyces_coelicolor_A3</td>\n",
              "      <td>SCO5089</td>\n",
              "      <td>5532449</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>4</td>\n",
              "      <td>0.50</td>\n",
              "      <td>CGG</td>\n",
              "      <td>CCATGGCTTTCTCCAGTTCT</td>\n",
              "      <td>TTTCTCCAGTTCT</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>361</th>\n",
              "      <td>Streptomyces_coelicolor_A3</td>\n",
              "      <td>SCO5089</td>\n",
              "      <td>5532449</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>21</td>\n",
              "      <td>0.65</td>\n",
              "      <td>TGG</td>\n",
              "      <td>GGTGGTCAGCAGGGTTGCCA</td>\n",
              "      <td>AGCAGGGTTGCCA</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>362</th>\n",
              "      <td>Streptomyces_coelicolor_A3</td>\n",
              "      <td>SCO5089</td>\n",
              "      <td>5532449</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>30</td>\n",
              "      <td>0.60</td>\n",
              "      <td>GGG</td>\n",
              "      <td>CAGATCGTCGGTGGTCAGCA</td>\n",
              "      <td>TCGGTGGTCAGCA</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>363</th>\n",
              "      <td>Streptomyces_coelicolor_A3</td>\n",
              "      <td>SCO5089</td>\n",
              "      <td>5532449</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>31</td>\n",
              "      <td>0.65</td>\n",
              "      <td>AGG</td>\n",
              "      <td>GCAGATCGTCGGTGGTCAGC</td>\n",
              "      <td>GTCGGTGGTCAGC</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>444</th>\n",
              "      <td>Streptomyces_coelicolor_A3</td>\n",
              "      <td>SCO5090</td>\n",
              "      <td>5532706</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>32</td>\n",
              "      <td>0.65</td>\n",
              "      <td>CGG</td>\n",
              "      <td>TGGACCACCCGGTGTTCTCC</td>\n",
              "      <td>CCCGGTGTTCTCC</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>27</th>\n",
              "      <td>Streptomyces_coelicolor_A3</td>\n",
              "      <td>SCO5087</td>\n",
              "      <td>5529801</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>41</td>\n",
              "      <td>0.70</td>\n",
              "      <td>CGG</td>\n",
              "      <td>CCCCTCGATGCCGGATCCAC</td>\n",
              "      <td>ATGCCGGATCCAC</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>28</th>\n",
              "      <td>Streptomyces_coelicolor_A3</td>\n",
              "      <td>SCO5087</td>\n",
              "      <td>5529801</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>50</td>\n",
              "      <td>0.65</td>\n",
              "      <td>CGG</td>\n",
              "      <td>GATACGGGACCCCTCGATGC</td>\n",
              "      <td>GACCCCTCGATGC</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>29</th>\n",
              "      <td>Streptomyces_coelicolor_A3</td>\n",
              "      <td>SCO5087</td>\n",
              "      <td>5529801</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>65</td>\n",
              "      <td>0.65</td>\n",
              "      <td>GGG</td>\n",
              "      <td>GGAGGCTCGAAGGCCGATAC</td>\n",
              "      <td>CGAAGGCCGATAC</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>30</th>\n",
              "      <td>Streptomyces_coelicolor_A3</td>\n",
              "      <td>SCO5087</td>\n",
              "      <td>5529801</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>66</td>\n",
              "      <td>0.60</td>\n",
              "      <td>CGG</td>\n",
              "      <td>AGGAGGCTCGAAGGCCGATA</td>\n",
              "      <td>TCGAAGGCCGATA</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>31</th>\n",
              "      <td>Streptomyces_coelicolor_A3</td>\n",
              "      <td>SCO5087</td>\n",
              "      <td>5529801</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>75</td>\n",
              "      <td>0.65</td>\n",
              "      <td>AGG</td>\n",
              "      <td>GTGGCTCGAAGGAGGCTCGA</td>\n",
              "      <td>GAAGGAGGCTCGA</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>369</th>\n",
              "      <td>Streptomyces_coelicolor_A3</td>\n",
              "      <td>SCO5089</td>\n",
              "      <td>5532449</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>75</td>\n",
              "      <td>0.70</td>\n",
              "      <td>CGG</td>\n",
              "      <td>GTCCGTCCCGTCCGTCTCAC</td>\n",
              "      <td>CCGTCCGTCTCAC</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "                    strain_name locus_tag  gene_loc  gene_strand  \\\n",
              "11   Streptomyces_coelicolor_A3   SCO5087   5529801            1   \n",
              "12   Streptomyces_coelicolor_A3   SCO5087   5529801            1   \n",
              "13   Streptomyces_coelicolor_A3   SCO5087   5529801            1   \n",
              "434  Streptomyces_coelicolor_A3   SCO5090   5532706            1   \n",
              "14   Streptomyces_coelicolor_A3   SCO5087   5529801            1   \n",
              "15   Streptomyces_coelicolor_A3   SCO5087   5529801            1   \n",
              "351  Streptomyces_coelicolor_A3   SCO5089   5532449            1   \n",
              "16   Streptomyces_coelicolor_A3   SCO5087   5529801            1   \n",
              "17   Streptomyces_coelicolor_A3   SCO5087   5529801            1   \n",
              "18   Streptomyces_coelicolor_A3   SCO5087   5529801            1   \n",
              "19   Streptomyces_coelicolor_A3   SCO5087   5529801            1   \n",
              "21   Streptomyces_coelicolor_A3   SCO5087   5529801            1   \n",
              "22   Streptomyces_coelicolor_A3   SCO5087   5529801            1   \n",
              "23   Streptomyces_coelicolor_A3   SCO5087   5529801            1   \n",
              "440  Streptomyces_coelicolor_A3   SCO5090   5532706            1   \n",
              "353  Streptomyces_coelicolor_A3   SCO5089   5532449            1   \n",
              "354  Streptomyces_coelicolor_A3   SCO5089   5532449            1   \n",
              "355  Streptomyces_coelicolor_A3   SCO5089   5532449            1   \n",
              "356  Streptomyces_coelicolor_A3   SCO5089   5532449            1   \n",
              "441  Streptomyces_coelicolor_A3   SCO5090   5532706            1   \n",
              "357  Streptomyces_coelicolor_A3   SCO5089   5532449            1   \n",
              "358  Streptomyces_coelicolor_A3   SCO5089   5532449            1   \n",
              "359  Streptomyces_coelicolor_A3   SCO5089   5532449            1   \n",
              "360  Streptomyces_coelicolor_A3   SCO5089   5532449            1   \n",
              "361  Streptomyces_coelicolor_A3   SCO5089   5532449            1   \n",
              "362  Streptomyces_coelicolor_A3   SCO5089   5532449            1   \n",
              "363  Streptomyces_coelicolor_A3   SCO5089   5532449            1   \n",
              "444  Streptomyces_coelicolor_A3   SCO5090   5532706            1   \n",
              "27   Streptomyces_coelicolor_A3   SCO5087   5529801            1   \n",
              "28   Streptomyces_coelicolor_A3   SCO5087   5529801            1   \n",
              "29   Streptomyces_coelicolor_A3   SCO5087   5529801            1   \n",
              "30   Streptomyces_coelicolor_A3   SCO5087   5529801            1   \n",
              "31   Streptomyces_coelicolor_A3   SCO5087   5529801            1   \n",
              "369  Streptomyces_coelicolor_A3   SCO5089   5532449            1   \n",
              "\n",
              "     sgrna_strand  sgrna_loc    gc  pam                 sgrna  \\\n",
              "11             -1        -95  0.60  GGG  TCCGAAGTCGCACTGGTGAC   \n",
              "12             -1        -94  0.60  CGG  CTCCGAAGTCGCACTGGTGA   \n",
              "13             -1        -88  0.65  TGG  GCAGGACTCCGAAGTCGCAC   \n",
              "434            -1        -73  0.65  CGG  CGGGTATGGACACGCCGTAC   \n",
              "14             -1        -70  0.65  AGG  AGGGGAACACATGGCCACGC   \n",
              "15             -1        -61  0.65  TGG  GAGGCAGGGAGGGGAACACA   \n",
              "351            -1        -56  0.70  CGG  CCGCCGAGTTGAAGCCCCAG   \n",
              "16             -1        -52  0.70  GGG  AGGGACCACGAGGCAGGGAG   \n",
              "17             -1        -51  0.70  GGG  GAGGGACCACGAGGCAGGGA   \n",
              "18             -1        -50  0.70  AGG  TGAGGGACCACGAGGCAGGG   \n",
              "19             -1        -47  0.70  GGG  GCGTGAGGGACCACGAGGCA   \n",
              "21             -1        -42  0.70  AGG  TGAGCGCGTGAGGGACCACG   \n",
              "22             -1        -33  0.65  GGG  GCCCAAAGCTGAGCGCGTGA   \n",
              "23             -1        -32  0.70  AGG  CGCCCAAAGCTGAGCGCGTG   \n",
              "440            -1        -24  0.60  CGG  ATCAGGTCGAGCAGCTCTCG   \n",
              "353            -1        -23  0.60  CGG  GTTACGGGGTCGGTGCGAAA   \n",
              "354            -1        -13  0.65  CGG  TCTCGGGGTCGTTACGGGGT   \n",
              "355            -1         -9  0.60  GGG  CAGTTCTCGGGGTCGTTACG   \n",
              "356            -1         -8  0.60  GGG  CCAGTTCTCGGGGTCGTTAC   \n",
              "441            -1         -7  0.70  AGG  CGGCCAGTGCGCCGTTGATC   \n",
              "357            -1         -7  0.55  CGG  TCCAGTTCTCGGGGTCGTTA   \n",
              "358            -1          2  0.50  GGG  ATGGCTTTCTCCAGTTCTCG   \n",
              "359            -1          3  0.50  GGG  CATGGCTTTCTCCAGTTCTC   \n",
              "360            -1          4  0.50  CGG  CCATGGCTTTCTCCAGTTCT   \n",
              "361            -1         21  0.65  TGG  GGTGGTCAGCAGGGTTGCCA   \n",
              "362            -1         30  0.60  GGG  CAGATCGTCGGTGGTCAGCA   \n",
              "363            -1         31  0.65  AGG  GCAGATCGTCGGTGGTCAGC   \n",
              "444            -1         32  0.65  CGG  TGGACCACCCGGTGTTCTCC   \n",
              "27             -1         41  0.70  CGG  CCCCTCGATGCCGGATCCAC   \n",
              "28             -1         50  0.65  CGG  GATACGGGACCCCTCGATGC   \n",
              "29             -1         65  0.65  GGG  GGAGGCTCGAAGGCCGATAC   \n",
              "30             -1         66  0.60  CGG  AGGAGGCTCGAAGGCCGATA   \n",
              "31             -1         75  0.65  AGG  GTGGCTCGAAGGAGGCTCGA   \n",
              "369            -1         75  0.70  CGG  GTCCGTCCCGTCCGTCTCAC   \n",
              "\n",
              "    sgrna_seed_sequence  off_target_count  \n",
              "11        TCGCACTGGTGAC                 0  \n",
              "12        GTCGCACTGGTGA                 0  \n",
              "13        TCCGAAGTCGCAC                 0  \n",
              "434       GGACACGCCGTAC                 0  \n",
              "14        CACATGGCCACGC                 0  \n",
              "15        GGAGGGGAACACA                 0  \n",
              "351       GTTGAAGCCCCAG                 0  \n",
              "16        ACGAGGCAGGGAG                 0  \n",
              "17        CACGAGGCAGGGA                 0  \n",
              "18        CCACGAGGCAGGG                 0  \n",
              "19        GGACCACGAGGCA                 0  \n",
              "21        GTGAGGGACCACG                 0  \n",
              "22        GCTGAGCGCGTGA                 0  \n",
              "23        AGCTGAGCGCGTG                 0  \n",
              "440       CGAGCAGCTCTCG                 0  \n",
              "353       GGTCGGTGCGAAA                 0  \n",
              "354       GTCGTTACGGGGT                 0  \n",
              "355       CGGGGTCGTTACG                 0  \n",
              "356       TCGGGGTCGTTAC                 0  \n",
              "441       TGCGCCGTTGATC                 0  \n",
              "357       CTCGGGGTCGTTA                 0  \n",
              "358       TCTCCAGTTCTCG                 0  \n",
              "359       TTCTCCAGTTCTC                 0  \n",
              "360       TTTCTCCAGTTCT                 0  \n",
              "361       AGCAGGGTTGCCA                 0  \n",
              "362       TCGGTGGTCAGCA                 0  \n",
              "363       GTCGGTGGTCAGC                 0  \n",
              "444       CCCGGTGTTCTCC                 0  \n",
              "27        ATGCCGGATCCAC                 0  \n",
              "28        GACCCCTCGATGC                 0  \n",
              "29        CGAAGGCCGATAC                 0  \n",
              "30        TCGAAGGCCGATA                 0  \n",
              "31        GAAGGAGGCTCGA                 0  \n",
              "369       CCGTCCGTCTCAC                 0  "
            ]
          },
          "execution_count": 3,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "\n",
        "# Initialize SgRNAArgs with desired parameters\n",
        "args = SgRNAargs(path_to_genome, \n",
        "                genes_to_KO,\n",
        "                step=['find', 'filter'],\n",
        "                gc_upper = gc_upper,\n",
        "                gc_lower = gc_lower,\n",
        "                off_target_seed = off_target_seed,\n",
        "                off_target_upper = off_target_upper,\n",
        "                cas_type='cas9',\n",
        "                extension_to_promoter_region=200,\n",
        "                target_non_template_strand = True\n",
        "                  )\n",
        "\n",
        "sgrna_df = extract_sgRNAs_for_crispri(args)\n",
        "sgrna_df"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 4,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>strain_name</th>\n",
              "      <th>locus_tag</th>\n",
              "      <th>gene_loc</th>\n",
              "      <th>gene_strand</th>\n",
              "      <th>sgrna_strand</th>\n",
              "      <th>sgrna_loc</th>\n",
              "      <th>gc</th>\n",
              "      <th>pam</th>\n",
              "      <th>sgrna</th>\n",
              "      <th>sgrna_seed_sequence</th>\n",
              "      <th>off_target_count</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>11</th>\n",
              "      <td>Streptomyces_coelicolor_A3</td>\n",
              "      <td>SCO5087</td>\n",
              "      <td>5529801</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>-95</td>\n",
              "      <td>0.60</td>\n",
              "      <td>GGG</td>\n",
              "      <td>TCCGAAGTCGCACTGGTGAC</td>\n",
              "      <td>TCGCACTGGTGAC</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>12</th>\n",
              "      <td>Streptomyces_coelicolor_A3</td>\n",
              "      <td>SCO5087</td>\n",
              "      <td>5529801</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>-94</td>\n",
              "      <td>0.60</td>\n",
              "      <td>CGG</td>\n",
              "      <td>CTCCGAAGTCGCACTGGTGA</td>\n",
              "      <td>GTCGCACTGGTGA</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>13</th>\n",
              "      <td>Streptomyces_coelicolor_A3</td>\n",
              "      <td>SCO5087</td>\n",
              "      <td>5529801</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>-88</td>\n",
              "      <td>0.65</td>\n",
              "      <td>TGG</td>\n",
              "      <td>GCAGGACTCCGAAGTCGCAC</td>\n",
              "      <td>TCCGAAGTCGCAC</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>434</th>\n",
              "      <td>Streptomyces_coelicolor_A3</td>\n",
              "      <td>SCO5090</td>\n",
              "      <td>5532706</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>-73</td>\n",
              "      <td>0.65</td>\n",
              "      <td>CGG</td>\n",
              "      <td>CGGGTATGGACACGCCGTAC</td>\n",
              "      <td>GGACACGCCGTAC</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>14</th>\n",
              "      <td>Streptomyces_coelicolor_A3</td>\n",
              "      <td>SCO5087</td>\n",
              "      <td>5529801</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>-70</td>\n",
              "      <td>0.65</td>\n",
              "      <td>AGG</td>\n",
              "      <td>AGGGGAACACATGGCCACGC</td>\n",
              "      <td>CACATGGCCACGC</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>15</th>\n",
              "      <td>Streptomyces_coelicolor_A3</td>\n",
              "      <td>SCO5087</td>\n",
              "      <td>5529801</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>-61</td>\n",
              "      <td>0.65</td>\n",
              "      <td>TGG</td>\n",
              "      <td>GAGGCAGGGAGGGGAACACA</td>\n",
              "      <td>GGAGGGGAACACA</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>351</th>\n",
              "      <td>Streptomyces_coelicolor_A3</td>\n",
              "      <td>SCO5089</td>\n",
              "      <td>5532449</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>-56</td>\n",
              "      <td>0.70</td>\n",
              "      <td>CGG</td>\n",
              "      <td>CCGCCGAGTTGAAGCCCCAG</td>\n",
              "      <td>GTTGAAGCCCCAG</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>440</th>\n",
              "      <td>Streptomyces_coelicolor_A3</td>\n",
              "      <td>SCO5090</td>\n",
              "      <td>5532706</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>-24</td>\n",
              "      <td>0.60</td>\n",
              "      <td>CGG</td>\n",
              "      <td>ATCAGGTCGAGCAGCTCTCG</td>\n",
              "      <td>CGAGCAGCTCTCG</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>353</th>\n",
              "      <td>Streptomyces_coelicolor_A3</td>\n",
              "      <td>SCO5089</td>\n",
              "      <td>5532449</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>-23</td>\n",
              "      <td>0.60</td>\n",
              "      <td>CGG</td>\n",
              "      <td>GTTACGGGGTCGGTGCGAAA</td>\n",
              "      <td>GGTCGGTGCGAAA</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>354</th>\n",
              "      <td>Streptomyces_coelicolor_A3</td>\n",
              "      <td>SCO5089</td>\n",
              "      <td>5532449</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>-13</td>\n",
              "      <td>0.65</td>\n",
              "      <td>CGG</td>\n",
              "      <td>TCTCGGGGTCGTTACGGGGT</td>\n",
              "      <td>GTCGTTACGGGGT</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>355</th>\n",
              "      <td>Streptomyces_coelicolor_A3</td>\n",
              "      <td>SCO5089</td>\n",
              "      <td>5532449</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>-9</td>\n",
              "      <td>0.60</td>\n",
              "      <td>GGG</td>\n",
              "      <td>CAGTTCTCGGGGTCGTTACG</td>\n",
              "      <td>CGGGGTCGTTACG</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>356</th>\n",
              "      <td>Streptomyces_coelicolor_A3</td>\n",
              "      <td>SCO5089</td>\n",
              "      <td>5532449</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>-8</td>\n",
              "      <td>0.60</td>\n",
              "      <td>GGG</td>\n",
              "      <td>CCAGTTCTCGGGGTCGTTAC</td>\n",
              "      <td>TCGGGGTCGTTAC</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>441</th>\n",
              "      <td>Streptomyces_coelicolor_A3</td>\n",
              "      <td>SCO5090</td>\n",
              "      <td>5532706</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>-7</td>\n",
              "      <td>0.70</td>\n",
              "      <td>AGG</td>\n",
              "      <td>CGGCCAGTGCGCCGTTGATC</td>\n",
              "      <td>TGCGCCGTTGATC</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>444</th>\n",
              "      <td>Streptomyces_coelicolor_A3</td>\n",
              "      <td>SCO5090</td>\n",
              "      <td>5532706</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>32</td>\n",
              "      <td>0.65</td>\n",
              "      <td>CGG</td>\n",
              "      <td>TGGACCACCCGGTGTTCTCC</td>\n",
              "      <td>CCCGGTGTTCTCC</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "                    strain_name locus_tag  gene_loc  gene_strand  \\\n",
              "11   Streptomyces_coelicolor_A3   SCO5087   5529801            1   \n",
              "12   Streptomyces_coelicolor_A3   SCO5087   5529801            1   \n",
              "13   Streptomyces_coelicolor_A3   SCO5087   5529801            1   \n",
              "434  Streptomyces_coelicolor_A3   SCO5090   5532706            1   \n",
              "14   Streptomyces_coelicolor_A3   SCO5087   5529801            1   \n",
              "15   Streptomyces_coelicolor_A3   SCO5087   5529801            1   \n",
              "351  Streptomyces_coelicolor_A3   SCO5089   5532449            1   \n",
              "440  Streptomyces_coelicolor_A3   SCO5090   5532706            1   \n",
              "353  Streptomyces_coelicolor_A3   SCO5089   5532449            1   \n",
              "354  Streptomyces_coelicolor_A3   SCO5089   5532449            1   \n",
              "355  Streptomyces_coelicolor_A3   SCO5089   5532449            1   \n",
              "356  Streptomyces_coelicolor_A3   SCO5089   5532449            1   \n",
              "441  Streptomyces_coelicolor_A3   SCO5090   5532706            1   \n",
              "444  Streptomyces_coelicolor_A3   SCO5090   5532706            1   \n",
              "\n",
              "     sgrna_strand  sgrna_loc    gc  pam                 sgrna  \\\n",
              "11             -1        -95  0.60  GGG  TCCGAAGTCGCACTGGTGAC   \n",
              "12             -1        -94  0.60  CGG  CTCCGAAGTCGCACTGGTGA   \n",
              "13             -1        -88  0.65  TGG  GCAGGACTCCGAAGTCGCAC   \n",
              "434            -1        -73  0.65  CGG  CGGGTATGGACACGCCGTAC   \n",
              "14             -1        -70  0.65  AGG  AGGGGAACACATGGCCACGC   \n",
              "15             -1        -61  0.65  TGG  GAGGCAGGGAGGGGAACACA   \n",
              "351            -1        -56  0.70  CGG  CCGCCGAGTTGAAGCCCCAG   \n",
              "440            -1        -24  0.60  CGG  ATCAGGTCGAGCAGCTCTCG   \n",
              "353            -1        -23  0.60  CGG  GTTACGGGGTCGGTGCGAAA   \n",
              "354            -1        -13  0.65  CGG  TCTCGGGGTCGTTACGGGGT   \n",
              "355            -1         -9  0.60  GGG  CAGTTCTCGGGGTCGTTACG   \n",
              "356            -1         -8  0.60  GGG  CCAGTTCTCGGGGTCGTTAC   \n",
              "441            -1         -7  0.70  AGG  CGGCCAGTGCGCCGTTGATC   \n",
              "444            -1         32  0.65  CGG  TGGACCACCCGGTGTTCTCC   \n",
              "\n",
              "    sgrna_seed_sequence  off_target_count  \n",
              "11        TCGCACTGGTGAC                 0  \n",
              "12        GTCGCACTGGTGA                 0  \n",
              "13        TCCGAAGTCGCAC                 0  \n",
              "434       GGACACGCCGTAC                 0  \n",
              "14        CACATGGCCACGC                 0  \n",
              "15        GGAGGGGAACACA                 0  \n",
              "351       GTTGAAGCCCCAG                 0  \n",
              "440       CGAGCAGCTCTCG                 0  \n",
              "353       GGTCGGTGCGAAA                 0  \n",
              "354       GTCGTTACGGGGT                 0  \n",
              "355       CGGGGTCGTTACG                 0  \n",
              "356       TCGGGGTCGTTAC                 0  \n",
              "441       TGCGCCGTTGATC                 0  \n",
              "444       CCCGGTGTTCTCC                 0  "
            ]
          },
          "execution_count": 4,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "# Filter the DataFrame to retain only up to 5 sgRNA sequences per locus_tag\n",
        "filtered_df = sgrna_df.groupby('locus_tag').head(number_of_sgRNAs_per_group)\n",
        "filtered_df"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "## Output"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 5,
      "metadata": {},
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "SCO5087_loc_-95\n"
          ]
        },
        {
          "data": {
            "text/plain": [
              "[Contig(o11279),\n",
              " Contig(o11279),\n",
              " Contig(o11279),\n",
              " Contig(o11279),\n",
              " Contig(o11279),\n",
              " Contig(o11279),\n",
              " Contig(o11279),\n",
              " Contig(o11279),\n",
              " Contig(o11279),\n",
              " Contig(o11279),\n",
              " Contig(o11279),\n",
              " Contig(o11279),\n",
              " Contig(o11279),\n",
              " Contig(o11279)]"
            ]
          },
          "execution_count": 5,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "# MAke oligoes\n",
        "list_of_ssDNAs = make_ssDNA_oligos(filtered_df, upstream_ovh = up_homology,\n",
        "                      downstream_ovh=dw_homology)\n",
        "print(list_of_ssDNAs[0].name)\n",
        "\n",
        "# cut plasmid\n",
        "from Bio.Restriction import NcoI\n",
        "linearized_plasmid = sorted(clean_plasmid.cut(NcoI), key=lambda x: len(x), reverse=True)[0]\n",
        "#print(linearized_plasmid)\n",
        "\n",
        "sgRNA_vectors = assemble_plasmids_by_ssDNA_bridging(list_of_ssDNAs,linearized_plasmid)\n",
        "sgRNA_vectors"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 6,
      "metadata": {},
      "outputs": [],
      "source": [
        "\n",
        "# Constructing a meaningful name, ID, and description for the assembled plasmid using user input\n",
        "targeting_info = []\n",
        "for index, row in filtered_df.iterrows():\n",
        "    formatted_str = f\"CRISPRi_{row['locus_tag']}_p{row['sgrna_loc']}\"\n",
        "    targeting_info.append(formatted_str)\n",
        "\n",
        "for i in range(len(sgRNA_vectors)):\n",
        "    sgRNA_vectors[i].name = f'p{targeting_info[i]}_#{i+1}'\n",
        "    sgRNA_vectors[i].id = sgRNA_vectors[i].name  # Using the same value for ID as for name for simplicity\n",
        "    sgRNA_vectors[i].description = f'Assembled plasmid targeting {\", \".join(genes_to_KO)} for single gene KNOCK-DOWN, assembled using StreptoAIM.'\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 7,
      "metadata": {},
      "outputs": [],
      "source": [
        "print_plasmids = False\n",
        "\n",
        "if print_plasmids: \n",
        "    for vector in sgRNA_vectors: \n",
        "        vector.write(f\"../../data/plasmids/sgRNA_plasmids_pCRISPR–dCas9/{vector.id}.gb\")"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "### IDT primers"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 8,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>Name</th>\n",
              "      <th>Sequence</th>\n",
              "      <th>Concentration</th>\n",
              "      <th>Purification</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>SCO5087_loc_-95</td>\n",
              "      <td>CGGTTGGTAGGATCGACGGCTCCGAAGTCGCACTGGTGACGTTTTA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>SCO5087_loc_-94</td>\n",
              "      <td>CGGTTGGTAGGATCGACGGCCTCCGAAGTCGCACTGGTGAGTTTTA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>SCO5087_loc_-88</td>\n",
              "      <td>CGGTTGGTAGGATCGACGGCGCAGGACTCCGAAGTCGCACGTTTTA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>SCO5090_loc_-73</td>\n",
              "      <td>CGGTTGGTAGGATCGACGGCCGGGTATGGACACGCCGTACGTTTTA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>SCO5087_loc_-70</td>\n",
              "      <td>CGGTTGGTAGGATCGACGGCAGGGGAACACATGGCCACGCGTTTTA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>5</th>\n",
              "      <td>SCO5087_loc_-61</td>\n",
              "      <td>CGGTTGGTAGGATCGACGGCGAGGCAGGGAGGGGAACACAGTTTTA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>6</th>\n",
              "      <td>SCO5089_loc_-56</td>\n",
              "      <td>CGGTTGGTAGGATCGACGGCCCGCCGAGTTGAAGCCCCAGGTTTTA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>7</th>\n",
              "      <td>SCO5090_loc_-24</td>\n",
              "      <td>CGGTTGGTAGGATCGACGGCATCAGGTCGAGCAGCTCTCGGTTTTA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>8</th>\n",
              "      <td>SCO5089_loc_-23</td>\n",
              "      <td>CGGTTGGTAGGATCGACGGCGTTACGGGGTCGGTGCGAAAGTTTTA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>9</th>\n",
              "      <td>SCO5089_loc_-13</td>\n",
              "      <td>CGGTTGGTAGGATCGACGGCTCTCGGGGTCGTTACGGGGTGTTTTA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>10</th>\n",
              "      <td>SCO5089_loc_-9</td>\n",
              "      <td>CGGTTGGTAGGATCGACGGCCAGTTCTCGGGGTCGTTACGGTTTTA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>11</th>\n",
              "      <td>SCO5089_loc_-8</td>\n",
              "      <td>CGGTTGGTAGGATCGACGGCCCAGTTCTCGGGGTCGTTACGTTTTA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>12</th>\n",
              "      <td>SCO5090_loc_-7</td>\n",
              "      <td>CGGTTGGTAGGATCGACGGCCGGCCAGTGCGCCGTTGATCGTTTTA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>13</th>\n",
              "      <td>SCO5090_loc_32</td>\n",
              "      <td>CGGTTGGTAGGATCGACGGCTGGACCACCCGGTGTTCTCCGTTTTA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "               Name                                           Sequence  \\\n",
              "0   SCO5087_loc_-95  CGGTTGGTAGGATCGACGGCTCCGAAGTCGCACTGGTGACGTTTTA...   \n",
              "1   SCO5087_loc_-94  CGGTTGGTAGGATCGACGGCCTCCGAAGTCGCACTGGTGAGTTTTA...   \n",
              "2   SCO5087_loc_-88  CGGTTGGTAGGATCGACGGCGCAGGACTCCGAAGTCGCACGTTTTA...   \n",
              "3   SCO5090_loc_-73  CGGTTGGTAGGATCGACGGCCGGGTATGGACACGCCGTACGTTTTA...   \n",
              "4   SCO5087_loc_-70  CGGTTGGTAGGATCGACGGCAGGGGAACACATGGCCACGCGTTTTA...   \n",
              "5   SCO5087_loc_-61  CGGTTGGTAGGATCGACGGCGAGGCAGGGAGGGGAACACAGTTTTA...   \n",
              "6   SCO5089_loc_-56  CGGTTGGTAGGATCGACGGCCCGCCGAGTTGAAGCCCCAGGTTTTA...   \n",
              "7   SCO5090_loc_-24  CGGTTGGTAGGATCGACGGCATCAGGTCGAGCAGCTCTCGGTTTTA...   \n",
              "8   SCO5089_loc_-23  CGGTTGGTAGGATCGACGGCGTTACGGGGTCGGTGCGAAAGTTTTA...   \n",
              "9   SCO5089_loc_-13  CGGTTGGTAGGATCGACGGCTCTCGGGGTCGTTACGGGGTGTTTTA...   \n",
              "10   SCO5089_loc_-9  CGGTTGGTAGGATCGACGGCCAGTTCTCGGGGTCGTTACGGTTTTA...   \n",
              "11   SCO5089_loc_-8  CGGTTGGTAGGATCGACGGCCCAGTTCTCGGGGTCGTTACGTTTTA...   \n",
              "12   SCO5090_loc_-7  CGGTTGGTAGGATCGACGGCCGGCCAGTGCGCCGTTGATCGTTTTA...   \n",
              "13   SCO5090_loc_32  CGGTTGGTAGGATCGACGGCTGGACCACCCGGTGTTCTCCGTTTTA...   \n",
              "\n",
              "   Concentration Purification  \n",
              "0           25nm          STD  \n",
              "1           25nm          STD  \n",
              "2           25nm          STD  \n",
              "3           25nm          STD  \n",
              "4           25nm          STD  \n",
              "5           25nm          STD  \n",
              "6           25nm          STD  \n",
              "7           25nm          STD  \n",
              "8           25nm          STD  \n",
              "9           25nm          STD  \n",
              "10          25nm          STD  \n",
              "11          25nm          STD  \n",
              "12          25nm          STD  \n",
              "13          25nm          STD  "
            ]
          },
          "execution_count": 8,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "idt_primers=primers_to_IDT(list_of_ssDNAs)\n",
        "idt_primers"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 9,
      "metadata": {},
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "Project structure for 'CRISPRi_workflow_2024-06-28T09:32:25.460595' created successfully.\n"
          ]
        },
        {
          "name": "stderr",
          "output_type": "stream",
          "text": [
            "/Users/lucaslevassor/opt/anaconda3/envs/streptocad/lib/python3.11/site-packages/Bio/SeqIO/InsdcIO.py:728: BiopythonWarning: Increasing length of locus line to allow long name. This will result in fields that are not in usual positions.\n",
            "  warnings.warn(\n"
          ]
        }
      ],
      "source": [
        "\n",
        "\n",
        "input_files = [\n",
        "    {\"name\": \"input_genome.gb\", \"content\": genome},\n",
        "    {\"name\": \"input_plasmid.gb\", \"content\": clean_plasmid}\n",
        "]\n",
        "\n",
        "output_files = [\n",
        "    {\"name\": \"cBEST_w_sgRNAs.gb\", \"content\": sgRNA_vectors}, # LIST OF Dseqrecords\n",
        "    {\"name\": \"full_idt.csv\", \"content\": idt_primers},\n",
        "    {\"name\": \"sgrna_df.csv\", \"content\": sgrna_df},\n",
        "    {\"name\": \"filtered_df.csv\", \"content\": filtered_df}\n",
        "]\n",
        "\n",
        "input_values = {\n",
        "    \"genes_to_knockout\": genes_to_KO,\n",
        "\n",
        "\n",
        "    \"filtering_metrics\": {\n",
        "        \"gc_upper\": gc_upper,\n",
        "        \"gc_lower\": gc_lower,\n",
        "        \"off_target_seed\": off_target_seed,\n",
        "        \"off_target_upper\": off_target_upper,\n",
        "        \"cas_type\": cas_type,\n",
        "        \"number_of_sgRNAs_per_group\": number_of_sgRNAs_per_group,\n",
        "        'extension_to_promoter_region':extension_to_promoter_region,\n",
        "\n",
        "    },\n",
        "    \"overlapping_sequences\": {\n",
        "        \"up_homology\": str(up_homology),\n",
        "        \"dw_homology\": str(dw_homology)\n",
        "    }\n",
        "}\n",
        "\n",
        "\n",
        "# Paths to Markdown files\n",
        "markdown_file_paths = [\n",
        "    \"../../protocols/conjugation_protcol.md\",\n",
        "    \"../../protocols/single_target_crispr_plasmid_protcol.md\"\n",
        "\n",
        "]\n",
        "\n",
        "# Data and time\n",
        "timestamp = datetime.utcnow().isoformat()\n",
        "\n",
        "generate_project_directory_structure(\n",
        "    project_name=f\"CRISPRi_workflow_{timestamp}\",\n",
        "    input_files=input_files,\n",
        "    output_files=output_files,\n",
        "    input_values=input_values,\n",
        "    markdown_file_paths=markdown_file_paths\n",
        "\n",
        ")"
      ]
    }
  ],
  "metadata": {
    "kernelspec": {
      "display_name": "constrain",
      "language": "python",
      "name": "python3"
    },
    "language_info": {
      "codemirror_mode": {
        "name": "ipython",
        "version": 3
      },
      "file_extension": ".py",
      "mimetype": "text/x-python",
      "name": "python",
      "nbconvert_exporter": "python",
      "pygments_lexer": "ipython3",
      "version": "3.11.9"
    }
  },
  "nbformat": 4,
  "nbformat_minor": 2
}
