{
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "# W5-CRISPR-Cas9-inframe_deletion"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "In Workflow 5, CRISPR-Cas9 combines ssDNA bridging and Gibson cloning to create in-frame deletions (Figure 6A). This ensures that the downstream genes are not disrupted, which is crucial for functional studies of gene knockouts. This workflow can be used to study gene function by knocking out specific genes in streptomycetes and observing the resulting phenotypic changes. It can also be used to create markerless deletions, which are important for constructing clean genetic backgrounds for further genetic manipulations. To get started, download the pCRISPR-Cas9 file and a genome file (e.g., S. coelicolor A3), then select the genes to delete. StreptoAIM will generate the necessary primers and plasmid constructs (S5).\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 28,
      "metadata": {},
      "outputs": [],
      "source": [
        "import sys\n",
        "import os\n",
        "from Bio.Restriction import StuI\n",
        "from pydna.dseqrecord import Dseqrecord\n",
        "import pandas as pd\n",
        "from datetime import datetime\n",
        "\n",
        "\n",
        "# Ensure the src directory is in the Python path\n",
        "project_root = os.path.abspath(os.path.join(os.getcwd(), '../../'))\n",
        "if project_root not in sys.path:\n",
        "    sys.path.insert(0, project_root)\n",
        "\n",
        "from streptocad.sequence_loading.sequence_loading import (\n",
        "    load_and_process_gene_sequences,\n",
        "    load_and_process_genome_sequences,\n",
        "    load_and_process_plasmid, \n",
        "    check_and_convert_input,\n",
        "    annotate_dseqrecord,\n",
        "    process_specified_gene_sequences_from_record)\n",
        "\n",
        "from streptocad.utils import polymerase_dict, create_primer_df_from_dict,ProjectDirectory,extract_metadata_to_dataframe\n",
        "from streptocad.primers.primer_generation import create_idt_order_dataframe\n",
        "from streptocad.cloning.ssDNA_bridging import assemble_plasmids_by_ssDNA_bridging,make_ssDNA_oligos\n",
        "from streptocad.crispr.guideRNAcas3_9_12 import extract_sgRNAs, SgRNAargs\n",
        "from streptocad.cloning.gibson_cloning import (\n",
        "    find_up_dw_repair_templates,\n",
        "    assemble_multiple_plasmids_with_repair_templates_for_deletion,\n",
        "    update_primer_names\n",
        ")\n",
        "\n",
        "from streptocad.cloning.plasmid_processing import check_plasmid_restriction_sites, determine_workflow_order_for_plasmids\n",
        "from streptocad.primers.primer_generation import checking_primers, primers_to_IDT, find_best_check_primers_from_genome"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "## INPUT"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 29,
      "metadata": {},
      "outputs": [],
      "source": [
        "# Inputs\n",
        "# 1 Add genome of choice (genbank, fasta)\n",
        "path_to_genome = '../../data/genomes/Streptomyces_coelicolor_A3_chromosome.gb'\n",
        "genome = load_and_process_genome_sequences(path_to_genome)[0]\n",
        "\n",
        "# 2 Add plasmid \n",
        "path_to_plasmid = '../../data/plasmids/pCRISPR–Cas9_plasmid_addgene.gbk'\n",
        "clean_plasmid = load_and_process_plasmid(path_to_plasmid)\n",
        "\n",
        "# 3 Choose genes to knock out (list)\n",
        "genes_to_KO = ['SCO5087', 'SCO5089', 'SCO5090','SCO5091', 'SCO5092' ]\n",
        "#genes_to_KO = ['SCO5087']#, 'SCO5089', 'SCO5090','SCO5091', 'SCO5092' ]\n",
        "\n",
        "#genes_to_KO = ['80000-100000', '4000-7000', '9000-14000','15000-20000']\n",
        "\n",
        "\n",
        "#### Advanced settings ####\n",
        "# 4 Filtering metrics for sgRNAs\n",
        "gc_upper = 0.72\n",
        "gc_lower = 0.2\n",
        "off_target_seed = 13\n",
        "off_target_upper = 10\n",
        "cas_type='cas9'\n",
        "number_of_sgRNAs_per_group = 2\n",
        "\n",
        "# 5 Choose polymerase and target melting temperature\n",
        "chosen_polymerase = polymerase_dict['Phusion High-Fidelity DNA Polymerase (GC Buffer)']\n",
        "melting_temperature = 65\n",
        "primer_concentration = 0.4 \n",
        "primer_number_increment = 23\n",
        "flanking_region = 500\n",
        "\n",
        "# 6 Choose overlapping sequences for our plasmid we can use the following\n",
        "#As per the article **\"CRISPR–Cas9, CRISPRi and CRISPR-BEST-mediated genetic manipulation in streptomycetes\"** we need the following oligoes: \n",
        "#CGGTTGGTAGGATCGACGGC **-N20-** GTTTTAGAGCTAGAAATAGC\n",
        "up_homology = Dseqrecord('CGGTTGGTAGGATCGACGGC')\n",
        "dw_homology = Dseqrecord('GTTTTAGAGCTAGAAATAGC')\n",
        "\n",
        "# 7 extra settings\n",
        "#### Advanced settings ####\n",
        "repair_length = 1000\n",
        "overlap_length = 40 "
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 30,
      "metadata": {},
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "pCRISPR-Cas9\n"
          ]
        }
      ],
      "source": [
        "print(clean_plasmid.id)"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "# Computation"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 31,
      "metadata": {},
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "False\n"
          ]
        },
        {
          "data": {
            "text/plain": [
              "25824"
            ]
          },
          "execution_count": 31,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "target_dict, genes_to_KO, annotation_input = check_and_convert_input(genes_to_KO)\n",
        "\n",
        "print(annotation_input)\n",
        "if annotation_input == True:\n",
        "    genome = annotate_dseqrecord(genome, target_dict)\n",
        "\n",
        "\n",
        "len(genome.features)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 32,
      "metadata": {},
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "Pam was found outside designated locus_tag: SCO5087. To incorporate this extent borders. Skipping to next locus tag.\n",
            "sgRNA generated were outside the designated border in SCO5087. To incorporate this extent borders. Skipping to next locus tag.\n",
            "Pam was found outside designated locus_tag: SCO5087. To incorporate this extent borders. Skipping to next locus tag.\n",
            "sgRNA generated were outside the designated border in SCO5087. To incorporate this extent borders. Skipping to next locus tag.\n",
            "sgRNA generated were outside the designated border in SCO5087. To incorporate this extent borders. Skipping to next locus tag.\n",
            "sgRNA generated were outside the designated border in SCO5089. To incorporate this extent borders. Skipping to next locus tag.\n",
            "sgRNA generated were outside the designated border in SCO5089. To incorporate this extent borders. Skipping to next locus tag.\n",
            "Pam was found outside designated locus_tag: SCO5090. To incorporate this extent borders. Skipping to next locus tag.\n",
            "sgRNA generated were outside the designated border in SCO5090. To incorporate this extent borders. Skipping to next locus tag.\n",
            "sgRNA generated were outside the designated border in SCO5090. To incorporate this extent borders. Skipping to next locus tag.\n",
            "sgRNA generated were outside the designated border in SCO5090. To incorporate this extent borders. Skipping to next locus tag.\n",
            "sgRNA generated were outside the designated border in SCO5090. To incorporate this extent borders. Skipping to next locus tag.\n",
            "sgRNA generated were outside the designated border in SCO5090. To incorporate this extent borders. Skipping to next locus tag.\n",
            "sgRNA generated were outside the designated border in SCO5091. To incorporate this extent borders. Skipping to next locus tag.\n",
            "sgRNA generated were outside the designated border in SCO5091. To incorporate this extent borders. Skipping to next locus tag.\n",
            "sgRNA generated were outside the designated border in SCO5091. To incorporate this extent borders. Skipping to next locus tag.\n",
            "sgRNA generated were outside the designated border in SCO5091. To incorporate this extent borders. Skipping to next locus tag.\n",
            "sgRNA generated were outside the designated border in SCO5091. To incorporate this extent borders. Skipping to next locus tag.\n",
            "Pam was found outside designated locus_tag: SCO5091. To incorporate this extent borders. Skipping to next locus tag.\n",
            "sgRNA generated were outside the designated border in SCO5091. To incorporate this extent borders. Skipping to next locus tag.\n",
            "sgRNA generated were outside the designated border in SCO5091. To incorporate this extent borders. Skipping to next locus tag.\n",
            "sgRNA generated were outside the designated border in SCO5092. To incorporate this extent borders. Skipping to next locus tag.\n",
            "sgRNA generated were outside the designated border in SCO5092. To incorporate this extent borders. Skipping to next locus tag.\n",
            "sgRNA generated were outside the designated border in SCO5092. To incorporate this extent borders. Skipping to next locus tag.\n",
            "sgRNA generated were outside the designated border in SCO5092. To incorporate this extent borders. Skipping to next locus tag.\n",
            "sgRNA generated were outside the designated border in SCO5092. To incorporate this extent borders. Skipping to next locus tag.\n",
            "sgRNA generated were outside the designated border in SCO5092. To incorporate this extent borders. Skipping to next locus tag.\n"
          ]
        },
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>strain_name</th>\n",
              "      <th>locus_tag</th>\n",
              "      <th>gene_loc</th>\n",
              "      <th>gene_strand</th>\n",
              "      <th>sgrna_strand</th>\n",
              "      <th>sgrna_loc</th>\n",
              "      <th>gc</th>\n",
              "      <th>pam</th>\n",
              "      <th>sgrna</th>\n",
              "      <th>sgrna_seed_sequence</th>\n",
              "      <th>off_target_count</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>483</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5090</td>\n",
              "      <td>5532706</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>546</td>\n",
              "      <td>0.65</td>\n",
              "      <td>CGG</td>\n",
              "      <td>GCAGCGACCTGTGGCAGGAA</td>\n",
              "      <td>CCTGTGGCAGGAA</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>485</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5090</td>\n",
              "      <td>5532706</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>537</td>\n",
              "      <td>0.65</td>\n",
              "      <td>TGG</td>\n",
              "      <td>TCATCGAGCGCAGCGACCTG</td>\n",
              "      <td>GCGCAGCGACCTG</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>489</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5090</td>\n",
              "      <td>5532706</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>487</td>\n",
              "      <td>0.65</td>\n",
              "      <td>GGG</td>\n",
              "      <td>CACGTTCGAGGACACGCTCC</td>\n",
              "      <td>GAGGACACGCTCC</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>490</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5090</td>\n",
              "      <td>5532706</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>486</td>\n",
              "      <td>0.60</td>\n",
              "      <td>CGG</td>\n",
              "      <td>TCACGTTCGAGGACACGCTC</td>\n",
              "      <td>CGAGGACACGCTC</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>491</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5090</td>\n",
              "      <td>5532706</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>475</td>\n",
              "      <td>0.60</td>\n",
              "      <td>AGG</td>\n",
              "      <td>GGAGCTGGTCTTCACGTTCG</td>\n",
              "      <td>GTCTTCACGTTCG</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>...</th>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>449</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5090</td>\n",
              "      <td>5532706</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>890</td>\n",
              "      <td>0.65</td>\n",
              "      <td>CGG</td>\n",
              "      <td>ACCCATCTGCGCGAGGTGCT</td>\n",
              "      <td>TGCGCGAGGTGCT</td>\n",
              "      <td>3</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>200</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5087</td>\n",
              "      <td>5529801</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>706</td>\n",
              "      <td>0.65</td>\n",
              "      <td>CGG</td>\n",
              "      <td>CAGCGCCGACGTGATGTTCG</td>\n",
              "      <td>GACGTGATGTTCG</td>\n",
              "      <td>3</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>650</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5091</td>\n",
              "      <td>5533653</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>649</td>\n",
              "      <td>0.65</td>\n",
              "      <td>GGG</td>\n",
              "      <td>CGAGGTGATCGACGCCAACC</td>\n",
              "      <td>ATCGACGCCAACC</td>\n",
              "      <td>3</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>191</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5087</td>\n",
              "      <td>5529801</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>838</td>\n",
              "      <td>0.70</td>\n",
              "      <td>TGG</td>\n",
              "      <td>TACGCGGGACGGCTTCGTGC</td>\n",
              "      <td>GACGGCTTCGTGC</td>\n",
              "      <td>3</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>753</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5092</td>\n",
              "      <td>5534558</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>178</td>\n",
              "      <td>0.65</td>\n",
              "      <td>AGG</td>\n",
              "      <td>AGGAGTTGGCCGTACGAGCC</td>\n",
              "      <td>GGCCGTACGAGCC</td>\n",
              "      <td>3</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>423 rows × 11 columns</p>\n",
              "</div>"
            ],
            "text/plain": [
              "     strain_name locus_tag  gene_loc  gene_strand  sgrna_strand  sgrna_loc  \\\n",
              "483  NC_003888.3   SCO5090   5532706            1             1        546   \n",
              "485  NC_003888.3   SCO5090   5532706            1             1        537   \n",
              "489  NC_003888.3   SCO5090   5532706            1             1        487   \n",
              "490  NC_003888.3   SCO5090   5532706            1             1        486   \n",
              "491  NC_003888.3   SCO5090   5532706            1             1        475   \n",
              "..           ...       ...       ...          ...           ...        ...   \n",
              "449  NC_003888.3   SCO5090   5532706            1             1        890   \n",
              "200  NC_003888.3   SCO5087   5529801            1             1        706   \n",
              "650  NC_003888.3   SCO5091   5533653            1             1        649   \n",
              "191  NC_003888.3   SCO5087   5529801            1             1        838   \n",
              "753  NC_003888.3   SCO5092   5534558            1            -1        178   \n",
              "\n",
              "       gc  pam                 sgrna sgrna_seed_sequence  off_target_count  \n",
              "483  0.65  CGG  GCAGCGACCTGTGGCAGGAA       CCTGTGGCAGGAA                 0  \n",
              "485  0.65  TGG  TCATCGAGCGCAGCGACCTG       GCGCAGCGACCTG                 0  \n",
              "489  0.65  GGG  CACGTTCGAGGACACGCTCC       GAGGACACGCTCC                 0  \n",
              "490  0.60  CGG  TCACGTTCGAGGACACGCTC       CGAGGACACGCTC                 0  \n",
              "491  0.60  AGG  GGAGCTGGTCTTCACGTTCG       GTCTTCACGTTCG                 0  \n",
              "..    ...  ...                   ...                 ...               ...  \n",
              "449  0.65  CGG  ACCCATCTGCGCGAGGTGCT       TGCGCGAGGTGCT                 3  \n",
              "200  0.65  CGG  CAGCGCCGACGTGATGTTCG       GACGTGATGTTCG                 3  \n",
              "650  0.65  GGG  CGAGGTGATCGACGCCAACC       ATCGACGCCAACC                 3  \n",
              "191  0.70  TGG  TACGCGGGACGGCTTCGTGC       GACGGCTTCGTGC                 3  \n",
              "753  0.65  AGG  AGGAGTTGGCCGTACGAGCC       GGCCGTACGAGCC                 3  \n",
              "\n",
              "[423 rows x 11 columns]"
            ]
          },
          "execution_count": 32,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "# Initialize SgRNAargs with desired parameters\n",
        "args = SgRNAargs(genome, \n",
        "                genes_to_KO,\n",
        "                step=['find', 'filter'],\n",
        "                gc_upper = gc_upper,\n",
        "                gc_lower = gc_lower,\n",
        "                off_target_seed = off_target_seed,\n",
        "                off_target_upper = off_target_upper,\n",
        "                cas_type='cas9'\n",
        "                )\n",
        "\n",
        "sgrna_df = extract_sgRNAs(args)\n",
        "sgrna_df"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 33,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>strain_name</th>\n",
              "      <th>locus_tag</th>\n",
              "      <th>gene_loc</th>\n",
              "      <th>gene_strand</th>\n",
              "      <th>sgrna_strand</th>\n",
              "      <th>sgrna_loc</th>\n",
              "      <th>gc</th>\n",
              "      <th>pam</th>\n",
              "      <th>sgrna</th>\n",
              "      <th>sgrna_seed_sequence</th>\n",
              "      <th>off_target_count</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>483</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5090</td>\n",
              "      <td>5532706</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>546</td>\n",
              "      <td>0.65</td>\n",
              "      <td>CGG</td>\n",
              "      <td>GCAGCGACCTGTGGCAGGAA</td>\n",
              "      <td>CCTGTGGCAGGAA</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>485</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5090</td>\n",
              "      <td>5532706</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>537</td>\n",
              "      <td>0.65</td>\n",
              "      <td>TGG</td>\n",
              "      <td>TCATCGAGCGCAGCGACCTG</td>\n",
              "      <td>GCGCAGCGACCTG</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>549</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5091</td>\n",
              "      <td>5533653</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>48</td>\n",
              "      <td>0.60</td>\n",
              "      <td>CGG</td>\n",
              "      <td>CTCGTAGGCGTAGACACCCT</td>\n",
              "      <td>GCGTAGACACCCT</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>554</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5091</td>\n",
              "      <td>5533653</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>78</td>\n",
              "      <td>0.70</td>\n",
              "      <td>CGG</td>\n",
              "      <td>GTTGCTGACGCACCAGCCGC</td>\n",
              "      <td>ACGCACCAGCCGC</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>843</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5092</td>\n",
              "      <td>5534558</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>25</td>\n",
              "      <td>0.65</td>\n",
              "      <td>GGG</td>\n",
              "      <td>AGCCGACCAGGGAATGCTCC</td>\n",
              "      <td>CAGGGAATGCTCC</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>760</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5092</td>\n",
              "      <td>5534558</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>267</td>\n",
              "      <td>0.70</td>\n",
              "      <td>TGG</td>\n",
              "      <td>GGACTTGCGCGCGAAGCGCA</td>\n",
              "      <td>CGCGCGAAGCGCA</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>330</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5089</td>\n",
              "      <td>5532449</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>61</td>\n",
              "      <td>0.70</td>\n",
              "      <td>CGG</td>\n",
              "      <td>CGTGGAGTGCGCCGGTGAGA</td>\n",
              "      <td>TGCGCCGGTGAGA</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>222</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5087</td>\n",
              "      <td>5529801</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>556</td>\n",
              "      <td>0.60</td>\n",
              "      <td>TGG</td>\n",
              "      <td>CCGGCACATGTTCGACTACC</td>\n",
              "      <td>ATGTTCGACTACC</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>207</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5087</td>\n",
              "      <td>5529801</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>656</td>\n",
              "      <td>0.70</td>\n",
              "      <td>CGG</td>\n",
              "      <td>ACCTCGGGCCTGGACTCCGT</td>\n",
              "      <td>GCCTGGACTCCGT</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>329</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5089</td>\n",
              "      <td>5532449</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>65</td>\n",
              "      <td>0.70</td>\n",
              "      <td>CGG</td>\n",
              "      <td>GAGTGCGCCGGTGAGACGGA</td>\n",
              "      <td>CCGGTGAGACGGA</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "     strain_name locus_tag  gene_loc  gene_strand  sgrna_strand  sgrna_loc  \\\n",
              "483  NC_003888.3   SCO5090   5532706            1             1        546   \n",
              "485  NC_003888.3   SCO5090   5532706            1             1        537   \n",
              "549  NC_003888.3   SCO5091   5533653            1            -1         48   \n",
              "554  NC_003888.3   SCO5091   5533653            1            -1         78   \n",
              "843  NC_003888.3   SCO5092   5534558            1             1         25   \n",
              "760  NC_003888.3   SCO5092   5534558            1            -1        267   \n",
              "330  NC_003888.3   SCO5089   5532449            1             1         61   \n",
              "222  NC_003888.3   SCO5087   5529801            1             1        556   \n",
              "207  NC_003888.3   SCO5087   5529801            1             1        656   \n",
              "329  NC_003888.3   SCO5089   5532449            1             1         65   \n",
              "\n",
              "       gc  pam                 sgrna sgrna_seed_sequence  off_target_count  \n",
              "483  0.65  CGG  GCAGCGACCTGTGGCAGGAA       CCTGTGGCAGGAA                 0  \n",
              "485  0.65  TGG  TCATCGAGCGCAGCGACCTG       GCGCAGCGACCTG                 0  \n",
              "549  0.60  CGG  CTCGTAGGCGTAGACACCCT       GCGTAGACACCCT                 0  \n",
              "554  0.70  CGG  GTTGCTGACGCACCAGCCGC       ACGCACCAGCCGC                 0  \n",
              "843  0.65  GGG  AGCCGACCAGGGAATGCTCC       CAGGGAATGCTCC                 0  \n",
              "760  0.70  TGG  GGACTTGCGCGCGAAGCGCA       CGCGCGAAGCGCA                 0  \n",
              "330  0.70  CGG  CGTGGAGTGCGCCGGTGAGA       TGCGCCGGTGAGA                 0  \n",
              "222  0.60  TGG  CCGGCACATGTTCGACTACC       ATGTTCGACTACC                 0  \n",
              "207  0.70  CGG  ACCTCGGGCCTGGACTCCGT       GCCTGGACTCCGT                 0  \n",
              "329  0.70  CGG  GAGTGCGCCGGTGAGACGGA       CCGGTGAGACGGA                 0  "
            ]
          },
          "execution_count": 33,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "# Filter the DataFrame to retain only up to 5 sgRNA sequences per locus_tag\n",
        "filtered_df = sgrna_df.groupby('locus_tag').head(number_of_sgRNAs_per_group)\n",
        "filtered_df"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "## Output"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 34,
      "metadata": {},
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "SCO5090_loc_546\n"
          ]
        },
        {
          "data": {
            "text/plain": [
              "[Contig(o11279),\n",
              " Contig(o11279),\n",
              " Contig(o11279),\n",
              " Contig(o11279),\n",
              " Contig(o11279),\n",
              " Contig(o11279),\n",
              " Contig(o11279),\n",
              " Contig(o11279),\n",
              " Contig(o11279),\n",
              " Contig(o11279)]"
            ]
          },
          "execution_count": 34,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "# MAke oligoes\n",
        "list_of_ssDNAs = make_ssDNA_oligos(filtered_df, upstream_ovh = up_homology,\n",
        "                      downstream_ovh=dw_homology)\n",
        "print(list_of_ssDNAs[0].name)\n",
        "\n",
        "# cut plasmid\n",
        "from Bio.Restriction import NcoI\n",
        "linearized_plasmid = sorted(clean_plasmid.cut(NcoI), key=lambda x: len(x), reverse=True)[0]\n",
        "#print(linearized_plasmid)\n",
        "\n",
        "sgRNA_vectors = assemble_plasmids_by_ssDNA_bridging(list_of_ssDNAs,linearized_plasmid)\n",
        "sgRNA_vectors"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 35,
      "metadata": {},
      "outputs": [],
      "source": [
        "# Constructing a meaningful name, ID, and description for the assembled plasmid using user input\n",
        "targeting_info = []\n",
        "for index, row in filtered_df.iterrows():\n",
        "    formatted_str = f\"pCas9_{row['locus_tag']}({row['sgrna_loc']})\"\n",
        "    targeting_info.append(formatted_str)\n",
        "\n",
        "for i in range(len(sgRNA_vectors)):\n",
        "    sgRNA_vectors[i].name = f'{targeting_info[i]}'\n",
        "    sgRNA_vectors[i].id = sgRNA_vectors[i].name  \n",
        "    sgRNA_vectors[i].description = f'CRISPR-Cas9 targeting {\", \".join(genes_to_KO)} for single gene knockout, assembled using StreptoCAD.'"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 36,
      "metadata": {},
      "outputs": [],
      "source": [
        "print_plasmids = False\n",
        "\n",
        "if print_plasmids: \n",
        "    for vector in sgRNA_vectors: \n",
        "        vector.write(f\"../../data/plasmids/sgRNA_plasmids_pCRISPR-Cas9/{vector.id}.gb\")"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "### IDT primers"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 37,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>Name</th>\n",
              "      <th>Sequence</th>\n",
              "      <th>Concentration</th>\n",
              "      <th>Purification</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>SCO5090_loc_546</td>\n",
              "      <td>CGGTTGGTAGGATCGACGGCGCAGCGACCTGTGGCAGGAAGTTTTA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>SCO5090_loc_537</td>\n",
              "      <td>CGGTTGGTAGGATCGACGGCTCATCGAGCGCAGCGACCTGGTTTTA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>SCO5091_loc_48</td>\n",
              "      <td>CGGTTGGTAGGATCGACGGCCTCGTAGGCGTAGACACCCTGTTTTA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>SCO5091_loc_78</td>\n",
              "      <td>CGGTTGGTAGGATCGACGGCGTTGCTGACGCACCAGCCGCGTTTTA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>SCO5092_loc_25</td>\n",
              "      <td>CGGTTGGTAGGATCGACGGCAGCCGACCAGGGAATGCTCCGTTTTA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>5</th>\n",
              "      <td>SCO5092_loc_267</td>\n",
              "      <td>CGGTTGGTAGGATCGACGGCGGACTTGCGCGCGAAGCGCAGTTTTA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>6</th>\n",
              "      <td>SCO5089_loc_61</td>\n",
              "      <td>CGGTTGGTAGGATCGACGGCCGTGGAGTGCGCCGGTGAGAGTTTTA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>7</th>\n",
              "      <td>SCO5087_loc_556</td>\n",
              "      <td>CGGTTGGTAGGATCGACGGCCCGGCACATGTTCGACTACCGTTTTA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>8</th>\n",
              "      <td>SCO5087_loc_656</td>\n",
              "      <td>CGGTTGGTAGGATCGACGGCACCTCGGGCCTGGACTCCGTGTTTTA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>9</th>\n",
              "      <td>SCO5089_loc_65</td>\n",
              "      <td>CGGTTGGTAGGATCGACGGCGAGTGCGCCGGTGAGACGGAGTTTTA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "              Name                                           Sequence  \\\n",
              "0  SCO5090_loc_546  CGGTTGGTAGGATCGACGGCGCAGCGACCTGTGGCAGGAAGTTTTA...   \n",
              "1  SCO5090_loc_537  CGGTTGGTAGGATCGACGGCTCATCGAGCGCAGCGACCTGGTTTTA...   \n",
              "2   SCO5091_loc_48  CGGTTGGTAGGATCGACGGCCTCGTAGGCGTAGACACCCTGTTTTA...   \n",
              "3   SCO5091_loc_78  CGGTTGGTAGGATCGACGGCGTTGCTGACGCACCAGCCGCGTTTTA...   \n",
              "4   SCO5092_loc_25  CGGTTGGTAGGATCGACGGCAGCCGACCAGGGAATGCTCCGTTTTA...   \n",
              "5  SCO5092_loc_267  CGGTTGGTAGGATCGACGGCGGACTTGCGCGCGAAGCGCAGTTTTA...   \n",
              "6   SCO5089_loc_61  CGGTTGGTAGGATCGACGGCCGTGGAGTGCGCCGGTGAGAGTTTTA...   \n",
              "7  SCO5087_loc_556  CGGTTGGTAGGATCGACGGCCCGGCACATGTTCGACTACCGTTTTA...   \n",
              "8  SCO5087_loc_656  CGGTTGGTAGGATCGACGGCACCTCGGGCCTGGACTCCGTGTTTTA...   \n",
              "9   SCO5089_loc_65  CGGTTGGTAGGATCGACGGCGAGTGCGCCGGTGAGACGGAGTTTTA...   \n",
              "\n",
              "  Concentration Purification  \n",
              "0          25nm          STD  \n",
              "1          25nm          STD  \n",
              "2          25nm          STD  \n",
              "3          25nm          STD  \n",
              "4          25nm          STD  \n",
              "5          25nm          STD  \n",
              "6          25nm          STD  \n",
              "7          25nm          STD  \n",
              "8          25nm          STD  \n",
              "9          25nm          STD  "
            ]
          },
          "execution_count": 37,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "idt_primers=primers_to_IDT(list_of_ssDNAs)\n",
        "idt_primers"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "# In frame deletion"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 38,
      "metadata": {},
      "outputs": [],
      "source": [
        "in_frame_deletion = True\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 39,
      "metadata": {},
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "pCas9_SCO5090(546)\n",
            "pCas9_SCO5090(537)\n",
            "pCas9_SCO5091(48)\n",
            "pCas9_SCO5091(78)\n",
            "pCas9_SCO5092(25)\n",
            "pCas9_SCO5092(267)\n",
            "pCas9_SCO5089(61)\n",
            "pCas9_SCO5087(556)\n",
            "pCas9_SCO5087(656)\n",
            "pCas9_SCO5089(65)\n",
            "\n",
            "Processing gene: SCO5087\n",
            "Checking plasmid: pCas9_SCO5090(546)\n",
            "Checking plasmid: pCas9_SCO5090(537)\n",
            "Checking plasmid: pCas9_SCO5091(48)\n",
            "Checking plasmid: pCas9_SCO5091(78)\n",
            "Checking plasmid: pCas9_SCO5092(25)\n",
            "Checking plasmid: pCas9_SCO5092(267)\n",
            "Checking plasmid: pCas9_SCO5089(61)\n",
            "Checking plasmid: pCas9_SCO5087(556)\n",
            "Match found in plasmid: pCas9_SCO5087(556)\n",
            "Repair template match found for gene: SCO5087\n",
            "Record added for gene: SCO5087\n",
            "Checking plasmid: pCas9_SCO5087(656)\n",
            "Match found in plasmid: pCas9_SCO5087(656)\n",
            "Repair template match found for gene: SCO5087\n",
            "Record added for gene: SCO5087\n",
            "Checking plasmid: pCas9_SCO5089(65)\n",
            "\n",
            "Processing gene: SCO5089\n",
            "Checking plasmid: pCas9_SCO5090(546)\n",
            "Checking plasmid: pCas9_SCO5090(537)\n",
            "Checking plasmid: pCas9_SCO5091(48)\n",
            "Checking plasmid: pCas9_SCO5091(78)\n",
            "Checking plasmid: pCas9_SCO5092(25)\n",
            "Checking plasmid: pCas9_SCO5092(267)\n",
            "Checking plasmid: pCas9_SCO5089(61)\n",
            "Match found in plasmid: pCas9_SCO5089(61)\n",
            "Repair template match found for gene: SCO5089\n",
            "Record added for gene: SCO5089\n",
            "Checking plasmid: pCas9_SCO5087(556)\n",
            "Checking plasmid: pCas9_SCO5087(656)\n",
            "Checking plasmid: pCas9_SCO5089(65)\n",
            "Match found in plasmid: pCas9_SCO5089(65)\n",
            "Repair template match found for gene: SCO5089\n",
            "Record added for gene: SCO5089\n",
            "\n",
            "Processing gene: SCO5090\n",
            "Checking plasmid: pCas9_SCO5090(546)\n",
            "Match found in plasmid: pCas9_SCO5090(546)\n",
            "Repair template match found for gene: SCO5090\n",
            "Record added for gene: SCO5090\n",
            "Checking plasmid: pCas9_SCO5090(537)\n",
            "Match found in plasmid: pCas9_SCO5090(537)\n",
            "Repair template match found for gene: SCO5090\n",
            "Record added for gene: SCO5090\n",
            "Checking plasmid: pCas9_SCO5091(48)\n",
            "Checking plasmid: pCas9_SCO5091(78)\n",
            "Checking plasmid: pCas9_SCO5092(25)\n",
            "Checking plasmid: pCas9_SCO5092(267)\n",
            "Checking plasmid: pCas9_SCO5089(61)\n",
            "Checking plasmid: pCas9_SCO5087(556)\n",
            "Checking plasmid: pCas9_SCO5087(656)\n",
            "Checking plasmid: pCas9_SCO5089(65)\n",
            "\n",
            "Processing gene: SCO5091\n",
            "Checking plasmid: pCas9_SCO5090(546)\n",
            "Checking plasmid: pCas9_SCO5090(537)\n",
            "Checking plasmid: pCas9_SCO5091(48)\n",
            "Match found in plasmid: pCas9_SCO5091(48)\n",
            "Repair template match found for gene: SCO5091\n",
            "Record added for gene: SCO5091\n",
            "Checking plasmid: pCas9_SCO5091(78)\n",
            "Match found in plasmid: pCas9_SCO5091(78)\n",
            "Repair template match found for gene: SCO5091\n",
            "Record added for gene: SCO5091\n",
            "Checking plasmid: pCas9_SCO5092(25)\n",
            "Checking plasmid: pCas9_SCO5092(267)\n",
            "Checking plasmid: pCas9_SCO5089(61)\n",
            "Checking plasmid: pCas9_SCO5087(556)\n",
            "Checking plasmid: pCas9_SCO5087(656)\n",
            "Checking plasmid: pCas9_SCO5089(65)\n",
            "\n",
            "Processing gene: SCO5092\n",
            "Checking plasmid: pCas9_SCO5090(546)\n",
            "Checking plasmid: pCas9_SCO5090(537)\n",
            "Checking plasmid: pCas9_SCO5091(48)\n",
            "Checking plasmid: pCas9_SCO5091(78)\n",
            "Checking plasmid: pCas9_SCO5092(25)\n",
            "Match found in plasmid: pCas9_SCO5092(25)\n",
            "Repair template match found for gene: SCO5092\n",
            "Record added for gene: SCO5092\n",
            "Checking plasmid: pCas9_SCO5092(267)\n",
            "Match found in plasmid: pCas9_SCO5092(267)\n",
            "Repair template match found for gene: SCO5092\n",
            "Record added for gene: SCO5092\n",
            "Checking plasmid: pCas9_SCO5089(61)\n",
            "Checking plasmid: pCas9_SCO5087(556)\n",
            "Checking plasmid: pCas9_SCO5087(656)\n",
            "Checking plasmid: pCas9_SCO5089(65)\n",
            "[{'gene_name': 'SCO5087', 'name': 'pCas9_SCO5087(556)', 'contig': Dseqrecord(o13279), 'up_forwar_p_name': 'SCO5087_repair_up_forwar_p', 'up_reverse_p_name': 'SCO5087_repair_up_reverse_p', 'up_forwar_p': SCO5087_repair_up_forwar_p 56-mer:5'-AAGGCCGCTTTTGCG..CAG-3', 'up_reverse_p': SCO5087_repair_up_reverse_p 36-mer:5'-GACTCCGGTGATGAG..GAC-3', 'up_forwar_primer_str': 'AAGGCCGCTTTTGCGGGATCTCGTCGAAGGCACTAGAAGGGACGCCGGAAATCCAG', 'up_reverse_primer_str': 'GACTCCGGTGATGAGGACGCCCCATCTCCCTTCGAC', 'up_forwar_p_anneal': Seq('GACGCCGGAAATCCAG'), 'up_reverse_p_anneal': Seq('CCCATCTCCCTTCGAC'), 'tm_up_forwar_p': 56, 'tm_up_reverse_p': 54, 'ta_up': 58, 'dw_forwar_p_name': 'SCO5087_repair_dw_forwar_p', 'dw_reverse_p_name': 'SCO5087_repair_dw_reverse_p', 'dw_forwar_p': SCO5087_repair_dw_forwar_p 36-mer:5'-GGCGGTCGAAGGGAG..CGG-3', 'dw_reverse_p': SCO5087_repair_dw_reverse_p 55-mer:5'-TCCCCGTCCGGGACC..CCC-3', 'dw_forwar_primer_str': 'GGCGGTCGAAGGGAGATGGGGCGTCCTCATCACCGG', 'dw_reverse_primer_str': 'TCCCCGTCCGGGACCCGCGCGGTCGATCCCCGCATATAGGGAGTAGAGGCGGCCC', 'dw_forwar_p_anneal': Seq('GCGTCCTCATCACCGG'), 'dw_reverse_p_anneal': Seq('GAGTAGAGGCGGCCC'), 'tm_dw_forwar_p': 58, 'tm_dw_reverse_p': 57, 'ta_dw': 61}, {'gene_name': 'SCO5087', 'name': 'pCas9_SCO5087(656)', 'contig': Dseqrecord(o13279), 'up_forwar_p_name': 'SCO5087_repair_up_forwar_p', 'up_reverse_p_name': 'SCO5087_repair_up_reverse_p', 'up_forwar_p': f1000 56-mer:5'-AAGGCCGCTTTTGCG..CAG-3', 'up_reverse_p': r1000 36-mer:5'-GACTCCGGTGATGAG..GAC-3', 'up_forwar_primer_str': 'AAGGCCGCTTTTGCGGGATCTCGTCGAAGGCACTAGAAGGGACGCCGGAAATCCAG', 'up_reverse_primer_str': 'GACTCCGGTGATGAGGACGCCCCATCTCCCTTCGAC', 'up_forwar_p_anneal': Seq('GACGCCGGAAATCCAG'), 'up_reverse_p_anneal': Seq('CCCATCTCCCTTCGAC'), 'tm_up_forwar_p': 56, 'tm_up_reverse_p': 54, 'ta_up': 58, 'dw_forwar_p_name': 'SCO5087_repair_dw_forwar_p', 'dw_reverse_p_name': 'SCO5087_repair_dw_reverse_p', 'dw_forwar_p': f1000 36-mer:5'-GGCGGTCGAAGGGAG..CGG-3', 'dw_reverse_p': r1000 55-mer:5'-TCCCCGTCCGGGACC..CCC-3', 'dw_forwar_primer_str': 'GGCGGTCGAAGGGAGATGGGGCGTCCTCATCACCGG', 'dw_reverse_primer_str': 'TCCCCGTCCGGGACCCGCGCGGTCGATCCCCGCATATAGGGAGTAGAGGCGGCCC', 'dw_forwar_p_anneal': Seq('GCGTCCTCATCACCGG'), 'dw_reverse_p_anneal': Seq('GAGTAGAGGCGGCCC'), 'tm_dw_forwar_p': 58, 'tm_dw_reverse_p': 57, 'ta_dw': 61}, {'gene_name': 'SCO5089', 'name': 'pCas9_SCO5089(61)', 'contig': Dseqrecord(o13279), 'up_forwar_p_name': 'SCO5089_repair_up_forwar_p', 'up_reverse_p_name': 'SCO5089_repair_up_reverse_p', 'up_forwar_p': SCO5089_repair_up_forwar_p 54-mer:5'-AAGGCCGCTTTTGCG..AGG-3', 'up_reverse_p': SCO5089_repair_up_reverse_p 38-mer:5'-CCGGTGTTCTCCCGG..TCG-3', 'up_forwar_primer_str': 'AAGGCCGCTTTTGCGGGATCTCGTCGAAGGCACTAGAAGGCCAGGACGCGAAGG', 'up_reverse_primer_str': 'CCGGTGTTCTCCCGGCCGGCGGCTTTCTCCAGTTCTCG', 'up_forwar_p_anneal': Seq('CCAGGACGCGAAGG'), 'up_reverse_p_anneal': Seq('GGCTTTCTCCAGTTCTCG'), 'tm_up_forwar_p': 54, 'tm_up_reverse_p': 56, 'ta_up': 58, 'dw_forwar_p_name': 'SCO5089_repair_dw_forwar_p', 'dw_reverse_p_name': 'SCO5089_repair_dw_reverse_p', 'dw_forwar_p': SCO5089_repair_dw_forwar_p 33-mer:5'-CCCGAGAACTGGAGA..AGA-3', 'dw_reverse_p': SCO5089_repair_dw_reverse_p 54-mer:5'-TCCCCGTCCGGGACC..CGT-3', 'dw_forwar_primer_str': 'CCCGAGAACTGGAGAAAGCCGCCGGCCGGGAGA', 'dw_reverse_primer_str': 'TCCCCGTCCGGGACCCGCGCGGTCGATCCCCGCATATAGGGGGGGCCTGCTCGT', 'dw_forwar_p_anneal': Seq('GCCGGCCGGGAGA'), 'dw_reverse_p_anneal': Seq('GGGGGCCTGCTCGT'), 'tm_dw_forwar_p': 59, 'tm_dw_reverse_p': 60, 'ta_dw': 63}, {'gene_name': 'SCO5089', 'name': 'pCas9_SCO5089(65)', 'contig': Dseqrecord(o13279), 'up_forwar_p_name': 'SCO5089_repair_up_forwar_p', 'up_reverse_p_name': 'SCO5089_repair_up_reverse_p', 'up_forwar_p': f1000 54-mer:5'-AAGGCCGCTTTTGCG..AGG-3', 'up_reverse_p': r1000 38-mer:5'-CCGGTGTTCTCCCGG..TCG-3', 'up_forwar_primer_str': 'AAGGCCGCTTTTGCGGGATCTCGTCGAAGGCACTAGAAGGCCAGGACGCGAAGG', 'up_reverse_primer_str': 'CCGGTGTTCTCCCGGCCGGCGGCTTTCTCCAGTTCTCG', 'up_forwar_p_anneal': Seq('CCAGGACGCGAAGG'), 'up_reverse_p_anneal': Seq('GGCTTTCTCCAGTTCTCG'), 'tm_up_forwar_p': 54, 'tm_up_reverse_p': 56, 'ta_up': 58, 'dw_forwar_p_name': 'SCO5089_repair_dw_forwar_p', 'dw_reverse_p_name': 'SCO5089_repair_dw_reverse_p', 'dw_forwar_p': f1000 33-mer:5'-CCCGAGAACTGGAGA..AGA-3', 'dw_reverse_p': r1000 54-mer:5'-TCCCCGTCCGGGACC..CGT-3', 'dw_forwar_primer_str': 'CCCGAGAACTGGAGAAAGCCGCCGGCCGGGAGA', 'dw_reverse_primer_str': 'TCCCCGTCCGGGACCCGCGCGGTCGATCCCCGCATATAGGGGGGGCCTGCTCGT', 'dw_forwar_p_anneal': Seq('GCCGGCCGGGAGA'), 'dw_reverse_p_anneal': Seq('GGGGGCCTGCTCGT'), 'tm_dw_forwar_p': 59, 'tm_dw_reverse_p': 60, 'ta_dw': 63}, {'gene_name': 'SCO5090', 'name': 'pCas9_SCO5090(546)', 'contig': Dseqrecord(o13279), 'up_forwar_p_name': 'SCO5090_repair_up_forwar_p', 'up_reverse_p_name': 'SCO5090_repair_up_reverse_p', 'up_forwar_p': SCO5090_repair_up_forwar_p 53-mer:5'-AAGGCCGCTTTTGCG..CCG-3', 'up_reverse_p': SCO5090_repair_up_reverse_p 35-mer:5'-GACCTCACGGACCTC..AGT-3', 'up_forwar_primer_str': 'AAGGCCGCTTTTGCGGGATCTCGTCGAAGGCACTAGAAGGGCCACGCCCGCCG', 'up_reverse_primer_str': 'GACCTCACGGACCTCGACCGGCCGCCTCGGCCAGT', 'up_forwar_p_anneal': Seq('GCCACGCCCGCCG'), 'up_reverse_p_anneal': Seq('GCCGCCTCGGCCAGT'), 'tm_up_forwar_p': 63, 'tm_up_reverse_p': 64, 'ta_up': 66, 'dw_forwar_p_name': 'SCO5090_repair_dw_forwar_p', 'dw_reverse_p_name': 'SCO5090_repair_dw_reverse_p', 'dw_forwar_p': SCO5090_repair_dw_forwar_p 34-mer:5'-GGCGCACTGGCCGAG..CGT-3', 'dw_reverse_p': SCO5090_repair_dw_reverse_p 53-mer:5'-TCCCCGTCCGGGACC..ACT-3', 'dw_forwar_primer_str': 'GGCGCACTGGCCGAGGCGGCCGGTCGAGGTCCGT', 'dw_reverse_primer_str': 'TCCCCGTCCGGGACCCGCGCGGTCGATCCCCGCATATAGGACCGTGCGGGACT', 'dw_forwar_p_anneal': Seq('CGGTCGAGGTCCGT'), 'dw_reverse_p_anneal': Seq('ACCGTGCGGGACT'), 'tm_dw_forwar_p': 56, 'tm_dw_reverse_p': 55, 'ta_dw': 59}, {'gene_name': 'SCO5090', 'name': 'pCas9_SCO5090(537)', 'contig': Dseqrecord(o13279), 'up_forwar_p_name': 'SCO5090_repair_up_forwar_p', 'up_reverse_p_name': 'SCO5090_repair_up_reverse_p', 'up_forwar_p': f1000 53-mer:5'-AAGGCCGCTTTTGCG..CCG-3', 'up_reverse_p': r1000 35-mer:5'-GACCTCACGGACCTC..AGT-3', 'up_forwar_primer_str': 'AAGGCCGCTTTTGCGGGATCTCGTCGAAGGCACTAGAAGGGCCACGCCCGCCG', 'up_reverse_primer_str': 'GACCTCACGGACCTCGACCGGCCGCCTCGGCCAGT', 'up_forwar_p_anneal': Seq('GCCACGCCCGCCG'), 'up_reverse_p_anneal': Seq('GCCGCCTCGGCCAGT'), 'tm_up_forwar_p': 63, 'tm_up_reverse_p': 64, 'ta_up': 66, 'dw_forwar_p_name': 'SCO5090_repair_dw_forwar_p', 'dw_reverse_p_name': 'SCO5090_repair_dw_reverse_p', 'dw_forwar_p': f1000 34-mer:5'-GGCGCACTGGCCGAG..CGT-3', 'dw_reverse_p': r1000 53-mer:5'-TCCCCGTCCGGGACC..ACT-3', 'dw_forwar_primer_str': 'GGCGCACTGGCCGAGGCGGCCGGTCGAGGTCCGT', 'dw_reverse_primer_str': 'TCCCCGTCCGGGACCCGCGCGGTCGATCCCCGCATATAGGACCGTGCGGGACT', 'dw_forwar_p_anneal': Seq('CGGTCGAGGTCCGT'), 'dw_reverse_p_anneal': Seq('ACCGTGCGGGACT'), 'tm_dw_forwar_p': 56, 'tm_dw_reverse_p': 55, 'ta_dw': 59}, {'gene_name': 'SCO5091', 'name': 'pCas9_SCO5091(48)', 'contig': Dseqrecord(o13279), 'up_forwar_p_name': 'SCO5091_repair_up_forwar_p', 'up_reverse_p_name': 'SCO5091_repair_up_reverse_p', 'up_forwar_p': SCO5091_repair_up_forwar_p 54-mer:5'-AAGGCCGCTTTTGCG..AGA-3', 'up_reverse_p': SCO5091_repair_up_reverse_p 33-mer:5'-GGCTGCCATCTTCGA..CGG-3', 'up_forwar_primer_str': 'AAGGCCGCTTTTGCGGGATCTCGTCGAAGGCACTAGAAGGGACACGCCGCGAGA', 'up_reverse_primer_str': 'GGCTGCCATCTTCGAACTCCGACGCCGGCCCGG', 'up_forwar_p_anneal': Seq('GACACGCCGCGAGA'), 'up_reverse_p_anneal': Seq('GACGCCGGCCCGG'), 'tm_up_forwar_p': 57, 'tm_up_reverse_p': 61, 'ta_up': 61, 'dw_forwar_p_name': 'SCO5091_repair_dw_forwar_p', 'dw_reverse_p_name': 'SCO5091_repair_dw_reverse_p', 'dw_forwar_p': SCO5091_repair_dw_forwar_p 39-mer:5'-GCCGCCGCCGGGCCG..AGC-3', 'dw_reverse_p': SCO5091_repair_dw_reverse_p 54-mer:5'-TCCCCGTCCGGGACC..CTG-3', 'dw_forwar_primer_str': 'GCCGCCGCCGGGCCGGCGTCGGAGTTCGAAGATGGCAGC', 'dw_reverse_primer_str': 'TCCCCGTCCGGGACCCGCGCGGTCGATCCCCGCATATAGGAGGCCGAGACCCTG', 'dw_forwar_p_anneal': Seq('GGAGTTCGAAGATGGCAGC'), 'dw_reverse_p_anneal': Seq('AGGCCGAGACCCTG'), 'tm_dw_forwar_p': 60, 'tm_dw_reverse_p': 55, 'ta_dw': 59}, {'gene_name': 'SCO5091', 'name': 'pCas9_SCO5091(78)', 'contig': Dseqrecord(o13279), 'up_forwar_p_name': 'SCO5091_repair_up_forwar_p', 'up_reverse_p_name': 'SCO5091_repair_up_reverse_p', 'up_forwar_p': f1000 54-mer:5'-AAGGCCGCTTTTGCG..AGA-3', 'up_reverse_p': r1000 33-mer:5'-GGCTGCCATCTTCGA..CGG-3', 'up_forwar_primer_str': 'AAGGCCGCTTTTGCGGGATCTCGTCGAAGGCACTAGAAGGGACACGCCGCGAGA', 'up_reverse_primer_str': 'GGCTGCCATCTTCGAACTCCGACGCCGGCCCGG', 'up_forwar_p_anneal': Seq('GACACGCCGCGAGA'), 'up_reverse_p_anneal': Seq('GACGCCGGCCCGG'), 'tm_up_forwar_p': 57, 'tm_up_reverse_p': 61, 'ta_up': 61, 'dw_forwar_p_name': 'SCO5091_repair_dw_forwar_p', 'dw_reverse_p_name': 'SCO5091_repair_dw_reverse_p', 'dw_forwar_p': f1000 39-mer:5'-GCCGCCGCCGGGCCG..AGC-3', 'dw_reverse_p': r1000 54-mer:5'-TCCCCGTCCGGGACC..CTG-3', 'dw_forwar_primer_str': 'GCCGCCGCCGGGCCGGCGTCGGAGTTCGAAGATGGCAGC', 'dw_reverse_primer_str': 'TCCCCGTCCGGGACCCGCGCGGTCGATCCCCGCATATAGGAGGCCGAGACCCTG', 'dw_forwar_p_anneal': Seq('GGAGTTCGAAGATGGCAGC'), 'dw_reverse_p_anneal': Seq('AGGCCGAGACCCTG'), 'tm_dw_forwar_p': 60, 'tm_dw_reverse_p': 55, 'ta_dw': 59}, {'gene_name': 'SCO5092', 'name': 'pCas9_SCO5092(25)', 'contig': Dseqrecord(o13279), 'up_forwar_p_name': 'SCO5092_repair_up_forwar_p', 'up_reverse_p_name': 'SCO5092_repair_up_reverse_p', 'up_forwar_p': SCO5092_repair_up_forwar_p 53-mer:5'-AAGGCCGCTTTTGCG..ACG-3', 'up_reverse_p': SCO5092_repair_up_reverse_p 39-mer:5'-GACGGCCTTCCCGGC..CGA-3', 'up_forwar_primer_str': 'AAGGCCGCTTTTGCGGGATCTCGTCGAAGGCACTAGAAGGACCGTCGCGGACG', 'up_reverse_primer_str': 'GACGGCCTTCCCGGCCGGCCCTTCGAACTCCCTAGGCGA', 'up_forwar_p_anneal': Seq('ACCGTCGCGGACG'), 'up_reverse_p_anneal': Seq('CTTCGAACTCCCTAGGCGA'), 'tm_up_forwar_p': 57, 'tm_up_reverse_p': 60, 'ta_up': 60, 'dw_forwar_p_name': 'SCO5092_repair_dw_forwar_p', 'dw_reverse_p_name': 'SCO5092_repair_dw_reverse_p', 'dw_forwar_p': SCO5092_repair_dw_forwar_p 33-mer:5'-CTCGCCTAGGGAGTT..GAA-3', 'dw_reverse_p': SCO5092_repair_dw_reverse_p 57-mer:5'-TCCCCGTCCGGGACC..GCG-3', 'dw_forwar_primer_str': 'CTCGCCTAGGGAGTTCGAAGGGCCGGCCGGGAA', 'dw_reverse_primer_str': 'TCCCCGTCCGGGACCCGCGCGGTCGATCCCCGCATATAGGACATCTCCGGGGTCGCG', 'dw_forwar_p_anneal': Seq('GGCCGGCCGGGAA'), 'dw_reverse_p_anneal': Seq('ACATCTCCGGGGTCGCG'), 'tm_dw_forwar_p': 60, 'tm_dw_reverse_p': 63, 'ta_dw': 64}, {'gene_name': 'SCO5092', 'name': 'pCas9_SCO5092(267)', 'contig': Dseqrecord(o13279), 'up_forwar_p_name': 'SCO5092_repair_up_forwar_p', 'up_reverse_p_name': 'SCO5092_repair_up_reverse_p', 'up_forwar_p': f1000 53-mer:5'-AAGGCCGCTTTTGCG..ACG-3', 'up_reverse_p': r1000 39-mer:5'-GACGGCCTTCCCGGC..CGA-3', 'up_forwar_primer_str': 'AAGGCCGCTTTTGCGGGATCTCGTCGAAGGCACTAGAAGGACCGTCGCGGACG', 'up_reverse_primer_str': 'GACGGCCTTCCCGGCCGGCCCTTCGAACTCCCTAGGCGA', 'up_forwar_p_anneal': Seq('ACCGTCGCGGACG'), 'up_reverse_p_anneal': Seq('CTTCGAACTCCCTAGGCGA'), 'tm_up_forwar_p': 57, 'tm_up_reverse_p': 60, 'ta_up': 60, 'dw_forwar_p_name': 'SCO5092_repair_dw_forwar_p', 'dw_reverse_p_name': 'SCO5092_repair_dw_reverse_p', 'dw_forwar_p': f1000 33-mer:5'-CTCGCCTAGGGAGTT..GAA-3', 'dw_reverse_p': r1000 57-mer:5'-TCCCCGTCCGGGACC..GCG-3', 'dw_forwar_primer_str': 'CTCGCCTAGGGAGTTCGAAGGGCCGGCCGGGAA', 'dw_reverse_primer_str': 'TCCCCGTCCGGGACCCGCGCGGTCGATCCCCGCATATAGGACATCTCCGGGGTCGCG', 'dw_forwar_p_anneal': Seq('GGCCGGCCGGGAA'), 'dw_reverse_p_anneal': Seq('ACATCTCCGGGGTCGCG'), 'tm_dw_forwar_p': 60, 'tm_dw_reverse_p': 63, 'ta_dw': 64}]\n"
          ]
        }
      ],
      "source": [
        "if in_frame_deletion:\n",
        "    \n",
        "\n",
        "    # Make repair templates\n",
        "    repair_templates_data = find_up_dw_repair_templates(genome, \n",
        "                                                        genes_to_KO, \n",
        "                                                        target_tm=melting_temperature, \n",
        "                                                        primer_tm_kwargs={'conc':primer_concentration, 'prodcode':chosen_polymerase} , \n",
        "                                                        repair_length=repair_length, )\n",
        "\n",
        "    # Digest the plasmids\n",
        "    processed_records = [Dseqrecord(record, circular=True).cut(StuI)[0] for record in sgRNA_vectors]\n",
        "\n",
        "    # Rename them appropriately\n",
        "    for i in range(len(processed_records)):\n",
        "        processed_records[i].name = sgRNA_vectors[i].name\n",
        "        print(processed_records[i].name)\n",
        "\n",
        "    # Assembly \n",
        "    assembly_data = assemble_multiple_plasmids_with_repair_templates_for_deletion(genes_to_KO, processed_records, \n",
        "                                                                                repair_templates_data, \n",
        "                                                                                overlap=overlap_length)\n",
        "    # updating the primer names to something systematic.\n",
        "    update_primer_names(assembly_data)\n",
        "    print(assembly_data)\n",
        "\n",
        "    # Parse through the primer df\n",
        "    primer_df = create_primer_df_from_dict(assembly_data)\n",
        "    unique_df = primer_df.drop_duplicates(keep='first')\n",
        "\n",
        "    # IDT df\n",
        "    idt_df = create_idt_order_dataframe(unique_df, concentration=\"25nm\", purification=\"STD\")\n",
        "\n",
        "    # Contigs\n",
        "    assembled_contigs = []\n",
        "    for data in assembly_data: \n",
        "        contig_record = data['contig']\n",
        "        contig_record.id = f\"{data['name']}_w_rep\"\n",
        "        contig_record.name = f\"{data['name']}_w_rep\"\n",
        "        assembled_contigs.append(contig_record)\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 40,
      "metadata": {},
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "   template   direction f_primer_anneal(5-3) r_primer_anneal(5-3)  f_tm  r_tm  \\\n",
            "0   SCO5087    upstream     GACGCCGGAAATCCAG     CCCATCTCCCTTCGAC    56    54   \n",
            "1   SCO5087  downstream     GCGTCCTCATCACCGG      GAGTAGAGGCGGCCC    58    57   \n",
            "4   SCO5089    upstream       CCAGGACGCGAAGG   GGCTTTCTCCAGTTCTCG    54    56   \n",
            "5   SCO5089  downstream        GCCGGCCGGGAGA       GGGGGCCTGCTCGT    59    60   \n",
            "8   SCO5090    upstream        GCCACGCCCGCCG      GCCGCCTCGGCCAGT    63    64   \n",
            "9   SCO5090  downstream       CGGTCGAGGTCCGT        ACCGTGCGGGACT    56    55   \n",
            "12  SCO5091    upstream       GACACGCCGCGAGA        GACGCCGGCCCGG    57    61   \n",
            "13  SCO5091  downstream  GGAGTTCGAAGATGGCAGC       AGGCCGAGACCCTG    60    55   \n",
            "16  SCO5092    upstream        ACCGTCGCGGACG  CTTCGAACTCCCTAGGCGA    57    60   \n",
            "17  SCO5092  downstream        GGCCGGCCGGGAA    ACATCTCCGGGGTCGCG    60    63   \n",
            "\n",
            "    ta                            f_primer_sequences(5-3)  \\\n",
            "0   58  AAGGCCGCTTTTGCGGGATCTCGTCGAAGGCACTAGAAGGGACGCC...   \n",
            "1   61               GGCGGTCGAAGGGAGATGGGGCGTCCTCATCACCGG   \n",
            "4   58  AAGGCCGCTTTTGCGGGATCTCGTCGAAGGCACTAGAAGGCCAGGA...   \n",
            "5   63                  CCCGAGAACTGGAGAAAGCCGCCGGCCGGGAGA   \n",
            "8   66  AAGGCCGCTTTTGCGGGATCTCGTCGAAGGCACTAGAAGGGCCACG...   \n",
            "9   59                 GGCGCACTGGCCGAGGCGGCCGGTCGAGGTCCGT   \n",
            "12  61  AAGGCCGCTTTTGCGGGATCTCGTCGAAGGCACTAGAAGGGACACG...   \n",
            "13  59            GCCGCCGCCGGGCCGGCGTCGGAGTTCGAAGATGGCAGC   \n",
            "16  60  AAGGCCGCTTTTGCGGGATCTCGTCGAAGGCACTAGAAGGACCGTC...   \n",
            "17  64                  CTCGCCTAGGGAGTTCGAAGGGCCGGCCGGGAA   \n",
            "\n",
            "                              r_primer_sequences(5-3)  \\\n",
            "0                GACTCCGGTGATGAGGACGCCCCATCTCCCTTCGAC   \n",
            "1   TCCCCGTCCGGGACCCGCGCGGTCGATCCCCGCATATAGGGAGTAG...   \n",
            "4              CCGGTGTTCTCCCGGCCGGCGGCTTTCTCCAGTTCTCG   \n",
            "5   TCCCCGTCCGGGACCCGCGCGGTCGATCCCCGCATATAGGGGGGGC...   \n",
            "8                 GACCTCACGGACCTCGACCGGCCGCCTCGGCCAGT   \n",
            "9   TCCCCGTCCGGGACCCGCGCGGTCGATCCCCGCATATAGGACCGTG...   \n",
            "12                  GGCTGCCATCTTCGAACTCCGACGCCGGCCCGG   \n",
            "13  TCCCCGTCCGGGACCCGCGCGGTCGATCCCCGCATATAGGAGGCCG...   \n",
            "16            GACGGCCTTCCCGGCCGGCCCTTCGAACTCCCTAGGCGA   \n",
            "17  TCCCCGTCCGGGACCCGCGCGGTCGATCCCCGCATATAGGACATCT...   \n",
            "\n",
            "                 f_primer_name                r_primer_name  \n",
            "0   SCO5087_repair_up_forwar_p  SCO5087_repair_up_reverse_p  \n",
            "1   SCO5087_repair_dw_forwar_p  SCO5087_repair_dw_reverse_p  \n",
            "4   SCO5089_repair_up_forwar_p  SCO5089_repair_up_reverse_p  \n",
            "5   SCO5089_repair_dw_forwar_p  SCO5089_repair_dw_reverse_p  \n",
            "8   SCO5090_repair_up_forwar_p  SCO5090_repair_up_reverse_p  \n",
            "9   SCO5090_repair_dw_forwar_p  SCO5090_repair_dw_reverse_p  \n",
            "12  SCO5091_repair_up_forwar_p  SCO5091_repair_up_reverse_p  \n",
            "13  SCO5091_repair_dw_forwar_p  SCO5091_repair_dw_reverse_p  \n",
            "16  SCO5092_repair_up_forwar_p  SCO5092_repair_up_reverse_p  \n",
            "17  SCO5092_repair_dw_forwar_p  SCO5092_repair_dw_reverse_p  \n",
            "        sgRNA plasmid              plasmid_name  \\\n",
            "0  pCas9_SCO5090(546)  pCas9_SCO5090(546)_w_rep   \n",
            "1  pCas9_SCO5090(537)  pCas9_SCO5090(537)_w_rep   \n",
            "2   pCas9_SCO5091(48)   pCas9_SCO5091(48)_w_rep   \n",
            "3   pCas9_SCO5091(78)   pCas9_SCO5091(78)_w_rep   \n",
            "4   pCas9_SCO5092(25)   pCas9_SCO5092(25)_w_rep   \n",
            "5  pCas9_SCO5092(267)  pCas9_SCO5092(267)_w_rep   \n",
            "6   pCas9_SCO5089(61)   pCas9_SCO5089(61)_w_rep   \n",
            "7  pCas9_SCO5087(556)  pCas9_SCO5087(556)_w_rep   \n",
            "8  pCas9_SCO5087(656)  pCas9_SCO5087(656)_w_rep   \n",
            "9   pCas9_SCO5089(65)   pCas9_SCO5089(65)_w_rep   \n",
            "\n",
            "         which workflow to proceed with  sgRNA plasmid #StuI sites  \\\n",
            "0  Proceed with sgRNA integration first                          1   \n",
            "1  Proceed with sgRNA integration first                          1   \n",
            "2  Proceed with sgRNA integration first                          1   \n",
            "3  Proceed with sgRNA integration first                          1   \n",
            "4  Proceed with sgRNA integration first                          1   \n",
            "5  Proceed with sgRNA integration first                          1   \n",
            "6  Proceed with sgRNA integration first                          1   \n",
            "7  Proceed with sgRNA integration first                          1   \n",
            "8  Proceed with sgRNA integration first                          1   \n",
            "9  Proceed with sgRNA integration first                          1   \n",
            "\n",
            "   repair template plasmid #NcoI sites  \n",
            "0                                    4  \n",
            "1                                    4  \n",
            "2                                    2  \n",
            "3                                    2  \n",
            "4                                    2  \n",
            "5                                    2  \n",
            "6                                    1  \n",
            "7                                    0  \n",
            "8                                    0  \n",
            "9                                    1  \n"
          ]
        }
      ],
      "source": [
        "if in_frame_deletion:\n",
        "    print(unique_df)\n",
        "    workflow_df = determine_workflow_order_for_plasmids(sgRNA_vectors, \n",
        "                                                            assembled_contigs,\n",
        "                                                            [\"StuI\"], [ \"NcoI\"])\n",
        "    print(workflow_df )"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 41,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>plasmid_name</th>\n",
              "      <th>date</th>\n",
              "      <th>original_plasmid</th>\n",
              "      <th>integration</th>\n",
              "      <th>size</th>\n",
              "      <th>sgRNA plasmid</th>\n",
              "      <th>which workflow to proceed with</th>\n",
              "      <th>sgRNA plasmid #StuI sites</th>\n",
              "      <th>repair template plasmid #NcoI sites</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>pCas9_SCO5087(556)_w_rep</td>\n",
              "      <td>2025-08-12</td>\n",
              "      <td>pCRISPR-Cas9</td>\n",
              "      <td>sgRNA_SCO5090(546)</td>\n",
              "      <td>13279</td>\n",
              "      <td>pCas9_SCO5087(556)</td>\n",
              "      <td>Proceed with sgRNA integration first</td>\n",
              "      <td>1</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>pCas9_SCO5087(656)_w_rep</td>\n",
              "      <td>2025-08-12</td>\n",
              "      <td>pCRISPR-Cas9</td>\n",
              "      <td>sgRNA_SCO5090(537)</td>\n",
              "      <td>13279</td>\n",
              "      <td>pCas9_SCO5087(656)</td>\n",
              "      <td>Proceed with sgRNA integration first</td>\n",
              "      <td>1</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>pCas9_SCO5089(61)_w_rep</td>\n",
              "      <td>2025-08-12</td>\n",
              "      <td>pCRISPR-Cas9</td>\n",
              "      <td>sgRNA_SCO5091(48)</td>\n",
              "      <td>13279</td>\n",
              "      <td>pCas9_SCO5089(61)</td>\n",
              "      <td>Proceed with sgRNA integration first</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>pCas9_SCO5089(65)_w_rep</td>\n",
              "      <td>2025-08-12</td>\n",
              "      <td>pCRISPR-Cas9</td>\n",
              "      <td>sgRNA_SCO5091(78)</td>\n",
              "      <td>13279</td>\n",
              "      <td>pCas9_SCO5089(65)</td>\n",
              "      <td>Proceed with sgRNA integration first</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>pCas9_SCO5090(546)_w_rep</td>\n",
              "      <td>2025-08-12</td>\n",
              "      <td>pCRISPR-Cas9</td>\n",
              "      <td>sgRNA_SCO5092(25)</td>\n",
              "      <td>13279</td>\n",
              "      <td>pCas9_SCO5090(546)</td>\n",
              "      <td>Proceed with sgRNA integration first</td>\n",
              "      <td>1</td>\n",
              "      <td>4</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>5</th>\n",
              "      <td>pCas9_SCO5090(537)_w_rep</td>\n",
              "      <td>2025-08-12</td>\n",
              "      <td>pCRISPR-Cas9</td>\n",
              "      <td>sgRNA_SCO5092(267)</td>\n",
              "      <td>13279</td>\n",
              "      <td>pCas9_SCO5090(537)</td>\n",
              "      <td>Proceed with sgRNA integration first</td>\n",
              "      <td>1</td>\n",
              "      <td>4</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>6</th>\n",
              "      <td>pCas9_SCO5091(48)_w_rep</td>\n",
              "      <td>2025-08-12</td>\n",
              "      <td>pCRISPR-Cas9</td>\n",
              "      <td>sgRNA_SCO5089(61)</td>\n",
              "      <td>13279</td>\n",
              "      <td>pCas9_SCO5091(48)</td>\n",
              "      <td>Proceed with sgRNA integration first</td>\n",
              "      <td>1</td>\n",
              "      <td>2</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>7</th>\n",
              "      <td>pCas9_SCO5091(78)_w_rep</td>\n",
              "      <td>2025-08-12</td>\n",
              "      <td>pCRISPR-Cas9</td>\n",
              "      <td>sgRNA_SCO5087(556)</td>\n",
              "      <td>13279</td>\n",
              "      <td>pCas9_SCO5091(78)</td>\n",
              "      <td>Proceed with sgRNA integration first</td>\n",
              "      <td>1</td>\n",
              "      <td>2</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>8</th>\n",
              "      <td>pCas9_SCO5092(25)_w_rep</td>\n",
              "      <td>2025-08-12</td>\n",
              "      <td>pCRISPR-Cas9</td>\n",
              "      <td>sgRNA_SCO5087(656)</td>\n",
              "      <td>13279</td>\n",
              "      <td>pCas9_SCO5092(25)</td>\n",
              "      <td>Proceed with sgRNA integration first</td>\n",
              "      <td>1</td>\n",
              "      <td>2</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>9</th>\n",
              "      <td>pCas9_SCO5092(267)_w_rep</td>\n",
              "      <td>2025-08-12</td>\n",
              "      <td>pCRISPR-Cas9</td>\n",
              "      <td>sgRNA_SCO5089(65)</td>\n",
              "      <td>13279</td>\n",
              "      <td>pCas9_SCO5092(267)</td>\n",
              "      <td>Proceed with sgRNA integration first</td>\n",
              "      <td>1</td>\n",
              "      <td>2</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "               plasmid_name        date original_plasmid         integration  \\\n",
              "0  pCas9_SCO5087(556)_w_rep  2025-08-12     pCRISPR-Cas9  sgRNA_SCO5090(546)   \n",
              "1  pCas9_SCO5087(656)_w_rep  2025-08-12     pCRISPR-Cas9  sgRNA_SCO5090(537)   \n",
              "2   pCas9_SCO5089(61)_w_rep  2025-08-12     pCRISPR-Cas9   sgRNA_SCO5091(48)   \n",
              "3   pCas9_SCO5089(65)_w_rep  2025-08-12     pCRISPR-Cas9   sgRNA_SCO5091(78)   \n",
              "4  pCas9_SCO5090(546)_w_rep  2025-08-12     pCRISPR-Cas9   sgRNA_SCO5092(25)   \n",
              "5  pCas9_SCO5090(537)_w_rep  2025-08-12     pCRISPR-Cas9  sgRNA_SCO5092(267)   \n",
              "6   pCas9_SCO5091(48)_w_rep  2025-08-12     pCRISPR-Cas9   sgRNA_SCO5089(61)   \n",
              "7   pCas9_SCO5091(78)_w_rep  2025-08-12     pCRISPR-Cas9  sgRNA_SCO5087(556)   \n",
              "8   pCas9_SCO5092(25)_w_rep  2025-08-12     pCRISPR-Cas9  sgRNA_SCO5087(656)   \n",
              "9  pCas9_SCO5092(267)_w_rep  2025-08-12     pCRISPR-Cas9   sgRNA_SCO5089(65)   \n",
              "\n",
              "    size       sgRNA plasmid        which workflow to proceed with  \\\n",
              "0  13279  pCas9_SCO5087(556)  Proceed with sgRNA integration first   \n",
              "1  13279  pCas9_SCO5087(656)  Proceed with sgRNA integration first   \n",
              "2  13279   pCas9_SCO5089(61)  Proceed with sgRNA integration first   \n",
              "3  13279   pCas9_SCO5089(65)  Proceed with sgRNA integration first   \n",
              "4  13279  pCas9_SCO5090(546)  Proceed with sgRNA integration first   \n",
              "5  13279  pCas9_SCO5090(537)  Proceed with sgRNA integration first   \n",
              "6  13279   pCas9_SCO5091(48)  Proceed with sgRNA integration first   \n",
              "7  13279   pCas9_SCO5091(78)  Proceed with sgRNA integration first   \n",
              "8  13279   pCas9_SCO5092(25)  Proceed with sgRNA integration first   \n",
              "9  13279  pCas9_SCO5092(267)  Proceed with sgRNA integration first   \n",
              "\n",
              "   sgRNA plasmid #StuI sites  repair template plasmid #NcoI sites  \n",
              "0                          1                                    0  \n",
              "1                          1                                    0  \n",
              "2                          1                                    1  \n",
              "3                          1                                    1  \n",
              "4                          1                                    4  \n",
              "5                          1                                    4  \n",
              "6                          1                                    2  \n",
              "7                          1                                    2  \n",
              "8                          1                                    2  \n",
              "9                          1                                    2  "
            ]
          },
          "execution_count": 41,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "if in_frame_deletion:\n",
        "    integration_names = filtered_df.apply(lambda row: f\"sgRNA_{row['locus_tag']}({row['sgrna_loc']})\", axis=1).tolist()\n",
        "    plasmid_metadata_df = extract_metadata_to_dataframe(assembled_contigs,\n",
        "                                                        clean_plasmid,\n",
        "                                                        integration_names)\n",
        "    workflow_df = determine_workflow_order_for_plasmids(sgRNA_vectors, \n",
        "                                                            assembled_contigs,\n",
        "                                                            [\"StuI\",], [ \"NcoI\"])\n",
        "    \n",
        "    plasmid_metadata_df = pd.merge(plasmid_metadata_df, workflow_df, on='plasmid_name', how='inner')  # or 'left'/'right' depending on what you need\n",
        "\n",
        "else: \n",
        "    integration_names = filtered_df.apply(lambda row: f\"sgRNA_{row['locus_tag']}({row['sgrna_loc']})\", axis=1).tolist()\n",
        "    plasmid_metadata_df = extract_metadata_to_dataframe(sgRNA_vectors,\n",
        "                                                        clean_plasmid,\n",
        "                                                        integration_names)\n",
        "\n",
        "plasmid_metadata_df"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 42,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>locus tag</th>\n",
              "      <th>f_primer_name</th>\n",
              "      <th>r_primer_name</th>\n",
              "      <th>f_primer_sequences(5-3)</th>\n",
              "      <th>r_primer_sequences(5-3)</th>\n",
              "      <th>f_tm</th>\n",
              "      <th>r_tm</th>\n",
              "      <th>ta</th>\n",
              "      <th>flanking_region</th>\n",
              "      <th>annealing_temperature</th>\n",
              "      <th>...</th>\n",
              "      <th>heterodimer_tm</th>\n",
              "      <th>heterodimer_deltaG (kcal/mol)</th>\n",
              "      <th>hairpin_forward_structure_found</th>\n",
              "      <th>hairpin_forward_tm</th>\n",
              "      <th>hairpin_forward_deltaG (kcal/mol)</th>\n",
              "      <th>hairpin_reverse_structure_found</th>\n",
              "      <th>hairpin_reverse_tm</th>\n",
              "      <th>hairpin_reverse_deltaG (kcal/mol)</th>\n",
              "      <th>f_tm</th>\n",
              "      <th>r_tm</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>SCO5091</td>\n",
              "      <td>SCO5091_fwd_checking_primer</td>\n",
              "      <td>SCO5091_rev_checking_primer</td>\n",
              "      <td>AGCCGCCCGAGGAGCTGG</td>\n",
              "      <td>CAGGCACCCGCCGCCGAG</td>\n",
              "      <td>69</td>\n",
              "      <td>70</td>\n",
              "      <td>72</td>\n",
              "      <td>502</td>\n",
              "      <td>72</td>\n",
              "      <td>...</td>\n",
              "      <td>-19.574171</td>\n",
              "      <td>-0.275126</td>\n",
              "      <td>False</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>False</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>69</td>\n",
              "      <td>70</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>SCO5092</td>\n",
              "      <td>SCO5092_fwd_checking_primer</td>\n",
              "      <td>SCO5092_rev_checking_primer</td>\n",
              "      <td>ACGTCGGCGAGCGACAGG</td>\n",
              "      <td>CACCCCCCGCACCCACCG</td>\n",
              "      <td>66</td>\n",
              "      <td>70</td>\n",
              "      <td>69</td>\n",
              "      <td>508</td>\n",
              "      <td>69</td>\n",
              "      <td>...</td>\n",
              "      <td>2.724266</td>\n",
              "      <td>1.657074</td>\n",
              "      <td>False</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>False</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>66</td>\n",
              "      <td>70</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>SCO5090</td>\n",
              "      <td>SCO5090_fwd_checking_primer</td>\n",
              "      <td>SCO5090_rev_checking_primer</td>\n",
              "      <td>CTCTACTCCGGCGGCGGC</td>\n",
              "      <td>CGACATGACGACGTCCCCCGC</td>\n",
              "      <td>67</td>\n",
              "      <td>69</td>\n",
              "      <td>70</td>\n",
              "      <td>509</td>\n",
              "      <td>70</td>\n",
              "      <td>...</td>\n",
              "      <td>4.251557</td>\n",
              "      <td>0.508299</td>\n",
              "      <td>False</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>False</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>67</td>\n",
              "      <td>69</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>SCO5087</td>\n",
              "      <td>SCO5087_fwd_checking_primer</td>\n",
              "      <td>SCO5087_rev_checking_primer</td>\n",
              "      <td>CTTTCCGCCGGTCGAGGC</td>\n",
              "      <td>GCGGATCGTACGGCGGGC</td>\n",
              "      <td>65</td>\n",
              "      <td>68</td>\n",
              "      <td>68</td>\n",
              "      <td>524</td>\n",
              "      <td>68</td>\n",
              "      <td>...</td>\n",
              "      <td>32.653761</td>\n",
              "      <td>-1.899417</td>\n",
              "      <td>False</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>False</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>65</td>\n",
              "      <td>68</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>SCO5089</td>\n",
              "      <td>SCO5089_fwd_checking_primer</td>\n",
              "      <td>SCO5089_rev_checking_primer</td>\n",
              "      <td>CTGGAGGACAGCGCCGCG</td>\n",
              "      <td>GTTCCTGCCACAGGTCGCTGC</td>\n",
              "      <td>67</td>\n",
              "      <td>68</td>\n",
              "      <td>70</td>\n",
              "      <td>543</td>\n",
              "      <td>70</td>\n",
              "      <td>...</td>\n",
              "      <td>7.747596</td>\n",
              "      <td>-1.793463</td>\n",
              "      <td>False</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>False</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>67</td>\n",
              "      <td>68</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>5 rows × 25 columns</p>\n",
              "</div>"
            ],
            "text/plain": [
              "  locus tag                f_primer_name                r_primer_name  \\\n",
              "0   SCO5091  SCO5091_fwd_checking_primer  SCO5091_rev_checking_primer   \n",
              "1   SCO5092  SCO5092_fwd_checking_primer  SCO5092_rev_checking_primer   \n",
              "2   SCO5090  SCO5090_fwd_checking_primer  SCO5090_rev_checking_primer   \n",
              "3   SCO5087  SCO5087_fwd_checking_primer  SCO5087_rev_checking_primer   \n",
              "4   SCO5089  SCO5089_fwd_checking_primer  SCO5089_rev_checking_primer   \n",
              "\n",
              "  f_primer_sequences(5-3) r_primer_sequences(5-3)  f_tm  r_tm  ta  \\\n",
              "0      AGCCGCCCGAGGAGCTGG      CAGGCACCCGCCGCCGAG    69    70  72   \n",
              "1      ACGTCGGCGAGCGACAGG      CACCCCCCGCACCCACCG    66    70  69   \n",
              "2      CTCTACTCCGGCGGCGGC   CGACATGACGACGTCCCCCGC    67    69  70   \n",
              "3      CTTTCCGCCGGTCGAGGC      GCGGATCGTACGGCGGGC    65    68  68   \n",
              "4      CTGGAGGACAGCGCCGCG   GTTCCTGCCACAGGTCGCTGC    67    68  70   \n",
              "\n",
              "  flanking_region  annealing_temperature  ... heterodimer_tm  \\\n",
              "0             502                     72  ...     -19.574171   \n",
              "1             508                     69  ...       2.724266   \n",
              "2             509                     70  ...       4.251557   \n",
              "3             524                     68  ...      32.653761   \n",
              "4             543                     70  ...       7.747596   \n",
              "\n",
              "   heterodimer_deltaG (kcal/mol)  hairpin_forward_structure_found  \\\n",
              "0                      -0.275126                            False   \n",
              "1                       1.657074                            False   \n",
              "2                       0.508299                            False   \n",
              "3                      -1.899417                            False   \n",
              "4                      -1.793463                            False   \n",
              "\n",
              "   hairpin_forward_tm  hairpin_forward_deltaG (kcal/mol)  \\\n",
              "0                 0.0                                0.0   \n",
              "1                 0.0                                0.0   \n",
              "2                 0.0                                0.0   \n",
              "3                 0.0                                0.0   \n",
              "4                 0.0                                0.0   \n",
              "\n",
              "   hairpin_reverse_structure_found  hairpin_reverse_tm  \\\n",
              "0                            False                 0.0   \n",
              "1                            False                 0.0   \n",
              "2                            False                 0.0   \n",
              "3                            False                 0.0   \n",
              "4                            False                 0.0   \n",
              "\n",
              "   hairpin_reverse_deltaG (kcal/mol)  f_tm  r_tm  \n",
              "0                                0.0    69    70  \n",
              "1                                0.0    66    70  \n",
              "2                                0.0    67    69  \n",
              "3                                0.0    65    68  \n",
              "4                                0.0    67    68  \n",
              "\n",
              "[5 rows x 25 columns]"
            ]
          },
          "execution_count": 42,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "# Getting checking primers\n",
        "checking_primers_df = find_best_check_primers_from_genome(genome, \n",
        "                                       genes_to_KO, \n",
        "                                       flanking_region=flanking_region,\n",
        "                                       target_tm = melting_temperature, \n",
        "                                        primer_concentration = primer_concentration, \n",
        "                                        polymerase = chosen_polymerase, \n",
        "                                        limit=18)\n",
        "checking_primers_df\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 43,
      "metadata": {},
      "outputs": [],
      "source": [
        "# make column labels unique inside each DF\n",
        "unique_df = unique_df.loc[:, ~unique_df.columns.duplicated(keep='first')]\n",
        "checking_primers_df_copy = checking_primers_df.copy()\n",
        "# rename\n",
        "checking_primers_df_copy = checking_primers_df_copy.rename(columns={'locus tag': 'template'})\n",
        "checking_primers_df_copy = checking_primers_df_copy.loc[:, ~checking_primers_df_copy.columns.duplicated(keep='first')]\n",
        "\n",
        "# now safe to concat rows\n",
        "pcr_table = pd.concat([unique_df, checking_primers_df_copy], ignore_index=True)\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 44,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>Name</th>\n",
              "      <th>Sequence</th>\n",
              "      <th>Concentration</th>\n",
              "      <th>Purification</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>SCO5091_fwd_checking_primer</td>\n",
              "      <td>AGCCGCCCGAGGAGCTGG</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>SCO5092_fwd_checking_primer</td>\n",
              "      <td>ACGTCGGCGAGCGACAGG</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>SCO5090_fwd_checking_primer</td>\n",
              "      <td>CTCTACTCCGGCGGCGGC</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>SCO5087_fwd_checking_primer</td>\n",
              "      <td>CTTTCCGCCGGTCGAGGC</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>SCO5089_fwd_checking_primer</td>\n",
              "      <td>CTGGAGGACAGCGCCGCG</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>5</th>\n",
              "      <td>SCO5091_rev_checking_primer</td>\n",
              "      <td>CAGGCACCCGCCGCCGAG</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>6</th>\n",
              "      <td>SCO5092_rev_checking_primer</td>\n",
              "      <td>CACCCCCCGCACCCACCG</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>7</th>\n",
              "      <td>SCO5090_rev_checking_primer</td>\n",
              "      <td>CGACATGACGACGTCCCCCGC</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>8</th>\n",
              "      <td>SCO5087_rev_checking_primer</td>\n",
              "      <td>GCGGATCGTACGGCGGGC</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>9</th>\n",
              "      <td>SCO5089_rev_checking_primer</td>\n",
              "      <td>GTTCCTGCCACAGGTCGCTGC</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "                          Name               Sequence Concentration  \\\n",
              "0  SCO5091_fwd_checking_primer     AGCCGCCCGAGGAGCTGG          25nm   \n",
              "1  SCO5092_fwd_checking_primer     ACGTCGGCGAGCGACAGG          25nm   \n",
              "2  SCO5090_fwd_checking_primer     CTCTACTCCGGCGGCGGC          25nm   \n",
              "3  SCO5087_fwd_checking_primer     CTTTCCGCCGGTCGAGGC          25nm   \n",
              "4  SCO5089_fwd_checking_primer     CTGGAGGACAGCGCCGCG          25nm   \n",
              "5  SCO5091_rev_checking_primer     CAGGCACCCGCCGCCGAG          25nm   \n",
              "6  SCO5092_rev_checking_primer     CACCCCCCGCACCCACCG          25nm   \n",
              "7  SCO5090_rev_checking_primer  CGACATGACGACGTCCCCCGC          25nm   \n",
              "8  SCO5087_rev_checking_primer     GCGGATCGTACGGCGGGC          25nm   \n",
              "9  SCO5089_rev_checking_primer  GTTCCTGCCACAGGTCGCTGC          25nm   \n",
              "\n",
              "  Purification  \n",
              "0          STD  \n",
              "1          STD  \n",
              "2          STD  \n",
              "3          STD  \n",
              "4          STD  \n",
              "5          STD  \n",
              "6          STD  \n",
              "7          STD  \n",
              "8          STD  \n",
              "9          STD  "
            ]
          },
          "execution_count": 44,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "# making the primers into dseqrecords\n",
        "checking_primers_df_idt = create_idt_order_dataframe(checking_primers_df)\n",
        "checking_primers_df_idt\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 45,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>Name</th>\n",
              "      <th>Sequence</th>\n",
              "      <th>Concentration</th>\n",
              "      <th>Purification</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>SCO5087_repair_up_forwar_p</td>\n",
              "      <td>AAGGCCGCTTTTGCGGGATCTCGTCGAAGGCACTAGAAGGGACGCC...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>SCO5087_repair_dw_forwar_p</td>\n",
              "      <td>GGCGGTCGAAGGGAGATGGGGCGTCCTCATCACCGG</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>SCO5089_repair_up_forwar_p</td>\n",
              "      <td>AAGGCCGCTTTTGCGGGATCTCGTCGAAGGCACTAGAAGGCCAGGA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>SCO5089_repair_dw_forwar_p</td>\n",
              "      <td>CCCGAGAACTGGAGAAAGCCGCCGGCCGGGAGA</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>SCO5090_repair_up_forwar_p</td>\n",
              "      <td>AAGGCCGCTTTTGCGGGATCTCGTCGAAGGCACTAGAAGGGCCACG...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>5</th>\n",
              "      <td>SCO5090_repair_dw_forwar_p</td>\n",
              "      <td>GGCGCACTGGCCGAGGCGGCCGGTCGAGGTCCGT</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>6</th>\n",
              "      <td>SCO5091_repair_up_forwar_p</td>\n",
              "      <td>AAGGCCGCTTTTGCGGGATCTCGTCGAAGGCACTAGAAGGGACACG...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>7</th>\n",
              "      <td>SCO5091_repair_dw_forwar_p</td>\n",
              "      <td>GCCGCCGCCGGGCCGGCGTCGGAGTTCGAAGATGGCAGC</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>8</th>\n",
              "      <td>SCO5092_repair_up_forwar_p</td>\n",
              "      <td>AAGGCCGCTTTTGCGGGATCTCGTCGAAGGCACTAGAAGGACCGTC...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>9</th>\n",
              "      <td>SCO5092_repair_dw_forwar_p</td>\n",
              "      <td>CTCGCCTAGGGAGTTCGAAGGGCCGGCCGGGAA</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>10</th>\n",
              "      <td>SCO5087_repair_up_reverse_p</td>\n",
              "      <td>GACTCCGGTGATGAGGACGCCCCATCTCCCTTCGAC</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>11</th>\n",
              "      <td>SCO5087_repair_dw_reverse_p</td>\n",
              "      <td>TCCCCGTCCGGGACCCGCGCGGTCGATCCCCGCATATAGGGAGTAG...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>12</th>\n",
              "      <td>SCO5089_repair_up_reverse_p</td>\n",
              "      <td>CCGGTGTTCTCCCGGCCGGCGGCTTTCTCCAGTTCTCG</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>13</th>\n",
              "      <td>SCO5089_repair_dw_reverse_p</td>\n",
              "      <td>TCCCCGTCCGGGACCCGCGCGGTCGATCCCCGCATATAGGGGGGGC...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>14</th>\n",
              "      <td>SCO5090_repair_up_reverse_p</td>\n",
              "      <td>GACCTCACGGACCTCGACCGGCCGCCTCGGCCAGT</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>15</th>\n",
              "      <td>SCO5090_repair_dw_reverse_p</td>\n",
              "      <td>TCCCCGTCCGGGACCCGCGCGGTCGATCCCCGCATATAGGACCGTG...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>16</th>\n",
              "      <td>SCO5091_repair_up_reverse_p</td>\n",
              "      <td>GGCTGCCATCTTCGAACTCCGACGCCGGCCCGG</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>17</th>\n",
              "      <td>SCO5091_repair_dw_reverse_p</td>\n",
              "      <td>TCCCCGTCCGGGACCCGCGCGGTCGATCCCCGCATATAGGAGGCCG...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>18</th>\n",
              "      <td>SCO5092_repair_up_reverse_p</td>\n",
              "      <td>GACGGCCTTCCCGGCCGGCCCTTCGAACTCCCTAGGCGA</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>19</th>\n",
              "      <td>SCO5092_repair_dw_reverse_p</td>\n",
              "      <td>TCCCCGTCCGGGACCCGCGCGGTCGATCCCCGCATATAGGACATCT...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "                           Name  \\\n",
              "0    SCO5087_repair_up_forwar_p   \n",
              "1    SCO5087_repair_dw_forwar_p   \n",
              "2    SCO5089_repair_up_forwar_p   \n",
              "3    SCO5089_repair_dw_forwar_p   \n",
              "4    SCO5090_repair_up_forwar_p   \n",
              "5    SCO5090_repair_dw_forwar_p   \n",
              "6    SCO5091_repair_up_forwar_p   \n",
              "7    SCO5091_repair_dw_forwar_p   \n",
              "8    SCO5092_repair_up_forwar_p   \n",
              "9    SCO5092_repair_dw_forwar_p   \n",
              "10  SCO5087_repair_up_reverse_p   \n",
              "11  SCO5087_repair_dw_reverse_p   \n",
              "12  SCO5089_repair_up_reverse_p   \n",
              "13  SCO5089_repair_dw_reverse_p   \n",
              "14  SCO5090_repair_up_reverse_p   \n",
              "15  SCO5090_repair_dw_reverse_p   \n",
              "16  SCO5091_repair_up_reverse_p   \n",
              "17  SCO5091_repair_dw_reverse_p   \n",
              "18  SCO5092_repair_up_reverse_p   \n",
              "19  SCO5092_repair_dw_reverse_p   \n",
              "\n",
              "                                             Sequence Concentration  \\\n",
              "0   AAGGCCGCTTTTGCGGGATCTCGTCGAAGGCACTAGAAGGGACGCC...          25nm   \n",
              "1                GGCGGTCGAAGGGAGATGGGGCGTCCTCATCACCGG          25nm   \n",
              "2   AAGGCCGCTTTTGCGGGATCTCGTCGAAGGCACTAGAAGGCCAGGA...          25nm   \n",
              "3                   CCCGAGAACTGGAGAAAGCCGCCGGCCGGGAGA          25nm   \n",
              "4   AAGGCCGCTTTTGCGGGATCTCGTCGAAGGCACTAGAAGGGCCACG...          25nm   \n",
              "5                  GGCGCACTGGCCGAGGCGGCCGGTCGAGGTCCGT          25nm   \n",
              "6   AAGGCCGCTTTTGCGGGATCTCGTCGAAGGCACTAGAAGGGACACG...          25nm   \n",
              "7             GCCGCCGCCGGGCCGGCGTCGGAGTTCGAAGATGGCAGC          25nm   \n",
              "8   AAGGCCGCTTTTGCGGGATCTCGTCGAAGGCACTAGAAGGACCGTC...          25nm   \n",
              "9                   CTCGCCTAGGGAGTTCGAAGGGCCGGCCGGGAA          25nm   \n",
              "10               GACTCCGGTGATGAGGACGCCCCATCTCCCTTCGAC          25nm   \n",
              "11  TCCCCGTCCGGGACCCGCGCGGTCGATCCCCGCATATAGGGAGTAG...          25nm   \n",
              "12             CCGGTGTTCTCCCGGCCGGCGGCTTTCTCCAGTTCTCG          25nm   \n",
              "13  TCCCCGTCCGGGACCCGCGCGGTCGATCCCCGCATATAGGGGGGGC...          25nm   \n",
              "14                GACCTCACGGACCTCGACCGGCCGCCTCGGCCAGT          25nm   \n",
              "15  TCCCCGTCCGGGACCCGCGCGGTCGATCCCCGCATATAGGACCGTG...          25nm   \n",
              "16                  GGCTGCCATCTTCGAACTCCGACGCCGGCCCGG          25nm   \n",
              "17  TCCCCGTCCGGGACCCGCGCGGTCGATCCCCGCATATAGGAGGCCG...          25nm   \n",
              "18            GACGGCCTTCCCGGCCGGCCCTTCGAACTCCCTAGGCGA          25nm   \n",
              "19  TCCCCGTCCGGGACCCGCGCGGTCGATCCCCGCATATAGGACATCT...          25nm   \n",
              "\n",
              "   Purification  \n",
              "0           STD  \n",
              "1           STD  \n",
              "2           STD  \n",
              "3           STD  \n",
              "4           STD  \n",
              "5           STD  \n",
              "6           STD  \n",
              "7           STD  \n",
              "8           STD  \n",
              "9           STD  \n",
              "10          STD  \n",
              "11          STD  \n",
              "12          STD  \n",
              "13          STD  \n",
              "14          STD  \n",
              "15          STD  \n",
              "16          STD  \n",
              "17          STD  \n",
              "18          STD  \n",
              "19          STD  "
            ]
          },
          "execution_count": 45,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "idt_df"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 46,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>Name</th>\n",
              "      <th>Sequence</th>\n",
              "      <th>Concentration</th>\n",
              "      <th>Purification</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>SCO5090_loc_546</td>\n",
              "      <td>CGGTTGGTAGGATCGACGGCGCAGCGACCTGTGGCAGGAAGTTTTA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>SCO5090_loc_537</td>\n",
              "      <td>CGGTTGGTAGGATCGACGGCTCATCGAGCGCAGCGACCTGGTTTTA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>SCO5091_loc_48</td>\n",
              "      <td>CGGTTGGTAGGATCGACGGCCTCGTAGGCGTAGACACCCTGTTTTA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>SCO5091_loc_78</td>\n",
              "      <td>CGGTTGGTAGGATCGACGGCGTTGCTGACGCACCAGCCGCGTTTTA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>SCO5092_loc_25</td>\n",
              "      <td>CGGTTGGTAGGATCGACGGCAGCCGACCAGGGAATGCTCCGTTTTA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>5</th>\n",
              "      <td>SCO5092_loc_267</td>\n",
              "      <td>CGGTTGGTAGGATCGACGGCGGACTTGCGCGCGAAGCGCAGTTTTA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>6</th>\n",
              "      <td>SCO5089_loc_61</td>\n",
              "      <td>CGGTTGGTAGGATCGACGGCCGTGGAGTGCGCCGGTGAGAGTTTTA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>7</th>\n",
              "      <td>SCO5087_loc_556</td>\n",
              "      <td>CGGTTGGTAGGATCGACGGCCCGGCACATGTTCGACTACCGTTTTA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>8</th>\n",
              "      <td>SCO5087_loc_656</td>\n",
              "      <td>CGGTTGGTAGGATCGACGGCACCTCGGGCCTGGACTCCGTGTTTTA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>9</th>\n",
              "      <td>SCO5089_loc_65</td>\n",
              "      <td>CGGTTGGTAGGATCGACGGCGAGTGCGCCGGTGAGACGGAGTTTTA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>SCO5087_repair_up_forwar_p</td>\n",
              "      <td>AAGGCCGCTTTTGCGGGATCTCGTCGAAGGCACTAGAAGGGACGCC...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>SCO5087_repair_dw_forwar_p</td>\n",
              "      <td>GGCGGTCGAAGGGAGATGGGGCGTCCTCATCACCGG</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>SCO5089_repair_up_forwar_p</td>\n",
              "      <td>AAGGCCGCTTTTGCGGGATCTCGTCGAAGGCACTAGAAGGCCAGGA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>SCO5089_repair_dw_forwar_p</td>\n",
              "      <td>CCCGAGAACTGGAGAAAGCCGCCGGCCGGGAGA</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>SCO5090_repair_up_forwar_p</td>\n",
              "      <td>AAGGCCGCTTTTGCGGGATCTCGTCGAAGGCACTAGAAGGGCCACG...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>5</th>\n",
              "      <td>SCO5090_repair_dw_forwar_p</td>\n",
              "      <td>GGCGCACTGGCCGAGGCGGCCGGTCGAGGTCCGT</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>6</th>\n",
              "      <td>SCO5091_repair_up_forwar_p</td>\n",
              "      <td>AAGGCCGCTTTTGCGGGATCTCGTCGAAGGCACTAGAAGGGACACG...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>7</th>\n",
              "      <td>SCO5091_repair_dw_forwar_p</td>\n",
              "      <td>GCCGCCGCCGGGCCGGCGTCGGAGTTCGAAGATGGCAGC</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>8</th>\n",
              "      <td>SCO5092_repair_up_forwar_p</td>\n",
              "      <td>AAGGCCGCTTTTGCGGGATCTCGTCGAAGGCACTAGAAGGACCGTC...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>9</th>\n",
              "      <td>SCO5092_repair_dw_forwar_p</td>\n",
              "      <td>CTCGCCTAGGGAGTTCGAAGGGCCGGCCGGGAA</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>10</th>\n",
              "      <td>SCO5087_repair_up_reverse_p</td>\n",
              "      <td>GACTCCGGTGATGAGGACGCCCCATCTCCCTTCGAC</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>11</th>\n",
              "      <td>SCO5087_repair_dw_reverse_p</td>\n",
              "      <td>TCCCCGTCCGGGACCCGCGCGGTCGATCCCCGCATATAGGGAGTAG...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>12</th>\n",
              "      <td>SCO5089_repair_up_reverse_p</td>\n",
              "      <td>CCGGTGTTCTCCCGGCCGGCGGCTTTCTCCAGTTCTCG</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>13</th>\n",
              "      <td>SCO5089_repair_dw_reverse_p</td>\n",
              "      <td>TCCCCGTCCGGGACCCGCGCGGTCGATCCCCGCATATAGGGGGGGC...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>14</th>\n",
              "      <td>SCO5090_repair_up_reverse_p</td>\n",
              "      <td>GACCTCACGGACCTCGACCGGCCGCCTCGGCCAGT</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>15</th>\n",
              "      <td>SCO5090_repair_dw_reverse_p</td>\n",
              "      <td>TCCCCGTCCGGGACCCGCGCGGTCGATCCCCGCATATAGGACCGTG...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>16</th>\n",
              "      <td>SCO5091_repair_up_reverse_p</td>\n",
              "      <td>GGCTGCCATCTTCGAACTCCGACGCCGGCCCGG</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>17</th>\n",
              "      <td>SCO5091_repair_dw_reverse_p</td>\n",
              "      <td>TCCCCGTCCGGGACCCGCGCGGTCGATCCCCGCATATAGGAGGCCG...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>18</th>\n",
              "      <td>SCO5092_repair_up_reverse_p</td>\n",
              "      <td>GACGGCCTTCCCGGCCGGCCCTTCGAACTCCCTAGGCGA</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>19</th>\n",
              "      <td>SCO5092_repair_dw_reverse_p</td>\n",
              "      <td>TCCCCGTCCGGGACCCGCGCGGTCGATCCCCGCATATAGGACATCT...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>SCO5091_fwd_checking_primer</td>\n",
              "      <td>AGCCGCCCGAGGAGCTGG</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>SCO5092_fwd_checking_primer</td>\n",
              "      <td>ACGTCGGCGAGCGACAGG</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>SCO5090_fwd_checking_primer</td>\n",
              "      <td>CTCTACTCCGGCGGCGGC</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>SCO5087_fwd_checking_primer</td>\n",
              "      <td>CTTTCCGCCGGTCGAGGC</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>SCO5089_fwd_checking_primer</td>\n",
              "      <td>CTGGAGGACAGCGCCGCG</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>5</th>\n",
              "      <td>SCO5091_rev_checking_primer</td>\n",
              "      <td>CAGGCACCCGCCGCCGAG</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>6</th>\n",
              "      <td>SCO5092_rev_checking_primer</td>\n",
              "      <td>CACCCCCCGCACCCACCG</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>7</th>\n",
              "      <td>SCO5090_rev_checking_primer</td>\n",
              "      <td>CGACATGACGACGTCCCCCGC</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>8</th>\n",
              "      <td>SCO5087_rev_checking_primer</td>\n",
              "      <td>GCGGATCGTACGGCGGGC</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>9</th>\n",
              "      <td>SCO5089_rev_checking_primer</td>\n",
              "      <td>GTTCCTGCCACAGGTCGCTGC</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "                           Name  \\\n",
              "0               SCO5090_loc_546   \n",
              "1               SCO5090_loc_537   \n",
              "2                SCO5091_loc_48   \n",
              "3                SCO5091_loc_78   \n",
              "4                SCO5092_loc_25   \n",
              "5               SCO5092_loc_267   \n",
              "6                SCO5089_loc_61   \n",
              "7               SCO5087_loc_556   \n",
              "8               SCO5087_loc_656   \n",
              "9                SCO5089_loc_65   \n",
              "0    SCO5087_repair_up_forwar_p   \n",
              "1    SCO5087_repair_dw_forwar_p   \n",
              "2    SCO5089_repair_up_forwar_p   \n",
              "3    SCO5089_repair_dw_forwar_p   \n",
              "4    SCO5090_repair_up_forwar_p   \n",
              "5    SCO5090_repair_dw_forwar_p   \n",
              "6    SCO5091_repair_up_forwar_p   \n",
              "7    SCO5091_repair_dw_forwar_p   \n",
              "8    SCO5092_repair_up_forwar_p   \n",
              "9    SCO5092_repair_dw_forwar_p   \n",
              "10  SCO5087_repair_up_reverse_p   \n",
              "11  SCO5087_repair_dw_reverse_p   \n",
              "12  SCO5089_repair_up_reverse_p   \n",
              "13  SCO5089_repair_dw_reverse_p   \n",
              "14  SCO5090_repair_up_reverse_p   \n",
              "15  SCO5090_repair_dw_reverse_p   \n",
              "16  SCO5091_repair_up_reverse_p   \n",
              "17  SCO5091_repair_dw_reverse_p   \n",
              "18  SCO5092_repair_up_reverse_p   \n",
              "19  SCO5092_repair_dw_reverse_p   \n",
              "0   SCO5091_fwd_checking_primer   \n",
              "1   SCO5092_fwd_checking_primer   \n",
              "2   SCO5090_fwd_checking_primer   \n",
              "3   SCO5087_fwd_checking_primer   \n",
              "4   SCO5089_fwd_checking_primer   \n",
              "5   SCO5091_rev_checking_primer   \n",
              "6   SCO5092_rev_checking_primer   \n",
              "7   SCO5090_rev_checking_primer   \n",
              "8   SCO5087_rev_checking_primer   \n",
              "9   SCO5089_rev_checking_primer   \n",
              "\n",
              "                                             Sequence Concentration  \\\n",
              "0   CGGTTGGTAGGATCGACGGCGCAGCGACCTGTGGCAGGAAGTTTTA...          25nm   \n",
              "1   CGGTTGGTAGGATCGACGGCTCATCGAGCGCAGCGACCTGGTTTTA...          25nm   \n",
              "2   CGGTTGGTAGGATCGACGGCCTCGTAGGCGTAGACACCCTGTTTTA...          25nm   \n",
              "3   CGGTTGGTAGGATCGACGGCGTTGCTGACGCACCAGCCGCGTTTTA...          25nm   \n",
              "4   CGGTTGGTAGGATCGACGGCAGCCGACCAGGGAATGCTCCGTTTTA...          25nm   \n",
              "5   CGGTTGGTAGGATCGACGGCGGACTTGCGCGCGAAGCGCAGTTTTA...          25nm   \n",
              "6   CGGTTGGTAGGATCGACGGCCGTGGAGTGCGCCGGTGAGAGTTTTA...          25nm   \n",
              "7   CGGTTGGTAGGATCGACGGCCCGGCACATGTTCGACTACCGTTTTA...          25nm   \n",
              "8   CGGTTGGTAGGATCGACGGCACCTCGGGCCTGGACTCCGTGTTTTA...          25nm   \n",
              "9   CGGTTGGTAGGATCGACGGCGAGTGCGCCGGTGAGACGGAGTTTTA...          25nm   \n",
              "0   AAGGCCGCTTTTGCGGGATCTCGTCGAAGGCACTAGAAGGGACGCC...          25nm   \n",
              "1                GGCGGTCGAAGGGAGATGGGGCGTCCTCATCACCGG          25nm   \n",
              "2   AAGGCCGCTTTTGCGGGATCTCGTCGAAGGCACTAGAAGGCCAGGA...          25nm   \n",
              "3                   CCCGAGAACTGGAGAAAGCCGCCGGCCGGGAGA          25nm   \n",
              "4   AAGGCCGCTTTTGCGGGATCTCGTCGAAGGCACTAGAAGGGCCACG...          25nm   \n",
              "5                  GGCGCACTGGCCGAGGCGGCCGGTCGAGGTCCGT          25nm   \n",
              "6   AAGGCCGCTTTTGCGGGATCTCGTCGAAGGCACTAGAAGGGACACG...          25nm   \n",
              "7             GCCGCCGCCGGGCCGGCGTCGGAGTTCGAAGATGGCAGC          25nm   \n",
              "8   AAGGCCGCTTTTGCGGGATCTCGTCGAAGGCACTAGAAGGACCGTC...          25nm   \n",
              "9                   CTCGCCTAGGGAGTTCGAAGGGCCGGCCGGGAA          25nm   \n",
              "10               GACTCCGGTGATGAGGACGCCCCATCTCCCTTCGAC          25nm   \n",
              "11  TCCCCGTCCGGGACCCGCGCGGTCGATCCCCGCATATAGGGAGTAG...          25nm   \n",
              "12             CCGGTGTTCTCCCGGCCGGCGGCTTTCTCCAGTTCTCG          25nm   \n",
              "13  TCCCCGTCCGGGACCCGCGCGGTCGATCCCCGCATATAGGGGGGGC...          25nm   \n",
              "14                GACCTCACGGACCTCGACCGGCCGCCTCGGCCAGT          25nm   \n",
              "15  TCCCCGTCCGGGACCCGCGCGGTCGATCCCCGCATATAGGACCGTG...          25nm   \n",
              "16                  GGCTGCCATCTTCGAACTCCGACGCCGGCCCGG          25nm   \n",
              "17  TCCCCGTCCGGGACCCGCGCGGTCGATCCCCGCATATAGGAGGCCG...          25nm   \n",
              "18            GACGGCCTTCCCGGCCGGCCCTTCGAACTCCCTAGGCGA          25nm   \n",
              "19  TCCCCGTCCGGGACCCGCGCGGTCGATCCCCGCATATAGGACATCT...          25nm   \n",
              "0                                  AGCCGCCCGAGGAGCTGG          25nm   \n",
              "1                                  ACGTCGGCGAGCGACAGG          25nm   \n",
              "2                                  CTCTACTCCGGCGGCGGC          25nm   \n",
              "3                                  CTTTCCGCCGGTCGAGGC          25nm   \n",
              "4                                  CTGGAGGACAGCGCCGCG          25nm   \n",
              "5                                  CAGGCACCCGCCGCCGAG          25nm   \n",
              "6                                  CACCCCCCGCACCCACCG          25nm   \n",
              "7                               CGACATGACGACGTCCCCCGC          25nm   \n",
              "8                                  GCGGATCGTACGGCGGGC          25nm   \n",
              "9                               GTTCCTGCCACAGGTCGCTGC          25nm   \n",
              "\n",
              "   Purification  \n",
              "0           STD  \n",
              "1           STD  \n",
              "2           STD  \n",
              "3           STD  \n",
              "4           STD  \n",
              "5           STD  \n",
              "6           STD  \n",
              "7           STD  \n",
              "8           STD  \n",
              "9           STD  \n",
              "0           STD  \n",
              "1           STD  \n",
              "2           STD  \n",
              "3           STD  \n",
              "4           STD  \n",
              "5           STD  \n",
              "6           STD  \n",
              "7           STD  \n",
              "8           STD  \n",
              "9           STD  \n",
              "10          STD  \n",
              "11          STD  \n",
              "12          STD  \n",
              "13          STD  \n",
              "14          STD  \n",
              "15          STD  \n",
              "16          STD  \n",
              "17          STD  \n",
              "18          STD  \n",
              "19          STD  \n",
              "0           STD  \n",
              "1           STD  \n",
              "2           STD  \n",
              "3           STD  \n",
              "4           STD  \n",
              "5           STD  \n",
              "6           STD  \n",
              "7           STD  \n",
              "8           STD  \n",
              "9           STD  "
            ]
          },
          "execution_count": 46,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "if in_frame_deletion: \n",
        "    full_idt = pd.concat([idt_primers, idt_df, checking_primers_df_idt])\n",
        "else: \n",
        "    full_idt = pd.concat([idt_primers, checking_primers_df_idt])\n",
        "\n",
        "full_idt"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "## Folder with all the generated I/O"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 47,
      "metadata": {},
      "outputs": [],
      "source": [
        "input_files = [\n",
        "    {\"name\": \"input_genome.gb\", \"content\": genome},\n",
        "    {\"name\": \"input_plasmid.gb\", \"content\": clean_plasmid}\n",
        "]\n",
        "\n",
        "if in_frame_deletion: \n",
        "    output_files = [\n",
        "        {\"name\": \"cBEST_w_sgRNAs.gb\", \"content\": assembled_contigs}, # LIST OF Dseqrecords\n",
        "        {\"name\": \"primer_df.csv\", \"content\": primer_df},\n",
        "        {\"name\": \"full_idt.csv\", \"content\": full_idt},\n",
        "        {\"name\": \"sgrna_df.csv\", \"content\": sgrna_df},\n",
        "        {\"name\": \"filtered_df.csv\", \"content\": filtered_df},\n",
        "        {\"name\": \"plasmid_metadata_df.csv\", \"content\": plasmid_metadata_df},\n",
        "\n",
        "    ]\n",
        "else: \n",
        "    output_files = [\n",
        "        {\"name\": \"cBEST_w_sgRNAs.gb\", \"content\": sgRNA_vectors}, # LIST OF Dseqrecords\n",
        "        {\"name\": \"full_idt.csv\", \"content\": full_idt},\n",
        "        {\"name\": \"sgrna_df.csv\", \"content\": sgrna_df},\n",
        "        {\"name\": \"filtered_df.csv\", \"content\": filtered_df},\n",
        "        {\"name\": \"plasmid_metadata_df.csv\", \"content\": plasmid_metadata_df},\n",
        "        {\"name\": \"workflow_order_df.csv\", \"content\": workflow_df},\n",
        "\n",
        "        workflow_df\n",
        "\n",
        "    ]\n",
        "\n",
        "input_values = {\n",
        "    \"genes_to_knockout\": genes_to_KO,\n",
        "\n",
        "\n",
        "    \"filtering_metrics\": {\n",
        "        \"gc_upper\": gc_upper,\n",
        "        \"gc_lower\": gc_lower,\n",
        "        \"off_target_seed\": off_target_seed,\n",
        "        \"off_target_upper\": off_target_upper,\n",
        "        \"cas_type\": cas_type,\n",
        "        \"number_of_sgRNAs_per_group\": number_of_sgRNAs_per_group,\n",
        "\n",
        "    },  \n",
        "    \"polymerase_settings\": {\n",
        "        \"chosen_polymerase\": chosen_polymerase,\n",
        "        \"melting_temperature\": melting_temperature,\n",
        "        \"primer_concentration\": primer_concentration,\n",
        "        \"primer_number_increment\": primer_number_increment,\n",
        "        \"flanking_region\": flanking_region\n",
        "    },\n",
        "    \"overlapping_sequences\": {\n",
        "        \"up_homology\": str(up_homology),\n",
        "        \"dw_homology\": str(dw_homology)\n",
        "    }\n",
        "}\n",
        "\n",
        "\n",
        "# Paths to Markdown files\n",
        "markdown_file_paths = [\n",
        "    \"../../protocols/conjugation_protcol.md\",\n",
        "    \"../../protocols/single_target_crispr_plasmid_protcol.md\"\n",
        "\n",
        "]\n",
        "\n",
        "\n",
        "timestamp = datetime.utcnow().isoformat()\n",
        "\n",
        "# Create project directory structure\n",
        "project_directory = ProjectDirectory(\n",
        "    project_name=f\"CRISPR_cas9_plasmid_workflow_{timestamp}\",\n",
        "    input_files=input_files,\n",
        "    output_files=output_files,\n",
        "    input_values=input_values,\n",
        "    markdown_file_paths=markdown_file_paths\n",
        ")\n",
        "\n",
        "# DO You want to save the folder? \n",
        "save_zip_folder = False \n",
        "\n",
        "if save_zip_folder: \n",
        "    # Generate the project directory structure and get the zip content\n",
        "    zip_content = project_directory.create_directory_structure(create_directories=False)\n",
        "\n",
        "    # Save the zip file to disk (optional)\n",
        "    with open(\"project_structure.zip\", \"wb\") as f:\n",
        "        f.write(zip_content)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {},
      "outputs": [],
      "source": []
    }
  ],
  "metadata": {
    "kernelspec": {
      "display_name": "streptocad",
      "language": "python",
      "name": "python3"
    },
    "language_info": {
      "codemirror_mode": {
        "name": "ipython",
        "version": 3
      },
      "file_extension": ".py",
      "mimetype": "text/x-python",
      "name": "python",
      "nbconvert_exporter": "python",
      "pygments_lexer": "ipython3",
      "version": "3.11.11"
    }
  },
  "nbformat": 4,
  "nbformat_minor": 2
}
