{
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "# W5-CRISPR-Cas9-inframe_deletion"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "In Workflow 5, CRISPR-Cas9 combines ssDNA bridging and Gibson cloning to create in-frame deletions (Figure 6A). This ensures that the downstream genes are not disrupted, which is crucial for functional studies of gene knockouts. This workflow can be used to study gene function by knocking out specific genes in streptomycetes and observing the resulting phenotypic changes. It can also be used to create markerless deletions, which are important for constructing clean genetic backgrounds for further genetic manipulations. To get started, download the pCRISPR-Cas9 file and a genome file (e.g., S. coelicolor A3), then select the genes to delete. StreptoAIM will generate the necessary primers and plasmid constructs (S5).\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 1,
      "metadata": {},
      "outputs": [],
      "source": [
        "import sys\n",
        "import os\n",
        "from Bio.Restriction import StuI\n",
        "from pydna.dseqrecord import Dseqrecord\n",
        "import pandas as pd\n",
        "from datetime import datetime\n",
        "\n",
        "\n",
        "# Ensure the src directory is in the Python path\n",
        "project_root = os.path.abspath(os.path.join(os.getcwd(), '../../'))\n",
        "if project_root not in sys.path:\n",
        "    sys.path.insert(0, project_root)\n",
        "\n",
        "from streptocad.sequence_loading.sequence_loading import (\n",
        "    load_and_process_gene_sequences,\n",
        "    load_and_process_genome_sequences,\n",
        "    load_and_process_plasmid, \n",
        "    check_and_convert_input,\n",
        "    annotate_dseqrecord,\n",
        "    process_specified_gene_sequences_from_record)\n",
        "\n",
        "from streptocad.utils import polymerase_dict, create_primer_df_from_dict,ProjectDirectory,extract_metadata_to_dataframe\n",
        "from streptocad.primers.primer_generation import create_idt_order_dataframe\n",
        "from streptocad.cloning.ssDNA_bridging import assemble_plasmids_by_ssDNA_bridging,make_ssDNA_oligos\n",
        "from streptocad.crispr.guideRNAcas3_9_12 import extract_sgRNAs, SgRNAargs\n",
        "from streptocad.cloning.gibson_cloning import (\n",
        "    find_up_dw_repair_templates,\n",
        "    assemble_multiple_plasmids_with_repair_templates_for_deletion,\n",
        "    update_primer_names\n",
        ")\n",
        "\n",
        "from streptocad.cloning.plasmid_processing import check_plasmid_restriction_sites, determine_workflow_order_for_plasmids\n",
        "from streptocad.primers.primer_generation import checking_primers, primers_to_IDT, find_best_check_primers_from_genome\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "## INPUT"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 2,
      "metadata": {},
      "outputs": [],
      "source": [
        "# Inputs\n",
        "# 1 Add genome of choice (genbank, fasta)\n",
        "path_to_genome = '../../data/genomes/Streptomyces_coelicolor_A3_chromosome.gb'\n",
        "genome = load_and_process_genome_sequences(path_to_genome)[0]\n",
        "\n",
        "# 2 Add plasmid \n",
        "path_to_plasmid = '../../data/plasmids/pCRISPR–Cas9_plasmid_addgene.gbk'\n",
        "clean_plasmid = load_and_process_plasmid(path_to_plasmid)\n",
        "\n",
        "# 3 Choose genes to knock out (list)\n",
        "#genes_to_KO = ['SCO5087', 'SCO5089', 'SCO5090','SCO5091', 'SCO5092' ]\n",
        "genes_to_KO = ['80000-100000', '4000-7000', '9000-14000','15000-20000']\n",
        "\n",
        "\n",
        "#### Advanced settings ####\n",
        "# 4 Filtering metrics for sgRNAs\n",
        "gc_upper = 0.72\n",
        "gc_lower = 0.2\n",
        "off_target_seed = 13\n",
        "off_target_upper = 10\n",
        "cas_type='cas9'\n",
        "number_of_sgRNAs_per_group = 2\n",
        "\n",
        "# 5 Choose polymerase and target melting temperature\n",
        "chosen_polymerase = polymerase_dict['Phusion High-Fidelity DNA Polymerase (GC Buffer)']\n",
        "melting_temperature = 65\n",
        "primer_concentration = 0.4 \n",
        "primer_number_increment = 23\n",
        "flanking_region = 500\n",
        "\n",
        "# 6 Choose overlapping sequences for our plasmid we can use the following\n",
        "#As per the article **\"CRISPR–Cas9, CRISPRi and CRISPR-BEST-mediated genetic manipulation in streptomycetes\"** we need the following oligoes: \n",
        "#CGGTTGGTAGGATCGACGGC **-N20-** GTTTTAGAGCTAGAAATAGC\n",
        "up_homology = Dseqrecord('CGGTTGGTAGGATCGACGGC')\n",
        "dw_homology = Dseqrecord('GTTTTAGAGCTAGAAATAGC')\n",
        "\n",
        "# 7 extra settings\n",
        "#### Advanced settings ####\n",
        "repair_length = 1000\n",
        "overlap_length = 40 "
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 3,
      "metadata": {},
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "pCRISPR-Cas9\n"
          ]
        }
      ],
      "source": [
        "print(clean_plasmid.id)"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "# Computation"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 4,
      "metadata": {},
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "True\n",
            "this is the target dict [{'region1': [80000, 100000]}, {'region2': [4000, 7000]}, {'region3': [9000, 14000]}, {'region4': [15000, 20000]}]\n",
            "region1 [80000, 100000]\n",
            "region2 [4000, 7000]\n",
            "region3 [9000, 14000]\n",
            "region4 [15000, 20000]\n"
          ]
        },
        {
          "data": {
            "text/plain": [
              "25828"
            ]
          },
          "execution_count": 4,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "target_dict, genes_to_KO, annotation_input = check_and_convert_input(genes_to_KO)\n",
        "\n",
        "print(annotation_input)\n",
        "if annotation_input == True:\n",
        "    genome = annotate_dseqrecord(genome, target_dict)\n",
        "\n",
        "\n",
        "len(genome.features)\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 5,
      "metadata": {},
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "sgRNA generated were outside the designated border in region1. To incorporate this extent borders. Skipping to next locus tag.\n",
            "sgRNA generated were outside the designated border in region1. To incorporate this extent borders. Skipping to next locus tag.\n",
            "sgRNA generated were outside the designated border in region2. To incorporate this extent borders. Skipping to next locus tag.\n",
            "sgRNA generated were outside the designated border in region2. To incorporate this extent borders. Skipping to next locus tag.\n",
            "sgRNA generated were outside the designated border in region2. To incorporate this extent borders. Skipping to next locus tag.\n",
            "sgRNA generated were outside the designated border in region2. To incorporate this extent borders. Skipping to next locus tag.\n",
            "sgRNA generated were outside the designated border in region2. To incorporate this extent borders. Skipping to next locus tag.\n",
            "sgRNA generated were outside the designated border in region2. To incorporate this extent borders. Skipping to next locus tag.\n",
            "sgRNA generated were outside the designated border in region3. To incorporate this extent borders. Skipping to next locus tag.\n",
            "sgRNA generated were outside the designated border in region3. To incorporate this extent borders. Skipping to next locus tag.\n",
            "sgRNA generated were outside the designated border in region3. To incorporate this extent borders. Skipping to next locus tag.\n",
            "sgRNA generated were outside the designated border in region3. To incorporate this extent borders. Skipping to next locus tag.\n",
            "sgRNA generated were outside the designated border in region3. To incorporate this extent borders. Skipping to next locus tag.\n",
            "sgRNA generated were outside the designated border in region4. To incorporate this extent borders. Skipping to next locus tag.\n",
            "sgRNA generated were outside the designated border in region4. To incorporate this extent borders. Skipping to next locus tag.\n",
            "sgRNA generated were outside the designated border in region4. To incorporate this extent borders. Skipping to next locus tag.\n",
            "sgRNA generated were outside the designated border in region4. To incorporate this extent borders. Skipping to next locus tag.\n",
            "sgRNA generated were outside the designated border in region4. To incorporate this extent borders. Skipping to next locus tag.\n",
            "sgRNA generated were outside the designated border in region4. To incorporate this extent borders. Skipping to next locus tag.\n"
          ]
        },
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>strain_name</th>\n",
              "      <th>locus_tag</th>\n",
              "      <th>gene_loc</th>\n",
              "      <th>gene_strand</th>\n",
              "      <th>sgrna_strand</th>\n",
              "      <th>sgrna_loc</th>\n",
              "      <th>gc</th>\n",
              "      <th>pam</th>\n",
              "      <th>sgrna</th>\n",
              "      <th>sgrna_seed_sequence</th>\n",
              "      <th>off_target_count</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>3517</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>region1</td>\n",
              "      <td>80001</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>6574</td>\n",
              "      <td>0.60</td>\n",
              "      <td>TGG</td>\n",
              "      <td>GCCACATTCACTACGACCCC</td>\n",
              "      <td>TCACTACGACCCC</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2538</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>region1</td>\n",
              "      <td>80001</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>16046</td>\n",
              "      <td>0.65</td>\n",
              "      <td>GGG</td>\n",
              "      <td>CCCTGATCCGAGTGTGTGCC</td>\n",
              "      <td>CCGAGTGTGTGCC</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2539</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>region1</td>\n",
              "      <td>80001</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>16045</td>\n",
              "      <td>0.60</td>\n",
              "      <td>CGG</td>\n",
              "      <td>ACCCTGATCCGAGTGTGTGC</td>\n",
              "      <td>TCCGAGTGTGTGC</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2540</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>region1</td>\n",
              "      <td>80001</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>16018</td>\n",
              "      <td>0.55</td>\n",
              "      <td>GGG</td>\n",
              "      <td>ACGACAATCCGGTGATCAGC</td>\n",
              "      <td>TCCGGTGATCAGC</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2541</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>region1</td>\n",
              "      <td>80001</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>16017</td>\n",
              "      <td>0.50</td>\n",
              "      <td>CGG</td>\n",
              "      <td>TACGACAATCCGGTGATCAG</td>\n",
              "      <td>ATCCGGTGATCAG</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>...</th>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>191</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>region1</td>\n",
              "      <td>80001</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>1988</td>\n",
              "      <td>0.65</td>\n",
              "      <td>CGG</td>\n",
              "      <td>ATCCTCGCCGTGACCGTCAC</td>\n",
              "      <td>CCGTGACCGTCAC</td>\n",
              "      <td>7</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>172</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>region1</td>\n",
              "      <td>80001</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>1765</td>\n",
              "      <td>0.65</td>\n",
              "      <td>TGG</td>\n",
              "      <td>TGGTAGAGCGCAGCATCGGC</td>\n",
              "      <td>GCGCAGCATCGGC</td>\n",
              "      <td>8</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3954</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>region1</td>\n",
              "      <td>80001</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>2235</td>\n",
              "      <td>0.70</td>\n",
              "      <td>GGG</td>\n",
              "      <td>AGTAGACGGTGCCGTGGCCG</td>\n",
              "      <td>GGTGCCGTGGCCG</td>\n",
              "      <td>8</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>179</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>region1</td>\n",
              "      <td>80001</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>1842</td>\n",
              "      <td>0.70</td>\n",
              "      <td>AGG</td>\n",
              "      <td>CACCCTGGGCATCGACGTCG</td>\n",
              "      <td>GGCATCGACGTCG</td>\n",
              "      <td>8</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3996</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>region1</td>\n",
              "      <td>80001</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>1929</td>\n",
              "      <td>0.65</td>\n",
              "      <td>AGG</td>\n",
              "      <td>TCCTCTTGGCCTGGTCGAGG</td>\n",
              "      <td>GGCCTGGTCGAGG</td>\n",
              "      <td>9</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>4297 rows × 11 columns</p>\n",
              "</div>"
            ],
            "text/plain": [
              "      strain_name locus_tag  gene_loc  gene_strand  sgrna_strand  sgrna_loc  \\\n",
              "3517  NC_003888.3   region1     80001            1             1       6574   \n",
              "2538  NC_003888.3   region1     80001            1             1      16046   \n",
              "2539  NC_003888.3   region1     80001            1             1      16045   \n",
              "2540  NC_003888.3   region1     80001            1             1      16018   \n",
              "2541  NC_003888.3   region1     80001            1             1      16017   \n",
              "...           ...       ...       ...          ...           ...        ...   \n",
              "191   NC_003888.3   region1     80001            1            -1       1988   \n",
              "172   NC_003888.3   region1     80001            1            -1       1765   \n",
              "3954  NC_003888.3   region1     80001            1             1       2235   \n",
              "179   NC_003888.3   region1     80001            1            -1       1842   \n",
              "3996  NC_003888.3   region1     80001            1             1       1929   \n",
              "\n",
              "        gc  pam                 sgrna sgrna_seed_sequence  off_target_count  \n",
              "3517  0.60  TGG  GCCACATTCACTACGACCCC       TCACTACGACCCC                 0  \n",
              "2538  0.65  GGG  CCCTGATCCGAGTGTGTGCC       CCGAGTGTGTGCC                 0  \n",
              "2539  0.60  CGG  ACCCTGATCCGAGTGTGTGC       TCCGAGTGTGTGC                 0  \n",
              "2540  0.55  GGG  ACGACAATCCGGTGATCAGC       TCCGGTGATCAGC                 0  \n",
              "2541  0.50  CGG  TACGACAATCCGGTGATCAG       ATCCGGTGATCAG                 0  \n",
              "...    ...  ...                   ...                 ...               ...  \n",
              "191   0.65  CGG  ATCCTCGCCGTGACCGTCAC       CCGTGACCGTCAC                 7  \n",
              "172   0.65  TGG  TGGTAGAGCGCAGCATCGGC       GCGCAGCATCGGC                 8  \n",
              "3954  0.70  GGG  AGTAGACGGTGCCGTGGCCG       GGTGCCGTGGCCG                 8  \n",
              "179   0.70  AGG  CACCCTGGGCATCGACGTCG       GGCATCGACGTCG                 8  \n",
              "3996  0.65  AGG  TCCTCTTGGCCTGGTCGAGG       GGCCTGGTCGAGG                 9  \n",
              "\n",
              "[4297 rows x 11 columns]"
            ]
          },
          "execution_count": 5,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "\n",
        "# Initialize SgRNAargs with desired parameters\n",
        "args = SgRNAargs(genome, \n",
        "                genes_to_KO,\n",
        "                step=['find', 'filter'],\n",
        "                gc_upper = gc_upper,\n",
        "                gc_lower = gc_lower,\n",
        "                off_target_seed = off_target_seed,\n",
        "                off_target_upper = off_target_upper,\n",
        "                cas_type='cas9'\n",
        "                )\n",
        "\n",
        "sgrna_df = extract_sgRNAs(args)\n",
        "sgrna_df"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 6,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>strain_name</th>\n",
              "      <th>locus_tag</th>\n",
              "      <th>gene_loc</th>\n",
              "      <th>gene_strand</th>\n",
              "      <th>sgrna_strand</th>\n",
              "      <th>sgrna_loc</th>\n",
              "      <th>gc</th>\n",
              "      <th>pam</th>\n",
              "      <th>sgrna</th>\n",
              "      <th>sgrna_seed_sequence</th>\n",
              "      <th>off_target_count</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>3517</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>region1</td>\n",
              "      <td>80001</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>6574</td>\n",
              "      <td>0.60</td>\n",
              "      <td>TGG</td>\n",
              "      <td>GCCACATTCACTACGACCCC</td>\n",
              "      <td>TCACTACGACCCC</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2538</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>region1</td>\n",
              "      <td>80001</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>16046</td>\n",
              "      <td>0.65</td>\n",
              "      <td>GGG</td>\n",
              "      <td>CCCTGATCCGAGTGTGTGCC</td>\n",
              "      <td>CCGAGTGTGTGCC</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>5691</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>region3</td>\n",
              "      <td>9001</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>3477</td>\n",
              "      <td>0.70</td>\n",
              "      <td>GGG</td>\n",
              "      <td>GCCGGTCGGCCATATCCCCT</td>\n",
              "      <td>GGCCATATCCCCT</td>\n",
              "      <td>1</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>7003</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>region4</td>\n",
              "      <td>15001</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>265</td>\n",
              "      <td>0.65</td>\n",
              "      <td>AGG</td>\n",
              "      <td>ACGGCGTACAGGTCGCGTAC</td>\n",
              "      <td>ACAGGTCGCGTAC</td>\n",
              "      <td>1</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4772</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>region2</td>\n",
              "      <td>4001</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>346</td>\n",
              "      <td>0.70</td>\n",
              "      <td>TGG</td>\n",
              "      <td>GTCCCTCGACTCACTGGCGC</td>\n",
              "      <td>GACTCACTGGCGC</td>\n",
              "      <td>1</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>5251</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>region3</td>\n",
              "      <td>9001</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>2460</td>\n",
              "      <td>0.40</td>\n",
              "      <td>TGG</td>\n",
              "      <td>TAGTTGTGTGGGATTGATCT</td>\n",
              "      <td>GTGGGATTGATCT</td>\n",
              "      <td>1</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4773</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>region2</td>\n",
              "      <td>4001</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>340</td>\n",
              "      <td>0.65</td>\n",
              "      <td>TGG</td>\n",
              "      <td>GTGCTGGTCCCTCGACTCAC</td>\n",
              "      <td>TCCCTCGACTCAC</td>\n",
              "      <td>1</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>7004</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>region4</td>\n",
              "      <td>15001</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>254</td>\n",
              "      <td>0.65</td>\n",
              "      <td>AGG</td>\n",
              "      <td>TCACGTCACCGACGGCGTAC</td>\n",
              "      <td>ACCGACGGCGTAC</td>\n",
              "      <td>1</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "      strain_name locus_tag  gene_loc  gene_strand  sgrna_strand  sgrna_loc  \\\n",
              "3517  NC_003888.3   region1     80001            1             1       6574   \n",
              "2538  NC_003888.3   region1     80001            1             1      16046   \n",
              "5691  NC_003888.3   region3      9001            1             1       3477   \n",
              "7003  NC_003888.3   region4     15001            1             1        265   \n",
              "4772  NC_003888.3   region2      4001            1             1        346   \n",
              "5251  NC_003888.3   region3      9001            1            -1       2460   \n",
              "4773  NC_003888.3   region2      4001            1             1        340   \n",
              "7004  NC_003888.3   region4     15001            1             1        254   \n",
              "\n",
              "        gc  pam                 sgrna sgrna_seed_sequence  off_target_count  \n",
              "3517  0.60  TGG  GCCACATTCACTACGACCCC       TCACTACGACCCC                 0  \n",
              "2538  0.65  GGG  CCCTGATCCGAGTGTGTGCC       CCGAGTGTGTGCC                 0  \n",
              "5691  0.70  GGG  GCCGGTCGGCCATATCCCCT       GGCCATATCCCCT                 1  \n",
              "7003  0.65  AGG  ACGGCGTACAGGTCGCGTAC       ACAGGTCGCGTAC                 1  \n",
              "4772  0.70  TGG  GTCCCTCGACTCACTGGCGC       GACTCACTGGCGC                 1  \n",
              "5251  0.40  TGG  TAGTTGTGTGGGATTGATCT       GTGGGATTGATCT                 1  \n",
              "4773  0.65  TGG  GTGCTGGTCCCTCGACTCAC       TCCCTCGACTCAC                 1  \n",
              "7004  0.65  AGG  TCACGTCACCGACGGCGTAC       ACCGACGGCGTAC                 1  "
            ]
          },
          "execution_count": 6,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "# Filter the DataFrame to retain only up to 5 sgRNA sequences per locus_tag\n",
        "filtered_df = sgrna_df.groupby('locus_tag').head(number_of_sgRNAs_per_group)\n",
        "filtered_df"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "## Output"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 7,
      "metadata": {},
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "region1_loc_6574\n"
          ]
        },
        {
          "data": {
            "text/plain": [
              "[Contig(o11279),\n",
              " Contig(o11279),\n",
              " Contig(o11279),\n",
              " Contig(o11279),\n",
              " Contig(o11279),\n",
              " Contig(o11279),\n",
              " Contig(o11279),\n",
              " Contig(o11279)]"
            ]
          },
          "execution_count": 7,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "# MAke oligoes\n",
        "list_of_ssDNAs = make_ssDNA_oligos(filtered_df, upstream_ovh = up_homology,\n",
        "                      downstream_ovh=dw_homology)\n",
        "print(list_of_ssDNAs[0].name)\n",
        "\n",
        "# cut plasmid\n",
        "from Bio.Restriction import NcoI\n",
        "linearized_plasmid = sorted(clean_plasmid.cut(NcoI), key=lambda x: len(x), reverse=True)[0]\n",
        "#print(linearized_plasmid)\n",
        "\n",
        "sgRNA_vectors = assemble_plasmids_by_ssDNA_bridging(list_of_ssDNAs,linearized_plasmid)\n",
        "sgRNA_vectors"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 8,
      "metadata": {},
      "outputs": [],
      "source": [
        "\n",
        "# Constructing a meaningful name, ID, and description for the assembled plasmid using user input\n",
        "targeting_info = []\n",
        "for index, row in filtered_df.iterrows():\n",
        "    formatted_str = f\"pCas9_{row['locus_tag']}({row['sgrna_loc']})\"\n",
        "    targeting_info.append(formatted_str)\n",
        "\n",
        "for i in range(len(sgRNA_vectors)):\n",
        "    sgRNA_vectors[i].name = f'{targeting_info[i]}'\n",
        "    sgRNA_vectors[i].id = sgRNA_vectors[i].name  \n",
        "    sgRNA_vectors[i].description = f'CRISPR-Cas9 targeting {\", \".join(genes_to_KO)} for single gene knockout, assembled using StreptoCAD.'\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 9,
      "metadata": {},
      "outputs": [],
      "source": [
        "print_plasmids = False\n",
        "\n",
        "if print_plasmids: \n",
        "    for vector in sgRNA_vectors: \n",
        "        vector.write(f\"../../data/plasmids/sgRNA_plasmids_pCRISPR-Cas9/{vector.id}.gb\")"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "### IDT primers"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 10,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>Name</th>\n",
              "      <th>Sequence</th>\n",
              "      <th>Concentration</th>\n",
              "      <th>Purification</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>region1_loc_6574</td>\n",
              "      <td>CGGTTGGTAGGATCGACGGCGCCACATTCACTACGACCCCGTTTTA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>region1_loc_16046</td>\n",
              "      <td>CGGTTGGTAGGATCGACGGCCCCTGATCCGAGTGTGTGCCGTTTTA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>region3_loc_3477</td>\n",
              "      <td>CGGTTGGTAGGATCGACGGCGCCGGTCGGCCATATCCCCTGTTTTA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>region4_loc_265</td>\n",
              "      <td>CGGTTGGTAGGATCGACGGCACGGCGTACAGGTCGCGTACGTTTTA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>region2_loc_346</td>\n",
              "      <td>CGGTTGGTAGGATCGACGGCGTCCCTCGACTCACTGGCGCGTTTTA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>5</th>\n",
              "      <td>region3_loc_2460</td>\n",
              "      <td>CGGTTGGTAGGATCGACGGCTAGTTGTGTGGGATTGATCTGTTTTA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>6</th>\n",
              "      <td>region2_loc_340</td>\n",
              "      <td>CGGTTGGTAGGATCGACGGCGTGCTGGTCCCTCGACTCACGTTTTA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>7</th>\n",
              "      <td>region4_loc_254</td>\n",
              "      <td>CGGTTGGTAGGATCGACGGCTCACGTCACCGACGGCGTACGTTTTA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "                Name                                           Sequence  \\\n",
              "0   region1_loc_6574  CGGTTGGTAGGATCGACGGCGCCACATTCACTACGACCCCGTTTTA...   \n",
              "1  region1_loc_16046  CGGTTGGTAGGATCGACGGCCCCTGATCCGAGTGTGTGCCGTTTTA...   \n",
              "2   region3_loc_3477  CGGTTGGTAGGATCGACGGCGCCGGTCGGCCATATCCCCTGTTTTA...   \n",
              "3    region4_loc_265  CGGTTGGTAGGATCGACGGCACGGCGTACAGGTCGCGTACGTTTTA...   \n",
              "4    region2_loc_346  CGGTTGGTAGGATCGACGGCGTCCCTCGACTCACTGGCGCGTTTTA...   \n",
              "5   region3_loc_2460  CGGTTGGTAGGATCGACGGCTAGTTGTGTGGGATTGATCTGTTTTA...   \n",
              "6    region2_loc_340  CGGTTGGTAGGATCGACGGCGTGCTGGTCCCTCGACTCACGTTTTA...   \n",
              "7    region4_loc_254  CGGTTGGTAGGATCGACGGCTCACGTCACCGACGGCGTACGTTTTA...   \n",
              "\n",
              "  Concentration Purification  \n",
              "0          25nm          STD  \n",
              "1          25nm          STD  \n",
              "2          25nm          STD  \n",
              "3          25nm          STD  \n",
              "4          25nm          STD  \n",
              "5          25nm          STD  \n",
              "6          25nm          STD  \n",
              "7          25nm          STD  "
            ]
          },
          "execution_count": 10,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "idt_primers=primers_to_IDT(list_of_ssDNAs)\n",
        "idt_primers"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "# In frame deletion"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 11,
      "metadata": {},
      "outputs": [],
      "source": [
        "in_frame_deletion = True"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 12,
      "metadata": {},
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "[{'gene_name': 'region1', 'name': 'pCas9_region1(6574)', 'contig': Dseqrecord(o13279), 'up_forwar_p_name': 'primer_1', 'up_reverse_p_name': 'primer_2', 'up_forwar_p': primer_1 59-mer:5'-AAGGCCGCTTTTGCG..AGC-3', 'up_reverse_p': primer_2 36-mer:5'-GCCCGGCCGACTCGT..ACG-3', 'up_forwar_primer_str': 'AAGGCCGCTTTTGCGGGATCTCGTCGAAGGCACTAGAAGGGATGACGACTGTGACGAGC', 'up_reverse_primer_str': 'GCCCGGCCGACTCGTTGGGCATCTGCCGGGAGAACG', 'up_forwar_p_anneal': Seq('GATGACGACTGTGACGAGC'), 'up_reverse_p_anneal': Seq('ATCTGCCGGGAGAACG'), 'tm_up_forwar_p': 59, 'tm_up_reverse_p': 57, 'ta_up': 61, 'dw_forwar_p_name': 'primer_3', 'dw_reverse_p_name': 'primer_4', 'dw_forwar_p': primer_3 34-mer:5'-ATGCCGTTCTCCCGG..CGG-3', 'dw_reverse_p': primer_4 53-mer:5'-TCCCCGTCCGGGACC..CCG-3', 'dw_forwar_primer_str': 'ATGCCGTTCTCCCGGCAGATGCCCAACGAGTCGG', 'dw_reverse_primer_str': 'TCCCCGTCCGGGACCCGCGCGGTCGATCCCCGCATATAGGCGGCGAGAGGCCG', 'dw_forwar_p_anneal': Seq('GCCCAACGAGTCGG'), 'dw_reverse_p_anneal': Seq('CGGCGAGAGGCCG'), 'tm_dw_forwar_p': 55, 'tm_dw_reverse_p': 57, 'ta_dw': 58}, {'gene_name': 'region1', 'name': 'pCas9_region1(16046)', 'contig': Dseqrecord(o13279), 'up_forwar_p_name': 'primer_1', 'up_reverse_p_name': 'primer_2', 'up_forwar_p': f1000 59-mer:5'-AAGGCCGCTTTTGCG..AGC-3', 'up_reverse_p': r1000 36-mer:5'-GCCCGGCCGACTCGT..ACG-3', 'up_forwar_primer_str': 'AAGGCCGCTTTTGCGGGATCTCGTCGAAGGCACTAGAAGGGATGACGACTGTGACGAGC', 'up_reverse_primer_str': 'GCCCGGCCGACTCGTTGGGCATCTGCCGGGAGAACG', 'up_forwar_p_anneal': Seq('GATGACGACTGTGACGAGC'), 'up_reverse_p_anneal': Seq('ATCTGCCGGGAGAACG'), 'tm_up_forwar_p': 59, 'tm_up_reverse_p': 57, 'ta_up': 61, 'dw_forwar_p_name': 'primer_3', 'dw_reverse_p_name': 'primer_4', 'dw_forwar_p': f1000 34-mer:5'-ATGCCGTTCTCCCGG..CGG-3', 'dw_reverse_p': r1000 53-mer:5'-TCCCCGTCCGGGACC..CCG-3', 'dw_forwar_primer_str': 'ATGCCGTTCTCCCGGCAGATGCCCAACGAGTCGG', 'dw_reverse_primer_str': 'TCCCCGTCCGGGACCCGCGCGGTCGATCCCCGCATATAGGCGGCGAGAGGCCG', 'dw_forwar_p_anneal': Seq('GCCCAACGAGTCGG'), 'dw_reverse_p_anneal': Seq('CGGCGAGAGGCCG'), 'tm_dw_forwar_p': 55, 'tm_dw_reverse_p': 57, 'ta_dw': 58}, {'gene_name': 'region2', 'name': 'pCas9_region2(346)', 'contig': Dseqrecord(o13279), 'up_forwar_p_name': 'primer_5', 'up_reverse_p_name': 'primer_6', 'up_forwar_p': primer_5 53-mer:5'-AAGGCCGCTTTTGCG..GCG-3', 'up_reverse_p': primer_6 38-mer:5'-CCACAGCGCGTGATA..CAC-3', 'up_forwar_primer_str': 'AAGGCCGCTTTTGCGGGATCTCGTCGAAGGCACTAGAAGGCAGCAACGCCGCG', 'up_reverse_primer_str': 'CCACAGCGCGTGATAGACAGAGACAACAAGCCCGTCAC', 'up_forwar_p_anneal': Seq('CAGCAACGCCGCG'), 'up_reverse_p_anneal': Seq('AGACAACAAGCCCGTCAC'), 'tm_up_forwar_p': 57, 'tm_up_reverse_p': 59, 'ta_up': 60, 'dw_forwar_p_name': 'primer_7', 'dw_reverse_p_name': 'primer_8', 'dw_forwar_p': primer_7 38-mer:5'-CCGTGACGGGCTTGT..TGT-3', 'dw_reverse_p': primer_8 57-mer:5'-TCCCCGTCCGGGACC..AGG-3', 'dw_forwar_primer_str': 'CCGTGACGGGCTTGTTGTCTCTGTCTATCACGCGCTGT', 'dw_reverse_primer_str': 'TCCCCGTCCGGGACCCGCGCGGTCGATCCCCGCATATAGGTGGCCATGATGTCCAGG', 'dw_forwar_p_anneal': Seq('CTGTCTATCACGCGCTGT'), 'dw_reverse_p_anneal': Seq('TGGCCATGATGTCCAGG'), 'tm_dw_forwar_p': 59, 'tm_dw_reverse_p': 58, 'ta_dw': 62}, {'gene_name': 'region2', 'name': 'pCas9_region2(340)', 'contig': Dseqrecord(o13279), 'up_forwar_p_name': 'primer_5', 'up_reverse_p_name': 'primer_6', 'up_forwar_p': f1000 53-mer:5'-AAGGCCGCTTTTGCG..GCG-3', 'up_reverse_p': r1000 38-mer:5'-CCACAGCGCGTGATA..CAC-3', 'up_forwar_primer_str': 'AAGGCCGCTTTTGCGGGATCTCGTCGAAGGCACTAGAAGGCAGCAACGCCGCG', 'up_reverse_primer_str': 'CCACAGCGCGTGATAGACAGAGACAACAAGCCCGTCAC', 'up_forwar_p_anneal': Seq('CAGCAACGCCGCG'), 'up_reverse_p_anneal': Seq('AGACAACAAGCCCGTCAC'), 'tm_up_forwar_p': 57, 'tm_up_reverse_p': 59, 'ta_up': 60, 'dw_forwar_p_name': 'primer_7', 'dw_reverse_p_name': 'primer_8', 'dw_forwar_p': f1000 38-mer:5'-CCGTGACGGGCTTGT..TGT-3', 'dw_reverse_p': r1000 57-mer:5'-TCCCCGTCCGGGACC..AGG-3', 'dw_forwar_primer_str': 'CCGTGACGGGCTTGTTGTCTCTGTCTATCACGCGCTGT', 'dw_reverse_primer_str': 'TCCCCGTCCGGGACCCGCGCGGTCGATCCCCGCATATAGGTGGCCATGATGTCCAGG', 'dw_forwar_p_anneal': Seq('CTGTCTATCACGCGCTGT'), 'dw_reverse_p_anneal': Seq('TGGCCATGATGTCCAGG'), 'tm_dw_forwar_p': 59, 'tm_dw_reverse_p': 58, 'ta_dw': 62}, {'gene_name': 'region3', 'name': 'pCas9_region3(3477)', 'contig': Dseqrecord(o13279), 'up_forwar_p_name': 'primer_9', 'up_reverse_p_name': 'primer_10', 'up_forwar_p': primer_9 54-mer:5'-AAGGCCGCTTTTGCG..GGC-3', 'up_reverse_p': primer_10 33-mer:5'-CATCGTGAACTGACC..CGG-3', 'up_forwar_primer_str': 'AAGGCCGCTTTTGCGGGATCTCGTCGAAGGCACTAGAAGGTCCTCGAGGACGGC', 'up_reverse_primer_str': 'CATCGTGAACTGACCCCTCGTCGGCGAGGTCGG', 'up_forwar_p_anneal': Seq('TCCTCGAGGACGGC'), 'up_reverse_p_anneal': Seq('TCGGCGAGGTCGG'), 'tm_up_forwar_p': 55, 'tm_up_reverse_p': 55, 'ta_up': 59, 'dw_forwar_p_name': 'primer_11', 'dw_reverse_p_name': 'primer_12', 'dw_forwar_p': primer_11 37-mer:5'-CACGAGGCCGACCTC..ACG-3', 'dw_reverse_p': primer_12 55-mer:5'-TCCCCGTCCGGGACC..GCC-3', 'dw_forwar_primer_str': 'CACGAGGCCGACCTCGCCGACGAGGGGTCAGTTCACG', 'dw_reverse_primer_str': 'TCCCCGTCCGGGACCCGCGCGGTCGATCCCCGCATATAGGCAACATCGCGCTGCC', 'dw_forwar_p_anneal': Seq('CGAGGGGTCAGTTCACG'), 'dw_reverse_p_anneal': Seq('CAACATCGCGCTGCC'), 'tm_dw_forwar_p': 58, 'tm_dw_reverse_p': 57, 'ta_dw': 61}, {'gene_name': 'region3', 'name': 'pCas9_region3(2460)', 'contig': Dseqrecord(o13279), 'up_forwar_p_name': 'primer_9', 'up_reverse_p_name': 'primer_10', 'up_forwar_p': f1000 54-mer:5'-AAGGCCGCTTTTGCG..GGC-3', 'up_reverse_p': r1000 33-mer:5'-CATCGTGAACTGACC..CGG-3', 'up_forwar_primer_str': 'AAGGCCGCTTTTGCGGGATCTCGTCGAAGGCACTAGAAGGTCCTCGAGGACGGC', 'up_reverse_primer_str': 'CATCGTGAACTGACCCCTCGTCGGCGAGGTCGG', 'up_forwar_p_anneal': Seq('TCCTCGAGGACGGC'), 'up_reverse_p_anneal': Seq('TCGGCGAGGTCGG'), 'tm_up_forwar_p': 55, 'tm_up_reverse_p': 55, 'ta_up': 59, 'dw_forwar_p_name': 'primer_11', 'dw_reverse_p_name': 'primer_12', 'dw_forwar_p': f1000 37-mer:5'-CACGAGGCCGACCTC..ACG-3', 'dw_reverse_p': r1000 55-mer:5'-TCCCCGTCCGGGACC..GCC-3', 'dw_forwar_primer_str': 'CACGAGGCCGACCTCGCCGACGAGGGGTCAGTTCACG', 'dw_reverse_primer_str': 'TCCCCGTCCGGGACCCGCGCGGTCGATCCCCGCATATAGGCAACATCGCGCTGCC', 'dw_forwar_p_anneal': Seq('CGAGGGGTCAGTTCACG'), 'dw_reverse_p_anneal': Seq('CAACATCGCGCTGCC'), 'tm_dw_forwar_p': 58, 'tm_dw_reverse_p': 57, 'ta_dw': 61}, {'gene_name': 'region4', 'name': 'pCas9_region4(265)', 'contig': Dseqrecord(o13279), 'up_forwar_p_name': 'primer_13', 'up_reverse_p_name': 'primer_14', 'up_forwar_p': primer_13 57-mer:5'-AAGGCCGCTTTTGCG..ACG-3', 'up_reverse_p': primer_14 35-mer:5'-GGTAACAGGCTGACC..GCC-3', 'up_forwar_primer_str': 'AAGGCCGCTTTTGCGGGATCTCGTCGAAGGCACTAGAAGGCGAGGGGTCAGTTCACG', 'up_reverse_primer_str': 'GGTAACAGGCTGACCAGTTCCAACATCGCGCTGCC', 'up_forwar_p_anneal': Seq('CGAGGGGTCAGTTCACG'), 'up_reverse_p_anneal': Seq('CAACATCGCGCTGCC'), 'tm_up_forwar_p': 58, 'tm_up_reverse_p': 57, 'ta_up': 61, 'dw_forwar_p_name': 'primer_15', 'dw_reverse_p_name': 'primer_16', 'dw_forwar_p': primer_15 40-mer:5'-TGCCGGGCAGCGCGA..ACC-3', 'dw_reverse_p': primer_16 53-mer:5'-TCCCCGTCCGGGACC..CGG-3', 'dw_forwar_primer_str': 'TGCCGGGCAGCGCGATGTTGGAACTGGTCAGCCTGTTACC', 'dw_reverse_primer_str': 'TCCCCGTCCGGGACCCGCGCGGTCGATCCCCGCATATAGGGCTTGCGGGGCGG', 'dw_forwar_p_anneal': Seq('GAACTGGTCAGCCTGTTACC'), 'dw_reverse_p_anneal': Seq('GCTTGCGGGGCGG'), 'tm_dw_forwar_p': 60, 'tm_dw_reverse_p': 59, 'ta_dw': 62}, {'gene_name': 'region4', 'name': 'pCas9_region4(254)', 'contig': Dseqrecord(o13279), 'up_forwar_p_name': 'primer_13', 'up_reverse_p_name': 'primer_14', 'up_forwar_p': f1000 57-mer:5'-AAGGCCGCTTTTGCG..ACG-3', 'up_reverse_p': r1000 35-mer:5'-GGTAACAGGCTGACC..GCC-3', 'up_forwar_primer_str': 'AAGGCCGCTTTTGCGGGATCTCGTCGAAGGCACTAGAAGGCGAGGGGTCAGTTCACG', 'up_reverse_primer_str': 'GGTAACAGGCTGACCAGTTCCAACATCGCGCTGCC', 'up_forwar_p_anneal': Seq('CGAGGGGTCAGTTCACG'), 'up_reverse_p_anneal': Seq('CAACATCGCGCTGCC'), 'tm_up_forwar_p': 58, 'tm_up_reverse_p': 57, 'ta_up': 61, 'dw_forwar_p_name': 'primer_15', 'dw_reverse_p_name': 'primer_16', 'dw_forwar_p': f1000 40-mer:5'-TGCCGGGCAGCGCGA..ACC-3', 'dw_reverse_p': r1000 53-mer:5'-TCCCCGTCCGGGACC..CGG-3', 'dw_forwar_primer_str': 'TGCCGGGCAGCGCGATGTTGGAACTGGTCAGCCTGTTACC', 'dw_reverse_primer_str': 'TCCCCGTCCGGGACCCGCGCGGTCGATCCCCGCATATAGGGCTTGCGGGGCGG', 'dw_forwar_p_anneal': Seq('GAACTGGTCAGCCTGTTACC'), 'dw_reverse_p_anneal': Seq('GCTTGCGGGGCGG'), 'tm_dw_forwar_p': 60, 'tm_dw_reverse_p': 59, 'ta_dw': 62}]\n"
          ]
        }
      ],
      "source": [
        "if in_frame_deletion:\n",
        "    \n",
        "\n",
        "    # Make repair templates\n",
        "    repair_templates_data = find_up_dw_repair_templates(genome, \n",
        "                                                        genes_to_KO, \n",
        "                                                        target_tm=melting_temperature, \n",
        "                                                        \n",
        "                                                        primer_tm_kwargs={'conc':primer_concentration, 'prodcode':chosen_polymerase} )\n",
        "\n",
        "\n",
        "    # Digest the plasmids\n",
        "    processed_records = [Dseqrecord(record, circular=True).cut(StuI)[0] for record in sgRNA_vectors]\n",
        "\n",
        "    # Rename them appropriately\n",
        "    for i in range(len(processed_records)):\n",
        "        processed_records[i].name = sgRNA_vectors[i].name\n",
        "\n",
        "    # Assembly \n",
        "    assembly_data = assemble_multiple_plasmids_with_repair_templates_for_deletion(genes_to_KO, processed_records, \n",
        "                                                                                repair_templates_data, \n",
        "                                                                                overlap=40)\n",
        "    # updating the primer names to something systematic.\n",
        "    update_primer_names(assembly_data)\n",
        "\n",
        "    # Parse through the primer df\n",
        "    primer_df = create_primer_df_from_dict(assembly_data)\n",
        "\n",
        "    # Rename them appropriately\n",
        "    for i in range(len(processed_records)):\n",
        "        processed_records[i].name = sgRNA_vectors[i].name\n",
        "\n",
        "    # Assembly \n",
        "    assembly_data = assemble_multiple_plasmids_with_repair_templates_for_deletion(genes_to_KO, processed_records, \n",
        "                                                                                repair_templates_data, \n",
        "                                                                                overlap=40)\n",
        "    # updating the primer names to something systematic.\n",
        "    update_primer_names(assembly_data)\n",
        "    print(assembly_data)\n",
        "\n",
        "    # Parse through the primer df\n",
        "    primer_df = create_primer_df_from_dict(assembly_data)\n",
        "    unique_df = primer_df.drop_duplicates(keep='first')\n",
        "\n",
        "    # IDT df\n",
        "    idt_df = create_idt_order_dataframe(unique_df, concentration=\"25nm\", purification=\"STD\")\n",
        "\n",
        "    # Contigs\n",
        "    assembled_contigs = []\n",
        "    for data in assembly_data: \n",
        "        contig_record = data['contig']\n",
        "        contig_record.id = f\"{data['name']}_w_rep\"\n",
        "        contig_record.name = f\"{data['name']}_w_rep\"\n",
        "        assembled_contigs.append(contig_record)\n",
        "\n",
        "\n",
        "    print_plasmids = False\n",
        "\n",
        "    if print_plasmids:\n",
        "        # Write to genbank files\n",
        "        output_directory = \"../../data/plasmids/sgRNA_plasmids_pCRISPR-Cas9_with_repair_templates/\"\n",
        "        for contig in assembled_contigs: \n",
        "            contig.write(f\"{output_directory}/{contig.id}.gb\")"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 13,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>sgRNA plasmid</th>\n",
              "      <th>repair template plasmid</th>\n",
              "      <th>which workflow to proceed with</th>\n",
              "      <th>sgRNA plasmid #StuI sites</th>\n",
              "      <th>repair template plasmid #NcoI sites</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>pCas9_region1(6574)</td>\n",
              "      <td>pCas9_region1(6574)_w_rep</td>\n",
              "      <td>Proceed with sgRNA integration first</td>\n",
              "      <td>1</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>pCas9_region1(16046)</td>\n",
              "      <td>pCas9_region1(16046)_w_rep</td>\n",
              "      <td>Proceed with sgRNA integration first</td>\n",
              "      <td>1</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>pCas9_region3(3477)</td>\n",
              "      <td>pCas9_region3(3477)_w_rep</td>\n",
              "      <td>Proceed with sgRNA integration first</td>\n",
              "      <td>1</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>pCas9_region4(265)</td>\n",
              "      <td>pCas9_region4(265)_w_rep</td>\n",
              "      <td>Proceed with sgRNA integration first</td>\n",
              "      <td>1</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>pCas9_region2(346)</td>\n",
              "      <td>pCas9_region2(346)_w_rep</td>\n",
              "      <td>Proceed with sgRNA integration first</td>\n",
              "      <td>1</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>5</th>\n",
              "      <td>pCas9_region3(2460)</td>\n",
              "      <td>pCas9_region3(2460)_w_rep</td>\n",
              "      <td>Proceed with sgRNA integration first</td>\n",
              "      <td>1</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>6</th>\n",
              "      <td>pCas9_region2(340)</td>\n",
              "      <td>pCas9_region2(340)_w_rep</td>\n",
              "      <td>Proceed with sgRNA integration first</td>\n",
              "      <td>1</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>7</th>\n",
              "      <td>pCas9_region4(254)</td>\n",
              "      <td>pCas9_region4(254)_w_rep</td>\n",
              "      <td>Proceed with sgRNA integration first</td>\n",
              "      <td>1</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "          sgRNA plasmid     repair template plasmid  \\\n",
              "0   pCas9_region1(6574)   pCas9_region1(6574)_w_rep   \n",
              "1  pCas9_region1(16046)  pCas9_region1(16046)_w_rep   \n",
              "2   pCas9_region3(3477)   pCas9_region3(3477)_w_rep   \n",
              "3    pCas9_region4(265)    pCas9_region4(265)_w_rep   \n",
              "4    pCas9_region2(346)    pCas9_region2(346)_w_rep   \n",
              "5   pCas9_region3(2460)   pCas9_region3(2460)_w_rep   \n",
              "6    pCas9_region2(340)    pCas9_region2(340)_w_rep   \n",
              "7    pCas9_region4(254)    pCas9_region4(254)_w_rep   \n",
              "\n",
              "         which workflow to proceed with  sgRNA plasmid #StuI sites  \\\n",
              "0  Proceed with sgRNA integration first                          1   \n",
              "1  Proceed with sgRNA integration first                          1   \n",
              "2  Proceed with sgRNA integration first                          1   \n",
              "3  Proceed with sgRNA integration first                          1   \n",
              "4  Proceed with sgRNA integration first                          1   \n",
              "5  Proceed with sgRNA integration first                          1   \n",
              "6  Proceed with sgRNA integration first                          1   \n",
              "7  Proceed with sgRNA integration first                          1   \n",
              "\n",
              "   repair template plasmid #NcoI sites  \n",
              "0                                    0  \n",
              "1                                    0  \n",
              "2                                    0  \n",
              "3                                    0  \n",
              "4                                    0  \n",
              "5                                    0  \n",
              "6                                    0  \n",
              "7                                    0  "
            ]
          },
          "execution_count": 13,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "workflow_df = determine_workflow_order_for_plasmids(sgRNA_vectors, \n",
        "                                                            assembled_contigs,\n",
        "                                                            [\"StuI\", \"NcoI\"])\n",
        "workflow_df"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 14,
      "metadata": {},
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "Determining the order of workflows :           sgRNA plasmid     repair template plasmid  \\\n",
            "0   pCas9_region1(6574)   pCas9_region1(6574)_w_rep   \n",
            "1  pCas9_region1(16046)  pCas9_region1(16046)_w_rep   \n",
            "2   pCas9_region3(3477)   pCas9_region3(3477)_w_rep   \n",
            "3    pCas9_region4(265)    pCas9_region4(265)_w_rep   \n",
            "4    pCas9_region2(346)    pCas9_region2(346)_w_rep   \n",
            "5   pCas9_region3(2460)   pCas9_region3(2460)_w_rep   \n",
            "6    pCas9_region2(340)    pCas9_region2(340)_w_rep   \n",
            "7    pCas9_region4(254)    pCas9_region4(254)_w_rep   \n",
            "\n",
            "         which workflow to proceed with  sgRNA plasmid #StuI sites  \\\n",
            "0  Proceed with sgRNA integration first                          1   \n",
            "1  Proceed with sgRNA integration first                          1   \n",
            "2  Proceed with sgRNA integration first                          1   \n",
            "3  Proceed with sgRNA integration first                          1   \n",
            "4  Proceed with sgRNA integration first                          1   \n",
            "5  Proceed with sgRNA integration first                          1   \n",
            "6  Proceed with sgRNA integration first                          1   \n",
            "7  Proceed with sgRNA integration first                          1   \n",
            "\n",
            "   repair template plasmid #NcoI sites  \n",
            "0                                    0  \n",
            "1                                    0  \n",
            "2                                    0  \n",
            "3                                    0  \n",
            "4                                    0  \n",
            "5                                    0  \n",
            "6                                    0  \n",
            "7                                    0  \n"
          ]
        }
      ],
      "source": [
        "if in_frame_deletion:\n",
        "    integration_names = filtered_df.apply(lambda row: f\"sgRNA_{row['locus_tag']}({row['sgrna_loc']})\", axis=1).tolist()\n",
        "    plasmid_metadata_df = extract_metadata_to_dataframe(assembled_contigs,\n",
        "                                                        clean_plasmid,\n",
        "                                                        integration_names)\n",
        "    workflow_df = determine_workflow_order_for_plasmids(sgRNA_vectors, \n",
        "                                                            assembled_contigs,\n",
        "                                                            [\"StuI\", \"NcoI\"])\n",
        "    \n",
        "    print(f'Determining the order of workflows : {workflow_df}')\n",
        "\n",
        "else: \n",
        "    integration_names = filtered_df.apply(lambda row: f\"sgRNA_{row['locus_tag']}({row['sgrna_loc']})\", axis=1).tolist()\n",
        "    plasmid_metadata_df = extract_metadata_to_dataframe(sgRNA_vectors,\n",
        "                                                        clean_plasmid,\n",
        "                                                        integration_names)\n",
        "\n",
        "    plasmid_metadata_df"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 15,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>locus tag</th>\n",
              "      <th>f_primer_name</th>\n",
              "      <th>r_primer_name</th>\n",
              "      <th>f_primer_sequences(5-3)</th>\n",
              "      <th>r_primer_sequences(5-3)</th>\n",
              "      <th>f_tm</th>\n",
              "      <th>r_tm</th>\n",
              "      <th>ta</th>\n",
              "      <th>flanking_region</th>\n",
              "      <th>annealing_temperature</th>\n",
              "      <th>...</th>\n",
              "      <th>homodimer_reverse_tm</th>\n",
              "      <th>homodimer_reverse_deltaG (kcal/mol)</th>\n",
              "      <th>heterodimer_tm</th>\n",
              "      <th>heterodimer_deltaG (kcal/mol)</th>\n",
              "      <th>hairpin_forward_structure_found</th>\n",
              "      <th>hairpin_forward_tm</th>\n",
              "      <th>hairpin_forward_deltaG (kcal/mol)</th>\n",
              "      <th>hairpin_reverse_structure_found</th>\n",
              "      <th>hairpin_reverse_tm</th>\n",
              "      <th>hairpin_reverse_deltaG (kcal/mol)</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>region2</td>\n",
              "      <td>region2_fwd_checking_primer</td>\n",
              "      <td>region2_rev_checking_primer</td>\n",
              "      <td>GACCGCGCTGGGTG</td>\n",
              "      <td>TCCGGCCGCATCG</td>\n",
              "      <td>59</td>\n",
              "      <td>57</td>\n",
              "      <td>60</td>\n",
              "      <td>500</td>\n",
              "      <td>60</td>\n",
              "      <td>...</td>\n",
              "      <td>22.848051</td>\n",
              "      <td>-5.018625</td>\n",
              "      <td>6.535424</td>\n",
              "      <td>-0.061707</td>\n",
              "      <td>False</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>False</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>region4</td>\n",
              "      <td>region4_fwd_checking_primer</td>\n",
              "      <td>region4_rev_checking_primer</td>\n",
              "      <td>TTGCTGCAACCTGCCTTG</td>\n",
              "      <td>TAGGTGCCTCCGTCGA</td>\n",
              "      <td>61</td>\n",
              "      <td>58</td>\n",
              "      <td>62</td>\n",
              "      <td>500</td>\n",
              "      <td>62</td>\n",
              "      <td>...</td>\n",
              "      <td>-13.794142</td>\n",
              "      <td>-0.897729</td>\n",
              "      <td>-1.321907</td>\n",
              "      <td>0.948660</td>\n",
              "      <td>False</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>False</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>region3</td>\n",
              "      <td>region3_fwd_checking_primer</td>\n",
              "      <td>region3_rev_checking_primer</td>\n",
              "      <td>ATCCCGCAGCTGCAGG</td>\n",
              "      <td>GAGGAGCGGCAATGGTCATC</td>\n",
              "      <td>61</td>\n",
              "      <td>63</td>\n",
              "      <td>65</td>\n",
              "      <td>500</td>\n",
              "      <td>65</td>\n",
              "      <td>...</td>\n",
              "      <td>1.445721</td>\n",
              "      <td>-1.026208</td>\n",
              "      <td>9.252334</td>\n",
              "      <td>-1.213024</td>\n",
              "      <td>False</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>False</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>region1</td>\n",
              "      <td>region1_fwd_checking_primer</td>\n",
              "      <td>region1_rev_checking_primer</td>\n",
              "      <td>GCCGCTGGCCAAGC</td>\n",
              "      <td>AGGCGTGCGCGTGG</td>\n",
              "      <td>60</td>\n",
              "      <td>62</td>\n",
              "      <td>63</td>\n",
              "      <td>700</td>\n",
              "      <td>63</td>\n",
              "      <td>...</td>\n",
              "      <td>22.176370</td>\n",
              "      <td>-2.921991</td>\n",
              "      <td>11.333376</td>\n",
              "      <td>-0.877576</td>\n",
              "      <td>False</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>False</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>4 rows × 23 columns</p>\n",
              "</div>"
            ],
            "text/plain": [
              "  locus tag                f_primer_name                r_primer_name  \\\n",
              "0   region2  region2_fwd_checking_primer  region2_rev_checking_primer   \n",
              "1   region4  region4_fwd_checking_primer  region4_rev_checking_primer   \n",
              "2   region3  region3_fwd_checking_primer  region3_rev_checking_primer   \n",
              "3   region1  region1_fwd_checking_primer  region1_rev_checking_primer   \n",
              "\n",
              "  f_primer_sequences(5-3) r_primer_sequences(5-3) f_tm r_tm  ta  \\\n",
              "0          GACCGCGCTGGGTG           TCCGGCCGCATCG   59   57  60   \n",
              "1      TTGCTGCAACCTGCCTTG        TAGGTGCCTCCGTCGA   61   58  62   \n",
              "2        ATCCCGCAGCTGCAGG    GAGGAGCGGCAATGGTCATC   61   63  65   \n",
              "3          GCCGCTGGCCAAGC          AGGCGTGCGCGTGG   60   62  63   \n",
              "\n",
              "  flanking_region  annealing_temperature  ... homodimer_reverse_tm  \\\n",
              "0             500                     60  ...            22.848051   \n",
              "1             500                     62  ...           -13.794142   \n",
              "2             500                     65  ...             1.445721   \n",
              "3             700                     63  ...            22.176370   \n",
              "\n",
              "   homodimer_reverse_deltaG (kcal/mol)  heterodimer_tm  \\\n",
              "0                            -5.018625        6.535424   \n",
              "1                            -0.897729       -1.321907   \n",
              "2                            -1.026208        9.252334   \n",
              "3                            -2.921991       11.333376   \n",
              "\n",
              "   heterodimer_deltaG (kcal/mol)  hairpin_forward_structure_found  \\\n",
              "0                      -0.061707                            False   \n",
              "1                       0.948660                            False   \n",
              "2                      -1.213024                            False   \n",
              "3                      -0.877576                            False   \n",
              "\n",
              "   hairpin_forward_tm  hairpin_forward_deltaG (kcal/mol)  \\\n",
              "0                 0.0                                0.0   \n",
              "1                 0.0                                0.0   \n",
              "2                 0.0                                0.0   \n",
              "3                 0.0                                0.0   \n",
              "\n",
              "   hairpin_reverse_structure_found  hairpin_reverse_tm  \\\n",
              "0                            False                 0.0   \n",
              "1                            False                 0.0   \n",
              "2                            False                 0.0   \n",
              "3                            False                 0.0   \n",
              "\n",
              "   hairpin_reverse_deltaG (kcal/mol)  \n",
              "0                                0.0  \n",
              "1                                0.0  \n",
              "2                                0.0  \n",
              "3                                0.0  \n",
              "\n",
              "[4 rows x 23 columns]"
            ]
          },
          "execution_count": 15,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "# Getting checking primers\n",
        "checking_primers_df = find_best_check_primers_from_genome(genome, \n",
        "                                       genes_to_KO, \n",
        "                                       flanking_region=flanking_region,\n",
        "                                       target_tm = melting_temperature, \n",
        "                                        primer_concentration = primer_concentration, \n",
        "                                        polymerase = chosen_polymerase)\n",
        "checking_primers_df\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 16,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>Name</th>\n",
              "      <th>Sequence</th>\n",
              "      <th>Concentration</th>\n",
              "      <th>Purification</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>region2_fwd_checking_primer</td>\n",
              "      <td>GACCGCGCTGGGTG</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>region4_fwd_checking_primer</td>\n",
              "      <td>TTGCTGCAACCTGCCTTG</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>region3_fwd_checking_primer</td>\n",
              "      <td>ATCCCGCAGCTGCAGG</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>region1_fwd_checking_primer</td>\n",
              "      <td>GCCGCTGGCCAAGC</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>region2_rev_checking_primer</td>\n",
              "      <td>TCCGGCCGCATCG</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>5</th>\n",
              "      <td>region4_rev_checking_primer</td>\n",
              "      <td>TAGGTGCCTCCGTCGA</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>6</th>\n",
              "      <td>region3_rev_checking_primer</td>\n",
              "      <td>GAGGAGCGGCAATGGTCATC</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>7</th>\n",
              "      <td>region1_rev_checking_primer</td>\n",
              "      <td>AGGCGTGCGCGTGG</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "                          Name              Sequence Concentration  \\\n",
              "0  region2_fwd_checking_primer        GACCGCGCTGGGTG          25nm   \n",
              "1  region4_fwd_checking_primer    TTGCTGCAACCTGCCTTG          25nm   \n",
              "2  region3_fwd_checking_primer      ATCCCGCAGCTGCAGG          25nm   \n",
              "3  region1_fwd_checking_primer        GCCGCTGGCCAAGC          25nm   \n",
              "4  region2_rev_checking_primer         TCCGGCCGCATCG          25nm   \n",
              "5  region4_rev_checking_primer      TAGGTGCCTCCGTCGA          25nm   \n",
              "6  region3_rev_checking_primer  GAGGAGCGGCAATGGTCATC          25nm   \n",
              "7  region1_rev_checking_primer        AGGCGTGCGCGTGG          25nm   \n",
              "\n",
              "  Purification  \n",
              "0          STD  \n",
              "1          STD  \n",
              "2          STD  \n",
              "3          STD  \n",
              "4          STD  \n",
              "5          STD  \n",
              "6          STD  \n",
              "7          STD  "
            ]
          },
          "execution_count": 16,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "# making the primers into dseqrecords\n",
        "checking_primers_df_idt = create_idt_order_dataframe(checking_primers_df)\n",
        "checking_primers_df_idt\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 17,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>Name</th>\n",
              "      <th>Sequence</th>\n",
              "      <th>Concentration</th>\n",
              "      <th>Purification</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>region1_loc_6574</td>\n",
              "      <td>CGGTTGGTAGGATCGACGGCGCCACATTCACTACGACCCCGTTTTA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>region1_loc_16046</td>\n",
              "      <td>CGGTTGGTAGGATCGACGGCCCCTGATCCGAGTGTGTGCCGTTTTA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>region3_loc_3477</td>\n",
              "      <td>CGGTTGGTAGGATCGACGGCGCCGGTCGGCCATATCCCCTGTTTTA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>region4_loc_265</td>\n",
              "      <td>CGGTTGGTAGGATCGACGGCACGGCGTACAGGTCGCGTACGTTTTA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>region2_loc_346</td>\n",
              "      <td>CGGTTGGTAGGATCGACGGCGTCCCTCGACTCACTGGCGCGTTTTA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>5</th>\n",
              "      <td>region3_loc_2460</td>\n",
              "      <td>CGGTTGGTAGGATCGACGGCTAGTTGTGTGGGATTGATCTGTTTTA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>6</th>\n",
              "      <td>region2_loc_340</td>\n",
              "      <td>CGGTTGGTAGGATCGACGGCGTGCTGGTCCCTCGACTCACGTTTTA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>7</th>\n",
              "      <td>region4_loc_254</td>\n",
              "      <td>CGGTTGGTAGGATCGACGGCTCACGTCACCGACGGCGTACGTTTTA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>primer_1</td>\n",
              "      <td>AAGGCCGCTTTTGCGGGATCTCGTCGAAGGCACTAGAAGGGATGAC...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>primer_3</td>\n",
              "      <td>ATGCCGTTCTCCCGGCAGATGCCCAACGAGTCGG</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>primer_5</td>\n",
              "      <td>AAGGCCGCTTTTGCGGGATCTCGTCGAAGGCACTAGAAGGCAGCAA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>primer_7</td>\n",
              "      <td>CCGTGACGGGCTTGTTGTCTCTGTCTATCACGCGCTGT</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>primer_9</td>\n",
              "      <td>AAGGCCGCTTTTGCGGGATCTCGTCGAAGGCACTAGAAGGTCCTCG...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>5</th>\n",
              "      <td>primer_11</td>\n",
              "      <td>CACGAGGCCGACCTCGCCGACGAGGGGTCAGTTCACG</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>6</th>\n",
              "      <td>primer_13</td>\n",
              "      <td>AAGGCCGCTTTTGCGGGATCTCGTCGAAGGCACTAGAAGGCGAGGG...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>7</th>\n",
              "      <td>primer_15</td>\n",
              "      <td>TGCCGGGCAGCGCGATGTTGGAACTGGTCAGCCTGTTACC</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>8</th>\n",
              "      <td>primer_2</td>\n",
              "      <td>GCCCGGCCGACTCGTTGGGCATCTGCCGGGAGAACG</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>9</th>\n",
              "      <td>primer_4</td>\n",
              "      <td>TCCCCGTCCGGGACCCGCGCGGTCGATCCCCGCATATAGGCGGCGA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>10</th>\n",
              "      <td>primer_6</td>\n",
              "      <td>CCACAGCGCGTGATAGACAGAGACAACAAGCCCGTCAC</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>11</th>\n",
              "      <td>primer_8</td>\n",
              "      <td>TCCCCGTCCGGGACCCGCGCGGTCGATCCCCGCATATAGGTGGCCA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>12</th>\n",
              "      <td>primer_10</td>\n",
              "      <td>CATCGTGAACTGACCCCTCGTCGGCGAGGTCGG</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>13</th>\n",
              "      <td>primer_12</td>\n",
              "      <td>TCCCCGTCCGGGACCCGCGCGGTCGATCCCCGCATATAGGCAACAT...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>14</th>\n",
              "      <td>primer_14</td>\n",
              "      <td>GGTAACAGGCTGACCAGTTCCAACATCGCGCTGCC</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>15</th>\n",
              "      <td>primer_16</td>\n",
              "      <td>TCCCCGTCCGGGACCCGCGCGGTCGATCCCCGCATATAGGGCTTGC...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>region2_fwd_checking_primer</td>\n",
              "      <td>GACCGCGCTGGGTG</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>region4_fwd_checking_primer</td>\n",
              "      <td>TTGCTGCAACCTGCCTTG</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>region3_fwd_checking_primer</td>\n",
              "      <td>ATCCCGCAGCTGCAGG</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>region1_fwd_checking_primer</td>\n",
              "      <td>GCCGCTGGCCAAGC</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>region2_rev_checking_primer</td>\n",
              "      <td>TCCGGCCGCATCG</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>5</th>\n",
              "      <td>region4_rev_checking_primer</td>\n",
              "      <td>TAGGTGCCTCCGTCGA</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>6</th>\n",
              "      <td>region3_rev_checking_primer</td>\n",
              "      <td>GAGGAGCGGCAATGGTCATC</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>7</th>\n",
              "      <td>region1_rev_checking_primer</td>\n",
              "      <td>AGGCGTGCGCGTGG</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "                           Name  \\\n",
              "0              region1_loc_6574   \n",
              "1             region1_loc_16046   \n",
              "2              region3_loc_3477   \n",
              "3               region4_loc_265   \n",
              "4               region2_loc_346   \n",
              "5              region3_loc_2460   \n",
              "6               region2_loc_340   \n",
              "7               region4_loc_254   \n",
              "0                      primer_1   \n",
              "1                      primer_3   \n",
              "2                      primer_5   \n",
              "3                      primer_7   \n",
              "4                      primer_9   \n",
              "5                     primer_11   \n",
              "6                     primer_13   \n",
              "7                     primer_15   \n",
              "8                      primer_2   \n",
              "9                      primer_4   \n",
              "10                     primer_6   \n",
              "11                     primer_8   \n",
              "12                    primer_10   \n",
              "13                    primer_12   \n",
              "14                    primer_14   \n",
              "15                    primer_16   \n",
              "0   region2_fwd_checking_primer   \n",
              "1   region4_fwd_checking_primer   \n",
              "2   region3_fwd_checking_primer   \n",
              "3   region1_fwd_checking_primer   \n",
              "4   region2_rev_checking_primer   \n",
              "5   region4_rev_checking_primer   \n",
              "6   region3_rev_checking_primer   \n",
              "7   region1_rev_checking_primer   \n",
              "\n",
              "                                             Sequence Concentration  \\\n",
              "0   CGGTTGGTAGGATCGACGGCGCCACATTCACTACGACCCCGTTTTA...          25nm   \n",
              "1   CGGTTGGTAGGATCGACGGCCCCTGATCCGAGTGTGTGCCGTTTTA...          25nm   \n",
              "2   CGGTTGGTAGGATCGACGGCGCCGGTCGGCCATATCCCCTGTTTTA...          25nm   \n",
              "3   CGGTTGGTAGGATCGACGGCACGGCGTACAGGTCGCGTACGTTTTA...          25nm   \n",
              "4   CGGTTGGTAGGATCGACGGCGTCCCTCGACTCACTGGCGCGTTTTA...          25nm   \n",
              "5   CGGTTGGTAGGATCGACGGCTAGTTGTGTGGGATTGATCTGTTTTA...          25nm   \n",
              "6   CGGTTGGTAGGATCGACGGCGTGCTGGTCCCTCGACTCACGTTTTA...          25nm   \n",
              "7   CGGTTGGTAGGATCGACGGCTCACGTCACCGACGGCGTACGTTTTA...          25nm   \n",
              "0   AAGGCCGCTTTTGCGGGATCTCGTCGAAGGCACTAGAAGGGATGAC...          25nm   \n",
              "1                  ATGCCGTTCTCCCGGCAGATGCCCAACGAGTCGG          25nm   \n",
              "2   AAGGCCGCTTTTGCGGGATCTCGTCGAAGGCACTAGAAGGCAGCAA...          25nm   \n",
              "3              CCGTGACGGGCTTGTTGTCTCTGTCTATCACGCGCTGT          25nm   \n",
              "4   AAGGCCGCTTTTGCGGGATCTCGTCGAAGGCACTAGAAGGTCCTCG...          25nm   \n",
              "5               CACGAGGCCGACCTCGCCGACGAGGGGTCAGTTCACG          25nm   \n",
              "6   AAGGCCGCTTTTGCGGGATCTCGTCGAAGGCACTAGAAGGCGAGGG...          25nm   \n",
              "7            TGCCGGGCAGCGCGATGTTGGAACTGGTCAGCCTGTTACC          25nm   \n",
              "8                GCCCGGCCGACTCGTTGGGCATCTGCCGGGAGAACG          25nm   \n",
              "9   TCCCCGTCCGGGACCCGCGCGGTCGATCCCCGCATATAGGCGGCGA...          25nm   \n",
              "10             CCACAGCGCGTGATAGACAGAGACAACAAGCCCGTCAC          25nm   \n",
              "11  TCCCCGTCCGGGACCCGCGCGGTCGATCCCCGCATATAGGTGGCCA...          25nm   \n",
              "12                  CATCGTGAACTGACCCCTCGTCGGCGAGGTCGG          25nm   \n",
              "13  TCCCCGTCCGGGACCCGCGCGGTCGATCCCCGCATATAGGCAACAT...          25nm   \n",
              "14                GGTAACAGGCTGACCAGTTCCAACATCGCGCTGCC          25nm   \n",
              "15  TCCCCGTCCGGGACCCGCGCGGTCGATCCCCGCATATAGGGCTTGC...          25nm   \n",
              "0                                      GACCGCGCTGGGTG          25nm   \n",
              "1                                  TTGCTGCAACCTGCCTTG          25nm   \n",
              "2                                    ATCCCGCAGCTGCAGG          25nm   \n",
              "3                                      GCCGCTGGCCAAGC          25nm   \n",
              "4                                       TCCGGCCGCATCG          25nm   \n",
              "5                                    TAGGTGCCTCCGTCGA          25nm   \n",
              "6                                GAGGAGCGGCAATGGTCATC          25nm   \n",
              "7                                      AGGCGTGCGCGTGG          25nm   \n",
              "\n",
              "   Purification  \n",
              "0           STD  \n",
              "1           STD  \n",
              "2           STD  \n",
              "3           STD  \n",
              "4           STD  \n",
              "5           STD  \n",
              "6           STD  \n",
              "7           STD  \n",
              "0           STD  \n",
              "1           STD  \n",
              "2           STD  \n",
              "3           STD  \n",
              "4           STD  \n",
              "5           STD  \n",
              "6           STD  \n",
              "7           STD  \n",
              "8           STD  \n",
              "9           STD  \n",
              "10          STD  \n",
              "11          STD  \n",
              "12          STD  \n",
              "13          STD  \n",
              "14          STD  \n",
              "15          STD  \n",
              "0           STD  \n",
              "1           STD  \n",
              "2           STD  \n",
              "3           STD  \n",
              "4           STD  \n",
              "5           STD  \n",
              "6           STD  \n",
              "7           STD  "
            ]
          },
          "execution_count": 17,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "if in_frame_deletion: \n",
        "    full_idt = pd.concat([idt_primers, idt_df, checking_primers_df_idt])\n",
        "else: \n",
        "    full_idt = pd.concat([idt_primers, checking_primers_df_idt])\n",
        "\n",
        "full_idt"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "## Folder with all the generated I/O"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 18,
      "metadata": {},
      "outputs": [],
      "source": [
        "input_files = [\n",
        "    {\"name\": \"input_genome.gb\", \"content\": genome},\n",
        "    {\"name\": \"input_plasmid.gb\", \"content\": clean_plasmid}\n",
        "]\n",
        "\n",
        "if in_frame_deletion: \n",
        "    output_files = [\n",
        "        {\"name\": \"cBEST_w_sgRNAs.gb\", \"content\": assembled_contigs}, # LIST OF Dseqrecords\n",
        "        {\"name\": \"primer_df.csv\", \"content\": primer_df},\n",
        "        {\"name\": \"full_idt.csv\", \"content\": full_idt},\n",
        "        {\"name\": \"sgrna_df.csv\", \"content\": sgrna_df},\n",
        "        {\"name\": \"filtered_df.csv\", \"content\": filtered_df},\n",
        "        {\"name\": \"plasmid_metadata_df.csv\", \"content\": plasmid_metadata_df},\n",
        "\n",
        "    ]\n",
        "else: \n",
        "    output_files = [\n",
        "        {\"name\": \"cBEST_w_sgRNAs.gb\", \"content\": sgRNA_vectors}, # LIST OF Dseqrecords\n",
        "        {\"name\": \"full_idt.csv\", \"content\": full_idt},\n",
        "        {\"name\": \"sgrna_df.csv\", \"content\": sgrna_df},\n",
        "        {\"name\": \"filtered_df.csv\", \"content\": filtered_df},\n",
        "        {\"name\": \"plasmid_metadata_df.csv\", \"content\": plasmid_metadata_df},\n",
        "        {\"name\": \"workflow_order_df.csv\", \"content\": workflow_df},\n",
        "\n",
        "        workflow_df\n",
        "\n",
        "    ]\n",
        "\n",
        "input_values = {\n",
        "    \"genes_to_knockout\": genes_to_KO,\n",
        "\n",
        "\n",
        "    \"filtering_metrics\": {\n",
        "        \"gc_upper\": gc_upper,\n",
        "        \"gc_lower\": gc_lower,\n",
        "        \"off_target_seed\": off_target_seed,\n",
        "        \"off_target_upper\": off_target_upper,\n",
        "        \"cas_type\": cas_type,\n",
        "        \"number_of_sgRNAs_per_group\": number_of_sgRNAs_per_group,\n",
        "\n",
        "    },  \n",
        "    \"polymerase_settings\": {\n",
        "        \"chosen_polymerase\": chosen_polymerase,\n",
        "        \"melting_temperature\": melting_temperature,\n",
        "        \"primer_concentration\": primer_concentration,\n",
        "        \"primer_number_increment\": primer_number_increment,\n",
        "        \"flanking_region\": flanking_region\n",
        "    },\n",
        "    \"overlapping_sequences\": {\n",
        "        \"up_homology\": str(up_homology),\n",
        "        \"dw_homology\": str(dw_homology)\n",
        "    }\n",
        "}\n",
        "\n",
        "\n",
        "# Paths to Markdown files\n",
        "markdown_file_paths = [\n",
        "    \"../../protocols/conjugation_protcol.md\",\n",
        "    \"../../protocols/single_target_crispr_plasmid_protcol.md\"\n",
        "\n",
        "]\n",
        "\n",
        "\n",
        "timestamp = datetime.utcnow().isoformat()\n",
        "\n",
        "# Create project directory structure\n",
        "project_directory = ProjectDirectory(\n",
        "    project_name=f\"CRISPR_cas9_plasmid_workflow_{timestamp}\",\n",
        "    input_files=input_files,\n",
        "    output_files=output_files,\n",
        "    input_values=input_values,\n",
        "    markdown_file_paths=markdown_file_paths\n",
        ")\n",
        "\n",
        "# DO You want to save the folder? \n",
        "save_zip_folder = False \n",
        "\n",
        "if save_zip_folder: \n",
        "    # Generate the project directory structure and get the zip content\n",
        "    zip_content = project_directory.create_directory_structure(create_directories=False)\n",
        "\n",
        "    # Save the zip file to disk (optional)\n",
        "    with open(\"project_structure.zip\", \"wb\") as f:\n",
        "        f.write(zip_content)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {},
      "outputs": [],
      "source": []
    }
  ],
  "metadata": {
    "kernelspec": {
      "display_name": "constrain",
      "language": "python",
      "name": "python3"
    },
    "language_info": {
      "codemirror_mode": {
        "name": "ipython",
        "version": 3
      },
      "file_extension": ".py",
      "mimetype": "text/x-python",
      "name": "python",
      "nbconvert_exporter": "python",
      "pygments_lexer": "ipython3",
      "version": "3.11.9"
    }
  },
  "nbformat": 4,
  "nbformat_minor": 2
}
