{
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "# W5-CRISPR-Cas9-inframe_deletion"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "In Workflow 5, CRISPR-Cas9 combines ssDNA bridging and Gibson cloning to create in-frame deletions (Figure 6A). This ensures that the downstream genes are not disrupted, which is crucial for functional studies of gene knockouts. This workflow can be used to study gene function by knocking out specific genes in streptomycetes and observing the resulting phenotypic changes. It can also be used to create markerless deletions, which are important for constructing clean genetic backgrounds for further genetic manipulations. To get started, download the pCRISPR-Cas9 file and a genome file (e.g., S. coelicolor A3), then select the genes to delete. StreptoAIM will generate the necessary primers and plasmid constructs (S5).\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 1,
      "metadata": {},
      "outputs": [],
      "source": [
        "import sys\n",
        "import os\n",
        "from Bio.Restriction import StuI\n",
        "from pydna.dseqrecord import Dseqrecord\n",
        "import pandas as pd\n",
        "from datetime import datetime\n",
        "\n",
        "\n",
        "# Ensure the src directory is in the Python path\n",
        "project_root = os.path.abspath(os.path.join(os.getcwd(), '../../'))\n",
        "if project_root not in sys.path:\n",
        "    sys.path.insert(0, project_root)\n",
        "\n",
        "from streptocad.sequence_loading.sequence_loading import (\n",
        "    load_and_process_gene_sequences,\n",
        "    load_and_process_genome_sequences,\n",
        "    load_and_process_plasmid, \n",
        "    check_and_convert_input,\n",
        "    annotate_dseqrecord,\n",
        "    process_specified_gene_sequences_from_record)\n",
        "\n",
        "from streptocad.utils import polymerase_dict, create_primer_df_from_dict,ProjectDirectory,extract_metadata_to_dataframe\n",
        "from streptocad.primers.primer_generation import create_idt_order_dataframe\n",
        "from streptocad.cloning.ssDNA_bridging import assemble_plasmids_by_ssDNA_bridging,make_ssDNA_oligos\n",
        "from streptocad.crispr.guideRNAcas3_9_12 import extract_sgRNAs, SgRNAargs\n",
        "from streptocad.cloning.gibson_cloning import (\n",
        "    find_up_dw_repair_templates,\n",
        "    assemble_multiple_plasmids_with_repair_templates_for_deletion,\n",
        "    update_primer_names\n",
        ")\n",
        "from streptocad.primers.primer_generation import checking_primers, primers_to_IDT, find_best_check_primers_from_genome\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "## INPUT"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 2,
      "metadata": {},
      "outputs": [],
      "source": [
        "# Inputs\n",
        "# 1 Add genome of choice (genbank, fasta)\n",
        "path_to_genome = '../../data/genomes/Streptomyces_coelicolor_A3_chromosome.gb'\n",
        "genome = load_and_process_genome_sequences(path_to_genome)[0]\n",
        "\n",
        "# 2 Add plasmid \n",
        "path_to_plasmid = '../../data/plasmids/pCRISPR–Cas9_plasmid_addgene.gbk'\n",
        "clean_plasmid = load_and_process_plasmid(path_to_plasmid)\n",
        "\n",
        "# 3 Choose genes to knock out (list)\n",
        "#genes_to_KO = ['SCO5087', 'SCO5089', 'SCO5090','SCO5091', 'SCO5092' ]\n",
        "genes_to_KO = ['80000-100000', '4000-7000', '9000-14000','15000-20000']\n",
        "\n",
        "\n",
        "#### Advanced settings ####\n",
        "# 4 Filtering metrics for sgRNAs\n",
        "gc_upper = 0.72\n",
        "gc_lower = 0.2\n",
        "off_target_seed = 13\n",
        "off_target_upper = 10\n",
        "cas_type='cas9'\n",
        "number_of_sgRNAs_per_group = 2\n",
        "\n",
        "# 5 Choose polymerase and target melting temperature\n",
        "chosen_polymerase = polymerase_dict['Phusion High-Fidelity DNA Polymerase (GC Buffer)']\n",
        "\n",
        "\n",
        "melting_temperature = 65\n",
        "primer_concentration = 0.4 \n",
        "primer_number_increment = 23\n",
        "flanking_region = 500\n",
        "\n",
        "# 6 Choose overlapping sequences for our plasmid we can use the following\n",
        "#As per the article **\"CRISPR–Cas9, CRISPRi and CRISPR-BEST-mediated genetic manipulation in streptomycetes\"** we need the following oligoes: \n",
        "#CGGTTGGTAGGATCGACGGC **-N20-** GTTTTAGAGCTAGAAATAGC\n",
        "up_homology = Dseqrecord('CGGTTGGTAGGATCGACGGC')\n",
        "dw_homology = Dseqrecord('GTTTTAGAGCTAGAAATAGC')"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "# Computation"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 3,
      "metadata": {},
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "True\n",
            "this is the target dict [{'region_1_(80000-100000)': [80000, 100000]}, {'region_2_(4000-7000)': [4000, 7000]}, {'region_3_(9000-14000)': [9000, 14000]}, {'region_4_(15000-20000)': [15000, 20000]}]\n",
            "region_1_(80000-100000) [80000, 100000]\n",
            "region_2_(4000-7000) [4000, 7000]\n",
            "region_3_(9000-14000) [9000, 14000]\n",
            "region_4_(15000-20000) [15000, 20000]\n"
          ]
        },
        {
          "data": {
            "text/plain": [
              "25828"
            ]
          },
          "execution_count": 3,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "target_dict, genes_to_KO, annotation_input = check_and_convert_input(genes_to_KO)\n",
        "\n",
        "print(annotation_input)\n",
        "if annotation_input == True:\n",
        "    genome = annotate_dseqrecord(genome, target_dict)\n",
        "\n",
        "\n",
        "len(genome.features)\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 4,
      "metadata": {},
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "sgRNA generated were outside the designated border in region_1_(80000-100000). To incorporate this extent borders. Skipping to next locus tag.\n",
            "sgRNA generated were outside the designated border in region_1_(80000-100000). To incorporate this extent borders. Skipping to next locus tag.\n",
            "sgRNA generated were outside the designated border in region_2_(4000-7000). To incorporate this extent borders. Skipping to next locus tag.\n",
            "sgRNA generated were outside the designated border in region_2_(4000-7000). To incorporate this extent borders. Skipping to next locus tag.\n",
            "sgRNA generated were outside the designated border in region_2_(4000-7000). To incorporate this extent borders. Skipping to next locus tag.\n",
            "sgRNA generated were outside the designated border in region_2_(4000-7000). To incorporate this extent borders. Skipping to next locus tag.\n",
            "sgRNA generated were outside the designated border in region_2_(4000-7000). To incorporate this extent borders. Skipping to next locus tag.\n",
            "sgRNA generated were outside the designated border in region_2_(4000-7000). To incorporate this extent borders. Skipping to next locus tag.\n",
            "sgRNA generated were outside the designated border in region_3_(9000-14000). To incorporate this extent borders. Skipping to next locus tag.\n",
            "sgRNA generated were outside the designated border in region_3_(9000-14000). To incorporate this extent borders. Skipping to next locus tag.\n",
            "sgRNA generated were outside the designated border in region_3_(9000-14000). To incorporate this extent borders. Skipping to next locus tag.\n",
            "sgRNA generated were outside the designated border in region_3_(9000-14000). To incorporate this extent borders. Skipping to next locus tag.\n",
            "sgRNA generated were outside the designated border in region_3_(9000-14000). To incorporate this extent borders. Skipping to next locus tag.\n",
            "sgRNA generated were outside the designated border in region_4_(15000-20000). To incorporate this extent borders. Skipping to next locus tag.\n",
            "sgRNA generated were outside the designated border in region_4_(15000-20000). To incorporate this extent borders. Skipping to next locus tag.\n",
            "sgRNA generated were outside the designated border in region_4_(15000-20000). To incorporate this extent borders. Skipping to next locus tag.\n",
            "sgRNA generated were outside the designated border in region_4_(15000-20000). To incorporate this extent borders. Skipping to next locus tag.\n",
            "sgRNA generated were outside the designated border in region_4_(15000-20000). To incorporate this extent borders. Skipping to next locus tag.\n",
            "sgRNA generated were outside the designated border in region_4_(15000-20000). To incorporate this extent borders. Skipping to next locus tag.\n"
          ]
        },
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>strain_name</th>\n",
              "      <th>locus_tag</th>\n",
              "      <th>gene_loc</th>\n",
              "      <th>gene_strand</th>\n",
              "      <th>sgrna_strand</th>\n",
              "      <th>sgrna_loc</th>\n",
              "      <th>gc</th>\n",
              "      <th>pam</th>\n",
              "      <th>sgrna</th>\n",
              "      <th>sgrna_seed_sequence</th>\n",
              "      <th>off_target_count</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>3517</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>region_1_(80000-100000)</td>\n",
              "      <td>80001</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>6574</td>\n",
              "      <td>0.60</td>\n",
              "      <td>TGG</td>\n",
              "      <td>GCCACATTCACTACGACCCC</td>\n",
              "      <td>TCACTACGACCCC</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2538</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>region_1_(80000-100000)</td>\n",
              "      <td>80001</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>16046</td>\n",
              "      <td>0.65</td>\n",
              "      <td>GGG</td>\n",
              "      <td>CCCTGATCCGAGTGTGTGCC</td>\n",
              "      <td>CCGAGTGTGTGCC</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2539</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>region_1_(80000-100000)</td>\n",
              "      <td>80001</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>16045</td>\n",
              "      <td>0.60</td>\n",
              "      <td>CGG</td>\n",
              "      <td>ACCCTGATCCGAGTGTGTGC</td>\n",
              "      <td>TCCGAGTGTGTGC</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2540</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>region_1_(80000-100000)</td>\n",
              "      <td>80001</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>16018</td>\n",
              "      <td>0.55</td>\n",
              "      <td>GGG</td>\n",
              "      <td>ACGACAATCCGGTGATCAGC</td>\n",
              "      <td>TCCGGTGATCAGC</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2541</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>region_1_(80000-100000)</td>\n",
              "      <td>80001</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>16017</td>\n",
              "      <td>0.50</td>\n",
              "      <td>CGG</td>\n",
              "      <td>TACGACAATCCGGTGATCAG</td>\n",
              "      <td>ATCCGGTGATCAG</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>...</th>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>191</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>region_1_(80000-100000)</td>\n",
              "      <td>80001</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>1988</td>\n",
              "      <td>0.65</td>\n",
              "      <td>CGG</td>\n",
              "      <td>ATCCTCGCCGTGACCGTCAC</td>\n",
              "      <td>CCGTGACCGTCAC</td>\n",
              "      <td>7</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>172</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>region_1_(80000-100000)</td>\n",
              "      <td>80001</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>1765</td>\n",
              "      <td>0.65</td>\n",
              "      <td>TGG</td>\n",
              "      <td>TGGTAGAGCGCAGCATCGGC</td>\n",
              "      <td>GCGCAGCATCGGC</td>\n",
              "      <td>8</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3954</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>region_1_(80000-100000)</td>\n",
              "      <td>80001</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>2235</td>\n",
              "      <td>0.70</td>\n",
              "      <td>GGG</td>\n",
              "      <td>AGTAGACGGTGCCGTGGCCG</td>\n",
              "      <td>GGTGCCGTGGCCG</td>\n",
              "      <td>8</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>179</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>region_1_(80000-100000)</td>\n",
              "      <td>80001</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>1842</td>\n",
              "      <td>0.70</td>\n",
              "      <td>AGG</td>\n",
              "      <td>CACCCTGGGCATCGACGTCG</td>\n",
              "      <td>GGCATCGACGTCG</td>\n",
              "      <td>8</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3996</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>region_1_(80000-100000)</td>\n",
              "      <td>80001</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>1929</td>\n",
              "      <td>0.65</td>\n",
              "      <td>AGG</td>\n",
              "      <td>TCCTCTTGGCCTGGTCGAGG</td>\n",
              "      <td>GGCCTGGTCGAGG</td>\n",
              "      <td>9</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>4297 rows × 11 columns</p>\n",
              "</div>"
            ],
            "text/plain": [
              "      strain_name                locus_tag  gene_loc  gene_strand  \\\n",
              "3517  NC_003888.3  region_1_(80000-100000)     80001            1   \n",
              "2538  NC_003888.3  region_1_(80000-100000)     80001            1   \n",
              "2539  NC_003888.3  region_1_(80000-100000)     80001            1   \n",
              "2540  NC_003888.3  region_1_(80000-100000)     80001            1   \n",
              "2541  NC_003888.3  region_1_(80000-100000)     80001            1   \n",
              "...           ...                      ...       ...          ...   \n",
              "191   NC_003888.3  region_1_(80000-100000)     80001            1   \n",
              "172   NC_003888.3  region_1_(80000-100000)     80001            1   \n",
              "3954  NC_003888.3  region_1_(80000-100000)     80001            1   \n",
              "179   NC_003888.3  region_1_(80000-100000)     80001            1   \n",
              "3996  NC_003888.3  region_1_(80000-100000)     80001            1   \n",
              "\n",
              "      sgrna_strand  sgrna_loc    gc  pam                 sgrna  \\\n",
              "3517             1       6574  0.60  TGG  GCCACATTCACTACGACCCC   \n",
              "2538             1      16046  0.65  GGG  CCCTGATCCGAGTGTGTGCC   \n",
              "2539             1      16045  0.60  CGG  ACCCTGATCCGAGTGTGTGC   \n",
              "2540             1      16018  0.55  GGG  ACGACAATCCGGTGATCAGC   \n",
              "2541             1      16017  0.50  CGG  TACGACAATCCGGTGATCAG   \n",
              "...            ...        ...   ...  ...                   ...   \n",
              "191             -1       1988  0.65  CGG  ATCCTCGCCGTGACCGTCAC   \n",
              "172             -1       1765  0.65  TGG  TGGTAGAGCGCAGCATCGGC   \n",
              "3954             1       2235  0.70  GGG  AGTAGACGGTGCCGTGGCCG   \n",
              "179             -1       1842  0.70  AGG  CACCCTGGGCATCGACGTCG   \n",
              "3996             1       1929  0.65  AGG  TCCTCTTGGCCTGGTCGAGG   \n",
              "\n",
              "     sgrna_seed_sequence  off_target_count  \n",
              "3517       TCACTACGACCCC                 0  \n",
              "2538       CCGAGTGTGTGCC                 0  \n",
              "2539       TCCGAGTGTGTGC                 0  \n",
              "2540       TCCGGTGATCAGC                 0  \n",
              "2541       ATCCGGTGATCAG                 0  \n",
              "...                  ...               ...  \n",
              "191        CCGTGACCGTCAC                 7  \n",
              "172        GCGCAGCATCGGC                 8  \n",
              "3954       GGTGCCGTGGCCG                 8  \n",
              "179        GGCATCGACGTCG                 8  \n",
              "3996       GGCCTGGTCGAGG                 9  \n",
              "\n",
              "[4297 rows x 11 columns]"
            ]
          },
          "execution_count": 4,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "\n",
        "# Initialize SgRNAargs with desired parameters\n",
        "args = SgRNAargs(genome, \n",
        "                genes_to_KO,\n",
        "                step=['find', 'filter'],\n",
        "                gc_upper = gc_upper,\n",
        "                gc_lower = gc_lower,\n",
        "                off_target_seed = off_target_seed,\n",
        "                off_target_upper = off_target_upper,\n",
        "                cas_type='cas9'\n",
        "                )\n",
        "\n",
        "sgrna_df = extract_sgRNAs(args)\n",
        "sgrna_df"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 5,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>strain_name</th>\n",
              "      <th>locus_tag</th>\n",
              "      <th>gene_loc</th>\n",
              "      <th>gene_strand</th>\n",
              "      <th>sgrna_strand</th>\n",
              "      <th>sgrna_loc</th>\n",
              "      <th>gc</th>\n",
              "      <th>pam</th>\n",
              "      <th>sgrna</th>\n",
              "      <th>sgrna_seed_sequence</th>\n",
              "      <th>off_target_count</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>3517</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>region_1_(80000-100000)</td>\n",
              "      <td>80001</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>6574</td>\n",
              "      <td>0.60</td>\n",
              "      <td>TGG</td>\n",
              "      <td>GCCACATTCACTACGACCCC</td>\n",
              "      <td>TCACTACGACCCC</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2538</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>region_1_(80000-100000)</td>\n",
              "      <td>80001</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>16046</td>\n",
              "      <td>0.65</td>\n",
              "      <td>GGG</td>\n",
              "      <td>CCCTGATCCGAGTGTGTGCC</td>\n",
              "      <td>CCGAGTGTGTGCC</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>5691</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>region_3_(9000-14000)</td>\n",
              "      <td>9001</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>3477</td>\n",
              "      <td>0.70</td>\n",
              "      <td>GGG</td>\n",
              "      <td>GCCGGTCGGCCATATCCCCT</td>\n",
              "      <td>GGCCATATCCCCT</td>\n",
              "      <td>1</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>7003</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>region_4_(15000-20000)</td>\n",
              "      <td>15001</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>265</td>\n",
              "      <td>0.65</td>\n",
              "      <td>AGG</td>\n",
              "      <td>ACGGCGTACAGGTCGCGTAC</td>\n",
              "      <td>ACAGGTCGCGTAC</td>\n",
              "      <td>1</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4772</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>region_2_(4000-7000)</td>\n",
              "      <td>4001</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>346</td>\n",
              "      <td>0.70</td>\n",
              "      <td>TGG</td>\n",
              "      <td>GTCCCTCGACTCACTGGCGC</td>\n",
              "      <td>GACTCACTGGCGC</td>\n",
              "      <td>1</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>5251</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>region_3_(9000-14000)</td>\n",
              "      <td>9001</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>2460</td>\n",
              "      <td>0.40</td>\n",
              "      <td>TGG</td>\n",
              "      <td>TAGTTGTGTGGGATTGATCT</td>\n",
              "      <td>GTGGGATTGATCT</td>\n",
              "      <td>1</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4773</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>region_2_(4000-7000)</td>\n",
              "      <td>4001</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>340</td>\n",
              "      <td>0.65</td>\n",
              "      <td>TGG</td>\n",
              "      <td>GTGCTGGTCCCTCGACTCAC</td>\n",
              "      <td>TCCCTCGACTCAC</td>\n",
              "      <td>1</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>7004</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>region_4_(15000-20000)</td>\n",
              "      <td>15001</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>254</td>\n",
              "      <td>0.65</td>\n",
              "      <td>AGG</td>\n",
              "      <td>TCACGTCACCGACGGCGTAC</td>\n",
              "      <td>ACCGACGGCGTAC</td>\n",
              "      <td>1</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "      strain_name                locus_tag  gene_loc  gene_strand  \\\n",
              "3517  NC_003888.3  region_1_(80000-100000)     80001            1   \n",
              "2538  NC_003888.3  region_1_(80000-100000)     80001            1   \n",
              "5691  NC_003888.3    region_3_(9000-14000)      9001            1   \n",
              "7003  NC_003888.3   region_4_(15000-20000)     15001            1   \n",
              "4772  NC_003888.3     region_2_(4000-7000)      4001            1   \n",
              "5251  NC_003888.3    region_3_(9000-14000)      9001            1   \n",
              "4773  NC_003888.3     region_2_(4000-7000)      4001            1   \n",
              "7004  NC_003888.3   region_4_(15000-20000)     15001            1   \n",
              "\n",
              "      sgrna_strand  sgrna_loc    gc  pam                 sgrna  \\\n",
              "3517             1       6574  0.60  TGG  GCCACATTCACTACGACCCC   \n",
              "2538             1      16046  0.65  GGG  CCCTGATCCGAGTGTGTGCC   \n",
              "5691             1       3477  0.70  GGG  GCCGGTCGGCCATATCCCCT   \n",
              "7003             1        265  0.65  AGG  ACGGCGTACAGGTCGCGTAC   \n",
              "4772             1        346  0.70  TGG  GTCCCTCGACTCACTGGCGC   \n",
              "5251            -1       2460  0.40  TGG  TAGTTGTGTGGGATTGATCT   \n",
              "4773             1        340  0.65  TGG  GTGCTGGTCCCTCGACTCAC   \n",
              "7004             1        254  0.65  AGG  TCACGTCACCGACGGCGTAC   \n",
              "\n",
              "     sgrna_seed_sequence  off_target_count  \n",
              "3517       TCACTACGACCCC                 0  \n",
              "2538       CCGAGTGTGTGCC                 0  \n",
              "5691       GGCCATATCCCCT                 1  \n",
              "7003       ACAGGTCGCGTAC                 1  \n",
              "4772       GACTCACTGGCGC                 1  \n",
              "5251       GTGGGATTGATCT                 1  \n",
              "4773       TCCCTCGACTCAC                 1  \n",
              "7004       ACCGACGGCGTAC                 1  "
            ]
          },
          "execution_count": 5,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "# Filter the DataFrame to retain only up to 5 sgRNA sequences per locus_tag\n",
        "filtered_df = sgrna_df.groupby('locus_tag').head(number_of_sgRNAs_per_group)\n",
        "filtered_df"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "## Output"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 6,
      "metadata": {},
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "region_1_(80000-100000)_loc_6574\n"
          ]
        },
        {
          "data": {
            "text/plain": [
              "[Contig(o11279),\n",
              " Contig(o11279),\n",
              " Contig(o11279),\n",
              " Contig(o11279),\n",
              " Contig(o11279),\n",
              " Contig(o11279),\n",
              " Contig(o11279),\n",
              " Contig(o11279)]"
            ]
          },
          "execution_count": 6,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "# MAke oligoes\n",
        "list_of_ssDNAs = make_ssDNA_oligos(filtered_df, upstream_ovh = up_homology,\n",
        "                      downstream_ovh=dw_homology)\n",
        "print(list_of_ssDNAs[0].name)\n",
        "\n",
        "# cut plasmid\n",
        "from Bio.Restriction import NcoI\n",
        "linearized_plasmid = sorted(clean_plasmid.cut(NcoI), key=lambda x: len(x), reverse=True)[0]\n",
        "#print(linearized_plasmid)\n",
        "\n",
        "sgRNA_vectors = assemble_plasmids_by_ssDNA_bridging(list_of_ssDNAs,linearized_plasmid)\n",
        "sgRNA_vectors"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 7,
      "metadata": {},
      "outputs": [],
      "source": [
        "\n",
        "# Constructing a meaningful name, ID, and description for the assembled plasmid using user input\n",
        "targeting_info = []\n",
        "for index, row in filtered_df.iterrows():\n",
        "    formatted_str = f\"pCas9_{row['locus_tag']}({row['sgrna_loc']})\"\n",
        "    targeting_info.append(formatted_str)\n",
        "\n",
        "for i in range(len(sgRNA_vectors)):\n",
        "    sgRNA_vectors[i].name = f'{targeting_info[i]}'\n",
        "    sgRNA_vectors[i].id = sgRNA_vectors[i].name  \n",
        "    sgRNA_vectors[i].description = f'CRISPR-Cas9 targeting {\", \".join(genes_to_KO)} for single gene knockout, assembled using StreptoCAD.'\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 8,
      "metadata": {},
      "outputs": [],
      "source": [
        "print_plasmids = False\n",
        "\n",
        "if print_plasmids: \n",
        "    for vector in sgRNA_vectors: \n",
        "        vector.write(f\"../../data/plasmids/sgRNA_plasmids_pCRISPR-Cas9/{vector.id}.gb\")"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "### IDT primers"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 9,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>Name</th>\n",
              "      <th>Sequence</th>\n",
              "      <th>Concentration</th>\n",
              "      <th>Purification</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>region_1_(80000-100000)_loc_6574</td>\n",
              "      <td>CGGTTGGTAGGATCGACGGCGCCACATTCACTACGACCCCGTTTTA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>region_1_(80000-100000)_loc_16046</td>\n",
              "      <td>CGGTTGGTAGGATCGACGGCCCCTGATCCGAGTGTGTGCCGTTTTA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>region_3_(9000-14000)_loc_3477</td>\n",
              "      <td>CGGTTGGTAGGATCGACGGCGCCGGTCGGCCATATCCCCTGTTTTA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>region_4_(15000-20000)_loc_265</td>\n",
              "      <td>CGGTTGGTAGGATCGACGGCACGGCGTACAGGTCGCGTACGTTTTA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>region_2_(4000-7000)_loc_346</td>\n",
              "      <td>CGGTTGGTAGGATCGACGGCGTCCCTCGACTCACTGGCGCGTTTTA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>5</th>\n",
              "      <td>region_3_(9000-14000)_loc_2460</td>\n",
              "      <td>CGGTTGGTAGGATCGACGGCTAGTTGTGTGGGATTGATCTGTTTTA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>6</th>\n",
              "      <td>region_2_(4000-7000)_loc_340</td>\n",
              "      <td>CGGTTGGTAGGATCGACGGCGTGCTGGTCCCTCGACTCACGTTTTA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>7</th>\n",
              "      <td>region_4_(15000-20000)_loc_254</td>\n",
              "      <td>CGGTTGGTAGGATCGACGGCTCACGTCACCGACGGCGTACGTTTTA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "                                Name  \\\n",
              "0   region_1_(80000-100000)_loc_6574   \n",
              "1  region_1_(80000-100000)_loc_16046   \n",
              "2     region_3_(9000-14000)_loc_3477   \n",
              "3     region_4_(15000-20000)_loc_265   \n",
              "4       region_2_(4000-7000)_loc_346   \n",
              "5     region_3_(9000-14000)_loc_2460   \n",
              "6       region_2_(4000-7000)_loc_340   \n",
              "7     region_4_(15000-20000)_loc_254   \n",
              "\n",
              "                                            Sequence Concentration  \\\n",
              "0  CGGTTGGTAGGATCGACGGCGCCACATTCACTACGACCCCGTTTTA...          25nm   \n",
              "1  CGGTTGGTAGGATCGACGGCCCCTGATCCGAGTGTGTGCCGTTTTA...          25nm   \n",
              "2  CGGTTGGTAGGATCGACGGCGCCGGTCGGCCATATCCCCTGTTTTA...          25nm   \n",
              "3  CGGTTGGTAGGATCGACGGCACGGCGTACAGGTCGCGTACGTTTTA...          25nm   \n",
              "4  CGGTTGGTAGGATCGACGGCGTCCCTCGACTCACTGGCGCGTTTTA...          25nm   \n",
              "5  CGGTTGGTAGGATCGACGGCTAGTTGTGTGGGATTGATCTGTTTTA...          25nm   \n",
              "6  CGGTTGGTAGGATCGACGGCGTGCTGGTCCCTCGACTCACGTTTTA...          25nm   \n",
              "7  CGGTTGGTAGGATCGACGGCTCACGTCACCGACGGCGTACGTTTTA...          25nm   \n",
              "\n",
              "  Purification  \n",
              "0          STD  \n",
              "1          STD  \n",
              "2          STD  \n",
              "3          STD  \n",
              "4          STD  \n",
              "5          STD  \n",
              "6          STD  \n",
              "7          STD  "
            ]
          },
          "execution_count": 9,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "idt_primers=primers_to_IDT(list_of_ssDNAs)\n",
        "idt_primers"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "# In frame deletion"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 10,
      "metadata": {},
      "outputs": [],
      "source": [
        "in_frame_deletion = False"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 11,
      "metadata": {},
      "outputs": [],
      "source": [
        "if in_frame_deletion:\n",
        "    # Make repair templates\n",
        "    repair_templates_data = find_up_dw_repair_templates(genome, \n",
        "                                                        genes_to_KO, \n",
        "                                                        target_tm=melting_temperature, \n",
        "                                                        \n",
        "                                                        primer_tm_kwargs={'conc':primer_concentration, 'prodcode':chosen_polymerase} )\n",
        "\n",
        "\n",
        "    # Digest the plasmids\n",
        "    processed_records = [Dseqrecord(record, circular=True).cut(StuI)[0] for record in sgRNA_vectors]\n",
        "\n",
        "    # Rename them appropriately\n",
        "    for i in range(len(processed_records)):\n",
        "        processed_records[i].name = sgRNA_vectors[i].name\n",
        "\n",
        "    # Assembly \n",
        "    assembly_data = assemble_multiple_plasmids_with_repair_templates_for_deletion(genes_to_KO, processed_records, \n",
        "                                                                                repair_templates_data, \n",
        "                                                                                overlap=40)\n",
        "    # updating the primer names to something systematic.\n",
        "    update_primer_names(assembly_data)\n",
        "    print(assembly_data)\n",
        "\n",
        "    # Parse through the primer df\n",
        "    primer_df = create_primer_df_from_dict(assembly_data)\n",
        "\n",
        "    # Digest the plasmids\n",
        "    processed_records = [Dseqrecord(record, circular=True).cut(StuI)[0] for record in sgRNA_vectors]\n",
        "\n",
        "    # Rename them appropriately\n",
        "    for i in range(len(processed_records)):\n",
        "        processed_records[i].name = sgRNA_vectors[i].name\n",
        "\n",
        "    # Assembly \n",
        "    ##### THE PROBLEM IS HERE\n",
        "    assembly_data = assemble_multiple_plasmids_with_repair_templates_for_deletion(genes_to_KO, processed_records, \n",
        "                                                                                repair_templates_data, \n",
        "                                                                                overlap=40)\n",
        "    # updating the primer names to something systematic.\n",
        "    update_primer_names(assembly_data)\n",
        "    print(assembly_data)\n",
        "\n",
        "    # Parse through the primer df\n",
        "    primer_df = create_primer_df_from_dict(assembly_data)\n",
        "    unique_df = primer_df.drop_duplicates(keep='first')\n",
        "\n",
        "    # IDT df\n",
        "    idt_df = create_idt_order_dataframe(unique_df, concentration=\"25nm\", purification=\"STD\")\n",
        "\n",
        "    # Contigs\n",
        "    assembled_contigs = []\n",
        "    for data in assembly_data: \n",
        "        contig_record = data['contig']\n",
        "        contig_record.id = f\"{data['name']}_w_rep\"\n",
        "        contig_record.name = f\"{data['name']}_w_rep\"\n",
        "        assembled_contigs.append(contig_record)\n",
        "\n",
        "\n",
        "    print_plasmids = False\n",
        "\n",
        "    if print_plasmids:\n",
        "        # Write to genbank files\n",
        "        output_directory = \"../../data/plasmids/sgRNA_plasmids_pCRISPR-Cas9_with_repair_templates/\"\n",
        "        for contig in assembled_contigs: \n",
        "            contig.write(f\"{output_directory}/{contig.id}.gb\")"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 12,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>plasmid_name</th>\n",
              "      <th>date</th>\n",
              "      <th>original_plasmid</th>\n",
              "      <th>integration</th>\n",
              "      <th>size</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>pCas9_region_1_(80000-100000)(6574)</td>\n",
              "      <td>2024-07-23</td>\n",
              "      <td>CRISPR-Cas9</td>\n",
              "      <td>sgRNA_region_1_(80000-100000)(6574)</td>\n",
              "      <td>11279</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>pCas9_region_1_(80000-100000)(16046)</td>\n",
              "      <td>2024-07-23</td>\n",
              "      <td>CRISPR-Cas9</td>\n",
              "      <td>sgRNA_region_1_(80000-100000)(16046)</td>\n",
              "      <td>11279</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>pCas9_region_3_(9000-14000)(3477)</td>\n",
              "      <td>2024-07-23</td>\n",
              "      <td>CRISPR-Cas9</td>\n",
              "      <td>sgRNA_region_3_(9000-14000)(3477)</td>\n",
              "      <td>11279</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>pCas9_region_4_(15000-20000)(265)</td>\n",
              "      <td>2024-07-23</td>\n",
              "      <td>CRISPR-Cas9</td>\n",
              "      <td>sgRNA_region_4_(15000-20000)(265)</td>\n",
              "      <td>11279</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>pCas9_region_2_(4000-7000)(346)</td>\n",
              "      <td>2024-07-23</td>\n",
              "      <td>CRISPR-Cas9</td>\n",
              "      <td>sgRNA_region_2_(4000-7000)(346)</td>\n",
              "      <td>11279</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>5</th>\n",
              "      <td>pCas9_region_3_(9000-14000)(2460)</td>\n",
              "      <td>2024-07-23</td>\n",
              "      <td>CRISPR-Cas9</td>\n",
              "      <td>sgRNA_region_3_(9000-14000)(2460)</td>\n",
              "      <td>11279</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>6</th>\n",
              "      <td>pCas9_region_2_(4000-7000)(340)</td>\n",
              "      <td>2024-07-23</td>\n",
              "      <td>CRISPR-Cas9</td>\n",
              "      <td>sgRNA_region_2_(4000-7000)(340)</td>\n",
              "      <td>11279</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>7</th>\n",
              "      <td>pCas9_region_4_(15000-20000)(254)</td>\n",
              "      <td>2024-07-23</td>\n",
              "      <td>CRISPR-Cas9</td>\n",
              "      <td>sgRNA_region_4_(15000-20000)(254)</td>\n",
              "      <td>11279</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "                           plasmid_name        date original_plasmid  \\\n",
              "0   pCas9_region_1_(80000-100000)(6574)  2024-07-23      CRISPR-Cas9   \n",
              "1  pCas9_region_1_(80000-100000)(16046)  2024-07-23      CRISPR-Cas9   \n",
              "2     pCas9_region_3_(9000-14000)(3477)  2024-07-23      CRISPR-Cas9   \n",
              "3     pCas9_region_4_(15000-20000)(265)  2024-07-23      CRISPR-Cas9   \n",
              "4       pCas9_region_2_(4000-7000)(346)  2024-07-23      CRISPR-Cas9   \n",
              "5     pCas9_region_3_(9000-14000)(2460)  2024-07-23      CRISPR-Cas9   \n",
              "6       pCas9_region_2_(4000-7000)(340)  2024-07-23      CRISPR-Cas9   \n",
              "7     pCas9_region_4_(15000-20000)(254)  2024-07-23      CRISPR-Cas9   \n",
              "\n",
              "                            integration   size  \n",
              "0   sgRNA_region_1_(80000-100000)(6574)  11279  \n",
              "1  sgRNA_region_1_(80000-100000)(16046)  11279  \n",
              "2     sgRNA_region_3_(9000-14000)(3477)  11279  \n",
              "3     sgRNA_region_4_(15000-20000)(265)  11279  \n",
              "4       sgRNA_region_2_(4000-7000)(346)  11279  \n",
              "5     sgRNA_region_3_(9000-14000)(2460)  11279  \n",
              "6       sgRNA_region_2_(4000-7000)(340)  11279  \n",
              "7     sgRNA_region_4_(15000-20000)(254)  11279  "
            ]
          },
          "execution_count": 12,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "if in_frame_deletion:\n",
        "    integration_names = filtered_df.apply(lambda row: f\"sgRNA_{row['locus_tag']}({row['sgrna_loc']})\", axis=1).tolist()\n",
        "    plasmid_metadata_df = extract_metadata_to_dataframe(assembled_contigs,\n",
        "                                                        'CRISPR-Cas9',\n",
        "                                                        integration_names)\n",
        "else: \n",
        "    integration_names = filtered_df.apply(lambda row: f\"sgRNA_{row['locus_tag']}({row['sgrna_loc']})\", axis=1).tolist()\n",
        "    plasmid_metadata_df = extract_metadata_to_dataframe(sgRNA_vectors,\n",
        "                                                        clean_plasmid,\n",
        "                                                        integration_names)\n",
        "\n",
        "plasmid_metadata_df"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 13,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>locus tag</th>\n",
              "      <th>f_primer_name</th>\n",
              "      <th>r_primer_name</th>\n",
              "      <th>f_primer_sequences(5-3)</th>\n",
              "      <th>r_primer_sequences(5-3)</th>\n",
              "      <th>f_tm</th>\n",
              "      <th>r_tm</th>\n",
              "      <th>ta</th>\n",
              "      <th>flanking_region</th>\n",
              "      <th>annealing_temperature</th>\n",
              "      <th>...</th>\n",
              "      <th>homodimer_reverse_tm</th>\n",
              "      <th>homodimer_reverse_deltaG (kcal/mol)</th>\n",
              "      <th>heterodimer_tm</th>\n",
              "      <th>heterodimer_deltaG (kcal/mol)</th>\n",
              "      <th>hairpin_forward_structure_found</th>\n",
              "      <th>hairpin_forward_tm</th>\n",
              "      <th>hairpin_forward_deltaG (kcal/mol)</th>\n",
              "      <th>hairpin_reverse_structure_found</th>\n",
              "      <th>hairpin_reverse_tm</th>\n",
              "      <th>hairpin_reverse_deltaG (kcal/mol)</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>region_4_(15000-20000)</td>\n",
              "      <td>region_4_(15000-20000)_fwd_checking_primer</td>\n",
              "      <td>region_4_(15000-20000)_rev_checking_primer</td>\n",
              "      <td>TTGCTGCAACCTGCCTTG</td>\n",
              "      <td>TAGGTGCCTCCGTCGA</td>\n",
              "      <td>61</td>\n",
              "      <td>58</td>\n",
              "      <td>62</td>\n",
              "      <td>500</td>\n",
              "      <td>62</td>\n",
              "      <td>...</td>\n",
              "      <td>-13.794142</td>\n",
              "      <td>-0.897729</td>\n",
              "      <td>-1.321907</td>\n",
              "      <td>0.948660</td>\n",
              "      <td>False</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>False</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>region_3_(9000-14000)</td>\n",
              "      <td>region_3_(9000-14000)_fwd_checking_primer</td>\n",
              "      <td>region_3_(9000-14000)_rev_checking_primer</td>\n",
              "      <td>ATCCCGCAGCTGCAGG</td>\n",
              "      <td>GAGGAGCGGCAATGGTCATC</td>\n",
              "      <td>61</td>\n",
              "      <td>63</td>\n",
              "      <td>65</td>\n",
              "      <td>500</td>\n",
              "      <td>65</td>\n",
              "      <td>...</td>\n",
              "      <td>1.445721</td>\n",
              "      <td>-1.026208</td>\n",
              "      <td>9.252334</td>\n",
              "      <td>-1.213024</td>\n",
              "      <td>False</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>False</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>region_2_(4000-7000)</td>\n",
              "      <td>region_2_(4000-7000)_fwd_checking_primer</td>\n",
              "      <td>region_2_(4000-7000)_rev_checking_primer</td>\n",
              "      <td>GACCGCGCTGGGTG</td>\n",
              "      <td>TCCGGCCGCATCG</td>\n",
              "      <td>59</td>\n",
              "      <td>57</td>\n",
              "      <td>60</td>\n",
              "      <td>500</td>\n",
              "      <td>60</td>\n",
              "      <td>...</td>\n",
              "      <td>22.848051</td>\n",
              "      <td>-5.018625</td>\n",
              "      <td>6.535424</td>\n",
              "      <td>-0.061707</td>\n",
              "      <td>False</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>False</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>region_1_(80000-100000)</td>\n",
              "      <td>region_1_(80000-100000)_fwd_checking_primer</td>\n",
              "      <td>region_1_(80000-100000)_rev_checking_primer</td>\n",
              "      <td>GCCGCTGGCCAAGC</td>\n",
              "      <td>AGGCGTGCGCGTGG</td>\n",
              "      <td>60</td>\n",
              "      <td>62</td>\n",
              "      <td>63</td>\n",
              "      <td>700</td>\n",
              "      <td>63</td>\n",
              "      <td>...</td>\n",
              "      <td>22.176370</td>\n",
              "      <td>-2.921991</td>\n",
              "      <td>11.333376</td>\n",
              "      <td>-0.877576</td>\n",
              "      <td>False</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>False</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>4 rows × 23 columns</p>\n",
              "</div>"
            ],
            "text/plain": [
              "                 locus tag                                f_primer_name  \\\n",
              "0   region_4_(15000-20000)   region_4_(15000-20000)_fwd_checking_primer   \n",
              "1    region_3_(9000-14000)    region_3_(9000-14000)_fwd_checking_primer   \n",
              "2     region_2_(4000-7000)     region_2_(4000-7000)_fwd_checking_primer   \n",
              "3  region_1_(80000-100000)  region_1_(80000-100000)_fwd_checking_primer   \n",
              "\n",
              "                                 r_primer_name f_primer_sequences(5-3)  \\\n",
              "0   region_4_(15000-20000)_rev_checking_primer      TTGCTGCAACCTGCCTTG   \n",
              "1    region_3_(9000-14000)_rev_checking_primer        ATCCCGCAGCTGCAGG   \n",
              "2     region_2_(4000-7000)_rev_checking_primer          GACCGCGCTGGGTG   \n",
              "3  region_1_(80000-100000)_rev_checking_primer          GCCGCTGGCCAAGC   \n",
              "\n",
              "  r_primer_sequences(5-3) f_tm r_tm  ta flanking_region  \\\n",
              "0        TAGGTGCCTCCGTCGA   61   58  62             500   \n",
              "1    GAGGAGCGGCAATGGTCATC   61   63  65             500   \n",
              "2           TCCGGCCGCATCG   59   57  60             500   \n",
              "3          AGGCGTGCGCGTGG   60   62  63             700   \n",
              "\n",
              "   annealing_temperature  ... homodimer_reverse_tm  \\\n",
              "0                     62  ...           -13.794142   \n",
              "1                     65  ...             1.445721   \n",
              "2                     60  ...            22.848051   \n",
              "3                     63  ...            22.176370   \n",
              "\n",
              "   homodimer_reverse_deltaG (kcal/mol)  heterodimer_tm  \\\n",
              "0                            -0.897729       -1.321907   \n",
              "1                            -1.026208        9.252334   \n",
              "2                            -5.018625        6.535424   \n",
              "3                            -2.921991       11.333376   \n",
              "\n",
              "   heterodimer_deltaG (kcal/mol)  hairpin_forward_structure_found  \\\n",
              "0                       0.948660                            False   \n",
              "1                      -1.213024                            False   \n",
              "2                      -0.061707                            False   \n",
              "3                      -0.877576                            False   \n",
              "\n",
              "   hairpin_forward_tm  hairpin_forward_deltaG (kcal/mol)  \\\n",
              "0                 0.0                                0.0   \n",
              "1                 0.0                                0.0   \n",
              "2                 0.0                                0.0   \n",
              "3                 0.0                                0.0   \n",
              "\n",
              "   hairpin_reverse_structure_found  hairpin_reverse_tm  \\\n",
              "0                            False                 0.0   \n",
              "1                            False                 0.0   \n",
              "2                            False                 0.0   \n",
              "3                            False                 0.0   \n",
              "\n",
              "   hairpin_reverse_deltaG (kcal/mol)  \n",
              "0                                0.0  \n",
              "1                                0.0  \n",
              "2                                0.0  \n",
              "3                                0.0  \n",
              "\n",
              "[4 rows x 23 columns]"
            ]
          },
          "execution_count": 13,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "# Getting checking primers\n",
        "checking_primers_df = find_best_check_primers_from_genome(genome, \n",
        "                                       genes_to_KO, \n",
        "                                       flanking_region=flanking_region,\n",
        "                                       target_tm = melting_temperature, \n",
        "                                        primer_concentration = primer_concentration, \n",
        "                                        polymerase = chosen_polymerase)\n",
        "checking_primers_df\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 14,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>Name</th>\n",
              "      <th>Sequence</th>\n",
              "      <th>Concentration</th>\n",
              "      <th>Purification</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>region_4_(15000-20000)_fwd_checking_primer</td>\n",
              "      <td>TTGCTGCAACCTGCCTTG</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>region_3_(9000-14000)_fwd_checking_primer</td>\n",
              "      <td>ATCCCGCAGCTGCAGG</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>region_2_(4000-7000)_fwd_checking_primer</td>\n",
              "      <td>GACCGCGCTGGGTG</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>region_1_(80000-100000)_fwd_checking_primer</td>\n",
              "      <td>GCCGCTGGCCAAGC</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>region_4_(15000-20000)_rev_checking_primer</td>\n",
              "      <td>TAGGTGCCTCCGTCGA</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>5</th>\n",
              "      <td>region_3_(9000-14000)_rev_checking_primer</td>\n",
              "      <td>GAGGAGCGGCAATGGTCATC</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>6</th>\n",
              "      <td>region_2_(4000-7000)_rev_checking_primer</td>\n",
              "      <td>TCCGGCCGCATCG</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>7</th>\n",
              "      <td>region_1_(80000-100000)_rev_checking_primer</td>\n",
              "      <td>AGGCGTGCGCGTGG</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "                                          Name              Sequence  \\\n",
              "0   region_4_(15000-20000)_fwd_checking_primer    TTGCTGCAACCTGCCTTG   \n",
              "1    region_3_(9000-14000)_fwd_checking_primer      ATCCCGCAGCTGCAGG   \n",
              "2     region_2_(4000-7000)_fwd_checking_primer        GACCGCGCTGGGTG   \n",
              "3  region_1_(80000-100000)_fwd_checking_primer        GCCGCTGGCCAAGC   \n",
              "4   region_4_(15000-20000)_rev_checking_primer      TAGGTGCCTCCGTCGA   \n",
              "5    region_3_(9000-14000)_rev_checking_primer  GAGGAGCGGCAATGGTCATC   \n",
              "6     region_2_(4000-7000)_rev_checking_primer         TCCGGCCGCATCG   \n",
              "7  region_1_(80000-100000)_rev_checking_primer        AGGCGTGCGCGTGG   \n",
              "\n",
              "  Concentration Purification  \n",
              "0          25nm          STD  \n",
              "1          25nm          STD  \n",
              "2          25nm          STD  \n",
              "3          25nm          STD  \n",
              "4          25nm          STD  \n",
              "5          25nm          STD  \n",
              "6          25nm          STD  \n",
              "7          25nm          STD  "
            ]
          },
          "execution_count": 14,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "# making the primers into dseqrecords\n",
        "checking_primers_df_idt = create_idt_order_dataframe(checking_primers_df)\n",
        "checking_primers_df_idt\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 15,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>Name</th>\n",
              "      <th>Sequence</th>\n",
              "      <th>Concentration</th>\n",
              "      <th>Purification</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>region_4_(15000-20000)_fwd_checking_primer</td>\n",
              "      <td>TTGCTGCAACCTGCCTTG</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>region_3_(9000-14000)_fwd_checking_primer</td>\n",
              "      <td>ATCCCGCAGCTGCAGG</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>region_2_(4000-7000)_fwd_checking_primer</td>\n",
              "      <td>GACCGCGCTGGGTG</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>region_1_(80000-100000)_fwd_checking_primer</td>\n",
              "      <td>GCCGCTGGCCAAGC</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>region_4_(15000-20000)_rev_checking_primer</td>\n",
              "      <td>TAGGTGCCTCCGTCGA</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>5</th>\n",
              "      <td>region_3_(9000-14000)_rev_checking_primer</td>\n",
              "      <td>GAGGAGCGGCAATGGTCATC</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>6</th>\n",
              "      <td>region_2_(4000-7000)_rev_checking_primer</td>\n",
              "      <td>TCCGGCCGCATCG</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>7</th>\n",
              "      <td>region_1_(80000-100000)_rev_checking_primer</td>\n",
              "      <td>AGGCGTGCGCGTGG</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "                                          Name              Sequence  \\\n",
              "0   region_4_(15000-20000)_fwd_checking_primer    TTGCTGCAACCTGCCTTG   \n",
              "1    region_3_(9000-14000)_fwd_checking_primer      ATCCCGCAGCTGCAGG   \n",
              "2     region_2_(4000-7000)_fwd_checking_primer        GACCGCGCTGGGTG   \n",
              "3  region_1_(80000-100000)_fwd_checking_primer        GCCGCTGGCCAAGC   \n",
              "4   region_4_(15000-20000)_rev_checking_primer      TAGGTGCCTCCGTCGA   \n",
              "5    region_3_(9000-14000)_rev_checking_primer  GAGGAGCGGCAATGGTCATC   \n",
              "6     region_2_(4000-7000)_rev_checking_primer         TCCGGCCGCATCG   \n",
              "7  region_1_(80000-100000)_rev_checking_primer        AGGCGTGCGCGTGG   \n",
              "\n",
              "  Concentration Purification  \n",
              "0          25nm          STD  \n",
              "1          25nm          STD  \n",
              "2          25nm          STD  \n",
              "3          25nm          STD  \n",
              "4          25nm          STD  \n",
              "5          25nm          STD  \n",
              "6          25nm          STD  \n",
              "7          25nm          STD  "
            ]
          },
          "execution_count": 15,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "if in_frame_deletion: \n",
        "    full_idt = pd.concat([idt_df, checking_primers_df_idt])\n",
        "else: \n",
        "    full_idt = pd.concat([checking_primers_df_idt])\n",
        "\n",
        "full_idt"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "## Folder with all the generated I/O"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 16,
      "metadata": {},
      "outputs": [],
      "source": [
        "input_files = [\n",
        "    {\"name\": \"input_genome.gb\", \"content\": genome},\n",
        "    {\"name\": \"input_plasmid.gb\", \"content\": clean_plasmid}\n",
        "]\n",
        "\n",
        "if in_frame_deletion: \n",
        "    output_files = [\n",
        "        {\"name\": \"cBEST_w_sgRNAs.gb\", \"content\": assembled_contigs}, # LIST OF Dseqrecords\n",
        "        {\"name\": \"primer_df.csv\", \"content\": primer_df},\n",
        "        {\"name\": \"full_idt.csv\", \"content\": full_idt},\n",
        "        {\"name\": \"sgrna_df.csv\", \"content\": sgrna_df},\n",
        "        {\"name\": \"filtered_df.csv\", \"content\": filtered_df},\n",
        "        {\"name\": \"plasmid_metadata_df.csv\", \"content\": plasmid_metadata_df},\n",
        "\n",
        "    ]\n",
        "else: \n",
        "    output_files = [\n",
        "        {\"name\": \"cBEST_w_sgRNAs.gb\", \"content\": sgRNA_vectors}, # LIST OF Dseqrecords\n",
        "        {\"name\": \"full_idt.csv\", \"content\": full_idt},\n",
        "        {\"name\": \"sgrna_df.csv\", \"content\": sgrna_df},\n",
        "        {\"name\": \"filtered_df.csv\", \"content\": filtered_df},\n",
        "        {\"name\": \"plasmid_metadata_df.csv\", \"content\": plasmid_metadata_df},\n",
        "\n",
        "    ]\n",
        "\n",
        "input_values = {\n",
        "    \"genes_to_knockout\": genes_to_KO,\n",
        "\n",
        "\n",
        "    \"filtering_metrics\": {\n",
        "        \"gc_upper\": gc_upper,\n",
        "        \"gc_lower\": gc_lower,\n",
        "        \"off_target_seed\": off_target_seed,\n",
        "        \"off_target_upper\": off_target_upper,\n",
        "        \"cas_type\": cas_type,\n",
        "        \"number_of_sgRNAs_per_group\": number_of_sgRNAs_per_group,\n",
        "\n",
        "    },  \n",
        "    \"polymerase_settings\": {\n",
        "        \"chosen_polymerase\": chosen_polymerase,\n",
        "        \"melting_temperature\": melting_temperature,\n",
        "        \"primer_concentration\": primer_concentration,\n",
        "        \"primer_number_increment\": primer_number_increment,\n",
        "        \"flanking_region\": flanking_region\n",
        "    },\n",
        "    \"overlapping_sequences\": {\n",
        "        \"up_homology\": str(up_homology),\n",
        "        \"dw_homology\": str(dw_homology)\n",
        "    }\n",
        "}\n",
        "\n",
        "\n",
        "# Paths to Markdown files\n",
        "markdown_file_paths = [\n",
        "    \"../../protocols/conjugation_protcol.md\",\n",
        "    \"../../protocols/single_target_crispr_plasmid_protcol.md\"\n",
        "\n",
        "]\n",
        "\n",
        "\n",
        "timestamp = datetime.utcnow().isoformat()\n",
        "\n",
        "# Create project directory structure\n",
        "project_directory = ProjectDirectory(\n",
        "    project_name=f\"CRISPR_cas9_plasmid_workflow_{timestamp}\",\n",
        "    input_files=input_files,\n",
        "    output_files=output_files,\n",
        "    input_values=input_values,\n",
        "    markdown_file_paths=markdown_file_paths\n",
        ")\n",
        "\n",
        "# DO You want to save the folder? \n",
        "save_zip_folder = False \n",
        "\n",
        "if save_zip_folder: \n",
        "    # Generate the project directory structure and get the zip content\n",
        "    zip_content = project_directory.create_directory_structure(create_directories=False)\n",
        "\n",
        "    # Save the zip file to disk (optional)\n",
        "    with open(\"project_structure.zip\", \"wb\") as f:\n",
        "        f.write(zip_content)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {},
      "outputs": [],
      "source": []
    }
  ],
  "metadata": {
    "kernelspec": {
      "display_name": "constrain",
      "language": "python",
      "name": "python3"
    },
    "language_info": {
      "codemirror_mode": {
        "name": "ipython",
        "version": 3
      },
      "file_extension": ".py",
      "mimetype": "text/x-python",
      "name": "python",
      "nbconvert_exporter": "python",
      "pygments_lexer": "ipython3",
      "version": "3.11.9"
    }
  },
  "nbformat": 4,
  "nbformat_minor": 2
}
