{
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "# W5-CRISPR-Cas9-inframe_deletion"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "In Workflow 5, CRISPR-Cas9 combines ssDNA bridging and Gibson cloning to create in-frame deletions (Figure 6A). This ensures that the downstream genes are not disrupted, which is crucial for functional studies of gene knockouts. This workflow can be used to study gene function by knocking out specific genes in streptomycetes and observing the resulting phenotypic changes. It can also be used to create markerless deletions, which are important for constructing clean genetic backgrounds for further genetic manipulations. To get started, download the pCRISPR-Cas9 file and a genome file (e.g., S. coelicolor A3), then select the genes to delete. StreptoAIM will generate the necessary primers and plasmid constructs (S5).\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 1,
      "metadata": {},
      "outputs": [],
      "source": [
        "import sys\n",
        "import os\n",
        "from Bio.Restriction import StuI\n",
        "from pydna.dseqrecord import Dseqrecord\n",
        "import pandas as pd\n",
        "from datetime import datetime\n",
        "\n",
        "\n",
        "# Ensure the src directory is in the Python path\n",
        "project_root = os.path.abspath(os.path.join(os.getcwd(), '../../'))\n",
        "if project_root not in sys.path:\n",
        "    sys.path.insert(0, project_root)\n",
        "\n",
        "from streptocad.sequence_loading.sequence_loading import (\n",
        "    load_and_process_plasmid,\n",
        "    load_and_process_genome_sequences\n",
        ")\n",
        "from streptocad.utils import polymerase_dict, create_primer_df_from_dict,generate_project_directory_structure\n",
        "from streptocad.primers.primer_generation import create_idt_order_dataframe\n",
        "from streptocad.cloning.ssDNA_bridging import assemble_plasmids_by_ssDNA_bridging,make_ssDNA_oligos\n",
        "from streptocad.crispr.guideRNAcas3_9_12 import extract_sgRNAs, SgRNAargs\n",
        "from streptocad.cloning.gibson_cloning import (\n",
        "    find_up_dw_repair_templates,\n",
        "    assemble_multiple_plasmids_with_repair_templates_for_deletion,\n",
        "    update_primer_names\n",
        ")\n",
        "from streptocad.primers.primer_generation import create_dseqrecords_from_df,checking_primers, primers_to_IDT\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "## INPUT"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 2,
      "metadata": {},
      "outputs": [],
      "source": [
        "# Inputs\n",
        "# 1 Add genome of choice (genbank, fasta)\n",
        "path_to_genome = '../../data/genomes/Streptomyces_coelicolor_A3_chromosome.gb'\n",
        "genome = load_and_process_genome_sequences(path_to_genome)[0]\n",
        "\n",
        "# 2 Add plasmid \n",
        "path_to_plasmid = '../../data/plasmids/pCRISPR–Cas9_plasmid_addgene.gbk'\n",
        "clean_plasmid = load_and_process_plasmid(path_to_plasmid)\n",
        "\n",
        "# 3 Choose genes to knock out (list)\n",
        "genes_to_KO = ['SCO5087', 'SCO5089', 'SCO5090']\n",
        "\n",
        "#### Advanced settings ####\n",
        "\n",
        "# 4 Filtering metrics for sgRNAs\n",
        "gc_upper = 0.72\n",
        "gc_lower = 0.2\n",
        "off_target_seed = 13\n",
        "off_target_upper = 10\n",
        "cas_type='cas9'\n",
        "number_of_sgRNAs_per_group = 2\n",
        "\n",
        "# 5 Choose polymerase and target melting temperature\n",
        "chosen_polymerase = polymerase_dict['Phusion High-Fidelity DNA Polymerase (GC Buffer)']\n",
        "melting_temperature = 65\n",
        "primer_concentration = 0.4 \n",
        "primer_number_increment = 23\n",
        "flanking_region = 500\n",
        "\n",
        "# 6 Choose overlapping sequences for our plasmid we can use the following\n",
        "#As per the article **\"CRISPR–Cas9, CRISPRi and CRISPR-BEST-mediated genetic manipulation in streptomycetes\"** we need the following oligoes: \n",
        "#CGGTTGGTAGGATCGACGGC **-N20-** GTTTTAGAGCTAGAAATAGC\n",
        "up_homology = Dseqrecord('CGGTTGGTAGGATCGACGGC')\n",
        "dw_homology = Dseqrecord('GTTTTAGAGCTAGAAATAGC')"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "# Computation"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 3,
      "metadata": {},
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "sgRNA generated were outside the designated border in SCO5087. To incorporate this extent borders. Skipping to next locus tag.\n",
            "sgRNA generated were outside the designated border in SCO5087. To incorporate this extent borders. Skipping to next locus tag.\n",
            "sgRNA generated were outside the designated border in SCO5087. To incorporate this extent borders. Skipping to next locus tag.\n",
            "sgRNA generated were outside the designated border in SCO5089. To incorporate this extent borders. Skipping to next locus tag.\n",
            "sgRNA generated were outside the designated border in SCO5089. To incorporate this extent borders. Skipping to next locus tag.\n",
            "sgRNA generated were outside the designated border in SCO5090. To incorporate this extent borders. Skipping to next locus tag.\n",
            "sgRNA generated were outside the designated border in SCO5090. To incorporate this extent borders. Skipping to next locus tag.\n",
            "sgRNA generated were outside the designated border in SCO5090. To incorporate this extent borders. Skipping to next locus tag.\n",
            "sgRNA generated were outside the designated border in SCO5090. To incorporate this extent borders. Skipping to next locus tag.\n",
            "sgRNA generated were outside the designated border in SCO5090. To incorporate this extent borders. Skipping to next locus tag.\n"
          ]
        },
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>strain_name</th>\n",
              "      <th>locus_tag</th>\n",
              "      <th>gene_loc</th>\n",
              "      <th>gene_strand</th>\n",
              "      <th>sgrna_strand</th>\n",
              "      <th>sgrna_loc</th>\n",
              "      <th>gc</th>\n",
              "      <th>pam</th>\n",
              "      <th>sgrna</th>\n",
              "      <th>sgrna_seed_sequence</th>\n",
              "      <th>off_target_count</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>336</th>\n",
              "      <td>Streptomyces_coelicolor_A3_chromosome</td>\n",
              "      <td>SCO5090</td>\n",
              "      <td>5532706</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>32</td>\n",
              "      <td>0.65</td>\n",
              "      <td>CGG</td>\n",
              "      <td>TGGACCACCCGGTGTTCTCC</td>\n",
              "      <td>CCCGGTGTTCTCC</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>332</th>\n",
              "      <td>Streptomyces_coelicolor_A3_chromosome</td>\n",
              "      <td>SCO5089</td>\n",
              "      <td>5532449</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>61</td>\n",
              "      <td>0.70</td>\n",
              "      <td>CGG</td>\n",
              "      <td>CGTGGAGTGCGCCGGTGAGA</td>\n",
              "      <td>TGCGCCGGTGAGA</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>331</th>\n",
              "      <td>Streptomyces_coelicolor_A3_chromosome</td>\n",
              "      <td>SCO5089</td>\n",
              "      <td>5532449</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>65</td>\n",
              "      <td>0.70</td>\n",
              "      <td>CGG</td>\n",
              "      <td>GAGTGCGCCGGTGAGACGGA</td>\n",
              "      <td>CCGGTGAGACGGA</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>330</th>\n",
              "      <td>Streptomyces_coelicolor_A3_chromosome</td>\n",
              "      <td>SCO5089</td>\n",
              "      <td>5532449</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>66</td>\n",
              "      <td>0.70</td>\n",
              "      <td>GGG</td>\n",
              "      <td>AGTGCGCCGGTGAGACGGAC</td>\n",
              "      <td>CGGTGAGACGGAC</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>328</th>\n",
              "      <td>Streptomyces_coelicolor_A3_chromosome</td>\n",
              "      <td>SCO5089</td>\n",
              "      <td>5532449</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>80</td>\n",
              "      <td>0.70</td>\n",
              "      <td>CGG</td>\n",
              "      <td>ACGGACGGGACGGACCTGTC</td>\n",
              "      <td>GGACGGACCTGTC</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>...</th>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>406</th>\n",
              "      <td>Streptomyces_coelicolor_A3_chromosome</td>\n",
              "      <td>SCO5090</td>\n",
              "      <td>5532706</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>691</td>\n",
              "      <td>0.70</td>\n",
              "      <td>TGG</td>\n",
              "      <td>GCCGGATCGAACGGGTGGTG</td>\n",
              "      <td>CGAACGGGTGGTG</td>\n",
              "      <td>2</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>78</th>\n",
              "      <td>Streptomyces_coelicolor_A3_chromosome</td>\n",
              "      <td>SCO5087</td>\n",
              "      <td>5529801</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>745</td>\n",
              "      <td>0.65</td>\n",
              "      <td>GGG</td>\n",
              "      <td>CGACGACGATCGGGGTGATC</td>\n",
              "      <td>GATCGGGGTGATC</td>\n",
              "      <td>2</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>452</th>\n",
              "      <td>Streptomyces_coelicolor_A3_chromosome</td>\n",
              "      <td>SCO5090</td>\n",
              "      <td>5532706</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>890</td>\n",
              "      <td>0.65</td>\n",
              "      <td>CGG</td>\n",
              "      <td>ACCCATCTGCGCGAGGTGCT</td>\n",
              "      <td>TGCGCGAGGTGCT</td>\n",
              "      <td>3</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>192</th>\n",
              "      <td>Streptomyces_coelicolor_A3_chromosome</td>\n",
              "      <td>SCO5087</td>\n",
              "      <td>5529801</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>838</td>\n",
              "      <td>0.70</td>\n",
              "      <td>TGG</td>\n",
              "      <td>TACGCGGGACGGCTTCGTGC</td>\n",
              "      <td>GACGGCTTCGTGC</td>\n",
              "      <td>3</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>201</th>\n",
              "      <td>Streptomyces_coelicolor_A3_chromosome</td>\n",
              "      <td>SCO5087</td>\n",
              "      <td>5529801</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>706</td>\n",
              "      <td>0.65</td>\n",
              "      <td>CGG</td>\n",
              "      <td>CAGCGCCGACGTGATGTTCG</td>\n",
              "      <td>GACGTGATGTTCG</td>\n",
              "      <td>3</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>279 rows × 11 columns</p>\n",
              "</div>"
            ],
            "text/plain": [
              "                               strain_name locus_tag  gene_loc  gene_strand  \\\n",
              "336  Streptomyces_coelicolor_A3_chromosome   SCO5090   5532706            1   \n",
              "332  Streptomyces_coelicolor_A3_chromosome   SCO5089   5532449            1   \n",
              "331  Streptomyces_coelicolor_A3_chromosome   SCO5089   5532449            1   \n",
              "330  Streptomyces_coelicolor_A3_chromosome   SCO5089   5532449            1   \n",
              "328  Streptomyces_coelicolor_A3_chromosome   SCO5089   5532449            1   \n",
              "..                                     ...       ...       ...          ...   \n",
              "406  Streptomyces_coelicolor_A3_chromosome   SCO5090   5532706            1   \n",
              "78   Streptomyces_coelicolor_A3_chromosome   SCO5087   5529801            1   \n",
              "452  Streptomyces_coelicolor_A3_chromosome   SCO5090   5532706            1   \n",
              "192  Streptomyces_coelicolor_A3_chromosome   SCO5087   5529801            1   \n",
              "201  Streptomyces_coelicolor_A3_chromosome   SCO5087   5529801            1   \n",
              "\n",
              "     sgrna_strand  sgrna_loc    gc  pam                 sgrna  \\\n",
              "336            -1         32  0.65  CGG  TGGACCACCCGGTGTTCTCC   \n",
              "332             1         61  0.70  CGG  CGTGGAGTGCGCCGGTGAGA   \n",
              "331             1         65  0.70  CGG  GAGTGCGCCGGTGAGACGGA   \n",
              "330             1         66  0.70  GGG  AGTGCGCCGGTGAGACGGAC   \n",
              "328             1         80  0.70  CGG  ACGGACGGGACGGACCTGTC   \n",
              "..            ...        ...   ...  ...                   ...   \n",
              "406            -1        691  0.70  TGG  GCCGGATCGAACGGGTGGTG   \n",
              "78             -1        745  0.65  GGG  CGACGACGATCGGGGTGATC   \n",
              "452             1        890  0.65  CGG  ACCCATCTGCGCGAGGTGCT   \n",
              "192             1        838  0.70  TGG  TACGCGGGACGGCTTCGTGC   \n",
              "201             1        706  0.65  CGG  CAGCGCCGACGTGATGTTCG   \n",
              "\n",
              "    sgrna_seed_sequence  off_target_count  \n",
              "336       CCCGGTGTTCTCC                 0  \n",
              "332       TGCGCCGGTGAGA                 0  \n",
              "331       CCGGTGAGACGGA                 0  \n",
              "330       CGGTGAGACGGAC                 0  \n",
              "328       GGACGGACCTGTC                 0  \n",
              "..                  ...               ...  \n",
              "406       CGAACGGGTGGTG                 2  \n",
              "78        GATCGGGGTGATC                 2  \n",
              "452       TGCGCGAGGTGCT                 3  \n",
              "192       GACGGCTTCGTGC                 3  \n",
              "201       GACGTGATGTTCG                 3  \n",
              "\n",
              "[279 rows x 11 columns]"
            ]
          },
          "execution_count": 3,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "\n",
        "# Initialize SgRNAargs with desired parameters\n",
        "args = SgRNAargs(path_to_genome, \n",
        "                genes_to_KO,\n",
        "                step=['find', 'filter'],\n",
        "                gc_upper = gc_upper,\n",
        "                gc_lower = gc_lower,\n",
        "                off_target_seed = off_target_seed,\n",
        "                off_target_upper = off_target_upper,\n",
        "                cas_type='cas9'\n",
        "                )\n",
        "\n",
        "sgrna_df = extract_sgRNAs(args)\n",
        "sgrna_df"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 4,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>strain_name</th>\n",
              "      <th>locus_tag</th>\n",
              "      <th>gene_loc</th>\n",
              "      <th>gene_strand</th>\n",
              "      <th>sgrna_strand</th>\n",
              "      <th>sgrna_loc</th>\n",
              "      <th>gc</th>\n",
              "      <th>pam</th>\n",
              "      <th>sgrna</th>\n",
              "      <th>sgrna_seed_sequence</th>\n",
              "      <th>off_target_count</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>336</th>\n",
              "      <td>Streptomyces_coelicolor_A3_chromosome</td>\n",
              "      <td>SCO5090</td>\n",
              "      <td>5532706</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>32</td>\n",
              "      <td>0.65</td>\n",
              "      <td>CGG</td>\n",
              "      <td>TGGACCACCCGGTGTTCTCC</td>\n",
              "      <td>CCCGGTGTTCTCC</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>332</th>\n",
              "      <td>Streptomyces_coelicolor_A3_chromosome</td>\n",
              "      <td>SCO5089</td>\n",
              "      <td>5532449</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>61</td>\n",
              "      <td>0.70</td>\n",
              "      <td>CGG</td>\n",
              "      <td>CGTGGAGTGCGCCGGTGAGA</td>\n",
              "      <td>TGCGCCGGTGAGA</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>331</th>\n",
              "      <td>Streptomyces_coelicolor_A3_chromosome</td>\n",
              "      <td>SCO5089</td>\n",
              "      <td>5532449</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>65</td>\n",
              "      <td>0.70</td>\n",
              "      <td>CGG</td>\n",
              "      <td>GAGTGCGCCGGTGAGACGGA</td>\n",
              "      <td>CCGGTGAGACGGA</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>347</th>\n",
              "      <td>Streptomyces_coelicolor_A3_chromosome</td>\n",
              "      <td>SCO5090</td>\n",
              "      <td>5532706</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>123</td>\n",
              "      <td>0.60</td>\n",
              "      <td>CGG</td>\n",
              "      <td>GTGCACGGTGGGTTCGAAGA</td>\n",
              "      <td>GTGGGTTCGAAGA</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>262</th>\n",
              "      <td>Streptomyces_coelicolor_A3_chromosome</td>\n",
              "      <td>SCO5087</td>\n",
              "      <td>5529801</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>271</td>\n",
              "      <td>0.70</td>\n",
              "      <td>AGG</td>\n",
              "      <td>CCCCTCGCCCTACCGTTCAC</td>\n",
              "      <td>CCCTACCGTTCAC</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>261</th>\n",
              "      <td>Streptomyces_coelicolor_A3_chromosome</td>\n",
              "      <td>SCO5087</td>\n",
              "      <td>5529801</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>277</td>\n",
              "      <td>0.65</td>\n",
              "      <td>CGG</td>\n",
              "      <td>GCCCTACCGTTCACAGGTCG</td>\n",
              "      <td>CGTTCACAGGTCG</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "                               strain_name locus_tag  gene_loc  gene_strand  \\\n",
              "336  Streptomyces_coelicolor_A3_chromosome   SCO5090   5532706            1   \n",
              "332  Streptomyces_coelicolor_A3_chromosome   SCO5089   5532449            1   \n",
              "331  Streptomyces_coelicolor_A3_chromosome   SCO5089   5532449            1   \n",
              "347  Streptomyces_coelicolor_A3_chromosome   SCO5090   5532706            1   \n",
              "262  Streptomyces_coelicolor_A3_chromosome   SCO5087   5529801            1   \n",
              "261  Streptomyces_coelicolor_A3_chromosome   SCO5087   5529801            1   \n",
              "\n",
              "     sgrna_strand  sgrna_loc    gc  pam                 sgrna  \\\n",
              "336            -1         32  0.65  CGG  TGGACCACCCGGTGTTCTCC   \n",
              "332             1         61  0.70  CGG  CGTGGAGTGCGCCGGTGAGA   \n",
              "331             1         65  0.70  CGG  GAGTGCGCCGGTGAGACGGA   \n",
              "347            -1        123  0.60  CGG  GTGCACGGTGGGTTCGAAGA   \n",
              "262             1        271  0.70  AGG  CCCCTCGCCCTACCGTTCAC   \n",
              "261             1        277  0.65  CGG  GCCCTACCGTTCACAGGTCG   \n",
              "\n",
              "    sgrna_seed_sequence  off_target_count  \n",
              "336       CCCGGTGTTCTCC                 0  \n",
              "332       TGCGCCGGTGAGA                 0  \n",
              "331       CCGGTGAGACGGA                 0  \n",
              "347       GTGGGTTCGAAGA                 0  \n",
              "262       CCCTACCGTTCAC                 0  \n",
              "261       CGTTCACAGGTCG                 0  "
            ]
          },
          "execution_count": 4,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "# Filter the DataFrame to retain only up to 5 sgRNA sequences per locus_tag\n",
        "filtered_df = sgrna_df.groupby('locus_tag').head(number_of_sgRNAs_per_group)\n",
        "filtered_df"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "## Output"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 5,
      "metadata": {},
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "SCO5090_loc_32\n"
          ]
        },
        {
          "data": {
            "text/plain": [
              "[Contig(o11279),\n",
              " Contig(o11279),\n",
              " Contig(o11279),\n",
              " Contig(o11279),\n",
              " Contig(o11279),\n",
              " Contig(o11279)]"
            ]
          },
          "execution_count": 5,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "# MAke oligoes\n",
        "list_of_ssDNAs = make_ssDNA_oligos(filtered_df, upstream_ovh = up_homology,\n",
        "                      downstream_ovh=dw_homology)\n",
        "print(list_of_ssDNAs[0].name)\n",
        "\n",
        "# cut plasmid\n",
        "from Bio.Restriction import NcoI\n",
        "linearized_plasmid = sorted(clean_plasmid.cut(NcoI), key=lambda x: len(x), reverse=True)[0]\n",
        "#print(linearized_plasmid)\n",
        "\n",
        "sgRNA_vectors = assemble_plasmids_by_ssDNA_bridging(list_of_ssDNAs,linearized_plasmid)\n",
        "sgRNA_vectors"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 6,
      "metadata": {},
      "outputs": [],
      "source": [
        "\n",
        "# Constructing a meaningful name, ID, and description for the assembled plasmid using user input\n",
        "targeting_info = []\n",
        "for index, row in filtered_df.iterrows():\n",
        "    formatted_str = f\"CRISPR-Cas9_plasmid_targeting_{row['locus_tag']}_position_{row['sgrna_loc']}\"\n",
        "    targeting_info.append(formatted_str)\n",
        "\n",
        "for i in range(len(sgRNA_vectors)):\n",
        "    sgRNA_vectors[i].name = f'{targeting_info[i]}_{i+1}'\n",
        "    sgRNA_vectors[i].id = sgRNA_vectors[i].name  \n",
        "    sgRNA_vectors[i].description = f'CRISPR-Cas9 targeting {\", \".join(genes_to_KO)} for single gene knockout, assembled using StreptoAIM.'\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 7,
      "metadata": {},
      "outputs": [],
      "source": [
        "print_plasmids = False\n",
        "\n",
        "if print_plasmids: \n",
        "    for vector in sgRNA_vectors: \n",
        "        vector.write(f\"../../data/plasmids/sgRNA_plasmids_pCRISPR-Cas9/{vector.id}.gb\")"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "### IDT primers"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 8,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>Name</th>\n",
              "      <th>Sequence</th>\n",
              "      <th>Concentration</th>\n",
              "      <th>Purification</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>SCO5090_loc_32</td>\n",
              "      <td>CGGTTGGTAGGATCGACGGCTGGACCACCCGGTGTTCTCCGTTTTA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>SCO5089_loc_61</td>\n",
              "      <td>CGGTTGGTAGGATCGACGGCCGTGGAGTGCGCCGGTGAGAGTTTTA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>SCO5089_loc_65</td>\n",
              "      <td>CGGTTGGTAGGATCGACGGCGAGTGCGCCGGTGAGACGGAGTTTTA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>SCO5090_loc_123</td>\n",
              "      <td>CGGTTGGTAGGATCGACGGCGTGCACGGTGGGTTCGAAGAGTTTTA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>SCO5087_loc_271</td>\n",
              "      <td>CGGTTGGTAGGATCGACGGCCCCCTCGCCCTACCGTTCACGTTTTA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>5</th>\n",
              "      <td>SCO5087_loc_277</td>\n",
              "      <td>CGGTTGGTAGGATCGACGGCGCCCTACCGTTCACAGGTCGGTTTTA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "              Name                                           Sequence  \\\n",
              "0   SCO5090_loc_32  CGGTTGGTAGGATCGACGGCTGGACCACCCGGTGTTCTCCGTTTTA...   \n",
              "1   SCO5089_loc_61  CGGTTGGTAGGATCGACGGCCGTGGAGTGCGCCGGTGAGAGTTTTA...   \n",
              "2   SCO5089_loc_65  CGGTTGGTAGGATCGACGGCGAGTGCGCCGGTGAGACGGAGTTTTA...   \n",
              "3  SCO5090_loc_123  CGGTTGGTAGGATCGACGGCGTGCACGGTGGGTTCGAAGAGTTTTA...   \n",
              "4  SCO5087_loc_271  CGGTTGGTAGGATCGACGGCCCCCTCGCCCTACCGTTCACGTTTTA...   \n",
              "5  SCO5087_loc_277  CGGTTGGTAGGATCGACGGCGCCCTACCGTTCACAGGTCGGTTTTA...   \n",
              "\n",
              "  Concentration Purification  \n",
              "0          25nm          STD  \n",
              "1          25nm          STD  \n",
              "2          25nm          STD  \n",
              "3          25nm          STD  \n",
              "4          25nm          STD  \n",
              "5          25nm          STD  "
            ]
          },
          "execution_count": 8,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "idt_primers=primers_to_IDT(list_of_ssDNAs)\n",
        "idt_primers"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "# In frame deletion"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 9,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/plain": [
              "[{'name': 'SCO5087',\n",
              "  'up_repair': Amplicon(1000),\n",
              "  'up_forwar_p': f1000 16-mer:5'-GACGCCGGAAATCCAG-3',\n",
              "  'up_reverse_p': r1000 16-mer:5'-CCCATCTCCCTTCGAC-3',\n",
              "  'tm_up_forwar_p': 56,\n",
              "  'tm_up_reverse_p': 54,\n",
              "  'ta_up': 58,\n",
              "  'location_up_start': 5528800,\n",
              "  'location_up_end': 5529800,\n",
              "  'dw_repair': Amplicon(1000),\n",
              "  'dw_forwar_p': f1000 16-mer:5'-GCGTCCTCATCACCGG-3',\n",
              "  'dw_reverse_p': r1000 15-mer:5'-GAGTAGAGGCGGCCC-3',\n",
              "  'tm_dw_forwar_p': 58,\n",
              "  'tm_dw_reverse_p': 57,\n",
              "  'ta_dw': 61,\n",
              "  'location_dw_start': 5531204,\n",
              "  'location_dw_end': 5532204},\n",
              " {'name': 'SCO5089',\n",
              "  'up_repair': Amplicon(1000),\n",
              "  'up_forwar_p': f1000 14-mer:5'-CCAGGACGCGAAGG-3',\n",
              "  'up_reverse_p': r1000 18-mer:5'-GGCTTTCTCCAGTTCTCG-3',\n",
              "  'tm_up_forwar_p': 54,\n",
              "  'tm_up_reverse_p': 56,\n",
              "  'ta_up': 58,\n",
              "  'location_up_start': 5531448,\n",
              "  'location_up_end': 5532448,\n",
              "  'dw_repair': Amplicon(1000),\n",
              "  'dw_forwar_p': f1000 13-mer:5'-GCCGGCCGGGAGA-3',\n",
              "  'dw_reverse_p': r1000 14-mer:5'-GGGGGCCTGCTCGT-3',\n",
              "  'tm_dw_forwar_p': 59,\n",
              "  'tm_dw_reverse_p': 60,\n",
              "  'ta_dw': 63,\n",
              "  'location_dw_start': 5532709,\n",
              "  'location_dw_end': 5533709},\n",
              " {'name': 'SCO5090',\n",
              "  'up_repair': Amplicon(1000),\n",
              "  'up_forwar_p': f1000 13-mer:5'-GCCACGCCCGCCG-3',\n",
              "  'up_reverse_p': r1000 15-mer:5'-GCCGCCTCGGCCAGT-3',\n",
              "  'tm_up_forwar_p': 63,\n",
              "  'tm_up_reverse_p': 64,\n",
              "  'ta_up': 66,\n",
              "  'location_up_start': 5531705,\n",
              "  'location_up_end': 5532705,\n",
              "  'dw_repair': Amplicon(1000),\n",
              "  'dw_forwar_p': f1000 14-mer:5'-CGGTCGAGGTCCGT-3',\n",
              "  'dw_reverse_p': r1000 13-mer:5'-ACCGTGCGGGACT-3',\n",
              "  'tm_dw_forwar_p': 56,\n",
              "  'tm_dw_reverse_p': 55,\n",
              "  'ta_dw': 59,\n",
              "  'location_dw_start': 5533656,\n",
              "  'location_dw_end': 5534656}]"
            ]
          },
          "execution_count": 9,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "# Make repair templates\n",
        "repair_templates_data = find_up_dw_repair_templates(genome, \n",
        "                                                    genes_to_KO, \n",
        "                                                    target_tm=melting_temperature, \n",
        "                                                    \n",
        "                                                    primer_tm_kwargs={'conc':primer_concentration, 'prodcode':chosen_polymerase} )\n",
        "\n",
        "repair_templates_data"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 10,
      "metadata": {},
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "[{'gene_name': 'SCO5087', 'name': 'CRISPR-Cas9_plasmid_targeting_SCO5087_position_271_5', 'contig': Dseqrecord(o13279), 'up_forwar_p_name': 'primer_1', 'up_reverse_p_name': 'primer_2', 'up_forwar_p': primer_1 56-mer:5'-AAGGCCGCTTTTGCG..CAG-3', 'up_reverse_p': primer_2 36-mer:5'-GACTCCGGTGATGAG..GAC-3', 'up_forwar_primer_str': 'AAGGCCGCTTTTGCGGGATCTCGTCGAAGGCACTAGAAGGGACGCCGGAAATCCAG', 'up_reverse_primer_str': 'GACTCCGGTGATGAGGACGCCCCATCTCCCTTCGAC', 'up_forwar_p_anneal': Seq('GACGCCGGAAATCCAG'), 'up_reverse_p_anneal': Seq('CCCATCTCCCTTCGAC'), 'tm_up_forwar_p': 56, 'tm_up_reverse_p': 54, 'ta_up': 58, 'dw_forwar_p_name': 'primer_3', 'dw_reverse_p_name': 'primer_4', 'dw_forwar_p': primer_3 36-mer:5'-GGCGGTCGAAGGGAG..CGG-3', 'dw_reverse_p': primer_4 55-mer:5'-TCCCCGTCCGGGACC..CCC-3', 'dw_forwar_primer_str': 'GGCGGTCGAAGGGAGATGGGGCGTCCTCATCACCGG', 'dw_reverse_primer_str': 'TCCCCGTCCGGGACCCGCGCGGTCGATCCCCGCATATAGGGAGTAGAGGCGGCCC', 'dw_forwar_p_anneal': Seq('GCGTCCTCATCACCGG'), 'dw_reverse_p_anneal': Seq('GAGTAGAGGCGGCCC'), 'tm_dw_forwar_p': 58, 'tm_dw_reverse_p': 57, 'ta_dw': 61}, {'gene_name': 'SCO5087', 'name': 'CRISPR-Cas9_plasmid_targeting_SCO5087_position_277_6', 'contig': Dseqrecord(o13279), 'up_forwar_p_name': 'primer_1', 'up_reverse_p_name': 'primer_2', 'up_forwar_p': f1000 56-mer:5'-AAGGCCGCTTTTGCG..CAG-3', 'up_reverse_p': r1000 36-mer:5'-GACTCCGGTGATGAG..GAC-3', 'up_forwar_primer_str': 'AAGGCCGCTTTTGCGGGATCTCGTCGAAGGCACTAGAAGGGACGCCGGAAATCCAG', 'up_reverse_primer_str': 'GACTCCGGTGATGAGGACGCCCCATCTCCCTTCGAC', 'up_forwar_p_anneal': Seq('GACGCCGGAAATCCAG'), 'up_reverse_p_anneal': Seq('CCCATCTCCCTTCGAC'), 'tm_up_forwar_p': 56, 'tm_up_reverse_p': 54, 'ta_up': 58, 'dw_forwar_p_name': 'primer_3', 'dw_reverse_p_name': 'primer_4', 'dw_forwar_p': f1000 36-mer:5'-GGCGGTCGAAGGGAG..CGG-3', 'dw_reverse_p': r1000 55-mer:5'-TCCCCGTCCGGGACC..CCC-3', 'dw_forwar_primer_str': 'GGCGGTCGAAGGGAGATGGGGCGTCCTCATCACCGG', 'dw_reverse_primer_str': 'TCCCCGTCCGGGACCCGCGCGGTCGATCCCCGCATATAGGGAGTAGAGGCGGCCC', 'dw_forwar_p_anneal': Seq('GCGTCCTCATCACCGG'), 'dw_reverse_p_anneal': Seq('GAGTAGAGGCGGCCC'), 'tm_dw_forwar_p': 58, 'tm_dw_reverse_p': 57, 'ta_dw': 61}, {'gene_name': 'SCO5089', 'name': 'CRISPR-Cas9_plasmid_targeting_SCO5089_position_61_2', 'contig': Dseqrecord(o13279), 'up_forwar_p_name': 'primer_5', 'up_reverse_p_name': 'primer_6', 'up_forwar_p': primer_5 54-mer:5'-AAGGCCGCTTTTGCG..AGG-3', 'up_reverse_p': primer_6 38-mer:5'-CCGGTGTTCTCCCGG..TCG-3', 'up_forwar_primer_str': 'AAGGCCGCTTTTGCGGGATCTCGTCGAAGGCACTAGAAGGCCAGGACGCGAAGG', 'up_reverse_primer_str': 'CCGGTGTTCTCCCGGCCGGCGGCTTTCTCCAGTTCTCG', 'up_forwar_p_anneal': Seq('CCAGGACGCGAAGG'), 'up_reverse_p_anneal': Seq('GGCTTTCTCCAGTTCTCG'), 'tm_up_forwar_p': 54, 'tm_up_reverse_p': 56, 'ta_up': 58, 'dw_forwar_p_name': 'primer_7', 'dw_reverse_p_name': 'primer_8', 'dw_forwar_p': primer_7 33-mer:5'-CCCGAGAACTGGAGA..AGA-3', 'dw_reverse_p': primer_8 54-mer:5'-TCCCCGTCCGGGACC..CGT-3', 'dw_forwar_primer_str': 'CCCGAGAACTGGAGAAAGCCGCCGGCCGGGAGA', 'dw_reverse_primer_str': 'TCCCCGTCCGGGACCCGCGCGGTCGATCCCCGCATATAGGGGGGGCCTGCTCGT', 'dw_forwar_p_anneal': Seq('GCCGGCCGGGAGA'), 'dw_reverse_p_anneal': Seq('GGGGGCCTGCTCGT'), 'tm_dw_forwar_p': 59, 'tm_dw_reverse_p': 60, 'ta_dw': 63}, {'gene_name': 'SCO5089', 'name': 'CRISPR-Cas9_plasmid_targeting_SCO5089_position_65_3', 'contig': Dseqrecord(o13279), 'up_forwar_p_name': 'primer_5', 'up_reverse_p_name': 'primer_6', 'up_forwar_p': f1000 54-mer:5'-AAGGCCGCTTTTGCG..AGG-3', 'up_reverse_p': r1000 38-mer:5'-CCGGTGTTCTCCCGG..TCG-3', 'up_forwar_primer_str': 'AAGGCCGCTTTTGCGGGATCTCGTCGAAGGCACTAGAAGGCCAGGACGCGAAGG', 'up_reverse_primer_str': 'CCGGTGTTCTCCCGGCCGGCGGCTTTCTCCAGTTCTCG', 'up_forwar_p_anneal': Seq('CCAGGACGCGAAGG'), 'up_reverse_p_anneal': Seq('GGCTTTCTCCAGTTCTCG'), 'tm_up_forwar_p': 54, 'tm_up_reverse_p': 56, 'ta_up': 58, 'dw_forwar_p_name': 'primer_7', 'dw_reverse_p_name': 'primer_8', 'dw_forwar_p': f1000 33-mer:5'-CCCGAGAACTGGAGA..AGA-3', 'dw_reverse_p': r1000 54-mer:5'-TCCCCGTCCGGGACC..CGT-3', 'dw_forwar_primer_str': 'CCCGAGAACTGGAGAAAGCCGCCGGCCGGGAGA', 'dw_reverse_primer_str': 'TCCCCGTCCGGGACCCGCGCGGTCGATCCCCGCATATAGGGGGGGCCTGCTCGT', 'dw_forwar_p_anneal': Seq('GCCGGCCGGGAGA'), 'dw_reverse_p_anneal': Seq('GGGGGCCTGCTCGT'), 'tm_dw_forwar_p': 59, 'tm_dw_reverse_p': 60, 'ta_dw': 63}, {'gene_name': 'SCO5090', 'name': 'CRISPR-Cas9_plasmid_targeting_SCO5090_position_32_1', 'contig': Dseqrecord(o13279), 'up_forwar_p_name': 'primer_9', 'up_reverse_p_name': 'primer_10', 'up_forwar_p': primer_9 53-mer:5'-AAGGCCGCTTTTGCG..CCG-3', 'up_reverse_p': primer_10 35-mer:5'-GACCTCACGGACCTC..AGT-3', 'up_forwar_primer_str': 'AAGGCCGCTTTTGCGGGATCTCGTCGAAGGCACTAGAAGGGCCACGCCCGCCG', 'up_reverse_primer_str': 'GACCTCACGGACCTCGACCGGCCGCCTCGGCCAGT', 'up_forwar_p_anneal': Seq('GCCACGCCCGCCG'), 'up_reverse_p_anneal': Seq('GCCGCCTCGGCCAGT'), 'tm_up_forwar_p': 63, 'tm_up_reverse_p': 64, 'ta_up': 66, 'dw_forwar_p_name': 'primer_11', 'dw_reverse_p_name': 'primer_12', 'dw_forwar_p': primer_11 34-mer:5'-GGCGCACTGGCCGAG..CGT-3', 'dw_reverse_p': primer_12 53-mer:5'-TCCCCGTCCGGGACC..ACT-3', 'dw_forwar_primer_str': 'GGCGCACTGGCCGAGGCGGCCGGTCGAGGTCCGT', 'dw_reverse_primer_str': 'TCCCCGTCCGGGACCCGCGCGGTCGATCCCCGCATATAGGACCGTGCGGGACT', 'dw_forwar_p_anneal': Seq('CGGTCGAGGTCCGT'), 'dw_reverse_p_anneal': Seq('ACCGTGCGGGACT'), 'tm_dw_forwar_p': 56, 'tm_dw_reverse_p': 55, 'ta_dw': 59}, {'gene_name': 'SCO5090', 'name': 'CRISPR-Cas9_plasmid_targeting_SCO5090_position_123_4', 'contig': Dseqrecord(o13279), 'up_forwar_p_name': 'primer_9', 'up_reverse_p_name': 'primer_10', 'up_forwar_p': f1000 53-mer:5'-AAGGCCGCTTTTGCG..CCG-3', 'up_reverse_p': r1000 35-mer:5'-GACCTCACGGACCTC..AGT-3', 'up_forwar_primer_str': 'AAGGCCGCTTTTGCGGGATCTCGTCGAAGGCACTAGAAGGGCCACGCCCGCCG', 'up_reverse_primer_str': 'GACCTCACGGACCTCGACCGGCCGCCTCGGCCAGT', 'up_forwar_p_anneal': Seq('GCCACGCCCGCCG'), 'up_reverse_p_anneal': Seq('GCCGCCTCGGCCAGT'), 'tm_up_forwar_p': 63, 'tm_up_reverse_p': 64, 'ta_up': 66, 'dw_forwar_p_name': 'primer_11', 'dw_reverse_p_name': 'primer_12', 'dw_forwar_p': f1000 34-mer:5'-GGCGCACTGGCCGAG..CGT-3', 'dw_reverse_p': r1000 53-mer:5'-TCCCCGTCCGGGACC..ACT-3', 'dw_forwar_primer_str': 'GGCGCACTGGCCGAGGCGGCCGGTCGAGGTCCGT', 'dw_reverse_primer_str': 'TCCCCGTCCGGGACCCGCGCGGTCGATCCCCGCATATAGGACCGTGCGGGACT', 'dw_forwar_p_anneal': Seq('CGGTCGAGGTCCGT'), 'dw_reverse_p_anneal': Seq('ACCGTGCGGGACT'), 'tm_dw_forwar_p': 56, 'tm_dw_reverse_p': 55, 'ta_dw': 59}]\n"
          ]
        },
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>template</th>\n",
              "      <th>direction</th>\n",
              "      <th>f_primer_anneal(5-3)</th>\n",
              "      <th>r_primer_anneal(5-3)</th>\n",
              "      <th>f_tm</th>\n",
              "      <th>r_tm</th>\n",
              "      <th>ta</th>\n",
              "      <th>f_primer_sequences(5-3)</th>\n",
              "      <th>r_primer_sequences(5-3)</th>\n",
              "      <th>f_primer_name</th>\n",
              "      <th>r_primer_name</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>SCO5087</td>\n",
              "      <td>upstream</td>\n",
              "      <td>GACGCCGGAAATCCAG</td>\n",
              "      <td>CCCATCTCCCTTCGAC</td>\n",
              "      <td>56</td>\n",
              "      <td>54</td>\n",
              "      <td>58</td>\n",
              "      <td>AAGGCCGCTTTTGCGGGATCTCGTCGAAGGCACTAGAAGGGACGCC...</td>\n",
              "      <td>GACTCCGGTGATGAGGACGCCCCATCTCCCTTCGAC</td>\n",
              "      <td>primer_1</td>\n",
              "      <td>primer_2</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>SCO5087</td>\n",
              "      <td>downstream</td>\n",
              "      <td>GCGTCCTCATCACCGG</td>\n",
              "      <td>GAGTAGAGGCGGCCC</td>\n",
              "      <td>58</td>\n",
              "      <td>57</td>\n",
              "      <td>61</td>\n",
              "      <td>GGCGGTCGAAGGGAGATGGGGCGTCCTCATCACCGG</td>\n",
              "      <td>TCCCCGTCCGGGACCCGCGCGGTCGATCCCCGCATATAGGGAGTAG...</td>\n",
              "      <td>primer_3</td>\n",
              "      <td>primer_4</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>SCO5087</td>\n",
              "      <td>upstream</td>\n",
              "      <td>GACGCCGGAAATCCAG</td>\n",
              "      <td>CCCATCTCCCTTCGAC</td>\n",
              "      <td>56</td>\n",
              "      <td>54</td>\n",
              "      <td>58</td>\n",
              "      <td>AAGGCCGCTTTTGCGGGATCTCGTCGAAGGCACTAGAAGGGACGCC...</td>\n",
              "      <td>GACTCCGGTGATGAGGACGCCCCATCTCCCTTCGAC</td>\n",
              "      <td>primer_1</td>\n",
              "      <td>primer_2</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>SCO5087</td>\n",
              "      <td>downstream</td>\n",
              "      <td>GCGTCCTCATCACCGG</td>\n",
              "      <td>GAGTAGAGGCGGCCC</td>\n",
              "      <td>58</td>\n",
              "      <td>57</td>\n",
              "      <td>61</td>\n",
              "      <td>GGCGGTCGAAGGGAGATGGGGCGTCCTCATCACCGG</td>\n",
              "      <td>TCCCCGTCCGGGACCCGCGCGGTCGATCCCCGCATATAGGGAGTAG...</td>\n",
              "      <td>primer_3</td>\n",
              "      <td>primer_4</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>SCO5089</td>\n",
              "      <td>upstream</td>\n",
              "      <td>CCAGGACGCGAAGG</td>\n",
              "      <td>GGCTTTCTCCAGTTCTCG</td>\n",
              "      <td>54</td>\n",
              "      <td>56</td>\n",
              "      <td>58</td>\n",
              "      <td>AAGGCCGCTTTTGCGGGATCTCGTCGAAGGCACTAGAAGGCCAGGA...</td>\n",
              "      <td>CCGGTGTTCTCCCGGCCGGCGGCTTTCTCCAGTTCTCG</td>\n",
              "      <td>primer_5</td>\n",
              "      <td>primer_6</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>5</th>\n",
              "      <td>SCO5089</td>\n",
              "      <td>downstream</td>\n",
              "      <td>GCCGGCCGGGAGA</td>\n",
              "      <td>GGGGGCCTGCTCGT</td>\n",
              "      <td>59</td>\n",
              "      <td>60</td>\n",
              "      <td>63</td>\n",
              "      <td>CCCGAGAACTGGAGAAAGCCGCCGGCCGGGAGA</td>\n",
              "      <td>TCCCCGTCCGGGACCCGCGCGGTCGATCCCCGCATATAGGGGGGGC...</td>\n",
              "      <td>primer_7</td>\n",
              "      <td>primer_8</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>6</th>\n",
              "      <td>SCO5089</td>\n",
              "      <td>upstream</td>\n",
              "      <td>CCAGGACGCGAAGG</td>\n",
              "      <td>GGCTTTCTCCAGTTCTCG</td>\n",
              "      <td>54</td>\n",
              "      <td>56</td>\n",
              "      <td>58</td>\n",
              "      <td>AAGGCCGCTTTTGCGGGATCTCGTCGAAGGCACTAGAAGGCCAGGA...</td>\n",
              "      <td>CCGGTGTTCTCCCGGCCGGCGGCTTTCTCCAGTTCTCG</td>\n",
              "      <td>primer_5</td>\n",
              "      <td>primer_6</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>7</th>\n",
              "      <td>SCO5089</td>\n",
              "      <td>downstream</td>\n",
              "      <td>GCCGGCCGGGAGA</td>\n",
              "      <td>GGGGGCCTGCTCGT</td>\n",
              "      <td>59</td>\n",
              "      <td>60</td>\n",
              "      <td>63</td>\n",
              "      <td>CCCGAGAACTGGAGAAAGCCGCCGGCCGGGAGA</td>\n",
              "      <td>TCCCCGTCCGGGACCCGCGCGGTCGATCCCCGCATATAGGGGGGGC...</td>\n",
              "      <td>primer_7</td>\n",
              "      <td>primer_8</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>8</th>\n",
              "      <td>SCO5090</td>\n",
              "      <td>upstream</td>\n",
              "      <td>GCCACGCCCGCCG</td>\n",
              "      <td>GCCGCCTCGGCCAGT</td>\n",
              "      <td>63</td>\n",
              "      <td>64</td>\n",
              "      <td>66</td>\n",
              "      <td>AAGGCCGCTTTTGCGGGATCTCGTCGAAGGCACTAGAAGGGCCACG...</td>\n",
              "      <td>GACCTCACGGACCTCGACCGGCCGCCTCGGCCAGT</td>\n",
              "      <td>primer_9</td>\n",
              "      <td>primer_10</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>9</th>\n",
              "      <td>SCO5090</td>\n",
              "      <td>downstream</td>\n",
              "      <td>CGGTCGAGGTCCGT</td>\n",
              "      <td>ACCGTGCGGGACT</td>\n",
              "      <td>56</td>\n",
              "      <td>55</td>\n",
              "      <td>59</td>\n",
              "      <td>GGCGCACTGGCCGAGGCGGCCGGTCGAGGTCCGT</td>\n",
              "      <td>TCCCCGTCCGGGACCCGCGCGGTCGATCCCCGCATATAGGACCGTG...</td>\n",
              "      <td>primer_11</td>\n",
              "      <td>primer_12</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>10</th>\n",
              "      <td>SCO5090</td>\n",
              "      <td>upstream</td>\n",
              "      <td>GCCACGCCCGCCG</td>\n",
              "      <td>GCCGCCTCGGCCAGT</td>\n",
              "      <td>63</td>\n",
              "      <td>64</td>\n",
              "      <td>66</td>\n",
              "      <td>AAGGCCGCTTTTGCGGGATCTCGTCGAAGGCACTAGAAGGGCCACG...</td>\n",
              "      <td>GACCTCACGGACCTCGACCGGCCGCCTCGGCCAGT</td>\n",
              "      <td>primer_9</td>\n",
              "      <td>primer_10</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>11</th>\n",
              "      <td>SCO5090</td>\n",
              "      <td>downstream</td>\n",
              "      <td>CGGTCGAGGTCCGT</td>\n",
              "      <td>ACCGTGCGGGACT</td>\n",
              "      <td>56</td>\n",
              "      <td>55</td>\n",
              "      <td>59</td>\n",
              "      <td>GGCGCACTGGCCGAGGCGGCCGGTCGAGGTCCGT</td>\n",
              "      <td>TCCCCGTCCGGGACCCGCGCGGTCGATCCCCGCATATAGGACCGTG...</td>\n",
              "      <td>primer_11</td>\n",
              "      <td>primer_12</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "   template   direction f_primer_anneal(5-3) r_primer_anneal(5-3)  f_tm  r_tm  \\\n",
              "0   SCO5087    upstream     GACGCCGGAAATCCAG     CCCATCTCCCTTCGAC    56    54   \n",
              "1   SCO5087  downstream     GCGTCCTCATCACCGG      GAGTAGAGGCGGCCC    58    57   \n",
              "2   SCO5087    upstream     GACGCCGGAAATCCAG     CCCATCTCCCTTCGAC    56    54   \n",
              "3   SCO5087  downstream     GCGTCCTCATCACCGG      GAGTAGAGGCGGCCC    58    57   \n",
              "4   SCO5089    upstream       CCAGGACGCGAAGG   GGCTTTCTCCAGTTCTCG    54    56   \n",
              "5   SCO5089  downstream        GCCGGCCGGGAGA       GGGGGCCTGCTCGT    59    60   \n",
              "6   SCO5089    upstream       CCAGGACGCGAAGG   GGCTTTCTCCAGTTCTCG    54    56   \n",
              "7   SCO5089  downstream        GCCGGCCGGGAGA       GGGGGCCTGCTCGT    59    60   \n",
              "8   SCO5090    upstream        GCCACGCCCGCCG      GCCGCCTCGGCCAGT    63    64   \n",
              "9   SCO5090  downstream       CGGTCGAGGTCCGT        ACCGTGCGGGACT    56    55   \n",
              "10  SCO5090    upstream        GCCACGCCCGCCG      GCCGCCTCGGCCAGT    63    64   \n",
              "11  SCO5090  downstream       CGGTCGAGGTCCGT        ACCGTGCGGGACT    56    55   \n",
              "\n",
              "    ta                            f_primer_sequences(5-3)  \\\n",
              "0   58  AAGGCCGCTTTTGCGGGATCTCGTCGAAGGCACTAGAAGGGACGCC...   \n",
              "1   61               GGCGGTCGAAGGGAGATGGGGCGTCCTCATCACCGG   \n",
              "2   58  AAGGCCGCTTTTGCGGGATCTCGTCGAAGGCACTAGAAGGGACGCC...   \n",
              "3   61               GGCGGTCGAAGGGAGATGGGGCGTCCTCATCACCGG   \n",
              "4   58  AAGGCCGCTTTTGCGGGATCTCGTCGAAGGCACTAGAAGGCCAGGA...   \n",
              "5   63                  CCCGAGAACTGGAGAAAGCCGCCGGCCGGGAGA   \n",
              "6   58  AAGGCCGCTTTTGCGGGATCTCGTCGAAGGCACTAGAAGGCCAGGA...   \n",
              "7   63                  CCCGAGAACTGGAGAAAGCCGCCGGCCGGGAGA   \n",
              "8   66  AAGGCCGCTTTTGCGGGATCTCGTCGAAGGCACTAGAAGGGCCACG...   \n",
              "9   59                 GGCGCACTGGCCGAGGCGGCCGGTCGAGGTCCGT   \n",
              "10  66  AAGGCCGCTTTTGCGGGATCTCGTCGAAGGCACTAGAAGGGCCACG...   \n",
              "11  59                 GGCGCACTGGCCGAGGCGGCCGGTCGAGGTCCGT   \n",
              "\n",
              "                              r_primer_sequences(5-3) f_primer_name  \\\n",
              "0                GACTCCGGTGATGAGGACGCCCCATCTCCCTTCGAC      primer_1   \n",
              "1   TCCCCGTCCGGGACCCGCGCGGTCGATCCCCGCATATAGGGAGTAG...      primer_3   \n",
              "2                GACTCCGGTGATGAGGACGCCCCATCTCCCTTCGAC      primer_1   \n",
              "3   TCCCCGTCCGGGACCCGCGCGGTCGATCCCCGCATATAGGGAGTAG...      primer_3   \n",
              "4              CCGGTGTTCTCCCGGCCGGCGGCTTTCTCCAGTTCTCG      primer_5   \n",
              "5   TCCCCGTCCGGGACCCGCGCGGTCGATCCCCGCATATAGGGGGGGC...      primer_7   \n",
              "6              CCGGTGTTCTCCCGGCCGGCGGCTTTCTCCAGTTCTCG      primer_5   \n",
              "7   TCCCCGTCCGGGACCCGCGCGGTCGATCCCCGCATATAGGGGGGGC...      primer_7   \n",
              "8                 GACCTCACGGACCTCGACCGGCCGCCTCGGCCAGT      primer_9   \n",
              "9   TCCCCGTCCGGGACCCGCGCGGTCGATCCCCGCATATAGGACCGTG...     primer_11   \n",
              "10                GACCTCACGGACCTCGACCGGCCGCCTCGGCCAGT      primer_9   \n",
              "11  TCCCCGTCCGGGACCCGCGCGGTCGATCCCCGCATATAGGACCGTG...     primer_11   \n",
              "\n",
              "   r_primer_name  \n",
              "0       primer_2  \n",
              "1       primer_4  \n",
              "2       primer_2  \n",
              "3       primer_4  \n",
              "4       primer_6  \n",
              "5       primer_8  \n",
              "6       primer_6  \n",
              "7       primer_8  \n",
              "8      primer_10  \n",
              "9      primer_12  \n",
              "10     primer_10  \n",
              "11     primer_12  "
            ]
          },
          "execution_count": 10,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "\n",
        "# Digest the plasmids\n",
        "processed_records = [Dseqrecord(record, circular=True).cut(StuI)[0] for record in sgRNA_vectors]\n",
        "\n",
        "# Rename them appropriately\n",
        "for i in range(len(processed_records)):\n",
        "    processed_records[i].name = sgRNA_vectors[i].name\n",
        "\n",
        "# Assembly \n",
        "##### THE PROBLEM IS HERE\n",
        "assembly_data = assemble_multiple_plasmids_with_repair_templates_for_deletion(genes_to_KO, processed_records, \n",
        "                                                                              repair_templates_data, \n",
        "                                                                              overlap=40)\n",
        "# updating the primer names to something systematic.\n",
        "update_primer_names(assembly_data)\n",
        "print(assembly_data)\n",
        "\n",
        "# Parse through the primer df\n",
        "primer_df = create_primer_df_from_dict(assembly_data)\n",
        "\n",
        "primer_df"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 11,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>template</th>\n",
              "      <th>direction</th>\n",
              "      <th>f_primer_anneal(5-3)</th>\n",
              "      <th>r_primer_anneal(5-3)</th>\n",
              "      <th>f_tm</th>\n",
              "      <th>r_tm</th>\n",
              "      <th>ta</th>\n",
              "      <th>f_primer_sequences(5-3)</th>\n",
              "      <th>r_primer_sequences(5-3)</th>\n",
              "      <th>f_primer_name</th>\n",
              "      <th>r_primer_name</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>SCO5087</td>\n",
              "      <td>upstream</td>\n",
              "      <td>GACGCCGGAAATCCAG</td>\n",
              "      <td>CCCATCTCCCTTCGAC</td>\n",
              "      <td>56</td>\n",
              "      <td>54</td>\n",
              "      <td>58</td>\n",
              "      <td>AAGGCCGCTTTTGCGGGATCTCGTCGAAGGCACTAGAAGGGACGCC...</td>\n",
              "      <td>GACTCCGGTGATGAGGACGCCCCATCTCCCTTCGAC</td>\n",
              "      <td>primer_1</td>\n",
              "      <td>primer_2</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>SCO5087</td>\n",
              "      <td>downstream</td>\n",
              "      <td>GCGTCCTCATCACCGG</td>\n",
              "      <td>GAGTAGAGGCGGCCC</td>\n",
              "      <td>58</td>\n",
              "      <td>57</td>\n",
              "      <td>61</td>\n",
              "      <td>GGCGGTCGAAGGGAGATGGGGCGTCCTCATCACCGG</td>\n",
              "      <td>TCCCCGTCCGGGACCCGCGCGGTCGATCCCCGCATATAGGGAGTAG...</td>\n",
              "      <td>primer_3</td>\n",
              "      <td>primer_4</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>SCO5089</td>\n",
              "      <td>upstream</td>\n",
              "      <td>CCAGGACGCGAAGG</td>\n",
              "      <td>GGCTTTCTCCAGTTCTCG</td>\n",
              "      <td>54</td>\n",
              "      <td>56</td>\n",
              "      <td>58</td>\n",
              "      <td>AAGGCCGCTTTTGCGGGATCTCGTCGAAGGCACTAGAAGGCCAGGA...</td>\n",
              "      <td>CCGGTGTTCTCCCGGCCGGCGGCTTTCTCCAGTTCTCG</td>\n",
              "      <td>primer_5</td>\n",
              "      <td>primer_6</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>5</th>\n",
              "      <td>SCO5089</td>\n",
              "      <td>downstream</td>\n",
              "      <td>GCCGGCCGGGAGA</td>\n",
              "      <td>GGGGGCCTGCTCGT</td>\n",
              "      <td>59</td>\n",
              "      <td>60</td>\n",
              "      <td>63</td>\n",
              "      <td>CCCGAGAACTGGAGAAAGCCGCCGGCCGGGAGA</td>\n",
              "      <td>TCCCCGTCCGGGACCCGCGCGGTCGATCCCCGCATATAGGGGGGGC...</td>\n",
              "      <td>primer_7</td>\n",
              "      <td>primer_8</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>8</th>\n",
              "      <td>SCO5090</td>\n",
              "      <td>upstream</td>\n",
              "      <td>GCCACGCCCGCCG</td>\n",
              "      <td>GCCGCCTCGGCCAGT</td>\n",
              "      <td>63</td>\n",
              "      <td>64</td>\n",
              "      <td>66</td>\n",
              "      <td>AAGGCCGCTTTTGCGGGATCTCGTCGAAGGCACTAGAAGGGCCACG...</td>\n",
              "      <td>GACCTCACGGACCTCGACCGGCCGCCTCGGCCAGT</td>\n",
              "      <td>primer_9</td>\n",
              "      <td>primer_10</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>9</th>\n",
              "      <td>SCO5090</td>\n",
              "      <td>downstream</td>\n",
              "      <td>CGGTCGAGGTCCGT</td>\n",
              "      <td>ACCGTGCGGGACT</td>\n",
              "      <td>56</td>\n",
              "      <td>55</td>\n",
              "      <td>59</td>\n",
              "      <td>GGCGCACTGGCCGAGGCGGCCGGTCGAGGTCCGT</td>\n",
              "      <td>TCCCCGTCCGGGACCCGCGCGGTCGATCCCCGCATATAGGACCGTG...</td>\n",
              "      <td>primer_11</td>\n",
              "      <td>primer_12</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "  template   direction f_primer_anneal(5-3) r_primer_anneal(5-3)  f_tm  r_tm  \\\n",
              "0  SCO5087    upstream     GACGCCGGAAATCCAG     CCCATCTCCCTTCGAC    56    54   \n",
              "1  SCO5087  downstream     GCGTCCTCATCACCGG      GAGTAGAGGCGGCCC    58    57   \n",
              "4  SCO5089    upstream       CCAGGACGCGAAGG   GGCTTTCTCCAGTTCTCG    54    56   \n",
              "5  SCO5089  downstream        GCCGGCCGGGAGA       GGGGGCCTGCTCGT    59    60   \n",
              "8  SCO5090    upstream        GCCACGCCCGCCG      GCCGCCTCGGCCAGT    63    64   \n",
              "9  SCO5090  downstream       CGGTCGAGGTCCGT        ACCGTGCGGGACT    56    55   \n",
              "\n",
              "   ta                            f_primer_sequences(5-3)  \\\n",
              "0  58  AAGGCCGCTTTTGCGGGATCTCGTCGAAGGCACTAGAAGGGACGCC...   \n",
              "1  61               GGCGGTCGAAGGGAGATGGGGCGTCCTCATCACCGG   \n",
              "4  58  AAGGCCGCTTTTGCGGGATCTCGTCGAAGGCACTAGAAGGCCAGGA...   \n",
              "5  63                  CCCGAGAACTGGAGAAAGCCGCCGGCCGGGAGA   \n",
              "8  66  AAGGCCGCTTTTGCGGGATCTCGTCGAAGGCACTAGAAGGGCCACG...   \n",
              "9  59                 GGCGCACTGGCCGAGGCGGCCGGTCGAGGTCCGT   \n",
              "\n",
              "                             r_primer_sequences(5-3) f_primer_name  \\\n",
              "0               GACTCCGGTGATGAGGACGCCCCATCTCCCTTCGAC      primer_1   \n",
              "1  TCCCCGTCCGGGACCCGCGCGGTCGATCCCCGCATATAGGGAGTAG...      primer_3   \n",
              "4             CCGGTGTTCTCCCGGCCGGCGGCTTTCTCCAGTTCTCG      primer_5   \n",
              "5  TCCCCGTCCGGGACCCGCGCGGTCGATCCCCGCATATAGGGGGGGC...      primer_7   \n",
              "8                GACCTCACGGACCTCGACCGGCCGCCTCGGCCAGT      primer_9   \n",
              "9  TCCCCGTCCGGGACCCGCGCGGTCGATCCCCGCATATAGGACCGTG...     primer_11   \n",
              "\n",
              "  r_primer_name  \n",
              "0      primer_2  \n",
              "1      primer_4  \n",
              "4      primer_6  \n",
              "5      primer_8  \n",
              "8     primer_10  \n",
              "9     primer_12  "
            ]
          },
          "execution_count": 11,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "#Find_unique primers\n",
        "unique_df = primer_df.drop_duplicates(keep='first')\n",
        "unique_df"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 12,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>Name</th>\n",
              "      <th>Sequence</th>\n",
              "      <th>Concentration</th>\n",
              "      <th>Purification</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>primer_1</td>\n",
              "      <td>AAGGCCGCTTTTGCGGGATCTCGTCGAAGGCACTAGAAGGGACGCC...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>primer_3</td>\n",
              "      <td>GGCGGTCGAAGGGAGATGGGGCGTCCTCATCACCGG</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>primer_5</td>\n",
              "      <td>AAGGCCGCTTTTGCGGGATCTCGTCGAAGGCACTAGAAGGCCAGGA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>primer_7</td>\n",
              "      <td>CCCGAGAACTGGAGAAAGCCGCCGGCCGGGAGA</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>primer_9</td>\n",
              "      <td>AAGGCCGCTTTTGCGGGATCTCGTCGAAGGCACTAGAAGGGCCACG...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>5</th>\n",
              "      <td>primer_11</td>\n",
              "      <td>GGCGCACTGGCCGAGGCGGCCGGTCGAGGTCCGT</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>6</th>\n",
              "      <td>primer_2</td>\n",
              "      <td>GACTCCGGTGATGAGGACGCCCCATCTCCCTTCGAC</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>7</th>\n",
              "      <td>primer_4</td>\n",
              "      <td>TCCCCGTCCGGGACCCGCGCGGTCGATCCCCGCATATAGGGAGTAG...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>8</th>\n",
              "      <td>primer_6</td>\n",
              "      <td>CCGGTGTTCTCCCGGCCGGCGGCTTTCTCCAGTTCTCG</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>9</th>\n",
              "      <td>primer_8</td>\n",
              "      <td>TCCCCGTCCGGGACCCGCGCGGTCGATCCCCGCATATAGGGGGGGC...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>10</th>\n",
              "      <td>primer_10</td>\n",
              "      <td>GACCTCACGGACCTCGACCGGCCGCCTCGGCCAGT</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>11</th>\n",
              "      <td>primer_12</td>\n",
              "      <td>TCCCCGTCCGGGACCCGCGCGGTCGATCCCCGCATATAGGACCGTG...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "         Name                                           Sequence  \\\n",
              "0    primer_1  AAGGCCGCTTTTGCGGGATCTCGTCGAAGGCACTAGAAGGGACGCC...   \n",
              "1    primer_3               GGCGGTCGAAGGGAGATGGGGCGTCCTCATCACCGG   \n",
              "2    primer_5  AAGGCCGCTTTTGCGGGATCTCGTCGAAGGCACTAGAAGGCCAGGA...   \n",
              "3    primer_7                  CCCGAGAACTGGAGAAAGCCGCCGGCCGGGAGA   \n",
              "4    primer_9  AAGGCCGCTTTTGCGGGATCTCGTCGAAGGCACTAGAAGGGCCACG...   \n",
              "5   primer_11                 GGCGCACTGGCCGAGGCGGCCGGTCGAGGTCCGT   \n",
              "6    primer_2               GACTCCGGTGATGAGGACGCCCCATCTCCCTTCGAC   \n",
              "7    primer_4  TCCCCGTCCGGGACCCGCGCGGTCGATCCCCGCATATAGGGAGTAG...   \n",
              "8    primer_6             CCGGTGTTCTCCCGGCCGGCGGCTTTCTCCAGTTCTCG   \n",
              "9    primer_8  TCCCCGTCCGGGACCCGCGCGGTCGATCCCCGCATATAGGGGGGGC...   \n",
              "10  primer_10                GACCTCACGGACCTCGACCGGCCGCCTCGGCCAGT   \n",
              "11  primer_12  TCCCCGTCCGGGACCCGCGCGGTCGATCCCCGCATATAGGACCGTG...   \n",
              "\n",
              "   Concentration Purification  \n",
              "0           25nm          STD  \n",
              "1           25nm          STD  \n",
              "2           25nm          STD  \n",
              "3           25nm          STD  \n",
              "4           25nm          STD  \n",
              "5           25nm          STD  \n",
              "6           25nm          STD  \n",
              "7           25nm          STD  \n",
              "8           25nm          STD  \n",
              "9           25nm          STD  \n",
              "10          25nm          STD  \n",
              "11          25nm          STD  "
            ]
          },
          "execution_count": 12,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "idt_df = create_idt_order_dataframe(unique_df, concentration=\"25nm\", purification=\"STD\")\n",
        "idt_df"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 13,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/plain": [
              "[Dseqrecord(o13279),\n",
              " Dseqrecord(o13279),\n",
              " Dseqrecord(o13279),\n",
              " Dseqrecord(o13279),\n",
              " Dseqrecord(o13279),\n",
              " Dseqrecord(o13279)]"
            ]
          },
          "execution_count": 13,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "# Contigs\n",
        "assembled_contigs = []\n",
        "for data in assembly_data: \n",
        "    contig_record = data['contig']\n",
        "    contig_record.id = f\"{data['name']}_w_rep\"\n",
        "    assembled_contigs.append(contig_record)\n",
        "assembled_contigs\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 14,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/plain": [
              "[Dseqrecord(o13279),\n",
              " Dseqrecord(o13279),\n",
              " Dseqrecord(o13279),\n",
              " Dseqrecord(o13279),\n",
              " Dseqrecord(o13279),\n",
              " Dseqrecord(o13279)]"
            ]
          },
          "execution_count": 14,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "assembled_contigs"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 15,
      "metadata": {},
      "outputs": [],
      "source": [
        "print_plasmids = False\n",
        "\n",
        "if print_plasmids:\n",
        "    # Write to genbank files\n",
        "    output_directory = \"../../data/plasmids/sgRNA_plasmids_pCRISPR-Cas9_with_repair_templates/\"\n",
        "    for contig in assembled_contigs: \n",
        "        contig.write(f\"{output_directory}/{contig.id}.gb\")"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 16,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>Locus Tag</th>\n",
              "      <th>f_primer_name</th>\n",
              "      <th>r_primer_name</th>\n",
              "      <th>f_primer_sequences(5-3)</th>\n",
              "      <th>r_primer_sequences(5-3)</th>\n",
              "      <th>f_tm</th>\n",
              "      <th>r_tm</th>\n",
              "      <th>ta</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>SCO5087</td>\n",
              "      <td>SCO5087_fwd_checking_primer</td>\n",
              "      <td>SCO5087_rev_checking_primer</td>\n",
              "      <td>GACGATTCGGCCCGTG</td>\n",
              "      <td>CAGGGCGTCCAGGC</td>\n",
              "      <td>59</td>\n",
              "      <td>58</td>\n",
              "      <td>61</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>SCO5089</td>\n",
              "      <td>SCO5089_fwd_checking_primer</td>\n",
              "      <td>SCO5089_rev_checking_primer</td>\n",
              "      <td>CGTACGGCGAACTCGC</td>\n",
              "      <td>GTCCCCGGACGGC</td>\n",
              "      <td>59</td>\n",
              "      <td>57</td>\n",
              "      <td>60</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>SCO5090</td>\n",
              "      <td>SCO5090_fwd_checking_primer</td>\n",
              "      <td>SCO5090_rev_checking_primer</td>\n",
              "      <td>GGCGGCGGCCC</td>\n",
              "      <td>GACGTCCCCCGCGAAC</td>\n",
              "      <td>57</td>\n",
              "      <td>62</td>\n",
              "      <td>61</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "  Locus Tag                f_primer_name                r_primer_name  \\\n",
              "0   SCO5087  SCO5087_fwd_checking_primer  SCO5087_rev_checking_primer   \n",
              "1   SCO5089  SCO5089_fwd_checking_primer  SCO5089_rev_checking_primer   \n",
              "2   SCO5090  SCO5090_fwd_checking_primer  SCO5090_rev_checking_primer   \n",
              "\n",
              "  f_primer_sequences(5-3) r_primer_sequences(5-3)  f_tm  r_tm  ta  \n",
              "0        GACGATTCGGCCCGTG          CAGGGCGTCCAGGC    59    58  61  \n",
              "1        CGTACGGCGAACTCGC           GTCCCCGGACGGC    59    57  60  \n",
              "2             GGCGGCGGCCC        GACGTCCCCCGCGAAC    57    62  61  "
            ]
          },
          "execution_count": 16,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "# Getting checking primers\n",
        "checking_primers_df = checking_primers(path_to_genome, genes_to_KO, \n",
        "                                       flanking_region=flanking_region,\n",
        "                                       target_tm = melting_temperature, \n",
        "                                        primer_concentration = primer_concentration, \n",
        "                                        polymerase = chosen_polymerase)\n",
        "checking_primers_df\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 17,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>Name</th>\n",
              "      <th>Sequence</th>\n",
              "      <th>Concentration</th>\n",
              "      <th>Purification</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>SCO5087_fwd_checking_primer</td>\n",
              "      <td>GACGATTCGGCCCGTG</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>SCO5089_fwd_checking_primer</td>\n",
              "      <td>CGTACGGCGAACTCGC</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>SCO5090_fwd_checking_primer</td>\n",
              "      <td>GGCGGCGGCCC</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>SCO5087_rev_checking_primer</td>\n",
              "      <td>CAGGGCGTCCAGGC</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>SCO5089_rev_checking_primer</td>\n",
              "      <td>GTCCCCGGACGGC</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>5</th>\n",
              "      <td>SCO5090_rev_checking_primer</td>\n",
              "      <td>GACGTCCCCCGCGAAC</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "                          Name          Sequence Concentration Purification\n",
              "0  SCO5087_fwd_checking_primer  GACGATTCGGCCCGTG          25nm          STD\n",
              "1  SCO5089_fwd_checking_primer  CGTACGGCGAACTCGC          25nm          STD\n",
              "2  SCO5090_fwd_checking_primer       GGCGGCGGCCC          25nm          STD\n",
              "3  SCO5087_rev_checking_primer    CAGGGCGTCCAGGC          25nm          STD\n",
              "4  SCO5089_rev_checking_primer     GTCCCCGGACGGC          25nm          STD\n",
              "5  SCO5090_rev_checking_primer  GACGTCCCCCGCGAAC          25nm          STD"
            ]
          },
          "execution_count": 17,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "# making the primers into dseqrecords\n",
        "checking_primers_df_idt = create_idt_order_dataframe(checking_primers_df)\n",
        "checking_primers_df_idt\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 18,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>Name</th>\n",
              "      <th>Sequence</th>\n",
              "      <th>Concentration</th>\n",
              "      <th>Purification</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>primer_1</td>\n",
              "      <td>AAGGCCGCTTTTGCGGGATCTCGTCGAAGGCACTAGAAGGGACGCC...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>primer_3</td>\n",
              "      <td>GGCGGTCGAAGGGAGATGGGGCGTCCTCATCACCGG</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>primer_5</td>\n",
              "      <td>AAGGCCGCTTTTGCGGGATCTCGTCGAAGGCACTAGAAGGCCAGGA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>primer_7</td>\n",
              "      <td>CCCGAGAACTGGAGAAAGCCGCCGGCCGGGAGA</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>primer_9</td>\n",
              "      <td>AAGGCCGCTTTTGCGGGATCTCGTCGAAGGCACTAGAAGGGCCACG...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>5</th>\n",
              "      <td>primer_11</td>\n",
              "      <td>GGCGCACTGGCCGAGGCGGCCGGTCGAGGTCCGT</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>6</th>\n",
              "      <td>primer_2</td>\n",
              "      <td>GACTCCGGTGATGAGGACGCCCCATCTCCCTTCGAC</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>7</th>\n",
              "      <td>primer_4</td>\n",
              "      <td>TCCCCGTCCGGGACCCGCGCGGTCGATCCCCGCATATAGGGAGTAG...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>8</th>\n",
              "      <td>primer_6</td>\n",
              "      <td>CCGGTGTTCTCCCGGCCGGCGGCTTTCTCCAGTTCTCG</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>9</th>\n",
              "      <td>primer_8</td>\n",
              "      <td>TCCCCGTCCGGGACCCGCGCGGTCGATCCCCGCATATAGGGGGGGC...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>10</th>\n",
              "      <td>primer_10</td>\n",
              "      <td>GACCTCACGGACCTCGACCGGCCGCCTCGGCCAGT</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>11</th>\n",
              "      <td>primer_12</td>\n",
              "      <td>TCCCCGTCCGGGACCCGCGCGGTCGATCCCCGCATATAGGACCGTG...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>SCO5087_fwd_checking_primer</td>\n",
              "      <td>GACGATTCGGCCCGTG</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>SCO5089_fwd_checking_primer</td>\n",
              "      <td>CGTACGGCGAACTCGC</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>SCO5090_fwd_checking_primer</td>\n",
              "      <td>GGCGGCGGCCC</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>SCO5087_rev_checking_primer</td>\n",
              "      <td>CAGGGCGTCCAGGC</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>SCO5089_rev_checking_primer</td>\n",
              "      <td>GTCCCCGGACGGC</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>5</th>\n",
              "      <td>SCO5090_rev_checking_primer</td>\n",
              "      <td>GACGTCCCCCGCGAAC</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "                           Name  \\\n",
              "0                      primer_1   \n",
              "1                      primer_3   \n",
              "2                      primer_5   \n",
              "3                      primer_7   \n",
              "4                      primer_9   \n",
              "5                     primer_11   \n",
              "6                      primer_2   \n",
              "7                      primer_4   \n",
              "8                      primer_6   \n",
              "9                      primer_8   \n",
              "10                    primer_10   \n",
              "11                    primer_12   \n",
              "0   SCO5087_fwd_checking_primer   \n",
              "1   SCO5089_fwd_checking_primer   \n",
              "2   SCO5090_fwd_checking_primer   \n",
              "3   SCO5087_rev_checking_primer   \n",
              "4   SCO5089_rev_checking_primer   \n",
              "5   SCO5090_rev_checking_primer   \n",
              "\n",
              "                                             Sequence Concentration  \\\n",
              "0   AAGGCCGCTTTTGCGGGATCTCGTCGAAGGCACTAGAAGGGACGCC...          25nm   \n",
              "1                GGCGGTCGAAGGGAGATGGGGCGTCCTCATCACCGG          25nm   \n",
              "2   AAGGCCGCTTTTGCGGGATCTCGTCGAAGGCACTAGAAGGCCAGGA...          25nm   \n",
              "3                   CCCGAGAACTGGAGAAAGCCGCCGGCCGGGAGA          25nm   \n",
              "4   AAGGCCGCTTTTGCGGGATCTCGTCGAAGGCACTAGAAGGGCCACG...          25nm   \n",
              "5                  GGCGCACTGGCCGAGGCGGCCGGTCGAGGTCCGT          25nm   \n",
              "6                GACTCCGGTGATGAGGACGCCCCATCTCCCTTCGAC          25nm   \n",
              "7   TCCCCGTCCGGGACCCGCGCGGTCGATCCCCGCATATAGGGAGTAG...          25nm   \n",
              "8              CCGGTGTTCTCCCGGCCGGCGGCTTTCTCCAGTTCTCG          25nm   \n",
              "9   TCCCCGTCCGGGACCCGCGCGGTCGATCCCCGCATATAGGGGGGGC...          25nm   \n",
              "10                GACCTCACGGACCTCGACCGGCCGCCTCGGCCAGT          25nm   \n",
              "11  TCCCCGTCCGGGACCCGCGCGGTCGATCCCCGCATATAGGACCGTG...          25nm   \n",
              "0                                    GACGATTCGGCCCGTG          25nm   \n",
              "1                                    CGTACGGCGAACTCGC          25nm   \n",
              "2                                         GGCGGCGGCCC          25nm   \n",
              "3                                      CAGGGCGTCCAGGC          25nm   \n",
              "4                                       GTCCCCGGACGGC          25nm   \n",
              "5                                    GACGTCCCCCGCGAAC          25nm   \n",
              "\n",
              "   Purification  \n",
              "0           STD  \n",
              "1           STD  \n",
              "2           STD  \n",
              "3           STD  \n",
              "4           STD  \n",
              "5           STD  \n",
              "6           STD  \n",
              "7           STD  \n",
              "8           STD  \n",
              "9           STD  \n",
              "10          STD  \n",
              "11          STD  \n",
              "0           STD  \n",
              "1           STD  \n",
              "2           STD  \n",
              "3           STD  \n",
              "4           STD  \n",
              "5           STD  "
            ]
          },
          "execution_count": 18,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "\n",
        "full_idt = pd.concat([idt_df, checking_primers_df_idt])\n",
        "full_idt"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {},
      "outputs": [],
      "source": [
        "\n",
        "\n",
        "input_files = [\n",
        "    {\"name\": \"input_genome.gb\", \"content\": genome},\n",
        "    {\"name\": \"input_plasmid.gb\", \"content\": clean_plasmid}\n",
        "]\n",
        "\n",
        "output_files = [\n",
        "    {\"name\": \"cBEST_w_sgRNAs.gb\", \"content\": assembled_contigs}, # LIST OF Dseqrecords\n",
        "    {\"name\": \"primer_df.csv\", \"content\": primer_df},\n",
        "    {\"name\": \"full_idt.csv\", \"content\": full_idt},\n",
        "    {\"name\": \"sgrna_df.csv\", \"content\": sgrna_df},\n",
        "    {\"name\": \"filtered_df.csv\", \"content\": filtered_df}\n",
        "]\n",
        "\n",
        "input_values = {\n",
        "    \"genes_to_knockout\": genes_to_KO,\n",
        "\n",
        "\n",
        "    \"filtering_metrics\": {\n",
        "        \"gc_upper\": gc_upper,\n",
        "        \"gc_lower\": gc_lower,\n",
        "        \"off_target_seed\": off_target_seed,\n",
        "        \"off_target_upper\": off_target_upper,\n",
        "        \"cas_type\": cas_type,\n",
        "        \"number_of_sgRNAs_per_group\": number_of_sgRNAs_per_group,\n",
        "\n",
        "    },  \n",
        "    \"polymerase_settings\": {\n",
        "        \"chosen_polymerase\": chosen_polymerase,\n",
        "        \"melting_temperature\": melting_temperature,\n",
        "        \"primer_concentration\": primer_concentration,\n",
        "        \"primer_number_increment\": primer_number_increment,\n",
        "        \"flanking_region\": flanking_region\n",
        "    },\n",
        "    \"overlapping_sequences\": {\n",
        "        \"up_homology\": str(up_homology),\n",
        "        \"dw_homology\": str(dw_homology)\n",
        "    }\n",
        "}\n",
        "\n",
        "\n",
        "# Paths to Markdown files\n",
        "markdown_file_paths = [\n",
        "    \"../../protocols/conjugation_protcol.md\",\n",
        "    \"../../protocols/single_target_crispr_plasmid_protcol.md\"\n",
        "\n",
        "]\n",
        "\n",
        "# Data and time\n",
        "timestamp = datetime.utcnow().isoformat()\n",
        "\n",
        "generate_project_directory_structure(\n",
        "    project_name=f\"CRISPR_cas9_inframe_deletion_workflow_{timestamp}\",\n",
        "    input_files=input_files,\n",
        "    output_files=output_files,\n",
        "    input_values=input_values,\n",
        "    markdown_file_paths=markdown_file_paths\n",
        "\n",
        ")"
      ]
    }
  ],
  "metadata": {
    "kernelspec": {
      "display_name": "constrain",
      "language": "python",
      "name": "python3"
    },
    "language_info": {
      "codemirror_mode": {
        "name": "ipython",
        "version": 3
      },
      "file_extension": ".py",
      "mimetype": "text/x-python",
      "name": "python",
      "nbconvert_exporter": "python",
      "pygments_lexer": "ipython3",
      "version": "3.11.9"
    }
  },
  "nbformat": 4,
  "nbformat_minor": 2
}
