{
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "# W5-CRISPR-Cas9-inframe_deletion"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "In Workflow 5, CRISPR-Cas9 combines ssDNA bridging and Gibson cloning to create in-frame deletions (Figure 6A). This ensures that the downstream genes are not disrupted, which is crucial for functional studies of gene knockouts. This workflow can be used to study gene function by knocking out specific genes in streptomycetes and observing the resulting phenotypic changes. It can also be used to create markerless deletions, which are important for constructing clean genetic backgrounds for further genetic manipulations. To get started, download the pCRISPR-Cas9 file and a genome file (e.g., S. coelicolor A3), then select the genes to delete. StreptoAIM will generate the necessary primers and plasmid constructs (S5).\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 1,
      "metadata": {},
      "outputs": [],
      "source": [
        "import sys\n",
        "import os\n",
        "from Bio.Restriction import StuI\n",
        "from pydna.dseqrecord import Dseqrecord\n",
        "import pandas as pd\n",
        "from datetime import datetime\n",
        "\n",
        "\n",
        "# Ensure the src directory is in the Python path\n",
        "project_root = os.path.abspath(os.path.join(os.getcwd(), '../../'))\n",
        "if project_root not in sys.path:\n",
        "    sys.path.insert(0, project_root)\n",
        "\n",
        "from streptocad.sequence_loading.sequence_loading import (\n",
        "    load_and_process_gene_sequences,\n",
        "    load_and_process_genome_sequences,\n",
        "    load_and_process_plasmid, \n",
        "    check_and_convert_input,\n",
        "    annotate_dseqrecord,\n",
        "    process_specified_gene_sequences_from_record)\n",
        "\n",
        "from streptocad.utils import polymerase_dict, create_primer_df_from_dict,ProjectDirectory,extract_metadata_to_dataframe\n",
        "from streptocad.primers.primer_generation import create_idt_order_dataframe\n",
        "from streptocad.cloning.ssDNA_bridging import assemble_plasmids_by_ssDNA_bridging,make_ssDNA_oligos\n",
        "from streptocad.crispr.guideRNAcas3_9_12 import extract_sgRNAs, SgRNAargs\n",
        "from streptocad.cloning.gibson_cloning import (\n",
        "    find_up_dw_repair_templates,\n",
        "    assemble_multiple_plasmids_with_repair_templates_for_deletion,\n",
        "    update_primer_names\n",
        ")\n",
        "\n",
        "from streptocad.cloning.plasmid_processing import check_plasmid_restriction_sites, determine_workflow_order_for_plasmids\n",
        "from streptocad.primers.primer_generation import checking_primers, primers_to_IDT, find_best_check_primers_from_genome"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "## INPUT"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 2,
      "metadata": {},
      "outputs": [],
      "source": [
        "# Inputs\n",
        "# 1 Add genome of choice (genbank, fasta)\n",
        "path_to_genome = '../../data/genomes/Streptomyces_coelicolor_A3_chromosome.gb'\n",
        "genome = load_and_process_genome_sequences(path_to_genome)[0]\n",
        "\n",
        "# 2 Add plasmid \n",
        "path_to_plasmid = '../../data/plasmids/pCRISPR–Cas9_plasmid_addgene.gbk'\n",
        "clean_plasmid = load_and_process_plasmid(path_to_plasmid)\n",
        "\n",
        "# 3 Choose genes to knock out (list)\n",
        "#genes_to_KO = ['SCO5087', 'SCO5089', 'SCO5090','SCO5091', 'SCO5092' ]\n",
        "genes_to_KO = ['SCO5087']#, 'SCO5089', 'SCO5090','SCO5091', 'SCO5092' ]\n",
        "\n",
        "#genes_to_KO = ['80000-100000', '4000-7000', '9000-14000','15000-20000']\n",
        "\n",
        "\n",
        "#### Advanced settings ####\n",
        "# 4 Filtering metrics for sgRNAs\n",
        "gc_upper = 0.72\n",
        "gc_lower = 0.2\n",
        "off_target_seed = 13\n",
        "off_target_upper = 10\n",
        "cas_type='cas9'\n",
        "number_of_sgRNAs_per_group = 2\n",
        "\n",
        "# 5 Choose polymerase and target melting temperature\n",
        "chosen_polymerase = polymerase_dict['Phusion High-Fidelity DNA Polymerase (GC Buffer)']\n",
        "melting_temperature = 65\n",
        "primer_concentration = 0.4 \n",
        "primer_number_increment = 23\n",
        "flanking_region = 500\n",
        "\n",
        "# 6 Choose overlapping sequences for our plasmid we can use the following\n",
        "#As per the article **\"CRISPR–Cas9, CRISPRi and CRISPR-BEST-mediated genetic manipulation in streptomycetes\"** we need the following oligoes: \n",
        "#CGGTTGGTAGGATCGACGGC **-N20-** GTTTTAGAGCTAGAAATAGC\n",
        "up_homology = Dseqrecord('CGGTTGGTAGGATCGACGGC')\n",
        "dw_homology = Dseqrecord('GTTTTAGAGCTAGAAATAGC')\n",
        "\n",
        "# 7 extra settings\n",
        "#### Advanced settings ####\n",
        "repair_length = 1000\n",
        "overlap_length = 40 "
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 3,
      "metadata": {},
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "pCRISPR-Cas9\n"
          ]
        }
      ],
      "source": [
        "print(clean_plasmid.id)"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "# Computation"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 4,
      "metadata": {},
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "False\n"
          ]
        },
        {
          "data": {
            "text/plain": [
              "25824"
            ]
          },
          "execution_count": 4,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "target_dict, genes_to_KO, annotation_input = check_and_convert_input(genes_to_KO)\n",
        "\n",
        "print(annotation_input)\n",
        "if annotation_input == True:\n",
        "    genome = annotate_dseqrecord(genome, target_dict)\n",
        "\n",
        "\n",
        "len(genome.features)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 5,
      "metadata": {},
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "Pam was found outside designated locus_tag: SCO5087. To incorporate this extent borders. Skipping to next locus tag.\n",
            "sgRNA generated were outside the designated border in SCO5087. To incorporate this extent borders. Skipping to next locus tag.\n",
            "Pam was found outside designated locus_tag: SCO5087. To incorporate this extent borders. Skipping to next locus tag.\n",
            "sgRNA generated were outside the designated border in SCO5087. To incorporate this extent borders. Skipping to next locus tag.\n",
            "sgRNA generated were outside the designated border in SCO5087. To incorporate this extent borders. Skipping to next locus tag.\n"
          ]
        },
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>strain_name</th>\n",
              "      <th>locus_tag</th>\n",
              "      <th>gene_loc</th>\n",
              "      <th>gene_strand</th>\n",
              "      <th>sgrna_strand</th>\n",
              "      <th>sgrna_loc</th>\n",
              "      <th>gc</th>\n",
              "      <th>pam</th>\n",
              "      <th>sgrna</th>\n",
              "      <th>sgrna_seed_sequence</th>\n",
              "      <th>off_target_count</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>163</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5087</td>\n",
              "      <td>5529801</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>1172</td>\n",
              "      <td>0.65</td>\n",
              "      <td>GGG</td>\n",
              "      <td>TCGATGGTCGGCCACTCGCT</td>\n",
              "      <td>TCGGCCACTCGCT</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>165</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5087</td>\n",
              "      <td>5529801</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>1160</td>\n",
              "      <td>0.50</td>\n",
              "      <td>CGG</td>\n",
              "      <td>AGCTCGATCAAGTCGATGGT</td>\n",
              "      <td>TCAAGTCGATGGT</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>170</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5087</td>\n",
              "      <td>5529801</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>1093</td>\n",
              "      <td>0.65</td>\n",
              "      <td>CGG</td>\n",
              "      <td>GAACGACCGCCACGAGACAG</td>\n",
              "      <td>CGCCACGAGACAG</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>176</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5087</td>\n",
              "      <td>5529801</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>997</td>\n",
              "      <td>0.65</td>\n",
              "      <td>GGG</td>\n",
              "      <td>CGAGATGGCCGAGACCATCC</td>\n",
              "      <td>GCCGAGACCATCC</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>177</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5087</td>\n",
              "      <td>5529801</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>996</td>\n",
              "      <td>0.65</td>\n",
              "      <td>CGG</td>\n",
              "      <td>GCGAGATGGCCGAGACCATC</td>\n",
              "      <td>GGCCGAGACCATC</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>...</th>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>255</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5087</td>\n",
              "      <td>5529801</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>308</td>\n",
              "      <td>0.70</td>\n",
              "      <td>GGG</td>\n",
              "      <td>GACTTCGACCCGGTCGCCGA</td>\n",
              "      <td>ACCCGGTCGCCGA</td>\n",
              "      <td>2</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>173</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5087</td>\n",
              "      <td>5529801</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>1055</td>\n",
              "      <td>0.50</td>\n",
              "      <td>CGG</td>\n",
              "      <td>ATCGACTACATCAACGCGCA</td>\n",
              "      <td>ACATCAACGCGCA</td>\n",
              "      <td>2</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>78</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5087</td>\n",
              "      <td>5529801</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>745</td>\n",
              "      <td>0.65</td>\n",
              "      <td>GGG</td>\n",
              "      <td>CGACGACGATCGGGGTGATC</td>\n",
              "      <td>GATCGGGGTGATC</td>\n",
              "      <td>2</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>200</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5087</td>\n",
              "      <td>5529801</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>706</td>\n",
              "      <td>0.65</td>\n",
              "      <td>CGG</td>\n",
              "      <td>CAGCGCCGACGTGATGTTCG</td>\n",
              "      <td>GACGTGATGTTCG</td>\n",
              "      <td>3</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>191</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5087</td>\n",
              "      <td>5529801</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>838</td>\n",
              "      <td>0.70</td>\n",
              "      <td>TGG</td>\n",
              "      <td>TACGCGGGACGGCTTCGTGC</td>\n",
              "      <td>GACGGCTTCGTGC</td>\n",
              "      <td>3</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>147 rows × 11 columns</p>\n",
              "</div>"
            ],
            "text/plain": [
              "     strain_name locus_tag  gene_loc  gene_strand  sgrna_strand  sgrna_loc  \\\n",
              "163  NC_003888.3   SCO5087   5529801            1             1       1172   \n",
              "165  NC_003888.3   SCO5087   5529801            1             1       1160   \n",
              "170  NC_003888.3   SCO5087   5529801            1             1       1093   \n",
              "176  NC_003888.3   SCO5087   5529801            1             1        997   \n",
              "177  NC_003888.3   SCO5087   5529801            1             1        996   \n",
              "..           ...       ...       ...          ...           ...        ...   \n",
              "255  NC_003888.3   SCO5087   5529801            1             1        308   \n",
              "173  NC_003888.3   SCO5087   5529801            1             1       1055   \n",
              "78   NC_003888.3   SCO5087   5529801            1            -1        745   \n",
              "200  NC_003888.3   SCO5087   5529801            1             1        706   \n",
              "191  NC_003888.3   SCO5087   5529801            1             1        838   \n",
              "\n",
              "       gc  pam                 sgrna sgrna_seed_sequence  off_target_count  \n",
              "163  0.65  GGG  TCGATGGTCGGCCACTCGCT       TCGGCCACTCGCT                 0  \n",
              "165  0.50  CGG  AGCTCGATCAAGTCGATGGT       TCAAGTCGATGGT                 0  \n",
              "170  0.65  CGG  GAACGACCGCCACGAGACAG       CGCCACGAGACAG                 0  \n",
              "176  0.65  GGG  CGAGATGGCCGAGACCATCC       GCCGAGACCATCC                 0  \n",
              "177  0.65  CGG  GCGAGATGGCCGAGACCATC       GGCCGAGACCATC                 0  \n",
              "..    ...  ...                   ...                 ...               ...  \n",
              "255  0.70  GGG  GACTTCGACCCGGTCGCCGA       ACCCGGTCGCCGA                 2  \n",
              "173  0.50  CGG  ATCGACTACATCAACGCGCA       ACATCAACGCGCA                 2  \n",
              "78   0.65  GGG  CGACGACGATCGGGGTGATC       GATCGGGGTGATC                 2  \n",
              "200  0.65  CGG  CAGCGCCGACGTGATGTTCG       GACGTGATGTTCG                 3  \n",
              "191  0.70  TGG  TACGCGGGACGGCTTCGTGC       GACGGCTTCGTGC                 3  \n",
              "\n",
              "[147 rows x 11 columns]"
            ]
          },
          "execution_count": 5,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "# Initialize SgRNAargs with desired parameters\n",
        "args = SgRNAargs(genome, \n",
        "                genes_to_KO,\n",
        "                step=['find', 'filter'],\n",
        "                gc_upper = gc_upper,\n",
        "                gc_lower = gc_lower,\n",
        "                off_target_seed = off_target_seed,\n",
        "                off_target_upper = off_target_upper,\n",
        "                cas_type='cas9'\n",
        "                )\n",
        "\n",
        "sgrna_df = extract_sgRNAs(args)\n",
        "sgrna_df"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 6,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>strain_name</th>\n",
              "      <th>locus_tag</th>\n",
              "      <th>gene_loc</th>\n",
              "      <th>gene_strand</th>\n",
              "      <th>sgrna_strand</th>\n",
              "      <th>sgrna_loc</th>\n",
              "      <th>gc</th>\n",
              "      <th>pam</th>\n",
              "      <th>sgrna</th>\n",
              "      <th>sgrna_seed_sequence</th>\n",
              "      <th>off_target_count</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>163</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5087</td>\n",
              "      <td>5529801</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>1172</td>\n",
              "      <td>0.65</td>\n",
              "      <td>GGG</td>\n",
              "      <td>TCGATGGTCGGCCACTCGCT</td>\n",
              "      <td>TCGGCCACTCGCT</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>165</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5087</td>\n",
              "      <td>5529801</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>1160</td>\n",
              "      <td>0.50</td>\n",
              "      <td>CGG</td>\n",
              "      <td>AGCTCGATCAAGTCGATGGT</td>\n",
              "      <td>TCAAGTCGATGGT</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "     strain_name locus_tag  gene_loc  gene_strand  sgrna_strand  sgrna_loc  \\\n",
              "163  NC_003888.3   SCO5087   5529801            1             1       1172   \n",
              "165  NC_003888.3   SCO5087   5529801            1             1       1160   \n",
              "\n",
              "       gc  pam                 sgrna sgrna_seed_sequence  off_target_count  \n",
              "163  0.65  GGG  TCGATGGTCGGCCACTCGCT       TCGGCCACTCGCT                 0  \n",
              "165  0.50  CGG  AGCTCGATCAAGTCGATGGT       TCAAGTCGATGGT                 0  "
            ]
          },
          "execution_count": 6,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "# Filter the DataFrame to retain only up to 5 sgRNA sequences per locus_tag\n",
        "filtered_df = sgrna_df.groupby('locus_tag').head(number_of_sgRNAs_per_group)\n",
        "filtered_df"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "## Output"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 7,
      "metadata": {},
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "SCO5087_loc_1172\n"
          ]
        },
        {
          "data": {
            "text/plain": [
              "[Contig(o11279), Contig(o11279)]"
            ]
          },
          "execution_count": 7,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "# MAke oligoes\n",
        "list_of_ssDNAs = make_ssDNA_oligos(filtered_df, upstream_ovh = up_homology,\n",
        "                      downstream_ovh=dw_homology)\n",
        "print(list_of_ssDNAs[0].name)\n",
        "\n",
        "# cut plasmid\n",
        "from Bio.Restriction import NcoI\n",
        "linearized_plasmid = sorted(clean_plasmid.cut(NcoI), key=lambda x: len(x), reverse=True)[0]\n",
        "#print(linearized_plasmid)\n",
        "\n",
        "sgRNA_vectors = assemble_plasmids_by_ssDNA_bridging(list_of_ssDNAs,linearized_plasmid)\n",
        "sgRNA_vectors"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 8,
      "metadata": {},
      "outputs": [],
      "source": [
        "# Constructing a meaningful name, ID, and description for the assembled plasmid using user input\n",
        "targeting_info = []\n",
        "for index, row in filtered_df.iterrows():\n",
        "    formatted_str = f\"pCas9_{row['locus_tag']}({row['sgrna_loc']})\"\n",
        "    targeting_info.append(formatted_str)\n",
        "\n",
        "for i in range(len(sgRNA_vectors)):\n",
        "    sgRNA_vectors[i].name = f'{targeting_info[i]}'\n",
        "    sgRNA_vectors[i].id = sgRNA_vectors[i].name  \n",
        "    sgRNA_vectors[i].description = f'CRISPR-Cas9 targeting {\", \".join(genes_to_KO)} for single gene knockout, assembled using StreptoCAD.'"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 9,
      "metadata": {},
      "outputs": [],
      "source": [
        "print_plasmids = False\n",
        "\n",
        "if print_plasmids: \n",
        "    for vector in sgRNA_vectors: \n",
        "        vector.write(f\"../../data/plasmids/sgRNA_plasmids_pCRISPR-Cas9/{vector.id}.gb\")"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "### IDT primers"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 10,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>Name</th>\n",
              "      <th>Sequence</th>\n",
              "      <th>Concentration</th>\n",
              "      <th>Purification</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>SCO5087_loc_1172</td>\n",
              "      <td>CGGTTGGTAGGATCGACGGCTCGATGGTCGGCCACTCGCTGTTTTA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>SCO5087_loc_1160</td>\n",
              "      <td>CGGTTGGTAGGATCGACGGCAGCTCGATCAAGTCGATGGTGTTTTA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "               Name                                           Sequence  \\\n",
              "0  SCO5087_loc_1172  CGGTTGGTAGGATCGACGGCTCGATGGTCGGCCACTCGCTGTTTTA...   \n",
              "1  SCO5087_loc_1160  CGGTTGGTAGGATCGACGGCAGCTCGATCAAGTCGATGGTGTTTTA...   \n",
              "\n",
              "  Concentration Purification  \n",
              "0          25nm          STD  \n",
              "1          25nm          STD  "
            ]
          },
          "execution_count": 10,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "idt_primers=primers_to_IDT(list_of_ssDNAs)\n",
        "idt_primers"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "# In frame deletion"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 11,
      "metadata": {},
      "outputs": [],
      "source": [
        "in_frame_deletion = True\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 12,
      "metadata": {},
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "pCas9_SCO5087(1172)\n",
            "pCas9_SCO5087(1160)\n",
            "\n",
            "Processing gene: SCO5087\n",
            "Checking plasmid: pCas9_SCO5087(1172)\n",
            "Match found in plasmid: pCas9_SCO5087(1172)\n",
            "Repair template match found for gene: SCO5087\n",
            "Record added for gene: SCO5087\n",
            "Checking plasmid: pCas9_SCO5087(1160)\n",
            "Match found in plasmid: pCas9_SCO5087(1160)\n",
            "Repair template match found for gene: SCO5087\n",
            "Record added for gene: SCO5087\n",
            "[{'gene_name': 'SCO5087', 'name': 'pCas9_SCO5087(1172)', 'contig': Dseqrecord(o13279), 'up_forwar_p_name': 'SCO5087_repair_up_forwar_p', 'up_reverse_p_name': 'SCO5087_repair_up_reverse_p', 'up_forwar_p': SCO5087_repair_up_forwar_p 56-mer:5'-AAGGCCGCTTTTGCG..CAG-3', 'up_reverse_p': SCO5087_repair_up_reverse_p 36-mer:5'-GACTCCGGTGATGAG..GAC-3', 'up_forwar_primer_str': 'AAGGCCGCTTTTGCGGGATCTCGTCGAAGGCACTAGAAGGGACGCCGGAAATCCAG', 'up_reverse_primer_str': 'GACTCCGGTGATGAGGACGCCCCATCTCCCTTCGAC', 'up_forwar_p_anneal': Seq('GACGCCGGAAATCCAG'), 'up_reverse_p_anneal': Seq('CCCATCTCCCTTCGAC'), 'tm_up_forwar_p': 56, 'tm_up_reverse_p': 54, 'ta_up': 58, 'dw_forwar_p_name': 'SCO5087_repair_dw_forwar_p', 'dw_reverse_p_name': 'SCO5087_repair_dw_reverse_p', 'dw_forwar_p': SCO5087_repair_dw_forwar_p 36-mer:5'-GGCGGTCGAAGGGAG..CGG-3', 'dw_reverse_p': SCO5087_repair_dw_reverse_p 55-mer:5'-TCCCCGTCCGGGACC..CCC-3', 'dw_forwar_primer_str': 'GGCGGTCGAAGGGAGATGGGGCGTCCTCATCACCGG', 'dw_reverse_primer_str': 'TCCCCGTCCGGGACCCGCGCGGTCGATCCCCGCATATAGGGAGTAGAGGCGGCCC', 'dw_forwar_p_anneal': Seq('GCGTCCTCATCACCGG'), 'dw_reverse_p_anneal': Seq('GAGTAGAGGCGGCCC'), 'tm_dw_forwar_p': 58, 'tm_dw_reverse_p': 57, 'ta_dw': 61}, {'gene_name': 'SCO5087', 'name': 'pCas9_SCO5087(1160)', 'contig': Dseqrecord(o13279), 'up_forwar_p_name': 'SCO5087_repair_up_forwar_p', 'up_reverse_p_name': 'SCO5087_repair_up_reverse_p', 'up_forwar_p': f1000 56-mer:5'-AAGGCCGCTTTTGCG..CAG-3', 'up_reverse_p': r1000 36-mer:5'-GACTCCGGTGATGAG..GAC-3', 'up_forwar_primer_str': 'AAGGCCGCTTTTGCGGGATCTCGTCGAAGGCACTAGAAGGGACGCCGGAAATCCAG', 'up_reverse_primer_str': 'GACTCCGGTGATGAGGACGCCCCATCTCCCTTCGAC', 'up_forwar_p_anneal': Seq('GACGCCGGAAATCCAG'), 'up_reverse_p_anneal': Seq('CCCATCTCCCTTCGAC'), 'tm_up_forwar_p': 56, 'tm_up_reverse_p': 54, 'ta_up': 58, 'dw_forwar_p_name': 'SCO5087_repair_dw_forwar_p', 'dw_reverse_p_name': 'SCO5087_repair_dw_reverse_p', 'dw_forwar_p': f1000 36-mer:5'-GGCGGTCGAAGGGAG..CGG-3', 'dw_reverse_p': r1000 55-mer:5'-TCCCCGTCCGGGACC..CCC-3', 'dw_forwar_primer_str': 'GGCGGTCGAAGGGAGATGGGGCGTCCTCATCACCGG', 'dw_reverse_primer_str': 'TCCCCGTCCGGGACCCGCGCGGTCGATCCCCGCATATAGGGAGTAGAGGCGGCCC', 'dw_forwar_p_anneal': Seq('GCGTCCTCATCACCGG'), 'dw_reverse_p_anneal': Seq('GAGTAGAGGCGGCCC'), 'tm_dw_forwar_p': 58, 'tm_dw_reverse_p': 57, 'ta_dw': 61}]\n"
          ]
        }
      ],
      "source": [
        "if in_frame_deletion:\n",
        "    \n",
        "\n",
        "    # Make repair templates\n",
        "    repair_templates_data = find_up_dw_repair_templates(genome, \n",
        "                                                        genes_to_KO, \n",
        "                                                        target_tm=melting_temperature, \n",
        "                                                        primer_tm_kwargs={'conc':primer_concentration, 'prodcode':chosen_polymerase} , \n",
        "                                                        repair_length=repair_length)\n",
        "\n",
        "    # Digest the plasmids\n",
        "    processed_records = [Dseqrecord(record, circular=True).cut(StuI)[0] for record in sgRNA_vectors]\n",
        "\n",
        "    # Rename them appropriately\n",
        "    for i in range(len(processed_records)):\n",
        "        processed_records[i].name = sgRNA_vectors[i].name\n",
        "        print(processed_records[i].name)\n",
        "\n",
        "    # Assembly \n",
        "    assembly_data = assemble_multiple_plasmids_with_repair_templates_for_deletion(genes_to_KO, processed_records, \n",
        "                                                                                repair_templates_data, \n",
        "                                                                                overlap=overlap_length)\n",
        "    # updating the primer names to something systematic.\n",
        "    update_primer_names(assembly_data)\n",
        "    print(assembly_data)\n",
        "\n",
        "    # Parse through the primer df\n",
        "    primer_df = create_primer_df_from_dict(assembly_data)\n",
        "    unique_df = primer_df.drop_duplicates(keep='first')\n",
        "\n",
        "    # IDT df\n",
        "    idt_df = create_idt_order_dataframe(unique_df, concentration=\"25nm\", purification=\"STD\")\n",
        "\n",
        "    # Contigs\n",
        "    assembled_contigs = []\n",
        "    for data in assembly_data: \n",
        "        contig_record = data['contig']\n",
        "        contig_record.id = f\"{data['name']}_w_rep\"\n",
        "        contig_record.name = f\"{data['name']}_w_rep\"\n",
        "        assembled_contigs.append(contig_record)\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 13,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>template</th>\n",
              "      <th>direction</th>\n",
              "      <th>f_primer_anneal(5-3)</th>\n",
              "      <th>r_primer_anneal(5-3)</th>\n",
              "      <th>f_tm</th>\n",
              "      <th>r_tm</th>\n",
              "      <th>ta</th>\n",
              "      <th>f_primer_sequences(5-3)</th>\n",
              "      <th>r_primer_sequences(5-3)</th>\n",
              "      <th>f_primer_name</th>\n",
              "      <th>r_primer_name</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>SCO5087</td>\n",
              "      <td>upstream</td>\n",
              "      <td>GACGCCGGAAATCCAG</td>\n",
              "      <td>CCCATCTCCCTTCGAC</td>\n",
              "      <td>56</td>\n",
              "      <td>54</td>\n",
              "      <td>58</td>\n",
              "      <td>AAGGCCGCTTTTGCGGGATCTCGTCGAAGGCACTAGAAGGGACGCC...</td>\n",
              "      <td>GACTCCGGTGATGAGGACGCCCCATCTCCCTTCGAC</td>\n",
              "      <td>SCO5087_repair_up_forwar_p</td>\n",
              "      <td>SCO5087_repair_up_reverse_p</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>SCO5087</td>\n",
              "      <td>downstream</td>\n",
              "      <td>GCGTCCTCATCACCGG</td>\n",
              "      <td>GAGTAGAGGCGGCCC</td>\n",
              "      <td>58</td>\n",
              "      <td>57</td>\n",
              "      <td>61</td>\n",
              "      <td>GGCGGTCGAAGGGAGATGGGGCGTCCTCATCACCGG</td>\n",
              "      <td>TCCCCGTCCGGGACCCGCGCGGTCGATCCCCGCATATAGGGAGTAG...</td>\n",
              "      <td>SCO5087_repair_dw_forwar_p</td>\n",
              "      <td>SCO5087_repair_dw_reverse_p</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "  template   direction f_primer_anneal(5-3) r_primer_anneal(5-3)  f_tm  r_tm  \\\n",
              "0  SCO5087    upstream     GACGCCGGAAATCCAG     CCCATCTCCCTTCGAC    56    54   \n",
              "1  SCO5087  downstream     GCGTCCTCATCACCGG      GAGTAGAGGCGGCCC    58    57   \n",
              "\n",
              "   ta                            f_primer_sequences(5-3)  \\\n",
              "0  58  AAGGCCGCTTTTGCGGGATCTCGTCGAAGGCACTAGAAGGGACGCC...   \n",
              "1  61               GGCGGTCGAAGGGAGATGGGGCGTCCTCATCACCGG   \n",
              "\n",
              "                             r_primer_sequences(5-3)  \\\n",
              "0               GACTCCGGTGATGAGGACGCCCCATCTCCCTTCGAC   \n",
              "1  TCCCCGTCCGGGACCCGCGCGGTCGATCCCCGCATATAGGGAGTAG...   \n",
              "\n",
              "                f_primer_name                r_primer_name  \n",
              "0  SCO5087_repair_up_forwar_p  SCO5087_repair_up_reverse_p  \n",
              "1  SCO5087_repair_dw_forwar_p  SCO5087_repair_dw_reverse_p  "
            ]
          },
          "execution_count": 13,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "unique_df"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {},
      "outputs": [],
      "source": []
    },
    {
      "cell_type": "code",
      "execution_count": 14,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>sgRNA plasmid</th>\n",
              "      <th>plasmid_name</th>\n",
              "      <th>which workflow to proceed with</th>\n",
              "      <th>sgRNA plasmid #StuI sites</th>\n",
              "      <th>repair template plasmid #NcoI sites</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>pCas9_SCO5087(1172)</td>\n",
              "      <td>pCas9_SCO5087(1172)_w_rep</td>\n",
              "      <td>Proceed with sgRNA integration first</td>\n",
              "      <td>1</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>pCas9_SCO5087(1160)</td>\n",
              "      <td>pCas9_SCO5087(1160)_w_rep</td>\n",
              "      <td>Proceed with sgRNA integration first</td>\n",
              "      <td>1</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "         sgRNA plasmid               plasmid_name  \\\n",
              "0  pCas9_SCO5087(1172)  pCas9_SCO5087(1172)_w_rep   \n",
              "1  pCas9_SCO5087(1160)  pCas9_SCO5087(1160)_w_rep   \n",
              "\n",
              "         which workflow to proceed with  sgRNA plasmid #StuI sites  \\\n",
              "0  Proceed with sgRNA integration first                          1   \n",
              "1  Proceed with sgRNA integration first                          1   \n",
              "\n",
              "   repair template plasmid #NcoI sites  \n",
              "0                                    0  \n",
              "1                                    0  "
            ]
          },
          "execution_count": 14,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "workflow_df = determine_workflow_order_for_plasmids(sgRNA_vectors, \n",
        "                                                            assembled_contigs,\n",
        "                                                            [\"StuI\"], [ \"NcoI\"])\n",
        "workflow_df"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 15,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>plasmid_name</th>\n",
              "      <th>date</th>\n",
              "      <th>original_plasmid</th>\n",
              "      <th>integration</th>\n",
              "      <th>size</th>\n",
              "      <th>sgRNA plasmid</th>\n",
              "      <th>which workflow to proceed with</th>\n",
              "      <th>sgRNA plasmid #StuI sites</th>\n",
              "      <th>repair template plasmid #NcoI sites</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>pCas9_SCO5087(1172)_w_rep</td>\n",
              "      <td>2024-10-11</td>\n",
              "      <td>pCRISPR-Cas9</td>\n",
              "      <td>sgRNA_SCO5087(1172)</td>\n",
              "      <td>13279</td>\n",
              "      <td>pCas9_SCO5087(1172)</td>\n",
              "      <td>Proceed with sgRNA integration first</td>\n",
              "      <td>1</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>pCas9_SCO5087(1160)_w_rep</td>\n",
              "      <td>2024-10-11</td>\n",
              "      <td>pCRISPR-Cas9</td>\n",
              "      <td>sgRNA_SCO5087(1160)</td>\n",
              "      <td>13279</td>\n",
              "      <td>pCas9_SCO5087(1160)</td>\n",
              "      <td>Proceed with sgRNA integration first</td>\n",
              "      <td>1</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "                plasmid_name        date original_plasmid  \\\n",
              "0  pCas9_SCO5087(1172)_w_rep  2024-10-11     pCRISPR-Cas9   \n",
              "1  pCas9_SCO5087(1160)_w_rep  2024-10-11     pCRISPR-Cas9   \n",
              "\n",
              "           integration   size        sgRNA plasmid  \\\n",
              "0  sgRNA_SCO5087(1172)  13279  pCas9_SCO5087(1172)   \n",
              "1  sgRNA_SCO5087(1160)  13279  pCas9_SCO5087(1160)   \n",
              "\n",
              "         which workflow to proceed with  sgRNA plasmid #StuI sites  \\\n",
              "0  Proceed with sgRNA integration first                          1   \n",
              "1  Proceed with sgRNA integration first                          1   \n",
              "\n",
              "   repair template plasmid #NcoI sites  \n",
              "0                                    0  \n",
              "1                                    0  "
            ]
          },
          "execution_count": 15,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "if in_frame_deletion:\n",
        "    integration_names = filtered_df.apply(lambda row: f\"sgRNA_{row['locus_tag']}({row['sgrna_loc']})\", axis=1).tolist()\n",
        "    plasmid_metadata_df = extract_metadata_to_dataframe(assembled_contigs,\n",
        "                                                        clean_plasmid,\n",
        "                                                        integration_names)\n",
        "    workflow_df = determine_workflow_order_for_plasmids(sgRNA_vectors, \n",
        "                                                            assembled_contigs,\n",
        "                                                            [\"StuI\",], [ \"NcoI\"])\n",
        "    \n",
        "    plasmid_metadata_df = pd.merge(plasmid_metadata_df, workflow_df, on='plasmid_name', how='inner')  # or 'left'/'right' depending on what you need\n",
        "\n",
        "else: \n",
        "    integration_names = filtered_df.apply(lambda row: f\"sgRNA_{row['locus_tag']}({row['sgrna_loc']})\", axis=1).tolist()\n",
        "    plasmid_metadata_df = extract_metadata_to_dataframe(sgRNA_vectors,\n",
        "                                                        clean_plasmid,\n",
        "                                                        integration_names)\n",
        "\n",
        "plasmid_metadata_df"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 16,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>locus tag</th>\n",
              "      <th>f_primer_name</th>\n",
              "      <th>r_primer_name</th>\n",
              "      <th>f_primer_sequences(5-3)</th>\n",
              "      <th>r_primer_sequences(5-3)</th>\n",
              "      <th>f_tm</th>\n",
              "      <th>r_tm</th>\n",
              "      <th>ta</th>\n",
              "      <th>flanking_region</th>\n",
              "      <th>annealing_temperature</th>\n",
              "      <th>...</th>\n",
              "      <th>homodimer_reverse_tm</th>\n",
              "      <th>homodimer_reverse_deltaG (kcal/mol)</th>\n",
              "      <th>heterodimer_tm</th>\n",
              "      <th>heterodimer_deltaG (kcal/mol)</th>\n",
              "      <th>hairpin_forward_structure_found</th>\n",
              "      <th>hairpin_forward_tm</th>\n",
              "      <th>hairpin_forward_deltaG (kcal/mol)</th>\n",
              "      <th>hairpin_reverse_structure_found</th>\n",
              "      <th>hairpin_reverse_tm</th>\n",
              "      <th>hairpin_reverse_deltaG (kcal/mol)</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>SCO5087</td>\n",
              "      <td>SCO5087_fwd_checking_primer</td>\n",
              "      <td>SCO5087_rev_checking_primer</td>\n",
              "      <td>GACGATTCGGCCCGTG</td>\n",
              "      <td>CAGGGCGTCCAGGC</td>\n",
              "      <td>59</td>\n",
              "      <td>58</td>\n",
              "      <td>61</td>\n",
              "      <td>500</td>\n",
              "      <td>61</td>\n",
              "      <td>...</td>\n",
              "      <td>9.706246</td>\n",
              "      <td>-3.415427</td>\n",
              "      <td>1.918567</td>\n",
              "      <td>-2.396641</td>\n",
              "      <td>False</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>False</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>1 rows × 23 columns</p>\n",
              "</div>"
            ],
            "text/plain": [
              "  locus tag                f_primer_name                r_primer_name  \\\n",
              "0   SCO5087  SCO5087_fwd_checking_primer  SCO5087_rev_checking_primer   \n",
              "\n",
              "  f_primer_sequences(5-3) r_primer_sequences(5-3) f_tm r_tm  ta  \\\n",
              "0        GACGATTCGGCCCGTG          CAGGGCGTCCAGGC   59   58  61   \n",
              "\n",
              "  flanking_region  annealing_temperature  ... homodimer_reverse_tm  \\\n",
              "0             500                     61  ...             9.706246   \n",
              "\n",
              "   homodimer_reverse_deltaG (kcal/mol)  heterodimer_tm  \\\n",
              "0                            -3.415427        1.918567   \n",
              "\n",
              "   heterodimer_deltaG (kcal/mol)  hairpin_forward_structure_found  \\\n",
              "0                      -2.396641                            False   \n",
              "\n",
              "   hairpin_forward_tm  hairpin_forward_deltaG (kcal/mol)  \\\n",
              "0                 0.0                                0.0   \n",
              "\n",
              "   hairpin_reverse_structure_found  hairpin_reverse_tm  \\\n",
              "0                            False                 0.0   \n",
              "\n",
              "   hairpin_reverse_deltaG (kcal/mol)  \n",
              "0                                0.0  \n",
              "\n",
              "[1 rows x 23 columns]"
            ]
          },
          "execution_count": 16,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "# Getting checking primers\n",
        "checking_primers_df = find_best_check_primers_from_genome(genome, \n",
        "                                       genes_to_KO, \n",
        "                                       flanking_region=flanking_region,\n",
        "                                       target_tm = melting_temperature, \n",
        "                                        primer_concentration = primer_concentration, \n",
        "                                        polymerase = chosen_polymerase)\n",
        "checking_primers_df\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 17,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>template</th>\n",
              "      <th>direction</th>\n",
              "      <th>f_primer_anneal(5-3)</th>\n",
              "      <th>r_primer_anneal(5-3)</th>\n",
              "      <th>f_tm</th>\n",
              "      <th>r_tm</th>\n",
              "      <th>ta</th>\n",
              "      <th>f_primer_sequences(5-3)</th>\n",
              "      <th>r_primer_sequences(5-3)</th>\n",
              "      <th>f_primer_name</th>\n",
              "      <th>...</th>\n",
              "      <th>homodimer_reverse_tm</th>\n",
              "      <th>homodimer_reverse_deltaG (kcal/mol)</th>\n",
              "      <th>heterodimer_tm</th>\n",
              "      <th>heterodimer_deltaG (kcal/mol)</th>\n",
              "      <th>hairpin_forward_structure_found</th>\n",
              "      <th>hairpin_forward_tm</th>\n",
              "      <th>hairpin_forward_deltaG (kcal/mol)</th>\n",
              "      <th>hairpin_reverse_structure_found</th>\n",
              "      <th>hairpin_reverse_tm</th>\n",
              "      <th>hairpin_reverse_deltaG (kcal/mol)</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>SCO5087</td>\n",
              "      <td>upstream</td>\n",
              "      <td>GACGCCGGAAATCCAG</td>\n",
              "      <td>CCCATCTCCCTTCGAC</td>\n",
              "      <td>56</td>\n",
              "      <td>54</td>\n",
              "      <td>58</td>\n",
              "      <td>AAGGCCGCTTTTGCGGGATCTCGTCGAAGGCACTAGAAGGGACGCC...</td>\n",
              "      <td>GACTCCGGTGATGAGGACGCCCCATCTCCCTTCGAC</td>\n",
              "      <td>SCO5087_repair_up_forwar_p</td>\n",
              "      <td>...</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>SCO5087</td>\n",
              "      <td>downstream</td>\n",
              "      <td>GCGTCCTCATCACCGG</td>\n",
              "      <td>GAGTAGAGGCGGCCC</td>\n",
              "      <td>58</td>\n",
              "      <td>57</td>\n",
              "      <td>61</td>\n",
              "      <td>GGCGGTCGAAGGGAGATGGGGCGTCCTCATCACCGG</td>\n",
              "      <td>TCCCCGTCCGGGACCCGCGCGGTCGATCCCCGCATATAGGGAGTAG...</td>\n",
              "      <td>SCO5087_repair_dw_forwar_p</td>\n",
              "      <td>...</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>SCO5087</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>59</td>\n",
              "      <td>58</td>\n",
              "      <td>61</td>\n",
              "      <td>GACGATTCGGCCCGTG</td>\n",
              "      <td>CAGGGCGTCCAGGC</td>\n",
              "      <td>SCO5087_fwd_checking_primer</td>\n",
              "      <td>...</td>\n",
              "      <td>9.706246</td>\n",
              "      <td>-3.415427</td>\n",
              "      <td>1.918567</td>\n",
              "      <td>-2.396641</td>\n",
              "      <td>False</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>False</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>3 rows × 26 columns</p>\n",
              "</div>"
            ],
            "text/plain": [
              "  template   direction f_primer_anneal(5-3) r_primer_anneal(5-3) f_tm r_tm  \\\n",
              "0  SCO5087    upstream     GACGCCGGAAATCCAG     CCCATCTCCCTTCGAC   56   54   \n",
              "1  SCO5087  downstream     GCGTCCTCATCACCGG      GAGTAGAGGCGGCCC   58   57   \n",
              "0  SCO5087         NaN                  NaN                  NaN   59   58   \n",
              "\n",
              "   ta                            f_primer_sequences(5-3)  \\\n",
              "0  58  AAGGCCGCTTTTGCGGGATCTCGTCGAAGGCACTAGAAGGGACGCC...   \n",
              "1  61               GGCGGTCGAAGGGAGATGGGGCGTCCTCATCACCGG   \n",
              "0  61                                   GACGATTCGGCCCGTG   \n",
              "\n",
              "                             r_primer_sequences(5-3)  \\\n",
              "0               GACTCCGGTGATGAGGACGCCCCATCTCCCTTCGAC   \n",
              "1  TCCCCGTCCGGGACCCGCGCGGTCGATCCCCGCATATAGGGAGTAG...   \n",
              "0                                     CAGGGCGTCCAGGC   \n",
              "\n",
              "                 f_primer_name  ... homodimer_reverse_tm  \\\n",
              "0   SCO5087_repair_up_forwar_p  ...                  NaN   \n",
              "1   SCO5087_repair_dw_forwar_p  ...                  NaN   \n",
              "0  SCO5087_fwd_checking_primer  ...             9.706246   \n",
              "\n",
              "  homodimer_reverse_deltaG (kcal/mol)  heterodimer_tm  \\\n",
              "0                                 NaN             NaN   \n",
              "1                                 NaN             NaN   \n",
              "0                           -3.415427        1.918567   \n",
              "\n",
              "  heterodimer_deltaG (kcal/mol)  hairpin_forward_structure_found  \\\n",
              "0                           NaN                              NaN   \n",
              "1                           NaN                              NaN   \n",
              "0                     -2.396641                            False   \n",
              "\n",
              "   hairpin_forward_tm  hairpin_forward_deltaG (kcal/mol)  \\\n",
              "0                 NaN                                NaN   \n",
              "1                 NaN                                NaN   \n",
              "0                 0.0                                0.0   \n",
              "\n",
              "   hairpin_reverse_structure_found  hairpin_reverse_tm  \\\n",
              "0                              NaN                 NaN   \n",
              "1                              NaN                 NaN   \n",
              "0                            False                 0.0   \n",
              "\n",
              "   hairpin_reverse_deltaG (kcal/mol)  \n",
              "0                                NaN  \n",
              "1                                NaN  \n",
              "0                                0.0  \n",
              "\n",
              "[3 rows x 26 columns]"
            ]
          },
          "execution_count": 17,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "# Create a copy of checking_primers_df\n",
        "checking_primers_df_copy = checking_primers_df.copy()\n",
        "checking_primers_df_copy = checking_primers_df_copy.rename(columns={'locus tag': 'template'})\n",
        "\n",
        "# Concatenate unique_df and the modified checking_primers_df_copy\n",
        "pcr_table = pd.concat([unique_df, checking_primers_df_copy])\n",
        "\n",
        "# Display the concatenated table\n",
        "pcr_table"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 18,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>Name</th>\n",
              "      <th>Sequence</th>\n",
              "      <th>Concentration</th>\n",
              "      <th>Purification</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>SCO5087_fwd_checking_primer</td>\n",
              "      <td>GACGATTCGGCCCGTG</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>SCO5087_rev_checking_primer</td>\n",
              "      <td>CAGGGCGTCCAGGC</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "                          Name          Sequence Concentration Purification\n",
              "0  SCO5087_fwd_checking_primer  GACGATTCGGCCCGTG          25nm          STD\n",
              "1  SCO5087_rev_checking_primer    CAGGGCGTCCAGGC          25nm          STD"
            ]
          },
          "execution_count": 18,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "# making the primers into dseqrecords\n",
        "checking_primers_df_idt = create_idt_order_dataframe(checking_primers_df)\n",
        "checking_primers_df_idt\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 19,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>Name</th>\n",
              "      <th>Sequence</th>\n",
              "      <th>Concentration</th>\n",
              "      <th>Purification</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>SCO5087_repair_up_forwar_p</td>\n",
              "      <td>AAGGCCGCTTTTGCGGGATCTCGTCGAAGGCACTAGAAGGGACGCC...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>SCO5087_repair_dw_forwar_p</td>\n",
              "      <td>GGCGGTCGAAGGGAGATGGGGCGTCCTCATCACCGG</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>SCO5087_repair_up_reverse_p</td>\n",
              "      <td>GACTCCGGTGATGAGGACGCCCCATCTCCCTTCGAC</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>SCO5087_repair_dw_reverse_p</td>\n",
              "      <td>TCCCCGTCCGGGACCCGCGCGGTCGATCCCCGCATATAGGGAGTAG...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "                          Name  \\\n",
              "0   SCO5087_repair_up_forwar_p   \n",
              "1   SCO5087_repair_dw_forwar_p   \n",
              "2  SCO5087_repair_up_reverse_p   \n",
              "3  SCO5087_repair_dw_reverse_p   \n",
              "\n",
              "                                            Sequence Concentration  \\\n",
              "0  AAGGCCGCTTTTGCGGGATCTCGTCGAAGGCACTAGAAGGGACGCC...          25nm   \n",
              "1               GGCGGTCGAAGGGAGATGGGGCGTCCTCATCACCGG          25nm   \n",
              "2               GACTCCGGTGATGAGGACGCCCCATCTCCCTTCGAC          25nm   \n",
              "3  TCCCCGTCCGGGACCCGCGCGGTCGATCCCCGCATATAGGGAGTAG...          25nm   \n",
              "\n",
              "  Purification  \n",
              "0          STD  \n",
              "1          STD  \n",
              "2          STD  \n",
              "3          STD  "
            ]
          },
          "execution_count": 19,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "idt_df"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 20,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>Name</th>\n",
              "      <th>Sequence</th>\n",
              "      <th>Concentration</th>\n",
              "      <th>Purification</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>SCO5087_loc_1172</td>\n",
              "      <td>CGGTTGGTAGGATCGACGGCTCGATGGTCGGCCACTCGCTGTTTTA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>SCO5087_loc_1160</td>\n",
              "      <td>CGGTTGGTAGGATCGACGGCAGCTCGATCAAGTCGATGGTGTTTTA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>SCO5087_repair_up_forwar_p</td>\n",
              "      <td>AAGGCCGCTTTTGCGGGATCTCGTCGAAGGCACTAGAAGGGACGCC...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>SCO5087_repair_dw_forwar_p</td>\n",
              "      <td>GGCGGTCGAAGGGAGATGGGGCGTCCTCATCACCGG</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>SCO5087_repair_up_reverse_p</td>\n",
              "      <td>GACTCCGGTGATGAGGACGCCCCATCTCCCTTCGAC</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>SCO5087_repair_dw_reverse_p</td>\n",
              "      <td>TCCCCGTCCGGGACCCGCGCGGTCGATCCCCGCATATAGGGAGTAG...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>SCO5087_fwd_checking_primer</td>\n",
              "      <td>GACGATTCGGCCCGTG</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>SCO5087_rev_checking_primer</td>\n",
              "      <td>CAGGGCGTCCAGGC</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "                          Name  \\\n",
              "0             SCO5087_loc_1172   \n",
              "1             SCO5087_loc_1160   \n",
              "0   SCO5087_repair_up_forwar_p   \n",
              "1   SCO5087_repair_dw_forwar_p   \n",
              "2  SCO5087_repair_up_reverse_p   \n",
              "3  SCO5087_repair_dw_reverse_p   \n",
              "0  SCO5087_fwd_checking_primer   \n",
              "1  SCO5087_rev_checking_primer   \n",
              "\n",
              "                                            Sequence Concentration  \\\n",
              "0  CGGTTGGTAGGATCGACGGCTCGATGGTCGGCCACTCGCTGTTTTA...          25nm   \n",
              "1  CGGTTGGTAGGATCGACGGCAGCTCGATCAAGTCGATGGTGTTTTA...          25nm   \n",
              "0  AAGGCCGCTTTTGCGGGATCTCGTCGAAGGCACTAGAAGGGACGCC...          25nm   \n",
              "1               GGCGGTCGAAGGGAGATGGGGCGTCCTCATCACCGG          25nm   \n",
              "2               GACTCCGGTGATGAGGACGCCCCATCTCCCTTCGAC          25nm   \n",
              "3  TCCCCGTCCGGGACCCGCGCGGTCGATCCCCGCATATAGGGAGTAG...          25nm   \n",
              "0                                   GACGATTCGGCCCGTG          25nm   \n",
              "1                                     CAGGGCGTCCAGGC          25nm   \n",
              "\n",
              "  Purification  \n",
              "0          STD  \n",
              "1          STD  \n",
              "0          STD  \n",
              "1          STD  \n",
              "2          STD  \n",
              "3          STD  \n",
              "0          STD  \n",
              "1          STD  "
            ]
          },
          "execution_count": 20,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "if in_frame_deletion: \n",
        "    full_idt = pd.concat([idt_primers, idt_df, checking_primers_df_idt])\n",
        "else: \n",
        "    full_idt = pd.concat([idt_primers, checking_primers_df_idt])\n",
        "\n",
        "full_idt"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "## Folder with all the generated I/O"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 21,
      "metadata": {},
      "outputs": [],
      "source": [
        "input_files = [\n",
        "    {\"name\": \"input_genome.gb\", \"content\": genome},\n",
        "    {\"name\": \"input_plasmid.gb\", \"content\": clean_plasmid}\n",
        "]\n",
        "\n",
        "if in_frame_deletion: \n",
        "    output_files = [\n",
        "        {\"name\": \"cBEST_w_sgRNAs.gb\", \"content\": assembled_contigs}, # LIST OF Dseqrecords\n",
        "        {\"name\": \"primer_df.csv\", \"content\": primer_df},\n",
        "        {\"name\": \"full_idt.csv\", \"content\": full_idt},\n",
        "        {\"name\": \"sgrna_df.csv\", \"content\": sgrna_df},\n",
        "        {\"name\": \"filtered_df.csv\", \"content\": filtered_df},\n",
        "        {\"name\": \"plasmid_metadata_df.csv\", \"content\": plasmid_metadata_df},\n",
        "\n",
        "    ]\n",
        "else: \n",
        "    output_files = [\n",
        "        {\"name\": \"cBEST_w_sgRNAs.gb\", \"content\": sgRNA_vectors}, # LIST OF Dseqrecords\n",
        "        {\"name\": \"full_idt.csv\", \"content\": full_idt},\n",
        "        {\"name\": \"sgrna_df.csv\", \"content\": sgrna_df},\n",
        "        {\"name\": \"filtered_df.csv\", \"content\": filtered_df},\n",
        "        {\"name\": \"plasmid_metadata_df.csv\", \"content\": plasmid_metadata_df},\n",
        "        {\"name\": \"workflow_order_df.csv\", \"content\": workflow_df},\n",
        "\n",
        "        workflow_df\n",
        "\n",
        "    ]\n",
        "\n",
        "input_values = {\n",
        "    \"genes_to_knockout\": genes_to_KO,\n",
        "\n",
        "\n",
        "    \"filtering_metrics\": {\n",
        "        \"gc_upper\": gc_upper,\n",
        "        \"gc_lower\": gc_lower,\n",
        "        \"off_target_seed\": off_target_seed,\n",
        "        \"off_target_upper\": off_target_upper,\n",
        "        \"cas_type\": cas_type,\n",
        "        \"number_of_sgRNAs_per_group\": number_of_sgRNAs_per_group,\n",
        "\n",
        "    },  \n",
        "    \"polymerase_settings\": {\n",
        "        \"chosen_polymerase\": chosen_polymerase,\n",
        "        \"melting_temperature\": melting_temperature,\n",
        "        \"primer_concentration\": primer_concentration,\n",
        "        \"primer_number_increment\": primer_number_increment,\n",
        "        \"flanking_region\": flanking_region\n",
        "    },\n",
        "    \"overlapping_sequences\": {\n",
        "        \"up_homology\": str(up_homology),\n",
        "        \"dw_homology\": str(dw_homology)\n",
        "    }\n",
        "}\n",
        "\n",
        "\n",
        "# Paths to Markdown files\n",
        "markdown_file_paths = [\n",
        "    \"../../protocols/conjugation_protcol.md\",\n",
        "    \"../../protocols/single_target_crispr_plasmid_protcol.md\"\n",
        "\n",
        "]\n",
        "\n",
        "\n",
        "timestamp = datetime.utcnow().isoformat()\n",
        "\n",
        "# Create project directory structure\n",
        "project_directory = ProjectDirectory(\n",
        "    project_name=f\"CRISPR_cas9_plasmid_workflow_{timestamp}\",\n",
        "    input_files=input_files,\n",
        "    output_files=output_files,\n",
        "    input_values=input_values,\n",
        "    markdown_file_paths=markdown_file_paths\n",
        ")\n",
        "\n",
        "# DO You want to save the folder? \n",
        "save_zip_folder = False \n",
        "\n",
        "if save_zip_folder: \n",
        "    # Generate the project directory structure and get the zip content\n",
        "    zip_content = project_directory.create_directory_structure(create_directories=False)\n",
        "\n",
        "    # Save the zip file to disk (optional)\n",
        "    with open(\"project_structure.zip\", \"wb\") as f:\n",
        "        f.write(zip_content)"
      ]
    }
  ],
  "metadata": {
    "kernelspec": {
      "display_name": "constrain",
      "language": "python",
      "name": "python3"
    },
    "language_info": {
      "codemirror_mode": {
        "name": "ipython",
        "version": 3
      },
      "file_extension": ".py",
      "mimetype": "text/x-python",
      "name": "python",
      "nbconvert_exporter": "python",
      "pygments_lexer": "ipython3",
      "version": "3.11.9"
    }
  },
  "nbformat": 4,
  "nbformat_minor": 2
}
