{
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "# W6-CRISPR-Cas3-in-frame-deletion"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "In Workflow 6, CRISPR-Cas3 is paired with a PCR and Gibson protocol to achieve in-frame deletions22(Figure 7A). Just like in workflow 5, this workflow can be used to knock out specific genes in streptomycetes but with the crucial difference of using Cas3 which has been shown to have higher efficiencies for targeted gene deletions in for example BGCs and random-sized deletion in streptomyces22. Due to homologies, and high GC content, this protocol relies on two steps of Gibson cloning to incorporate the sgRNA and the upstream and downstream repair regions. To get started, users can download the pCRISPR-Cas3 file and the genome file (S. coelicolor A3). StreptoAIM will then generate the necessary primers, plasmid files, and checking primers to verify full gene deletions (S5). \n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 1,
      "metadata": {},
      "outputs": [],
      "source": [
        "import sys\n",
        "import os\n",
        "from Bio.Restriction import StuI\n",
        "import pandas as pd\n",
        "from datetime import datetime\n",
        "\n",
        "# Ensure the src directory is in the Python path\n",
        "project_root = os.path.abspath(os.path.join(os.getcwd(), '../../'))\n",
        "if project_root not in sys.path:\n",
        "    sys.path.insert(0, project_root)\n",
        "\n",
        "\n",
        "from pydna.dseqrecord import Dseqrecord\n",
        "\n",
        "from streptocad.sequence_loading.sequence_loading import (\n",
        "    load_and_process_plasmid,\n",
        "    load_and_process_genome_sequences\n",
        ")\n",
        "\n",
        "from streptocad.utils import polymerase_dict, create_primer_df_from_dict,ProjectDirectory,extract_metadata_to_dataframe\n",
        "from streptocad.primers.primer_generation import create_idt_order_dataframe, make_primer_records, primers_to_IDT, checking_primers\n",
        "from streptocad.crispr.guideRNAcas3_9_12 import extract_sgRNAs, SgRNAargs\n",
        "from streptocad.cloning.cas3_plasmid_cloning import generate_cas3_primers, cas3_plasmid_pcrs, assemble_cas3_plasmids\n",
        "from streptocad.cloning.gibson_cloning import (\n",
        "    find_up_dw_repair_templates,\n",
        "    assemble_multiple_plasmids_with_repair_templates_for_deletion,\n",
        "    update_primer_names\n",
        ")"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "## INPUT"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 2,
      "metadata": {},
      "outputs": [],
      "source": [
        "# Inputs\n",
        "# 1 Add genome of choice (genbank, fasta)\n",
        "path_to_genome = '../../data/genomes/Streptomyces_coelicolor_A3_chromosome.gb'\n",
        "genome = load_and_process_genome_sequences(path_to_genome)[0]\n",
        "\n",
        "# 2 Add plasmid \n",
        "path_to_plasmid = '../../data/plasmids/pCRISPR_Cas3.gbk'\n",
        "clean_plasmid = load_and_process_plasmid(path_to_plasmid)\n",
        "\n",
        "# 3 Choose genes to knock out (list)\n",
        "genes_to_KO = ['SCO5087', 'SCO5089', 'SCO5090']\n",
        "\n",
        "#### Advanced settings ####\n",
        "# 4 Filtering metrics for sgRNAs\n",
        "gc_upper = 0.8\n",
        "gc_lower = 0.2\n",
        "off_target_seed = 8\n",
        "off_target_upper = 10\n",
        "cas_type='cas3'\n",
        "number_of_sgRNAs_per_group = 2\n",
        "\n",
        "# 5 Choose polymerase and target melting temperature\n",
        "chosen_polymerase = polymerase_dict['Phusion High-Fidelity DNA Polymerase (GC Buffer)']\n",
        "melting_temperature = 65\n",
        "primer_concentration = 0.4 \n",
        "primer_number_increment = 23\n",
        "flanking_region = 500\n",
        "\n",
        "# 6 Choose overlapping sequences for our plasmid we can use the following\n",
        "#As per the article **\"CRISPR–Cas9, CRISPRi and CRISPR-BEST-mediated genetic manipulation in streptomycetes\"** we need the following oligoes: \n",
        "#CGGTTGGTAGGATCGACGGC **-N20-** GTTTTAGAGCTAGAAATAGC\n",
        "up_homology = Dseqrecord('CGGTTGGTAGGATCGACGGC')\n",
        "dw_homology = Dseqrecord('GTTTTAGAGCTAGAAATAGC')"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "# Computation"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 3,
      "metadata": {},
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "sgRNA generated were too smallSCO5090. To incorporate this extent borders. Skipping to next locus tag.\n"
          ]
        },
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>strain_name</th>\n",
              "      <th>locus_tag</th>\n",
              "      <th>gene_loc</th>\n",
              "      <th>gene_strand</th>\n",
              "      <th>sgrna_strand</th>\n",
              "      <th>sgrna_loc</th>\n",
              "      <th>gc</th>\n",
              "      <th>pam</th>\n",
              "      <th>sgrna</th>\n",
              "      <th>sgrna_seed_sequence</th>\n",
              "      <th>off_target_count</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>28</th>\n",
              "      <td>Streptomyces_coelicolor_A3_chromosome</td>\n",
              "      <td>SCO5087</td>\n",
              "      <td>5529801</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>131</td>\n",
              "      <td>0.558824</td>\n",
              "      <td>TTC</td>\n",
              "      <td>AATCCGAACAGCTCCTTCGATGCGTTCGTCCGGT</td>\n",
              "      <td>AATCCGAA</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>Streptomyces_coelicolor_A3_chromosome</td>\n",
              "      <td>SCO5087</td>\n",
              "      <td>5529801</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>163</td>\n",
              "      <td>0.617647</td>\n",
              "      <td>TTC</td>\n",
              "      <td>GGATTGAAGCGCAGAGTCGTCATCACGGGCGTCG</td>\n",
              "      <td>GGATTGAA</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>6</th>\n",
              "      <td>Streptomyces_coelicolor_A3_chromosome</td>\n",
              "      <td>SCO5087</td>\n",
              "      <td>5529801</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>306</td>\n",
              "      <td>0.735294</td>\n",
              "      <td>TTC</td>\n",
              "      <td>ACAGGTCGCGGCGGAGGCCGACTTCGACCCGGTC</td>\n",
              "      <td>ACAGGTCG</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>Streptomyces_coelicolor_A3_chromosome</td>\n",
              "      <td>SCO5087</td>\n",
              "      <td>5529801</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>94</td>\n",
              "      <td>0.705882</td>\n",
              "      <td>TTC</td>\n",
              "      <td>GAGCCTCCTTCGAGCCACGGGGCCGACGATGACG</td>\n",
              "      <td>GAGCCTCC</td>\n",
              "      <td>1</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>24</th>\n",
              "      <td>Streptomyces_coelicolor_A3_chromosome</td>\n",
              "      <td>SCO5087</td>\n",
              "      <td>5529801</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>962</td>\n",
              "      <td>0.676471</td>\n",
              "      <td>TTC</td>\n",
              "      <td>AGGCCCGTCATGTGGTACGCGTTGCAGCGCGTCG</td>\n",
              "      <td>AGGCCCGT</td>\n",
              "      <td>1</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>14</th>\n",
              "      <td>Streptomyces_coelicolor_A3_chromosome</td>\n",
              "      <td>SCO5087</td>\n",
              "      <td>5529801</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>850</td>\n",
              "      <td>0.705882</td>\n",
              "      <td>TTC</td>\n",
              "      <td>GACGGTACGCGGGACGGCTTCGTGCTGGCCGAAG</td>\n",
              "      <td>GACGGTAC</td>\n",
              "      <td>1</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>26</th>\n",
              "      <td>Streptomyces_coelicolor_A3_chromosome</td>\n",
              "      <td>SCO5087</td>\n",
              "      <td>5529801</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>201</td>\n",
              "      <td>0.794118</td>\n",
              "      <td>TTC</td>\n",
              "      <td>CCAGAACTGCCGGGTGCCGTTCCCGCCGGGGGCG</td>\n",
              "      <td>CCAGAACT</td>\n",
              "      <td>1</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>Streptomyces_coelicolor_A3_chromosome</td>\n",
              "      <td>SCO5087</td>\n",
              "      <td>5529801</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>105</td>\n",
              "      <td>0.735294</td>\n",
              "      <td>TTC</td>\n",
              "      <td>GAGCCACGGGGCCGACGATGACGACGACCACCGG</td>\n",
              "      <td>GAGCCACG</td>\n",
              "      <td>2</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>29</th>\n",
              "      <td>Streptomyces_coelicolor_A3_chromosome</td>\n",
              "      <td>SCO5087</td>\n",
              "      <td>5529801</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>113</td>\n",
              "      <td>0.647059</td>\n",
              "      <td>TTC</td>\n",
              "      <td>GATGCGTTCGTCCGGTGGTCGTCGTCATCGTCGG</td>\n",
              "      <td>GATGCGTT</td>\n",
              "      <td>2</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>19</th>\n",
              "      <td>Streptomyces_coelicolor_A3_chromosome</td>\n",
              "      <td>SCO5087</td>\n",
              "      <td>5529801</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>1390</td>\n",
              "      <td>0.735294</td>\n",
              "      <td>TTC</td>\n",
              "      <td>CAGAGCGCGATGGTGCTGCGCGACGCCGAGACCG</td>\n",
              "      <td>CAGAGCGC</td>\n",
              "      <td>3</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>34</th>\n",
              "      <td>Streptomyces_coelicolor_A3_chromosome</td>\n",
              "      <td>SCO5090</td>\n",
              "      <td>5532706</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>199</td>\n",
              "      <td>0.764706</td>\n",
              "      <td>TTC</td>\n",
              "      <td>CGGATCTGGGCGCGCGTGGGCGGCCGGGTGAAGA</td>\n",
              "      <td>CGGATCTG</td>\n",
              "      <td>3</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>38</th>\n",
              "      <td>Streptomyces_coelicolor_A3_chromosome</td>\n",
              "      <td>SCO5090</td>\n",
              "      <td>5532706</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>505</td>\n",
              "      <td>0.735294</td>\n",
              "      <td>TTC</td>\n",
              "      <td>ACGTTCGAGGACACGCTCCGGGTGCCGTCCGGGG</td>\n",
              "      <td>ACGTTCGA</td>\n",
              "      <td>4</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>45</th>\n",
              "      <td>Streptomyces_coelicolor_A3_chromosome</td>\n",
              "      <td>SCO5090</td>\n",
              "      <td>5532706</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>543</td>\n",
              "      <td>0.647059</td>\n",
              "      <td>TTC</td>\n",
              "      <td>CTGCCACAGGTCGCTGCGCTCGATGAAGGCGTAG</td>\n",
              "      <td>CTGCCACA</td>\n",
              "      <td>4</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>44</th>\n",
              "      <td>Streptomyces_coelicolor_A3_chromosome</td>\n",
              "      <td>SCO5090</td>\n",
              "      <td>5532706</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>563</td>\n",
              "      <td>0.735294</td>\n",
              "      <td>TTC</td>\n",
              "      <td>CGCACGTGCGGCAGCCGTTCCTGCCACAGGTCGC</td>\n",
              "      <td>CGCACGTG</td>\n",
              "      <td>4</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>43</th>\n",
              "      <td>Streptomyces_coelicolor_A3_chromosome</td>\n",
              "      <td>SCO5090</td>\n",
              "      <td>5532706</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>828</td>\n",
              "      <td>0.794118</td>\n",
              "      <td>TTC</td>\n",
              "      <td>CACGCGGGCCGGGTCGATCGCCACGGTGTGGCGG</td>\n",
              "      <td>CACGCGGG</td>\n",
              "      <td>4</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>Streptomyces_coelicolor_A3_chromosome</td>\n",
              "      <td>SCO5087</td>\n",
              "      <td>5529801</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>235</td>\n",
              "      <td>0.735294</td>\n",
              "      <td>TTC</td>\n",
              "      <td>TGGGAACTGCTCACCTCGGGGCGGACGGCGACGC</td>\n",
              "      <td>TGGGAACT</td>\n",
              "      <td>4</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>5</th>\n",
              "      <td>Streptomyces_coelicolor_A3_chromosome</td>\n",
              "      <td>SCO5087</td>\n",
              "      <td>5529801</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>286</td>\n",
              "      <td>0.735294</td>\n",
              "      <td>TTC</td>\n",
              "      <td>GACCCCTCGCCCTACCGTTCACAGGTCGCGGCGG</td>\n",
              "      <td>GACCCCTC</td>\n",
              "      <td>4</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>Streptomyces_coelicolor_A3_chromosome</td>\n",
              "      <td>SCO5087</td>\n",
              "      <td>5529801</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>283</td>\n",
              "      <td>0.676471</td>\n",
              "      <td>TTC</td>\n",
              "      <td>TTCGACCCCTCGCCCTACCGTTCACAGGTCGCGG</td>\n",
              "      <td>TTCGACCC</td>\n",
              "      <td>5</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>33</th>\n",
              "      <td>Streptomyces_coelicolor_A3_chromosome</td>\n",
              "      <td>SCO5090</td>\n",
              "      <td>5532706</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>145</td>\n",
              "      <td>0.676471</td>\n",
              "      <td>TTC</td>\n",
              "      <td>GAACCCACCGTGCACGTACAGGTCCTGGAGCGCG</td>\n",
              "      <td>GAACCCAC</td>\n",
              "      <td>5</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>53</th>\n",
              "      <td>Streptomyces_coelicolor_A3_chromosome</td>\n",
              "      <td>SCO5090</td>\n",
              "      <td>5532706</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>87</td>\n",
              "      <td>0.764706</td>\n",
              "      <td>TTC</td>\n",
              "      <td>GACGCGTGCGACCAGCTCGTAGAGCCGCCGCGCC</td>\n",
              "      <td>GACGCGTG</td>\n",
              "      <td>6</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>23</th>\n",
              "      <td>Streptomyces_coelicolor_A3_chromosome</td>\n",
              "      <td>SCO5087</td>\n",
              "      <td>5529801</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>1073</td>\n",
              "      <td>0.676471</td>\n",
              "      <td>TTC</td>\n",
              "      <td>TGCCGGGTGCCGGAGCCGTGCGCGTTGATGTAGT</td>\n",
              "      <td>TGCCGGGT</td>\n",
              "      <td>7</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>46</th>\n",
              "      <td>Streptomyces_coelicolor_A3_chromosome</td>\n",
              "      <td>SCO5090</td>\n",
              "      <td>5532706</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>404</td>\n",
              "      <td>0.764706</td>\n",
              "      <td>TTC</td>\n",
              "      <td>CCGCTGTTCGCGTCCAGCGCCTCGCGCAGTCCCG</td>\n",
              "      <td>CCGCTGTT</td>\n",
              "      <td>8</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>41</th>\n",
              "      <td>Streptomyces_coelicolor_A3_chromosome</td>\n",
              "      <td>SCO5090</td>\n",
              "      <td>5532706</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>720</td>\n",
              "      <td>0.705882</td>\n",
              "      <td>TTC</td>\n",
              "      <td>GATCCGGCTCTGCGTCCCCGCGAGGAGCATCGTG</td>\n",
              "      <td>GATCCGGC</td>\n",
              "      <td>8</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>30</th>\n",
              "      <td>Streptomyces_coelicolor_A3_chromosome</td>\n",
              "      <td>SCO5087</td>\n",
              "      <td>5529801</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>104</td>\n",
              "      <td>0.735294</td>\n",
              "      <td>TTC</td>\n",
              "      <td>GTCCGGTGGTCGTCGTCATCGTCGGCCCCGTGGC</td>\n",
              "      <td>GTCCGGTG</td>\n",
              "      <td>8</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>52</th>\n",
              "      <td>Streptomyces_coelicolor_A3_chromosome</td>\n",
              "      <td>SCO5090</td>\n",
              "      <td>5532706</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>108</td>\n",
              "      <td>0.705882</td>\n",
              "      <td>TTC</td>\n",
              "      <td>GAAGACGGCCGGCCAGTCTTCGACGCGTGCGACC</td>\n",
              "      <td>GAAGACGG</td>\n",
              "      <td>9</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>17</th>\n",
              "      <td>Streptomyces_coelicolor_A3_chromosome</td>\n",
              "      <td>SCO5087</td>\n",
              "      <td>5529801</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>1325</td>\n",
              "      <td>0.735294</td>\n",
              "      <td>TTC</td>\n",
              "      <td>CGCTGGAGGCCAGGGAGCGGAAGCTGCGGTCCGT</td>\n",
              "      <td>CGCTGGAG</td>\n",
              "      <td>9</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>20</th>\n",
              "      <td>Streptomyces_coelicolor_A3_chromosome</td>\n",
              "      <td>SCO5087</td>\n",
              "      <td>5529801</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>1307</td>\n",
              "      <td>0.647059</td>\n",
              "      <td>TTC</td>\n",
              "      <td>CGCTCCCTGGCCTCCAGCGGAACGTAGTCGAGAT</td>\n",
              "      <td>CGCTCCCT</td>\n",
              "      <td>10</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>11</th>\n",
              "      <td>Streptomyces_coelicolor_A3_chromosome</td>\n",
              "      <td>SCO5087</td>\n",
              "      <td>5529801</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>586</td>\n",
              "      <td>0.705882</td>\n",
              "      <td>TTC</td>\n",
              "      <td>GACTACCTGGTGCCCAGCGTCATGCCGGCCGAGG</td>\n",
              "      <td>GACTACCT</td>\n",
              "      <td>10</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>12</th>\n",
              "      <td>Streptomyces_coelicolor_A3_chromosome</td>\n",
              "      <td>SCO5087</td>\n",
              "      <td>5529801</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>742</td>\n",
              "      <td>0.764706</td>\n",
              "      <td>TTC</td>\n",
              "      <td>GCGGGAGCCGCCGACACCCCGATCACCCCGATCG</td>\n",
              "      <td>GCGGGAGC</td>\n",
              "      <td>10</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "                              strain_name locus_tag  gene_loc  gene_strand  \\\n",
              "28  Streptomyces_coelicolor_A3_chromosome   SCO5087   5529801            1   \n",
              "2   Streptomyces_coelicolor_A3_chromosome   SCO5087   5529801            1   \n",
              "6   Streptomyces_coelicolor_A3_chromosome   SCO5087   5529801            1   \n",
              "0   Streptomyces_coelicolor_A3_chromosome   SCO5087   5529801            1   \n",
              "24  Streptomyces_coelicolor_A3_chromosome   SCO5087   5529801            1   \n",
              "14  Streptomyces_coelicolor_A3_chromosome   SCO5087   5529801            1   \n",
              "26  Streptomyces_coelicolor_A3_chromosome   SCO5087   5529801            1   \n",
              "1   Streptomyces_coelicolor_A3_chromosome   SCO5087   5529801            1   \n",
              "29  Streptomyces_coelicolor_A3_chromosome   SCO5087   5529801            1   \n",
              "19  Streptomyces_coelicolor_A3_chromosome   SCO5087   5529801            1   \n",
              "34  Streptomyces_coelicolor_A3_chromosome   SCO5090   5532706            1   \n",
              "38  Streptomyces_coelicolor_A3_chromosome   SCO5090   5532706            1   \n",
              "45  Streptomyces_coelicolor_A3_chromosome   SCO5090   5532706            1   \n",
              "44  Streptomyces_coelicolor_A3_chromosome   SCO5090   5532706            1   \n",
              "43  Streptomyces_coelicolor_A3_chromosome   SCO5090   5532706            1   \n",
              "3   Streptomyces_coelicolor_A3_chromosome   SCO5087   5529801            1   \n",
              "5   Streptomyces_coelicolor_A3_chromosome   SCO5087   5529801            1   \n",
              "4   Streptomyces_coelicolor_A3_chromosome   SCO5087   5529801            1   \n",
              "33  Streptomyces_coelicolor_A3_chromosome   SCO5090   5532706            1   \n",
              "53  Streptomyces_coelicolor_A3_chromosome   SCO5090   5532706            1   \n",
              "23  Streptomyces_coelicolor_A3_chromosome   SCO5087   5529801            1   \n",
              "46  Streptomyces_coelicolor_A3_chromosome   SCO5090   5532706            1   \n",
              "41  Streptomyces_coelicolor_A3_chromosome   SCO5090   5532706            1   \n",
              "30  Streptomyces_coelicolor_A3_chromosome   SCO5087   5529801            1   \n",
              "52  Streptomyces_coelicolor_A3_chromosome   SCO5090   5532706            1   \n",
              "17  Streptomyces_coelicolor_A3_chromosome   SCO5087   5529801            1   \n",
              "20  Streptomyces_coelicolor_A3_chromosome   SCO5087   5529801            1   \n",
              "11  Streptomyces_coelicolor_A3_chromosome   SCO5087   5529801            1   \n",
              "12  Streptomyces_coelicolor_A3_chromosome   SCO5087   5529801            1   \n",
              "\n",
              "    sgrna_strand  sgrna_loc        gc  pam  \\\n",
              "28             1        131  0.558824  TTC   \n",
              "2             -1        163  0.617647  TTC   \n",
              "6             -1        306  0.735294  TTC   \n",
              "0             -1         94  0.705882  TTC   \n",
              "24             1        962  0.676471  TTC   \n",
              "14            -1        850  0.705882  TTC   \n",
              "26             1        201  0.794118  TTC   \n",
              "1             -1        105  0.735294  TTC   \n",
              "29             1        113  0.647059  TTC   \n",
              "19            -1       1390  0.735294  TTC   \n",
              "34            -1        199  0.764706  TTC   \n",
              "38            -1        505  0.735294  TTC   \n",
              "45             1        543  0.647059  TTC   \n",
              "44             1        563  0.735294  TTC   \n",
              "43             1        828  0.794118  TTC   \n",
              "3             -1        235  0.735294  TTC   \n",
              "5             -1        286  0.735294  TTC   \n",
              "4             -1        283  0.676471  TTC   \n",
              "33            -1        145  0.676471  TTC   \n",
              "53             1         87  0.764706  TTC   \n",
              "23             1       1073  0.676471  TTC   \n",
              "46             1        404  0.764706  TTC   \n",
              "41            -1        720  0.705882  TTC   \n",
              "30             1        104  0.735294  TTC   \n",
              "52             1        108  0.705882  TTC   \n",
              "17            -1       1325  0.735294  TTC   \n",
              "20             1       1307  0.647059  TTC   \n",
              "11            -1        586  0.705882  TTC   \n",
              "12            -1        742  0.764706  TTC   \n",
              "\n",
              "                                 sgrna sgrna_seed_sequence  off_target_count  \n",
              "28  AATCCGAACAGCTCCTTCGATGCGTTCGTCCGGT            AATCCGAA                 0  \n",
              "2   GGATTGAAGCGCAGAGTCGTCATCACGGGCGTCG            GGATTGAA                 0  \n",
              "6   ACAGGTCGCGGCGGAGGCCGACTTCGACCCGGTC            ACAGGTCG                 0  \n",
              "0   GAGCCTCCTTCGAGCCACGGGGCCGACGATGACG            GAGCCTCC                 1  \n",
              "24  AGGCCCGTCATGTGGTACGCGTTGCAGCGCGTCG            AGGCCCGT                 1  \n",
              "14  GACGGTACGCGGGACGGCTTCGTGCTGGCCGAAG            GACGGTAC                 1  \n",
              "26  CCAGAACTGCCGGGTGCCGTTCCCGCCGGGGGCG            CCAGAACT                 1  \n",
              "1   GAGCCACGGGGCCGACGATGACGACGACCACCGG            GAGCCACG                 2  \n",
              "29  GATGCGTTCGTCCGGTGGTCGTCGTCATCGTCGG            GATGCGTT                 2  \n",
              "19  CAGAGCGCGATGGTGCTGCGCGACGCCGAGACCG            CAGAGCGC                 3  \n",
              "34  CGGATCTGGGCGCGCGTGGGCGGCCGGGTGAAGA            CGGATCTG                 3  \n",
              "38  ACGTTCGAGGACACGCTCCGGGTGCCGTCCGGGG            ACGTTCGA                 4  \n",
              "45  CTGCCACAGGTCGCTGCGCTCGATGAAGGCGTAG            CTGCCACA                 4  \n",
              "44  CGCACGTGCGGCAGCCGTTCCTGCCACAGGTCGC            CGCACGTG                 4  \n",
              "43  CACGCGGGCCGGGTCGATCGCCACGGTGTGGCGG            CACGCGGG                 4  \n",
              "3   TGGGAACTGCTCACCTCGGGGCGGACGGCGACGC            TGGGAACT                 4  \n",
              "5   GACCCCTCGCCCTACCGTTCACAGGTCGCGGCGG            GACCCCTC                 4  \n",
              "4   TTCGACCCCTCGCCCTACCGTTCACAGGTCGCGG            TTCGACCC                 5  \n",
              "33  GAACCCACCGTGCACGTACAGGTCCTGGAGCGCG            GAACCCAC                 5  \n",
              "53  GACGCGTGCGACCAGCTCGTAGAGCCGCCGCGCC            GACGCGTG                 6  \n",
              "23  TGCCGGGTGCCGGAGCCGTGCGCGTTGATGTAGT            TGCCGGGT                 7  \n",
              "46  CCGCTGTTCGCGTCCAGCGCCTCGCGCAGTCCCG            CCGCTGTT                 8  \n",
              "41  GATCCGGCTCTGCGTCCCCGCGAGGAGCATCGTG            GATCCGGC                 8  \n",
              "30  GTCCGGTGGTCGTCGTCATCGTCGGCCCCGTGGC            GTCCGGTG                 8  \n",
              "52  GAAGACGGCCGGCCAGTCTTCGACGCGTGCGACC            GAAGACGG                 9  \n",
              "17  CGCTGGAGGCCAGGGAGCGGAAGCTGCGGTCCGT            CGCTGGAG                 9  \n",
              "20  CGCTCCCTGGCCTCCAGCGGAACGTAGTCGAGAT            CGCTCCCT                10  \n",
              "11  GACTACCTGGTGCCCAGCGTCATGCCGGCCGAGG            GACTACCT                10  \n",
              "12  GCGGGAGCCGCCGACACCCCGATCACCCCGATCG            GCGGGAGC                10  "
            ]
          },
          "execution_count": 3,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "\n",
        "# Initialize SgRNAargs with desired parameters\n",
        "args = SgRNAargs(path_to_genome, \n",
        "                genes_to_KO,\n",
        "                step=['find', 'filter'],\n",
        "                gc_upper = gc_upper,\n",
        "                gc_lower = gc_lower,\n",
        "                off_target_seed = off_target_seed,\n",
        "                off_target_upper = off_target_upper,\n",
        "                cas_type=cas_type\n",
        "                )\n",
        "\n",
        "sgrna_df = extract_sgRNAs(args)\n",
        "sgrna_df"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 4,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>strain_name</th>\n",
              "      <th>locus_tag</th>\n",
              "      <th>gene_loc</th>\n",
              "      <th>gene_strand</th>\n",
              "      <th>sgrna_strand</th>\n",
              "      <th>sgrna_loc</th>\n",
              "      <th>gc</th>\n",
              "      <th>pam</th>\n",
              "      <th>sgrna</th>\n",
              "      <th>sgrna_seed_sequence</th>\n",
              "      <th>off_target_count</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>28</th>\n",
              "      <td>Streptomyces_coelicolor_A3_chromosome</td>\n",
              "      <td>SCO5087</td>\n",
              "      <td>5529801</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>131</td>\n",
              "      <td>0.558824</td>\n",
              "      <td>TTC</td>\n",
              "      <td>AATCCGAACAGCTCCTTCGATGCGTTCGTCCGGT</td>\n",
              "      <td>AATCCGAA</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>Streptomyces_coelicolor_A3_chromosome</td>\n",
              "      <td>SCO5087</td>\n",
              "      <td>5529801</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>163</td>\n",
              "      <td>0.617647</td>\n",
              "      <td>TTC</td>\n",
              "      <td>GGATTGAAGCGCAGAGTCGTCATCACGGGCGTCG</td>\n",
              "      <td>GGATTGAA</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>34</th>\n",
              "      <td>Streptomyces_coelicolor_A3_chromosome</td>\n",
              "      <td>SCO5090</td>\n",
              "      <td>5532706</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>199</td>\n",
              "      <td>0.764706</td>\n",
              "      <td>TTC</td>\n",
              "      <td>CGGATCTGGGCGCGCGTGGGCGGCCGGGTGAAGA</td>\n",
              "      <td>CGGATCTG</td>\n",
              "      <td>3</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>38</th>\n",
              "      <td>Streptomyces_coelicolor_A3_chromosome</td>\n",
              "      <td>SCO5090</td>\n",
              "      <td>5532706</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>505</td>\n",
              "      <td>0.735294</td>\n",
              "      <td>TTC</td>\n",
              "      <td>ACGTTCGAGGACACGCTCCGGGTGCCGTCCGGGG</td>\n",
              "      <td>ACGTTCGA</td>\n",
              "      <td>4</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "                              strain_name locus_tag  gene_loc  gene_strand  \\\n",
              "28  Streptomyces_coelicolor_A3_chromosome   SCO5087   5529801            1   \n",
              "2   Streptomyces_coelicolor_A3_chromosome   SCO5087   5529801            1   \n",
              "34  Streptomyces_coelicolor_A3_chromosome   SCO5090   5532706            1   \n",
              "38  Streptomyces_coelicolor_A3_chromosome   SCO5090   5532706            1   \n",
              "\n",
              "    sgrna_strand  sgrna_loc        gc  pam  \\\n",
              "28             1        131  0.558824  TTC   \n",
              "2             -1        163  0.617647  TTC   \n",
              "34            -1        199  0.764706  TTC   \n",
              "38            -1        505  0.735294  TTC   \n",
              "\n",
              "                                 sgrna sgrna_seed_sequence  off_target_count  \n",
              "28  AATCCGAACAGCTCCTTCGATGCGTTCGTCCGGT            AATCCGAA                 0  \n",
              "2   GGATTGAAGCGCAGAGTCGTCATCACGGGCGTCG            GGATTGAA                 0  \n",
              "34  CGGATCTGGGCGCGCGTGGGCGGCCGGGTGAAGA            CGGATCTG                 3  \n",
              "38  ACGTTCGAGGACACGCTCCGGGTGCCGTCCGGGG            ACGTTCGA                 4  "
            ]
          },
          "execution_count": 4,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "# Filter the DataFrame to retain only up to 5 sgRNA sequences per locus_tag\n",
        "filtered_df = sgrna_df.groupby('locus_tag').head(number_of_sgRNAs_per_group)\n",
        "filtered_df"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "## Output"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 5,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>strain_name</th>\n",
              "      <th>locus_tag</th>\n",
              "      <th>gene_loc</th>\n",
              "      <th>gene_strand</th>\n",
              "      <th>sgrna_strand</th>\n",
              "      <th>sgrna_loc</th>\n",
              "      <th>gc</th>\n",
              "      <th>pam</th>\n",
              "      <th>sgrna</th>\n",
              "      <th>sgrna_seed_sequence</th>\n",
              "      <th>off_target_count</th>\n",
              "      <th>Fwd Primer</th>\n",
              "      <th>Rev Primer</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>28</th>\n",
              "      <td>Streptomyces_coelicolor_A3_chromosome</td>\n",
              "      <td>SCO5087</td>\n",
              "      <td>5529801</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>131</td>\n",
              "      <td>0.558824</td>\n",
              "      <td>TTC</td>\n",
              "      <td>AATCCGAACAGCTCCTTCGATGCGTTCGTCCGGT</td>\n",
              "      <td>AATCCGAA</td>\n",
              "      <td>0</td>\n",
              "      <td>AATCCGAACAGCTCCTTCGATGCGTTCGTCCGGTGTCGCcCggCaa...</td>\n",
              "      <td>ACCGGACGAACGCATCGAAGGAGCTGTTCGGATTGTTTCAATCCAC...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>Streptomyces_coelicolor_A3_chromosome</td>\n",
              "      <td>SCO5087</td>\n",
              "      <td>5529801</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>163</td>\n",
              "      <td>0.617647</td>\n",
              "      <td>TTC</td>\n",
              "      <td>GGATTGAAGCGCAGAGTCGTCATCACGGGCGTCG</td>\n",
              "      <td>GGATTGAA</td>\n",
              "      <td>0</td>\n",
              "      <td>GGATTGAAGCGCAGAGTCGTCATCACGGGCGTCGGTCGCcCggCaa...</td>\n",
              "      <td>CGACGCCCGTGATGACGACTCTGCGCTTCAATCCGTTTCAATCCAC...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>34</th>\n",
              "      <td>Streptomyces_coelicolor_A3_chromosome</td>\n",
              "      <td>SCO5090</td>\n",
              "      <td>5532706</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>199</td>\n",
              "      <td>0.764706</td>\n",
              "      <td>TTC</td>\n",
              "      <td>CGGATCTGGGCGCGCGTGGGCGGCCGGGTGAAGA</td>\n",
              "      <td>CGGATCTG</td>\n",
              "      <td>3</td>\n",
              "      <td>CGGATCTGGGCGCGCGTGGGCGGCCGGGTGAAGAGTCGCcCggCaa...</td>\n",
              "      <td>TCTTCACCCGGCCGCCCACGCGCGCCCAGATCCGGTTTCAATCCAC...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>38</th>\n",
              "      <td>Streptomyces_coelicolor_A3_chromosome</td>\n",
              "      <td>SCO5090</td>\n",
              "      <td>5532706</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>505</td>\n",
              "      <td>0.735294</td>\n",
              "      <td>TTC</td>\n",
              "      <td>ACGTTCGAGGACACGCTCCGGGTGCCGTCCGGGG</td>\n",
              "      <td>ACGTTCGA</td>\n",
              "      <td>4</td>\n",
              "      <td>ACGTTCGAGGACACGCTCCGGGTGCCGTCCGGGGGTCGCcCggCaa...</td>\n",
              "      <td>CCCCGGACGGCACCCGGAGCGTGTCCTCGAACGTGTTTCAATCCAC...</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "                              strain_name locus_tag  gene_loc  gene_strand  \\\n",
              "28  Streptomyces_coelicolor_A3_chromosome   SCO5087   5529801            1   \n",
              "2   Streptomyces_coelicolor_A3_chromosome   SCO5087   5529801            1   \n",
              "34  Streptomyces_coelicolor_A3_chromosome   SCO5090   5532706            1   \n",
              "38  Streptomyces_coelicolor_A3_chromosome   SCO5090   5532706            1   \n",
              "\n",
              "    sgrna_strand  sgrna_loc        gc  pam  \\\n",
              "28             1        131  0.558824  TTC   \n",
              "2             -1        163  0.617647  TTC   \n",
              "34            -1        199  0.764706  TTC   \n",
              "38            -1        505  0.735294  TTC   \n",
              "\n",
              "                                 sgrna sgrna_seed_sequence  off_target_count  \\\n",
              "28  AATCCGAACAGCTCCTTCGATGCGTTCGTCCGGT            AATCCGAA                 0   \n",
              "2   GGATTGAAGCGCAGAGTCGTCATCACGGGCGTCG            GGATTGAA                 0   \n",
              "34  CGGATCTGGGCGCGCGTGGGCGGCCGGGTGAAGA            CGGATCTG                 3   \n",
              "38  ACGTTCGAGGACACGCTCCGGGTGCCGTCCGGGG            ACGTTCGA                 4   \n",
              "\n",
              "                                           Fwd Primer  \\\n",
              "28  AATCCGAACAGCTCCTTCGATGCGTTCGTCCGGTGTCGCcCggCaa...   \n",
              "2   GGATTGAAGCGCAGAGTCGTCATCACGGGCGTCGGTCGCcCggCaa...   \n",
              "34  CGGATCTGGGCGCGCGTGGGCGGCCGGGTGAAGAGTCGCcCggCaa...   \n",
              "38  ACGTTCGAGGACACGCTCCGGGTGCCGTCCGGGGGTCGCcCggCaa...   \n",
              "\n",
              "                                           Rev Primer  \n",
              "28  ACCGGACGAACGCATCGAAGGAGCTGTTCGGATTGTTTCAATCCAC...  \n",
              "2   CGACGCCCGTGATGACGACTCTGCGCTTCAATCCGTTTCAATCCAC...  \n",
              "34  TCTTCACCCGGCCGCCCACGCGCGCCCAGATCCGGTTTCAATCCAC...  \n",
              "38  CCCCGGACGGCACCCGGAGCGTGTCCTCGAACGTGTTTCAATCCAC...  "
            ]
          },
          "execution_count": 5,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "filtered_df_w_primers = generate_cas3_primers(filtered_df)\n",
        "filtered_df_w_primers"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 6,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/plain": [
              "[[Amplicon(1489), Amplicon(460)],\n",
              " [Amplicon(1489), Amplicon(460)],\n",
              " [Amplicon(1489), Amplicon(460)],\n",
              " [Amplicon(1489), Amplicon(460)]]"
            ]
          },
          "execution_count": 6,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "amplicons = cas3_plasmid_pcrs(clean_plasmid, filtered_df)\n",
        "\n",
        "amplicons"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 7,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/plain": [
              "[Contig(o12766), Contig(o12766), Contig(o12766), Contig(o12766)]"
            ]
          },
          "execution_count": 7,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "# cut plasmid\n",
        "assembled_cas3_plasmids = assemble_cas3_plasmids(clean_plasmid, amplicons)\n",
        "assembled_cas3_plasmids"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 8,
      "metadata": {},
      "outputs": [],
      "source": [
        "\n",
        "# Constructing a meaningful name, ID, and description for the assembled plasmid using user input\n",
        "targeting_info = []\n",
        "for index, row in filtered_df.iterrows():\n",
        "    formatted_str = f\"pCas3_{row['locus_tag']}({row['sgrna_loc']})\"\n",
        "    targeting_info.append(formatted_str)\n",
        "\n",
        "for i in range(len(assembled_cas3_plasmids)):\n",
        "    assembled_cas3_plasmids[i].name = f'{targeting_info[i]}'\n",
        "    assembled_cas3_plasmids[i].id = assembled_cas3_plasmids[i].name  # Using the same value for ID as for name for simplicity\n",
        "    assembled_cas3_plasmids[i].description = f'Assembled plasmid targeting {\", \".join(genes_to_KO)} for single gene knockout, assembled using StreptoCAD.'\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 9,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/plain": [
              "'pCas3_SCO5087(131)'"
            ]
          },
          "execution_count": 9,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "assembled_cas3_plasmids[0].name"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 10,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/plain": [
              "'Assembled plasmid targeting SCO5087, SCO5089, SCO5090 for single gene knockout, assembled using StreptoCAD.'"
            ]
          },
          "execution_count": 10,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "assembled_cas3_plasmids[0].description"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 11,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/plain": [
              "'pCas3_SCO5087(131)'"
            ]
          },
          "execution_count": 11,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "assembled_cas3_plasmids[0].id"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 12,
      "metadata": {},
      "outputs": [],
      "source": [
        "print_plasmids = False\n",
        "\n",
        "if print_plasmids: \n",
        "    for vector in assembled_cas3_plasmids: \n",
        "        vector.write(f\"../../data/plasmids/sgRNA_plasmids_pCRISPR–Cas3/{vector.id}.gb\")"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "### IDT primers"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 13,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/plain": [
              "[Dseqrecord(-52),\n",
              " Dseqrecord(-54),\n",
              " Dseqrecord(-52),\n",
              " Dseqrecord(-54),\n",
              " Dseqrecord(-52),\n",
              " Dseqrecord(-54),\n",
              " Dseqrecord(-52),\n",
              " Dseqrecord(-54)]"
            ]
          },
          "execution_count": 13,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "primer_records = make_primer_records(filtered_df_w_primers)\n",
        "primer_records"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 14,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>Name</th>\n",
              "      <th>Sequence</th>\n",
              "      <th>Concentration</th>\n",
              "      <th>Purification</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>SCO5087_loc_131_fwd</td>\n",
              "      <td>AATCCGAACAGCTCCTTCGATGCGTTCGTCCGGTGTCGCcCggCaa...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>SCO5087_loc_131_rev</td>\n",
              "      <td>ACCGGACGAACGCATCGAAGGAGCTGTTCGGATTGTTTCAATCCAC...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>SCO5087_loc_163_fwd</td>\n",
              "      <td>GGATTGAAGCGCAGAGTCGTCATCACGGGCGTCGGTCGCcCggCaa...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>SCO5087_loc_163_rev</td>\n",
              "      <td>CGACGCCCGTGATGACGACTCTGCGCTTCAATCCGTTTCAATCCAC...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>SCO5090_loc_199_fwd</td>\n",
              "      <td>CGGATCTGGGCGCGCGTGGGCGGCCGGGTGAAGAGTCGCcCggCaa...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>5</th>\n",
              "      <td>SCO5090_loc_199_rev</td>\n",
              "      <td>TCTTCACCCGGCCGCCCACGCGCGCCCAGATCCGGTTTCAATCCAC...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>6</th>\n",
              "      <td>SCO5090_loc_505_fwd</td>\n",
              "      <td>ACGTTCGAGGACACGCTCCGGGTGCCGTCCGGGGGTCGCcCggCaa...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>7</th>\n",
              "      <td>SCO5090_loc_505_rev</td>\n",
              "      <td>CCCCGGACGGCACCCGGAGCGTGTCCTCGAACGTGTTTCAATCCAC...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "                  Name                                           Sequence  \\\n",
              "0  SCO5087_loc_131_fwd  AATCCGAACAGCTCCTTCGATGCGTTCGTCCGGTGTCGCcCggCaa...   \n",
              "1  SCO5087_loc_131_rev  ACCGGACGAACGCATCGAAGGAGCTGTTCGGATTGTTTCAATCCAC...   \n",
              "2  SCO5087_loc_163_fwd  GGATTGAAGCGCAGAGTCGTCATCACGGGCGTCGGTCGCcCggCaa...   \n",
              "3  SCO5087_loc_163_rev  CGACGCCCGTGATGACGACTCTGCGCTTCAATCCGTTTCAATCCAC...   \n",
              "4  SCO5090_loc_199_fwd  CGGATCTGGGCGCGCGTGGGCGGCCGGGTGAAGAGTCGCcCggCaa...   \n",
              "5  SCO5090_loc_199_rev  TCTTCACCCGGCCGCCCACGCGCGCCCAGATCCGGTTTCAATCCAC...   \n",
              "6  SCO5090_loc_505_fwd  ACGTTCGAGGACACGCTCCGGGTGCCGTCCGGGGGTCGCcCggCaa...   \n",
              "7  SCO5090_loc_505_rev  CCCCGGACGGCACCCGGAGCGTGTCCTCGAACGTGTTTCAATCCAC...   \n",
              "\n",
              "  Concentration Purification  \n",
              "0          25nm          STD  \n",
              "1          25nm          STD  \n",
              "2          25nm          STD  \n",
              "3          25nm          STD  \n",
              "4          25nm          STD  \n",
              "5          25nm          STD  \n",
              "6          25nm          STD  \n",
              "7          25nm          STD  "
            ]
          },
          "execution_count": 14,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "idt_primers_cas3=primers_to_IDT(primer_records)\n",
        "idt_primers_cas3"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "# In frame deletion"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 15,
      "metadata": {},
      "outputs": [],
      "source": [
        "in_frame_deletion = True "
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 16,
      "metadata": {},
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "repair_templates_data\n",
            "Digest the plasmids\n",
            "Rename them appropriately\n",
            "assembly_data\n"
          ]
        }
      ],
      "source": [
        "if in_frame_deletion:\n",
        "    # Make repair templates\n",
        "    print('repair_templates_data')\n",
        "    repair_templates_data = find_up_dw_repair_templates(genome, \n",
        "                                                        genes_to_KO, \n",
        "                                                        target_tm=melting_temperature, \n",
        "                                                        primer_tm_kwargs={'conc':primer_concentration, 'prodcode':chosen_polymerase} )\n",
        "    # Digest the plasmids\n",
        "    print('Digest the plasmids')\n",
        "    processed_records = [Dseqrecord(record, circular=True).cut(StuI)[0] for record in assembled_cas3_plasmids]\n",
        "\n",
        "    # Rename them appropriately\n",
        "    print('Rename them appropriately')\n",
        "\n",
        "    for i in range(len(processed_records)):\n",
        "        processed_records[i].name = assembled_cas3_plasmids[i].name\n",
        "\n",
        "    # Assembly \n",
        "    print('assembly_data')\n",
        "\n",
        "    assembly_data = assemble_multiple_plasmids_with_repair_templates_for_deletion(genes_to_KO, processed_records, \n",
        "                                                                                repair_templates_data, \n",
        "                                                                                overlap=40)\n",
        "    # updating the primer names to something systematic.\n",
        "    update_primer_names(assembly_data)\n",
        "\n",
        "    # Parse through the primer df\n",
        "    primer_df = create_primer_df_from_dict(assembly_data)\n",
        "\n",
        "    #Find_unique primers\n",
        "    unique_df = primer_df.drop_duplicates(keep='first')\n",
        "    idt_df = create_idt_order_dataframe(unique_df, concentration=\"25nm\", purification=\"STD\")\n",
        "\n",
        "    # Contigs\n",
        "    assembled_contigs = []\n",
        "    for data in assembly_data: \n",
        "        contig_record = data['contig']\n",
        "        contig_record.id = f\"{data['name']}_w_rep\"\n",
        "        contig_record.name = f\"{data['name']}_w_rep\"\n",
        "\n",
        "        assembled_contigs.append(contig_record)\n",
        "\n",
        "    print_plasmids = False\n",
        "\n",
        "    if print_plasmids:\n",
        "        # Write to genbank files\n",
        "        output_directory = \"../../data/plasmids/sgRNA_plasmids_pCRISPR–Cas3_with_repair_templates/\"\n",
        "        for contig in assembled_contigs: \n",
        "            contig.write(f\"{output_directory}/{contig.id}.gb\")"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 17,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>plasmid_name</th>\n",
              "      <th>date</th>\n",
              "      <th>original_plasmid</th>\n",
              "      <th>integration</th>\n",
              "      <th>size</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>pCas3_SCO5087(131)_w_rep</td>\n",
              "      <td>2024-07-18</td>\n",
              "      <td>CRISPR-Cas9</td>\n",
              "      <td>sgRNA_SCO5087(131)</td>\n",
              "      <td>6425</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>pCas3_SCO5087(163)_w_rep</td>\n",
              "      <td>2024-07-18</td>\n",
              "      <td>CRISPR-Cas9</td>\n",
              "      <td>sgRNA_SCO5087(163)</td>\n",
              "      <td>6425</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>pCas3_SCO5090(199)_w_rep</td>\n",
              "      <td>2024-07-18</td>\n",
              "      <td>CRISPR-Cas9</td>\n",
              "      <td>sgRNA_SCO5090(199)</td>\n",
              "      <td>6425</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>pCas3_SCO5090(505)_w_rep</td>\n",
              "      <td>2024-07-18</td>\n",
              "      <td>CRISPR-Cas9</td>\n",
              "      <td>sgRNA_SCO5090(505)</td>\n",
              "      <td>6425</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "               plasmid_name        date original_plasmid         integration  \\\n",
              "0  pCas3_SCO5087(131)_w_rep  2024-07-18      CRISPR-Cas9  sgRNA_SCO5087(131)   \n",
              "1  pCas3_SCO5087(163)_w_rep  2024-07-18      CRISPR-Cas9  sgRNA_SCO5087(163)   \n",
              "2  pCas3_SCO5090(199)_w_rep  2024-07-18      CRISPR-Cas9  sgRNA_SCO5090(199)   \n",
              "3  pCas3_SCO5090(505)_w_rep  2024-07-18      CRISPR-Cas9  sgRNA_SCO5090(505)   \n",
              "\n",
              "   size  \n",
              "0  6425  \n",
              "1  6425  \n",
              "2  6425  \n",
              "3  6425  "
            ]
          },
          "execution_count": 17,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "if in_frame_deletion:\n",
        "    integration_names = filtered_df.apply(lambda row: f\"sgRNA_{row['locus_tag']}({row['sgrna_loc']})\", axis=1).tolist()\n",
        "    plasmid_metadata_df = extract_metadata_to_dataframe(assembled_contigs,\n",
        "                                                        'CRISPR-Cas9',\n",
        "                                                        integration_names)\n",
        "else: \n",
        "    integration_names = filtered_df.apply(lambda row: f\"sgRNA_{row['locus_tag']}({row['sgrna_loc']})\", axis=1).tolist()\n",
        "    plasmid_metadata_df = extract_metadata_to_dataframe(assembled_cas3_plasmids,\n",
        "                                                        'CRISPR-Cas9',\n",
        "                                                        integration_names)\n",
        "\n",
        "plasmid_metadata_df"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 18,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>Locus Tag</th>\n",
              "      <th>f_primer_name</th>\n",
              "      <th>r_primer_name</th>\n",
              "      <th>f_primer_sequences(5-3)</th>\n",
              "      <th>r_primer_sequences(5-3)</th>\n",
              "      <th>f_tm</th>\n",
              "      <th>r_tm</th>\n",
              "      <th>ta</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>SCO5087</td>\n",
              "      <td>SCO5087_fwd_checking_primer</td>\n",
              "      <td>SCO5087_rev_checking_primer</td>\n",
              "      <td>GACGATTCGGCCCGTG</td>\n",
              "      <td>CAGGGCGTCCAGGC</td>\n",
              "      <td>59</td>\n",
              "      <td>58</td>\n",
              "      <td>61</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>SCO5089</td>\n",
              "      <td>SCO5089_fwd_checking_primer</td>\n",
              "      <td>SCO5089_rev_checking_primer</td>\n",
              "      <td>CGTACGGCGAACTCGC</td>\n",
              "      <td>GTCCCCGGACGGC</td>\n",
              "      <td>59</td>\n",
              "      <td>57</td>\n",
              "      <td>60</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>SCO5090</td>\n",
              "      <td>SCO5090_fwd_checking_primer</td>\n",
              "      <td>SCO5090_rev_checking_primer</td>\n",
              "      <td>GGCGGCGGCCC</td>\n",
              "      <td>GACGTCCCCCGCGAAC</td>\n",
              "      <td>57</td>\n",
              "      <td>62</td>\n",
              "      <td>61</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "  Locus Tag                f_primer_name                r_primer_name  \\\n",
              "0   SCO5087  SCO5087_fwd_checking_primer  SCO5087_rev_checking_primer   \n",
              "1   SCO5089  SCO5089_fwd_checking_primer  SCO5089_rev_checking_primer   \n",
              "2   SCO5090  SCO5090_fwd_checking_primer  SCO5090_rev_checking_primer   \n",
              "\n",
              "  f_primer_sequences(5-3) r_primer_sequences(5-3)  f_tm  r_tm  ta  \n",
              "0        GACGATTCGGCCCGTG          CAGGGCGTCCAGGC    59    58  61  \n",
              "1        CGTACGGCGAACTCGC           GTCCCCGGACGGC    59    57  60  \n",
              "2             GGCGGCGGCCC        GACGTCCCCCGCGAAC    57    62  61  "
            ]
          },
          "execution_count": 18,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "# Getting checking primers\n",
        "checking_primers_df = checking_primers(path_to_genome, genes_to_KO, \n",
        "                                       flanking_region=flanking_region,\n",
        "                                       target_tm = melting_temperature, \n",
        "                                       primer_concentration = primer_concentration, \n",
        "                                       polymerase = chosen_polymerase)\n",
        "checking_primers_df"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 19,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>Name</th>\n",
              "      <th>Sequence</th>\n",
              "      <th>Concentration</th>\n",
              "      <th>Purification</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>SCO5087_fwd_checking_primer</td>\n",
              "      <td>GACGATTCGGCCCGTG</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>SCO5089_fwd_checking_primer</td>\n",
              "      <td>CGTACGGCGAACTCGC</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>SCO5090_fwd_checking_primer</td>\n",
              "      <td>GGCGGCGGCCC</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>SCO5087_rev_checking_primer</td>\n",
              "      <td>CAGGGCGTCCAGGC</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>SCO5089_rev_checking_primer</td>\n",
              "      <td>GTCCCCGGACGGC</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>5</th>\n",
              "      <td>SCO5090_rev_checking_primer</td>\n",
              "      <td>GACGTCCCCCGCGAAC</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "                          Name          Sequence Concentration Purification\n",
              "0  SCO5087_fwd_checking_primer  GACGATTCGGCCCGTG          25nm          STD\n",
              "1  SCO5089_fwd_checking_primer  CGTACGGCGAACTCGC          25nm          STD\n",
              "2  SCO5090_fwd_checking_primer       GGCGGCGGCCC          25nm          STD\n",
              "3  SCO5087_rev_checking_primer    CAGGGCGTCCAGGC          25nm          STD\n",
              "4  SCO5089_rev_checking_primer     GTCCCCGGACGGC          25nm          STD\n",
              "5  SCO5090_rev_checking_primer  GACGTCCCCCGCGAAC          25nm          STD"
            ]
          },
          "execution_count": 19,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "# making the primers into dseqrecords\n",
        "checking_primers_df_idt = create_idt_order_dataframe(checking_primers_df)\n",
        "checking_primers_df_idt\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 22,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>Name</th>\n",
              "      <th>Sequence</th>\n",
              "      <th>Concentration</th>\n",
              "      <th>Purification</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>SCO5087_loc_131_fwd</td>\n",
              "      <td>AATCCGAACAGCTCCTTCGATGCGTTCGTCCGGTGTCGCcCggCaa...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>SCO5087_loc_131_rev</td>\n",
              "      <td>ACCGGACGAACGCATCGAAGGAGCTGTTCGGATTGTTTCAATCCAC...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>SCO5087_loc_163_fwd</td>\n",
              "      <td>GGATTGAAGCGCAGAGTCGTCATCACGGGCGTCGGTCGCcCggCaa...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>SCO5087_loc_163_rev</td>\n",
              "      <td>CGACGCCCGTGATGACGACTCTGCGCTTCAATCCGTTTCAATCCAC...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>SCO5090_loc_199_fwd</td>\n",
              "      <td>CGGATCTGGGCGCGCGTGGGCGGCCGGGTGAAGAGTCGCcCggCaa...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>5</th>\n",
              "      <td>SCO5090_loc_199_rev</td>\n",
              "      <td>TCTTCACCCGGCCGCCCACGCGCGCCCAGATCCGGTTTCAATCCAC...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>6</th>\n",
              "      <td>SCO5090_loc_505_fwd</td>\n",
              "      <td>ACGTTCGAGGACACGCTCCGGGTGCCGTCCGGGGGTCGCcCggCaa...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>7</th>\n",
              "      <td>SCO5090_loc_505_rev</td>\n",
              "      <td>CCCCGGACGGCACCCGGAGCGTGTCCTCGAACGTGTTTCAATCCAC...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>8</th>\n",
              "      <td>primer_1</td>\n",
              "      <td>ATGCAGGAGAAGTCGGTGCTGAACAACCAGCACAAGCAGGGACGCC...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>9</th>\n",
              "      <td>primer_3</td>\n",
              "      <td>GGCGGTCGAAGGGAGATGGGGCGTCCTCATCACCGG</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>10</th>\n",
              "      <td>primer_5</td>\n",
              "      <td>ATGCAGGAGAAGTCGGTGCTGAACAACCAGCACAAGCAGGGCCACG...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>11</th>\n",
              "      <td>primer_7</td>\n",
              "      <td>GGCGCACTGGCCGAGGCGGCCGGTCGAGGTCCGT</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>12</th>\n",
              "      <td>primer_2</td>\n",
              "      <td>GACTCCGGTGATGAGGACGCCCCATCTCCCTTCGAC</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>13</th>\n",
              "      <td>primer_4</td>\n",
              "      <td>TCCCGCAGGCTGCGACGGCCCTCCACCCGGTGGTAGAAGGGAGTAG...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>14</th>\n",
              "      <td>primer_6</td>\n",
              "      <td>GACCTCACGGACCTCGACCGGCCGCCTCGGCCAGT</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>15</th>\n",
              "      <td>primer_8</td>\n",
              "      <td>TCCCGCAGGCTGCGACGGCCCTCCACCCGGTGGTAGAAGGACCGTG...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>16</th>\n",
              "      <td>SCO5087_fwd_checking_primer</td>\n",
              "      <td>GACGATTCGGCCCGTG</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>17</th>\n",
              "      <td>SCO5089_fwd_checking_primer</td>\n",
              "      <td>CGTACGGCGAACTCGC</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>18</th>\n",
              "      <td>SCO5090_fwd_checking_primer</td>\n",
              "      <td>GGCGGCGGCCC</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>19</th>\n",
              "      <td>SCO5087_rev_checking_primer</td>\n",
              "      <td>CAGGGCGTCCAGGC</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>20</th>\n",
              "      <td>SCO5089_rev_checking_primer</td>\n",
              "      <td>GTCCCCGGACGGC</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>21</th>\n",
              "      <td>SCO5090_rev_checking_primer</td>\n",
              "      <td>GACGTCCCCCGCGAAC</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "                           Name  \\\n",
              "0           SCO5087_loc_131_fwd   \n",
              "1           SCO5087_loc_131_rev   \n",
              "2           SCO5087_loc_163_fwd   \n",
              "3           SCO5087_loc_163_rev   \n",
              "4           SCO5090_loc_199_fwd   \n",
              "5           SCO5090_loc_199_rev   \n",
              "6           SCO5090_loc_505_fwd   \n",
              "7           SCO5090_loc_505_rev   \n",
              "8                      primer_1   \n",
              "9                      primer_3   \n",
              "10                     primer_5   \n",
              "11                     primer_7   \n",
              "12                     primer_2   \n",
              "13                     primer_4   \n",
              "14                     primer_6   \n",
              "15                     primer_8   \n",
              "16  SCO5087_fwd_checking_primer   \n",
              "17  SCO5089_fwd_checking_primer   \n",
              "18  SCO5090_fwd_checking_primer   \n",
              "19  SCO5087_rev_checking_primer   \n",
              "20  SCO5089_rev_checking_primer   \n",
              "21  SCO5090_rev_checking_primer   \n",
              "\n",
              "                                             Sequence Concentration  \\\n",
              "0   AATCCGAACAGCTCCTTCGATGCGTTCGTCCGGTGTCGCcCggCaa...          25nm   \n",
              "1   ACCGGACGAACGCATCGAAGGAGCTGTTCGGATTGTTTCAATCCAC...          25nm   \n",
              "2   GGATTGAAGCGCAGAGTCGTCATCACGGGCGTCGGTCGCcCggCaa...          25nm   \n",
              "3   CGACGCCCGTGATGACGACTCTGCGCTTCAATCCGTTTCAATCCAC...          25nm   \n",
              "4   CGGATCTGGGCGCGCGTGGGCGGCCGGGTGAAGAGTCGCcCggCaa...          25nm   \n",
              "5   TCTTCACCCGGCCGCCCACGCGCGCCCAGATCCGGTTTCAATCCAC...          25nm   \n",
              "6   ACGTTCGAGGACACGCTCCGGGTGCCGTCCGGGGGTCGCcCggCaa...          25nm   \n",
              "7   CCCCGGACGGCACCCGGAGCGTGTCCTCGAACGTGTTTCAATCCAC...          25nm   \n",
              "8   ATGCAGGAGAAGTCGGTGCTGAACAACCAGCACAAGCAGGGACGCC...          25nm   \n",
              "9                GGCGGTCGAAGGGAGATGGGGCGTCCTCATCACCGG          25nm   \n",
              "10  ATGCAGGAGAAGTCGGTGCTGAACAACCAGCACAAGCAGGGCCACG...          25nm   \n",
              "11                 GGCGCACTGGCCGAGGCGGCCGGTCGAGGTCCGT          25nm   \n",
              "12               GACTCCGGTGATGAGGACGCCCCATCTCCCTTCGAC          25nm   \n",
              "13  TCCCGCAGGCTGCGACGGCCCTCCACCCGGTGGTAGAAGGGAGTAG...          25nm   \n",
              "14                GACCTCACGGACCTCGACCGGCCGCCTCGGCCAGT          25nm   \n",
              "15  TCCCGCAGGCTGCGACGGCCCTCCACCCGGTGGTAGAAGGACCGTG...          25nm   \n",
              "16                                   GACGATTCGGCCCGTG          25nm   \n",
              "17                                   CGTACGGCGAACTCGC          25nm   \n",
              "18                                        GGCGGCGGCCC          25nm   \n",
              "19                                     CAGGGCGTCCAGGC          25nm   \n",
              "20                                      GTCCCCGGACGGC          25nm   \n",
              "21                                   GACGTCCCCCGCGAAC          25nm   \n",
              "\n",
              "   Purification  \n",
              "0           STD  \n",
              "1           STD  \n",
              "2           STD  \n",
              "3           STD  \n",
              "4           STD  \n",
              "5           STD  \n",
              "6           STD  \n",
              "7           STD  \n",
              "8           STD  \n",
              "9           STD  \n",
              "10          STD  \n",
              "11          STD  \n",
              "12          STD  \n",
              "13          STD  \n",
              "14          STD  \n",
              "15          STD  \n",
              "16          STD  \n",
              "17          STD  \n",
              "18          STD  \n",
              "19          STD  \n",
              "20          STD  \n",
              "21          STD  "
            ]
          },
          "execution_count": 22,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "if in_frame_deletion: \n",
        "    full_idt = pd.concat([idt_primers_cas3,idt_df, checking_primers_df_idt],ignore_index=True)\n",
        "else: \n",
        "    full_idt = pd.concat([idt_primers_cas3,checking_primers_df_idt],ignore_index=True)\n",
        "\n",
        "full_idt"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "## Folder with all the generated I/O"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 21,
      "metadata": {},
      "outputs": [],
      "source": [
        "\n",
        "input_files = [\n",
        "    {\"name\": \"input_genome.gb\", \"content\": genome},\n",
        "    {\"name\": \"input_plasmid.gb\", \"content\": clean_plasmid}\n",
        "]\n",
        "\n",
        "\n",
        "if in_frame_deletion: \n",
        "    output_files = [\n",
        "        {\"name\": \"BEST_w_sgRNAs.gb\", \"content\": assembled_contigs}, # LIST OF Dseqrecords\n",
        "        {\"name\": \"primer_df.csv\", \"content\": primer_df},\n",
        "        {\"name\": \"full_idt.csv\", \"content\": full_idt},\n",
        "        {\"name\": \"sgrna_df.csv\", \"content\": sgrna_df},\n",
        "        {\"name\": \"filtered_sgrna_df.csv\", \"content\": filtered_df},\n",
        "        {\"name\": \"plasmid_metadata_df.csv\", \"content\": plasmid_metadata_df},\n",
        "\n",
        "    ]\n",
        "else: \n",
        "    output_files = [\n",
        "        {\"name\": \"cBEST_w_sgRNAs.gb\", \"content\": assembled_cas3_plasmids}, # LIST OF Dseqrecords\n",
        "        {\"name\": \"full_idt.csv\", \"content\": full_idt},\n",
        "        {\"name\": \"sgrna_df.csv\", \"content\": sgrna_df},\n",
        "        {\"name\": \"filtered_df.csv\", \"content\": filtered_df},\n",
        "        {\"name\": \"plasmid_metadata_df.csv\", \"content\": plasmid_metadata_df},\n",
        "    ]\n",
        "         \n",
        "input_values = {\n",
        "    \"genes_to_knockout\": genes_to_KO,\n",
        "\n",
        "    \"filtering_metrics\": {\n",
        "        \"gc_upper\": gc_upper,\n",
        "        \"gc_lower\": gc_lower,\n",
        "        \"off_target_seed\": off_target_seed,\n",
        "        \"off_target_upper\": off_target_upper,\n",
        "        \"cas_type\": cas_type,\n",
        "        \"number_of_sgRNAs_per_group\": number_of_sgRNAs_per_group,\n",
        "\n",
        "    },  \n",
        "    \"polymerase_settings\": {\n",
        "        \"chosen_polymerase\": chosen_polymerase,\n",
        "        \"melting_temperature\": melting_temperature,\n",
        "        \"primer_concentration\": primer_concentration,\n",
        "        \"primer_number_increment\": primer_number_increment,\n",
        "        \"flanking_region\": flanking_region\n",
        "    },\n",
        "    \"overlapping_sequences\": {\n",
        "        \"up_homology\": str(up_homology),\n",
        "        \"dw_homology\": str(dw_homology)\n",
        "    }\n",
        "}\n",
        "\n",
        "# Paths to Markdown files\n",
        "markdown_file_paths = [\n",
        "    \"../../protocols/conjugation_protcol.md\",\n",
        "    \"../../protocols/single_target_crispr_plasmid_protcol.md\"\n",
        "\n",
        "]\n",
        "\n",
        "timestamp = datetime.utcnow().isoformat()\n",
        "\n",
        "# Create project directory structure\n",
        "project_directory = ProjectDirectory(\n",
        "    project_name=f\"CRISPR_cas9_inframe_deletion_workflow_{timestamp}\",\n",
        "    input_files=input_files,\n",
        "    output_files=output_files,\n",
        "    input_values=input_values,\n",
        "    markdown_file_paths=markdown_file_paths\n",
        ")\n",
        "\n",
        "# DO You want to save the folder? \n",
        "save_zip_folder = False \n",
        "\n",
        "if save_zip_folder: \n",
        "    # Generate the project directory structure and get the zip content\n",
        "    zip_content = project_directory.create_directory_structure(create_directories=False)\n",
        "\n",
        "    # Save the zip file to disk (optional)\n",
        "    with open(\"project_structure.zip\", \"wb\") as f:\n",
        "        f.write(zip_content)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {},
      "outputs": [],
      "source": []
    }
  ],
  "metadata": {
    "kernelspec": {
      "display_name": "constrain",
      "language": "python",
      "name": "python3"
    },
    "language_info": {
      "codemirror_mode": {
        "name": "ipython",
        "version": 3
      },
      "file_extension": ".py",
      "mimetype": "text/x-python",
      "name": "python",
      "nbconvert_exporter": "python",
      "pygments_lexer": "ipython3",
      "version": "3.8.13"
    }
  },
  "nbformat": 4,
  "nbformat_minor": 2
}
