{
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "# W6-CRISPR-Cas3-in-frame-deletion"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "In Workflow 6, CRISPR-Cas3 is paired with a PCR and Gibson protocol to achieve in-frame deletions22(Figure 7A). Just like in workflow 5, this workflow can be used to knock out specific genes in streptomycetes but with the crucial difference of using Cas3 which has been shown to have higher efficiencies for targeted gene deletions in for example BGCs and random-sized deletion in streptomyces22. Due to homologies, and high GC content, this protocol relies on two steps of Gibson cloning to incorporate the sgRNA and the upstream and downstream repair regions. To get started, users can download the pCRISPR-Cas3 file and the genome file (S. coelicolor A3). StreptoAIM will then generate the necessary primers, plasmid files, and checking primers to verify full gene deletions (S5). \n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 1,
      "metadata": {},
      "outputs": [],
      "source": [
        "import sys\n",
        "import os\n",
        "from Bio.Restriction import * \n",
        "import pandas as pd\n",
        "from datetime import datetime\n",
        "\n",
        "# Ensure the src directory is in the Python path\n",
        "project_root = os.path.abspath(os.path.join(os.getcwd(), '../../'))\n",
        "if project_root not in sys.path:\n",
        "    sys.path.insert(0, project_root)\n",
        "\n",
        "\n",
        "from pydna.dseqrecord import Dseqrecord\n",
        "\n",
        "from streptocad.sequence_loading.sequence_loading import (\n",
        "    load_and_process_plasmid,\n",
        "    load_and_process_genome_sequences, \n",
        "    check_and_convert_input, \n",
        "    annotate_dseqrecord)\n",
        "\n",
        "from streptocad.utils import polymerase_dict, create_primer_df_from_dict,ProjectDirectory,extract_metadata_to_dataframe\n",
        "from streptocad.primers.primer_generation import create_idt_order_dataframe, make_primer_records, primers_to_IDT, find_best_check_primers_from_genome\n",
        "from streptocad.crispr.guideRNAcas3_9_12 import extract_sgRNAs, SgRNAargs\n",
        "from streptocad.cloning.cas3_plasmid_cloning import generate_cas3_protospacer_primers, cas3_plasmid_pcrs, assemble_cas3_plasmids\n",
        "from streptocad.cloning.gibson_cloning import (\n",
        "    find_up_dw_repair_templates,\n",
        "    assemble_multiple_plasmids_with_repair_templates_for_deletion,\n",
        "    update_primer_names\n",
        ")\n",
        "from streptocad.cloning.plasmid_processing import determine_workflow_order_for_plasmids"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "## INPUT"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 2,
      "metadata": {},
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "pCRISPR-Cas3\n"
          ]
        }
      ],
      "source": [
        "# Inputs\n",
        "# 1 Add genome of choice (genbank, fasta)\n",
        "path_to_genome = '../../data/genomes/Streptomyces_coelicolor_A3_chromosome.gb'\n",
        "genome = load_and_process_genome_sequences(path_to_genome)[0]\n",
        "\n",
        "# 2 Add plasmid \n",
        "path_to_plasmid = '../../data/plasmids/pCRISPR_Cas3.gbk'\n",
        "clean_plasmid = load_and_process_plasmid(path_to_plasmid)\n",
        "print(clean_plasmid.name)\n",
        "\n",
        "# 3 Choose genes to knock out (list)\n",
        "genes_to_KO = ['SCO5087', 'SCO5088',]\n",
        "#genes_to_KO = ['80000-100000', '4000-7000', '9000-14000','15000-20000']\n",
        "\n",
        "#### Advanced settings ####\n",
        "# 4 Filtering metrics for sgRNAs\n",
        "gc_upper = 0.8\n",
        "gc_lower = 0.2\n",
        "off_target_seed = 8\n",
        "off_target_upper = 10\n",
        "cas_type='cas3'\n",
        "number_of_sgRNAs_per_group = 2\n",
        "\n",
        "# 5 Choose polymerase and target melting temperature\n",
        "chosen_polymerase = polymerase_dict['Phusion High-Fidelity DNA Polymerase (GC Buffer)']\n",
        "melting_temperature = 65\n",
        "primer_concentration = 0.4 \n",
        "primer_number_increment = 23\n",
        "flanking_region = 500\n",
        "\n",
        "# 6 Choose overlapping sequences for our plasmid we can use the following\n",
        "# the part of the primers that anneal to the generate the crRNA and protospacer\n",
        "forward_protospacer_overhang = 'GTCGCCCGGCAAAACCGG'\n",
        "reverse_protospacer_overhang = 'GTTTCAATCCACGCGCCCGT'\n",
        "\n",
        "# The universal part of the \n",
        "forward_backbone_overhang = 'GAGCTCATAAGTTCCTATTCCGAAG'\n",
        "reverse_backbone_overhang = 'AAGAAGTGGGTGTCGGACGC'\n",
        "\n",
        "Enzyme_for_repair_template_integration = EcoRI\n",
        "repair_length = 1000\n",
        "overlap_length = 40 \n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "# Computation"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 3,
      "metadata": {},
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "False\n"
          ]
        },
        {
          "data": {
            "text/plain": [
              "25824"
            ]
          },
          "execution_count": 3,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "target_dict, genes_to_KO, annotation_input = check_and_convert_input(genes_to_KO)\n",
        "\n",
        "print(annotation_input)\n",
        "if annotation_input == True:\n",
        "    genome = annotate_dseqrecord(genome, target_dict)\n",
        "\n",
        "\n",
        "len(genome.features)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 4,
      "metadata": {},
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "sgRNA generated were too smallSCO5088. To incorporate this extent borders. Skipping to next locus tag.\n"
          ]
        },
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>strain_name</th>\n",
              "      <th>locus_tag</th>\n",
              "      <th>gene_loc</th>\n",
              "      <th>gene_strand</th>\n",
              "      <th>sgrna_strand</th>\n",
              "      <th>sgrna_loc</th>\n",
              "      <th>gc</th>\n",
              "      <th>pam</th>\n",
              "      <th>sgrna</th>\n",
              "      <th>sgrna_seed_sequence</th>\n",
              "      <th>off_target_count</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5087</td>\n",
              "      <td>5529801</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>163</td>\n",
              "      <td>0.617647</td>\n",
              "      <td>TTC</td>\n",
              "      <td>GGATTGAAGCGCAGAGTCGTCATCACGGGCGTCG</td>\n",
              "      <td>GGATTGAA</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>6</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5087</td>\n",
              "      <td>5529801</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>306</td>\n",
              "      <td>0.735294</td>\n",
              "      <td>TTC</td>\n",
              "      <td>ACAGGTCGCGGCGGAGGCCGACTTCGACCCGGTC</td>\n",
              "      <td>ACAGGTCG</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>28</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5087</td>\n",
              "      <td>5529801</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>131</td>\n",
              "      <td>0.558824</td>\n",
              "      <td>TTC</td>\n",
              "      <td>AATCCGAACAGCTCCTTCGATGCGTTCGTCCGGT</td>\n",
              "      <td>AATCCGAA</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5087</td>\n",
              "      <td>5529801</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>94</td>\n",
              "      <td>0.705882</td>\n",
              "      <td>TTC</td>\n",
              "      <td>GAGCCTCCTTCGAGCCACGGGGCCGACGATGACG</td>\n",
              "      <td>GAGCCTCC</td>\n",
              "      <td>1</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>14</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5087</td>\n",
              "      <td>5529801</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>850</td>\n",
              "      <td>0.705882</td>\n",
              "      <td>TTC</td>\n",
              "      <td>GACGGTACGCGGGACGGCTTCGTGCTGGCCGAAG</td>\n",
              "      <td>GACGGTAC</td>\n",
              "      <td>1</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>24</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5087</td>\n",
              "      <td>5529801</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>962</td>\n",
              "      <td>0.676471</td>\n",
              "      <td>TTC</td>\n",
              "      <td>AGGCCCGTCATGTGGTACGCGTTGCAGCGCGTCG</td>\n",
              "      <td>AGGCCCGT</td>\n",
              "      <td>1</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>26</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5087</td>\n",
              "      <td>5529801</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>201</td>\n",
              "      <td>0.794118</td>\n",
              "      <td>TTC</td>\n",
              "      <td>CCAGAACTGCCGGGTGCCGTTCCCGCCGGGGGCG</td>\n",
              "      <td>CCAGAACT</td>\n",
              "      <td>1</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5087</td>\n",
              "      <td>5529801</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>105</td>\n",
              "      <td>0.735294</td>\n",
              "      <td>TTC</td>\n",
              "      <td>GAGCCACGGGGCCGACGATGACGACGACCACCGG</td>\n",
              "      <td>GAGCCACG</td>\n",
              "      <td>2</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>29</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5087</td>\n",
              "      <td>5529801</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>113</td>\n",
              "      <td>0.647059</td>\n",
              "      <td>TTC</td>\n",
              "      <td>GATGCGTTCGTCCGGTGGTCGTCGTCATCGTCGG</td>\n",
              "      <td>GATGCGTT</td>\n",
              "      <td>2</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>19</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5087</td>\n",
              "      <td>5529801</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>1390</td>\n",
              "      <td>0.735294</td>\n",
              "      <td>TTC</td>\n",
              "      <td>CAGAGCGCGATGGTGCTGCGCGACGCCGAGACCG</td>\n",
              "      <td>CAGAGCGC</td>\n",
              "      <td>3</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>43</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5088</td>\n",
              "      <td>5531201</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>1219</td>\n",
              "      <td>0.705882</td>\n",
              "      <td>TTC</td>\n",
              "      <td>AACTCGGCGGCGGTCCTGCGCCGTTTCGCACCGA</td>\n",
              "      <td>AACTCGGC</td>\n",
              "      <td>3</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>5</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5087</td>\n",
              "      <td>5529801</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>286</td>\n",
              "      <td>0.735294</td>\n",
              "      <td>TTC</td>\n",
              "      <td>GACCCCTCGCCCTACCGTTCACAGGTCGCGGCGG</td>\n",
              "      <td>GACCCCTC</td>\n",
              "      <td>4</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5087</td>\n",
              "      <td>5529801</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>235</td>\n",
              "      <td>0.735294</td>\n",
              "      <td>TTC</td>\n",
              "      <td>TGGGAACTGCTCACCTCGGGGCGGACGGCGACGC</td>\n",
              "      <td>TGGGAACT</td>\n",
              "      <td>4</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>34</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5088</td>\n",
              "      <td>5531201</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>370</td>\n",
              "      <td>0.676471</td>\n",
              "      <td>TTC</td>\n",
              "      <td>ACCCACCGGGAGTTCCGCAAGCTGTGGTCCGAGG</td>\n",
              "      <td>ACCCACCG</td>\n",
              "      <td>4</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5087</td>\n",
              "      <td>5529801</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>283</td>\n",
              "      <td>0.676471</td>\n",
              "      <td>TTC</td>\n",
              "      <td>TTCGACCCCTCGCCCTACCGTTCACAGGTCGCGG</td>\n",
              "      <td>TTCGACCC</td>\n",
              "      <td>5</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>36</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5088</td>\n",
              "      <td>5531201</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>436</td>\n",
              "      <td>0.617647</td>\n",
              "      <td>TTC</td>\n",
              "      <td>GCCTGGTTCTACGCGGTGAACACCGGCCAGATCT</td>\n",
              "      <td>GCCTGGTT</td>\n",
              "      <td>6</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>23</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5087</td>\n",
              "      <td>5529801</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>1073</td>\n",
              "      <td>0.676471</td>\n",
              "      <td>TTC</td>\n",
              "      <td>TGCCGGGTGCCGGAGCCGTGCGCGTTGATGTAGT</td>\n",
              "      <td>TGCCGGGT</td>\n",
              "      <td>7</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>51</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5088</td>\n",
              "      <td>5531201</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>257</td>\n",
              "      <td>0.794118</td>\n",
              "      <td>TTC</td>\n",
              "      <td>GCGTCCTGGAGGGCCCAGTCCGCGGCGGTCAGCG</td>\n",
              "      <td>GCGTCCTG</td>\n",
              "      <td>7</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>37</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5088</td>\n",
              "      <td>5531201</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>445</td>\n",
              "      <td>0.647059</td>\n",
              "      <td>TTC</td>\n",
              "      <td>TACGCGGTGAACACCGGCCAGATCTCCATCCGGC</td>\n",
              "      <td>TACGCGGT</td>\n",
              "      <td>8</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>47</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5088</td>\n",
              "      <td>5531201</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>842</td>\n",
              "      <td>0.794118</td>\n",
              "      <td>TTC</td>\n",
              "      <td>AGCGCCAGCCGGATCGCGCGTTCCAGGCCCGCGG</td>\n",
              "      <td>AGCGCCAG</td>\n",
              "      <td>8</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>35</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5088</td>\n",
              "      <td>5531201</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>385</td>\n",
              "      <td>0.676471</td>\n",
              "      <td>TTC</td>\n",
              "      <td>CGCAAGCTGTGGTCCGAGGGCCCCAAGTCGGTCA</td>\n",
              "      <td>CGCAAGCT</td>\n",
              "      <td>8</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>39</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5088</td>\n",
              "      <td>5531201</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>716</td>\n",
              "      <td>0.764706</td>\n",
              "      <td>TTC</td>\n",
              "      <td>CGGGCGAGGGCGGGGCCATCCTCGTCCTGGAGGA</td>\n",
              "      <td>CGGGCGAG</td>\n",
              "      <td>8</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>30</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5087</td>\n",
              "      <td>5529801</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>104</td>\n",
              "      <td>0.735294</td>\n",
              "      <td>TTC</td>\n",
              "      <td>GTCCGGTGGTCGTCGTCATCGTCGGCCCCGTGGC</td>\n",
              "      <td>GTCCGGTG</td>\n",
              "      <td>8</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>50</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5088</td>\n",
              "      <td>5531201</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>416</td>\n",
              "      <td>0.617647</td>\n",
              "      <td>TTC</td>\n",
              "      <td>ACCGCGTAGAACCAGGCGAAGGACTCGTACACGC</td>\n",
              "      <td>ACCGCGTA</td>\n",
              "      <td>9</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>17</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5087</td>\n",
              "      <td>5529801</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>1325</td>\n",
              "      <td>0.735294</td>\n",
              "      <td>TTC</td>\n",
              "      <td>CGCTGGAGGCCAGGGAGCGGAAGCTGCGGTCCGT</td>\n",
              "      <td>CGCTGGAG</td>\n",
              "      <td>9</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>33</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5088</td>\n",
              "      <td>5531201</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>364</td>\n",
              "      <td>0.617647</td>\n",
              "      <td>TTC</td>\n",
              "      <td>GACTTCACCCACCGGGAGTTCCGCAAGCTGTGGT</td>\n",
              "      <td>GACTTCAC</td>\n",
              "      <td>9</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>20</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5087</td>\n",
              "      <td>5529801</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>1307</td>\n",
              "      <td>0.647059</td>\n",
              "      <td>TTC</td>\n",
              "      <td>CGCTCCCTGGCCTCCAGCGGAACGTAGTCGAGAT</td>\n",
              "      <td>CGCTCCCT</td>\n",
              "      <td>10</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>12</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5087</td>\n",
              "      <td>5529801</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>742</td>\n",
              "      <td>0.764706</td>\n",
              "      <td>TTC</td>\n",
              "      <td>GCGGGAGCCGCCGACACCCCGATCACCCCGATCG</td>\n",
              "      <td>GCGGGAGC</td>\n",
              "      <td>10</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>11</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5087</td>\n",
              "      <td>5529801</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>586</td>\n",
              "      <td>0.705882</td>\n",
              "      <td>TTC</td>\n",
              "      <td>GACTACCTGGTGCCCAGCGTCATGCCGGCCGAGG</td>\n",
              "      <td>GACTACCT</td>\n",
              "      <td>10</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "    strain_name locus_tag  gene_loc  gene_strand  sgrna_strand  sgrna_loc  \\\n",
              "2   NC_003888.3   SCO5087   5529801            1            -1        163   \n",
              "6   NC_003888.3   SCO5087   5529801            1            -1        306   \n",
              "28  NC_003888.3   SCO5087   5529801            1             1        131   \n",
              "0   NC_003888.3   SCO5087   5529801            1            -1         94   \n",
              "14  NC_003888.3   SCO5087   5529801            1            -1        850   \n",
              "24  NC_003888.3   SCO5087   5529801            1             1        962   \n",
              "26  NC_003888.3   SCO5087   5529801            1             1        201   \n",
              "1   NC_003888.3   SCO5087   5529801            1            -1        105   \n",
              "29  NC_003888.3   SCO5087   5529801            1             1        113   \n",
              "19  NC_003888.3   SCO5087   5529801            1            -1       1390   \n",
              "43  NC_003888.3   SCO5088   5531201            1            -1       1219   \n",
              "5   NC_003888.3   SCO5087   5529801            1            -1        286   \n",
              "3   NC_003888.3   SCO5087   5529801            1            -1        235   \n",
              "34  NC_003888.3   SCO5088   5531201            1            -1        370   \n",
              "4   NC_003888.3   SCO5087   5529801            1            -1        283   \n",
              "36  NC_003888.3   SCO5088   5531201            1            -1        436   \n",
              "23  NC_003888.3   SCO5087   5529801            1             1       1073   \n",
              "51  NC_003888.3   SCO5088   5531201            1             1        257   \n",
              "37  NC_003888.3   SCO5088   5531201            1            -1        445   \n",
              "47  NC_003888.3   SCO5088   5531201            1             1        842   \n",
              "35  NC_003888.3   SCO5088   5531201            1            -1        385   \n",
              "39  NC_003888.3   SCO5088   5531201            1            -1        716   \n",
              "30  NC_003888.3   SCO5087   5529801            1             1        104   \n",
              "50  NC_003888.3   SCO5088   5531201            1             1        416   \n",
              "17  NC_003888.3   SCO5087   5529801            1            -1       1325   \n",
              "33  NC_003888.3   SCO5088   5531201            1            -1        364   \n",
              "20  NC_003888.3   SCO5087   5529801            1             1       1307   \n",
              "12  NC_003888.3   SCO5087   5529801            1            -1        742   \n",
              "11  NC_003888.3   SCO5087   5529801            1            -1        586   \n",
              "\n",
              "          gc  pam                               sgrna sgrna_seed_sequence  \\\n",
              "2   0.617647  TTC  GGATTGAAGCGCAGAGTCGTCATCACGGGCGTCG            GGATTGAA   \n",
              "6   0.735294  TTC  ACAGGTCGCGGCGGAGGCCGACTTCGACCCGGTC            ACAGGTCG   \n",
              "28  0.558824  TTC  AATCCGAACAGCTCCTTCGATGCGTTCGTCCGGT            AATCCGAA   \n",
              "0   0.705882  TTC  GAGCCTCCTTCGAGCCACGGGGCCGACGATGACG            GAGCCTCC   \n",
              "14  0.705882  TTC  GACGGTACGCGGGACGGCTTCGTGCTGGCCGAAG            GACGGTAC   \n",
              "24  0.676471  TTC  AGGCCCGTCATGTGGTACGCGTTGCAGCGCGTCG            AGGCCCGT   \n",
              "26  0.794118  TTC  CCAGAACTGCCGGGTGCCGTTCCCGCCGGGGGCG            CCAGAACT   \n",
              "1   0.735294  TTC  GAGCCACGGGGCCGACGATGACGACGACCACCGG            GAGCCACG   \n",
              "29  0.647059  TTC  GATGCGTTCGTCCGGTGGTCGTCGTCATCGTCGG            GATGCGTT   \n",
              "19  0.735294  TTC  CAGAGCGCGATGGTGCTGCGCGACGCCGAGACCG            CAGAGCGC   \n",
              "43  0.705882  TTC  AACTCGGCGGCGGTCCTGCGCCGTTTCGCACCGA            AACTCGGC   \n",
              "5   0.735294  TTC  GACCCCTCGCCCTACCGTTCACAGGTCGCGGCGG            GACCCCTC   \n",
              "3   0.735294  TTC  TGGGAACTGCTCACCTCGGGGCGGACGGCGACGC            TGGGAACT   \n",
              "34  0.676471  TTC  ACCCACCGGGAGTTCCGCAAGCTGTGGTCCGAGG            ACCCACCG   \n",
              "4   0.676471  TTC  TTCGACCCCTCGCCCTACCGTTCACAGGTCGCGG            TTCGACCC   \n",
              "36  0.617647  TTC  GCCTGGTTCTACGCGGTGAACACCGGCCAGATCT            GCCTGGTT   \n",
              "23  0.676471  TTC  TGCCGGGTGCCGGAGCCGTGCGCGTTGATGTAGT            TGCCGGGT   \n",
              "51  0.794118  TTC  GCGTCCTGGAGGGCCCAGTCCGCGGCGGTCAGCG            GCGTCCTG   \n",
              "37  0.647059  TTC  TACGCGGTGAACACCGGCCAGATCTCCATCCGGC            TACGCGGT   \n",
              "47  0.794118  TTC  AGCGCCAGCCGGATCGCGCGTTCCAGGCCCGCGG            AGCGCCAG   \n",
              "35  0.676471  TTC  CGCAAGCTGTGGTCCGAGGGCCCCAAGTCGGTCA            CGCAAGCT   \n",
              "39  0.764706  TTC  CGGGCGAGGGCGGGGCCATCCTCGTCCTGGAGGA            CGGGCGAG   \n",
              "30  0.735294  TTC  GTCCGGTGGTCGTCGTCATCGTCGGCCCCGTGGC            GTCCGGTG   \n",
              "50  0.617647  TTC  ACCGCGTAGAACCAGGCGAAGGACTCGTACACGC            ACCGCGTA   \n",
              "17  0.735294  TTC  CGCTGGAGGCCAGGGAGCGGAAGCTGCGGTCCGT            CGCTGGAG   \n",
              "33  0.617647  TTC  GACTTCACCCACCGGGAGTTCCGCAAGCTGTGGT            GACTTCAC   \n",
              "20  0.647059  TTC  CGCTCCCTGGCCTCCAGCGGAACGTAGTCGAGAT            CGCTCCCT   \n",
              "12  0.764706  TTC  GCGGGAGCCGCCGACACCCCGATCACCCCGATCG            GCGGGAGC   \n",
              "11  0.705882  TTC  GACTACCTGGTGCCCAGCGTCATGCCGGCCGAGG            GACTACCT   \n",
              "\n",
              "    off_target_count  \n",
              "2                  0  \n",
              "6                  0  \n",
              "28                 0  \n",
              "0                  1  \n",
              "14                 1  \n",
              "24                 1  \n",
              "26                 1  \n",
              "1                  2  \n",
              "29                 2  \n",
              "19                 3  \n",
              "43                 3  \n",
              "5                  4  \n",
              "3                  4  \n",
              "34                 4  \n",
              "4                  5  \n",
              "36                 6  \n",
              "23                 7  \n",
              "51                 7  \n",
              "37                 8  \n",
              "47                 8  \n",
              "35                 8  \n",
              "39                 8  \n",
              "30                 8  \n",
              "50                 9  \n",
              "17                 9  \n",
              "33                 9  \n",
              "20                10  \n",
              "12                10  \n",
              "11                10  "
            ]
          },
          "execution_count": 4,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "# Initialize SgRNAargs with desired parameters\n",
        "args = SgRNAargs(genome, \n",
        "                genes_to_KO,\n",
        "                step=['find', 'filter'],\n",
        "                gc_upper = gc_upper,\n",
        "                gc_lower = gc_lower,\n",
        "                off_target_seed = off_target_seed,\n",
        "                off_target_upper = off_target_upper,\n",
        "                cas_type=cas_type\n",
        "                )\n",
        "\n",
        "sgrna_df = extract_sgRNAs(args)\n",
        "sgrna_df"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 5,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>strain_name</th>\n",
              "      <th>locus_tag</th>\n",
              "      <th>gene_loc</th>\n",
              "      <th>gene_strand</th>\n",
              "      <th>sgrna_strand</th>\n",
              "      <th>sgrna_loc</th>\n",
              "      <th>gc</th>\n",
              "      <th>pam</th>\n",
              "      <th>sgrna</th>\n",
              "      <th>sgrna_seed_sequence</th>\n",
              "      <th>off_target_count</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5087</td>\n",
              "      <td>5529801</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>163</td>\n",
              "      <td>0.617647</td>\n",
              "      <td>TTC</td>\n",
              "      <td>GGATTGAAGCGCAGAGTCGTCATCACGGGCGTCG</td>\n",
              "      <td>GGATTGAA</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>6</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5087</td>\n",
              "      <td>5529801</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>306</td>\n",
              "      <td>0.735294</td>\n",
              "      <td>TTC</td>\n",
              "      <td>ACAGGTCGCGGCGGAGGCCGACTTCGACCCGGTC</td>\n",
              "      <td>ACAGGTCG</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>43</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5088</td>\n",
              "      <td>5531201</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>1219</td>\n",
              "      <td>0.705882</td>\n",
              "      <td>TTC</td>\n",
              "      <td>AACTCGGCGGCGGTCCTGCGCCGTTTCGCACCGA</td>\n",
              "      <td>AACTCGGC</td>\n",
              "      <td>3</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>34</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5088</td>\n",
              "      <td>5531201</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>370</td>\n",
              "      <td>0.676471</td>\n",
              "      <td>TTC</td>\n",
              "      <td>ACCCACCGGGAGTTCCGCAAGCTGTGGTCCGAGG</td>\n",
              "      <td>ACCCACCG</td>\n",
              "      <td>4</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "    strain_name locus_tag  gene_loc  gene_strand  sgrna_strand  sgrna_loc  \\\n",
              "2   NC_003888.3   SCO5087   5529801            1            -1        163   \n",
              "6   NC_003888.3   SCO5087   5529801            1            -1        306   \n",
              "43  NC_003888.3   SCO5088   5531201            1            -1       1219   \n",
              "34  NC_003888.3   SCO5088   5531201            1            -1        370   \n",
              "\n",
              "          gc  pam                               sgrna sgrna_seed_sequence  \\\n",
              "2   0.617647  TTC  GGATTGAAGCGCAGAGTCGTCATCACGGGCGTCG            GGATTGAA   \n",
              "6   0.735294  TTC  ACAGGTCGCGGCGGAGGCCGACTTCGACCCGGTC            ACAGGTCG   \n",
              "43  0.705882  TTC  AACTCGGCGGCGGTCCTGCGCCGTTTCGCACCGA            AACTCGGC   \n",
              "34  0.676471  TTC  ACCCACCGGGAGTTCCGCAAGCTGTGGTCCGAGG            ACCCACCG   \n",
              "\n",
              "    off_target_count  \n",
              "2                  0  \n",
              "6                  0  \n",
              "43                 3  \n",
              "34                 4  "
            ]
          },
          "execution_count": 5,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "# Filter the DataFrame to retain only up to 5 sgRNA sequences per locus_tag\n",
        "filtered_df = sgrna_df.groupby('locus_tag').head(number_of_sgRNAs_per_group)\n",
        "filtered_df"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "## Output"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 6,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>strain_name</th>\n",
              "      <th>locus_tag</th>\n",
              "      <th>gene_loc</th>\n",
              "      <th>gene_strand</th>\n",
              "      <th>sgrna_strand</th>\n",
              "      <th>sgrna_loc</th>\n",
              "      <th>gc</th>\n",
              "      <th>pam</th>\n",
              "      <th>sgrna</th>\n",
              "      <th>sgrna_seed_sequence</th>\n",
              "      <th>off_target_count</th>\n",
              "      <th>Fwd Primer</th>\n",
              "      <th>Rev Primer</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5087</td>\n",
              "      <td>5529801</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>163</td>\n",
              "      <td>0.617647</td>\n",
              "      <td>TTC</td>\n",
              "      <td>GGATTGAAGCGCAGAGTCGTCATCACGGGCGTCG</td>\n",
              "      <td>GGATTGAA</td>\n",
              "      <td>0</td>\n",
              "      <td>GGATTGAAGCGCAGAGTCGTCATCACGGGCGTCGGTCGCCCGGCAA...</td>\n",
              "      <td>CGACGCCCGTGATGACGACTCTGCGCTTCAATCCGTTTCAATCCAC...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>6</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5087</td>\n",
              "      <td>5529801</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>306</td>\n",
              "      <td>0.735294</td>\n",
              "      <td>TTC</td>\n",
              "      <td>ACAGGTCGCGGCGGAGGCCGACTTCGACCCGGTC</td>\n",
              "      <td>ACAGGTCG</td>\n",
              "      <td>0</td>\n",
              "      <td>ACAGGTCGCGGCGGAGGCCGACTTCGACCCGGTCGTCGCCCGGCAA...</td>\n",
              "      <td>GACCGGGTCGAAGTCGGCCTCCGCCGCGACCTGTGTTTCAATCCAC...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>43</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5088</td>\n",
              "      <td>5531201</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>1219</td>\n",
              "      <td>0.705882</td>\n",
              "      <td>TTC</td>\n",
              "      <td>AACTCGGCGGCGGTCCTGCGCCGTTTCGCACCGA</td>\n",
              "      <td>AACTCGGC</td>\n",
              "      <td>3</td>\n",
              "      <td>AACTCGGCGGCGGTCCTGCGCCGTTTCGCACCGAGTCGCCCGGCAA...</td>\n",
              "      <td>TCGGTGCGAAACGGCGCAGGACCGCCGCCGAGTTGTTTCAATCCAC...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>34</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO5088</td>\n",
              "      <td>5531201</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>370</td>\n",
              "      <td>0.676471</td>\n",
              "      <td>TTC</td>\n",
              "      <td>ACCCACCGGGAGTTCCGCAAGCTGTGGTCCGAGG</td>\n",
              "      <td>ACCCACCG</td>\n",
              "      <td>4</td>\n",
              "      <td>ACCCACCGGGAGTTCCGCAAGCTGTGGTCCGAGGGTCGCCCGGCAA...</td>\n",
              "      <td>CCTCGGACCACAGCTTGCGGAACTCCCGGTGGGTGTTTCAATCCAC...</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "    strain_name locus_tag  gene_loc  gene_strand  sgrna_strand  sgrna_loc  \\\n",
              "2   NC_003888.3   SCO5087   5529801            1            -1        163   \n",
              "6   NC_003888.3   SCO5087   5529801            1            -1        306   \n",
              "43  NC_003888.3   SCO5088   5531201            1            -1       1219   \n",
              "34  NC_003888.3   SCO5088   5531201            1            -1        370   \n",
              "\n",
              "          gc  pam                               sgrna sgrna_seed_sequence  \\\n",
              "2   0.617647  TTC  GGATTGAAGCGCAGAGTCGTCATCACGGGCGTCG            GGATTGAA   \n",
              "6   0.735294  TTC  ACAGGTCGCGGCGGAGGCCGACTTCGACCCGGTC            ACAGGTCG   \n",
              "43  0.705882  TTC  AACTCGGCGGCGGTCCTGCGCCGTTTCGCACCGA            AACTCGGC   \n",
              "34  0.676471  TTC  ACCCACCGGGAGTTCCGCAAGCTGTGGTCCGAGG            ACCCACCG   \n",
              "\n",
              "    off_target_count                                         Fwd Primer  \\\n",
              "2                  0  GGATTGAAGCGCAGAGTCGTCATCACGGGCGTCGGTCGCCCGGCAA...   \n",
              "6                  0  ACAGGTCGCGGCGGAGGCCGACTTCGACCCGGTCGTCGCCCGGCAA...   \n",
              "43                 3  AACTCGGCGGCGGTCCTGCGCCGTTTCGCACCGAGTCGCCCGGCAA...   \n",
              "34                 4  ACCCACCGGGAGTTCCGCAAGCTGTGGTCCGAGGGTCGCCCGGCAA...   \n",
              "\n",
              "                                           Rev Primer  \n",
              "2   CGACGCCCGTGATGACGACTCTGCGCTTCAATCCGTTTCAATCCAC...  \n",
              "6   GACCGGGTCGAAGTCGGCCTCCGCCGCGACCTGTGTTTCAATCCAC...  \n",
              "43  TCGGTGCGAAACGGCGCAGGACCGCCGCCGAGTTGTTTCAATCCAC...  \n",
              "34  CCTCGGACCACAGCTTGCGGAACTCCCGGTGGGTGTTTCAATCCAC...  "
            ]
          },
          "execution_count": 6,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "filtered_df_w_primers = generate_cas3_protospacer_primers(filtered_df,\n",
        "                                                          fwd_overhang= forward_protospacer_overhang, \n",
        "                                                          rev_overhang=reverse_protospacer_overhang)\n",
        "filtered_df_w_primers"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 7,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/plain": [
              "[[Amplicon(1489), Amplicon(460)],\n",
              " [Amplicon(1489), Amplicon(460)],\n",
              " [Amplicon(1489), Amplicon(460)],\n",
              " [Amplicon(1489), Amplicon(460)]]"
            ]
          },
          "execution_count": 7,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "amplicons = cas3_plasmid_pcrs(clean_plasmid, \n",
        "                              filtered_df, \n",
        "                              universal_fwd_seq=forward_backbone_overhang, \n",
        "                              universal_rev_seq=reverse_backbone_overhang\n",
        "                              )\n",
        "\n",
        "amplicons"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 8,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/plain": [
              "[[Amplicon(1489), Amplicon(460)],\n",
              " [Amplicon(1489), Amplicon(460)],\n",
              " [Amplicon(1489), Amplicon(460)],\n",
              " [Amplicon(1489), Amplicon(460)]]"
            ]
          },
          "execution_count": 8,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "amplicons = cas3_plasmid_pcrs(clean_plasmid, filtered_df)\n",
        "\n",
        "amplicons"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 9,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/plain": [
              "[Contig(o12766), Contig(o12766), Contig(o12766), Contig(o12766)]"
            ]
          },
          "execution_count": 9,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "# cut plasmid and incorporate amplicons\n",
        "assembled_cas3_plasmids = assemble_cas3_plasmids(clean_plasmid, amplicons)\n",
        "assembled_cas3_plasmids"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 10,
      "metadata": {},
      "outputs": [],
      "source": [
        "# Constructing a meaningful name, ID, and description for the assembled plasmid using user input\n",
        "targeting_info = []\n",
        "for index, row in filtered_df.iterrows():\n",
        "    formatted_str = f\"pCas3_{row['locus_tag']}({row['sgrna_loc']})\"\n",
        "    targeting_info.append(formatted_str)\n",
        "\n",
        "for i in range(len(assembled_cas3_plasmids)):\n",
        "    assembled_cas3_plasmids[i].name = f'{targeting_info[i]}'\n",
        "    assembled_cas3_plasmids[i].id = assembled_cas3_plasmids[i].name  \n",
        "    assembled_cas3_plasmids[i].description = f'CRISPR-Cas3 targeting {\", \".join(genes_to_KO)} for single gene knockout, assembled using StreptoCAD.'\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 11,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/plain": [
              "'pCas3_SCO5087(163)'"
            ]
          },
          "execution_count": 11,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "assembled_cas3_plasmids[0].name"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 12,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/plain": [
              "'CRISPR-Cas3 targeting SCO5087, SCO5088 for single gene knockout, assembled using StreptoCAD.'"
            ]
          },
          "execution_count": 12,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "assembled_cas3_plasmids[0].description"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 13,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/plain": [
              "'pCas3_SCO5087(163)'"
            ]
          },
          "execution_count": 13,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "assembled_cas3_plasmids[0].id"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 14,
      "metadata": {},
      "outputs": [],
      "source": [
        "print_plasmids = False\n",
        "\n",
        "if print_plasmids: \n",
        "    for vector in assembled_cas3_plasmids: \n",
        "        vector.write(f\"../../data/plasmids/sgRNA_plasmids_pCRISPRâ€“Cas3/{vector.id}.gb\")"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "### IDT primers"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 15,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/plain": [
              "[Dseqrecord(-52),\n",
              " Dseqrecord(-54),\n",
              " Dseqrecord(-52),\n",
              " Dseqrecord(-54),\n",
              " Dseqrecord(-52),\n",
              " Dseqrecord(-54),\n",
              " Dseqrecord(-52),\n",
              " Dseqrecord(-54)]"
            ]
          },
          "execution_count": 15,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "primer_records = make_primer_records(filtered_df_w_primers)\n",
        "primer_records"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 16,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>Name</th>\n",
              "      <th>Sequence</th>\n",
              "      <th>Concentration</th>\n",
              "      <th>Purification</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>SCO5087_loc_163_fwd</td>\n",
              "      <td>GGATTGAAGCGCAGAGTCGTCATCACGGGCGTCGGTCGCCCGGCAA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>SCO5087_loc_163_rev</td>\n",
              "      <td>CGACGCCCGTGATGACGACTCTGCGCTTCAATCCGTTTCAATCCAC...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>SCO5087_loc_306_fwd</td>\n",
              "      <td>ACAGGTCGCGGCGGAGGCCGACTTCGACCCGGTCGTCGCCCGGCAA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>SCO5087_loc_306_rev</td>\n",
              "      <td>GACCGGGTCGAAGTCGGCCTCCGCCGCGACCTGTGTTTCAATCCAC...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>SCO5088_loc_1219_fwd</td>\n",
              "      <td>AACTCGGCGGCGGTCCTGCGCCGTTTCGCACCGAGTCGCCCGGCAA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>5</th>\n",
              "      <td>SCO5088_loc_1219_rev</td>\n",
              "      <td>TCGGTGCGAAACGGCGCAGGACCGCCGCCGAGTTGTTTCAATCCAC...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>6</th>\n",
              "      <td>SCO5088_loc_370_fwd</td>\n",
              "      <td>ACCCACCGGGAGTTCCGCAAGCTGTGGTCCGAGGGTCGCCCGGCAA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>7</th>\n",
              "      <td>SCO5088_loc_370_rev</td>\n",
              "      <td>CCTCGGACCACAGCTTGCGGAACTCCCGGTGGGTGTTTCAATCCAC...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "                   Name                                           Sequence  \\\n",
              "0   SCO5087_loc_163_fwd  GGATTGAAGCGCAGAGTCGTCATCACGGGCGTCGGTCGCCCGGCAA...   \n",
              "1   SCO5087_loc_163_rev  CGACGCCCGTGATGACGACTCTGCGCTTCAATCCGTTTCAATCCAC...   \n",
              "2   SCO5087_loc_306_fwd  ACAGGTCGCGGCGGAGGCCGACTTCGACCCGGTCGTCGCCCGGCAA...   \n",
              "3   SCO5087_loc_306_rev  GACCGGGTCGAAGTCGGCCTCCGCCGCGACCTGTGTTTCAATCCAC...   \n",
              "4  SCO5088_loc_1219_fwd  AACTCGGCGGCGGTCCTGCGCCGTTTCGCACCGAGTCGCCCGGCAA...   \n",
              "5  SCO5088_loc_1219_rev  TCGGTGCGAAACGGCGCAGGACCGCCGCCGAGTTGTTTCAATCCAC...   \n",
              "6   SCO5088_loc_370_fwd  ACCCACCGGGAGTTCCGCAAGCTGTGGTCCGAGGGTCGCCCGGCAA...   \n",
              "7   SCO5088_loc_370_rev  CCTCGGACCACAGCTTGCGGAACTCCCGGTGGGTGTTTCAATCCAC...   \n",
              "\n",
              "  Concentration Purification  \n",
              "0          25nm          STD  \n",
              "1          25nm          STD  \n",
              "2          25nm          STD  \n",
              "3          25nm          STD  \n",
              "4          25nm          STD  \n",
              "5          25nm          STD  \n",
              "6          25nm          STD  \n",
              "7          25nm          STD  "
            ]
          },
          "execution_count": 16,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "idt_primers_cas3=primers_to_IDT(primer_records)\n",
        "idt_primers_cas3"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "# In frame deletion"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 17,
      "metadata": {},
      "outputs": [],
      "source": [
        "in_frame_deletion = True \n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 18,
      "metadata": {},
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "[Contig(o12766), Contig(o12766), Contig(o12766), Contig(o12766)]\n"
          ]
        }
      ],
      "source": [
        "print(assembled_cas3_plasmids)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 19,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/plain": [
              "(Dseqrecord(-12770),)"
            ]
          },
          "execution_count": 19,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "Dseqrecord(assembled_cas3_plasmids[3], circular=True).cut(EcoRI)"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "#### Getting repair templates"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 20,
      "metadata": {},
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "pCas3_SCO5087(163)\n",
            "pCas3_SCO5087(306)\n",
            "pCas3_SCO5088(1219)\n",
            "pCas3_SCO5088(370)\n",
            "\n",
            "Processing gene: SCO5087\n",
            "Checking plasmid: pCas3_SCO5087(163)\n",
            "Match found in plasmid: pCas3_SCO5087(163)\n",
            "Repair template match found for gene: SCO5087\n",
            "Record added for gene: SCO5087\n",
            "Checking plasmid: pCas3_SCO5087(306)\n",
            "Match found in plasmid: pCas3_SCO5087(306)\n",
            "Repair template match found for gene: SCO5087\n",
            "Record added for gene: SCO5087\n",
            "Checking plasmid: pCas3_SCO5088(1219)\n",
            "Checking plasmid: pCas3_SCO5088(370)\n",
            "\n",
            "Processing gene: SCO5088\n",
            "Checking plasmid: pCas3_SCO5087(163)\n",
            "Checking plasmid: pCas3_SCO5087(306)\n",
            "Checking plasmid: pCas3_SCO5088(1219)\n",
            "Match found in plasmid: pCas3_SCO5088(1219)\n",
            "Repair template match found for gene: SCO5088\n",
            "Record added for gene: SCO5088\n",
            "Checking plasmid: pCas3_SCO5088(370)\n",
            "Match found in plasmid: pCas3_SCO5088(370)\n",
            "Repair template match found for gene: SCO5088\n",
            "Record added for gene: SCO5088\n"
          ]
        },
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>template</th>\n",
              "      <th>direction</th>\n",
              "      <th>f_primer_anneal(5-3)</th>\n",
              "      <th>r_primer_anneal(5-3)</th>\n",
              "      <th>f_tm</th>\n",
              "      <th>r_tm</th>\n",
              "      <th>ta</th>\n",
              "      <th>f_primer_sequences(5-3)</th>\n",
              "      <th>r_primer_sequences(5-3)</th>\n",
              "      <th>f_primer_name</th>\n",
              "      <th>r_primer_name</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>SCO5087</td>\n",
              "      <td>upstream</td>\n",
              "      <td>GACGCCGGAAATCCAG</td>\n",
              "      <td>CCCATCTCCCTTCGAC</td>\n",
              "      <td>56</td>\n",
              "      <td>54</td>\n",
              "      <td>58</td>\n",
              "      <td>AGAAGCTTTCTAGAGGATCCCCGGGTACCGAGCTCGAATTGACGCC...</td>\n",
              "      <td>GACTCCGGTGATGAGGACGCCCCATCTCCCTTCGAC</td>\n",
              "      <td>SCO5087_repair_up_forwar_p</td>\n",
              "      <td>SCO5087_repair_up_reverse_p</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>SCO5087</td>\n",
              "      <td>downstream</td>\n",
              "      <td>GCGTCCTCATCACCGG</td>\n",
              "      <td>GAGTAGAGGCGGCCC</td>\n",
              "      <td>58</td>\n",
              "      <td>57</td>\n",
              "      <td>61</td>\n",
              "      <td>GGCGGTCGAAGGGAGATGGGGCGTCCTCATCACCGG</td>\n",
              "      <td>AACGTCTGGAAAGACGACAAAACTTTAGATCTGGGGAATTGAGTAG...</td>\n",
              "      <td>SCO5087_repair_dw_forwar_p</td>\n",
              "      <td>SCO5087_repair_dw_reverse_p</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>SCO5088</td>\n",
              "      <td>upstream</td>\n",
              "      <td>ACCCGGACACCCTC</td>\n",
              "      <td>GCGGCCGCCCCCG</td>\n",
              "      <td>55</td>\n",
              "      <td>66</td>\n",
              "      <td>59</td>\n",
              "      <td>AGAAGCTTTCTAGAGGATCCCCGGGTACCGAGCTCGAATTACCCGG...</td>\n",
              "      <td>TTCTCCAGTTCTCGGGGTCGGCGGCCGCCCCCG</td>\n",
              "      <td>SCO5088_repair_up_forwar_p</td>\n",
              "      <td>SCO5088_repair_up_reverse_p</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>5</th>\n",
              "      <td>SCO5088</td>\n",
              "      <td>downstream</td>\n",
              "      <td>CGACCCCGAGAACTGG</td>\n",
              "      <td>TACACGATGCTCCTCGC</td>\n",
              "      <td>58</td>\n",
              "      <td>58</td>\n",
              "      <td>61</td>\n",
              "      <td>GAGACCGCGGGGGCGGCCGCCGACCCCGAGAACTGG</td>\n",
              "      <td>AACGTCTGGAAAGACGACAAAACTTTAGATCTGGGGAATTTACACG...</td>\n",
              "      <td>SCO5088_repair_dw_forwar_p</td>\n",
              "      <td>SCO5088_repair_dw_reverse_p</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "  template   direction f_primer_anneal(5-3) r_primer_anneal(5-3)  f_tm  r_tm  \\\n",
              "0  SCO5087    upstream     GACGCCGGAAATCCAG     CCCATCTCCCTTCGAC    56    54   \n",
              "1  SCO5087  downstream     GCGTCCTCATCACCGG      GAGTAGAGGCGGCCC    58    57   \n",
              "4  SCO5088    upstream       ACCCGGACACCCTC        GCGGCCGCCCCCG    55    66   \n",
              "5  SCO5088  downstream     CGACCCCGAGAACTGG    TACACGATGCTCCTCGC    58    58   \n",
              "\n",
              "   ta                            f_primer_sequences(5-3)  \\\n",
              "0  58  AGAAGCTTTCTAGAGGATCCCCGGGTACCGAGCTCGAATTGACGCC...   \n",
              "1  61               GGCGGTCGAAGGGAGATGGGGCGTCCTCATCACCGG   \n",
              "4  59  AGAAGCTTTCTAGAGGATCCCCGGGTACCGAGCTCGAATTACCCGG...   \n",
              "5  61               GAGACCGCGGGGGCGGCCGCCGACCCCGAGAACTGG   \n",
              "\n",
              "                             r_primer_sequences(5-3)  \\\n",
              "0               GACTCCGGTGATGAGGACGCCCCATCTCCCTTCGAC   \n",
              "1  AACGTCTGGAAAGACGACAAAACTTTAGATCTGGGGAATTGAGTAG...   \n",
              "4                  TTCTCCAGTTCTCGGGGTCGGCGGCCGCCCCCG   \n",
              "5  AACGTCTGGAAAGACGACAAAACTTTAGATCTGGGGAATTTACACG...   \n",
              "\n",
              "                f_primer_name                r_primer_name  \n",
              "0  SCO5087_repair_up_forwar_p  SCO5087_repair_up_reverse_p  \n",
              "1  SCO5087_repair_dw_forwar_p  SCO5087_repair_dw_reverse_p  \n",
              "4  SCO5088_repair_up_forwar_p  SCO5088_repair_up_reverse_p  \n",
              "5  SCO5088_repair_dw_forwar_p  SCO5088_repair_dw_reverse_p  "
            ]
          },
          "execution_count": 20,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "if in_frame_deletion:\n",
        "    \n",
        "\n",
        "    # Make repair templates\n",
        "    repair_templates_data = find_up_dw_repair_templates(genome, \n",
        "                                                        genes_to_KO, \n",
        "                                                        target_tm=melting_temperature, \n",
        "                                                        primer_tm_kwargs={'conc':primer_concentration, 'prodcode':chosen_polymerase} , \n",
        "                                                        repair_length=repair_length)\n",
        "\n",
        "    # Digest the plasmids\n",
        "    processed_records = [sorted(Dseqrecord(record, circular=True).cut(Enzyme_for_repair_template_integration), key=lambda x: len(x), reverse=True)[0] for record in assembled_cas3_plasmids]\n",
        "\n",
        "    # Rename them appropriately\n",
        "    for i in range(len(processed_records)):\n",
        "        processed_records[i].name = assembled_cas3_plasmids[i].name\n",
        "        print(processed_records[i].name)\n",
        "\n",
        "    # Assembly \n",
        "    assembly_data = assemble_multiple_plasmids_with_repair_templates_for_deletion(genes_to_KO, \n",
        "                                                                                processed_records, \n",
        "                                                                                repair_templates_data, \n",
        "                                                                                overlap=overlap_length)\n",
        "    # updating the primer names to something systematic.\n",
        "    update_primer_names(assembly_data)\n",
        "\n",
        "    # Parse through the primer df\n",
        "    primer_df = create_primer_df_from_dict(assembly_data)\n",
        "\n",
        "    #Find_unique primers\n",
        "    unique_df = primer_df.drop_duplicates(keep='first')\n",
        "    idt_df = create_idt_order_dataframe(unique_df, concentration=\"25nm\", purification=\"STD\")\n",
        "\n",
        "    # Contigs\n",
        "    assembled_contigs = []\n",
        "    for data in assembly_data: \n",
        "        contig_record = data['contig']\n",
        "        contig_record.id = f\"{data['name']}_w_rep\"\n",
        "        contig_record.name = f\"{data['name']}_w_rep\"\n",
        "        assembled_contigs.append(contig_record)\n",
        "\n",
        "\n",
        "# print our df that contains primers for the adding repair templates to the vecors.         \n",
        "unique_df"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "Check for restriction enzyme sites"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 21,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>plasmid_name</th>\n",
              "      <th>date</th>\n",
              "      <th>original_plasmid</th>\n",
              "      <th>integration</th>\n",
              "      <th>size</th>\n",
              "      <th>sgRNA plasmid</th>\n",
              "      <th>which workflow to proceed with</th>\n",
              "      <th>sgRNA plasmid #EcoRI sites</th>\n",
              "      <th>repair template plasmid #NcoI sites</th>\n",
              "      <th>repair template plasmid #BstBI sites</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>pCas3_SCO5087(163)_w_rep</td>\n",
              "      <td>2024-10-11</td>\n",
              "      <td>pCRISPR-Cas3</td>\n",
              "      <td>sgRNA_SCO5087(163)</td>\n",
              "      <td>14770</td>\n",
              "      <td>pCas3_SCO5087(163)</td>\n",
              "      <td>Proceed with sgRNA integration first</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>1</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>pCas3_SCO5087(306)_w_rep</td>\n",
              "      <td>2024-10-11</td>\n",
              "      <td>pCRISPR-Cas3</td>\n",
              "      <td>sgRNA_SCO5087(306)</td>\n",
              "      <td>14770</td>\n",
              "      <td>pCas3_SCO5087(306)</td>\n",
              "      <td>Proceed with sgRNA integration first</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>1</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>pCas3_SCO5088(1219)_w_rep</td>\n",
              "      <td>2024-10-11</td>\n",
              "      <td>pCRISPR-Cas3</td>\n",
              "      <td>sgRNA_SCO5088(1219)</td>\n",
              "      <td>14770</td>\n",
              "      <td>pCas3_SCO5088(1219)</td>\n",
              "      <td>Proceed with sgRNA integration first</td>\n",
              "      <td>0</td>\n",
              "      <td>2</td>\n",
              "      <td>2</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>pCas3_SCO5088(370)_w_rep</td>\n",
              "      <td>2024-10-11</td>\n",
              "      <td>pCRISPR-Cas3</td>\n",
              "      <td>sgRNA_SCO5088(370)</td>\n",
              "      <td>14770</td>\n",
              "      <td>pCas3_SCO5088(370)</td>\n",
              "      <td>Proceed with sgRNA integration first</td>\n",
              "      <td>0</td>\n",
              "      <td>2</td>\n",
              "      <td>2</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "                plasmid_name        date original_plasmid  \\\n",
              "0   pCas3_SCO5087(163)_w_rep  2024-10-11     pCRISPR-Cas3   \n",
              "1   pCas3_SCO5087(306)_w_rep  2024-10-11     pCRISPR-Cas3   \n",
              "2  pCas3_SCO5088(1219)_w_rep  2024-10-11     pCRISPR-Cas3   \n",
              "3   pCas3_SCO5088(370)_w_rep  2024-10-11     pCRISPR-Cas3   \n",
              "\n",
              "           integration   size        sgRNA plasmid  \\\n",
              "0   sgRNA_SCO5087(163)  14770   pCas3_SCO5087(163)   \n",
              "1   sgRNA_SCO5087(306)  14770   pCas3_SCO5087(306)   \n",
              "2  sgRNA_SCO5088(1219)  14770  pCas3_SCO5088(1219)   \n",
              "3   sgRNA_SCO5088(370)  14770   pCas3_SCO5088(370)   \n",
              "\n",
              "         which workflow to proceed with  sgRNA plasmid #EcoRI sites  \\\n",
              "0  Proceed with sgRNA integration first                           0   \n",
              "1  Proceed with sgRNA integration first                           0   \n",
              "2  Proceed with sgRNA integration first                           0   \n",
              "3  Proceed with sgRNA integration first                           0   \n",
              "\n",
              "   repair template plasmid #NcoI sites  repair template plasmid #BstBI sites  \n",
              "0                                    0                                     1  \n",
              "1                                    0                                     1  \n",
              "2                                    2                                     2  \n",
              "3                                    2                                     2  "
            ]
          },
          "execution_count": 21,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "if in_frame_deletion:\n",
        "    integration_names = filtered_df.apply(lambda row: f\"sgRNA_{row['locus_tag']}({row['sgrna_loc']})\", axis=1).tolist()\n",
        "    plasmid_metadata_df = extract_metadata_to_dataframe(assembled_contigs,\n",
        "                                                        clean_plasmid,\n",
        "                                                        integration_names)\n",
        "    workflow_df = determine_workflow_order_for_plasmids(assembled_cas3_plasmids, \n",
        "                                                        assembled_contigs,\n",
        "                                                        [Enzyme_for_repair_template_integration,], [ \"NcoI\",\"BstBI\" ]) \n",
        "    \n",
        "    plasmid_metadata_df = pd.merge(plasmid_metadata_df, workflow_df, on='plasmid_name', how='inner')  # or 'left'/'right' depending on what you need\n",
        "\n",
        "else: \n",
        "    integration_names = filtered_df.apply(lambda row: f\"sgRNA_{row['locus_tag']}({row['sgrna_loc']})\", axis=1).tolist()\n",
        "    plasmid_metadata_df = extract_metadata_to_dataframe(assembled_cas3_plasmids,\n",
        "                                                        clean_plasmid,\n",
        "                                                        integration_names)\n",
        "\n",
        "plasmid_metadata_df"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 22,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>locus tag</th>\n",
              "      <th>f_primer_name</th>\n",
              "      <th>r_primer_name</th>\n",
              "      <th>f_primer_sequences(5-3)</th>\n",
              "      <th>r_primer_sequences(5-3)</th>\n",
              "      <th>f_tm</th>\n",
              "      <th>r_tm</th>\n",
              "      <th>ta</th>\n",
              "      <th>flanking_region</th>\n",
              "      <th>annealing_temperature</th>\n",
              "      <th>...</th>\n",
              "      <th>homodimer_reverse_tm</th>\n",
              "      <th>homodimer_reverse_deltaG (kcal/mol)</th>\n",
              "      <th>heterodimer_tm</th>\n",
              "      <th>heterodimer_deltaG (kcal/mol)</th>\n",
              "      <th>hairpin_forward_structure_found</th>\n",
              "      <th>hairpin_forward_tm</th>\n",
              "      <th>hairpin_forward_deltaG (kcal/mol)</th>\n",
              "      <th>hairpin_reverse_structure_found</th>\n",
              "      <th>hairpin_reverse_tm</th>\n",
              "      <th>hairpin_reverse_deltaG (kcal/mol)</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>SCO5088</td>\n",
              "      <td>SCO5088_fwd_checking_primer</td>\n",
              "      <td>SCO5088_rev_checking_primer</td>\n",
              "      <td>CGCATCCACGCCGAGA</td>\n",
              "      <td>CAGCGTCCGGCGG</td>\n",
              "      <td>61</td>\n",
              "      <td>58</td>\n",
              "      <td>62</td>\n",
              "      <td>500</td>\n",
              "      <td>62</td>\n",
              "      <td>...</td>\n",
              "      <td>22.507821</td>\n",
              "      <td>-4.506029</td>\n",
              "      <td>22.023864</td>\n",
              "      <td>-4.603239</td>\n",
              "      <td>False</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>False</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>SCO5087</td>\n",
              "      <td>SCO5087_fwd_checking_primer</td>\n",
              "      <td>SCO5087_rev_checking_primer</td>\n",
              "      <td>GACGATTCGGCCCGTG</td>\n",
              "      <td>CAGGGCGTCCAGGC</td>\n",
              "      <td>59</td>\n",
              "      <td>58</td>\n",
              "      <td>61</td>\n",
              "      <td>500</td>\n",
              "      <td>61</td>\n",
              "      <td>...</td>\n",
              "      <td>9.706246</td>\n",
              "      <td>-3.415427</td>\n",
              "      <td>1.918567</td>\n",
              "      <td>-2.396641</td>\n",
              "      <td>False</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>False</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>2 rows Ã— 23 columns</p>\n",
              "</div>"
            ],
            "text/plain": [
              "  locus tag                f_primer_name                r_primer_name  \\\n",
              "0   SCO5088  SCO5088_fwd_checking_primer  SCO5088_rev_checking_primer   \n",
              "1   SCO5087  SCO5087_fwd_checking_primer  SCO5087_rev_checking_primer   \n",
              "\n",
              "  f_primer_sequences(5-3) r_primer_sequences(5-3) f_tm r_tm  ta  \\\n",
              "0        CGCATCCACGCCGAGA           CAGCGTCCGGCGG   61   58  62   \n",
              "1        GACGATTCGGCCCGTG          CAGGGCGTCCAGGC   59   58  61   \n",
              "\n",
              "  flanking_region  annealing_temperature  ... homodimer_reverse_tm  \\\n",
              "0             500                     62  ...            22.507821   \n",
              "1             500                     61  ...             9.706246   \n",
              "\n",
              "   homodimer_reverse_deltaG (kcal/mol)  heterodimer_tm  \\\n",
              "0                            -4.506029       22.023864   \n",
              "1                            -3.415427        1.918567   \n",
              "\n",
              "   heterodimer_deltaG (kcal/mol)  hairpin_forward_structure_found  \\\n",
              "0                      -4.603239                            False   \n",
              "1                      -2.396641                            False   \n",
              "\n",
              "   hairpin_forward_tm  hairpin_forward_deltaG (kcal/mol)  \\\n",
              "0                 0.0                                0.0   \n",
              "1                 0.0                                0.0   \n",
              "\n",
              "   hairpin_reverse_structure_found  hairpin_reverse_tm  \\\n",
              "0                            False                 0.0   \n",
              "1                            False                 0.0   \n",
              "\n",
              "   hairpin_reverse_deltaG (kcal/mol)  \n",
              "0                                0.0  \n",
              "1                                0.0  \n",
              "\n",
              "[2 rows x 23 columns]"
            ]
          },
          "execution_count": 22,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "# Getting checking primers\n",
        "checking_primers_df = find_best_check_primers_from_genome(genome, \n",
        "                                       genes_to_KO, \n",
        "                                       flanking_region=flanking_region,\n",
        "                                       target_tm = melting_temperature, \n",
        "                                        primer_concentration = primer_concentration, \n",
        "                                        polymerase = chosen_polymerase)\n",
        "checking_primers_df"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 23,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>Name</th>\n",
              "      <th>Sequence</th>\n",
              "      <th>Concentration</th>\n",
              "      <th>Purification</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>SCO5088_fwd_checking_primer</td>\n",
              "      <td>CGCATCCACGCCGAGA</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>SCO5087_fwd_checking_primer</td>\n",
              "      <td>GACGATTCGGCCCGTG</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>SCO5088_rev_checking_primer</td>\n",
              "      <td>CAGCGTCCGGCGG</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>SCO5087_rev_checking_primer</td>\n",
              "      <td>CAGGGCGTCCAGGC</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "                          Name          Sequence Concentration Purification\n",
              "0  SCO5088_fwd_checking_primer  CGCATCCACGCCGAGA          25nm          STD\n",
              "1  SCO5087_fwd_checking_primer  GACGATTCGGCCCGTG          25nm          STD\n",
              "2  SCO5088_rev_checking_primer     CAGCGTCCGGCGG          25nm          STD\n",
              "3  SCO5087_rev_checking_primer    CAGGGCGTCCAGGC          25nm          STD"
            ]
          },
          "execution_count": 23,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "# making the primers into dseqrecords\n",
        "checking_primers_df_idt = create_idt_order_dataframe(checking_primers_df)\n",
        "checking_primers_df_idt\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 24,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>Name</th>\n",
              "      <th>Sequence</th>\n",
              "      <th>Concentration</th>\n",
              "      <th>Purification</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>SCO5087_loc_163_fwd</td>\n",
              "      <td>GGATTGAAGCGCAGAGTCGTCATCACGGGCGTCGGTCGCCCGGCAA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>SCO5087_loc_163_rev</td>\n",
              "      <td>CGACGCCCGTGATGACGACTCTGCGCTTCAATCCGTTTCAATCCAC...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>SCO5087_loc_306_fwd</td>\n",
              "      <td>ACAGGTCGCGGCGGAGGCCGACTTCGACCCGGTCGTCGCCCGGCAA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>SCO5087_loc_306_rev</td>\n",
              "      <td>GACCGGGTCGAAGTCGGCCTCCGCCGCGACCTGTGTTTCAATCCAC...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>SCO5088_loc_1219_fwd</td>\n",
              "      <td>AACTCGGCGGCGGTCCTGCGCCGTTTCGCACCGAGTCGCCCGGCAA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>5</th>\n",
              "      <td>SCO5088_loc_1219_rev</td>\n",
              "      <td>TCGGTGCGAAACGGCGCAGGACCGCCGCCGAGTTGTTTCAATCCAC...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>6</th>\n",
              "      <td>SCO5088_loc_370_fwd</td>\n",
              "      <td>ACCCACCGGGAGTTCCGCAAGCTGTGGTCCGAGGGTCGCCCGGCAA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>7</th>\n",
              "      <td>SCO5088_loc_370_rev</td>\n",
              "      <td>CCTCGGACCACAGCTTGCGGAACTCCCGGTGGGTGTTTCAATCCAC...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>8</th>\n",
              "      <td>SCO5087_repair_up_forwar_p</td>\n",
              "      <td>AGAAGCTTTCTAGAGGATCCCCGGGTACCGAGCTCGAATTGACGCC...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>9</th>\n",
              "      <td>SCO5087_repair_dw_forwar_p</td>\n",
              "      <td>GGCGGTCGAAGGGAGATGGGGCGTCCTCATCACCGG</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>10</th>\n",
              "      <td>SCO5088_repair_up_forwar_p</td>\n",
              "      <td>AGAAGCTTTCTAGAGGATCCCCGGGTACCGAGCTCGAATTACCCGG...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>11</th>\n",
              "      <td>SCO5088_repair_dw_forwar_p</td>\n",
              "      <td>GAGACCGCGGGGGCGGCCGCCGACCCCGAGAACTGG</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>12</th>\n",
              "      <td>SCO5087_repair_up_reverse_p</td>\n",
              "      <td>GACTCCGGTGATGAGGACGCCCCATCTCCCTTCGAC</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>13</th>\n",
              "      <td>SCO5087_repair_dw_reverse_p</td>\n",
              "      <td>AACGTCTGGAAAGACGACAAAACTTTAGATCTGGGGAATTGAGTAG...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>14</th>\n",
              "      <td>SCO5088_repair_up_reverse_p</td>\n",
              "      <td>TTCTCCAGTTCTCGGGGTCGGCGGCCGCCCCCG</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>15</th>\n",
              "      <td>SCO5088_repair_dw_reverse_p</td>\n",
              "      <td>AACGTCTGGAAAGACGACAAAACTTTAGATCTGGGGAATTTACACG...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>16</th>\n",
              "      <td>SCO5088_fwd_checking_primer</td>\n",
              "      <td>CGCATCCACGCCGAGA</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>17</th>\n",
              "      <td>SCO5087_fwd_checking_primer</td>\n",
              "      <td>GACGATTCGGCCCGTG</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>18</th>\n",
              "      <td>SCO5088_rev_checking_primer</td>\n",
              "      <td>CAGCGTCCGGCGG</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>19</th>\n",
              "      <td>SCO5087_rev_checking_primer</td>\n",
              "      <td>CAGGGCGTCCAGGC</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "                           Name  \\\n",
              "0           SCO5087_loc_163_fwd   \n",
              "1           SCO5087_loc_163_rev   \n",
              "2           SCO5087_loc_306_fwd   \n",
              "3           SCO5087_loc_306_rev   \n",
              "4          SCO5088_loc_1219_fwd   \n",
              "5          SCO5088_loc_1219_rev   \n",
              "6           SCO5088_loc_370_fwd   \n",
              "7           SCO5088_loc_370_rev   \n",
              "8    SCO5087_repair_up_forwar_p   \n",
              "9    SCO5087_repair_dw_forwar_p   \n",
              "10   SCO5088_repair_up_forwar_p   \n",
              "11   SCO5088_repair_dw_forwar_p   \n",
              "12  SCO5087_repair_up_reverse_p   \n",
              "13  SCO5087_repair_dw_reverse_p   \n",
              "14  SCO5088_repair_up_reverse_p   \n",
              "15  SCO5088_repair_dw_reverse_p   \n",
              "16  SCO5088_fwd_checking_primer   \n",
              "17  SCO5087_fwd_checking_primer   \n",
              "18  SCO5088_rev_checking_primer   \n",
              "19  SCO5087_rev_checking_primer   \n",
              "\n",
              "                                             Sequence Concentration  \\\n",
              "0   GGATTGAAGCGCAGAGTCGTCATCACGGGCGTCGGTCGCCCGGCAA...          25nm   \n",
              "1   CGACGCCCGTGATGACGACTCTGCGCTTCAATCCGTTTCAATCCAC...          25nm   \n",
              "2   ACAGGTCGCGGCGGAGGCCGACTTCGACCCGGTCGTCGCCCGGCAA...          25nm   \n",
              "3   GACCGGGTCGAAGTCGGCCTCCGCCGCGACCTGTGTTTCAATCCAC...          25nm   \n",
              "4   AACTCGGCGGCGGTCCTGCGCCGTTTCGCACCGAGTCGCCCGGCAA...          25nm   \n",
              "5   TCGGTGCGAAACGGCGCAGGACCGCCGCCGAGTTGTTTCAATCCAC...          25nm   \n",
              "6   ACCCACCGGGAGTTCCGCAAGCTGTGGTCCGAGGGTCGCCCGGCAA...          25nm   \n",
              "7   CCTCGGACCACAGCTTGCGGAACTCCCGGTGGGTGTTTCAATCCAC...          25nm   \n",
              "8   AGAAGCTTTCTAGAGGATCCCCGGGTACCGAGCTCGAATTGACGCC...          25nm   \n",
              "9                GGCGGTCGAAGGGAGATGGGGCGTCCTCATCACCGG          25nm   \n",
              "10  AGAAGCTTTCTAGAGGATCCCCGGGTACCGAGCTCGAATTACCCGG...          25nm   \n",
              "11               GAGACCGCGGGGGCGGCCGCCGACCCCGAGAACTGG          25nm   \n",
              "12               GACTCCGGTGATGAGGACGCCCCATCTCCCTTCGAC          25nm   \n",
              "13  AACGTCTGGAAAGACGACAAAACTTTAGATCTGGGGAATTGAGTAG...          25nm   \n",
              "14                  TTCTCCAGTTCTCGGGGTCGGCGGCCGCCCCCG          25nm   \n",
              "15  AACGTCTGGAAAGACGACAAAACTTTAGATCTGGGGAATTTACACG...          25nm   \n",
              "16                                   CGCATCCACGCCGAGA          25nm   \n",
              "17                                   GACGATTCGGCCCGTG          25nm   \n",
              "18                                      CAGCGTCCGGCGG          25nm   \n",
              "19                                     CAGGGCGTCCAGGC          25nm   \n",
              "\n",
              "   Purification  \n",
              "0           STD  \n",
              "1           STD  \n",
              "2           STD  \n",
              "3           STD  \n",
              "4           STD  \n",
              "5           STD  \n",
              "6           STD  \n",
              "7           STD  \n",
              "8           STD  \n",
              "9           STD  \n",
              "10          STD  \n",
              "11          STD  \n",
              "12          STD  \n",
              "13          STD  \n",
              "14          STD  \n",
              "15          STD  \n",
              "16          STD  \n",
              "17          STD  \n",
              "18          STD  \n",
              "19          STD  "
            ]
          },
          "execution_count": 24,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "if in_frame_deletion: \n",
        "    full_idt = pd.concat([idt_primers_cas3,idt_df, checking_primers_df_idt],ignore_index=True)\n",
        "else: \n",
        "    full_idt = pd.concat([idt_primers_cas3,checking_primers_df_idt],ignore_index=True)\n",
        "\n",
        "full_idt"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "## Folder with all the generated I/O"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 25,
      "metadata": {},
      "outputs": [
        {
          "name": "stderr",
          "output_type": "stream",
          "text": [
            "/Users/lucaslevassor/opt/anaconda3/envs/streptocad/lib/python3.11/site-packages/Bio/SeqIO/InsdcIO.py:728: BiopythonWarning: Increasing length of locus line to allow long name. This will result in fields that are not in usual positions.\n",
            "  warnings.warn(\n"
          ]
        }
      ],
      "source": [
        "input_files = [\n",
        "    {\"name\": \"input_genome.gb\", \"content\": genome},\n",
        "    {\"name\": \"input_plasmid.gb\", \"content\": clean_plasmid}\n",
        "]\n",
        "\n",
        "\n",
        "if in_frame_deletion: \n",
        "    output_files = [\n",
        "        {\"name\": \"BEST_w_sgRNAs.gb\", \"content\": assembled_contigs}, # LIST OF Dseqrecords\n",
        "        {\"name\": \"primer_df.csv\", \"content\": primer_df},\n",
        "        {\"name\": \"full_idt.csv\", \"content\": full_idt},\n",
        "        {\"name\": \"sgrna_df.csv\", \"content\": sgrna_df},\n",
        "        {\"name\": \"filtered_sgrna_df.csv\", \"content\": filtered_df},\n",
        "        {\"name\": \"plasmid_metadata_df.csv\", \"content\": plasmid_metadata_df},\n",
        "\n",
        "    ]\n",
        "else: \n",
        "    output_files = [\n",
        "        {\"name\": \"cBEST_w_sgRNAs.gb\", \"content\": assembled_cas3_plasmids}, # LIST OF Dseqrecords\n",
        "        {\"name\": \"full_idt.csv\", \"content\": full_idt},\n",
        "        {\"name\": \"sgrna_df.csv\", \"content\": sgrna_df},\n",
        "        {\"name\": \"filtered_df.csv\", \"content\": filtered_df},\n",
        "        {\"name\": \"plasmid_metadata_df.csv\", \"content\": plasmid_metadata_df},\n",
        "        {\"name\": \"workflow_order_df.csv\", \"content\": workflow_df},\n",
        "\n",
        "    ]\n",
        "         \n",
        "input_values = {\n",
        "    \"genes_to_knockout\": genes_to_KO,\n",
        "\n",
        "    \"filtering_metrics\": {\n",
        "        \"gc_upper\": gc_upper,\n",
        "        \"gc_lower\": gc_lower,\n",
        "        \"off_target_seed\": off_target_seed,\n",
        "        \"off_target_upper\": off_target_upper,\n",
        "        \"cas_type\": cas_type,\n",
        "        \"number_of_sgRNAs_per_group\": number_of_sgRNAs_per_group,\n",
        "\n",
        "    },  \n",
        "    \"polymerase_settings\": {\n",
        "        \"chosen_polymerase\": chosen_polymerase,\n",
        "        \"melting_temperature\": melting_temperature,\n",
        "        \"primer_concentration\": primer_concentration,\n",
        "        \"primer_number_increment\": primer_number_increment,\n",
        "        \"flanking_region\": flanking_region\n",
        "    },\n",
        "    \"overlapping_sequences\": {\n",
        "        \"up_homology\": str(forward_backbone_overhang),\n",
        "        \"dw_homology\": str(reverse_backbone_overhang)\n",
        "    }\n",
        "}\n",
        "\n",
        "# Paths to Markdown files\n",
        "markdown_file_paths = [\n",
        "    \"../../protocols/conjugation_protcol.md\",\n",
        "    \"../../protocols/single_target_crispr_plasmid_protcol.md\"\n",
        "\n",
        "]\n",
        "\n",
        "timestamp = datetime.utcnow().isoformat()\n",
        "\n",
        "# Create project directory structure\n",
        "project_directory = ProjectDirectory(\n",
        "    project_name=f\"CRISPR_cas3_inframe_deletion_workflow_{timestamp}\",\n",
        "    input_files=input_files,\n",
        "    output_files=output_files,\n",
        "    input_values=input_values,\n",
        "    markdown_file_paths=markdown_file_paths\n",
        ")\n",
        "\n",
        "# DO You want to save the folder? \n",
        "save_zip_folder = True \n",
        "\n",
        "if save_zip_folder: \n",
        "    # Generate the project directory structure and get the zip content\n",
        "    zip_content = project_directory.create_directory_structure(create_directories=False)\n",
        "\n",
        "    # Save the zip file to disk (optional)\n",
        "    with open(\"project_structure.zip\", \"wb\") as f:\n",
        "        f.write(zip_content)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {},
      "outputs": [],
      "source": []
    }
  ],
  "metadata": {
    "kernelspec": {
      "display_name": "constrain",
      "language": "python",
      "name": "python3"
    },
    "language_info": {
      "codemirror_mode": {
        "name": "ipython",
        "version": 3
      },
      "file_extension": ".py",
      "mimetype": "text/x-python",
      "name": "python",
      "nbconvert_exporter": "python",
      "pygments_lexer": "ipython3",
      "version": "3.11.9"
    }
  },
  "nbformat": 4,
  "nbformat_minor": 2
}
