{
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "# W6-CRISPR-Cas3-in-frame-deletion"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "In Workflow 6, CRISPR-Cas3 is paired with a PCR and Gibson protocol to achieve in-frame deletions22(Figure 7A). Just like in workflow 5, this workflow can be used to knock out specific genes in streptomycetes but with the crucial difference of using Cas3 which has been shown to have higher efficiencies for targeted gene deletions in for example BGCs and random-sized deletion in streptomyces22. Due to homologies, and high GC content, this protocol relies on two steps of Gibson cloning to incorporate the sgRNA and the upstream and downstream repair regions. To get started, users can download the pCRISPR-Cas3 file and the genome file (S. coelicolor A3). StreptoAIM will then generate the necessary primers, plasmid files, and checking primers to verify full gene deletions (S5). \n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 24,
      "metadata": {},
      "outputs": [],
      "source": [
        "import sys\n",
        "import os\n",
        "from Bio.Restriction import StuI\n",
        "import pandas as pd\n",
        "from datetime import datetime\n",
        "\n",
        "# Ensure the src directory is in the Python path\n",
        "project_root = os.path.abspath(os.path.join(os.getcwd(), '../../'))\n",
        "if project_root not in sys.path:\n",
        "    sys.path.insert(0, project_root)\n",
        "\n",
        "\n",
        "from pydna.dseqrecord import Dseqrecord\n",
        "\n",
        "from streptocad.sequence_loading.sequence_loading import (\n",
        "    load_and_process_plasmid,\n",
        "    load_and_process_genome_sequences, \n",
        "    check_and_convert_input, \n",
        "    annotate_dseqrecord\n",
        "    \n",
        ")\n",
        "\n",
        "from streptocad.utils import polymerase_dict, create_primer_df_from_dict,ProjectDirectory,extract_metadata_to_dataframe\n",
        "from streptocad.primers.primer_generation import create_idt_order_dataframe, make_primer_records, primers_to_IDT, find_best_check_primers_from_genome\n",
        "from streptocad.crispr.guideRNAcas3_9_12 import extract_sgRNAs, SgRNAargs\n",
        "from streptocad.cloning.cas3_plasmid_cloning import generate_cas3_primers, cas3_plasmid_pcrs, assemble_cas3_plasmids\n",
        "from streptocad.cloning.gibson_cloning import (\n",
        "    find_up_dw_repair_templates,\n",
        "    assemble_multiple_plasmids_with_repair_templates_for_deletion,\n",
        "    update_primer_names\n",
        ")"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "## INPUT"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 25,
      "metadata": {},
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "pCRISPR-Cas3\n"
          ]
        }
      ],
      "source": [
        "# Inputs\n",
        "# 1 Add genome of choice (genbank, fasta)\n",
        "path_to_genome = '../../data/genomes/Streptomyces_coelicolor_A3_chromosome.gb'\n",
        "genome = load_and_process_genome_sequences(path_to_genome)[0]\n",
        "\n",
        "# 2 Add plasmid \n",
        "path_to_plasmid = '../../data/plasmids/pCRISPR_Cas3.gbk'\n",
        "clean_plasmid = load_and_process_plasmid(path_to_plasmid)\n",
        "print(clean_plasmid.name)\n",
        "\n",
        "# 3 Choose genes to knock out (list)\n",
        "genes_to_KO = ['SCO5087', 'SCO5089', 'SCO5090']\n",
        "genes_to_KO = ['80000-100000', '4000-7000', '9000-14000','15000-20000']\n",
        "\n",
        "#### Advanced settings ####\n",
        "# 4 Filtering metrics for sgRNAs\n",
        "gc_upper = 0.8\n",
        "gc_lower = 0.2\n",
        "off_target_seed = 8\n",
        "off_target_upper = 10\n",
        "cas_type='cas3'\n",
        "number_of_sgRNAs_per_group = 2\n",
        "\n",
        "# 5 Choose polymerase and target melting temperature\n",
        "chosen_polymerase = polymerase_dict['Phusion High-Fidelity DNA Polymerase (GC Buffer)']\n",
        "melting_temperature = 65\n",
        "primer_concentration = 0.4 \n",
        "primer_number_increment = 23\n",
        "flanking_region = 500\n",
        "\n",
        "# 6 Choose overlapping sequences for our plasmid we can use the following\n",
        "#As per the article **\"CRISPRâ€“Cas9, CRISPRi and CRISPR-BEST-mediated genetic manipulation in streptomycetes\"** we need the following oligoes: \n",
        "#CGGTTGGTAGGATCGACGGC **-N20-** GTTTTAGAGCTAGAAATAGC\n",
        "up_homology = Dseqrecord('CGGTTGGTAGGATCGACGGC')\n",
        "dw_homology = Dseqrecord('GTTTTAGAGCTAGAAATAGC')"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 26,
      "metadata": {},
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "pCRISPR-Cas3\n"
          ]
        }
      ],
      "source": [
        "print(clean_plasmid.id)\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "# Computation"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 4,
      "metadata": {},
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "True\n",
            "this is the target dict [{'region_1_(80000-100000)': [80000, 100000]}, {'region_2_(4000-7000)': [4000, 7000]}, {'region_3_(9000-14000)': [9000, 14000]}, {'region_4_(15000-20000)': [15000, 20000]}]\n",
            "region_1_(80000-100000) [80000, 100000]\n",
            "region_2_(4000-7000) [4000, 7000]\n",
            "region_3_(9000-14000) [9000, 14000]\n",
            "region_4_(15000-20000) [15000, 20000]\n"
          ]
        },
        {
          "data": {
            "text/plain": [
              "25828"
            ]
          },
          "execution_count": 4,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "target_dict, genes_to_KO, annotation_input = check_and_convert_input(genes_to_KO)\n",
        "\n",
        "print(annotation_input)\n",
        "if annotation_input == True:\n",
        "    genome = annotate_dseqrecord(genome, target_dict)\n",
        "\n",
        "\n",
        "len(genome.features)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 5,
      "metadata": {},
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "sgRNA generated were too smallregion_1_(80000-100000). To incorporate this extent borders. Skipping to next locus tag.\n"
          ]
        },
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>strain_name</th>\n",
              "      <th>locus_tag</th>\n",
              "      <th>gene_loc</th>\n",
              "      <th>gene_strand</th>\n",
              "      <th>sgrna_strand</th>\n",
              "      <th>sgrna_loc</th>\n",
              "      <th>gc</th>\n",
              "      <th>pam</th>\n",
              "      <th>sgrna</th>\n",
              "      <th>sgrna_seed_sequence</th>\n",
              "      <th>off_target_count</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>356</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>region_1_(80000-100000)</td>\n",
              "      <td>80001</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>9290</td>\n",
              "      <td>0.705882</td>\n",
              "      <td>TTC</td>\n",
              "      <td>TCCCATGTGCCGTCGACCGCCCACATCCGCAGCC</td>\n",
              "      <td>TCCCATGT</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>364</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>region_1_(80000-100000)</td>\n",
              "      <td>80001</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>8869</td>\n",
              "      <td>0.617647</td>\n",
              "      <td>TTC</td>\n",
              "      <td>GTGTTAGCTAAGGCCTCCGCCCGCGAAGGAAAGG</td>\n",
              "      <td>GTGTTAGC</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>63</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>region_1_(80000-100000)</td>\n",
              "      <td>80001</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>5345</td>\n",
              "      <td>0.676471</td>\n",
              "      <td>TTC</td>\n",
              "      <td>GGCAATCGAACAGGGCTGGAGCGGACGGTGACGG</td>\n",
              "      <td>GGCAATCG</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>142</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>region_1_(80000-100000)</td>\n",
              "      <td>80001</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>11939</td>\n",
              "      <td>0.558824</td>\n",
              "      <td>TTC</td>\n",
              "      <td>CTCAGGACTCTTTCGGGACGTCTTGCACGACGTT</td>\n",
              "      <td>CTCAGGAC</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>146</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>region_1_(80000-100000)</td>\n",
              "      <td>80001</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>12029</td>\n",
              "      <td>0.617647</td>\n",
              "      <td>TTC</td>\n",
              "      <td>ACAACAAGGTCGCGGTCGTGACGGGTGGAAGTGC</td>\n",
              "      <td>ACAACAAG</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>...</th>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>426</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>region_1_(80000-100000)</td>\n",
              "      <td>80001</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>3753</td>\n",
              "      <td>0.588235</td>\n",
              "      <td>TTC</td>\n",
              "      <td>CGGAGCGCTCGCTGAGTTATATACGGATCTGCGC</td>\n",
              "      <td>CGGAGCGC</td>\n",
              "      <td>10</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>689</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>region_4_(15000-20000)</td>\n",
              "      <td>15001</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>3850</td>\n",
              "      <td>0.529412</td>\n",
              "      <td>TTC</td>\n",
              "      <td>ACCGAGTTCTTCGCCTTCGTCGACAATGACACCA</td>\n",
              "      <td>ACCGAGTT</td>\n",
              "      <td>10</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>49</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>region_1_(80000-100000)</td>\n",
              "      <td>80001</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>4477</td>\n",
              "      <td>0.647059</td>\n",
              "      <td>TTC</td>\n",
              "      <td>GAGGTGCATTTCACTGCGATTGGCGAGCCGCCCC</td>\n",
              "      <td>GAGGTGCA</td>\n",
              "      <td>10</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>239</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>region_1_(80000-100000)</td>\n",
              "      <td>80001</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>19760</td>\n",
              "      <td>0.647059</td>\n",
              "      <td>TTC</td>\n",
              "      <td>GGCTACTGTCCGCAGACCGTGGTGCTGGACGACT</td>\n",
              "      <td>GGCTACTG</td>\n",
              "      <td>10</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>263</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>region_1_(80000-100000)</td>\n",
              "      <td>80001</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>18615</td>\n",
              "      <td>0.705882</td>\n",
              "      <td>TTC</td>\n",
              "      <td>CTGGTCTTCCGCCGACGTACCCGGAACGCGCTCG</td>\n",
              "      <td>CTGGTCTT</td>\n",
              "      <td>10</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>463 rows Ã— 11 columns</p>\n",
              "</div>"
            ],
            "text/plain": [
              "     strain_name                locus_tag  gene_loc  gene_strand  \\\n",
              "356  NC_003888.3  region_1_(80000-100000)     80001            1   \n",
              "364  NC_003888.3  region_1_(80000-100000)     80001            1   \n",
              "63   NC_003888.3  region_1_(80000-100000)     80001            1   \n",
              "142  NC_003888.3  region_1_(80000-100000)     80001            1   \n",
              "146  NC_003888.3  region_1_(80000-100000)     80001            1   \n",
              "..           ...                      ...       ...          ...   \n",
              "426  NC_003888.3  region_1_(80000-100000)     80001            1   \n",
              "689  NC_003888.3   region_4_(15000-20000)     15001            1   \n",
              "49   NC_003888.3  region_1_(80000-100000)     80001            1   \n",
              "239  NC_003888.3  region_1_(80000-100000)     80001            1   \n",
              "263  NC_003888.3  region_1_(80000-100000)     80001            1   \n",
              "\n",
              "     sgrna_strand  sgrna_loc        gc  pam  \\\n",
              "356             1       9290  0.705882  TTC   \n",
              "364             1       8869  0.617647  TTC   \n",
              "63             -1       5345  0.676471  TTC   \n",
              "142            -1      11939  0.558824  TTC   \n",
              "146            -1      12029  0.617647  TTC   \n",
              "..            ...        ...       ...  ...   \n",
              "426             1       3753  0.588235  TTC   \n",
              "689             1       3850  0.529412  TTC   \n",
              "49             -1       4477  0.647059  TTC   \n",
              "239             1      19760  0.647059  TTC   \n",
              "263             1      18615  0.705882  TTC   \n",
              "\n",
              "                                  sgrna sgrna_seed_sequence  off_target_count  \n",
              "356  TCCCATGTGCCGTCGACCGCCCACATCCGCAGCC            TCCCATGT                 0  \n",
              "364  GTGTTAGCTAAGGCCTCCGCCCGCGAAGGAAAGG            GTGTTAGC                 0  \n",
              "63   GGCAATCGAACAGGGCTGGAGCGGACGGTGACGG            GGCAATCG                 0  \n",
              "142  CTCAGGACTCTTTCGGGACGTCTTGCACGACGTT            CTCAGGAC                 0  \n",
              "146  ACAACAAGGTCGCGGTCGTGACGGGTGGAAGTGC            ACAACAAG                 0  \n",
              "..                                  ...                 ...               ...  \n",
              "426  CGGAGCGCTCGCTGAGTTATATACGGATCTGCGC            CGGAGCGC                10  \n",
              "689  ACCGAGTTCTTCGCCTTCGTCGACAATGACACCA            ACCGAGTT                10  \n",
              "49   GAGGTGCATTTCACTGCGATTGGCGAGCCGCCCC            GAGGTGCA                10  \n",
              "239  GGCTACTGTCCGCAGACCGTGGTGCTGGACGACT            GGCTACTG                10  \n",
              "263  CTGGTCTTCCGCCGACGTACCCGGAACGCGCTCG            CTGGTCTT                10  \n",
              "\n",
              "[463 rows x 11 columns]"
            ]
          },
          "execution_count": 5,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "\n",
        "# Initialize SgRNAargs with desired parameters\n",
        "args = SgRNAargs(genome, \n",
        "                genes_to_KO,\n",
        "                step=['find', 'filter'],\n",
        "                gc_upper = gc_upper,\n",
        "                gc_lower = gc_lower,\n",
        "                off_target_seed = off_target_seed,\n",
        "                off_target_upper = off_target_upper,\n",
        "                cas_type=cas_type\n",
        "                )\n",
        "\n",
        "sgrna_df = extract_sgRNAs(args)\n",
        "sgrna_df"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 6,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>strain_name</th>\n",
              "      <th>locus_tag</th>\n",
              "      <th>gene_loc</th>\n",
              "      <th>gene_strand</th>\n",
              "      <th>sgrna_strand</th>\n",
              "      <th>sgrna_loc</th>\n",
              "      <th>gc</th>\n",
              "      <th>pam</th>\n",
              "      <th>sgrna</th>\n",
              "      <th>sgrna_seed_sequence</th>\n",
              "      <th>off_target_count</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>356</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>region_1_(80000-100000)</td>\n",
              "      <td>80001</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>9290</td>\n",
              "      <td>0.705882</td>\n",
              "      <td>TTC</td>\n",
              "      <td>TCCCATGTGCCGTCGACCGCCCACATCCGCAGCC</td>\n",
              "      <td>TCCCATGT</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>364</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>region_1_(80000-100000)</td>\n",
              "      <td>80001</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>8869</td>\n",
              "      <td>0.617647</td>\n",
              "      <td>TTC</td>\n",
              "      <td>GTGTTAGCTAAGGCCTCCGCCCGCGAAGGAAAGG</td>\n",
              "      <td>GTGTTAGC</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>630</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>region_3_(9000-14000)</td>\n",
              "      <td>9001</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>305</td>\n",
              "      <td>0.647059</td>\n",
              "      <td>TTC</td>\n",
              "      <td>ACCATTGCTCGGCTTCCGCTTCGGCGTCCCACAG</td>\n",
              "      <td>ACCATTGC</td>\n",
              "      <td>1</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>643</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>region_4_(15000-20000)</td>\n",
              "      <td>15001</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>1743</td>\n",
              "      <td>0.676471</td>\n",
              "      <td>TTC</td>\n",
              "      <td>TGCAAGTAGCCGAACACGGCCAGGCCGACGGCAG</td>\n",
              "      <td>TGCAAGTA</td>\n",
              "      <td>1</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>684</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>region_4_(15000-20000)</td>\n",
              "      <td>15001</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>4374</td>\n",
              "      <td>0.588235</td>\n",
              "      <td>TTC</td>\n",
              "      <td>CCTCGTTCTGAATGGAGCTCGCCATGGGTCAGTC</td>\n",
              "      <td>CCTCGTTC</td>\n",
              "      <td>1</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>611</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>region_3_(9000-14000)</td>\n",
              "      <td>9001</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>2495</td>\n",
              "      <td>0.647059</td>\n",
              "      <td>TTC</td>\n",
              "      <td>GGTTTGGTTGTTCGAGGGCTTGGCCTGCGGTCCG</td>\n",
              "      <td>GGTTTGGT</td>\n",
              "      <td>1</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>512</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>region_2_(4000-7000)</td>\n",
              "      <td>4001</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>2196</td>\n",
              "      <td>0.735294</td>\n",
              "      <td>TTC</td>\n",
              "      <td>CTGCAGGAGCCGCTGCTGGCCGGGCAACAGGGTG</td>\n",
              "      <td>CTGCAGGA</td>\n",
              "      <td>1</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>486</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>region_2_(4000-7000)</td>\n",
              "      <td>4001</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>1244</td>\n",
              "      <td>0.617647</td>\n",
              "      <td>TTC</td>\n",
              "      <td>GAAAGCATGGCCTCGTCGACGCCACGGCAGGATT</td>\n",
              "      <td>GAAAGCAT</td>\n",
              "      <td>1</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "     strain_name                locus_tag  gene_loc  gene_strand  \\\n",
              "356  NC_003888.3  region_1_(80000-100000)     80001            1   \n",
              "364  NC_003888.3  region_1_(80000-100000)     80001            1   \n",
              "630  NC_003888.3    region_3_(9000-14000)      9001            1   \n",
              "643  NC_003888.3   region_4_(15000-20000)     15001            1   \n",
              "684  NC_003888.3   region_4_(15000-20000)     15001            1   \n",
              "611  NC_003888.3    region_3_(9000-14000)      9001            1   \n",
              "512  NC_003888.3     region_2_(4000-7000)      4001            1   \n",
              "486  NC_003888.3     region_2_(4000-7000)      4001            1   \n",
              "\n",
              "     sgrna_strand  sgrna_loc        gc  pam  \\\n",
              "356             1       9290  0.705882  TTC   \n",
              "364             1       8869  0.617647  TTC   \n",
              "630             1        305  0.647059  TTC   \n",
              "643            -1       1743  0.676471  TTC   \n",
              "684             1       4374  0.588235  TTC   \n",
              "611             1       2495  0.647059  TTC   \n",
              "512             1       2196  0.735294  TTC   \n",
              "486            -1       1244  0.617647  TTC   \n",
              "\n",
              "                                  sgrna sgrna_seed_sequence  off_target_count  \n",
              "356  TCCCATGTGCCGTCGACCGCCCACATCCGCAGCC            TCCCATGT                 0  \n",
              "364  GTGTTAGCTAAGGCCTCCGCCCGCGAAGGAAAGG            GTGTTAGC                 0  \n",
              "630  ACCATTGCTCGGCTTCCGCTTCGGCGTCCCACAG            ACCATTGC                 1  \n",
              "643  TGCAAGTAGCCGAACACGGCCAGGCCGACGGCAG            TGCAAGTA                 1  \n",
              "684  CCTCGTTCTGAATGGAGCTCGCCATGGGTCAGTC            CCTCGTTC                 1  \n",
              "611  GGTTTGGTTGTTCGAGGGCTTGGCCTGCGGTCCG            GGTTTGGT                 1  \n",
              "512  CTGCAGGAGCCGCTGCTGGCCGGGCAACAGGGTG            CTGCAGGA                 1  \n",
              "486  GAAAGCATGGCCTCGTCGACGCCACGGCAGGATT            GAAAGCAT                 1  "
            ]
          },
          "execution_count": 6,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "# Filter the DataFrame to retain only up to 5 sgRNA sequences per locus_tag\n",
        "filtered_df = sgrna_df.groupby('locus_tag').head(number_of_sgRNAs_per_group)\n",
        "filtered_df"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "## Output"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 7,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>strain_name</th>\n",
              "      <th>locus_tag</th>\n",
              "      <th>gene_loc</th>\n",
              "      <th>gene_strand</th>\n",
              "      <th>sgrna_strand</th>\n",
              "      <th>sgrna_loc</th>\n",
              "      <th>gc</th>\n",
              "      <th>pam</th>\n",
              "      <th>sgrna</th>\n",
              "      <th>sgrna_seed_sequence</th>\n",
              "      <th>off_target_count</th>\n",
              "      <th>Fwd Primer</th>\n",
              "      <th>Rev Primer</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>356</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>region_1_(80000-100000)</td>\n",
              "      <td>80001</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>9290</td>\n",
              "      <td>0.705882</td>\n",
              "      <td>TTC</td>\n",
              "      <td>TCCCATGTGCCGTCGACCGCCCACATCCGCAGCC</td>\n",
              "      <td>TCCCATGT</td>\n",
              "      <td>0</td>\n",
              "      <td>TCCCATGTGCCGTCGACCGCCCACATCCGCAGCCGTCGCcCggCaa...</td>\n",
              "      <td>GGCTGCGGATGTGGGCGGTCGACGGCACATGGGAGTTTCAATCCAC...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>364</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>region_1_(80000-100000)</td>\n",
              "      <td>80001</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>8869</td>\n",
              "      <td>0.617647</td>\n",
              "      <td>TTC</td>\n",
              "      <td>GTGTTAGCTAAGGCCTCCGCCCGCGAAGGAAAGG</td>\n",
              "      <td>GTGTTAGC</td>\n",
              "      <td>0</td>\n",
              "      <td>GTGTTAGCTAAGGCCTCCGCCCGCGAAGGAAAGGGTCGCcCggCaa...</td>\n",
              "      <td>CCTTTCCTTCGCGGGCGGAGGCCTTAGCTAACACGTTTCAATCCAC...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>630</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>region_3_(9000-14000)</td>\n",
              "      <td>9001</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>305</td>\n",
              "      <td>0.647059</td>\n",
              "      <td>TTC</td>\n",
              "      <td>ACCATTGCTCGGCTTCCGCTTCGGCGTCCCACAG</td>\n",
              "      <td>ACCATTGC</td>\n",
              "      <td>1</td>\n",
              "      <td>ACCATTGCTCGGCTTCCGCTTCGGCGTCCCACAGGTCGCcCggCaa...</td>\n",
              "      <td>CTGTGGGACGCCGAAGCGGAAGCCGAGCAATGGTGTTTCAATCCAC...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>643</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>region_4_(15000-20000)</td>\n",
              "      <td>15001</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>1743</td>\n",
              "      <td>0.676471</td>\n",
              "      <td>TTC</td>\n",
              "      <td>TGCAAGTAGCCGAACACGGCCAGGCCGACGGCAG</td>\n",
              "      <td>TGCAAGTA</td>\n",
              "      <td>1</td>\n",
              "      <td>TGCAAGTAGCCGAACACGGCCAGGCCGACGGCAGGTCGCcCggCaa...</td>\n",
              "      <td>CTGCCGTCGGCCTGGCCGTGTTCGGCTACTTGCAGTTTCAATCCAC...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>684</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>region_4_(15000-20000)</td>\n",
              "      <td>15001</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>4374</td>\n",
              "      <td>0.588235</td>\n",
              "      <td>TTC</td>\n",
              "      <td>CCTCGTTCTGAATGGAGCTCGCCATGGGTCAGTC</td>\n",
              "      <td>CCTCGTTC</td>\n",
              "      <td>1</td>\n",
              "      <td>CCTCGTTCTGAATGGAGCTCGCCATGGGTCAGTCGTCGCcCggCaa...</td>\n",
              "      <td>GACTGACCCATGGCGAGCTCCATTCAGAACGAGGGTTTCAATCCAC...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>611</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>region_3_(9000-14000)</td>\n",
              "      <td>9001</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>2495</td>\n",
              "      <td>0.647059</td>\n",
              "      <td>TTC</td>\n",
              "      <td>GGTTTGGTTGTTCGAGGGCTTGGCCTGCGGTCCG</td>\n",
              "      <td>GGTTTGGT</td>\n",
              "      <td>1</td>\n",
              "      <td>GGTTTGGTTGTTCGAGGGCTTGGCCTGCGGTCCGGTCGCcCggCaa...</td>\n",
              "      <td>CGGACCGCAGGCCAAGCCCTCGAACAACCAAACCGTTTCAATCCAC...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>512</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>region_2_(4000-7000)</td>\n",
              "      <td>4001</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>2196</td>\n",
              "      <td>0.735294</td>\n",
              "      <td>TTC</td>\n",
              "      <td>CTGCAGGAGCCGCTGCTGGCCGGGCAACAGGGTG</td>\n",
              "      <td>CTGCAGGA</td>\n",
              "      <td>1</td>\n",
              "      <td>CTGCAGGAGCCGCTGCTGGCCGGGCAACAGGGTGGTCGCcCggCaa...</td>\n",
              "      <td>CACCCTGTTGCCCGGCCAGCAGCGGCTCCTGCAGGTTTCAATCCAC...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>486</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>region_2_(4000-7000)</td>\n",
              "      <td>4001</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>1244</td>\n",
              "      <td>0.617647</td>\n",
              "      <td>TTC</td>\n",
              "      <td>GAAAGCATGGCCTCGTCGACGCCACGGCAGGATT</td>\n",
              "      <td>GAAAGCAT</td>\n",
              "      <td>1</td>\n",
              "      <td>GAAAGCATGGCCTCGTCGACGCCACGGCAGGATTGTCGCcCggCaa...</td>\n",
              "      <td>AATCCTGCCGTGGCGTCGACGAGGCCATGCTTTCGTTTCAATCCAC...</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "     strain_name                locus_tag  gene_loc  gene_strand  \\\n",
              "356  NC_003888.3  region_1_(80000-100000)     80001            1   \n",
              "364  NC_003888.3  region_1_(80000-100000)     80001            1   \n",
              "630  NC_003888.3    region_3_(9000-14000)      9001            1   \n",
              "643  NC_003888.3   region_4_(15000-20000)     15001            1   \n",
              "684  NC_003888.3   region_4_(15000-20000)     15001            1   \n",
              "611  NC_003888.3    region_3_(9000-14000)      9001            1   \n",
              "512  NC_003888.3     region_2_(4000-7000)      4001            1   \n",
              "486  NC_003888.3     region_2_(4000-7000)      4001            1   \n",
              "\n",
              "     sgrna_strand  sgrna_loc        gc  pam  \\\n",
              "356             1       9290  0.705882  TTC   \n",
              "364             1       8869  0.617647  TTC   \n",
              "630             1        305  0.647059  TTC   \n",
              "643            -1       1743  0.676471  TTC   \n",
              "684             1       4374  0.588235  TTC   \n",
              "611             1       2495  0.647059  TTC   \n",
              "512             1       2196  0.735294  TTC   \n",
              "486            -1       1244  0.617647  TTC   \n",
              "\n",
              "                                  sgrna sgrna_seed_sequence  off_target_count  \\\n",
              "356  TCCCATGTGCCGTCGACCGCCCACATCCGCAGCC            TCCCATGT                 0   \n",
              "364  GTGTTAGCTAAGGCCTCCGCCCGCGAAGGAAAGG            GTGTTAGC                 0   \n",
              "630  ACCATTGCTCGGCTTCCGCTTCGGCGTCCCACAG            ACCATTGC                 1   \n",
              "643  TGCAAGTAGCCGAACACGGCCAGGCCGACGGCAG            TGCAAGTA                 1   \n",
              "684  CCTCGTTCTGAATGGAGCTCGCCATGGGTCAGTC            CCTCGTTC                 1   \n",
              "611  GGTTTGGTTGTTCGAGGGCTTGGCCTGCGGTCCG            GGTTTGGT                 1   \n",
              "512  CTGCAGGAGCCGCTGCTGGCCGGGCAACAGGGTG            CTGCAGGA                 1   \n",
              "486  GAAAGCATGGCCTCGTCGACGCCACGGCAGGATT            GAAAGCAT                 1   \n",
              "\n",
              "                                            Fwd Primer  \\\n",
              "356  TCCCATGTGCCGTCGACCGCCCACATCCGCAGCCGTCGCcCggCaa...   \n",
              "364  GTGTTAGCTAAGGCCTCCGCCCGCGAAGGAAAGGGTCGCcCggCaa...   \n",
              "630  ACCATTGCTCGGCTTCCGCTTCGGCGTCCCACAGGTCGCcCggCaa...   \n",
              "643  TGCAAGTAGCCGAACACGGCCAGGCCGACGGCAGGTCGCcCggCaa...   \n",
              "684  CCTCGTTCTGAATGGAGCTCGCCATGGGTCAGTCGTCGCcCggCaa...   \n",
              "611  GGTTTGGTTGTTCGAGGGCTTGGCCTGCGGTCCGGTCGCcCggCaa...   \n",
              "512  CTGCAGGAGCCGCTGCTGGCCGGGCAACAGGGTGGTCGCcCggCaa...   \n",
              "486  GAAAGCATGGCCTCGTCGACGCCACGGCAGGATTGTCGCcCggCaa...   \n",
              "\n",
              "                                            Rev Primer  \n",
              "356  GGCTGCGGATGTGGGCGGTCGACGGCACATGGGAGTTTCAATCCAC...  \n",
              "364  CCTTTCCTTCGCGGGCGGAGGCCTTAGCTAACACGTTTCAATCCAC...  \n",
              "630  CTGTGGGACGCCGAAGCGGAAGCCGAGCAATGGTGTTTCAATCCAC...  \n",
              "643  CTGCCGTCGGCCTGGCCGTGTTCGGCTACTTGCAGTTTCAATCCAC...  \n",
              "684  GACTGACCCATGGCGAGCTCCATTCAGAACGAGGGTTTCAATCCAC...  \n",
              "611  CGGACCGCAGGCCAAGCCCTCGAACAACCAAACCGTTTCAATCCAC...  \n",
              "512  CACCCTGTTGCCCGGCCAGCAGCGGCTCCTGCAGGTTTCAATCCAC...  \n",
              "486  AATCCTGCCGTGGCGTCGACGAGGCCATGCTTTCGTTTCAATCCAC...  "
            ]
          },
          "execution_count": 7,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "filtered_df_w_primers = generate_cas3_primers(filtered_df)\n",
        "filtered_df_w_primers"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 8,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/plain": [
              "[[Amplicon(1489), Amplicon(460)],\n",
              " [Amplicon(1489), Amplicon(460)],\n",
              " [Amplicon(1489), Amplicon(460)],\n",
              " [Amplicon(1489), Amplicon(460)],\n",
              " [Amplicon(1489), Amplicon(460)],\n",
              " [Amplicon(1489), Amplicon(460)],\n",
              " [Amplicon(1489), Amplicon(460)],\n",
              " [Amplicon(1489), Amplicon(460)]]"
            ]
          },
          "execution_count": 8,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "amplicons = cas3_plasmid_pcrs(clean_plasmid, filtered_df)\n",
        "\n",
        "amplicons"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 9,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/plain": [
              "[Contig(o12766),\n",
              " Contig(o12766),\n",
              " Contig(o12766),\n",
              " Contig(o12766),\n",
              " Contig(o12766),\n",
              " Contig(o12766),\n",
              " Contig(o12766),\n",
              " Contig(o12766)]"
            ]
          },
          "execution_count": 9,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "# cut plasmid\n",
        "assembled_cas3_plasmids = assemble_cas3_plasmids(clean_plasmid, amplicons)\n",
        "assembled_cas3_plasmids"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 10,
      "metadata": {},
      "outputs": [],
      "source": [
        "\n",
        "# Constructing a meaningful name, ID, and description for the assembled plasmid using user input\n",
        "targeting_info = []\n",
        "for index, row in filtered_df.iterrows():\n",
        "    formatted_str = f\"pCas3_{row['locus_tag']}({row['sgrna_loc']})\"\n",
        "    targeting_info.append(formatted_str)\n",
        "\n",
        "for i in range(len(assembled_cas3_plasmids)):\n",
        "    assembled_cas3_plasmids[i].name = f'{targeting_info[i]}'\n",
        "    assembled_cas3_plasmids[i].id = assembled_cas3_plasmids[i].name  # Using the same value for ID as for name for simplicity\n",
        "    assembled_cas3_plasmids[i].description = f'Assembled plasmid targeting {\", \".join(genes_to_KO)} for single gene knockout, assembled using StreptoCAD.'\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 11,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/plain": [
              "'pCas3_region_1_(80000-100000)(9290)'"
            ]
          },
          "execution_count": 11,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "assembled_cas3_plasmids[0].name"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 12,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/plain": [
              "'Assembled plasmid targeting region_1_(80000-100000), region_2_(4000-7000), region_3_(9000-14000), region_4_(15000-20000) for single gene knockout, assembled using StreptoCAD.'"
            ]
          },
          "execution_count": 12,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "assembled_cas3_plasmids[0].description"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 13,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/plain": [
              "'pCas3_region_1_(80000-100000)(9290)'"
            ]
          },
          "execution_count": 13,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "assembled_cas3_plasmids[0].id"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 14,
      "metadata": {},
      "outputs": [],
      "source": [
        "print_plasmids = False\n",
        "\n",
        "if print_plasmids: \n",
        "    for vector in assembled_cas3_plasmids: \n",
        "        vector.write(f\"../../data/plasmids/sgRNA_plasmids_pCRISPRâ€“Cas3/{vector.id}.gb\")"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "### IDT primers"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 15,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/plain": [
              "[Dseqrecord(-52),\n",
              " Dseqrecord(-54),\n",
              " Dseqrecord(-52),\n",
              " Dseqrecord(-54),\n",
              " Dseqrecord(-52),\n",
              " Dseqrecord(-54),\n",
              " Dseqrecord(-52),\n",
              " Dseqrecord(-54),\n",
              " Dseqrecord(-52),\n",
              " Dseqrecord(-54),\n",
              " Dseqrecord(-52),\n",
              " Dseqrecord(-54),\n",
              " Dseqrecord(-52),\n",
              " Dseqrecord(-54),\n",
              " Dseqrecord(-52),\n",
              " Dseqrecord(-54)]"
            ]
          },
          "execution_count": 15,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "primer_records = make_primer_records(filtered_df_w_primers)\n",
        "primer_records"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 16,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>Name</th>\n",
              "      <th>Sequence</th>\n",
              "      <th>Concentration</th>\n",
              "      <th>Purification</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>region_1_(80000-100000)_loc_9290_fwd</td>\n",
              "      <td>TCCCATGTGCCGTCGACCGCCCACATCCGCAGCCGTCGCcCggCaa...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>region_1_(80000-100000)_loc_9290_rev</td>\n",
              "      <td>GGCTGCGGATGTGGGCGGTCGACGGCACATGGGAGTTTCAATCCAC...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>region_1_(80000-100000)_loc_8869_fwd</td>\n",
              "      <td>GTGTTAGCTAAGGCCTCCGCCCGCGAAGGAAAGGGTCGCcCggCaa...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>region_1_(80000-100000)_loc_8869_rev</td>\n",
              "      <td>CCTTTCCTTCGCGGGCGGAGGCCTTAGCTAACACGTTTCAATCCAC...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>region_3_(9000-14000)_loc_305_fwd</td>\n",
              "      <td>ACCATTGCTCGGCTTCCGCTTCGGCGTCCCACAGGTCGCcCggCaa...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>5</th>\n",
              "      <td>region_3_(9000-14000)_loc_305_rev</td>\n",
              "      <td>CTGTGGGACGCCGAAGCGGAAGCCGAGCAATGGTGTTTCAATCCAC...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>6</th>\n",
              "      <td>region_4_(15000-20000)_loc_1743_fwd</td>\n",
              "      <td>TGCAAGTAGCCGAACACGGCCAGGCCGACGGCAGGTCGCcCggCaa...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>7</th>\n",
              "      <td>region_4_(15000-20000)_loc_1743_rev</td>\n",
              "      <td>CTGCCGTCGGCCTGGCCGTGTTCGGCTACTTGCAGTTTCAATCCAC...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>8</th>\n",
              "      <td>region_4_(15000-20000)_loc_4374_fwd</td>\n",
              "      <td>CCTCGTTCTGAATGGAGCTCGCCATGGGTCAGTCGTCGCcCggCaa...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>9</th>\n",
              "      <td>region_4_(15000-20000)_loc_4374_rev</td>\n",
              "      <td>GACTGACCCATGGCGAGCTCCATTCAGAACGAGGGTTTCAATCCAC...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>10</th>\n",
              "      <td>region_3_(9000-14000)_loc_2495_fwd</td>\n",
              "      <td>GGTTTGGTTGTTCGAGGGCTTGGCCTGCGGTCCGGTCGCcCggCaa...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>11</th>\n",
              "      <td>region_3_(9000-14000)_loc_2495_rev</td>\n",
              "      <td>CGGACCGCAGGCCAAGCCCTCGAACAACCAAACCGTTTCAATCCAC...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>12</th>\n",
              "      <td>region_2_(4000-7000)_loc_2196_fwd</td>\n",
              "      <td>CTGCAGGAGCCGCTGCTGGCCGGGCAACAGGGTGGTCGCcCggCaa...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>13</th>\n",
              "      <td>region_2_(4000-7000)_loc_2196_rev</td>\n",
              "      <td>CACCCTGTTGCCCGGCCAGCAGCGGCTCCTGCAGGTTTCAATCCAC...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>14</th>\n",
              "      <td>region_2_(4000-7000)_loc_1244_fwd</td>\n",
              "      <td>GAAAGCATGGCCTCGTCGACGCCACGGCAGGATTGTCGCcCggCaa...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>15</th>\n",
              "      <td>region_2_(4000-7000)_loc_1244_rev</td>\n",
              "      <td>AATCCTGCCGTGGCGTCGACGAGGCCATGCTTTCGTTTCAATCCAC...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "                                    Name  \\\n",
              "0   region_1_(80000-100000)_loc_9290_fwd   \n",
              "1   region_1_(80000-100000)_loc_9290_rev   \n",
              "2   region_1_(80000-100000)_loc_8869_fwd   \n",
              "3   region_1_(80000-100000)_loc_8869_rev   \n",
              "4      region_3_(9000-14000)_loc_305_fwd   \n",
              "5      region_3_(9000-14000)_loc_305_rev   \n",
              "6    region_4_(15000-20000)_loc_1743_fwd   \n",
              "7    region_4_(15000-20000)_loc_1743_rev   \n",
              "8    region_4_(15000-20000)_loc_4374_fwd   \n",
              "9    region_4_(15000-20000)_loc_4374_rev   \n",
              "10    region_3_(9000-14000)_loc_2495_fwd   \n",
              "11    region_3_(9000-14000)_loc_2495_rev   \n",
              "12     region_2_(4000-7000)_loc_2196_fwd   \n",
              "13     region_2_(4000-7000)_loc_2196_rev   \n",
              "14     region_2_(4000-7000)_loc_1244_fwd   \n",
              "15     region_2_(4000-7000)_loc_1244_rev   \n",
              "\n",
              "                                             Sequence Concentration  \\\n",
              "0   TCCCATGTGCCGTCGACCGCCCACATCCGCAGCCGTCGCcCggCaa...          25nm   \n",
              "1   GGCTGCGGATGTGGGCGGTCGACGGCACATGGGAGTTTCAATCCAC...          25nm   \n",
              "2   GTGTTAGCTAAGGCCTCCGCCCGCGAAGGAAAGGGTCGCcCggCaa...          25nm   \n",
              "3   CCTTTCCTTCGCGGGCGGAGGCCTTAGCTAACACGTTTCAATCCAC...          25nm   \n",
              "4   ACCATTGCTCGGCTTCCGCTTCGGCGTCCCACAGGTCGCcCggCaa...          25nm   \n",
              "5   CTGTGGGACGCCGAAGCGGAAGCCGAGCAATGGTGTTTCAATCCAC...          25nm   \n",
              "6   TGCAAGTAGCCGAACACGGCCAGGCCGACGGCAGGTCGCcCggCaa...          25nm   \n",
              "7   CTGCCGTCGGCCTGGCCGTGTTCGGCTACTTGCAGTTTCAATCCAC...          25nm   \n",
              "8   CCTCGTTCTGAATGGAGCTCGCCATGGGTCAGTCGTCGCcCggCaa...          25nm   \n",
              "9   GACTGACCCATGGCGAGCTCCATTCAGAACGAGGGTTTCAATCCAC...          25nm   \n",
              "10  GGTTTGGTTGTTCGAGGGCTTGGCCTGCGGTCCGGTCGCcCggCaa...          25nm   \n",
              "11  CGGACCGCAGGCCAAGCCCTCGAACAACCAAACCGTTTCAATCCAC...          25nm   \n",
              "12  CTGCAGGAGCCGCTGCTGGCCGGGCAACAGGGTGGTCGCcCggCaa...          25nm   \n",
              "13  CACCCTGTTGCCCGGCCAGCAGCGGCTCCTGCAGGTTTCAATCCAC...          25nm   \n",
              "14  GAAAGCATGGCCTCGTCGACGCCACGGCAGGATTGTCGCcCggCaa...          25nm   \n",
              "15  AATCCTGCCGTGGCGTCGACGAGGCCATGCTTTCGTTTCAATCCAC...          25nm   \n",
              "\n",
              "   Purification  \n",
              "0           STD  \n",
              "1           STD  \n",
              "2           STD  \n",
              "3           STD  \n",
              "4           STD  \n",
              "5           STD  \n",
              "6           STD  \n",
              "7           STD  \n",
              "8           STD  \n",
              "9           STD  \n",
              "10          STD  \n",
              "11          STD  \n",
              "12          STD  \n",
              "13          STD  \n",
              "14          STD  \n",
              "15          STD  "
            ]
          },
          "execution_count": 16,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "idt_primers_cas3=primers_to_IDT(primer_records)\n",
        "idt_primers_cas3"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "# In frame deletion"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 17,
      "metadata": {},
      "outputs": [],
      "source": [
        "in_frame_deletion = False "
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 18,
      "metadata": {},
      "outputs": [],
      "source": [
        "if in_frame_deletion:\n",
        "    # Make repair templates\n",
        "    print('repair_templates_data')\n",
        "    repair_templates_data = find_up_dw_repair_templates(genome, \n",
        "                                                        genes_to_KO, \n",
        "                                                        target_tm=melting_temperature, \n",
        "                                                        primer_tm_kwargs={'conc':primer_concentration, 'prodcode':chosen_polymerase} )\n",
        "    # Digest the plasmids\n",
        "    print('Digest the plasmids')\n",
        "    processed_records = [Dseqrecord(record, circular=True).cut(StuI)[0] for record in assembled_cas3_plasmids]\n",
        "\n",
        "    # Rename them appropriately\n",
        "    print('Rename them appropriately')\n",
        "\n",
        "    for i in range(len(processed_records)):\n",
        "        processed_records[i].name = assembled_cas3_plasmids[i].name\n",
        "\n",
        "    # Assembly \n",
        "    print('assembly_data')\n",
        "\n",
        "    assembly_data = assemble_multiple_plasmids_with_repair_templates_for_deletion(genes_to_KO, processed_records, \n",
        "                                                                                repair_templates_data, \n",
        "                                                                                overlap=40)\n",
        "    # updating the primer names to something systematic.\n",
        "    update_primer_names(assembly_data)\n",
        "\n",
        "    # Parse through the primer df\n",
        "    primer_df = create_primer_df_from_dict(assembly_data)\n",
        "\n",
        "    #Find_unique primers\n",
        "    unique_df = primer_df.drop_duplicates(keep='first')\n",
        "    idt_df = create_idt_order_dataframe(unique_df, concentration=\"25nm\", purification=\"STD\")\n",
        "\n",
        "    # Contigs\n",
        "    assembled_contigs = []\n",
        "    for data in assembly_data: \n",
        "        contig_record = data['contig']\n",
        "        contig_record.id = f\"{data['name']}_w_rep\"\n",
        "        contig_record.name = f\"{data['name']}_w_rep\"\n",
        "\n",
        "        assembled_contigs.append(contig_record)\n",
        "\n",
        "    print_plasmids = False\n",
        "\n",
        "    if print_plasmids:\n",
        "        # Write to genbank files\n",
        "        output_directory = \"../../data/plasmids/sgRNA_plasmids_pCRISPRâ€“Cas3_with_repair_templates/\"\n",
        "        for contig in assembled_contigs: \n",
        "            contig.write(f\"{output_directory}/{contig.id}.gb\")"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 19,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>plasmid_name</th>\n",
              "      <th>date</th>\n",
              "      <th>original_plasmid</th>\n",
              "      <th>integration</th>\n",
              "      <th>size</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>pCas3_region_1_(80000-100000)(9290)</td>\n",
              "      <td>2024-07-23</td>\n",
              "      <td>pCRISPR-Cas3</td>\n",
              "      <td>sgRNA_region_1_(80000-100000)(9290)</td>\n",
              "      <td>12766</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>pCas3_region_1_(80000-100000)(8869)</td>\n",
              "      <td>2024-07-23</td>\n",
              "      <td>pCRISPR-Cas3</td>\n",
              "      <td>sgRNA_region_1_(80000-100000)(8869)</td>\n",
              "      <td>12766</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>pCas3_region_3_(9000-14000)(305)</td>\n",
              "      <td>2024-07-23</td>\n",
              "      <td>pCRISPR-Cas3</td>\n",
              "      <td>sgRNA_region_3_(9000-14000)(305)</td>\n",
              "      <td>12766</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>pCas3_region_4_(15000-20000)(1743)</td>\n",
              "      <td>2024-07-23</td>\n",
              "      <td>pCRISPR-Cas3</td>\n",
              "      <td>sgRNA_region_4_(15000-20000)(1743)</td>\n",
              "      <td>12766</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>pCas3_region_4_(15000-20000)(4374)</td>\n",
              "      <td>2024-07-23</td>\n",
              "      <td>pCRISPR-Cas3</td>\n",
              "      <td>sgRNA_region_4_(15000-20000)(4374)</td>\n",
              "      <td>12766</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>5</th>\n",
              "      <td>pCas3_region_3_(9000-14000)(2495)</td>\n",
              "      <td>2024-07-23</td>\n",
              "      <td>pCRISPR-Cas3</td>\n",
              "      <td>sgRNA_region_3_(9000-14000)(2495)</td>\n",
              "      <td>12766</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>6</th>\n",
              "      <td>pCas3_region_2_(4000-7000)(2196)</td>\n",
              "      <td>2024-07-23</td>\n",
              "      <td>pCRISPR-Cas3</td>\n",
              "      <td>sgRNA_region_2_(4000-7000)(2196)</td>\n",
              "      <td>12766</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>7</th>\n",
              "      <td>pCas3_region_2_(4000-7000)(1244)</td>\n",
              "      <td>2024-07-23</td>\n",
              "      <td>pCRISPR-Cas3</td>\n",
              "      <td>sgRNA_region_2_(4000-7000)(1244)</td>\n",
              "      <td>12766</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "                          plasmid_name        date original_plasmid  \\\n",
              "0  pCas3_region_1_(80000-100000)(9290)  2024-07-23     pCRISPR-Cas3   \n",
              "1  pCas3_region_1_(80000-100000)(8869)  2024-07-23     pCRISPR-Cas3   \n",
              "2     pCas3_region_3_(9000-14000)(305)  2024-07-23     pCRISPR-Cas3   \n",
              "3   pCas3_region_4_(15000-20000)(1743)  2024-07-23     pCRISPR-Cas3   \n",
              "4   pCas3_region_4_(15000-20000)(4374)  2024-07-23     pCRISPR-Cas3   \n",
              "5    pCas3_region_3_(9000-14000)(2495)  2024-07-23     pCRISPR-Cas3   \n",
              "6     pCas3_region_2_(4000-7000)(2196)  2024-07-23     pCRISPR-Cas3   \n",
              "7     pCas3_region_2_(4000-7000)(1244)  2024-07-23     pCRISPR-Cas3   \n",
              "\n",
              "                           integration   size  \n",
              "0  sgRNA_region_1_(80000-100000)(9290)  12766  \n",
              "1  sgRNA_region_1_(80000-100000)(8869)  12766  \n",
              "2     sgRNA_region_3_(9000-14000)(305)  12766  \n",
              "3   sgRNA_region_4_(15000-20000)(1743)  12766  \n",
              "4   sgRNA_region_4_(15000-20000)(4374)  12766  \n",
              "5    sgRNA_region_3_(9000-14000)(2495)  12766  \n",
              "6     sgRNA_region_2_(4000-7000)(2196)  12766  \n",
              "7     sgRNA_region_2_(4000-7000)(1244)  12766  "
            ]
          },
          "execution_count": 19,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "if in_frame_deletion:\n",
        "    integration_names = filtered_df.apply(lambda row: f\"sgRNA_{row['locus_tag']}({row['sgrna_loc']})\", axis=1).tolist()\n",
        "    plasmid_metadata_df = extract_metadata_to_dataframe(assembled_contigs,\n",
        "                                                        clean_plasmid,\n",
        "                                                        integration_names)\n",
        "else: \n",
        "    integration_names = filtered_df.apply(lambda row: f\"sgRNA_{row['locus_tag']}({row['sgrna_loc']})\", axis=1).tolist()\n",
        "    plasmid_metadata_df = extract_metadata_to_dataframe(assembled_cas3_plasmids,\n",
        "                                                        clean_plasmid,\n",
        "                                                        integration_names)\n",
        "\n",
        "plasmid_metadata_df"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 20,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>locus tag</th>\n",
              "      <th>f_primer_name</th>\n",
              "      <th>r_primer_name</th>\n",
              "      <th>f_primer_sequences(5-3)</th>\n",
              "      <th>r_primer_sequences(5-3)</th>\n",
              "      <th>f_tm</th>\n",
              "      <th>r_tm</th>\n",
              "      <th>ta</th>\n",
              "      <th>flanking_region</th>\n",
              "      <th>annealing_temperature</th>\n",
              "      <th>...</th>\n",
              "      <th>homodimer_reverse_tm</th>\n",
              "      <th>homodimer_reverse_deltaG (kcal/mol)</th>\n",
              "      <th>heterodimer_tm</th>\n",
              "      <th>heterodimer_deltaG (kcal/mol)</th>\n",
              "      <th>hairpin_forward_structure_found</th>\n",
              "      <th>hairpin_forward_tm</th>\n",
              "      <th>hairpin_forward_deltaG (kcal/mol)</th>\n",
              "      <th>hairpin_reverse_structure_found</th>\n",
              "      <th>hairpin_reverse_tm</th>\n",
              "      <th>hairpin_reverse_deltaG (kcal/mol)</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>region_3_(9000-14000)</td>\n",
              "      <td>region_3_(9000-14000)_fwd_checking_primer</td>\n",
              "      <td>region_3_(9000-14000)_rev_checking_primer</td>\n",
              "      <td>ATCCCGCAGCTGCAGG</td>\n",
              "      <td>GAGGAGCGGCAATGGTCATC</td>\n",
              "      <td>61</td>\n",
              "      <td>63</td>\n",
              "      <td>65</td>\n",
              "      <td>500</td>\n",
              "      <td>65</td>\n",
              "      <td>...</td>\n",
              "      <td>1.445721</td>\n",
              "      <td>-1.026208</td>\n",
              "      <td>9.252334</td>\n",
              "      <td>-1.213024</td>\n",
              "      <td>False</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>False</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>region_2_(4000-7000)</td>\n",
              "      <td>region_2_(4000-7000)_fwd_checking_primer</td>\n",
              "      <td>region_2_(4000-7000)_rev_checking_primer</td>\n",
              "      <td>GACCGCGCTGGGTG</td>\n",
              "      <td>TCCGGCCGCATCG</td>\n",
              "      <td>59</td>\n",
              "      <td>57</td>\n",
              "      <td>60</td>\n",
              "      <td>500</td>\n",
              "      <td>60</td>\n",
              "      <td>...</td>\n",
              "      <td>22.848051</td>\n",
              "      <td>-5.018625</td>\n",
              "      <td>6.535424</td>\n",
              "      <td>-0.061707</td>\n",
              "      <td>False</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>False</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>region_4_(15000-20000)</td>\n",
              "      <td>region_4_(15000-20000)_fwd_checking_primer</td>\n",
              "      <td>region_4_(15000-20000)_rev_checking_primer</td>\n",
              "      <td>TTGCTGCAACCTGCCTTG</td>\n",
              "      <td>TAGGTGCCTCCGTCGA</td>\n",
              "      <td>61</td>\n",
              "      <td>58</td>\n",
              "      <td>62</td>\n",
              "      <td>500</td>\n",
              "      <td>62</td>\n",
              "      <td>...</td>\n",
              "      <td>-13.794142</td>\n",
              "      <td>-0.897729</td>\n",
              "      <td>-1.321907</td>\n",
              "      <td>0.948660</td>\n",
              "      <td>False</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>False</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>region_1_(80000-100000)</td>\n",
              "      <td>region_1_(80000-100000)_fwd_checking_primer</td>\n",
              "      <td>region_1_(80000-100000)_rev_checking_primer</td>\n",
              "      <td>GCCGCTGGCCAAGC</td>\n",
              "      <td>AGGCGTGCGCGTGG</td>\n",
              "      <td>60</td>\n",
              "      <td>62</td>\n",
              "      <td>63</td>\n",
              "      <td>700</td>\n",
              "      <td>63</td>\n",
              "      <td>...</td>\n",
              "      <td>22.176370</td>\n",
              "      <td>-2.921991</td>\n",
              "      <td>11.333376</td>\n",
              "      <td>-0.877576</td>\n",
              "      <td>False</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>False</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>4 rows Ã— 23 columns</p>\n",
              "</div>"
            ],
            "text/plain": [
              "                 locus tag                                f_primer_name  \\\n",
              "0    region_3_(9000-14000)    region_3_(9000-14000)_fwd_checking_primer   \n",
              "1     region_2_(4000-7000)     region_2_(4000-7000)_fwd_checking_primer   \n",
              "2   region_4_(15000-20000)   region_4_(15000-20000)_fwd_checking_primer   \n",
              "3  region_1_(80000-100000)  region_1_(80000-100000)_fwd_checking_primer   \n",
              "\n",
              "                                 r_primer_name f_primer_sequences(5-3)  \\\n",
              "0    region_3_(9000-14000)_rev_checking_primer        ATCCCGCAGCTGCAGG   \n",
              "1     region_2_(4000-7000)_rev_checking_primer          GACCGCGCTGGGTG   \n",
              "2   region_4_(15000-20000)_rev_checking_primer      TTGCTGCAACCTGCCTTG   \n",
              "3  region_1_(80000-100000)_rev_checking_primer          GCCGCTGGCCAAGC   \n",
              "\n",
              "  r_primer_sequences(5-3) f_tm r_tm  ta flanking_region  \\\n",
              "0    GAGGAGCGGCAATGGTCATC   61   63  65             500   \n",
              "1           TCCGGCCGCATCG   59   57  60             500   \n",
              "2        TAGGTGCCTCCGTCGA   61   58  62             500   \n",
              "3          AGGCGTGCGCGTGG   60   62  63             700   \n",
              "\n",
              "   annealing_temperature  ... homodimer_reverse_tm  \\\n",
              "0                     65  ...             1.445721   \n",
              "1                     60  ...            22.848051   \n",
              "2                     62  ...           -13.794142   \n",
              "3                     63  ...            22.176370   \n",
              "\n",
              "   homodimer_reverse_deltaG (kcal/mol)  heterodimer_tm  \\\n",
              "0                            -1.026208        9.252334   \n",
              "1                            -5.018625        6.535424   \n",
              "2                            -0.897729       -1.321907   \n",
              "3                            -2.921991       11.333376   \n",
              "\n",
              "   heterodimer_deltaG (kcal/mol)  hairpin_forward_structure_found  \\\n",
              "0                      -1.213024                            False   \n",
              "1                      -0.061707                            False   \n",
              "2                       0.948660                            False   \n",
              "3                      -0.877576                            False   \n",
              "\n",
              "   hairpin_forward_tm  hairpin_forward_deltaG (kcal/mol)  \\\n",
              "0                 0.0                                0.0   \n",
              "1                 0.0                                0.0   \n",
              "2                 0.0                                0.0   \n",
              "3                 0.0                                0.0   \n",
              "\n",
              "   hairpin_reverse_structure_found  hairpin_reverse_tm  \\\n",
              "0                            False                 0.0   \n",
              "1                            False                 0.0   \n",
              "2                            False                 0.0   \n",
              "3                            False                 0.0   \n",
              "\n",
              "   hairpin_reverse_deltaG (kcal/mol)  \n",
              "0                                0.0  \n",
              "1                                0.0  \n",
              "2                                0.0  \n",
              "3                                0.0  \n",
              "\n",
              "[4 rows x 23 columns]"
            ]
          },
          "execution_count": 20,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "# Getting checking primers\n",
        "checking_primers_df = find_best_check_primers_from_genome(genome, \n",
        "                                       genes_to_KO, \n",
        "                                       flanking_region=flanking_region,\n",
        "                                       target_tm = melting_temperature, \n",
        "                                        primer_concentration = primer_concentration, \n",
        "                                        polymerase = chosen_polymerase)\n",
        "checking_primers_df"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 21,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>Name</th>\n",
              "      <th>Sequence</th>\n",
              "      <th>Concentration</th>\n",
              "      <th>Purification</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>region_3_(9000-14000)_fwd_checking_primer</td>\n",
              "      <td>ATCCCGCAGCTGCAGG</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>region_2_(4000-7000)_fwd_checking_primer</td>\n",
              "      <td>GACCGCGCTGGGTG</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>region_4_(15000-20000)_fwd_checking_primer</td>\n",
              "      <td>TTGCTGCAACCTGCCTTG</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>region_1_(80000-100000)_fwd_checking_primer</td>\n",
              "      <td>GCCGCTGGCCAAGC</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>region_3_(9000-14000)_rev_checking_primer</td>\n",
              "      <td>GAGGAGCGGCAATGGTCATC</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>5</th>\n",
              "      <td>region_2_(4000-7000)_rev_checking_primer</td>\n",
              "      <td>TCCGGCCGCATCG</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>6</th>\n",
              "      <td>region_4_(15000-20000)_rev_checking_primer</td>\n",
              "      <td>TAGGTGCCTCCGTCGA</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>7</th>\n",
              "      <td>region_1_(80000-100000)_rev_checking_primer</td>\n",
              "      <td>AGGCGTGCGCGTGG</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "                                          Name              Sequence  \\\n",
              "0    region_3_(9000-14000)_fwd_checking_primer      ATCCCGCAGCTGCAGG   \n",
              "1     region_2_(4000-7000)_fwd_checking_primer        GACCGCGCTGGGTG   \n",
              "2   region_4_(15000-20000)_fwd_checking_primer    TTGCTGCAACCTGCCTTG   \n",
              "3  region_1_(80000-100000)_fwd_checking_primer        GCCGCTGGCCAAGC   \n",
              "4    region_3_(9000-14000)_rev_checking_primer  GAGGAGCGGCAATGGTCATC   \n",
              "5     region_2_(4000-7000)_rev_checking_primer         TCCGGCCGCATCG   \n",
              "6   region_4_(15000-20000)_rev_checking_primer      TAGGTGCCTCCGTCGA   \n",
              "7  region_1_(80000-100000)_rev_checking_primer        AGGCGTGCGCGTGG   \n",
              "\n",
              "  Concentration Purification  \n",
              "0          25nm          STD  \n",
              "1          25nm          STD  \n",
              "2          25nm          STD  \n",
              "3          25nm          STD  \n",
              "4          25nm          STD  \n",
              "5          25nm          STD  \n",
              "6          25nm          STD  \n",
              "7          25nm          STD  "
            ]
          },
          "execution_count": 21,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "# making the primers into dseqrecords\n",
        "checking_primers_df_idt = create_idt_order_dataframe(checking_primers_df)\n",
        "checking_primers_df_idt\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 22,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>Name</th>\n",
              "      <th>Sequence</th>\n",
              "      <th>Concentration</th>\n",
              "      <th>Purification</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>region_1_(80000-100000)_loc_9290_fwd</td>\n",
              "      <td>TCCCATGTGCCGTCGACCGCCCACATCCGCAGCCGTCGCcCggCaa...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>region_1_(80000-100000)_loc_9290_rev</td>\n",
              "      <td>GGCTGCGGATGTGGGCGGTCGACGGCACATGGGAGTTTCAATCCAC...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>region_1_(80000-100000)_loc_8869_fwd</td>\n",
              "      <td>GTGTTAGCTAAGGCCTCCGCCCGCGAAGGAAAGGGTCGCcCggCaa...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>region_1_(80000-100000)_loc_8869_rev</td>\n",
              "      <td>CCTTTCCTTCGCGGGCGGAGGCCTTAGCTAACACGTTTCAATCCAC...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>region_3_(9000-14000)_loc_305_fwd</td>\n",
              "      <td>ACCATTGCTCGGCTTCCGCTTCGGCGTCCCACAGGTCGCcCggCaa...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>5</th>\n",
              "      <td>region_3_(9000-14000)_loc_305_rev</td>\n",
              "      <td>CTGTGGGACGCCGAAGCGGAAGCCGAGCAATGGTGTTTCAATCCAC...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>6</th>\n",
              "      <td>region_4_(15000-20000)_loc_1743_fwd</td>\n",
              "      <td>TGCAAGTAGCCGAACACGGCCAGGCCGACGGCAGGTCGCcCggCaa...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>7</th>\n",
              "      <td>region_4_(15000-20000)_loc_1743_rev</td>\n",
              "      <td>CTGCCGTCGGCCTGGCCGTGTTCGGCTACTTGCAGTTTCAATCCAC...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>8</th>\n",
              "      <td>region_4_(15000-20000)_loc_4374_fwd</td>\n",
              "      <td>CCTCGTTCTGAATGGAGCTCGCCATGGGTCAGTCGTCGCcCggCaa...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>9</th>\n",
              "      <td>region_4_(15000-20000)_loc_4374_rev</td>\n",
              "      <td>GACTGACCCATGGCGAGCTCCATTCAGAACGAGGGTTTCAATCCAC...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>10</th>\n",
              "      <td>region_3_(9000-14000)_loc_2495_fwd</td>\n",
              "      <td>GGTTTGGTTGTTCGAGGGCTTGGCCTGCGGTCCGGTCGCcCggCaa...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>11</th>\n",
              "      <td>region_3_(9000-14000)_loc_2495_rev</td>\n",
              "      <td>CGGACCGCAGGCCAAGCCCTCGAACAACCAAACCGTTTCAATCCAC...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>12</th>\n",
              "      <td>region_2_(4000-7000)_loc_2196_fwd</td>\n",
              "      <td>CTGCAGGAGCCGCTGCTGGCCGGGCAACAGGGTGGTCGCcCggCaa...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>13</th>\n",
              "      <td>region_2_(4000-7000)_loc_2196_rev</td>\n",
              "      <td>CACCCTGTTGCCCGGCCAGCAGCGGCTCCTGCAGGTTTCAATCCAC...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>14</th>\n",
              "      <td>region_2_(4000-7000)_loc_1244_fwd</td>\n",
              "      <td>GAAAGCATGGCCTCGTCGACGCCACGGCAGGATTGTCGCcCggCaa...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>15</th>\n",
              "      <td>region_2_(4000-7000)_loc_1244_rev</td>\n",
              "      <td>AATCCTGCCGTGGCGTCGACGAGGCCATGCTTTCGTTTCAATCCAC...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>16</th>\n",
              "      <td>region_3_(9000-14000)_fwd_checking_primer</td>\n",
              "      <td>ATCCCGCAGCTGCAGG</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>17</th>\n",
              "      <td>region_2_(4000-7000)_fwd_checking_primer</td>\n",
              "      <td>GACCGCGCTGGGTG</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>18</th>\n",
              "      <td>region_4_(15000-20000)_fwd_checking_primer</td>\n",
              "      <td>TTGCTGCAACCTGCCTTG</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>19</th>\n",
              "      <td>region_1_(80000-100000)_fwd_checking_primer</td>\n",
              "      <td>GCCGCTGGCCAAGC</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>20</th>\n",
              "      <td>region_3_(9000-14000)_rev_checking_primer</td>\n",
              "      <td>GAGGAGCGGCAATGGTCATC</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>21</th>\n",
              "      <td>region_2_(4000-7000)_rev_checking_primer</td>\n",
              "      <td>TCCGGCCGCATCG</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>22</th>\n",
              "      <td>region_4_(15000-20000)_rev_checking_primer</td>\n",
              "      <td>TAGGTGCCTCCGTCGA</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>23</th>\n",
              "      <td>region_1_(80000-100000)_rev_checking_primer</td>\n",
              "      <td>AGGCGTGCGCGTGG</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "                                           Name  \\\n",
              "0          region_1_(80000-100000)_loc_9290_fwd   \n",
              "1          region_1_(80000-100000)_loc_9290_rev   \n",
              "2          region_1_(80000-100000)_loc_8869_fwd   \n",
              "3          region_1_(80000-100000)_loc_8869_rev   \n",
              "4             region_3_(9000-14000)_loc_305_fwd   \n",
              "5             region_3_(9000-14000)_loc_305_rev   \n",
              "6           region_4_(15000-20000)_loc_1743_fwd   \n",
              "7           region_4_(15000-20000)_loc_1743_rev   \n",
              "8           region_4_(15000-20000)_loc_4374_fwd   \n",
              "9           region_4_(15000-20000)_loc_4374_rev   \n",
              "10           region_3_(9000-14000)_loc_2495_fwd   \n",
              "11           region_3_(9000-14000)_loc_2495_rev   \n",
              "12            region_2_(4000-7000)_loc_2196_fwd   \n",
              "13            region_2_(4000-7000)_loc_2196_rev   \n",
              "14            region_2_(4000-7000)_loc_1244_fwd   \n",
              "15            region_2_(4000-7000)_loc_1244_rev   \n",
              "16    region_3_(9000-14000)_fwd_checking_primer   \n",
              "17     region_2_(4000-7000)_fwd_checking_primer   \n",
              "18   region_4_(15000-20000)_fwd_checking_primer   \n",
              "19  region_1_(80000-100000)_fwd_checking_primer   \n",
              "20    region_3_(9000-14000)_rev_checking_primer   \n",
              "21     region_2_(4000-7000)_rev_checking_primer   \n",
              "22   region_4_(15000-20000)_rev_checking_primer   \n",
              "23  region_1_(80000-100000)_rev_checking_primer   \n",
              "\n",
              "                                             Sequence Concentration  \\\n",
              "0   TCCCATGTGCCGTCGACCGCCCACATCCGCAGCCGTCGCcCggCaa...          25nm   \n",
              "1   GGCTGCGGATGTGGGCGGTCGACGGCACATGGGAGTTTCAATCCAC...          25nm   \n",
              "2   GTGTTAGCTAAGGCCTCCGCCCGCGAAGGAAAGGGTCGCcCggCaa...          25nm   \n",
              "3   CCTTTCCTTCGCGGGCGGAGGCCTTAGCTAACACGTTTCAATCCAC...          25nm   \n",
              "4   ACCATTGCTCGGCTTCCGCTTCGGCGTCCCACAGGTCGCcCggCaa...          25nm   \n",
              "5   CTGTGGGACGCCGAAGCGGAAGCCGAGCAATGGTGTTTCAATCCAC...          25nm   \n",
              "6   TGCAAGTAGCCGAACACGGCCAGGCCGACGGCAGGTCGCcCggCaa...          25nm   \n",
              "7   CTGCCGTCGGCCTGGCCGTGTTCGGCTACTTGCAGTTTCAATCCAC...          25nm   \n",
              "8   CCTCGTTCTGAATGGAGCTCGCCATGGGTCAGTCGTCGCcCggCaa...          25nm   \n",
              "9   GACTGACCCATGGCGAGCTCCATTCAGAACGAGGGTTTCAATCCAC...          25nm   \n",
              "10  GGTTTGGTTGTTCGAGGGCTTGGCCTGCGGTCCGGTCGCcCggCaa...          25nm   \n",
              "11  CGGACCGCAGGCCAAGCCCTCGAACAACCAAACCGTTTCAATCCAC...          25nm   \n",
              "12  CTGCAGGAGCCGCTGCTGGCCGGGCAACAGGGTGGTCGCcCggCaa...          25nm   \n",
              "13  CACCCTGTTGCCCGGCCAGCAGCGGCTCCTGCAGGTTTCAATCCAC...          25nm   \n",
              "14  GAAAGCATGGCCTCGTCGACGCCACGGCAGGATTGTCGCcCggCaa...          25nm   \n",
              "15  AATCCTGCCGTGGCGTCGACGAGGCCATGCTTTCGTTTCAATCCAC...          25nm   \n",
              "16                                   ATCCCGCAGCTGCAGG          25nm   \n",
              "17                                     GACCGCGCTGGGTG          25nm   \n",
              "18                                 TTGCTGCAACCTGCCTTG          25nm   \n",
              "19                                     GCCGCTGGCCAAGC          25nm   \n",
              "20                               GAGGAGCGGCAATGGTCATC          25nm   \n",
              "21                                      TCCGGCCGCATCG          25nm   \n",
              "22                                   TAGGTGCCTCCGTCGA          25nm   \n",
              "23                                     AGGCGTGCGCGTGG          25nm   \n",
              "\n",
              "   Purification  \n",
              "0           STD  \n",
              "1           STD  \n",
              "2           STD  \n",
              "3           STD  \n",
              "4           STD  \n",
              "5           STD  \n",
              "6           STD  \n",
              "7           STD  \n",
              "8           STD  \n",
              "9           STD  \n",
              "10          STD  \n",
              "11          STD  \n",
              "12          STD  \n",
              "13          STD  \n",
              "14          STD  \n",
              "15          STD  \n",
              "16          STD  \n",
              "17          STD  \n",
              "18          STD  \n",
              "19          STD  \n",
              "20          STD  \n",
              "21          STD  \n",
              "22          STD  \n",
              "23          STD  "
            ]
          },
          "execution_count": 22,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "if in_frame_deletion: \n",
        "    full_idt = pd.concat([idt_primers_cas3,idt_df, checking_primers_df_idt],ignore_index=True)\n",
        "else: \n",
        "    full_idt = pd.concat([idt_primers_cas3,checking_primers_df_idt],ignore_index=True)\n",
        "\n",
        "full_idt"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "## Folder with all the generated I/O"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 23,
      "metadata": {},
      "outputs": [],
      "source": [
        "\n",
        "input_files = [\n",
        "    {\"name\": \"input_genome.gb\", \"content\": genome},\n",
        "    {\"name\": \"input_plasmid.gb\", \"content\": clean_plasmid}\n",
        "]\n",
        "\n",
        "\n",
        "if in_frame_deletion: \n",
        "    output_files = [\n",
        "        {\"name\": \"BEST_w_sgRNAs.gb\", \"content\": assembled_contigs}, # LIST OF Dseqrecords\n",
        "        {\"name\": \"primer_df.csv\", \"content\": primer_df},\n",
        "        {\"name\": \"full_idt.csv\", \"content\": full_idt},\n",
        "        {\"name\": \"sgrna_df.csv\", \"content\": sgrna_df},\n",
        "        {\"name\": \"filtered_sgrna_df.csv\", \"content\": filtered_df},\n",
        "        {\"name\": \"plasmid_metadata_df.csv\", \"content\": plasmid_metadata_df},\n",
        "\n",
        "    ]\n",
        "else: \n",
        "    output_files = [\n",
        "        {\"name\": \"cBEST_w_sgRNAs.gb\", \"content\": assembled_cas3_plasmids}, # LIST OF Dseqrecords\n",
        "        {\"name\": \"full_idt.csv\", \"content\": full_idt},\n",
        "        {\"name\": \"sgrna_df.csv\", \"content\": sgrna_df},\n",
        "        {\"name\": \"filtered_df.csv\", \"content\": filtered_df},\n",
        "        {\"name\": \"plasmid_metadata_df.csv\", \"content\": plasmid_metadata_df},\n",
        "    ]\n",
        "         \n",
        "input_values = {\n",
        "    \"genes_to_knockout\": genes_to_KO,\n",
        "\n",
        "    \"filtering_metrics\": {\n",
        "        \"gc_upper\": gc_upper,\n",
        "        \"gc_lower\": gc_lower,\n",
        "        \"off_target_seed\": off_target_seed,\n",
        "        \"off_target_upper\": off_target_upper,\n",
        "        \"cas_type\": cas_type,\n",
        "        \"number_of_sgRNAs_per_group\": number_of_sgRNAs_per_group,\n",
        "\n",
        "    },  \n",
        "    \"polymerase_settings\": {\n",
        "        \"chosen_polymerase\": chosen_polymerase,\n",
        "        \"melting_temperature\": melting_temperature,\n",
        "        \"primer_concentration\": primer_concentration,\n",
        "        \"primer_number_increment\": primer_number_increment,\n",
        "        \"flanking_region\": flanking_region\n",
        "    },\n",
        "    \"overlapping_sequences\": {\n",
        "        \"up_homology\": str(up_homology),\n",
        "        \"dw_homology\": str(dw_homology)\n",
        "    }\n",
        "}\n",
        "\n",
        "# Paths to Markdown files\n",
        "markdown_file_paths = [\n",
        "    \"../../protocols/conjugation_protcol.md\",\n",
        "    \"../../protocols/single_target_crispr_plasmid_protcol.md\"\n",
        "\n",
        "]\n",
        "\n",
        "timestamp = datetime.utcnow().isoformat()\n",
        "\n",
        "# Create project directory structure\n",
        "project_directory = ProjectDirectory(\n",
        "    project_name=f\"CRISPR_cas9_inframe_deletion_workflow_{timestamp}\",\n",
        "    input_files=input_files,\n",
        "    output_files=output_files,\n",
        "    input_values=input_values,\n",
        "    markdown_file_paths=markdown_file_paths\n",
        ")\n",
        "\n",
        "# DO You want to save the folder? \n",
        "save_zip_folder = False \n",
        "\n",
        "if save_zip_folder: \n",
        "    # Generate the project directory structure and get the zip content\n",
        "    zip_content = project_directory.create_directory_structure(create_directories=False)\n",
        "\n",
        "    # Save the zip file to disk (optional)\n",
        "    with open(\"project_structure.zip\", \"wb\") as f:\n",
        "        f.write(zip_content)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {},
      "outputs": [],
      "source": []
    }
  ],
  "metadata": {
    "kernelspec": {
      "display_name": "constrain",
      "language": "python",
      "name": "python3"
    },
    "language_info": {
      "codemirror_mode": {
        "name": "ipython",
        "version": 3
      },
      "file_extension": ".py",
      "mimetype": "text/x-python",
      "name": "python",
      "nbconvert_exporter": "python",
      "pygments_lexer": "ipython3",
      "version": "3.8.13"
    }
  },
  "nbformat": 4,
  "nbformat_minor": 2
}
