{
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "# W6-CRISPR-Cas3-in-frame-deletion"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "In Workflow 6, CRISPR-Cas3 is paired with a PCR and Gibson protocol to achieve in-frame deletions22(Figure 7A). Just like in workflow 5, this workflow can be used to knock out specific genes in streptomycetes but with the crucial difference of using Cas3 which has been shown to have higher efficiencies for targeted gene deletions in for example BGCs and random-sized deletion in streptomyces22. Due to homologies, and high GC content, this protocol relies on two steps of Gibson cloning to incorporate the sgRNA and the upstream and downstream repair regions. To get started, users can download the pCRISPR-Cas3 file and the genome file (S. coelicolor A3). StreptoAIM will then generate the necessary primers, plasmid files, and checking primers to verify full gene deletions (S5). \n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 1,
      "metadata": {},
      "outputs": [],
      "source": [
        "import sys\n",
        "import os\n",
        "from Bio.Restriction import * \n",
        "import pandas as pd\n",
        "from datetime import datetime\n",
        "\n",
        "# Ensure the src directory is in the Python path\n",
        "project_root = os.path.abspath(os.path.join(os.getcwd(), '../../'))\n",
        "if project_root not in sys.path:\n",
        "    sys.path.insert(0, project_root)\n",
        "\n",
        "\n",
        "from pydna.dseqrecord import Dseqrecord\n",
        "\n",
        "from streptocad.sequence_loading.sequence_loading import (\n",
        "    load_and_process_plasmid,\n",
        "    load_and_process_genome_sequences, \n",
        "    check_and_convert_input, \n",
        "    annotate_dseqrecord)\n",
        "\n",
        "from streptocad.utils import polymerase_dict, create_primer_df_from_dict,ProjectDirectory,extract_metadata_to_dataframe\n",
        "from streptocad.primers.primer_generation import create_idt_order_dataframe, make_primer_records, primers_to_IDT, find_best_check_primers_from_genome\n",
        "from streptocad.crispr.guideRNAcas3_9_12 import extract_sgRNAs, SgRNAargs\n",
        "from streptocad.cloning.cas3_plasmid_cloning import generate_cas3_protospacer_primers, cas3_plasmid_pcrs, assemble_cas3_plasmids\n",
        "from streptocad.cloning.gibson_cloning import (\n",
        "    find_up_dw_repair_templates,\n",
        "    assemble_multiple_plasmids_with_repair_templates_for_deletion,\n",
        "    update_primer_names\n",
        ")\n",
        "from streptocad.cloning.plasmid_processing import determine_workflow_order_for_plasmids"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 2,
      "metadata": {},
      "outputs": [],
      "source": [
        "from pydna.dseqrecord import Dseqrecord\n",
        "\n",
        "def find_locus_tags(record: Dseqrecord, max_tags: int = None):\n",
        "    \"\"\"\n",
        "    Iterate through a Dseqrecord and return a list of unique locus tags found in its features,\n",
        "    stopping after `max_tags` tags if specified.\n",
        "\n",
        "    :param record: pydna.dseqrecord.Dseqrecord object\n",
        "    :param max_tags: maximum number of locus tags to collect (None for no limit)\n",
        "    :return: list of locus_tag strings\n",
        "    \"\"\"\n",
        "    tags = []\n",
        "    seen = set()\n",
        "    count = 0\n",
        "\n",
        "    # Ensure the record has features\n",
        "    if not hasattr(record, 'features'):\n",
        "        return tags\n",
        "\n",
        "    for feature in record.features:\n",
        "        if max_tags is not None and count >= max_tags:\n",
        "            break\n",
        "\n",
        "        qualifiers = getattr(feature, 'qualifiers', {})\n",
        "\n",
        "        for tag in qualifiers.get('locus_tag', []):\n",
        "            if max_tags is not None and count >= max_tags:\n",
        "                break\n",
        "            # only append new tags\n",
        "            if tag not in seen:\n",
        "                seen.add(tag)\n",
        "                tags.append(tag)\n",
        "                count += 1\n",
        "\n",
        "        if max_tags is not None and count >= max_tags:\n",
        "            break\n",
        "\n",
        "    return tags\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "## INPUT"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 3,
      "metadata": {},
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "pCRISPR-Cas3\n"
          ]
        }
      ],
      "source": [
        "# Inputs\n",
        "# 1 Add genome of choice (genbank, fasta)\n",
        "path_to_genome = '../../../../data/genomes/Streptomyces_coelicolor_A3_chromosome.gb'\n",
        "genome = load_and_process_genome_sequences(path_to_genome)[0]\n",
        "\n",
        "# 2 Add plasmid \n",
        "path_to_plasmid = '../../../../data/plasmids/pCRISPR_Cas3.gbk'\n",
        "clean_plasmid = load_and_process_plasmid(path_to_plasmid)\n",
        "print(clean_plasmid.name)\n",
        "\n",
        "# 3 Choose genes to knock out (list)\n",
        "locus_tags = find_locus_tags(genome, 150)\n",
        "genes_to_KO = locus_tags\n",
        "\n",
        "\n",
        "#### Advanced settings ####\n",
        "# 4 Filtering metrics for sgRNAs\n",
        "gc_upper = 0.8\n",
        "gc_lower = 0.2\n",
        "off_target_seed = 8\n",
        "off_target_upper = 10\n",
        "cas_type='cas3'\n",
        "number_of_sgRNAs_per_group = 1\n",
        "\n",
        "# 5 Choose polymerase and target melting temperature\n",
        "chosen_polymerase = polymerase_dict['Phusion High-Fidelity DNA Polymerase (GC Buffer)']\n",
        "melting_temperature = 65\n",
        "primer_concentration = 0.4 \n",
        "primer_number_increment = 23\n",
        "flanking_region = 500\n",
        "# 6 Choose overlapping sequences for our plasmid we can use the following\n",
        "# the part of the primers that anneal to the generate the crRNA and protospacer\n",
        "forward_protospacer_overhang = 'GTCGCCCGGCAAAACCGG'\n",
        "reverse_protospacer_overhang = 'GTTTCAATCCACGCGCCCGT'\n",
        "\n",
        "# The universal part of the \n",
        "forward_backbone_overhang = 'GAGCTCATAAGTTCCTATTCCGAAG'\n",
        "reverse_backbone_overhang = 'AAGAAGTGGGTGTCGGACGC'\n",
        "\n",
        "Enzyme_for_repair_template_integration = EcoRI\n",
        "repair_length = 1000\n",
        "overlap_length = 40 \n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "# Computation"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 4,
      "metadata": {},
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "False\n"
          ]
        }
      ],
      "source": [
        "target_dict, genes_to_KO, annotation_input = check_and_convert_input(genes_to_KO)\n",
        "\n",
        "print(annotation_input)\n",
        "if annotation_input == True:\n",
        "    genome = annotate_dseqrecord(genome, target_dict)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 5,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>strain_name</th>\n",
              "      <th>locus_tag</th>\n",
              "      <th>gene_loc</th>\n",
              "      <th>gene_strand</th>\n",
              "      <th>sgrna_strand</th>\n",
              "      <th>sgrna_loc</th>\n",
              "      <th>gc</th>\n",
              "      <th>pam</th>\n",
              "      <th>sgrna</th>\n",
              "      <th>sgrna_seed_sequence</th>\n",
              "      <th>off_target_count</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>1196</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO0088</td>\n",
              "      <td>74917</td>\n",
              "      <td>-1</td>\n",
              "      <td>1</td>\n",
              "      <td>836</td>\n",
              "      <td>0.676471</td>\n",
              "      <td>TTC</td>\n",
              "      <td>ACATGGAGGCGGGAGTCGTCCGGTATCCCGACGG</td>\n",
              "      <td>ACATGGAG</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2173</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO0144</td>\n",
              "      <td>137197</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>109</td>\n",
              "      <td>0.676471</td>\n",
              "      <td>TTC</td>\n",
              "      <td>AAAGGCAGCCTGACCATCCCGCTGGCCCACGTCC</td>\n",
              "      <td>AAAGGCAG</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2176</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO0144</td>\n",
              "      <td>137197</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>282</td>\n",
              "      <td>0.588235</td>\n",
              "      <td>TTC</td>\n",
              "      <td>GAAGGCTGTCGTGATCGAACTCTCCGACGAGCAG</td>\n",
              "      <td>GAAGGCTG</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>764</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO0062</td>\n",
              "      <td>47789</td>\n",
              "      <td>-1</td>\n",
              "      <td>-1</td>\n",
              "      <td>844</td>\n",
              "      <td>0.647059</td>\n",
              "      <td>TTC</td>\n",
              "      <td>CGGATTTGTACAGGGCCCGGATCGCTCCCTGGCT</td>\n",
              "      <td>CGGATTTG</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>754</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO0062</td>\n",
              "      <td>47789</td>\n",
              "      <td>-1</td>\n",
              "      <td>1</td>\n",
              "      <td>325</td>\n",
              "      <td>0.676471</td>\n",
              "      <td>TTC</td>\n",
              "      <td>TACTCACGCCTCGCCCTCGGCGTCCAGGAAGTGG</td>\n",
              "      <td>TACTCACG</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>...</th>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2144</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO0142</td>\n",
              "      <td>134295</td>\n",
              "      <td>-1</td>\n",
              "      <td>1</td>\n",
              "      <td>601</td>\n",
              "      <td>0.705882</td>\n",
              "      <td>TTC</td>\n",
              "      <td>CTGGTCAGCGACACCACCCGGCACCAGCTGGTGG</td>\n",
              "      <td>CTGGTCAG</td>\n",
              "      <td>10</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>879</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO0067</td>\n",
              "      <td>54097</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>313</td>\n",
              "      <td>0.617647</td>\n",
              "      <td>TTC</td>\n",
              "      <td>GCCACGATCATCATGCCCGTCAAGAGGCAGGCGT</td>\n",
              "      <td>GCCACGAT</td>\n",
              "      <td>10</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1794</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO0126</td>\n",
              "      <td>106637</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>767</td>\n",
              "      <td>0.735294</td>\n",
              "      <td>TTC</td>\n",
              "      <td>CCGGTGTCCACGCCGCCGCTGATGACCAGGTCCG</td>\n",
              "      <td>CCGGTGTC</td>\n",
              "      <td>10</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2011</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO0135</td>\n",
              "      <td>127571</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>82</td>\n",
              "      <td>0.676471</td>\n",
              "      <td>TTC</td>\n",
              "      <td>GCGGACTCGTCCAGCTGGAACGGCGTGGTGGAAC</td>\n",
              "      <td>GCGGACTC</td>\n",
              "      <td>10</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2031</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO0136</td>\n",
              "      <td>128634</td>\n",
              "      <td>-1</td>\n",
              "      <td>1</td>\n",
              "      <td>364</td>\n",
              "      <td>0.764706</td>\n",
              "      <td>TTC</td>\n",
              "      <td>CTGGTCAGCCTGGCCCTGGCCCGGTTCACGCCGC</td>\n",
              "      <td>CTGGTCAG</td>\n",
              "      <td>10</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>1362 rows × 11 columns</p>\n",
              "</div>"
            ],
            "text/plain": [
              "      strain_name locus_tag  gene_loc  gene_strand  sgrna_strand  sgrna_loc  \\\n",
              "1196  NC_003888.3   SCO0088     74917           -1             1        836   \n",
              "2173  NC_003888.3   SCO0144    137197            1            -1        109   \n",
              "2176  NC_003888.3   SCO0144    137197            1            -1        282   \n",
              "764   NC_003888.3   SCO0062     47789           -1            -1        844   \n",
              "754   NC_003888.3   SCO0062     47789           -1             1        325   \n",
              "...           ...       ...       ...          ...           ...        ...   \n",
              "2144  NC_003888.3   SCO0142    134295           -1             1        601   \n",
              "879   NC_003888.3   SCO0067     54097            1            -1        313   \n",
              "1794  NC_003888.3   SCO0126    106637            1             1        767   \n",
              "2011  NC_003888.3   SCO0135    127571            1            -1         82   \n",
              "2031  NC_003888.3   SCO0136    128634           -1             1        364   \n",
              "\n",
              "            gc  pam                               sgrna sgrna_seed_sequence  \\\n",
              "1196  0.676471  TTC  ACATGGAGGCGGGAGTCGTCCGGTATCCCGACGG            ACATGGAG   \n",
              "2173  0.676471  TTC  AAAGGCAGCCTGACCATCCCGCTGGCCCACGTCC            AAAGGCAG   \n",
              "2176  0.588235  TTC  GAAGGCTGTCGTGATCGAACTCTCCGACGAGCAG            GAAGGCTG   \n",
              "764   0.647059  TTC  CGGATTTGTACAGGGCCCGGATCGCTCCCTGGCT            CGGATTTG   \n",
              "754   0.676471  TTC  TACTCACGCCTCGCCCTCGGCGTCCAGGAAGTGG            TACTCACG   \n",
              "...        ...  ...                                 ...                 ...   \n",
              "2144  0.705882  TTC  CTGGTCAGCGACACCACCCGGCACCAGCTGGTGG            CTGGTCAG   \n",
              "879   0.617647  TTC  GCCACGATCATCATGCCCGTCAAGAGGCAGGCGT            GCCACGAT   \n",
              "1794  0.735294  TTC  CCGGTGTCCACGCCGCCGCTGATGACCAGGTCCG            CCGGTGTC   \n",
              "2011  0.676471  TTC  GCGGACTCGTCCAGCTGGAACGGCGTGGTGGAAC            GCGGACTC   \n",
              "2031  0.764706  TTC  CTGGTCAGCCTGGCCCTGGCCCGGTTCACGCCGC            CTGGTCAG   \n",
              "\n",
              "      off_target_count  \n",
              "1196                 0  \n",
              "2173                 0  \n",
              "2176                 0  \n",
              "764                  0  \n",
              "754                  0  \n",
              "...                ...  \n",
              "2144                10  \n",
              "879                 10  \n",
              "1794                10  \n",
              "2011                10  \n",
              "2031                10  \n",
              "\n",
              "[1362 rows x 11 columns]"
            ]
          },
          "execution_count": 5,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "# Initialize SgRNAargs with desired parameters\n",
        "args = SgRNAargs(genome, \n",
        "                genes_to_KO,\n",
        "                step=['find', 'filter'],\n",
        "                gc_upper = gc_upper,\n",
        "                gc_lower = gc_lower,\n",
        "                off_target_seed = off_target_seed,\n",
        "                off_target_upper = off_target_upper,\n",
        "                cas_type=cas_type\n",
        "                )\n",
        "\n",
        "sgrna_df = extract_sgRNAs(args)\n",
        "sgrna_df"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 6,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>strain_name</th>\n",
              "      <th>locus_tag</th>\n",
              "      <th>gene_loc</th>\n",
              "      <th>gene_strand</th>\n",
              "      <th>sgrna_strand</th>\n",
              "      <th>sgrna_loc</th>\n",
              "      <th>gc</th>\n",
              "      <th>pam</th>\n",
              "      <th>sgrna</th>\n",
              "      <th>sgrna_seed_sequence</th>\n",
              "      <th>off_target_count</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>1196</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO0088</td>\n",
              "      <td>74917</td>\n",
              "      <td>-1</td>\n",
              "      <td>1</td>\n",
              "      <td>836</td>\n",
              "      <td>0.676471</td>\n",
              "      <td>TTC</td>\n",
              "      <td>ACATGGAGGCGGGAGTCGTCCGGTATCCCGACGG</td>\n",
              "      <td>ACATGGAG</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2173</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO0144</td>\n",
              "      <td>137197</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>109</td>\n",
              "      <td>0.676471</td>\n",
              "      <td>TTC</td>\n",
              "      <td>AAAGGCAGCCTGACCATCCCGCTGGCCCACGTCC</td>\n",
              "      <td>AAAGGCAG</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>764</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO0062</td>\n",
              "      <td>47789</td>\n",
              "      <td>-1</td>\n",
              "      <td>-1</td>\n",
              "      <td>844</td>\n",
              "      <td>0.647059</td>\n",
              "      <td>TTC</td>\n",
              "      <td>CGGATTTGTACAGGGCCCGGATCGCTCCCTGGCT</td>\n",
              "      <td>CGGATTTG</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>732</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO0061</td>\n",
              "      <td>46430</td>\n",
              "      <td>-1</td>\n",
              "      <td>1</td>\n",
              "      <td>882</td>\n",
              "      <td>0.588235</td>\n",
              "      <td>TTC</td>\n",
              "      <td>GATATACGCGTACGAGGTGGACGGCAACGGCAAC</td>\n",
              "      <td>GATATACG</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>792</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO0063</td>\n",
              "      <td>48889</td>\n",
              "      <td>-1</td>\n",
              "      <td>-1</td>\n",
              "      <td>288</td>\n",
              "      <td>0.735294</td>\n",
              "      <td>TTC</td>\n",
              "      <td>GAACTTGCCGACGTCGGCGAAGCGGCCGACGCCG</td>\n",
              "      <td>GAACTTGC</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>...</th>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>285</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO0015</td>\n",
              "      <td>16070</td>\n",
              "      <td>-1</td>\n",
              "      <td>-1</td>\n",
              "      <td>596</td>\n",
              "      <td>0.676471</td>\n",
              "      <td>TTC</td>\n",
              "      <td>TGCAAGTAGCCGAACACGGCCAGGCCGACGGCAG</td>\n",
              "      <td>TGCAAGTA</td>\n",
              "      <td>1</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1120</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO0083</td>\n",
              "      <td>69799</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>69</td>\n",
              "      <td>0.676471</td>\n",
              "      <td>TTC</td>\n",
              "      <td>ATCCGACACGATCCACGGCGACGTCCCCACACCG</td>\n",
              "      <td>ATCCGACA</td>\n",
              "      <td>1</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2140</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO0142</td>\n",
              "      <td>134295</td>\n",
              "      <td>-1</td>\n",
              "      <td>1</td>\n",
              "      <td>157</td>\n",
              "      <td>0.647059</td>\n",
              "      <td>TTC</td>\n",
              "      <td>TGGACGTTCTGCTGCGTCAACTGCCTGCACGCCC</td>\n",
              "      <td>TGGACGTT</td>\n",
              "      <td>1</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>251</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO0012</td>\n",
              "      <td>14044</td>\n",
              "      <td>-1</td>\n",
              "      <td>-1</td>\n",
              "      <td>114</td>\n",
              "      <td>0.764706</td>\n",
              "      <td>TTC</td>\n",
              "      <td>CAACGGAGCCTCCCGGCCCGGCTCCCGGTACACC</td>\n",
              "      <td>CAACGGAG</td>\n",
              "      <td>1</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>181</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO0006</td>\n",
              "      <td>9304</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>272</td>\n",
              "      <td>0.705882</td>\n",
              "      <td>TTC</td>\n",
              "      <td>TTGGCAAGGGCGTCGATAGCGGGGGTGCGCACGG</td>\n",
              "      <td>TTGGCAAG</td>\n",
              "      <td>1</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>100 rows × 11 columns</p>\n",
              "</div>"
            ],
            "text/plain": [
              "      strain_name locus_tag  gene_loc  gene_strand  sgrna_strand  sgrna_loc  \\\n",
              "1196  NC_003888.3   SCO0088     74917           -1             1        836   \n",
              "2173  NC_003888.3   SCO0144    137197            1            -1        109   \n",
              "764   NC_003888.3   SCO0062     47789           -1            -1        844   \n",
              "732   NC_003888.3   SCO0061     46430           -1             1        882   \n",
              "792   NC_003888.3   SCO0063     48889           -1            -1        288   \n",
              "...           ...       ...       ...          ...           ...        ...   \n",
              "285   NC_003888.3   SCO0015     16070           -1            -1        596   \n",
              "1120  NC_003888.3   SCO0083     69799            1             1         69   \n",
              "2140  NC_003888.3   SCO0142    134295           -1             1        157   \n",
              "251   NC_003888.3   SCO0012     14044           -1            -1        114   \n",
              "181   NC_003888.3   SCO0006      9304            1             1        272   \n",
              "\n",
              "            gc  pam                               sgrna sgrna_seed_sequence  \\\n",
              "1196  0.676471  TTC  ACATGGAGGCGGGAGTCGTCCGGTATCCCGACGG            ACATGGAG   \n",
              "2173  0.676471  TTC  AAAGGCAGCCTGACCATCCCGCTGGCCCACGTCC            AAAGGCAG   \n",
              "764   0.647059  TTC  CGGATTTGTACAGGGCCCGGATCGCTCCCTGGCT            CGGATTTG   \n",
              "732   0.588235  TTC  GATATACGCGTACGAGGTGGACGGCAACGGCAAC            GATATACG   \n",
              "792   0.735294  TTC  GAACTTGCCGACGTCGGCGAAGCGGCCGACGCCG            GAACTTGC   \n",
              "...        ...  ...                                 ...                 ...   \n",
              "285   0.676471  TTC  TGCAAGTAGCCGAACACGGCCAGGCCGACGGCAG            TGCAAGTA   \n",
              "1120  0.676471  TTC  ATCCGACACGATCCACGGCGACGTCCCCACACCG            ATCCGACA   \n",
              "2140  0.647059  TTC  TGGACGTTCTGCTGCGTCAACTGCCTGCACGCCC            TGGACGTT   \n",
              "251   0.764706  TTC  CAACGGAGCCTCCCGGCCCGGCTCCCGGTACACC            CAACGGAG   \n",
              "181   0.705882  TTC  TTGGCAAGGGCGTCGATAGCGGGGGTGCGCACGG            TTGGCAAG   \n",
              "\n",
              "      off_target_count  \n",
              "1196                 0  \n",
              "2173                 0  \n",
              "764                  0  \n",
              "732                  0  \n",
              "792                  0  \n",
              "...                ...  \n",
              "285                  1  \n",
              "1120                 1  \n",
              "2140                 1  \n",
              "251                  1  \n",
              "181                  1  \n",
              "\n",
              "[100 rows x 11 columns]"
            ]
          },
          "execution_count": 6,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "# Filter the DataFrame to retain only up to 5 sgRNA sequences per locus_tag\n",
        "filtered_df = sgrna_df.groupby('locus_tag').head(number_of_sgRNAs_per_group)[:100]\n",
        "filtered_df"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "## Output"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 7,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>strain_name</th>\n",
              "      <th>locus_tag</th>\n",
              "      <th>gene_loc</th>\n",
              "      <th>gene_strand</th>\n",
              "      <th>sgrna_strand</th>\n",
              "      <th>sgrna_loc</th>\n",
              "      <th>gc</th>\n",
              "      <th>pam</th>\n",
              "      <th>sgrna</th>\n",
              "      <th>sgrna_seed_sequence</th>\n",
              "      <th>off_target_count</th>\n",
              "      <th>Fwd Primer</th>\n",
              "      <th>Rev Primer</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>1196</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO0088</td>\n",
              "      <td>74917</td>\n",
              "      <td>-1</td>\n",
              "      <td>1</td>\n",
              "      <td>836</td>\n",
              "      <td>0.676471</td>\n",
              "      <td>TTC</td>\n",
              "      <td>ACATGGAGGCGGGAGTCGTCCGGTATCCCGACGG</td>\n",
              "      <td>ACATGGAG</td>\n",
              "      <td>0</td>\n",
              "      <td>ACATGGAGGCGGGAGTCGTCCGGTATCCCGACGGGTCGCCCGGCAA...</td>\n",
              "      <td>CCGTCGGGATACCGGACGACTCCCGCCTCCATGTGTTTCAATCCAC...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2173</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO0144</td>\n",
              "      <td>137197</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>109</td>\n",
              "      <td>0.676471</td>\n",
              "      <td>TTC</td>\n",
              "      <td>AAAGGCAGCCTGACCATCCCGCTGGCCCACGTCC</td>\n",
              "      <td>AAAGGCAG</td>\n",
              "      <td>0</td>\n",
              "      <td>AAAGGCAGCCTGACCATCCCGCTGGCCCACGTCCGTCGCCCGGCAA...</td>\n",
              "      <td>GGACGTGGGCCAGCGGGATGGTCAGGCTGCCTTTGTTTCAATCCAC...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>764</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO0062</td>\n",
              "      <td>47789</td>\n",
              "      <td>-1</td>\n",
              "      <td>-1</td>\n",
              "      <td>844</td>\n",
              "      <td>0.647059</td>\n",
              "      <td>TTC</td>\n",
              "      <td>CGGATTTGTACAGGGCCCGGATCGCTCCCTGGCT</td>\n",
              "      <td>CGGATTTG</td>\n",
              "      <td>0</td>\n",
              "      <td>CGGATTTGTACAGGGCCCGGATCGCTCCCTGGCTGTCGCCCGGCAA...</td>\n",
              "      <td>AGCCAGGGAGCGATCCGGGCCCTGTACAAATCCGGTTTCAATCCAC...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>732</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO0061</td>\n",
              "      <td>46430</td>\n",
              "      <td>-1</td>\n",
              "      <td>1</td>\n",
              "      <td>882</td>\n",
              "      <td>0.588235</td>\n",
              "      <td>TTC</td>\n",
              "      <td>GATATACGCGTACGAGGTGGACGGCAACGGCAAC</td>\n",
              "      <td>GATATACG</td>\n",
              "      <td>0</td>\n",
              "      <td>GATATACGCGTACGAGGTGGACGGCAACGGCAACGTCGCCCGGCAA...</td>\n",
              "      <td>GTTGCCGTTGCCGTCCACCTCGTACGCGTATATCGTTTCAATCCAC...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>792</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO0063</td>\n",
              "      <td>48889</td>\n",
              "      <td>-1</td>\n",
              "      <td>-1</td>\n",
              "      <td>288</td>\n",
              "      <td>0.735294</td>\n",
              "      <td>TTC</td>\n",
              "      <td>GAACTTGCCGACGTCGGCGAAGCGGCCGACGCCG</td>\n",
              "      <td>GAACTTGC</td>\n",
              "      <td>0</td>\n",
              "      <td>GAACTTGCCGACGTCGGCGAAGCGGCCGACGCCGGTCGCCCGGCAA...</td>\n",
              "      <td>CGGCGTCGGCCGCTTCGCCGACGTCGGCAAGTTCGTTTCAATCCAC...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>...</th>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>285</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO0015</td>\n",
              "      <td>16070</td>\n",
              "      <td>-1</td>\n",
              "      <td>-1</td>\n",
              "      <td>596</td>\n",
              "      <td>0.676471</td>\n",
              "      <td>TTC</td>\n",
              "      <td>TGCAAGTAGCCGAACACGGCCAGGCCGACGGCAG</td>\n",
              "      <td>TGCAAGTA</td>\n",
              "      <td>1</td>\n",
              "      <td>TGCAAGTAGCCGAACACGGCCAGGCCGACGGCAGGTCGCCCGGCAA...</td>\n",
              "      <td>CTGCCGTCGGCCTGGCCGTGTTCGGCTACTTGCAGTTTCAATCCAC...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1120</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO0083</td>\n",
              "      <td>69799</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>69</td>\n",
              "      <td>0.676471</td>\n",
              "      <td>TTC</td>\n",
              "      <td>ATCCGACACGATCCACGGCGACGTCCCCACACCG</td>\n",
              "      <td>ATCCGACA</td>\n",
              "      <td>1</td>\n",
              "      <td>ATCCGACACGATCCACGGCGACGTCCCCACACCGGTCGCCCGGCAA...</td>\n",
              "      <td>CGGTGTGGGGACGTCGCCGTGGATCGTGTCGGATGTTTCAATCCAC...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2140</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO0142</td>\n",
              "      <td>134295</td>\n",
              "      <td>-1</td>\n",
              "      <td>1</td>\n",
              "      <td>157</td>\n",
              "      <td>0.647059</td>\n",
              "      <td>TTC</td>\n",
              "      <td>TGGACGTTCTGCTGCGTCAACTGCCTGCACGCCC</td>\n",
              "      <td>TGGACGTT</td>\n",
              "      <td>1</td>\n",
              "      <td>TGGACGTTCTGCTGCGTCAACTGCCTGCACGCCCGTCGCCCGGCAA...</td>\n",
              "      <td>GGGCGTGCAGGCAGTTGACGCAGCAGAACGTCCAGTTTCAATCCAC...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>251</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO0012</td>\n",
              "      <td>14044</td>\n",
              "      <td>-1</td>\n",
              "      <td>-1</td>\n",
              "      <td>114</td>\n",
              "      <td>0.764706</td>\n",
              "      <td>TTC</td>\n",
              "      <td>CAACGGAGCCTCCCGGCCCGGCTCCCGGTACACC</td>\n",
              "      <td>CAACGGAG</td>\n",
              "      <td>1</td>\n",
              "      <td>CAACGGAGCCTCCCGGCCCGGCTCCCGGTACACCGTCGCCCGGCAA...</td>\n",
              "      <td>GGTGTACCGGGAGCCGGGCCGGGAGGCTCCGTTGGTTTCAATCCAC...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>181</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO0006</td>\n",
              "      <td>9304</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>272</td>\n",
              "      <td>0.705882</td>\n",
              "      <td>TTC</td>\n",
              "      <td>TTGGCAAGGGCGTCGATAGCGGGGGTGCGCACGG</td>\n",
              "      <td>TTGGCAAG</td>\n",
              "      <td>1</td>\n",
              "      <td>TTGGCAAGGGCGTCGATAGCGGGGGTGCGCACGGGTCGCCCGGCAA...</td>\n",
              "      <td>CCGTGCGCACCCCCGCTATCGACGCCCTTGCCAAGTTTCAATCCAC...</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>100 rows × 13 columns</p>\n",
              "</div>"
            ],
            "text/plain": [
              "      strain_name locus_tag  gene_loc  gene_strand  sgrna_strand  sgrna_loc  \\\n",
              "1196  NC_003888.3   SCO0088     74917           -1             1        836   \n",
              "2173  NC_003888.3   SCO0144    137197            1            -1        109   \n",
              "764   NC_003888.3   SCO0062     47789           -1            -1        844   \n",
              "732   NC_003888.3   SCO0061     46430           -1             1        882   \n",
              "792   NC_003888.3   SCO0063     48889           -1            -1        288   \n",
              "...           ...       ...       ...          ...           ...        ...   \n",
              "285   NC_003888.3   SCO0015     16070           -1            -1        596   \n",
              "1120  NC_003888.3   SCO0083     69799            1             1         69   \n",
              "2140  NC_003888.3   SCO0142    134295           -1             1        157   \n",
              "251   NC_003888.3   SCO0012     14044           -1            -1        114   \n",
              "181   NC_003888.3   SCO0006      9304            1             1        272   \n",
              "\n",
              "            gc  pam                               sgrna sgrna_seed_sequence  \\\n",
              "1196  0.676471  TTC  ACATGGAGGCGGGAGTCGTCCGGTATCCCGACGG            ACATGGAG   \n",
              "2173  0.676471  TTC  AAAGGCAGCCTGACCATCCCGCTGGCCCACGTCC            AAAGGCAG   \n",
              "764   0.647059  TTC  CGGATTTGTACAGGGCCCGGATCGCTCCCTGGCT            CGGATTTG   \n",
              "732   0.588235  TTC  GATATACGCGTACGAGGTGGACGGCAACGGCAAC            GATATACG   \n",
              "792   0.735294  TTC  GAACTTGCCGACGTCGGCGAAGCGGCCGACGCCG            GAACTTGC   \n",
              "...        ...  ...                                 ...                 ...   \n",
              "285   0.676471  TTC  TGCAAGTAGCCGAACACGGCCAGGCCGACGGCAG            TGCAAGTA   \n",
              "1120  0.676471  TTC  ATCCGACACGATCCACGGCGACGTCCCCACACCG            ATCCGACA   \n",
              "2140  0.647059  TTC  TGGACGTTCTGCTGCGTCAACTGCCTGCACGCCC            TGGACGTT   \n",
              "251   0.764706  TTC  CAACGGAGCCTCCCGGCCCGGCTCCCGGTACACC            CAACGGAG   \n",
              "181   0.705882  TTC  TTGGCAAGGGCGTCGATAGCGGGGGTGCGCACGG            TTGGCAAG   \n",
              "\n",
              "      off_target_count                                         Fwd Primer  \\\n",
              "1196                 0  ACATGGAGGCGGGAGTCGTCCGGTATCCCGACGGGTCGCCCGGCAA...   \n",
              "2173                 0  AAAGGCAGCCTGACCATCCCGCTGGCCCACGTCCGTCGCCCGGCAA...   \n",
              "764                  0  CGGATTTGTACAGGGCCCGGATCGCTCCCTGGCTGTCGCCCGGCAA...   \n",
              "732                  0  GATATACGCGTACGAGGTGGACGGCAACGGCAACGTCGCCCGGCAA...   \n",
              "792                  0  GAACTTGCCGACGTCGGCGAAGCGGCCGACGCCGGTCGCCCGGCAA...   \n",
              "...                ...                                                ...   \n",
              "285                  1  TGCAAGTAGCCGAACACGGCCAGGCCGACGGCAGGTCGCCCGGCAA...   \n",
              "1120                 1  ATCCGACACGATCCACGGCGACGTCCCCACACCGGTCGCCCGGCAA...   \n",
              "2140                 1  TGGACGTTCTGCTGCGTCAACTGCCTGCACGCCCGTCGCCCGGCAA...   \n",
              "251                  1  CAACGGAGCCTCCCGGCCCGGCTCCCGGTACACCGTCGCCCGGCAA...   \n",
              "181                  1  TTGGCAAGGGCGTCGATAGCGGGGGTGCGCACGGGTCGCCCGGCAA...   \n",
              "\n",
              "                                             Rev Primer  \n",
              "1196  CCGTCGGGATACCGGACGACTCCCGCCTCCATGTGTTTCAATCCAC...  \n",
              "2173  GGACGTGGGCCAGCGGGATGGTCAGGCTGCCTTTGTTTCAATCCAC...  \n",
              "764   AGCCAGGGAGCGATCCGGGCCCTGTACAAATCCGGTTTCAATCCAC...  \n",
              "732   GTTGCCGTTGCCGTCCACCTCGTACGCGTATATCGTTTCAATCCAC...  \n",
              "792   CGGCGTCGGCCGCTTCGCCGACGTCGGCAAGTTCGTTTCAATCCAC...  \n",
              "...                                                 ...  \n",
              "285   CTGCCGTCGGCCTGGCCGTGTTCGGCTACTTGCAGTTTCAATCCAC...  \n",
              "1120  CGGTGTGGGGACGTCGCCGTGGATCGTGTCGGATGTTTCAATCCAC...  \n",
              "2140  GGGCGTGCAGGCAGTTGACGCAGCAGAACGTCCAGTTTCAATCCAC...  \n",
              "251   GGTGTACCGGGAGCCGGGCCGGGAGGCTCCGTTGGTTTCAATCCAC...  \n",
              "181   CCGTGCGCACCCCCGCTATCGACGCCCTTGCCAAGTTTCAATCCAC...  \n",
              "\n",
              "[100 rows x 13 columns]"
            ]
          },
          "execution_count": 7,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "filtered_df_w_primers = generate_cas3_protospacer_primers(filtered_df,\n",
        "                                                          fwd_overhang= forward_protospacer_overhang, \n",
        "                                                          rev_overhang=reverse_protospacer_overhang)\n",
        "filtered_df_w_primers"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 8,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/plain": [
              "100"
            ]
          },
          "execution_count": 8,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "amplicons = cas3_plasmid_pcrs(clean_plasmid, \n",
        "                              filtered_df, \n",
        "                              universal_fwd_seq=forward_backbone_overhang, \n",
        "                              universal_rev_seq=reverse_backbone_overhang\n",
        "                              )\n",
        "\n",
        "len(amplicons)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 9,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/plain": [
              "100"
            ]
          },
          "execution_count": 9,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "amplicons = cas3_plasmid_pcrs(clean_plasmid, filtered_df)\n",
        "\n",
        "len(amplicons)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 10,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/plain": [
              "100"
            ]
          },
          "execution_count": 10,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "# cut plasmid and incorporate amplicons\n",
        "assembled_cas3_plasmids = assemble_cas3_plasmids(clean_plasmid, amplicons)\n",
        "len(assembled_cas3_plasmids)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 11,
      "metadata": {},
      "outputs": [],
      "source": [
        "# Constructing a meaningful name, ID, and description for the assembled plasmid using user input\n",
        "targeting_info = []\n",
        "for index, row in filtered_df.iterrows():\n",
        "    formatted_str = f\"pCas3_{row['locus_tag']}({row['sgrna_loc']})\"\n",
        "    targeting_info.append(formatted_str)\n",
        "\n",
        "for i in range(len(assembled_cas3_plasmids)):\n",
        "    assembled_cas3_plasmids[i].name = f'{targeting_info[i]}'\n",
        "    assembled_cas3_plasmids[i].id = assembled_cas3_plasmids[i].name  \n",
        "    assembled_cas3_plasmids[i].description = f'CRISPR-Cas3 targeting {\", \".join(genes_to_KO)} for single gene knockout, assembled using StreptoCAD.'\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 12,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/plain": [
              "'pCas3_SCO0088(836)'"
            ]
          },
          "execution_count": 12,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "assembled_cas3_plasmids[0].name"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 13,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/plain": [
              "'CRISPR-Cas3 targeting SCO0001, SCO0002, SCO0003, SCO0004, SCO0005, SCO0006, SCO0007, SCO0008, SCO0009, SCO0010, SCO0011, SCO0012, SCO0013, SCO0014, SCO0015, SCO0016, SCO0017, SCO0018, SCO0019, SCO0020, SCO0021, SCO0022, SCO0025, SCO0026, SCO0027, SCO0028, SCO0029, SCO0030, SCO0031, SCO0033, SCO0034, SCO0035, SCO0036, SCO0037, SCO0038, SCO0039, SCO0040, SCO0041, SCO0042, SCO0043, SCO0044, SCO0045, SCO0046, SCO0047, SCO0048, SCO0049, SCO0050, SCO0051, SCO0053, SCO0055, SCO0056, SCO0057, SCO0058, SCO0059, SCO0060, SCO0061, SCO0062, SCO0063, SCO0064, SCO0065, SCO0066, SCO0067, SCO0068, SCO0069, SCO0070, SCO0071, SCO0072, SCO0073, SCO0074, SCO0075, SCO0076, SCO0079, SCO0080, SCO0081, SCO0083, SCO0084, SCO0085, SCO0086, SCO0087, SCO0088, SCO0089, SCO0090, SCO0091, SCO0092, SCO0095, SCO0096, SCO0097, SCO0098, SCO0099, SCO0101, SCO0102, SCO0103, SCO0104, SCO0105, SCO0106, SCO0107, SCO0108, SCO0109, SCO0110, SCO0111, SCO0112, SCO0113, SCO0114, SCO0116, SCO0117, SCO0118, SCO0119, SCO0120, SCO0121, SCO0122, SCO0123, SCO0124, SCO0125, SCO0126, SCO0127, SCO0128, SCO0129, SCO0130, SCO0131, SCO0132, SCO0133, SCO0134, SCO0135, SCO0136, SCO0137, SCO0138, SCO0139, SCO0140, SCO0141, SCO0142, SCO0143, SCO0144, SCO0145, SCO0146, SCO0147, SCO0148, SCO0149, SCO0150, SCO0151, SCO0152, SCO0153, SCO0154, SCO0155, SCO0156, SCO0157, SCO0158, SCO0159, SCO0160, SCO0161, SCO0162 for single gene knockout, assembled using StreptoCAD.'"
            ]
          },
          "execution_count": 13,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "assembled_cas3_plasmids[0].description"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 14,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/plain": [
              "'pCas3_SCO0088(836)'"
            ]
          },
          "execution_count": 14,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "assembled_cas3_plasmids[0].id"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 15,
      "metadata": {},
      "outputs": [],
      "source": [
        "print_plasmids = False\n",
        "\n",
        "if print_plasmids: \n",
        "    for vector in assembled_cas3_plasmids: \n",
        "        vector.write(f\"../../data/plasmids/sgRNA_plasmids_pCRISPR–Cas3/{vector.id}.gb\")"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "### IDT primers"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 16,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/plain": [
              "[Dseqrecord(-52),\n",
              " Dseqrecord(-54),\n",
              " Dseqrecord(-52),\n",
              " Dseqrecord(-54),\n",
              " Dseqrecord(-52),\n",
              " Dseqrecord(-54),\n",
              " Dseqrecord(-52),\n",
              " Dseqrecord(-54),\n",
              " Dseqrecord(-52),\n",
              " Dseqrecord(-54),\n",
              " Dseqrecord(-52),\n",
              " Dseqrecord(-54),\n",
              " Dseqrecord(-52),\n",
              " Dseqrecord(-54),\n",
              " Dseqrecord(-52),\n",
              " Dseqrecord(-54),\n",
              " Dseqrecord(-52),\n",
              " Dseqrecord(-54),\n",
              " Dseqrecord(-52),\n",
              " Dseqrecord(-54),\n",
              " Dseqrecord(-52),\n",
              " Dseqrecord(-54),\n",
              " Dseqrecord(-52),\n",
              " Dseqrecord(-54),\n",
              " Dseqrecord(-52),\n",
              " Dseqrecord(-54),\n",
              " Dseqrecord(-52),\n",
              " Dseqrecord(-54),\n",
              " Dseqrecord(-52),\n",
              " Dseqrecord(-54),\n",
              " Dseqrecord(-52),\n",
              " Dseqrecord(-54),\n",
              " Dseqrecord(-52),\n",
              " Dseqrecord(-54),\n",
              " Dseqrecord(-52),\n",
              " Dseqrecord(-54),\n",
              " Dseqrecord(-52),\n",
              " Dseqrecord(-54),\n",
              " Dseqrecord(-52),\n",
              " Dseqrecord(-54),\n",
              " Dseqrecord(-52),\n",
              " Dseqrecord(-54),\n",
              " Dseqrecord(-52),\n",
              " Dseqrecord(-54),\n",
              " Dseqrecord(-52),\n",
              " Dseqrecord(-54),\n",
              " Dseqrecord(-52),\n",
              " Dseqrecord(-54),\n",
              " Dseqrecord(-52),\n",
              " Dseqrecord(-54),\n",
              " Dseqrecord(-52),\n",
              " Dseqrecord(-54),\n",
              " Dseqrecord(-52),\n",
              " Dseqrecord(-54),\n",
              " Dseqrecord(-52),\n",
              " Dseqrecord(-54),\n",
              " Dseqrecord(-52),\n",
              " Dseqrecord(-54),\n",
              " Dseqrecord(-52),\n",
              " Dseqrecord(-54),\n",
              " Dseqrecord(-52),\n",
              " Dseqrecord(-54),\n",
              " Dseqrecord(-52),\n",
              " Dseqrecord(-54),\n",
              " Dseqrecord(-52),\n",
              " Dseqrecord(-54),\n",
              " Dseqrecord(-52),\n",
              " Dseqrecord(-54),\n",
              " Dseqrecord(-52),\n",
              " Dseqrecord(-54),\n",
              " Dseqrecord(-52),\n",
              " Dseqrecord(-54),\n",
              " Dseqrecord(-52),\n",
              " Dseqrecord(-54),\n",
              " Dseqrecord(-52),\n",
              " Dseqrecord(-54),\n",
              " Dseqrecord(-52),\n",
              " Dseqrecord(-54),\n",
              " Dseqrecord(-52),\n",
              " Dseqrecord(-54),\n",
              " Dseqrecord(-52),\n",
              " Dseqrecord(-54),\n",
              " Dseqrecord(-52),\n",
              " Dseqrecord(-54),\n",
              " Dseqrecord(-52),\n",
              " Dseqrecord(-54),\n",
              " Dseqrecord(-52),\n",
              " Dseqrecord(-54),\n",
              " Dseqrecord(-52),\n",
              " Dseqrecord(-54),\n",
              " Dseqrecord(-52),\n",
              " Dseqrecord(-54),\n",
              " Dseqrecord(-52),\n",
              " Dseqrecord(-54),\n",
              " Dseqrecord(-52),\n",
              " Dseqrecord(-54),\n",
              " Dseqrecord(-52),\n",
              " Dseqrecord(-54),\n",
              " Dseqrecord(-52),\n",
              " Dseqrecord(-54),\n",
              " Dseqrecord(-52),\n",
              " Dseqrecord(-54),\n",
              " Dseqrecord(-52),\n",
              " Dseqrecord(-54),\n",
              " Dseqrecord(-52),\n",
              " Dseqrecord(-54),\n",
              " Dseqrecord(-52),\n",
              " Dseqrecord(-54),\n",
              " Dseqrecord(-52),\n",
              " Dseqrecord(-54),\n",
              " Dseqrecord(-52),\n",
              " Dseqrecord(-54),\n",
              " Dseqrecord(-52),\n",
              " Dseqrecord(-54),\n",
              " Dseqrecord(-52),\n",
              " Dseqrecord(-54),\n",
              " Dseqrecord(-52),\n",
              " Dseqrecord(-54),\n",
              " Dseqrecord(-52),\n",
              " Dseqrecord(-54),\n",
              " Dseqrecord(-52),\n",
              " Dseqrecord(-54),\n",
              " Dseqrecord(-52),\n",
              " Dseqrecord(-54),\n",
              " Dseqrecord(-52),\n",
              " Dseqrecord(-54),\n",
              " Dseqrecord(-52),\n",
              " Dseqrecord(-54),\n",
              " Dseqrecord(-52),\n",
              " Dseqrecord(-54),\n",
              " Dseqrecord(-52),\n",
              " Dseqrecord(-54),\n",
              " Dseqrecord(-52),\n",
              " Dseqrecord(-54),\n",
              " Dseqrecord(-52),\n",
              " Dseqrecord(-54),\n",
              " Dseqrecord(-52),\n",
              " Dseqrecord(-54),\n",
              " Dseqrecord(-52),\n",
              " Dseqrecord(-54),\n",
              " Dseqrecord(-52),\n",
              " Dseqrecord(-54),\n",
              " Dseqrecord(-52),\n",
              " Dseqrecord(-54),\n",
              " Dseqrecord(-52),\n",
              " Dseqrecord(-54),\n",
              " Dseqrecord(-52),\n",
              " Dseqrecord(-54),\n",
              " Dseqrecord(-52),\n",
              " Dseqrecord(-54),\n",
              " Dseqrecord(-52),\n",
              " Dseqrecord(-54),\n",
              " Dseqrecord(-52),\n",
              " Dseqrecord(-54),\n",
              " Dseqrecord(-52),\n",
              " Dseqrecord(-54),\n",
              " Dseqrecord(-52),\n",
              " Dseqrecord(-54),\n",
              " Dseqrecord(-52),\n",
              " Dseqrecord(-54),\n",
              " Dseqrecord(-52),\n",
              " Dseqrecord(-54),\n",
              " Dseqrecord(-52),\n",
              " Dseqrecord(-54),\n",
              " Dseqrecord(-52),\n",
              " Dseqrecord(-54),\n",
              " Dseqrecord(-52),\n",
              " Dseqrecord(-54),\n",
              " Dseqrecord(-52),\n",
              " Dseqrecord(-54),\n",
              " Dseqrecord(-52),\n",
              " Dseqrecord(-54),\n",
              " Dseqrecord(-52),\n",
              " Dseqrecord(-54),\n",
              " Dseqrecord(-52),\n",
              " Dseqrecord(-54),\n",
              " Dseqrecord(-52),\n",
              " Dseqrecord(-54),\n",
              " Dseqrecord(-52),\n",
              " Dseqrecord(-54),\n",
              " Dseqrecord(-52),\n",
              " Dseqrecord(-54),\n",
              " Dseqrecord(-52),\n",
              " Dseqrecord(-54),\n",
              " Dseqrecord(-52),\n",
              " Dseqrecord(-54),\n",
              " Dseqrecord(-52),\n",
              " Dseqrecord(-54),\n",
              " Dseqrecord(-52),\n",
              " Dseqrecord(-54),\n",
              " Dseqrecord(-52),\n",
              " Dseqrecord(-54),\n",
              " Dseqrecord(-52),\n",
              " Dseqrecord(-54),\n",
              " Dseqrecord(-52),\n",
              " Dseqrecord(-54),\n",
              " Dseqrecord(-52),\n",
              " Dseqrecord(-54),\n",
              " Dseqrecord(-52),\n",
              " Dseqrecord(-54)]"
            ]
          },
          "execution_count": 16,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "primer_records = make_primer_records(filtered_df_w_primers)\n",
        "primer_records"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 17,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>Name</th>\n",
              "      <th>Sequence</th>\n",
              "      <th>Concentration</th>\n",
              "      <th>Purification</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>SCO0088_loc_836_fwd</td>\n",
              "      <td>ACATGGAGGCGGGAGTCGTCCGGTATCCCGACGGGTCGCCCGGCAA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>SCO0088_loc_836_rev</td>\n",
              "      <td>CCGTCGGGATACCGGACGACTCCCGCCTCCATGTGTTTCAATCCAC...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>SCO0144_loc_109_fwd</td>\n",
              "      <td>AAAGGCAGCCTGACCATCCCGCTGGCCCACGTCCGTCGCCCGGCAA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>SCO0144_loc_109_rev</td>\n",
              "      <td>GGACGTGGGCCAGCGGGATGGTCAGGCTGCCTTTGTTTCAATCCAC...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>SCO0062_loc_844_fwd</td>\n",
              "      <td>CGGATTTGTACAGGGCCCGGATCGCTCCCTGGCTGTCGCCCGGCAA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>...</th>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>195</th>\n",
              "      <td>SCO0142_loc_157_rev</td>\n",
              "      <td>GGGCGTGCAGGCAGTTGACGCAGCAGAACGTCCAGTTTCAATCCAC...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>196</th>\n",
              "      <td>SCO0012_loc_114_fwd</td>\n",
              "      <td>CAACGGAGCCTCCCGGCCCGGCTCCCGGTACACCGTCGCCCGGCAA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>197</th>\n",
              "      <td>SCO0012_loc_114_rev</td>\n",
              "      <td>GGTGTACCGGGAGCCGGGCCGGGAGGCTCCGTTGGTTTCAATCCAC...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>198</th>\n",
              "      <td>SCO0006_loc_272_fwd</td>\n",
              "      <td>TTGGCAAGGGCGTCGATAGCGGGGGTGCGCACGGGTCGCCCGGCAA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>199</th>\n",
              "      <td>SCO0006_loc_272_rev</td>\n",
              "      <td>CCGTGCGCACCCCCGCTATCGACGCCCTTGCCAAGTTTCAATCCAC...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>200 rows × 4 columns</p>\n",
              "</div>"
            ],
            "text/plain": [
              "                    Name                                           Sequence  \\\n",
              "0    SCO0088_loc_836_fwd  ACATGGAGGCGGGAGTCGTCCGGTATCCCGACGGGTCGCCCGGCAA...   \n",
              "1    SCO0088_loc_836_rev  CCGTCGGGATACCGGACGACTCCCGCCTCCATGTGTTTCAATCCAC...   \n",
              "2    SCO0144_loc_109_fwd  AAAGGCAGCCTGACCATCCCGCTGGCCCACGTCCGTCGCCCGGCAA...   \n",
              "3    SCO0144_loc_109_rev  GGACGTGGGCCAGCGGGATGGTCAGGCTGCCTTTGTTTCAATCCAC...   \n",
              "4    SCO0062_loc_844_fwd  CGGATTTGTACAGGGCCCGGATCGCTCCCTGGCTGTCGCCCGGCAA...   \n",
              "..                   ...                                                ...   \n",
              "195  SCO0142_loc_157_rev  GGGCGTGCAGGCAGTTGACGCAGCAGAACGTCCAGTTTCAATCCAC...   \n",
              "196  SCO0012_loc_114_fwd  CAACGGAGCCTCCCGGCCCGGCTCCCGGTACACCGTCGCCCGGCAA...   \n",
              "197  SCO0012_loc_114_rev  GGTGTACCGGGAGCCGGGCCGGGAGGCTCCGTTGGTTTCAATCCAC...   \n",
              "198  SCO0006_loc_272_fwd  TTGGCAAGGGCGTCGATAGCGGGGGTGCGCACGGGTCGCCCGGCAA...   \n",
              "199  SCO0006_loc_272_rev  CCGTGCGCACCCCCGCTATCGACGCCCTTGCCAAGTTTCAATCCAC...   \n",
              "\n",
              "    Concentration Purification  \n",
              "0            25nm          STD  \n",
              "1            25nm          STD  \n",
              "2            25nm          STD  \n",
              "3            25nm          STD  \n",
              "4            25nm          STD  \n",
              "..            ...          ...  \n",
              "195          25nm          STD  \n",
              "196          25nm          STD  \n",
              "197          25nm          STD  \n",
              "198          25nm          STD  \n",
              "199          25nm          STD  \n",
              "\n",
              "[200 rows x 4 columns]"
            ]
          },
          "execution_count": 17,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "idt_primers_cas3=primers_to_IDT(primer_records)\n",
        "idt_primers_cas3"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": []
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "# In frame deletion"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 18,
      "metadata": {},
      "outputs": [],
      "source": [
        "in_frame_deletion = False \n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 19,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/plain": [
              "(Dseqrecord(-12770),)"
            ]
          },
          "execution_count": 19,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "Dseqrecord(assembled_cas3_plasmids[3], circular=True).cut(EcoRI)"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "#### Getting repair templates"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 20,
      "metadata": {},
      "outputs": [],
      "source": [
        "if in_frame_deletion:\n",
        "    \n",
        "\n",
        "    # Make repair templates\n",
        "    repair_templates_data = find_up_dw_repair_templates(genome, \n",
        "                                                        genes_to_KO, \n",
        "                                                        target_tm=melting_temperature, \n",
        "                                                        primer_tm_kwargs={'conc':primer_concentration, 'prodcode':chosen_polymerase} , \n",
        "                                                        repair_length=repair_length)\n",
        "\n",
        "    # Digest the plasmids\n",
        "    processed_records = [sorted(Dseqrecord(record, circular=True).cut(Enzyme_for_repair_template_integration), key=lambda x: len(x), reverse=True)[0] for record in assembled_cas3_plasmids]\n",
        "\n",
        "    # Rename them appropriately\n",
        "    for i in range(len(processed_records)):\n",
        "        processed_records[i].name = assembled_cas3_plasmids[i].name\n",
        "        print(processed_records[i].name)\n",
        "\n",
        "    # Assembly \n",
        "    assembly_data = assemble_multiple_plasmids_with_repair_templates_for_deletion(genes_to_KO, \n",
        "                                                                                processed_records, \n",
        "                                                                                repair_templates_data, \n",
        "                                                                                overlap=overlap_length)\n",
        "    # updating the primer names to something systematic.\n",
        "    update_primer_names(assembly_data)\n",
        "\n",
        "    # Parse through the primer df\n",
        "    primer_df = create_primer_df_from_dict(assembly_data)\n",
        "\n",
        "    #Find_unique primers\n",
        "    unique_df = primer_df.drop_duplicates(keep='first')\n",
        "    idt_df = create_idt_order_dataframe(unique_df, concentration=\"25nm\", purification=\"STD\")\n",
        "\n",
        "    # Contigs\n",
        "    assembled_contigs = []\n",
        "    for data in assembly_data: \n",
        "        contig_record = data['contig']\n",
        "        contig_record.id = f\"{data['name']}_w_rep\"\n",
        "        contig_record.name = f\"{data['name']}_w_rep\"\n",
        "        assembled_contigs.append(contig_record)\n",
        "\n",
        "\n",
        "    # print our df that contains primers for the adding repair templates to the vecors.         \n",
        "    print(unique_df)"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "Check for restriction enzyme sites"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 21,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>plasmid_name</th>\n",
              "      <th>date</th>\n",
              "      <th>original_plasmid</th>\n",
              "      <th>integration</th>\n",
              "      <th>size</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>pCas3_SCO0088(836)</td>\n",
              "      <td>2025-06-10</td>\n",
              "      <td>pCRISPR-Cas3</td>\n",
              "      <td>sgRNA_SCO0088(836)</td>\n",
              "      <td>12766</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>pCas3_SCO0144(109)</td>\n",
              "      <td>2025-06-10</td>\n",
              "      <td>pCRISPR-Cas3</td>\n",
              "      <td>sgRNA_SCO0144(109)</td>\n",
              "      <td>12766</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>pCas3_SCO0062(844)</td>\n",
              "      <td>2025-06-10</td>\n",
              "      <td>pCRISPR-Cas3</td>\n",
              "      <td>sgRNA_SCO0062(844)</td>\n",
              "      <td>12766</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>pCas3_SCO0061(882)</td>\n",
              "      <td>2025-06-10</td>\n",
              "      <td>pCRISPR-Cas3</td>\n",
              "      <td>sgRNA_SCO0061(882)</td>\n",
              "      <td>12766</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>pCas3_SCO0063(288)</td>\n",
              "      <td>2025-06-10</td>\n",
              "      <td>pCRISPR-Cas3</td>\n",
              "      <td>sgRNA_SCO0063(288)</td>\n",
              "      <td>12766</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>...</th>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>95</th>\n",
              "      <td>pCas3_SCO0015(596)</td>\n",
              "      <td>2025-06-10</td>\n",
              "      <td>pCRISPR-Cas3</td>\n",
              "      <td>sgRNA_SCO0015(596)</td>\n",
              "      <td>12766</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>96</th>\n",
              "      <td>pCas3_SCO0083(69)</td>\n",
              "      <td>2025-06-10</td>\n",
              "      <td>pCRISPR-Cas3</td>\n",
              "      <td>sgRNA_SCO0083(69)</td>\n",
              "      <td>12766</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>97</th>\n",
              "      <td>pCas3_SCO0142(157)</td>\n",
              "      <td>2025-06-10</td>\n",
              "      <td>pCRISPR-Cas3</td>\n",
              "      <td>sgRNA_SCO0142(157)</td>\n",
              "      <td>12766</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>98</th>\n",
              "      <td>pCas3_SCO0012(114)</td>\n",
              "      <td>2025-06-10</td>\n",
              "      <td>pCRISPR-Cas3</td>\n",
              "      <td>sgRNA_SCO0012(114)</td>\n",
              "      <td>12766</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>99</th>\n",
              "      <td>pCas3_SCO0006(272)</td>\n",
              "      <td>2025-06-10</td>\n",
              "      <td>pCRISPR-Cas3</td>\n",
              "      <td>sgRNA_SCO0006(272)</td>\n",
              "      <td>12766</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>100 rows × 5 columns</p>\n",
              "</div>"
            ],
            "text/plain": [
              "          plasmid_name        date original_plasmid         integration   size\n",
              "0   pCas3_SCO0088(836)  2025-06-10     pCRISPR-Cas3  sgRNA_SCO0088(836)  12766\n",
              "1   pCas3_SCO0144(109)  2025-06-10     pCRISPR-Cas3  sgRNA_SCO0144(109)  12766\n",
              "2   pCas3_SCO0062(844)  2025-06-10     pCRISPR-Cas3  sgRNA_SCO0062(844)  12766\n",
              "3   pCas3_SCO0061(882)  2025-06-10     pCRISPR-Cas3  sgRNA_SCO0061(882)  12766\n",
              "4   pCas3_SCO0063(288)  2025-06-10     pCRISPR-Cas3  sgRNA_SCO0063(288)  12766\n",
              "..                 ...         ...              ...                 ...    ...\n",
              "95  pCas3_SCO0015(596)  2025-06-10     pCRISPR-Cas3  sgRNA_SCO0015(596)  12766\n",
              "96   pCas3_SCO0083(69)  2025-06-10     pCRISPR-Cas3   sgRNA_SCO0083(69)  12766\n",
              "97  pCas3_SCO0142(157)  2025-06-10     pCRISPR-Cas3  sgRNA_SCO0142(157)  12766\n",
              "98  pCas3_SCO0012(114)  2025-06-10     pCRISPR-Cas3  sgRNA_SCO0012(114)  12766\n",
              "99  pCas3_SCO0006(272)  2025-06-10     pCRISPR-Cas3  sgRNA_SCO0006(272)  12766\n",
              "\n",
              "[100 rows x 5 columns]"
            ]
          },
          "execution_count": 21,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "if in_frame_deletion:\n",
        "    integration_names = filtered_df.apply(lambda row: f\"sgRNA_{row['locus_tag']}({row['sgrna_loc']})\", axis=1).tolist()\n",
        "    plasmid_metadata_df = extract_metadata_to_dataframe(assembled_contigs,\n",
        "                                                        clean_plasmid,\n",
        "                                                        integration_names)\n",
        "    workflow_df = determine_workflow_order_for_plasmids(assembled_cas3_plasmids, \n",
        "                                                        assembled_contigs,\n",
        "                                                        [Enzyme_for_repair_template_integration,], [ \"NcoI\",\"BstBI\" ]) \n",
        "    \n",
        "    plasmid_metadata_df = pd.merge(plasmid_metadata_df, workflow_df, on='plasmid_name', how='inner')  # or 'left'/'right' depending on what you need\n",
        "\n",
        "else: \n",
        "    integration_names = filtered_df.apply(lambda row: f\"sgRNA_{row['locus_tag']}({row['sgrna_loc']})\", axis=1).tolist()\n",
        "    plasmid_metadata_df = extract_metadata_to_dataframe(assembled_cas3_plasmids,\n",
        "                                                        clean_plasmid,\n",
        "                                                        integration_names)\n",
        "\n",
        "plasmid_metadata_df"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 22,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>Name</th>\n",
              "      <th>Sequence</th>\n",
              "      <th>Concentration</th>\n",
              "      <th>Purification</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>SCO0088_loc_836_fwd</td>\n",
              "      <td>ACATGGAGGCGGGAGTCGTCCGGTATCCCGACGGGTCGCCCGGCAA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>SCO0088_loc_836_rev</td>\n",
              "      <td>CCGTCGGGATACCGGACGACTCCCGCCTCCATGTGTTTCAATCCAC...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>SCO0144_loc_109_fwd</td>\n",
              "      <td>AAAGGCAGCCTGACCATCCCGCTGGCCCACGTCCGTCGCCCGGCAA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>SCO0144_loc_109_rev</td>\n",
              "      <td>GGACGTGGGCCAGCGGGATGGTCAGGCTGCCTTTGTTTCAATCCAC...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>SCO0062_loc_844_fwd</td>\n",
              "      <td>CGGATTTGTACAGGGCCCGGATCGCTCCCTGGCTGTCGCCCGGCAA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>...</th>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>195</th>\n",
              "      <td>SCO0142_loc_157_rev</td>\n",
              "      <td>GGGCGTGCAGGCAGTTGACGCAGCAGAACGTCCAGTTTCAATCCAC...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>196</th>\n",
              "      <td>SCO0012_loc_114_fwd</td>\n",
              "      <td>CAACGGAGCCTCCCGGCCCGGCTCCCGGTACACCGTCGCCCGGCAA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>197</th>\n",
              "      <td>SCO0012_loc_114_rev</td>\n",
              "      <td>GGTGTACCGGGAGCCGGGCCGGGAGGCTCCGTTGGTTTCAATCCAC...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>198</th>\n",
              "      <td>SCO0006_loc_272_fwd</td>\n",
              "      <td>TTGGCAAGGGCGTCGATAGCGGGGGTGCGCACGGGTCGCCCGGCAA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>199</th>\n",
              "      <td>SCO0006_loc_272_rev</td>\n",
              "      <td>CCGTGCGCACCCCCGCTATCGACGCCCTTGCCAAGTTTCAATCCAC...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>200 rows × 4 columns</p>\n",
              "</div>"
            ],
            "text/plain": [
              "                    Name                                           Sequence  \\\n",
              "0    SCO0088_loc_836_fwd  ACATGGAGGCGGGAGTCGTCCGGTATCCCGACGGGTCGCCCGGCAA...   \n",
              "1    SCO0088_loc_836_rev  CCGTCGGGATACCGGACGACTCCCGCCTCCATGTGTTTCAATCCAC...   \n",
              "2    SCO0144_loc_109_fwd  AAAGGCAGCCTGACCATCCCGCTGGCCCACGTCCGTCGCCCGGCAA...   \n",
              "3    SCO0144_loc_109_rev  GGACGTGGGCCAGCGGGATGGTCAGGCTGCCTTTGTTTCAATCCAC...   \n",
              "4    SCO0062_loc_844_fwd  CGGATTTGTACAGGGCCCGGATCGCTCCCTGGCTGTCGCCCGGCAA...   \n",
              "..                   ...                                                ...   \n",
              "195  SCO0142_loc_157_rev  GGGCGTGCAGGCAGTTGACGCAGCAGAACGTCCAGTTTCAATCCAC...   \n",
              "196  SCO0012_loc_114_fwd  CAACGGAGCCTCCCGGCCCGGCTCCCGGTACACCGTCGCCCGGCAA...   \n",
              "197  SCO0012_loc_114_rev  GGTGTACCGGGAGCCGGGCCGGGAGGCTCCGTTGGTTTCAATCCAC...   \n",
              "198  SCO0006_loc_272_fwd  TTGGCAAGGGCGTCGATAGCGGGGGTGCGCACGGGTCGCCCGGCAA...   \n",
              "199  SCO0006_loc_272_rev  CCGTGCGCACCCCCGCTATCGACGCCCTTGCCAAGTTTCAATCCAC...   \n",
              "\n",
              "    Concentration Purification  \n",
              "0            25nm          STD  \n",
              "1            25nm          STD  \n",
              "2            25nm          STD  \n",
              "3            25nm          STD  \n",
              "4            25nm          STD  \n",
              "..            ...          ...  \n",
              "195          25nm          STD  \n",
              "196          25nm          STD  \n",
              "197          25nm          STD  \n",
              "198          25nm          STD  \n",
              "199          25nm          STD  \n",
              "\n",
              "[200 rows x 4 columns]"
            ]
          },
          "execution_count": 22,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "\n",
        "full_idt = idt_primers_cas3\n",
        "full_idt"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "## Folder with all the generated I/O"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 23,
      "metadata": {},
      "outputs": [
        {
          "ename": "NameError",
          "evalue": "name 'primer_number_increment' is not defined",
          "output_type": "error",
          "traceback": [
            "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
            "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
            "Cell \u001b[0;32mIn[23], line 43\u001b[0m\n\u001b[1;32m     17\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m: \n\u001b[1;32m     18\u001b[0m     output_files \u001b[38;5;241m=\u001b[39m [\n\u001b[1;32m     19\u001b[0m         {\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mname\u001b[39m\u001b[38;5;124m\"\u001b[39m: \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mcBEST_w_sgRNAs.gb\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mcontent\u001b[39m\u001b[38;5;124m\"\u001b[39m: assembled_cas3_plasmids}, \u001b[38;5;66;03m# LIST OF Dseqrecords\u001b[39;00m\n\u001b[1;32m     20\u001b[0m         {\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mname\u001b[39m\u001b[38;5;124m\"\u001b[39m: \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mfull_idt.csv\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mcontent\u001b[39m\u001b[38;5;124m\"\u001b[39m: full_idt},\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m     24\u001b[0m \n\u001b[1;32m     25\u001b[0m     ]\n\u001b[1;32m     27\u001b[0m input_values \u001b[38;5;241m=\u001b[39m {\n\u001b[1;32m     28\u001b[0m     \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mgenes_to_knockout\u001b[39m\u001b[38;5;124m\"\u001b[39m: genes_to_KO,\n\u001b[1;32m     29\u001b[0m \n\u001b[1;32m     30\u001b[0m     \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mfiltering_metrics\u001b[39m\u001b[38;5;124m\"\u001b[39m: {\n\u001b[1;32m     31\u001b[0m         \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mgc_upper\u001b[39m\u001b[38;5;124m\"\u001b[39m: gc_upper,\n\u001b[1;32m     32\u001b[0m         \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mgc_lower\u001b[39m\u001b[38;5;124m\"\u001b[39m: gc_lower,\n\u001b[1;32m     33\u001b[0m         \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124moff_target_seed\u001b[39m\u001b[38;5;124m\"\u001b[39m: off_target_seed,\n\u001b[1;32m     34\u001b[0m         \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124moff_target_upper\u001b[39m\u001b[38;5;124m\"\u001b[39m: off_target_upper,\n\u001b[1;32m     35\u001b[0m         \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mcas_type\u001b[39m\u001b[38;5;124m\"\u001b[39m: cas_type,\n\u001b[1;32m     36\u001b[0m         \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mnumber_of_sgRNAs_per_group\u001b[39m\u001b[38;5;124m\"\u001b[39m: number_of_sgRNAs_per_group,\n\u001b[1;32m     37\u001b[0m \n\u001b[1;32m     38\u001b[0m     },  \n\u001b[1;32m     39\u001b[0m     \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mpolymerase_settings\u001b[39m\u001b[38;5;124m\"\u001b[39m: {\n\u001b[1;32m     40\u001b[0m         \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mchosen_polymerase\u001b[39m\u001b[38;5;124m\"\u001b[39m: chosen_polymerase,\n\u001b[1;32m     41\u001b[0m         \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mmelting_temperature\u001b[39m\u001b[38;5;124m\"\u001b[39m: melting_temperature,\n\u001b[1;32m     42\u001b[0m         \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mprimer_concentration\u001b[39m\u001b[38;5;124m\"\u001b[39m: primer_concentration,\n\u001b[0;32m---> 43\u001b[0m         \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mprimer_number_increment\u001b[39m\u001b[38;5;124m\"\u001b[39m: \u001b[43mprimer_number_increment\u001b[49m,\n\u001b[1;32m     44\u001b[0m         \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mflanking_region\u001b[39m\u001b[38;5;124m\"\u001b[39m: flanking_region\n\u001b[1;32m     45\u001b[0m     },\n\u001b[1;32m     46\u001b[0m     \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124moverlapping_sequences\u001b[39m\u001b[38;5;124m\"\u001b[39m: {\n\u001b[1;32m     47\u001b[0m         \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mup_homology\u001b[39m\u001b[38;5;124m\"\u001b[39m: \u001b[38;5;28mstr\u001b[39m(forward_backbone_overhang),\n\u001b[1;32m     48\u001b[0m         \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mdw_homology\u001b[39m\u001b[38;5;124m\"\u001b[39m: \u001b[38;5;28mstr\u001b[39m(reverse_backbone_overhang)\n\u001b[1;32m     49\u001b[0m     }\n\u001b[1;32m     50\u001b[0m }\n\u001b[1;32m     52\u001b[0m \u001b[38;5;66;03m# Paths to Markdown files\u001b[39;00m\n\u001b[1;32m     53\u001b[0m markdown_file_paths \u001b[38;5;241m=\u001b[39m [\n\u001b[1;32m     54\u001b[0m     \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m../../protocols/conjugation_protcol.md\u001b[39m\u001b[38;5;124m\"\u001b[39m,\n\u001b[1;32m     55\u001b[0m     \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m../../protocols/single_target_crispr_plasmid_protcol.md\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m     56\u001b[0m \n\u001b[1;32m     57\u001b[0m ]\n",
            "\u001b[0;31mNameError\u001b[0m: name 'primer_number_increment' is not defined"
          ]
        }
      ],
      "source": [
        "input_files = [\n",
        "    {\"name\": \"input_genome.gb\", \"content\": genome},\n",
        "    {\"name\": \"input_plasmid.gb\", \"content\": clean_plasmid}\n",
        "]\n",
        "\n",
        "\n",
        "if in_frame_deletion: \n",
        "    output_files = [\n",
        "        {\"name\": \"BEST_w_sgRNAs.gb\", \"content\": assembled_contigs}, # LIST OF Dseqrecords\n",
        "        {\"name\": \"primer_df.csv\", \"content\": primer_df},\n",
        "        {\"name\": \"full_idt.csv\", \"content\": full_idt},\n",
        "        {\"name\": \"sgrna_df.csv\", \"content\": sgrna_df},\n",
        "        {\"name\": \"filtered_sgrna_df.csv\", \"content\": filtered_df},\n",
        "        {\"name\": \"plasmid_metadata_df.csv\", \"content\": plasmid_metadata_df},\n",
        "\n",
        "    ]\n",
        "else: \n",
        "    output_files = [\n",
        "        {\"name\": \"cBEST_w_sgRNAs.gb\", \"content\": assembled_cas3_plasmids}, # LIST OF Dseqrecords\n",
        "        {\"name\": \"full_idt.csv\", \"content\": full_idt},\n",
        "        {\"name\": \"sgrna_df.csv\", \"content\": sgrna_df},\n",
        "        {\"name\": \"filtered_df.csv\", \"content\": filtered_df},\n",
        "        {\"name\": \"plasmid_metadata_df.csv\", \"content\": plasmid_metadata_df},\n",
        "\n",
        "    ]\n",
        "         \n",
        "input_values = {\n",
        "    \"genes_to_knockout\": genes_to_KO,\n",
        "\n",
        "    \"filtering_metrics\": {\n",
        "        \"gc_upper\": gc_upper,\n",
        "        \"gc_lower\": gc_lower,\n",
        "        \"off_target_seed\": off_target_seed,\n",
        "        \"off_target_upper\": off_target_upper,\n",
        "        \"cas_type\": cas_type,\n",
        "        \"number_of_sgRNAs_per_group\": number_of_sgRNAs_per_group,\n",
        "\n",
        "    },  \n",
        "    \"polymerase_settings\": {\n",
        "        \"chosen_polymerase\": chosen_polymerase,\n",
        "        \"melting_temperature\": melting_temperature,\n",
        "        \"primer_concentration\": primer_concentration,\n",
        "        \"flanking_region\": flanking_region\n",
        "    },\n",
        "    \"overlapping_sequences\": {\n",
        "        \"up_homology\": str(forward_backbone_overhang),\n",
        "        \"dw_homology\": str(reverse_backbone_overhang)\n",
        "    }\n",
        "}\n",
        "\n",
        "# Paths to Markdown files\n",
        "markdown_file_paths = [\n",
        "    \"../../protocols/conjugation_protcol.md\",\n",
        "    \"../../protocols/single_target_crispr_plasmid_protcol.md\"\n",
        "\n",
        "]\n",
        "\n",
        "timestamp = datetime.utcnow().isoformat()\n",
        "\n",
        "# Create project directory structure\n",
        "project_directory = ProjectDirectory(\n",
        "    project_name=f\"CRISPR_cas3_inframe_deletion_workflow_{timestamp}\",\n",
        "    input_files=input_files,\n",
        "    output_files=output_files,\n",
        "    input_values=input_values,\n",
        "    markdown_file_paths=markdown_file_paths\n",
        ")\n",
        "\n",
        "# DO You want to save the folder? \n",
        "save_zip_folder = False \n",
        "\n",
        "if save_zip_folder: \n",
        "    # Generate the project directory structure and get the zip content\n",
        "    zip_content = project_directory.create_directory_structure(create_directories=False)\n",
        "\n",
        "    # Save the zip file to disk (optional)\n",
        "    with open(\"project_structure.zip\", \"wb\") as f:\n",
        "        f.write(zip_content)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {},
      "outputs": [],
      "source": []
    }
  ],
  "metadata": {
    "kernelspec": {
      "display_name": "streptocad",
      "language": "python",
      "name": "python3"
    },
    "language_info": {
      "codemirror_mode": {
        "name": "ipython",
        "version": 3
      },
      "file_extension": ".py",
      "mimetype": "text/x-python",
      "name": "python",
      "nbconvert_exporter": "python",
      "pygments_lexer": "ipython3",
      "version": "3.11.11"
    }
  },
  "nbformat": 4,
  "nbformat_minor": 2
}
