{
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "# W6-CRISPR-Cas3-in-frame-deletion"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "In Workflow 6, CRISPR-Cas3 is paired with a PCR and Gibson protocol to achieve in-frame deletions22(Figure 7A). Just like in workflow 5, this workflow can be used to knock out specific genes in streptomycetes but with the crucial difference of using Cas3 which has been shown to have higher efficiencies for targeted gene deletions in for example BGCs and random-sized deletion in streptomyces22. Due to homologies, and high GC content, this protocol relies on two steps of Gibson cloning to incorporate the sgRNA and the upstream and downstream repair regions. To get started, users can download the pCRISPR-Cas3 file and the genome file (S. coelicolor A3). StreptoAIM will then generate the necessary primers, plasmid files, and checking primers to verify full gene deletions (S5). \n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 24,
      "metadata": {},
      "outputs": [],
      "source": [
        "import sys\n",
        "import os\n",
        "from Bio.Restriction import * \n",
        "import pandas as pd\n",
        "from datetime import datetime\n",
        "\n",
        "# Ensure the src directory is in the Python path\n",
        "project_root = os.path.abspath(os.path.join(os.getcwd(), '../../'))\n",
        "if project_root not in sys.path:\n",
        "    sys.path.insert(0, project_root)\n",
        "\n",
        "\n",
        "from pydna.dseqrecord import Dseqrecord\n",
        "\n",
        "from streptocad.sequence_loading.sequence_loading import (\n",
        "    load_and_process_plasmid,\n",
        "    load_and_process_genome_sequences, \n",
        "    check_and_convert_input, \n",
        "    annotate_dseqrecord)\n",
        "\n",
        "from streptocad.utils import polymerase_dict, create_primer_df_from_dict,ProjectDirectory,extract_metadata_to_dataframe\n",
        "from streptocad.primers.primer_generation import create_idt_order_dataframe, make_primer_records, primers_to_IDT, find_best_check_primers_from_genome\n",
        "from streptocad.crispr.guideRNAcas3_9_12 import extract_sgRNAs, SgRNAargs\n",
        "from streptocad.cloning.cas3_plasmid_cloning import generate_cas3_protospacer_primers, cas3_plasmid_pcrs, assemble_cas3_plasmids\n",
        "from streptocad.cloning.gibson_cloning import (\n",
        "    find_up_dw_repair_templates,\n",
        "    assemble_multiple_plasmids_with_repair_templates_for_deletion,\n",
        "    update_primer_names\n",
        ")\n",
        "from streptocad.cloning.plasmid_processing import determine_workflow_order_for_plasmids"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 25,
      "metadata": {},
      "outputs": [],
      "source": [
        "from pydna.dseqrecord import Dseqrecord\n",
        "\n",
        "def find_locus_tags(record: Dseqrecord, max_tags: int = None):\n",
        "    \"\"\"\n",
        "    Iterate through a Dseqrecord and return a list of unique locus tags found in its features,\n",
        "    stopping after `max_tags` tags if specified.\n",
        "\n",
        "    :param record: pydna.dseqrecord.Dseqrecord object\n",
        "    :param max_tags: maximum number of locus tags to collect (None for no limit)\n",
        "    :return: list of locus_tag strings\n",
        "    \"\"\"\n",
        "    tags = []\n",
        "    seen = set()\n",
        "    count = 0\n",
        "\n",
        "    # Ensure the record has features\n",
        "    if not hasattr(record, 'features'):\n",
        "        return tags\n",
        "\n",
        "    for feature in record.features:\n",
        "        if max_tags is not None and count >= max_tags:\n",
        "            break\n",
        "\n",
        "        qualifiers = getattr(feature, 'qualifiers', {})\n",
        "\n",
        "        for tag in qualifiers.get('locus_tag', []):\n",
        "            if max_tags is not None and count >= max_tags:\n",
        "                break\n",
        "            # only append new tags\n",
        "            if tag not in seen:\n",
        "                seen.add(tag)\n",
        "                tags.append(tag)\n",
        "                count += 1\n",
        "\n",
        "        if max_tags is not None and count >= max_tags:\n",
        "            break\n",
        "\n",
        "    return tags\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "## INPUT"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 26,
      "metadata": {},
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "pCRISPR-Cas3\n"
          ]
        }
      ],
      "source": [
        "# Inputs\n",
        "# 1 Add genome of choice (genbank, fasta)\n",
        "path_to_genome = '../../../../data/genomes/Streptomyces_coelicolor_A3_chromosome.gb'\n",
        "genome = load_and_process_genome_sequences(path_to_genome)[0]\n",
        "\n",
        "# 2 Add plasmid \n",
        "path_to_plasmid = '../../../../data/plasmids/pCRISPR_Cas3.gbk'\n",
        "clean_plasmid = load_and_process_plasmid(path_to_plasmid)\n",
        "print(clean_plasmid.name)\n",
        "\n",
        "# 3 Choose genes to knock out (list)\n",
        "locus_tags = find_locus_tags(genome, 10)\n",
        "genes_to_KO = locus_tags\n",
        "\n",
        "\n",
        "#### Advanced settings ####\n",
        "# 4 Filtering metrics for sgRNAs\n",
        "gc_upper = 0.8\n",
        "gc_lower = 0.2\n",
        "off_target_seed = 8\n",
        "off_target_upper = 10\n",
        "cas_type='cas3'\n",
        "number_of_sgRNAs_per_group = 1\n",
        "\n",
        "# 5 Choose polymerase and target melting temperature\n",
        "chosen_polymerase = polymerase_dict['Phusion High-Fidelity DNA Polymerase (GC Buffer)']\n",
        "melting_temperature = 65\n",
        "primer_concentration = 0.4 \n",
        "primer_number_increment = 23\n",
        "flanking_region = 500\n",
        "\n",
        "# 6 Choose overlapping sequences for our plasmid we can use the following\n",
        "# the part of the primers that anneal to the generate the crRNA and protospacer\n",
        "forward_protospacer_overhang = 'GTCGCCCGGCAAAACCGG'\n",
        "reverse_protospacer_overhang = 'GTTTCAATCCACGCGCCCGT'\n",
        "\n",
        "# The universal part of the \n",
        "forward_backbone_overhang = 'GAGCTCATAAGTTCCTATTCCGAAG'\n",
        "reverse_backbone_overhang = 'AAGAAGTGGGTGTCGGACGC'\n",
        "\n",
        "Enzyme_for_repair_template_integration = EcoRI\n",
        "repair_length = 1000\n",
        "overlap_length = 40 \n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "# Computation"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 27,
      "metadata": {},
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "False\n"
          ]
        }
      ],
      "source": [
        "target_dict, genes_to_KO, annotation_input = check_and_convert_input(genes_to_KO)\n",
        "\n",
        "print(annotation_input)\n",
        "if annotation_input == True:\n",
        "    genome = annotate_dseqrecord(genome, target_dict)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 28,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>strain_name</th>\n",
              "      <th>locus_tag</th>\n",
              "      <th>gene_loc</th>\n",
              "      <th>gene_strand</th>\n",
              "      <th>sgrna_strand</th>\n",
              "      <th>sgrna_loc</th>\n",
              "      <th>gc</th>\n",
              "      <th>pam</th>\n",
              "      <th>sgrna</th>\n",
              "      <th>sgrna_seed_sequence</th>\n",
              "      <th>off_target_count</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>61</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO0002</td>\n",
              "      <td>1252</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>348</td>\n",
              "      <td>0.705882</td>\n",
              "      <td>TTC</td>\n",
              "      <td>CAACGGCCCAGGGACAGTCCGGCGACCAGTGACG</td>\n",
              "      <td>CAACGGCC</td>\n",
              "      <td>1</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>100</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO0003</td>\n",
              "      <td>3869</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>1216</td>\n",
              "      <td>0.558824</td>\n",
              "      <td>TTC</td>\n",
              "      <td>TCTGATTGGCTAGCCACGCTCCCAGAGCAAATCC</td>\n",
              "      <td>TCTGATTG</td>\n",
              "      <td>1</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>108</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO0003</td>\n",
              "      <td>3869</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>166</td>\n",
              "      <td>0.764706</td>\n",
              "      <td>TTC</td>\n",
              "      <td>CCCAGACACCACGCACGCCAGGGCCCGCCATCGC</td>\n",
              "      <td>CCCAGACA</td>\n",
              "      <td>1</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>38</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO0002</td>\n",
              "      <td>1252</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>2202</td>\n",
              "      <td>0.617647</td>\n",
              "      <td>TTC</td>\n",
              "      <td>TCTTTGCCGCGTGAGCCATCGTCCGATGTCGTCG</td>\n",
              "      <td>TCTTTGCC</td>\n",
              "      <td>1</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>72</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO0003</td>\n",
              "      <td>3869</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>803</td>\n",
              "      <td>0.588235</td>\n",
              "      <td>TTC</td>\n",
              "      <td>TGTGTGACCGGTTGTGGCGACAACGCGTCTTGGA</td>\n",
              "      <td>TGTGTGAC</td>\n",
              "      <td>1</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>...</th>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>179</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO0006</td>\n",
              "      <td>9304</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>501</td>\n",
              "      <td>0.764706</td>\n",
              "      <td>TTC</td>\n",
              "      <td>GATGGCGAGGGTCTTGGCGCTGGCGGCGGGTGGG</td>\n",
              "      <td>GATGGCGA</td>\n",
              "      <td>10</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>217</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO0010</td>\n",
              "      <td>12366</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>862</td>\n",
              "      <td>0.647059</td>\n",
              "      <td>TTC</td>\n",
              "      <td>GTCGTGACGGTCCCGCGGATTCGTCACGTCTGGA</td>\n",
              "      <td>GTCGTGAC</td>\n",
              "      <td>10</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>67</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO0003</td>\n",
              "      <td>3869</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>383</td>\n",
              "      <td>0.647059</td>\n",
              "      <td>TTC</td>\n",
              "      <td>CTGCCGTCATGCTGGGATGCCGTCTGTGTACGGC</td>\n",
              "      <td>CTGCCGTC</td>\n",
              "      <td>10</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>35</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO0002</td>\n",
              "      <td>1252</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>2274</td>\n",
              "      <td>0.705882</td>\n",
              "      <td>TTC</td>\n",
              "      <td>GACGGGCGTCACACCCAGCGCGGTCAGTCGTTCC</td>\n",
              "      <td>GACGGGCG</td>\n",
              "      <td>10</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>172</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO0006</td>\n",
              "      <td>9304</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>841</td>\n",
              "      <td>0.764706</td>\n",
              "      <td>TTC</td>\n",
              "      <td>CCCTCCCGCCTCGGCGACCAAGAGCCCTTCCGCG</td>\n",
              "      <td>CCCTCCCG</td>\n",
              "      <td>10</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>124 rows × 11 columns</p>\n",
              "</div>"
            ],
            "text/plain": [
              "     strain_name locus_tag  gene_loc  gene_strand  sgrna_strand  sgrna_loc  \\\n",
              "61   NC_003888.3   SCO0002      1252            1             1        348   \n",
              "100  NC_003888.3   SCO0003      3869            1             1       1216   \n",
              "108  NC_003888.3   SCO0003      3869            1             1        166   \n",
              "38   NC_003888.3   SCO0002      1252            1             1       2202   \n",
              "72   NC_003888.3   SCO0003      3869            1            -1        803   \n",
              "..           ...       ...       ...          ...           ...        ...   \n",
              "179  NC_003888.3   SCO0006      9304            1             1        501   \n",
              "217  NC_003888.3   SCO0010     12366            1            -1        862   \n",
              "67   NC_003888.3   SCO0003      3869            1            -1        383   \n",
              "35   NC_003888.3   SCO0002      1252            1             1       2274   \n",
              "172  NC_003888.3   SCO0006      9304            1            -1        841   \n",
              "\n",
              "           gc  pam                               sgrna sgrna_seed_sequence  \\\n",
              "61   0.705882  TTC  CAACGGCCCAGGGACAGTCCGGCGACCAGTGACG            CAACGGCC   \n",
              "100  0.558824  TTC  TCTGATTGGCTAGCCACGCTCCCAGAGCAAATCC            TCTGATTG   \n",
              "108  0.764706  TTC  CCCAGACACCACGCACGCCAGGGCCCGCCATCGC            CCCAGACA   \n",
              "38   0.617647  TTC  TCTTTGCCGCGTGAGCCATCGTCCGATGTCGTCG            TCTTTGCC   \n",
              "72   0.588235  TTC  TGTGTGACCGGTTGTGGCGACAACGCGTCTTGGA            TGTGTGAC   \n",
              "..        ...  ...                                 ...                 ...   \n",
              "179  0.764706  TTC  GATGGCGAGGGTCTTGGCGCTGGCGGCGGGTGGG            GATGGCGA   \n",
              "217  0.647059  TTC  GTCGTGACGGTCCCGCGGATTCGTCACGTCTGGA            GTCGTGAC   \n",
              "67   0.647059  TTC  CTGCCGTCATGCTGGGATGCCGTCTGTGTACGGC            CTGCCGTC   \n",
              "35   0.705882  TTC  GACGGGCGTCACACCCAGCGCGGTCAGTCGTTCC            GACGGGCG   \n",
              "172  0.764706  TTC  CCCTCCCGCCTCGGCGACCAAGAGCCCTTCCGCG            CCCTCCCG   \n",
              "\n",
              "     off_target_count  \n",
              "61                  1  \n",
              "100                 1  \n",
              "108                 1  \n",
              "38                  1  \n",
              "72                  1  \n",
              "..                ...  \n",
              "179                10  \n",
              "217                10  \n",
              "67                 10  \n",
              "35                 10  \n",
              "172                10  \n",
              "\n",
              "[124 rows x 11 columns]"
            ]
          },
          "execution_count": 28,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "# Initialize SgRNAargs with desired parameters\n",
        "args = SgRNAargs(genome, \n",
        "                genes_to_KO,\n",
        "                step=['find', 'filter'],\n",
        "                gc_upper = gc_upper,\n",
        "                gc_lower = gc_lower,\n",
        "                off_target_seed = off_target_seed,\n",
        "                off_target_upper = off_target_upper,\n",
        "                cas_type=cas_type\n",
        "                )\n",
        "\n",
        "sgrna_df = extract_sgRNAs(args)\n",
        "sgrna_df"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 29,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>strain_name</th>\n",
              "      <th>locus_tag</th>\n",
              "      <th>gene_loc</th>\n",
              "      <th>gene_strand</th>\n",
              "      <th>sgrna_strand</th>\n",
              "      <th>sgrna_loc</th>\n",
              "      <th>gc</th>\n",
              "      <th>pam</th>\n",
              "      <th>sgrna</th>\n",
              "      <th>sgrna_seed_sequence</th>\n",
              "      <th>off_target_count</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>61</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO0002</td>\n",
              "      <td>1252</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>348</td>\n",
              "      <td>0.705882</td>\n",
              "      <td>TTC</td>\n",
              "      <td>CAACGGCCCAGGGACAGTCCGGCGACCAGTGACG</td>\n",
              "      <td>CAACGGCC</td>\n",
              "      <td>1</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>100</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO0003</td>\n",
              "      <td>3869</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>1216</td>\n",
              "      <td>0.558824</td>\n",
              "      <td>TTC</td>\n",
              "      <td>TCTGATTGGCTAGCCACGCTCCCAGAGCAAATCC</td>\n",
              "      <td>TCTGATTG</td>\n",
              "      <td>1</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>136</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO0005</td>\n",
              "      <td>7187</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>1273</td>\n",
              "      <td>0.735294</td>\n",
              "      <td>TTC</td>\n",
              "      <td>TGCCAGCACGTCGCCGGCCACACCGGCTCCAACC</td>\n",
              "      <td>TGCCAGCA</td>\n",
              "      <td>1</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>181</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO0006</td>\n",
              "      <td>9304</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>272</td>\n",
              "      <td>0.705882</td>\n",
              "      <td>TTC</td>\n",
              "      <td>TTGGCAAGGGCGTCGATAGCGGGGGTGCGCACGG</td>\n",
              "      <td>TTGGCAAG</td>\n",
              "      <td>1</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>196</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO0009</td>\n",
              "      <td>11617</td>\n",
              "      <td>-1</td>\n",
              "      <td>1</td>\n",
              "      <td>226</td>\n",
              "      <td>0.676471</td>\n",
              "      <td>TTC</td>\n",
              "      <td>GACCCACTGTGCGAGAGCCGCGAGACCTCGTTGG</td>\n",
              "      <td>GACCCACT</td>\n",
              "      <td>1</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>220</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO0010</td>\n",
              "      <td>12366</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>1066</td>\n",
              "      <td>0.647059</td>\n",
              "      <td>TTC</td>\n",
              "      <td>TGCGCTCTGGTGGTCTGACGTTCAGCCCGACCAC</td>\n",
              "      <td>TGCGCTCT</td>\n",
              "      <td>2</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>190</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO0008</td>\n",
              "      <td>11010</td>\n",
              "      <td>-1</td>\n",
              "      <td>1</td>\n",
              "      <td>196</td>\n",
              "      <td>0.705882</td>\n",
              "      <td>TTC</td>\n",
              "      <td>CGGGTAGATCACCAGGTCGCGGGTCACCAGGGCG</td>\n",
              "      <td>CGGGTAGA</td>\n",
              "      <td>2</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>186</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO0007</td>\n",
              "      <td>10554</td>\n",
              "      <td>-1</td>\n",
              "      <td>-1</td>\n",
              "      <td>291</td>\n",
              "      <td>0.794118</td>\n",
              "      <td>TTC</td>\n",
              "      <td>CCACTCCCCCACCTGACGCGCGCCCTGCCCCACC</td>\n",
              "      <td>CCACTCCC</td>\n",
              "      <td>3</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>122</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO0004</td>\n",
              "      <td>6226</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>861</td>\n",
              "      <td>0.676471</td>\n",
              "      <td>TTC</td>\n",
              "      <td>TGCGTCAGCGCAGACCAGCGTCTCCTCGTGCAGC</td>\n",
              "      <td>TGCGTCAG</td>\n",
              "      <td>4</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>13</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO0001</td>\n",
              "      <td>446</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>89</td>\n",
              "      <td>0.676471</td>\n",
              "      <td>TTC</td>\n",
              "      <td>ACACCCGCCGTCTGATACTGCGTCACCCGGCACG</td>\n",
              "      <td>ACACCCGC</td>\n",
              "      <td>4</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "     strain_name locus_tag  gene_loc  gene_strand  sgrna_strand  sgrna_loc  \\\n",
              "61   NC_003888.3   SCO0002      1252            1             1        348   \n",
              "100  NC_003888.3   SCO0003      3869            1             1       1216   \n",
              "136  NC_003888.3   SCO0005      7187            1            -1       1273   \n",
              "181  NC_003888.3   SCO0006      9304            1             1        272   \n",
              "196  NC_003888.3   SCO0009     11617           -1             1        226   \n",
              "220  NC_003888.3   SCO0010     12366            1             1       1066   \n",
              "190  NC_003888.3   SCO0008     11010           -1             1        196   \n",
              "186  NC_003888.3   SCO0007     10554           -1            -1        291   \n",
              "122  NC_003888.3   SCO0004      6226            1             1        861   \n",
              "13   NC_003888.3   SCO0001       446            1             1         89   \n",
              "\n",
              "           gc  pam                               sgrna sgrna_seed_sequence  \\\n",
              "61   0.705882  TTC  CAACGGCCCAGGGACAGTCCGGCGACCAGTGACG            CAACGGCC   \n",
              "100  0.558824  TTC  TCTGATTGGCTAGCCACGCTCCCAGAGCAAATCC            TCTGATTG   \n",
              "136  0.735294  TTC  TGCCAGCACGTCGCCGGCCACACCGGCTCCAACC            TGCCAGCA   \n",
              "181  0.705882  TTC  TTGGCAAGGGCGTCGATAGCGGGGGTGCGCACGG            TTGGCAAG   \n",
              "196  0.676471  TTC  GACCCACTGTGCGAGAGCCGCGAGACCTCGTTGG            GACCCACT   \n",
              "220  0.647059  TTC  TGCGCTCTGGTGGTCTGACGTTCAGCCCGACCAC            TGCGCTCT   \n",
              "190  0.705882  TTC  CGGGTAGATCACCAGGTCGCGGGTCACCAGGGCG            CGGGTAGA   \n",
              "186  0.794118  TTC  CCACTCCCCCACCTGACGCGCGCCCTGCCCCACC            CCACTCCC   \n",
              "122  0.676471  TTC  TGCGTCAGCGCAGACCAGCGTCTCCTCGTGCAGC            TGCGTCAG   \n",
              "13   0.676471  TTC  ACACCCGCCGTCTGATACTGCGTCACCCGGCACG            ACACCCGC   \n",
              "\n",
              "     off_target_count  \n",
              "61                  1  \n",
              "100                 1  \n",
              "136                 1  \n",
              "181                 1  \n",
              "196                 1  \n",
              "220                 2  \n",
              "190                 2  \n",
              "186                 3  \n",
              "122                 4  \n",
              "13                  4  "
            ]
          },
          "execution_count": 29,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "# Filter the DataFrame to retain only up to 5 sgRNA sequences per locus_tag\n",
        "filtered_df = sgrna_df.groupby('locus_tag').head(number_of_sgRNAs_per_group)\n",
        "filtered_df"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "## Output"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 30,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>strain_name</th>\n",
              "      <th>locus_tag</th>\n",
              "      <th>gene_loc</th>\n",
              "      <th>gene_strand</th>\n",
              "      <th>sgrna_strand</th>\n",
              "      <th>sgrna_loc</th>\n",
              "      <th>gc</th>\n",
              "      <th>pam</th>\n",
              "      <th>sgrna</th>\n",
              "      <th>sgrna_seed_sequence</th>\n",
              "      <th>off_target_count</th>\n",
              "      <th>Fwd Primer</th>\n",
              "      <th>Rev Primer</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>61</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO0002</td>\n",
              "      <td>1252</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>348</td>\n",
              "      <td>0.705882</td>\n",
              "      <td>TTC</td>\n",
              "      <td>CAACGGCCCAGGGACAGTCCGGCGACCAGTGACG</td>\n",
              "      <td>CAACGGCC</td>\n",
              "      <td>1</td>\n",
              "      <td>CAACGGCCCAGGGACAGTCCGGCGACCAGTGACGGTCGCCCGGCAA...</td>\n",
              "      <td>CGTCACTGGTCGCCGGACTGTCCCTGGGCCGTTGGTTTCAATCCAC...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>100</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO0003</td>\n",
              "      <td>3869</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>1216</td>\n",
              "      <td>0.558824</td>\n",
              "      <td>TTC</td>\n",
              "      <td>TCTGATTGGCTAGCCACGCTCCCAGAGCAAATCC</td>\n",
              "      <td>TCTGATTG</td>\n",
              "      <td>1</td>\n",
              "      <td>TCTGATTGGCTAGCCACGCTCCCAGAGCAAATCCGTCGCCCGGCAA...</td>\n",
              "      <td>GGATTTGCTCTGGGAGCGTGGCTAGCCAATCAGAGTTTCAATCCAC...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>136</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO0005</td>\n",
              "      <td>7187</td>\n",
              "      <td>1</td>\n",
              "      <td>-1</td>\n",
              "      <td>1273</td>\n",
              "      <td>0.735294</td>\n",
              "      <td>TTC</td>\n",
              "      <td>TGCCAGCACGTCGCCGGCCACACCGGCTCCAACC</td>\n",
              "      <td>TGCCAGCA</td>\n",
              "      <td>1</td>\n",
              "      <td>TGCCAGCACGTCGCCGGCCACACCGGCTCCAACCGTCGCCCGGCAA...</td>\n",
              "      <td>GGTTGGAGCCGGTGTGGCCGGCGACGTGCTGGCAGTTTCAATCCAC...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>181</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO0006</td>\n",
              "      <td>9304</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>272</td>\n",
              "      <td>0.705882</td>\n",
              "      <td>TTC</td>\n",
              "      <td>TTGGCAAGGGCGTCGATAGCGGGGGTGCGCACGG</td>\n",
              "      <td>TTGGCAAG</td>\n",
              "      <td>1</td>\n",
              "      <td>TTGGCAAGGGCGTCGATAGCGGGGGTGCGCACGGGTCGCCCGGCAA...</td>\n",
              "      <td>CCGTGCGCACCCCCGCTATCGACGCCCTTGCCAAGTTTCAATCCAC...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>196</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO0009</td>\n",
              "      <td>11617</td>\n",
              "      <td>-1</td>\n",
              "      <td>1</td>\n",
              "      <td>226</td>\n",
              "      <td>0.676471</td>\n",
              "      <td>TTC</td>\n",
              "      <td>GACCCACTGTGCGAGAGCCGCGAGACCTCGTTGG</td>\n",
              "      <td>GACCCACT</td>\n",
              "      <td>1</td>\n",
              "      <td>GACCCACTGTGCGAGAGCCGCGAGACCTCGTTGGGTCGCCCGGCAA...</td>\n",
              "      <td>CCAACGAGGTCTCGCGGCTCTCGCACAGTGGGTCGTTTCAATCCAC...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>220</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO0010</td>\n",
              "      <td>12366</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>1066</td>\n",
              "      <td>0.647059</td>\n",
              "      <td>TTC</td>\n",
              "      <td>TGCGCTCTGGTGGTCTGACGTTCAGCCCGACCAC</td>\n",
              "      <td>TGCGCTCT</td>\n",
              "      <td>2</td>\n",
              "      <td>TGCGCTCTGGTGGTCTGACGTTCAGCCCGACCACGTCGCCCGGCAA...</td>\n",
              "      <td>GTGGTCGGGCTGAACGTCAGACCACCAGAGCGCAGTTTCAATCCAC...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>190</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO0008</td>\n",
              "      <td>11010</td>\n",
              "      <td>-1</td>\n",
              "      <td>1</td>\n",
              "      <td>196</td>\n",
              "      <td>0.705882</td>\n",
              "      <td>TTC</td>\n",
              "      <td>CGGGTAGATCACCAGGTCGCGGGTCACCAGGGCG</td>\n",
              "      <td>CGGGTAGA</td>\n",
              "      <td>2</td>\n",
              "      <td>CGGGTAGATCACCAGGTCGCGGGTCACCAGGGCGGTCGCCCGGCAA...</td>\n",
              "      <td>CGCCCTGGTGACCCGCGACCTGGTGATCTACCCGGTTTCAATCCAC...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>186</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO0007</td>\n",
              "      <td>10554</td>\n",
              "      <td>-1</td>\n",
              "      <td>-1</td>\n",
              "      <td>291</td>\n",
              "      <td>0.794118</td>\n",
              "      <td>TTC</td>\n",
              "      <td>CCACTCCCCCACCTGACGCGCGCCCTGCCCCACC</td>\n",
              "      <td>CCACTCCC</td>\n",
              "      <td>3</td>\n",
              "      <td>CCACTCCCCCACCTGACGCGCGCCCTGCCCCACCGTCGCCCGGCAA...</td>\n",
              "      <td>GGTGGGGCAGGGCGCGCGTCAGGTGGGGGAGTGGGTTTCAATCCAC...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>122</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO0004</td>\n",
              "      <td>6226</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>861</td>\n",
              "      <td>0.676471</td>\n",
              "      <td>TTC</td>\n",
              "      <td>TGCGTCAGCGCAGACCAGCGTCTCCTCGTGCAGC</td>\n",
              "      <td>TGCGTCAG</td>\n",
              "      <td>4</td>\n",
              "      <td>TGCGTCAGCGCAGACCAGCGTCTCCTCGTGCAGCGTCGCCCGGCAA...</td>\n",
              "      <td>GCTGCACGAGGAGACGCTGGTCTGCGCTGACGCAGTTTCAATCCAC...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>13</th>\n",
              "      <td>NC_003888.3</td>\n",
              "      <td>SCO0001</td>\n",
              "      <td>446</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>89</td>\n",
              "      <td>0.676471</td>\n",
              "      <td>TTC</td>\n",
              "      <td>ACACCCGCCGTCTGATACTGCGTCACCCGGCACG</td>\n",
              "      <td>ACACCCGC</td>\n",
              "      <td>4</td>\n",
              "      <td>ACACCCGCCGTCTGATACTGCGTCACCCGGCACGGTCGCCCGGCAA...</td>\n",
              "      <td>CGTGCCGGGTGACGCAGTATCAGACGGCGGGTGTGTTTCAATCCAC...</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "     strain_name locus_tag  gene_loc  gene_strand  sgrna_strand  sgrna_loc  \\\n",
              "61   NC_003888.3   SCO0002      1252            1             1        348   \n",
              "100  NC_003888.3   SCO0003      3869            1             1       1216   \n",
              "136  NC_003888.3   SCO0005      7187            1            -1       1273   \n",
              "181  NC_003888.3   SCO0006      9304            1             1        272   \n",
              "196  NC_003888.3   SCO0009     11617           -1             1        226   \n",
              "220  NC_003888.3   SCO0010     12366            1             1       1066   \n",
              "190  NC_003888.3   SCO0008     11010           -1             1        196   \n",
              "186  NC_003888.3   SCO0007     10554           -1            -1        291   \n",
              "122  NC_003888.3   SCO0004      6226            1             1        861   \n",
              "13   NC_003888.3   SCO0001       446            1             1         89   \n",
              "\n",
              "           gc  pam                               sgrna sgrna_seed_sequence  \\\n",
              "61   0.705882  TTC  CAACGGCCCAGGGACAGTCCGGCGACCAGTGACG            CAACGGCC   \n",
              "100  0.558824  TTC  TCTGATTGGCTAGCCACGCTCCCAGAGCAAATCC            TCTGATTG   \n",
              "136  0.735294  TTC  TGCCAGCACGTCGCCGGCCACACCGGCTCCAACC            TGCCAGCA   \n",
              "181  0.705882  TTC  TTGGCAAGGGCGTCGATAGCGGGGGTGCGCACGG            TTGGCAAG   \n",
              "196  0.676471  TTC  GACCCACTGTGCGAGAGCCGCGAGACCTCGTTGG            GACCCACT   \n",
              "220  0.647059  TTC  TGCGCTCTGGTGGTCTGACGTTCAGCCCGACCAC            TGCGCTCT   \n",
              "190  0.705882  TTC  CGGGTAGATCACCAGGTCGCGGGTCACCAGGGCG            CGGGTAGA   \n",
              "186  0.794118  TTC  CCACTCCCCCACCTGACGCGCGCCCTGCCCCACC            CCACTCCC   \n",
              "122  0.676471  TTC  TGCGTCAGCGCAGACCAGCGTCTCCTCGTGCAGC            TGCGTCAG   \n",
              "13   0.676471  TTC  ACACCCGCCGTCTGATACTGCGTCACCCGGCACG            ACACCCGC   \n",
              "\n",
              "     off_target_count                                         Fwd Primer  \\\n",
              "61                  1  CAACGGCCCAGGGACAGTCCGGCGACCAGTGACGGTCGCCCGGCAA...   \n",
              "100                 1  TCTGATTGGCTAGCCACGCTCCCAGAGCAAATCCGTCGCCCGGCAA...   \n",
              "136                 1  TGCCAGCACGTCGCCGGCCACACCGGCTCCAACCGTCGCCCGGCAA...   \n",
              "181                 1  TTGGCAAGGGCGTCGATAGCGGGGGTGCGCACGGGTCGCCCGGCAA...   \n",
              "196                 1  GACCCACTGTGCGAGAGCCGCGAGACCTCGTTGGGTCGCCCGGCAA...   \n",
              "220                 2  TGCGCTCTGGTGGTCTGACGTTCAGCCCGACCACGTCGCCCGGCAA...   \n",
              "190                 2  CGGGTAGATCACCAGGTCGCGGGTCACCAGGGCGGTCGCCCGGCAA...   \n",
              "186                 3  CCACTCCCCCACCTGACGCGCGCCCTGCCCCACCGTCGCCCGGCAA...   \n",
              "122                 4  TGCGTCAGCGCAGACCAGCGTCTCCTCGTGCAGCGTCGCCCGGCAA...   \n",
              "13                  4  ACACCCGCCGTCTGATACTGCGTCACCCGGCACGGTCGCCCGGCAA...   \n",
              "\n",
              "                                            Rev Primer  \n",
              "61   CGTCACTGGTCGCCGGACTGTCCCTGGGCCGTTGGTTTCAATCCAC...  \n",
              "100  GGATTTGCTCTGGGAGCGTGGCTAGCCAATCAGAGTTTCAATCCAC...  \n",
              "136  GGTTGGAGCCGGTGTGGCCGGCGACGTGCTGGCAGTTTCAATCCAC...  \n",
              "181  CCGTGCGCACCCCCGCTATCGACGCCCTTGCCAAGTTTCAATCCAC...  \n",
              "196  CCAACGAGGTCTCGCGGCTCTCGCACAGTGGGTCGTTTCAATCCAC...  \n",
              "220  GTGGTCGGGCTGAACGTCAGACCACCAGAGCGCAGTTTCAATCCAC...  \n",
              "190  CGCCCTGGTGACCCGCGACCTGGTGATCTACCCGGTTTCAATCCAC...  \n",
              "186  GGTGGGGCAGGGCGCGCGTCAGGTGGGGGAGTGGGTTTCAATCCAC...  \n",
              "122  GCTGCACGAGGAGACGCTGGTCTGCGCTGACGCAGTTTCAATCCAC...  \n",
              "13   CGTGCCGGGTGACGCAGTATCAGACGGCGGGTGTGTTTCAATCCAC...  "
            ]
          },
          "execution_count": 30,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "filtered_df_w_primers = generate_cas3_protospacer_primers(filtered_df,\n",
        "                                                          fwd_overhang= forward_protospacer_overhang, \n",
        "                                                          rev_overhang=reverse_protospacer_overhang)\n",
        "filtered_df_w_primers"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 31,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/plain": [
              "10"
            ]
          },
          "execution_count": 31,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "amplicons = cas3_plasmid_pcrs(clean_plasmid, \n",
        "                              filtered_df, \n",
        "                              universal_fwd_seq=forward_backbone_overhang, \n",
        "                              universal_rev_seq=reverse_backbone_overhang\n",
        "                              )\n",
        "\n",
        "len(amplicons)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 32,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/plain": [
              "10"
            ]
          },
          "execution_count": 32,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "amplicons = cas3_plasmid_pcrs(clean_plasmid, filtered_df)\n",
        "\n",
        "len(amplicons)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {},
      "outputs": [],
      "source": [
        "# cut plasmid and incorporate amplicons\n",
        "assembled_cas3_plasmids = assemble_cas3_plasmids(clean_plasmid, amplicons)\n",
        "len(assembled_cas3_plasmids)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 11,
      "metadata": {},
      "outputs": [],
      "source": [
        "# Constructing a meaningful name, ID, and description for the assembled plasmid using user input\n",
        "targeting_info = []\n",
        "for index, row in filtered_df.iterrows():\n",
        "    formatted_str = f\"pCas3_{row['locus_tag']}({row['sgrna_loc']})\"\n",
        "    targeting_info.append(formatted_str)\n",
        "\n",
        "for i in range(len(assembled_cas3_plasmids)):\n",
        "    assembled_cas3_plasmids[i].name = f'{targeting_info[i]}'\n",
        "    assembled_cas3_plasmids[i].id = assembled_cas3_plasmids[i].name  \n",
        "    assembled_cas3_plasmids[i].description = f'CRISPR-Cas3 targeting {\", \".join(genes_to_KO)} for single gene knockout, assembled using StreptoCAD.'\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 12,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/plain": [
              "'pCas3_SCO0002(348)'"
            ]
          },
          "execution_count": 12,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "assembled_cas3_plasmids[0].name"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 13,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/plain": [
              "'CRISPR-Cas3 targeting SCO0001, SCO0002, SCO0003, SCO0004, SCO0005, SCO0006, SCO0007, SCO0008, SCO0009, SCO0010 for single gene knockout, assembled using StreptoCAD.'"
            ]
          },
          "execution_count": 13,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "assembled_cas3_plasmids[0].description"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 14,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/plain": [
              "'pCas3_SCO0002(348)'"
            ]
          },
          "execution_count": 14,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "assembled_cas3_plasmids[0].id"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 15,
      "metadata": {},
      "outputs": [],
      "source": [
        "print_plasmids = False\n",
        "\n",
        "if print_plasmids: \n",
        "    for vector in assembled_cas3_plasmids: \n",
        "        vector.write(f\"../../data/plasmids/sgRNA_plasmids_pCRISPR–Cas3/{vector.id}.gb\")"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "### IDT primers"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 16,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/plain": [
              "[Dseqrecord(-52),\n",
              " Dseqrecord(-54),\n",
              " Dseqrecord(-52),\n",
              " Dseqrecord(-54),\n",
              " Dseqrecord(-52),\n",
              " Dseqrecord(-54),\n",
              " Dseqrecord(-52),\n",
              " Dseqrecord(-54),\n",
              " Dseqrecord(-52),\n",
              " Dseqrecord(-54),\n",
              " Dseqrecord(-52),\n",
              " Dseqrecord(-54),\n",
              " Dseqrecord(-52),\n",
              " Dseqrecord(-54),\n",
              " Dseqrecord(-52),\n",
              " Dseqrecord(-54),\n",
              " Dseqrecord(-52),\n",
              " Dseqrecord(-54),\n",
              " Dseqrecord(-52),\n",
              " Dseqrecord(-54)]"
            ]
          },
          "execution_count": 16,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "primer_records = make_primer_records(filtered_df_w_primers)\n",
        "primer_records"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>Name</th>\n",
              "      <th>Sequence</th>\n",
              "      <th>Concentration</th>\n",
              "      <th>Purification</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>SCO0002_loc_348_fwd</td>\n",
              "      <td>CAACGGCCCAGGGACAGTCCGGCGACCAGTGACGGTCGCCCGGCAA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>SCO0002_loc_348_rev</td>\n",
              "      <td>CGTCACTGGTCGCCGGACTGTCCCTGGGCCGTTGGTTTCAATCCAC...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>SCO0003_loc_1216_fwd</td>\n",
              "      <td>TCTGATTGGCTAGCCACGCTCCCAGAGCAAATCCGTCGCCCGGCAA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>SCO0003_loc_1216_rev</td>\n",
              "      <td>GGATTTGCTCTGGGAGCGTGGCTAGCCAATCAGAGTTTCAATCCAC...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>SCO0005_loc_1273_fwd</td>\n",
              "      <td>TGCCAGCACGTCGCCGGCCACACCGGCTCCAACCGTCGCCCGGCAA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>5</th>\n",
              "      <td>SCO0005_loc_1273_rev</td>\n",
              "      <td>GGTTGGAGCCGGTGTGGCCGGCGACGTGCTGGCAGTTTCAATCCAC...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>6</th>\n",
              "      <td>SCO0006_loc_272_fwd</td>\n",
              "      <td>TTGGCAAGGGCGTCGATAGCGGGGGTGCGCACGGGTCGCCCGGCAA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>7</th>\n",
              "      <td>SCO0006_loc_272_rev</td>\n",
              "      <td>CCGTGCGCACCCCCGCTATCGACGCCCTTGCCAAGTTTCAATCCAC...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>8</th>\n",
              "      <td>SCO0009_loc_226_fwd</td>\n",
              "      <td>GACCCACTGTGCGAGAGCCGCGAGACCTCGTTGGGTCGCCCGGCAA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>9</th>\n",
              "      <td>SCO0009_loc_226_rev</td>\n",
              "      <td>CCAACGAGGTCTCGCGGCTCTCGCACAGTGGGTCGTTTCAATCCAC...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>10</th>\n",
              "      <td>SCO0010_loc_1066_fwd</td>\n",
              "      <td>TGCGCTCTGGTGGTCTGACGTTCAGCCCGACCACGTCGCCCGGCAA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>11</th>\n",
              "      <td>SCO0010_loc_1066_rev</td>\n",
              "      <td>GTGGTCGGGCTGAACGTCAGACCACCAGAGCGCAGTTTCAATCCAC...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>12</th>\n",
              "      <td>SCO0008_loc_196_fwd</td>\n",
              "      <td>CGGGTAGATCACCAGGTCGCGGGTCACCAGGGCGGTCGCCCGGCAA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>13</th>\n",
              "      <td>SCO0008_loc_196_rev</td>\n",
              "      <td>CGCCCTGGTGACCCGCGACCTGGTGATCTACCCGGTTTCAATCCAC...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>14</th>\n",
              "      <td>SCO0007_loc_291_fwd</td>\n",
              "      <td>CCACTCCCCCACCTGACGCGCGCCCTGCCCCACCGTCGCCCGGCAA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>15</th>\n",
              "      <td>SCO0007_loc_291_rev</td>\n",
              "      <td>GGTGGGGCAGGGCGCGCGTCAGGTGGGGGAGTGGGTTTCAATCCAC...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>16</th>\n",
              "      <td>SCO0004_loc_861_fwd</td>\n",
              "      <td>TGCGTCAGCGCAGACCAGCGTCTCCTCGTGCAGCGTCGCCCGGCAA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>17</th>\n",
              "      <td>SCO0004_loc_861_rev</td>\n",
              "      <td>GCTGCACGAGGAGACGCTGGTCTGCGCTGACGCAGTTTCAATCCAC...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>18</th>\n",
              "      <td>SCO0001_loc_89_fwd</td>\n",
              "      <td>ACACCCGCCGTCTGATACTGCGTCACCCGGCACGGTCGCCCGGCAA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>19</th>\n",
              "      <td>SCO0001_loc_89_rev</td>\n",
              "      <td>CGTGCCGGGTGACGCAGTATCAGACGGCGGGTGTGTTTCAATCCAC...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "                    Name                                           Sequence  \\\n",
              "0    SCO0002_loc_348_fwd  CAACGGCCCAGGGACAGTCCGGCGACCAGTGACGGTCGCCCGGCAA...   \n",
              "1    SCO0002_loc_348_rev  CGTCACTGGTCGCCGGACTGTCCCTGGGCCGTTGGTTTCAATCCAC...   \n",
              "2   SCO0003_loc_1216_fwd  TCTGATTGGCTAGCCACGCTCCCAGAGCAAATCCGTCGCCCGGCAA...   \n",
              "3   SCO0003_loc_1216_rev  GGATTTGCTCTGGGAGCGTGGCTAGCCAATCAGAGTTTCAATCCAC...   \n",
              "4   SCO0005_loc_1273_fwd  TGCCAGCACGTCGCCGGCCACACCGGCTCCAACCGTCGCCCGGCAA...   \n",
              "5   SCO0005_loc_1273_rev  GGTTGGAGCCGGTGTGGCCGGCGACGTGCTGGCAGTTTCAATCCAC...   \n",
              "6    SCO0006_loc_272_fwd  TTGGCAAGGGCGTCGATAGCGGGGGTGCGCACGGGTCGCCCGGCAA...   \n",
              "7    SCO0006_loc_272_rev  CCGTGCGCACCCCCGCTATCGACGCCCTTGCCAAGTTTCAATCCAC...   \n",
              "8    SCO0009_loc_226_fwd  GACCCACTGTGCGAGAGCCGCGAGACCTCGTTGGGTCGCCCGGCAA...   \n",
              "9    SCO0009_loc_226_rev  CCAACGAGGTCTCGCGGCTCTCGCACAGTGGGTCGTTTCAATCCAC...   \n",
              "10  SCO0010_loc_1066_fwd  TGCGCTCTGGTGGTCTGACGTTCAGCCCGACCACGTCGCCCGGCAA...   \n",
              "11  SCO0010_loc_1066_rev  GTGGTCGGGCTGAACGTCAGACCACCAGAGCGCAGTTTCAATCCAC...   \n",
              "12   SCO0008_loc_196_fwd  CGGGTAGATCACCAGGTCGCGGGTCACCAGGGCGGTCGCCCGGCAA...   \n",
              "13   SCO0008_loc_196_rev  CGCCCTGGTGACCCGCGACCTGGTGATCTACCCGGTTTCAATCCAC...   \n",
              "14   SCO0007_loc_291_fwd  CCACTCCCCCACCTGACGCGCGCCCTGCCCCACCGTCGCCCGGCAA...   \n",
              "15   SCO0007_loc_291_rev  GGTGGGGCAGGGCGCGCGTCAGGTGGGGGAGTGGGTTTCAATCCAC...   \n",
              "16   SCO0004_loc_861_fwd  TGCGTCAGCGCAGACCAGCGTCTCCTCGTGCAGCGTCGCCCGGCAA...   \n",
              "17   SCO0004_loc_861_rev  GCTGCACGAGGAGACGCTGGTCTGCGCTGACGCAGTTTCAATCCAC...   \n",
              "18    SCO0001_loc_89_fwd  ACACCCGCCGTCTGATACTGCGTCACCCGGCACGGTCGCCCGGCAA...   \n",
              "19    SCO0001_loc_89_rev  CGTGCCGGGTGACGCAGTATCAGACGGCGGGTGTGTTTCAATCCAC...   \n",
              "\n",
              "   Concentration Purification  \n",
              "0           25nm          STD  \n",
              "1           25nm          STD  \n",
              "2           25nm          STD  \n",
              "3           25nm          STD  \n",
              "4           25nm          STD  \n",
              "5           25nm          STD  \n",
              "6           25nm          STD  \n",
              "7           25nm          STD  \n",
              "8           25nm          STD  \n",
              "9           25nm          STD  \n",
              "10          25nm          STD  \n",
              "11          25nm          STD  \n",
              "12          25nm          STD  \n",
              "13          25nm          STD  \n",
              "14          25nm          STD  \n",
              "15          25nm          STD  \n",
              "16          25nm          STD  \n",
              "17          25nm          STD  \n",
              "18          25nm          STD  \n",
              "19          25nm          STD  "
            ]
          },
          "execution_count": 17,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "idt_primers_cas3=primers_to_IDT(primer_records)\n",
        "idt_primers_cas3"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": []
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "# In frame deletion"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 18,
      "metadata": {},
      "outputs": [],
      "source": [
        "in_frame_deletion = False \n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 19,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/plain": [
              "(Dseqrecord(-12770),)"
            ]
          },
          "execution_count": 19,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "Dseqrecord(assembled_cas3_plasmids[3], circular=True).cut(EcoRI)"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "#### Getting repair templates"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 20,
      "metadata": {},
      "outputs": [],
      "source": [
        "if in_frame_deletion:\n",
        "    \n",
        "\n",
        "    # Make repair templates\n",
        "    repair_templates_data = find_up_dw_repair_templates(genome, \n",
        "                                                        genes_to_KO, \n",
        "                                                        target_tm=melting_temperature, \n",
        "                                                        primer_tm_kwargs={'conc':primer_concentration, 'prodcode':chosen_polymerase} , \n",
        "                                                        repair_length=repair_length)\n",
        "\n",
        "    # Digest the plasmids\n",
        "    processed_records = [sorted(Dseqrecord(record, circular=True).cut(Enzyme_for_repair_template_integration), key=lambda x: len(x), reverse=True)[0] for record in assembled_cas3_plasmids]\n",
        "\n",
        "    # Rename them appropriately\n",
        "    for i in range(len(processed_records)):\n",
        "        processed_records[i].name = assembled_cas3_plasmids[i].name\n",
        "        print(processed_records[i].name)\n",
        "\n",
        "    # Assembly \n",
        "    assembly_data = assemble_multiple_plasmids_with_repair_templates_for_deletion(genes_to_KO, \n",
        "                                                                                processed_records, \n",
        "                                                                                repair_templates_data, \n",
        "                                                                                overlap=overlap_length)\n",
        "    # updating the primer names to something systematic.\n",
        "    update_primer_names(assembly_data)\n",
        "\n",
        "    # Parse through the primer df\n",
        "    primer_df = create_primer_df_from_dict(assembly_data)\n",
        "\n",
        "    #Find_unique primers\n",
        "    unique_df = primer_df.drop_duplicates(keep='first')\n",
        "    idt_df = create_idt_order_dataframe(unique_df, concentration=\"25nm\", purification=\"STD\")\n",
        "\n",
        "    # Contigs\n",
        "    assembled_contigs = []\n",
        "    for data in assembly_data: \n",
        "        contig_record = data['contig']\n",
        "        contig_record.id = f\"{data['name']}_w_rep\"\n",
        "        contig_record.name = f\"{data['name']}_w_rep\"\n",
        "        assembled_contigs.append(contig_record)\n",
        "\n",
        "\n",
        "    # print our df that contains primers for the adding repair templates to the vecors.         \n",
        "    print(unique_df)"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "Check for restriction enzyme sites"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 21,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>plasmid_name</th>\n",
              "      <th>date</th>\n",
              "      <th>original_plasmid</th>\n",
              "      <th>integration</th>\n",
              "      <th>size</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>pCas3_SCO0002(348)</td>\n",
              "      <td>2025-06-10</td>\n",
              "      <td>pCRISPR-Cas3</td>\n",
              "      <td>sgRNA_SCO0002(348)</td>\n",
              "      <td>12766</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>pCas3_SCO0003(1216)</td>\n",
              "      <td>2025-06-10</td>\n",
              "      <td>pCRISPR-Cas3</td>\n",
              "      <td>sgRNA_SCO0003(1216)</td>\n",
              "      <td>12766</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>pCas3_SCO0005(1273)</td>\n",
              "      <td>2025-06-10</td>\n",
              "      <td>pCRISPR-Cas3</td>\n",
              "      <td>sgRNA_SCO0005(1273)</td>\n",
              "      <td>12766</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>pCas3_SCO0006(272)</td>\n",
              "      <td>2025-06-10</td>\n",
              "      <td>pCRISPR-Cas3</td>\n",
              "      <td>sgRNA_SCO0006(272)</td>\n",
              "      <td>12766</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>pCas3_SCO0009(226)</td>\n",
              "      <td>2025-06-10</td>\n",
              "      <td>pCRISPR-Cas3</td>\n",
              "      <td>sgRNA_SCO0009(226)</td>\n",
              "      <td>12766</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>5</th>\n",
              "      <td>pCas3_SCO0010(1066)</td>\n",
              "      <td>2025-06-10</td>\n",
              "      <td>pCRISPR-Cas3</td>\n",
              "      <td>sgRNA_SCO0010(1066)</td>\n",
              "      <td>12766</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>6</th>\n",
              "      <td>pCas3_SCO0008(196)</td>\n",
              "      <td>2025-06-10</td>\n",
              "      <td>pCRISPR-Cas3</td>\n",
              "      <td>sgRNA_SCO0008(196)</td>\n",
              "      <td>12766</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>7</th>\n",
              "      <td>pCas3_SCO0007(291)</td>\n",
              "      <td>2025-06-10</td>\n",
              "      <td>pCRISPR-Cas3</td>\n",
              "      <td>sgRNA_SCO0007(291)</td>\n",
              "      <td>12766</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>8</th>\n",
              "      <td>pCas3_SCO0004(861)</td>\n",
              "      <td>2025-06-10</td>\n",
              "      <td>pCRISPR-Cas3</td>\n",
              "      <td>sgRNA_SCO0004(861)</td>\n",
              "      <td>12766</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>9</th>\n",
              "      <td>pCas3_SCO0001(89)</td>\n",
              "      <td>2025-06-10</td>\n",
              "      <td>pCRISPR-Cas3</td>\n",
              "      <td>sgRNA_SCO0001(89)</td>\n",
              "      <td>12766</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "          plasmid_name        date original_plasmid          integration  \\\n",
              "0   pCas3_SCO0002(348)  2025-06-10     pCRISPR-Cas3   sgRNA_SCO0002(348)   \n",
              "1  pCas3_SCO0003(1216)  2025-06-10     pCRISPR-Cas3  sgRNA_SCO0003(1216)   \n",
              "2  pCas3_SCO0005(1273)  2025-06-10     pCRISPR-Cas3  sgRNA_SCO0005(1273)   \n",
              "3   pCas3_SCO0006(272)  2025-06-10     pCRISPR-Cas3   sgRNA_SCO0006(272)   \n",
              "4   pCas3_SCO0009(226)  2025-06-10     pCRISPR-Cas3   sgRNA_SCO0009(226)   \n",
              "5  pCas3_SCO0010(1066)  2025-06-10     pCRISPR-Cas3  sgRNA_SCO0010(1066)   \n",
              "6   pCas3_SCO0008(196)  2025-06-10     pCRISPR-Cas3   sgRNA_SCO0008(196)   \n",
              "7   pCas3_SCO0007(291)  2025-06-10     pCRISPR-Cas3   sgRNA_SCO0007(291)   \n",
              "8   pCas3_SCO0004(861)  2025-06-10     pCRISPR-Cas3   sgRNA_SCO0004(861)   \n",
              "9    pCas3_SCO0001(89)  2025-06-10     pCRISPR-Cas3    sgRNA_SCO0001(89)   \n",
              "\n",
              "    size  \n",
              "0  12766  \n",
              "1  12766  \n",
              "2  12766  \n",
              "3  12766  \n",
              "4  12766  \n",
              "5  12766  \n",
              "6  12766  \n",
              "7  12766  \n",
              "8  12766  \n",
              "9  12766  "
            ]
          },
          "execution_count": 21,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "if in_frame_deletion:\n",
        "    integration_names = filtered_df.apply(lambda row: f\"sgRNA_{row['locus_tag']}({row['sgrna_loc']})\", axis=1).tolist()\n",
        "    plasmid_metadata_df = extract_metadata_to_dataframe(assembled_contigs,\n",
        "                                                        clean_plasmid,\n",
        "                                                        integration_names)\n",
        "    workflow_df = determine_workflow_order_for_plasmids(assembled_cas3_plasmids, \n",
        "                                                        assembled_contigs,\n",
        "                                                        [Enzyme_for_repair_template_integration,], [ \"NcoI\",\"BstBI\" ]) \n",
        "    \n",
        "    plasmid_metadata_df = pd.merge(plasmid_metadata_df, workflow_df, on='plasmid_name', how='inner')  # or 'left'/'right' depending on what you need\n",
        "\n",
        "else: \n",
        "    integration_names = filtered_df.apply(lambda row: f\"sgRNA_{row['locus_tag']}({row['sgrna_loc']})\", axis=1).tolist()\n",
        "    plasmid_metadata_df = extract_metadata_to_dataframe(assembled_cas3_plasmids,\n",
        "                                                        clean_plasmid,\n",
        "                                                        integration_names)\n",
        "\n",
        "plasmid_metadata_df"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 22,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>Name</th>\n",
              "      <th>Sequence</th>\n",
              "      <th>Concentration</th>\n",
              "      <th>Purification</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>SCO0002_loc_348_fwd</td>\n",
              "      <td>CAACGGCCCAGGGACAGTCCGGCGACCAGTGACGGTCGCCCGGCAA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>SCO0002_loc_348_rev</td>\n",
              "      <td>CGTCACTGGTCGCCGGACTGTCCCTGGGCCGTTGGTTTCAATCCAC...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>SCO0003_loc_1216_fwd</td>\n",
              "      <td>TCTGATTGGCTAGCCACGCTCCCAGAGCAAATCCGTCGCCCGGCAA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>SCO0003_loc_1216_rev</td>\n",
              "      <td>GGATTTGCTCTGGGAGCGTGGCTAGCCAATCAGAGTTTCAATCCAC...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>SCO0005_loc_1273_fwd</td>\n",
              "      <td>TGCCAGCACGTCGCCGGCCACACCGGCTCCAACCGTCGCCCGGCAA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>5</th>\n",
              "      <td>SCO0005_loc_1273_rev</td>\n",
              "      <td>GGTTGGAGCCGGTGTGGCCGGCGACGTGCTGGCAGTTTCAATCCAC...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>6</th>\n",
              "      <td>SCO0006_loc_272_fwd</td>\n",
              "      <td>TTGGCAAGGGCGTCGATAGCGGGGGTGCGCACGGGTCGCCCGGCAA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>7</th>\n",
              "      <td>SCO0006_loc_272_rev</td>\n",
              "      <td>CCGTGCGCACCCCCGCTATCGACGCCCTTGCCAAGTTTCAATCCAC...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>8</th>\n",
              "      <td>SCO0009_loc_226_fwd</td>\n",
              "      <td>GACCCACTGTGCGAGAGCCGCGAGACCTCGTTGGGTCGCCCGGCAA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>9</th>\n",
              "      <td>SCO0009_loc_226_rev</td>\n",
              "      <td>CCAACGAGGTCTCGCGGCTCTCGCACAGTGGGTCGTTTCAATCCAC...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>10</th>\n",
              "      <td>SCO0010_loc_1066_fwd</td>\n",
              "      <td>TGCGCTCTGGTGGTCTGACGTTCAGCCCGACCACGTCGCCCGGCAA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>11</th>\n",
              "      <td>SCO0010_loc_1066_rev</td>\n",
              "      <td>GTGGTCGGGCTGAACGTCAGACCACCAGAGCGCAGTTTCAATCCAC...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>12</th>\n",
              "      <td>SCO0008_loc_196_fwd</td>\n",
              "      <td>CGGGTAGATCACCAGGTCGCGGGTCACCAGGGCGGTCGCCCGGCAA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>13</th>\n",
              "      <td>SCO0008_loc_196_rev</td>\n",
              "      <td>CGCCCTGGTGACCCGCGACCTGGTGATCTACCCGGTTTCAATCCAC...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>14</th>\n",
              "      <td>SCO0007_loc_291_fwd</td>\n",
              "      <td>CCACTCCCCCACCTGACGCGCGCCCTGCCCCACCGTCGCCCGGCAA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>15</th>\n",
              "      <td>SCO0007_loc_291_rev</td>\n",
              "      <td>GGTGGGGCAGGGCGCGCGTCAGGTGGGGGAGTGGGTTTCAATCCAC...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>16</th>\n",
              "      <td>SCO0004_loc_861_fwd</td>\n",
              "      <td>TGCGTCAGCGCAGACCAGCGTCTCCTCGTGCAGCGTCGCCCGGCAA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>17</th>\n",
              "      <td>SCO0004_loc_861_rev</td>\n",
              "      <td>GCTGCACGAGGAGACGCTGGTCTGCGCTGACGCAGTTTCAATCCAC...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>18</th>\n",
              "      <td>SCO0001_loc_89_fwd</td>\n",
              "      <td>ACACCCGCCGTCTGATACTGCGTCACCCGGCACGGTCGCCCGGCAA...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>19</th>\n",
              "      <td>SCO0001_loc_89_rev</td>\n",
              "      <td>CGTGCCGGGTGACGCAGTATCAGACGGCGGGTGTGTTTCAATCCAC...</td>\n",
              "      <td>25nm</td>\n",
              "      <td>STD</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "                    Name                                           Sequence  \\\n",
              "0    SCO0002_loc_348_fwd  CAACGGCCCAGGGACAGTCCGGCGACCAGTGACGGTCGCCCGGCAA...   \n",
              "1    SCO0002_loc_348_rev  CGTCACTGGTCGCCGGACTGTCCCTGGGCCGTTGGTTTCAATCCAC...   \n",
              "2   SCO0003_loc_1216_fwd  TCTGATTGGCTAGCCACGCTCCCAGAGCAAATCCGTCGCCCGGCAA...   \n",
              "3   SCO0003_loc_1216_rev  GGATTTGCTCTGGGAGCGTGGCTAGCCAATCAGAGTTTCAATCCAC...   \n",
              "4   SCO0005_loc_1273_fwd  TGCCAGCACGTCGCCGGCCACACCGGCTCCAACCGTCGCCCGGCAA...   \n",
              "5   SCO0005_loc_1273_rev  GGTTGGAGCCGGTGTGGCCGGCGACGTGCTGGCAGTTTCAATCCAC...   \n",
              "6    SCO0006_loc_272_fwd  TTGGCAAGGGCGTCGATAGCGGGGGTGCGCACGGGTCGCCCGGCAA...   \n",
              "7    SCO0006_loc_272_rev  CCGTGCGCACCCCCGCTATCGACGCCCTTGCCAAGTTTCAATCCAC...   \n",
              "8    SCO0009_loc_226_fwd  GACCCACTGTGCGAGAGCCGCGAGACCTCGTTGGGTCGCCCGGCAA...   \n",
              "9    SCO0009_loc_226_rev  CCAACGAGGTCTCGCGGCTCTCGCACAGTGGGTCGTTTCAATCCAC...   \n",
              "10  SCO0010_loc_1066_fwd  TGCGCTCTGGTGGTCTGACGTTCAGCCCGACCACGTCGCCCGGCAA...   \n",
              "11  SCO0010_loc_1066_rev  GTGGTCGGGCTGAACGTCAGACCACCAGAGCGCAGTTTCAATCCAC...   \n",
              "12   SCO0008_loc_196_fwd  CGGGTAGATCACCAGGTCGCGGGTCACCAGGGCGGTCGCCCGGCAA...   \n",
              "13   SCO0008_loc_196_rev  CGCCCTGGTGACCCGCGACCTGGTGATCTACCCGGTTTCAATCCAC...   \n",
              "14   SCO0007_loc_291_fwd  CCACTCCCCCACCTGACGCGCGCCCTGCCCCACCGTCGCCCGGCAA...   \n",
              "15   SCO0007_loc_291_rev  GGTGGGGCAGGGCGCGCGTCAGGTGGGGGAGTGGGTTTCAATCCAC...   \n",
              "16   SCO0004_loc_861_fwd  TGCGTCAGCGCAGACCAGCGTCTCCTCGTGCAGCGTCGCCCGGCAA...   \n",
              "17   SCO0004_loc_861_rev  GCTGCACGAGGAGACGCTGGTCTGCGCTGACGCAGTTTCAATCCAC...   \n",
              "18    SCO0001_loc_89_fwd  ACACCCGCCGTCTGATACTGCGTCACCCGGCACGGTCGCCCGGCAA...   \n",
              "19    SCO0001_loc_89_rev  CGTGCCGGGTGACGCAGTATCAGACGGCGGGTGTGTTTCAATCCAC...   \n",
              "\n",
              "   Concentration Purification  \n",
              "0           25nm          STD  \n",
              "1           25nm          STD  \n",
              "2           25nm          STD  \n",
              "3           25nm          STD  \n",
              "4           25nm          STD  \n",
              "5           25nm          STD  \n",
              "6           25nm          STD  \n",
              "7           25nm          STD  \n",
              "8           25nm          STD  \n",
              "9           25nm          STD  \n",
              "10          25nm          STD  \n",
              "11          25nm          STD  \n",
              "12          25nm          STD  \n",
              "13          25nm          STD  \n",
              "14          25nm          STD  \n",
              "15          25nm          STD  \n",
              "16          25nm          STD  \n",
              "17          25nm          STD  \n",
              "18          25nm          STD  \n",
              "19          25nm          STD  "
            ]
          },
          "execution_count": 22,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "full_idt = idt_primers_cas3\n",
        "full_idt"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "## Folder with all the generated I/O"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 23,
      "metadata": {},
      "outputs": [],
      "source": [
        "input_files = [\n",
        "    {\"name\": \"input_genome.gb\", \"content\": genome},\n",
        "    {\"name\": \"input_plasmid.gb\", \"content\": clean_plasmid}\n",
        "]\n",
        "\n",
        "\n",
        "if in_frame_deletion: \n",
        "    output_files = [\n",
        "        {\"name\": \"BEST_w_sgRNAs.gb\", \"content\": assembled_contigs}, # LIST OF Dseqrecords\n",
        "        {\"name\": \"primer_df.csv\", \"content\": primer_df},\n",
        "        {\"name\": \"full_idt.csv\", \"content\": full_idt},\n",
        "        {\"name\": \"sgrna_df.csv\", \"content\": sgrna_df},\n",
        "        {\"name\": \"filtered_sgrna_df.csv\", \"content\": filtered_df},\n",
        "        {\"name\": \"plasmid_metadata_df.csv\", \"content\": plasmid_metadata_df},\n",
        "\n",
        "    ]\n",
        "else: \n",
        "    output_files = [\n",
        "        {\"name\": \"cBEST_w_sgRNAs.gb\", \"content\": assembled_cas3_plasmids}, # LIST OF Dseqrecords\n",
        "        {\"name\": \"full_idt.csv\", \"content\": full_idt},\n",
        "        {\"name\": \"sgrna_df.csv\", \"content\": sgrna_df},\n",
        "        {\"name\": \"filtered_df.csv\", \"content\": filtered_df},\n",
        "        {\"name\": \"plasmid_metadata_df.csv\", \"content\": plasmid_metadata_df},\n",
        "\n",
        "    ]\n",
        "         \n",
        "input_values = {\n",
        "    \"genes_to_knockout\": genes_to_KO,\n",
        "\n",
        "    \"filtering_metrics\": {\n",
        "        \"gc_upper\": gc_upper,\n",
        "        \"gc_lower\": gc_lower,\n",
        "        \"off_target_seed\": off_target_seed,\n",
        "        \"off_target_upper\": off_target_upper,\n",
        "        \"cas_type\": cas_type,\n",
        "        \"number_of_sgRNAs_per_group\": number_of_sgRNAs_per_group,\n",
        "\n",
        "    },  \n",
        "    \"polymerase_settings\": {\n",
        "        \"chosen_polymerase\": chosen_polymerase,\n",
        "        \"melting_temperature\": melting_temperature,\n",
        "        \"primer_concentration\": primer_concentration,\n",
        "        \"primer_number_increment\": primer_number_increment,\n",
        "        \"flanking_region\": flanking_region\n",
        "    },\n",
        "    \"overlapping_sequences\": {\n",
        "        \"up_homology\": str(forward_backbone_overhang),\n",
        "        \"dw_homology\": str(reverse_backbone_overhang)\n",
        "    }\n",
        "}\n",
        "\n",
        "# Paths to Markdown files\n",
        "markdown_file_paths = [\n",
        "    \"../../protocols/conjugation_protcol.md\",\n",
        "    \"../../protocols/single_target_crispr_plasmid_protcol.md\"\n",
        "\n",
        "]\n",
        "\n",
        "timestamp = datetime.utcnow().isoformat()\n",
        "\n",
        "# Create project directory structure\n",
        "project_directory = ProjectDirectory(\n",
        "    project_name=f\"CRISPR_cas3_inframe_deletion_workflow_{timestamp}\",\n",
        "    input_files=input_files,\n",
        "    output_files=output_files,\n",
        "    input_values=input_values,\n",
        "    markdown_file_paths=markdown_file_paths\n",
        ")\n",
        "\n",
        "# DO You want to save the folder? \n",
        "save_zip_folder = False \n",
        "\n",
        "if save_zip_folder: \n",
        "    # Generate the project directory structure and get the zip content\n",
        "    zip_content = project_directory.create_directory_structure(create_directories=False)\n",
        "\n",
        "    # Save the zip file to disk (optional)\n",
        "    with open(\"project_structure.zip\", \"wb\") as f:\n",
        "        f.write(zip_content)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {},
      "outputs": [],
      "source": []
    }
  ],
  "metadata": {
    "kernelspec": {
      "display_name": "streptocad",
      "language": "python",
      "name": "python3"
    },
    "language_info": {
      "codemirror_mode": {
        "name": "ipython",
        "version": 3
      },
      "file_extension": ".py",
      "mimetype": "text/x-python",
      "name": "python",
      "nbconvert_exporter": "python",
      "pygments_lexer": "ipython3",
      "version": "3.11.11"
    }
  },
  "nbformat": 4,
  "nbformat_minor": 2
}
